diff --git a/packages/ui/components/button/Button.tsx b/packages/ui/components/button/Button.tsx
index 8ac50e6faf..b6ebd7f08e 100644
--- a/packages/ui/components/button/Button.tsx
+++ b/packages/ui/components/button/Button.tsx
@@ -125,7 +125,7 @@ export const buttonClasses = cva(
           "border-default",
           "text-error",
           // Hover state
-          "dark:hover:text-red-100",
+          "dark:hover:text-red-400",
           "hover:border-semantic-error",
           "hover:bg-error",
           // Focus state
