diff --git a/packages/app-store/zohocrm/lib/CrmService.ts b/packages/app-store/zohocrm/lib/CrmService.ts
index 1bdf522a35..22f58d1545 100644
--- a/packages/app-store/zohocrm/lib/CrmService.ts
+++ b/packages/app-store/zohocrm/lib/CrmService.ts
@@ -236,7 +236,7 @@ class ZohoCrmCrmService implements CRM {
         );
         if (!zohoCrmTokenInfo.data.error) {
           // set expiry date as offset from current time.
-          zohoCrmTokenInfo.data.expiryDate = Math.round(Date.now() + 60 * 60);
+          zohoCrmTokenInfo.data.expiryDate = Math.round(Date.now() + 60 * 60 * 1000);
 
           await prisma.credential.update({
             where: {
