diff --git a/.github/CODEOWNERS b/.github/CODEOWNERS
index 84af29155b..25910495e1 100644
--- a/.github/CODEOWNERS
+++ b/.github/CODEOWNERS
@@ -37,3 +37,9 @@
 /packages/trpc/server/routers/viewer/bookings/get.handler.ts @calcom/Foundation
 /packages/trpc/server/routers/viewer/slots/**/* @calcom/Foundation
 /packages/ui/**/* @calcom/Consumer @calcom/Foundation
+
+# Exclude test files from requiring foundation/consumer approval
+**/*.spec.*
+**/*.e2e.*
+**/*.test.*
+**/*.integration-test.*
