diff --git a/packages/lib/domainManager/organization.ts b/packages/lib/domainManager/organization.ts
index 66bd6f7baabf6b..a122fa0e16b15a 100644
--- a/packages/lib/domainManager/organization.ts
+++ b/packages/lib/domainManager/organization.ts
@@ -24,7 +24,6 @@ export const deleteDomain = async (slug: string) => {
     isDnsRecordDeleted = await deleteDnsRecord(domain);
   }
   return isDomainDeleted && isDnsRecordDeleted;
-  return false;
 };
 
 export const createDomain = async (slug: string) => {
