diff --git a/apps/web/app/(use-page-wrapper)/settings/(settings-layout)/SettingsLayoutAppDirClient.tsx b/apps/web/app/(use-page-wrapper)/settings/(settings-layout)/SettingsLayoutAppDirClient.tsx
index 171315ce28bbd3..00bfb75a3ca759 100644
--- a/apps/web/app/(use-page-wrapper)/settings/(settings-layout)/SettingsLayoutAppDirClient.tsx
+++ b/apps/web/app/(use-page-wrapper)/settings/(settings-layout)/SettingsLayoutAppDirClient.tsx
@@ -142,16 +142,16 @@ const getTabs = (orgBranding: OrganizationBranding | null) => {
           href: "/settings/developer/webhooks",
           trackingMetadata: { section: "developer", page: "webhooks" },
         },
+        {
+          name: "oAuth",
+          href: "/settings/developer/oauth",
+          trackingMetadata: { section: "developer", page: "oauth_clients" },
+        },
         {
           name: "api_keys",
           href: "/settings/developer/api-keys",
           trackingMetadata: { section: "developer", page: "api_keys" },
         },
-        {
-          name: "oAuth",
-          href: "/settings/developer/oauth",
-          trackingMetadata: { section: "developer", page: "oauth_clients" }
-        },
         {
           name: "admin_api",
           href: "/settings/organizations/admin-api",
diff --git a/apps/web/modules/ee/api-keys/components/ApiKeyDialogForm.tsx b/apps/web/modules/ee/api-keys/components/ApiKeyDialogForm.tsx
index 8b8d26890c7707..9decae3fcb9253 100644
--- a/apps/web/modules/ee/api-keys/components/ApiKeyDialogForm.tsx
+++ b/apps/web/modules/ee/api-keys/components/ApiKeyDialogForm.tsx
@@ -164,7 +164,7 @@ export default function ApiKeyDialogForm({
                 <Link
                   target="_blank"
                   rel="noopener noreferrer"
-                  href="https://cal.com/platform"
+                  href="https://cal.com/integrate"
                   className="border-subtle relative flex w-full items-start rounded-[10px] border p-4 text-sm">
                   {t("api_key_modal_subtitle_platform")}
                 </Link>
diff --git a/apps/web/public/static/locales/en/common.json b/apps/web/public/static/locales/en/common.json
index a3b52231b236ec..198f53066a6a99 100644
--- a/apps/web/public/static/locales/en/common.json
+++ b/apps/web/public/static/locales/en/common.json
@@ -1448,7 +1448,7 @@
   "test_failed": "Test failed",
   "provide_api_key": "Provide API key",
   "api_key_modal_subtitle": "API keys allow you to make API calls for your own account.",
-  "api_key_modal_subtitle_platform": "Create a marketplace or platform instead.",
+  "api_key_modal_subtitle_platform": "Create an OAuth app and allow user to schedule.",
   "api_keys_subtitle": "Generate API keys to use for accessing your own account.",
   "create_api_key": "Create an API key",
   "personal_note": "Name this key",
