diff --git a/apps/web/app/(use-page-wrapper)/apps/routing-forms/forms/[[...pages]]/Forms.tsx b/apps/web/app/(use-page-wrapper)/apps/routing-forms/forms/[[...pages]]/Forms.tsx
index d60500e9dc8690..bbb702400cfba5 100644
--- a/apps/web/app/(use-page-wrapper)/apps/routing-forms/forms/[[...pages]]/Forms.tsx
+++ b/apps/web/app/(use-page-wrapper)/apps/routing-forms/forms/[[...pages]]/Forms.tsx
@@ -149,6 +149,7 @@ export default function RoutingForms({ appUrl }: { appUrl: string }) {
   return (
     <LicenseRequired>
       <ShellMain
+        disableSticky={true}
         heading={t("routing")}
         CTA={
           hasPaidPlan && forms?.length ? (
diff --git a/apps/web/app/(use-page-wrapper)/settings/organizations/(org-user-only)/[id]/members/layout.tsx b/apps/web/app/(use-page-wrapper)/settings/organizations/(org-user-only)/[id]/members/layout.tsx
index eef834cf8b7336..3ca093d60e7098 100644
--- a/apps/web/app/(use-page-wrapper)/settings/organizations/(org-user-only)/[id]/members/layout.tsx
+++ b/apps/web/app/(use-page-wrapper)/settings/organizations/(org-user-only)/[id]/members/layout.tsx
@@ -10,6 +10,7 @@ export default async function Layout({ children }: { children: React.ReactNode }
   return (
     <Shell
       withoutMain={false}
+      disableSticky={true}
       heading={t("organization_members")}
       subtitle={t("organization_description")}
       headerClassName="hidden md:block"
diff --git a/apps/web/app/(use-page-wrapper)/settings/platform/members/layout.tsx b/apps/web/app/(use-page-wrapper)/settings/platform/members/layout.tsx
index 94dc169f9f8143..0f880c807d78b7 100644
--- a/apps/web/app/(use-page-wrapper)/settings/platform/members/layout.tsx
+++ b/apps/web/app/(use-page-wrapper)/settings/platform/members/layout.tsx
@@ -23,6 +23,7 @@ export default async function Layout({ children }: { children: React.ReactNode }
           />
         </div>
       }
+      disableSticky={true}
       title={t("platform_members")}
       subtitle={t("platform_members_description")}
       withoutMain={false}
diff --git a/apps/web/modules/bookings/components/BookingListContainer.tsx b/apps/web/modules/bookings/components/BookingListContainer.tsx
index a17c43cc068c57..d6769b9672666c 100644
--- a/apps/web/modules/bookings/components/BookingListContainer.tsx
+++ b/apps/web/modules/bookings/components/BookingListContainer.tsx
@@ -28,7 +28,6 @@ import {
 import type { BookingListingStatus, BookingsGetOutput, RowData } from "../types";
 import { BookingDetailsSheet } from "./BookingDetailsSheet";
 import { BookingList } from "./BookingList";
-import { BookingsCsvDownload } from "./BookingsCsvDownload";
 import { ViewToggleButton } from "./ViewToggleButton";
 
 interface FilterButtonProps {
@@ -189,7 +188,7 @@ function BookingListInner({
         <div className="hidden grow md:block" />
 
         <DataTableSegment.Select />
-        <BookingsCsvDownload status={status} />
+        {/* <BookingsCsvDownload status={status} /> */}
         {bookingsV3Enabled && <ViewToggleButton bookingsV3Enabled={bookingsV3Enabled} />}
       </div>
       {displayedFilterCount > 0 && showFilters && (
diff --git a/apps/web/modules/ee/workflows/views/WorkflowsPage.tsx b/apps/web/modules/ee/workflows/views/WorkflowsPage.tsx
index 78841ddc98847f..67eadcee7f529e 100644
--- a/apps/web/modules/ee/workflows/views/WorkflowsPage.tsx
+++ b/apps/web/modules/ee/workflows/views/WorkflowsPage.tsx
@@ -56,6 +56,7 @@ function WorkflowsPage({ filteredList }: PageProps) {
     <Shell withoutMain>
       <LicenseRequired>
         <ShellMain
+          disableSticky={true}
           heading={t("workflows")}
           subtitle={t("workflows_to_automate_notifications")}
           title={t("workflows")}
diff --git a/apps/web/modules/shell/Shell.tsx b/apps/web/modules/shell/Shell.tsx
index 74c54a00c54885..f861cb3df7b1fa 100644
--- a/apps/web/modules/shell/Shell.tsx
+++ b/apps/web/modules/shell/Shell.tsx
@@ -87,6 +87,7 @@ export type LayoutProps = {
   afterHeading?: ReactNode;
   smallHeading?: boolean;
   isPlatformUser?: boolean;
+  disableSticky?: boolean;
 };
 
 const KBarWrapper = ({ children, withKBar = false }: { withKBar: boolean; children: React.ReactNode }) =>
@@ -132,8 +133,9 @@ export function ShellMain(props: LayoutProps) {
       {(props.heading || !!props.backPath) && (
         <div
           className={classNames(
-            "bg-default sticky top-0 z-10 mb-0 flex items-center md:mb-6 md:mt-0",
-            props.smallHeading ? "lg:mb-7" : "lg:mb-8"
+            "bg-default mb-0 flex items-center md:mb-6 md:mt-0",
+            props.smallHeading ? "lg:mb-7" : "lg:mb-8",
+            !props.disableSticky && "sticky top-0 z-10"
           )}>
           {!!props.backPath && (
             <Button
