diff --git a/apps/web/app/(use-page-wrapper)/apps/routing-forms/forms/[[...pages]]/Forms.tsx b/apps/web/app/(use-page-wrapper)/apps/routing-forms/forms/[[...pages]]/Forms.tsx
index 26beaa28b7b4a1..d60500e9dc8690 100644
--- a/apps/web/app/(use-page-wrapper)/apps/routing-forms/forms/[[...pages]]/Forms.tsx
+++ b/apps/web/app/(use-page-wrapper)/apps/routing-forms/forms/[[...pages]]/Forms.tsx
@@ -169,7 +169,7 @@ export default function RoutingForms({ appUrl }: { appUrl: string }) {
                 <Button color="primary" href={`${WEBAPP_URL}/settings/teams/new`}>
                   {t("upgrade")}
                 </Button>
-                <Button color="minimal" href="https://go.cal.com/teams-video" target="_blank">
+                <Button color="minimal" href="https://cal.com/routing" target="_blank">
                   {t("learn_more")}
                 </Button>
               </ButtonGroup>
diff --git a/apps/web/modules/ee/teams/components/TeamsListing.tsx b/apps/web/modules/ee/teams/components/TeamsListing.tsx
index 063243e9730608..a0a83f4e49167e 100644
--- a/apps/web/modules/ee/teams/components/TeamsListing.tsx
+++ b/apps/web/modules/ee/teams/components/TeamsListing.tsx
@@ -144,7 +144,7 @@ export function TeamsListing({
                   <Button color="primary" href={`${WEBAPP_URL}/settings/teams/new`}>
                     {t("create_team")}
                   </Button>
-                  <Button color="minimal" href="https://go.cal.com/teams-video" target="_blank">
+                  <Button color="minimal" href="https://cal.com/teams" target="_blank">
                     {t("learn_more")}
                   </Button>
                 </ButtonGroup>
