diff --git a/apps/web/app/api/teams/create/route.ts b/apps/web/app/api/teams/create/route.ts
index 38f5ef15057107..0d90c93b9aa8e1 100644
--- a/apps/web/app/api/teams/create/route.ts
+++ b/apps/web/app/api/teams/create/route.ts
@@ -133,7 +133,7 @@ async function getHandler(req: NextRequest) {
 
   // redirect to team screen
   return NextResponse.redirect(
-    new URL(`/settings/teams/${team.id}/onboard-members?event=team_created`, req.nextUrl.origin),
+    new URL(`/settings/teams/${team.id}/onboard-members?event=team_created`, WEBAPP_URL),
     { status: 302 }
   );
 }
