diff --git a/packages/dayjs/index.ts b/packages/dayjs/index.ts
index 6248cf3efc9d79..799464b1e0a1bd 100644
--- a/packages/dayjs/index.ts
+++ b/packages/dayjs/index.ts
@@ -18,9 +18,9 @@ dayjs.extend(isBetween);
 dayjs.extend(isToday);
 dayjs.extend(localizedFormat);
 dayjs.extend(relativeTime);
+dayjs.extend(utc);
 dayjs.extend(timeZone);
 dayjs.extend(toArray);
-dayjs.extend(utc);
 dayjs.extend(minmax);
 dayjs.extend(duration);
 
