diff --git a/apps/web/lib/signup/getServerSideProps.tsx b/apps/web/lib/signup/getServerSideProps.tsx
index 617c48fb5a43d0..2596a24eb4af82 100644
--- a/apps/web/lib/signup/getServerSideProps.tsx
+++ b/apps/web/lib/signup/getServerSideProps.tsx
@@ -156,9 +156,18 @@ export const getServerSideProps = async (ctx: GetServerSidePropsContext) => {
 
   let username = guessUsernameFromEmail(verificationToken.identifier);
 
+  if (!verificationToken?.team) {
+    return {
+      redirect: {
+        permanent: false,
+        destination: `/auth/error?error=Verification Token is not associated with any team`,
+      },
+    } as const;
+  }
+
   const tokenTeam = {
-    ...verificationToken?.team,
-    metadata: teamMetadataSchema.parse(verificationToken?.team?.metadata),
+    ...verificationToken.team,
+    metadata: teamMetadataSchema.parse(verificationToken.team.metadata ?? null),
   };
 
   const isATeamInOrganization = tokenTeam?.parentId !== null;
