diff --git a/packages/app-store-cli/src/build.ts b/packages/app-store-cli/src/build.ts
index 5e25269a61d382..f397263ba47798 100644
--- a/packages/app-store-cli/src/build.ts
+++ b/packages/app-store-cli/src/build.ts
@@ -2,8 +2,7 @@ import { spawnSync } from "node:child_process";
 import fs from "node:fs";
 import path from "node:path";
 import process from "node:process";
-import type { AppMeta } from "@calcom/types/App";
-import { AppMetaSchema } from "@calcom/types/AppMetaSchema";
+import { AppMetaSchema, type AppMetaType } from "@calcom/types/AppMetaSchema";
 import chokidar from "chokidar";
 // eslint-disable-next-line no-restricted-imports
 import { debounce } from "lodash";
@@ -39,7 +38,7 @@ const getVariableName = (appName: string) => appName.replace(/[-./]/g, "_");
 // INFO: Handle stripe separately as it's an old app with different dirName than slug/appId
 const getAppId = (app: { name: string }) => (app.name === "stripepayment" ? "stripe" : app.name);
 
-type App = Partial<AppMeta> & {
+type App = Partial<AppMetaType> & {
   name: string;
   path: string;
 };
@@ -82,7 +81,7 @@ function generateFiles() {
     for (let i = 0; i < appDirs.length; i++) {
       const configPath = path.join(APP_STORE_PATH, appDirs[i].path, "config.json");
       const metadataPath = path.join(APP_STORE_PATH, appDirs[i].path, "_metadata.ts");
-      let app: AppMetaSchema;
+      let app: Partial<AppMetaType>;
 
       if (fs.existsSync(configPath)) {
         try {
