diff --git a/packages/app-store-cli/src/build.ts b/packages/app-store-cli/src/build.ts
index d25dafa6b89445..5e25269a61d382 100644
--- a/packages/app-store-cli/src/build.ts
+++ b/packages/app-store-cli/src/build.ts
@@ -492,7 +492,7 @@ function generateFiles() {
   const redirectAppSlugs: string[] = [];
   forEachAppDir((app) => {
     // Exclude templates - they are not actual apps
-    if (app.externalLink && !app.path.startsWith("templates/")) {
+    if (app.externalLink && !app.path.replace(/\\/g, "/").startsWith("templates/")) {
       redirectAppSlugs.push(app.name);
     }
   });
