diff --git a/apps/web/modules/event-types/components/tabs/apps/EventAppsTab.tsx b/apps/web/modules/event-types/components/tabs/apps/EventAppsTab.tsx
index 7a638138a4b960..4a962fac95707e 100644
--- a/apps/web/modules/event-types/components/tabs/apps/EventAppsTab.tsx
+++ b/apps/web/modules/event-types/components/tabs/apps/EventAppsTab.tsx
@@ -116,7 +116,7 @@ export const EventAppsTab = ({
               title={
                 <ServerTrans
                   t={t}
-                  i18nKey={`${lockedText}_${isManagedEventType ? "for_members" : "by_team_admins"}`}
+                  i18nKey={`${lockedText}_${isManagedEventType ? "for_members" : "by_team_admin"}`}
                 />
               }
               actions={<div className="flex h-full items-center">{appsDisableProps.LockedIcon}</div>}
@@ -124,7 +124,7 @@ export const EventAppsTab = ({
                 <ServerTrans
                   t={t}
                   i18nKey={`apps_${lockedText}_${
-                    isManagedEventType ? "for_members" : "by_team_admins"
+                    isManagedEventType ? "for_members" : "by_team_admin"
                   }_description`}
                 />
               }
diff --git a/apps/web/modules/event-types/components/tabs/webhooks/EventWebhooksTab.tsx b/apps/web/modules/event-types/components/tabs/webhooks/EventWebhooksTab.tsx
index b844c3870fdcc2..816f56b51eca1a 100644
--- a/apps/web/modules/event-types/components/tabs/webhooks/EventWebhooksTab.tsx
+++ b/apps/web/modules/event-types/components/tabs/webhooks/EventWebhooksTab.tsx
@@ -128,7 +128,7 @@ export const EventWebhooksTab = ({ eventType }: Pick<EventTypeSetupProps, "event
                     title={
                       <ServerTrans
                         t={t}
-                        i18nKey={`${lockedText}_${isManagedEventType ? "for_members" : "by_team_admins"}`}
+                        i18nKey={`${lockedText}_${isManagedEventType ? "for_members" : "by_team_admin"}`}
                       />
                     }
                     actions={
@@ -138,7 +138,7 @@ export const EventWebhooksTab = ({ eventType }: Pick<EventTypeSetupProps, "event
                       <ServerTrans
                         t={t}
                         i18nKey={`webhooks_${lockedText}_${
-                          isManagedEventType ? "for_members" : "by_team_admins"
+                          isManagedEventType ? "for_members" : "by_team_admin"
                         }_description`}
                       />
                     }
diff --git a/apps/web/modules/event-types/components/tabs/workflows/EventWorkflowsTab.tsx b/apps/web/modules/event-types/components/tabs/workflows/EventWorkflowsTab.tsx
index b73e98a1c1eaef..ef14b9f6b9fee3 100644
--- a/apps/web/modules/event-types/components/tabs/workflows/EventWorkflowsTab.tsx
+++ b/apps/web/modules/event-types/components/tabs/workflows/EventWorkflowsTab.tsx
@@ -253,7 +253,7 @@ function EventWorkflowsTab(props: Props) {
               title={
                 <ServerTrans
                   t={t}
-                  i18nKey={`${lockedText}_${isManagedEventType ? "for_members" : "by_team_admins"}`}
+                  i18nKey={`${lockedText}_${isManagedEventType ? "for_members" : "by_team_admin"}`}
                 />
               }
               actions={<div className="flex h-full items-center">{workflowsDisableProps.LockedIcon}</div>}
@@ -261,7 +261,7 @@ function EventWorkflowsTab(props: Props) {
                 <ServerTrans
                   t={t}
                   i18nKey={`workflows_${lockedText}_${
-                    isManagedEventType ? "for_members" : "by_team_admins"
+                    isManagedEventType ? "for_members" : "by_team_admin"
                   }_description`}
                 />
               }
