diff --git a/plugins/checklist/spec/system/rich_editor_extension_spec.rb b/plugins/checklist/spec/system/rich_editor_extension_spec.rb
index 0b79b59ea2510..a17485c898c61 100644
--- a/plugins/checklist/spec/system/rich_editor_extension_spec.rb
+++ b/plugins/checklist/spec/system/rich_editor_extension_spec.rb
@@ -99,7 +99,6 @@ def click_checklist_toolbar_option
       expect(checklist).to have_items(count: 2)
 
       rich.send_keys(:home)
-      wait_for_timeout
       rich.send_keys(:backspace)
 
       expect(checklist).to have_checkboxes(count: 1)
@@ -130,10 +129,9 @@ def click_checklist_toolbar_option
       rich.send_keys("Only item")
 
       expect(checklist).to have_checkboxes(count: 1)
-      expect(checklist).to have_items(count: 1)
 
       rich.send_keys(:home)
-      wait_for_timeout
+      expect(checklist).to have_items(count: 1)
       rich.send_keys(:backspace)
 
       expect(checklist).to have_no_checkboxes
