diff --git a/app/controllers/categories_controller.rb b/app/controllers/categories_controller.rb
index b58638ae27..9a642337a0 100644
--- a/app/controllers/categories_controller.rb
+++ b/app/controllers/categories_controller.rb
@@ -206,12 +206,14 @@ class CategoriesController < ApplicationController
       category_params.delete(:position)
 
       old_custom_fields = cat.custom_fields.dup
-      if category_params[:custom_fields]
-        category_params[:custom_fields].each do |key, value|
-          if value.nil? || value == ""
-            cat.custom_fields.delete(key)
+      category_params[:custom_fields]&.each do |key, value|
+        if value.nil? || value == ""
+          cat.custom_fields.delete(key)
+        else
+          cat.custom_fields[key] = if value.is_a?(TrueClass) || value.is_a?(FalseClass)
+            value.to_s
           else
-            cat.custom_fields[key] = value
+            value
           end
         end
       end
diff --git a/frontend/discourse/admin/controllers/edit-category/tabs.js b/frontend/discourse/admin/controllers/edit-category/tabs.js
index 798ed4e982..b737d12bd5 100644
--- a/frontend/discourse/admin/controllers/edit-category/tabs.js
+++ b/frontend/discourse/admin/controllers/edit-category/tabs.js
@@ -111,13 +111,6 @@ export default class EditCategoryTabsController extends Controller {
       data.category_setting = { ...(this.model.category_setting ?? {}) };
       data.custom_fields = { ...(this.model.custom_fields ?? {}) };
 
-      Object.entries(data.custom_fields).forEach(([key, value]) => {
-        // Coerce to boolean otherwise FormKit checkboxes do not work correctly.
-        if (value === "true" || value === "false") {
-          data.custom_fields[key] = value === "true";
-        }
-      });
-
       if (this.siteSettings.enable_category_type_setup) {
         data.category_type_site_settings = {};
 
diff --git a/frontend/discourse/app/form-kit/components/fk/control/checkbox.gjs b/frontend/discourse/app/form-kit/components/fk/control/checkbox.gjs
index 548498f3dd..017ed932ff 100644
--- a/frontend/discourse/app/form-kit/components/fk/control/checkbox.gjs
+++ b/frontend/discourse/app/form-kit/components/fk/control/checkbox.gjs
@@ -4,22 +4,22 @@ import { action } from "@ember/object";
 import FKLabel from "discourse/form-kit/components/fk/label";
 import FKOptional from "discourse/form-kit/components/fk/optional";
 import FKTooltip from "discourse/form-kit/components/fk/tooltip";
-import { eq } from "discourse/truth-helpers";
+import { eq, or } from "discourse/truth-helpers";
 import { i18n } from "discourse-i18n";
 
 export default class FKControlCheckbox extends Component {
   static controlType = "checkbox";
 
   @action
-  handleInput() {
-    this.args.field.set(!this.args.field.value);
+  handleInput(event) {
+    this.args.field.set(event.target.checked);
   }
 
   <template>
     <FKLabel class="form-kit__control-checkbox-label">
       <input
         type="checkbox"
-        checked={{eq @field.value true}}
+        checked={{or (eq @field.value true) (eq @field.value "true")}}
         class="form-kit__control-checkbox"
         disabled={{@field.disabled}}
         ...attributes
diff --git a/frontend/discourse/tests/integration/components/form-kit/controls/checkbox-test.gjs b/frontend/discourse/tests/integration/components/form-kit/controls/checkbox-test.gjs
index a4385f45bc..4caf312308 100644
--- a/frontend/discourse/tests/integration/components/form-kit/controls/checkbox-test.gjs
+++ b/frontend/discourse/tests/integration/components/form-kit/controls/checkbox-test.gjs
@@ -35,6 +35,31 @@ module(
       assert.deepEqual(data, { foo: true });
     });
 
+    test("when value is string 'false'", async function (assert) {
+      let data = { foo: "false" };
+      const mutateData = (x) => (data = x);
+
+      await render(
+        <template>
+          <Form @onSubmit={{mutateData}} @data={{data}} as |form|>
+            <form.Field @name="foo" @title="Foo" as |field|>
+              <field.Checkbox />
+            </form.Field>
+          </Form>
+        </template>
+      );
+
+      assert.form().field("foo").hasValue(false);
+
+      await formKit().field("foo").toggle();
+
+      assert.form().field("foo").hasValue(true);
+
+      await formKit().submit();
+
+      assert.deepEqual(data, { foo: true });
+    });
+
     test("when disabled", async function (assert) {
       await render(
         <template>
@@ -79,6 +104,31 @@ module(
       assert.form().field("foo").hasTitle("Foo (optional)");
     });
 
+    test("when value is string 'true'", async function (assert) {
+      let data = { foo: "true" };
+      const mutateData = (x) => (data = x);
+
+      await render(
+        <template>
+          <Form @onSubmit={{mutateData}} @data={{data}} as |form|>
+            <form.Field @name="foo" @title="Foo" as |field|>
+              <field.Checkbox />
+            </form.Field>
+          </Form>
+        </template>
+      );
+
+      assert.form().field("foo").hasValue(true);
+
+      await formKit().field("foo").toggle();
+
+      assert.form().field("foo").hasValue(false);
+
+      await formKit().submit();
+
+      assert.deepEqual(data, { foo: false });
+    });
+
     test("required", async function (assert) {
       await render(
         <template>
diff --git a/spec/requests/categories_controller_spec.rb b/spec/requests/categories_controller_spec.rb
index 584038303a..fdd526eb9a 100644
--- a/spec/requests/categories_controller_spec.rb
+++ b/spec/requests/categories_controller_spec.rb
@@ -914,6 +914,7 @@ RSpec.describe CategoriesController do
                 custom_fields: {
                   "dancing" => "frogs",
                   "running" => %w[turtle salamander],
+                  "enable_thingy" => true,
                 },
                 minimum_required_tags: "",
                 allow_global_tags: "true",
@@ -933,6 +934,7 @@ RSpec.describe CategoriesController do
           expect(category.custom_fields).to eq(
             "dancing" => "frogs",
             "running" => %w[turtle salamander],
+            "enable_thingy" => "true",
           )
           expect(category.minimum_required_tags).to eq(0)
           expect(category.allow_global_tags).to eq(true)
