diff --git a/spec/system/page_objects/components/composer.rb b/spec/system/page_objects/components/composer.rb
index 865fb74013746..543ccfcbadf7a 100644
--- a/spec/system/page_objects/components/composer.rb
+++ b/spec/system/page_objects/components/composer.rb
@@ -155,7 +155,7 @@ def locale
 
       def set_locale(locale)
         Components::DMenu.new(POST_LANGUAGE_SELECTOR).expand
-        find("#{POST_LANGUAGE_SELECTOR} button", text: locale).click
+        within("#d-menu-portals", visible: false) { find("button", text: locale).click }
       end
 
       def switch_category(category_name)
