diff --git a/lib/discourse.rb b/lib/discourse.rb
index de3734e521ba1..66ce2269b3c72 100644
--- a/lib/discourse.rb
+++ b/lib/discourse.rb
@@ -461,13 +461,16 @@ def self.find_plugin_js_assets(args)
     plugins.each do |plugin|
       if plugin.js_asset_exists?
         if ENV["ROLLUP_PLUGIN_COMPILER"] == "1"
-          assets << {
-            name: Plugin::JsManager.digested_logical_path_for(plugin.directory_name, "main"),
-            imports: Plugin::JsManager.import_paths_for(plugin.directory_name, "main"),
-            plugin: plugin,
-            type_module: true,
-            importmap_name: "discourse/plugins/#{plugin.directory_name}",
-          }
+          if logical_path =
+               Plugin::JsManager.digested_logical_path_for(plugin.directory_name, "main")
+            assets << {
+              name: logical_path,
+              imports: Plugin::JsManager.import_paths_for(plugin.directory_name, "main"),
+              plugin: plugin,
+              type_module: true,
+              importmap_name: "discourse/plugins/#{plugin.directory_name}",
+            }
+          end
         else
           assets << { name: "plugins/#{plugin.directory_name}", plugin: plugin }
         end
@@ -482,15 +485,16 @@ def self.find_plugin_js_assets(args)
       end
 
       if args[:include_admin_asset] && plugin.admin_js_asset_exists?
-        if ENV["ROLLUP_PLUGIN_COMPILER"] == "1" &&
-             logical_path =
+        if ENV["ROLLUP_PLUGIN_COMPILER"] == "1"
+          if logical_path =
                Plugin::JsManager.digested_logical_path_for(plugin.directory_name, "admin")
-          assets << {
-            name: logical_path,
-            imports: Plugin::JsManager.import_paths_for(plugin.directory_name, "admin"),
-            plugin: plugin,
-            type_module: true,
-          }
+            assets << {
+              name: logical_path,
+              imports: Plugin::JsManager.import_paths_for(plugin.directory_name, "admin"),
+              plugin: plugin,
+              type_module: true,
+            }
+          end
         else
           assets << { name: "plugins/#{plugin.directory_name}_admin", plugin: plugin }
         end
@@ -498,15 +502,16 @@ def self.find_plugin_js_assets(args)
 
       if args[:include_test_assets_for]&.include?(plugin.directory_name) &&
            plugin.test_js_asset_exists?
-        if ENV["ROLLUP_PLUGIN_COMPILER"] == "1" &&
-             logical_path =
+        if ENV["ROLLUP_PLUGIN_COMPILER"] == "1"
+          if logical_path =
                Plugin::JsManager.digested_logical_path_for(plugin.directory_name, "test")
-          assets << {
-            name: logical_path,
-            imports: Plugin::JsManager.import_paths_for(plugin.directory_name, "test"),
-            plugin: plugin,
-            type_module: true,
-          }
+            assets << {
+              name: logical_path,
+              imports: Plugin::JsManager.import_paths_for(plugin.directory_name, "test"),
+              plugin: plugin,
+              type_module: true,
+            }
+          end
         else
           assets << { name: "plugins/test/#{plugin.directory_name}_tests", plugin: plugin }
         end
diff --git a/lib/plugin/js_manager.rb b/lib/plugin/js_manager.rb
index 8bd76490f190a..b58e01f730999 100644
--- a/lib/plugin/js_manager.rb
+++ b/lib/plugin/js_manager.rb
@@ -2,15 +2,20 @@
 
 module Plugin
   class JsManager
-    @@manifest_data = {}
+    @manifest_data = {}
 
     def self.read_manifest(plugin_name)
       manifest_path = "#{Rails.root}/app/assets/generated/#{plugin_name}/manifest.json"
+      JSON.parse(File.read(manifest_path))
+    rescue Errno::ENOENT
+      {}
+    end
 
+    def self.parsed_manifest(plugin_name)
       if Rails.env.production?
-        @@manifest_data[plugin_name] ||= JSON.parse(File.read(manifest_path))
+        @manifest_data[plugin_name] ||= read_manifest(plugin_name)
       else
-        JSON.parse(File.read(manifest_path))
+        read_manifest(plugin_name)
       end
     end
 
