diff --git a/plugins/discourse-solved/assets/javascripts/discourse/initializers/solved-category-change-confirmation.gjs b/plugins/discourse-solved/assets/javascripts/discourse/initializers/solved-category-change-confirmation.gjs
index fbfe659b3e7b0..c17e1284ac70a 100644
--- a/plugins/discourse-solved/assets/javascripts/discourse/initializers/solved-category-change-confirmation.gjs
+++ b/plugins/discourse-solved/assets/javascripts/discourse/initializers/solved-category-change-confirmation.gjs
@@ -27,7 +27,7 @@ export default {
                 .split("|")
                 .filter(Boolean);
 
-              return tags.some((t) => solvedTags.includes(t.name));
+              return (tags || []).some((t) => solvedTags.includes(t.name));
             }
 
             @action
