diff --git a/app/assets/stylesheets/common/base/uc-modernize-foundation-theme.scss b/app/assets/stylesheets/common/base/uc-modernize-foundation-theme.scss
index 4007f315a1..9044f159ac 100644
--- a/app/assets/stylesheets/common/base/uc-modernize-foundation-theme.scss
+++ b/app/assets/stylesheets/common/base/uc-modernize-foundation-theme.scss
@@ -72,7 +72,7 @@
   }
 
   .btn.no-text.btn-icon {
-    &:not(.user-menu-tab, .user-menu-tab-active) {
+    &:not(.user-menu-tab, .user-menu-tab-active, .dropdown-select-box-header) {
       width: var(--button-height);
     }
   }
diff --git a/frontend/discourse/admin/components/admin-flag-item.gjs b/frontend/discourse/admin/components/admin-flag-item.gjs
index f882d71f17..bf608d22fc 100644
--- a/frontend/discourse/admin/components/admin-flag-item.gjs
+++ b/frontend/discourse/admin/components/admin-flag-item.gjs
@@ -178,7 +178,7 @@ export default class AdminFlagItem extends Component {
               @title={{i18n "admin.config_areas.flags.more_options.title"}}
               @icon="ellipsis-vertical"
               @onRegisterApi={{this.onRegisterApi}}
-              @class="btn-default"
+              @triggerClass="btn-default"
             >
               <:content>
                 <DropdownMenu as |dropdown|>
diff --git a/frontend/discourse/admin/components/api-key-item.gjs b/frontend/discourse/admin/components/api-key-item.gjs
index f847c32ed6..663ac138b6 100644
--- a/frontend/discourse/admin/components/api-key-item.gjs
+++ b/frontend/discourse/admin/components/api-key-item.gjs
@@ -127,7 +127,7 @@ export default class ApiKeyItem extends Component {
             @title={{i18n "admin.config_areas.user_fields.more_options.title"}}
             @icon="ellipsis-vertical"
             @onRegisterApi={{this.onRegisterApi}}
-            @class="btn-default"
+            @triggerClass="btn-default"
           >
             <:content>
               <DropdownMenu as |dropdown|>
diff --git a/frontend/discourse/app/components/bulk-select-topics-dropdown.gjs b/frontend/discourse/app/components/bulk-select-topics-dropdown.gjs
index ea25152bcc..351bcc2f30 100644
--- a/frontend/discourse/app/components/bulk-select-topics-dropdown.gjs
+++ b/frontend/discourse/app/components/bulk-select-topics-dropdown.gjs
@@ -317,7 +317,7 @@ export default class BulkSelectTopicsDropdown extends Component {
       @autofocus={{true}}
       @identifier="bulk-select-topics-dropdown"
       @onRegisterApi={{this.onRegisterApi}}
-      @class="btn-default"
+      @triggerClass="btn-default"
     >
       <:trigger>
         <span class="d-button-label">
diff --git a/frontend/discourse/app/components/post-language-selector.gjs b/frontend/discourse/app/components/post-language-selector.gjs
index 1ccaf38b3c..c1e4e72fe1 100644
--- a/frontend/discourse/app/components/post-language-selector.gjs
+++ b/frontend/discourse/app/components/post-language-selector.gjs
@@ -47,7 +47,7 @@ export default class PostLanguageSelector extends Component {
       @label={{this.selectedLanguage}}
       @modalForMobile={{true}}
       @onRegisterApi={{this.onRegisterApi}}
-      @class="btn-transparent post-language-selector"
+      @triggerClass="btn-transparent"
     >
       <:content>
         <DropdownMenu as |dropdown|>
