diff --git a/frontend/discourse/app/services/scroll-direction.js b/frontend/discourse/app/services/scroll-direction.js
index 865b3b87580f0..7f96bb64aca60 100644
--- a/frontend/discourse/app/services/scroll-direction.js
+++ b/frontend/discourse/app/services/scroll-direction.js
@@ -34,6 +34,7 @@ export default class ScrollDirection extends Service {
 
   willDestroy() {
     window.removeEventListener("scroll", this.onScroll);
+    this.router.off("routeWillChange", this.routeWillChange);
     this.router.off("routeDidChange", this.routeDidChange);
   }
 
