diff --git a/app/serializers/site_category_serializer.rb b/app/serializers/site_category_serializer.rb
index d1d3ea6044e9a..beb9c0cb3863c 100644
--- a/app/serializers/site_category_serializer.rb
+++ b/app/serializers/site_category_serializer.rb
@@ -51,13 +51,28 @@ def name
   end
 
   def description
-    translated_description =
-      if (ContentLocalization.show_translated_category?(object, scope))
-        object.get_localization&.description
-      else
-        object.description
-      end
+    localized_description || object.description
+  end
+
+  def description_text
+    return super if object.uncategorized?
+    return ERB::Util.html_escape(localized_description).html_safe if localized_description
+    object.description_text
+  end
+
+  def description_excerpt
+    return super if object.uncategorized?
+    return PrettyText.excerpt(localized_description, 300) if localized_description
+    object.description_excerpt
+  end
 
-    translated_description || object.description
+  private
+
+  def localized_description
+    return @localized_description if defined?(@localized_description)
+    @localized_description =
+      if ContentLocalization.show_translated_category?(object, scope)
+        object.get_localization&.description.presence
+      end
   end
 end
diff --git a/spec/serializers/category_serializer_spec.rb b/spec/serializers/category_serializer_spec.rb
index cecc8937a04af..2ea0311b9ed6b 100644
--- a/spec/serializers/category_serializer_spec.rb
+++ b/spec/serializers/category_serializer_spec.rb
@@ -138,7 +138,7 @@
         category: category_with_localization,
         locale: "ja",
         name: "日本語名",
-        description: "日本語の説明",
+        description: "<p>最初の段落</p><p>二番目の段落</p>",
       )
     end
 
@@ -153,7 +153,7 @@
       expect(json[:description]).to eq("Original Description")
     end
 
-    it "returns translated name and description for SiteCategorySerializer when enabled" do
+    it "returns translated attributes for SiteCategorySerializer when enabled" do
       SiteSetting.content_localization_enabled = true
       user.update!(locale: "ja")
       I18n.with_locale("ja") do
@@ -164,7 +164,26 @@
             root: false,
           ).as_json
         expect(json[:name]).to eq("日本語名")
-        expect(json[:description]).to eq("日本語の説明")
+        expect(json[:description]).to eq("<p>最初の段落</p><p>二番目の段落</p>")
+        expect(json[:description_text]).to eq("&lt;p&gt;最初の段落&lt;/p&gt;&lt;p&gt;二番目の段落&lt;/p&gt;")
+        expect(json[:description_excerpt]).to eq("最初の段落 二番目の段落")
+      end
+    end
+
+    it "returns untranslated attributes for SiteCategorySerializer when disabled" do
+      SiteSetting.content_localization_enabled = false
+      user.update!(locale: "ja")
+      I18n.with_locale("ja") do
+        json =
+          SiteCategorySerializer.new(
+            category_with_localization,
+            scope: Guardian.new(user),
+            root: false,
+          ).as_json
+        expect(json[:name]).to eq("Original Name")
+        expect(json[:description]).to eq("Original Description")
+        expect(json[:description_text]).to eq(category_with_localization.description_text)
+        expect(json[:description_excerpt]).to eq(category_with_localization.description_excerpt)
       end
     end
 
