diff --git a/frontend/discourse/select-kit/components/select-kit.js b/frontend/discourse/select-kit/components/select-kit.js
index 456b9ff31f21f..b25371a0ed274 100644
--- a/frontend/discourse/select-kit/components/select-kit.js
+++ b/frontend/discourse/select-kit/components/select-kit.js
@@ -976,7 +976,7 @@ export default class SelectKit extends Component {
       this.cleanupFloatingUi?.();
       this.cleanupFloatingUi = autoUpdate(
         this.getHeader(),
-        this._mainElement(),
+        this._bodyElement(),
         () => this.updateFloatingUiPosition()
       );
     } else {
