diff --git a/frontend/discourse/app/components/reviewable-claimed-topic.gjs b/frontend/discourse/app/components/reviewable-claimed-topic.gjs
index 81e9055b15..64d5ee6003 100644
--- a/frontend/discourse/app/components/reviewable-claimed-topic.gjs
+++ b/frontend/discourse/app/components/reviewable-claimed-topic.gjs
@@ -15,7 +15,7 @@ export default class ReviewableClaimedTopic extends Component {
   }
 
   get isRefresh() {
-    return this.siteSettings.reviewable_ui_refresh;
+    return this.currentUser?.use_reviewable_ui_refresh;
   }
 
   @action
