diff --git a/plugins/discourse-ai/db/fixtures/agents/603_ai_agents.rb b/plugins/discourse-ai/db/fixtures/agents/603_ai_agents.rb
index c184528e22a1c..c14d327cf41d0 100644
--- a/plugins/discourse-ai/db/fixtures/agents/603_ai_agents.rb
+++ b/plugins/discourse-ai/db/fixtures/agents/603_ai_agents.rb
@@ -1,5 +1,13 @@
 # frozen_string_literal: true
 
+# during transition period prior to rename of persona table
+# it is a view, we can not write to it, post migration will take care
+ai_agents_is_view =
+  DB.query_single(
+    "SELECT 1 FROM pg_class WHERE relname = 'ai_agents' AND relkind IN ('v')",
+  ).present?
+return if ai_agents_is_view
+
 summarization_agents = [DiscourseAi::Agents::Summarizer, DiscourseAi::Agents::ShortSummarizer]
 
 def from_setting(setting_name)
