diff --git a/public/app/features/panel/panellinks/link_srv.ts b/public/app/features/panel/panellinks/link_srv.ts
index a76765e686e..3c1dd1fcfaa 100644
--- a/public/app/features/panel/panellinks/link_srv.ts
+++ b/public/app/features/panel/panellinks/link_srv.ts
@@ -84,7 +84,7 @@ const isRecordOrArray = (value: unknown): value is Record<string, unknown> | unk
   typeof value === 'object' && value !== null;
 
 const getVariableValueProperties = (variable: TypedVariableModel): string[] => {
-  if (!('options' in variable) || !variable.options[0].properties) {
+  if (!('options' in variable) || !variable.options?.[0]?.properties) {
     return [];
   }
 
