diff --git a/public/app/features/panel/components/VizTypePicker/VisualizationSuggestions.tsx b/public/app/features/panel/components/VizTypePicker/VisualizationSuggestions.tsx
index d72788004b3..58401a8c698 100644
--- a/public/app/features/panel/components/VizTypePicker/VisualizationSuggestions.tsx
+++ b/public/app/features/panel/components/VizTypePicker/VisualizationSuggestions.tsx
@@ -299,7 +299,7 @@ export function VisualizationSuggestions({ onChange, data, panel, searchQuery, i
                   suggestion={suggestion}
                   width={width}
                   tabIndex={index}
-                  onClick={() => applySuggestion(suggestion, index)}
+                  onClick={() => applySuggestion(suggestion, index, false, true)}
                 />
               </div>
             ))}
