diff --git a/public/app/features/alerting/unified/NotificationPoliciesPage.test.tsx b/public/app/features/alerting/unified/NotificationPoliciesPage.test.tsx
index 0cc324cff7a58..441b669bf7b95 100644
--- a/public/app/features/alerting/unified/NotificationPoliciesPage.test.tsx
+++ b/public/app/features/alerting/unified/NotificationPoliciesPage.test.tsx
@@ -712,7 +712,7 @@ describe('alertingNavigationV2 respects alertingMultiplePolicies', () => {
       );
       await screen.findByTestId('search-query-input');
 
-      expect(screen.getAllByTestId('am-root-route-container').length).toBeGreaterThan(0);
+      expect((await screen.findAllByTestId('am-root-route-container')).length).toBeGreaterThan(0);
     });
   });
 });
