diff --git a/public/app/plugins/panel/heatmap/fields.test.ts b/public/app/plugins/panel/heatmap/fields.test.ts
index 80c77311619..631fab1fcd7 100644
--- a/public/app/plugins/panel/heatmap/fields.test.ts
+++ b/public/app/plugins/panel/heatmap/fields.test.ts
@@ -1,5 +1,6 @@
-import { createTheme } from '@grafana/data';
+import { createTheme, dateTime, FieldType } from '@grafana/data';
 
+import { prepareHeatmapData } from './fields';
 import { Options } from './panelcfg.gen';
 
 const theme = createTheme();
@@ -7,8 +8,36 @@ const theme = createTheme();
 describe('Heatmap data', () => {
   const options: Options = {} as Options;
 
-  it('simple test stub', () => {
-    expect(theme).toBeDefined();
-    expect(options).toBeDefined();
+  const tpl = {
+    frames: [],
+    annotations: [],
+    options,
+    palette: [],
+    theme,
+    replaceVariables: undefined,
+    timeRange: { from: dateTime(), to: dateTime(), raw: { from: 'now-1h', to: 'now' } },
+  };
+
+  it('omit empty series array', () => {
+    const info = prepareHeatmapData({
+      ...tpl,
+      frames: [],
+    });
+
+    expect(info).toEqual({});
+  });
+
+  it('omit frame.length: 0', () => {
+    const info = prepareHeatmapData({
+      ...tpl,
+      frames: [
+        {
+          fields: [{ name: '', config: {}, type: FieldType.time, values: [] }],
+          length: 0,
+        },
+      ],
+    });
+
+    expect(info).toEqual({});
   });
 });
diff --git a/public/app/plugins/panel/heatmap/fields.ts b/public/app/plugins/panel/heatmap/fields.ts
index 62e1b5992d8..b698688e59a 100644
--- a/public/app/plugins/panel/heatmap/fields.ts
+++ b/public/app/plugins/panel/heatmap/fields.ts
@@ -89,6 +89,11 @@ export function prepareHeatmapData({
   replaceVariables = (v) => v,
   timeRange,
 }: PrepareHeatmapDataOptions): HeatmapData {
+  // exclude empty frames
+  frames = frames.filter(
+    (frame) => frame.length > 0 && frame.fields.length > 0 && frame.fields.every((field) => field.values.length > 0)
+  );
+
   if (!frames?.length) {
     return {};
   }
