diff --git a/.github/actions/change-detection/action.yml b/.github/actions/change-detection/action.yml
index a040c37af56..73071e86c98 100644
--- a/.github/actions/change-detection/action.yml
+++ b/.github/actions/change-detection/action.yml
@@ -78,6 +78,7 @@ runs:
             - '.github/actions/change-detection/**'
             - '**.cue'
             - 'devenv/docker/blocks/*_tests/**'
+            - 'scripts/ci/backend-tests/**'
             - 'kindsv2/**'
             - '${{ inputs.self }}'
           frontend:
diff --git a/scripts/ci/backend-tests/pkgs-with-tests-named.sh b/scripts/ci/backend-tests/pkgs-with-tests-named.sh
index 82a677cab24..b2f586a086a 100755
--- a/scripts/ci/backend-tests/pkgs-with-tests-named.sh
+++ b/scripts/ci/backend-tests/pkgs-with-tests-named.sh
@@ -60,7 +60,7 @@ fi
 readarray -t PACKAGES <<< "$(go list -f '{{.Dir}}' -e "${dirs[@]}")"
 
 for i in "${!PACKAGES[@]}"; do
-    readarray -t PKG_FILES <<< "$(find "${PACKAGES[$i]}" -type f -name '*_test.go')"
+    readarray -t PKG_FILES <<< "$(find "${PACKAGES[$i]}" -maxdepth 1 -type f -name '*_test.go')"
     if [ ${#PKG_FILES[@]} -eq 0 ] || [ ${#PKG_FILES[@]} -eq 1 ] && [ -z "${PKG_FILES[0]}" ]; then
         unset "PACKAGES[$i]"
         continue
