diff --git a/public/app/features/panel/components/VizTypePicker/VisualizationSuggestionCard.test.tsx b/public/app/features/panel/components/VizTypePicker/VisualizationSuggestionCard.test.tsx
index c10cb478945..c0d4b1da24b 100644
--- a/public/app/features/panel/components/VizTypePicker/VisualizationSuggestionCard.test.tsx
+++ b/public/app/features/panel/components/VizTypePicker/VisualizationSuggestionCard.test.tsx
@@ -55,7 +55,7 @@ describe('VisualizationSuggestionCard', () => {
   it('should render a panel renderer card when no imgSrc is provided', () => {
     render(<VisualizationSuggestionCard data={mockData} suggestion={baseSuggestion} width={100} />);
 
-    expect(screen.getByRole('button', { name: 'Time series' })).toBeInTheDocument();
+    expect(screen.getByLabelText('Time series')).toBeInTheDocument();
     expect(screen.getByTestId('panel-renderer')).toBeInTheDocument();
   });
 
@@ -88,6 +88,6 @@ describe('VisualizationSuggestionCard', () => {
 
     render(<VisualizationSuggestionCard data={mockData} suggestion={baseSuggestion} width={200} />);
 
-    expect(screen.getByRole('button', { name: 'Time series' })).toBeInTheDocument();
+    expect(screen.getByLabelText('Time series')).toBeInTheDocument();
   });
 });
diff --git a/public/app/features/panel/components/VizTypePicker/VisualizationSuggestionCard.tsx b/public/app/features/panel/components/VizTypePicker/VisualizationSuggestionCard.tsx
index 7984fd9d423..e6370216adb 100644
--- a/public/app/features/panel/components/VizTypePicker/VisualizationSuggestionCard.tsx
+++ b/public/app/features/panel/components/VizTypePicker/VisualizationSuggestionCard.tsx
@@ -9,7 +9,7 @@ import { Tooltip, useStyles2 } from '@grafana/ui';
 
 import { PanelRenderer } from '../PanelRenderer';
 
-export interface Props extends HTMLAttributes<HTMLButtonElement> {
+export interface Props extends HTMLAttributes<HTMLDivElement> {
   data: PanelData;
   width: number;
   suggestion: PanelPluginVisualizationSuggestion;
@@ -26,18 +26,17 @@ export function VisualizationSuggestionCard({ data, suggestion, width, className
     className: cx(className, styles.vizBox),
     'data-testid': selectors.components.VisualizationPreview.card(suggestion.name),
     style: outerStyles,
-    tabIndex: -1, // selection is handled by parent container
     ...restProps,
-  } satisfies HTMLAttributes<HTMLButtonElement> & { 'data-testid': string };
+  } satisfies HTMLAttributes<HTMLDivElement> & { 'data-testid': string };
 
   let content: ReactNode;
 
   if (cardOptions.imgSrc) {
     content = (
-      <button {...commonButtonProps} className={cx(commonButtonProps.className, styles.imgBox)}>
+      <div {...commonButtonProps} className={cx(commonButtonProps.className, styles.imgBox)}>
         <div className={styles.name}>{suggestion.name}</div>
         <img className={styles.img} src={cardOptions.imgSrc} alt={suggestion.name} />
-      </button>
+      </div>
     );
   } else {
     let preview = suggestion;
@@ -47,7 +46,7 @@ export function VisualizationSuggestionCard({ data, suggestion, width, className
     }
 
     content = (
-      <button {...commonButtonProps}>
+      <div {...commonButtonProps}>
         {/* to use inert in React 18, we have to do this hacky object spread thing. https://stackoverflow.com/questions/72720469/error-when-using-inert-attribute-with-typescript */}
         <div style={innerStyles} className={styles.renderContainer} {...{ inert: '' }}>
           <PanelRenderer
@@ -60,7 +59,7 @@ export function VisualizationSuggestionCard({ data, suggestion, width, className
             fieldConfig={preview.fieldConfig}
           />
         </div>
-      </button>
+      </div>
     );
   }
 
