diff --git a/public/app/features/alerting/unified/notifications/NotificationDetailPage.tsx b/public/app/features/alerting/unified/notifications/NotificationDetailPage.tsx
index ca400cf480183..d179ac9c875b8 100644
--- a/public/app/features/alerting/unified/notifications/NotificationDetailPage.tsx
+++ b/public/app/features/alerting/unified/notifications/NotificationDetailPage.tsx
@@ -43,7 +43,7 @@ function NotificationDetailPage() {
   const pageNav = { text: pageTitle };
 
   return (
-    <AlertingPageWrapper navId="alerts-notifications" pageNav={pageNav} isLoading={false}>
+    <AlertingPageWrapper navId="alerts-history" pageNav={pageNav} isLoading={false}>
       {uuid ? (
         <NotificationDetail uuid={uuid} timestamp={timestamp} onTitleChange={setPageTitle} />
       ) : (
