diff --git a/public/app/features/dashboard-scene/scene/UnconfiguredPanel.tsx b/public/app/features/dashboard-scene/scene/UnconfiguredPanel.tsx
index 07f15cd8999..cfa1aeec28d 100644
--- a/public/app/features/dashboard-scene/scene/UnconfiguredPanel.tsx
+++ b/public/app/features/dashboard-scene/scene/UnconfiguredPanel.tsx
@@ -1,5 +1,5 @@
 import { css } from '@emotion/css';
-import { useCallback, useEffect, useState } from 'react';
+import { useCallback, useState } from 'react';
 
 import { CoreApp, GrafanaTheme2, PanelPlugin, PanelProps } from '@grafana/data';
 import { Trans, t } from '@grafana/i18n';
@@ -19,7 +19,6 @@ import {
   useStyles2,
 } from '@grafana/ui';
 
-import { NEW_PANEL_TITLE } from '../../dashboard/utils/dashboard';
 import { DashboardInteractions } from '../utils/interactions';
 import { findVizPanelByKey, getVizPanelKeyForPanelId } from '../utils/utils';
 
@@ -64,18 +63,6 @@ function UnconfiguredPanelComp(props: PanelProps) {
     dashboard.onShowAddLibraryPanelDrawer(panel.getRef());
   };
 
-  useEffect(() => {
-    if (!panel || !config.featureToggles.newVizSuggestions) {
-      return;
-    }
-
-    if (panelContext.app === CoreApp.PanelEditor) {
-      panel.setState({ title: '' });
-    } else if (!panel.state.title) {
-      panel.setState({ title: NEW_PANEL_TITLE });
-    }
-  }, [panel, panelContext.app]);
-
   const MenuActions = () => (
     <Menu>
       <Menu.Item
diff --git a/public/app/features/dashboard-scene/utils/utils.ts b/public/app/features/dashboard-scene/utils/utils.ts
index 361564f85a3..8526e9a8bd1 100644
--- a/public/app/features/dashboard-scene/utils/utils.ts
+++ b/public/app/features/dashboard-scene/utils/utils.ts
@@ -268,10 +268,7 @@ export function getDefaultPluginId(): string {
 export function getDefaultVizPanel(): VizPanel {
   const defaultPluginId = getDefaultPluginId();
 
-  const newPanelTitle =
-    config.featureToggles.newVizSuggestions && defaultPluginId === UNCONFIGURED_PANEL_PLUGIN_ID
-      ? ''
-      : t('dashboard.new-panel-title', 'New panel');
+  const newPanelTitle = t('dashboard.new-panel-title', 'New panel');
 
   const datasourceSettings = getDataSourceSrv().getInstanceSettings(null);
 
