diff --git a/tests/utils-shared/src/main/java/org/keycloak/testsuite/util/oauth/oid4vc/CredentialIssuerMetadataRequest.java b/tests/utils-shared/src/main/java/org/keycloak/testsuite/util/oauth/oid4vc/CredentialIssuerMetadataRequest.java
index df0538ced5..3d2062d3bf 100644
--- a/tests/utils-shared/src/main/java/org/keycloak/testsuite/util/oauth/oid4vc/CredentialIssuerMetadataRequest.java
+++ b/tests/utils-shared/src/main/java/org/keycloak/testsuite/util/oauth/oid4vc/CredentialIssuerMetadataRequest.java
@@ -9,7 +9,7 @@ import org.apache.http.client.methods.CloseableHttpResponse;
 
 public class CredentialIssuerMetadataRequest extends AbstractHttpGetRequest<CredentialIssuerMetadataRequest, CredentialIssuerMetadataResponse> {
 
-    public CredentialIssuerMetadataRequest(AbstractOAuthClient<?> client) {
+    CredentialIssuerMetadataRequest(AbstractOAuthClient<?> client) {
         super(client);
     }
 
diff --git a/tests/utils-shared/src/main/java/org/keycloak/testsuite/util/oauth/oid4vc/CredentialOfferRequest.java b/tests/utils-shared/src/main/java/org/keycloak/testsuite/util/oauth/oid4vc/CredentialOfferRequest.java
index a839c31b2f..b7af1c5acf 100644
--- a/tests/utils-shared/src/main/java/org/keycloak/testsuite/util/oauth/oid4vc/CredentialOfferRequest.java
+++ b/tests/utils-shared/src/main/java/org/keycloak/testsuite/util/oauth/oid4vc/CredentialOfferRequest.java
@@ -13,12 +13,12 @@ public class CredentialOfferRequest extends AbstractHttpGetRequest<CredentialOff
 
     private final CredentialOfferURI credOfferURI;
 
-    public CredentialOfferRequest(AbstractOAuthClient<?> client, CredentialOfferURI credOfferUri) {
+    CredentialOfferRequest(AbstractOAuthClient<?> client, CredentialOfferURI credOfferUri) {
         super(client);
         this.credOfferURI = credOfferUri;
     }
 
-    public CredentialOfferRequest(AbstractOAuthClient<?> client, String nonce) {
+    CredentialOfferRequest(AbstractOAuthClient<?> client, String nonce) {
         super(client);
         credOfferURI = new CredentialOfferURI();
         credOfferURI.setIssuer(client.getEndpoints().getOid4vcCredentialOffer());
diff --git a/tests/utils-shared/src/main/java/org/keycloak/testsuite/util/oauth/oid4vc/CredentialOfferUriRequest.java b/tests/utils-shared/src/main/java/org/keycloak/testsuite/util/oauth/oid4vc/CredentialOfferUriRequest.java
index fc6fa7ef17..b12a575c95 100644
--- a/tests/utils-shared/src/main/java/org/keycloak/testsuite/util/oauth/oid4vc/CredentialOfferUriRequest.java
+++ b/tests/utils-shared/src/main/java/org/keycloak/testsuite/util/oauth/oid4vc/CredentialOfferUriRequest.java
@@ -16,7 +16,7 @@ public class CredentialOfferUriRequest extends AbstractHttpGetRequest<Credential
     private String username;
     private String clientId;
 
-    public CredentialOfferUriRequest(AbstractOAuthClient<?> client, String credConfigId) {
+    CredentialOfferUriRequest(AbstractOAuthClient<?> client, String credConfigId) {
         super(client);
         this.credConfigId = credConfigId;
     }
diff --git a/tests/utils-shared/src/main/java/org/keycloak/testsuite/util/oauth/oid4vc/OID4VCClient.java b/tests/utils-shared/src/main/java/org/keycloak/testsuite/util/oauth/oid4vc/OID4VCClient.java
index 42e04c9f14..1c46b6f50e 100644
--- a/tests/utils-shared/src/main/java/org/keycloak/testsuite/util/oauth/oid4vc/OID4VCClient.java
+++ b/tests/utils-shared/src/main/java/org/keycloak/testsuite/util/oauth/oid4vc/OID4VCClient.java
@@ -34,7 +34,7 @@ public class OID4VCClient {
     }
 
     public Oid4vcCredentialRequest credentialRequest() {
-        return new Oid4vcCredentialRequest(client, new CredentialRequest());
+        return credentialRequest(new CredentialRequest());
     }
 
     public Oid4vcCredentialRequest credentialRequest(CredentialRequest credRequest) {
diff --git a/tests/utils-shared/src/main/java/org/keycloak/testsuite/util/oauth/oid4vc/Oid4vcCredentialRequest.java b/tests/utils-shared/src/main/java/org/keycloak/testsuite/util/oauth/oid4vc/Oid4vcCredentialRequest.java
index 83e81131dc..ecba7ba23e 100644
--- a/tests/utils-shared/src/main/java/org/keycloak/testsuite/util/oauth/oid4vc/Oid4vcCredentialRequest.java
+++ b/tests/utils-shared/src/main/java/org/keycloak/testsuite/util/oauth/oid4vc/Oid4vcCredentialRequest.java
@@ -16,7 +16,7 @@ public class Oid4vcCredentialRequest extends AbstractHttpPostRequest<Oid4vcCrede
 
     private final CredentialRequest credRequest;
 
-    public Oid4vcCredentialRequest(AbstractOAuthClient<?> client, CredentialRequest credRequest) {
+    Oid4vcCredentialRequest(AbstractOAuthClient<?> client, CredentialRequest credRequest) {
         super(client);
         this.credRequest = credRequest;
     }
diff --git a/tests/utils-shared/src/main/java/org/keycloak/testsuite/util/oauth/oid4vc/Oid4vcNonceRequest.java b/tests/utils-shared/src/main/java/org/keycloak/testsuite/util/oauth/oid4vc/Oid4vcNonceRequest.java
index 296e14a0f0..e992842390 100644
--- a/tests/utils-shared/src/main/java/org/keycloak/testsuite/util/oauth/oid4vc/Oid4vcNonceRequest.java
+++ b/tests/utils-shared/src/main/java/org/keycloak/testsuite/util/oauth/oid4vc/Oid4vcNonceRequest.java
@@ -9,7 +9,7 @@ import org.apache.http.client.methods.CloseableHttpResponse;
 
 public class Oid4vcNonceRequest extends AbstractHttpPostRequest<Oid4vcNonceRequest, Oid4vcNonceResponse> {
 
-    public Oid4vcNonceRequest(AbstractOAuthClient<?> client) {
+    Oid4vcNonceRequest(AbstractOAuthClient<?> client) {
         super(client);
     }
 
diff --git a/tests/utils-shared/src/main/java/org/keycloak/testsuite/util/oauth/oid4vc/PreAuthorizedCodeGrantRequest.java b/tests/utils-shared/src/main/java/org/keycloak/testsuite/util/oauth/oid4vc/PreAuthorizedCodeGrantRequest.java
index a8a4d7e7ee..fdb409e881 100644
--- a/tests/utils-shared/src/main/java/org/keycloak/testsuite/util/oauth/oid4vc/PreAuthorizedCodeGrantRequest.java
+++ b/tests/utils-shared/src/main/java/org/keycloak/testsuite/util/oauth/oid4vc/PreAuthorizedCodeGrantRequest.java
@@ -17,7 +17,7 @@ public class PreAuthorizedCodeGrantRequest extends AbstractHttpPostRequest<PreAu
 
     private final String preAuthorizedCode;
 
-    public PreAuthorizedCodeGrantRequest(AbstractOAuthClient<?> client, String preAuthorizedCode) {
+    PreAuthorizedCodeGrantRequest(AbstractOAuthClient<?> client, String preAuthorizedCode) {
         super(client);
         this.preAuthorizedCode = preAuthorizedCode;
     }
diff --git a/testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/oid4vc/issuance/signing/OID4VCAuthorizationDetailsFlowTestBase.java b/testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/oid4vc/issuance/signing/OID4VCAuthorizationDetailsFlowTestBase.java
index 1f24c627de..76ab85e1e3 100644
--- a/testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/oid4vc/issuance/signing/OID4VCAuthorizationDetailsFlowTestBase.java
+++ b/testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/oid4vc/issuance/signing/OID4VCAuthorizationDetailsFlowTestBase.java
@@ -49,6 +49,7 @@ import org.keycloak.testsuite.util.oauth.oid4vc.CredentialIssuerMetadataResponse
 import org.keycloak.testsuite.util.oauth.oid4vc.CredentialOfferResponse;
 import org.keycloak.testsuite.util.oauth.oid4vc.CredentialOfferUriResponse;
 import org.keycloak.testsuite.util.oauth.oid4vc.Oid4vcCredentialResponse;
+import org.keycloak.util.JsonSerialization;
 
 import org.apache.http.HttpHeaders;
 import org.apache.http.HttpStatus;
