diff --git a/services/src/main/java/org/keycloak/services/managers/AuthenticationManager.java b/services/src/main/java/org/keycloak/services/managers/AuthenticationManager.java
index 5ff72ff6ad..bd98c0afbc 100755
--- a/services/src/main/java/org/keycloak/services/managers/AuthenticationManager.java
+++ b/services/src/main/java/org/keycloak/services/managers/AuthenticationManager.java
@@ -1088,6 +1088,7 @@ public class AuthenticationManager {
 
         if (authSession.getAuthNote(END_AFTER_REQUIRED_ACTIONS) != null) {
             LoginFormsProvider infoPage = session.getProvider(LoginFormsProvider.class).setAuthenticationSession(authSession)
+                    .setAttribute("messageHeader", Messages.ACCOUNT_UPDATED_TITLE)
                     .setSuccess(Messages.ACCOUNT_UPDATED);
             if (authSession.getAuthNote(SET_REDIRECT_URI_AFTER_REQUIRED_ACTIONS) != null) {
                 if (authSession.getRedirectUri() != null) {
diff --git a/services/src/main/java/org/keycloak/services/messages/Messages.java b/services/src/main/java/org/keycloak/services/messages/Messages.java
index a3e4de7edc..710ec6f82e 100755
--- a/services/src/main/java/org/keycloak/services/messages/Messages.java
+++ b/services/src/main/java/org/keycloak/services/messages/Messages.java
@@ -140,6 +140,8 @@ public class Messages {
 
     public static final String ACCOUNT_UPDATED = "accountUpdatedMessage";
 
+    public static final String ACCOUNT_UPDATED_TITLE = "accountUpdatedTitle";
+
     public static final String ACCOUNT_PASSWORD_UPDATED = "accountPasswordUpdatedMessage";
 
     public static final String NO_ACCESS = "noAccessMessage";
diff --git a/tests/base/src/test/java/org/keycloak/tests/admin/user/UserEmailTest.java b/tests/base/src/test/java/org/keycloak/tests/admin/user/UserEmailTest.java
index e0279d5520..3b4a568cee 100644
--- a/tests/base/src/test/java/org/keycloak/tests/admin/user/UserEmailTest.java
+++ b/tests/base/src/test/java/org/keycloak/tests/admin/user/UserEmailTest.java
@@ -614,7 +614,7 @@ public class UserEmailTest extends AbstractUserTest {
 
         passwordUpdatePage.changePassword("new-pass", "new-pass");
 
-        assertEquals("Your account has been updated.", driver.findElement(By.id("kc-page-title")).getText());
+        assertEquals("Account updated", driver.findElement(By.id("kc-page-title")).getText());
 
         String pageSource = driver.page().getPageSource();
 
@@ -696,7 +696,7 @@ public class UserEmailTest extends AbstractUserTest {
 
         passwordUpdatePage.changePassword("new-pass", "new-pass");
 
-        assertEquals("Your account has been updated.", driver.findElement(By.id("kc-page-title")).getText());
+        assertEquals("Account updated", driver.findElement(By.id("kc-page-title")).getText());
 
         String pageSource = driver.page().getPageSource();
 
diff --git a/themes/src/main/resources/theme/base/login/messages/messages_en.properties b/themes/src/main/resources/theme/base/login/messages/messages_en.properties
index c347d4e80c..2cb321c801 100644
--- a/themes/src/main/resources/theme/base/login/messages/messages_en.properties
+++ b/themes/src/main/resources/theme/base/login/messages/messages_en.properties
@@ -311,6 +311,7 @@ linkIdpMessage=You need to verify your email address to link your account with {
 emailSentMessage=You should receive an email shortly with further instructions.
 emailSendErrorMessage=Failed to send email, please try again later.
 
+accountUpdatedTitle=Account updated
 accountUpdatedMessage=Your account has been updated.
 accountPasswordUpdatedMessage=Your password has been updated.
 
