diff --git a/js/apps/admin-ui/src/utils/useSetTimeout.ts b/js/apps/admin-ui/src/utils/useSetTimeout.ts
index 0c605fd8ab..502075f0de 100644
--- a/js/apps/admin-ui/src/utils/useSetTimeout.ts
+++ b/js/apps/admin-ui/src/utils/useSetTimeout.ts
@@ -4,13 +4,14 @@ export default function useSetTimeout() {
   const didUnmountRef = useRef(false);
   const { current: scheduledTimers } = useRef(new Set<number>());
 
-  useEffect(
-    () => () => {
+  useEffect(() => {
+    didUnmountRef.current = false;
+
+    return () => {
       didUnmountRef.current = true;
       clearAll();
-    },
-    [],
-  );
+    };
+  }, []);
 
   function clearAll() {
     scheduledTimers.forEach((timer) => clearTimeout(timer));
