diff --git a/core/src/main/java/org/keycloak/jose/jwk/JWKBuilder.java b/core/src/main/java/org/keycloak/jose/jwk/JWKBuilder.java
index 1b84bb26b268..d799c73809a0 100644
--- a/core/src/main/java/org/keycloak/jose/jwk/JWKBuilder.java
+++ b/core/src/main/java/org/keycloak/jose/jwk/JWKBuilder.java
@@ -39,7 +39,7 @@
  */
 public class JWKBuilder {
 
-    public static final String DEFAULT_PUBLIC_KEY_USE = "sig";
+    public static final KeyUse DEFAULT_PUBLIC_KEY_USE = KeyUse.SIG;
 
     private String kid;
 
@@ -105,13 +105,17 @@ public JWK rsa(Key key, List<X509Certificate> certificates, KeyUse keyUse) {
 
     public JWK rsa(Key key, KeyUse keyUse) {
         JWK k = rsa(key);
-        String keyUseString = keyUse == null ? DEFAULT_PUBLIC_KEY_USE : keyUse.getSpecName();
+        String keyUseString = keyUse == null ? DEFAULT_PUBLIC_KEY_USE.getSpecName() : keyUse.getSpecName();
         if (KeyUse.ENC == keyUse) keyUseString = "enc";
         k.setPublicKeyUse(keyUseString);
         return k;
     }
 
     public JWK ec(Key key) {
+        return ec(key, DEFAULT_PUBLIC_KEY_USE);
+    }
+
+    public JWK ec(Key key, KeyUse keyUse) {
         ECPublicKey ecKey = (ECPublicKey) key;
 
         ECPublicJWK k = new ECPublicJWK();
@@ -122,7 +126,7 @@ public JWK ec(Key key) {
         k.setKeyId(kid);
         k.setKeyType(KeyType.EC);
         k.setAlgorithm(algorithm);
-        k.setPublicKeyUse(DEFAULT_PUBLIC_KEY_USE);
+        k.setPublicKeyUse(keyUse == null ? DEFAULT_PUBLIC_KEY_USE.getSpecName() : keyUse.getSpecName());
         k.setCrv("P-" + fieldSize);
         k.setX(Base64Url.encode(toIntegerBytes(ecKey.getW().getAffineX(), fieldSize)));
         k.setY(Base64Url.encode(toIntegerBytes(ecKey.getW().getAffineY(), fieldSize)));
diff --git a/js/apps/account-ui/package.json b/js/apps/account-ui/package.json
index ea1533a80f47..bcb5965f77fe 100644
--- a/js/apps/account-ui/package.json
+++ b/js/apps/account-ui/package.json
@@ -10,21 +10,21 @@
     "@patternfly/patternfly": "^4.224.2",
     "@patternfly/react-core": "^4.276.8",
     "@patternfly/react-icons": "^4.93.6",
-    "i18next": "^22.4.13",
+    "i18next": "^22.4.14",
     "i18next-http-backend": "^2.2.0",
     "keycloak-js": "999.0.0-SNAPSHOT",
     "keycloak-masthead": "999.0.0-SNAPSHOT",
     "lodash-es": "^4.17.21",
     "react": "^18.2.0",
     "react-dom": "^18.2.0",
-    "react-hook-form": "^7.43.8",
+    "react-hook-form": "^7.43.9",
     "react-i18next": "^12.2.0",
-    "react-router-dom": "6.9.0",
+    "react-router-dom": "6.10.0",
     "ui-shared": "999.0.0-SNAPSHOT"
   },
   "devDependencies": {
     "@types/lodash-es": "^4.17.7",
-    "@types/react": "^18.0.31",
+    "@types/react": "^18.0.32",
     "@types/react-dom": "^18.0.11",
     "@vitejs/plugin-react-swc": "^3.2.0",
     "vite": "^4.2.1",
diff --git a/js/apps/admin-ui/cypress/e2e/clients_test.spec.ts b/js/apps/admin-ui/cypress/e2e/clients_test.spec.ts
index 1321c48d7281..f77210b2d722 100644
--- a/js/apps/admin-ui/cypress/e2e/clients_test.spec.ts
+++ b/js/apps/admin-ui/cypress/e2e/clients_test.spec.ts
@@ -542,7 +542,8 @@ describe("Clients test", () => {
     it("Should delete attribute from client role", () => {
       commonPage.tableUtils().clickRowItemLink(itemId);
       rolesTab.goToAttributesTab();
-      attributesTab.deleteAttribute(1);
+      attributesTab.deleteAttribute(0);
+      attributesTab.assertEmpty();
       commonPage
         .masthead()
         .checkNotificationMessage("The role has been saved", true);
diff --git a/js/apps/admin-ui/cypress/e2e/group_test.spec.ts b/js/apps/admin-ui/cypress/e2e/group_test.spec.ts
index 6b31a49fa44e..2fcc5014b100 100644
--- a/js/apps/admin-ui/cypress/e2e/group_test.spec.ts
+++ b/js/apps/admin-ui/cypress/e2e/group_test.spec.ts
@@ -420,7 +420,8 @@ describe("Group test", () => {
     });
 
     it("Remove attribute", () => {
-      attributesTab.deleteAttribute(1).assertRowItemsEqualTo(1);
+      attributesTab.deleteAttribute(0);
+      attributesTab.assertEmpty();
       groupPage.assertNotificationGroupUpdated();
     });
 
@@ -429,7 +430,7 @@ describe("Group test", () => {
         .addAttribute("key", "value")
         .addAnAttributeButton()
         .revert()
-        .assertRowItemsEqualTo(1);
+        .assertEmpty();
     });
   });
 
diff --git a/js/apps/admin-ui/cypress/e2e/realm_roles_test.spec.ts b/js/apps/admin-ui/cypress/e2e/realm_roles_test.spec.ts
index 1669f0fdcea9..f65b30c3a171 100644
--- a/js/apps/admin-ui/cypress/e2e/realm_roles_test.spec.ts
+++ b/js/apps/admin-ui/cypress/e2e/realm_roles_test.spec.ts
@@ -248,10 +248,10 @@ describe("Realm roles test", () => {
       listingPage.itemExist(editRoleName).goToItemDetails(editRoleName);
 
       createRealmRolePage.goToAttributesTab();
-      keyValue.fillKeyValue({ key: "one", value: "1" }).validateRows(2);
+      keyValue.fillKeyValue({ key: "one", value: "1" }).validateRows(1);
       keyValue.save();
       masthead.checkNotificationMessage("The role has been saved", true);
-      keyValue.validateRows(2);
+      keyValue.validateRows(1);
     });
 
     it("should add attribute multiple", () => {
@@ -259,17 +259,17 @@ describe("Realm roles test", () => {
 
       createRealmRolePage.goToAttributesTab();
       keyValue
-        .fillKeyValue({ key: "two", value: "2" }, 1)
-        .fillKeyValue({ key: "three", value: "3" }, 2)
+        .fillKeyValue({ key: "two", value: "2" })
+        .fillKeyValue({ key: "three", value: "3" })
         .save()
-        .validateRows(4);
+        .validateRows(3);
     });
 
     it("should delete attribute", () => {
       listingPage.itemExist(editRoleName).goToItemDetails(editRoleName);
       createRealmRolePage.goToAttributesTab();
 
-      keyValue.deleteRow(1).save().validateRows(3);
+      keyValue.deleteRow(1).save().validateRows(2);
     });
   });
 
diff --git a/js/apps/admin-ui/cypress/support/pages/admin-ui/manage/AttributesTab.ts b/js/apps/admin-ui/cypress/support/pages/admin-ui/manage/AttributesTab.ts
index cdffecdbca37..97fd9b7aeb81 100644
--- a/js/apps/admin-ui/cypress/support/pages/admin-ui/manage/AttributesTab.ts
+++ b/js/apps/admin-ui/cypress/support/pages/admin-ui/manage/AttributesTab.ts
@@ -2,9 +2,10 @@ export default class AttributesTab {
   private saveAttributeBtn = "save-attributes";
   private addAttributeBtn = "attributes-add-row";
   private attributesTab = "attributes";
-  private attributeRow = "row";
-  private keyInput = (index: number) => `attributes[${index}].key`;
-  private valueInput = (index: number) => `attributes[${index}].value`;
+  private keyInput = "attributes-key";
+  private valueInput = "attributes-value";
+  private removeBtn = "attributes-remove";
+  private emptyState = "attributes-empty-state";
 
   public goToAttributesTab() {
     cy.findByTestId(this.attributesTab).click();
@@ -13,14 +14,15 @@ export default class AttributesTab {
   }
 
   public addAttribute(key: string, value: string) {
-    cy.findAllByTestId(this.attributeRow)
+    cy.findByTestId(this.addAttributeBtn).click();
+
+    cy.findAllByTestId(this.keyInput)
       .its("length")
-      .then((index) => {
-        cy.findByTestId(this.keyInput(index - 1)).type(key, { force: true });
-        cy.findByTestId(this.valueInput(index - 1)).type(value, {
-          force: true,
-        });
+      .then((length) => {
+        this.keyInputAt(length - 1).type(key, { force: true });
+        this.valueInputAt(length - 1).type(value, { force: true });
       });
+
     return this;
   }
 
@@ -35,7 +37,7 @@ export default class AttributesTab {
   }
 
   public deleteAttributeButton(row: number) {
-    cy.findByTestId(`attributes[${row - 1}].remove`).click({ force: true });
+    this.removeButtonAt(row).click({ force: true });
     return this;
   }
 
@@ -48,19 +50,27 @@ export default class AttributesTab {
     this.deleteAttributeButton(rowIndex);
     this.save();
 
-    cy.findAllByTestId(`attributes[${rowIndex - 1}].key`).should(
-      "have.value",
-      ""
-    );
-    cy.findAllByTestId(`attributes[${rowIndex - 1}].value`).should(
-      "have.value",
-      ""
-    );
     return this;
   }
 
+  public assertEmpty() {
+    cy.findByTestId(this.emptyState).should("exist");
+  }
+
   public assertRowItemsEqualTo(amount: number) {
-    cy.findAllByTestId("row").its("length").should("be.eq", amount);
+    cy.findAllByTestId(this.keyInput).its("length").should("be.eq", amount);
     return this;
   }
+
+  private keyInputAt(index: number) {
+    return cy.findAllByTestId(this.keyInput).eq(index);
+  }
+
+  private valueInputAt(index: number) {
+    return cy.findAllByTestId(this.valueInput).eq(index);
+  }
+
+  private removeButtonAt(index: number) {
+    return cy.findAllByTestId(this.removeBtn).eq(index);
+  }
 }
diff --git a/js/apps/admin-ui/cypress/support/pages/admin-ui/manage/KeyValueInput.ts b/js/apps/admin-ui/cypress/support/pages/admin-ui/manage/KeyValueInput.ts
index ceb14d75890a..5510e129d49c 100644
--- a/js/apps/admin-ui/cypress/support/pages/admin-ui/manage/KeyValueInput.ts
+++ b/js/apps/admin-ui/cypress/support/pages/admin-ui/manage/KeyValueInput.ts
@@ -7,22 +7,26 @@ export default class KeyValueInput {
     this.name = name;
   }
 
-  fillKeyValue({ key, value }: KeyValueType, index = 0) {
-    cy.findByTestId(`${this.name}[${index}].key`).clear();
-    cy.findByTestId(`${this.name}[${index}].key`).type(key);
-    cy.findByTestId(`${this.name}[${index}].value`).clear();
-    cy.findByTestId(`${this.name}[${index}].value`).type(value);
+  fillKeyValue({ key, value }: KeyValueType) {
     cy.findByTestId(`${this.name}-add-row`).click();
+
+    cy.findAllByTestId(`${this.name}-key`)
+      .its("length")
+      .then((length) => {
+        this.keyInputAt(length - 1).type(key);
+        this.valueInputAt(length - 1).type(value);
+      });
+
     return this;
   }
 
   deleteRow(index: number) {
-    cy.findByTestId(`${this.name}[${index}].remove`).click();
+    cy.findAllByTestId(`${this.name}-remove`).eq(index).click();
     return this;
   }
 
   validateRows(numberOfRows: number) {
-    cy.findAllByTestId("row").should("have.length", numberOfRows);
+    cy.findAllByTestId(`${this.name}-key`).should("have.length", numberOfRows);
     return this;
   }
 
@@ -30,4 +34,12 @@ export default class KeyValueInput {
     cy.findByTestId("save-attributes").click();
     return this;
   }
+
+  keyInputAt(index: number) {
+    return cy.findAllByTestId(`${this.name}-key`).eq(index);
+  }
+
+  valueInputAt(index: number) {
+    return cy.findAllByTestId(`${this.name}-value`).eq(index);
+  }
 }
diff --git a/js/apps/admin-ui/cypress/support/pages/admin-ui/manage/LegacyKeyValueInput.ts b/js/apps/admin-ui/cypress/support/pages/admin-ui/manage/LegacyKeyValueInput.ts
new file mode 100644
index 000000000000..cf3be1246292
--- /dev/null
+++ b/js/apps/admin-ui/cypress/support/pages/admin-ui/manage/LegacyKeyValueInput.ts
@@ -0,0 +1,33 @@
+import type { KeyValueType } from "../../../../../src/components/key-value-form/key-value-convert";
+
+export default class LegacyKeyValueInput {
+  private name: string;
+
+  constructor(name: string) {
+    this.name = name;
+  }
+
+  fillKeyValue({ key, value }: KeyValueType, index = 0) {
+    cy.findByTestId(`${this.name}.${index}.key`).clear();
+    cy.findByTestId(`${this.name}.${index}.key`).type(key);
+    cy.findByTestId(`${this.name}.${index}.value`).clear();
+    cy.findByTestId(`${this.name}.${index}.value`).type(value);
+    cy.findByTestId(`${this.name}-add-row`).click();
+    return this;
+  }
+
+  deleteRow(index: number) {
+    cy.findByTestId(`${this.name}.${index}.remove`).click();
+    return this;
+  }
+
+  validateRows(numberOfRows: number) {
+    cy.findAllByTestId("row").should("have.length", numberOfRows);
+    return this;
+  }
+
+  save() {
+    cy.findByTestId("save-attributes").click();
+    return this;
+  }
+}
diff --git a/js/apps/admin-ui/cypress/support/pages/admin-ui/manage/identity_providers/AddMapperPage.ts b/js/apps/admin-ui/cypress/support/pages/admin-ui/manage/identity_providers/AddMapperPage.ts
index 0708e7780d03..46b8c67b17a8 100644
--- a/js/apps/admin-ui/cypress/support/pages/admin-ui/manage/identity_providers/AddMapperPage.ts
+++ b/js/apps/admin-ui/cypress/support/pages/admin-ui/manage/identity_providers/AddMapperPage.ts
@@ -1,4 +1,4 @@
-import KeyValueInput from "../KeyValueInput";
+import LegacyKeyValueInput from "../LegacyKeyValueInput";
 
 export default class AddMapperPage {
   private mappersTab = "mappers-tab";
@@ -23,8 +23,8 @@ export default class AddMapperPage {
   private mappersUrl = "/oidc/mappers";
   private regexAttributeValuesSwitch = "are.attribute.values.regex";
   private syncmodeSelectToggle = "#syncMode";
-  private attributesKeyInput = '[data-testid="config.attributes[0].key"]';
-  private attributesValueInput = '[data-testid="config.attributes[0].value"]';
+  private attributesKeyInput = '[data-testid="config.attributes.0.key"]';
+  private attributesValueInput = '[data-testid="config.attributes.0.value"]';
   private template = "template";
   private target = "#target";
 
@@ -392,7 +392,7 @@ export default class AddMapperPage {
 
     cy.findByTestId(this.idpMapperSelect).contains("Claim to Role").click();
 
-    const keyValue = new KeyValueInput("config.claims");
+    const keyValue = new LegacyKeyValueInput("config.claims");
 
     keyValue.fillKeyValue({ key: "key", value: "value" });
 
diff --git a/js/apps/admin-ui/cypress/support/pages/admin-ui/manage/realm_settings/UserProfile.ts b/js/apps/admin-ui/cypress/support/pages/admin-ui/manage/realm_settings/UserProfile.ts
index 682bb495b660..c3cd37d2b669 100644
--- a/js/apps/admin-ui/cypress/support/pages/admin-ui/manage/realm_settings/UserProfile.ts
+++ b/js/apps/admin-ui/cypress/support/pages/admin-ui/manage/realm_settings/UserProfile.ts
@@ -17,8 +17,9 @@ export default class UserProfile {
   private newAttributeRequiredFor = 'input[name="roles"]';
   private newAttributeRequiredWhen = 'input[name="requiredWhen"]';
   private newAttributeEmptyValidators = ".kc-emptyValidators";
-  private newAttributeAnnotationKey = 'input[name="annotations[0].key"]';
-  private newAttributeAnnotationValue = 'input[name="annotations[0].value"]';
+  private newAttributeAnnotationBtn = "annotations-add-row";
+  private newAttributeAnnotationKey = "annotations-key";
+  private newAttributeAnnotationValue = "annotations-value";
   private validatorRolesList = "#validator";
   private validatorsList = 'tbody [data-label="name"]';
   private saveNewAttributeBtn = "attribute-create";
@@ -106,8 +107,9 @@ export default class UserProfile {
     cy.get(this.newAttributeRequiredFor).first().check({ force: true });
     cy.get(this.newAttributeRequiredWhen).first().check();
     cy.get(this.newAttributeEmptyValidators).contains("No validators.");
-    cy.get(this.newAttributeAnnotationKey).type("test");
-    cy.get(this.newAttributeAnnotationValue).type("123");
+    cy.findByTestId(this.newAttributeAnnotationBtn).click();
+    cy.findByTestId(this.newAttributeAnnotationKey).type("test");
+    cy.findByTestId(this.newAttributeAnnotationValue).type("123");
     return this;
   }
 
diff --git a/js/apps/admin-ui/package.json b/js/apps/admin-ui/package.json
index 58eb27380036..ca1306915685 100644
--- a/js/apps/admin-ui/package.json
+++ b/js/apps/admin-ui/package.json
@@ -70,7 +70,7 @@
     "dagre": "^0.8.5",
     "file-saver": "^2.0.5",
     "flat": "^5.0.2",
-    "i18next": "^22.4.13",
+    "i18next": "^22.4.14",
     "i18next-http-backend": "^2.2.0",
     "keycloak-js": "999.0.0-SNAPSHOT",
     "lodash-es": "^4.17.21",
@@ -78,9 +78,9 @@
     "react-dom": "^18.2.0",
     "react-dropzone": "^14.2.3",
     "react-error-boundary": "^3.1.4",
-    "react-hook-form": "^7.43.8",
+    "react-hook-form": "^7.43.9",
     "react-i18next": "^12.2.0",
-    "react-router-dom": "6.9.0",
+    "react-router-dom": "6.10.0",
     "reactflow": "^11.7.0",
     "ui-shared": "999.0.0-SNAPSHOT",
     "use-react-router-breadcrumbs": "^4.0.1"
@@ -94,7 +94,7 @@
     "@types/file-saver": "^2.0.5",
     "@types/flat": "^5.0.2",
     "@types/lodash-es": "^4.17.7",
-    "@types/react": "^18.0.31",
+    "@types/react": "^18.0.32",
     "@types/react-dom": "^18.0.11",
     "@vitejs/plugin-react-swc": "^3.2.0",
     "cypress": "^12.9.0",
diff --git a/js/apps/admin-ui/public/resources/en/common.json b/js/apps/admin-ui/public/resources/en/common.json
index a40bf43da451..bede38e4a205 100644
--- a/js/apps/admin-ui/public/resources/en/common.json
+++ b/js/apps/admin-ui/public/resources/en/common.json
@@ -172,10 +172,13 @@
     "years": "Years"
   },
   "attributes": "Attributes",
+  "missingAttributes": "No attributes have been defined yet. Click the below button to add attributes, key and value are required for a key pair.",
   "addAttribute": "Add an attribute",
   "removeAttribute": "Remove attribute",
   "keyPlaceholder": "Type a key",
   "valuePlaceholder": "Type a value",
+  "keyError": "A key must be provided.",
+  "valueError": "A value must be provided.",
   "credentials": "Credentials",
   "clientId": "Client ID",
   "clientName": "Name",
diff --git a/js/apps/admin-ui/src/clients/authorization/KeyBasedAttributeInput.tsx b/js/apps/admin-ui/src/clients/authorization/KeyBasedAttributeInput.tsx
index 6271edac8b19..48e881ffb4d4 100644
--- a/js/apps/admin-ui/src/clients/authorization/KeyBasedAttributeInput.tsx
+++ b/js/apps/admin-ui/src/clients/authorization/KeyBasedAttributeInput.tsx
@@ -103,14 +103,14 @@ const ValueInput = ({
     <Td>
       {resources || attributeValues?.length ? (
         <Controller
-          name={`${name}[${rowIndex}].value`}
+          name={`${name}.${rowIndex}.value`}
           defaultValue={[]}
           control={control}
           render={({ field }) => (
             <Select
               toggleId={`${attribute.id}-value`}
               className="kc-attribute-value-selectable"
-              name={`${name}[${rowIndex}].value`}
+              name={`${name}.${rowIndex}.value`}
               chipGroupProps={{
                 numChips: 1,
                 expandedText: t("common:hide"),
@@ -171,7 +171,7 @@ export const KeyBasedAttributeInput = ({
     }
   }, [fields]);
 
-  const watchLastValue = watch(`${name}[${fields.length - 1}].value`, "");
+  const watchLastValue = watch(`${name}.${fields.length - 1}.value`, "");
 
   return (
     <TableComposable
@@ -190,14 +190,14 @@ export const KeyBasedAttributeInput = ({
           <Tr key={attribute.id} data-testid="attribute-row">
             <Td>
               <Controller
-                name={`${name}[${rowIndex}].key`}
+                name={`${name}.${rowIndex}.key`}
                 defaultValue=""
                 control={control}
                 render={({ field }) => (
                   <Select
-                    toggleId={`${name}[${rowIndex}].key`}
+                    toggleId={`${name}.${rowIndex}.key`}
                     className="kc-attribute-key-selectable"
-                    name={`${name}[${rowIndex}].key`}
+                    name={`${name}.${rowIndex}.key`}
                     onToggle={(open) => toggleKeySelect(rowIndex, open)}
                     isOpen={isKeyOpenArray[rowIndex]}
                     variant={SelectVariant.typeahead}
diff --git a/js/apps/admin-ui/src/components/dynamic/MapComponent.tsx b/js/apps/admin-ui/src/components/dynamic/MapComponent.tsx
index eee343e262ff..8fa486808ba2 100644
--- a/js/apps/admin-ui/src/components/dynamic/MapComponent.tsx
+++ b/js/apps/admin-ui/src/components/dynamic/MapComponent.tsx
@@ -84,11 +84,11 @@ export const MapComponent = ({ name, label, helpText }: ComponentProps) => {
           <Flex key={attribute.id} data-testid="row">
             <FlexItem grow={{ default: "grow" }}>
               <TextInput
-                name={`${fieldName}[${index}].key`}
+                name={`${fieldName}.${index}.key`}
                 placeholder={t("common:keyPlaceholder")}
                 aria-label={t("key")}
                 defaultValue={attribute.key}
-                data-testid={`${fieldName}[${index}].key`}
+                data-testid={`${fieldName}.${index}.key`}
                 onChange={(value) => updateKey(index, value)}
                 onBlur={() => update()}
               />
@@ -98,11 +98,11 @@ export const MapComponent = ({ name, label, helpText }: ComponentProps) => {
               spacer={{ default: "spacerNone" }}
             >
               <TextInput
-                name={`${fieldName}[${index}].value`}
+                name={`${fieldName}.${index}.value`}
                 placeholder={t("common:valuePlaceholder")}
                 aria-label={t("common:value")}
                 defaultValue={attribute.value}
-                data-testid={`${fieldName}[${index}].value`}
+                data-testid={`${fieldName}.${index}.value`}
                 onChange={(value) => updateValue(index, value)}
                 onBlur={() => update()}
               />
@@ -113,7 +113,7 @@ export const MapComponent = ({ name, label, helpText }: ComponentProps) => {
                 title={t("common:removeAttribute")}
                 isDisabled={map.length === 1}
                 onClick={() => remove(index)}
-                data-testid={`${fieldName}[${index}].remove`}
+                data-testid={`${fieldName}.${index}.remove`}
               >
                 <MinusCircleIcon />
               </Button>
diff --git a/js/apps/admin-ui/src/components/key-value-form/KeyValueInput.tsx b/js/apps/admin-ui/src/components/key-value-form/KeyValueInput.tsx
index 1e45a334523c..2e9e2bd88c03 100644
--- a/js/apps/admin-ui/src/components/key-value-form/KeyValueInput.tsx
+++ b/js/apps/admin-ui/src/components/key-value-form/KeyValueInput.tsx
@@ -2,16 +2,20 @@ import {
   ActionList,
   ActionListItem,
   Button,
-  Flex,
-  FlexItem,
+  EmptyState,
+  EmptyStateBody,
+  Grid,
+  GridItem,
+  HelperText,
+  HelperTextItem,
+  InputGroup,
 } from "@patternfly/react-core";
 import { MinusCircleIcon, PlusCircleIcon } from "@patternfly/react-icons";
-import { useEffect } from "react";
-import { useFieldArray, useFormContext, useWatch } from "react-hook-form";
+import { Fragment } from "react";
+import { useFieldArray, useFormContext } from "react-hook-form";
 import { useTranslation } from "react-i18next";
 
 import { KeycloakTextInput } from "../keycloak-text-input/KeycloakTextInput";
-import { KeyValueType } from "./key-value-convert";
 
 type KeyValueInputProps = {
   name: string;
@@ -19,83 +23,82 @@ type KeyValueInputProps = {
 
 export const KeyValueInput = ({ name }: KeyValueInputProps) => {
   const { t } = useTranslation("common");
-  const { control, register } = useFormContext();
-  const { fields, append, remove } = useFieldArray({
+  const {
     control,
-    name,
-  });
+    register,
+    formState: { errors },
+  } = useFormContext();
 
-  const watchFields = useWatch({
+  const { fields, append, remove } = useFieldArray({
     control,
     name,
-    defaultValue: [{ key: "", value: "" }],
   });
 
-  const isValid =
-    Array.isArray(watchFields) &&
-    watchFields.every(
-      ({ key, value }: KeyValueType) =>
-        // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
-        key?.trim().length !== 0 && value?.trim().length !== 0
-    );
+  const appendNew = () => append({ key: "", value: "" });
 
-  useEffect(() => {
-    if (!fields.length) {
-      append({ key: "", value: "" }, { shouldFocus: false });
-    }
-  }, [fields]);
-
-  return (
+  return fields.length > 0 ? (
     <>
-      <Flex direction={{ default: "column" }}>
-        <Flex>
-          <FlexItem
-            grow={{ default: "grow" }}
-            spacer={{ default: "spacerNone" }}
-          >
-            <strong>{t("key")}</strong>
-          </FlexItem>
-          <FlexItem grow={{ default: "grow" }}>
-            <strong>{t("value")}</strong>
-          </FlexItem>
-        </Flex>
-        {fields.map((attribute, index) => (
-          <Flex key={attribute.id} data-testid="row">
-            <FlexItem grow={{ default: "grow" }}>
-              <KeycloakTextInput
-                placeholder={t("keyPlaceholder")}
-                aria-label={t("key")}
-                defaultValue=""
-                data-testid={`${name}[${index}].key`}
-                {...register(`${name}[${index}].key`)}
-              />
-            </FlexItem>
-            <FlexItem
-              grow={{ default: "grow" }}
-              spacer={{ default: "spacerNone" }}
-            >
-              <KeycloakTextInput
-                placeholder={t("valuePlaceholder")}
-                aria-label={t("value")}
-                defaultValue=""
-                data-testid={`${name}[${index}].value`}
-                {...register(`${name}[${index}].value`)}
-              />
-            </FlexItem>
-            <FlexItem>
-              <Button
-                variant="link"
-                title={t("removeAttribute")}
-                isDisabled={watchFields.length === 1}
-                onClick={() => remove(index)}
-                data-testid={`${name}[${index}].remove`}
-              >
-                <MinusCircleIcon />
-              </Button>
-            </FlexItem>
-          </Flex>
-        ))}
-      </Flex>
+      <Grid hasGutter>
+        <GridItem className="pf-c-form__label" span={6}>
+          <span className="pf-c-form__label-text">{t("key")}</span>
+        </GridItem>
+        <GridItem className="pf-c-form__label" span={6}>
+          <span className="pf-c-form__label-text">{t("value")}</span>
+        </GridItem>
+        {fields.map((attribute, index) => {
+          const keyError = !!(errors as any)[name]?.[index]?.key;
+          const valueError = !!(errors as any)[name]?.[index]?.value;
+
+          return (
+            <Fragment key={attribute.id}>
+              <GridItem span={6}>
+                <KeycloakTextInput
+                  placeholder={t("keyPlaceholder")}
+                  aria-label={t("key")}
+                  data-testid={`${name}-key`}
+                  {...register(`${name}.${index}.key`, { required: true })}
+                  validated={keyError ? "error" : "default"}
+                  isRequired
+                />
+                {keyError && (
+                  <HelperText>
+                    <HelperTextItem variant="error">
+                      {t("keyError")}
+                    </HelperTextItem>
+                  </HelperText>
+                )}
+              </GridItem>
+              <GridItem span={6}>
+                <InputGroup>
+                  <KeycloakTextInput
+                    placeholder={t("valuePlaceholder")}
+                    aria-label={t("value")}
+                    data-testid={`${name}-value`}
+                    {...register(`${name}.${index}.value`, { required: true })}
+                    validated={valueError ? "error" : "default"}
+                    isRequired
+                  />
+                  <Button
+                    variant="link"
+                    title={t("removeAttribute")}
+                    onClick={() => remove(index)}
+                    data-testid={`${name}-remove`}
+                  >
+                    <MinusCircleIcon />
+                  </Button>
+                </InputGroup>
+                {valueError && (
+                  <HelperText>
+                    <HelperTextItem variant="error">
+                      {t("valueError")}
+                    </HelperTextItem>
+                  </HelperText>
+                )}
+              </GridItem>
+            </Fragment>
+          );
+        })}
+      </Grid>
       <ActionList>
         <ActionListItem>
           <Button
@@ -103,13 +106,29 @@ export const KeyValueInput = ({ name }: KeyValueInputProps) => {
             className="pf-u-px-0 pf-u-mt-sm"
             variant="link"
             icon={<PlusCircleIcon />}
-            isDisabled={!isValid}
-            onClick={() => append({ key: "", value: "" })}
+            onClick={appendNew}
           >
             {t("addAttribute")}
           </Button>
         </ActionListItem>
       </ActionList>
     </>
+  ) : (
+    <EmptyState
+      data-testid={`${name}-empty-state`}
+      className="pf-u-p-0"
+      variant="xs"
+    >
+      <EmptyStateBody>{t("missingAttributes")}</EmptyStateBody>
+      <Button
+        data-testid={`${name}-add-row`}
+        variant="link"
+        icon={<PlusCircleIcon />}
+        isSmall
+        onClick={appendNew}
+      >
+        {t("addAttribute")}
+      </Button>
+    </EmptyState>
   );
 };
diff --git a/js/apps/admin-ui/src/components/key-value-form/key-value-convert.test.ts b/js/apps/admin-ui/src/components/key-value-form/key-value-convert.test.ts
index 36edb14101da..ea2d68c4e635 100644
--- a/js/apps/admin-ui/src/components/key-value-form/key-value-convert.test.ts
+++ b/js/apps/admin-ui/src/components/key-value-form/key-value-convert.test.ts
@@ -41,18 +41,6 @@ describe("Tests the convert functions for attribute input", () => {
     expect(result).toEqual({ theKey: ["theValue"] });
   });
 
-  it("convert empty object to attributes", () => {
-    const given: {
-      [key: string]: string[];
-    } = {};
-
-    //when
-    const result = arrayToKeyValue(given);
-
-    //then
-    expect(result).toEqual([{ key: "", value: "" }]);
-  });
-
   it("convert object to attributes", () => {
     const given = { one: ["1"], two: ["2"] };
 
@@ -63,7 +51,6 @@ describe("Tests the convert functions for attribute input", () => {
     expect(result).toEqual([
       { key: "one", value: "1" },
       { key: "two", value: "2" },
-      { key: "", value: "" },
     ]);
   });
 
diff --git a/js/apps/admin-ui/src/components/key-value-form/key-value-convert.ts b/js/apps/admin-ui/src/components/key-value-form/key-value-convert.ts
index a8045d75cda9..ef0181fd7956 100644
--- a/js/apps/admin-ui/src/components/key-value-form/key-value-convert.ts
+++ b/js/apps/admin-ui/src/components/key-value-form/key-value-convert.ts
@@ -22,5 +22,5 @@ export function arrayToKeyValue<T>(attributes: Record<string, string[]> = {}) {
     value.map<KeyValueType>((value) => ({ key, value }))
   );
 
-  return result.concat({ key: "", value: "" }) as PathValue<T, Path<T>>;
+  return result as PathValue<T, Path<T>>;
 }
diff --git a/js/apps/admin-ui/src/components/multi-line-input/MultiLineInput.tsx b/js/apps/admin-ui/src/components/multi-line-input/MultiLineInput.tsx
index 7cb6b85b3faa..28910f9313d3 100644
--- a/js/apps/admin-ui/src/components/multi-line-input/MultiLineInput.tsx
+++ b/js/apps/admin-ui/src/components/multi-line-input/MultiLineInput.tsx
@@ -85,7 +85,7 @@ export const MultiLineInput = ({
             <TextInput
               data-testid={name + index}
               onChange={(value) => updateValue(index, value)}
-              name={`${name}[${index}].value`}
+              name={`${name}.${index}.value`}
               value={value}
               isDisabled={isDisabled}
               {...rest}
diff --git a/js/apps/admin-ui/src/util.test.ts b/js/apps/admin-ui/src/util.test.ts
index 32dee9725a1a..2119b75244f9 100644
--- a/js/apps/admin-ui/src/util.test.ts
+++ b/js/apps/admin-ui/src/util.test.ts
@@ -26,10 +26,7 @@ describe("Tests the form convert util functions", () => {
     expect(values).toEqual({
       name: "client",
       other: { one: "1", two: "2" },
-      attributes: [
-        { key: "one", value: "1" },
-        { key: "", value: "" },
-      ],
+      attributes: [{ key: "one", value: "1" }],
     });
   });
 
diff --git a/js/libs/keycloak-masthead/package.json b/js/libs/keycloak-masthead/package.json
index 669b179abebf..8449a7d63acf 100644
--- a/js/libs/keycloak-masthead/package.json
+++ b/js/libs/keycloak-masthead/package.json
@@ -47,12 +47,12 @@
     "react-dom": "^18.2.0"
   },
   "devDependencies": {
-    "@types/react": "^18.0.31",
+    "@types/react": "^18.0.32",
     "@types/react-dom": "^18.0.11",
     "@vitejs/plugin-react-swc": "^3.2.0",
     "rollup-plugin-peer-deps-external": "^2.2.4",
     "vite": "^4.2.1",
     "vite-plugin-checker": "^0.5.6",
-    "vite-plugin-dts": "^2.1.0"
+    "vite-plugin-dts": "^2.2.0"
   }
 }
diff --git a/js/libs/ui-shared/package.json b/js/libs/ui-shared/package.json
index 2646c50677a6..91c56b294483 100644
--- a/js/libs/ui-shared/package.json
+++ b/js/libs/ui-shared/package.json
@@ -36,17 +36,17 @@
   },
   "dependencies": {
     "@patternfly/react-core": "^4.276.8",
-    "react-hook-form": "7.43.8",
+    "react-hook-form": "7.43.9",
     "react": "^18.2.0",
     "react-dom": "^18.2.0"
   },
   "devDependencies": {
-    "@types/react": "^18.0.31",
+    "@types/react": "^18.0.32",
     "@types/react-dom": "^18.0.11",
     "@vitejs/plugin-react-swc": "^3.2.0",
     "rollup-plugin-peer-deps-external": "^2.2.4",
     "vite": "^4.2.1",
     "vite-plugin-checker": "^0.5.6",
-    "vite-plugin-dts": "^2.1.0"
+    "vite-plugin-dts": "^2.2.0"
   }
 }
diff --git a/js/package-lock.json b/js/package-lock.json
index 371e807cc132..d71b6e070c39 100644
--- a/js/package-lock.json
+++ b/js/package-lock.json
@@ -20,8 +20,8 @@
         "@typescript-eslint/parser": "^5.57.0",
         "eslint": "^8.37.0",
         "eslint-config-prettier": "^8.8.0",
-        "eslint-import-resolver-typescript": "^3.5.3",
-        "eslint-plugin-cypress": "^2.13.1",
+        "eslint-import-resolver-typescript": "^3.5.4",
+        "eslint-plugin-cypress": "^2.13.2",
         "eslint-plugin-import": "^2.27.5",
         "eslint-plugin-lodash": "^7.4.0",
         "eslint-plugin-mocha": "^10.1.0",
@@ -30,7 +30,7 @@
         "husky": "^8.0.3",
         "lint-staged": "^13.2.0",
         "prettier": "^2.8.7",
-        "typescript": "^5.0.2",
+        "typescript": "^5.0.3",
         "wireit": "^0.9.5"
       }
     },
@@ -39,21 +39,21 @@
         "@patternfly/patternfly": "^4.224.2",
         "@patternfly/react-core": "^4.276.8",
         "@patternfly/react-icons": "^4.93.6",
-        "i18next": "^22.4.13",
+        "i18next": "^22.4.14",
         "i18next-http-backend": "^2.2.0",
         "keycloak-js": "999.0.0-SNAPSHOT",
         "keycloak-masthead": "999.0.0-SNAPSHOT",
         "lodash-es": "^4.17.21",
         "react": "^18.2.0",
         "react-dom": "^18.2.0",
-        "react-hook-form": "^7.43.8",
+        "react-hook-form": "^7.43.9",
         "react-i18next": "^12.2.0",
-        "react-router-dom": "6.9.0",
+        "react-router-dom": "6.10.0",
         "ui-shared": "999.0.0-SNAPSHOT"
       },
       "devDependencies": {
         "@types/lodash-es": "^4.17.7",
-        "@types/react": "^18.0.31",
+        "@types/react": "^18.0.32",
         "@types/react-dom": "^18.0.11",
         "@vitejs/plugin-react-swc": "^3.2.0",
         "vite": "^4.2.1",
@@ -72,7 +72,7 @@
         "dagre": "^0.8.5",
         "file-saver": "^2.0.5",
         "flat": "^5.0.2",
-        "i18next": "^22.4.13",
+        "i18next": "^22.4.14",
         "i18next-http-backend": "^2.2.0",
         "keycloak-js": "999.0.0-SNAPSHOT",
         "lodash-es": "^4.17.21",
@@ -80,9 +80,9 @@
         "react-dom": "^18.2.0",
         "react-dropzone": "^14.2.3",
         "react-error-boundary": "^3.1.4",
-        "react-hook-form": "^7.43.8",
+        "react-hook-form": "^7.43.9",
         "react-i18next": "^12.2.0",
-        "react-router-dom": "6.9.0",
+        "react-router-dom": "6.10.0",
         "reactflow": "^11.7.0",
         "ui-shared": "999.0.0-SNAPSHOT",
         "use-react-router-breadcrumbs": "^4.0.1"
@@ -96,7 +96,7 @@
         "@types/file-saver": "^2.0.5",
         "@types/flat": "^5.0.2",
         "@types/lodash-es": "^4.17.7",
-        "@types/react": "^18.0.31",
+        "@types/react": "^18.0.32",
         "@types/react-dom": "^18.0.11",
         "@vitejs/plugin-react-swc": "^3.2.0",
         "cypress": "^12.9.0",
@@ -168,13 +168,13 @@
         "react-dom": "^18.2.0"
       },
       "devDependencies": {
-        "@types/react": "^18.0.31",
+        "@types/react": "^18.0.32",
         "@types/react-dom": "^18.0.11",
         "@vitejs/plugin-react-swc": "^3.2.0",
         "rollup-plugin-peer-deps-external": "^2.2.4",
         "vite": "^4.2.1",
         "vite-plugin-checker": "^0.5.6",
-        "vite-plugin-dts": "^2.1.0"
+        "vite-plugin-dts": "^2.2.0"
       }
     },
     "libs/ui-shared": {
@@ -183,16 +183,16 @@
         "@patternfly/react-core": "^4.276.8",
         "react": "^18.2.0",
         "react-dom": "^18.2.0",
-        "react-hook-form": "7.43.8"
+        "react-hook-form": "7.43.9"
       },
       "devDependencies": {
-        "@types/react": "^18.0.31",
+        "@types/react": "^18.0.32",
         "@types/react-dom": "^18.0.11",
         "@vitejs/plugin-react-swc": "^3.2.0",
         "rollup-plugin-peer-deps-external": "^2.2.4",
         "vite": "^4.2.1",
         "vite-plugin-checker": "^0.5.6",
-        "vite-plugin-dts": "^2.1.0"
+        "vite-plugin-dts": "^2.2.0"
       }
     },
     "node_modules/@4tw/cypress-drag-drop": {
@@ -1503,9 +1503,9 @@
       }
     },
     "node_modules/@remix-run/router": {
-      "version": "1.4.0",
-      "resolved": "https://registry.npmjs.org/@remix-run/router/-/router-1.4.0.tgz",
-      "integrity": "sha512-BJ9SxXux8zAg991UmT8slpwpsd31K1dHHbD3Ba4VzD+liLQ4WAMSxQp2d2ZPRPfN0jN2NPRowcSSoM7lCaF08Q==",
+      "version": "1.5.0",
+      "resolved": "https://registry.npmjs.org/@remix-run/router/-/router-1.5.0.tgz",
+      "integrity": "sha512-bkUDCp8o1MvFO+qxkODcbhSqRa6P2GXgrGZVpt0dCXNW2HCSCqYI0ZoAqEOSAjRWmmlKcYgFvN4B4S+zo/f8kg==",
       "engines": {
         "node": ">=14"
       }
@@ -2540,9 +2540,9 @@
       "integrity": "sha512-JCB8C6SnDoQf0cNycqd/35A7MjcnK+ZTqE7judS6o7utxUCg6imJg3QK2qzHKszlTjcj2cn+NwMB2i96ubpj7w=="
     },
     "node_modules/@types/react": {
-      "version": "18.0.31",
-      "resolved": "https://registry.npmjs.org/@types/react/-/react-18.0.31.tgz",
-      "integrity": "sha512-EEG67of7DsvRDU6BLLI0p+k1GojDLz9+lZsnCpCRTa/lOokvyPBvp8S5x+A24hME3yyQuIipcP70KJ6H7Qupww==",
+      "version": "18.0.32",
+      "resolved": "https://registry.npmjs.org/@types/react/-/react-18.0.32.tgz",
+      "integrity": "sha512-gYGXdtPQ9Cj0w2Fwqg5/ak6BcK3Z15YgjSqtyDizWUfx7mQ8drs0NBUzRRsAdoFVTO8kJ8L2TL8Skm7OFPnLUw==",
       "dependencies": {
         "@types/prop-types": "*",
         "@types/scheduler": "*",
@@ -5021,18 +5021,18 @@
       }
     },
     "node_modules/eslint-import-resolver-typescript": {
-      "version": "3.5.3",
-      "resolved": "https://registry.npmjs.org/eslint-import-resolver-typescript/-/eslint-import-resolver-typescript-3.5.3.tgz",
-      "integrity": "sha512-njRcKYBc3isE42LaTcJNVANR3R99H9bAxBDMNDr2W7yq5gYPxbU3MkdhsQukxZ/Xg9C2vcyLlDsbKfRDg0QvCQ==",
+      "version": "3.5.4",
+      "resolved": "https://registry.npmjs.org/eslint-import-resolver-typescript/-/eslint-import-resolver-typescript-3.5.4.tgz",
+      "integrity": "sha512-9xUpnedEmSfG57sN1UvWPiEhfJ8bPt0Wg2XysA7Mlc79iFGhmJtRUg9LxtkK81FhMUui0YuR2E8iUsVhePkh4A==",
       "dev": true,
       "dependencies": {
         "debug": "^4.3.4",
-        "enhanced-resolve": "^5.10.0",
-        "get-tsconfig": "^4.2.0",
-        "globby": "^13.1.2",
-        "is-core-module": "^2.10.0",
+        "enhanced-resolve": "^5.12.0",
+        "get-tsconfig": "^4.5.0",
+        "globby": "^13.1.3",
+        "is-core-module": "^2.11.0",
         "is-glob": "^4.0.3",
-        "synckit": "^0.8.4"
+        "synckit": "^0.8.5"
       },
       "engines": {
         "node": "^14.18.0 || >=16.0.0"
@@ -5103,9 +5103,9 @@
       }
     },
     "node_modules/eslint-plugin-cypress": {
-      "version": "2.13.1",
-      "resolved": "https://registry.npmjs.org/eslint-plugin-cypress/-/eslint-plugin-cypress-2.13.1.tgz",
-      "integrity": "sha512-THjc7IT3S9H4KwmRhzAhMGQaEqy78/7W75He/gBhJEH0vIuAY16vOI4YSliDo/ZY+Wm6DtvMHR+8uVvICcI3Lw==",
+      "version": "2.13.2",
+      "resolved": "https://registry.npmjs.org/eslint-plugin-cypress/-/eslint-plugin-cypress-2.13.2.tgz",
+      "integrity": "sha512-LlwjnBTzuKuC0A4H0RxVjs0YeAWK+CD1iM9Dp8un3lzT713ePQxfpPstCD+9HSAss8emuE3b2hCNUST+NrUwKw==",
       "dev": true,
       "dependencies": {
         "globals": "^11.12.0"
@@ -5921,9 +5921,9 @@
       }
     },
     "node_modules/get-tsconfig": {
-      "version": "4.4.0",
-      "resolved": "https://registry.npmjs.org/get-tsconfig/-/get-tsconfig-4.4.0.tgz",
-      "integrity": "sha512-0Gdjo/9+FzsYhXCEFueo2aY1z1tpXrxWZzP7k8ul9qt1U5o8rYJwTJYmaeHdrVosYIVYkOy2iwCJ9FdpocJhPQ==",
+      "version": "4.5.0",
+      "resolved": "https://registry.npmjs.org/get-tsconfig/-/get-tsconfig-4.5.0.tgz",
+      "integrity": "sha512-MjhiaIWCJ1sAU4pIQ5i5OfOuHHxVo1oYeNsWTON7jxYkod8pHocXeh+SSbmu5OZZZK73B6cbJ2XADzXehLyovQ==",
       "dev": true,
       "funding": {
         "url": "https://github.com/privatenumber/get-tsconfig?sponsor=1"
@@ -6280,9 +6280,9 @@
       }
     },
     "node_modules/i18next": {
-      "version": "22.4.13",
-      "resolved": "https://registry.npmjs.org/i18next/-/i18next-22.4.13.tgz",
-      "integrity": "sha512-GX7flMHRRqQA0I1yGLmaZ4Hwt1JfLqagk8QPDPZsqekbKtXsuIngSVWM/s3SLgNkrEXjA+0sMGNuOEkkmyqmWg==",
+      "version": "22.4.14",
+      "resolved": "https://registry.npmjs.org/i18next/-/i18next-22.4.14.tgz",
+      "integrity": "sha512-VtLPtbdwGn0+DAeE00YkiKKXadkwg+rBUV+0v8v0ikEjwdiJ0gmYChVE4GIa9HXymY6wKapkL93vGT7xpq6aTw==",
       "funding": [
         {
           "type": "individual",
@@ -8841,9 +8841,9 @@
       }
     },
     "node_modules/react-hook-form": {
-      "version": "7.43.8",
-      "resolved": "https://registry.npmjs.org/react-hook-form/-/react-hook-form-7.43.8.tgz",
-      "integrity": "sha512-BQm+Ge5KjTk1EchDBRhdP8Pkb7MArO2jFF+UWYr3rtvh6197khi22uloLqlWeuY02ItlCzPunPsFt1/q9wQKnw==",
+      "version": "7.43.9",
+      "resolved": "https://registry.npmjs.org/react-hook-form/-/react-hook-form-7.43.9.tgz",
+      "integrity": "sha512-AUDN3Pz2NSeoxQ7Hs6OhQhDr6gtF9YRuutGDwPQqhSUAHJSgGl2VeY3qN19MG0SucpjgDiuMJ4iC5T5uB+eaNQ==",
       "engines": {
         "node": ">=12.22.0"
       },
@@ -8896,11 +8896,11 @@
       }
     },
     "node_modules/react-router": {
-      "version": "6.9.0",
-      "resolved": "https://registry.npmjs.org/react-router/-/react-router-6.9.0.tgz",
-      "integrity": "sha512-51lKevGNUHrt6kLuX3e/ihrXoXCa9ixY/nVWRLlob4r/l0f45x3SzBvYJe3ctleLUQQ5fVa4RGgJOTH7D9Umhw==",
+      "version": "6.10.0",
+      "resolved": "https://registry.npmjs.org/react-router/-/react-router-6.10.0.tgz",
+      "integrity": "sha512-Nrg0BWpQqrC3ZFFkyewrflCud9dio9ME3ojHCF/WLsprJVzkq3q3UeEhMCAW1dobjeGbWgjNn/PVF6m46ANxXQ==",
       "dependencies": {
-        "@remix-run/router": "1.4.0"
+        "@remix-run/router": "1.5.0"
       },
       "engines": {
         "node": ">=14"
@@ -8910,12 +8910,12 @@
       }
     },
     "node_modules/react-router-dom": {
-      "version": "6.9.0",
-      "resolved": "https://registry.npmjs.org/react-router-dom/-/react-router-dom-6.9.0.tgz",
-      "integrity": "sha512-/seUAPY01VAuwkGyVBPCn1OXfVbaWGGu4QN9uj0kCPcTyNYgL1ldZpxZUpRU7BLheKQI4Twtl/OW2nHRF1u26Q==",
+      "version": "6.10.0",
+      "resolved": "https://registry.npmjs.org/react-router-dom/-/react-router-dom-6.10.0.tgz",
+      "integrity": "sha512-E5dfxRPuXKJqzwSe/qGcqdwa18QiWC6f3H3cWXM24qj4N0/beCIf/CWTipop2xm7mR0RCS99NnaqPNjHtrAzCg==",
       "dependencies": {
-        "@remix-run/router": "1.4.0",
-        "react-router": "6.9.0"
+        "@remix-run/router": "1.5.0",
+        "react-router": "6.10.0"
       },
       "engines": {
         "node": ">=14"
@@ -10090,9 +10090,9 @@
       }
     },
     "node_modules/typescript": {
-      "version": "5.0.2",
-      "resolved": "https://registry.npmjs.org/typescript/-/typescript-5.0.2.tgz",
-      "integrity": "sha512-wVORMBGO/FAs/++blGNeAVdbNKtIh1rbBL2EyQ1+J9lClJ93KiiKe8PmFIVdXhHcyv44SL9oglmfeSsndo0jRw==",
+      "version": "5.0.3",
+      "resolved": "https://registry.npmjs.org/typescript/-/typescript-5.0.3.tgz",
+      "integrity": "sha512-xv8mOEDnigb/tN9PSMTwSEqAnUvkoXMQlicOb0IUVDBSQCgBSaAAROUZYy2IcUy5qU6XajK5jjjO7TMWqBTKZA==",
       "dev": true,
       "bin": {
         "tsc": "bin/tsc",
@@ -10435,9 +10435,9 @@
       }
     },
     "node_modules/vite-plugin-dts": {
-      "version": "2.1.0",
-      "resolved": "https://registry.npmjs.org/vite-plugin-dts/-/vite-plugin-dts-2.1.0.tgz",
-      "integrity": "sha512-Vw0FdCuM3VLR4hTFHh0yMEzfwI7NyFvPIMFwvE+Q0t4qtoHIfYOP/JXs7nTnHuQk87FSjlhGeIJ1fLBcktgPgA==",
+      "version": "2.2.0",
+      "resolved": "https://registry.npmjs.org/vite-plugin-dts/-/vite-plugin-dts-2.2.0.tgz",
+      "integrity": "sha512-XmZtv02I7eGWm3DrZbLo1AdJK5gCisk9GqJBpY4N63pDYR6AVUnlyjFP5FCBvSBUfgE0Ppl90bKgtJU9k3AzFw==",
       "dev": true,
       "dependencies": {
         "@babel/parser": "^7.20.15",
diff --git a/js/package.json b/js/package.json
index 43d3d99f97fb..bfcd737052bd 100644
--- a/js/package.json
+++ b/js/package.json
@@ -18,8 +18,8 @@
     "@typescript-eslint/parser": "^5.57.0",
     "eslint": "^8.37.0",
     "eslint-config-prettier": "^8.8.0",
-    "eslint-import-resolver-typescript": "^3.5.3",
-    "eslint-plugin-cypress": "^2.13.1",
+    "eslint-import-resolver-typescript": "^3.5.4",
+    "eslint-plugin-cypress": "^2.13.2",
     "eslint-plugin-import": "^2.27.5",
     "eslint-plugin-lodash": "^7.4.0",
     "eslint-plugin-mocha": "^10.1.0",
@@ -28,7 +28,7 @@
     "husky": "^8.0.3",
     "lint-staged": "^13.2.0",
     "prettier": "^2.8.7",
-    "typescript": "^5.0.2",
+    "typescript": "^5.0.3",
     "wireit": "^0.9.5"
   },
   "lint-staged": {
diff --git a/services/src/main/java/org/keycloak/protocol/oidc/OIDCLoginProtocolService.java b/services/src/main/java/org/keycloak/protocol/oidc/OIDCLoginProtocolService.java
index 0cb195eacc0c..081cabb28bc2 100644
--- a/services/src/main/java/org/keycloak/protocol/oidc/OIDCLoginProtocolService.java
+++ b/services/src/main/java/org/keycloak/protocol/oidc/OIDCLoginProtocolService.java
@@ -219,7 +219,7 @@ public Response certs() {
                     if (k.getType().equals(KeyType.RSA)) {
                         return b.rsa(k.getPublicKey(), certificates, k.getUse());
                     } else if (k.getType().equals(KeyType.EC)) {
-                        return b.ec(k.getPublicKey());
+                        return b.ec(k.getPublicKey(), k.getUse());
                     }
                     return null;
                 })
