diff --git a/docs/guides/server/containers.adoc b/docs/guides/server/containers.adoc
index 6e9617f617dc..2781a4562e8c 100644
--- a/docs/guides/server/containers.adoc
+++ b/docs/guides/server/containers.adoc
@@ -193,14 +193,7 @@ Opening up `https://localhost:9000/metrics` leads to a page containing operation
 === Known issues with Docker
 
 * If a `RUN dnf install` command seems to be taking an excessive amount of time, then likely your Docker systemd service has the file limit setting `LimitNOFILE` configured incorrectly.
-Either update the service configuration to use a better value, such as 1024000, or directly use `ulimit` in the RUN command:
-
-[source, dockerfile]
-----
-...
-RUN ulimit -n 1024000 && dnf install --installroot ...
-...
-----
+Either update the service configuration to use a better value, such as 1024000, or use `--ulimit` in the https://docs.docker.com/reference/cli/docker/buildx/build/#ulimit[docker build command], e.g. `--ulimit nofile=1024000`.
 
 * If you are including provider JARs and your container fails a `start --optimized` with a notification that a provider JAR has changed, this is due to Docker truncating
 or otherwise modifying file modification timestamps from what the `build` command recorded to what is seen at runtime.
diff --git a/js/apps/admin-ui/maven-resources/theme/keycloak.v2/admin/messages/messages_en.properties b/js/apps/admin-ui/maven-resources/theme/keycloak.v2/admin/messages/messages_en.properties
index 3b035b93272e..d7055732ada6 100644
--- a/js/apps/admin-ui/maven-resources/theme/keycloak.v2/admin/messages/messages_en.properties
+++ b/js/apps/admin-ui/maven-resources/theme/keycloak.v2/admin/messages/messages_en.properties
@@ -1398,9 +1398,11 @@ deleteClientPolicyConditionConfirm=This action will permanently delete {{conditi
 selectATheme=Select a theme
 themeColors=Theme colors
 defaults=Reset to defaults
-themeColorsLight=Theme colors [light]
-themeColorsDark=Theme colors [dark]
-themePreviewInfo=In order to preview the theme colors, the current themen needs to be set to the one you want to preview, so we have automatically switched you to the one you want to preview.
+quickTheme=Quick Theme
+themeMode=Theme mode
+lightMode=Light
+darkMode=Dark
+themePreviewInfo=In order to preview the theme colors, the current theme needs to be set to the one you want to preview, so we have automatically switched you to the one you want to preview.
 backgroundImage=Login background image
 favicon=Favicon
 errorColor=Error color
diff --git a/js/apps/admin-ui/src/realm-settings/routes/ThemesTab.tsx b/js/apps/admin-ui/src/realm-settings/routes/ThemesTab.tsx
index 57d109f7bb7c..b81952d04a4b 100644
--- a/js/apps/admin-ui/src/realm-settings/routes/ThemesTab.tsx
+++ b/js/apps/admin-ui/src/realm-settings/routes/ThemesTab.tsx
@@ -3,7 +3,7 @@ import type { Path } from "react-router-dom";
 import { generateEncodedPath } from "../../utils/generateEncodedPath";
 import type { AppRouteObject } from "../../routes";
 
-export type ThemesTabType = "settings" | "lightColors" | "darkColors";
+export type ThemesTabType = "settings" | "quickTheme";
 
 export type ThemesParams = {
   realm: string;
diff --git a/js/apps/admin-ui/src/realm-settings/themes/ThemeColors.tsx b/js/apps/admin-ui/src/realm-settings/themes/ThemeColors.tsx
index f362c5562766..d20827b314d6 100644
--- a/js/apps/admin-ui/src/realm-settings/themes/ThemeColors.tsx
+++ b/js/apps/admin-ui/src/realm-settings/themes/ThemeColors.tsx
@@ -11,8 +11,10 @@ import {
   Text,
   TextContent,
   TextInputProps,
+  ToggleGroup,
+  ToggleGroupItem,
 } from "@patternfly/react-core";
-import { useEffect, useMemo } from "react";
+import { useEffect, useMemo, useState } from "react";
 import {
   FormProvider,
   useForm,
@@ -74,10 +76,14 @@ const switchTheme = (theme: ThemeType) => {
 type ThemeColorsProps = {
   realm: RealmRepresentation;
   save: (realm: ThemeRealmRepresentation) => void;
-  theme: "light" | "dark";
+  theme?: "light" | "dark";
 };
 
-export const ThemeColors = ({ realm, save, theme }: ThemeColorsProps) => {
+export const ThemeColors = ({
+  realm,
+  save,
+  theme: initialTheme,
+}: ThemeColorsProps) => {
   const { t } = useTranslation();
   const form = useForm();
   const { handleSubmit, watch } = form;
@@ -86,9 +92,17 @@ export const ThemeColors = ({ realm, save, theme }: ThemeColorsProps) => {
   const [open, toggle, setOpen] = useToggle();
 
   const mediaQuery = window.matchMedia("(prefers-color-scheme: dark)");
+  const getDarkModeFromRealm = () => {
+    const darkMode = realm.attributes?.darkMode;
+    if (darkMode === "false") return "light";
+    if (darkMode === "true") return "dark";
+    return mediaQuery.matches ? "dark" : "light";
+  };
+  const originalTheme = useMemo(() => getDarkModeFromRealm(), []);
+  const [theme, setTheme] = useState<ThemeType>(initialTheme || originalTheme);
   const mapping = useMemo(
     () => (theme === "light" ? lightTheme() : darkTheme()),
-    [],
+    [theme],
   );
 
   const reset = () => {
@@ -136,9 +150,9 @@ export const ThemeColors = ({ realm, save, theme }: ThemeColorsProps) => {
     setupForm();
     switchTheme(theme);
     return () => {
-      switchTheme(mediaQuery.matches ? "dark" : "light");
+      switchTheme(originalTheme);
     };
-  }, [realm]);
+  }, [realm, theme]);
 
   return (
     <>
@@ -162,6 +176,22 @@ export const ThemeColors = ({ realm, save, theme }: ThemeColorsProps) => {
           <FlexItem>
             <FormAccess isHorizontal role="manage-realm">
               <FormProvider {...form}>
+                {!initialTheme && (
+                  <FormGroup label={t("themeMode")}>
+                    <ToggleGroup aria-label={t("themeMode")}>
+                      <ToggleGroupItem
+                        text={t("lightMode")}
+                        isSelected={theme === "light"}
+                        onChange={() => setTheme("light")}
+                      />
+                      <ToggleGroupItem
+                        text={t("darkMode")}
+                        isSelected={theme === "dark"}
+                        onChange={() => setTheme("dark")}
+                      />
+                    </ToggleGroup>
+                  </FormGroup>
+                )}
                 <FormGroup label={t("favicon")}>
                   <ImageUpload name="favicon" />
                 </FormGroup>
diff --git a/js/apps/admin-ui/src/realm-settings/themes/ThemesTab.tsx b/js/apps/admin-ui/src/realm-settings/themes/ThemesTab.tsx
index 0661f7008192..d3da620172f6 100644
--- a/js/apps/admin-ui/src/realm-settings/themes/ThemesTab.tsx
+++ b/js/apps/admin-ui/src/realm-settings/themes/ThemesTab.tsx
@@ -152,8 +152,7 @@ styles=css/login.css css/theme-styles.css
   });
 
   const settingsTab = useRoutableTab(toThemesTab(param("settings")));
-  const lightColorsTab = useRoutableTab(toThemesTab(param("lightColors")));
-  const darkColorsTab = useRoutableTab(toThemesTab(param("darkColors")));
+  const quickThemeTab = useRoutableTab(toThemesTab(param("quickTheme")));
 
   if (!isFeatureEnabled(Feature.QuickTheme)) {
     return <ThemeSettingsTab realm={realm} save={save} />;
@@ -177,23 +176,13 @@ styles=css/login.css css/theme-styles.css
         <ThemeSettingsTab realm={realm} save={save} />
       </Tab>
       <Tab
-        id="lightColors"
-        title={<TabTitleText>{t("themeColorsLight")}</TabTitleText>}
-        data-testid="lightColors-tab"
-        {...lightColorsTab}
+        id="quickTheme"
+        title={<TabTitleText>{t("quickTheme")}</TabTitleText>}
+        data-testid="quickTheme-tab"
+        {...quickThemeTab}
       >
         <LogoContext>
-          <ThemeColors realm={realm} save={saveTheme} theme="light" />
-        </LogoContext>
-      </Tab>
-      <Tab
-        id="darkColors"
-        title={<TabTitleText>{t("themeColorsDark")}</TabTitleText>}
-        data-testid="darkColors-tab"
-        {...darkColorsTab}
-      >
-        <LogoContext>
-          <ThemeColors realm={realm} save={saveTheme} theme="dark" />
+          <ThemeColors realm={realm} save={saveTheme} />
         </LogoContext>
       </Tab>
     </RoutableTabs>
diff --git a/operator/README.md b/operator/README.md
index f9588b2efb56..16d14f6cb9a7 100644
--- a/operator/README.md
+++ b/operator/README.md
@@ -48,7 +48,7 @@ Vanilla minikube does not support Network Policies, and Cilium implements the CN
 Another CNI implementation may work too.
 
 ```bash
-minikube start --addons ingress --cni cilium
+minikube start --addons ingress --cni cilium --cpus=max
 ```
 
 Enable the Minikube Docker daemon:
@@ -113,7 +113,7 @@ minikube addons enable ingress
 To avoid skipping tests that are depending on custom Keycloak images, you need to build those first:
 
 ```bash
-./build-testing-docker-images.sh [SOURCE KEYCLOAK IMAGE TAG] [SOURCE KEYCLOAK IMAGE]
+./scripts/build-testing-docker-images.sh [SOURCE KEYCLOAK IMAGE TAG] [SOURCE KEYCLOAK IMAGE]
 ```
 
 And run the tests passing an extra Java property:
diff --git a/operator/scripts/olm-testing.sh b/operator/scripts/olm-testing.sh
index 12e1b1e2a234..9e856e220238 100755
--- a/operator/scripts/olm-testing.sh
+++ b/operator/scripts/olm-testing.sh
@@ -17,7 +17,7 @@ VERSION="86400000.0.0"
 (
   cd $SCRIPT_DIR/../../quarkus/container
   
-  docker build --build-arg KEYCLOAK_DIST=$(ls keycloak-*.tar.gz) . -t "$REGISTRY/${UUID}keycloak:${VERSION}"
+  docker build --ulimit nofile=1024000 --build-arg KEYCLOAK_DIST=$(ls keycloak-*.tar.gz) . -t "$REGISTRY/${UUID}keycloak:${VERSION}"
   docker push "$REGISTRY/${UUID}keycloak:${VERSION}"
 )
 
diff --git a/operator/src/main/resources/application.properties b/operator/src/main/resources/application.properties
index 85f9b8558e1b..263a646271c6 100644
--- a/operator/src/main/resources/application.properties
+++ b/operator/src/main/resources/application.properties
@@ -30,3 +30,5 @@ quarkus.operator-sdk.bundle.channels=fast
 
 quarkus.operator-sdk.namespaces=JOSDK_WATCH_CURRENT
 quarkus.operator-sdk.generate-with-watched-namespaces=JOSDK_WATCH_CURRENT
+
+quarkus.docker.additional-args=--ulimit,nofile=1024000
diff --git a/services/src/main/java/org/keycloak/url/HostnameV2ProviderFactory.java b/services/src/main/java/org/keycloak/url/HostnameV2ProviderFactory.java
index 8a055108ceb9..035b337ae27e 100644
--- a/services/src/main/java/org/keycloak/url/HostnameV2ProviderFactory.java
+++ b/services/src/main/java/org/keycloak/url/HostnameV2ProviderFactory.java
@@ -105,7 +105,7 @@ private void validateAndSetHostname(String hostname) {
     private URI validateAndCreateUri(String uri, String validationFailedMessage) {
         URI result;
         try {
-            result = URI.create(uri);
+            result = URI.create(uri.endsWith("/") ? uri : uri + "/");
         }
         catch (IllegalArgumentException e) {
             throw new IllegalArgumentException(validationFailedMessage, e);
diff --git a/services/src/test/java/org/keycloak/url/HostnameV2ProviderFactoryTest.java b/services/src/test/java/org/keycloak/url/HostnameV2ProviderFactoryTest.java
index 70c2966895ee..f96687e385dc 100644
--- a/services/src/test/java/org/keycloak/url/HostnameV2ProviderFactoryTest.java
+++ b/services/src/test/java/org/keycloak/url/HostnameV2ProviderFactoryTest.java
@@ -21,6 +21,7 @@
 import java.util.HashMap;
 import java.util.Map;
 
+import org.keycloak.urls.UrlType;
 import org.keycloak.utils.ScopeUtil;
 
 import org.junit.Test;
@@ -40,7 +41,7 @@ public void hostnameUrlValidationTest() throws IOException{
         assertHostname("my-example.com", true);
         assertHostname("192.196.0.0", true);
         assertHostname("[2001:0000:130F:0000:0000:09C0:876A:130B]", true);
-        
+
         assertHostname("https://my-example.com?auth.this", false);
         assertHostname("my-example.com/auth.this", false);
         assertHostname("https://my-example.com:8080#fragment", false);
@@ -49,7 +50,7 @@ public void hostnameUrlValidationTest() throws IOException{
         assertHostname("?my-example.com", false);
         assertHostname("192.196.0.5555", false);
     }
-    
+
     @Test
     public void hostnameUrlExpected() throws IOException {
         Map<String, String> values = new HashMap<>();
@@ -60,6 +61,21 @@ public void hostnameUrlExpected() throws IOException {
                 assertThrows(IllegalArgumentException.class, () -> factory.init(ScopeUtil.createScope(values))).getMessage());
     }
 
+    @Test
+    public void hostnameUrlBaseUri() throws IOException {
+        Map<String, String> values = new HashMap<>();
+        values.put("hostname", "https://full/path");
+        assertBaseUriEndsWithSlash(values);
+        values.put("hostname", "https://full/path/");
+        assertBaseUriEndsWithSlash(values);
+    }
+
+    private void assertBaseUriEndsWithSlash(Map<String, String> values) {
+        HostnameV2ProviderFactory factory = new HostnameV2ProviderFactory();
+        factory.init(ScopeUtil.createScope(values));
+        assertEquals("https://full/path/", factory.create(null).getBaseUri(null, UrlType.FRONTEND).toString());
+    }
+
     private void assertHostname(String hostname, boolean valid) {
         Map<String, String> values = new HashMap<>();
         values.put("hostname", hostname);
