diff --git a/quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/KeycloakMain.java b/quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/KeycloakMain.java
index e9147516dbc2..c12829701909 100644
--- a/quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/KeycloakMain.java
+++ b/quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/KeycloakMain.java
@@ -18,17 +18,21 @@
 package org.keycloak.quarkus.runtime;
 
 import java.util.List;
+import java.util.Properties;
 import java.util.concurrent.ForkJoinPool;
 
 import jakarta.enterprise.context.ApplicationScoped;
 
+import org.keycloak.common.Profile;
 import org.keycloak.common.Version;
 import org.keycloak.infinispan.util.InfinispanUtils;
 import org.keycloak.quarkus.runtime.cli.ExecutionExceptionHandler;
 import org.keycloak.quarkus.runtime.cli.Picocli;
 import org.keycloak.quarkus.runtime.cli.command.AbstractNonServerCommand;
 import org.keycloak.quarkus.runtime.cli.command.DryRunMixin;
+import org.keycloak.quarkus.runtime.configuration.Configuration;
 import org.keycloak.quarkus.runtime.configuration.PersistedConfigSource;
+import org.keycloak.quarkus.runtime.configuration.mappers.PropertyMappers;
 import org.keycloak.quarkus.runtime.integration.jaxrs.QuarkusKeycloakApplication;
 
 import io.quarkus.arc.Arc;
@@ -77,6 +81,15 @@ public void exit(int exitCode) {
         main(args, picocli);
     }
 
+    public static void reset(Properties systemProperties) {
+        System.setProperties((Properties) systemProperties.clone());
+        PropertyMappers.reset();
+        PersistedConfigSource.getInstance().getConfigValueProperties().clear();
+        Profile.reset();
+        Configuration.resetConfig();
+        ExecutionExceptionHandler.resetExceptionTransformers();
+    }
+
     public static void main(String[] args, Picocli picocli) {
         List<String> cliArgs = List.of(args.length == 0 ? new String[] {"-h"} : args);
 
diff --git a/quarkus/runtime/src/test/java/org/keycloak/quarkus/runtime/configuration/AbstractConfigurationTest.java b/quarkus/runtime/src/test/java/org/keycloak/quarkus/runtime/configuration/AbstractConfigurationTest.java
index d6d193a68e84..a3340ac92715 100644
--- a/quarkus/runtime/src/test/java/org/keycloak/quarkus/runtime/configuration/AbstractConfigurationTest.java
+++ b/quarkus/runtime/src/test/java/org/keycloak/quarkus/runtime/configuration/AbstractConfigurationTest.java
@@ -23,10 +23,8 @@
 import java.util.function.Function;
 
 import org.keycloak.Config;
-import org.keycloak.common.Profile;
 import org.keycloak.quarkus.runtime.Environment;
-import org.keycloak.quarkus.runtime.cli.ExecutionExceptionHandler;
-import org.keycloak.quarkus.runtime.configuration.mappers.PropertyMappers;
+import org.keycloak.quarkus.runtime.KeycloakMain;
 
 import io.smallrye.config.ConfigValue;
 import io.smallrye.config.ConfigValue.ConfigValueBuilder;
@@ -71,17 +69,9 @@ public static void setSystemProperty(String key, String value, Runnable runnable
 
     @BeforeClass
     public static void resetConfiguration() {
-        System.setProperties((Properties) SYSTEM_PROPERTIES.clone());
+        KeycloakMain.reset(SYSTEM_PROPERTIES);
         Environment.setHomeDir(Paths.get("src/test/resources/"));
-
         KcEnvConfigSource.ENV_OVERRIDE.clear();
-
-        PropertyMappers.reset();
-        ConfigArgsConfigSource.setCliArgs();
-        PersistedConfigSource.getInstance().getConfigValueProperties().clear();
-        Profile.reset();
-        Configuration.resetConfig();
-        ExecutionExceptionHandler.resetExceptionTransformers();
     }
 
     @After
diff --git a/quarkus/tests/junit5/src/main/java/org/keycloak/Keycloak.java b/quarkus/tests/junit5/src/main/java/org/keycloak/Keycloak.java
index f45c927792bd..ccc3b9b62905 100644
--- a/quarkus/tests/junit5/src/main/java/org/keycloak/Keycloak.java
+++ b/quarkus/tests/junit5/src/main/java/org/keycloak/Keycloak.java
@@ -22,6 +22,7 @@
 import java.nio.file.Path;
 import java.util.ArrayList;
 import java.util.List;
+import java.util.Properties;
 import java.util.concurrent.TimeoutException;
 
 import org.keycloak.common.Version;
@@ -32,6 +33,7 @@
 import org.keycloak.config.SecurityOptions;
 import org.keycloak.platform.Platform;
 import org.keycloak.quarkus.runtime.Environment;
+import org.keycloak.quarkus.runtime.KeycloakMain;
 import org.keycloak.quarkus.runtime.cli.Picocli;
 import org.keycloak.quarkus.runtime.configuration.Configuration;
 import org.keycloak.quarkus.runtime.configuration.IgnoredArtifacts;
@@ -173,6 +175,7 @@ public static Builder builder() {
     private Path homeDir;
     private List<Dependency> dependencies;
     private boolean fipsEnabled;
+    private Properties systemProperties;
 
     public Keycloak() {
         this(null, Version.VERSION, List.of(), false);
@@ -190,6 +193,7 @@ public Keycloak(Path homeDir, String version, List<Dependency> dependencies, boo
     }
 
     private Keycloak start(List<String> args) {
+        systemProperties = (Properties) System.getProperties().clone();
         QuarkusBootstrap.Builder builder = QuarkusBootstrap.builder()
                 .setExistingModel(applicationModel)
                 .setApplicationRoot(applicationModel.getApplicationModule().getModuleDir().toPath())
@@ -216,8 +220,15 @@ private Keycloak start(List<String> args) {
     }
 
     public void stop() throws TimeoutException {
-        if (isRunning()) {
-            closeApplication();
+        try {
+            if (isRunning()) {
+                closeApplication();
+            }
+        } finally {
+            if (systemProperties != null) {
+                KeycloakMain.reset(systemProperties);
+                systemProperties = null;
+            }
         }
     }
 
