diff --git a/operator/src/main/java/org/keycloak/operator/controllers/KeycloakClientBaseController.java b/operator/src/main/java/org/keycloak/operator/controllers/KeycloakClientBaseController.java
index b24d67530e13..1dc5be593cf3 100644
--- a/operator/src/main/java/org/keycloak/operator/controllers/KeycloakClientBaseController.java
+++ b/operator/src/main/java/org/keycloak/operator/controllers/KeycloakClientBaseController.java
@@ -67,6 +67,7 @@
 import org.keycloak.operator.crds.v2alpha1.deployment.Keycloak;
 import org.keycloak.operator.crds.v2alpha1.deployment.KeycloakStatusAggregator;
 import org.keycloak.operator.crds.v2alpha1.deployment.KeycloakStatusCondition;
+import org.keycloak.operator.crds.v2alpha1.deployment.spec.FeatureSpec;
 import org.keycloak.operator.crds.v2alpha1.deployment.spec.HttpSpec;
 import org.keycloak.representations.admin.v2.BaseClientRepresentation;
 
@@ -95,6 +96,7 @@
 public abstract class KeycloakClientBaseController<R extends CustomResource<? extends KeycloakClientSpec<S>, KeycloakClientStatus>, T extends BaseClientRepresentation, S extends BaseClientRepresentation>
         implements Reconciler<R>, Cleaner<R> {
 
+    public static final String CLIENT_ADMIN_API_V2 = "client-admin-api:v2";
     private static final String CLIENT_API_VERSION = "v2";
     private static final String HTTPS = "https";
 
@@ -153,6 +155,12 @@ public UpdateControl<R> reconcile(R resource, Context<R> context) throws Excepti
 
         KeycloakClientStatusAggregator statusAggregator = new KeycloakClientStatusAggregator(resource);
 
+        if (!hasFeatureEnabled(keycloak)) {
+            statusAggregator.setCondition(KeycloakClientStatusCondition.HAS_ERRORS, Boolean.TRUE, "Cannot create/update because the server does not have %s enabled".formatted(CLIENT_ADMIN_API_V2));
+            resource.setStatus(statusAggregator.build());
+            return UpdateControl.patchStatus(resource);
+        }
+
         S client = resource.getSpec().getClient();
         // first convert to the target representation - the spec representation is specialized
         var map = context.getClient().getKubernetesSerialization().convertValue(client, Map.class);
@@ -197,6 +205,15 @@ public UpdateControl<R> reconcile(R resource, Context<R> context) throws Excepti
         return updateControl;
     }
 
+    // TODO: this doesn't mesh well with the current feature concept
+    // we specifically need v2 enabled, so we can't simply check for client-admin-api
+    // - the behavior is also version dependent later verions of keycloak presumably will have client-admin-api:v2
+    //   enabled by default, so we'd need to check specifically for that feature being disabled, or remove this check altogether
+    private boolean hasFeatureEnabled(Keycloak keycloak) {
+        return Optional.ofNullable(keycloak.getSpec().getFeatureSpec()).map(FeatureSpec::getEnabledFeatures)
+                .filter(ef -> ef.contains(CLIENT_ADMIN_API_V2)).isPresent();
+    }
+
     abstract boolean prepareRepresentation(S crRepresentation, T targetRepresentation, Context<?> context);
 
     abstract Class<T> getTargetRepresentation();
@@ -216,6 +233,14 @@ public DeleteControl cleanup(R resource, Context<R> context) throws Exception {
             return DeleteControl.defaultDelete();
         }
 
+        if (!hasFeatureEnabled(keycloak)) {
+            // TODO: this behavior is not very straight-forward. For now just log an error
+            // in the server and move on
+            Log.error("Cannot delete Client $s/%s because the server does not have %s enabled.".formatted(
+                    resource.getMetadata().getNamespace(), resource.getMetadata().getName(), CLIENT_ADMIN_API_V2));
+            return DeleteControl.defaultDelete();
+        }
+
         invoke(resource, context, keycloak, client -> {
             try {
                 client.deleteClient();
diff --git a/operator/src/test/java/org/keycloak/operator/testsuite/integration/KeycloakClientTest.java b/operator/src/test/java/org/keycloak/operator/testsuite/integration/KeycloakClientTest.java
index 447499a3a774..8cd9c49d4d71 100644
--- a/operator/src/test/java/org/keycloak/operator/testsuite/integration/KeycloakClientTest.java
+++ b/operator/src/test/java/org/keycloak/operator/testsuite/integration/KeycloakClientTest.java
@@ -59,7 +59,6 @@
 
 import static org.junit.jupiter.api.Assertions.assertNull;
 
-@DisabledIfApiServerTest
 @Tag(BaseOperatorTest.SLOW)
 @QuarkusTest
 public class KeycloakClientTest extends BaseOperatorTest {
@@ -93,6 +92,7 @@ public void afterEach() {
         k8sclient.secrets().withName(CLIENT_TLS_SECRET).delete();
     }
 
+    @DisabledIfApiServerTest
     @Test
     public void testBasicSamlClientCreationAndDeletionHttp() throws InterruptedException {
         var kc = getTestKeycloakDeployment(false);
@@ -125,12 +125,12 @@ public void testBasicSamlClientCreationAndDeletionHttp() throws InterruptedExcep
         assertNull(k8sclient.resource(client).get());
     }
 
-    @Test
+    @DisabledIfApiServerTest
     public void testBasicOIDCClientCreationAndDeletionHttp() throws InterruptedException {
         helpTestBasicOIDCClientCreationAndDeletion(false);
     }
 
-    @Test
+    @DisabledIfApiServerTest
     public void testBasicOIDCClientCreationAndDeletionHttps() throws InterruptedException {
         helpTestBasicOIDCClientCreationAndDeletion(true);
     }
@@ -220,5 +220,25 @@ private void deployKeycloakWithAdminApiV2(boolean https, Keycloak kc) {
         initCustomBootstrapAdminServiceAccount(kc);
         deployKeycloak(k8sclient, kc, true);
     }
+    
+    @Test
+    public void testFeatureRequired() {
+        var kc = getTestKeycloakDeployment(true);
+        K8sUtils.deployKeycloak(k8sclient, kc, false);
+
+        KeycloakSAMLClient client = new KeycloakSAMLClientBuilder().withNewMetadata().withName("new-client")
+                .endMetadata().withNewSpec().withRealm("master").withKeycloakCRName(kc.getMetadata().getName())
+                .withNewClient().withEnabled(true).endClient().endSpec().build();
+
+        K8sUtils.set(k8sclient, client);
+
+        Awaitility.await().ignoreExceptions()
+                .until(() -> Optional.ofNullable(k8sclient.resource(client).get().getStatus())
+                        .filter(s -> s.getConditions().stream()
+                                .anyMatch(c -> Boolean.TRUE.equals(c.getStatus())
+                                        && KeycloakClientStatusCondition.HAS_ERRORS.equals(c.getType())
+                                        && c.getMessage().contains(KeycloakClientBaseController.CLIENT_ADMIN_API_V2)))
+                        .isPresent());
+    }
 
 }
