diff --git a/quarkus/tests/junit5/src/main/java/org/keycloak/it/utils/Maven.java b/quarkus/tests/junit5/src/main/java/org/keycloak/it/utils/Maven.java
index 857c4b2d6a55..2f707a586322 100644
--- a/quarkus/tests/junit5/src/main/java/org/keycloak/it/utils/Maven.java
+++ b/quarkus/tests/junit5/src/main/java/org/keycloak/it/utils/Maven.java
@@ -45,6 +45,8 @@
 
 public final class Maven {
 
+    private static BootstrapMavenContext context;
+
     public static Path resolveArtifact(String groupId, String artifactId) {
         return getArtifact(groupId, artifactId).getFile().toPath();
     }
@@ -62,6 +64,11 @@ public static Artifact getArtifact(String groupId, String artifactId) {
                             .setRepositories(remoteRepositories));
             List<Dependency> dependencies = new ArrayList<>(projectDescriptor.getDependencies());
             dependencies.addAll(projectDescriptor.getManagedDependencies());
+
+            if (groupId.equals(project.getGroupId()) &&  artifactId.equals(project.getArtifactId())) {
+                return projectDescriptor.getArtifact();
+            }
+
             Artifact artifact = resolveArtifact(groupId, artifactId, dependencies);
 
             if (artifact == null) {
@@ -148,11 +155,14 @@ public static Path getKeycloakQuarkusModulePath() {
         throw new RuntimeException("Failed to find keycloak-parent module.");
     }
 
-    private static BootstrapMavenContext bootstrapCurrentMavenContext() throws BootstrapMavenException, URISyntaxException {
-        Path classPathDir = Paths.get(Thread.currentThread().getContextClassLoader().getResource(".").toURI());
-        Path projectDir = BuildToolHelper.getProjectDir(classPathDir);
-        return new BootstrapMavenContext(
-                BootstrapMavenContext.config().setPreferPomsFromWorkspace(true).setWorkspaceModuleParentHierarchy(true)
-                        .setCurrentProject(projectDir.toString()));
+    public static BootstrapMavenContext bootstrapCurrentMavenContext() throws BootstrapMavenException, URISyntaxException {
+        if (context == null) {
+            Path classPathDir = Paths.get(Thread.currentThread().getContextClassLoader().getResource(".").toURI());
+            Path projectDir = BuildToolHelper.getProjectDir(classPathDir);
+            context = new BootstrapMavenContext(
+                    BootstrapMavenContext.config().setPreferPomsFromWorkspace(true).setWorkspaceModuleParentHierarchy(true)
+                            .setCurrentProject(projectDir.toString()));
+        }
+        return context;
     }
 }
