diff --git a/js/apps/admin-ui/src/context/auth/AdminClient.tsx b/js/apps/admin-ui/src/context/auth/AdminClient.tsx
index de65f1930e6f..e4a95a55f1a2 100644
--- a/js/apps/admin-ui/src/context/auth/AdminClient.tsx
+++ b/js/apps/admin-ui/src/context/auth/AdminClient.tsx
@@ -38,10 +38,11 @@ export function useFetch<T>(
   deps?: DependencyList
 ) {
   const onError = useErrorHandler();
-  const controller = new AbortController();
-  const { signal } = controller;
 
   useEffect(() => {
+    const controller = new AbortController();
+    const { signal } = controller;
+
     adminClientCall()
       .then((result) => {
         if (!signal.aborted) {
