diff --git a/README.md b/README.md
index 37825892cf..e7ecae6c96 100644
--- a/README.md
+++ b/README.md
@@ -42,7 +42,7 @@ Leo supports `let` and `const` keywords for variable definition.
 **Allocated** variables define private variables in the constraint system. Their value is constrained in the circuit on initialization.
 
 **Constant** variables do not define a variable in the constraint system. Their value is constrained in the circuit on computation with an **allocated** variable. 
-**Constant** variables can be mutable. They do not have the same functionality as `const` variables in other languages.
+**Constant** variables cannot be mutable. They have the same functionality as `const` variables in other languages.
 ```rust
 function addOne() -> {
     let a = 0u8;   // allocated, value enforced on this line
diff --git a/compiler/src/constraints/statement.rs b/compiler/src/constraints/statement.rs
index be20365b03..17974bfda9 100644
--- a/compiler/src/constraints/statement.rs
+++ b/compiler/src/constraints/statement.rs
@@ -276,8 +276,13 @@ impl<F: Field + PrimeField, G: GroupType<F>> ConstrainedProgram<F, G> {
             expression,
         )?;
 
-        if let Declare::Let = declare {
-            value.allocate_value(cs, span)?;
+        match declare {
+            Declare::Let => value.allocate_value(cs, span)?,
+            Declare::Const => {
+                if variable.mutable {
+                    return Err(StatementError::immutable_assign(variable.to_string(), span));
+                }
+            }
         }
 
         self.store_definition(function_scope, variable, value)
diff --git a/compiler/src/constraints/value.rs b/compiler/src/constraints/value.rs
index b003af82cb..c423d0681d 100644
--- a/compiler/src/constraints/value.rs
+++ b/compiler/src/constraints/value.rs
@@ -1,12 +1,8 @@
 //! The in memory stored value for a defined name in a resolved Leo program.
 
 use crate::{
-    constraints::{
-        boolean::{allocate_bool, new_bool_constant},
-        field::allocate_field,
-        group::allocate_group,
-    },
-    errors::ValueError,
+    constraints::boolean::{allocate_bool, new_bool_constant},
+    errors::{ExpressionError, FieldError, ValueError},
     new_scope,
     FieldType,
     GroupType,
@@ -14,7 +10,6 @@ use crate::{
 };
 use leo_types::{Circuit, Function, Identifier, Span, Type};
 
-use crate::errors::ExpressionError;
 use snarkos_errors::gadgets::SynthesisError;
 use snarkos_models::{
     curves::{Field, PrimeField},
@@ -167,16 +162,14 @@ impl<F: Field + PrimeField, G: GroupType<F>> ConstrainedValue<F, G> {
                 *integer = Integer::allocate_type(&mut cs, integer_type, name, option, span)?;
             }
             ConstrainedValue::Field(field) => {
-                let option = field.get_value().map(|v| format!("{}", v));
-                let name = option.clone().map(|f| f.to_string()).unwrap_or(format!("[allocated]"));
+                let gadget = field
+                    .allocated(cs.ns(|| format!("allocate field {}:{}", span.line, span.start)))
+                    .map_err(|error| ValueError::FieldError(FieldError::synthesis_error(error, span)))?;
 
-                *field = allocate_field(&mut cs, name, option, span)?;
+                *field = FieldType::Allocated(gadget)
             }
             ConstrainedValue::Group(group) => {
-                let name = format!("{}", group); // may need to implement u256 -> decimal formatting
-                let option = Some(name.clone());
-
-                *group = allocate_group(&mut cs, name, option, span)?;
+                *group = group.to_allocated(cs, span)?;
             }
             // value wrappers
             ConstrainedValue::Array(array) => {
diff --git a/compiler/src/errors/constraints/field.rs b/compiler/src/errors/constraints/field.rs
index 05e1d2f437..3fc81f200d 100644
--- a/compiler/src/errors/constraints/field.rs
+++ b/compiler/src/errors/constraints/field.rs
@@ -36,7 +36,7 @@ impl FieldError {
     }
 
     pub fn missing_field(expected: String, span: Span) -> Self {
-        let message = format!("expected integer input `{}` not found", expected);
+        let message = format!("expected field input `{}` not found", expected);
 
         Self::new_from_span(message, span)
     }
@@ -46,4 +46,10 @@ impl FieldError {
 
         Self::new_from_span(message, span)
     }
+
+    pub fn synthesis_error(error: SynthesisError, span: Span) -> Self {
+        let message = format!("compilation failed due to field synthesis error `{}`", error);
+
+        Self::new_from_span(message, span)
+    }
 }
diff --git a/compiler/src/errors/constraints/group.rs b/compiler/src/errors/constraints/group.rs
index 5e7fdcefd6..6531cc356e 100644
--- a/compiler/src/errors/constraints/group.rs
+++ b/compiler/src/errors/constraints/group.rs
@@ -40,4 +40,10 @@ impl GroupError {
 
         Self::new_from_span(message, span)
     }
+
+    pub fn synthesis_error(error: SynthesisError, span: Span) -> Self {
+        let message = format!("compilation failed due to group synthesis error `{}`", error);
+
+        Self::new_from_span(message, span)
+    }
 }
diff --git a/compiler/src/field/mod.rs b/compiler/src/field/mod.rs
index 69a894e52f..c7837fda28 100644
--- a/compiler/src/field/mod.rs
+++ b/compiler/src/field/mod.rs
@@ -150,7 +150,9 @@ impl<F: Field + PrimeField> FieldType<F> {
     pub fn allocated<CS: ConstraintSystem<F>>(&self, mut cs: CS) -> Result<FpGadget<F>, SynthesisError> {
         match self {
             FieldType::Constant(constant) => FpGadget::alloc(&mut cs.ns(|| format!("{:?}", constant)), || Ok(constant)),
-            FieldType::Allocated(allocated) => Ok(allocated.clone()),
+            FieldType::Allocated(allocated) => FpGadget::alloc(&mut cs.ns(|| format!("{:?}", allocated)), || {
+                allocated.value.ok_or(SynthesisError::AssignmentMissing)
+            }),
         }
     }
 }
diff --git a/compiler/src/group/edwards_bls12.rs b/compiler/src/group/edwards_bls12.rs
index c6467639e4..9ea55eb730 100644
--- a/compiler/src/group/edwards_bls12.rs
+++ b/compiler/src/group/edwards_bls12.rs
@@ -10,7 +10,7 @@ use snarkos_gadgets::curves::edwards_bls12::EdwardsBlsGadget;
 use snarkos_models::{
     curves::AffineCurve,
     gadgets::{
-        curves::{FpGadget, GroupGadget},
+        curves::{FieldGadget, FpGadget, GroupGadget},
         r1cs::ConstraintSystem,
         utilities::{
             alloc::AllocGadget,
@@ -39,6 +39,12 @@ impl GroupType<Fq> for EdwardsGroupType {
         Ok(EdwardsGroupType::Constant(value))
     }
 
+    fn to_allocated<CS: ConstraintSystem<Fq>>(&self, mut cs: CS, span: Span) -> Result<Self, GroupError> {
+        self.allocated(cs.ns(|| format!("allocate affine point {}:{}", span.line, span.start)))
+            .map(|result| EdwardsGroupType::Allocated(result))
+            .map_err(|error| GroupError::synthesis_error(error, span))
+    }
+
     fn add<CS: ConstraintSystem<Fq>>(&self, cs: CS, other: &Self, span: Span) -> Result<Self, GroupError> {
         match (self, other) {
             (EdwardsGroupType::Constant(self_value), EdwardsGroupType::Constant(other_value)) => {
@@ -127,7 +133,19 @@ impl EdwardsGroupType {
                     || Ok(constant),
                 )
             }
-            EdwardsGroupType::Allocated(allocated) => Ok(allocated.clone()),
+            EdwardsGroupType::Allocated(allocated) => {
+                let x_value = allocated.x.get_value();
+                let y_value = allocated.y.get_value();
+
+                let x_allocated = FpGadget::alloc(cs.ns(|| format!("x")), || {
+                    x_value.ok_or(SynthesisError::AssignmentMissing)
+                })?;
+                let y_allocated = FpGadget::alloc(cs.ns(|| format!("y")), || {
+                    y_value.ok_or(SynthesisError::AssignmentMissing)
+                })?;
+
+                Ok(EdwardsBlsGadget::new(x_allocated, y_allocated))
+            }
         }
     }
 }
diff --git a/compiler/src/group/mod.rs b/compiler/src/group/mod.rs
index 5ff7c8aefb..c35b3aac9a 100644
--- a/compiler/src/group/mod.rs
+++ b/compiler/src/group/mod.rs
@@ -35,6 +35,8 @@ pub trait GroupType<F: Field>:
 {
     fn constant(string: String, span: Span) -> Result<Self, GroupError>;
 
+    fn to_allocated<CS: ConstraintSystem<F>>(&self, cs: CS, span: Span) -> Result<Self, GroupError>;
+
     fn add<CS: ConstraintSystem<F>>(&self, cs: CS, other: &Self, span: Span) -> Result<Self, GroupError>;
 
     fn sub<CS: ConstraintSystem<F>>(&self, cs: CS, other: &Self, span: Span) -> Result<Self, GroupError>;
diff --git a/compiler/tests/mutability/const.leo b/compiler/tests/mutability/const.leo
new file mode 100644
index 0000000000..a167625c02
--- /dev/null
+++ b/compiler/tests/mutability/const.leo
@@ -0,0 +1,5 @@
+// Constant variables are immutable by default.
+function main() {
+  const a = 1u32;
+  a = 0;
+}
\ No newline at end of file
diff --git a/compiler/tests/mutability/const_mut.leo b/compiler/tests/mutability/const_mut.leo
new file mode 100644
index 0000000000..06dc51c222
--- /dev/null
+++ b/compiler/tests/mutability/const_mut.leo
@@ -0,0 +1,4 @@
+// Adding the `mut` keyword to a constant variable is illegal
+function main() {
+  const mut a = 1u32;
+}
\ No newline at end of file
diff --git a/compiler/tests/mutability/mod.rs b/compiler/tests/mutability/mod.rs
index 41b03949a7..cc8dcdb2dd 100644
--- a/compiler/tests/mutability/mod.rs
+++ b/compiler/tests/mutability/mod.rs
@@ -47,6 +47,22 @@ fn test_let_mut() {
     mut_success(program);
 }
 
+#[test]
+fn test_const_fail() {
+    let bytes = include_bytes!("const.leo");
+    let program = parse_program(bytes).unwrap();
+
+    mut_fail(program);
+}
+
+#[test]
+fn test_const_mut_fail() {
+    let bytes = include_bytes!("const_mut.leo");
+    let program = parse_program(bytes).unwrap();
+
+    mut_fail(program);
+}
+
 #[test]
 fn test_array() {
     let bytes = include_bytes!("array.leo");
diff --git a/examples/pedersen_hash/src/main.leo b/examples/pedersen_hash/src/main.leo
index 5b417f604f..ea38b77e81 100644
--- a/examples/pedersen_hash/src/main.leo
+++ b/examples/pedersen_hash/src/main.leo
@@ -1,12 +1,12 @@
 circuit PedersenHash {
-    parameters: u32[512]
-    static function new(parameters: u32[512]) -> Self {
+    parameters: group[512]
+    static function new(parameters: group[512]) -> Self {
         return Self { parameters: parameters }
     }
-    function hash(bits: bool[512]) -> u32 {
-        let mut digest: u32 = 0;
+    function hash(bits: bool[512]) -> group {
+        let mut digest: group = 0;
         for i in 0..512 {
-            let base = if bits[i] ? parameters[i] : 0u32;
+            let base = if bits[i] ? parameters[i] : 0group;
             digest += base;
         }
         return digest
@@ -14,8 +14,8 @@ circuit PedersenHash {
 }
 
 // The 'pedersen_hash' main function.
-function main() -> u32 {
-    const parameters = [0u32; 512];
+function main() -> group {
+    let parameters = [0group; 512];
     const pedersen = PedersenHash::new(parameters);
     let input: bool[512] = [true; 512];
     return pedersen.hash(input)
