diff --git a/leo/commands/build.rs b/leo/commands/build.rs
index 394be25ee3..222de665b8 100644
--- a/leo/commands/build.rs
+++ b/leo/commands/build.rs
@@ -180,6 +180,9 @@ impl Command for Build {
             None
         };
 
+        // Unset the Leo panic hook.
+        let _ = std::panic::take_hook();
+
         // Change the cwd to the build directory to compile aleo files.
         std::env::set_current_dir(&build_directory)
             .map_err(|err| PackageError::failed_to_set_cwd(build_directory.display(), err))?;
diff --git a/leo/commands/deploy.rs b/leo/commands/deploy.rs
index 5adabaa96f..e2b39991c3 100644
--- a/leo/commands/deploy.rs
+++ b/leo/commands/deploy.rs
@@ -16,6 +16,7 @@
 
 use crate::commands::ALEO_CLI_COMMAND;
 use crate::{commands::Command, context::Context};
+
 use leo_errors::{CliError, PackageError, Result};
 use leo_package::build::BuildDirectory;
 
@@ -49,6 +50,9 @@ impl Command for Deploy {
         std::env::set_current_dir(&build_directory)
             .map_err(|err| PackageError::failed_to_set_cwd(build_directory.display(), err))?;
 
+        // Unset the Leo panic hook.
+        let _ = std::panic::take_hook();
+
         // Call the `aleo node` command from the Aleo SDK.
         println!();
         let command = AleoDeploy::try_parse_from([ALEO_CLI_COMMAND]).map_err(CliError::failed_to_parse_aleo_node)?;
diff --git a/leo/commands/new.rs b/leo/commands/new.rs
index b438bca449..3b2ec8f1f6 100644
--- a/leo/commands/new.rs
+++ b/leo/commands/new.rs
@@ -66,7 +66,7 @@ impl Command for New {
         // Initialize the Leo package in the directory created by `aleo new`.
         Package::initialize(&self.name, &package_path)?;
 
-        // Change the cwd to the Leo package directory. to compile aleo files.
+        // Change the cwd to the Leo package directory to compile aleo files.
         std::env::set_current_dir(&package_path)
             .map_err(|err| PackageError::failed_to_set_cwd(package_path.display(), err))?;
 
diff --git a/leo/commands/run.rs b/leo/commands/run.rs
index de641013c5..6f0dfeefbb 100644
--- a/leo/commands/run.rs
+++ b/leo/commands/run.rs
@@ -86,6 +86,10 @@ impl Command for Run {
         if self.compiler_options.offline {
             arguments.push(String::from("--offline"));
         }
+
+        // Unset the Leo panic hook
+        let _ = std::panic::take_hook();
+
         println!();
         let command = AleoRun::try_parse_from(&arguments).map_err(CliError::failed_to_parse_aleo_run)?;
         let res = command.parse().map_err(CliError::failed_to_execute_aleo_run)?;
