diff --git a/leo/main.rs b/leo/main.rs
index 61aa8a7b0c..7c8c124904 100644
--- a/leo/main.rs
+++ b/leo/main.rs
@@ -22,9 +22,8 @@ use clap::{App, AppSettings, Arg};
 fn main() -> Result<(), CLIError> {
     let app = App::new("leo")
         .version(include_str!("./leo-version"))
-        .before_help("Hello world")
-        // .about("Leo compiler and package manager")
-        // .author("The Aleo Team <hello@aleo.org>")
+        .about("Leo compiler and package manager")
+        .author("The Aleo Team <hello@aleo.org>")
         .settings(&[
             AppSettings::ColoredHelp,
             AppSettings::DisableHelpSubcommand,
