diff --git a/compiler/passes/src/const_propagation/ast.rs b/compiler/passes/src/const_propagation/ast.rs
index b0224b75f7..6554c8cb50 100644
--- a/compiler/passes/src/const_propagation/ast.rs
+++ b/compiler/passes/src/const_propagation/ast.rs
@@ -88,12 +88,11 @@ impl AstReconstructor for ConstPropagationVisitor<'_> {
             *arg = self.reconstruct_expression(std::mem::take(arg)).0;
         });
         for member in input.members.iter_mut() {
-            if let Some(expr) = std::mem::take(&mut member.expression) {
-                let (new_expr, value_opt) = self.reconstruct_expression(expr);
-                member.expression = Some(new_expr);
-                if let Some(value) = value_opt {
-                    values.push(value);
-                }
+            let expression = member.expression.take().unwrap_or_else(|| member.identifier.into());
+            let (new_expr, value_opt) = self.reconstruct_expression(expression);
+            member.expression = Some(new_expr);
+            if let Some(value) = value_opt {
+                values.push(value);
             }
         }
 
diff --git a/compiler/passes/src/type_checking/ast.rs b/compiler/passes/src/type_checking/ast.rs
index 92b771dd44..247415f1bf 100644
--- a/compiler/passes/src/type_checking/ast.rs
+++ b/compiler/passes/src/type_checking/ast.rs
@@ -1025,24 +1025,8 @@ impl AstVisitor for TypeCheckingVisitor<'_> {
 
         // Async functions return a single future.
         let mut ret = if func.variant == Variant::AsyncFunction {
-            // Type check after resolving the input types.
-            let actual = Type::Future(FutureType::new(
-                Vec::new(),
-                Some(Location::new(callee_program, input.function.name)),
-                false,
-            ));
-            match expected {
-                Some(Type::Future(_)) | None => {
-                    // If the expected type is a `Future` or if it's not set, then just return the
-                    // actual type of the future from the expression itself
-                    actual
-                }
-                Some(_) => {
-                    // Otherwise, error out. There is a mismatch in types.
-                    self.maybe_assert_type(&actual, expected, input.span());
-                    Type::Unit
-                }
-            }
+            // Async functions always return futures.
+            Type::Future(FutureType::new(Vec::new(), Some(Location::new(callee_program, input.function.name)), false))
         } else if func.variant == Variant::AsyncTransition {
             // Fully infer future type.
             let Some(inputs) = self
diff --git a/leo/cli/commands/build.rs b/leo/cli/commands/build.rs
index a2074f4893..a2e0c1fc41 100644
--- a/leo/cli/commands/build.rs
+++ b/leo/cli/commands/build.rs
@@ -215,7 +215,7 @@ fn compile_leo_file(
         NetworkName::CanaryV0 => Program::<CanaryV0>::from_str(&bytecode)?.to_checksum().iter().join(", "),
     };
 
-    tracing::info!("    \n{} statements before dead code elimination.", compiler.statements_before_dce);
+    tracing::info!("    {} statements before dead code elimination.", compiler.statements_before_dce);
     tracing::info!("    {} statements after dead code elimination.", compiler.statements_after_dce);
     tracing::info!("    The program checksum is: '[{checksum}]'.");
 
diff --git a/leo/cli/commands/clean.rs b/leo/cli/commands/clean.rs
index 828b07c403..80cd32f37c 100644
--- a/leo/cli/commands/clean.rs
+++ b/leo/cli/commands/clean.rs
@@ -14,6 +14,8 @@
 // You should have received a copy of the GNU General Public License
 // along with the Leo library. If not, see <https://www.gnu.org/licenses/>.
 
+use anyhow::anyhow;
+
 use super::*;
 
 /// Clean outputs folder command
@@ -35,17 +37,27 @@ impl Command for LeoClean {
     fn apply(self, context: Context, _: Self::Input) -> Result<Self::Output> {
         let path = context.dir()?;
 
+        let manifest_path = path.join(leo_package::MANIFEST_FILENAME);
+
+        if !manifest_path.exists() {
+            return Err(anyhow!(
+                "{} doesn't exist - this doesn't appear to be a Leo package.",
+                leo_package::MANIFEST_FILENAME
+            )
+            .into());
+        }
+
         // Removes the outputs/ directory.
         let outputs_path = path.join(leo_package::OUTPUTS_DIRECTORY);
-        std::fs::remove_dir_all(&outputs_path)
-            .map_err(|e| PackageError::failed_to_remove_directory(outputs_path.display(), e))?;
-        tracing::info!("🧹 Cleaned the outputs directory {}", outputs_path.display().to_string().dimmed());
+        if std::fs::remove_dir_all(&outputs_path).is_ok() {
+            tracing::info!("🧹 Cleaned the outputs directory {}", outputs_path.display().to_string().dimmed());
+        }
 
         // Removes the build/ directory.
         let build_path = path.join(leo_package::BUILD_DIRECTORY);
-        std::fs::remove_dir_all(&build_path)
-            .map_err(|e| PackageError::failed_to_remove_directory(build_path.display(), e))?;
-        tracing::info!("🧹 Cleaned the build directory {}", build_path.display().to_string().dimmed());
+        if std::fs::remove_dir_all(&build_path).is_ok() {
+            tracing::info!("🧹 Cleaned the build directory {}", build_path.display().to_string().dimmed());
+        }
 
         Ok(())
     }
diff --git a/tests/expectations/compiler/bugs/b28778.out b/tests/expectations/compiler/bugs/b28778.out
new file mode 100644
index 0000000000..c9c2dad369
--- /dev/null
+++ b/tests/expectations/compiler/bugs/b28778.out
@@ -0,0 +1,10 @@
+program b28788.aleo;
+
+struct Foo:
+    x as u32;
+
+function main:
+    input r0 as u32.public;
+    input r1 as u32.private;
+    cast 42u32 into r2 as Foo;
+    output r2.x as u32.private;
diff --git a/tests/expectations/compiler/bugs/b28793.out b/tests/expectations/compiler/bugs/b28793.out
new file mode 100644
index 0000000000..c97ca8cffd
--- /dev/null
+++ b/tests/expectations/compiler/bugs/b28793.out
@@ -0,0 +1,5 @@
+Error [ETYC0372003]: Expected type `()` but type `Future<Fn()>` was found
+    --> compiler-test:4:16
+     |
+   4 |         return bar();
+     |                ^^^^^
diff --git a/tests/expectations/compiler/finalize/unknown_mapping_operation_fail.out b/tests/expectations/compiler/finalize/unknown_mapping_operation_fail.out
index e6f02251eb..81150de355 100644
--- a/tests/expectations/compiler/finalize/unknown_mapping_operation_fail.out
+++ b/tests/expectations/compiler/finalize/unknown_mapping_operation_fail.out
@@ -1,8 +1,3 @@
-Error [ETYC0372117]: Expected type `()` but type `Future<Fn()>` was found.
-    --> compiler-test:7:17
-     |
-   7 |          return finalize_mint_public(receiver, amount);
-     |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 Error [ETYC0372003]: Expected type `()` but type `Future<Fn(address,u64)>` was found
     --> compiler-test:7:17
      |
diff --git a/tests/expectations/compiler/type_inference/futures_fail.out b/tests/expectations/compiler/type_inference/futures_fail.out
index fe8d12283f..cfbfa58279 100644
--- a/tests/expectations/compiler/type_inference/futures_fail.out
+++ b/tests/expectations/compiler/type_inference/futures_fail.out
@@ -3,11 +3,6 @@ Error [ETYC0372003]: Expected type `u32` but type `Future<Fn()>` was found
      |
    7 |         let f: u32 = child.aleo/foo();
      |                      ^^^^^^^^^^^^^^^^
-Error [ETYC0372117]: Expected type `u8` but type `Future<Fn()>` was found.
-    --> compiler-test:8:21
-     |
-   8 |         let r: u8 = foo(f);
-     |                     ^^^^^^
 Error [ETYC0372117]: Expected type `Future<Fn()>` but type `u32` was found.
     --> compiler-test:8:25
      |
@@ -35,11 +30,6 @@ Error [ETYC0372003]: Expected type `(u8, u8)` but type `Future<Fn()>` was found
      |
   13 |         let f: (u8, u8) = child.aleo/bar();
      |                           ^^^^^^^^^^^^^^^^
-Error [ETYC0372117]: Expected type `bool` but type `Future<Fn()>` was found.
-    --> compiler-test:14:23
-     |
-  14 |         let r: bool = foo(f);
-     |                       ^^^^^^
 Error [ETYC0372117]: Expected type `Future<Fn()>` but type `(u8, u8)` was found.
     --> compiler-test:14:27
      |
diff --git a/tests/tests/compiler/bugs/b28778.leo b/tests/tests/compiler/bugs/b28778.leo
new file mode 100644
index 0000000000..a2eb362bcf
--- /dev/null
+++ b/tests/tests/compiler/bugs/b28778.leo
@@ -0,0 +1,13 @@
+program b28788.aleo {
+    const x: u32 = 42;
+
+    struct Foo {
+        x: u32,
+    }
+
+    transition main(public a: u32, b: u32) -> u32 {
+        // Previously the `x` would lead to a panic in codegen.
+        let f = Foo { x };
+        return f.x;
+    }
+}
diff --git a/tests/tests/compiler/bugs/b28793.leo b/tests/tests/compiler/bugs/b28793.leo
new file mode 100644
index 0000000000..a15b451c5c
--- /dev/null
+++ b/tests/tests/compiler/bugs/b28793.leo
@@ -0,0 +1,10 @@
+program b28610.aleo {
+    // This should produce only one type error.
+    async transition foo() {
+        return bar();
+    }
+
+    async function bar() {
+        assert(true);
+    }
+}
