diff --git a/.github/workflows/leo.yml b/.github/workflows/release-tests.yml
similarity index 86%
rename from .github/workflows/leo.yml
rename to .github/workflows/release-tests.yml
index 750e79d6d8..f97ee3384e 100644
--- a/.github/workflows/leo.yml
+++ b/.github/workflows/release-tests.yml
@@ -1,4 +1,4 @@
-name: Leo Programs
+name: Leo Release Tests
 on:
   pull_request:
   push:
@@ -9,7 +9,7 @@ env:
 
 jobs:
   new:
-    name: Hello Leo (from 'leo new')
+    name: Hello Leo ('leo new hello-world')
     runs-on: ubuntu-latest
     steps:
       - name: Checkout
@@ -39,16 +39,16 @@ jobs:
           command: install
           args: --path .
 
-      - name: 'leo new'
+      - name: 'leo new hello-world'
         run: |
           cd ..
-          leo new hello_world
+          leo new hello-world
           ls -la
-          cd hello_world && ls -la
+          cd hello-world && ls -la
           leo run
 
   init:
-    name: Hello Leo (from 'leo init')
+    name: Hello Leo ('leo init')
     runs-on: ubuntu-latest
     steps:
       - name: Checkout
@@ -80,7 +80,7 @@ jobs:
 
       - name: 'leo init'
         run: |
-          cd .. && mkdir hello_world && cd hello_world
+          cd .. && mkdir hello-world && cd hello-world
           leo init
           ls -la
           leo run
diff --git a/Cargo.lock b/Cargo.lock
index 859062f9eb..c18c6695ce 100644
--- a/Cargo.lock
+++ b/Cargo.lock
@@ -1326,6 +1326,7 @@ version = "1.0.2"
 name = "leo-package"
 version = "1.0.2"
 dependencies = [
+ "lazy_static",
  "serde",
  "serde_json",
  "thiserror",
diff --git a/ast/src/circuits/circuit_variable_definition.rs b/ast/src/circuits/circuit_variable_definition.rs
index 533fff7455..8bd462a099 100644
--- a/ast/src/circuits/circuit_variable_definition.rs
+++ b/ast/src/circuits/circuit_variable_definition.rs
@@ -14,7 +14,12 @@
 // You should have received a copy of the GNU General Public License
 // along with the Leo library. If not, see <https://www.gnu.org/licenses/>.
 
-use crate::{ast::Rule, common::Identifier, types::Type, SpanDef};
+use crate::{
+    ast::Rule,
+    common::{Identifier, Mutable},
+    types::Type,
+    SpanDef,
+};
 
 use pest::Span;
 use pest_ast::FromPest;
@@ -23,8 +28,9 @@ use serde::Serialize;
 #[derive(Clone, Debug, FromPest, PartialEq, Serialize)]
 #[pest_ast(rule(Rule::circuit_variable_definition))]
 pub struct CircuitVariableDefinition<'ast> {
+    pub mutable: Option<Mutable>,
     pub identifier: Identifier<'ast>,
-    pub _type: Type<'ast>,
+    pub type_: Type<'ast>,
     #[pest_ast(outer())]
     #[serde(with = "SpanDef")]
     pub span: Span<'ast>,
diff --git a/ast/src/console/console_assert.rs b/ast/src/console/console_assert.rs
index 8f96890025..d64dc0eaa7 100644
--- a/ast/src/console/console_assert.rs
+++ b/ast/src/console/console_assert.rs
@@ -1,3 +1,19 @@
+// Copyright (C) 2019-2020 Aleo Systems Inc.
+// This file is part of the Leo library.
+
+// The Leo library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// The Leo library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with the Leo library. If not, see <https://www.gnu.org/licenses/>.
+
 use crate::{ast::Rule, expressions::Expression, SpanDef};
 
 use pest::Span;
diff --git a/ast/src/console/console_error.rs b/ast/src/console/console_error.rs
index db7cfbd180..45f729a846 100644
--- a/ast/src/console/console_error.rs
+++ b/ast/src/console/console_error.rs
@@ -1,3 +1,19 @@
+// Copyright (C) 2019-2020 Aleo Systems Inc.
+// This file is part of the Leo library.
+
+// The Leo library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// The Leo library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with the Leo library. If not, see <https://www.gnu.org/licenses/>.
+
 use crate::{ast::Rule, console::FormattedString, SpanDef};
 
 use pest::Span;
diff --git a/ast/src/console/console_function.rs b/ast/src/console/console_function.rs
index c137c30c0d..0b48af1a43 100644
--- a/ast/src/console/console_function.rs
+++ b/ast/src/console/console_function.rs
@@ -1,3 +1,19 @@
+// Copyright (C) 2019-2020 Aleo Systems Inc.
+// This file is part of the Leo library.
+
+// The Leo library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// The Leo library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with the Leo library. If not, see <https://www.gnu.org/licenses/>.
+
 use crate::{
     ast::Rule,
     console::{ConsoleAssert, ConsoleDebug, ConsoleError, ConsoleLog},
diff --git a/ast/src/console/console_function_call.rs b/ast/src/console/console_function_call.rs
index f77cb329dc..2c14af2d76 100644
--- a/ast/src/console/console_function_call.rs
+++ b/ast/src/console/console_function_call.rs
@@ -1,3 +1,19 @@
+// Copyright (C) 2019-2020 Aleo Systems Inc.
+// This file is part of the Leo library.
+
+// The Leo library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// The Leo library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with the Leo library. If not, see <https://www.gnu.org/licenses/>.
+
 use crate::{
     ast::Rule,
     common::LineEnd,
diff --git a/ast/src/console/mod.rs b/ast/src/console/mod.rs
index 6743e500c5..bf4eb3b03e 100644
--- a/ast/src/console/mod.rs
+++ b/ast/src/console/mod.rs
@@ -1,3 +1,19 @@
+// Copyright (C) 2019-2020 Aleo Systems Inc.
+// This file is part of the Leo library.
+
+// The Leo library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// The Leo library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with the Leo library. If not, see <https://www.gnu.org/licenses/>.
+
 pub mod console_assert;
 pub use console_assert::*;
 
diff --git a/ast/src/leo.pest b/ast/src/leo.pest
index 5af328b93b..163838bb5e 100644
--- a/ast/src/leo.pest
+++ b/ast/src/leo.pest
@@ -311,7 +311,7 @@ circuit = { "circuit " ~ identifier ~ "{" ~ NEWLINE* ~ circuit_member* ~ NEWLINE
 circuit_variable = { identifier ~ ":" ~ expression }
 
 // Declared in circuits/circuit_variable_definition.rs
-circuit_variable_definition = { identifier ~ ":" ~ type_ ~ ","?}
+circuit_variable_definition = { mutable? ~ identifier ~ ":" ~ type_ ~ ","?}
 
 // Declared in circuits/circuit_function.rs
 circuit_function = { static_? ~ function }
diff --git a/compiler/src/console/assert.rs b/compiler/src/console/assert.rs
index 0c7e00b2d8..e105ea878a 100644
--- a/compiler/src/console/assert.rs
+++ b/compiler/src/console/assert.rs
@@ -1,3 +1,19 @@
+// Copyright (C) 2019-2020 Aleo Systems Inc.
+// This file is part of the Leo library.
+
+// The Leo library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// The Leo library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with the Leo library. If not, see <https://www.gnu.org/licenses/>.
+
 //! Enforces an assert equals statement in a compiled Leo program.
 
 use crate::{errors::ConsoleError, program::ConstrainedProgram, value::ConstrainedValue, GroupType};
diff --git a/compiler/src/console/console.rs b/compiler/src/console/console.rs
index 637dc2bd37..3b96a101b5 100644
--- a/compiler/src/console/console.rs
+++ b/compiler/src/console/console.rs
@@ -1,3 +1,19 @@
+// Copyright (C) 2019-2020 Aleo Systems Inc.
+// This file is part of the Leo library.
+
+// The Leo library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// The Leo library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with the Leo library. If not, see <https://www.gnu.org/licenses/>.
+
 //! Evaluates a macro in a compiled Leo program.
 
 use crate::{errors::ConsoleError, program::ConstrainedProgram, GroupType};
diff --git a/compiler/src/console/mod.rs b/compiler/src/console/mod.rs
index 3ed19d6067..6e3be58884 100644
--- a/compiler/src/console/mod.rs
+++ b/compiler/src/console/mod.rs
@@ -1,3 +1,19 @@
+// Copyright (C) 2019-2020 Aleo Systems Inc.
+// This file is part of the Leo library.
+
+// The Leo library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// The Leo library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with the Leo library. If not, see <https://www.gnu.org/licenses/>.
+
 pub mod assert;
 pub use assert::*;
 
diff --git a/compiler/src/errors/console.rs b/compiler/src/errors/console.rs
index 87a4d03970..95117d8243 100644
--- a/compiler/src/errors/console.rs
+++ b/compiler/src/errors/console.rs
@@ -1,3 +1,19 @@
+// Copyright (C) 2019-2020 Aleo Systems Inc.
+// This file is part of the Leo library.
+
+// The Leo library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// The Leo library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with the Leo library. If not, see <https://www.gnu.org/licenses/>.
+
 use crate::errors::ExpressionError;
 use leo_typed::{Error as FormattedError, Span};
 
diff --git a/compiler/src/errors/statement.rs b/compiler/src/errors/statement.rs
index 84ae11c969..2c4f483b97 100644
--- a/compiler/src/errors/statement.rs
+++ b/compiler/src/errors/statement.rs
@@ -96,6 +96,12 @@ impl StatementError {
         Self::new_from_span(message, span)
     }
 
+    pub fn immutable_circuit_variable(name: String, span: Span) -> Self {
+        let message = format!("Circuit member variable `{}` is immutable", name);
+
+        Self::new_from_span(message, span)
+    }
+
     pub fn indicator_calculation(name: String, span: Span) -> Self {
         let message = format!(
             "Constraint system failed to evaluate branch selection indicator `{}`",
@@ -168,8 +174,8 @@ impl StatementError {
         Self::new_from_span(message, span)
     }
 
-    pub fn undefined_circuit_object(name: String, span: Span) -> Self {
-        let message = format!("Attempted to assign to unknown circuit object `{}`", name);
+    pub fn undefined_circuit_variable(name: String, span: Span) -> Self {
+        let message = format!("Attempted to assign to unknown circuit member variable `{}`", name);
 
         Self::new_from_span(message, span)
     }
diff --git a/compiler/src/expression/circuit/access.rs b/compiler/src/expression/circuit/access.rs
index 716ff73dba..7ad1eb5653 100644
--- a/compiler/src/expression/circuit/access.rs
+++ b/compiler/src/expression/circuit/access.rs
@@ -77,9 +77,9 @@ impl<F: Field + PrimeField, G: GroupType<F>> ConstrainedProgram<F, G> {
                         for stored_member in members {
                             let circuit_scope = new_scope(file_scope.clone(), circuit_name.to_string());
                             let self_keyword = new_scope(circuit_scope, SELF_KEYWORD.to_string());
-                            let field = new_scope(self_keyword, stored_member.0.to_string());
+                            let variable = new_scope(self_keyword, stored_member.0.to_string());
 
-                            self.store(field, stored_member.1.clone());
+                            self.store(variable, stored_member.1.clone());
                         }
                     }
                     ConstrainedValue::Static(value) => {
diff --git a/compiler/src/expression/circuit/circuit.rs b/compiler/src/expression/circuit/circuit.rs
index b27196e49c..b180428ae4 100644
--- a/compiler/src/expression/circuit/circuit.rs
+++ b/compiler/src/expression/circuit/circuit.rs
@@ -55,22 +55,27 @@ impl<F: Field + PrimeField, G: GroupType<F>> ConstrainedProgram<F, G> {
 
         for member in circuit.members.clone().into_iter() {
             match member {
-                CircuitMember::CircuitVariable(identifier, _type) => {
+                CircuitMember::CircuitVariable(is_mutable, identifier, type_) => {
                     let matched_variable = members
                         .clone()
                         .into_iter()
                         .find(|variable| variable.identifier.eq(&identifier));
                     match matched_variable {
                         Some(variable) => {
-                            // Resolve and enforce circuit object
-                            let variable_value = self.enforce_expression(
+                            // Resolve and enforce circuit variable
+                            let mut variable_value = self.enforce_expression(
                                 cs,
                                 file_scope.clone(),
                                 function_scope.clone(),
-                                Some(_type.clone()),
+                                Some(type_.clone()),
                                 variable.expression,
                             )?;
 
+                            // Add mutability to circuit variable
+                            if is_mutable {
+                                variable_value = ConstrainedValue::Mutable(Box::new(variable_value))
+                            }
+
                             resolved_members.push(ConstrainedCircuitMember(identifier, variable_value))
                         }
                         None => return Err(ExpressionError::expected_circuit_member(identifier.to_string(), span)),
diff --git a/compiler/src/statement/assign/circuit_variable.rs b/compiler/src/statement/assign/circuit_variable.rs
index 82b08f0d31..8848261e39 100644
--- a/compiler/src/statement/assign/circuit_variable.rs
+++ b/compiler/src/statement/assign/circuit_variable.rs
@@ -33,19 +33,20 @@ impl<F: Field + PrimeField, G: GroupType<F>> ConstrainedProgram<F, G> {
         cs: &mut CS,
         indicator: Option<Boolean>,
         circuit_name: String,
-        object_name: Identifier,
+        variable_name: Identifier,
         mut new_value: ConstrainedValue<F, G>,
         span: Span,
     ) -> Result<(), StatementError> {
         let condition = indicator.unwrap_or(Boolean::Constant(true));
 
+        // Get the mutable circuit by name
         match self.get_mutable_assignee(circuit_name, span.clone())? {
             ConstrainedValue::CircuitExpression(_variable, members) => {
                 // Modify the circuit variable in place
-                let matched_variable = members.into_iter().find(|object| object.0 == object_name);
+                let matched_variable = members.into_iter().find(|member| member.0 == variable_name);
 
                 match matched_variable {
-                    Some(object) => match &object.1 {
+                    Some(member) => match &member.1 {
                         ConstrainedValue::Function(_circuit_identifier, function) => {
                             return Err(StatementError::immutable_circuit_function(
                                 function.identifier.to_string(),
@@ -55,27 +56,35 @@ impl<F: Field + PrimeField, G: GroupType<F>> ConstrainedProgram<F, G> {
                         ConstrainedValue::Static(_value) => {
                             return Err(StatementError::immutable_circuit_function("static".into(), span));
                         }
-                        _ => {
-                            new_value.resolve_type(Some(object.1.to_type(span.clone())?), span.clone())?;
+                        ConstrainedValue::Mutable(value) => {
+                            new_value.resolve_type(Some(value.to_type(span.clone())?), span.clone())?;
 
                             let name_unique = format!("select {} {}:{}", new_value, span.line, span.start);
                             let selected_value = ConstrainedValue::conditionally_select(
                                 cs.ns(|| name_unique),
                                 &condition,
                                 &new_value,
-                                &object.1,
+                                &member.1,
                             )
                             .map_err(|_| {
-                                StatementError::select_fail(new_value.to_string(), object.1.to_string(), span)
+                                StatementError::select_fail(new_value.to_string(), member.1.to_string(), span)
                             })?;
 
-                            object.1 = selected_value.to_owned();
+                            member.1 = selected_value.to_owned();
+                        }
+                        _ => {
+                            return Err(StatementError::immutable_circuit_variable(variable_name.name, span));
                         }
                     },
-                    None => return Err(StatementError::undefined_circuit_object(object_name.to_string(), span)),
+                    None => {
+                        return Err(StatementError::undefined_circuit_variable(
+                            variable_name.to_string(),
+                            span,
+                        ));
+                    }
                 }
             }
-            _ => return Err(StatementError::undefined_circuit(object_name.to_string(), span)),
+            _ => return Err(StatementError::undefined_circuit(variable_name.to_string(), span)),
         }
 
         Ok(())
diff --git a/compiler/src/value/value.rs b/compiler/src/value/value.rs
index b4452d4195..c229fa2e9c 100644
--- a/compiler/src/value/value.rs
+++ b/compiler/src/value/value.rs
@@ -132,6 +132,7 @@ impl<F: Field + PrimeField, G: GroupType<F>> ConstrainedValue<F, G> {
                 Type::Tuple(types)
             }
             ConstrainedValue::CircuitExpression(id, _members) => Type::Circuit(id.clone()),
+            ConstrainedValue::Mutable(value) => return value.to_type(span),
             value => return Err(ValueError::implicit(value.to_string(), span)),
         })
     }
diff --git a/compiler/tests/console/mod.rs b/compiler/tests/console/mod.rs
index d2773c7660..2716f815a5 100644
--- a/compiler/tests/console/mod.rs
+++ b/compiler/tests/console/mod.rs
@@ -1,3 +1,19 @@
+// Copyright (C) 2019-2020 Aleo Systems Inc.
+// This file is part of the Leo library.
+
+// The Leo library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// The Leo library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with the Leo library. If not, see <https://www.gnu.org/licenses/>.
+
 use crate::{assert_satisfied, expect_compiler_error, generate_main_input, parse_program};
 use leo_typed::InputValue;
 
diff --git a/compiler/tests/mutability/circuit_function_mut.leo b/compiler/tests/mutability/circuit_function_mut.leo
new file mode 100644
index 0000000000..eba1d02c0e
--- /dev/null
+++ b/compiler/tests/mutability/circuit_function_mut.leo
@@ -0,0 +1,9 @@
+// Adding the `mut` keyword makes a circuit variable mutable.
+circuit Foo {
+    function bar() {}
+}
+
+function main() {
+    let mut a = Foo { x: 1 };
+    a.bar = 0;
+}
\ No newline at end of file
diff --git a/compiler/tests/mutability/circuit_static_function_mut.leo b/compiler/tests/mutability/circuit_static_function_mut.leo
new file mode 100644
index 0000000000..0717f51c44
--- /dev/null
+++ b/compiler/tests/mutability/circuit_static_function_mut.leo
@@ -0,0 +1,9 @@
+// Adding the `mut` keyword makes a circuit variable mutable.
+circuit Foo {
+    static function bar() {}
+}
+
+function main() {
+    let mut a = Foo { x: 1 };
+    a.bar = 0;
+}
\ No newline at end of file
diff --git a/compiler/tests/mutability/circuit_variable_mut.leo b/compiler/tests/mutability/circuit_variable_mut.leo
new file mode 100644
index 0000000000..f87668c48e
--- /dev/null
+++ b/compiler/tests/mutability/circuit_variable_mut.leo
@@ -0,0 +1,11 @@
+// Adding the `mut` keyword makes a circuit variable mutable.
+circuit Foo {
+    mut x: u32
+}
+
+function main() {
+    let mut a = Foo { x: 1 };
+    a.x = 0;
+
+    console.assert(a.x == 0u32);
+}
\ No newline at end of file
diff --git a/compiler/tests/mutability/mod.rs b/compiler/tests/mutability/mod.rs
index 3c00bab145..ea380447bf 100644
--- a/compiler/tests/mutability/mod.rs
+++ b/compiler/tests/mutability/mod.rs
@@ -78,9 +78,33 @@ fn test_circuit_mut() {
     let bytes = include_bytes!("circuit_mut.leo");
     let program = parse_program(bytes).unwrap();
 
+    expect_compiler_error(program);
+}
+
+#[test]
+fn test_circuit_variable_mut() {
+    let bytes = include_bytes!("circuit_variable_mut.leo");
+    let program = parse_program(bytes).unwrap();
+
     assert_satisfied(program);
 }
 
+#[test]
+fn test_circuit_function_mut() {
+    let bytes = include_bytes!("circuit_function_mut.leo");
+    let program = parse_program(bytes).unwrap();
+
+    expect_compiler_error(program);
+}
+
+#[test]
+fn test_circuit_static_function_mut() {
+    let bytes = include_bytes!("circuit_static_function_mut.leo");
+    let program = parse_program(bytes).unwrap();
+
+    expect_compiler_error(program);
+}
+
 #[test]
 fn test_function_input() {
     let bytes = include_bytes!("function_input.leo");
diff --git a/leo/cli.rs b/leo/cli.rs
index 42f9b4683b..acad5a2f58 100644
--- a/leo/cli.rs
+++ b/leo/cli.rs
@@ -14,7 +14,7 @@
 // You should have received a copy of the GNU General Public License
 // along with the Leo library. If not, see <https://www.gnu.org/licenses/>.
 
-use crate::{cli_types::*, errors::CLIError, logger};
+use crate::{cli_types::*, errors::CLIError, logger, updater::Updater};
 
 use clap::{App, AppSettings, Arg, ArgMatches, SubCommand};
 
@@ -90,6 +90,10 @@ pub trait CLI {
             false => logger::init_logger("leo", 1),
         }
 
+        if arguments.subcommand().0 != "update" {
+            Updater::print_cli();
+        }
+
         let options = Self::parse(arguments)?;
         let _output = Self::output(options)?;
         Ok(())
diff --git a/leo/commands/build.rs b/leo/commands/build.rs
index 429f3463a4..888003d8ce 100644
--- a/leo/commands/build.rs
+++ b/leo/commands/build.rs
@@ -25,7 +25,7 @@ use leo_package::{
     inputs::*,
     outputs::{ChecksumFile, CircuitFile, OutputsDirectory, OUTPUTS_DIRECTORY_NAME},
     root::Manifest,
-    source::{LibFile, MainFile, LIB_FILE_NAME, MAIN_FILE_NAME, SOURCE_DIRECTORY_NAME},
+    source::{LibraryFile, MainFile, LIBRARY_FILENAME, MAIN_FILENAME, SOURCE_DIRECTORY_NAME},
 };
 
 use snarkos_curves::{bls12_377::Bls12_377, edwards_bls12::Fq};
@@ -81,11 +81,11 @@ impl CLI for BuildCommand {
         let start = Instant::now();
 
         // Compile the package starting with the lib.leo file
-        if LibFile::exists_at(&package_path) {
+        if LibraryFile::exists_at(&package_path) {
             // Construct the path to the library file in the source directory
             let mut lib_file_path = package_path.clone();
             lib_file_path.push(SOURCE_DIRECTORY_NAME);
-            lib_file_path.push(LIB_FILE_NAME);
+            lib_file_path.push(LIBRARY_FILENAME);
 
             // Log compilation of library file to console
             tracing::info!("Compiling library... ({:?})", lib_file_path);
@@ -107,7 +107,7 @@ impl CLI for BuildCommand {
             // Construct the path to the main file in the source directory
             let mut main_file_path = package_path.clone();
             main_file_path.push(SOURCE_DIRECTORY_NAME);
-            main_file_path.push(MAIN_FILE_NAME);
+            main_file_path.push(MAIN_FILENAME);
 
             // Load the input file at `package_name.in`
             let input_string = InputFile::new(&package_name).read_from(&path)?;
diff --git a/leo/commands/deploy.rs b/leo/commands/deploy.rs
index 44f73b0ef9..a04b8d9c33 100644
--- a/leo/commands/deploy.rs
+++ b/leo/commands/deploy.rs
@@ -22,7 +22,7 @@ use crate::{
 };
 use leo_package::{
     root::Manifest,
-    source::{MAIN_FILE_NAME, SOURCE_DIRECTORY_NAME},
+    source::{MAIN_FILENAME, SOURCE_DIRECTORY_NAME},
 };
 
 use clap::ArgMatches;
@@ -67,7 +67,7 @@ impl CLI for DeployCommand {
             None => {
                 let mut main_file_path = path.clone();
                 main_file_path.push(SOURCE_DIRECTORY_NAME);
-                main_file_path.push(MAIN_FILE_NAME);
+                main_file_path.push(MAIN_FILENAME);
 
                 Err(CLIError::RunError(RunError::MainFileDoesNotExist(
                     main_file_path.into_os_string(),
diff --git a/leo/commands/init.rs b/leo/commands/init.rs
index c5b2f43791..8f38690aa5 100644
--- a/leo/commands/init.rs
+++ b/leo/commands/init.rs
@@ -19,11 +19,7 @@ use crate::{
     cli_types::*,
     errors::{CLIError, InitError},
 };
-use leo_package::{
-    inputs::*,
-    root::{Gitignore, Manifest, README},
-    source::{LibFile, MainFile, SourceDirectory},
-};
+use leo_package::LeoPackage;
 
 use clap::ArgMatches;
 use std::env::current_dir;
@@ -62,60 +58,12 @@ impl CLI for InitCommand {
             .to_string_lossy()
             .to_string();
 
-        // Verify the directory exists
+        // Verify the directory does not exist
         if !path.exists() {
             return Err(InitError::DirectoryDoesNotExist(path.as_os_str().to_owned()).into());
         }
 
-        // Verify a manifest file does not already exist
-        if Manifest::exists_at(&path) {
-            return Err(InitError::PackageAlreadyExists(path.as_os_str().to_owned()).into());
-        }
-
-        // Create the manifest file
-        Manifest::new(&package_name).write_to(&path)?;
-
-        // Create the .gitignore file
-        Gitignore::new().write_to(&path)?;
-
-        // Create the README.md file
-        README::new(&package_name).write_to(&path)?;
-
-        // Create the source directory
-        SourceDirectory::create(&path)?;
-
-        // Create a new library or binary file
-
-        if options {
-            // Verify the library file does not exist
-            if !LibFile::exists_at(&path) {
-                // Create the library file in the source directory
-                LibFile::new(&package_name).write_to(&path)?;
-            }
-        } else {
-            // Create the input directory
-            InputsDirectory::create(&path)?;
-
-            // Verify the input file does not exist
-            let input_file = InputFile::new(&package_name);
-            if !input_file.exists_at(&path) {
-                // Create the input file in the inputs directory
-                input_file.write_to(&path)?;
-            }
-
-            // Verify the state file does not exist
-            let state_file = StateFile::new(&package_name);
-            if !state_file.exists_at(&path) {
-                // Create the state file in the inputs directory
-                state_file.write_to(&path)?;
-            }
-
-            // Verify the main file does not exist
-            if !MainFile::exists_at(&path) {
-                // Create the main file in the source directory
-                MainFile::new(&package_name).write_to(&path)?;
-            }
-        }
+        LeoPackage::initialize(&package_name, options, &path)?;
 
         tracing::info!("Successfully initialized package \"{}\"\n", package_name);
 
diff --git a/leo/commands/lint.rs b/leo/commands/lint.rs
index 2c78f22cc1..4487f79b9f 100644
--- a/leo/commands/lint.rs
+++ b/leo/commands/lint.rs
@@ -22,7 +22,7 @@ use crate::{
 };
 use leo_package::{
     root::Manifest,
-    source::{MAIN_FILE_NAME, SOURCE_DIRECTORY_NAME},
+    source::{MAIN_FILENAME, SOURCE_DIRECTORY_NAME},
 };
 
 use clap::ArgMatches;
@@ -67,7 +67,7 @@ impl CLI for LintCommand {
             None => {
                 let mut main_file_path = path.clone();
                 main_file_path.push(SOURCE_DIRECTORY_NAME);
-                main_file_path.push(MAIN_FILE_NAME);
+                main_file_path.push(MAIN_FILENAME);
 
                 Err(CLIError::RunError(RunError::MainFileDoesNotExist(
                     main_file_path.into_os_string(),
diff --git a/leo/commands/new.rs b/leo/commands/new.rs
index 74e3d52251..d210ff6ebd 100644
--- a/leo/commands/new.rs
+++ b/leo/commands/new.rs
@@ -19,11 +19,7 @@ use crate::{
     cli_types::*,
     errors::{CLIError, NewError},
 };
-use leo_package::{
-    inputs::*,
-    root::{Gitignore, Manifest, README},
-    source::{LibFile, MainFile, SourceDirectory},
-};
+use leo_package::LeoPackage;
 
 use clap::ArgMatches;
 use std::{env::current_dir, fs};
@@ -89,35 +85,7 @@ impl CLI for NewCommand {
         fs::create_dir_all(&path)
             .map_err(|error| NewError::CreatingRootDirectory(path.as_os_str().to_owned(), error))?;
 
-        // Create the manifest file
-        Manifest::new(&package_name).write_to(&path)?;
-
-        // Create the .gitignore file
-        Gitignore::new().write_to(&path)?;
-
-        // Create the README.md file
-        README::new(&package_name).write_to(&path)?;
-
-        // Create the source directory
-        SourceDirectory::create(&path)?;
-
-        // Create a new library or binary file
-        if options.1 {
-            // Create the library file in the source directory
-            LibFile::new(&package_name).write_to(&path)?;
-        } else {
-            // Create the input directory
-            InputsDirectory::create(&path)?;
-
-            // Create the input file in the inputs directory
-            InputFile::new(&package_name).write_to(&path)?;
-
-            // Create the state file in the inputs directory
-            StateFile::new(&package_name).write_to(&path)?;
-
-            // Create the main file in the source directory
-            MainFile::new(&package_name).write_to(&path)?;
-        }
+        LeoPackage::initialize(&package_name, options.1, &path)?;
 
         tracing::info!("Successfully initialized package \"{}\"\n", package_name);
 
diff --git a/leo/commands/remove.rs b/leo/commands/remove.rs
index 46e1ac5687..4b26ba3e9e 100644
--- a/leo/commands/remove.rs
+++ b/leo/commands/remove.rs
@@ -14,37 +14,35 @@
 // You should have received a copy of the GNU General Public License
 // along with the Leo library. If not, see <https://www.gnu.org/licenses/>.
 
-use crate::{
-    cli::*,
-    cli_types::*,
-    commands::BuildCommand,
-    errors::{CLIError, RunError},
-};
-use leo_package::{
-    root::Manifest,
-    source::{MAIN_FILE_NAME, SOURCE_DIRECTORY_NAME},
-};
+use crate::{cli::*, cli_types::*, errors::CLIError};
+use leo_package::LeoPackage;
 
 use clap::ArgMatches;
-use std::{convert::TryFrom, env::current_dir};
+use std::env::current_dir;
 
 #[derive(Debug)]
 pub struct RemoveCommand;
 
 impl CLI for RemoveCommand {
-    type Options = ();
+    type Options = Option<String>;
     type Output = ();
 
-    const ABOUT: AboutType = "Uninstall a package from the current package (*)";
-    const ARGUMENTS: &'static [ArgumentType] = &[];
+    const ABOUT: AboutType = "Uninstall a package from the current package";
+    const ARGUMENTS: &'static [ArgumentType] = &[
+        // (name, description, required, index)
+        ("NAME", "Removes the package from the current directory", true, 1u64),
+    ];
     const FLAGS: &'static [FlagType] = &[];
     const NAME: NameType = "remove";
     const OPTIONS: &'static [OptionType] = &[];
     const SUBCOMMANDS: &'static [SubCommandType] = &[];
 
     #[cfg_attr(tarpaulin, skip)]
-    fn parse(_arguments: &ArgMatches) -> Result<Self::Options, CLIError> {
-        Ok(())
+    fn parse(arguments: &ArgMatches) -> Result<Self::Options, CLIError> {
+        Ok(match arguments.value_of("NAME") {
+            Some(name) => Some(name.to_string()),
+            None => unreachable!(),
+        })
     }
 
     #[cfg_attr(tarpaulin, skip)]
@@ -55,24 +53,11 @@ impl CLI for RemoveCommand {
 
         let path = current_dir()?;
 
-        match BuildCommand::output(options)? {
-            Some((_program, _checksum_differs)) => {
-                // Get the package name
-                let _package_name = Manifest::try_from(&path)?.get_package_name();
-
-                tracing::info!("Unimplemented - `leo remove`");
-
-                Ok(())
-            }
-            None => {
-                let mut main_file_path = path.clone();
-                main_file_path.push(SOURCE_DIRECTORY_NAME);
-                main_file_path.push(MAIN_FILE_NAME);
-
-                Err(CLIError::RunError(RunError::MainFileDoesNotExist(
-                    main_file_path.into_os_string(),
-                )))
-            }
+        if let Some(package_name) = options {
+            LeoPackage::remove_imported_package(&package_name, &path)?;
+            tracing::info!("Successfully removed package \"{}\"\n", package_name);
         }
+
+        Ok(())
     }
 }
diff --git a/leo/commands/setup.rs b/leo/commands/setup.rs
index 80247a79f4..c1afbd4127 100644
--- a/leo/commands/setup.rs
+++ b/leo/commands/setup.rs
@@ -24,7 +24,7 @@ use leo_compiler::{compiler::Compiler, group::targets::edwards_bls12::EdwardsGro
 use leo_package::{
     outputs::{ProvingKeyFile, VerificationKeyFile},
     root::Manifest,
-    source::{MAIN_FILE_NAME, SOURCE_DIRECTORY_NAME},
+    source::{MAIN_FILENAME, SOURCE_DIRECTORY_NAME},
 };
 
 use snarkos_algorithms::snark::groth16::{Groth16, Parameters, PreparedVerifyingKey, VerifyingKey};
@@ -148,7 +148,7 @@ impl CLI for SetupCommand {
             None => {
                 let mut main_file_path = path.clone();
                 main_file_path.push(SOURCE_DIRECTORY_NAME);
-                main_file_path.push(MAIN_FILE_NAME);
+                main_file_path.push(MAIN_FILENAME);
 
                 Err(CLIError::RunError(RunError::MainFileDoesNotExist(
                     main_file_path.into_os_string(),
diff --git a/leo/commands/test.rs b/leo/commands/test.rs
index a069b38d07..91e8dbe6dc 100644
--- a/leo/commands/test.rs
+++ b/leo/commands/test.rs
@@ -24,7 +24,7 @@ use leo_package::{
     inputs::*,
     outputs::{OutputsDirectory, OUTPUTS_DIRECTORY_NAME},
     root::Manifest,
-    source::{LibFile, MainFile, LIB_FILE_NAME, MAIN_FILE_NAME, SOURCE_DIRECTORY_NAME},
+    source::{LibraryFile, MainFile, LIBRARY_FILENAME, MAIN_FILENAME, SOURCE_DIRECTORY_NAME},
 };
 
 use snarkos_curves::edwards_bls12::Fq;
@@ -70,9 +70,9 @@ impl CLI for TestCommand {
 
         // Verify a main or library file exists
         if MainFile::exists_at(&package_path) {
-            file_path.push(MAIN_FILE_NAME);
-        } else if LibFile::exists_at(&package_path) {
-            file_path.push(LIB_FILE_NAME);
+            file_path.push(MAIN_FILENAME);
+        } else if LibraryFile::exists_at(&package_path) {
+            file_path.push(LIBRARY_FILENAME);
         } else {
             return Err(ProgramFileDoesNotExist(package_path.into()).into());
         }
diff --git a/leo/commands/update.rs b/leo/commands/update.rs
index 0258ba81d7..27381f30c8 100644
--- a/leo/commands/update.rs
+++ b/leo/commands/update.rs
@@ -14,48 +14,11 @@
 // You should have received a copy of the GNU General Public License
 // along with the Leo library. If not, see <https://www.gnu.org/licenses/>.
 
-use crate::{cli::CLI, cli_types::*};
-
-use self_update::{backends::github, cargo_crate_version, Status};
-
-const LEO_BIN_NAME: &str = "leo";
-const LEO_REPO_OWNER: &str = "AleoHQ";
-const LEO_REPO_NAME: &str = "leo";
+use crate::{cli::CLI, cli_types::*, updater::Updater};
 
 #[derive(Debug)]
 pub struct UpdateCommand;
 
-impl UpdateCommand {
-    /// Show all available releases for `leo`
-    pub fn show_available_releases() -> Result<(), self_update::errors::Error> {
-        let releases = github::ReleaseList::configure()
-            .repo_owner(LEO_REPO_OWNER)
-            .repo_name(LEO_REPO_NAME)
-            .build()?
-            .fetch()?;
-
-        tracing::info!("List of available Leo's versions");
-        for release in releases {
-            tracing::info!("* {}", release.version);
-        }
-        Ok(())
-    }
-
-    /// Update `leo` to the latest release
-    pub fn update_to_latest_release() -> Result<Status, self_update::errors::Error> {
-        let status = github::Update::configure()
-            .repo_owner(LEO_REPO_OWNER)
-            .repo_name(LEO_REPO_NAME)
-            .bin_name(LEO_BIN_NAME)
-            .show_download_progress(true)
-            .current_version(cargo_crate_version!())
-            .build()?
-            .update()?;
-
-        Ok(status)
-    }
-}
-
 impl CLI for UpdateCommand {
     type Options = (bool,);
     type Output = ();
@@ -78,14 +41,14 @@ impl CLI for UpdateCommand {
         let _enter = span.enter();
 
         match options {
-            (true,) => match UpdateCommand::show_available_releases() {
+            (true,) => match Updater::show_available_releases() {
                 Ok(_) => return Ok(()),
                 Err(e) => {
                     tracing::error!("Could not fetch that latest version of Leo");
                     tracing::error!("{}", e);
                 }
             },
-            (false,) => match UpdateCommand::update_to_latest_release() {
+            (false,) => match Updater::update_to_latest_release(true) {
                 Ok(status) => {
                     if status.uptodate() {
                         tracing::info!("Leo is already on the latest version: {}", status.version());
diff --git a/leo/config.rs b/leo/config.rs
index 48a041eba2..ff15ff4725 100644
--- a/leo/config.rs
+++ b/leo/config.rs
@@ -56,7 +56,7 @@ pub struct Config {
 
 impl Default for Config {
     fn default() -> Self {
-        Self { auto_update: true }
+        Self { auto_update: false }
     }
 }
 
diff --git a/leo/errors/cli.rs b/leo/errors/cli.rs
index ca160a609f..59cd13599f 100644
--- a/leo/errors/cli.rs
+++ b/leo/errors/cli.rs
@@ -54,7 +54,7 @@ pub enum CLIError {
     InputFileError(InputFileError),
 
     #[error("{}", _0)]
-    LibFileError(LibFileError),
+    LibraryFileError(LibraryFileError),
 
     #[error("{}", _0)]
     LoginError(LoginError),
@@ -74,6 +74,9 @@ pub enum CLIError {
     #[error("{}", _0)]
     OutputsDirectoryError(OutputsDirectoryError),
 
+    #[error("{}", _0)]
+    PackageError(PackageError),
+
     #[error("{}", _0)]
     ProofFileError(ProofFileError),
 
@@ -133,13 +136,14 @@ impl_cli_error!(
     InitError,
     InputsDirectoryError,
     InputFileError,
-    LibFileError,
+    LibraryFileError,
     LoginError,
     MainFileError,
     ManifestError,
     NewError,
     OutputFileError,
     OutputsDirectoryError,
+    PackageError,
     ProofFileError,
     ProvingKeyFileError,
     PublishError,
@@ -152,17 +156,24 @@ impl_cli_error!(
     ZipFileError
 );
 
+impl From<clap::Error> for CLIError {
+    fn from(error: clap::Error) -> Self {
+        tracing::error!("{}\n", error);
+        CLIError::Crate("clap", format!("{}", error))
+    }
+}
+
 impl From<leo_compiler::errors::CompilerError> for CLIError {
     fn from(error: leo_compiler::errors::CompilerError) -> Self {
         tracing::error!("{}\n", error);
-        CLIError::Crate("leo_compiler", "Program failed due to previous error".into())
+        CLIError::Crate("leo-compiler", "Program failed due to previous error".into())
     }
 }
 
 impl From<leo_input::errors::InputParserError> for CLIError {
     fn from(error: leo_input::errors::InputParserError) -> Self {
         tracing::error!("{}\n", error);
-        CLIError::Crate("leo_input", "Program failed due to previous error".into())
+        CLIError::Crate("leo-input", "Program failed due to previous error".into())
     }
 }
 
diff --git a/leo/errors/commands/init.rs b/leo/errors/commands/init.rs
index 95552cb02e..ff9042444e 100644
--- a/leo/errors/commands/init.rs
+++ b/leo/errors/commands/init.rs
@@ -29,9 +29,6 @@ pub enum InitError {
     #[error("{}", _0)]
     ManifestError(#[from] ManifestError),
 
-    #[error("package at path {:?} already exists", _0)]
-    PackageAlreadyExists(OsString),
-
     #[error("package name is missing - {:?}", _0)]
     ProjectNameInvalid(OsString),
 }
diff --git a/leo/errors/commands/new.rs b/leo/errors/commands/new.rs
index fd816f75db..ffa4586c4d 100644
--- a/leo/errors/commands/new.rs
+++ b/leo/errors/commands/new.rs
@@ -29,9 +29,6 @@ pub enum NewError {
     #[error("{}", _0)]
     ManifestError(#[from] ManifestError),
 
-    #[error("package at path {:?} already exists", _0)]
-    PackageAlreadyExists(OsString),
-
     #[error("package name is missing - {:?}", _0)]
     ProjectNameInvalid(OsString),
 }
diff --git a/leo/lib.rs b/leo/lib.rs
index 5de76c348c..1fa729229c 100644
--- a/leo/lib.rs
+++ b/leo/lib.rs
@@ -25,3 +25,4 @@ pub mod config;
 pub mod errors;
 pub mod logger;
 pub mod synthesizer;
+pub mod updater;
diff --git a/leo/main.rs b/leo/main.rs
index ae9ea633ec..7c8c124904 100644
--- a/leo/main.rs
+++ b/leo/main.rs
@@ -14,13 +14,13 @@
 // You should have received a copy of the GNU General Public License
 // along with the Leo library. If not, see <https://www.gnu.org/licenses/>.
 
-use leo_lang::{cli::*, commands::*, config::Config, errors::CLIError};
+use leo_lang::{cli::*, commands::*, errors::CLIError, logger, updater::Updater};
 
 use clap::{App, AppSettings, Arg};
 
 #[cfg_attr(tarpaulin, skip)]
 fn main() -> Result<(), CLIError> {
-    let arguments = App::new("leo")
+    let app = App::new("leo")
         .version(include_str!("./leo-version"))
         .about("Leo compiler and package manager")
         .author("The Aleo Team <hello@aleo.org>")
@@ -28,7 +28,6 @@ fn main() -> Result<(), CLIError> {
             AppSettings::ColoredHelp,
             AppSettings::DisableHelpSubcommand,
             AppSettings::DisableVersion,
-            AppSettings::SubcommandRequiredElseHelp,
         ])
         .args(&[Arg::with_name("debug")
             .short("d")
@@ -53,18 +52,10 @@ fn main() -> Result<(), CLIError> {
             LintCommand::new().display_order(14),
             UpdateCommand::new().display_order(15),
         ])
-        .set_term_width(0)
-        .get_matches();
+        .set_term_width(0);
 
-    let config = Config::read_config()?;
-
-    if config.auto_update {
-        if let Ok(status) = UpdateCommand::update_to_latest_release() {
-            if status.updated() {
-                tracing::info!("Leo has successfully updated to version: {}", status.version());
-            }
-        }
-    }
+    let mut help = app.clone();
+    let arguments = app.get_matches();
 
     match arguments.subcommand() {
         ("new", Some(arguments)) => NewCommand::process(arguments),
@@ -83,6 +74,17 @@ fn main() -> Result<(), CLIError> {
         ("clean", Some(arguments)) => CleanCommand::process(arguments),
         ("lint", Some(arguments)) => LintCommand::process(arguments),
         ("update", Some(arguments)) => UpdateCommand::process(arguments),
-        _ => unreachable!(),
+        _ => {
+            // Set logging environment
+            match arguments.is_present("debug") {
+                true => logger::init_logger("leo", 2),
+                false => logger::init_logger("leo", 1),
+            }
+
+            Updater::print_cli();
+
+            help.print_help()?;
+            Ok(())
+        }
     }
 }
diff --git a/leo/updater.rs b/leo/updater.rs
new file mode 100644
index 0000000000..5fdb7b01b0
--- /dev/null
+++ b/leo/updater.rs
@@ -0,0 +1,86 @@
+use crate::config::Config;
+
+use colored::Colorize;
+use self_update::{backends::github, version::bump_is_greater, Status};
+
+pub struct Updater;
+
+// TODO Add logic for users to easily select release versions.
+impl Updater {
+    const LEO_BIN_NAME: &'static str = "leo";
+    const LEO_REPO_NAME: &'static str = "leo";
+    const LEO_REPO_OWNER: &'static str = "AleoHQ";
+
+    /// Show all available releases for `leo`.
+    pub fn show_available_releases() -> Result<(), self_update::errors::Error> {
+        let releases = github::ReleaseList::configure()
+            .repo_owner(Self::LEO_REPO_OWNER)
+            .repo_name(Self::LEO_REPO_NAME)
+            .build()?
+            .fetch()?;
+
+        tracing::info!("List of available Leo's versions");
+        for release in releases {
+            tracing::info!("* {}", release.version);
+        }
+        Ok(())
+    }
+
+    /// Update `leo` to the latest release.
+    pub fn update_to_latest_release(show_output: bool) -> Result<Status, self_update::errors::Error> {
+        let status = github::Update::configure()
+            .repo_owner(Self::LEO_REPO_OWNER)
+            .repo_name(Self::LEO_REPO_NAME)
+            .bin_name(Self::LEO_BIN_NAME)
+            .current_version(&include_str!("./leo-version").replace('v', ""))
+            .show_download_progress(true)
+            .no_confirm(true)
+            .show_output(show_output)
+            .build()?
+            .update()?;
+
+        Ok(status)
+    }
+
+    /// Check if there is an available update for `leo` and return the newest release.
+    pub fn update_available() -> Result<Option<String>, self_update::errors::Error> {
+        let updater = github::Update::configure()
+            .repo_owner(Self::LEO_REPO_OWNER)
+            .repo_name(Self::LEO_REPO_NAME)
+            .bin_name(Self::LEO_BIN_NAME)
+            .current_version(&include_str!("./leo-version").replace('v', ""))
+            .build()?;
+
+        let current_version = updater.current_version();
+        let latest_release = updater.get_latest_release()?;
+
+        if bump_is_greater(&current_version, &latest_release.version)? {
+            Ok(Some(latest_release.version))
+        } else {
+            Ok(None)
+        }
+    }
+
+    /// Display the CLI message, if the Leo configuration allows.
+    pub fn print_cli() {
+        let config = Config::read_config().unwrap();
+
+        if config.auto_update {
+            // If the auto update configuration is on, attempt to update the version.
+            if let Ok(status) = Self::update_to_latest_release(false) {
+                if status.updated() {
+                    tracing::info!("Successfully updated to {}", status.version());
+                }
+            }
+        } else {
+            // If the auto update configuration is off, notify the user to update leo.
+            if let Some(latest_version) = Self::update_available().unwrap() {
+                let mut message = format!("{}", "🟢 A new version is available! Run".bold().green());
+                message += &format!("{}", " `leo update` ".bold().white());
+                message += &format!("{}", &(format!("to update to v{}.", latest_version)).bold().green());
+
+                tracing::info!("\n{}\n", message);
+            }
+        }
+    }
+}
diff --git a/package/Cargo.toml b/package/Cargo.toml
index 14afff3175..3bb2a74878 100644
--- a/package/Cargo.toml
+++ b/package/Cargo.toml
@@ -38,3 +38,6 @@ version = "2"
 
 [dependencies.zip]
 version = "0.5"
+
+[dev-dependencies.lazy_static]
+version = "1.3.0"
\ No newline at end of file
diff --git a/package/src/errors/imports/directory.rs b/package/src/errors/imports/directory.rs
index 8d1f8ded19..e36d3740dd 100644
--- a/package/src/errors/imports/directory.rs
+++ b/package/src/errors/imports/directory.rs
@@ -30,6 +30,9 @@ pub enum ImportsDirectoryError {
     #[error("file {:?} type getting: {}", _0, _1)]
     GettingFileType(OsString, io::Error),
 
+    #[error("package {:?} does not exist as an import", _0)]
+    ImportDoesNotExist(OsString),
+
     #[error("invalid file {:?} extension: {:?}", _0, _1)]
     InvalidFileExtension(OsString, OsString),
 
diff --git a/package/src/errors/mod.rs b/package/src/errors/mod.rs
index 84b6620f73..7f09f85707 100644
--- a/package/src/errors/mod.rs
+++ b/package/src/errors/mod.rs
@@ -23,6 +23,9 @@ pub use inputs::*;
 pub mod outputs;
 pub use outputs::*;
 
+pub mod package;
+pub use self::package::*;
+
 pub mod root;
 pub use self::root::*;
 
diff --git a/package/src/errors/package.rs b/package/src/errors/package.rs
new file mode 100644
index 0000000000..b1a86d6419
--- /dev/null
+++ b/package/src/errors/package.rs
@@ -0,0 +1,88 @@
+use std::{ffi::OsString, io};
+
+#[derive(Debug, Error)]
+pub enum PackageError {
+    #[error("{}: {}", _0, _1)]
+    Crate(&'static str, String),
+
+    #[error("`{}` creating: {}", _0, _1)]
+    Creating(&'static str, io::Error),
+
+    #[error("Failed to initialize package {:?} ({:?})", _0, _1)]
+    FailedToInitialize(String, OsString),
+
+    #[error("`{}` metadata: {}", _0, _1)]
+    Removing(&'static str, io::Error),
+}
+
+impl From<std::io::Error> for PackageError {
+    fn from(error: std::io::Error) -> Self {
+        PackageError::Crate("std::io", format!("{}", error))
+    }
+}
+
+impl From<crate::errors::GitignoreError> for PackageError {
+    fn from(error: crate::errors::GitignoreError) -> Self {
+        PackageError::Crate("leo-package", format!("{}", error))
+    }
+}
+
+impl From<crate::errors::InputFileError> for PackageError {
+    fn from(error: crate::errors::InputFileError) -> Self {
+        PackageError::Crate("leo-package", format!("{}", error))
+    }
+}
+
+impl From<crate::errors::InputsDirectoryError> for PackageError {
+    fn from(error: crate::errors::InputsDirectoryError) -> Self {
+        PackageError::Crate("leo-package", format!("{}", error))
+    }
+}
+
+impl From<crate::errors::ImportsDirectoryError> for PackageError {
+    fn from(error: crate::errors::ImportsDirectoryError) -> Self {
+        PackageError::Crate("leo-package", format!("{}", error))
+    }
+}
+
+impl From<crate::errors::OutputsDirectoryError> for PackageError {
+    fn from(error: crate::errors::OutputsDirectoryError) -> Self {
+        PackageError::Crate("leo-package", format!("{}", error))
+    }
+}
+
+impl From<crate::errors::READMEError> for PackageError {
+    fn from(error: crate::errors::READMEError) -> Self {
+        PackageError::Crate("leo-package", format!("{}", error))
+    }
+}
+
+impl From<crate::errors::SourceDirectoryError> for PackageError {
+    fn from(error: crate::errors::SourceDirectoryError) -> Self {
+        PackageError::Crate("leo-package", format!("{}", error))
+    }
+}
+
+impl From<crate::errors::StateFileError> for PackageError {
+    fn from(error: crate::errors::StateFileError) -> Self {
+        PackageError::Crate("leo-package", format!("{}", error))
+    }
+}
+
+impl From<crate::errors::LibraryFileError> for PackageError {
+    fn from(error: crate::errors::LibraryFileError) -> Self {
+        PackageError::Crate("leo-package", format!("{}", error))
+    }
+}
+
+impl From<crate::errors::ManifestError> for PackageError {
+    fn from(error: crate::errors::ManifestError) -> Self {
+        PackageError::Crate("leo-package", format!("{}", error))
+    }
+}
+
+impl From<crate::errors::MainFileError> for PackageError {
+    fn from(error: crate::errors::MainFileError) -> Self {
+        PackageError::Crate("leo-package", format!("{}", error))
+    }
+}
diff --git a/package/src/errors/source/lib.rs b/package/src/errors/source/lib.rs
index 2560ae4d8e..ee76aa2646 100644
--- a/package/src/errors/source/lib.rs
+++ b/package/src/errors/source/lib.rs
@@ -17,7 +17,7 @@
 use std::io;
 
 #[derive(Debug, Error)]
-pub enum LibFileError {
+pub enum LibraryFileError {
     #[error("{}: {}", _0, _1)]
     Crate(&'static str, String),
 
@@ -28,8 +28,8 @@ pub enum LibFileError {
     Writing(io::Error),
 }
 
-impl From<std::io::Error> for LibFileError {
+impl From<std::io::Error> for LibraryFileError {
     fn from(error: std::io::Error) -> Self {
-        LibFileError::Crate("std::io", format!("{}", error))
+        LibraryFileError::Crate("std::io", format!("{}", error))
     }
 }
diff --git a/package/src/imports/directory.rs b/package/src/imports/directory.rs
index d9f5be5868..0a84701c29 100644
--- a/package/src/imports/directory.rs
+++ b/package/src/imports/directory.rs
@@ -46,4 +46,22 @@ impl ImportsDirectory {
 
         Ok(())
     }
+
+    /// Removes an imported package in the imports directory at the provided path.
+    pub fn remove_import(path: &PathBuf, package_name: &str) -> Result<(), ImportsDirectoryError> {
+        let mut path = path.to_owned();
+        if path.is_dir() && !path.ends_with(IMPORTS_DIRECTORY_NAME) {
+            path.push(PathBuf::from(IMPORTS_DIRECTORY_NAME));
+        }
+
+        path.push(PathBuf::from(package_name));
+
+        if !path.exists() || !path.is_dir() {
+            return Err(ImportsDirectoryError::ImportDoesNotExist(package_name.into()));
+        }
+
+        fs::remove_dir_all(&path).map_err(ImportsDirectoryError::Removing)?;
+
+        Ok(())
+    }
 }
diff --git a/package/src/inputs/input.rs b/package/src/inputs/input.rs
index 06415b0e51..c7c91266ae 100644
--- a/package/src/inputs/input.rs
+++ b/package/src/inputs/input.rs
@@ -39,6 +39,10 @@ impl InputFile {
         }
     }
 
+    pub fn filename(&self) -> String {
+        format!("{}{}{}", INPUTS_DIRECTORY_NAME, self.package_name, INPUT_FILE_EXTENSION)
+    }
+
     pub fn exists_at(&self, path: &PathBuf) -> bool {
         let path = self.setup_file_path(path);
         path.exists()
diff --git a/package/src/inputs/state.rs b/package/src/inputs/state.rs
index df1a1fed17..a8c2043f5a 100644
--- a/package/src/inputs/state.rs
+++ b/package/src/inputs/state.rs
@@ -39,6 +39,10 @@ impl StateFile {
         }
     }
 
+    pub fn filename(&self) -> String {
+        format!("{}{}{}", INPUTS_DIRECTORY_NAME, self.package_name, STATE_FILE_EXTENSION)
+    }
+
     pub fn exists_at(&self, path: &PathBuf) -> bool {
         let path = self.setup_file_path(path);
         path.exists()
diff --git a/package/src/lib.rs b/package/src/lib.rs
index 01b7a4d290..0a7d6bc37d 100644
--- a/package/src/lib.rs
+++ b/package/src/lib.rs
@@ -23,5 +23,22 @@ pub use errors::*;
 pub mod imports;
 pub mod inputs;
 pub mod outputs;
+pub mod package;
 pub mod root;
 pub mod source;
+
+use std::path::PathBuf;
+
+pub struct LeoPackage;
+
+impl LeoPackage {
+    /// Initializes a Leo package at the given path.
+    pub fn initialize(package_name: &str, is_lib: bool, path: &PathBuf) -> Result<(), PackageError> {
+        package::Package::initialize(package_name, is_lib, path)
+    }
+
+    /// Removes an imported Leo package
+    pub fn remove_imported_package(package_name: &str, path: &PathBuf) -> Result<(), PackageError> {
+        package::Package::remove_imported_package(package_name, path)
+    }
+}
diff --git a/package/src/package.rs b/package/src/package.rs
new file mode 100644
index 0000000000..d5b1b7609f
--- /dev/null
+++ b/package/src/package.rs
@@ -0,0 +1,190 @@
+// Copyright (C) 2019-2020 Aleo Systems Inc.
+// This file is part of the Leo library.
+
+// The Leo library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// The Leo library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with the Leo library. If not, see <https://www.gnu.org/licenses/>.
+
+use crate::{
+    errors::PackageError,
+    imports::ImportsDirectory,
+    inputs::{InputFile, InputsDirectory, StateFile},
+    root::{Gitignore, Manifest, README},
+    source::{LibraryFile, MainFile, SourceDirectory},
+};
+
+use serde::Deserialize;
+use std::path::PathBuf;
+
+#[derive(Deserialize)]
+pub struct Package {
+    pub name: String,
+    pub version: String,
+    pub description: Option<String>,
+    pub license: Option<String>,
+}
+
+impl Package {
+    pub fn new(package_name: &str) -> Self {
+        Self {
+            name: package_name.to_owned(),
+            version: "0.1.0".to_owned(),
+            description: None,
+            license: None,
+        }
+    }
+
+    /// Returns `true` if a package is can be initialized at a given path.
+    pub fn can_initialize(package_name: &str, is_lib: bool, path: &PathBuf) -> bool {
+        let mut result = true;
+        let mut existing_files = vec![];
+
+        // Check if the manifest file already exists.
+        if Manifest::exists_at(&path) {
+            existing_files.push(Manifest::filename());
+            result = false;
+        }
+
+        if is_lib {
+            // Check if the library file already exists.
+            if LibraryFile::exists_at(&path) {
+                existing_files.push(LibraryFile::filename());
+                result = false;
+            }
+        } else {
+            // Check if the input file already exists.
+            let input_file = InputFile::new(&package_name);
+            if input_file.exists_at(&path) {
+                existing_files.push(input_file.filename());
+                result = false;
+            }
+
+            // Check if the state file already exists.
+            let state_file = StateFile::new(&package_name);
+            if state_file.exists_at(&path) {
+                existing_files.push(state_file.filename());
+                result = false;
+            }
+
+            // Check if the main file already exists.
+            if MainFile::exists_at(&path) {
+                existing_files.push(MainFile::filename());
+                result = false;
+            }
+        }
+
+        if existing_files.len() > 0 {
+            tracing::error!("File(s) {:?} already exist", existing_files);
+        }
+
+        return result;
+    }
+
+    /// Returns `true` if a package is initialized at the given path
+    pub fn is_initialized(package_name: &str, is_lib: bool, path: &PathBuf) -> bool {
+        // Check if the manifest file exists.
+        if !Manifest::exists_at(&path) {
+            return false;
+        }
+
+        if is_lib {
+            // Check if the library file exists.
+            if !LibraryFile::exists_at(&path) {
+                return false;
+            }
+        } else {
+            // Check if the input file exists.
+            let input_file = InputFile::new(&package_name);
+            if !input_file.exists_at(&path) {
+                return false;
+            }
+
+            // Check if the state file exists.
+            let state_file = StateFile::new(&package_name);
+            if !state_file.exists_at(&path) {
+                return false;
+            }
+
+            // Check if the main file exists.
+            if !MainFile::exists_at(&path) {
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    /// Creates a package at the given path
+    pub fn initialize(package_name: &str, is_lib: bool, path: &PathBuf) -> Result<(), PackageError> {
+        // First, verify that this directory is not already initialized as a Leo package.
+        {
+            if !Self::can_initialize(package_name, is_lib, path) {
+                return Err(
+                    PackageError::FailedToInitialize(package_name.to_owned(), path.as_os_str().to_owned()).into(),
+                );
+            }
+        }
+        // Next, initialize this directory as a Leo package.
+        {
+            // Create the manifest file.
+            Manifest::new(&package_name).write_to(&path)?;
+
+            // Verify that the .gitignore file does not exist.
+            if !Gitignore::exists_at(&path) {
+                // Create the .gitignore file.
+                Gitignore::new().write_to(&path)?;
+            }
+
+            // Verify that the README.md file does not exist.
+            if !README::exists_at(&path) {
+                // Create the README.md file.
+                README::new(package_name).write_to(&path)?;
+            }
+
+            // Create the source directory.
+            SourceDirectory::create(&path)?;
+
+            // Create a new library or binary file.
+            if is_lib {
+                // Create the library file in the source directory.
+                LibraryFile::new(&package_name).write_to(&path)?;
+            } else {
+                // Create the input directory.
+                InputsDirectory::create(&path)?;
+
+                // Create the input file in the inputs directory.
+                InputFile::new(&package_name).write_to(&path)?;
+
+                // Create the state file in the inputs directory.
+                StateFile::new(&package_name).write_to(&path)?;
+
+                // Create the main file in the source directory.
+                MainFile::new(&package_name).write_to(&path)?;
+            }
+        }
+        // Next, verify that a valid Leo package has been initialized in this directory
+        {
+            if !Self::is_initialized(package_name, is_lib, path) {
+                return Err(
+                    PackageError::FailedToInitialize(package_name.to_owned(), path.as_os_str().to_owned()).into(),
+                );
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Removes the package at the given path
+    pub fn remove_imported_package(package_name: &str, path: &PathBuf) -> Result<(), PackageError> {
+        Ok(ImportsDirectory::remove_import(path, package_name)?)
+    }
+}
diff --git a/package/src/root/gitignore.rs b/package/src/root/gitignore.rs
index 9ecb3f7adc..306dd5220a 100644
--- a/package/src/root/gitignore.rs
+++ b/package/src/root/gitignore.rs
@@ -21,7 +21,7 @@ use crate::errors::GitignoreError;
 use serde::Deserialize;
 use std::{fs::File, io::Write, path::PathBuf};
 
-pub static GITIGNORE_FILE_NAME: &str = ".gitignore";
+pub static GITIGNORE_FILENAME: &str = ".gitignore";
 
 #[derive(Deserialize)]
 pub struct Gitignore;
@@ -34,7 +34,7 @@ impl Gitignore {
     pub fn exists_at(path: &PathBuf) -> bool {
         let mut path = path.to_owned();
         if path.is_dir() {
-            path.push(PathBuf::from(GITIGNORE_FILE_NAME));
+            path.push(PathBuf::from(GITIGNORE_FILENAME));
         }
         path.exists()
     }
@@ -42,7 +42,7 @@ impl Gitignore {
     pub fn write_to(self, path: &PathBuf) -> Result<(), GitignoreError> {
         let mut path = path.to_owned();
         if path.is_dir() {
-            path.push(PathBuf::from(GITIGNORE_FILE_NAME));
+            path.push(PathBuf::from(GITIGNORE_FILENAME));
         }
 
         let mut file = File::create(&path)?;
diff --git a/package/src/root/manifest.rs b/package/src/root/manifest.rs
index 293e993683..bb33dd63ac 100644
--- a/package/src/root/manifest.rs
+++ b/package/src/root/manifest.rs
@@ -14,7 +14,7 @@
 // You should have received a copy of the GNU General Public License
 // along with the Leo library. If not, see <https://www.gnu.org/licenses/>.
 
-use crate::errors::ManifestError;
+use crate::{errors::ManifestError, package::Package};
 
 use serde::Deserialize;
 use std::{
@@ -24,44 +24,35 @@ use std::{
     path::PathBuf,
 };
 
-pub const MANIFEST_FILE_NAME: &str = "Leo.toml";
+pub const MANIFEST_FILENAME: &str = "Leo.toml";
 
 #[derive(Clone, Deserialize)]
 pub struct Remote {
     pub author: String,
 }
 
-#[derive(Deserialize)]
-pub struct Package {
-    pub name: String,
-    pub version: String,
-    pub description: Option<String>,
-    pub license: Option<String>,
-    pub remote: Option<Remote>,
-}
-
 #[derive(Deserialize)]
 pub struct Manifest {
     pub package: Package,
+    pub remote: Option<Remote>,
 }
 
 impl Manifest {
     pub fn new(package_name: &str) -> Self {
         Self {
-            package: Package {
-                name: package_name.to_owned(),
-                version: "0.1.0".to_owned(),
-                description: None,
-                license: None,
-                remote: None,
-            },
+            package: Package::new(package_name),
+            remote: None,
         }
     }
 
+    pub fn filename() -> String {
+        MANIFEST_FILENAME.to_string()
+    }
+
     pub fn exists_at(path: &PathBuf) -> bool {
         let mut path = path.to_owned();
         if path.is_dir() {
-            path.push(PathBuf::from(MANIFEST_FILE_NAME));
+            path.push(PathBuf::from(MANIFEST_FILENAME));
         }
         path.exists()
     }
@@ -83,18 +74,18 @@ impl Manifest {
     }
 
     pub fn get_package_remote(&self) -> Option<Remote> {
-        self.package.remote.clone()
+        self.remote.clone()
     }
 
     pub fn write_to(self, path: &PathBuf) -> Result<(), ManifestError> {
         let mut path = path.to_owned();
         if path.is_dir() {
-            path.push(PathBuf::from(MANIFEST_FILE_NAME));
+            path.push(PathBuf::from(MANIFEST_FILENAME));
         }
 
-        let mut file = File::create(&path).map_err(|error| ManifestError::Creating(MANIFEST_FILE_NAME, error))?;
+        let mut file = File::create(&path).map_err(|error| ManifestError::Creating(MANIFEST_FILENAME, error))?;
         file.write_all(self.template().as_bytes())
-            .map_err(|error| ManifestError::Writing(MANIFEST_FILE_NAME, error))
+            .map_err(|error| ManifestError::Writing(MANIFEST_FILENAME, error))
     }
 
     fn template(&self) -> String {
@@ -119,18 +110,18 @@ impl TryFrom<&PathBuf> for Manifest {
     fn try_from(path: &PathBuf) -> Result<Self, Self::Error> {
         let mut path = path.to_owned();
         if path.is_dir() {
-            path.push(PathBuf::from(MANIFEST_FILE_NAME));
+            path.push(PathBuf::from(MANIFEST_FILENAME));
         }
 
-        let mut file = File::open(path.clone()).map_err(|error| ManifestError::Opening(MANIFEST_FILE_NAME, error))?;
+        let mut file = File::open(path.clone()).map_err(|error| ManifestError::Opening(MANIFEST_FILENAME, error))?;
         let size = file
             .metadata()
-            .map_err(|error| ManifestError::Metadata(MANIFEST_FILE_NAME, error))?
+            .map_err(|error| ManifestError::Metadata(MANIFEST_FILENAME, error))?
             .len() as usize;
 
         let mut buffer = String::with_capacity(size);
         file.read_to_string(&mut buffer)
-            .map_err(|error| ManifestError::Reading(MANIFEST_FILE_NAME, error))?;
+            .map_err(|error| ManifestError::Reading(MANIFEST_FILENAME, error))?;
 
         // Determine if the old remote format is being used, and update to new convention
 
@@ -185,12 +176,12 @@ author = "{author}"
 
         // Rewrite the toml file if it has been updated
         if buffer != new_toml {
-            let mut file = File::create(&path).map_err(|error| ManifestError::Creating(MANIFEST_FILE_NAME, error))?;
+            let mut file = File::create(&path).map_err(|error| ManifestError::Creating(MANIFEST_FILENAME, error))?;
             file.write_all(new_toml.as_bytes())
-                .map_err(|error| ManifestError::Writing(MANIFEST_FILE_NAME, error))?;
+                .map_err(|error| ManifestError::Writing(MANIFEST_FILENAME, error))?;
         }
 
         // Read the toml file
-        Ok(toml::from_str(&new_toml).map_err(|error| ManifestError::Parsing(MANIFEST_FILE_NAME, error))?)
+        Ok(toml::from_str(&new_toml).map_err(|error| ManifestError::Parsing(MANIFEST_FILENAME, error))?)
     }
 }
diff --git a/package/src/root/readme.rs b/package/src/root/readme.rs
index f59adfaafd..8631211e82 100644
--- a/package/src/root/readme.rs
+++ b/package/src/root/readme.rs
@@ -21,7 +21,7 @@ use crate::errors::READMEError;
 use serde::Deserialize;
 use std::{fs::File, io::Write, path::PathBuf};
 
-pub static README_FILE_NAME: &str = "README.md";
+pub static README_FILENAME: &str = "README.md";
 
 #[derive(Deserialize)]
 pub struct README {
@@ -42,7 +42,7 @@ impl README {
     pub fn exists_at(path: &PathBuf) -> bool {
         let mut path = path.to_owned();
         if path.is_dir() {
-            path.push(PathBuf::from(README_FILE_NAME));
+            path.push(PathBuf::from(README_FILENAME));
         }
         path.exists()
     }
@@ -50,7 +50,7 @@ impl README {
     pub fn write_to(self, path: &PathBuf) -> Result<(), READMEError> {
         let mut path = path.to_owned();
         if path.is_dir() {
-            path.push(PathBuf::from(README_FILE_NAME));
+            path.push(PathBuf::from(README_FILENAME));
         }
 
         let mut file = File::create(&path)?;
diff --git a/package/src/root/zip.rs b/package/src/root/zip.rs
index 055452debf..7277fb3dd6 100644
--- a/package/src/root/zip.rs
+++ b/package/src/root/zip.rs
@@ -28,7 +28,7 @@ use crate::{
         PROVING_KEY_FILE_EXTENSION,
         VERIFICATION_KEY_FILE_EXTENSION,
     },
-    root::{MANIFEST_FILE_NAME, README_FILE_NAME},
+    root::{MANIFEST_FILENAME, README_FILENAME},
     source::{SOURCE_DIRECTORY_NAME, SOURCE_FILE_EXTENSION},
 };
 
@@ -172,8 +172,7 @@ fn is_included(path: &Path) -> bool {
     }
 
     // Allow the README.md and Leo.toml files in the root directory
-    if (path.ends_with(README_FILE_NAME) | path.ends_with(MANIFEST_FILE_NAME)) & (path.parent() == Some(Path::new("")))
-    {
+    if (path.ends_with(README_FILENAME) | path.ends_with(MANIFEST_FILENAME)) & (path.parent() == Some(Path::new(""))) {
         return true;
     }
 
diff --git a/package/src/source/lib.rs b/package/src/source/library.rs
similarity index 79%
rename from package/src/source/lib.rs
rename to package/src/source/library.rs
index 426822a4f9..0531f0ed67 100644
--- a/package/src/source/lib.rs
+++ b/package/src/source/library.rs
@@ -16,43 +16,47 @@
 
 //! The `lib.leo` file.
 
-use crate::{errors::LibFileError, source::directory::SOURCE_DIRECTORY_NAME};
+use crate::{errors::LibraryFileError, source::directory::SOURCE_DIRECTORY_NAME};
 
 use serde::Deserialize;
 use std::{fs::File, io::Write, path::PathBuf};
 
-pub static LIB_FILE_NAME: &str = "lib.leo";
+pub static LIBRARY_FILENAME: &str = "lib.leo";
 
 #[derive(Deserialize)]
-pub struct LibFile {
+pub struct LibraryFile {
     pub package_name: String,
 }
 
-impl LibFile {
+impl LibraryFile {
     pub fn new(package_name: &str) -> Self {
         Self {
             package_name: package_name.to_string(),
         }
     }
 
+    pub fn filename() -> String {
+        format!("{}{}", SOURCE_DIRECTORY_NAME, LIBRARY_FILENAME)
+    }
+
     pub fn exists_at(path: &PathBuf) -> bool {
         let mut path = path.to_owned();
         if path.is_dir() {
             if !path.ends_with(SOURCE_DIRECTORY_NAME) {
                 path.push(PathBuf::from(SOURCE_DIRECTORY_NAME));
             }
-            path.push(PathBuf::from(LIB_FILE_NAME));
+            path.push(PathBuf::from(LIBRARY_FILENAME));
         }
         path.exists()
     }
 
-    pub fn write_to(self, path: &PathBuf) -> Result<(), LibFileError> {
+    pub fn write_to(self, path: &PathBuf) -> Result<(), LibraryFileError> {
         let mut path = path.to_owned();
         if path.is_dir() {
             if !path.ends_with(SOURCE_DIRECTORY_NAME) {
                 path.push(PathBuf::from(SOURCE_DIRECTORY_NAME));
             }
-            path.push(PathBuf::from(LIB_FILE_NAME));
+            path.push(PathBuf::from(LIBRARY_FILENAME));
         }
 
         let mut file = File::create(&path)?;
diff --git a/package/src/source/main.rs b/package/src/source/main.rs
index 2fd5faf6e9..8136f12f5c 100644
--- a/package/src/source/main.rs
+++ b/package/src/source/main.rs
@@ -21,7 +21,7 @@ use crate::{errors::MainFileError, source::directory::SOURCE_DIRECTORY_NAME};
 use serde::Deserialize;
 use std::{fs::File, io::Write, path::PathBuf};
 
-pub static MAIN_FILE_NAME: &str = "main.leo";
+pub static MAIN_FILENAME: &str = "main.leo";
 
 #[derive(Deserialize)]
 pub struct MainFile {
@@ -35,13 +35,17 @@ impl MainFile {
         }
     }
 
+    pub fn filename() -> String {
+        format!("{}{}", SOURCE_DIRECTORY_NAME, MAIN_FILENAME)
+    }
+
     pub fn exists_at(path: &PathBuf) -> bool {
         let mut path = path.to_owned();
         if path.is_dir() {
             if !path.ends_with(SOURCE_DIRECTORY_NAME) {
                 path.push(PathBuf::from(SOURCE_DIRECTORY_NAME));
             }
-            path.push(PathBuf::from(MAIN_FILE_NAME));
+            path.push(PathBuf::from(MAIN_FILENAME));
         }
         path.exists()
     }
@@ -52,7 +56,7 @@ impl MainFile {
             if !path.ends_with(SOURCE_DIRECTORY_NAME) {
                 path.push(PathBuf::from(SOURCE_DIRECTORY_NAME));
             }
-            path.push(PathBuf::from(MAIN_FILE_NAME));
+            path.push(PathBuf::from(MAIN_FILENAME));
         }
 
         let mut file = File::create(&path)?;
diff --git a/package/src/source/mod.rs b/package/src/source/mod.rs
index 35f5b284ac..7c849f7f94 100644
--- a/package/src/source/mod.rs
+++ b/package/src/source/mod.rs
@@ -17,8 +17,8 @@
 pub mod directory;
 pub use directory::*;
 
-pub mod lib;
-pub use lib::*;
+pub mod library;
+pub use library::*;
 
 pub mod main;
 pub use main::*;
diff --git a/package/tests/initialize.rs b/package/tests/initialize.rs
new file mode 100644
index 0000000000..62dc23eb85
--- /dev/null
+++ b/package/tests/initialize.rs
@@ -0,0 +1,192 @@
+// Copyright (C) 2019-2020 Aleo Systems Inc.
+// This file is part of the Leo library.
+
+// The Leo library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// The Leo library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with the Leo library. If not, see <https://www.gnu.org/licenses/>.
+
+use lazy_static::lazy_static;
+use std::{
+    cell::RefCell,
+    env,
+    fs,
+    path::PathBuf,
+    sync::atomic::{AtomicUsize, Ordering},
+};
+
+const PACKAGE_TEST_DIRECTORY: &str = "package-testing";
+
+thread_local! {
+    /// Establish a test id for each test.
+    static TEST_ID: RefCell<Option<usize>> = RefCell::new(None);
+}
+
+lazy_static! {
+    /// Create a testing directory for packages in `target/`
+    static ref TEST_DIR: PathBuf = {
+        let mut path = env::current_exe().unwrap();
+        path.pop(); // Remove executable name
+        path.pop(); // Remove 'debug'
+
+        // Attempt to point at the `target` directory
+        if path.file_name().and_then(|s| s.to_str()) != Some("target") {
+            path.pop();
+        }
+
+        path.push(PACKAGE_TEST_DIRECTORY);
+        fs::create_dir_all(&path).unwrap();
+
+        path
+    };
+}
+
+/// Create a new directory for each test based on the ID of the test.
+fn test_dir() -> PathBuf {
+    static NEXT_ID: AtomicUsize = AtomicUsize::new(0);
+
+    let id = NEXT_ID.fetch_add(1, Ordering::Relaxed);
+    TEST_ID.with(|n| *n.borrow_mut() = Some(id));
+
+    let path: PathBuf = TEST_DIR.join(&format!("t{}", id)).into();
+
+    if path.exists() {
+        if let Err(e) = fs::remove_dir_all(&path) {
+            panic!("failed to remove {:?}: {:?}", &path, e)
+        }
+    }
+
+    fs::create_dir_all(&path).unwrap();
+
+    path
+}
+
+// Tests for package initialization
+mod initialize_package {
+    use super::*;
+    use leo_package::{
+        inputs::{InputFile, InputsDirectory, StateFile},
+        package::Package,
+        root::Manifest,
+        source::{LibraryFile, MainFile, SourceDirectory},
+    };
+
+    const TEST_PACKAGE_NAME: &str = "test-package";
+
+    #[test]
+    fn initialize_valid_package() {
+        let test_directory = test_dir();
+
+        // Ensure a package can be initialized at the `test_directory`
+        assert!(Package::can_initialize(TEST_PACKAGE_NAME, false, &test_directory));
+
+        // Initialize a package at the `test_directory`
+        assert!(Package::initialize(TEST_PACKAGE_NAME, false, &test_directory).is_ok());
+
+        // Ensure a package is initialized at the `test_directory`
+        assert!(Package::is_initialized(TEST_PACKAGE_NAME, false, &test_directory));
+    }
+
+    #[test]
+    #[ignore]
+    fn initialize_fails_with_invalid_package_names() {
+        unimplemented!()
+    }
+
+    #[test]
+    fn initialize_fails_with_existing_manifest() {
+        let test_directory = test_dir();
+
+        // Ensure a package can be initialized at the `test_directory`
+        assert!(Package::can_initialize(TEST_PACKAGE_NAME, false, &test_directory));
+
+        // Manually add a manifest file to the `test_directory`
+        Manifest::new(TEST_PACKAGE_NAME).write_to(&test_directory).unwrap();
+
+        // Attempt to initialize a package at the `test_directory`
+        assert!(Package::initialize(TEST_PACKAGE_NAME, false, &test_directory).is_err());
+
+        // Ensure package is not initialized at the `test_directory`
+        assert!(!Package::is_initialized(TEST_PACKAGE_NAME, false, &test_directory));
+    }
+
+    #[test]
+    fn initialize_fails_with_existing_library_file() {
+        let test_directory = test_dir();
+
+        // Ensure a package can be initialized at the `test_directory`
+        assert!(Package::can_initialize(TEST_PACKAGE_NAME, true, &test_directory));
+
+        // Manually add a source directory and a library file to the `test_directory`
+        SourceDirectory::create(&test_directory).unwrap();
+        LibraryFile::new(TEST_PACKAGE_NAME).write_to(&test_directory).unwrap();
+
+        // Attempt to initialize a package at the `test_directory`
+        assert!(Package::initialize(TEST_PACKAGE_NAME, true, &test_directory).is_err());
+
+        // Ensure package is not initialized at the `test_directory`
+        assert!(!Package::is_initialized(TEST_PACKAGE_NAME, true, &test_directory));
+    }
+
+    #[test]
+    fn initialize_fails_with_existing_input_file() {
+        let test_directory = test_dir();
+
+        // Ensure a package can be initialized at the `test_directory`
+        assert!(Package::can_initialize(TEST_PACKAGE_NAME, false, &test_directory));
+
+        // Manually add an inputs directory and an input file to the `test_directory`
+        InputsDirectory::create(&test_directory).unwrap();
+        InputFile::new(TEST_PACKAGE_NAME).write_to(&test_directory).unwrap();
+
+        // Attempt to initialize a package at the `test_directory`
+        assert!(Package::initialize(TEST_PACKAGE_NAME, false, &test_directory).is_err());
+
+        // Ensure package is not initialized at the `test_directory`
+        assert!(!Package::is_initialized(TEST_PACKAGE_NAME, false, &test_directory));
+    }
+
+    #[test]
+    fn initialize_fails_with_existing_state_file() {
+        let test_directory = test_dir();
+
+        // Ensure a package can be initialized at the `test_directory`
+        assert!(Package::can_initialize(TEST_PACKAGE_NAME, false, &test_directory));
+
+        // Manually add an inputs directory and a state file to the `test_directory`
+        InputsDirectory::create(&test_directory).unwrap();
+        StateFile::new(TEST_PACKAGE_NAME).write_to(&test_directory).unwrap();
+
+        // Attempt to initialize a package at the `test_directory`
+        assert!(Package::initialize(TEST_PACKAGE_NAME, false, &test_directory).is_err());
+
+        // Ensure package is not initialized at the `test_directory`
+        assert!(!Package::is_initialized(TEST_PACKAGE_NAME, false, &test_directory));
+    }
+
+    #[test]
+    fn initialize_fails_with_existing_main_file() {
+        let test_directory = test_dir();
+
+        // Ensure a package can be initialized at the `test_directory`
+        assert!(Package::can_initialize(TEST_PACKAGE_NAME, false, &test_directory));
+
+        // Manually add a source directory and a main file to the `test_directory`
+        SourceDirectory::create(&test_directory).unwrap();
+        MainFile::new(TEST_PACKAGE_NAME).write_to(&test_directory).unwrap();
+
+        // Attempt to initialize a package at the `test_directory`
+        assert!(Package::initialize(TEST_PACKAGE_NAME, false, &test_directory).is_err());
+
+        // Ensure package is not initialized at the `test_directory`
+        assert!(!Package::is_initialized(TEST_PACKAGE_NAME, false, &test_directory));
+    }
+}
diff --git a/typed/src/circuits/circuit_member.rs b/typed/src/circuits/circuit_member.rs
index 9434fa9e9a..008806d3d3 100644
--- a/typed/src/circuits/circuit_member.rs
+++ b/typed/src/circuits/circuit_member.rs
@@ -26,15 +26,18 @@ use std::fmt;
 
 #[derive(Clone, PartialEq, Eq, Serialize, Deserialize)]
 pub enum CircuitMember {
-    CircuitVariable(Identifier, Type),
+    // (is_mutable, variable_name, variable_type)
+    CircuitVariable(bool, Identifier, Type),
+    // (is_static, function)
     CircuitFunction(bool, Function),
 }
 
 impl<'ast> From<AstCircuitVariableDefinition<'ast>> for CircuitMember {
     fn from(circuit_value: AstCircuitVariableDefinition<'ast>) -> Self {
         CircuitMember::CircuitVariable(
+            circuit_value.mutable.is_some(),
             Identifier::from(circuit_value.identifier),
-            Type::from(circuit_value._type),
+            Type::from(circuit_value.type_),
         )
     }
 }
@@ -60,9 +63,14 @@ impl<'ast> From<AstCircuitMember<'ast>> for CircuitMember {
 impl fmt::Display for CircuitMember {
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
         match self {
-            CircuitMember::CircuitVariable(ref identifier, ref _type) => write!(f, "{}: {}", identifier, _type),
-            CircuitMember::CircuitFunction(ref _static, ref function) => {
-                if *_static {
+            CircuitMember::CircuitVariable(ref mutable, ref identifier, ref type_) => {
+                if *mutable {
+                    write!(f, "mut ")?;
+                }
+                write!(f, "{}: {}", identifier, type_)
+            }
+            CircuitMember::CircuitFunction(ref static_, ref function) => {
+                if *static_ {
                     write!(f, "static ")?;
                 }
                 write!(f, "{}", function)
diff --git a/typed/src/console/console_function.rs b/typed/src/console/console_function.rs
index ca1faf5efa..95d0a0c37d 100644
--- a/typed/src/console/console_function.rs
+++ b/typed/src/console/console_function.rs
@@ -1,3 +1,19 @@
+// Copyright (C) 2019-2020 Aleo Systems Inc.
+// This file is part of the Leo library.
+
+// The Leo library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// The Leo library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with the Leo library. If not, see <https://www.gnu.org/licenses/>.
+
 use crate::{Expression, FormattedString};
 use leo_ast::console::{
     ConsoleAssert as AstConsoleAssert,
diff --git a/typed/src/console/console_function_call.rs b/typed/src/console/console_function_call.rs
index 0c74685b14..b2a9ef09f1 100644
--- a/typed/src/console/console_function_call.rs
+++ b/typed/src/console/console_function_call.rs
@@ -1,3 +1,19 @@
+// Copyright (C) 2019-2020 Aleo Systems Inc.
+// This file is part of the Leo library.
+
+// The Leo library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// The Leo library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with the Leo library. If not, see <https://www.gnu.org/licenses/>.
+
 use crate::{ConsoleFunction, Span};
 use leo_ast::console::ConsoleFunctionCall as AstConsoleFunctionCall;
 
diff --git a/typed/src/console/mod.rs b/typed/src/console/mod.rs
index 81b012e896..071f43d819 100644
--- a/typed/src/console/mod.rs
+++ b/typed/src/console/mod.rs
@@ -1,3 +1,19 @@
+// Copyright (C) 2019-2020 Aleo Systems Inc.
+// This file is part of the Leo library.
+
+// The Leo library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// The Leo library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with the Leo library. If not, see <https://www.gnu.org/licenses/>.
+
 pub mod console_function;
 pub use console_function::*;
 
