diff --git a/.circleci/config.yml b/.circleci/config.yml
index eb801f834d..5ce58312b2 100644
--- a/.circleci/config.yml
+++ b/.circleci/config.yml
@@ -186,18 +186,18 @@ jobs:
             export LEO=/home/circleci/project/project/bin/leo
             ./project/.circleci/leo-setup.sh
 
-  leo-add-remove:
-    docker:
-      - image: cimg/rust:1.54.0
-    resource_class: xlarge
-    steps:
-      - attach_workspace:
-          at: /home/circleci/project/
-      - run:
-          name: leo add & remove
-          command: |
-            export LEO=/home/circleci/project/project/bin/leo
-            ./project/.circleci/leo-add-remove.sh
+  # leo-add-remove:
+  #   docker:
+  #     - image: cimg/rust:1.54.0
+  #   resource_class: xlarge
+  #   steps:
+  #     - attach_workspace:
+  #         at: /home/circleci/project/
+  #     - run:
+  #         name: leo add & remove
+  #         command: |
+  #           export LEO=/home/circleci/project/project/bin/leo
+  #           ./project/.circleci/leo-add-remove.sh
 
   leo-check-constraints:
     docker:
@@ -270,9 +270,9 @@ workflows:
       - leo-setup:
           requires:
             - leo-executable
-      - leo-add-remove:
-          requires:
-            - leo-executable
+      # - leo-add-remove:
+      #     requires:
+      #       - leo-executable
       - leo-check-constraints:
           requires:
             - leo-executable
diff --git a/.circleci/r b/.circleci/r
deleted file mode 100644
index e69de29bb2..0000000000
diff --git a/.github/workflows/acl2.yml b/.github/workflows/acl2.yml
new file mode 100644
index 0000000000..d9fc9c4af4
--- /dev/null
+++ b/.github/workflows/acl2.yml
@@ -0,0 +1,95 @@
+name: Leo-ACL2
+on: 
+  pull_request:
+    push:
+      branches:
+        - master
+        - staging
+        - trying
+      paths-ignore:
+        - 'docs/**'
+        - 'documentation/**'
+env: 
+  RUST_BACKTRACE: 1
+
+# This job can only be run on linux (Ubuntu)
+jobs:
+  acl2:
+    name: leo-acl2
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v2
+
+      - name: Install Rust
+        uses: actions-rs/toolchain@v1
+        with:
+          profile: minimal
+          toolchain: stable
+          override: true
+
+      - name: Generate asts
+        run: |
+          cargo -q run -p leo-test-framework --bin tgc
+
+      # Pull the latest release from the leo-acl2-bin repo, and put it into the
+      # repo/acl2 directory. After it's done, unpack the tgz file locally.
+      - name: Pull tgc executable
+        run: |
+          mkdir acl2 && cd acl2;
+          wget $(curl -s https://api.github.com/repos/AleoHQ/leo-acl2-bin/releases/latest \
+              | grep "browser_download_url.*.tgz"  \
+              | cut -d : -f 2,3 \
+              | tr -d \" \
+              | xargs)
+
+          tar -xvzf $(ls)
+      
+      # Using the prepared ASTs and the pulled and unzipped tgc run theorem generation.
+      - name: Run tgc over ASTs
+        run: |
+          canonicalization_errors=();
+          type_inference_errors=();
+          for dir in `ls tmp/tgc`;
+          do
+            cd tmp/tgc/$dir; # enter the directory
+            ./../../../acl2/tgc canonicalization initial_ast.json canonicalization_ast.json canonicalization-theorem.lisp > canonicalization_result.out || canonicalization_errors+=("$dir");
+            # Disabling Type inference for now
+            # ./../../../acl2/tgc type-inference canonicalization_ast.json type_inferenced_ast.json type-inference-theorem.lisp > type_inference_result.out || type_inference_errors+=("$dir");
+            cd ../../..
+          done;
+
+          if [ ${#canonicalization_errors[@]} -eq 0 ]; then
+            echo "Canonicalization - Success!"
+          else
+            echo "Canonicalization Failures:";
+            for dir in ${canonicalization_errors[@]};
+            do
+              echo $dir;
+            done;
+
+            echo "Attaching logs:"
+            for dir in ${canonicalization_errors[@]};
+            do
+              cat tmp/tgc/$dir/canonicalization_result.out
+            done;
+            exit 1
+          fi
+
+          if [ ${#type_inference_errors[@]} -eq 0 ]; then
+            echo "Type Inference - Success!"
+          else
+            echo "Type Inference Failures:";
+            for dir in ${type_inference_errors[@]};
+            do
+              echo $dir;
+            done;
+
+            echo "Attaching logs:"
+            for dir in ${type_inference_errors[@]};
+            do
+              cat tmp/tgc/$dir/type_inference_result.out
+            done;
+
+            exit 1
+          fi
diff --git a/Cargo.lock b/Cargo.lock
index 1aa878be32..d76817cbf0 100644
--- a/Cargo.lock
+++ b/Cargo.lock
@@ -1320,6 +1320,7 @@ dependencies = [
  "console",
  "dirs",
  "from-pest",
+ "indexmap",
  "lazy_static",
  "leo-ast",
  "leo-compiler",
@@ -1359,6 +1360,7 @@ version = "1.5.3"
 name = "leo-package"
 version = "1.5.3"
 dependencies = [
+ "indexmap",
  "lazy_static",
  "leo-errors",
  "serde",
@@ -1423,9 +1425,15 @@ dependencies = [
 name = "leo-test-framework"
 version = "1.5.3"
 dependencies = [
+ "leo-asg",
+ "leo-ast",
+ "leo-compiler",
+ "leo-imports",
+ "leo-parser",
  "serde",
  "serde_json",
  "serde_yaml",
+ "structopt",
 ]
 
 [[package]]
diff --git a/Cargo.toml b/Cargo.toml
index fb3a68e932..852e27b752 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -117,6 +117,10 @@ version = "0.14.0"
 [dependencies.from-pest]
 version = "0.3.1"
 
+[dependencies.indexmap]
+version = "1.7"
+features = ["serde"]
+
 [dependencies.lazy_static]
 version = "1.4.0"
 
diff --git a/FORMAT_ABNF_GRAMMER.md b/FORMAT_ABNF_GRAMMER.md
deleted file mode 100644
index e69de29bb2..0000000000
diff --git a/asg/src/expression/ternary.rs b/asg/src/expression/ternary.rs
index 8f286d4559..4e26d254aa 100644
--- a/asg/src/expression/ternary.rs
+++ b/asg/src/expression/ternary.rs
@@ -87,7 +87,15 @@ impl<'a> FromAst<'a, leo_ast::TernaryExpression> for TernaryExpression<'a> {
         )?);
         let left: PartialType = if_true.get().get_type().unwrap().into();
 
-        let if_false = Cell::new(<&Expression<'a>>::from_ast(scope, &*value.if_false, expected_type)?);
+        let if_false = if expected_type.is_none() {
+            Cell::new(<&Expression<'a>>::from_ast(
+                scope,
+                &*value.if_false,
+                Some(left.clone()),
+            )?)
+        } else {
+            Cell::new(<&Expression<'a>>::from_ast(scope, &*value.if_false, expected_type)?)
+        };
         let right = if_false.get().get_type().unwrap().into();
 
         if left != right {
diff --git a/ast/src/reducer/canonicalization.rs b/ast/src/reducer/canonicalization.rs
index 63bf895b59..1f2ec23b74 100644
--- a/ast/src/reducer/canonicalization.rs
+++ b/ast/src/reducer/canonicalization.rs
@@ -291,7 +291,7 @@ impl Canonicalizer {
                     return Expression::Identifier(self.circuit_name.as_ref().unwrap().clone());
                 }
             }
-            _ => {}
+            _ => (),
         }
 
         expression.clone()
diff --git a/compiler/src/compiler.rs b/compiler/src/compiler.rs
index 7e8c825fb4..c7b27872b8 100644
--- a/compiler/src/compiler.rs
+++ b/compiler/src/compiler.rs
@@ -23,6 +23,7 @@ pub use leo_asg::{new_context, AsgContext as Context, AsgContext};
 use leo_asg::{Asg, AsgPass, Program as AsgProgram};
 use leo_ast::{Input, MainInput, Program as AstProgram};
 use leo_errors::{CompilerError, Result};
+use leo_imports::ImportParser;
 use leo_input::LeoInputParser;
 use leo_package::inputs::InputPairs;
 use leo_parser::parse_ast;
@@ -34,6 +35,7 @@ use snarkvm_r1cs::{ConstraintSynthesizer, ConstraintSystem, SynthesisError};
 
 use sha2::{Digest, Sha256};
 use std::{
+    collections::HashMap,
     fs,
     marker::PhantomData,
     path::{Path, PathBuf},
@@ -62,6 +64,7 @@ pub struct Compiler<'a, F: PrimeField, G: GroupType<F>> {
     context: AsgContext<'a>,
     asg: Option<AsgProgram<'a>>,
     options: CompilerOptions,
+    imports_map: HashMap<String, String>,
     ast_snapshot_options: AstSnapshotOptions,
     _engine: PhantomData<F>,
     _group: PhantomData<G>,
@@ -77,6 +80,7 @@ impl<'a, F: PrimeField, G: GroupType<F>> Compiler<'a, F, G> {
         output_directory: PathBuf,
         context: AsgContext<'a>,
         options: Option<CompilerOptions>,
+        imports_map: HashMap<String, String>,
         ast_snapshot_options: Option<AstSnapshotOptions>,
     ) -> Self {
         Self {
@@ -88,6 +92,7 @@ impl<'a, F: PrimeField, G: GroupType<F>> Compiler<'a, F, G> {
             asg: None,
             context,
             options: options.unwrap_or_default(),
+            imports_map,
             ast_snapshot_options: ast_snapshot_options.unwrap_or_default(),
             _engine: PhantomData,
             _group: PhantomData,
@@ -107,6 +112,7 @@ impl<'a, F: PrimeField, G: GroupType<F>> Compiler<'a, F, G> {
         output_directory: PathBuf,
         context: AsgContext<'a>,
         options: Option<CompilerOptions>,
+        imports_map: HashMap<String, String>,
         ast_snapshot_options: Option<AstSnapshotOptions>,
     ) -> Result<Self> {
         let mut compiler = Self::new(
@@ -115,6 +121,7 @@ impl<'a, F: PrimeField, G: GroupType<F>> Compiler<'a, F, G> {
             output_directory,
             context,
             options,
+            imports_map,
             ast_snapshot_options,
         );
 
@@ -146,6 +153,7 @@ impl<'a, F: PrimeField, G: GroupType<F>> Compiler<'a, F, G> {
         state_path: &Path,
         context: AsgContext<'a>,
         options: Option<CompilerOptions>,
+        imports_map: HashMap<String, String>,
         ast_snapshot_options: Option<AstSnapshotOptions>,
     ) -> Result<Self> {
         let mut compiler = Self::new(
@@ -154,6 +162,7 @@ impl<'a, F: PrimeField, G: GroupType<F>> Compiler<'a, F, G> {
             output_directory,
             context,
             options,
+            imports_map,
             ast_snapshot_options,
         );
 
@@ -239,7 +248,7 @@ impl<'a, F: PrimeField, G: GroupType<F>> Compiler<'a, F, G> {
             ast.to_json_file(self.output_directory.clone(), "initial_ast.json")?;
         }
 
-        // Preform canonicalization of AST always.
+        // Perform canonicalization of AST always.
         ast.canonicalize()?;
 
         if self.ast_snapshot_options.canonicalized {
@@ -256,7 +265,7 @@ impl<'a, F: PrimeField, G: GroupType<F>> Compiler<'a, F, G> {
         let asg = Asg::new(
             self.context,
             &self.program,
-            &mut leo_imports::ImportParser::new(self.main_file_path.clone()),
+            &mut ImportParser::new(self.main_file_path.clone(), self.imports_map.clone()),
         )?;
 
         if self.ast_snapshot_options.type_inferenced {
diff --git a/compiler/src/option.rs b/compiler/src/option.rs
index 0dc5ccca51..f163cec5ad 100644
--- a/compiler/src/option.rs
+++ b/compiler/src/option.rs
@@ -17,7 +17,7 @@
 ///
 /// Toggles compiler optimizations on the program.
 ///
-#[derive(Clone)]
+#[derive(Clone, Debug)]
 pub struct CompilerOptions {
     pub constant_folding_enabled: bool,
     pub dead_code_elimination_enabled: bool,
diff --git a/compiler/src/test.rs b/compiler/src/test.rs
index 244fa58961..bb8c688799 100644
--- a/compiler/src/test.rs
+++ b/compiler/src/test.rs
@@ -14,7 +14,10 @@
 // You should have received a copy of the GNU General Public License
 // along with the Leo library. If not, see <https://www.gnu.org/licenses/>.
 
-use std::path::{Path, PathBuf};
+use std::{
+    collections::HashMap,
+    path::{Path, PathBuf},
+};
 
 use leo_asg::*;
 use leo_ast::{Ast, Program};
@@ -50,6 +53,7 @@ fn new_compiler(path: PathBuf, theorem_options: Option<AstSnapshotOptions>) -> E
         output_dir,
         make_test_context(),
         None,
+        HashMap::new(),
         theorem_options,
     )
 }
diff --git a/errors/src/cli/cli_errors.rs b/errors/src/cli/cli_errors.rs
index 09a060439b..28f8b9b77e 100644
--- a/errors/src/cli/cli_errors.rs
+++ b/errors/src/cli/cli_errors.rs
@@ -157,7 +157,7 @@ create_errors!(
 
     /// For when the CLI cannot find the manifest file.
     @backtraced
-    mainifest_file_not_found {
+    manifest_file_not_found {
         args: (),
         msg: "Package manifest not found, try running `leo init`",
         help: None,
@@ -379,6 +379,27 @@ create_errors!(
         msg: format!("Old release version {} {}", current, latest),
         help: None,
     }
+
+    @backtraced
+    dependencies_are_not_installed {
+        args: (),
+        msg: "dependencies are not installed, please run `leo fetch` first",
+        help: None,
+    }
+
+    @backtraced
+    recursive_dependency_found {
+        args: (message: impl Display),
+        msg: format!("recursive dependency found \n{}", message),
+        help: None,
+    }
+
+    @backtraced
+    unable_to_read_imported_dependency_manifest {
+        args: (),
+        msg: "unable to parse imported dependency's manifest",
+        help: None,
+    }
 );
 
 impl CliError {
diff --git a/errors/src/package/package_errors.rs b/errors/src/package/package_errors.rs
index fca8c663f8..b51bb00f58 100644
--- a/errors/src/package/package_errors.rs
+++ b/errors/src/package/package_errors.rs
@@ -218,19 +218,19 @@ create_errors!(
         help: None,
     }
 
-    /// For when reading the prooving key failed.
+    /// For when reading the proving key failed.
     @backtraced
     failed_to_read_proving_key_file {
         args: (path: impl Debug),
-        msg: format!("Cannot read prooving key file from the provided file path - {:?}", path),
+        msg: format!("Cannot read proving key file from the provided file path - {:?}", path),
         help: None,
     }
 
-    /// For when removing the prooving key file failed.
+    /// For when removing the proving key file failed.
     @backtraced
     failed_to_remove_proving_key_file {
         args: (path: impl Debug),
-        msg: format!("Cannot remove prooving key file from the provided file path - {:?}", path),
+        msg: format!("Cannot remove proving key file from the provided file path - {:?}", path),
         help: None,
     }
 
@@ -242,6 +242,22 @@ create_errors!(
         help: None,
     }
 
+    /// For when reading the snapshot file failed.
+    @backtraced
+    failed_to_read_snapshot_file {
+        args: (path: impl Debug),
+        msg: format!("Cannot read snapshot file from the provided file path - {:?}", path),
+        help: None,
+    }
+
+    /// For when removing the snapshot file failed.
+    @backtraced
+    failed_to_remove_snapshot_file {
+        args: (path: impl Debug),
+        msg: format!("Cannot remove snapshot file from the provided file path - {:?}", path),
+        help: None,
+    }
+
     /// For when reading the verification key file failed.
     @backtraced
     failed_to_read_verification_key_file {
@@ -481,4 +497,66 @@ create_errors!(
         msg: format!("invalid project name {}", package),
         help: None,
     }
+
+    @backtraced
+    failed_to_create_lock_file {
+        args: (filename: impl Display, error: impl ErrorArg),
+        msg: format!("failed creating manifest file `{}` {}", filename, error),
+        help: None,
+    }
+
+    /// For when getting the lock file metadata failed.
+    @backtraced
+    failed_to_get_lock_file_metadata {
+        args: (filename: impl Display, error: impl ErrorArg),
+        msg: format!("failed getting lock file metadata `{}` {}", filename, error),
+        help: None,
+    }
+
+    /// For when opening the lock file failed.
+    @backtraced
+    failed_to_open_lock_file {
+        args: (filename: impl Display, error: impl ErrorArg),
+        msg: format!("failed openining lock file `{}` {}", filename, error),
+        help: None,
+    }
+
+    /// For when parsing the lock file failed.
+    @backtraced
+    failed_to_parse_lock_file {
+        args: (filename: impl Display, error: impl ErrorArg),
+        msg: format!("failed parsing lock file `{}` {}", filename, error),
+        help: None,
+    }
+
+    /// For when reading the lock file failed.
+    @backtraced
+    failed_to_read_lock_file {
+        args: (filename: impl Display, error: impl ErrorArg),
+        msg: format!("failed reading lock file `{}` {}", filename, error),
+        help: None,
+    }
+
+    /// For when writing the lock file failed.
+    @backtraced
+    failed_to_write_lock_file {
+        args: (filename: impl Display, error: impl ErrorArg),
+        msg: format!("failed writing lock file `{}` {}", filename, error),
+        help: None,
+    }
+
+    /// For when the lock file has an IO error.
+    @backtraced
+    io_error_lock_file {
+        args: (error: impl ErrorArg),
+        msg: format!("IO error lock file from the provided file path - {}", error),
+        help: None,
+    }
+
+    @backtraced
+    failed_to_serialize_lock_file {
+        args: (error: impl ErrorArg),
+        msg: format!("serialization failed: {}", error),
+        help: None,
+    }
 );
diff --git a/imports/src/parser/import_parser.rs b/imports/src/parser/import_parser.rs
index 0212800d1a..106992551a 100644
--- a/imports/src/parser/import_parser.rs
+++ b/imports/src/parser/import_parser.rs
@@ -18,7 +18,7 @@ use leo_asg::{AsgContext, ImportResolver, Program};
 use leo_errors::{ImportError, LeoError, Result, Span};
 
 use indexmap::{IndexMap, IndexSet};
-use std::path::PathBuf;
+use std::{collections::HashMap, path::PathBuf};
 
 /// Stores imported packages.
 ///
@@ -29,14 +29,16 @@ pub struct ImportParser<'a> {
     program_path: PathBuf,
     partial_imports: IndexSet<String>,
     imports: IndexMap<String, Program<'a>>,
+    pub imports_map: HashMap<String, String>,
 }
 
 impl<'a> ImportParser<'a> {
-    pub fn new(program_path: PathBuf) -> Self {
+    pub fn new(program_path: PathBuf, imports_map: HashMap<String, String>) -> Self {
         ImportParser {
             program_path,
             partial_imports: Default::default(),
             imports: Default::default(),
+            imports_map,
         }
     }
 }
diff --git a/imports/src/parser/parse_package.rs b/imports/src/parser/parse_package.rs
index 3f2079598b..0a2700c7ac 100644
--- a/imports/src/parser/parse_package.rs
+++ b/imports/src/parser/parse_package.rs
@@ -104,10 +104,19 @@ impl<'a> ImportParser<'a> {
                 .collect::<Result<Vec<_>, std::io::Error>>()
                 .map_err(|error| ImportError::directory_error(error, error_path, span))?;
 
+            // Keeping backward compatibilty for existing packages.
+            // If index_map contains key, use it or try to access directly.
+            // TODO: Remove when migration is possible.
+            let package_name = self
+                .imports_map
+                .get(package_name)
+                .unwrap_or(&package_name.to_string())
+                .clone();
+
             // Check if the imported package name is in the imports directory.
             let matched_import_entry = entries
                 .into_iter()
-                .find(|entry| entry.file_name().into_string().unwrap().eq(package_name));
+                .find(|entry| entry.file_name().into_string().unwrap().eq(&package_name));
 
             // Check if the package name was found in both the source and imports directory.
             match (matched_source_entry, matched_import_entry) {
diff --git a/leo/commands/build.rs b/leo/commands/build.rs
index 214f56a6f3..f38c2adb48 100644
--- a/leo/commands/build.rs
+++ b/leo/commands/build.rs
@@ -56,9 +56,9 @@ pub struct BuildOptions {
 impl Default for BuildOptions {
     fn default() -> Self {
         Self {
-            disable_constant_folding: true,
-            disable_code_elimination: true,
-            disable_all_optimizations: true,
+            disable_constant_folding: false,
+            disable_code_elimination: false,
+            disable_all_optimizations: false,
             enable_all_ast_snapshots: false,
             enable_initial_ast_snapshot: false,
             enable_canonicalized_ast_snapshot: false,
@@ -69,10 +69,10 @@ impl Default for BuildOptions {
 
 impl From<BuildOptions> for CompilerOptions {
     fn from(options: BuildOptions) -> Self {
-        if !options.disable_all_optimizations {
+        if options.disable_all_optimizations {
             CompilerOptions {
-                constant_folding_enabled: true,
-                dead_code_elimination_enabled: true,
+                constant_folding_enabled: false,
+                dead_code_elimination_enabled: false,
             }
         } else {
             CompilerOptions {
@@ -123,7 +123,14 @@ impl Command for Build {
 
     fn apply(self, context: Context, _: Self::Input) -> Result<Self::Output> {
         let path = context.dir()?;
-        let package_name = context.manifest()?.get_package_name();
+        let manifest = context.manifest().map_err(|_| CliError::manifest_file_not_found())?;
+        let package_name = manifest.get_package_name();
+        let imports_map = manifest.get_imports_map().unwrap_or_default();
+
+        // Error out if there are dependencies but no lock file found.
+        if !imports_map.is_empty() && !context.lock_file_exists()? {
+            return Err(CliError::dependencies_are_not_installed().into());
+        }
 
         // Sanitize the package path to the root directory.
         let mut package_path = path.clone();
@@ -159,6 +166,12 @@ impl Command for Build {
         // Log compilation of files to console
         tracing::info!("Compiling main program... ({:?})", main_file_path);
 
+        let imports_map = if context.lock_file_exists()? {
+            context.lock_file()?.to_import_map()
+        } else {
+            Default::default()
+        };
+
         // Load the program at `main_file_path`
         let program = Compiler::<Fq, EdwardsGroupType>::parse_program_with_input(
             package_name.clone(),
@@ -170,6 +183,7 @@ impl Command for Build {
             &state_path,
             thread_leaked_context(),
             Some(self.compiler_options.clone().into()),
+            imports_map,
             Some(self.compiler_options.into()),
         )?;
 
diff --git a/leo/commands/clean.rs b/leo/commands/clean.rs
index 54ba1344f3..4b1f85b67b 100644
--- a/leo/commands/clean.rs
+++ b/leo/commands/clean.rs
@@ -17,7 +17,9 @@
 use crate::{commands::Command, context::Context};
 use leo_compiler::OutputFile;
 use leo_errors::Result;
-use leo_package::outputs::{ChecksumFile, CircuitFile, ProofFile, ProvingKeyFile, VerificationKeyFile};
+use leo_package::outputs::{
+    ChecksumFile, CircuitFile, ProofFile, ProvingKeyFile, Snapshot, SnapshotFile, VerificationKeyFile,
+};
 
 use structopt::StructOpt;
 use tracing::span::Span;
@@ -61,6 +63,11 @@ impl Command for Clean {
         // Remove the proof from the output directory
         ProofFile::new(&package_name).remove(&path)?;
 
+        // Remove AST snapshots from the output directory
+        SnapshotFile::new(&package_name, Snapshot::Initial).remove(&path)?;
+        SnapshotFile::new(&package_name, Snapshot::TypeInference).remove(&path)?;
+        SnapshotFile::new(&package_name, Snapshot::Canonicalization).remove(&path)?;
+
         Ok(())
     }
 }
diff --git a/leo/commands/package/add.rs b/leo/commands/package/add.rs
index e5e7a4635b..0a27de6b97 100644
--- a/leo/commands/package/add.rs
+++ b/leo/commands/package/add.rs
@@ -14,6 +14,10 @@
 // You should have received a copy of the GNU General Public License
 // along with the Leo library. If not, see <https://www.gnu.org/licenses/>.
 
+// !!!!!!!!!!!!!!!!!!!!!!!!!!!!
+// COMMAND TEMPORARILY DISABLED
+// !!!!!!!!!!!!!!!!!!!!!!!!!!!!
+
 use crate::{api::Fetch, commands::Command, context::Context};
 use leo_errors::{CliError, Result};
 use leo_package::imports::{ImportsDirectory, IMPORTS_DIRECTORY_NAME};
@@ -21,6 +25,7 @@ use leo_package::imports::{ImportsDirectory, IMPORTS_DIRECTORY_NAME};
 use std::{
     fs::{create_dir_all, File},
     io::{Read, Write},
+    path::PathBuf,
 };
 use structopt::StructOpt;
 use tracing::Span;
@@ -76,7 +81,7 @@ impl Add {
 
 impl Command for Add {
     type Input = ();
-    type Output = ();
+    type Output = PathBuf;
 
     fn log_span(&self) -> Span {
         tracing::span!(tracing::Level::INFO, "Adding")
@@ -88,18 +93,20 @@ impl Command for Add {
 
     fn apply(self, context: Context, _: Self::Input) -> Result<Self::Output> {
         // Check that a manifest exists for the current package.
-        context.manifest().map_err(|_| CliError::mainifest_file_not_found())?;
+        context.manifest().map_err(|_| CliError::manifest_file_not_found())?;
 
         let (author, package_name) = self
             .try_read_arguments()
             .map_err(CliError::cli_bytes_conversion_error)?;
 
+        tracing::info!("Package: {}/{}", &author, &package_name);
+
         // Attempt to fetch the package.
         let reader = {
             let fetch = Fetch {
-                author,
+                author: author.clone(),
                 package_name: package_name.clone(),
-                version: self.version,
+                version: self.version.clone(),
             };
             let bytes = context
                 .api
@@ -114,7 +121,14 @@ impl Command for Add {
         {
             ImportsDirectory::create(&path)?;
             path.push(IMPORTS_DIRECTORY_NAME);
-            path.push(package_name);
+
+            // Dumb compatibility hack.
+            // TODO: Remove once `leo add` functionality is discussed.
+            if self.version.is_some() {
+                path.push(format!("{}-{}@{}", author, package_name, self.version.unwrap()));
+            } else {
+                path.push(package_name.clone());
+            }
             create_dir_all(&path).map_err(CliError::cli_io_error)?;
         };
 
@@ -142,8 +156,8 @@ impl Command for Add {
             }
         }
 
-        tracing::info!("Successfully added a package");
+        tracing::info!("Successfully added package {}/{}", author, package_name);
 
-        Ok(())
+        Ok(path)
     }
 }
diff --git a/leo/commands/package/fetch.rs b/leo/commands/package/fetch.rs
new file mode 100644
index 0000000000..07c29fb579
--- /dev/null
+++ b/leo/commands/package/fetch.rs
@@ -0,0 +1,144 @@
+// Copyright (C) 2019-2021 Aleo Systems Inc.
+// This file is part of the Leo library.
+
+// The Leo library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// The Leo library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with the Leo library. If not, see <https://www.gnu.org/licenses/>.
+
+use crate::{
+    commands::{package::Add, Command},
+    context::{create_context, Context},
+};
+
+use leo_package::root::{
+    lock_file::{LockFile, Package},
+    Dependency,
+};
+
+use indexmap::{set::IndexSet, IndexMap};
+use leo_errors::{CliError, Result};
+use structopt::StructOpt;
+use tracing::span::Span;
+
+/// Pull dependencies specified in Leo toml.
+#[derive(StructOpt, Debug)]
+#[structopt(setting = structopt::clap::AppSettings::ColoredHelp)]
+pub struct Fetch {}
+
+impl Command for Fetch {
+    /// Names of dependencies in the current branch of a dependency tree.
+    type Input = IndexSet<String>;
+    type Output = ();
+
+    fn log_span(&self) -> Span {
+        tracing::span!(tracing::Level::INFO, "Fetching")
+    }
+
+    fn prelude(&self, context: Context) -> Result<Self::Input> {
+        let package_name = context.manifest()?.get_package_name();
+
+        let mut set = IndexSet::new();
+        set.insert(package_name);
+
+        Ok(set)
+    }
+
+    fn apply(self, context: Context, tree: Self::Input) -> Result<Self::Output> {
+        let dependencies = context
+            .manifest()
+            .map_err(|_| CliError::manifest_file_not_found())?
+            .get_package_dependencies();
+
+        // If program has no dependencies in the Leo.toml, exit with success.
+        let dependencies = match dependencies {
+            Some(dependencies) => dependencies,
+            None => return Ok(()),
+        };
+
+        if dependencies.is_empty() {
+            return Ok(());
+        }
+
+        let mut lock_file = LockFile::new();
+        self.add_dependencies(context.clone(), tree, &mut lock_file, dependencies)?;
+        lock_file.write_to(&context.dir()?)?;
+
+        Ok(())
+    }
+}
+
+impl Fetch {
+    /// Pulls dependencies and fills in the lock file. Also checks for
+    /// recursive dependencies with dependency tree.
+    fn add_dependencies(
+        &self,
+        context: Context,
+        mut tree: IndexSet<String>,
+        lock_file: &mut LockFile,
+        dependencies: IndexMap<String, Dependency>,
+    ) -> Result<()> {
+        // Go through each dependency in Leo.toml and add it to the imports.
+        // While adding, pull dependencies of this package as well and check for recursion.
+        for (import_name, dependency) in dependencies.into_iter() {
+            let mut package = Package::from(&dependency);
+            package.import_name = Some(import_name);
+
+            // Pull the dependency first.
+            let path = Add::new(
+                None,
+                Some(package.author.clone()),
+                Some(package.name.clone()),
+                Some(package.version.clone()),
+            )
+            .apply(context.clone(), ())?;
+
+            // Try inserting a new dependency to the branch. If not inserted,
+            // then fail because this dependency was added on a higher level.
+            if !tree.insert(package.name.clone()) {
+                // Pretty format for the message - show dependency structure.
+                let mut message: Vec<String> = tree
+                    .into_iter()
+                    .enumerate()
+                    .map(|(i, val)| format!("{}└─{}", " ".repeat(i * 2), val))
+                    .collect();
+
+                message.push(format!("{}└─{} (FAILURE)", " ".repeat(message.len() * 2), package.name));
+
+                return Err(CliError::recursive_dependency_found(message.join("\n")).into());
+            }
+
+            // Check imported dependency's dependencies.
+            let imported_dependencies = create_context(path, None)?
+                .manifest()
+                .map_err(|_| CliError::unable_to_read_imported_dependency_manifest())?
+                .get_package_dependencies();
+
+            if let Some(dependencies) = imported_dependencies {
+                if !dependencies.is_empty() {
+                    // Fill in the lock file with imported dependency and information about its dependencies.
+                    package.add_dependencies(&dependencies);
+                    lock_file.add_package(package);
+
+                    // Recursively call this method for imported program.
+                    self.add_dependencies(context.clone(), tree.clone(), lock_file, dependencies)?;
+
+                    continue;
+                }
+            }
+
+            // If there are no dependencies for the new import, add a single record.
+            lock_file.add_package(package);
+        }
+
+        Ok(())
+    }
+}
diff --git a/leo/commands/package/mod.rs b/leo/commands/package/mod.rs
index ce8c637f59..cd1c3ccd69 100644
--- a/leo/commands/package/mod.rs
+++ b/leo/commands/package/mod.rs
@@ -20,6 +20,9 @@ pub use add::Add;
 pub mod clone;
 pub use clone::Clone;
 
+pub mod fetch;
+pub use fetch::Fetch;
+
 pub mod login;
 pub use login::Login;
 
diff --git a/leo/commands/package/remove.rs b/leo/commands/package/remove.rs
index 6d8ba229f9..06907265ac 100644
--- a/leo/commands/package/remove.rs
+++ b/leo/commands/package/remove.rs
@@ -14,6 +14,10 @@
 // You should have received a copy of the GNU General Public License
 // along with the Leo library. If not, see <https://www.gnu.org/licenses/>.
 
+// !!!!!!!!!!!!!!!!!!!!!!!!!!!!
+// COMMAND TEMPORARILY DISABLED
+// !!!!!!!!!!!!!!!!!!!!!!!!!!!!
+
 use crate::{commands::Command, context::Context};
 use leo_errors::Result;
 use leo_package::LeoPackage;
diff --git a/leo/commands/test.rs b/leo/commands/test.rs
index 773ac53488..aa7cef20e3 100644
--- a/leo/commands/test.rs
+++ b/leo/commands/test.rs
@@ -109,6 +109,7 @@ impl Command for Test {
                 output_directory.clone(),
                 thread_leaked_context(),
                 Some(self.compiler_options.clone().into()),
+                std::collections::HashMap::new(),
                 Some(self.compiler_options.clone().into()),
             )?;
 
diff --git a/leo/context.rs b/leo/context.rs
index d7c7beaeed..49bdf54d1e 100644
--- a/leo/context.rs
+++ b/leo/context.rs
@@ -16,7 +16,7 @@
 
 use crate::{api::Api, config};
 use leo_errors::{CliError, Result};
-use leo_package::root::Manifest;
+use leo_package::root::{LockFile, Manifest};
 
 use std::{convert::TryFrom, env::current_dir, path::PathBuf};
 
@@ -41,10 +41,20 @@ impl Context {
         }
     }
 
-    /// Get package manifest for current context
+    /// Get package manifest for current context.
     pub fn manifest(&self) -> Result<Manifest> {
         Ok(Manifest::try_from(self.dir()?.as_path())?)
     }
+
+    /// Get lock file for current context.
+    pub fn lock_file(&self) -> Result<LockFile> {
+        Ok(LockFile::try_from(self.dir()?.as_path())?)
+    }
+
+    /// Check if lock file exists.
+    pub fn lock_file_exists(&self) -> Result<bool> {
+        Ok(LockFile::exists_at(&self.dir()?))
+    }
 }
 
 /// Create a new context for the current directory.
diff --git a/leo/main.rs b/leo/main.rs
index 38a25f722c..eb500655af 100644
--- a/leo/main.rs
+++ b/leo/main.rs
@@ -22,7 +22,7 @@ pub mod logger;
 pub mod updater;
 
 use commands::{
-    package::{Add, Clone, Login, Logout, Publish, Remove},
+    package::{Clone, Fetch, Login, Logout, Publish},
     Build, Clean, Command, Deploy, Init, Lint, New, Prove, Run, Setup, Test, Update, Watch,
 };
 use leo_errors::Result;
@@ -119,10 +119,15 @@ enum CommandOpts {
         command: Test,
     },
 
-    #[structopt(about = "Import a package from the Aleo Package Manager")]
-    Add {
+    // #[structopt(about = "Import a package from the Aleo Package Manager")]
+    // Add {
+    //     #[structopt(flatten)]
+    //     command: Add,
+    // },
+    #[structopt(about = "Pull dependencies from Aleo Package Manager")]
+    Fetch {
         #[structopt(flatten)]
-        command: Add,
+        command: Fetch,
     },
 
     #[structopt(about = "Clone a package from the Aleo Package Manager")]
@@ -149,12 +154,11 @@ enum CommandOpts {
         command: Publish,
     },
 
-    #[structopt(about = "Uninstall a package from the current package")]
-    Remove {
-        #[structopt(flatten)]
-        command: Remove,
-    },
-
+    // #[structopt(about = "Uninstall a package from the current package")]
+    // Remove {
+    //     #[structopt(flatten)]
+    //     command: Remove,
+    // },
     #[structopt(about = "Lints the Leo files in the package (*)")]
     Lint {
         #[structopt(flatten)]
@@ -204,13 +208,13 @@ fn run_with_args(opt: Opt) -> Result<()> {
         CommandOpts::Watch { command } => command.try_execute(context),
         CommandOpts::Update { command } => command.try_execute(context),
 
-        CommandOpts::Add { command } => command.try_execute(context),
+        // CommandOpts::Add { command } => command.try_execute(context),
+        CommandOpts::Fetch { command } => command.try_execute(context),
         CommandOpts::Clone { command } => command.try_execute(context),
         CommandOpts::Login { command } => command.try_execute(context),
         CommandOpts::Logout { command } => command.try_execute(context),
         CommandOpts::Publish { command } => command.try_execute(context),
-        CommandOpts::Remove { command } => command.try_execute(context),
-
+        // CommandOpts::Remove { command } => command.try_execute(context),
         CommandOpts::Lint { command } => command.try_execute(context),
         CommandOpts::Deploy { command } => command.try_execute(context),
     }
@@ -231,6 +235,7 @@ mod cli_tests {
     use crate::{run_with_args, Opt};
     use leo_errors::{CliError, Result};
 
+    use snarkvm_utilities::Write;
     use std::path::PathBuf;
     use structopt::StructOpt;
     use test_dir::{DirBuilder, FileType, TestDir};
@@ -347,6 +352,7 @@ mod cli_tests {
     }
 
     #[test]
+    #[ignore]
     fn test_import() {
         let dir = testdir("test");
         let path = dir.path("test");
@@ -388,4 +394,32 @@ mod cli_tests {
         assert!(run_cmd("leo test -f examples/silly-sudoku/src/lib.leo", path).is_ok());
         assert!(run_cmd("leo test -f examples/silly-sudoku/src/main.leo", path).is_ok());
     }
+
+    #[test]
+    fn test_install() {
+        let dir = testdir("test");
+        let path = dir.path("test");
+
+        assert!(run_cmd("leo new install", &Some(path.clone())).is_ok());
+
+        let install_path = &Some(path.join("install"));
+
+        let mut file = std::fs::OpenOptions::new()
+            .write(true)
+            .append(true)
+            .open(path.join("install/Leo.toml"))
+            .unwrap();
+
+        assert!(run_cmd("leo fetch", install_path).is_ok());
+        assert!(file
+            .write_all(
+                br#"
+            sudoku = {author = "justice-league", package = "u8u32", version = "0.1.0"}
+        "#
+            )
+            .is_ok());
+
+        assert!(run_cmd("leo fetch", install_path).is_ok());
+        assert!(run_cmd("leo build", install_path).is_ok());
+    }
 }
diff --git a/package/Cargo.toml b/package/Cargo.toml
index 2e79acfb60..48a1ca3f2c 100644
--- a/package/Cargo.toml
+++ b/package/Cargo.toml
@@ -21,6 +21,10 @@ edition = "2018"
 path = "../errors"
 version = "1.5.3"
 
+[dependencies.indexmap]
+version = "1.7"
+features = ["serde"]
+
 [dependencies.serde]
 version = "1.0"
 features = [ "derive" ]
diff --git a/package/README.md b/package/README.md
index dc16bf3844..aa62f7c2a7 100644
--- a/package/README.md
+++ b/package/README.md
@@ -4,6 +4,27 @@
 [![Authors](https://img.shields.io/badge/authors-Aleo-orange.svg)](../AUTHORS)
 [![License](https://img.shields.io/badge/License-GPLv3-blue.svg)](./LICENSE.md)
 
-This module defines the structure of a Leo project package.
+## Description
 
-Each file in this directory mirrors a corresponding file generated in a new Leo project package.
+This module defines the structure of a Leo project package. And describes behavior of package internals, such
+as Leo Manifest (Leo.toml), Lock File (Leo.lock), source files and imports. 
+
+Mainly used by Leo binary.
+
+## Structure
+
+Each directory in this crate mirrors a corresponding file generated in a new Leo project package:
+
+```bash
+package/src
+├── errors  # crate level error definitions
+├── imports # program imports management
+├── inputs  # program inputs directory
+├── outputs # program outputs directory
+├── root    # program root: Leo.toml, Leo.lock 
+└── source  # source files directory
+```
+
+## Testing
+
+Package features functional tests in the `tests/` directory.
diff --git a/package/src/outputs/ast_snapshot.rs b/package/src/outputs/ast_snapshot.rs
new file mode 100644
index 0000000000..e48a9977ec
--- /dev/null
+++ b/package/src/outputs/ast_snapshot.rs
@@ -0,0 +1,103 @@
+// Copyright (C) 2019-2021 Aleo Systems Inc.
+// This file is part of the Leo library.
+
+// The Leo library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// The Leo library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with the Leo library. If not, see <https://www.gnu.org/licenses/>.
+
+//! The serialized circuit output file.
+
+use crate::outputs::OUTPUTS_DIRECTORY_NAME;
+use leo_errors::{PackageError, Result};
+
+use serde::Deserialize;
+use std::{borrow::Cow, fmt, fs, path::Path};
+
+/// Enum to handle all 3 types of snapshots.
+#[derive(Deserialize)]
+pub enum Snapshot {
+    Initial,
+    TypeInference,
+    Canonicalization,
+}
+
+impl fmt::Display for Snapshot {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(
+            f,
+            "{}",
+            match self {
+                Self::Initial => "initial_ast",
+                Self::TypeInference => "type_inferenced_ast",
+                Self::Canonicalization => "canonicalization_ast",
+            }
+        )
+    }
+}
+
+pub static AST_SNAPSHOT_FILE_EXTENSION: &str = ".json";
+
+/// Generic Snapshot file wrapper. Each package can have up to 3
+/// different snapshots: initial_ast, canonicalization_ast and type_inferenced_ast;
+#[derive(Deserialize)]
+pub struct SnapshotFile {
+    pub package_name: String,
+    pub snapshot: Snapshot,
+}
+
+impl SnapshotFile {
+    pub fn new(package_name: &str, snapshot: Snapshot) -> Self {
+        Self {
+            package_name: package_name.to_string(),
+            snapshot,
+        }
+    }
+
+    pub fn exists_at(&self, path: &Path) -> bool {
+        let path = self.snapshot_file_path(path);
+        path.exists()
+    }
+
+    /// Reads the serialized circuit from the given file path if it exists.
+    pub fn read_from(&self, path: &Path) -> Result<String> {
+        let path = self.snapshot_file_path(path);
+
+        let result =
+            fs::read_to_string(&path).map_err(|_| PackageError::failed_to_read_snapshot_file(path.into_owned()))?;
+
+        Ok(result)
+    }
+
+    /// Removes the serialized circuit at the given path if it exists. Returns `true` on success,
+    /// `false` if the file doesn't exist, and `Error` if the file system fails during operation.
+    pub fn remove(&self, path: &Path) -> Result<bool> {
+        let path = self.snapshot_file_path(path);
+        if !path.exists() {
+            return Ok(false);
+        }
+
+        fs::remove_file(&path).map_err(|_| PackageError::failed_to_remove_snapshot_file(path.into_owned()))?;
+        Ok(true)
+    }
+
+    fn snapshot_file_path<'a>(&self, path: &'a Path) -> Cow<'a, Path> {
+        let mut path = Cow::from(path);
+        if path.is_dir() {
+            if !path.ends_with(OUTPUTS_DIRECTORY_NAME) {
+                path.to_mut().push(OUTPUTS_DIRECTORY_NAME);
+            }
+            path.to_mut()
+                .push(format!("{}{}", self.snapshot, AST_SNAPSHOT_FILE_EXTENSION));
+        }
+        path
+    }
+}
diff --git a/package/src/outputs/mod.rs b/package/src/outputs/mod.rs
index f4c02b8f18..d448130673 100644
--- a/package/src/outputs/mod.rs
+++ b/package/src/outputs/mod.rs
@@ -14,6 +14,9 @@
 // You should have received a copy of the GNU General Public License
 // along with the Leo library. If not, see <https://www.gnu.org/licenses/>.
 
+pub mod ast_snapshot;
+pub use self::ast_snapshot::*;
+
 pub mod circuit;
 pub use self::circuit::*;
 
diff --git a/package/src/root/lock_file.rs b/package/src/root/lock_file.rs
new file mode 100644
index 0000000000..82b00c5f8f
--- /dev/null
+++ b/package/src/root/lock_file.rs
@@ -0,0 +1,160 @@
+// Copyright (C) 2019-2021 Aleo Systems Inc.
+// This file is part of the Leo library.
+
+// The Leo library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// The Leo library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with the Leo library. If not, see <https://www.gnu.org/licenses/>.
+
+use crate::root::Dependency;
+use leo_errors::{PackageError, Result};
+
+use indexmap::IndexMap;
+use serde::{Deserialize, Serialize};
+use std::{
+    borrow::Cow,
+    collections::HashMap,
+    convert::TryFrom,
+    fmt::{self, Display},
+    fs::File,
+    io::{Read, Write},
+    path::Path,
+};
+
+pub const LOCKFILE_FILENAME: &str = "Leo.lock";
+
+/// Lock-file struct, contains all information about imported dependencies
+/// and their relationships.
+#[derive(Default, Debug, Clone, Serialize, Deserialize)]
+pub struct LockFile {
+    pub package: Vec<Package>,
+}
+
+/// Single dependency record.
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct Package {
+    pub name: String,
+    pub version: String,
+    pub author: String,
+    pub import_name: Option<String>,
+    #[serde(skip_serializing_if = "IndexMap::is_empty", default)]
+    pub dependencies: IndexMap<String, String>,
+}
+
+impl LockFile {
+    pub fn new() -> Self {
+        LockFile { package: vec![] }
+    }
+
+    /// Check if LockFile exists in a directory.
+    pub fn exists_at(path: &Path) -> bool {
+        let mut path = Cow::from(path);
+        if path.is_dir() {
+            path.to_mut().push(LOCKFILE_FILENAME);
+        }
+        path.exists()
+    }
+
+    /// Add Package record to the lock file. Chainable.
+    pub fn add_package(&mut self, package: Package) -> &mut Self {
+        self.package.push(package);
+        self
+    }
+
+    /// Print LockFile as toml.
+    pub fn to_string(&self) -> Result<String> {
+        Ok(toml::to_string(self).map_err(PackageError::failed_to_serialize_lock_file)?)
+    }
+
+    /// Form a HashMap of kind:
+    /// ``` imported_name => package_name ```
+    /// for all imported packages.
+    pub fn to_import_map(&self) -> HashMap<String, String> {
+        let mut result = HashMap::new();
+        for package in self.package.iter() {
+            match &package.import_name {
+                Some(name) => result.insert(name.clone(), package.to_string()),
+                None => result.insert(package.name.clone(), package.to_string()),
+            };
+        }
+
+        result
+    }
+
+    /// Write Leo.lock to the given location.
+    pub fn write_to(self, path: &Path) -> Result<()> {
+        let mut path = Cow::from(path);
+        if path.is_dir() {
+            path.to_mut().push(LOCKFILE_FILENAME);
+        }
+
+        File::create(&path)
+            .map_err(|error| PackageError::failed_to_create_lock_file(LOCKFILE_FILENAME, error))?
+            .write_all(self.to_string()?.as_bytes())
+            .map_err(|error| PackageError::failed_to_write_lock_file(LOCKFILE_FILENAME, error))?;
+
+        Ok(())
+    }
+}
+
+impl TryFrom<&Path> for LockFile {
+    type Error = PackageError;
+
+    fn try_from(path: &Path) -> Result<Self, Self::Error> {
+        let mut path = Cow::from(path);
+        if path.is_dir() {
+            path.to_mut().push(LOCKFILE_FILENAME);
+        }
+
+        let mut file = File::open(path.clone())
+            .map_err(|error| PackageError::failed_to_open_lock_file(LOCKFILE_FILENAME, error))?;
+        let size = file
+            .metadata()
+            .map_err(|error| PackageError::failed_to_get_lock_file_metadata(LOCKFILE_FILENAME, error))?
+            .len() as usize;
+
+        let mut buffer = String::with_capacity(size);
+        file.read_to_string(&mut buffer)
+            .map_err(|error| PackageError::failed_to_read_lock_file(LOCKFILE_FILENAME, error))?;
+
+        toml::from_str(&buffer).map_err(|error| PackageError::failed_to_parse_lock_file(LOCKFILE_FILENAME, error))
+    }
+}
+
+impl Package {
+    /// Fill dependencies from Leo Manifest data.
+    pub fn add_dependencies(&mut self, dependencies: &IndexMap<String, Dependency>) {
+        for (import_name, dependency) in dependencies.iter() {
+            self.dependencies
+                .insert(import_name.clone(), Package::from(dependency).to_string());
+        }
+    }
+}
+
+impl Display for Package {
+    /// Form an path identifier for a package. It is the path under which package is stored
+    /// inside the `imports/` directory.
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{}-{}@{}", self.author, self.name, self.version)
+    }
+}
+
+impl From<&Dependency> for Package {
+    fn from(dependency: &Dependency) -> Package {
+        Package {
+            name: dependency.package.clone(),
+            author: dependency.author.clone(),
+            version: dependency.version.clone(),
+            dependencies: Default::default(),
+            import_name: None,
+        }
+    }
+}
diff --git a/package/src/root/manifest.rs b/package/src/root/manifest.rs
index fc08033cff..c72624bc96 100644
--- a/package/src/root/manifest.rs
+++ b/package/src/root/manifest.rs
@@ -17,9 +17,11 @@
 use crate::package::Package;
 use leo_errors::{PackageError, Result};
 
+use indexmap::IndexMap;
 use serde::Deserialize;
 use std::{
     borrow::Cow,
+    collections::HashMap,
     convert::TryFrom,
     fs::File,
     io::{Read, Write},
@@ -34,10 +36,18 @@ pub struct Remote {
     pub author: String,
 }
 
+#[derive(Clone, Debug, Deserialize)]
+pub struct Dependency {
+    pub author: String,
+    pub version: String,
+    pub package: String,
+}
+
 #[derive(Deserialize)]
 pub struct Manifest {
     pub project: Package,
     pub remote: Option<Remote>,
+    pub dependencies: Option<IndexMap<String, Dependency>>,
 }
 
 impl Manifest {
@@ -45,6 +55,7 @@ impl Manifest {
         Ok(Self {
             project: Package::new(package_name)?,
             remote: author.map(|author| Remote { author }),
+            dependencies: Some(IndexMap::<String, Dependency>::new()),
         })
     }
 
@@ -72,6 +83,27 @@ impl Manifest {
         self.project.description.clone()
     }
 
+    pub fn get_package_dependencies(&self) -> Option<IndexMap<String, Dependency>> {
+        self.dependencies.clone()
+    }
+
+    /// Get HashMap of kind:
+    ///     import name => import directory
+    /// Which then used in AST/ASG to resolve import paths.
+    pub fn get_imports_map(&self) -> Option<HashMap<String, String>> {
+        self.dependencies.clone().map(|dependencies| {
+            dependencies
+                .into_iter()
+                .map(|(name, dependency)| {
+                    (
+                        name,
+                        format!("{}-{}@{}", dependency.author, dependency.package, dependency.version),
+                    )
+                })
+                .collect()
+        })
+    }
+
     pub fn get_package_license(&self) -> Option<String> {
         self.project.license.clone()
     }
@@ -109,6 +141,14 @@ license = "MIT"
 
 [remote]
 author = "{author}" # Add your Aleo Package Manager username or team name.
+
+[target]
+curve = "bls12_377"
+proving_system = "groth16"
+
+[dependencies]
+# Define dependencies here in format:
+# name = {{ package = "package-name", author = "author", version = "version" }}
 "#,
             name = self.project.name,
             author = author
diff --git a/package/src/root/mod.rs b/package/src/root/mod.rs
index 6b90439ca1..3a6c2b95a0 100644
--- a/package/src/root/mod.rs
+++ b/package/src/root/mod.rs
@@ -17,6 +17,9 @@
 pub mod gitignore;
 pub use self::gitignore::*;
 
+pub mod lock_file;
+pub use self::lock_file::*;
+
 pub mod manifest;
 pub use self::manifest::*;
 
diff --git a/package/src/root/zip.rs b/package/src/root/zip.rs
index bf95a9aaf7..b40811ae47 100644
--- a/package/src/root/zip.rs
+++ b/package/src/root/zip.rs
@@ -156,10 +156,8 @@ impl ZipFile {
 
 /// Check if the file path should be included in the package zip file.
 fn is_included(path: &Path) -> bool {
-    // excluded directories: `output`, `imports`
-    if path.ends_with(OUTPUTS_DIRECTORY_NAME.trim_end_matches('/'))
-        | path.ends_with(IMPORTS_DIRECTORY_NAME.trim_end_matches('/'))
-    {
+    // DO NOT include `imports` and `outputs` directories.
+    if path.starts_with(IMPORTS_DIRECTORY_NAME) || path.starts_with(OUTPUTS_DIRECTORY_NAME) {
         return false;
     }
 
diff --git a/test-framework/.gitignore b/test-framework/.gitignore
new file mode 100644
index 0000000000..ef4cd56d9d
--- /dev/null
+++ b/test-framework/.gitignore
@@ -0,0 +1 @@
+!src/bin
diff --git a/test-framework/Cargo.toml b/test-framework/Cargo.toml
index fa5440225b..b471480d24 100644
--- a/test-framework/Cargo.toml
+++ b/test-framework/Cargo.toml
@@ -26,3 +26,28 @@ version = "1.0"
 
 [dependencies.serde_yaml]
 version = "0.8"
+
+# List of dependencies for tgc binary;
+
+[dependencies.leo-ast]
+path = "../ast"
+version = "1.5.2"
+
+[dependencies.leo-parser]
+path = "../parser"
+version = "1.5.2"
+
+[dependencies.leo-imports]
+path = "../imports"
+version = "1.5.2"
+
+[dependencies.leo-asg]
+path = "../asg"
+version = "1.5.2"
+
+[dependencies.leo-compiler]
+path = "../compiler"
+version = "1.5.2"
+
+[dependencies.structopt]
+version = "0.3"
diff --git a/test-framework/src/bin/tgc.rs b/test-framework/src/bin/tgc.rs
new file mode 100644
index 0000000000..fc62bcb413
--- /dev/null
+++ b/test-framework/src/bin/tgc.rs
@@ -0,0 +1,161 @@
+// Copyright (C) 2019-2021 Aleo Systems Inc.
+// This file is part of the Leo library.
+
+// The Leo library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// The Leo library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with the Leo library. If not, see <https://www.gnu.org/licenses/>.
+
+use leo_asg::Asg;
+use leo_compiler::{compiler::thread_leaked_context, TypeInferencePhase};
+use leo_imports::ImportParser;
+use leo_test_framework::{
+    fetch::find_tests,
+    test::{extract_test_config, TestExpectationMode as Expectation},
+};
+
+use std::{error::Error, fs, path::PathBuf};
+use structopt::{clap::AppSettings, StructOpt};
+
+#[derive(StructOpt)]
+#[structopt(name = "ast-stages-generator", author = "The Aleo Team <hello@aleo.org>", setting = AppSettings::ColoredHelp)]
+struct Opt {
+    #[structopt(
+        short,
+        long,
+        help = "Path to the output folder (auto generated)",
+        default_value = "tmp/tgc"
+    )]
+    path: PathBuf,
+
+    #[structopt(short, long, help = "Run only for test that match pattern")]
+    filter: Option<String>,
+
+    #[structopt(short, long, help = "Skip tests matching pattern")]
+    skip: Option<Vec<String>>,
+}
+
+fn main() {
+    handle_error(run_with_args(Opt::from_args()));
+}
+
+fn run_with_args(opt: Opt) -> Result<(), Box<dyn Error>> {
+    // Variable that stores all the tests.
+    let mut tests = Vec::new();
+    let mut test_dir = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
+    test_dir.push("../tests/");
+    find_tests(&test_dir, &mut tests);
+
+    if !opt.path.exists() {
+        fs::create_dir_all(&opt.path)?;
+    }
+
+    // Prepare directory for placing results.
+    'main_loop: for (index, (path, text)) in tests.iter().enumerate() {
+        if let Some(config) = extract_test_config(text) {
+            // Skip namespaces that we don't need; also skip failure tests.
+            if config.namespace != "Compile" || config.expectation == Expectation::Fail {
+                continue;
+            }
+
+            let mut test_name = path
+                .split("tests/")
+                .last()
+                .unwrap()
+                .replace(std::path::MAIN_SEPARATOR, "_");
+
+            // Filter out the tests that do not match pattern, if pattern is set.
+            if let Some(filter) = &opt.filter {
+                if !test_name.contains(filter) {
+                    continue;
+                }
+            }
+
+            // If skip flag is used, don't run tests matching the pattern.
+            if let Some(skips) = &opt.skip {
+                for skip_pattern in skips {
+                    if test_name.contains(skip_pattern) {
+                        println!("Skipping: {} because of {}", test_name, skip_pattern);
+                        continue 'main_loop;
+                    }
+                }
+            }
+
+            test_name.push_str(&format!("_{}", index));
+
+            // Create directory for this file.
+            let mut target = PathBuf::from("tmp/tgc");
+            target.push(test_name);
+
+            if !target.exists() {
+                fs::create_dir_all(target.clone())?;
+            }
+
+            let cwd = config
+                .extra
+                .get("cwd")
+                .map(|val| {
+                    let mut cwd = PathBuf::from(path);
+                    cwd.pop();
+                    cwd.join(&val.as_str().unwrap())
+                })
+                .unwrap_or(PathBuf::from(path));
+
+            // Write all files into the directory.
+            let (initial, canonicalized, type_inferenced) = generate_asts(cwd, text)?;
+
+            target.push("initial_ast.json");
+            fs::write(target.clone(), initial)?;
+            target.pop();
+
+            target.push("canonicalization_ast.json");
+            fs::write(target.clone(), canonicalized)?;
+            target.pop();
+
+            target.push("type_inferenced_ast.json");
+            fs::write(target.clone(), type_inferenced)?;
+        }
+    }
+
+    Ok(())
+}
+
+/// Do what Compiler does - prepare 3 stages of AST: initial, canonicalized and type_inferenced
+fn generate_asts(path: PathBuf, text: &String) -> Result<(String, String, String), Box<dyn Error>> {
+    let mut ast = leo_parser::parse_ast(path.clone().into_os_string().into_string().unwrap(), text)?;
+    let initial = ast.to_json_string()?;
+
+    ast.canonicalize()?;
+    let canonicalized = ast.to_json_string()?;
+
+    let asg = Asg::new(
+        thread_leaked_context(),
+        &ast,
+        &mut ImportParser::new(path, Default::default()),
+    )?;
+
+    let type_inferenced = TypeInferencePhase::default()
+        .phase_ast(&ast.into_repr(), &asg.clone().into_repr())
+        .expect("Failed to produce type inference ast.")
+        .to_json_string()?;
+
+    Ok((initial, canonicalized, type_inferenced))
+}
+
+fn handle_error(res: Result<(), Box<dyn Error>>) {
+    match res {
+        Ok(_) => (),
+        Err(err) => {
+            eprintln!("Error: {}", err);
+            std::process::exit(1);
+        }
+    }
+}
diff --git a/test-framework/src/test.rs b/test-framework/src/test.rs
index ba129c39ab..dc3d0871f5 100644
--- a/test-framework/src/test.rs
+++ b/test-framework/src/test.rs
@@ -22,7 +22,7 @@ pub enum TestExpectationMode {
     Fail,
 }
 
-#[derive(serde::Serialize, serde::Deserialize)]
+#[derive(Debug, serde::Serialize, serde::Deserialize)]
 pub struct TestConfig {
     pub namespace: String,
     pub expectation: TestExpectationMode,
diff --git a/tests/compiler/statements/inputs/ternary_explicit_and_implicit.in b/tests/compiler/statements/inputs/ternary_explicit_and_implicit.in
new file mode 100644
index 0000000000..ef82e41af8
--- /dev/null
+++ b/tests/compiler/statements/inputs/ternary_explicit_and_implicit.in
@@ -0,0 +1,6 @@
+[main]
+x: u8 = 3;
+y: bool = true;
+
+[registers]
+a: u8 = 0;
\ No newline at end of file
diff --git a/tests/compiler/statements/ternary_explicit_and_implicit.leo b/tests/compiler/statements/ternary_explicit_and_implicit.leo
new file mode 100644
index 0000000000..d092bc2928
--- /dev/null
+++ b/tests/compiler/statements/ternary_explicit_and_implicit.leo
@@ -0,0 +1,9 @@
+/*
+namespace: Compile
+expectation: Pass
+input_file: inputs/ternary_explicit_and_implicit.in
+*/
+
+function main(x: u8, y: bool) -> u8 {
+    return y ? x : 2;
+}
diff --git a/tests/expectations/compiler/compiler/statements/assign_ternary.leo.out b/tests/expectations/compiler/compiler/statements/assign_ternary.leo.out
index 659cb7e8b4..9a641be8e0 100644
--- a/tests/expectations/compiler/compiler/statements/assign_ternary.leo.out
+++ b/tests/expectations/compiler/compiler/statements/assign_ternary.leo.out
@@ -2,4 +2,4 @@
 namespace: Compile
 expectation: Fail
 outputs:
-  - "Error [EASG0373021]: ternary sides had different types: left u32, right bool\n    --> compiler-test:4:13\n     |\n   4 |     let x = true ? x: true;\n     |             ^^^^^^^^^^^^^^"
+  - "Error [EASG0373026]: unexpected type, expected: 'u32', received: 'bool'\n    --> compiler-test:4:23\n     |\n   4 |     let x = true ? x: true;\n     |                       ^^^^"
diff --git a/tests/expectations/compiler/compiler/statements/ternary_explicit_and_implicit.leo.out b/tests/expectations/compiler/compiler/statements/ternary_explicit_and_implicit.leo.out
new file mode 100644
index 0000000000..2486b2233d
--- /dev/null
+++ b/tests/expectations/compiler/compiler/statements/ternary_explicit_and_implicit.leo.out
@@ -0,0 +1,21 @@
+---
+namespace: Compile
+expectation: Pass
+outputs:
+  - circuit:
+      num_public_variables: 0
+      num_private_variables: 25
+      num_constraints: 33
+      at: 00318b0f80b01b9e2d5a8154a536deb1545f04abd493a84071ef13dce1a9d92d
+      bt: 799ee4dc07f6320b8c3dcbd5ebdf6de8a4cf5af8b1edc45dbe5621c8a34fa38d
+      ct: 7b6ec9eacd65cfa47dfe7e73eb7c8fa4be804a9fb878fddf59fe5d06e7c04489
+    output:
+      - input_file: inputs/ternary_explicit_and_implicit.in
+        output:
+          registers:
+            a:
+              type: u8
+              value: "3"
+    initial_ast: 29b5d6188c9fe56eaee3024a900b92e4bd9445364551b85e924fe9a392bac35d
+    canonicalized_ast: 29b5d6188c9fe56eaee3024a900b92e4bd9445364551b85e924fe9a392bac35d
+    type_inferenced_ast: 02edde30e615a1ac3e87818daa6fca7e4d7c0c0c0a1a9d9159a0a4190c11ec8a
