diff --git a/.circleci/leo-add-remove.sh b/.circleci/leo-add-remove.sh
index de855a4ed8..8f4a2f6eed 100755
--- a/.circleci/leo-add-remove.sh
+++ b/.circleci/leo-add-remove.sh
@@ -1,6 +1,8 @@
 # leo add (w/o login) & remove
 
 $LEO new my-app && cd my-app || exit 1
-$LEO add howard/silly-sudoku
-$LEO remove silly-sudoku
+$LEO add justice-league/u8u32
+echo "import u8u32.u8_u32; function main() {}" > src/main.leo
+$LEO build
+$LEO remove u8u32
 $LEO clean
diff --git a/asg/src/error/mod.rs b/asg/src/error/mod.rs
index 8b3eccc9d1..5eb6f79c58 100644
--- a/asg/src/error/mod.rs
+++ b/asg/src/error/mod.rs
@@ -224,7 +224,6 @@ impl AsgConvertError {
     }
 
     pub fn unexpected_type(expected: &str, received: Option<&str>, span: &Span) -> Self {
-        // panic!(format!("unexpected type, expected: '{}', received: '{}'", expected, received.unwrap_or("unknown")));
         Self::new_from_span(
             format!(
                 "unexpected type, expected: '{}', received: '{}'",
diff --git a/asg/src/expression/tuple_init.rs b/asg/src/expression/tuple_init.rs
index 195b4fe4af..2dc40db252 100644
--- a/asg/src/expression/tuple_init.rs
+++ b/asg/src/expression/tuple_init.rs
@@ -94,7 +94,9 @@ impl<'a> FromAst<'a, leo_ast::TupleInitExpression> for TupleInitExpression<'a> {
         };
 
         if let Some(tuple_types) = tuple_types.as_ref() {
-            if tuple_types.len() != value.elements.len() {
+            // Expected type can be equal or less than actual size of a tuple.
+            // Size of expected tuple can be based on accessed index.
+            if tuple_types.len() > value.elements.len() {
                 return Err(AsgConvertError::unexpected_type(
                     &*format!("tuple of length {}", tuple_types.len()),
                     Some(&*format!("tuple of length {}", value.elements.len())),
diff --git a/imports/src/errors/import_parser.rs b/imports/src/errors/import_parser.rs
index 0c04467403..c91d06a006 100644
--- a/imports/src/errors/import_parser.rs
+++ b/imports/src/errors/import_parser.rs
@@ -94,10 +94,10 @@ impl ImportParserError {
     }
 
     ///
-    /// Failed to find a library file for the current package.
+    /// Failed to find a main file for the current package.
     ///
-    pub fn expected_lib_file(entry: String, span: &Span) -> Self {
-        let message = format!("Expected library file `{}`.", entry,);
+    pub fn expected_main_file(entry: String, span: &Span) -> Self {
+        let message = format!("Expected main file at `{}`.", entry,);
 
         Self::new_from_span(message, span)
     }
diff --git a/imports/src/parser/parse_package.rs b/imports/src/parser/parse_package.rs
index d7252844e7..a14f12a2cc 100644
--- a/imports/src/parser/parse_package.rs
+++ b/imports/src/parser/parse_package.rs
@@ -72,6 +72,7 @@ impl<'a> ImportParser<'a> {
 
         // Search for package name in `imports` directory
         let mut imports_directory = path.clone();
+        imports_directory.pop(); // path leads to src/ folder, imports is one level below
         imports_directory.push(IMPORTS_DIRECTORY_NAME);
 
         // Read from local `src` directory or the current path
diff --git a/imports/src/parser/parse_symbol.rs b/imports/src/parser/parse_symbol.rs
index e37e89a9f2..3a09506546 100644
--- a/imports/src/parser/parse_symbol.rs
+++ b/imports/src/parser/parse_symbol.rs
@@ -19,7 +19,7 @@ use leo_ast::{Program, Span};
 
 use std::fs::DirEntry;
 
-static LIBRARY_FILE: &str = "src/lib.leo";
+static MAIN_FILE: &str = "src/main.leo";
 
 impl<'a> ImportParser<'a> {
     ///
@@ -39,10 +39,10 @@ impl<'a> ImportParser<'a> {
 
         let mut file_path = package.path();
         if file_type.is_dir() {
-            file_path.push(LIBRARY_FILE);
+            file_path.push(MAIN_FILE);
 
             if !file_path.exists() {
-                return Err(ImportParserError::expected_lib_file(
+                return Err(ImportParserError::expected_main_file(
                     format!("{:?}", file_path.as_path()),
                     span,
                 ));
diff --git a/tests/compiler/tuples/access.leo b/tests/compiler/tuples/access.leo
new file mode 100644
index 0000000000..c75498876d
--- /dev/null
+++ b/tests/compiler/tuples/access.leo
@@ -0,0 +1,13 @@
+/*
+namespace: Compile
+expectation: Pass
+input_file: inputs/true_true.in
+*/
+
+function main(a: (bool, bool)) -> (bool, bool) {
+    let b = (false, false, true).0;
+    let c = (false, true, false).1;
+    let d = (true, false, false).2;
+
+    return (a.0 == b, a.1 == c == d);
+}
diff --git a/tests/expectations/compiler/compiler/tuples/access.leo.out b/tests/expectations/compiler/compiler/tuples/access.leo.out
new file mode 100644
index 0000000000..06cacdebe4
--- /dev/null
+++ b/tests/expectations/compiler/compiler/tuples/access.leo.out
@@ -0,0 +1,21 @@
+---
+namespace: Compile
+expectation: Pass
+outputs:
+  - circuit:
+      num_public_variables: 0
+      num_private_variables: 2
+      num_constraints: 2
+      at: 401937c524c61a28b4fab76d7a1f85bb628850012af62362a0922610372faf92
+      bt: cdf9a9cee4f2edf55111a95ae60bde9801080f6bde638a5c79273a39a2f9f7f5
+      ct: 643d5437104296e21d906ecb15b2c96ad278f20cfc4af53b12bb6069bd853726
+    output:
+      - input_file: inputs/true_true.in
+        output:
+          registers:
+            b:
+              type: bool
+              value: "true"
+            c:
+              type: bool
+              value: "false"
