diff --git a/Cargo.lock b/Cargo.lock
index 3e4bb95f50..bab77ee4e2 100644
--- a/Cargo.lock
+++ b/Cargo.lock
@@ -1153,6 +1153,7 @@ checksum = "55e2e4c765aa53a0424761bf9f41aa7a6ac1efa87238f59560640e27fca028f2"
 dependencies = [
  "autocfg",
  "hashbrown",
+ "serde",
 ]
 
 [[package]]
@@ -1299,6 +1300,7 @@ name = "leo-ast"
 version = "1.0.6"
 dependencies = [
  "criterion",
+ "indexmap",
  "leo-grammar",
  "leo-input",
  "pest",
@@ -1314,6 +1316,7 @@ version = "1.0.6"
 dependencies = [
  "bincode",
  "hex",
+ "indexmap",
  "leo-ast",
  "leo-core",
  "leo-gadgets",
@@ -1392,6 +1395,7 @@ dependencies = [
 name = "leo-imports"
 version = "1.0.6"
 dependencies = [
+ "indexmap",
  "leo-ast",
  "leo-grammar",
  "thiserror",
@@ -1478,6 +1482,7 @@ dependencies = [
 name = "leo-state"
 version = "1.0.6"
 dependencies = [
+ "indexmap",
  "leo-ast",
  "leo-input",
  "rand",
@@ -1497,6 +1502,7 @@ dependencies = [
 name = "leo-symbol-table"
 version = "1.0.6"
 dependencies = [
+ "indexmap",
  "leo-ast",
  "leo-core",
  "leo-grammar",
@@ -1509,6 +1515,7 @@ dependencies = [
 name = "leo-type-inference"
 version = "1.0.6"
 dependencies = [
+ "indexmap",
  "leo-ast",
  "leo-grammar",
  "leo-imports",
diff --git a/ast/Cargo.toml b/ast/Cargo.toml
index eec77394f0..06c5d997ac 100644
--- a/ast/Cargo.toml
+++ b/ast/Cargo.toml
@@ -42,6 +42,10 @@ default-features = false
 version = "1.1.3"
 default-features = false
 
+[dependencies.indexmap]
+version = "1.6.0"
+features = ["serde-1"]
+
 [dependencies.pest]
 version = "2.0"
 
diff --git a/ast/src/annotation.rs b/ast/src/annotation.rs
index 2a882c1fc0..50348b7dfe 100644
--- a/ast/src/annotation.rs
+++ b/ast/src/annotation.rs
@@ -20,14 +20,14 @@ use leo_grammar::{
     definitions::{AnnotatedDefinition, Definition},
 };
 
-use std::collections::HashMap;
+use indexmap::IndexMap;
 
 pub fn load_annotation(
     annotated_definition: AnnotatedDefinition,
     _imports: &mut Vec<ImportStatement>,
-    _circuits: &mut HashMap<Identifier, Circuit>,
-    _functions: &mut HashMap<Identifier, Function>,
-    tests: &mut HashMap<Identifier, TestFunction>,
+    _circuits: &mut IndexMap<Identifier, Circuit>,
+    _functions: &mut IndexMap<Identifier, Function>,
+    tests: &mut IndexMap<Identifier, TestFunction>,
     _expected: &mut Vec<FunctionInput>,
 ) {
     let ast_annotation = annotated_definition.annotation;
@@ -45,7 +45,7 @@ pub fn load_annotation(
     }
 }
 
-pub fn load_annotated_test(test: TestFunction, annotation: Annotation, tests: &mut HashMap<Identifier, TestFunction>) {
+pub fn load_annotated_test(test: TestFunction, annotation: Annotation, tests: &mut IndexMap<Identifier, TestFunction>) {
     let name = annotation.name;
     let ast_arguments = annotation.arguments;
 
@@ -57,7 +57,7 @@ pub fn load_annotated_test(test: TestFunction, annotation: Annotation, tests: &m
 pub fn load_annotated_test_context(
     mut test: TestFunction,
     ast_arguments: AnnotationArguments,
-    tests: &mut HashMap<Identifier, TestFunction>,
+    tests: &mut IndexMap<Identifier, TestFunction>,
 ) {
     let arguments = ast_arguments.arguments;
 
diff --git a/ast/src/input/macros.rs b/ast/src/input/macros.rs
index 16276d4e1d..75a04d8f3e 100644
--- a/ast/src/input/macros.rs
+++ b/ast/src/input/macros.rs
@@ -22,7 +22,7 @@ macro_rules! input_section_impl {
         #[derive(Clone, PartialEq, Eq, Default)]
         pub struct $name {
             is_present: bool,
-            values: HashMap<Parameter, Option<InputValue>>,
+            values: IndexMap<Parameter, Option<InputValue>>,
         }
 
         impl $name {
@@ -63,8 +63,8 @@ macro_rules! input_section_impl {
                 Ok(())
             }
 
-            /// Returns this section's hashmap of values
-            pub fn values(&self) -> HashMap<Parameter, Option<InputValue>> {
+            /// Returns this section's [IndexMap] of values
+            pub fn values(&self) -> IndexMap<Parameter, Option<InputValue>> {
                 self.values.clone()
             }
         }
diff --git a/ast/src/input/program_input/main_input.rs b/ast/src/input/program_input/main_input.rs
index 26f77c7400..d930a0b050 100644
--- a/ast/src/input/program_input/main_input.rs
+++ b/ast/src/input/program_input/main_input.rs
@@ -16,11 +16,12 @@
 
 use crate::InputValue;
 use leo_input::{definitions::Definition, InputParserError};
-use std::collections::HashMap;
+
+use indexmap::IndexMap;
 
 #[derive(Clone, PartialEq, Eq, Default)]
 pub struct MainInput {
-    input: HashMap<String, Option<InputValue>>,
+    input: IndexMap<String, Option<InputValue>>,
 }
 
 #[allow(clippy::len_without_is_empty)]
diff --git a/ast/src/input/program_input/registers.rs b/ast/src/input/program_input/registers.rs
index e3d7217d4c..e0981602ad 100644
--- a/ast/src/input/program_input/registers.rs
+++ b/ast/src/input/program_input/registers.rs
@@ -17,6 +17,6 @@
 use crate::{InputValue, Parameter};
 use leo_input::{definitions::Definition, InputParserError};
 
-use std::collections::HashMap;
+use indexmap::IndexMap;
 
 input_section_impl!(Registers);
diff --git a/ast/src/input/program_state/private_state/record.rs b/ast/src/input/program_state/private_state/record.rs
index f2e1645449..0aef9d50e1 100644
--- a/ast/src/input/program_state/private_state/record.rs
+++ b/ast/src/input/program_state/private_state/record.rs
@@ -17,6 +17,6 @@
 use crate::{InputValue, Parameter};
 use leo_input::{definitions::Definition, InputParserError};
 
-use std::collections::HashMap;
+use indexmap::IndexMap;
 
 input_section_impl!(Record);
diff --git a/ast/src/input/program_state/private_state/state_leaf.rs b/ast/src/input/program_state/private_state/state_leaf.rs
index e66c51f0ab..2fd1c01d72 100644
--- a/ast/src/input/program_state/private_state/state_leaf.rs
+++ b/ast/src/input/program_state/private_state/state_leaf.rs
@@ -17,6 +17,6 @@
 use crate::{InputValue, Parameter};
 use leo_input::{definitions::Definition, InputParserError};
 
-use std::collections::HashMap;
+use indexmap::IndexMap;
 
 input_section_impl!(StateLeaf);
diff --git a/ast/src/input/program_state/public_state/state.rs b/ast/src/input/program_state/public_state/state.rs
index e6afcb7d6e..04b1a37923 100644
--- a/ast/src/input/program_state/public_state/state.rs
+++ b/ast/src/input/program_state/public_state/state.rs
@@ -17,6 +17,6 @@
 use crate::{InputValue, Parameter};
 use leo_input::{definitions::Definition, InputParserError};
 
-use std::collections::HashMap;
+use indexmap::IndexMap;
 
 input_section_impl!(State);
diff --git a/ast/src/program.rs b/ast/src/program.rs
index ac016363e5..4794267217 100644
--- a/ast/src/program.rs
+++ b/ast/src/program.rs
@@ -20,18 +20,17 @@
 use crate::{load_annotation, Circuit, Function, FunctionInput, Identifier, ImportStatement, TestFunction};
 use leo_grammar::{definitions::Definition, files::File};
 
+use indexmap::IndexMap;
 use serde::{Deserialize, Serialize};
-use std::collections::HashMap;
-
 /// Stores the Leo program abstract syntax tree.
 #[derive(Debug, Clone, Eq, PartialEq, Serialize, Deserialize)]
 pub struct Program {
     pub name: String,
     pub expected_input: Vec<FunctionInput>,
     pub imports: Vec<ImportStatement>,
-    pub circuits: HashMap<Identifier, Circuit>,
-    pub functions: HashMap<Identifier, Function>,
-    pub tests: HashMap<Identifier, TestFunction>,
+    pub circuits: IndexMap<Identifier, Circuit>,
+    pub functions: IndexMap<Identifier, Function>,
+    pub tests: IndexMap<Identifier, TestFunction>,
 }
 
 const MAIN_FUNCTION_NAME: &str = "main";
@@ -40,9 +39,9 @@ impl<'ast> Program {
     //! Logic to convert from an abstract syntax tree (ast) representation to a Leo program.
     pub fn from(program_name: &str, program_ast: &File<'ast>) -> Self {
         let mut imports = vec![];
-        let mut circuits = HashMap::new();
-        let mut functions = HashMap::new();
-        let mut tests = HashMap::new();
+        let mut circuits = IndexMap::new();
+        let mut functions = IndexMap::new();
+        let mut tests = IndexMap::new();
         let mut expected_input = vec![];
 
         program_ast
@@ -94,9 +93,9 @@ impl Program {
             name,
             expected_input: vec![],
             imports: vec![],
-            circuits: HashMap::new(),
-            functions: HashMap::new(),
-            tests: HashMap::new(),
+            circuits: IndexMap::new(),
+            functions: IndexMap::new(),
+            tests: IndexMap::new(),
         }
     }
 
diff --git a/compiler/Cargo.toml b/compiler/Cargo.toml
index cbf0636741..fe242bf7d0 100644
--- a/compiler/Cargo.toml
+++ b/compiler/Cargo.toml
@@ -90,6 +90,10 @@ version = "1.0"
 [dependencies.hex]
 version = "0.4.2"
 
+[dependencies.indexmap]
+version = "1.6.0"
+features = ["serde-1"]
+
 [dependencies.pest]
 version = "2.0"
 
diff --git a/compiler/src/function/input/input_section.rs b/compiler/src/function/input/input_section.rs
index 88b3e77da6..c33bc39636 100644
--- a/compiler/src/function/input/input_section.rs
+++ b/compiler/src/function/input/input_section.rs
@@ -15,20 +15,21 @@
 // along with the Leo library. If not, see <https://www.gnu.org/licenses/>.
 
 use crate::{errors::FunctionError, ConstrainedCircuitMember, ConstrainedProgram, ConstrainedValue, GroupType};
-
 use leo_ast::{Identifier, InputValue, Parameter};
+
 use snarkos_models::{
     curves::{Field, PrimeField},
     gadgets::r1cs::ConstraintSystem,
 };
-use std::collections::HashMap;
+
+use indexmap::IndexMap;
 
 impl<F: Field + PrimeField, G: GroupType<F>> ConstrainedProgram<F, G> {
     pub fn allocate_input_section<CS: ConstraintSystem<F>>(
         &mut self,
         cs: &mut CS,
         identifier: Identifier,
-        section: HashMap<Parameter, Option<InputValue>>,
+        section: IndexMap<Parameter, Option<InputValue>>,
     ) -> Result<ConstrainedValue<F, G>, FunctionError> {
         let mut members = Vec::with_capacity(section.len());
 
diff --git a/compiler/src/program/program.rs b/compiler/src/program/program.rs
index 596c59b23e..75f36f4677 100644
--- a/compiler/src/program/program.rs
+++ b/compiler/src/program/program.rs
@@ -20,16 +20,16 @@ use crate::{value::ConstrainedValue, GroupType};
 
 use snarkos_models::curves::{Field, PrimeField};
 
-use std::collections::HashMap;
+use indexmap::IndexMap;
 
 pub struct ConstrainedProgram<F: Field + PrimeField, G: GroupType<F>> {
-    pub identifiers: HashMap<String, ConstrainedValue<F, G>>,
+    pub identifiers: IndexMap<String, ConstrainedValue<F, G>>,
 }
 
 impl<F: Field + PrimeField, G: GroupType<F>> Default for ConstrainedProgram<F, G> {
     fn default() -> Self {
         Self {
-            identifiers: HashMap::new(),
+            identifiers: IndexMap::new(),
         }
     }
 }
diff --git a/imports/Cargo.toml b/imports/Cargo.toml
index 3a4bd13771..909269250f 100644
--- a/imports/Cargo.toml
+++ b/imports/Cargo.toml
@@ -25,6 +25,10 @@ version = "1.0.6"
 path = "../grammar"
 version = "1.0.6"
 
+[dependencies.indexmap]
+version = "1.6.0"
+features = ["serde-1"]
+
 [dependencies.thiserror]
 version = "1.0"
 
diff --git a/imports/src/parser/import_parser.rs b/imports/src/parser/import_parser.rs
index 2466448ec9..3b4dc2fe7b 100644
--- a/imports/src/parser/import_parser.rs
+++ b/imports/src/parser/import_parser.rs
@@ -17,10 +17,8 @@
 use crate::errors::ImportParserError;
 use leo_ast::{Package, Program};
 
-use std::{
-    collections::{HashMap, HashSet},
-    env::current_dir,
-};
+use indexmap::{IndexMap, IndexSet};
+use std::env::current_dir;
 
 /// Stores imported packages.
 ///
@@ -28,8 +26,8 @@ use std::{
 /// directory, foreign in the imports directory, or part of the core package list.
 #[derive(Clone, Default)]
 pub struct ImportParser {
-    imports: HashMap<String, Program>,
-    core_packages: HashSet<Package>,
+    imports: IndexMap<String, Program>,
+    core_packages: IndexSet<Package>,
 }
 
 impl ImportParser {
diff --git a/state/Cargo.toml b/state/Cargo.toml
index e2a3f143ea..5e08e52011 100644
--- a/state/Cargo.toml
+++ b/state/Cargo.toml
@@ -52,6 +52,10 @@ default-features = false
 [dependencies.snarkos-utilities]
 version = "1.1.3"
 
+[dependencies.indexmap]
+version = "1.6.0"
+features = ["serde-1"]
+
 [dependencies.rand]
 version = "0.7"
 
diff --git a/state/src/utilities/input_value.rs b/state/src/utilities/input_value.rs
index 70dff7d22f..5e896ed564 100644
--- a/state/src/utilities/input_value.rs
+++ b/state/src/utilities/input_value.rs
@@ -17,13 +17,13 @@
 use crate::InputValueError;
 use leo_ast::{InputValue, Parameter};
 
-use std::collections::HashMap;
+use indexmap::IndexMap;
 
 /// Returns the input parameter with the given name.
 /// If a parameter with the given name does not exist, then an error is returned.
 pub fn find_input(
     name: String,
-    parameters: &HashMap<Parameter, Option<InputValue>>,
+    parameters: &IndexMap<Parameter, Option<InputValue>>,
 ) -> Result<InputValue, InputValueError> {
     let matched_parameter = parameters
         .iter()
diff --git a/symbol-table/Cargo.toml b/symbol-table/Cargo.toml
index 809f0de36f..aa9f5e1a50 100644
--- a/symbol-table/Cargo.toml
+++ b/symbol-table/Cargo.toml
@@ -33,6 +33,10 @@ version = "1.0.6"
 path = "../imports"
 version = "1.0.6"
 
+[dependencies.indexmap]
+version = "1.6.0"
+features = ["serde-1"]
+
 [dependencies.serde]
 version = "1.0"
 
diff --git a/symbol-table/src/symbol_table.rs b/symbol-table/src/symbol_table.rs
index 0e5a56b121..9c58914ada 100644
--- a/symbol-table/src/symbol_table.rs
+++ b/symbol-table/src/symbol_table.rs
@@ -19,7 +19,7 @@ use leo_ast::{Circuit, Function, Identifier, ImportStatement, ImportSymbol, Inpu
 use leo_core::CorePackageList;
 use leo_imports::ImportParser;
 
-use std::collections::{HashMap, HashSet};
+use indexmap::{IndexMap, IndexSet};
 
 pub const INPUT_VARIABLE_NAME: &str = "input";
 pub const RECORD_VARIABLE_NAME: &str = "record";
@@ -35,13 +35,13 @@ pub const STATE_LEAF_VARIABLE_NAME: &str = "state_leaf";
 #[derive(Clone, Default)]
 pub struct SymbolTable {
     /// Maps name -> parameter type.
-    names: HashMap<String, UserDefinedType>,
+    names: IndexMap<String, UserDefinedType>,
 
     /// Maps circuit name -> circuit type.
-    circuits: HashMap<String, CircuitType>,
+    circuits: IndexMap<String, CircuitType>,
 
     /// Maps function name -> function type.
-    functions: HashMap<String, FunctionType>,
+    functions: IndexMap<String, FunctionType>,
 
     /// The parent of this symbol table.
     parent: Option<Box<SymbolTable>>,
@@ -214,7 +214,7 @@ impl SymbolTable {
     /// If a circuit name has no duplicates, then it is inserted into the symbol table.
     /// Types defined later in the program cannot have the same name.
     ///
-    pub fn check_circuit_names(&mut self, circuits: &HashMap<Identifier, Circuit>) -> Result<(), SymbolTableError> {
+    pub fn check_circuit_names(&mut self, circuits: &IndexMap<Identifier, Circuit>) -> Result<(), SymbolTableError> {
         // Iterate over circuit names and definitions.
         for (identifier, circuit) in circuits.iter() {
             // Attempt to insert the circuit name into the symbol table.
@@ -230,7 +230,7 @@ impl SymbolTable {
     /// If a function name has no duplicates, then it is inserted into the symbol table.
     /// Types defined later in the program cannot have the same name.
     ///
-    pub fn check_function_names(&mut self, functions: &HashMap<Identifier, Function>) -> Result<(), SymbolTableError> {
+    pub fn check_function_names(&mut self, functions: &IndexMap<Identifier, Function>) -> Result<(), SymbolTableError> {
         // Iterate over function names and definitions.
         for (identifier, function) in functions.iter() {
             // Attempt to insert the function name into the symbol table.
@@ -326,7 +326,7 @@ impl SymbolTable {
 
         // Import all symbols from an imported file for now.
         // Keep track of which import files have already been checked.
-        let mut checked = HashSet::new();
+        let mut checked = IndexSet::new();
 
         // Iterate over each imported symbol.
         for (name, symbol) in imported_symbols.symbols {
@@ -416,7 +416,7 @@ impl SymbolTable {
     /// symbol table. Variables defined later in the program can lookup the definition
     /// and refer to its expected types
     ///
-    pub fn check_types_circuits(&mut self, circuits: &HashMap<Identifier, Circuit>) -> Result<(), SymbolTableError> {
+    pub fn check_types_circuits(&mut self, circuits: &IndexMap<Identifier, Circuit>) -> Result<(), SymbolTableError> {
         // Iterate over circuit names and definitions.
         for circuit in circuits.values() {
             // Get the identifier of the circuit.
@@ -439,7 +439,10 @@ impl SymbolTable {
     /// symbol table. Variables defined later in the program can lookup the definition
     /// and refer to its expected types
     ///
-    pub fn check_types_functions(&mut self, functions: &HashMap<Identifier, Function>) -> Result<(), SymbolTableError> {
+    pub fn check_types_functions(
+        &mut self,
+        functions: &IndexMap<Identifier, Function>,
+    ) -> Result<(), SymbolTableError> {
         // Iterate over function names and definitions.
         for function in functions.values() {
             // Get the identifier of the function.
diff --git a/symbol-table/src/types/circuits/circuit.rs b/symbol-table/src/types/circuits/circuit.rs
index df67e3495d..b0b123879c 100644
--- a/symbol-table/src/types/circuits/circuit.rs
+++ b/symbol-table/src/types/circuits/circuit.rs
@@ -17,11 +17,9 @@
 use crate::{types::circuits::CircuitVariableType, FunctionType, SymbolTable, Type, TypeError};
 use leo_ast::{Circuit, CircuitMember, Identifier, InputValue, Parameter, Span};
 
+use indexmap::IndexMap;
 use serde::{Deserialize, Serialize};
-use std::{
-    collections::HashMap,
-    hash::{Hash, Hasher},
-};
+use std::hash::{Hash, Hasher};
 
 /// Stores circuit definition details.
 ///
@@ -133,7 +131,7 @@ impl CircuitType {
     pub fn from_input_section(
         table: &SymbolTable,
         name: String,
-        section: HashMap<Parameter, Option<InputValue>>,
+        section: IndexMap<Parameter, Option<InputValue>>,
     ) -> Result<Self, TypeError> {
         // Create a new `CircuitVariableType` for each section pair.
         let mut variables = Vec::new();
diff --git a/type-inference/Cargo.toml b/type-inference/Cargo.toml
index 4728d0e3d9..2841ba3f58 100644
--- a/type-inference/Cargo.toml
+++ b/type-inference/Cargo.toml
@@ -33,6 +33,10 @@ version = "1.0.6"
 path = "../symbol-table"
 version = "1.0.6"
 
+[dependencies.indexmap]
+version = "1.6.0"
+features = ["serde-1"]
+
 [dependencies.serde_json]
 version = "1.0"
 
diff --git a/type-inference/src/objects/variable_table.rs b/type-inference/src/objects/variable_table.rs
index c48dcff583..9d15007cf3 100644
--- a/type-inference/src/objects/variable_table.rs
+++ b/type-inference/src/objects/variable_table.rs
@@ -16,11 +16,12 @@
 
 use crate::VariableTableError;
 use leo_symbol_table::{FunctionInputType, Type};
-use std::collections::HashMap;
+
+use indexmap::IndexMap;
 
 /// Mapping of variable names to types
 #[derive(Clone)]
-pub struct VariableTable(pub HashMap<String, Type>);
+pub struct VariableTable(pub IndexMap<String, Type>);
 
 impl VariableTable {
     ///
@@ -67,6 +68,6 @@ impl VariableTable {
 
 impl Default for VariableTable {
     fn default() -> Self {
-        Self(HashMap::new())
+        Self(IndexMap::new())
     }
 }
