diff --git a/Cargo.lock b/Cargo.lock
index d21daf157d..1b50f3c03c 100644
--- a/Cargo.lock
+++ b/Cargo.lock
@@ -1314,6 +1314,7 @@ dependencies = [
  "snarkvm-gadgets",
  "snarkvm-r1cs",
  "snarkvm-utilities",
+ "tempfile",
  "thiserror",
  "tracing",
 ]
diff --git a/asg/src/expression/circuit_init.rs b/asg/src/expression/circuit_init.rs
index a36fa5a43c..3eced2d008 100644
--- a/asg/src/expression/circuit_init.rs
+++ b/asg/src/expression/circuit_init.rs
@@ -67,7 +67,7 @@ impl<'a> ExpressionNode<'a> for CircuitInitExpression<'a> {
     }
 
     fn is_mut_ref(&self) -> bool {
-        false
+        true
     }
 
     fn const_value(&self) -> Option<ConstValue> {
diff --git a/ast/src/common/array_dimensions.rs b/ast/src/common/array_dimensions.rs
index b0079c902f..4dc7188464 100644
--- a/ast/src/common/array_dimensions.rs
+++ b/ast/src/common/array_dimensions.rs
@@ -26,18 +26,6 @@ use std::fmt;
 pub struct ArrayDimensions(pub Vec<PositiveNumber>);
 
 impl ArrayDimensions {
-    ///
-    /// Creates a new `PositiveNumber` from the given `usize` and `Span`.
-    /// Appends the new `PositiveNumber` to the array dimensions.
-    ///
-    pub fn push_usize(&mut self, number: usize) {
-        let positive_number = PositiveNumber {
-            value: number.to_string().into(),
-        };
-
-        self.0.push(positive_number)
-    }
-
     ///
     /// Appends a vector of array dimensions to the self array dimensions.
     ///
diff --git a/compiler/Cargo.toml b/compiler/Cargo.toml
index 1c7a47cea4..01c90ad6aa 100644
--- a/compiler/Cargo.toml
+++ b/compiler/Cargo.toml
@@ -114,6 +114,9 @@ default-features = false
 version = "0.2.1"
 default-features = false
 
+[dev-dependencies.tempfile]
+version = "3.0.4"
+
 [features]
 default = [ ]
 ci_skip = [ "leo-ast/ci_skip" ]
diff --git a/compiler/src/errors/compiler.rs b/compiler/src/errors/compiler.rs
index f9c5f9919e..f66cdd488a 100644
--- a/compiler/src/errors/compiler.rs
+++ b/compiler/src/errors/compiler.rs
@@ -14,16 +14,14 @@
 // You should have received a copy of the GNU General Public License
 // along with the Leo library. If not, see <https://www.gnu.org/licenses/>.
 
-use crate::errors::{FunctionError, ImportError, OutputBytesError, OutputFileError};
+use crate::errors::FunctionError;
 use leo_asg::{AsgConvertError, FormattedError};
 use leo_ast::{CanonicalizeError, LeoError};
-use leo_imports::ImportParserError;
 use leo_input::InputParserError;
 use leo_parser::SyntaxError;
 use leo_state::LocalDataVerificationError;
 
-use bincode::Error as SerdeError;
-use std::{ffi::OsString, path::PathBuf};
+use std::path::PathBuf;
 
 #[derive(Debug, Error)]
 pub enum CompilerError {
@@ -33,12 +31,6 @@ pub enum CompilerError {
     #[error("{}", _0)]
     AsgPassError(FormattedError),
 
-    #[error("{}", _0)]
-    ImportError(#[from] ImportError),
-
-    #[error("{}", _0)]
-    ImportParserError(#[from] ImportParserError),
-
     #[error("{}", _0)]
     InputParserError(#[from] InputParserError),
 
@@ -51,30 +43,15 @@ pub enum CompilerError {
     #[error("Cannot read from the provided file path '{:?}': {}", _0, _1)]
     FileReadError(PathBuf, std::io::Error),
 
-    #[error("Cannot parse file string `{:?}`", _0)]
-    FileStringError(OsString),
-
     #[error("{}", _0)]
     LocalDataVerificationError(#[from] LocalDataVerificationError),
 
-    #[error("`main` function not found")]
-    NoMain,
-
     #[error("`main` must be a function")]
     NoMainFunction,
 
     #[error("Failed to find input files for the current test")]
     NoTestInput,
 
-    #[error("{}", _0)]
-    OutputError(#[from] OutputFileError),
-
-    #[error("{}", _0)]
-    OutputStringError(#[from] OutputBytesError),
-
-    #[error("{}", _0)]
-    SerdeError(#[from] SerdeError),
-
     #[error("{}", _0)]
     AsgConvertError(#[from] AsgConvertError),
 
diff --git a/compiler/src/errors/expression.rs b/compiler/src/errors/expression.rs
index e472688f5b..e093b5dd2f 100644
--- a/compiler/src/errors/expression.rs
+++ b/compiler/src/errors/expression.rs
@@ -15,7 +15,7 @@
 // along with the Leo library. If not, see <https://www.gnu.org/licenses/>.
 
 use crate::errors::{AddressError, BooleanError, FieldError, FunctionError, GroupError, IntegerError, ValueError};
-use leo_ast::{ArrayDimensions, FormattedError, Identifier, LeoError, PositiveNumber, Span};
+use leo_ast::{FormattedError, Identifier, LeoError, Span};
 use snarkvm_r1cs::SynthesisError;
 
 #[derive(Debug, Error)]
@@ -115,24 +115,6 @@ impl ExpressionError {
         Self::new_from_span(message, span)
     }
 
-    pub fn invalid_dimensions(expected: &ArrayDimensions, actual: &ArrayDimensions, span: &Span) -> Self {
-        let message = format!(
-            "expected array dimensions {}, found array dimensions {}",
-            expected, actual
-        );
-
-        Self::new_from_span(message, span)
-    }
-
-    pub fn invalid_first_dimension(expected: &PositiveNumber, actual: &PositiveNumber, span: &Span) -> Self {
-        let message = format!(
-            "expected array dimension {}, found array dimension {}",
-            expected, actual
-        );
-
-        Self::new_from_span(message, span)
-    }
-
     pub fn invalid_index(actual: String, span: &Span) -> Self {
         let message = format!("index must resolve to an integer, found `{}`", actual);
 
@@ -145,48 +127,18 @@ impl ExpressionError {
         Self::new_from_span(message, span)
     }
 
-    pub fn invalid_spread(actual: String, span: &Span) -> Self {
-        let message = format!("spread should contain an array, found `{}`", actual);
-
-        Self::new_from_span(message, span)
-    }
-
-    pub fn invalid_member_access(member: String, span: &Span) -> Self {
-        let message = format!("non-static member `{}` must be accessed using `.` syntax", member);
-
-        Self::new_from_span(message, span)
-    }
-
     pub fn invalid_static_access(member: String, span: &Span) -> Self {
         let message = format!("static member `{}` must be accessed using `::` syntax", member);
 
         Self::new_from_span(message, span)
     }
 
-    pub fn function_no_return(function: String, span: &Span) -> Self {
-        let message = format!("inline function call to `{}` did not return", function);
-
-        Self::new_from_span(message, span)
-    }
-
-    pub fn self_keyword(span: &Span) -> Self {
-        let message = "cannot call keyword `Self` outside of a circuit function".to_string();
-
-        Self::new_from_span(message, span)
-    }
-
     pub fn undefined_array(actual: String, span: &Span) -> Self {
         let message = format!("array `{}` must be declared before it is used in an expression", actual);
 
         Self::new_from_span(message, span)
     }
 
-    pub fn undefined_tuple(actual: String, span: &Span) -> Self {
-        let message = format!("tuple `{}` must be declared before it is used in an expression", actual);
-
-        Self::new_from_span(message, span)
-    }
-
     pub fn undefined_circuit(actual: String, span: &Span) -> Self {
         let message = format!(
             "circuit `{}` must be declared before it is used in an expression",
@@ -196,21 +148,6 @@ impl ExpressionError {
         Self::new_from_span(message, span)
     }
 
-    pub fn undefined_first_dimension(span: &Span) -> Self {
-        let message = "the first dimension of the array must be a number".to_string();
-
-        Self::new_from_span(message, span)
-    }
-
-    pub fn undefined_function(function: String, span: &Span) -> Self {
-        let message = format!(
-            "function `{}` must be declared before it is used in an inline expression",
-            function
-        );
-
-        Self::new_from_span(message, span)
-    }
-
     pub fn undefined_identifier(identifier: Identifier) -> Self {
         let message = format!("Cannot find value `{}` in this scope", identifier.name);
 
@@ -222,22 +159,4 @@ impl ExpressionError {
 
         Self::new_from_span(message, span)
     }
-
-    pub fn undefined_static_access(circuit: String, member: String, span: &Span) -> Self {
-        let message = format!("Circuit `{}` has no static member `{}`", circuit, member);
-
-        Self::new_from_span(message, span)
-    }
-
-    pub fn unexpected_array(expected: String, span: &Span) -> Self {
-        let message = format!("expected type `{}`, found array with elements", expected);
-
-        Self::new_from_span(message, span)
-    }
-
-    pub fn unexpected_tuple(expected: String, actual: String, span: &Span) -> Self {
-        let message = format!("expected type `{}`, found tuple with values `{}`", expected, actual);
-
-        Self::new_from_span(message, span)
-    }
 }
diff --git a/compiler/src/errors/function.rs b/compiler/src/errors/function.rs
index 791c654377..d4ab00c0d5 100644
--- a/compiler/src/errors/function.rs
+++ b/compiler/src/errors/function.rs
@@ -128,18 +128,6 @@ impl FunctionError {
         Self::new_from_span(message, span)
     }
 
-    pub fn return_arguments_length(expected: usize, actual: usize, span: &Span) -> Self {
-        let message = format!("function expected {} returns, found {} returns", expected, actual);
-
-        Self::new_from_span(message, span)
-    }
-
-    pub fn return_argument_type(expected: String, actual: String, span: &Span) -> Self {
-        let message = format!("Expected function return type `{}`, found `{}`", expected, actual);
-
-        Self::new_from_span(message, span)
-    }
-
     pub fn input_not_found(expected: String, span: &Span) -> Self {
         let message = format!("main function input {} not found", expected);
 
diff --git a/compiler/src/errors/import.rs b/compiler/src/errors/import.rs
index dde6ce1644..022ec5d1b9 100644
--- a/compiler/src/errors/import.rs
+++ b/compiler/src/errors/import.rs
@@ -14,7 +14,7 @@
 // You should have received a copy of the GNU General Public License
 // along with the Leo library. If not, see <https://www.gnu.org/licenses/>.
 
-use leo_ast::{FormattedError, Identifier, ImportSymbol, LeoError, Span};
+use leo_ast::{FormattedError, LeoError};
 
 #[derive(Debug, Error)]
 pub enum ImportError {
@@ -23,25 +23,3 @@ pub enum ImportError {
 }
 
 impl LeoError for ImportError {}
-
-impl ImportError {
-    fn new_from_span(message: String, span: &Span) -> Self {
-        ImportError::Error(FormattedError::new_from_span(message, span))
-    }
-
-    pub fn unknown_package(identifier: Identifier) -> Self {
-        let message = format!(
-            "cannot find imported package `{}` in source files or import directory",
-            identifier.name
-        );
-
-        Self::new_from_span(message, &identifier.span)
-    }
-
-    pub fn unknown_symbol(symbol: ImportSymbol, file: String) -> Self {
-        let message = format!("cannot find imported symbol `{}` in imported file `{}`", symbol, file);
-        let error = FormattedError::new_from_span(message, &symbol.span);
-
-        ImportError::Error(error)
-    }
-}
diff --git a/compiler/src/errors/output_file.rs b/compiler/src/errors/output_file.rs
index c8b18185d6..014327c3ad 100644
--- a/compiler/src/errors/output_file.rs
+++ b/compiler/src/errors/output_file.rs
@@ -21,9 +21,6 @@ pub enum OutputFileError {
     #[error("{}: {}", _0, _1)]
     Crate(&'static str, String),
 
-    #[error("creating: {}", _0)]
-    Creating(io::Error),
-
     #[error("Cannot read from the provided file path - {:?}", _0)]
     FileReadError(PathBuf),
 
diff --git a/compiler/src/errors/statement.rs b/compiler/src/errors/statement.rs
index 6a7d49ca35..fd42cbd5d8 100644
--- a/compiler/src/errors/statement.rs
+++ b/compiler/src/errors/statement.rs
@@ -49,12 +49,6 @@ impl StatementError {
         StatementError::Error(FormattedError::new_from_span(message, span))
     }
 
-    pub fn arguments_type(expected: &Type, actual: &Type, span: &Span) -> Self {
-        let message = format!("expected return argument type `{}`, found type `{}`", expected, actual);
-
-        Self::new_from_span(message, span)
-    }
-
     pub fn array_assign_index(span: &Span) -> Self {
         let message = "Cannot assign single index to array of values".to_string();
 
@@ -103,24 +97,6 @@ impl StatementError {
         Self::new_from_span(message, span)
     }
 
-    pub fn immutable_assign(name: String, span: &Span) -> Self {
-        let message = format!("Cannot assign to immutable variable `{}`", name);
-
-        Self::new_from_span(message, span)
-    }
-
-    pub fn immutable_circuit_function(name: String, span: &Span) -> Self {
-        let message = format!("Cannot mutate circuit function, `{}`", name);
-
-        Self::new_from_span(message, span)
-    }
-
-    pub fn immutable_circuit_variable(name: String, span: &Span) -> Self {
-        let message = format!("Circuit member variable `{}` is immutable", name);
-
-        Self::new_from_span(message, span)
-    }
-
     pub fn indicator_calculation(name: String, span: &Span) -> Self {
         let message = format!(
             "Constraint system failed to evaluate branch selection indicator `{}`",
@@ -139,15 +115,6 @@ impl StatementError {
         Self::new_from_span(message, span)
     }
 
-    pub fn invalid_number_of_returns(expected: usize, actual: usize, span: &Span) -> Self {
-        let message = format!(
-            "Function return statement expected {} return values, found {} values",
-            expected, actual
-        );
-
-        Self::new_from_span(message, span)
-    }
-
     pub fn multiple_definition(value: String, span: &Span) -> Self {
         let message = format!("cannot assign multiple variables to a single value: {}", value,);
 
@@ -194,12 +161,6 @@ impl StatementError {
         Self::new_from_span(message, span)
     }
 
-    pub fn tuple_type(type_: String, span: &Span) -> Self {
-        let message = format!("Expected tuple type, found type `{}`", type_);
-
-        Self::new_from_span(message, span)
-    }
-
     pub fn unassigned(span: &Span) -> Self {
         let message = "Expected assignment of return values for expression".to_string();
 
diff --git a/compiler/src/errors/value/address.rs b/compiler/src/errors/value/address.rs
index 73a499705c..2b8d4d8022 100644
--- a/compiler/src/errors/value/address.rs
+++ b/compiler/src/errors/value/address.rs
@@ -16,7 +16,6 @@
 
 use leo_ast::{FormattedError, LeoError, Span};
 use snarkvm_dpc::AccountError;
-use snarkvm_r1cs::SynthesisError;
 
 #[derive(Debug, Error)]
 pub enum AddressError {
@@ -37,21 +36,6 @@ impl AddressError {
         Self::new_from_span(message, span)
     }
 
-    pub fn cannot_enforce(operation: String, error: SynthesisError, span: &Span) -> Self {
-        let message = format!(
-            "the address operation `{:?}` failed due to the synthesis error `{}`",
-            operation, error,
-        );
-
-        Self::new_from_span(message, span)
-    }
-
-    pub fn cannot_evaluate(operation: String, span: &Span) -> Self {
-        let message = format!("no implementation found for `{}`", operation);
-
-        Self::new_from_span(message, span)
-    }
-
     pub fn invalid_address(actual: String, span: &Span) -> Self {
         let message = format!("expected address input type, found `{}`", actual);
 
diff --git a/compiler/src/errors/value/field.rs b/compiler/src/errors/value/field.rs
index 87de99b995..9ec8a5f152 100644
--- a/compiler/src/errors/value/field.rs
+++ b/compiler/src/errors/value/field.rs
@@ -62,10 +62,4 @@ impl FieldError {
 
         Self::new_from_span(message, span)
     }
-
-    pub fn synthesis_error(error: SynthesisError, span: &Span) -> Self {
-        let message = format!("compilation failed due to field synthesis error `{:?}`", error);
-
-        Self::new_from_span(message, span)
-    }
 }
diff --git a/compiler/src/errors/value/integer.rs b/compiler/src/errors/value/integer.rs
index 8fecda685c..457575b1dc 100644
--- a/compiler/src/errors/value/integer.rs
+++ b/compiler/src/errors/value/integer.rs
@@ -14,7 +14,7 @@
 // You should have received a copy of the GNU General Public License
 // along with the Leo library. If not, see <https://www.gnu.org/licenses/>.
 
-use leo_ast::{FormattedError, IntegerType, LeoError, Span, Type};
+use leo_ast::{FormattedError, LeoError, Span};
 
 use snarkvm_gadgets::errors::SignedIntegerError;
 use snarkvm_r1cs::SynthesisError;
@@ -32,30 +32,12 @@ impl IntegerError {
         IntegerError::Error(FormattedError::new_from_span(message, span))
     }
 
-    pub fn cannot_enforce(operation: String, error: SynthesisError, span: &Span) -> Self {
-        let message = format!(
-            "the integer operation `{}` failed due to the synthesis error `{:?}`",
-            operation, error,
-        );
-
-        Self::new_from_span(message, span)
-    }
-
     pub fn signed(error: SignedIntegerError, span: &Span) -> Self {
         let message = format!("integer operation failed due to the signed integer error `{:?}`", error);
 
         Self::new_from_span(message, span)
     }
 
-    pub fn signed_error(operation: String, error: SignedIntegerError, span: &Span) -> Self {
-        let message = format!(
-            "the integer operation `{}` failed due to the signed integer error `{:?}`",
-            operation, error
-        );
-
-        Self::new_from_span(message, span)
-    }
-
     pub fn synthesis(error: SynthesisError, span: &Span) -> Self {
         let message = format!("integer operation failed due to the synthesis error `{}`", error);
 
@@ -77,32 +59,12 @@ impl IntegerError {
         Self::new_from_span(message, span)
     }
 
-    pub fn invalid_index(span: &Span) -> Self {
-        let message =
-            "index must be a constant value unsigned integer. allocated indices produce a circuit of unknown size"
-                .to_string();
-
-        Self::new_from_span(message, span)
-    }
-
     pub fn invalid_integer(actual: String, span: &Span) -> Self {
         let message = format!("failed to parse `{}` as expected integer type", actual);
 
         Self::new_from_span(message, span)
     }
 
-    pub fn invalid_integer_type(expected: &IntegerType, actual: &IntegerType, span: &Span) -> Self {
-        let message = format!("expected integer type {} found integer type {}", expected, actual);
-
-        Self::new_from_span(message, span)
-    }
-
-    pub fn invalid_type(actual: &Type, span: &Span) -> Self {
-        let message = format!("expected type {}, found type IntegerType", actual);
-
-        Self::new_from_span(message, span)
-    }
-
     pub fn missing_integer(expected: String, span: &Span) -> Self {
         let message = format!("expected integer input `{}` not found", expected);
 
diff --git a/compiler/src/errors/value/value.rs b/compiler/src/errors/value/value.rs
index 90b9ef29b2..bd34b788c8 100644
--- a/compiler/src/errors/value/value.rs
+++ b/compiler/src/errors/value/value.rs
@@ -15,7 +15,7 @@
 // along with the Leo library. If not, see <https://www.gnu.org/licenses/>.
 
 use crate::errors::{AddressError, BooleanError, FieldError, GroupError, IntegerError};
-use leo_ast::{FormattedError, LeoError, Span};
+use leo_ast::{FormattedError, LeoError};
 
 #[derive(Debug, Error)]
 pub enum ValueError {
@@ -39,21 +39,3 @@ pub enum ValueError {
 }
 
 impl LeoError for ValueError {}
-
-impl ValueError {
-    fn new_from_span(message: String, span: &Span) -> Self {
-        ValueError::Error(FormattedError::new_from_span(message, span))
-    }
-
-    pub fn implicit(value: String, span: &Span) -> Self {
-        let message = format!("explicit type needed for `{}`", value);
-
-        Self::new_from_span(message, span)
-    }
-
-    pub fn implicit_group(span: &Span) -> Self {
-        let message = "group coordinates should be in (x, y)group format".to_string();
-
-        Self::new_from_span(message, span)
-    }
-}
diff --git a/compiler/src/output/output_file.rs b/compiler/src/output/output_file.rs
index f9dd6eeda1..b9b2570baf 100644
--- a/compiler/src/output/output_file.rs
+++ b/compiler/src/output/output_file.rs
@@ -42,19 +42,6 @@ impl OutputFile {
         }
     }
 
-    pub fn exists_at(&self, path: &Path) -> bool {
-        let path = self.setup_file_path(path);
-        path.exists()
-    }
-
-    /// Reads the output register variables from the given file path if it exists.
-    pub fn read_from(&self, path: &Path) -> Result<String, OutputFileError> {
-        let path = self.setup_file_path(path);
-
-        let output = fs::read_to_string(&path).map_err(|_| OutputFileError::FileReadError(path.into_owned()))?;
-        Ok(output)
-    }
-
     /// Writes output to a file.
     pub fn write(&self, path: &Path, bytes: &[u8]) -> Result<(), OutputFileError> {
         // create output file
@@ -88,3 +75,26 @@ impl OutputFile {
         path
     }
 }
+
+#[cfg(test)]
+mod test_output_file {
+    use crate::{OutputFile, OUTPUTS_DIRECTORY_NAME};
+    use std::{error::Error, fs};
+
+    #[test]
+    fn test_all() -> Result<(), Box<dyn Error>> {
+        let dir = tempfile::tempdir()?;
+        let file = OutputFile::new("test");
+        let path = dir.path();
+
+        assert!(file.write(path, Default::default()).is_err());
+        assert!(!(file.remove(path)?));
+
+        fs::create_dir(dir.path().join(OUTPUTS_DIRECTORY_NAME))?;
+
+        assert!(file.write(path, Default::default()).is_ok());
+        assert!(file.remove(path)?);
+
+        Ok(())
+    }
+}
diff --git a/compiler/tests/circuits/mod.rs b/compiler/tests/circuits/mod.rs
index 9395853529..093e2263e1 100644
--- a/compiler/tests/circuits/mod.rs
+++ b/compiler/tests/circuits/mod.rs
@@ -308,3 +308,11 @@ fn test_duplicate_name_context() {
 
     assert_satisfied(program);
 }
+
+#[test]
+fn test_mutable_call_immutable_context() {
+    let program_string = include_str!("mutable_call_immutable_context.leo");
+    let program = parse_program(program_string).unwrap();
+
+    assert_satisfied(program);
+}
diff --git a/compiler/tests/circuits/mutable_call_immutable_context.leo b/compiler/tests/circuits/mutable_call_immutable_context.leo
new file mode 100644
index 0000000000..d68e1866c5
--- /dev/null
+++ b/compiler/tests/circuits/mutable_call_immutable_context.leo
@@ -0,0 +1,12 @@
+circuit TestMe {
+    x: u8,
+
+    function test_me(mut self) -> u8 {
+        self.x += 1;
+        return self.x
+    }
+}
+
+function main () {
+    const t = TestMe {x: 6u8}.test_me();
+}
\ No newline at end of file
diff --git a/compiler/tests/field/mod.rs b/compiler/tests/field/mod.rs
index 2a768450d6..f5ccb6c116 100644
--- a/compiler/tests/field/mod.rs
+++ b/compiler/tests/field/mod.rs
@@ -74,6 +74,14 @@ fn test_field() {
     assert_satisfied(program)
 }
 
+#[test]
+fn test_no_space_between_literal() {
+    let program_string = include_str!("no_space_between_literal.leo");
+    let mut program = parse_program(program_string).unwrap();
+
+    expect_compiler_error(program)
+}
+
 #[test]
 fn test_add() {
     use std::ops::Add;
diff --git a/compiler/tests/field/no_space_between_literal.leo b/compiler/tests/field/no_space_between_literal.leo
new file mode 100644
index 0000000000..9ac9dcd5db
--- /dev/null
+++ b/compiler/tests/field/no_space_between_literal.leo
@@ -0,0 +1,3 @@
+function main() {
+  const f = 1 field;
+}
\ No newline at end of file
diff --git a/compiler/tests/group/mod.rs b/compiler/tests/group/mod.rs
index f40cad73f6..f4cf84448c 100644
--- a/compiler/tests/group/mod.rs
+++ b/compiler/tests/group/mod.rs
@@ -45,6 +45,14 @@ pub fn group_element_to_input_value(g: EdwardsAffine) -> GroupValue {
     })
 }
 
+#[test]
+fn test_no_space_between_literal() {
+    let program_string = include_str!("no_space_between_literal.leo");
+    let mut program = parse_program(program_string).unwrap();
+
+    expect_compiler_error(program)
+}
+
 #[test]
 fn test_one() {
     let program_string = include_str!("one.leo");
diff --git a/compiler/tests/group/no_space_between_literal.leo b/compiler/tests/group/no_space_between_literal.leo
new file mode 100644
index 0000000000..af53e5dffe
--- /dev/null
+++ b/compiler/tests/group/no_space_between_literal.leo
@@ -0,0 +1,3 @@
+function main() {
+  const g = (0,1) group;
+}
\ No newline at end of file
diff --git a/compiler/tests/integers/i128/mod.rs b/compiler/tests/integers/i128/mod.rs
index e62e07260e..ead011f14e 100644
--- a/compiler/tests/integers/i128/mod.rs
+++ b/compiler/tests/integers/i128/mod.rs
@@ -132,3 +132,11 @@ fn test_i128_assert_eq() {
 fn test_i128_ternary() {
     TestI128::test_ternary();
 }
+
+#[test]
+fn test_no_space_between_literal() {
+    let program_string = include_str!("no_space_between_literal.leo");
+    let program = parse_program(program_string);
+
+    assert!(program.is_err());
+}
diff --git a/compiler/tests/integers/i128/no_space_between_literal.leo b/compiler/tests/integers/i128/no_space_between_literal.leo
new file mode 100644
index 0000000000..3142fb5724
--- /dev/null
+++ b/compiler/tests/integers/i128/no_space_between_literal.leo
@@ -0,0 +1,3 @@
+function main() {
+  const i = 1 i128;
+}
\ No newline at end of file
diff --git a/compiler/tests/integers/i16/mod.rs b/compiler/tests/integers/i16/mod.rs
index 49b45f5b3e..6f415a0877 100644
--- a/compiler/tests/integers/i16/mod.rs
+++ b/compiler/tests/integers/i16/mod.rs
@@ -131,3 +131,11 @@ fn test_i16_console_assert() {
 fn test_i16_ternary() {
     TestI16::test_ternary();
 }
+
+#[test]
+fn test_no_space_between_literal() {
+    let program_string = include_str!("no_space_between_literal.leo");
+    let program = parse_program(program_string);
+
+    assert!(program.is_err());
+}
diff --git a/compiler/tests/integers/i16/no_space_between_literal.leo b/compiler/tests/integers/i16/no_space_between_literal.leo
new file mode 100644
index 0000000000..96c6b2a914
--- /dev/null
+++ b/compiler/tests/integers/i16/no_space_between_literal.leo
@@ -0,0 +1,3 @@
+function main() {
+  const i = 1 i16;
+}
\ No newline at end of file
diff --git a/compiler/tests/integers/i32/mod.rs b/compiler/tests/integers/i32/mod.rs
index ee9e25701a..b8a6925ac2 100644
--- a/compiler/tests/integers/i32/mod.rs
+++ b/compiler/tests/integers/i32/mod.rs
@@ -131,3 +131,11 @@ fn test_i32_console_assert() {
 fn test_i32_ternary() {
     TestI32::test_ternary();
 }
+
+#[test]
+fn test_no_space_between_literal() {
+    let program_string = include_str!("no_space_between_literal.leo");
+    let program = parse_program(program_string);
+
+    assert!(program.is_err());
+}
diff --git a/compiler/tests/integers/i32/no_space_between_literal.leo b/compiler/tests/integers/i32/no_space_between_literal.leo
new file mode 100644
index 0000000000..845a7f403e
--- /dev/null
+++ b/compiler/tests/integers/i32/no_space_between_literal.leo
@@ -0,0 +1,3 @@
+function main() {
+  const i = 1 i32;
+}
\ No newline at end of file
diff --git a/compiler/tests/integers/i64/mod.rs b/compiler/tests/integers/i64/mod.rs
index 066ac9956f..ccfb06ce68 100644
--- a/compiler/tests/integers/i64/mod.rs
+++ b/compiler/tests/integers/i64/mod.rs
@@ -132,3 +132,11 @@ fn test_i64_console_assert() {
 fn test_i64_ternary() {
     TestI64::test_ternary();
 }
+
+#[test]
+fn test_no_space_between_literal() {
+    let program_string = include_str!("no_space_between_literal.leo");
+    let program = parse_program(program_string);
+
+    assert!(program.is_err());
+}
diff --git a/compiler/tests/integers/i64/no_space_between_literal.leo b/compiler/tests/integers/i64/no_space_between_literal.leo
new file mode 100644
index 0000000000..7fb6ea9e62
--- /dev/null
+++ b/compiler/tests/integers/i64/no_space_between_literal.leo
@@ -0,0 +1,3 @@
+function main() {
+  const i = 1 i64;
+}
\ No newline at end of file
diff --git a/compiler/tests/integers/i8/mod.rs b/compiler/tests/integers/i8/mod.rs
index 9133a72248..5073dda460 100644
--- a/compiler/tests/integers/i8/mod.rs
+++ b/compiler/tests/integers/i8/mod.rs
@@ -131,3 +131,11 @@ fn test_i8_console_assert() {
 fn test_i8_ternary() {
     TestI8::test_ternary();
 }
+
+#[test]
+fn test_no_space_between_literal() {
+    let program_string = include_str!("no_space_between_literal.leo");
+    let program = parse_program(program_string);
+
+    assert!(program.is_err());
+}
diff --git a/compiler/tests/integers/i8/no_space_between_literal.leo b/compiler/tests/integers/i8/no_space_between_literal.leo
new file mode 100644
index 0000000000..7a8eda7ef9
--- /dev/null
+++ b/compiler/tests/integers/i8/no_space_between_literal.leo
@@ -0,0 +1,3 @@
+function main() {
+  const i = 1 i8;
+}
\ No newline at end of file
diff --git a/compiler/tests/integers/u128/mod.rs b/compiler/tests/integers/u128/mod.rs
index 2c110af0a5..38677dcb29 100644
--- a/compiler/tests/integers/u128/mod.rs
+++ b/compiler/tests/integers/u128/mod.rs
@@ -116,3 +116,11 @@ fn test_u128_console_assert() {
 fn test_u128_ternary() {
     TestU128::test_ternary();
 }
+
+#[test]
+fn test_no_space_between_literal() {
+    let program_string = include_str!("no_space_between_literal.leo");
+    let program = parse_program(program_string);
+
+    assert!(program.is_err());
+}
diff --git a/compiler/tests/integers/u128/no_space_between_literal.leo b/compiler/tests/integers/u128/no_space_between_literal.leo
new file mode 100644
index 0000000000..fc48e7f1ae
--- /dev/null
+++ b/compiler/tests/integers/u128/no_space_between_literal.leo
@@ -0,0 +1,3 @@
+function main() {
+  const i = 1 u128;
+}
\ No newline at end of file
diff --git a/compiler/tests/integers/u16/mod.rs b/compiler/tests/integers/u16/mod.rs
index b4b202b9da..33f88452c1 100644
--- a/compiler/tests/integers/u16/mod.rs
+++ b/compiler/tests/integers/u16/mod.rs
@@ -116,3 +116,11 @@ fn test_u16_console_assert() {
 fn test_u16_ternary() {
     TestU16::test_ternary();
 }
+
+#[test]
+fn test_no_space_between_literal() {
+    let program_string = include_str!("no_space_between_literal.leo");
+    let program = parse_program(program_string);
+
+    assert!(program.is_err());
+}
diff --git a/compiler/tests/integers/u16/no_space_between_literal.leo b/compiler/tests/integers/u16/no_space_between_literal.leo
new file mode 100644
index 0000000000..34e125cfc6
--- /dev/null
+++ b/compiler/tests/integers/u16/no_space_between_literal.leo
@@ -0,0 +1,3 @@
+function main() {
+  const i = 1 u16;
+}
\ No newline at end of file
diff --git a/compiler/tests/integers/u32/mod.rs b/compiler/tests/integers/u32/mod.rs
index 920fc6ed5b..7f2d76d198 100644
--- a/compiler/tests/integers/u32/mod.rs
+++ b/compiler/tests/integers/u32/mod.rs
@@ -116,3 +116,11 @@ fn test_u32_console_assert() {
 fn test_u32_ternary() {
     TestU32::test_ternary();
 }
+
+#[test]
+fn test_no_space_between_literal() {
+    let program_string = include_str!("no_space_between_literal.leo");
+    let program = parse_program(program_string);
+
+    assert!(program.is_err());
+}
diff --git a/compiler/tests/integers/u32/no_space_between_literal.leo b/compiler/tests/integers/u32/no_space_between_literal.leo
new file mode 100644
index 0000000000..83b3caca55
--- /dev/null
+++ b/compiler/tests/integers/u32/no_space_between_literal.leo
@@ -0,0 +1,3 @@
+function main() {
+  const i = 1 u32;
+}
\ No newline at end of file
diff --git a/compiler/tests/integers/u64/mod.rs b/compiler/tests/integers/u64/mod.rs
index ec86c868f1..1a2af24ddd 100644
--- a/compiler/tests/integers/u64/mod.rs
+++ b/compiler/tests/integers/u64/mod.rs
@@ -116,3 +116,11 @@ fn test_u64_console_assert() {
 fn test_u64_ternary() {
     TestU64::test_ternary();
 }
+
+#[test]
+fn test_no_space_between_literal() {
+    let program_string = include_str!("no_space_between_literal.leo");
+    let program = parse_program(program_string);
+
+    assert!(program.is_err());
+}
diff --git a/compiler/tests/integers/u64/no_space_between_literal.leo b/compiler/tests/integers/u64/no_space_between_literal.leo
new file mode 100644
index 0000000000..852fd5c2aa
--- /dev/null
+++ b/compiler/tests/integers/u64/no_space_between_literal.leo
@@ -0,0 +1,3 @@
+function main() {
+  const i = 1 u64;
+}
\ No newline at end of file
diff --git a/compiler/tests/integers/u8/mod.rs b/compiler/tests/integers/u8/mod.rs
index a61e28246d..f1fb2b0fab 100644
--- a/compiler/tests/integers/u8/mod.rs
+++ b/compiler/tests/integers/u8/mod.rs
@@ -116,3 +116,11 @@ fn test_u8_console_assert() {
 fn test_u8_ternary() {
     TestU8::test_ternary();
 }
+
+#[test]
+fn test_no_space_between_literal() {
+    let program_string = include_str!("no_space_between_literal.leo");
+    let program = parse_program(program_string);
+
+    assert!(program.is_err());
+}
diff --git a/compiler/tests/integers/u8/no_space_between_literal.leo b/compiler/tests/integers/u8/no_space_between_literal.leo
new file mode 100644
index 0000000000..64eba2a083
--- /dev/null
+++ b/compiler/tests/integers/u8/no_space_between_literal.leo
@@ -0,0 +1,3 @@
+function main() {
+  const i = 1 u8;
+}
\ No newline at end of file
diff --git a/grammar/abnf-grammar.md b/grammar/abnf-grammar.md
index eb587deea0..29b757e6a4 100644
--- a/grammar/abnf-grammar.md
+++ b/grammar/abnf-grammar.md
@@ -494,7 +494,7 @@ Line terminators form whitespace, along with spaces and horizontal tabs.
 whitespace = space / horizontal-tab / newline
 ```
 
-Go to: _[horizontal-tab](#user-content-horizontal-tab), [newline](#user-content-newline), [space](#user-content-space)_;
+Go to: _[newline](#user-content-newline), [horizontal-tab](#user-content-horizontal-tab), [space](#user-content-space)_;
 
 
 There are two kinds of comments in Leo, as in other languages.
@@ -512,7 +512,7 @@ the ones used in the Java language specification.
 comment = block-comment / end-of-line-comment
 ```
 
-Go to: _[block-comment](#user-content-block-comment), [end-of-line-comment](#user-content-end-of-line-comment)_;
+Go to: _[end-of-line-comment](#user-content-end-of-line-comment), [block-comment](#user-content-block-comment)_;
 
 
 <a name="block-comment"></a>
@@ -529,7 +529,7 @@ rest-of-block-comment = "*" rest-of-block-comment-after-star
                       / not-star rest-of-block-comment
 ```
 
-Go to: _[not-star](#user-content-not-star), [rest-of-block-comment-after-star](#user-content-rest-of-block-comment-after-star), [rest-of-block-comment](#user-content-rest-of-block-comment)_;
+Go to: _[rest-of-block-comment-after-star](#user-content-rest-of-block-comment-after-star), [not-star](#user-content-not-star), [rest-of-block-comment](#user-content-rest-of-block-comment)_;
 
 
 <a name="rest-of-block-comment-after-star"></a>
@@ -539,7 +539,7 @@ rest-of-block-comment-after-star = "/"
                                  / not-star-or-slash rest-of-block-comment
 ```
 
-Go to: _[rest-of-block-comment-after-star](#user-content-rest-of-block-comment-after-star), [rest-of-block-comment](#user-content-rest-of-block-comment), [not-star-or-slash](#user-content-not-star-or-slash)_;
+Go to: _[rest-of-block-comment-after-star](#user-content-rest-of-block-comment-after-star), [not-star-or-slash](#user-content-not-star-or-slash), [rest-of-block-comment](#user-content-rest-of-block-comment)_;
 
 
 <a name="end-of-line-comment"></a>
@@ -614,7 +614,7 @@ lowercase-letter = %x61-7A   ; a-z
 letter = uppercase-letter / lowercase-letter
 ```
 
-Go to: _[lowercase-letter](#user-content-lowercase-letter), [uppercase-letter](#user-content-uppercase-letter)_;
+Go to: _[uppercase-letter](#user-content-uppercase-letter), [lowercase-letter](#user-content-lowercase-letter)_;
 
 
 An identifier is a non-empty sequence of letters, digits, and underscores,
@@ -784,12 +784,8 @@ Go to: _[integer](#user-content-integer)_;
 There are two kinds of group literals.
 One is a single integer followed by the type of group elements,
 which denotes the scalar product of the generator point by the integer.
-The other is a pair of integer coordinates,
-which are reduced modulo the prime to identify a point,
-which must be on the elliptic curve.
-It is also allowed to omit one (not both) coordinates,
-with an indication of how to infer the missing coordinate
-(i.e. sign high, sign low, or inferred).
+The other kind is not a token because it allows some whitespace inside;
+therefore, it is defined in the syntactic grammar.
 
 <a name="product-group-literal"></a>
 ```abnf
@@ -799,97 +795,6 @@ product-group-literal = integer "group"
 Go to: _[integer](#user-content-integer)_;
 
 
-<a name="group-coordinate"></a>
-```abnf
-group-coordinate = integer / "+" / "-" / "_"
-```
-
-Go to: _[integer](#user-content-integer)_;
-
-
-<a name="affine-group-literal"></a>
-```abnf
-affine-group-literal = "(" group-coordinate "," group-coordinate ")" "group"
-```
-
-Go to: _[group-coordinate](#user-content-group-coordinate)_;
-
-
-<a name="group-literal"></a>
-```abnf
-group-literal = product-group-literal / affine-group-literal
-```
-
-Go to: _[product-group-literal](#user-content-product-group-literal), [affine-group-literal](#user-content-affine-group-literal)_;
-
-
-Note that the rule for group literals above
-allows no whitespace between coordinates.
-If we want to allow whitespace,
-e.g. '(3, 4)group' as opposed to requiring '(3,4)group',
-then we should define affine group literals
-in the syntactic grammar instead of in the lexical grammar.
-We should have a notion of atomic literal in the lexical grammar,
-and (non-atomic) literal in the syntactic grammar.
-The lexical grammar should define a token for ')group'
-if we want no whitespace between the closing parenthesis and 'group'.
-More precisely, the rule for 'literal' below in the lexical grammar
-would be replaced with
-
-
-
-```
-atomic-literal = ... / product-group-literal
-```
-
-
-
-where the '...' stands for all the '...-literal' alternatives
-in the current rule for 'literal' below, except 'group-literal'.
-Furthermore, the rule for 'symbol' below in the lexical grammar
-would be extended to
-
-
-
-```
-symbol = ... / ")group"
-```
-
-
-
-where '...' stands for the current definiens of the rule.
-We would also have to adjust the rule for 'token' below in the lexical grammar
-to reference 'atomic-literal' instead of 'literal' in the definiens.
-We would then add to the syntactic grammar the following rules
-
-
-
-```
-affine-group-literal = "(" group-coordinate "," group-coordinate ")group"
-```
-
-
-
-```
-literal = atomic-literal / affine-group-literal
-```
-
-
-
-which would now define literals in the syntactic grammar.
-Note that now an affine group literal would have the form
-
-
-
-```
-( <ow> <coordinate> <ow> , <ow> <coordinate> <ow> )group
-```
-
-
-
-where <ow> is optional whitespace;
-however, no whitespace is allowed between the closing ')' and 'group'.
-
 Boolean literals are the usual two.
 
 <a name="boolean-literal"></a>
@@ -910,18 +815,18 @@ Go to: _[address](#user-content-address)_;
 
 The ones above are all the literals, as defined by the following rule.
 
-<a name="literal"></a>
+<a name="atomic-literal"></a>
 ```abnf
-literal = untyped-literal
-        / unsigned-literal
-        / signed-literal
-        / field-literal
-        / group-literal
-        / boolean-literal
-        / address-literal
+atomic-literal = untyped-literal
+               / unsigned-literal
+               / signed-literal
+               / field-literal
+               / product-group-literal
+               / boolean-literal
+               / address-literal
 ```
 
-Go to: _[unsigned-literal](#user-content-unsigned-literal), [untyped-literal](#user-content-untyped-literal), [field-literal](#user-content-field-literal), [group-literal](#user-content-group-literal), [address-literal](#user-content-address-literal), [boolean-literal](#user-content-boolean-literal), [signed-literal](#user-content-signed-literal)_;
+Go to: _[unsigned-literal](#user-content-unsigned-literal), [field-literal](#user-content-field-literal), [address-literal](#user-content-address-literal), [signed-literal](#user-content-signed-literal), [product-group-literal](#user-content-product-group-literal), [boolean-literal](#user-content-boolean-literal), [untyped-literal](#user-content-untyped-literal)_;
 
 
 After defining the (mostly) alphanumeric tokens above,
@@ -929,6 +834,14 @@ it remains to define tokens for non-alphanumeric symbols such as "+" and "(".
 Different programming languages used different terminologies for these,
 e.g. operators, separators, punctuators, etc.
 Here we use 'symbol', for all of them, but we can do something different.
+We also include a token consisting of
+a closing parenthesis immediately followed by 'group':
+as defined in the syntactic grammar,
+this is the final part of an affine group literal.
+Even though it includes letters,
+it seems appropriate to still consider it a symbol,
+particularly since it starts with a symbol.
+
 We could give names to all of these symbols,
 via rules such as
 
@@ -969,6 +882,7 @@ symbol = "!" / "&&" / "||"
        / "{" / "}"
        / "," / "." / ".." / "..." / ";" / ":" / "::" / "?"
        / "->" / "_"
+       / ")group"
 ```
 
 Everything defined above, other than comments and whitespace,
@@ -978,14 +892,14 @@ is a token, as defined by the following rule.
 ```abnf
 token = keyword
       / identifier
-      / literal
+      / atomic-literal
       / package-name
       / formatted-string
       / annotation-name
       / symbol
 ```
 
-Go to: _[literal](#user-content-literal), [annotation-name](#user-content-annotation-name), [formatted-string](#user-content-formatted-string), [symbol](#user-content-symbol), [identifier](#user-content-identifier), [keyword](#user-content-keyword), [package-name](#user-content-package-name)_;
+Go to: _[identifier](#user-content-identifier), [atomic-literal](#user-content-atomic-literal), [package-name](#user-content-package-name), [symbol](#user-content-symbol), [annotation-name](#user-content-annotation-name), [formatted-string](#user-content-formatted-string), [keyword](#user-content-keyword)_;
 
 
 
@@ -1021,7 +935,7 @@ signed-type = "i8" / "i16" / "i32" / "i64" / "i128"
 integer-type = unsigned-type / signed-type
 ```
 
-Go to: _[unsigned-type](#user-content-unsigned-type), [signed-type](#user-content-signed-type)_;
+Go to: _[signed-type](#user-content-signed-type), [unsigned-type](#user-content-unsigned-type)_;
 
 
 The integer types, along with the field and group types,
@@ -1063,7 +977,7 @@ address-type = "address"
 scalar-type =  boolean-type / arithmetic-type / address-type
 ```
 
-Go to: _[arithmetic-type](#user-content-arithmetic-type), [address-type](#user-content-address-type), [boolean-type](#user-content-boolean-type)_;
+Go to: _[arithmetic-type](#user-content-arithmetic-type), [boolean-type](#user-content-boolean-type), [address-type](#user-content-address-type)_;
 
 
 Circuit types are denoted by identifiers and the keyword 'Self'.
@@ -1125,7 +1039,7 @@ i.e. types whose values contain (sub)values
 aggregate-type = tuple-type / array-type / circuit-type
 ```
 
-Go to: _[array-type](#user-content-array-type), [tuple-type](#user-content-tuple-type), [circuit-type](#user-content-circuit-type)_;
+Go to: _[circuit-type](#user-content-circuit-type), [tuple-type](#user-content-tuple-type), [array-type](#user-content-array-type)_;
 
 
 Scalar and aggregate types form all the types.
@@ -1135,7 +1049,58 @@ Scalar and aggregate types form all the types.
 type = scalar-type / aggregate-type
 ```
 
-Go to: _[scalar-type](#user-content-scalar-type), [aggregate-type](#user-content-aggregate-type)_;
+Go to: _[aggregate-type](#user-content-aggregate-type), [scalar-type](#user-content-scalar-type)_;
+
+
+The lexical grammar above defines product group literals.
+The other kind of group literal is a pair of integer coordinates,
+which are reduced modulo the prime to identify a point,
+which must be on the elliptic curve.
+It is also allowed to omit one (not both) coordinates,
+with an indication of how to infer the missing coordinate
+(i.e. sign high, sign low, or inferred).
+This is an affine group literal,
+because it consists of affine point coordinates.
+
+<a name="group-coordinate"></a>
+```abnf
+group-coordinate = integer / "+" / "-" / "_"
+```
+
+Go to: _[integer](#user-content-integer)_;
+
+
+<a name="affine-group-literal"></a>
+```abnf
+affine-group-literal = "(" group-coordinate "," group-coordinate ")" "group"
+```
+
+Go to: _[group-coordinate](#user-content-group-coordinate)_;
+
+
+A literal is either an atomic one or an affine group literal.
+Here 'atomic' refers to being a token or not,
+since no whitespace is allowed within a token.
+
+<a name="literal"></a>
+```abnf
+literal = atomic-literal / affine-group-literal
+```
+
+Go to: _[atomic-literal](#user-content-atomic-literal), [affine-group-literal](#user-content-affine-group-literal)_;
+
+
+The following rule is not directly referenced in the rules for expressions
+(which reference 'literal' instead),
+but it is useful to establish terminology:
+a group literal is either a product group literal or an affine group literal.
+
+<a name="group-literal"></a>
+```abnf
+group-literal = product-group-literal / affine-group-literal
+```
+
+Go to: _[product-group-literal](#user-content-product-group-literal), [affine-group-literal](#user-content-affine-group-literal)_;
 
 
 As often done in grammatical language syntax specifications,
@@ -1166,7 +1131,7 @@ primary-expression = identifier
                    / function-call
 ```
 
-Go to: _[function-call](#user-content-function-call), [expression](#user-content-expression), [circuit-expression](#user-content-circuit-expression), [identifier](#user-content-identifier), [literal](#user-content-literal), [array-expression](#user-content-array-expression), [tuple-expression](#user-content-tuple-expression)_;
+Go to: _[function-call](#user-content-function-call), [literal](#user-content-literal), [array-expression](#user-content-array-expression), [circuit-expression](#user-content-circuit-expression), [identifier](#user-content-identifier), [expression](#user-content-expression), [tuple-expression](#user-content-tuple-expression)_;
 
 
 There are tuple expressions to construct and deconstruct tuples.
@@ -1235,7 +1200,7 @@ Go to: _[expression](#user-content-expression)_;
 array-repeat-construction = "[" expression ";" array-dimensions "]"
 ```
 
-Go to: _[expression](#user-content-expression), [array-dimensions](#user-content-array-dimensions)_;
+Go to: _[array-dimensions](#user-content-array-dimensions), [expression](#user-content-expression)_;
 
 
 <a name="array-construction"></a>
@@ -1243,7 +1208,7 @@ Go to: _[expression](#user-content-expression), [array-dimensions](#user-content
 array-construction = array-inline-construction / array-repeat-construction
 ```
 
-Go to: _[array-inline-construction](#user-content-array-inline-construction), [array-repeat-construction](#user-content-array-repeat-construction)_;
+Go to: _[array-repeat-construction](#user-content-array-repeat-construction), [array-inline-construction](#user-content-array-inline-construction)_;
 
 
 <a name="array-expression"></a>
@@ -1273,7 +1238,7 @@ circuit-construction = circuit-type "{"
                        "}"
 ```
 
-Go to: _[circuit-inline-element](#user-content-circuit-inline-element), [circuit-type](#user-content-circuit-type)_;
+Go to: _[circuit-type](#user-content-circuit-type), [circuit-inline-element](#user-content-circuit-inline-element)_;
 
 
 <a name="circuit-inline-element"></a>
@@ -1331,7 +1296,7 @@ unary-expression = postfix-expression
                  / "-" unary-expression
 ```
 
-Go to: _[postfix-expression](#user-content-postfix-expression), [unary-expression](#user-content-unary-expression)_;
+Go to: _[unary-expression](#user-content-unary-expression), [postfix-expression](#user-content-postfix-expression)_;
 
 
 Next in the operator precedence is casting.
@@ -1375,7 +1340,7 @@ additive-expression = multiplicative-expression
                     / additive-expression "-" multiplicative-expression
 ```
 
-Go to: _[additive-expression](#user-content-additive-expression), [multiplicative-expression](#user-content-multiplicative-expression)_;
+Go to: _[multiplicative-expression](#user-content-multiplicative-expression), [additive-expression](#user-content-additive-expression)_;
 
 
 Next in the precedence order are ordering relations.
@@ -1425,7 +1390,7 @@ disjunctive-expression = conjunctive-expression
                        / disjunctive-expression "||" conjunctive-expression
 ```
 
-Go to: _[disjunctive-expression](#user-content-disjunctive-expression), [conjunctive-expression](#user-content-conjunctive-expression)_;
+Go to: _[conjunctive-expression](#user-content-conjunctive-expression), [disjunctive-expression](#user-content-disjunctive-expression)_;
 
 
 Finally we have conditional expressions.
@@ -1438,7 +1403,7 @@ conditional-expression = disjunctive-expression
                          ":" conditional-expression
 ```
 
-Go to: _[conditional-expression](#user-content-conditional-expression), [expression](#user-content-expression), [disjunctive-expression](#user-content-disjunctive-expression)_;
+Go to: _[conditional-expression](#user-content-conditional-expression), [disjunctive-expression](#user-content-disjunctive-expression), [expression](#user-content-expression)_;
 
 
 These are all the expressions.
@@ -1471,7 +1436,7 @@ statement = expression-statement
           / block
 ```
 
-Go to: _[variable-definition-statement](#user-content-variable-definition-statement), [loop-statement](#user-content-loop-statement), [block](#user-content-block), [assignment-statement](#user-content-assignment-statement), [console-statement](#user-content-console-statement), [conditional-statement](#user-content-conditional-statement), [return-statement](#user-content-return-statement), [expression-statement](#user-content-expression-statement)_;
+Go to: _[variable-definition-statement](#user-content-variable-definition-statement), [conditional-statement](#user-content-conditional-statement), [console-statement](#user-content-console-statement), [return-statement](#user-content-return-statement), [expression-statement](#user-content-expression-statement), [loop-statement](#user-content-loop-statement), [assignment-statement](#user-content-assignment-statement), [block](#user-content-block)_;
 
 
 <a name="block"></a>
@@ -1514,7 +1479,7 @@ variable-definition-statement = ( "let" / "const" )
                                 [ ":" type ] "=" expression ";"
 ```
 
-Go to: _[identifier-or-identifiers](#user-content-identifier-or-identifiers), [type](#user-content-type), [expression](#user-content-expression)_;
+Go to: _[expression](#user-content-expression), [type](#user-content-type), [identifier-or-identifiers](#user-content-identifier-or-identifiers)_;
 
 
 <a name="identifier-or-identifiers"></a>
@@ -1547,7 +1512,7 @@ conditional-statement = branch
                       / branch "else" conditional-statement
 ```
 
-Go to: _[block](#user-content-block), [branch](#user-content-branch), [conditional-statement](#user-content-conditional-statement)_;
+Go to: _[conditional-statement](#user-content-conditional-statement), [branch](#user-content-branch), [block](#user-content-block)_;
 
 
 A loop statement implicitly defines a loop variable
@@ -1576,7 +1541,7 @@ assignment-operator = "=" / "+=" / "-=" / "*=" / "/=" / "**="
 assignment-statement = expression assignment-operator expression ";"
 ```
 
-Go to: _[assignment-operator](#user-content-assignment-operator), [expression](#user-content-expression)_;
+Go to: _[expression](#user-content-expression), [assignment-operator](#user-content-assignment-operator)_;
 
 
 Console statements start with the 'console' keyword,
@@ -1632,7 +1597,7 @@ Go to: _[formatted-string](#user-content-formatted-string)_;
 print-call = print-function print-arguments
 ```
 
-Go to: _[print-function](#user-content-print-function), [print-arguments](#user-content-print-arguments)_;
+Go to: _[print-arguments](#user-content-print-arguments), [print-function](#user-content-print-function)_;
 
 
 An annotation consists of an annotation name (which starts with '@')
@@ -1645,7 +1610,7 @@ annotation = annotation-name
              [ "(" identifier *( "," identifier ) ")" ]
 ```
 
-Go to: _[identifier](#user-content-identifier), [annotation-name](#user-content-annotation-name)_;
+Go to: _[annotation-name](#user-content-annotation-name), [identifier](#user-content-identifier)_;
 
 
 A function declaration defines a function.
@@ -1667,7 +1632,7 @@ function-declaration = *annotation "function" identifier
                        block
 ```
 
-Go to: _[function-parameters](#user-content-function-parameters), [type](#user-content-type), [block](#user-content-block), [identifier](#user-content-identifier)_;
+Go to: _[type](#user-content-type), [function-parameters](#user-content-function-parameters), [identifier](#user-content-identifier), [block](#user-content-block)_;
 
 
 <a name="function-parameters"></a>
@@ -1678,7 +1643,7 @@ function-parameters = self-parameter [ "," input-parameter ]
                     / input-parameter
 ```
 
-Go to: _[input-parameter](#user-content-input-parameter), [function-inputs](#user-content-function-inputs), [self-parameter](#user-content-self-parameter)_;
+Go to: _[function-inputs](#user-content-function-inputs), [input-parameter](#user-content-input-parameter), [self-parameter](#user-content-self-parameter)_;
 
 
 <a name="self-parameter"></a>
@@ -1726,7 +1691,7 @@ Go to: _[member-variable-declaration](#user-content-member-variable-declaration)
 member-variable-declaration = identifier ":" type
 ```
 
-Go to: _[identifier](#user-content-identifier), [type](#user-content-type)_;
+Go to: _[type](#user-content-type), [identifier](#user-content-identifier)_;
 
 
 <a name="member-function-declaration"></a>
@@ -1794,7 +1759,7 @@ declaration = import-declaration
             / circuit-declaration
 ```
 
-Go to: _[import-declaration](#user-content-import-declaration), [function-declaration](#user-content-function-declaration), [circuit-declaration](#user-content-circuit-declaration)_;
+Go to: _[function-declaration](#user-content-function-declaration), [circuit-declaration](#user-content-circuit-declaration), [import-declaration](#user-content-import-declaration)_;
 
 
 <a name="file"></a>
diff --git a/grammar/abnf-grammar.txt b/grammar/abnf-grammar.txt
index a397d1630b..cca93278d3 100644
--- a/grammar/abnf-grammar.txt
+++ b/grammar/abnf-grammar.txt
@@ -540,60 +540,11 @@ field-literal = integer %s"field"
 ; There are two kinds of group literals.
 ; One is a single integer followed by the type of group elements,
 ; which denotes the scalar product of the generator point by the integer.
-; The other is a pair of integer coordinates,
-; which are reduced modulo the prime to identify a point,
-; which must be on the elliptic curve.
-; It is also allowed to omit one (not both) coordinates,
-; with an indication of how to infer the missing coordinate
-; (i.e. sign high, sign low, or inferred).
+; The other kind is not a token because it allows some whitespace inside;
+; therefore, it is defined in the syntactic grammar.
 
 product-group-literal = integer %s"group"
 
-group-coordinate = integer / "+" / "-" / "_"
-
-affine-group-literal = "(" group-coordinate "," group-coordinate ")" %s"group"
-
-group-literal = product-group-literal / affine-group-literal
-
-; Note that the rule for group literals above
-; allows no whitespace between coordinates.
-; If we want to allow whitespace,
-; e.g. '(3, 4)group' as opposed to requiring '(3,4)group',
-; then we should define affine group literals
-; in the syntactic grammar instead of in the lexical grammar.
-; We should have a notion of atomic literal in the lexical grammar,
-; and (non-atomic) literal in the syntactic grammar.
-; The lexical grammar should define a token for ')group'
-; if we want no whitespace between the closing parenthesis and 'group'.
-; More precisely, the rule for 'literal' below in the lexical grammar
-; would be replaced with
-;
-;  atomic-literal = ... / product-group-literal
-;
-; where the '...' stands for all the '...-literal' alternatives
-; in the current rule for 'literal' below, except 'group-literal'.
-; Furthermore, the rule for 'symbol' below in the lexical grammar
-; would be extended to
-;
-;  symbol = ... / %s")group"
-;
-; where '...' stands for the current definiens of the rule.
-; We would also have to adjust the rule for 'token' below in the lexical grammar
-; to reference 'atomic-literal' instead of 'literal' in the definiens.
-; We would then add to the syntactic grammar the following rules
-;
-;  affine-group-literal = "(" group-coordinate "," group-coordinate %s")group"
-;
-;  literal = atomic-literal / affine-group-literal
-;
-; which would now define literals in the syntactic grammar.
-; Note that now an affine group literal would have the form
-;
-;  ( <ow> <coordinate> <ow> , <ow> <coordinate> <ow> )group
-;
-; where <ow> is optional whitespace;
-; however, no whitespace is allowed between the closing ')' and 'group'.
-
 ; Boolean literals are the usual two.
 
 boolean-literal = %s"true" / %s"false"
@@ -605,19 +556,27 @@ address-literal = %s"address" "(" address ")"
 
 ; The ones above are all the literals, as defined by the following rule.
 
-literal = untyped-literal
-        / unsigned-literal
-        / signed-literal
-        / field-literal
-        / group-literal
-        / boolean-literal
-        / address-literal
+atomic-literal = untyped-literal
+               / unsigned-literal
+               / signed-literal
+               / field-literal
+               / product-group-literal
+               / boolean-literal
+               / address-literal
 
 ; After defining the (mostly) alphanumeric tokens above,
 ; it remains to define tokens for non-alphanumeric symbols such as "+" and "(".
 ; Different programming languages used different terminologies for these,
 ; e.g. operators, separators, punctuators, etc.
 ; Here we use 'symbol', for all of them, but we can do something different.
+; We also include a token consisting of
+; a closing parenthesis immediately followed by 'group':
+; as defined in the syntactic grammar,
+; this is the final part of an affine group literal.
+; Even though it includes letters,
+; it seems appropriate to still consider it a symbol,
+; particularly since it starts with a symbol.
+
 ; We could give names to all of these symbols,
 ; via rules such as
 ;
@@ -644,13 +603,14 @@ symbol = "!" / "&&" / "||"
        / "{" / "}"
        / "," / "." / ".." / "..." / ";" / ":" / "::" / "?"
        / "->" / "_"
+       / %s")group"
 
 ; Everything defined above, other than comments and whitespace,
 ; is a token, as defined by the following rule.
 
 token = keyword
       / identifier
-      / literal
+      / atomic-literal
       / package-name
       / formatted-string
       / annotation-name
@@ -729,6 +689,33 @@ aggregate-type = tuple-type / array-type / circuit-type
 
 type = scalar-type / aggregate-type
 
+; The lexical grammar above defines product group literals.
+; The other kind of group literal is a pair of integer coordinates,
+; which are reduced modulo the prime to identify a point,
+; which must be on the elliptic curve.
+; It is also allowed to omit one (not both) coordinates,
+; with an indication of how to infer the missing coordinate
+; (i.e. sign high, sign low, or inferred).
+; This is an affine group literal,
+; because it consists of affine point coordinates.
+
+group-coordinate = integer / "+" / "-" / "_"
+
+affine-group-literal = "(" group-coordinate "," group-coordinate ")" %s"group"
+
+; A literal is either an atomic one or an affine group literal.
+; Here 'atomic' refers to being a token or not,
+; since no whitespace is allowed within a token.
+
+literal = atomic-literal / affine-group-literal
+
+; The following rule is not directly referenced in the rules for expressions
+; (which reference 'literal' instead),
+; but it is useful to establish terminology:
+; a group literal is either a product group literal or an affine group literal.
+
+group-literal = product-group-literal / affine-group-literal
+
 ; As often done in grammatical language syntax specifications,
 ; we define rules for different kinds of expressions,
 ; which also defines the relative precedence
diff --git a/imports/src/errors/import_parser.rs b/imports/src/errors/import_parser.rs
index 87a3e8b03c..0c04467403 100644
--- a/imports/src/errors/import_parser.rs
+++ b/imports/src/errors/import_parser.rs
@@ -15,15 +15,12 @@
 // along with the Leo library. If not, see <https://www.gnu.org/licenses/>.
 use leo_asg::AsgConvertError;
 use leo_ast::{FormattedError, Identifier, LeoError, Span};
-use leo_parser::{DeprecatedError, SyntaxError};
+use leo_parser::SyntaxError;
 
 use std::{io, path::Path};
 
 #[derive(Debug, Error)]
 pub enum ImportParserError {
-    #[error("{}", _0)]
-    DeprecatedError(#[from] DeprecatedError),
-
     #[error("{}", _0)]
     Error(#[from] FormattedError),
 
@@ -40,7 +37,6 @@ impl Into<AsgConvertError> for ImportParserError {
         match self {
             ImportParserError::Error(x) => AsgConvertError::ImportError(x),
             ImportParserError::SyntaxError(x) => x.into(),
-            ImportParserError::DeprecatedError(x) => AsgConvertError::SyntaxError(SyntaxError::DeprecatedError(x)),
             ImportParserError::AsgConvertError(x) => x,
         }
     }
@@ -66,15 +62,6 @@ impl ImportParserError {
         Self::new_from_span(message, span)
     }
 
-    ///
-    /// A core package name has been imported twice.
-    ///
-    pub fn duplicate_core_package(identifier: Identifier) -> Self {
-        let message = format!("Duplicate core_package import `{}`.", identifier.name);
-
-        Self::new_from_span(message, &identifier.span)
-    }
-
     ///
     /// Failed to convert a file path into an os string.
     ///
@@ -106,15 +93,6 @@ impl ImportParserError {
         Self::new_from_span(message, span)
     }
 
-    ///
-    /// Failed to import all symbols at a package path.
-    ///
-    pub fn star(path: &Path, span: &Span) -> Self {
-        let message = format!("Cannot import `*` from path `{:?}`.", path);
-
-        Self::new_from_span(message, span)
-    }
-
     ///
     /// Failed to find a library file for the current package.
     ///
diff --git a/package/src/errors/imports/directory.rs b/package/src/errors/imports/directory.rs
index 05e6cfadf2..e027babdec 100644
--- a/package/src/errors/imports/directory.rs
+++ b/package/src/errors/imports/directory.rs
@@ -14,34 +14,16 @@
 // You should have received a copy of the GNU General Public License
 // along with the Leo library. If not, see <https://www.gnu.org/licenses/>.
 
-use std::{ffi::OsString, fs::FileType, io};
+use std::{ffi::OsString, io};
 
 #[derive(Debug, Error)]
 pub enum ImportsDirectoryError {
     #[error("creating: {}", _0)]
     Creating(io::Error),
 
-    #[error("file entry getting: {}", _0)]
-    GettingFileEntry(io::Error),
-
-    #[error("file {:?} extension getting", _0)]
-    GettingFileExtension(OsString),
-
-    #[error("file {:?} type getting: {}", _0, _1)]
-    GettingFileType(OsString, io::Error),
-
     #[error("package {:?} does not exist as an import", _0)]
     ImportDoesNotExist(OsString),
 
-    #[error("invalid file {:?} extension: {:?}", _0, _1)]
-    InvalidFileExtension(OsString, OsString),
-
-    #[error("invalid file {:?} type: {:?}", _0, _1)]
-    InvalidFileType(OsString, FileType),
-
-    #[error("reading: {}", _0)]
-    Reading(io::Error),
-
     #[error("removing: {}", _0)]
     Removing(io::Error),
 }
diff --git a/package/src/errors/inputs/directory.rs b/package/src/errors/inputs/directory.rs
index ee93d8c6d6..0a36e53836 100644
--- a/package/src/errors/inputs/directory.rs
+++ b/package/src/errors/inputs/directory.rs
@@ -26,9 +26,6 @@ pub enum InputsDirectoryError {
     #[error("file entry getting: {}", _0)]
     GettingFileEntry(io::Error),
 
-    #[error("file {:?} extension getting", _0)]
-    GettingFileExtension(OsString),
-
     #[error("file {:?} name getting", _0)]
     GettingFileName(OsString),
 
@@ -38,9 +35,6 @@ pub enum InputsDirectoryError {
     #[error("{}", _0)]
     InputFileError(#[from] InputFileError),
 
-    #[error("invalid file {:?} extension: {:?}", _0, _1)]
-    InvalidFileExtension(String, OsString),
-
     #[error("invalid file {:?} type: {:?}", _0, _1)]
     InvalidFileType(OsString, FileType),
 
diff --git a/package/src/errors/inputs/input.rs b/package/src/errors/inputs/input.rs
index 659fcf32a3..0dbbbb0a0c 100644
--- a/package/src/errors/inputs/input.rs
+++ b/package/src/errors/inputs/input.rs
@@ -14,21 +14,15 @@
 // You should have received a copy of the GNU General Public License
 // along with the Leo library. If not, see <https://www.gnu.org/licenses/>.
 
-use std::{io, path::PathBuf};
+use std::path::PathBuf;
 
 #[derive(Debug, Error)]
 pub enum InputFileError {
     #[error("{}: {}", _0, _1)]
     Crate(&'static str, String),
 
-    #[error("creating: {}", _0)]
-    Creating(io::Error),
-
     #[error("Cannot read from the provided file path - {:?}", _0)]
     FileReadError(PathBuf),
-
-    #[error("writing: {}", _0)]
-    Writing(io::Error),
 }
 
 impl From<std::io::Error> for InputFileError {
diff --git a/package/src/errors/inputs/state.rs b/package/src/errors/inputs/state.rs
index 8d56253b46..e944f93469 100644
--- a/package/src/errors/inputs/state.rs
+++ b/package/src/errors/inputs/state.rs
@@ -14,21 +14,15 @@
 // You should have received a copy of the GNU General Public License
 // along with the Leo library. If not, see <https://www.gnu.org/licenses/>.
 
-use std::{io, path::PathBuf};
+use std::path::PathBuf;
 
 #[derive(Debug, Error)]
 pub enum StateFileError {
     #[error("{}: {}", _0, _1)]
     Crate(&'static str, String),
 
-    #[error("creating: {}", _0)]
-    Creating(io::Error),
-
     #[error("Cannot read from the provided file path - {:?}", _0)]
     FileReadError(PathBuf),
-
-    #[error("writing: {}", _0)]
-    Writing(io::Error),
 }
 
 impl From<std::io::Error> for StateFileError {
diff --git a/package/src/errors/outputs/checksum.rs b/package/src/errors/outputs/checksum.rs
index e2a6bd4375..7d52981bd8 100644
--- a/package/src/errors/outputs/checksum.rs
+++ b/package/src/errors/outputs/checksum.rs
@@ -14,24 +14,18 @@
 // You should have received a copy of the GNU General Public License
 // along with the Leo library. If not, see <https://www.gnu.org/licenses/>.
 
-use std::{io, path::PathBuf};
+use std::path::PathBuf;
 
 #[derive(Debug, Error)]
 pub enum ChecksumFileError {
     #[error("{}: {}", _0, _1)]
     Crate(&'static str, String),
 
-    #[error("creating: {}", _0)]
-    Creating(io::Error),
-
     #[error("Cannot read from the provided file path - {:?}", _0)]
     FileReadError(PathBuf),
 
     #[error("Cannot remove the provided file - {:?}", _0)]
     FileRemovalError(PathBuf),
-
-    #[error("writing: {}", _0)]
-    Writing(io::Error),
 }
 
 impl From<std::io::Error> for ChecksumFileError {
diff --git a/package/src/errors/outputs/circuit.rs b/package/src/errors/outputs/circuit.rs
index ce006c88f2..69b172c05d 100644
--- a/package/src/errors/outputs/circuit.rs
+++ b/package/src/errors/outputs/circuit.rs
@@ -14,24 +14,18 @@
 // You should have received a copy of the GNU General Public License
 // along with the Leo library. If not, see <https://www.gnu.org/licenses/>.
 
-use std::{io, path::PathBuf};
+use std::path::PathBuf;
 
 #[derive(Debug, Error)]
 pub enum CircuitFileError {
     #[error("{}: {}", _0, _1)]
     Crate(&'static str, String),
 
-    #[error("creating: {}", _0)]
-    Creating(io::Error),
-
     #[error("Cannot read from the provided file path - {:?}", _0)]
     FileReadError(PathBuf),
 
     #[error("Cannot remove the provided file - {:?}", _0)]
     FileRemovalError(PathBuf),
-
-    #[error("writing: {}", _0)]
-    Writing(io::Error),
 }
 
 impl From<std::io::Error> for CircuitFileError {
diff --git a/package/src/errors/outputs/directory.rs b/package/src/errors/outputs/directory.rs
index 2e16be7e39..8945526cfa 100644
--- a/package/src/errors/outputs/directory.rs
+++ b/package/src/errors/outputs/directory.rs
@@ -14,31 +14,13 @@
 // You should have received a copy of the GNU General Public License
 // along with the Leo library. If not, see <https://www.gnu.org/licenses/>.
 
-use std::{ffi::OsString, fs::FileType, io};
+use std::io;
 
 #[derive(Debug, Error)]
 pub enum OutputsDirectoryError {
     #[error("creating: {}", _0)]
     Creating(io::Error),
 
-    #[error("file entry getting: {}", _0)]
-    GettingFileEntry(io::Error),
-
-    #[error("file {:?} extension getting", _0)]
-    GettingFileExtension(OsString),
-
-    #[error("file {:?} type getting: {}", _0, _1)]
-    GettingFileType(OsString, io::Error),
-
-    #[error("invalid file {:?} extension: {:?}", _0, _1)]
-    InvalidFileExtension(OsString, OsString),
-
-    #[error("invalid file {:?} type: {:?}", _0, _1)]
-    InvalidFileType(OsString, FileType),
-
-    #[error("reading: {}", _0)]
-    Reading(io::Error),
-
     #[error("removing: {}", _0)]
     Removing(io::Error),
 }
diff --git a/package/src/errors/outputs/proof.rs b/package/src/errors/outputs/proof.rs
index 461a3ab259..adce5b7bf5 100644
--- a/package/src/errors/outputs/proof.rs
+++ b/package/src/errors/outputs/proof.rs
@@ -14,24 +14,18 @@
 // You should have received a copy of the GNU General Public License
 // along with the Leo library. If not, see <https://www.gnu.org/licenses/>.
 
-use std::{io, path::PathBuf};
+use std::path::PathBuf;
 
 #[derive(Debug, Error)]
 pub enum ProofFileError {
     #[error("{}: {}", _0, _1)]
     Crate(&'static str, String),
 
-    #[error("creating: {}", _0)]
-    Creating(io::Error),
-
     #[error("Cannot read from the provided file path - {:?}", _0)]
     FileReadError(PathBuf),
 
     #[error("Cannot remove the provided file - {:?}", _0)]
     FileRemovalError(PathBuf),
-
-    #[error("writing: {}", _0)]
-    Writing(io::Error),
 }
 
 impl From<std::io::Error> for ProofFileError {
diff --git a/package/src/errors/outputs/proving_key.rs b/package/src/errors/outputs/proving_key.rs
index d78e62aec4..6431af92fd 100644
--- a/package/src/errors/outputs/proving_key.rs
+++ b/package/src/errors/outputs/proving_key.rs
@@ -14,24 +14,18 @@
 // You should have received a copy of the GNU General Public License
 // along with the Leo library. If not, see <https://www.gnu.org/licenses/>.
 
-use std::{io, path::PathBuf};
+use std::path::PathBuf;
 
 #[derive(Debug, Error)]
 pub enum ProvingKeyFileError {
     #[error("{}: {}", _0, _1)]
     Crate(&'static str, String),
 
-    #[error("creating: {}", _0)]
-    Creating(io::Error),
-
     #[error("Cannot read from the provided file path - {:?}", _0)]
     FileReadError(PathBuf),
 
     #[error("Cannot remove the provided file - {:?}", _0)]
     FileRemovalError(PathBuf),
-
-    #[error("writing: {}", _0)]
-    Writing(io::Error),
 }
 
 impl From<std::io::Error> for ProvingKeyFileError {
diff --git a/package/src/errors/outputs/verification_key.rs b/package/src/errors/outputs/verification_key.rs
index 51398e630c..6a9f437b8a 100644
--- a/package/src/errors/outputs/verification_key.rs
+++ b/package/src/errors/outputs/verification_key.rs
@@ -14,27 +14,18 @@
 // You should have received a copy of the GNU General Public License
 // along with the Leo library. If not, see <https://www.gnu.org/licenses/>.
 
-use std::{io, path::PathBuf};
+use std::path::PathBuf;
 
 #[derive(Debug, Error)]
 pub enum VerificationKeyFileError {
     #[error("{}: {}", _0, _1)]
     Crate(&'static str, String),
 
-    #[error("creating: {}", _0)]
-    Creating(io::Error),
-
     #[error("Cannot read from the provided file path - {:?}", _0)]
     FileReadError(PathBuf),
 
     #[error("Cannot remove the provided file - {:?}", _0)]
     FileRemovalError(PathBuf),
-
-    #[error("Verification key file was corrupted")]
-    IncorrectVerificationKey,
-
-    #[error("writing: {}", _0)]
-    Writing(io::Error),
 }
 
 impl From<std::io::Error> for VerificationKeyFileError {
diff --git a/package/src/errors/package.rs b/package/src/errors/package.rs
index f0a07ea1ee..47f4808924 100644
--- a/package/src/errors/package.rs
+++ b/package/src/errors/package.rs
@@ -13,24 +13,18 @@
 
 // You should have received a copy of the GNU General Public License
 // along with the Leo library. If not, see <https://www.gnu.org/licenses/>.
-use std::{ffi::OsString, io};
+use std::ffi::OsString;
 
 #[derive(Debug, Error)]
 pub enum PackageError {
     #[error("{}: {}", _0, _1)]
     Crate(&'static str, String),
 
-    #[error("`{}` creating: {}", _0, _1)]
-    Creating(&'static str, io::Error),
-
     #[error("Failed to initialize package {:?} ({:?})", _0, _1)]
     FailedToInitialize(String, OsString),
 
     #[error("Invalid project name: {:?}", _0)]
     InvalidPackageName(String),
-
-    #[error("`{}` metadata: {}", _0, _1)]
-    Removing(&'static str, io::Error),
 }
 
 impl From<std::io::Error> for PackageError {
diff --git a/package/src/errors/root/gitignore.rs b/package/src/errors/root/gitignore.rs
index bb8b5d2bca..bf047ddabb 100644
--- a/package/src/errors/root/gitignore.rs
+++ b/package/src/errors/root/gitignore.rs
@@ -14,18 +14,10 @@
 // You should have received a copy of the GNU General Public License
 // along with the Leo library. If not, see <https://www.gnu.org/licenses/>.
 
-use std::io;
-
 #[derive(Debug, Error)]
 pub enum GitignoreError {
     #[error("{}: {}", _0, _1)]
     Crate(&'static str, String),
-
-    #[error("creating: {}", _0)]
-    Creating(io::Error),
-
-    #[error("writing: {}", _0)]
-    Writing(io::Error),
 }
 
 impl From<std::io::Error> for GitignoreError {
diff --git a/package/src/errors/root/readme.rs b/package/src/errors/root/readme.rs
index c123ddd826..f5d99216c5 100644
--- a/package/src/errors/root/readme.rs
+++ b/package/src/errors/root/readme.rs
@@ -14,19 +14,11 @@
 // You should have received a copy of the GNU General Public License
 // along with the Leo library. If not, see <https://www.gnu.org/licenses/>.
 
-use std::io;
-
 #[allow(clippy::upper_case_acronyms)]
 #[derive(Debug, Error)]
 pub enum READMEError {
     #[error("{}: {}", _0, _1)]
     Crate(&'static str, String),
-
-    #[error("creating: {}", _0)]
-    Creating(io::Error),
-
-    #[error("writing: {}", _0)]
-    Writing(io::Error),
 }
 
 impl From<std::io::Error> for READMEError {
diff --git a/package/src/errors/root/zip.rs b/package/src/errors/root/zip.rs
index e9a9c15b82..af9f4e231b 100644
--- a/package/src/errors/root/zip.rs
+++ b/package/src/errors/root/zip.rs
@@ -14,8 +14,7 @@
 // You should have received a copy of the GNU General Public License
 // along with the Leo library. If not, see <https://www.gnu.org/licenses/>.
 
-use std::{io, path::PathBuf};
-use walkdir::Error as WalkDirError;
+use std::path::PathBuf;
 use zip::result::ZipError;
 
 #[derive(Debug, Error)]
@@ -23,21 +22,12 @@ pub enum ZipFileError {
     #[error("{}: {}", _0, _1)]
     Crate(&'static str, String),
 
-    #[error("creating: {}", _0)]
-    Creating(io::Error),
-
     #[error("Cannot read from the provided file path - {:?}", _0)]
     FileReadError(PathBuf),
 
     #[error("Cannot remove the provided file - {:?}", _0)]
     FileRemovalError(PathBuf),
 
-    #[error("writing: {}", _0)]
-    Writing(io::Error),
-
-    #[error("{}", _0)]
-    WalkDirError(#[from] WalkDirError),
-
     #[error("{}", _0)]
     ZipError(#[from] ZipError),
 }
diff --git a/package/src/errors/source/lib.rs b/package/src/errors/source/lib.rs
index 6b754373e7..149257085c 100644
--- a/package/src/errors/source/lib.rs
+++ b/package/src/errors/source/lib.rs
@@ -14,18 +14,10 @@
 // You should have received a copy of the GNU General Public License
 // along with the Leo library. If not, see <https://www.gnu.org/licenses/>.
 
-use std::io;
-
 #[derive(Debug, Error)]
 pub enum LibraryFileError {
     #[error("{}: {}", _0, _1)]
     Crate(&'static str, String),
-
-    #[error("creating: {}", _0)]
-    Creating(io::Error),
-
-    #[error("writing: {}", _0)]
-    Writing(io::Error),
 }
 
 impl From<std::io::Error> for LibraryFileError {
diff --git a/package/src/errors/source/main.rs b/package/src/errors/source/main.rs
index 07460b50e2..fc1de8d998 100644
--- a/package/src/errors/source/main.rs
+++ b/package/src/errors/source/main.rs
@@ -14,18 +14,10 @@
 // You should have received a copy of the GNU General Public License
 // along with the Leo library. If not, see <https://www.gnu.org/licenses/>.
 
-use std::io;
-
 #[derive(Debug, Error)]
 pub enum MainFileError {
     #[error("{}: {}", _0, _1)]
     Crate(&'static str, String),
-
-    #[error("creating: {}", _0)]
-    Creating(io::Error),
-
-    #[error("writing: {}", _0)]
-    Writing(io::Error),
 }
 
 impl From<std::io::Error> for MainFileError {
diff --git a/package/src/imports/directory.rs b/package/src/imports/directory.rs
index a1d7e850ea..1f01597d2e 100644
--- a/package/src/imports/directory.rs
+++ b/package/src/imports/directory.rs
@@ -33,20 +33,6 @@ impl ImportsDirectory {
         fs::create_dir_all(&path).map_err(ImportsDirectoryError::Creating)
     }
 
-    /// Removes the directory at the provided path.
-    pub fn remove(path: &Path) -> Result<(), ImportsDirectoryError> {
-        let mut path = Cow::from(path);
-        if path.is_dir() && !path.ends_with(IMPORTS_DIRECTORY_NAME) {
-            path.to_mut().push(IMPORTS_DIRECTORY_NAME);
-        }
-
-        if path.exists() {
-            fs::remove_dir_all(&path).map_err(ImportsDirectoryError::Removing)?;
-        }
-
-        Ok(())
-    }
-
     /// Removes an imported package in the imports directory at the provided path.
     pub fn remove_import(path: &Path, package_name: &str) -> Result<(), ImportsDirectoryError> {
         let mut path = Cow::from(path);
diff --git a/parser/src/parser/context.rs b/parser/src/parser/context.rs
index 91014c4f87..4ba080dab4 100644
--- a/parser/src/parser/context.rs
+++ b/parser/src/parser/context.rs
@@ -16,7 +16,7 @@
 
 use std::unimplemented;
 
-use crate::{tokenizer::*, SyntaxError, SyntaxResult, Token, KEYWORD_TOKENS};
+use crate::{tokenizer::*, unexpected_whitespace, SyntaxError, SyntaxResult, Token, KEYWORD_TOKENS};
 use leo_ast::*;
 use tendril::format_tendril;
 
@@ -164,10 +164,16 @@ impl ParserContext {
     /// Removes the next two tokens if they are a pair of [`GroupCoordinate`] and returns them,
     /// or [None] if the next token is not a [`GroupCoordinate`].
     ///
-    pub fn eat_group_partial(&mut self) -> Option<(GroupCoordinate, GroupCoordinate, Span)> {
+    pub fn eat_group_partial(&mut self) -> SyntaxResult<Option<(GroupCoordinate, GroupCoordinate, Span)>> {
         let mut i = self.tokens.len() - 1;
-        let start_span = self.tokens.get(i)?.span.clone();
-        let first = self.peek_group_coordinate(&mut i)?;
+        let start_span = match self.tokens.get(i) {
+            Some(span) => span.span.clone(),
+            None => return Ok(None),
+        };
+        let first = match self.peek_group_coordinate(&mut i) {
+            Some(coord) => coord,
+            None => return Ok(None),
+        };
         match self.tokens.get(i) {
             Some(SpannedToken {
                 token: Token::Comma, ..
@@ -175,19 +181,24 @@ impl ParserContext {
                 i -= 1;
             }
             _ => {
-                return None;
+                return Ok(None);
             }
         }
-        let second = self.peek_group_coordinate(&mut i)?;
+        let second = match self.peek_group_coordinate(&mut i) {
+            Some(coord) => coord,
+            None => return Ok(None),
+        };
+        let right_paren_span;
         match self.tokens.get(i) {
             Some(SpannedToken {
                 token: Token::RightParen,
-                ..
+                span,
             }) => {
+                right_paren_span = span.clone();
                 i -= 1;
             }
             _ => {
-                return None;
+                return Ok(None);
             }
         }
         let end_span;
@@ -200,12 +211,18 @@ impl ParserContext {
                 i -= 1;
             }
             _ => {
-                return None;
+                return Ok(None);
             }
         }
 
         self.tokens.drain((i + 1)..);
-        Some((first, second, start_span + end_span))
+        unexpected_whitespace(
+            &right_paren_span,
+            &end_span,
+            &format!("({},{})", first, second),
+            "group",
+        )?;
+        Ok(Some((first, second, start_span + end_span)))
     }
 
     ///
diff --git a/parser/src/parser/expression.rs b/parser/src/parser/expression.rs
index 3b5d88fc0b..042a0520b5 100644
--- a/parser/src/parser/expression.rs
+++ b/parser/src/parser/expression.rs
@@ -554,7 +554,7 @@ impl ParserContext {
     /// tuple initialization expression.
     ///
     pub fn parse_tuple_expression(&mut self, span: &Span) -> SyntaxResult<Expression> {
-        if let Some((left, right, span)) = self.eat_group_partial() {
+        if let Some((left, right, span)) = self.eat_group_partial()? {
             return Ok(Expression::Value(ValueExpression::Group(Box::new(GroupValue::Tuple(
                 GroupTuple {
                     span,
@@ -657,19 +657,28 @@ impl ParserContext {
                     Some(SpannedToken {
                         token: Token::Field,
                         span: type_span,
-                    }) => Expression::Value(ValueExpression::Field(value, span + type_span)),
+                    }) => {
+                        unexpected_whitespace(&span, &type_span, &value, "field")?;
+                        Expression::Value(ValueExpression::Field(value, span + type_span))
+                    }
                     Some(SpannedToken {
                         token: Token::Group,
                         span: type_span,
-                    }) => Expression::Value(ValueExpression::Group(Box::new(GroupValue::Single(
-                        value,
-                        span + type_span,
-                    )))),
-                    Some(SpannedToken { token, span: type_span }) => Expression::Value(ValueExpression::Integer(
-                        Self::token_to_int_type(token).expect("unknown int type token"),
-                        value,
-                        span + type_span,
-                    )),
+                    }) => {
+                        unexpected_whitespace(&span, &type_span, &value, "group")?;
+                        Expression::Value(ValueExpression::Group(Box::new(GroupValue::Single(
+                            value,
+                            span + type_span,
+                        ))))
+                    }
+                    Some(SpannedToken { token, span: type_span }) => {
+                        unexpected_whitespace(&span, &type_span, &value, &token.to_string())?;
+                        Expression::Value(ValueExpression::Integer(
+                            Self::token_to_int_type(token).expect("unknown int type token"),
+                            value,
+                            span + type_span,
+                        ))
+                    }
                     None => Expression::Value(ValueExpression::Implicit(value, span)),
                 }
             }
diff --git a/parser/src/parser/file.rs b/parser/src/parser/file.rs
index 98f81fd938..0451c6ab1d 100644
--- a/parser/src/parser/file.rs
+++ b/parser/src/parser/file.rs
@@ -91,12 +91,7 @@ impl ParserContext {
             )));
         }
 
-        if start.col_stop != name.span.col_start {
-            let mut error_span = &start + &name.span;
-            error_span.col_start = start.col_stop - 1;
-            error_span.col_stop = name.span.col_start - 1;
-            return Err(SyntaxError::unexpected_whitespace("@", &name.name, &error_span));
-        }
+        unexpected_whitespace(&start, &name.span, &name.name, "@")?;
 
         let end_span;
         let arguments = if self.eat(Token::LeftParen).is_some() {
diff --git a/parser/src/parser/mod.rs b/parser/src/parser/mod.rs
index c217089d89..90b914aa02 100644
--- a/parser/src/parser/mod.rs
+++ b/parser/src/parser/mod.rs
@@ -41,3 +41,14 @@ pub fn parse(path: &str, source: &str) -> SyntaxResult<Program> {
 
     tokens.parse_program()
 }
+
+pub fn unexpected_whitespace(left_span: &Span, right_span: &Span, left: &str, right: &str) -> SyntaxResult<()> {
+    if left_span.col_stop != right_span.col_start {
+        let mut error_span = left_span + right_span;
+        error_span.col_start = left_span.col_stop - 1;
+        error_span.col_stop = right_span.col_start - 1;
+        return Err(SyntaxError::unexpected_whitespace(left, right, &error_span));
+    }
+
+    Ok(())
+}
