diff --git a/Cargo.lock b/Cargo.lock
index fc4a7097a6..db3ca2875e 100644
--- a/Cargo.lock
+++ b/Cargo.lock
@@ -439,9 +439,9 @@ dependencies = [
 
 [[package]]
 name = "clap"
-version = "4.5.15"
+version = "4.5.16"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "11d8838454fda655dafd3accb2b6e2bea645b9e4078abe84a22ceb947235c5cc"
+checksum = "ed6719fffa43d0d87e5fd8caeab59be1554fb028cd30edc88fc4369b17971019"
 dependencies = [
  "clap_builder",
  "clap_derive",
@@ -2729,18 +2729,18 @@ checksum = "61697e0a1c7e512e84a621326239844a24d8207b4669b41bc18b32ea5cbf988b"
 
 [[package]]
 name = "serde"
-version = "1.0.207"
+version = "1.0.208"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "5665e14a49a4ea1b91029ba7d3bca9f299e1f7cfa194388ccc20f14743e784f2"
+checksum = "cff085d2cb684faa248efb494c39b68e522822ac0de72ccf08109abde717cfb2"
 dependencies = [
  "serde_derive",
 ]
 
 [[package]]
 name = "serde_derive"
-version = "1.0.207"
+version = "1.0.208"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "6aea2634c86b0e8ef2cfdc0c340baede54ec27b1e46febd7f80dffb2aa44a00e"
+checksum = "24008e81ff7613ed8e5ba0cfaf24e2c2f1e5b8a0495711e44fcd4882fca62bcf"
 dependencies = [
  "proc-macro2",
  "quote 1.0.36",
diff --git a/compiler/compiler/Cargo.toml b/compiler/compiler/Cargo.toml
index f3bd08df1f..1077d369b2 100644
--- a/compiler/compiler/Cargo.toml
+++ b/compiler/compiler/Cargo.toml
@@ -78,7 +78,7 @@ workspace = true
 version = "1.10.6"
 
 [dev-dependencies.serde]
-version = "1.0.207"
+version = "1.0.208"
 features = [ "derive" ]
 
 [dev-dependencies.serde_yaml]
diff --git a/compiler/span/Cargo.toml b/compiler/span/Cargo.toml
index 16909b116a..c7dde5b3c0 100644
--- a/compiler/span/Cargo.toml
+++ b/compiler/span/Cargo.toml
@@ -29,5 +29,5 @@ version = "0.2.1"
 version = "1.0.1"
 
 [dependencies.serde]
-version = "1.0.207"
+version = "1.0.208"
 features = [ "derive", "rc" ]
diff --git a/errors/Cargo.toml b/errors/Cargo.toml
index 14c494c19c..8b46fca15b 100644
--- a/errors/Cargo.toml
+++ b/errors/Cargo.toml
@@ -41,7 +41,7 @@ version = "2.2.0"
 version = "0.12.5"
 
 [dependencies.serde]
-version = "1.0.207"
+version = "1.0.208"
 features = [ "derive", "rc" ]
 
 [dependencies.thiserror]
diff --git a/leo/cli/commands/account.rs b/leo/cli/commands/account.rs
index 80e4bff989..c1b8b82bdc 100644
--- a/leo/cli/commands/account.rs
+++ b/leo/cli/commands/account.rs
@@ -88,9 +88,6 @@ pub enum Account {
         /// Message (Aleo value) to sign
         #[clap(short = 'm', long)]
         message: String,
-        /// Seed the RNG with a numeric value
-        #[clap(short = 's', long)]
-        seed: Option<u64>,
         /// When enabled, parses the message as bytes instead of Aleo literals
         #[clap(short = 'r', long)]
         raw: bool,
@@ -150,19 +147,13 @@ impl Command for Account {
                     NetworkName::CanaryV0 => import_account::<CanaryV0>(private_key, write, discreet, &ctx, endpoint),
                 }?
             }
-            Self::Sign { message, seed, raw, private_key, private_key_file, network } => {
+            Self::Sign { message, raw, private_key, private_key_file, network } => {
                 // Parse the network.
                 let network = NetworkName::try_from(network.as_str())?;
                 let result = match network {
-                    NetworkName::MainnetV0 => {
-                        sign_message::<MainnetV0>(message, seed, raw, private_key, private_key_file)
-                    }
-                    NetworkName::TestnetV0 => {
-                        sign_message::<TestnetV0>(message, seed, raw, private_key, private_key_file)
-                    }
-                    NetworkName::CanaryV0 => {
-                        sign_message::<MainnetV0>(message, seed, raw, private_key, private_key_file)
-                    }
+                    NetworkName::MainnetV0 => sign_message::<MainnetV0>(message, raw, private_key, private_key_file),
+                    NetworkName::TestnetV0 => sign_message::<TestnetV0>(message, raw, private_key, private_key_file),
+                    NetworkName::CanaryV0 => sign_message::<MainnetV0>(message, raw, private_key, private_key_file),
                 }?;
                 println!("{result}")
             }
@@ -271,7 +262,6 @@ fn print_keys<N: Network>(private_key: PrivateKey<N>, discreet: bool) -> Result<
 // Sign a message with an Aleo private key
 pub(crate) fn sign_message<N: Network>(
     message: String,
-    seed: Option<u64>,
     raw: bool,
     private_key: Option<String>,
     private_key_file: Option<String>,
@@ -301,13 +291,8 @@ pub(crate) fn sign_message<N: Network>(
             Err(CliError::cli_invalid_input("cannot specify both the '--private-key' and '--private-key-file' flags"))?
         }
     };
-    // Recover the seed.
-    let mut rng = match seed {
-        // Recover the field element deterministically.
-        Some(seed) => ChaChaRng::seed_from_u64(seed),
-        // Sample a random field element.
-        None => ChaChaRng::from_entropy(),
-    };
+    // Sample a random field element.
+    let mut rng = ChaChaRng::from_entropy();
 
     // Sign the message
     let signature = if raw {
@@ -393,41 +378,21 @@ mod tests {
     fn test_signature_raw() {
         let key = "APrivateKey1zkp61PAYmrYEKLtRWeWhUoDpFnGLNuHrCciSqN49T86dw3p".to_string();
         let message = "Hello, world!".to_string();
-        assert!(sign_message::<CurrentNetwork>(message, None, true, Some(key), None).is_ok());
+        assert!(sign_message::<CurrentNetwork>(message, true, Some(key), None).is_ok());
     }
 
     #[test]
     fn test_signature() {
         let key = "APrivateKey1zkp61PAYmrYEKLtRWeWhUoDpFnGLNuHrCciSqN49T86dw3p".to_string();
         let message = "5field".to_string();
-        assert!(sign_message::<CurrentNetwork>(message, None, false, Some(key), None).is_ok());
+        assert!(sign_message::<CurrentNetwork>(message, false, Some(key), None).is_ok());
     }
 
     #[test]
     fn test_signature_fail() {
         let key = "APrivateKey1zkp61PAYmrYEKLtRWeWhUoDpFnGLNuHrCciSqN49T86dw3p".to_string();
         let message = "not a literal value".to_string();
-        assert!(sign_message::<CurrentNetwork>(message, None, false, Some(key), None).is_err());
-    }
-
-    #[test]
-    fn test_seeded_signature_raw() {
-        let seed = Some(38868010450269069);
-        let key = "APrivateKey1zkp61PAYmrYEKLtRWeWhUoDpFnGLNuHrCciSqN49T86dw3p".to_string();
-        let message = "Hello, world!".to_string();
-        let expected = "sign175pmqldmkqw2nwp7wz7tfmpyqdnvzaq06mh8t2g22frsmrdtuvpf843p0wzazg27rwrjft8863vwn5a5cqgr97ldw69cyq53l0zlwqhesm5elrqqunzqzmac7kzutl6zk7mqht3c0m9kg4hklv7h2js0qmxavwnpuwyl4lzldl6prs4qeqy9wxyp8y44nnydg3h8sg6ue99qkevd26g";
-        let actual = sign_message::<CurrentNetwork>(message, seed, true, Some(key), None).unwrap();
-        assert_eq!(expected, actual);
-    }
-
-    #[test]
-    fn test_seeded_signature() {
-        let seed = Some(38868010450269069);
-        let key = "APrivateKey1zkp61PAYmrYEKLtRWeWhUoDpFnGLNuHrCciSqN49T86dw3p".to_string();
-        let message = "5field".to_string();
-        let expected = "sign1ad29myqy8gv6xve2r6tuly39m63l2mpfpyvqkwdl2umxqek6q5qxmy63zmhjx75x90sqxq69u5ntzp25kp59e0hp4hj8l8085sg7vqlesm5elrqqunzqzmac7kzutl6zk7mqht3c0m9kg4hklv7h2js0qmxavwnpuwyl4lzldl6prs4qeqy9wxyp8y44nnydg3h8sg6ue99qk7v46re";
-        let actual = sign_message::<CurrentNetwork>(message, seed, false, Some(key), None).unwrap();
-        assert_eq!(expected, actual);
+        assert!(sign_message::<CurrentNetwork>(message, false, Some(key), None).is_err());
     }
 
     #[test]
diff --git a/leo/package/src/lib.rs b/leo/package/src/lib.rs
index c8a1a56885..24e33d58b7 100644
--- a/leo/package/src/lib.rs
+++ b/leo/package/src/lib.rs
@@ -44,19 +44,13 @@ pub(crate) fn parse_file_paths(directory: ReadDir, file_paths: &mut Vec<PathBuf>
             parse_file_paths(directory, file_paths)?;
             continue;
         } else {
-            // Verify that the file has the Leo file extension
-            let file_extension = file_path
-                .extension()
-                .ok_or_else(|| PackageError::failed_to_get_leo_file_extension(file_path.as_os_str().to_owned()))?;
-            if file_extension != LEO_FILE_EXTENSION.trim_start_matches('.') {
-                return Err(PackageError::invalid_leo_file_extension(
-                    file_path.as_os_str().to_owned(),
-                    file_extension.to_owned(),
-                )
-                .into());
+            // If the extension doesn't match, we simply skip this file
+            // If there's no extension, we also skip this file
+            if let Some(file_extension) = file_path.extension() {
+                if file_extension == LEO_FILE_EXTENSION.trim_start_matches('.') {
+                    file_paths.push(file_path);
+                }
             }
-
-            file_paths.push(file_path);
         }
     }
 
