diff --git a/.circleci/token/run.sh b/.circleci/token/run.sh
index 812d1c2b85..bff3959a56 100644
--- a/.circleci/token/run.sh
+++ b/.circleci/token/run.sh
@@ -8,18 +8,18 @@ fi
 
 # The private key and address of Alice.
 # Swap these into .env, when running transactions as the first bidder.
-# NETWORK=mainnet
+# NETWORK=testnet
 # PRIVATE_KEY=APrivateKey1zkp8CZNn3yeCseEtxuVPbDCwSyhGW6yZKUYKfgXmcpoGPWH
 
 # The private key and address of Bob.
 # Swap these into program.json, when running transactions as the second bidder.
-# NETWORK=mainnet
+# NETWORK=testnet
 # PRIVATE_KEY=APrivateKey1zkp2RWGDcde3efb89rjhME1VYA8QMxcxep5DShNBR6n8Yjh
 
 
 # Swap in the private key of Alice.
 echo "
-NETWORK=mainnet
+NETWORK=testnet
 PRIVATE_KEY=APrivateKey1zkp8CZNn3yeCseEtxuVPbDCwSyhGW6yZKUYKfgXmcpoGPWH
 ENDPOINT=https://localhost:3030
 " > .env
@@ -54,7 +54,7 @@ leo run mint_public aleo13ssze66adjjkt795z9u5wpq8h6kn0y2657726h4h3e3wfnez4vqsm30
 
 # Swap in the private key of Bob.
 echo "
-NETWORK=mainnet
+NETWORK=testnet
 PRIVATE_KEY=APrivateKey1zkp2RWGDcde3efb89rjhME1VYA8QMxcxep5DShNBR6n8Yjh
 ENDPOINT=https://localhost:3030
 " > .env
@@ -89,7 +89,7 @@ leo run mint_private aleo17vy26rpdhqx4598y5gp7nvaa9rk7tnvl6ufhvvf4calsrrqdaqyshd
 
 # Swap in the private key of Alice.
 echo "
-NETWORK=mainnet
+NETWORK=testnet
 PRIVATE_KEY=APrivateKey1zkp8CZNn3yeCseEtxuVPbDCwSyhGW6yZKUYKfgXmcpoGPWH
 ENDPOINT=https://localhost:3030
 " > .env
@@ -124,7 +124,7 @@ leo run transfer_public aleo17vy26rpdhqx4598y5gp7nvaa9rk7tnvl6ufhvvf4calsrrqdaqy
 
 # Swap in the private key of Bob.
 echo "
-NETWORK=mainnet
+NETWORK=testnet
 PRIVATE_KEY=APrivateKey1zkp2RWGDcde3efb89rjhME1VYA8QMxcxep5DShNBR6n8Yjh
 ENDPOINT=https://localhost:3030
 " > .env
@@ -163,7 +163,7 @@ leo run transfer_private "{
 
 # Swap in the private key of Alice.
 echo "
-NETWORK=mainnet
+NETWORK=testnet
 PRIVATE_KEY=APrivateKey1zkp8CZNn3yeCseEtxuVPbDCwSyhGW6yZKUYKfgXmcpoGPWH
 ENDPOINT=https://localhost:3030
 " > .env
@@ -199,7 +199,7 @@ leo run transfer_public_to_private aleo17vy26rpdhqx4598y5gp7nvaa9rk7tnvl6ufhvvf4
 
 # Swap in the private key of Bob.
 echo "
-NETWORK=mainnet
+NETWORK=testnet
 PRIVATE_KEY=APrivateKey1zkp2RWGDcde3efb89rjhME1VYA8QMxcxep5DShNBR6n8Yjh
 ENDPOINT=https://localhost:3030
 " > .env
@@ -241,7 +241,7 @@ leo run transfer_private_to_public "{
 # Swap in the private key of Alice.
 # This is done to ensure that program.json is the same after every execution of ./run.sh.
 echo "
-NETWORK=mainnet
+NETWORK=testnet
 PRIVATE_KEY=APrivateKey1zkp8CZNn3yeCseEtxuVPbDCwSyhGW6yZKUYKfgXmcpoGPWH
 ENDPOINT=https://localhost:3030
 " > .env
diff --git a/Cargo.lock b/Cargo.lock
index b9078f57e3..24d036b4f7 100644
--- a/Cargo.lock
+++ b/Cargo.lock
@@ -1356,7 +1356,7 @@ dependencies = [
  "http 1.1.0",
  "hyper 1.3.1",
  "hyper-util",
- "rustls 0.23.10",
+ "rustls",
  "rustls-pki-types",
  "tokio",
  "tokio-rustls",
@@ -2514,27 +2514,15 @@ dependencies = [
  "windows-sys 0.52.0",
 ]
 
-[[package]]
-name = "rustls"
-version = "0.22.4"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "bf4ef73721ac7bcd79b2b315da7779d8fc09718c6b3d2d1b2d94850eb8c18432"
-dependencies = [
- "log",
- "ring",
- "rustls-pki-types",
- "rustls-webpki",
- "subtle",
- "zeroize",
-]
-
 [[package]]
 name = "rustls"
 version = "0.23.10"
 source = "registry+https://github.com/rust-lang/crates.io-index"
 checksum = "05cff451f60db80f490f3c182b77c35260baace73209e9cdbbe526bfe3a4d402"
 dependencies = [
+ "log",
  "once_cell",
+ "ring",
  "rustls-pki-types",
  "rustls-webpki",
  "subtle",
@@ -2722,18 +2710,18 @@ checksum = "61697e0a1c7e512e84a621326239844a24d8207b4669b41bc18b32ea5cbf988b"
 
 [[package]]
 name = "serde"
-version = "1.0.203"
+version = "1.0.204"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "7253ab4de971e72fb7be983802300c30b5a7f0c2e56fab8abfc6a214307c0094"
+checksum = "bc76f558e0cbb2a839d37354c575f1dc3fdc6546b5be373ba43d95f231bf7c12"
 dependencies = [
  "serde_derive",
 ]
 
 [[package]]
 name = "serde_derive"
-version = "1.0.203"
+version = "1.0.204"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "500cbc0ebeb6f46627f50f3f5811ccf6bf00643be300b4c3eabc0ef55dc5b5ba"
+checksum = "e0cd7e117be63d3c3678776753929474f3b04a43a080c744d6b0ae2a8c28e222"
 dependencies = [
  "proc-macro2",
  "quote 1.0.36",
@@ -4025,7 +4013,7 @@ version = "0.26.0"
 source = "registry+https://github.com/rust-lang/crates.io-index"
 checksum = "0c7bc40d0e5a97695bb96e27995cd3a08538541b0a846f65bba7a359f36700d4"
 dependencies = [
- "rustls 0.23.10",
+ "rustls",
  "rustls-pki-types",
  "tokio",
 ]
@@ -4232,17 +4220,16 @@ checksum = "8ecb6da28b8a351d773b68d5825ac39017e680750f980f3a1a85cd8dd28a47c1"
 
 [[package]]
 name = "ureq"
-version = "2.9.7"
+version = "2.10.0"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "d11a831e3c0b56e438a28308e7c810799e3c118417f342d30ecec080105395cd"
+checksum = "72139d247e5f97a3eff96229a7ae85ead5328a39efe76f8bf5a06313d505b6ea"
 dependencies = [
  "base64 0.22.1",
  "flate2",
  "log",
  "once_cell",
- "rustls 0.22.4",
+ "rustls",
  "rustls-pki-types",
- "rustls-webpki",
  "serde",
  "serde_json",
  "url",
diff --git a/Cargo.toml b/Cargo.toml
index facd11cdf6..f77a2a7961 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -64,7 +64,7 @@ noconfig = [ ]
 dialoguer = "0.11.0"
 num-format = "0.4.4"
 text-tables = "0.3.1"
-ureq = "2.9.7"
+ureq = "2.10.0"
 
 [dependencies.leo-ast]
 path = "./compiler/ast"
diff --git a/compiler/compiler/Cargo.toml b/compiler/compiler/Cargo.toml
index 4814314375..40153f806a 100644
--- a/compiler/compiler/Cargo.toml
+++ b/compiler/compiler/Cargo.toml
@@ -78,7 +78,7 @@ workspace = true
 version = "1.10.5"
 
 [dev-dependencies.serde]
-version = "1.0.203"
+version = "1.0.204"
 features = [ "derive" ]
 
 [dev-dependencies.serde_yaml]
diff --git a/compiler/parser/examples/parser.rs b/compiler/parser/examples/parser.rs
index cf093f412d..cd7450f061 100644
--- a/compiler/parser/examples/parser.rs
+++ b/compiler/parser/examples/parser.rs
@@ -39,8 +39,8 @@ struct Opt {
     #[clap(short, long)]
     print_stdout: bool,
 
-    /// The network to use. Defaults to mainnet.
-    #[clap(long, default_value = "mainnet")]
+    /// The network to use. Defaults to testnet.
+    #[clap(long, default_value = "testnet")]
     pub(crate) network: String,
 }
 
diff --git a/compiler/parser/src/parser/context.rs b/compiler/parser/src/parser/context.rs
index 8639412447..da9ade6085 100644
--- a/compiler/parser/src/parser/context.rs
+++ b/compiler/parser/src/parser/context.rs
@@ -39,7 +39,7 @@ pub(crate) struct ParserContext<'a, N: Network> {
     /// The previous token, i.e., if `p.tokens = ['3', *, '4']`,
     /// then after two `p.bump()`s, we'll have `p.token = '*'` and `p.prev_token = '3'`.
     pub(crate) prev_token: SpannedToken,
-    /// true if parsing an expression for if and loop statements -- means struct inits are not legal
+    /// True if parsing an expression for if and loop statements -- means struct inits are not legal.
     pub(crate) disallow_struct_construction: bool,
     /// The name of the program being parsed.
     pub(crate) program_name: Option<Symbol>,
@@ -95,7 +95,7 @@ impl<'a, N: Network> ParserContext<'a, N> {
         &self.token.token == tok
     }
 
-    /// Checks whether the current token is a `Token::Int(_)`.
+    /// Checks whether the current token is a `Token::Integer(_)`.
     pub(super) fn check_int(&self) -> bool {
         matches!(&self.token.token, Token::Integer(_))
     }
@@ -142,7 +142,7 @@ impl<'a, N: Network> ParserContext<'a, N> {
         Identifier { name, span, id: self.node_builder.next_id() }
     }
 
-    /// Eats the next token if its an identifier and returns it.
+    /// Eats the next token if it is an identifier and returns it.
     pub(super) fn eat_identifier(&mut self) -> Option<Identifier> {
         if let Token::Identifier(name) = self.token.token {
             self.bump();
diff --git a/compiler/parser/src/parser/file.rs b/compiler/parser/src/parser/file.rs
index 06d844a987..92e743991f 100644
--- a/compiler/parser/src/parser/file.rs
+++ b/compiler/parser/src/parser/file.rs
@@ -75,11 +75,12 @@ impl<N: Network> ParserContext<'_, N> {
         // Parse `foo`.
         let import_name = self.expect_identifier()?;
 
-        // Parse `.aleo`.
+        // Parse `.`.
         self.expect(&Token::Dot)?;
 
+        // Parse network, which currently must be `aleo`.
         if !self.eat(&Token::Aleo) {
-            // Throw error for non-aleo files.
+            // Throw error for non-aleo networks.
             return Err(ParserError::invalid_network(self.token.span).into());
         }
 
@@ -100,10 +101,10 @@ impl<N: Network> ParserContext<'_, N> {
         // Set the program name in the context.
         self.program_name = Some(name.name);
 
-        // Parse the program network.
+        // Parse the `.`.
         self.expect(&Token::Dot)?;
 
-        // Otherwise throw parser error
+        // Parse the program network, which must be `aleo`, otherwise throw parser error.
         self.expect(&Token::Aleo).map_err(|_| ParserError::invalid_network(self.token.span))?;
 
         // Construct the program id.
@@ -136,7 +137,7 @@ impl<N: Network> ParserContext<'_, N> {
                 Token::At | Token::Async | Token::Function | Token::Transition | Token::Inline => {
                     let (id, function) = self.parse_function()?;
 
-                    // Partition into transitions and functions so that don't have to sort later.
+                    // Partition into transitions and functions so that we don't have to sort later.
                     if function.variant.is_transition() {
                         transitions.push((id, function));
                     } else {
@@ -146,13 +147,16 @@ impl<N: Network> ParserContext<'_, N> {
                 Token::RightCurly => break,
                 _ => {
                     return Err(Self::unexpected_item(&self.token, &[
+                        Token::Const,
                         Token::Struct,
                         Token::Record,
                         Token::Mapping,
                         Token::At,
+                        Token::Async,
                         Token::Function,
                         Token::Transition,
                         Token::Inline,
+                        Token::RightCurly,
                     ])
                     .into());
                 }
diff --git a/compiler/parser/src/parser/mod.rs b/compiler/parser/src/parser/mod.rs
index 7a6ee0d571..33e350a893 100644
--- a/compiler/parser/src/parser/mod.rs
+++ b/compiler/parser/src/parser/mod.rs
@@ -14,10 +14,10 @@
 // You should have received a copy of the GNU General Public License
 // along with the Leo library. If not, see <https://www.gnu.org/licenses/>.
 
-//! The parser to convert Leo code text into an [`Program`] AST type.
+//! The parser to convert Leo code text into a [`Program`] AST type.
 //!
-//! This module contains the [`parse()`] method which calls the underlying [`tokenize()`]
-//! method to create a new program ast.
+//! This module contains the [`parse()`] function which calls the underlying [`tokenize()`]
+//! method to create a new program AST.
 
 use crate::{tokenizer::*, Token};
 
diff --git a/compiler/parser/src/tokenizer/lexer.rs b/compiler/parser/src/tokenizer/lexer.rs
index fbe5bb7864..a7c33af1e7 100644
--- a/compiler/parser/src/tokenizer/lexer.rs
+++ b/compiler/parser/src/tokenizer/lexer.rs
@@ -154,8 +154,15 @@ impl Token {
     //     }
     // }
 
-    /// Returns a tuple: [(integer length, integer token)] if an integer can be eaten, otherwise returns [`None`].
-    /// An integer can be eaten if its bytes are at the front of the given `input` string.
+    /// Returns a tuple: [(integer length, integer token)] if an integer can be eaten.
+    /// An integer can be eaten if its characters are at the front of the given `input` string.
+    /// If there is no input, this function returns an error.
+    /// If there is input but no integer, this function returns the tuple consisting of
+    /// length 0 and a dummy integer token that contains an empty string.
+    /// However, this function is always called when the next character is a digit.
+    /// This function eats a sequence of one or more digits and underscores
+    /// (starting from a digit, as explained above, given when it is called),
+    /// which corresponds to a numeral in the ABNF grammar.
     fn eat_integer(input: &mut Peekable<impl Iterator<Item = char>>) -> Result<(usize, Token)> {
         if input.peek().is_none() {
             return Err(ParserError::lexer_empty_input().into());
@@ -178,7 +185,7 @@ impl Token {
     }
 
     /// Returns a tuple: [(token length, token)] if the next token can be eaten, otherwise returns an error.
-    /// The next token can be eaten if the bytes at the front of the given `input` string can be scanned into a token.
+    /// The next token can be eaten if the characters at the front of the given `input` string can be scanned into a token.
     pub(crate) fn eat(input: &str) -> Result<(usize, Token)> {
         if input.is_empty() {
             return Err(ParserError::lexer_empty_input().into());
@@ -221,13 +228,13 @@ impl Token {
         // See the example with the different combinations for Mul, MulAssign, Pow, PowAssign below.
         let match_four = |
             input: &mut Peekable<_>,
-            first_token, // Mul '*'
-            second_char, // '='
-            second_token, // MulAssign '*='
-            third_char, // '*'
-            third_token, // Pow '**'
-            fourth_char, // '='
-            fourth_token // PowAssign '**='
+            first_token, // e.e. Mul '*'
+            second_char, // e.g. '='
+            second_token, // e.g. MulAssign '*='
+            third_char, // e.g. '*'
+            third_token, // e.g. Pow '**'
+            fourth_char, // e.g. '='
+            fourth_token // e.g. PowAssign '**='
         | {
             input.next();
             Ok(if input.next_if_eq(&second_char).is_some() {
@@ -252,7 +259,7 @@ impl Token {
                 // Find end string quotation mark.
                 // Instead of checking each `char` and pushing, we can avoid reallocations.
                 // This works because the code 34 of double quote cannot appear as a byte
-                // in middle of a multi-byte UTF-8 encoding of a character,
+                // in the middle of a multi-byte UTF-8 encoding of a character,
                 // because those bytes all have the high bit set to 1;
                 // in UTF-8, the byte 34 can only appear as the single-byte encoding of double quote.
                 let rest = &input_str[1..];
@@ -306,7 +313,7 @@ impl Token {
                 if input.next_if_eq(&'/').is_some() {
                     // Find the end of the comment line.
                     // This works because the code 10 of line feed cannot appear as a byte
-                    // in middle of a multi-byte UTF-8 encoding of a character,
+                    // in the middle of a multi-byte UTF-8 encoding of a character,
                     // because those bytes all have the high bit set to 1;
                     // in UTF-8, the byte 10 can only appear as the single-byte encoding of line feed.
                     let comment = match input_str.as_bytes().iter().position(|c| *c == b'\n') {
@@ -416,8 +423,8 @@ impl Token {
                     "record" => Token::Record,
                     "return" => Token::Return,
                     "scalar" => Token::Scalar,
-                    "signature" => Token::Signature,
                     "self" => Token::SelfLower,
+                    "signature" => Token::Signature,
                     "string" => Token::String,
                     "struct" => Token::Struct,
                     "transition" => Token::Transition,
diff --git a/compiler/parser/src/tokenizer/token.rs b/compiler/parser/src/tokenizer/token.rs
index ce76068b84..f810cea96c 100644
--- a/compiler/parser/src/tokenizer/token.rs
+++ b/compiler/parser/src/tokenizer/token.rs
@@ -29,20 +29,24 @@ use leo_span::{sym, Symbol};
 #[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
 pub enum Token {
     // Comments
-    CommentLine(String),
-    CommentBlock(String),
+    CommentLine(String),  // the string includes the starting '//' and the ending line feed
+    CommentBlock(String), // the string includes the starting '/*' and the ending '*/'
 
     // Whitespace (we do not distinguish among different kinds here)
     WhiteSpace,
 
     // Literals (= atomic literals and numerals in the ABNF grammar)
-    // The string in Integer(String) consists of digits optionally followed by a type
-    // The string in AddressLit(String) has the form `aleo1...`
+    // The string in Integer(String) consists of digits
+    // The string in AddressLit(String) has the form `aleo1...`.
     True,
     False,
-    Integer(String), // = numeric literal or numeral in the ABNF grammar
+    Integer(String), // = numeral (including tuple index) in the ABNF grammar
     AddressLit(String),
     StaticString(String),
+    // The numeric literals in the ABNF grammar, which consist of numerals followed by types,
+    // are represented not as single tokens here,
+    // but as two separate tokens (one for the numeral and one for the type),
+    // enforcing, during parsing, the absence of whitespace or comments between those two tokens.
 
     // Identifiers
     Identifier(Symbol),
diff --git a/compiler/passes/src/type_checking/check_program.rs b/compiler/passes/src/type_checking/check_program.rs
index 0c35db12d5..39c859fe03 100644
--- a/compiler/passes/src/type_checking/check_program.rs
+++ b/compiler/passes/src/type_checking/check_program.rs
@@ -153,6 +153,10 @@ impl<'a, N: Network> ProgramVisitor<'a> for TypeChecker<'a, N> {
                 };
             check_has_field(sym::owner, Type::Address);
         }
+        // For structs, check that there is at least one member.
+        else if input.members.is_empty() {
+            self.emit_err(TypeCheckerError::empty_struct(input.span()));
+        }
 
         if !(input.is_record && self.scope_state.is_stub) {
             for Member { mode, identifier, type_, span, .. } in input.members.iter() {
diff --git a/compiler/span/Cargo.toml b/compiler/span/Cargo.toml
index cff7b884e1..a869ee42ce 100644
--- a/compiler/span/Cargo.toml
+++ b/compiler/span/Cargo.toml
@@ -29,5 +29,5 @@ version = "0.2.1"
 version = "1.0.1"
 
 [dependencies.serde]
-version = "1.0.203"
+version = "1.0.204"
 features = [ "derive", "rc" ]
diff --git a/errors/Cargo.toml b/errors/Cargo.toml
index 16b6c1273f..ad56f8673d 100644
--- a/errors/Cargo.toml
+++ b/errors/Cargo.toml
@@ -41,7 +41,7 @@ version = "2.2.0"
 version = "0.12.5"
 
 [dependencies.serde]
-version = "1.0.203"
+version = "1.0.204"
 features = [ "derive", "rc" ]
 
 [dependencies.thiserror]
diff --git a/errors/src/common/mod.rs b/errors/src/common/mod.rs
index 3a580f97f4..41c10ce326 100644
--- a/errors/src/common/mod.rs
+++ b/errors/src/common/mod.rs
@@ -32,7 +32,7 @@ pub use self::traits::*;
 
 // Right now for cleanliness of calling error functions we say each argument implements one of the follow types rather than giving a specific type.
 // This allows us to just pass many types rather doing conversions cleaning up the code.
-// The args can be made cleaneronce https://github.com/rust-lang/rust/issues/41517 or https://github.com/rust-lang/rust/issues/63063 hits stable.
+// The args can be made cleaner once https://github.com/rust-lang/rust/issues/41517 or https://github.com/rust-lang/rust/issues/63063 hits stable.
 // Either of why would allows to generate a type alias for these trait implementing types.
 // pub(crate) type DisplayArg = impl std::fmt::Display;
 // pub(crate) type DebugArg = impl std::fmt::Debug;
diff --git a/errors/src/errors/mod.rs b/errors/src/errors/mod.rs
index f52f608696..c99242d405 100644
--- a/errors/src/errors/mod.rs
+++ b/errors/src/errors/mod.rs
@@ -60,16 +60,16 @@ pub enum LeoError {
     /// Represents an AST Error in a Leo Error.
     #[error(transparent)]
     AstError(#[from] AstError),
-    /// Represents an CLI Error in a Leo Error.
+    /// Represents a CLI Error in a Leo Error.
     #[error(transparent)]
     CliError(#[from] CliError),
-    /// Represents an Compiler Error in a Leo Error.
+    /// Represents a Compiler Error in a Leo Error.
     #[error(transparent)]
     CompilerError(#[from] CompilerError),
-    /// Represents an Package Error in a Leo Error.
+    /// Represents a Package Error in a Leo Error.
     #[error(transparent)]
     PackageError(#[from] PackageError),
-    /// Represents an Parser Error in a Leo Error.
+    /// Represents a Parser Error in a Leo Error.
     #[error(transparent)]
     ParserError(#[from] ParserError),
     /// Represents a Type Checker Error in a Leo Error.
@@ -85,7 +85,7 @@ pub enum LeoError {
     /// not re-displaying an error.
     #[error("")]
     LastErrorCode(i32),
-    /// Represents a Utils Error in a Leo Error
+    /// Represents a Utils Error in a Leo Error.
     #[error(transparent)]
     UtilError(#[from] UtilError),
     /// Anyhow errors.
@@ -133,14 +133,14 @@ impl LeoError {
     }
 }
 
-/// The LeoWarning type that contains all sub error types.
-/// This allows a unified error type throughout the Leo crates.
+/// The LeoWarning type that contains all sub warning types.
+/// This allows a unified warning type throughout the Leo crates.
 #[derive(Debug, Error)]
 pub enum LeoWarning {
-    /// Represents an Parser Error in a Leo Error.
+    /// Represents an Parser Warning in a Leo Warning.
     #[error(transparent)]
     ParserWarning(#[from] ParserWarning),
-    /// Represents a Type Checker Error in a Leo Error.
+    /// Represents a Type Checker Warning in a Leo Warning.
     #[error(transparent)]
     TypeCheckerWarning(#[from] TypeCheckerWarning),
 }
diff --git a/errors/src/errors/type_checker/type_checker_error.rs b/errors/src/errors/type_checker/type_checker_error.rs
index 6f67a8ecc0..37486d0784 100644
--- a/errors/src/errors/type_checker/type_checker_error.rs
+++ b/errors/src/errors/type_checker/type_checker_error.rs
@@ -879,4 +879,11 @@ create_messages!(
         msg: format!("The async function `{name}` does not exist."),
         help: Some(format!("Ensure that `{name}` is defined as an async function in the current program.")),
     }
+
+    @formatted
+    empty_struct {
+        args: (),
+        msg: "A struct must have at least one member.".to_string(),
+        help: None,
+    }
 );
diff --git a/errors/src/errors/utils/util_errors.rs b/errors/src/errors/utils/util_errors.rs
index da25660922..98c1366cc8 100644
--- a/errors/src/errors/utils/util_errors.rs
+++ b/errors/src/errors/utils/util_errors.rs
@@ -114,7 +114,7 @@ create_messages!(
     missing_network_error {
         args: (dependency: impl Display),
         msg: format!("Dependency {dependency} is missing a network specification"),
-        help: Some("Add a network specification to the dependency in the `program.json` file. Example: `network: \"mainnet\"`".to_string()),
+        help: Some("Add a network specification to the dependency in the `program.json` file. Example: `network: \"testnet\"`".to_string()),
     }
 
     @formatted
diff --git a/errors/src/lib.rs b/errors/src/lib.rs
index eac73c9d90..55bfc9f5b7 100644
--- a/errors/src/lib.rs
+++ b/errors/src/lib.rs
@@ -21,7 +21,7 @@
 #[macro_use]
 extern crate thiserror;
 
-/// Contains the common functionalities for defining errors..
+/// Contains the common functionalities for defining errors.
 #[macro_use]
 pub mod common;
 pub use self::common::*;
@@ -29,6 +29,6 @@ pub use self::common::*;
 /// Contains traits and types for channels through which errors go.
 pub mod emitter;
 
-/// Contains the errors and warnings for the Leo lang.
+/// Contains the errors and warnings for the Leo language.
 pub mod errors;
 pub use self::errors::*;
diff --git a/examples/auction/.env b/examples/auction/.env
index f5b42ea894..cb5fec963e 100644
--- a/examples/auction/.env
+++ b/examples/auction/.env
@@ -1,4 +1,4 @@
 
-NETWORK=mainnet
+NETWORK=testnet
 PRIVATE_KEY=APrivateKey1zkp8CZNn3yeCseEtxuVPbDCwSyhGW6yZKUYKfgXmcpoGPWH
 ENDPOINT=https://localhost:3030
diff --git a/examples/auction/run.sh b/examples/auction/run.sh
index d94296f936..7804637849 100755
--- a/examples/auction/run.sh
+++ b/examples/auction/run.sh
@@ -8,18 +8,18 @@ fi
 
 # The private key and address of the first bidder.
 # Swap these into program.json, when running transactions as the first bidder.
-# NETWORK=mainnet
+# NETWORK=testnet
 # PRIVATE_KEY=APrivateKey1zkp8CZNn3yeCseEtxuVPbDCwSyhGW6yZKUYKfgXmcpoGPWH
 
 # The private key and address of the second bidder.
 # Swap these into program.json, when running transactions as the second bidder.
-# NETWORK=mainnet
+# NETWORK=testnet
 # PRIVATE_KEY=APrivateKey1zkp2RWGDcde3efb89rjhME1VYA8QMxcxep5DShNBR6n8Yjh
 
 
 # The private key and address of the auctioneer.
 # Swap these into program.json, when running transactions as the auctioneer.
-# NETWORK=mainnet
+# NETWORK=testnet
 # PRIVATE_KEY=APrivateKey1zkp2GUmKbVsuc1NSj28pa1WTQuZaK5f1DQJAT6vPcHyWokG
 
 
@@ -38,7 +38,7 @@ echo "
 "
 # Swap in the private key and address of the first bidder to .env.
 echo "
-NETWORK=mainnet
+NETWORK=testnet
 PRIVATE_KEY=APrivateKey1zkp8CZNn3yeCseEtxuVPbDCwSyhGW6yZKUYKfgXmcpoGPWH
 ENDPOINT=https://localhost:3030
 " > .env
@@ -61,7 +61,7 @@ leo run place_bid aleo1rhgdu77hgyqd3xjj8ucu3jj9r2krwz6mnzyd80gncr5fxcwlh5rsvzp9p
 
 # Swap in the private key and address of the second bidder to .env.
 echo "
-NETWORK=mainnet
+NETWORK=testnet
 PRIVATE_KEY=APrivateKey1zkp2RWGDcde3efb89rjhME1VYA8QMxcxep5DShNBR6n8Yjh
 ENDPOINT=https://localhost:3030
 " > .env
@@ -84,7 +84,7 @@ leo run place_bid aleo1s3ws5tra87fjycnjrwsjcrnw2qxr8jfqqdugnf0xzqqw29q9m5pqem2u4
 
 # Swap in the private key and address of the auctioneer to .env.
 echo "
-NETWORK=mainnet
+NETWORK=testnet
 PRIVATE_KEY=APrivateKey1zkp2GUmKbVsuc1NSj28pa1WTQuZaK5f1DQJAT6vPcHyWokG
 ENDPOINT=https://localhost:3030
 " > .env
@@ -142,7 +142,7 @@ leo run finish "{
 
 # Restore the .env file to its original state.
 echo "
-NETWORK=mainnet
+NETWORK=testnet
 PRIVATE_KEY=APrivateKey1zkp8CZNn3yeCseEtxuVPbDCwSyhGW6yZKUYKfgXmcpoGPWH
 ENDPOINT=https://localhost:3030
 " > .env
diff --git a/examples/basic_bank/.env b/examples/basic_bank/.env
index f5b42ea894..cb5fec963e 100644
--- a/examples/basic_bank/.env
+++ b/examples/basic_bank/.env
@@ -1,4 +1,4 @@
 
-NETWORK=mainnet
+NETWORK=testnet
 PRIVATE_KEY=APrivateKey1zkp8CZNn3yeCseEtxuVPbDCwSyhGW6yZKUYKfgXmcpoGPWH
 ENDPOINT=https://localhost:3030
diff --git a/examples/basic_bank/run.sh b/examples/basic_bank/run.sh
index 7ba84051f8..109119bb78 100755
--- a/examples/basic_bank/run.sh
+++ b/examples/basic_bank/run.sh
@@ -8,17 +8,17 @@ fi
 
 # The private key and address of the bank.
 # Swap these into program.json, when running transactions as the bank.
-# NETWORK=mainnet
+# NETWORK=testnet
 # PRIVATE_KEY=APrivateKey1zkp8CZNn3yeCseEtxuVPbDCwSyhGW6yZKUYKfgXmcpoGPWH
 
 # The private key and address of the user.
 # Swap these into program.json, when running transactions as the user.
-# NETWORK=mainnet
+# NETWORK=testnet
 # PRIVATE_KEY=APrivateKey1zkp2RWGDcde3efb89rjhME1VYA8QMxcxep5DShNBR6n8Yjh
 
 # Swap in the private key and address of the bank to .env.
 echo "
-NETWORK=mainnet
+NETWORK=testnet
 PRIVATE_KEY=APrivateKey1zkp8CZNn3yeCseEtxuVPbDCwSyhGW6yZKUYKfgXmcpoGPWH
 ENDPOINT=https://localhost:3030
 " > .env
@@ -69,7 +69,7 @@ leo run issue aleo1s3ws5tra87fjycnjrwsjcrnw2qxr8jfqqdugnf0xzqqw29q9m5pqem2u4t 10
 
 # Swap in the private key and address of the user to .env.
 echo "
-NETWORK=mainnet
+NETWORK=testnet
 PRIVATE_KEY=APrivateKey1zkp2RWGDcde3efb89rjhME1VYA8QMxcxep5DShNBR6n8Yjh
 ENDPOINT=https://localhost:3030
 " > .env
@@ -164,7 +164,7 @@ echo "
 
 # Swap in the private key and address of the bank to .env.
 echo "
-NETWORK=mainnet
+NETWORK=testnet
 PRIVATE_KEY=APrivateKey1zkp8CZNn3yeCseEtxuVPbDCwSyhGW6yZKUYKfgXmcpoGPWH
 ENDPOINT=https://localhost:3030
 " > .env
diff --git a/examples/battleship/.env b/examples/battleship/.env
index f0216e5f1b..8df6a23e94 100644
--- a/examples/battleship/.env
+++ b/examples/battleship/.env
@@ -1,5 +1,5 @@
 
-NETWORK=mainnet
+NETWORK=testnet
 PRIVATE_KEY=APrivateKey1zkp2RWGDcde3efb89rjhME1VYA8QMxcxep5DShNBR6n8Yjh
 ENDPOINT=https://localhost:3030
 
diff --git a/examples/battleship/run.sh b/examples/battleship/run.sh
index 0f6d6eacae..ed66acecc0 100755
--- a/examples/battleship/run.sh
+++ b/examples/battleship/run.sh
@@ -17,7 +17,7 @@ echo "
 "
 
 echo "
-NETWORK=mainnet
+NETWORK=testnet
 PRIVATE_KEY=APrivateKey1zkp8CZNn3yeCseEtxuVPbDCwSyhGW6yZKUYKfgXmcpoGPWH
 ENDPOINT=https://localhost:3030
 " > .env
@@ -68,7 +68,7 @@ echo "
 "
 
 echo "
-NETWORK=mainnet
+NETWORK=testnet
 PRIVATE_KEY=APrivateKey1zkp2RWGDcde3efb89rjhME1VYA8QMxcxep5DShNBR6n8Yjh
 ENDPOINT=https://localhost:3030
 " > .env
@@ -117,7 +117,7 @@ echo "
 "
 
 echo "
-NETWORK=mainnet
+NETWORK=testnet
 PRIVATE_KEY=APrivateKey1zkp8CZNn3yeCseEtxuVPbDCwSyhGW6yZKUYKfgXmcpoGPWH
 ENDPOINT=https://localhost:3030
 " > .env
@@ -153,7 +153,7 @@ echo "
 "
 
 echo "
-NETWORK=mainnet
+NETWORK=testnet
 PRIVATE_KEY=APrivateKey1zkp2RWGDcde3efb89rjhME1VYA8QMxcxep5DShNBR6n8Yjh
 ENDPOINT=https://localhost:3030
 " > .env
@@ -189,7 +189,7 @@ echo "
 "
 
 echo "
-NETWORK=mainnet
+NETWORK=testnet
 PRIVATE_KEY=APrivateKey1zkp8CZNn3yeCseEtxuVPbDCwSyhGW6yZKUYKfgXmcpoGPWH
 ENDPOINT=https://localhost:3030
 " > .env
@@ -225,7 +225,7 @@ echo "
 "
 
 echo "
-NETWORK=mainnet
+NETWORK=testnet
 PRIVATE_KEY=APrivateKey1zkp2RWGDcde3efb89rjhME1VYA8QMxcxep5DShNBR6n8Yjh
 ENDPOINT=https://localhost:3030
 " > .env
@@ -254,7 +254,7 @@ echo "
 
 # Restore the .env file to its original state
 echo "
-NETWORK=mainnet
+NETWORK=testnet
 PRIVATE_KEY=APrivateKey1zkp8CZNn3yeCseEtxuVPbDCwSyhGW6yZKUYKfgXmcpoGPWH
 ENDPOINT=https://localhost:3030
 " > .env
diff --git a/examples/bubblesort/.env b/examples/bubblesort/.env
index ab80b96bf5..0a686ef86f 100644
--- a/examples/bubblesort/.env
+++ b/examples/bubblesort/.env
@@ -1,3 +1,3 @@
-NETWORK=mainnet
+NETWORK=testnet
 PRIVATE_KEY=APrivateKey1zkp8CZNn3yeCseEtxuVPbDCwSyhGW6yZKUYKfgXmcpoGPWH
 ENDPOINT=https://localhost:3030
diff --git a/examples/core/.env b/examples/core/.env
index ab80b96bf5..0a686ef86f 100644
--- a/examples/core/.env
+++ b/examples/core/.env
@@ -1,3 +1,3 @@
-NETWORK=mainnet
+NETWORK=testnet
 PRIVATE_KEY=APrivateKey1zkp8CZNn3yeCseEtxuVPbDCwSyhGW6yZKUYKfgXmcpoGPWH
 ENDPOINT=https://localhost:3030
diff --git a/examples/fibonacci/.env b/examples/fibonacci/.env
index ab80b96bf5..0a686ef86f 100644
--- a/examples/fibonacci/.env
+++ b/examples/fibonacci/.env
@@ -1,3 +1,3 @@
-NETWORK=mainnet
+NETWORK=testnet
 PRIVATE_KEY=APrivateKey1zkp8CZNn3yeCseEtxuVPbDCwSyhGW6yZKUYKfgXmcpoGPWH
 ENDPOINT=https://localhost:3030
diff --git a/examples/groups/.env b/examples/groups/.env
index ab80b96bf5..0a686ef86f 100644
--- a/examples/groups/.env
+++ b/examples/groups/.env
@@ -1,3 +1,3 @@
-NETWORK=mainnet
+NETWORK=testnet
 PRIVATE_KEY=APrivateKey1zkp8CZNn3yeCseEtxuVPbDCwSyhGW6yZKUYKfgXmcpoGPWH
 ENDPOINT=https://localhost:3030
diff --git a/examples/hackers-delight/ntzdebruijn/.env b/examples/hackers-delight/ntzdebruijn/.env
index ab80b96bf5..0a686ef86f 100644
--- a/examples/hackers-delight/ntzdebruijn/.env
+++ b/examples/hackers-delight/ntzdebruijn/.env
@@ -1,3 +1,3 @@
-NETWORK=mainnet
+NETWORK=testnet
 PRIVATE_KEY=APrivateKey1zkp8CZNn3yeCseEtxuVPbDCwSyhGW6yZKUYKfgXmcpoGPWH
 ENDPOINT=https://localhost:3030
diff --git a/examples/hackers-delight/ntzgaudet/.env b/examples/hackers-delight/ntzgaudet/.env
index ab80b96bf5..0a686ef86f 100644
--- a/examples/hackers-delight/ntzgaudet/.env
+++ b/examples/hackers-delight/ntzgaudet/.env
@@ -1,3 +1,3 @@
-NETWORK=mainnet
+NETWORK=testnet
 PRIVATE_KEY=APrivateKey1zkp8CZNn3yeCseEtxuVPbDCwSyhGW6yZKUYKfgXmcpoGPWH
 ENDPOINT=https://localhost:3030
diff --git a/examples/hackers-delight/ntzloops/.env b/examples/hackers-delight/ntzloops/.env
index ab80b96bf5..0a686ef86f 100644
--- a/examples/hackers-delight/ntzloops/.env
+++ b/examples/hackers-delight/ntzloops/.env
@@ -1,3 +1,3 @@
-NETWORK=mainnet
+NETWORK=testnet
 PRIVATE_KEY=APrivateKey1zkp8CZNn3yeCseEtxuVPbDCwSyhGW6yZKUYKfgXmcpoGPWH
 ENDPOINT=https://localhost:3030
diff --git a/examples/hackers-delight/ntzmasks/.env b/examples/hackers-delight/ntzmasks/.env
index ab80b96bf5..0a686ef86f 100644
--- a/examples/hackers-delight/ntzmasks/.env
+++ b/examples/hackers-delight/ntzmasks/.env
@@ -1,3 +1,3 @@
-NETWORK=mainnet
+NETWORK=testnet
 PRIVATE_KEY=APrivateKey1zkp8CZNn3yeCseEtxuVPbDCwSyhGW6yZKUYKfgXmcpoGPWH
 ENDPOINT=https://localhost:3030
diff --git a/examples/hackers-delight/ntzreisers/.env b/examples/hackers-delight/ntzreisers/.env
index ab80b96bf5..0a686ef86f 100644
--- a/examples/hackers-delight/ntzreisers/.env
+++ b/examples/hackers-delight/ntzreisers/.env
@@ -1,3 +1,3 @@
-NETWORK=mainnet
+NETWORK=testnet
 PRIVATE_KEY=APrivateKey1zkp8CZNn3yeCseEtxuVPbDCwSyhGW6yZKUYKfgXmcpoGPWH
 ENDPOINT=https://localhost:3030
diff --git a/examples/hackers-delight/ntzseals/.env b/examples/hackers-delight/ntzseals/.env
index ab80b96bf5..0a686ef86f 100644
--- a/examples/hackers-delight/ntzseals/.env
+++ b/examples/hackers-delight/ntzseals/.env
@@ -1,3 +1,3 @@
-NETWORK=mainnet
+NETWORK=testnet
 PRIVATE_KEY=APrivateKey1zkp8CZNn3yeCseEtxuVPbDCwSyhGW6yZKUYKfgXmcpoGPWH
 ENDPOINT=https://localhost:3030
diff --git a/examples/hackers-delight/ntzsearchtree/.env b/examples/hackers-delight/ntzsearchtree/.env
index ab80b96bf5..0a686ef86f 100644
--- a/examples/hackers-delight/ntzsearchtree/.env
+++ b/examples/hackers-delight/ntzsearchtree/.env
@@ -1,3 +1,3 @@
-NETWORK=mainnet
+NETWORK=testnet
 PRIVATE_KEY=APrivateKey1zkp8CZNn3yeCseEtxuVPbDCwSyhGW6yZKUYKfgXmcpoGPWH
 ENDPOINT=https://localhost:3030
diff --git a/examples/hackers-delight/ntzsmallvals/.env b/examples/hackers-delight/ntzsmallvals/.env
index ab80b96bf5..0a686ef86f 100644
--- a/examples/hackers-delight/ntzsmallvals/.env
+++ b/examples/hackers-delight/ntzsmallvals/.env
@@ -1,3 +1,3 @@
-NETWORK=mainnet
+NETWORK=testnet
 PRIVATE_KEY=APrivateKey1zkp8CZNn3yeCseEtxuVPbDCwSyhGW6yZKUYKfgXmcpoGPWH
 ENDPOINT=https://localhost:3030
diff --git a/examples/helloworld/.env b/examples/helloworld/.env
index ab80b96bf5..0a686ef86f 100644
--- a/examples/helloworld/.env
+++ b/examples/helloworld/.env
@@ -1,3 +1,3 @@
-NETWORK=mainnet
+NETWORK=testnet
 PRIVATE_KEY=APrivateKey1zkp8CZNn3yeCseEtxuVPbDCwSyhGW6yZKUYKfgXmcpoGPWH
 ENDPOINT=https://localhost:3030
diff --git a/examples/interest/.env b/examples/interest/.env
index ab80b96bf5..0a686ef86f 100644
--- a/examples/interest/.env
+++ b/examples/interest/.env
@@ -1,3 +1,3 @@
-NETWORK=mainnet
+NETWORK=testnet
 PRIVATE_KEY=APrivateKey1zkp8CZNn3yeCseEtxuVPbDCwSyhGW6yZKUYKfgXmcpoGPWH
 ENDPOINT=https://localhost:3030
diff --git a/examples/lottery/.env b/examples/lottery/.env
index ab80b96bf5..0a686ef86f 100644
--- a/examples/lottery/.env
+++ b/examples/lottery/.env
@@ -1,3 +1,3 @@
-NETWORK=mainnet
+NETWORK=testnet
 PRIVATE_KEY=APrivateKey1zkp8CZNn3yeCseEtxuVPbDCwSyhGW6yZKUYKfgXmcpoGPWH
 ENDPOINT=https://localhost:3030
diff --git a/examples/message/.env b/examples/message/.env
index ab80b96bf5..0a686ef86f 100644
--- a/examples/message/.env
+++ b/examples/message/.env
@@ -1,3 +1,3 @@
-NETWORK=mainnet
+NETWORK=testnet
 PRIVATE_KEY=APrivateKey1zkp8CZNn3yeCseEtxuVPbDCwSyhGW6yZKUYKfgXmcpoGPWH
 ENDPOINT=https://localhost:3030
diff --git a/examples/simple_token/.env b/examples/simple_token/.env
index ab80b96bf5..0a686ef86f 100644
--- a/examples/simple_token/.env
+++ b/examples/simple_token/.env
@@ -1,3 +1,3 @@
-NETWORK=mainnet
+NETWORK=testnet
 PRIVATE_KEY=APrivateKey1zkp8CZNn3yeCseEtxuVPbDCwSyhGW6yZKUYKfgXmcpoGPWH
 ENDPOINT=https://localhost:3030
diff --git a/examples/tictactoe/.env b/examples/tictactoe/.env
index ab80b96bf5..0a686ef86f 100644
--- a/examples/tictactoe/.env
+++ b/examples/tictactoe/.env
@@ -1,3 +1,3 @@
-NETWORK=mainnet
+NETWORK=testnet
 PRIVATE_KEY=APrivateKey1zkp8CZNn3yeCseEtxuVPbDCwSyhGW6yZKUYKfgXmcpoGPWH
 ENDPOINT=https://localhost:3030
diff --git a/examples/token/.env b/examples/token/.env
index f5b42ea894..cb5fec963e 100644
--- a/examples/token/.env
+++ b/examples/token/.env
@@ -1,4 +1,4 @@
 
-NETWORK=mainnet
+NETWORK=testnet
 PRIVATE_KEY=APrivateKey1zkp8CZNn3yeCseEtxuVPbDCwSyhGW6yZKUYKfgXmcpoGPWH
 ENDPOINT=https://localhost:3030
diff --git a/examples/token/run.sh b/examples/token/run.sh
index 69b9f81993..cbf5393654 100755
--- a/examples/token/run.sh
+++ b/examples/token/run.sh
@@ -8,17 +8,17 @@ fi
 
 # The private key and address of Alice.
 # Swap these into program.json, when running transactions as the first bidder.
-# NETWORK=mainnet
+# NETWORK=testnet
 # PRIVATE_KEY=APrivateKey1zkp8CZNn3yeCseEtxuVPbDCwSyhGW6yZKUYKfgXmcpoGPWH
 
 # The private key and address of Bob.
 # Swap these into program.json, when running transactions as the second bidder.
-# NETWORK=mainnet
+# NETWORK=testnet
 # PRIVATE_KEY=APrivateKey1zkp2RWGDcde3efb89rjhME1VYA8QMxcxep5DShNBR6n8Yjh
 
 # Swap in the private key of Alice.
 echo "
-NETWORK=mainnet
+NETWORK=testnet
 PRIVATE_KEY=APrivateKey1zkp8CZNn3yeCseEtxuVPbDCwSyhGW6yZKUYKfgXmcpoGPWH
 ENDPOINT=https://localhost:3030
 " > .env
@@ -53,7 +53,7 @@ leo run mint_public aleo1rhgdu77hgyqd3xjj8ucu3jj9r2krwz6mnzyd80gncr5fxcwlh5rsvzp
 
 # Swap in the private key of Bob.
 echo "
-NETWORK=mainnet
+NETWORK=testnet
 PRIVATE_KEY=APrivateKey1zkp2RWGDcde3efb89rjhME1VYA8QMxcxep5DShNBR6n8Yjh
 ENDPOINT=https://localhost:3030
 " > .env
@@ -88,7 +88,7 @@ leo run mint_private aleo1s3ws5tra87fjycnjrwsjcrnw2qxr8jfqqdugnf0xzqqw29q9m5pqem
 
 # Swap in the private key of Alice.
 echo "
-NETWORK=mainnet
+NETWORK=testnet
 PRIVATE_KEY=APrivateKey1zkp8CZNn3yeCseEtxuVPbDCwSyhGW6yZKUYKfgXmcpoGPWH
 ENDPOINT=https://localhost:3030
 " > .env
@@ -123,7 +123,7 @@ leo run transfer_public aleo1s3ws5tra87fjycnjrwsjcrnw2qxr8jfqqdugnf0xzqqw29q9m5p
 
 # Swap in the private key of Bob.
 echo "
-NETWORK=mainnet
+NETWORK=testnet
 PRIVATE_KEY=APrivateKey1zkp2RWGDcde3efb89rjhME1VYA8QMxcxep5DShNBR6n8Yjh
 ENDPOINT=https://localhost:3030
 " > .env
@@ -162,7 +162,7 @@ leo run transfer_private "{
 
 # Swap in the private key of Alice.
 echo "
-NETWORK=mainnet
+NETWORK=testnet
 PRIVATE_KEY=APrivateKey1zkp8CZNn3yeCseEtxuVPbDCwSyhGW6yZKUYKfgXmcpoGPWH
 ENDPOINT=https://localhost:3030
 " > .env
@@ -198,7 +198,7 @@ leo run transfer_public_to_private aleo1s3ws5tra87fjycnjrwsjcrnw2qxr8jfqqdugnf0x
 
 # Swap in the private key of Bob.
 echo "
-NETWORK=mainnet
+NETWORK=testnet
 PRIVATE_KEY=APrivateKey1zkp2RWGDcde3efb89rjhME1VYA8QMxcxep5DShNBR6n8Yjh
 ENDPOINT=https://localhost:3030
 " > .env
@@ -240,7 +240,7 @@ leo run transfer_private_to_public "{
 # Swap in the private key of Alice.
 # This is done to ensure that program.json is the same after every execution of ./run.sh.
 echo "
-NETWORK=mainnet
+NETWORK=testnet
 PRIVATE_KEY=APrivateKey1zkp8CZNn3yeCseEtxuVPbDCwSyhGW6yZKUYKfgXmcpoGPWH
 ENDPOINT=https://localhost:3030
 " > .env
diff --git a/examples/twoadicity/.env b/examples/twoadicity/.env
index ab80b96bf5..0a686ef86f 100644
--- a/examples/twoadicity/.env
+++ b/examples/twoadicity/.env
@@ -1,3 +1,3 @@
-NETWORK=mainnet
+NETWORK=testnet
 PRIVATE_KEY=APrivateKey1zkp8CZNn3yeCseEtxuVPbDCwSyhGW6yZKUYKfgXmcpoGPWH
 ENDPOINT=https://localhost:3030
diff --git a/examples/vote/.env b/examples/vote/.env
index ab80b96bf5..0a686ef86f 100644
--- a/examples/vote/.env
+++ b/examples/vote/.env
@@ -1,3 +1,3 @@
-NETWORK=mainnet
+NETWORK=testnet
 PRIVATE_KEY=APrivateKey1zkp8CZNn3yeCseEtxuVPbDCwSyhGW6yZKUYKfgXmcpoGPWH
 ENDPOINT=https://localhost:3030
diff --git a/leo/cli/commands/account.rs b/leo/cli/commands/account.rs
index e9a36159c7..80e4bff989 100644
--- a/leo/cli/commands/account.rs
+++ b/leo/cli/commands/account.rs
@@ -47,7 +47,7 @@ pub enum Account {
         /// Print sensitive information (such as private key) discreetly to an alternate screen
         #[clap(long)]
         discreet: bool,
-        #[clap(short = 'n', long, help = "Name of the network to use", default_value = "mainnet")]
+        #[clap(short = 'n', long, help = "Name of the network to use", default_value = "testnet")]
         network: String,
         #[clap(
             short = 'e',
@@ -67,7 +67,7 @@ pub enum Account {
         /// Print sensitive information (such as private key) discreetly to an alternate screen
         #[clap(long)]
         discreet: bool,
-        #[clap(short = 'n', long, help = "Name of the network to use", default_value = "mainnet")]
+        #[clap(short = 'n', long, help = "Name of the network to use", default_value = "testnet")]
         network: String,
         #[clap(
             short = 'e',
@@ -94,7 +94,7 @@ pub enum Account {
         /// When enabled, parses the message as bytes instead of Aleo literals
         #[clap(short = 'r', long)]
         raw: bool,
-        #[clap(short = 'n', long, help = "Name of the network to use", default_value = "mainnet")]
+        #[clap(short = 'n', long, help = "Name of the network to use", default_value = "testnet")]
         network: String,
     },
     /// Verify a message from an Aleo address.
@@ -111,7 +111,7 @@ pub enum Account {
         /// When enabled, parses the message as bytes instead of Aleo literals
         #[clap(short = 'r', long)]
         raw: bool,
-        #[clap(short = 'n', long, help = "Name of the network to use", default_value = "mainnet")]
+        #[clap(short = 'n', long, help = "Name of the network to use", default_value = "testnet")]
         network: String,
     },
 }
diff --git a/leo/cli/commands/add.rs b/leo/cli/commands/add.rs
index 54720807eb..57a6479cdb 100755
--- a/leo/cli/commands/add.rs
+++ b/leo/cli/commands/add.rs
@@ -28,7 +28,7 @@ pub struct Add {
     #[clap(short = 'l', long, help = "Path to local dependency")]
     pub(crate) local: Option<PathBuf>,
 
-    #[clap(short = 'n', long, help = "Name of the network to use", default_value = "mainnet")]
+    #[clap(short = 'n', long, help = "Name of the network to use", default_value = "testnet")]
     pub(crate) network: String,
 
     #[clap(short = 'c', long, help = "Clear all previous dependencies.", default_value = "false")]
diff --git a/leo/cli/commands/example.rs b/leo/cli/commands/example.rs
index 44863c4c98..86d1aac040 100644
--- a/leo/cli/commands/example.rs
+++ b/leo/cli/commands/example.rs
@@ -23,7 +23,7 @@ use std::fs;
 pub struct Example {
     #[clap(name = "NAME", help = "The example to initialize.")]
     pub(crate) name: String,
-    #[clap(short = 'n', long, help = "Name of the network to use", default_value = "mainnet")]
+    #[clap(short = 'n', long, help = "Name of the network to use", default_value = "testnet")]
     pub(crate) network: String,
     #[clap(
         short = 'e',
diff --git a/leo/cli/commands/new.rs b/leo/cli/commands/new.rs
index df5fe53062..00cb9a622d 100644
--- a/leo/cli/commands/new.rs
+++ b/leo/cli/commands/new.rs
@@ -24,7 +24,7 @@ use leo_retriever::NetworkName;
 pub struct New {
     #[clap(name = "NAME", help = "Set package name")]
     pub(crate) name: String,
-    #[clap(short = 'n', long, help = "Name of the network to use", default_value = "mainnet")]
+    #[clap(short = 'n', long, help = "Name of the network to use", default_value = "testnet")]
     pub(crate) network: String,
     #[clap(
         short = 'e',
diff --git a/tests/expectations/compiler/structs/empty_struct_fail.out b/tests/expectations/compiler/structs/empty_struct_fail.out
new file mode 100644
index 0000000000..d9815695c9
--- /dev/null
+++ b/tests/expectations/compiler/structs/empty_struct_fail.out
@@ -0,0 +1,5 @@
+---
+namespace: Compile
+expectation: Fail
+outputs:
+  - "Error [ETYC0372112]: A struct must have at least one member.\n    --> compiler-test:4:5\n     |\n   4 |     struct Bar {}\n     |     ^^^^^^^^^^^^^\nError [ETYC0372083]: A program must have at least one transition function.\n    --> compiler-test:1:1\n     |\n   1 | \n   2 | \n   3 | program test.aleo {\n     | ^^^^^^^^^^^^\n"
diff --git a/tests/expectations/parser/finalize/finalize_fail.out b/tests/expectations/parser/finalize/finalize_fail.out
index 459d84d50d..c648822a9a 100644
--- a/tests/expectations/parser/finalize/finalize_fail.out
+++ b/tests/expectations/parser/finalize/finalize_fail.out
@@ -2,4 +2,4 @@
 namespace: Parse
 expectation: Fail
 outputs:
-  - "Error [EPAR0370005]: expected 'struct', 'record', 'mapping', '@', 'function', 'transition', 'inline' -- found 'finalize'\n    --> test:6:7\n     |\n   6 |     } finalize {\n     |       ^^^^^^^^"
+  - "Error [EPAR0370005]: expected 'const', 'struct', 'record', 'mapping', '@', 'async', 'function', 'transition', 'inline', '}' -- found 'finalize'\n    --> test:6:7\n     |\n   6 |     } finalize {\n     |       ^^^^^^^^"
diff --git a/tests/expectations/parser/functions/ident_token_fail.out b/tests/expectations/parser/functions/ident_token_fail.out
index a034d20dc4..33283f860f 100644
--- a/tests/expectations/parser/functions/ident_token_fail.out
+++ b/tests/expectations/parser/functions/ident_token_fail.out
@@ -2,4 +2,4 @@
 namespace: Parse
 expectation: Fail
 outputs:
-  - "Error [EPAR0370005]: expected 'struct', 'record', 'mapping', '@', 'function', 'transition', 'inline' -- found '1'\n    --> test:4:5\n     |\n   4 |     1 main() {}}\n     |     ^"
+  - "Error [EPAR0370005]: expected 'const', 'struct', 'record', 'mapping', '@', 'async', 'function', 'transition', 'inline', '}' -- found '1'\n    --> test:4:5\n     |\n   4 |     1 main() {}}\n     |     ^"
diff --git a/tests/expectations/parser/functions/test_keyword_fail.out b/tests/expectations/parser/functions/test_keyword_fail.out
index b4e98eef56..de79c01227 100644
--- a/tests/expectations/parser/functions/test_keyword_fail.out
+++ b/tests/expectations/parser/functions/test_keyword_fail.out
@@ -2,4 +2,4 @@
 namespace: Parse
 expectation: Fail
 outputs:
-  - "Error [EPAR0370005]: expected 'struct', 'record', 'mapping', '@', 'function', 'transition', 'inline' -- found 'test'\n    --> test:4:5\n     |\n   4 |     test main() {}}\n     |     ^^^^"
+  - "Error [EPAR0370005]: expected 'const', 'struct', 'record', 'mapping', '@', 'async', 'function', 'transition', 'inline', '}' -- found 'test'\n    --> test:4:5\n     |\n   4 |     test main() {}}\n     |     ^^^^"
diff --git a/tests/expectations/parser/program/circuit_deprecated_fail.out b/tests/expectations/parser/program/circuit_deprecated_fail.out
index f3f19c147a..b1dd4715b0 100644
--- a/tests/expectations/parser/program/circuit_deprecated_fail.out
+++ b/tests/expectations/parser/program/circuit_deprecated_fail.out
@@ -2,4 +2,4 @@
 namespace: Parse
 expectation: Fail
 outputs:
-  - "Error [EPAR0370005]: expected 'struct', 'record', 'mapping', '@', 'function', 'transition', 'inline' -- found 'circuit'\n    --> test:5:5\n     |\n   5 |     circuit Foo {\n     |     ^^^^^^^"
+  - "Error [EPAR0370005]: expected 'const', 'struct', 'record', 'mapping', '@', 'async', 'function', 'transition', 'inline', '}' -- found 'circuit'\n    --> test:5:5\n     |\n   5 |     circuit Foo {\n     |     ^^^^^^^"
diff --git a/tests/expectations/parser/program/mapping_fail.out b/tests/expectations/parser/program/mapping_fail.out
index 233b09e314..c675f52513 100644
--- a/tests/expectations/parser/program/mapping_fail.out
+++ b/tests/expectations/parser/program/mapping_fail.out
@@ -2,4 +2,4 @@
 namespace: Parse
 expectation: Fail
 outputs:
-  - "Error [EPAR0370005]: expected 'struct', 'record', 'mapping', '@', 'function', 'transition', 'inline' -- found 'mappin'\n    --> test:4:5\n     |\n   4 |     mappin balances: address => u128;\n     |     ^^^^^^"
+  - "Error [EPAR0370005]: expected 'const', 'struct', 'record', 'mapping', '@', 'async', 'function', 'transition', 'inline', '}' -- found 'mappin'\n    --> test:4:5\n     |\n   4 |     mappin balances: address => u128;\n     |     ^^^^^^"
diff --git a/tests/tests/compiler/structs/empty_struct_fail.leo b/tests/tests/compiler/structs/empty_struct_fail.leo
new file mode 100644
index 0000000000..020290cc35
--- /dev/null
+++ b/tests/tests/compiler/structs/empty_struct_fail.leo
@@ -0,0 +1,12 @@
+/*
+namespace: Compile
+expectation: Fail
+*/
+
+program test.aleo {
+    struct Bar {}
+
+    function main() -> bool {
+        return true;
+    }
+}
diff --git a/utils/retriever/Cargo.toml b/utils/retriever/Cargo.toml
index 13ee323796..f0b4b663ee 100644
--- a/utils/retriever/Cargo.toml
+++ b/utils/retriever/Cargo.toml
@@ -69,4 +69,4 @@ version = "3.10.1"
 version = "0.8.14"
 
 [dependencies.ureq]
-version = "2.9"
+version = "2.10"
