diff --git a/asg/src/error/mod.rs b/asg/src/error/mod.rs
index 5eb6f79c58..3205e733ae 100644
--- a/asg/src/error/mod.rs
+++ b/asg/src/error/mod.rs
@@ -17,7 +17,7 @@
 //! Errors encountered when attempting to convert to an asg from an ast.
 
 use crate::Span;
-use leo_ast::{FormattedError, LeoError};
+use leo_ast::{AstError, FormattedError, LeoError};
 use leo_parser::SyntaxError;
 
 #[derive(Debug, Error)]
@@ -28,6 +28,9 @@ pub enum AsgConvertError {
     #[error("{}", _0)]
     ImportError(FormattedError),
 
+    #[error("{}", _0)]
+    AstError(#[from] AstError),
+
     #[error("{}", _0)]
     InternalError(String),
 
diff --git a/imports/src/errors/import_parser.rs b/imports/src/errors/import_parser.rs
index c91d06a006..c653d4dbf9 100644
--- a/imports/src/errors/import_parser.rs
+++ b/imports/src/errors/import_parser.rs
@@ -14,7 +14,7 @@
 // You should have received a copy of the GNU General Public License
 // along with the Leo library. If not, see <https://www.gnu.org/licenses/>.
 use leo_asg::AsgConvertError;
-use leo_ast::{FormattedError, Identifier, LeoError, Span};
+use leo_ast::{AstError, FormattedError, Identifier, LeoError, Span};
 use leo_parser::SyntaxError;
 
 use std::{io, path::Path};
@@ -26,6 +26,10 @@ pub enum ImportParserError {
 
     #[error("{}", _0)]
     SyntaxError(#[from] SyntaxError),
+
+    #[error("{}", _0)]
+    AstError(#[from] AstError),
+
     #[error("{}", _0)]
     AsgConvertError(#[from] AsgConvertError),
 }
@@ -37,6 +41,7 @@ impl Into<AsgConvertError> for ImportParserError {
         match self {
             ImportParserError::Error(x) => AsgConvertError::ImportError(x),
             ImportParserError::SyntaxError(x) => x.into(),
+            ImportParserError::AstError(x) => AsgConvertError::AstError(x),
             ImportParserError::AsgConvertError(x) => x,
         }
     }
diff --git a/imports/src/parser/parse_symbol.rs b/imports/src/parser/parse_symbol.rs
index 3a09506546..ce1f391f60 100644
--- a/imports/src/parser/parse_symbol.rs
+++ b/imports/src/parser/parse_symbol.rs
@@ -54,8 +54,10 @@ impl<'a> ImportParser<'a> {
         // Build the package abstract syntax tree.
         let program_string =
             &std::fs::read_to_string(&file_path).map_err(|x| ImportParserError::io_error(span, file_path_str, x))?;
-        let mut ast = leo_parser::parse(&file_path_str, &program_string)?;
-        ast.name = file_name;
-        Ok(ast)
+        let mut program = leo_parser::parse(&file_path_str, &program_string)?;
+        program.name = file_name;
+        let mut ast = leo_ast::Ast::new(program);
+        ast.canonicalize()?;
+        Ok(ast.into_repr())
     }
 }
diff --git a/tests/compiler/import/imports/lib.leo b/tests/compiler/import/imports/lib.leo
new file mode 100644
index 0000000000..68d62cfc4b
--- /dev/null
+++ b/tests/compiler/import/imports/lib.leo
@@ -0,0 +1,3 @@
+function ello() -> [char; 4] {
+  return "ello";
+}
\ No newline at end of file
diff --git a/tests/compiler/import/string_import.leo b/tests/compiler/import/string_import.leo
new file mode 100644
index 0000000000..b073b36fc9
--- /dev/null
+++ b/tests/compiler/import/string_import.leo
@@ -0,0 +1,12 @@
+// namespace: Compile
+// expectation: Pass
+// input_file: input/dummy.in
+// cwd: imports
+
+import lib.ello;
+
+function main(y: bool) -> bool {
+    let ello_str = ello();
+    console.log("{}", ello_str);
+    return ello_str == "ello" && y;
+}
\ No newline at end of file
