diff --git a/leo/commands/add.rs b/leo/commands/add.rs
index afe77d8feb..834609a2df 100644
--- a/leo/commands/add.rs
+++ b/leo/commands/add.rs
@@ -21,7 +21,13 @@
 //    leo add -a author -p package_name
 //
 
-use crate::{cli::CLI, cli_types::*, config::*, errors::AddError::*};
+use crate::{
+    cli::CLI,
+    cli_types::*,
+    config::*,
+    errors::{AddError::*, CLIError},
+};
+
 use leo_package::{
     imports::{ImportsDirectory, IMPORTS_DIRECTORY_NAME},
     root::Manifest,
@@ -105,14 +111,16 @@ impl CLI for AddCommand {
         }
     }
 
-    fn output(options: Self::Options) -> Result<Self::Output, crate::errors::CLIError> {
+    fn output(options: Self::Options) -> Result<Self::Output, CLIError> {
         // Begin "Adding" context for console logging
         let span = tracing::span!(tracing::Level::INFO, "Adding");
         let _enter = span.enter();
 
-        let token = read_token()?;
-
+        // token can only be there if user has signed into Aleo PM
+        // so we need to verify that token exists before doing anything else
+        let token = read_token().map_err(|_| -> CLIError { NotAuthorized.into() })?;
         let path = current_dir()?;
+
         // Enforce that the current directory is a leo package
         Manifest::try_from(path.as_path())?;
 
diff --git a/leo/commands/login.rs b/leo/commands/login.rs
index 90722b889a..bc0c96ecf9 100644
--- a/leo/commands/login.rs
+++ b/leo/commands/login.rs
@@ -25,12 +25,13 @@ use crate::{
     cli::CLI,
     cli_types::*,
     config::*,
-    errors::LoginError::{CannotGetToken, NoConnectionFound, NoCredentialsProvided, WrongLoginOrPassword},
+    errors::{CLIError, LoginError::*},
 };
 
 use std::collections::HashMap;
 
 pub const LOGIN_URL: &str = "v1/account/authenticate";
+pub const PROFILE_URL: &str = "v1/account/my_profile";
 
 #[derive(Debug)]
 pub struct LoginCommand;
@@ -75,7 +76,7 @@ impl CLI for LoginCommand {
         }
     }
 
-    fn output(options: Self::Options) -> Result<Self::Output, crate::errors::CLIError> {
+    fn output(options: Self::Options) -> Result<Self::Output, CLIError> {
         // Begin "Login" context for console logging
         let span = tracing::span!(tracing::Level::INFO, "Login");
         let _enter = span.enter();
@@ -86,13 +87,13 @@ impl CLI for LoginCommand {
 
             // Login using username and password
             (None, Some(username), Some(password)) => {
-                let client = reqwest::blocking::Client::new();
-                let url = format!("{}{}", PACKAGE_MANAGER_URL, LOGIN_URL);
-
+                // prepare JSON data to be sent
                 let mut json = HashMap::new();
                 json.insert("email_username", username);
                 json.insert("password", password);
 
+                let client = reqwest::blocking::Client::new();
+                let url = format!("{}{}", PACKAGE_MANAGER_URL, LOGIN_URL);
                 let response: HashMap<String, String> = match client.post(&url).json(&json).send() {
                     Ok(result) => match result.json() {
                         Ok(json) => json,
@@ -116,7 +117,11 @@ impl CLI for LoginCommand {
 
             // Login using stored JWT credentials.
             // TODO (raychu86) Package manager re-authentication from token
-            (_, _, _) => Some(read_token()?),
+            (_, _, _) => {
+                let token = read_token().map_err(|_| -> CLIError { NoCredentialsProvided.into() })?;
+
+                Some(token)
+            }
         };
 
         match token {
diff --git a/leo/errors/commands/add.rs b/leo/errors/commands/add.rs
index 7e7ab8cdd3..720497a17f 100644
--- a/leo/errors/commands/add.rs
+++ b/leo/errors/commands/add.rs
@@ -18,15 +18,18 @@ use std::ffi::OsString;
 
 #[derive(Debug, Error)]
 pub enum AddError {
-    #[error("connection unavailable {:?}", _0)]
+    #[error("Connection unavailable {:?}", _0)]
     ConnectionUnavailable(OsString),
 
-    #[error("missing author or package name")]
+    #[error("Missing author or package name: leo add author/package")]
     MissingAuthorOrPackageName,
 
-    #[error("invalid remote")]
+    #[error("Invalid remote")]
     InvalidRemote,
 
+    #[error("Not authorized in package manager. Use leo login to sign in")]
+    NotAuthorized,
+
     #[error("{:?}", _0)]
     ZipError(OsString),
 }
