diff --git a/Cargo.lock b/Cargo.lock
index f42438a027..e20224df43 100644
--- a/Cargo.lock
+++ b/Cargo.lock
@@ -71,7 +71,7 @@ dependencies = [
 [[package]]
 name = "aleo"
 version = "0.2.0"
-source = "git+https://github.com/AleoHQ/aleo.git?rev=46b8e50#46b8e5064969ffafff7472de13d4f86a6a44ef0b"
+source = "git+https://github.com/AleoHQ/aleo.git?rev=ac95490#ac95490f691d04c1d95dde6e7c46bc89ebdfa60e"
 dependencies = [
  "anyhow",
  "clap 3.2.17",
diff --git a/Cargo.toml b/Cargo.toml
index f09fbe81fd..d82a34ba87 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -62,7 +62,7 @@ version = "1.5.3"
 
 [dependencies.aleo]
 git = "https://github.com/AleoHQ/aleo.git"
-rev = "46b8e50"
+rev = "ac95490"
 
 [dependencies.snarkvm]
 git = "https://github.com/AleoHQ/snarkVM.git"
diff --git a/leo/commands/node.rs b/leo/commands/node.rs
index 3a8fc840bf..fad9138b7e 100644
--- a/leo/commands/node.rs
+++ b/leo/commands/node.rs
@@ -30,6 +30,7 @@ pub enum Node {
     /// Starts a local development node
     Start {
         /// Skips deploying the local program at genesis.
+        #[structopt(long)]
         nodeploy: bool,
     },
 }
