diff --git a/.github/workflows/ci.yml b/.github/workflows/ci.yml
index 7aab099146..1e7bbe013f 100644
--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -184,6 +184,7 @@ jobs:
       RUSTC_WRAPPER: sccache
       SCCACHE_CACHE_SIZE: 2G
       SCCACHE_DIR: /home/runner/.cache/sccache
+      SCCACHE_RECACHE: 1 # Uncomment this to clear cache, then comment it back out
     steps:
       - name: Checkout
         uses: actions/checkout@v2
diff --git a/.github/workflows/docs.yml b/.github/workflows/docs.yml
index 81b65c7daf..077031e844 100644
--- a/.github/workflows/docs.yml
+++ b/.github/workflows/docs.yml
@@ -12,6 +12,7 @@ jobs:
       RUSTC_WRAPPER: sccache
       SCCACHE_CACHE_SIZE: 2G
       SCCACHE_DIR: /home/runner/.cache/sccache
+      # SCCACHE_RECACHE: 1 # Uncomment this to clear cache, then comment it back out
     steps:
       - name: Checkout
         uses: actions/checkout@v2
@@ -21,7 +22,7 @@ jobs:
       - name: Install sccache
         env:
           LINK: https://github.com/mozilla/sccache/releases/download
-          SCCACHE_VERSION: v0.2.15
+          SCCACHE_VERSION: v0.7.6
         run: |
           SCCACHE_FILE=sccache-$SCCACHE_VERSION-x86_64-unknown-linux-musl
           mkdir -p $HOME/.local/bin
@@ -66,4 +67,4 @@ jobs:
         run: sccache --show-stats
         
       - name: Stop sccache server
-        run: sccache --stop-server || true
\ No newline at end of file
+        run: sccache --stop-server || true
diff --git a/Cargo.lock b/Cargo.lock
index 3b338e758a..aafafeddb3 100644
--- a/Cargo.lock
+++ b/Cargo.lock
@@ -1559,10 +1559,10 @@ name = "leo-compiler"
 version = "1.10.0"
 dependencies = [
  "aleo-std-storage",
- "disassembler",
  "dotenvy",
  "indexmap 1.9.3",
  "leo-ast",
+ "leo-disassembler",
  "leo-errors",
  "leo-package",
  "leo-parser",
@@ -1806,9 +1806,9 @@ checksum = "90ed8c1e510134f979dbc4f070f87d4313098b704861a105fe34231c70a3901c"
 
 [[package]]
 name = "memchr"
-version = "2.7.1"
+version = "2.7.2"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "523dc4f511e55ab87b694dc30d0f820d60906ef06413f93d4d7a1385599cc149"
+checksum = "6c8640c5d730cb13ebd907d8d04b52f55ac9a2eec55b440c8892f40d56c76c1d"
 
 [[package]]
 name = "mime"
@@ -2024,9 +2024,9 @@ checksum = "ff011a302c396a5197692431fc1948019154afc178baf7d8e37367442a4601cf"
 
 [[package]]
 name = "openssl-sys"
-version = "0.9.101"
+version = "0.9.102"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "dda2b0f344e78efc2facf7d195d098df0dd72151b26ab98da807afc26c198dff"
+checksum = "c597637d56fbc83893a35eb0dd04b2b8e7a50c91e64e9493e398b5df4fb45fa2"
 dependencies = [
  "cc",
  "libc",
@@ -2154,9 +2154,9 @@ checksum = "d231b230927b5e4ad203db57bbcbee2802f6bce620b1e4a9024a07d94e2907ec"
 
 [[package]]
 name = "platforms"
-version = "3.3.0"
+version = "3.4.0"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "626dec3cac7cc0e1577a2ec3fc496277ec2baa084bebad95bb6fdbfae235f84c"
+checksum = "db23d408679286588f4d4644f965003d056e3dd5abcaaa938116871d7ce2fee7"
 
 [[package]]
 name = "plotters"
@@ -2545,9 +2545,9 @@ dependencies = [
 
 [[package]]
 name = "rustls-pki-types"
-version = "1.4.0"
+version = "1.4.1"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "868e20fada228fefaf6b652e00cc73623d54f8171e7352c18bb281571f2d92da"
+checksum = "ecd36cc4259e3e4514335c4a138c6b43171a8d61d8f5c9348f9fc7529416f247"
 
 [[package]]
 name = "rustls-webpki"
@@ -3981,9 +3981,9 @@ checksum = "1f3ccbac311fea05f86f61904b462b55fb3df8837a366dfc601a0161d0532f20"
 
 [[package]]
 name = "tokio"
-version = "1.36.0"
+version = "1.37.0"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "61285f6515fa018fb2d1e46eb21223fff441ee8db5d0f1435e8ab4f5cdb80931"
+checksum = "1adbebffeca75fcfd058afa480fb6c0b81e165a0323f9c9d39c9697e37c46787"
 dependencies = [
  "backtrace",
  "bytes",
@@ -4669,9 +4669,9 @@ dependencies = [
 
 [[package]]
 name = "zstd-sys"
-version = "2.0.9+zstd.1.5.5"
+version = "2.0.10+zstd.1.5.6"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "9e16efa8a874a0481a574084d34cc26fdb3b99627480f785888deb6386506656"
+checksum = "c253a4914af5bafc8fa8c86ee400827e83cf6ec01195ec1f1ed8441bf00d65aa"
 dependencies = [
  "cc",
  "pkg-config",
diff --git a/compiler/compiler/Cargo.toml b/compiler/compiler/Cargo.toml
index af1cd8bb3c..da49b281f2 100644
--- a/compiler/compiler/Cargo.toml
+++ b/compiler/compiler/Cargo.toml
@@ -45,7 +45,7 @@ version = "0.10"
 version = "1.9"
 features = []
 
-[dev-dependencies.disassembler]
+[dev-dependencies.leo-disassembler]
 path = "../../utils/disassembler"
 
 [dev-dependencies.leo-test-framework]
diff --git a/compiler/compiler/tests/compile.rs b/compiler/compiler/tests/compile.rs
index 0065bdadc6..29712e8c61 100644
--- a/compiler/compiler/tests/compile.rs
+++ b/compiler/compiler/tests/compile.rs
@@ -28,8 +28,8 @@ use utilities::{
     CurrentNetwork,
 };
 
-use disassembler::disassemble_from_str;
 use leo_compiler::{CompilerOptions, OutputOptions};
+use leo_disassembler::disassemble_from_str;
 use leo_errors::{emitter::Handler, LeoError};
 use leo_span::symbol::create_session_if_not_set_then;
 use leo_test_framework::{
diff --git a/compiler/compiler/tests/execute.rs b/compiler/compiler/tests/execute.rs
index 0c3bb40711..9240b53560 100644
--- a/compiler/compiler/tests/execute.rs
+++ b/compiler/compiler/tests/execute.rs
@@ -43,8 +43,8 @@ use leo_test_framework::{
 use aleo_std_storage::StorageMode;
 use snarkvm::{console, prelude::*};
 
-use disassembler::disassemble_from_str;
 use indexmap::IndexMap;
+use leo_disassembler::disassemble_from_str;
 use leo_errors::LeoError;
 use leo_span::Symbol;
 use regex::Regex;
