diff --git a/Cargo.lock b/Cargo.lock
index 215d6b98d3..1365cd45e6 100644
--- a/Cargo.lock
+++ b/Cargo.lock
@@ -3901,18 +3901,18 @@ checksum = "5dc41925991e82af3c3e21e25a9aad92e72930af57fbcc4b07867a18d1cd0459"
 
 [[package]]
 name = "thiserror"
-version = "1.0.61"
+version = "1.0.62"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "c546c80d6be4bc6a00c0f01730c08df82eaa7a7a61f11d656526506112cc1709"
+checksum = "f2675633b1499176c2dff06b0856a27976a8f9d436737b4cf4f312d4d91d8bbb"
 dependencies = [
  "thiserror-impl",
 ]
 
 [[package]]
 name = "thiserror-impl"
-version = "1.0.61"
+version = "1.0.62"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "46c3384250002a6d5af4d114f2845d37b57521033f30d5c3f46c4d70e1197533"
+checksum = "d20468752b09f49e909e55a5d338caa8bedf615594e9d80bc4c565d30faf798c"
 dependencies = [
  "proc-macro2",
  "quote 1.0.36",
diff --git a/compiler/parser/src/parser/context.rs b/compiler/parser/src/parser/context.rs
index da9ade6085..3e2db39f7d 100644
--- a/compiler/parser/src/parser/context.rs
+++ b/compiler/parser/src/parser/context.rs
@@ -201,7 +201,7 @@ impl<'a, N: Network> ParserContext<'a, N> {
     }
 
     /// Parses a list of `T`s using `inner`
-    /// The opening and closing delimiters are `bra` and `ket`,
+    /// The opening and closing delimiters are specified in `delimiter`,
     /// and elements in the list are optionally separated by `sep`.
     /// When `(list, true)` is returned, `sep` was a terminator.
     pub(super) fn parse_list<T>(
diff --git a/compiler/parser/src/parser/file.rs b/compiler/parser/src/parser/file.rs
index afc79846a9..1a6e4c64d9 100644
--- a/compiler/parser/src/parser/file.rs
+++ b/compiler/parser/src/parser/file.rs
@@ -296,7 +296,7 @@ impl<N: Network> ParserContext<'_, N> {
         }
     }
 
-    /// Returns a [`Input`] AST node if the next tokens represent a function output.
+    /// Returns an [`Input`] AST node if the next tokens represent a function input.
     fn parse_input(&mut self) -> Result<functions::Input> {
         let mode = self.parse_mode()?;
         let name = self.expect_identifier()?;
@@ -307,7 +307,7 @@ impl<N: Network> ParserContext<'_, N> {
         Ok(functions::Input { identifier: name, mode, type_, span: name.span, id: self.node_builder.next_id() })
     }
 
-    /// Returns a [`Output`] AST node if the next tokens represent a function output.
+    /// Returns an [`Output`] AST node if the next tokens represent a function output.
     fn parse_output(&mut self) -> Result<Output> {
         let mode = self.parse_mode()?;
         let (type_, span) = self.parse_type()?;
@@ -377,11 +377,13 @@ impl<N: Network> ParserContext<'_, N> {
             }
         };
 
-        // Parse the function body. Allow empty blocks. `fn foo(a:u8);`
+        // Parse the function body, which must be a block,
+        // but we also allow no body and a semicolon instead (e.g. `fn foo(a:u8);`).
         let (_has_empty_block, block) = match &self.token.token {
             Token::LeftCurly => (false, self.parse_block()?),
             Token::Semicolon => {
                 let semicolon = self.expect(&Token::Semicolon)?;
+                // TODO: make a distinction between empty block and no block (i.e. semicolon)
                 (true, Block { statements: Vec::new(), span: semicolon, id: self.node_builder.next_id() })
             }
             _ => self.unexpected("block or semicolon")?,
diff --git a/errors/Cargo.toml b/errors/Cargo.toml
index 53e3b77cf4..e370e57dbf 100644
--- a/errors/Cargo.toml
+++ b/errors/Cargo.toml
@@ -45,4 +45,4 @@ version = "1.0.204"
 features = [ "derive", "rc" ]
 
 [dependencies.thiserror]
-version = "1.0.61"
+version = "1.0.62"
diff --git a/leo/cli/commands/execute.rs b/leo/cli/commands/execute.rs
index cee8dc6e84..0f9be2bb4e 100644
--- a/leo/cli/commands/execute.rs
+++ b/leo/cli/commands/execute.rs
@@ -172,8 +172,13 @@ fn handle_execute<A: Aleo>(
         // Initialize the VM.
         let vm = VM::from(store)?;
 
+        // Remove the `.aleo` extension from the program name, if it exists.
+        let program_name = match program_name.strip_suffix(".aleo") {
+            Some(name) => name.to_string(),
+            None => program_name,
+        };
         // Load the main program, and all of its imports.
-        let program_id = &ProgramID::<A::Network>::from_str(&format!("{}.aleo", program_name))?;
+        let program_id = &ProgramID::<A::Network>::from_str(&format!("{program_name}.aleo"))?;
         load_program_from_network(context.clone(), &mut vm.process().write(), program_id, network, endpoint)?;
 
         let fee_record = if let Some(record) = command.fee_options.record {
