diff --git a/errors/src/errors/utils/util_errors.rs b/errors/src/errors/utils/util_errors.rs
index b10902a7fe..ccade9b17f 100644
--- a/errors/src/errors/utils/util_errors.rs
+++ b/errors/src/errors/utils/util_errors.rs
@@ -75,6 +75,7 @@ create_messages!(
         help: None,
     }
 
+    // TODO: Unused, remove.
     @backtraced
     reqwest_error {
         args: (error: impl Display),
@@ -140,8 +141,8 @@ create_messages!(
 
     @backtraced
     failed_to_retrieve_from_endpoint {
-        args: (error: impl Display),
-        msg: format!("{error}"),
+        args: (url: impl Display, error: impl Display),
+        msg: format!("Failed to retrieve from endpoint `{url}`: {error}"),
         help: None,
     }
 
diff --git a/leo/package/src/lib.rs b/leo/package/src/lib.rs
index a02a66306c..909135e59e 100644
--- a/leo/package/src/lib.rs
+++ b/leo/package/src/lib.rs
@@ -163,7 +163,7 @@ pub fn fetch_from_network_plain(url: &str) -> Result<String, UtilError> {
         .get(url)
         .header("X-Leo-Version", env!("CARGO_PKG_VERSION"))
         .call()
-        .map_err(UtilError::failed_to_retrieve_from_endpoint)?;
+        .map_err(|e| UtilError::failed_to_retrieve_from_endpoint(url, e))?;
     match response.status().as_u16() {
         200..=299 => Ok(response.body_mut().read_to_string().unwrap()),
         301 => Err(UtilError::endpoint_moved_error(url)),
diff --git a/leo/package/src/package.rs b/leo/package/src/package.rs
index 8f5dce713e..6ddc7f93c5 100644
--- a/leo/package/src/package.rs
+++ b/leo/package/src/package.rs
@@ -341,6 +341,9 @@ impl Package {
     ) -> Result<()> {
         let name_symbol = symbol(&new.name)?;
 
+        // Get the existing dependencies.
+        let dependencies = map.clone().into_iter().map(|(name, (dep, _))| (name, dep)).collect();
+
         let program = match map.entry(name_symbol) {
             Entry::Occupied(occupied) => {
                 // We've already visited this dependency. Just make sure it's compatible with
@@ -360,7 +363,7 @@ impl Package {
                     (Some(path), Location::Local) if !no_local => {
                         // It's a local dependency.
                         if path.extension().and_then(|p| p.to_str()) == Some("aleo") && path.is_file() {
-                            Program::from_aleo_path(name_symbol, path)?
+                            Program::from_aleo_path(name_symbol, path, &dependencies)?
                         } else {
                             Program::from_package_path(name_symbol, path)?
                         }
diff --git a/leo/package/src/program.rs b/leo/package/src/program.rs
index 7fef0cf48e..d736b8a842 100644
--- a/leo/package/src/program.rs
+++ b/leo/package/src/program.rs
@@ -21,7 +21,7 @@ use leo_span::Symbol;
 
 use snarkvm::prelude::{Program as SvmProgram, TestnetV0};
 
-use indexmap::IndexSet;
+use indexmap::{IndexMap, IndexSet};
 use std::path::Path;
 
 /// Information about an Aleo program.
@@ -39,16 +39,16 @@ pub struct Program {
 impl Program {
     /// Given the location `path` of a `.aleo` file, read the filesystem
     /// to obtain a `Program`.
-    pub fn from_aleo_path<P: AsRef<Path>>(name: Symbol, path: P) -> Result<Self> {
-        Self::from_aleo_path_impl(name, path.as_ref())
+    pub fn from_aleo_path<P: AsRef<Path>>(name: Symbol, path: P, map: &IndexMap<Symbol, Dependency>) -> Result<Self> {
+        Self::from_aleo_path_impl(name, path.as_ref(), map)
     }
 
-    fn from_aleo_path_impl(name: Symbol, path: &Path) -> Result<Self> {
+    fn from_aleo_path_impl(name: Symbol, path: &Path, map: &IndexMap<Symbol, Dependency>) -> Result<Self> {
         let bytecode = std::fs::read_to_string(path).map_err(|e| {
             UtilError::util_file_io_error(format_args!("Trying to read aleo file at {}", path.display()), e)
         })?;
 
-        let dependencies = parse_dependencies_from_aleo(name, &bytecode)?;
+        let dependencies = parse_dependencies_from_aleo(name, &bytecode, map)?;
 
         Ok(Program {
             name,
@@ -180,7 +180,7 @@ impl Program {
                 let url = format!("{endpoint}/{network}/program/{name}.aleo/latest_edition");
                 fetch_from_network(&url).and_then(|contents| {
                     contents.parse::<u16>().map_err(|e| {
-                        UtilError::failed_to_retrieve_from_endpoint(format!("Failed to parse edition as u16: {e}"))
+                        UtilError::failed_to_retrieve_from_endpoint(url, format!("Failed to parse edition as u16: {e}"))
                     })
                 })
             }
@@ -227,9 +227,10 @@ impl Program {
                 let contents = fetch_from_network(&edition_url)
                     .or_else(|_| fetch_from_network(&no_edition_url))
                     .map_err(|err| {
-                        UtilError::failed_to_retrieve_from_endpoint(format_args!(
-                            "Failed to fetch program `{name}` from network `{network}`: {err}"
-                        ))
+                        UtilError::failed_to_retrieve_from_endpoint(
+                            edition_url,
+                            format_args!("Failed to fetch program `{name}` from network `{network}`: {err}"),
+                        )
                     })?;
 
                 // If the file already exists, compare it to the new contents.
@@ -254,7 +255,7 @@ impl Program {
             }
         };
 
-        let dependencies = parse_dependencies_from_aleo(name, &bytecode)?;
+        let dependencies = parse_dependencies_from_aleo(name, &bytecode, &IndexMap::new())?;
 
         Ok(Program {
             name,
@@ -282,14 +283,28 @@ fn canonicalize_dependency_path_relative_to(base: &Path, mut dependency: Depende
 }
 
 /// Parse the `.aleo` file's imports and construct `Dependency`s.
-fn parse_dependencies_from_aleo(name: Symbol, bytecode: &str) -> Result<IndexSet<Dependency>> {
+fn parse_dependencies_from_aleo(
+    name: Symbol,
+    bytecode: &str,
+    existing: &IndexMap<Symbol, Dependency>,
+) -> Result<IndexSet<Dependency>> {
+    // Parse the bytecode into an SVM program.
     let svm_program: SvmProgram<TestnetV0> = bytecode.parse().map_err(|_| UtilError::snarkvm_parsing_error(name))?;
     let dependencies = svm_program
         .imports()
         .keys()
         .map(|program_id| {
-            let name = program_id.to_string();
-            Dependency { name, location: Location::Network, path: None, edition: None }
+            // If the dependency already exists, use it.
+            // Otherwise, assume it's a network dependency.
+            if let Some(dependency) = existing.get(&Symbol::intern(&program_id.name().to_string())) {
+                dependency.clone()
+            } else {
+                println!(
+                    "Warning: Dependency for program `{program_id}` not found, assuming that it is network dependency."
+                );
+                let name = program_id.to_string();
+                Dependency { name, location: Location::Network, path: None, edition: None }
+            }
         })
         .collect();
     Ok(dependencies)
