diff --git a/ast/src/leo.pest b/ast/src/leo.pest
index 38bf0c1a91..2d763db9ff 100644
--- a/ast/src/leo.pest
+++ b/ast/src/leo.pest
@@ -131,7 +131,7 @@ operation_pow_assign = { "**=" }
 /// Types
 
 // Declared in types/type_.rs
-type_ = { type_self | type_tuple |  type_array | type_data | type_circuit }
+type_ = { type_self | type_tuple | type_array | type_data | type_circuit }
 
 // Declared in types/integer_type.rs
 type_integer = {
diff --git a/compiler/src/expression/array/array.rs b/compiler/src/expression/array/array.rs
index 6c3c2160ae..8c06345902 100644
--- a/compiler/src/expression/array/array.rs
+++ b/compiler/src/expression/array/array.rs
@@ -46,13 +46,13 @@ impl<F: Field + PrimeField, G: GroupType<F>> ConstrainedProgram<F, G> {
         if let Some(type_) = expected_type {
             match type_ {
                 Type::Array(ref type_, ref dimensions) => {
-                    let number = match dimensions.last() {
+                    let number = match dimensions.first() {
                         Some(number) => number.clone(),
                         None => return Err(ExpressionError::unexpected_array(type_.to_string(), span)),
                     };
 
                     expected_dimensions.push(number);
-                    expected_type = Some(type_.inner_dimension(dimensions).clone());
+                    expected_type = Some(type_.outer_dimension(dimensions).clone());
                 }
                 ref type_ => {
                     return Err(ExpressionError::unexpected_array(type_.to_string(), span));
@@ -69,9 +69,7 @@ impl<F: Field + PrimeField, G: GroupType<F>> ConstrainedProgram<F, G> {
                         match self.get(&array_name) {
                             Some(value) => match value {
                                 ConstrainedValue::Array(array) => result.extend(array.clone()),
-                                value => {
-                                    return Err(ExpressionError::invalid_spread(value.to_string(), span));
-                                }
+                                value => return Err(ExpressionError::invalid_spread(value.to_string(), span)),
                             },
                             None => return Err(ExpressionError::undefined_array(identifier.name, span)),
                         }
diff --git a/compiler/src/value/value.rs b/compiler/src/value/value.rs
index c229fa2e9c..57e4df3f89 100644
--- a/compiler/src/value/value.rs
+++ b/compiler/src/value/value.rs
@@ -112,7 +112,7 @@ impl<F: Field + PrimeField, G: GroupType<F>> ConstrainedValue<F, G> {
                 let array_type = types[0].to_type(span.clone())?;
                 let count = types.len();
 
-                // nested array type
+                // Nested array type
                 if let Type::Array(inner_type, inner_dimensions) = &array_type {
                     let mut dimensions = inner_dimensions.clone();
                     dimensions.push(count);
diff --git a/compiler/tests/array/input/initializer_fail.in b/compiler/tests/array/input/initializer_fail.in
index cf991c79b6..0ff699aa13 100644
--- a/compiler/tests/array/input/initializer_fail.in
+++ b/compiler/tests/array/input/initializer_fail.in
@@ -1,2 +1,2 @@
 [main]
-a: [u8; (3, 2)] = [0; (2, 2)];
\ No newline at end of file
+a: [u8; (3, 2)] = [0u8; (2, 2)];
diff --git a/compiler/tests/array/input/inline_fail.leo b/compiler/tests/array/input/inline_fail.leo
index ef826e9f0c..1e380be10d 100644
--- a/compiler/tests/array/input/inline_fail.leo
+++ b/compiler/tests/array/input/inline_fail.leo
@@ -1,3 +1,3 @@
 function main() {
     let a = [1u8, bool];
-}
\ No newline at end of file
+}
diff --git a/compiler/tests/array/input/input_nested_3x2.in b/compiler/tests/array/input/input_nested_3x2.in
new file mode 100644
index 0000000000..9eaafe9bdd
--- /dev/null
+++ b/compiler/tests/array/input/input_nested_3x2.in
@@ -0,0 +1,2 @@
+[main]
+a: [u8; (3, 2)] = [[0u8; 2]; 3];
diff --git a/compiler/tests/array/input/input_nested_3x2_fail.in b/compiler/tests/array/input/input_nested_3x2_fail.in
new file mode 100644
index 0000000000..d1c114a08b
--- /dev/null
+++ b/compiler/tests/array/input/input_nested_3x2_fail.in
@@ -0,0 +1,2 @@
+[main]
+a: [u8; (3, 2)] = [[0u8; 3]; 2)];
diff --git a/compiler/tests/array/input/input_tuple_3x2.in b/compiler/tests/array/input/input_tuple_3x2.in
new file mode 100644
index 0000000000..1c0bd495ec
--- /dev/null
+++ b/compiler/tests/array/input/input_tuple_3x2.in
@@ -0,0 +1,2 @@
+[main]
+a: [u8; (3, 2)] = [0u8; (3, 2)];
diff --git a/compiler/tests/array/input/input_tuple_3x2_fail.in b/compiler/tests/array/input/input_tuple_3x2_fail.in
new file mode 100644
index 0000000000..c281d9d464
--- /dev/null
+++ b/compiler/tests/array/input/input_tuple_3x2_fail.in
@@ -0,0 +1,2 @@
+[main]
+a: [u8; (3, 2)] = [0u8; (2, 3)];
diff --git a/compiler/tests/array/input_nested_3x2.leo b/compiler/tests/array/input_nested_3x2.leo
new file mode 100644
index 0000000000..88777e90f7
--- /dev/null
+++ b/compiler/tests/array/input_nested_3x2.leo
@@ -0,0 +1,3 @@
+function main(a: [u8; (3, 2)]) {
+    console.assert(a == [[0u8; 2]; 3)]);
+}
diff --git a/compiler/tests/array/input_nested_3x2_fail.leo b/compiler/tests/array/input_nested_3x2_fail.leo
new file mode 100644
index 0000000000..1f5bf7b334
--- /dev/null
+++ b/compiler/tests/array/input_nested_3x2_fail.leo
@@ -0,0 +1,3 @@
+function main(a: [u8; (3, 2)]) {
+    console.assert(a == [[0u8; 2]; 3)]); // This should be written the right way as this test is for the input file.
+}
diff --git a/compiler/tests/array/input_tuple_3x2.leo b/compiler/tests/array/input_tuple_3x2.leo
new file mode 100644
index 0000000000..4c3c001f41
--- /dev/null
+++ b/compiler/tests/array/input_tuple_3x2.leo
@@ -0,0 +1,3 @@
+function main(a: [u8; (3, 2)]) {
+    console.assert(a == [0u8; (3, 2)]);
+}
diff --git a/compiler/tests/array/input_tuple_3x2_fail.leo b/compiler/tests/array/input_tuple_3x2_fail.leo
new file mode 100644
index 0000000000..4c3c001f41
--- /dev/null
+++ b/compiler/tests/array/input_tuple_3x2_fail.leo
@@ -0,0 +1,3 @@
+function main(a: [u8; (3, 2)]) {
+    console.assert(a == [0u8; (3, 2)]);
+}
diff --git a/compiler/tests/array/mod.rs b/compiler/tests/array/mod.rs
index b9a337e756..cee31fec09 100644
--- a/compiler/tests/array/mod.rs
+++ b/compiler/tests/array/mod.rs
@@ -120,33 +120,45 @@ fn test_initializer_input_fail() {
 }
 
 #[test]
-fn test_spread() {
-    let program_bytes = include_bytes!("spread.leo");
-    let input_bytes = include_bytes!("input/three_ones.in");
+#[ignore]
+// TODO (howardwu): Add support for this syntax.
+fn test_input_nested_3x2() {
+    let program_bytes = include_bytes!("input_nested_3x2.leo");
+    let input_bytes = include_bytes!("input/input_nested_3x2.in");
     let program = parse_program_with_input(program_bytes, input_bytes).unwrap();
 
     assert_satisfied(program);
 }
 
 #[test]
-fn test_slice() {
-    let program_bytes = include_bytes!("slice.leo");
-    let input_bytes = include_bytes!("input/three_ones.in");
+fn test_input_nested_3x2_fail() {
+    let program_bytes = include_bytes!("input_nested_3x2_fail.leo");
+    let input_bytes = include_bytes!("input/input_nested_3x2_fail.in");
+    let syntax_error = parse_program_with_input(program_bytes, input_bytes).is_err();
+
+    assert!(syntax_error);
+}
+
+#[test]
+fn test_input_tuple_3x2() {
+    let program_bytes = include_bytes!("input_tuple_3x2.leo");
+    let input_bytes = include_bytes!("input/input_tuple_3x2.in");
     let program = parse_program_with_input(program_bytes, input_bytes).unwrap();
 
     assert_satisfied(program);
 }
 
 #[test]
-fn test_multi() {
-    let program_bytes = include_bytes!("multi.leo");
-    let program = parse_program(program_bytes).unwrap();
+fn test_input_tuple_3x2_fail() {
+    let program_bytes = include_bytes!("input_tuple_3x2_fail.leo");
+    let input_bytes = include_bytes!("input/input_tuple_3x2_fail.in");
+    let syntax_error = parse_program_with_input(program_bytes, input_bytes).is_err();
 
-    assert_satisfied(program);
+    assert!(syntax_error);
 }
 
 #[test]
-fn test_multi_fail() {
+fn test_multi_fail_initializer() {
     let program_bytes = include_bytes!("multi_fail_initializer.leo");
     let program = parse_program(program_bytes).unwrap();
 
@@ -176,3 +188,53 @@ fn test_multi_initializer_fail() {
 
     expect_compiler_error(program);
 }
+
+#[test]
+fn test_nested_3x2_value() {
+    let program_bytes = include_bytes!("nested_3x2_value.leo");
+    let program = parse_program(program_bytes).unwrap();
+
+    assert_satisfied(program);
+}
+
+#[test]
+fn test_nested_3x2_value_fail() {
+    let program_bytes = include_bytes!("nested_3x2_value_fail.leo");
+    let program = parse_program(program_bytes).unwrap();
+
+    expect_compiler_error(program);
+}
+
+#[test]
+fn test_tuple_3x2_value() {
+    let program_bytes = include_bytes!("tuple_3x2_value.leo");
+    let program = parse_program(program_bytes).unwrap();
+
+    assert_satisfied(program);
+}
+
+#[test]
+fn test_tuple_3x2_value_fail() {
+    let program_bytes = include_bytes!("tuple_3x2_value_fail.leo");
+    let program = parse_program(program_bytes).unwrap();
+
+    expect_compiler_error(program);
+}
+
+#[test]
+fn test_spread() {
+    let program_bytes = include_bytes!("spread.leo");
+    let input_bytes = include_bytes!("input/three_ones.in");
+    let program = parse_program_with_input(program_bytes, input_bytes).unwrap();
+
+    assert_satisfied(program);
+}
+
+#[test]
+fn test_slice() {
+    let program_bytes = include_bytes!("slice.leo");
+    let input_bytes = include_bytes!("input/three_ones.in");
+    let program = parse_program_with_input(program_bytes, input_bytes).unwrap();
+
+    assert_satisfied(program);
+}
diff --git a/compiler/tests/array/multi.leo b/compiler/tests/array/multi.leo
deleted file mode 100644
index 8602c39465..0000000000
--- a/compiler/tests/array/multi.leo
+++ /dev/null
@@ -1,8 +0,0 @@
-// Multidimensional array syntax in leo
-function main() {
-    const a = [[0u32, 0u32, 0u32], [0u32, 0u32, 0u32]]; // inline
-
-    const b: [u32; (3, 2)] = [[0; 3]; 2]; // initializer
-
-    console.assert(a == b);
-}
\ No newline at end of file
diff --git a/compiler/tests/array/nested_3x2_value.leo b/compiler/tests/array/nested_3x2_value.leo
new file mode 100644
index 0000000000..c5f64c997e
--- /dev/null
+++ b/compiler/tests/array/nested_3x2_value.leo
@@ -0,0 +1,8 @@
+// Multidimensional array syntax in leo
+function main() {
+    const a = [[0u32, 0u32], [0u32, 0u32], [0u32, 0u32]]; // inline
+
+    const b: [u32; (3, 2)] = [[0; 2]; 3]; // initializer
+
+    console.assert(a == b);
+}
\ No newline at end of file
diff --git a/compiler/tests/array/nested_3x2_value_fail.leo b/compiler/tests/array/nested_3x2_value_fail.leo
new file mode 100644
index 0000000000..a187a51991
--- /dev/null
+++ b/compiler/tests/array/nested_3x2_value_fail.leo
@@ -0,0 +1,4 @@
+// Multidimensional array syntax in leo
+function main() {
+    const a: [u32; (3, 2)] = [[0; 3]; 2]; // initializer (incorrectly reversed ordering)
+}
diff --git a/compiler/tests/array/tuple_3x2_value.leo b/compiler/tests/array/tuple_3x2_value.leo
new file mode 100644
index 0000000000..b6659539d5
--- /dev/null
+++ b/compiler/tests/array/tuple_3x2_value.leo
@@ -0,0 +1,8 @@
+// Multidimensional array syntax in leo
+function main() {
+    const a = [[0u32, 0u32], [0u32, 0u32], [0u32, 0u32]]; // inline
+
+    const b: [u32; (3, 2)] = [0; (3, 2)]; // initializer
+
+    console.assert(a == b);
+}
\ No newline at end of file
diff --git a/compiler/tests/array/tuple_3x2_value_fail.leo b/compiler/tests/array/tuple_3x2_value_fail.leo
new file mode 100644
index 0000000000..78593ab696
--- /dev/null
+++ b/compiler/tests/array/tuple_3x2_value_fail.leo
@@ -0,0 +1,4 @@
+// Multidimensional array syntax in leo
+function main() {
+    const a: [u32; (3, 2)] = [0; (2, 3)]; // initializer (incorrectly reversed ordering)
+}
diff --git a/leo/cli.rs b/leo/cli.rs
index acad5a2f58..5393d0cc66 100644
--- a/leo/cli.rs
+++ b/leo/cli.rs
@@ -33,11 +33,17 @@ pub trait CLI {
     fn new<'a, 'b>() -> App<'a, 'b> {
         let arguments = &Self::ARGUMENTS
             .iter()
-            .map(|a| Arg::with_name(a.0).help(a.1).required(a.2).index(a.3))
+            .map(|a| {
+                let mut args = Arg::with_name(a.0).help(a.1).required(a.3).index(a.4);
+                if a.2.len() > 0 {
+                    args = args.possible_values(a.2);
+                }
+                args
+            })
             .collect::<Vec<Arg<'static, 'static>>>();
         let flags = &Self::FLAGS
             .iter()
-            .map(|a| Arg::from_usage(a).global(true))
+            .map(|a| Arg::from_usage(a))
             .collect::<Vec<Arg<'static, 'static>>>();
         let options = &Self::OPTIONS
             .iter()
@@ -56,6 +62,22 @@ pub trait CLI {
                     .about(s.1)
                     .args(
                         &s.2.iter()
+                            .map(|a| {
+                                let mut args = Arg::with_name(a.0).help(a.1).required(a.3).index(a.4);
+                                if a.2.len() > 0 {
+                                    args = args.possible_values(a.2);
+                                }
+                                args
+                            })
+                            .collect::<Vec<Arg<'static, 'static>>>(),
+                    )
+                    .args(
+                        &s.3.iter()
+                            .map(|a| Arg::from_usage(a))
+                            .collect::<Vec<Arg<'static, 'static>>>(),
+                    )
+                    .args(
+                        &s.4.iter()
                             .map(|a| match a.2.len() > 0 {
                                 true => Arg::from_usage(a.0)
                                     .conflicts_with_all(a.1)
@@ -65,7 +87,7 @@ pub trait CLI {
                             })
                             .collect::<Vec<Arg<'static, 'static>>>(),
                     )
-                    .settings(s.3)
+                    .settings(s.5)
             })
             .collect::<Vec<App<'static, 'static>>>();
 
diff --git a/leo/cli_types.rs b/leo/cli_types.rs
index da7521e8d8..a2527635e7 100644
--- a/leo/cli_types.rs
+++ b/leo/cli_types.rs
@@ -24,10 +24,14 @@ pub type DescriptionType = &'static str;
 
 pub type RequiredType = bool;
 
+pub type PossibleValuesType = &'static [&'static str];
+
 pub type IndexType = u64;
 
-pub type ArgumentType = (NameType, DescriptionType, RequiredType, IndexType);
+pub type ArgumentType = (NameType, DescriptionType, PossibleValuesType, RequiredType, IndexType);
 
+// Format
+// "[flag] -f --flag 'Add flag description here'"
 pub type FlagType = &'static str;
 
 // Format
@@ -39,4 +43,11 @@ pub type OptionType = (
     &'static [&'static str],
 );
 
-pub type SubCommandType = (NameType, AboutType, &'static [OptionType], &'static [AppSettings]);
+pub type SubCommandType = (
+    NameType,
+    AboutType,
+    &'static [ArgumentType],
+    &'static [FlagType],
+    &'static [OptionType],
+    &'static [AppSettings],
+);
diff --git a/leo/commands/add.rs b/leo/commands/add.rs
index 0f12c6f3b7..f379ec2840 100644
--- a/leo/commands/add.rs
+++ b/leo/commands/add.rs
@@ -35,7 +35,7 @@ use std::{
     io::{Read, Write},
 };
 
-pub const ADD_URL: &str = "api/package/fetch";
+pub const ADD_URL: &str = "v1/package/fetch";
 
 #[derive(Debug)]
 pub struct AddCommand;
@@ -47,10 +47,11 @@ impl CLI for AddCommand {
 
     const ABOUT: AboutType = "Install a package from the Aleo Package Manager";
     const ARGUMENTS: &'static [ArgumentType] = &[
-        // (name, description, required, index)
+        // (name, description, possible_values, required, index)
         (
             "REMOTE",
             "Install a package from the Aleo Package Manager with the given remote",
+            &[],
             false,
             1u64,
         ),
diff --git a/leo/commands/build.rs b/leo/commands/build.rs
index 888003d8ce..2bfefe9549 100644
--- a/leo/commands/build.rs
+++ b/leo/commands/build.rs
@@ -186,9 +186,9 @@ impl CLI for BuildCommand {
             // Drop "Compiling" context for console logging
             drop(enter);
 
-            // Begin "Finished" context for console logging todo: @collin figure a way to get this output with tracing without dropping span
-            tracing::span!(tracing::Level::INFO, "Finished").in_scope(|| {
-                tracing::info!("Completed in {} milliseconds\n", start.elapsed().as_millis());
+            // Begin "Done" context for console logging todo: @collin figure a way to get this output with tracing without dropping span
+            tracing::span!(tracing::Level::INFO, "Done").in_scope(|| {
+                tracing::info!("Finished in {} milliseconds\n", start.elapsed().as_millis());
             });
 
             return Ok(Some((program, checksum_differs)));
diff --git a/leo/commands/clean.rs b/leo/commands/clean.rs
index d28c919c8d..63d1e0d71b 100644
--- a/leo/commands/clean.rs
+++ b/leo/commands/clean.rs
@@ -75,8 +75,8 @@ impl CLI for CleanCommand {
         // Drop "Compiling" context for console logging
         drop(enter);
 
-        // Begin "Finished" context for console logging
-        tracing::span!(tracing::Level::INFO, "Finished").in_scope(|| {
+        // Begin "Done" context for console logging
+        tracing::span!(tracing::Level::INFO, "Done").in_scope(|| {
             tracing::info!("Program workspace cleaned\n");
         });
 
diff --git a/leo/commands/login.rs b/leo/commands/login.rs
index d805ac8fc6..90722b889a 100644
--- a/leo/commands/login.rs
+++ b/leo/commands/login.rs
@@ -30,7 +30,7 @@ use crate::{
 
 use std::collections::HashMap;
 
-pub const LOGIN_URL: &str = "api/account/authenticate";
+pub const LOGIN_URL: &str = "v1/account/authenticate";
 
 #[derive(Debug)]
 pub struct LoginCommand;
@@ -42,10 +42,11 @@ impl CLI for LoginCommand {
 
     const ABOUT: AboutType = "Login to the Aleo Package Manager";
     const ARGUMENTS: &'static [ArgumentType] = &[
-        // (name, description, required, index)
+        // (name, description, possible_values, required, index)
         (
             "NAME",
             "Sets the authentication token for login to the package manager",
+            &[],
             false,
             1u64,
         ),
diff --git a/leo/commands/new.rs b/leo/commands/new.rs
index d210ff6ebd..9905c74927 100644
--- a/leo/commands/new.rs
+++ b/leo/commands/new.rs
@@ -33,10 +33,11 @@ impl CLI for NewCommand {
 
     const ABOUT: AboutType = "Create a new Leo package in a new directory";
     const ARGUMENTS: &'static [ArgumentType] = &[
-        // (name, description, required, index)
+        // (name, description, possible_values, required, index)
         (
             "NAME",
             "Sets the resulting package name, defaults to the directory name",
+            &[],
             true,
             1u64,
         ),
diff --git a/leo/commands/prove.rs b/leo/commands/prove.rs
index 3773edcbd9..7c6ff69c60 100644
--- a/leo/commands/prove.rs
+++ b/leo/commands/prove.rs
@@ -75,9 +75,9 @@ impl CLI for ProveCommand {
         // Drop "Prover" context for console logging
         drop(enter);
 
-        // Begin "Finished" context for console logging
-        tracing::span!(tracing::Level::INFO, "Finished").in_scope(|| {
-            tracing::info!("Completed in {:?} milliseconds\n", end);
+        // Begin "Done" context for console logging
+        tracing::span!(tracing::Level::INFO, "Done").in_scope(|| {
+            tracing::info!("Finished in {:?} milliseconds\n", end);
         });
 
         Ok((program_proof, prepared_verifying_key))
diff --git a/leo/commands/publish.rs b/leo/commands/publish.rs
index cc81281cab..a4efab3842 100644
--- a/leo/commands/publish.rs
+++ b/leo/commands/publish.rs
@@ -38,7 +38,7 @@ use reqwest::{
 use serde::Deserialize;
 use std::{convert::TryFrom, env::current_dir};
 
-const PUBLISH_URL: &str = "api/package/publish";
+pub const PUBLISH_URL: &str = "v1/package/publish";
 
 #[derive(Deserialize)]
 struct ResponseJson {
diff --git a/leo/commands/remove.rs b/leo/commands/remove.rs
index 4b26ba3e9e..2a8a99e91d 100644
--- a/leo/commands/remove.rs
+++ b/leo/commands/remove.rs
@@ -29,8 +29,14 @@ impl CLI for RemoveCommand {
 
     const ABOUT: AboutType = "Uninstall a package from the current package";
     const ARGUMENTS: &'static [ArgumentType] = &[
-        // (name, description, required, index)
-        ("NAME", "Removes the package from the current directory", true, 1u64),
+        // (name, description, possible_values, required, index)
+        (
+            "NAME",
+            "Removes the package from the current directory",
+            &[],
+            true,
+            1u64,
+        ),
     ];
     const FLAGS: &'static [FlagType] = &[];
     const NAME: NameType = "remove";
diff --git a/leo/commands/run.rs b/leo/commands/run.rs
index 76dcaf839c..14b3e759fa 100644
--- a/leo/commands/run.rs
+++ b/leo/commands/run.rs
@@ -48,7 +48,7 @@ impl CLI for RunCommand {
         let (proof, prepared_verifying_key) = ProveCommand::output(options)?;
 
         // Begin "Verifying" context for console logging
-        let span = tracing::span!(tracing::Level::INFO, "Verifier");
+        let span = tracing::span!(tracing::Level::INFO, "Verifying");
         let enter = span.enter();
 
         tracing::info!("Starting...");
@@ -76,9 +76,9 @@ impl CLI for RunCommand {
         // Drop "Verifying" context for console logging
         drop(enter);
 
-        // Begin "Finished" context for console logging
-        tracing::span!(tracing::Level::INFO, "Finished").in_scope(|| {
-            tracing::info!("Completed in {:?} milliseconds\n", end);
+        // Begin "Done" context for console logging
+        tracing::span!(tracing::Level::INFO, "Done").in_scope(|| {
+            tracing::info!("Finished in {:?} milliseconds\n", end);
         });
 
         Ok(())
diff --git a/leo/commands/setup.rs b/leo/commands/setup.rs
index c1afbd4127..5ee406cdcc 100644
--- a/leo/commands/setup.rs
+++ b/leo/commands/setup.rs
@@ -138,9 +138,9 @@ impl CLI for SetupCommand {
                 // Drop "Setup" context for console logging
                 drop(enter);
 
-                // Begin "Finished" context for console logging
-                tracing::span!(tracing::Level::INFO, "Finished").in_scope(|| {
-                    tracing::info!("Completed in {:?} milliseconds\n", end);
+                // Begin "Done" context for console logging
+                tracing::span!(tracing::Level::INFO, "Done").in_scope(|| {
+                    tracing::info!("Finished in {:?} milliseconds\n", end);
                 });
 
                 Ok((program, proving_key, prepared_verifying_key))
diff --git a/leo/commands/test.rs b/leo/commands/test.rs
index 91e8dbe6dc..33468bb076 100644
--- a/leo/commands/test.rs
+++ b/leo/commands/test.rs
@@ -110,8 +110,8 @@ impl CLI for TestCommand {
 
         // Set the result of the test command to passed if no tests failed.
         if failed == 0 {
-            // Begin "Finished" context for console logging
-            tracing::span!(tracing::Level::INFO, "Finished").in_scope(|| {
+            // Begin "Done" context for console logging
+            tracing::span!(tracing::Level::INFO, "Done").in_scope(|| {
                 tracing::info!(
                     "Tests passed in {} milliseconds. {} passed; {} failed;\n",
                     start.elapsed().as_millis(),
@@ -120,8 +120,8 @@ impl CLI for TestCommand {
                 );
             });
         } else {
-            // Begin "Finished" context for console logging
-            tracing::span!(tracing::Level::ERROR, "Finished").in_scope(|| {
+            // Begin "Done" context for console logging
+            tracing::span!(tracing::Level::ERROR, "Done").in_scope(|| {
                 tracing::error!(
                     "Tests failed in {} milliseconds. {} passed; {} failed;\n",
                     start.elapsed().as_millis(),
diff --git a/leo/commands/update.rs b/leo/commands/update.rs
index 27381f30c8..2290e57e60 100644
--- a/leo/commands/update.rs
+++ b/leo/commands/update.rs
@@ -14,25 +14,59 @@
 // You should have received a copy of the GNU General Public License
 // along with the Leo library. If not, see <https://www.gnu.org/licenses/>.
 
-use crate::{cli::CLI, cli_types::*, updater::Updater};
+use crate::{cli::CLI, cli_types::*, config::Config, updater::Updater};
+
+use clap::AppSettings;
 
 #[derive(Debug)]
 pub struct UpdateCommand;
 
 impl CLI for UpdateCommand {
-    type Options = (bool,);
+    // (show_all_versions, quiet)
+    type Options = Option<(bool, bool, bool)>;
     type Output = ();
 
     const ABOUT: AboutType = "Update Leo to the latest version";
     const ARGUMENTS: &'static [ArgumentType] = &[];
-    const FLAGS: &'static [FlagType] = &[("--list")];
+    const FLAGS: &'static [FlagType] = &[
+        "[list] -l --list 'List all available versions of Leo'",
+        "[quiet] -q --quiet 'Suppress outputs to terminal'",
+        "[studio] -s --studio 'For Aleo Studio only'",
+    ];
     const NAME: NameType = "update";
     const OPTIONS: &'static [OptionType] = &[];
-    const SUBCOMMANDS: &'static [SubCommandType] = &[];
+    const SUBCOMMANDS: &'static [SubCommandType] = &[
+        // (name, description, options, settings)
+        (
+            UpdateAutomatic::NAME,
+            UpdateAutomatic::ABOUT,
+            UpdateAutomatic::ARGUMENTS,
+            UpdateAutomatic::FLAGS,
+            &UpdateAutomatic::OPTIONS,
+            &[
+                AppSettings::ColoredHelp,
+                AppSettings::DisableHelpSubcommand,
+                AppSettings::DisableVersion,
+            ],
+        ),
+    ];
 
     fn parse(arguments: &clap::ArgMatches) -> Result<Self::Options, crate::errors::CLIError> {
+        match arguments.subcommand() {
+            ("automatic", Some(arguments)) => {
+                // Run the `automatic` subcommand
+                let options = UpdateAutomatic::parse(arguments)?;
+                let _output = UpdateAutomatic::output(options)?;
+                return Ok(None);
+            }
+            _ => {}
+        };
+
         let show_all_versions = arguments.is_present("list");
-        Ok((show_all_versions,))
+        let quiet = arguments.is_present("quiet");
+        let studio = arguments.is_present("studio");
+
+        Ok(Some((show_all_versions, quiet, studio)))
     }
 
     fn output(options: Self::Options) -> Result<Self::Output, crate::errors::CLIError> {
@@ -40,29 +74,123 @@ impl CLI for UpdateCommand {
         let span = tracing::span!(tracing::Level::INFO, "Updating");
         let _enter = span.enter();
 
-        match options {
-            (true,) => match Updater::show_available_releases() {
+        let (show_all_versions, quiet, studio) = match options {
+            Some(options) => options,
+            None => return Ok(()),
+        };
+
+        match show_all_versions {
+            true => match Updater::show_available_releases() {
                 Ok(_) => return Ok(()),
                 Err(e) => {
                     tracing::error!("Could not fetch that latest version of Leo");
                     tracing::error!("{}", e);
                 }
             },
-            (false,) => match Updater::update_to_latest_release(true) {
-                Ok(status) => {
-                    if status.uptodate() {
-                        tracing::info!("Leo is already on the latest version: {}", status.version());
-                    } else if status.updated() {
-                        tracing::info!("Leo has successfully updated to version: {}", status.version());
-                    }
+            false => {
+                let config = Config::read_config()?;
+
+                // If update is run with studio and the automatic update is off, finish quietly
+                if studio && !config.update.automatic {
                     return Ok(());
                 }
-                Err(e) => {
-                    tracing::error!("Could not update Leo to the latest version");
-                    tracing::error!("{}", e);
+
+                match Updater::update_to_latest_release(!quiet) {
+                    Ok(status) => {
+                        if !quiet {
+                            if status.uptodate() {
+                                tracing::info!("Leo is already on the latest version {}", status.version());
+                            } else if status.updated() {
+                                tracing::info!("Leo has successfully updated to version {}", status.version());
+                            }
+                        }
+                        return Ok(());
+                    }
+                    Err(e) => {
+                        if !quiet {
+                            tracing::error!("Could not update Leo to the latest version");
+                            tracing::error!("{}", e);
+                        }
+                    }
                 }
-            },
+            }
         }
         Ok(())
     }
 }
+
+//TODO (raychu86) Move this to dedicated file/module
+#[derive(Debug)]
+pub struct UpdateAutomatic;
+
+impl CLI for UpdateAutomatic {
+    // (is_automatic, quiet)
+    type Options = (Option<bool>, bool);
+    type Output = ();
+
+    const ABOUT: AboutType = "Setting for automatic updates of Leo";
+    const ARGUMENTS: &'static [ArgumentType] = &[
+        // (name, description, possible_values, required, index)
+        (
+            "automatic",
+            "Enable or disable automatic updates",
+            &["true", "false"],
+            false,
+            1u64,
+        ),
+    ];
+    const FLAGS: &'static [FlagType] = &["[quiet] -q --quiet 'Suppress outputs to terminal'"];
+    const NAME: NameType = "automatic";
+    const OPTIONS: &'static [OptionType] = &[];
+    const SUBCOMMANDS: &'static [SubCommandType] = &[];
+
+    fn parse(arguments: &clap::ArgMatches) -> Result<Self::Options, crate::errors::CLIError> {
+        let quiet = arguments.is_present("quiet");
+
+        match arguments.value_of("automatic") {
+            Some(automatic) => {
+                // TODO enforce that the possible values is true or false
+                let automatic = match automatic {
+                    "true" => Some(true),
+                    "false" => Some(false),
+                    _ => unreachable!(),
+                };
+
+                Ok((automatic, quiet))
+            }
+            None => Ok((None, quiet)),
+        }
+    }
+
+    fn output(options: Self::Options) -> Result<Self::Output, crate::errors::CLIError> {
+        // Begin "Settings" context for console logging
+        let span = tracing::span!(tracing::Level::INFO, "Settings");
+        let enter = span.enter();
+
+        // If a boolean value is provided, update the saved
+        // `automatic` configuration value to this boolean value.
+        if let Some(automatic) = options.0 {
+            Config::set_update_automatic(automatic)?;
+        }
+
+        // If --quiet is not enabled, log the output.
+        if !options.1 {
+            // Read the `automatic` value now.
+            let automatic = Config::read_config()?.update.automatic;
+
+            // Log the output.
+            tracing::debug!("automatic = {}", automatic);
+            match automatic {
+                true => tracing::info!("Automatic updates are enabled. Leo will update as new versions are released."),
+                false => {
+                    tracing::info!("Automatic updates are disabled. Leo will not update as new versions are released.")
+                }
+            };
+        }
+
+        // Drop "Settings" context for console logging.
+        drop(enter);
+
+        Ok(())
+    }
+}
diff --git a/leo/config.rs b/leo/config.rs
index ff15ff4725..a1db95587e 100644
--- a/leo/config.rs
+++ b/leo/config.rs
@@ -26,7 +26,7 @@ use std::{
     path::{Path, PathBuf},
 };
 
-pub const PACKAGE_MANAGER_URL: &str = "https://apm-backend-prod.herokuapp.com/";
+pub const PACKAGE_MANAGER_URL: &str = "https://api.aleo.pm/";
 
 pub const LEO_CREDENTIALS_FILE: &str = "credentials";
 pub const LEO_CONFIG_FILE: &str = "config.toml";
@@ -49,14 +49,27 @@ lazy_static! {
     };
 }
 
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct Update {
+    pub automatic: bool,
+}
+
+impl Default for Update {
+    fn default() -> Self {
+        Self { automatic: true }
+    }
+}
+
 #[derive(Clone, Debug, Serialize, Deserialize)]
 pub struct Config {
-    pub auto_update: bool,
+    pub update: Update,
 }
 
 impl Default for Config {
     fn default() -> Self {
-        Self { auto_update: false }
+        Self {
+            update: Update::default(),
+        }
     }
 }
 
@@ -76,10 +89,19 @@ impl Config {
         }
 
         let toml_string = match fs::read_to_string(&config_path) {
-            Ok(toml) => toml,
+            Ok(mut toml) => {
+                // If the config is using an incorrect format, rewrite it.
+                if let Err(_) = toml::from_str::<Config>(&toml) {
+                    let default_config_string = toml::to_string(&Config::default())?;
+                    fs::write(&config_path, default_config_string.clone())?;
+                    toml = default_config_string;
+                }
+
+                toml
+            }
             Err(_) => {
                 create_dir_all(&config_dir)?;
-                String::new()
+                toml::to_string(&Config::default())?
             }
         };
 
@@ -88,6 +110,21 @@ impl Config {
 
         Ok(config)
     }
+
+    /// Update the `automatic` configuration in the `config.toml` file.
+    pub fn set_update_automatic(automatic: bool) -> Result<(), CLIError> {
+        let mut config = Self::read_config()?;
+
+        if config.update.automatic != automatic {
+            config.update.automatic = automatic;
+
+            // Update the config file
+            let config_path = LEO_CONFIG_PATH.clone();
+            fs::write(&config_path, toml::to_string(&config)?)?;
+        }
+
+        Ok(())
+    }
 }
 
 pub fn write_token(token: &str) -> Result<(), io::Error> {
diff --git a/leo/logger.rs b/leo/logger.rs
index 4095f3c159..cd5de1f697 100644
--- a/leo/logger.rs
+++ b/leo/logger.rs
@@ -186,7 +186,7 @@ where
                 }
             }
 
-            write!(writer, "{:>12} ", colored_string(meta.level(), &message)).expect("Error writing event");
+            write!(writer, "{:>10} ", colored_string(meta.level(), &message)).expect("Error writing event");
         }
 
         ctx.format_fields(writer, event)?;
diff --git a/leo/main.rs b/leo/main.rs
index 7c8c124904..61aa8a7b0c 100644
--- a/leo/main.rs
+++ b/leo/main.rs
@@ -22,8 +22,9 @@ use clap::{App, AppSettings, Arg};
 fn main() -> Result<(), CLIError> {
     let app = App::new("leo")
         .version(include_str!("./leo-version"))
-        .about("Leo compiler and package manager")
-        .author("The Aleo Team <hello@aleo.org>")
+        .before_help("Hello world")
+        // .about("Leo compiler and package manager")
+        // .author("The Aleo Team <hello@aleo.org>")
         .settings(&[
             AppSettings::ColoredHelp,
             AppSettings::DisableHelpSubcommand,
diff --git a/leo/updater.rs b/leo/updater.rs
index 5fdb7b01b0..3b92b990bf 100644
--- a/leo/updater.rs
+++ b/leo/updater.rs
@@ -33,7 +33,7 @@ impl Updater {
             .repo_name(Self::LEO_REPO_NAME)
             .bin_name(Self::LEO_BIN_NAME)
             .current_version(&include_str!("./leo-version").replace('v', ""))
-            .show_download_progress(true)
+            .show_download_progress(show_output)
             .no_confirm(true)
             .show_output(show_output)
             .build()?
@@ -65,7 +65,7 @@ impl Updater {
     pub fn print_cli() {
         let config = Config::read_config().unwrap();
 
-        if config.auto_update {
+        if config.update.automatic {
             // If the auto update configuration is on, attempt to update the version.
             if let Ok(status) = Self::update_to_latest_release(false) {
                 if status.updated() {
diff --git a/typed/src/common/span.rs b/typed/src/common/span.rs
index ac58c59c8b..b5af258eb2 100644
--- a/typed/src/common/span.rs
+++ b/typed/src/common/span.rs
@@ -14,7 +14,7 @@
 // You should have received a copy of the GNU General Public License
 // along with the Leo library. If not, see <https://www.gnu.org/licenses/>.
 
-use pest::{Position, Span as AstSpan};
+use pest::Span as AstSpan;
 use serde::{Deserialize, Serialize};
 
 #[derive(Clone, Debug, PartialEq, Eq, Hash, Serialize, Deserialize)]
diff --git a/typed/src/expression.rs b/typed/src/expression.rs
index 8b654a7352..fb17a983b4 100644
--- a/typed/src/expression.rs
+++ b/typed/src/expression.rs
@@ -501,7 +501,7 @@ impl<'ast> From<ArrayInitializerExpression<'ast>> for Expression {
         // Initializes an arrays elements using the rust `vec!` macro.
         // If there are multiple array dimensions, then `elements` is used as the first expression in a `vec!` macro.
         // This creates a multi-dimensional array by chaining `vec!` macros.
-        for (i, dimension) in dimensions.into_iter().enumerate() {
+        for (i, dimension) in dimensions.into_iter().rev().enumerate() {
             if i == 0 {
                 elements = vec![expression.clone(); dimension];
             } else {
diff --git a/typed/src/input/input_value.rs b/typed/src/input/input_value.rs
index 6f9ad83eee..6a1fbb75a8 100644
--- a/typed/src/input/input_value.rs
+++ b/typed/src/input/input_value.rs
@@ -108,7 +108,7 @@ impl InputValue {
     ) -> Result<Self, InputParserError> {
         let mut array_dimensions = TypedExpression::get_input_array_dimensions(array_type.dimensions.clone());
 
-        // Return an error if the outer array dimension does not equal the number of array elements
+        // Return an error if the outer array dimension does not equal the number of array elements.
         if let Some(outer_dimension) = array_dimensions.pop() {
             array_type.dimensions = array_type.dimensions.next_dimension();
 
@@ -118,13 +118,13 @@ impl InputValue {
         };
 
         let inner_array_type = if array_dimensions.len() == 0 {
-            // this is a single array
+            // This is a single array
             match array_type.type_ {
                 ArrayElement::Basic(basic) => Type::Basic(basic),
                 ArrayElement::Tuple(tuple) => Type::Tuple(tuple),
             }
         } else {
-            // this is a multi-dimensional array
+            // This is a multi-dimensional array
             Type::Array(array_type)
         };
 
