diff --git a/parser/src/tokenizer/mod.rs b/parser/src/tokenizer/mod.rs
index 6177973939..52fded2834 100644
--- a/parser/src/tokenizer/mod.rs
+++ b/parser/src/tokenizer/mod.rs
@@ -84,7 +84,7 @@ pub(crate) fn tokenize(path: &str, input: StrTendril) -> Result<Vec<SpannedToken
                     break;
                 } else if token_len == 0 {
                     return Err(ParserError::unexpected_token(
-                        &input[index..index + 1],
+                        &input[index..].chars().next().unwrap(),
                         &Span::new(
                             line_no,
                             line_no,
