diff --git a/.circleci/leo-new.sh b/.circleci/leo-new.sh
index 132dbf2381..721dd2654e 100755
--- a/.circleci/leo-new.sh
+++ b/.circleci/leo-new.sh
@@ -3,5 +3,13 @@ $LEO new foo
 ls -la
 cd foo && ls -la
 
+# Try to run `leo build`.
+(
+  $LEO build || exit
+)
+
+
 # Try to run `leo run`.
-$LEO run
+(
+  $LEO run || exit
+)
diff --git a/examples/import_point/.gitignore b/examples/import_point/.gitignore
deleted file mode 100644
index b28696155d..0000000000
--- a/examples/import_point/.gitignore
+++ /dev/null
@@ -1,2 +0,0 @@
-outputs/
-build/
diff --git a/examples/import_point/README.md b/examples/import_point/README.md
deleted file mode 100644
index 038ce192d3..0000000000
--- a/examples/import_point/README.md
+++ /dev/null
@@ -1,14 +0,0 @@
-# Leo import example
-An example showing how to import symbols from file `point.leo`.
-
-## Build Guide
-
-To compile this program, run:
-```bash
-leo build
-```
-
-To run this program, run:
-```bash
-leo run
-```
\ No newline at end of file
diff --git a/examples/import_point/imports/point.leo b/examples/import_point/imports/point.leo
deleted file mode 100644
index a067764fa8..0000000000
--- a/examples/import_point/imports/point.leo
+++ /dev/null
@@ -1,18 +0,0 @@
-// A (x, y) coordinate point.
-circuit Point {
-    x: u32,
-    y: u32,
-}
-
-// Returns a new point.
-function new(x: u32, y: u32) -> Point {
-    return Point {x, y};
-}
-
-// Returns the sum of two points.
-function sum(a: Point, b: Point) -> Point {
-    return Point {
-        x: a.x + b.x,
-        y: a.y + b.y,
-    };
-}
\ No newline at end of file
diff --git a/examples/import_point/inputs/import_point.in b/examples/import_point/inputs/import_point.in
deleted file mode 100644
index 0a4927f66a..0000000000
--- a/examples/import_point/inputs/import_point.in
+++ /dev/null
@@ -1,4 +0,0 @@
-// The program input for import_point/src/main.leo
-[main]
-public a: u32 = 1u32;
-b: u32 = 2u32; // Input variable `b` is private by default.
diff --git a/examples/import_point/program.json b/examples/import_point/program.json
deleted file mode 100644
index 5139a4f6b5..0000000000
--- a/examples/import_point/program.json
+++ /dev/null
@@ -1,10 +0,0 @@
-{
-    "program": "import_point.aleo",
-    "version": "0.0.0",
-    "description": "",
-    "development": {
-        "private_key": "APrivateKey1zkpBvXdKZKaXXcLUnwAVFCQNp41jrX6JqTuJo1JShfPoRfx",
-        "address": "aleo1mgfq6g40l6zkhsm063n3uhr43qk5e0zsua5aszeq5080dsvlcvxsn0rrau"
-    },
-    "license": "MIT"
-}
diff --git a/examples/import_point/src/main.leo b/examples/import_point/src/main.leo
deleted file mode 100644
index f3e306e86e..0000000000
--- a/examples/import_point/src/main.leo
+++ /dev/null
@@ -1,9 +0,0 @@
-// Import the circuits and functions from `imports/point.leo`.
-// In testnet3, Leo imports all symbols from a file by default.
-import point.leo;
-
-// The main function.
-@program
-function main(public a: u32, b: u32) -> u32 {
-    return a + b;
-}
