diff --git a/Cargo.lock b/Cargo.lock
index 6576a9fc29..a7e74e16b8 100644
--- a/Cargo.lock
+++ b/Cargo.lock
@@ -981,9 +981,9 @@ dependencies = [
 
 [[package]]
 name = "fastrand"
-version = "2.1.0"
+version = "2.1.1"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "9fc0510504f03c51ada170672ac806f1f105a88aa97a5281117e1ddc3368e51a"
+checksum = "e8c02a5121d4ea3eb16a80748c74f5549a5665e4c21333c6098f283870fbdea6"
 
 [[package]]
 name = "fiat-crypto"
@@ -2372,9 +2372,9 @@ dependencies = [
 
 [[package]]
 name = "regex"
-version = "1.10.6"
+version = "1.11.0"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "4219d74c6b67a3654a9fbebc4b419e22126d13d2f3c4a07ee0cb61ff79a79619"
+checksum = "38200e5ee88914975b69f657f0801b6f6dccafd44fd9326302a4aaeecfacb1d8"
 dependencies = [
  "aho-corasick",
  "memchr",
@@ -2384,9 +2384,9 @@ dependencies = [
 
 [[package]]
 name = "regex-automata"
-version = "0.4.6"
+version = "0.4.8"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "86b83b8b9847f9bf95ef68afb0b8e6cdb80f498442f5179a29fad448fcc1eaea"
+checksum = "368758f23274712b504848e9d5a6f010445cc8b87a7cdb4d7cbee666c1288da3"
 dependencies = [
  "aho-corasick",
  "memchr",
@@ -2395,9 +2395,9 @@ dependencies = [
 
 [[package]]
 name = "regex-syntax"
-version = "0.8.3"
+version = "0.8.5"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "adad44e29e4c806119491a7f06f03de4d1af22c3a680dd47f1e6e179439d1f56"
+checksum = "2b15c43186be67a4fd63bee50d0303afffcef381492ebe2c5d87f324e1b8815c"
 
 [[package]]
 name = "reqwest"
@@ -2537,9 +2537,9 @@ dependencies = [
 
 [[package]]
 name = "rustix"
-version = "0.38.34"
+version = "0.38.37"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "70dc5ec042f7a43c4a73241207cecc9873a06d45debb38b329f8541d85c2730f"
+checksum = "8acb788b847c24f28525660c4d7758620a7210875711f79e7f663cc152726811"
 dependencies = [
  "bitflags 2.5.0",
  "errno",
@@ -3983,12 +3983,12 @@ dependencies = [
 
 [[package]]
 name = "tempfile"
-version = "3.12.0"
+version = "3.13.0"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "04cbcdd0c794ebb0d4cf35e88edd2f7d2c4c3e9a5a6dab322839b321c6a87a64"
+checksum = "f0f2c9fc62d0beef6951ccffd757e241266a2c833136efbe35af6cd2567dca5b"
 dependencies = [
  "cfg-if",
- "fastrand 2.1.0",
+ "fastrand 2.1.1",
  "once_cell",
  "rustix",
  "windows-sys 0.59.0",
@@ -4026,18 +4026,18 @@ checksum = "5dc41925991e82af3c3e21e25a9aad92e72930af57fbcc4b07867a18d1cd0459"
 
 [[package]]
 name = "thiserror"
-version = "1.0.63"
+version = "1.0.64"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "c0342370b38b6a11b6cc11d6a805569958d54cfa061a29969c3b5ce2ea405724"
+checksum = "d50af8abc119fb8bb6dbabcfa89656f46f84aa0ac7688088608076ad2b459a84"
 dependencies = [
  "thiserror-impl",
 ]
 
 [[package]]
 name = "thiserror-impl"
-version = "1.0.63"
+version = "1.0.64"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "a4558b58466b9ad7ca0f102865eccc95938dca1a74a856f2b57b6629050da261"
+checksum = "08904e7672f5eb876eaaf87e0ce17857500934f4981c4a0ab2b4aa98baac7fc3"
 dependencies = [
  "proc-macro2",
  "quote 1.0.36",
diff --git a/compiler/compiler/Cargo.toml b/compiler/compiler/Cargo.toml
index 15787dfb8f..372044fd91 100644
--- a/compiler/compiler/Cargo.toml
+++ b/compiler/compiler/Cargo.toml
@@ -75,7 +75,7 @@ default-features = false
 workspace = true
 
 [dev-dependencies.regex]
-version = "1.10.6"
+version = "1.11.0"
 
 [dev-dependencies.serde]
 version = "1.0.210"
@@ -85,7 +85,7 @@ features = [ "derive" ]
 version = "0.8.25"
 
 [dev-dependencies.tempfile]
-version = "3.12"
+version = "3.13"
 
 [features]
 default = [ ]
diff --git a/compiler/passes/src/type_checking/checker.rs b/compiler/passes/src/type_checking/checker.rs
index e21180cda4..0ae942d8b7 100644
--- a/compiler/passes/src/type_checking/checker.rs
+++ b/compiler/passes/src/type_checking/checker.rs
@@ -1179,7 +1179,7 @@ impl<'a, N: Network> TypeChecker<'a, N> {
         }
 
         // Type check the function's parameters.
-        function.input.iter().enumerate().for_each(|(_index, input_var)| {
+        function.input.iter().for_each(|input_var| {
             // Check that the type of input parameter is defined.
             self.assert_type_is_valid(input_var.type_(), input_var.span());
             // Check that the type of the input parameter is not a tuple.
@@ -1239,6 +1239,9 @@ impl<'a, N: Network> TypeChecker<'a, N> {
                     self.handler.emit_err(err);
                 }
             }
+
+            // Add the input to the type table.
+            self.type_table.insert(input_var.identifier().id(), input_var.type_().clone());
         });
 
         // Type check the function's return type.
diff --git a/compiler/span/src/span.rs b/compiler/span/src/span.rs
index 1c0cedb168..f3780b1d4d 100644
--- a/compiler/span/src/span.rs
+++ b/compiler/span/src/span.rs
@@ -18,7 +18,7 @@
 
 use core::ops::{Add, Sub};
 use serde::{Deserialize, Serialize};
-use std::{fmt, usize};
+use std::fmt;
 
 use crate::symbol::with_session_globals;
 
diff --git a/docs/grammar/src/main.rs b/docs/grammar/src/main.rs
index 5d13b45ec2..220ba10a0b 100644
--- a/docs/grammar/src/main.rs
+++ b/docs/grammar/src/main.rs
@@ -113,7 +113,7 @@ impl<'a> Processor<'a> {
                     let def = def.trim();
 
                     // try to find rule matching definition or fail
-                    let rule = self.rules.get(&def.to_string()).cloned().unwrap();
+                    let rule = self.rules.get(def).cloned().unwrap();
 
                     self.enter_scope(Scope::Definition(rule));
                 }
diff --git a/errors/Cargo.toml b/errors/Cargo.toml
index 6e8ab471b7..cfd81f4373 100644
--- a/errors/Cargo.toml
+++ b/errors/Cargo.toml
@@ -45,4 +45,4 @@ version = "1.0.210"
 features = [ "derive", "rc" ]
 
 [dependencies.thiserror]
-version = "1.0.63"
+version = "1.0.64"
diff --git a/leo/cli/helpers/logger.rs b/leo/cli/helpers/logger.rs
index 0edafed086..a5ee6f6c13 100644
--- a/leo/cli/helpers/logger.rs
+++ b/leo/cli/helpers/logger.rs
@@ -40,35 +40,6 @@ pub struct Format<F = Full, T = SystemTime> {
 }
 
 impl<F, T> Format<F, T> {
-    /// Use the full JSON format.
-    ///
-    /// The full format includes fields from all entered spans.
-    ///
-    /// # Example Output
-    ///
-    /// ```ignore,json
-    /// {"timestamp":"Feb 20 11:28:15.096","level":"INFO","target":"mycrate","fields":{"message":"some message", "key": "value"}}
-    /// ```
-    ///
-    /// # Options
-    ///
-    /// - [`Format::flatten_event`] can be used to enable flattening event fields into the root
-    /// object.
-    ///
-    /// [`Format::flatten_event`]: #method.flatten_event
-    #[cfg(feature = "json")]
-    pub fn json(self) -> Format<Json, T> {
-        Format {
-            format: Json::default(),
-            timer: self.timer,
-            ansi: self.ansi,
-            display_target: self.display_target,
-            display_level: self.display_level,
-            display_thread_id: self.display_thread_id,
-            display_thread_name: self.display_thread_name,
-        }
-    }
-
     /// Use the given [`timer`] for log message timestamps.
     ///
     /// See [`time`] for the provided timer implementations.
diff --git a/leo/package/src/root/env.rs b/leo/package/src/root/env.rs
index 483f05c1c8..44c8d47131 100644
--- a/leo/package/src/root/env.rs
+++ b/leo/package/src/root/env.rs
@@ -22,7 +22,7 @@ use snarkvm::console::account::PrivateKey;
 use snarkvm::prelude::{MainnetV0, Network, TestnetV0};
 
 use serde::Deserialize;
-use std::{borrow::Cow, fs::File, io::Write, path::Path};
+use std::{borrow::Cow, fmt, fs::File, io::Write, path::Path};
 
 pub static ENV_FILENAME: &str = ".env";
 
@@ -67,15 +67,15 @@ impl<N: Network> Env<N> {
     }
 }
 
-impl<N: Network> ToString for Env<N> {
-    fn to_string(&self) -> String {
+impl<N: Network> fmt::Display for Env<N> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
         // Get the network name.
         let network = match N::ID {
             MainnetV0::ID => NetworkName::MainnetV0,
             TestnetV0::ID => NetworkName::TestnetV0,
             _ => unimplemented!("Unsupported network"),
         };
-        // Return the formatted string.
-        format!("NETWORK={network}\nPRIVATE_KEY={}\nENDPOINT={}\n", self.private_key, self.endpoint)
+        // Write the formatted string.
+        write!(f, "NETWORK={network}\nPRIVATE_KEY={}\nENDPOINT={}\n", self.private_key, self.endpoint)
     }
 }
diff --git a/tests/expectations/compiler/function/function_input_else_assignment.out b/tests/expectations/compiler/function/function_input_else_assignment.out
new file mode 100644
index 0000000000..a34c7a3577
--- /dev/null
+++ b/tests/expectations/compiler/function/function_input_else_assignment.out
@@ -0,0 +1,18 @@
+---
+namespace: Compile
+expectation: Pass
+outputs:
+  - - compile:
+        - initial_symbol_table: 3a64910c315e257b67757ebc3478a2df7ff7aa2ae8ef8d258507e44b53942b82
+          type_checked_symbol_table: 69a0008fc51152856295a0ef75043069f34b5f0e1ce5b944e4d575a2d184a233
+          unrolled_symbol_table: 69a0008fc51152856295a0ef75043069f34b5f0e1ce5b944e4d575a2d184a233
+          initial_ast: 4c9ab22f340920e9a086b4b55e1e09065bf8839d7bf03d725003fc53cc6d5265
+          unrolled_ast: 4c9ab22f340920e9a086b4b55e1e09065bf8839d7bf03d725003fc53cc6d5265
+          ssa_ast: bf04c3808bc97d0b45f1e69cb8f9faed5f31e01cc03e359364bf22132b36216b
+          flattened_ast: cde3b5d0a6e6744e69645072cc44a21dd6a4ef77c087e4a0afab075abd3b7a70
+          destructured_ast: f10b829b1ac2863d3bc8cbf86b51c8947b5c1db6a0ebcb714d4c9f8f2dedb274
+          inlined_ast: f10b829b1ac2863d3bc8cbf86b51c8947b5c1db6a0ebcb714d4c9f8f2dedb274
+          dce_ast: 6fb7a52d2eeff93d53a26ddc38a0bdf71615b6aa7ab3ab22a894e96bd1d309f1
+          bytecode: e78a049b9529f1feeaa63bcf231b86c28e604487d185d98bbc0e4a2f201c3ec0
+          errors: ""
+          warnings: ""
diff --git a/tests/test-framework/Cargo.toml b/tests/test-framework/Cargo.toml
index d53d964bba..3db9c6455d 100644
--- a/tests/test-framework/Cargo.toml
+++ b/tests/test-framework/Cargo.toml
@@ -51,7 +51,7 @@ version = "0.8"
 version = "2.5"
 
 [dependencies.regex]
-version = "1.10"
+version = "1.11"
 
 [dev-dependencies.leo-compiler]
 path = "../../compiler/compiler"
diff --git a/tests/test-framework/src/fetch.rs b/tests/test-framework/src/fetch.rs
index 0df7c27076..71316ace54 100644
--- a/tests/test-framework/src/fetch.rs
+++ b/tests/test-framework/src/fetch.rs
@@ -28,10 +28,7 @@ pub fn find_tests(path: &Path) -> impl Iterator<Item = (PathBuf, String)> {
         // Check if the file is a .leo file.
         let is_leo_file = path.extension().filter(|s| *s == "leo").is_some();
         // Read the test filter from the environment.
-        let filter = match std::env::var("TEST_FILTER") {
-            Ok(filter) => filter,
-            Err(_) => String::new(),
-        };
+        let filter = std::env::var("TEST_FILTER").unwrap_or_default();
         // Check if the path contains the filter.
         let satisfies_filter = filter.is_empty() || path.to_string_lossy().contains(&filter);
         // If the file is a .leo file and satisfies the filter, return the path and the file contents.
diff --git a/tests/tests/compiler/function/function_input_else_assignment.leo b/tests/tests/compiler/function/function_input_else_assignment.leo
new file mode 100644
index 0000000000..b6ef0d847a
--- /dev/null
+++ b/tests/tests/compiler/function/function_input_else_assignment.leo
@@ -0,0 +1,13 @@
+/*
+namespace: Compile
+expectation: Pass
+*/
+
+program test.aleo {
+    transition foo(a: u8) {
+        if false {
+        } else {
+            a = 0u8;
+        }
+    }
+}
diff --git a/utils/retriever/Cargo.toml b/utils/retriever/Cargo.toml
index 6a860aee84..e7189c26aa 100644
--- a/utils/retriever/Cargo.toml
+++ b/utils/retriever/Cargo.toml
@@ -63,7 +63,7 @@ version = "3.1.1"
 version = "0.10.8"
 
 [dependencies.tempfile]
-version = "3.12.0"
+version = "3.13.0"
 
 [dependencies.toml]
 version = "0.8.19"
