diff --git a/.resources/release-version b/.resources/release-version
index 74a63a2249..01558e339a 100644
--- a/.resources/release-version
+++ b/.resources/release-version
@@ -1 +1 @@
-v1.5.2
\ No newline at end of file
+v1.5.3
\ No newline at end of file
diff --git a/Cargo.lock b/Cargo.lock
index 9ec1da59f0..1fc8200b29 100644
--- a/Cargo.lock
+++ b/Cargo.lock
@@ -1159,7 +1159,7 @@ checksum = "830d08ce1d1d941e6b30645f1a0eb5643013d835ce3779a5fc208261dbe10f55"
 
 [[package]]
 name = "leo-abnf"
-version = "1.5.2"
+version = "1.5.3"
 dependencies = [
  "abnf",
  "anyhow",
@@ -1167,7 +1167,7 @@ dependencies = [
 
 [[package]]
 name = "leo-asg"
-version = "1.5.2"
+version = "1.5.3"
 dependencies = [
  "criterion",
  "indexmap",
@@ -1183,14 +1183,14 @@ dependencies = [
 
 [[package]]
 name = "leo-asg-passes"
-version = "1.5.2"
+version = "1.5.3"
 dependencies = [
  "leo-asg",
 ]
 
 [[package]]
 name = "leo-ast"
-version = "1.5.2"
+version = "1.5.3"
 dependencies = [
  "anyhow",
  "criterion",
@@ -1205,7 +1205,7 @@ dependencies = [
 
 [[package]]
 name = "leo-compiler"
-version = "1.5.2"
+version = "1.5.3"
 dependencies = [
  "bincode",
  "hex",
@@ -1243,7 +1243,7 @@ dependencies = [
 
 [[package]]
 name = "leo-imports"
-version = "1.5.2"
+version = "1.5.3"
 dependencies = [
  "indexmap",
  "leo-asg",
@@ -1255,7 +1255,7 @@ dependencies = [
 
 [[package]]
 name = "leo-input"
-version = "1.5.2"
+version = "1.5.3"
 dependencies = [
  "from-pest",
  "pest",
@@ -1267,7 +1267,7 @@ dependencies = [
 
 [[package]]
 name = "leo-lang"
-version = "1.5.2"
+version = "1.5.3"
 dependencies = [
  "ansi_term 0.12.1",
  "anyhow",
@@ -1310,11 +1310,11 @@ dependencies = [
 
 [[package]]
 name = "leo-linter"
-version = "1.5.2"
+version = "1.5.3"
 
 [[package]]
 name = "leo-package"
-version = "1.5.2"
+version = "1.5.3"
 dependencies = [
  "lazy_static",
  "serde",
@@ -1327,7 +1327,7 @@ dependencies = [
 
 [[package]]
 name = "leo-parser"
-version = "1.5.2"
+version = "1.5.3"
 dependencies = [
  "criterion",
  "indexmap",
@@ -1344,7 +1344,7 @@ dependencies = [
 
 [[package]]
 name = "leo-state"
-version = "1.5.2"
+version = "1.5.3"
 dependencies = [
  "indexmap",
  "leo-ast",
@@ -1361,7 +1361,7 @@ dependencies = [
 
 [[package]]
 name = "leo-synthesizer"
-version = "1.5.2"
+version = "1.5.3"
 dependencies = [
  "hex",
  "num-bigint",
@@ -1376,7 +1376,7 @@ dependencies = [
 
 [[package]]
 name = "leo-test-framework"
-version = "1.5.2"
+version = "1.5.3"
 dependencies = [
  "serde",
  "serde_json",
diff --git a/Cargo.toml b/Cargo.toml
index 7a70d2a876..e7187b0020 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "leo-lang"
-version = "1.5.2"
+version = "1.5.3"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "The Leo programming language"
 homepage = "https://aleo.org"
@@ -43,35 +43,35 @@ members = [
 
 [dependencies.leo-ast]
 path = "./ast"
-version = "1.5.2"
+version = "1.5.3"
 
 [dependencies.leo-compiler]
 path = "./compiler"
-version = "1.5.2"
+version = "1.5.3"
 
 [dependencies.leo-imports]
 path = "./imports"
-version = "1.5.2"
+version = "1.5.3"
 
 [dependencies.leo-input]
 path = "./input"
-version = "1.5.2"
+version = "1.5.3"
 
 [dependencies.leo-package]
 path = "./package"
-version = "1.5.2"
+version = "1.5.3"
 
 [dependencies.leo-parser]
 path = "./parser"
-version = "1.5.2"
+version = "1.5.3"
 
 [dependencies.leo-state]
 path = "./state"
-version = "1.5.2"
+version = "1.5.3"
 
 [dependencies.leo-synthesizer]
 path = "./synthesizer"
-version = "1.5.2"
+version = "1.5.3"
 
 [dependencies.snarkvm-algorithms]
 version = "0.7.4"
diff --git a/asg-passes/Cargo.toml b/asg-passes/Cargo.toml
index 83b0a0a066..6b5ca6d4f6 100644
--- a/asg-passes/Cargo.toml
+++ b/asg-passes/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "leo-asg-passes"
-version = "1.5.2"
+version = "1.5.3"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "The Leo programming language"
 homepage = "https://aleo.org"
@@ -22,4 +22,4 @@ path = "src/lib.rs"
 
 [dependencies.leo-asg]
 path = "../asg"
-version = "1.5.2"
+version = "1.5.3"
diff --git a/asg/Cargo.toml b/asg/Cargo.toml
index cd95da9596..f41ef3b108 100644
--- a/asg/Cargo.toml
+++ b/asg/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "leo-asg"
-version = "1.5.2"
+version = "1.5.3"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "ASG of the Leo programming language"
 homepage = "https://aleo.org"
@@ -30,11 +30,11 @@ version = "1.7"
 version = "1.0"
 
 [dependencies.leo-ast]
-version = "1.5.2"
+version = "1.5.3"
 path = "../ast"
 
 [dependencies.leo-parser]
-version = "1.5.2"
+version = "1.5.3"
 path = "../parser"
 
 [dependencies.num-bigint]
diff --git a/asg/src/reducer/monoidal_director.rs b/asg/src/reducer/monoidal_director.rs
index ce34755a11..161183a19b 100644
--- a/asg/src/reducer/monoidal_director.rs
+++ b/asg/src/reducer/monoidal_director.rs
@@ -238,9 +238,7 @@ impl<'a, T: Monoid, R: MonoidalReducerStatement<'a, T>> MonoidalDirector<'a, T,
     pub fn reduce_console(&mut self, input: &ConsoleStatement<'a>) -> T {
         let argument = match &input.function {
             ConsoleFunction::Assert(e) => self.reduce_expression(e.get()),
-            ConsoleFunction::Debug(f) | ConsoleFunction::Error(f) | ConsoleFunction::Log(f) => {
-                self.reduce_formatted_string(f)
-            }
+            ConsoleFunction::Error(f) | ConsoleFunction::Log(f) => self.reduce_formatted_string(f),
         };
 
         self.reducer.reduce_console(input, argument)
diff --git a/asg/src/reducer/reconstructing_director.rs b/asg/src/reducer/reconstructing_director.rs
index 75127d477a..bd2900d3e8 100644
--- a/asg/src/reducer/reconstructing_director.rs
+++ b/asg/src/reducer/reconstructing_director.rs
@@ -259,7 +259,7 @@ impl<'a, R: ReconstructingReducerStatement<'a>> ReconstructingDirector<'a, R> {
                 let argument = self.reduce_expression(argument.get());
                 self.reducer.reduce_console_assert(input, argument)
             }
-            ConsoleFunction::Debug(f) | ConsoleFunction::Error(f) | ConsoleFunction::Log(f) => {
+            ConsoleFunction::Error(f) | ConsoleFunction::Log(f) => {
                 let formatted = self.reduce_formatted_string(f.clone());
                 self.reducer.reduce_console_log(input, formatted)
             }
diff --git a/asg/src/reducer/reconstructing_reducer.rs b/asg/src/reducer/reconstructing_reducer.rs
index 755c2eed1c..1a17b886c0 100644
--- a/asg/src/reducer/reconstructing_reducer.rs
+++ b/asg/src/reducer/reconstructing_reducer.rs
@@ -300,7 +300,6 @@ pub trait ReconstructingReducerStatement<'a>: ReconstructingReducerExpression<'a
             span: input.span,
             function: match input.function {
                 ConsoleFunction::Assert(_) => unimplemented!(),
-                ConsoleFunction::Debug(_) => ConsoleFunction::Debug(argument),
                 ConsoleFunction::Error(_) => ConsoleFunction::Error(argument),
                 ConsoleFunction::Log(_) => ConsoleFunction::Log(argument),
             },
diff --git a/asg/src/reducer/visitor_director.rs b/asg/src/reducer/visitor_director.rs
index dc6dae6929..2d238e67aa 100644
--- a/asg/src/reducer/visitor_director.rs
+++ b/asg/src/reducer/visitor_director.rs
@@ -336,9 +336,7 @@ impl<'a, R: StatementVisitor<'a>> VisitorDirector<'a, R> {
             VisitResult::VisitChildren => {
                 match &input.function {
                     ConsoleFunction::Assert(e) => self.visit_expression(e)?,
-                    ConsoleFunction::Debug(f) | ConsoleFunction::Error(f) | ConsoleFunction::Log(f) => {
-                        self.visit_formatted_string(f)?
-                    }
+                    ConsoleFunction::Error(f) | ConsoleFunction::Log(f) => self.visit_formatted_string(f)?,
                 }
                 Ok(())
             }
diff --git a/asg/src/statement/console.rs b/asg/src/statement/console.rs
index 37359f90b3..5708576425 100644
--- a/asg/src/statement/console.rs
+++ b/asg/src/statement/console.rs
@@ -30,7 +30,6 @@ pub struct ConsoleArgs<'a> {
 #[derive(Clone)]
 pub enum ConsoleFunction<'a> {
     Assert(Cell<&'a Expression<'a>>),
-    Debug(ConsoleArgs<'a>),
     Error(ConsoleArgs<'a>),
     Log(ConsoleArgs<'a>),
 }
@@ -89,7 +88,6 @@ impl<'a> FromAst<'a, leo_ast::ConsoleStatement> for ConsoleStatement<'a> {
                 AstConsoleFunction::Assert(expression) => ConsoleFunction::Assert(Cell::new(
                     <&Expression<'a>>::from_ast(scope, expression, Some(Type::Boolean.into()))?,
                 )),
-                AstConsoleFunction::Debug(args) => ConsoleFunction::Debug(ConsoleArgs::from_ast(scope, args, None)?),
                 AstConsoleFunction::Error(args) => ConsoleFunction::Error(ConsoleArgs::from_ast(scope, args, None)?),
                 AstConsoleFunction::Log(args) => ConsoleFunction::Log(ConsoleArgs::from_ast(scope, args, None)?),
             },
@@ -103,7 +101,6 @@ impl<'a> Into<leo_ast::ConsoleStatement> for &ConsoleStatement<'a> {
         leo_ast::ConsoleStatement {
             function: match &self.function {
                 Assert(e) => AstConsoleFunction::Assert(e.get().into()),
-                Debug(args) => AstConsoleFunction::Debug(args.into()),
                 Error(args) => AstConsoleFunction::Error(args.into()),
                 Log(args) => AstConsoleFunction::Log(args.into()),
             },
diff --git a/asg/tests/pass/console/debug.leo b/asg/tests/pass/console/debug.leo
deleted file mode 100644
index 54e6eb539f..0000000000
--- a/asg/tests/pass/console/debug.leo
+++ /dev/null
@@ -1,3 +0,0 @@
-function main() {
-    console.debug("hello debug");
-}
\ No newline at end of file
diff --git a/asg/tests/pass/console/mod.rs b/asg/tests/pass/console/mod.rs
index 4f1a243012..43d3c0e0cb 100644
--- a/asg/tests/pass/console/mod.rs
+++ b/asg/tests/pass/console/mod.rs
@@ -40,14 +40,6 @@ fn test_log_input() {
     load_asg(program_string).unwrap();
 }
 
-// Debug
-
-#[test]
-fn test_debug() {
-    let program_string = include_str!("debug.leo");
-    load_asg(program_string).unwrap();
-}
-
 // Error
 
 #[test]
diff --git a/ast/Cargo.toml b/ast/Cargo.toml
index 43092b85f3..c33cd7f912 100644
--- a/ast/Cargo.toml
+++ b/ast/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "leo-ast"
-version = "1.5.2"
+version = "1.5.3"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Core AST of the Leo programming language"
 homepage = "https://aleo.org"
@@ -19,7 +19,7 @@ edition = "2018"
 
 [dependencies.leo-input]
 path = "../input"
-version = "1.5.2"
+version = "1.5.3"
 
 [dependencies.indexmap]
 version = "1.7.0"
diff --git a/ast/src/errors/error.rs b/ast/src/errors/error.rs
index 0cf34f5d25..b2c1ebff5a 100644
--- a/ast/src/errors/error.rs
+++ b/ast/src/errors/error.rs
@@ -15,7 +15,6 @@
 // along with the Leo library. If not, see <https://www.gnu.org/licenses/>.
 
 use crate::{LeoError, Span};
-
 use std::{fmt, sync::Arc};
 
 pub const INDENT: &str = "    ";
@@ -77,20 +76,21 @@ impl fmt::Display for FormattedError {
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
         let underline = underline(self.col_start, self.col_stop);
 
-        write!(
-            f,
+        let error_message = format!(
             "{indent     }--> {path}:{line_start}:{start}\n\
-             {indent     } |\n",
+             {indent     } ",
             indent = INDENT,
             path = &*self.path,
             line_start = self.line_start,
             start = self.col_start,
-        )?;
+        );
+
+        write!(f, "{}", error_message)?;
 
         for (line_no, line) in self.content.lines().enumerate() {
             writeln!(
                 f,
-                "{line_no:width$} | {text}",
+                "|\n{line_no:width$} | {text}",
                 width = INDENT.len(),
                 line_no = self.line_start + line_no,
                 text = line,
diff --git a/ast/src/reducer/canonicalization.rs b/ast/src/reducer/canonicalization.rs
index 4c78012c30..edf300ada4 100644
--- a/ast/src/reducer/canonicalization.rs
+++ b/ast/src/reducer/canonicalization.rs
@@ -392,7 +392,7 @@ impl Canonicalizer {
                     ConsoleFunction::Assert(expression) => {
                         ConsoleFunction::Assert(self.canonicalize_expression(expression))
                     }
-                    ConsoleFunction::Debug(args) | ConsoleFunction::Error(args) | ConsoleFunction::Log(args) => {
+                    ConsoleFunction::Error(args) | ConsoleFunction::Log(args) => {
                         let parameters = args
                             .parameters
                             .iter()
@@ -406,7 +406,6 @@ impl Canonicalizer {
                         };
 
                         match &console_function_call.function {
-                            ConsoleFunction::Debug(_) => ConsoleFunction::Debug(console_args),
                             ConsoleFunction::Error(_) => ConsoleFunction::Error(console_args),
                             ConsoleFunction::Log(_) => ConsoleFunction::Log(console_args),
                             _ => unimplemented!(), // impossible
diff --git a/ast/src/reducer/reconstructing_director.rs b/ast/src/reducer/reconstructing_director.rs
index 80a61e0476..cf5a409d06 100644
--- a/ast/src/reducer/reconstructing_director.rs
+++ b/ast/src/reducer/reconstructing_director.rs
@@ -390,7 +390,7 @@ impl<R: ReconstructingReducer> ReconstructingDirector<R> {
     ) -> Result<ConsoleStatement, ReducerError> {
         let function = match &console_function_call.function {
             ConsoleFunction::Assert(expression) => ConsoleFunction::Assert(self.reduce_expression(expression)?),
-            ConsoleFunction::Debug(args) | ConsoleFunction::Error(args) | ConsoleFunction::Log(args) => {
+            ConsoleFunction::Error(args) | ConsoleFunction::Log(args) => {
                 let mut parameters = vec![];
                 for parameter in args.parameters.iter() {
                     parameters.push(self.reduce_expression(parameter)?);
@@ -403,7 +403,6 @@ impl<R: ReconstructingReducer> ReconstructingDirector<R> {
                 };
 
                 match &console_function_call.function {
-                    ConsoleFunction::Debug(_) => ConsoleFunction::Debug(formatted),
                     ConsoleFunction::Error(_) => ConsoleFunction::Error(formatted),
                     ConsoleFunction::Log(_) => ConsoleFunction::Log(formatted),
                     _ => return Err(ReducerError::impossible_console_assert_call(&args.span)),
diff --git a/ast/src/statements/console/console_function.rs b/ast/src/statements/console/console_function.rs
index 5919980af6..8c782a1e88 100644
--- a/ast/src/statements/console/console_function.rs
+++ b/ast/src/statements/console/console_function.rs
@@ -22,7 +22,6 @@ use std::fmt;
 #[derive(Clone, PartialEq, Eq, Serialize, Deserialize, Debug)]
 pub enum ConsoleFunction {
     Assert(Expression),
-    Debug(ConsoleArgs),
     Error(ConsoleArgs),
     Log(ConsoleArgs),
 }
@@ -31,7 +30,6 @@ impl fmt::Display for ConsoleFunction {
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
         match self {
             ConsoleFunction::Assert(assert) => write!(f, "assert({})", assert),
-            ConsoleFunction::Debug(debug) => write!(f, "debug({})", debug),
             ConsoleFunction::Error(error) => write!(f, "error{})", error),
             ConsoleFunction::Log(log) => write!(f, "log({})", log),
         }
@@ -42,18 +40,14 @@ impl Node for ConsoleFunction {
     fn span(&self) -> &Span {
         match self {
             ConsoleFunction::Assert(assert) => assert.span(),
-            ConsoleFunction::Debug(formatted) | ConsoleFunction::Error(formatted) | ConsoleFunction::Log(formatted) => {
-                &formatted.span
-            }
+            ConsoleFunction::Error(formatted) | ConsoleFunction::Log(formatted) => &formatted.span,
         }
     }
 
     fn set_span(&mut self, span: Span) {
         match self {
             ConsoleFunction::Assert(assert) => assert.set_span(span),
-            ConsoleFunction::Debug(formatted) | ConsoleFunction::Error(formatted) | ConsoleFunction::Log(formatted) => {
-                formatted.set_span(span)
-            }
+            ConsoleFunction::Error(formatted) | ConsoleFunction::Log(formatted) => formatted.set_span(span),
         }
     }
 }
diff --git a/compiler/Cargo.toml b/compiler/Cargo.toml
index 07c7508152..5725a5f453 100644
--- a/compiler/Cargo.toml
+++ b/compiler/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "leo-compiler"
-version = "1.5.2"
+version = "1.5.3"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Compiler of the Leo programming language"
 homepage = "https://aleo.org"
@@ -19,39 +19,39 @@ edition = "2018"
 
 [dependencies.leo-ast]
 path = "../ast"
-version = "1.5.2"
+version = "1.5.3"
 
 [dependencies.leo-imports]
 path = "../imports"
-version = "1.5.2"
+version = "1.5.3"
 
 [dependencies.leo-input]
 path = "../input"
-version = "1.5.2"
+version = "1.5.3"
 
 [dependencies.leo-package]
 path = "../package"
-version = "1.5.2"
+version = "1.5.3"
 
 [dependencies.leo-state]
 path = "../state"
-version = "1.5.2"
+version = "1.5.3"
 
 [dependencies.leo-asg]
 path = "../asg"
-version = "1.5.2"
+version = "1.5.3"
 
 [dependencies.leo-parser]
 path = "../parser"
-version = "1.5.2"
+version = "1.5.3"
 
 [dependencies.leo-asg-passes]
 path = "../asg-passes"
-version = "1.5.2"
+version = "1.5.3"
 
 [dependencies.leo-synthesizer]
 path = "../synthesizer"
-version = "1.5.2"
+version = "1.5.3"
 
 [dependencies.tendril]
 version = "0.4"
@@ -71,7 +71,7 @@ default-features = false
 [dependencies.snarkvm-gadgets]
 version = "0.7.5"
 default-features = false
-features = ["curves"]
+features = [ "curves" ]
 
 [dependencies.snarkvm-r1cs]
 version = "0.7.5"
diff --git a/compiler/src/console/console.rs b/compiler/src/console/console.rs
index e3aa490942..46f3dee52a 100644
--- a/compiler/src/console/console.rs
+++ b/compiler/src/console/console.rs
@@ -39,13 +39,6 @@ impl<'a, F: PrimeField, G: GroupType<F>> ConstrainedProgram<'a, F, G> {
                     &console.span.clone().unwrap_or_default(),
                 )?;
             }
-            ConsoleFunction::Debug(string) => {
-                let string = self.format(cs, string)?;
-
-                if get_indicator_value(indicator) {
-                    tracing::debug!("{}", string);
-                }
-            }
             ConsoleFunction::Error(string) => {
                 let string = self.format(cs, string)?;
 
diff --git a/compiler/src/phases/reducing_director.rs b/compiler/src/phases/reducing_director.rs
index ff432e3f27..2a934909a8 100644
--- a/compiler/src/phases/reducing_director.rs
+++ b/compiler/src/phases/reducing_director.rs
@@ -597,8 +597,7 @@ impl<R: ReconstructingReducer, O: CombinerOptions> CombineAstAsgDirector<R, O> {
             (AstConsoleFunction::Assert(ast_expression), AsgConsoleFunction::Assert(asg_expression)) => {
                 AstConsoleFunction::Assert(self.reduce_expression(&ast_expression, asg_expression.get())?)
             }
-            (AstConsoleFunction::Debug(ast_console_args), AsgConsoleFunction::Debug(asg_format))
-            | (AstConsoleFunction::Error(ast_console_args), AsgConsoleFunction::Error(asg_format))
+            (AstConsoleFunction::Error(ast_console_args), AsgConsoleFunction::Error(asg_format))
             | (AstConsoleFunction::Log(ast_console_args), AsgConsoleFunction::Log(asg_format)) => {
                 let mut parameters = vec![];
                 for (ast_parameter, asg_parameter) in
@@ -614,7 +613,6 @@ impl<R: ReconstructingReducer, O: CombinerOptions> CombineAstAsgDirector<R, O> {
                 };
 
                 match &ast.function {
-                    AstConsoleFunction::Debug(_) => AstConsoleFunction::Debug(args),
                     AstConsoleFunction::Error(_) => AstConsoleFunction::Error(args),
                     AstConsoleFunction::Log(_) => AstConsoleFunction::Log(args),
                     _ => return Err(ReducerError::impossible_console_assert_call(&ast_console_args.span)),
diff --git a/compiler/src/test.rs b/compiler/src/test.rs
index 5d7d7eb2a6..62c17db6f1 100644
--- a/compiler/src/test.rs
+++ b/compiler/src/test.rs
@@ -231,7 +231,9 @@ impl Namespace for CompileNamespace {
                 .unwrap_or_else(|_| "Error converting ast to string.".to_string()),
         );
 
-        std::fs::remove_dir_all(std::path::Path::new("/tmp/output")).expect("Error failed to clean up output dir.");
+        if std::fs::read_dir("/tmp/output").is_ok() {
+            std::fs::remove_dir_all(std::path::Path::new("/tmp/output")).expect("Error failed to clean up output dir.");
+        }
 
         let final_output = CompileOutput {
             circuit: last_circuit.unwrap(),
diff --git a/compiler/src/value/address/address.rs b/compiler/src/value/address/address.rs
index f706448c4d..a9fd35a42a 100644
--- a/compiler/src/value/address/address.rs
+++ b/compiler/src/value/address/address.rs
@@ -85,33 +85,57 @@ impl Address {
         Ok(ConstrainedValue::Address(address))
     }
 
-    pub(crate) fn alloc_helper<Fn: FnOnce() -> Result<T, SynthesisError>, T: Borrow<String>>(
+    pub(crate) fn alloc_helper<
+        F: PrimeField,
+        CS: ConstraintSystem<F>,
+        Fn: FnOnce() -> Result<T, SynthesisError>,
+        T: Borrow<String>,
+    >(
+        cs: CS,
         value_gen: Fn,
     ) -> Result<AleoAddress<Components>, SynthesisError> {
-        let address_string = match value_gen() {
-            Ok(value) => {
-                let string_value = value.borrow().clone();
-                Ok(string_value)
-            }
-            _ => Err(SynthesisError::AssignmentMissing),
-        }?;
-
-        AleoAddress::from_str(&address_string).map_err(|_| SynthesisError::AssignmentMissing)
+        if cs.is_in_setup_mode() {
+            Ok(AleoAddress::<Components>::default())
+        } else {
+            let address_string = value_gen()?.borrow().clone();
+            AleoAddress::from_str(&address_string).map_err(|_| SynthesisError::AssignmentMissing)
+        }
     }
 }
 
 impl<F: PrimeField> AllocGadget<String, F> for Address {
+    fn alloc_constant<Fn: FnOnce() -> Result<T, SynthesisError>, T: Borrow<String>, CS: ConstraintSystem<F>>(
+        _cs: CS,
+        value_gen: Fn,
+    ) -> Result<Self, SynthesisError> {
+        let address = {
+            let address_string = value_gen()?.borrow().clone();
+            AleoAddress::from_str(&address_string).map_err(|_| SynthesisError::AssignmentMissing)?
+        };
+        let mut address_bytes = vec![];
+        address
+            .write_le(&mut address_bytes)
+            .map_err(|_| SynthesisError::AssignmentMissing)?;
+
+        let bytes = UInt8::constant_vec(&address_bytes[..]);
+
+        Ok(Address {
+            address: Some(address),
+            bytes,
+        })
+    }
+
     fn alloc<Fn: FnOnce() -> Result<T, SynthesisError>, T: Borrow<String>, CS: ConstraintSystem<F>>(
-        cs: CS,
+        mut cs: CS,
         value_gen: Fn,
     ) -> Result<Self, SynthesisError> {
-        let address = Self::alloc_helper(value_gen)?;
+        let address = Self::alloc_helper(cs.ns(|| "allocate the address"), value_gen)?;
         let mut address_bytes = vec![];
         address
             .write_le(&mut address_bytes)
             .map_err(|_| SynthesisError::AssignmentMissing)?;
 
-        let bytes = UInt8::alloc_vec(cs, &address_bytes[..])?;
+        let bytes = UInt8::alloc_vec(cs.ns(|| "allocate the address bytes"), &address_bytes[..])?;
 
         Ok(Address {
             address: Some(address),
@@ -120,16 +144,16 @@ impl<F: PrimeField> AllocGadget<String, F> for Address {
     }
 
     fn alloc_input<Fn: FnOnce() -> Result<T, SynthesisError>, T: Borrow<String>, CS: ConstraintSystem<F>>(
-        cs: CS,
+        mut cs: CS,
         value_gen: Fn,
     ) -> Result<Self, SynthesisError> {
-        let address = Self::alloc_helper(value_gen)?;
+        let address = Self::alloc_helper(cs.ns(|| "allocate the address"), value_gen)?;
         let mut address_bytes = vec![];
         address
             .write_le(&mut address_bytes)
             .map_err(|_| SynthesisError::AssignmentMissing)?;
 
-        let bytes = UInt8::alloc_input_vec_le(cs, &address_bytes[..])?;
+        let bytes = UInt8::alloc_input_vec_le(cs.ns(|| "allocate the address bytes"), &address_bytes[..])?;
 
         Ok(Address {
             address: Some(address),
diff --git a/grammar/Cargo.toml b/grammar/Cargo.toml
index d9fc3e5b9c..1e7248fc79 100644
--- a/grammar/Cargo.toml
+++ b/grammar/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "leo-abnf"
-version = "1.5.2"
+version = "1.5.3"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "ABNF to Markdown converter for the Leo programming language"
 homepage = "https://aleo.org"
diff --git a/imports/Cargo.toml b/imports/Cargo.toml
index 2770edf50b..5e9ede2cb7 100644
--- a/imports/Cargo.toml
+++ b/imports/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "leo-imports"
-version = "1.5.2"
+version = "1.5.3"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Import parser for Leo program package dependencies"
 homepage = "https://aleo.org"
@@ -19,15 +19,15 @@ edition = "2018"
 
 [dependencies.leo-ast]
 path = "../ast"
-version = "1.5.2"
+version = "1.5.3"
 
 [dependencies.leo-asg]
 path = "../asg"
-version = "1.5.2"
+version = "1.5.3"
 
 [dependencies.leo-parser]
 path = "../parser"
-version = "1.5.2"
+version = "1.5.3"
 
 [dependencies.indexmap]
 version = "1.7.0"
diff --git a/input/Cargo.toml b/input/Cargo.toml
index 3c98793c16..448b9028e4 100644
--- a/input/Cargo.toml
+++ b/input/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "leo-input"
-version = "1.5.2"
+version = "1.5.3"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Input parser of the Leo programming language"
 homepage = "https://aleo.org"
diff --git a/leo/logger.rs b/leo/logger.rs
index b2f639ef01..f7f145a4c2 100644
--- a/leo/logger.rs
+++ b/leo/logger.rs
@@ -218,6 +218,11 @@ pub fn init_logger(_app_name: &'static str, verbosity: usize) -> Result<()> {
         Err(_) => return Err(anyhow::anyhow!("Error: Failed to enable ansi_support")),
     };
 
+    use tracing_subscriber::fmt::writer::MakeWriterExt;
+
+    let stderr = std::io::stderr.with_max_level(tracing::Level::WARN);
+    let mk_writer = stderr.or_else(std::io::stdout);
+
     let subscriber = FmtSubscriber::builder()
         // all spans/events with a level higher than TRACE (e.g, debug, info, warn, etc.)
         // will be written to stdout.
@@ -227,6 +232,7 @@ pub fn init_logger(_app_name: &'static str, verbosity: usize) -> Result<()> {
             2 => tracing::Level::DEBUG,
             _ => tracing::Level::TRACE
         })
+        .with_writer(mk_writer)
         .without_time()
         .with_target(false)
         .event_format(Format::default())
diff --git a/leo/main.rs b/leo/main.rs
index d20347160f..5b633f1634 100644
--- a/leo/main.rs
+++ b/leo/main.rs
@@ -39,6 +39,7 @@ use commands::{
 };
 
 use anyhow::Result;
+use colored::Colorize;
 use std::{path::PathBuf, process::exit};
 use structopt::{clap::AppSettings, StructOpt};
 
@@ -229,7 +230,20 @@ fn handle_error<T>(res: Result<T>) -> T {
     match res {
         Ok(t) => t,
         Err(err) => {
-            eprintln!("Error: {}", err);
+            eprintln!(
+                "{} {}",
+                "Error:".bold().red(),
+                err.to_string()
+                    .lines()
+                    .enumerate()
+                    .map(|(i, l)| if i == 0 {
+                        l.bold().red().to_string()
+                    } else {
+                        l.to_string()
+                    })
+                    .collect::<Vec<_>>()
+                    .join("\n")
+            );
             exit(1);
         }
     }
diff --git a/linter/Cargo.toml b/linter/Cargo.toml
index f01bd3ffbc..c69dba5d89 100644
--- a/linter/Cargo.toml
+++ b/linter/Cargo.toml
@@ -2,7 +2,7 @@ dependencies = { }
 
 [package]
 name = "leo-linter"
-version = "1.5.2"
+version = "1.5.3"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Linter of the Leo programming language"
 homepage = "https://aleo.org"
diff --git a/package/Cargo.toml b/package/Cargo.toml
index 15778e7ea7..807f015829 100644
--- a/package/Cargo.toml
+++ b/package/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "leo-package"
-version = "1.5.2"
+version = "1.5.3"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Package parser of the Leo programming language"
 homepage = "https://aleo.org"
diff --git a/parser/Cargo.toml b/parser/Cargo.toml
index 55c6125d5d..10529e693a 100644
--- a/parser/Cargo.toml
+++ b/parser/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "leo-parser"
-version = "1.5.2"
+version = "1.5.3"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "AST generated by pest from the Leo grammar rules"
 homepage = "https://aleo.org"
@@ -24,7 +24,7 @@ harness = false
 
 [dependencies.leo-ast]
 path = "../ast"
-version = "1.5.2"
+version = "1.5.3"
 
 [dependencies.lazy_static]
 version = "1.3.0"
diff --git a/parser/src/parser/statement.rs b/parser/src/parser/statement.rs
index d60d3e12bb..1b82145f43 100644
--- a/parser/src/parser/statement.rs
+++ b/parser/src/parser/statement.rs
@@ -277,13 +277,12 @@ impl ParserContext {
                 let expr = self.parse_expression()?;
                 ConsoleFunction::Assert(expr)
             }
-            "debug" => ConsoleFunction::Debug(self.parse_console_args()?),
             "error" => ConsoleFunction::Error(self.parse_console_args()?),
             "log" => ConsoleFunction::Log(self.parse_console_args()?),
             x => {
                 return Err(SyntaxError::unexpected_ident(
                     &x,
-                    &["assert", "debug", "error", "log"],
+                    &["assert", "error", "log"],
                     &function.span,
                 ));
             }
diff --git a/state/Cargo.toml b/state/Cargo.toml
index 8e3946214d..13a4c99a12 100644
--- a/state/Cargo.toml
+++ b/state/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "leo-state"
-version = "1.5.2"
+version = "1.5.3"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "State parser of the Leo programming language"
 homepage = "https://aleo.org"
@@ -19,11 +19,11 @@ edition = "2018"
 
 [dependencies.leo-input]
 path = "../input"
-version = "1.5.2"
+version = "1.5.3"
 
 [dependencies.leo-ast]
 path = "../ast"
-version = "1.5.2"
+version = "1.5.3"
 
 [dependencies.snarkvm-algorithms]
 version = "0.7.4"
diff --git a/synthesizer/Cargo.toml b/synthesizer/Cargo.toml
index 7daf23a580..4a84f46cb7 100644
--- a/synthesizer/Cargo.toml
+++ b/synthesizer/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "leo-synthesizer"
-version = "1.5.2"
+version = "1.5.3"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Circuit synthesizer of the Leo programming language"
 homepage = "https://aleo.org"
diff --git a/test-framework/Cargo.toml b/test-framework/Cargo.toml
index 28c8aabc0a..fa5440225b 100644
--- a/test-framework/Cargo.toml
+++ b/test-framework/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "leo-test-framework"
-version = "1.5.2"
+version = "1.5.3"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Leo testing framework"
 homepage = "https://aleo.org"
diff --git a/tests/compiler/address/branch.leo b/tests/compiler/address/branch.leo
new file mode 100644
index 0000000000..5b2023a43b
--- /dev/null
+++ b/tests/compiler/address/branch.leo
@@ -0,0 +1,15 @@
+/*
+namespace: Compile
+expectation: Pass
+input_file: inputs/branch.in
+*/
+
+
+function main (x: address, y: bool) -> bool {
+    let z = aleo18cw5zdez3zhypev3tnfhmwvhre9ramwle4up947gcyy5rnmjw5yqn93wsr;
+    if y {
+        z = aleo1f2gs8g0qpumlgzpvmkw3q07y6xrwsdr0lqsu9h9fgnh8d7e44v9qhpgpkj;
+    }
+
+    return z == aleo1f2gs8g0qpumlgzpvmkw3q07y6xrwsdr0lqsu9h9fgnh8d7e44v9qhpgpkj;
+}
\ No newline at end of file
diff --git a/tests/compiler/address/index.leo b/tests/compiler/address/index.leo
new file mode 100644
index 0000000000..5d7cb2b389
--- /dev/null
+++ b/tests/compiler/address/index.leo
@@ -0,0 +1,12 @@
+/*
+namespace: Compile
+expectation: Pass
+input_file: inputs/index.in
+*/
+
+function main (x: u32) -> bool {
+    const y = [aleo1x0rh2cudq93fhukrsce8sgvcphddv4qs0clph64stpg0hstfds9qjvxcg6; 3];
+    let z = y[x];
+
+	return z == y[0];
+}
\ No newline at end of file
diff --git a/tests/compiler/address/inputs/branch.in b/tests/compiler/address/inputs/branch.in
new file mode 100644
index 0000000000..24e6b338e0
--- /dev/null
+++ b/tests/compiler/address/inputs/branch.in
@@ -0,0 +1,6 @@
+[main]
+x: address = aleo1qnr4dkkvkgfqph0vzc3y6z2eu975wnpz2925ntjccd5cfqxtyu8sta57j8;
+y: bool = true;
+
+[registers]
+a: bool = false;
\ No newline at end of file
diff --git a/tests/compiler/address/inputs/index.in b/tests/compiler/address/inputs/index.in
new file mode 100644
index 0000000000..5e701ee7a6
--- /dev/null
+++ b/tests/compiler/address/inputs/index.in
@@ -0,0 +1,5 @@
+[main]
+x: u32 = 0;
+
+[registers]
+a: bool = false;
\ No newline at end of file
diff --git a/tests/compiler/console/debug.leo b/tests/compiler/console/debug.leo
deleted file mode 100644
index 4b24b6e5fc..0000000000
--- a/tests/compiler/console/debug.leo
+++ /dev/null
@@ -1,10 +0,0 @@
-/*
-namespace: Compile
-expectation: Pass
-input_file: input/dummy.in
-*/
-
-function main(y: bool) -> bool {
-    console.debug("hello debug");
-    return y == true;
-}
diff --git a/tests/compiler/function/duplicate_definition_fail.leo b/tests/compiler/function/duplicate_definition_fail.leo
index 2909fae2b1..43ec30e787 100644
--- a/tests/compiler/function/duplicate_definition_fail.leo
+++ b/tests/compiler/function/duplicate_definition_fail.leo
@@ -1,15 +1,14 @@
 /*
 namespace: Compile
 expectation: Fail
-input_file: input/integers.in
+input_file: input/dummy.in
 */
 
 
-function main(a: u32) -> u32 {
+function main() {
     console.log("{}", 1u8);
-    return 10u32; 
 }
 
 function main() {
-    console.log("{}", 2u8);  
+    console.log("{}", 2u8);
 }
diff --git a/tests/expectations/compiler/compiler/address/branch.leo.out b/tests/expectations/compiler/compiler/address/branch.leo.out
new file mode 100644
index 0000000000..a4c062bc76
--- /dev/null
+++ b/tests/expectations/compiler/compiler/address/branch.leo.out
@@ -0,0 +1,21 @@
+---
+namespace: Compile
+expectation: Pass
+outputs:
+  - circuit:
+      num_public_variables: 0
+      num_private_variables: 1024
+      num_constraints: 1536
+      at: 5afd1d58b6826912fe5cba06b60d9a7debdbad9e922b8a78ed49f7a7ca0ac65e
+      bt: db77c3470cf1b3c80c2e3b1b3ddc59d9891912489bfcf39b7f67c2f314ca7f6d
+      ct: d0993682df5b495f4a6784882e0f007dbc378adb35007d250e2c098975f4d32e
+    output:
+      - input_file: inputs/branch.in
+        output:
+          registers:
+            a:
+              type: bool
+              value: "true"
+    initial_ast: a7748573e4731753b76889c6d4c28e9589e114860a163da87957cf20b916f733
+    canonicalized_ast: a7748573e4731753b76889c6d4c28e9589e114860a163da87957cf20b916f733
+    type_inferenced_ast: af3663710ad6278c3d2e28a753d62084505746cd49de11949fe5e8e390ffcc60
diff --git a/tests/expectations/compiler/compiler/address/index.leo.out b/tests/expectations/compiler/compiler/address/index.leo.out
new file mode 100644
index 0000000000..b4cda33ee6
--- /dev/null
+++ b/tests/expectations/compiler/compiler/address/index.leo.out
@@ -0,0 +1,21 @@
+---
+namespace: Compile
+expectation: Pass
+outputs:
+  - circuit:
+      num_public_variables: 0
+      num_private_variables: 1694
+      num_constraints: 2719
+      at: f9a9bc8304df327c7b16f088b61ac518fb3a7b6457ccc35a1ab9eab4565a1981
+      bt: 045eba0fdd405714f788492fd5914b053e0ece622e9a542270d8cfe85c0ee291
+      ct: e8fa4a33a657c70a7cecc7e9b284157b369b7ea57879948e108daf93e778551c
+    output:
+      - input_file: inputs/index.in
+        output:
+          registers:
+            a:
+              type: bool
+              value: "true"
+    initial_ast: 8934e4c1d645f6f98fe59f48bbf687623e811c99e504e87c6989bc00af62194a
+    canonicalized_ast: 8934e4c1d645f6f98fe59f48bbf687623e811c99e504e87c6989bc00af62194a
+    type_inferenced_ast: 70af835aeaec1d5bc1c4a3186635260ff44743e0d3a7aa5ac9f2c98ec03bd23e
diff --git a/tests/expectations/compiler/compiler/array/variable_slice_fail.leo.out b/tests/expectations/compiler/compiler/array/variable_slice_fail.leo.out
index 307885a20d..b38b3bbbf2 100644
--- a/tests/expectations/compiler/compiler/array/variable_slice_fail.leo.out
+++ b/tests/expectations/compiler/compiler/array/variable_slice_fail.leo.out
@@ -2,4 +2,4 @@
 namespace: Compile
 expectation: Fail
 outputs:
-  - "    --> compiler-test:6:19\n     |\n   6 |         const x = a[i..10];\n     |                   ^^^^^^^^\n     |\n     = array size cannot be inferred, add explicit types"
+  - "    --> compiler-test:7:17\n     |\n   7 |         console.debug(\"{}\", x);\n     |                 ^^^^^\n     |\n     = expected identifier 'assert', 'error', 'log' -- got 'debug'"
diff --git a/tests/expectations/compiler/compiler/core/core_circuit_star_fail.leo.out b/tests/expectations/compiler/compiler/core/core_circuit_star_fail.leo.out
index 104faae420..9bb7631c2d 100644
--- a/tests/expectations/compiler/compiler/core/core_circuit_star_fail.leo.out
+++ b/tests/expectations/compiler/compiler/core/core_circuit_star_fail.leo.out
@@ -2,4 +2,4 @@
 namespace: Compile
 expectation: Fail
 outputs:
-  - "    --> :0:0\n     |\n     |\n     |\n     = failed to resolve import: 'core'"
+  - "    --> :0:0\n          |\n     |\n     = failed to resolve import: 'core'"
diff --git a/tests/expectations/compiler/compiler/core/core_package_invalid.leo.out b/tests/expectations/compiler/compiler/core/core_package_invalid.leo.out
index 104faae420..9bb7631c2d 100644
--- a/tests/expectations/compiler/compiler/core/core_package_invalid.leo.out
+++ b/tests/expectations/compiler/compiler/core/core_package_invalid.leo.out
@@ -2,4 +2,4 @@
 namespace: Compile
 expectation: Fail
 outputs:
-  - "    --> :0:0\n     |\n     |\n     |\n     = failed to resolve import: 'core'"
+  - "    --> :0:0\n          |\n     |\n     = failed to resolve import: 'core'"
diff --git a/tests/expectations/compiler/compiler/core/core_unstable_package_invalid.leo.out b/tests/expectations/compiler/compiler/core/core_unstable_package_invalid.leo.out
index 4fb07b9d05..2493893de3 100644
--- a/tests/expectations/compiler/compiler/core/core_unstable_package_invalid.leo.out
+++ b/tests/expectations/compiler/compiler/core/core_unstable_package_invalid.leo.out
@@ -2,4 +2,4 @@
 namespace: Compile
 expectation: Fail
 outputs:
-  - "    --> :0:0\n     |\n     |\n     |\n     = failed to resolve import: 'core.unstable'"
+  - "    --> :0:0\n          |\n     |\n     = failed to resolve import: 'core.unstable'"
diff --git a/tests/expectations/compiler/compiler/function/duplicate_definition_fail.leo.out b/tests/expectations/compiler/compiler/function/duplicate_definition_fail.leo.out
index 7b1e418ef4..f76592d7fd 100644
--- a/tests/expectations/compiler/compiler/function/duplicate_definition_fail.leo.out
+++ b/tests/expectations/compiler/compiler/function/duplicate_definition_fail.leo.out
@@ -2,4 +2,4 @@
 namespace: Compile
 expectation: Fail
 outputs:
-  - "    --> compiler-test:6:12\n     |\n   6 |     return 10u32; \n     |            ^^^^^\n     |\n     = unexpected type, expected: '()', received: 'u32'"
+  - "    --> compiler-test:8:1\n     |\n   8 | function main() {\n|\n   9 | ...\n|\n  10 | }\n     | ^\n     |\n     = a function named \"main\" already exists in this scope"
diff --git a/tests/expectations/compiler/compiler/function/multiple_returns_fail.leo.out b/tests/expectations/compiler/compiler/function/multiple_returns_fail.leo.out
index 5fcaab60f3..6738b2788e 100644
--- a/tests/expectations/compiler/compiler/function/multiple_returns_fail.leo.out
+++ b/tests/expectations/compiler/compiler/function/multiple_returns_fail.leo.out
@@ -2,4 +2,4 @@
 namespace: Compile
 expectation: Fail
 outputs:
-  - "    --> compiler-test:4:5\n     |\n   4 |     if true {\n   5 | ...\n   6 |     }\n     |     ^\n     |\n     = function 'main' failed to validate return path: 'cannot have asymmetrical return in if statement'"
+  - "    --> compiler-test:4:5\n     |\n   4 |     if true {\n|\n   5 | ...\n|\n   6 |     }\n     |     ^\n     |\n     = function 'main' failed to validate return path: 'cannot have asymmetrical return in if statement'"
diff --git a/tests/expectations/compiler/compiler/function/multiple_returns_fail_conditional.leo.out b/tests/expectations/compiler/compiler/function/multiple_returns_fail_conditional.leo.out
index 52b50d8a3e..ee103c4030 100644
--- a/tests/expectations/compiler/compiler/function/multiple_returns_fail_conditional.leo.out
+++ b/tests/expectations/compiler/compiler/function/multiple_returns_fail_conditional.leo.out
@@ -2,4 +2,4 @@
 namespace: Compile
 expectation: Fail
 outputs:
-  - "    --> compiler-test:3:1\n     |\n   3 | function main() -> bool {\n   4 | ...\n   5 | }\n   6 | \n   7 | \n   8 | \n   9 | \n  10 | \n     | ^\n     |\n     = function 'main' missing return for all paths"
+  - "    --> compiler-test:3:1\n     |\n   3 | function main() -> bool {\n|\n   4 | ...\n|\n   5 | }\n|\n   6 | \n|\n   7 | \n|\n   8 | \n|\n   9 | \n|\n  10 | \n     | ^\n     |\n     = function 'main' missing return for all paths"
diff --git a/tests/expectations/compiler/compiler/input_files/program_registers/registers_fail.leo.out b/tests/expectations/compiler/compiler/input_files/program_registers/registers_fail.leo.out
index e074c91490..6609ce03a3 100644
--- a/tests/expectations/compiler/compiler/input_files/program_registers/registers_fail.leo.out
+++ b/tests/expectations/compiler/compiler/input_files/program_registers/registers_fail.leo.out
@@ -2,4 +2,4 @@
 namespace: Compile
 expectation: Fail
 outputs:
-  - "    --> compiler-test:3:1\n     |\n   3 | function main() -> bool {\n   4 | ...\n   5 | }\n     | ^\n     |\n     = Mismatched types. Expected register output type `u8`, found type `bool`."
+  - "    --> compiler-test:3:1\n     |\n   3 | function main() -> bool {\n|\n   4 | ...\n|\n   5 | }\n     | ^\n     |\n     = Mismatched types. Expected register output type `u8`, found type `bool`."
diff --git a/tests/expectations/parser/parser/statement/console.leo.out b/tests/expectations/parser/parser/statement/console.leo.out
index 2ed38532c9..4ff3e93b83 100644
--- a/tests/expectations/parser/parser/statement/console.leo.out
+++ b/tests/expectations/parser/parser/statement/console.leo.out
@@ -80,73 +80,6 @@ outputs:
         col_stop: 18
         path: test
         content: "console.error(\"x\");"
-  - Console:
-      function:
-        Debug:
-          string:
-            - Scalar: 123
-            - Scalar: 125
-          parameters:
-            - Identifier: "{\"name\":\"x\",\"span\":\"{\\\"line_start\\\":1,\\\"line_stop\\\":1,\\\"col_start\\\":21,\\\"col_stop\\\":22,\\\"path\\\":\\\"test\\\",\\\"content\\\":\\\"console.debug(\\\\\\\"{}\\\\\\\", x);\\\"}\"}"
-          span:
-            line_start: 1
-            line_stop: 1
-            col_start: 15
-            col_stop: 22
-            path: test
-            content: "console.debug(\"{}\", x);"
-      span:
-        line_start: 1
-        line_stop: 1
-        col_start: 1
-        col_stop: 22
-        path: test
-        content: "console.debug(\"{}\", x);"
-  - Console:
-      function:
-        Debug:
-          string:
-            - Scalar: 123
-            - Scalar: 125
-            - Scalar: 123
-            - Scalar: 125
-          parameters:
-            - Identifier: "{\"name\":\"x\",\"span\":\"{\\\"line_start\\\":1,\\\"line_stop\\\":1,\\\"col_start\\\":23,\\\"col_stop\\\":24,\\\"path\\\":\\\"test\\\",\\\"content\\\":\\\"console.debug(\\\\\\\"{}{}\\\\\\\", x, y);\\\"}\"}"
-            - Identifier: "{\"name\":\"y\",\"span\":\"{\\\"line_start\\\":1,\\\"line_stop\\\":1,\\\"col_start\\\":26,\\\"col_stop\\\":27,\\\"path\\\":\\\"test\\\",\\\"content\\\":\\\"console.debug(\\\\\\\"{}{}\\\\\\\", x, y);\\\"}\"}"
-          span:
-            line_start: 1
-            line_stop: 1
-            col_start: 15
-            col_stop: 27
-            path: test
-            content: "console.debug(\"{}{}\", x, y);"
-      span:
-        line_start: 1
-        line_stop: 1
-        col_start: 1
-        col_stop: 27
-        path: test
-        content: "console.debug(\"{}{}\", x, y);"
-  - Console:
-      function:
-        Debug:
-          string:
-            - Scalar: 120
-          parameters: []
-          span:
-            line_start: 1
-            line_stop: 1
-            col_start: 15
-            col_stop: 18
-            path: test
-            content: "console.debug(\"x\");"
-      span:
-        line_start: 1
-        line_stop: 1
-        col_start: 1
-        col_stop: 18
-        path: test
-        content: "console.debug(\"x\");"
   - Console:
       function:
         Log:
diff --git a/tests/parser/statement/console.leo b/tests/parser/statement/console.leo
index ba1b5e2cf4..773ec668b0 100644
--- a/tests/parser/statement/console.leo
+++ b/tests/parser/statement/console.leo
@@ -13,13 +13,6 @@ console.error("{}{}", x, y);
 console.error("x");
 
 
-console.debug("{}", x);
-
-console.debug("{}{}", x, y);
-
-console.debug("x");
-
-
 console.log("{}", x);
 
 console.log("{}{}", x, y);
