diff --git a/examples b/examples
index e817c379b8..601a80730d 160000
--- a/examples
+++ b/examples
@@ -1 +1 @@
-Subproject commit e817c379b8075a4a4f22c2bfc2b62463e7a3ee40
+Subproject commit 601a80730d180c370a91fb72a0af39c2e20645b1
diff --git a/leo/cli/commands/common/options.rs b/leo/cli/commands/common/options.rs
index 0e40e70da7..01e95ac5eb 100644
--- a/leo/cli/commands/common/options.rs
+++ b/leo/cli/commands/common/options.rs
@@ -320,10 +320,6 @@ pub fn get_consensus_heights(network_name: NetworkName, is_devnet: bool) -> Vec<
 
 /// Validates a vector of heights as consensus heights.
 pub fn validate_consensus_heights(heights: &[u32]) -> anyhow::Result<()> {
-    // Check that the correct number of heights were provided.
-    let actual = heights.len();
-    let expected = ConsensusVersion::latest() as usize;
-    ensure!(actual == expected, "Expected exactly {expected} consensus heights, but found: {actual}");
     // Assert that the genesis height is 0.
     ensure!(heights[0] == 0, "Genesis height must be 0.");
     // Assert that the consensus heights are strictly increasing.
diff --git a/leo/cli/commands/devnet/mod.rs b/leo/cli/commands/devnet/mod.rs
index 4b5fffcb8b..793867c837 100644
--- a/leo/cli/commands/devnet/mod.rs
+++ b/leo/cli/commands/devnet/mod.rs
@@ -101,7 +101,7 @@ impl Command for LeoDevnet {
     }
 
     fn apply(self, _cx: Context, _: Self::Input) -> Result<Self::Output> {
-        self.handle_apply().map_err(|e| CliError::custom(format!("Failed start devnet: {e}")).into())
+        self.handle_apply().map_err(|e| CliError::custom(format!("Failed to start devnet: {e}")).into())
     }
 }
 
diff --git a/leo/package/src/package.rs b/leo/package/src/package.rs
index 9f0a62327e..64623d21ff 100644
--- a/leo/package/src/package.rs
+++ b/leo/package/src/package.rs
@@ -430,11 +430,6 @@ fn main_template(name: &str) -> String {
     format!(
         r#"// The '{name}' program.
 program {name}.aleo {{
-    transition main(public a: u32, b: u32) -> u32 {{
-        let c: u32 = a + b;
-        return c;
-    }}
-    
     // This is the constructor for the program.
     // The constructor allows you to manage program upgrades.
     // It is called when the program is deployed or upgraded.
@@ -443,9 +438,14 @@ program {name}.aleo {{
     //  - @admin(address="aleo1...")
     //  - @checksum(mapping="credits.aleo/fixme", key="0field")
     //  - @custom
-    // For more information, please refer to the documentation: <TODO>.
+    // For more information, please refer to the documentation: `https://docs.leo-lang.org/guides/upgradability`
     @noupgrade
     async constructor() {{}}
+    
+    transition main(public a: u32, b: u32) -> u32 {{
+        let c: u32 = a + b;
+        return c;
+    }}
 }}
 "#
     )
