diff --git a/interpreter/src/cursor.rs b/interpreter/src/cursor.rs
index 340d460520..3d59c79df7 100644
--- a/interpreter/src/cursor.rs
+++ b/interpreter/src/cursor.rs
@@ -654,8 +654,8 @@ impl Cursor {
             }
             Expression::ArrayAccess(array) if step == 1 => {
                 let span = array.span();
-                let array = self.pop_value()?;
                 let index = self.pop_value()?;
+                let array = self.pop_value()?;
 
                 let index_usize: usize = match index {
                     Value::U8(x) => x.into(),
diff --git a/tests/expectations/interpreter-leo/array_index.out b/tests/expectations/interpreter-leo/array_index.out
new file mode 100644
index 0000000000..cb97ca9209
--- /dev/null
+++ b/tests/expectations/interpreter-leo/array_index.out
@@ -0,0 +1 @@
+30field
diff --git a/tests/tests/interpreter-leo/array_index.leo b/tests/tests/interpreter-leo/array_index.leo
new file mode 100644
index 0000000000..f195e9295e
--- /dev/null
+++ b/tests/tests/interpreter-leo/array_index.leo
@@ -0,0 +1,7 @@
+
+program test.aleo {
+    transition main() -> field {
+        let array: [field; 2] = [30field, 40field];
+        return array[0u32];
+    }
+}
