diff --git a/Cargo.lock b/Cargo.lock
index 46338d8749..0b5a0c155f 100644
--- a/Cargo.lock
+++ b/Cargo.lock
@@ -46,9 +46,9 @@ dependencies = [
 
 [[package]]
 name = "anyhow"
-version = "1.0.41"
+version = "1.0.42"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "15af2628f6890fe2609a3b91bef4c83450512802e59489f9c1cb1fa5df064a61"
+checksum = "595d3cfa7a60d4555cb5067b99f07142a08ea778de5cf993f7b75c7d8fabc486"
 
 [[package]]
 name = "assert_cmd"
@@ -1222,6 +1222,7 @@ dependencies = [
  "leo-imports",
  "leo-input",
  "leo-package",
+ "leo-parser",
  "leo-state",
  "leo-synthesizer",
  "notify",
@@ -2506,9 +2507,9 @@ checksum = "8ea5119cdb4c55b55d432abb513a0429384878c15dde60cc77b1c99de1a95a6a"
 
 [[package]]
 name = "structopt"
-version = "0.3.21"
+version = "0.3.22"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "5277acd7ee46e63e5168a80734c9f6ee81b1367a7d8772a2d765df2a3705d28c"
+checksum = "69b041cdcb67226aca307e6e7be44c8806423d83e018bd662360a93dabce4d71"
 dependencies = [
  "clap",
  "lazy_static",
@@ -2517,9 +2518,9 @@ dependencies = [
 
 [[package]]
 name = "structopt-derive"
-version = "0.4.14"
+version = "0.4.15"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "5ba9cdfda491b814720b6b06e0cac513d922fc407582032e8706e9f137976f90"
+checksum = "7813934aecf5f51a54775e00068c237de98489463968231a51746bbbc03f9c10"
 dependencies = [
  "heck",
  "proc-macro-error",
diff --git a/Cargo.toml b/Cargo.toml
index db5ac5af9b..11ba58b260 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -61,6 +61,10 @@ version = "1.5.2"
 path = "./package"
 version = "1.5.2"
 
+[dependencies.leo-parser]
+path = "./parser"
+version = "1.5.2"
+
 [dependencies.leo-state]
 path = "./state"
 version = "1.5.2"
diff --git a/asg/tests/pass/address/equal.leo b/asg/tests/pass/address/equal.leo
index 9012a8c49d..856f5cc843 100644
--- a/asg/tests/pass/address/equal.leo
+++ b/asg/tests/pass/address/equal.leo
@@ -1,3 +1,3 @@
 function main(a: address, b: address, c: bool) {
-    console.assert(a == b == c);
+    console.assert((a == b) == c);
 }
\ No newline at end of file
diff --git a/asg/tests/pass/field/eq.leo b/asg/tests/pass/field/eq.leo
index e64307d8de..91d6d3c14d 100644
--- a/asg/tests/pass/field/eq.leo
+++ b/asg/tests/pass/field/eq.leo
@@ -1,3 +1,3 @@
 function main(a: field, b: field, c: bool) {
-    console.assert(a == b == c);
+    console.assert((a == b) == c);
 }
\ No newline at end of file
diff --git a/asg/tests/pass/group/eq.leo b/asg/tests/pass/group/eq.leo
index 89701d9c99..09f6210ac3 100644
--- a/asg/tests/pass/group/eq.leo
+++ b/asg/tests/pass/group/eq.leo
@@ -1,3 +1,3 @@
 function main(a: group, b: group, c: bool) {
-    console.assert(a == b == c);
+    console.assert((a == b) == c);
 }
\ No newline at end of file
diff --git a/asg/tests/pass/integers/i128/eq.leo b/asg/tests/pass/integers/i128/eq.leo
index f4beddc8fc..2491a1a5b4 100644
--- a/asg/tests/pass/integers/i128/eq.leo
+++ b/asg/tests/pass/integers/i128/eq.leo
@@ -1,3 +1,3 @@
 function main(a: i128, b: i128, c: bool) {
-    console.assert(a == b == c);
+    console.assert((a == b) == c);
 }
\ No newline at end of file
diff --git a/asg/tests/pass/integers/i128/ne.leo b/asg/tests/pass/integers/i128/ne.leo
index dae148b49c..beabf5672a 100644
--- a/asg/tests/pass/integers/i128/ne.leo
+++ b/asg/tests/pass/integers/i128/ne.leo
@@ -1,3 +1,3 @@
 function main(a: i128, b: i128, c: bool) {
-    console.assert(a != b == c);
+    console.assert((a != b) == c);
 }
\ No newline at end of file
diff --git a/asg/tests/pass/integers/i16/eq.leo b/asg/tests/pass/integers/i16/eq.leo
index 338e3e50f6..7577061061 100644
--- a/asg/tests/pass/integers/i16/eq.leo
+++ b/asg/tests/pass/integers/i16/eq.leo
@@ -1,3 +1,3 @@
 function main(a: i16, b: i16, c: bool) {
-    console.assert(a == b == c);
+    console.assert((a == b) == c);
 }
\ No newline at end of file
diff --git a/asg/tests/pass/integers/i16/ne.leo b/asg/tests/pass/integers/i16/ne.leo
index f0a9472678..9a749d22ba 100644
--- a/asg/tests/pass/integers/i16/ne.leo
+++ b/asg/tests/pass/integers/i16/ne.leo
@@ -1,3 +1,3 @@
 function main(a: i16, b: i16, c: bool) {
-    console.assert(a != b == c);
+    console.assert((a != b) == c);
 }
\ No newline at end of file
diff --git a/asg/tests/pass/integers/i32/eq.leo b/asg/tests/pass/integers/i32/eq.leo
index bac7af0be0..eb91dad4df 100644
--- a/asg/tests/pass/integers/i32/eq.leo
+++ b/asg/tests/pass/integers/i32/eq.leo
@@ -1,3 +1,3 @@
 function main(a: i32, b: i32, c: bool) {
-    console.assert(a == b == c);
+    console.assert((a == b) == c);
 }
\ No newline at end of file
diff --git a/asg/tests/pass/integers/i32/ne.leo b/asg/tests/pass/integers/i32/ne.leo
index dcc1a185d1..1ea8c38ef7 100644
--- a/asg/tests/pass/integers/i32/ne.leo
+++ b/asg/tests/pass/integers/i32/ne.leo
@@ -1,3 +1,3 @@
 function main(a: i32, b: i32, c: bool) {
-    console.assert(a != b == c);
+    console.assert((a != b) == c);
 }
\ No newline at end of file
diff --git a/asg/tests/pass/integers/i64/eq.leo b/asg/tests/pass/integers/i64/eq.leo
index 86a234c834..1ef39c659a 100644
--- a/asg/tests/pass/integers/i64/eq.leo
+++ b/asg/tests/pass/integers/i64/eq.leo
@@ -1,3 +1,3 @@
 function main(a: i64, b: i64, c: bool) {
-    console.assert(a == b == c);
+    console.assert((a == b) == c);
 }
\ No newline at end of file
diff --git a/asg/tests/pass/integers/i64/ne.leo b/asg/tests/pass/integers/i64/ne.leo
index a2642bb479..956b2a0a45 100644
--- a/asg/tests/pass/integers/i64/ne.leo
+++ b/asg/tests/pass/integers/i64/ne.leo
@@ -1,3 +1,3 @@
 function main(a: i64, b: i64, c: bool) {
-    console.assert(a != b == c);
+    console.assert((a != b) == c);
 }
\ No newline at end of file
diff --git a/asg/tests/pass/integers/i8/eq.leo b/asg/tests/pass/integers/i8/eq.leo
index 130cc64b89..e844a220b1 100644
--- a/asg/tests/pass/integers/i8/eq.leo
+++ b/asg/tests/pass/integers/i8/eq.leo
@@ -1,3 +1,3 @@
 function main(a: i8, b: i8, c: bool) {
-    console.assert(a == b == c);
+    console.assert((a == b) == c);
 }
\ No newline at end of file
diff --git a/asg/tests/pass/integers/i8/ne.leo b/asg/tests/pass/integers/i8/ne.leo
index e5028d404a..2ab15c8fe6 100644
--- a/asg/tests/pass/integers/i8/ne.leo
+++ b/asg/tests/pass/integers/i8/ne.leo
@@ -1,3 +1,3 @@
 function main(a: i8, b: i8, c: bool) {
-    console.assert(a != b == c);
+    console.assert((a != b) == c);
 }
\ No newline at end of file
diff --git a/asg/tests/pass/integers/u128/eq.leo b/asg/tests/pass/integers/u128/eq.leo
index 2c2acd923a..ad8faf0461 100644
--- a/asg/tests/pass/integers/u128/eq.leo
+++ b/asg/tests/pass/integers/u128/eq.leo
@@ -1,3 +1,3 @@
 function main(a: u128, b: u128, c: bool) {
-    console.assert(a == b == c);
+    console.assert((a == b) == c);
 }
\ No newline at end of file
diff --git a/asg/tests/pass/integers/u128/ne.leo b/asg/tests/pass/integers/u128/ne.leo
index da3467d01e..1752361c3f 100644
--- a/asg/tests/pass/integers/u128/ne.leo
+++ b/asg/tests/pass/integers/u128/ne.leo
@@ -1,3 +1,3 @@
 function main(a: u128, b: u128, c: bool) {
-    console.assert(a != b == c);
+    console.assert((a != b) == c);
 }
\ No newline at end of file
diff --git a/asg/tests/pass/integers/u16/eq.leo b/asg/tests/pass/integers/u16/eq.leo
index a4b4e78b49..589c6fb81f 100644
--- a/asg/tests/pass/integers/u16/eq.leo
+++ b/asg/tests/pass/integers/u16/eq.leo
@@ -1,3 +1,3 @@
 function main(a: u16, b: u16, c: bool) {
-    console.assert(a == b == c);
+    console.assert((a == b) == c);
 }
\ No newline at end of file
diff --git a/asg/tests/pass/integers/u16/ne.leo b/asg/tests/pass/integers/u16/ne.leo
index e90a304cfe..206c777be5 100644
--- a/asg/tests/pass/integers/u16/ne.leo
+++ b/asg/tests/pass/integers/u16/ne.leo
@@ -1,3 +1,3 @@
 function main(a: u16, b: u16, c: bool) {
-    console.assert(a != b == c);
+    console.assert((a != b) == c);
 }
\ No newline at end of file
diff --git a/asg/tests/pass/integers/u32/eq.leo b/asg/tests/pass/integers/u32/eq.leo
index ca427b3c42..1c90d7cf73 100644
--- a/asg/tests/pass/integers/u32/eq.leo
+++ b/asg/tests/pass/integers/u32/eq.leo
@@ -1,3 +1,3 @@
 function main(a: u32, b: u32, c: bool) {
-    console.assert(a == b == c);
+    console.assert((a == b) == c);
 }
\ No newline at end of file
diff --git a/asg/tests/pass/integers/u32/ne.leo b/asg/tests/pass/integers/u32/ne.leo
index 00ee1a9989..f086aa0639 100644
--- a/asg/tests/pass/integers/u32/ne.leo
+++ b/asg/tests/pass/integers/u32/ne.leo
@@ -1,3 +1,3 @@
 function main(a: u32, b: u32, c: bool) {
-    console.assert(a != b == c);
+    console.assert((a != b) == c);
 }
\ No newline at end of file
diff --git a/asg/tests/pass/integers/u64/eq.leo b/asg/tests/pass/integers/u64/eq.leo
index 990b2dad2d..e69a2802de 100644
--- a/asg/tests/pass/integers/u64/eq.leo
+++ b/asg/tests/pass/integers/u64/eq.leo
@@ -1,3 +1,3 @@
 function main(a: u64, b: u64, c: bool) {
-    console.assert(a == b == c);
+    console.assert((a == b) == c);
 }
\ No newline at end of file
diff --git a/asg/tests/pass/integers/u64/ne.leo b/asg/tests/pass/integers/u64/ne.leo
index e47acbb1de..4c94f15023 100644
--- a/asg/tests/pass/integers/u64/ne.leo
+++ b/asg/tests/pass/integers/u64/ne.leo
@@ -1,3 +1,3 @@
 function main(a: u64, b: u64, c: bool) {
-    console.assert(a != b == c);
+    console.assert((a != b) == c);
 }
\ No newline at end of file
diff --git a/asg/tests/pass/integers/u8/eq.leo b/asg/tests/pass/integers/u8/eq.leo
index c2a487b0e7..6fae190f8a 100644
--- a/asg/tests/pass/integers/u8/eq.leo
+++ b/asg/tests/pass/integers/u8/eq.leo
@@ -1,3 +1,3 @@
 function main(a: u8, b: u8, c: bool) {
-    console.assert(a == b == c);
+    console.assert((a == b) == c);
 }
\ No newline at end of file
diff --git a/asg/tests/pass/integers/u8/ne.leo b/asg/tests/pass/integers/u8/ne.leo
index e75194a2f2..1ef74d8b35 100644
--- a/asg/tests/pass/integers/u8/ne.leo
+++ b/asg/tests/pass/integers/u8/ne.leo
@@ -1,3 +1,3 @@
 function main(a: u8, b: u8, c: bool) {
-    console.assert(a != b == c);
+    console.assert((a != b) == c);
 }
\ No newline at end of file
diff --git a/ast/src/statements/conditional.rs b/ast/src/statements/conditional.rs
index e2b11d98af..bb78e83fff 100644
--- a/ast/src/statements/conditional.rs
+++ b/ast/src/statements/conditional.rs
@@ -31,7 +31,7 @@ impl fmt::Display for ConditionalStatement {
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
         write!(f, "if ({}) {}", self.condition, self.block)?;
         match self.next.clone() {
-            Some(n_or_e) => write!(f, " {}", n_or_e),
+            Some(n_or_e) => write!(f, " else {}", n_or_e),
             None => write!(f, ""),
         }
     }
diff --git a/ast/src/statements/console/formatted_string.rs b/ast/src/statements/console/formatted_string.rs
index f8f961f5cb..641362b35a 100644
--- a/ast/src/statements/console/formatted_string.rs
+++ b/ast/src/statements/console/formatted_string.rs
@@ -74,7 +74,7 @@ impl fmt::Display for FormatString {
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
         write!(
             f,
-            "{}",
+            "\"{}\", {}",
             self.parts
                 .iter()
                 .map(|x| match x {
@@ -82,7 +82,12 @@ impl fmt::Display for FormatString {
                     FormatStringPart::Container => "{}".to_string(),
                 })
                 .collect::<Vec<_>>()
-                .join("")
+                .join(""),
+            self.parameters
+                .iter()
+                .map(|p| p.to_string())
+                .collect::<Vec<_>>()
+                .join(",")
         )
     }
 }
diff --git a/compiler/src/compiler.rs b/compiler/src/compiler.rs
index 82de5ab734..a3f3475593 100644
--- a/compiler/src/compiler.rs
+++ b/compiler/src/compiler.rs
@@ -241,7 +241,7 @@ impl<'a, F: PrimeField, G: GroupType<F>> Compiler<'a, F, G> {
         let mut ast: leo_ast::Ast = parse_ast(self.main_file_path.to_str().unwrap_or_default(), program_string)?;
 
         if self.proof_options.initial {
-            ast.to_json_file(self.output_directory.clone(), "inital_ast.json")?;
+            ast.to_json_file(self.output_directory.clone(), "initial_ast.json")?;
         }
 
         // Preform compiler optimization via canonicalizing AST if its enabled.
diff --git a/grammar/abnf-grammar.txt b/grammar/abnf-grammar.txt
index 9015bf77c2..ecd8699580 100644
--- a/grammar/abnf-grammar.txt
+++ b/grammar/abnf-grammar.txt
@@ -453,7 +453,7 @@ identifier = letter *( letter / digit / "_" )   ; but not a keyword or aleo1...
 ; A package name consists of one or more segments separated by single dashes,
 ; where each segment is a non-empty sequence of lowercase letters and digits.
 
-package-name = 1*( lowercase-letter / digit )
+package-name = lowercase-letter *( lowercase-letter / digit )
                *( "-" 1*( lowercase-letter / digit ) )
 
 ; Annotations have names, which are identifiers immediately preceded by `@`.
@@ -848,8 +848,8 @@ ordering-expression = additive-expression
 ; the rule below makes them left-associative.
 
 equality-expression = ordering-expression
-                    / equality-expression "==" ordering-expression
-                    / equality-expression "!=" ordering-expression
+                    / ordering-expression "==" ordering-expression
+                    / ordering-expression "!=" ordering-expression
 
 ; Next come conjunctive expressions, left-associative.
 
@@ -962,7 +962,7 @@ assert-call = %s"assert" "(" expression ")"
 
 print-function = %s"debug" / %s"error" / %s"log"
 
-print-arguments = "(" [ string-literal *( "," expression ) ] ")"
+print-arguments = "(" string-literal  *( "," expression ) ")"
 
 print-call = print-function print-arguments
 
@@ -1020,7 +1020,7 @@ member-function-declaration = function-declaration
 ; this may be relaxed after the backward compatibility is removed,
 ; allowing member variables and member functions to be intermixed.
 
-circuit-declaration = *annotation %s"circuit" identifier
+circuit-declaration = %s"circuit" identifier
                       "{" [ member-variable-declarations ]
                       *member-function-declaration "}"
 
diff --git a/leo/commands/package/publish.rs b/leo/commands/package/publish.rs
index 213ce26340..a3d7f165c7 100644
--- a/leo/commands/package/publish.rs
+++ b/leo/commands/package/publish.rs
@@ -20,6 +20,7 @@ use leo_package::{
     outputs::OutputsDirectory,
     root::{ZipFile, AUTHOR_PLACEHOLDER},
 };
+use leo_parser::KEYWORD_TOKENS;
 
 use anyhow::{anyhow, Result};
 use structopt::StructOpt;
@@ -47,6 +48,10 @@ impl Command for Publish {
         let manifest = context.manifest()?;
 
         let package_name = manifest.get_package_name();
+        if KEYWORD_TOKENS.iter().any(|keyword| keyword.to_string() == package_name) {
+            return Err(anyhow!("Cannot name a package after a keyword"));
+        }
+
         let package_version = manifest.get_package_version();
 
         match (
diff --git a/parser/src/errors/syntax.rs b/parser/src/errors/syntax.rs
index 98fbc6bad8..306719b5a4 100644
--- a/parser/src/errors/syntax.rs
+++ b/parser/src/errors/syntax.rs
@@ -100,6 +100,10 @@ impl SyntaxError {
         )
     }
 
+    pub fn unexpected_statement(got: String, expected: &str, span: &Span) -> Self {
+        Self::new_from_span(format!("expected '{}', got '{}'", expected, got), span)
+    }
+
     pub fn unexpected_str(got: &Token, expected: &str, span: &Span) -> Self {
         Self::new_from_span(format!("expected '{}', got '{}'", expected, got.to_string()), span)
     }
diff --git a/parser/src/lib.rs b/parser/src/lib.rs
index 2e32c5ab4f..b896eee0b8 100644
--- a/parser/src/lib.rs
+++ b/parser/src/lib.rs
@@ -25,6 +25,7 @@
 extern crate thiserror;
 
 pub(crate) mod tokenizer;
+pub use tokenizer::KEYWORD_TOKENS;
 pub(crate) use tokenizer::*;
 
 pub mod errors;
diff --git a/parser/src/parser/expression.rs b/parser/src/parser/expression.rs
index f11208ce3e..789ecaf9de 100644
--- a/parser/src/parser/expression.rs
+++ b/parser/src/parser/expression.rs
@@ -187,7 +187,7 @@ impl ParserContext {
     ///
     pub fn parse_equality_expression(&mut self) -> SyntaxResult<Expression> {
         let mut expr = self.parse_ordering_expression()?;
-        while let Some(SpannedToken { token: op, .. }) = self.eat_any(&[Token::Eq, Token::NotEq]) {
+        if let Some(SpannedToken { token: op, .. }) = self.eat_any(&[Token::Eq, Token::NotEq]) {
             let right = self.parse_ordering_expression()?;
             expr = Expression::Binary(BinaryExpression {
                 span: expr.span() + right.span(),
diff --git a/parser/src/parser/file.rs b/parser/src/parser/file.rs
index e605165eed..80abacaddb 100644
--- a/parser/src/parser/file.rs
+++ b/parser/src/parser/file.rs
@@ -102,11 +102,20 @@ impl ParserContext {
         let end_span;
         let arguments = if self.eat(Token::LeftParen).is_some() {
             let mut args = Vec::new();
+            let mut comma = false;
             loop {
                 if let Some(end) = self.eat(Token::RightParen) {
+                    if comma {
+                        return Err(SyntaxError::unexpected(
+                            &Token::RightParen,
+                            &[Token::Ident("identifier".into()), Token::Int("number".into())],
+                            &end.span,
+                        ));
+                    }
                     end_span = end.span;
                     break;
                 }
+                comma = false;
                 if let Some(ident) = self.eat_identifier() {
                     args.push(ident.name);
                 } else if let Some((int, _)) = self.eat_int() {
@@ -115,10 +124,11 @@ impl ParserContext {
                     let token = self.peek()?;
                     return Err(SyntaxError::unexpected_str(&token.token, "ident or int", &token.span));
                 }
-                if self.eat(Token::Comma).is_none() {
+                if self.eat(Token::Comma).is_none() && !comma {
                     end_span = self.expect(Token::RightParen)?;
                     break;
                 }
+                comma = true;
             }
             args
         } else {
diff --git a/parser/src/parser/statement.rs b/parser/src/parser/statement.rs
index c1d1bb85a1..eff1b31e31 100644
--- a/parser/src/parser/statement.rs
+++ b/parser/src/parser/statement.rs
@@ -185,7 +185,17 @@ impl ParserContext {
         self.fuzzy_struct_state = false;
         let body = self.parse_block()?;
         let next = if self.eat(Token::Else).is_some() {
-            Some(Box::new(self.parse_statement()?))
+            let s = self.parse_statement()?;
+            match s {
+                Statement::Block(_) | Statement::Conditional(_) => Some(Box::new(s)),
+                s => {
+                    return Err(SyntaxError::unexpected_statement(
+                        s.to_string(),
+                        "Block or Conditional",
+                        s.span(),
+                    ));
+                }
+            }
         } else {
             None
         };
diff --git a/parser/src/tokenizer/mod.rs b/parser/src/tokenizer/mod.rs
index adfcbb22e5..5da374e575 100644
--- a/parser/src/tokenizer/mod.rs
+++ b/parser/src/tokenizer/mod.rs
@@ -22,6 +22,7 @@
 pub(crate) mod token;
 use std::sync::Arc;
 
+pub use self::token::KEYWORD_TOKENS;
 pub(crate) use self::token::*;
 
 pub(crate) mod lexer;
diff --git a/tests/compiler/_group/eq.leo b/tests/compiler/_group/eq.leo
index 89701d9c99..09f6210ac3 100644
--- a/tests/compiler/_group/eq.leo
+++ b/tests/compiler/_group/eq.leo
@@ -1,3 +1,3 @@
 function main(a: group, b: group, c: bool) {
-    console.assert(a == b == c);
+    console.assert((a == b) == c);
 }
\ No newline at end of file
diff --git a/tests/compiler/array/multi_initializer.leo b/tests/compiler/array/multi_initializer.leo
index 98faed5d05..bb523f4043 100644
--- a/tests/compiler/array/multi_initializer.leo
+++ b/tests/compiler/array/multi_initializer.leo
@@ -8,5 +8,5 @@ function main(y: bool) -> bool {
     const a: [u8; (2, 2, 2)] = [1u8; (2, 2, 2)];
     const b: [u8; (2, 2, 2)] = [[[1u8; 2]; 2]; 2];
 
-    return a == b == y;
+    return (a == b) == y;
 }
diff --git a/tests/compiler/array/nested_3x2_value.leo b/tests/compiler/array/nested_3x2_value.leo
index 85693cc6b7..0a3d67912e 100644
--- a/tests/compiler/array/nested_3x2_value.leo
+++ b/tests/compiler/array/nested_3x2_value.leo
@@ -9,5 +9,5 @@ function main(y: bool) -> bool {
     const a = [[0u32, 0u32], [0u32, 0u32], [0u32, 0u32]]; // inline
     const b: [u32; (3, 2)] = [[0; 2]; 3]; // initializer
 
-    return a == b == y;
+    return (a == b) == y;
 }
diff --git a/tests/compiler/array/slice_lower.leo b/tests/compiler/array/slice_lower.leo
index cea4bf1484..bc62c00e85 100644
--- a/tests/compiler/array/slice_lower.leo
+++ b/tests/compiler/array/slice_lower.leo
@@ -10,5 +10,5 @@ function main(y: bool) -> bool {
 
     const actual = arr[..2];     // Should produce [0, 1]
 
-    return expected == actual == y;
+    return (expected == actual) == y;
 }
diff --git a/tests/compiler/array/tuple_3x2_value.leo b/tests/compiler/array/tuple_3x2_value.leo
index 4e23eaaca2..72b7a588ad 100644
--- a/tests/compiler/array/tuple_3x2_value.leo
+++ b/tests/compiler/array/tuple_3x2_value.leo
@@ -9,5 +9,5 @@ function main(y: bool) -> bool {
     const a = [[0u32, 0u32], [0u32, 0u32], [0u32, 0u32]]; // inline
     const b: [u32; (3, 2)] = [0; (3, 2)]; // initializer
 
-    return a == b == y;
+    return (a == b) == y;
 }
diff --git a/tests/compiler/array/type_nested_value_nested_4x3x2.leo b/tests/compiler/array/type_nested_value_nested_4x3x2.leo
index 1845a03257..52746e70a7 100644
--- a/tests/compiler/array/type_nested_value_nested_4x3x2.leo
+++ b/tests/compiler/array/type_nested_value_nested_4x3x2.leo
@@ -12,5 +12,5 @@ function main(y: bool) -> bool {
 
     const b: [[[u8; 2]; 3]; 4] = [[[0; 2]; 3]; 4]; // initializer
 
-    return a == b == y;
+    return (a == b) == y;
 }
diff --git a/tests/compiler/array/type_nested_value_tuple_3x2.leo b/tests/compiler/array/type_nested_value_tuple_3x2.leo
index 7dd77e0baa..933b1959f5 100644
--- a/tests/compiler/array/type_nested_value_tuple_3x2.leo
+++ b/tests/compiler/array/type_nested_value_tuple_3x2.leo
@@ -8,5 +8,5 @@ function main(y: bool) -> bool {
     const a = [[0u8, 0u8], [0u8, 0u8], [0u8, 0u8]]; // inline
     const b: [[u8; 2]; 3] = [0; (3, 2)]; // initializer
 
-    return a == b == y;
+    return (a == b) == y;
 }
diff --git a/tests/compiler/array/type_nested_value_tuple_4x3x2.leo b/tests/compiler/array/type_nested_value_tuple_4x3x2.leo
index 1efc8658bb..0e6edc880f 100644
--- a/tests/compiler/array/type_nested_value_tuple_4x3x2.leo
+++ b/tests/compiler/array/type_nested_value_tuple_4x3x2.leo
@@ -12,5 +12,5 @@ function main(y: bool) -> bool {
 
     const b: [[[u8; 2]; 3]; 4] = [0; (4, 3, 2)]; // initializer
 
-    return a == b == y;
+    return (a == b) == y;
 }
diff --git a/tests/compiler/array/type_tuple_value_nested_3x2.leo b/tests/compiler/array/type_tuple_value_nested_3x2.leo
index 20c2fdea95..9beb94b036 100644
--- a/tests/compiler/array/type_tuple_value_nested_3x2.leo
+++ b/tests/compiler/array/type_tuple_value_nested_3x2.leo
@@ -9,5 +9,5 @@ function main(y: bool) -> bool {
 
     const b: [u8; (3, 2)] = [[0; 2]; 3]; // initializer
 
-    return a == b == y;
+    return (a == b) == y;
 }
diff --git a/tests/compiler/array/type_tuple_value_nested_4x3x2.leo b/tests/compiler/array/type_tuple_value_nested_4x3x2.leo
index 63d41b546e..61ce947207 100644
--- a/tests/compiler/array/type_tuple_value_nested_4x3x2.leo
+++ b/tests/compiler/array/type_tuple_value_nested_4x3x2.leo
@@ -12,5 +12,5 @@ function main(y: bool) -> bool {
 
     const b: [u8; (4, 3, 2)] = [[[0; 2]; 3]; 4]; // initializer
 
-    return a == b == y;
+    return (a == b) == y;
 }
diff --git a/tests/compiler/array/type_tuple_value_tuple_3x2.leo b/tests/compiler/array/type_tuple_value_tuple_3x2.leo
index 15754d3f4f..6a78f1793a 100644
--- a/tests/compiler/array/type_tuple_value_tuple_3x2.leo
+++ b/tests/compiler/array/type_tuple_value_tuple_3x2.leo
@@ -9,5 +9,5 @@ function main(y: bool) -> bool {
 
     const b: [u8; (3, 2)] = [0; (3, 2)]; // initializer
 
-    return a == b == y;
+    return (a == b) == y;
 }
diff --git a/tests/compiler/array/type_tuple_value_tuple_4x3x2.leo b/tests/compiler/array/type_tuple_value_tuple_4x3x2.leo
index 9fc4402f21..de0ca7f419 100644
--- a/tests/compiler/array/type_tuple_value_tuple_4x3x2.leo
+++ b/tests/compiler/array/type_tuple_value_tuple_4x3x2.leo
@@ -12,5 +12,5 @@ function main(y: bool) -> bool{
 
     const b: [u8; (4, 3, 2)] = [0; (4, 3, 2)]; // initializer
 
-    return a == b == y;
+    return (a == b) == y;
 }
diff --git a/tests/compiler/circuits/const_self_variable.leo b/tests/compiler/circuits/const_self_variable.leo
index 72322c9d2d..80a5474f13 100644
--- a/tests/compiler/circuits/const_self_variable.leo
+++ b/tests/compiler/circuits/const_self_variable.leo
@@ -15,5 +15,5 @@ circuit Foo {
 function main(y: bool) -> bool {
     let f  = Foo { a: 0u8 };
 
-    return 1u8 == f.use_a() == true;
+    return (1u8 == f.use_a()) == true;
 }
diff --git a/tests/compiler/circuits/inline_member_pass.leo b/tests/compiler/circuits/inline_member_pass.leo
index f7f22f000d..8d56383dab 100644
--- a/tests/compiler/circuits/inline_member_pass.leo
+++ b/tests/compiler/circuits/inline_member_pass.leo
@@ -25,5 +25,5 @@ function main(const x: u8, y: bool) -> bool {
     const a = Foo { x };
     const b = Foo::new(x);
 
-    return b.x == a.x == y;
+    return (b.x == a.x) == y;
 }
diff --git a/tests/compiler/circuits/member_function.leo b/tests/compiler/circuits/member_function.leo
index 660110c1dc..81b513d38c 100644
--- a/tests/compiler/circuits/member_function.leo
+++ b/tests/compiler/circuits/member_function.leo
@@ -15,5 +15,5 @@ circuit Foo {
 function main(y: bool) -> bool {
     const a = Foo { x: 1u32 };
 
-    return a.echo() == 1u32 == true;
+    return (a.echo() == 1u32) == true;
 }
diff --git a/tests/compiler/circuits/member_static_function.leo b/tests/compiler/circuits/member_static_function.leo
index 3c0fffd542..b9947f7007 100644
--- a/tests/compiler/circuits/member_static_function.leo
+++ b/tests/compiler/circuits/member_static_function.leo
@@ -13,5 +13,5 @@ circuit Foo {
 function main(y: bool) -> bool {
     const a = Foo::echo(1u32);
 
-    return a == 1u32 == y;
+    return (a == 1u32) == y;
 }
diff --git a/tests/compiler/circuits/member_variable.leo b/tests/compiler/circuits/member_variable.leo
index db779b49fa..53d49315c8 100644
--- a/tests/compiler/circuits/member_variable.leo
+++ b/tests/compiler/circuits/member_variable.leo
@@ -11,5 +11,5 @@ circuit Foo {
 function main(y: bool) -> bool {
     const a = Foo { x: 1u32 };
 
-    return a.x == 1u32 == y;
+    return (a.x == 1u32) == y;
 }
diff --git a/tests/compiler/circuits/member_variable_and_function.leo b/tests/compiler/circuits/member_variable_and_function.leo
index 33d48043bb..0b9a38c1b6 100644
--- a/tests/compiler/circuits/member_variable_and_function.leo
+++ b/tests/compiler/circuits/member_variable_and_function.leo
@@ -16,5 +16,5 @@ function main(y: bool) -> bool {
     const a = Foo { foo: 1 };
     const b = a.foo + Foo::bar();
 
-    return b == 2u32 == y;
+    return (b == 2u32) == y;
 }
diff --git a/tests/compiler/circuits/mut_self_variable.leo b/tests/compiler/circuits/mut_self_variable.leo
index ea90f9e3d3..4781cf6d20 100644
--- a/tests/compiler/circuits/mut_self_variable.leo
+++ b/tests/compiler/circuits/mut_self_variable.leo
@@ -18,5 +18,5 @@ function main(y: bool) -> bool {
     f.set_a(1u8);
     f.set_a(2u8);
 
-    return f.a == 2u8 == y;
+    return (f.a == 2u8) == y;
 }
diff --git a/tests/compiler/circuits/mut_self_variable_branch.leo b/tests/compiler/circuits/mut_self_variable_branch.leo
index 47148693b6..6e8cd83be7 100644
--- a/tests/compiler/circuits/mut_self_variable_branch.leo
+++ b/tests/compiler/circuits/mut_self_variable_branch.leo
@@ -22,5 +22,5 @@ function main(y: bool) -> bool {
     f.set_a(false, 2u8);
     f.set_a(true, 2u8);
 
-    return f.a == 2u8 == y;
+    return (f.a == 2u8) == y;
 }
diff --git a/tests/compiler/circuits/mut_variable.leo b/tests/compiler/circuits/mut_variable.leo
index 86d26c2cc6..640010a4b4 100644
--- a/tests/compiler/circuits/mut_variable.leo
+++ b/tests/compiler/circuits/mut_variable.leo
@@ -14,5 +14,5 @@ function main(y: bool) -> bool {
     f.a = 1u8;
     f.a = 2u8;
 
-    return f.a == 2u8 == true;
+    return (f.a == 2u8) == true;
 }
diff --git a/tests/compiler/circuits/mutable_call_immutable_context.leo b/tests/compiler/circuits/mutable_call_immutable_context.leo
index 477245a939..acb90ce0fc 100644
--- a/tests/compiler/circuits/mutable_call_immutable_context.leo
+++ b/tests/compiler/circuits/mutable_call_immutable_context.leo
@@ -25,5 +25,5 @@ function main(y: bool) -> bool {
     const t = TestMe {x: 6u8}.test_me();
     const u = my_fn().test_me();
     const v = TestMe::new().test_me();
-    return v == 2u8 == y;
+    return (v == 2u8) == y;
 }
diff --git a/tests/compiler/circuits/self_member.leo b/tests/compiler/circuits/self_member.leo
index 4cf1f74cf8..fc92fc989c 100644
--- a/tests/compiler/circuits/self_member.leo
+++ b/tests/compiler/circuits/self_member.leo
@@ -16,5 +16,5 @@ function main(y: bool) -> bool {
     const a = Foo { f: 1u32 };
     const b = a.bar();
 
-    return b == 1u32 == y;
+    return (b == 1u32) == y;
 }
diff --git a/tests/compiler/function/iteration.leo b/tests/compiler/function/iteration.leo
index 109cbb9b45..1c0b310fa5 100644
--- a/tests/compiler/function/iteration.leo
+++ b/tests/compiler/function/iteration.leo
@@ -15,5 +15,5 @@ function main(y: bool) -> bool {
         a += one();
     }
 
-    return a == 10u32 == y;
+    return (a == 10u32) == y;
 }
diff --git a/tests/compiler/function/iteration_repeated.leo b/tests/compiler/function/iteration_repeated.leo
index 9bff4980b0..bf443867bd 100644
--- a/tests/compiler/function/iteration_repeated.leo
+++ b/tests/compiler/function/iteration_repeated.leo
@@ -17,5 +17,5 @@ function iteration() -> u32 {
 function main(y: bool) -> bool {
     const total = iteration() + iteration();
 
-    return total == 20 == y;
+    return (total == 20) == y;
 }
diff --git a/tests/compiler/function/repeated.leo b/tests/compiler/function/repeated.leo
index 1834ac04f3..9486d09fba 100644
--- a/tests/compiler/function/repeated.leo
+++ b/tests/compiler/function/repeated.leo
@@ -11,5 +11,5 @@ function one() -> bool {
 function main(y: bool) -> bool {
     const a = one() && one();
 
-    return a == true == y;
+    return (a == true) == y;
 }
diff --git a/tests/compiler/function/return.leo b/tests/compiler/function/return.leo
index 2dd1966020..c6ad8a6366 100644
--- a/tests/compiler/function/return.leo
+++ b/tests/compiler/function/return.leo
@@ -9,5 +9,5 @@ function one() -> u32 {
 }
 
 function main(y: bool) -> bool {
-    return one() == 1u32 == y;
+    return (one() == 1u32) == y;
 }
diff --git a/tests/compiler/function/value_unchanged.leo b/tests/compiler/function/value_unchanged.leo
index 287ddf1f52..df1f5e03dd 100644
--- a/tests/compiler/function/value_unchanged.leo
+++ b/tests/compiler/function/value_unchanged.leo
@@ -21,5 +21,5 @@ function main(y: bool) -> bool {
     const a = 1u32;
     bad_mutate(a);
 
-    return a == 1u32 == y; // <- value `a` is still `1u32`
+    return (a == 1u32) == y; // <- value `a` is still `1u32`
 }
diff --git a/tests/compiler/import/alias.leo b/tests/compiler/import/alias.leo
index 12bb1929a0..6d3db96c1b 100644
--- a/tests/compiler/import/alias.leo
+++ b/tests/compiler/import/alias.leo
@@ -6,5 +6,5 @@
 import test-import.foo as bar;
 
 function main(y: bool) -> bool {
-    return bar() == 1u32 == y;
+    return (bar() == 1u32) == y;
 }
diff --git a/tests/compiler/import/basic.leo b/tests/compiler/import/basic.leo
index 8961bcf9a9..e1996bfb90 100644
--- a/tests/compiler/import/basic.leo
+++ b/tests/compiler/import/basic.leo
@@ -6,5 +6,5 @@
 import test-import.foo;
 
 function main(y: bool) -> bool {
-    return foo() == 1u32 == true;
+    return (foo() == 1u32) == true;
 }
diff --git a/tests/compiler/import/many_import.leo b/tests/compiler/import/many_import.leo
index 01c8acfbf4..f0d1958377 100644
--- a/tests/compiler/import/many_import.leo
+++ b/tests/compiler/import/many_import.leo
@@ -27,5 +27,5 @@ function main(y: bool) -> bool {
 
     const car = Car { c: 1u32 };
 
-    return car.c == 1u32 == y;
+    return (car.c == 1u32) == y;
 }
diff --git a/tests/compiler/import/many_import_star.leo b/tests/compiler/import/many_import_star.leo
index 9b49a8f9c9..0076078903 100644
--- a/tests/compiler/import/many_import_star.leo
+++ b/tests/compiler/import/many_import_star.leo
@@ -20,5 +20,5 @@ function main(y: bool) -> bool {
 
     const car = Car { c: 1u32 };
 
-    return car.c == 1u32 == y;
+    return (car.c == 1u32) == y;
 }
diff --git a/tests/compiler/import/multiple.leo b/tests/compiler/import/multiple.leo
index aaa6148957..9c4eb1e5a7 100644
--- a/tests/compiler/import/multiple.leo
+++ b/tests/compiler/import/multiple.leo
@@ -11,5 +11,5 @@ import test-import.(
 function main(y: bool) -> bool {
     const a = Point { x: 1u32, y: 0u32 };
 
-    return a.x == 1u32 == y;
+    return (a.x == 1u32) == y;
 }
diff --git a/tests/compiler/import/star.leo b/tests/compiler/import/star.leo
index 82d888e29f..295ec2e199 100644
--- a/tests/compiler/import/star.leo
+++ b/tests/compiler/import/star.leo
@@ -8,5 +8,5 @@ import test-import.*;
 function main(y: bool) -> bool {
     const a = Point { x: 1u32, y: 0u32 };
 
-    return foo() == 1u32 == y;
+    return (foo() == 1u32) == y;
 }
diff --git a/tests/compiler/input_files/program_input_constants/main.leo b/tests/compiler/input_files/program_input_constants/main.leo
index c7df484e15..2ef2c8df2e 100644
--- a/tests/compiler/input_files/program_input_constants/main.leo
+++ b/tests/compiler/input_files/program_input_constants/main.leo
@@ -5,5 +5,5 @@ input_file: input/main.in
 */
 
 function main(const a: bool, b: bool) -> bool {
-    return a == true == b;
+    return (a == true) == b;
 }
diff --git a/tests/compiler/input_files/program_input_constants/main_array.leo b/tests/compiler/input_files/program_input_constants/main_array.leo
index c9e4ba20d1..8de1cc7550 100644
--- a/tests/compiler/input_files/program_input_constants/main_array.leo
+++ b/tests/compiler/input_files/program_input_constants/main_array.leo
@@ -6,5 +6,5 @@ input_file: input/main_array.in
 
 function main(const x: [i16; 1], y: bool) -> bool {
     console.log("{}", x);
-    return x[0] == 0 == y;
+    return (x[0] == 0) == y;
 }
diff --git a/tests/compiler/input_files/program_input_constants/main_multiple.leo b/tests/compiler/input_files/program_input_constants/main_multiple.leo
index 7f57b8d26c..e06775ea4b 100644
--- a/tests/compiler/input_files/program_input_constants/main_multiple.leo
+++ b/tests/compiler/input_files/program_input_constants/main_multiple.leo
@@ -5,5 +5,5 @@ input_file: input/main_multiple.in
 */
 
 function main(const a: bool, const b: bool, y: bool) -> bool {
-    return a != b == y;
+    return (a != b) == y;
 }
diff --git a/tests/compiler/integers/i128/eq.leo b/tests/compiler/integers/i128/eq.leo
index a319c3e672..22cc373cca 100644
--- a/tests/compiler/integers/i128/eq.leo
+++ b/tests/compiler/integers/i128/eq.leo
@@ -13,5 +13,5 @@ inputs:
 */
 
 function main(a: i128, b: i128, c: bool) -> bool {
-    return a == b == c;
+    return (a == b) == c;
 }
diff --git a/tests/compiler/integers/i128/ne.leo b/tests/compiler/integers/i128/ne.leo
index c8cc08abb9..626cb7dc27 100644
--- a/tests/compiler/integers/i128/ne.leo
+++ b/tests/compiler/integers/i128/ne.leo
@@ -21,5 +21,5 @@ inputs:
 */
 
 function main(a: i128, b: i128, c: bool) -> bool{
-    return a != b == c;
+    return (a != b) == c;
 }
diff --git a/tests/compiler/integers/i128/negate_min.leo b/tests/compiler/integers/i128/negate_min.leo
index 1997fb3bc8..d3163b0e0f 100644
--- a/tests/compiler/integers/i128/negate_min.leo
+++ b/tests/compiler/integers/i128/negate_min.leo
@@ -7,5 +7,5 @@ input_file: ../input/dummy.in
 function main(y: bool) -> bool{
     const a = -128i128;
     const b = -a;
-    return b == -a == y;
+    return (b == -a) == y;
 }
diff --git a/tests/compiler/integers/i128/negate_zero.leo b/tests/compiler/integers/i128/negate_zero.leo
index b78fff141c..29426913af 100644
--- a/tests/compiler/integers/i128/negate_zero.leo
+++ b/tests/compiler/integers/i128/negate_zero.leo
@@ -13,5 +13,5 @@ inputs:
 function main(y: bool) -> bool {
     const a = 0i128;
 
-    return -a == 0i128 == y;
+    return (-a == 0i128) == y;
 }
diff --git a/tests/compiler/integers/i16/eq.leo b/tests/compiler/integers/i16/eq.leo
index 91980ee899..32fd8636f4 100644
--- a/tests/compiler/integers/i16/eq.leo
+++ b/tests/compiler/integers/i16/eq.leo
@@ -13,5 +13,5 @@ inputs:
 */
 
 function main(a: i16, b: i16, c: bool) -> bool {
-    return a == b == c;
+    return (a == b) == c;
 }
diff --git a/tests/compiler/integers/i16/ne.leo b/tests/compiler/integers/i16/ne.leo
index 676bc78a0d..9d1d3ed5b4 100644
--- a/tests/compiler/integers/i16/ne.leo
+++ b/tests/compiler/integers/i16/ne.leo
@@ -21,5 +21,5 @@ inputs:
 */
 
 function main(a: i16, b: i16, c: bool) -> bool{
-    return a != b == c;
+    return (a != b) == c;
 }
diff --git a/tests/compiler/integers/i16/negate_min.leo b/tests/compiler/integers/i16/negate_min.leo
index 511535385a..8f89f8f091 100644
--- a/tests/compiler/integers/i16/negate_min.leo
+++ b/tests/compiler/integers/i16/negate_min.leo
@@ -7,5 +7,5 @@ input_file: ../input/dummy.in
 function main(y: bool) -> bool {
     const a = -128i16;
     const b = -a;
-    return b == -a == y;
+    return (b == -a) == y;
 }
diff --git a/tests/compiler/integers/i16/negate_zero.leo b/tests/compiler/integers/i16/negate_zero.leo
index 1c1c9eef8b..ad27821561 100644
--- a/tests/compiler/integers/i16/negate_zero.leo
+++ b/tests/compiler/integers/i16/negate_zero.leo
@@ -13,5 +13,5 @@ inputs:
 function main(y: bool) -> bool {
     const a = 0i16;
 
-    return -a == 0i16 == y;
+    return (-a == 0i16) == y;
 }
diff --git a/tests/compiler/integers/i32/eq.leo b/tests/compiler/integers/i32/eq.leo
index 93453b81d6..7f5b36d261 100644
--- a/tests/compiler/integers/i32/eq.leo
+++ b/tests/compiler/integers/i32/eq.leo
@@ -13,5 +13,5 @@ inputs:
 */
 
 function main(a: i32, b: i32, c: bool) -> bool {
-    return a == b == c;
+    return (a == b) == c;
 }
diff --git a/tests/compiler/integers/i32/ne.leo b/tests/compiler/integers/i32/ne.leo
index 60b30017af..502ca22281 100644
--- a/tests/compiler/integers/i32/ne.leo
+++ b/tests/compiler/integers/i32/ne.leo
@@ -21,5 +21,5 @@ inputs:
 */
 
 function main(a: i32, b: i32, c: bool) -> bool{
-    return a != b == c;
+    return (a != b) == c;
 }
diff --git a/tests/compiler/integers/i32/negate_min.leo b/tests/compiler/integers/i32/negate_min.leo
index c81f0b3d93..0f3d6f306d 100644
--- a/tests/compiler/integers/i32/negate_min.leo
+++ b/tests/compiler/integers/i32/negate_min.leo
@@ -7,5 +7,5 @@ input_file: ../input/dummy.in
 function main(y: bool) -> bool{
     const a = -128i32;
     const b = -a;
-    return b == -a == y;
+    return (b == -a) == y;
 }
diff --git a/tests/compiler/integers/i32/negate_zero.leo b/tests/compiler/integers/i32/negate_zero.leo
index 0b224e64a1..fe9084ea56 100644
--- a/tests/compiler/integers/i32/negate_zero.leo
+++ b/tests/compiler/integers/i32/negate_zero.leo
@@ -13,5 +13,5 @@ inputs:
 function main(y: bool) -> bool {
     const a = 0i32;
 
-    return -a == 0i32 == y;
+    return (-a == 0i32) == y;
 }
diff --git a/tests/compiler/integers/i64/eq.leo b/tests/compiler/integers/i64/eq.leo
index 5c8c13d794..b317353fef 100644
--- a/tests/compiler/integers/i64/eq.leo
+++ b/tests/compiler/integers/i64/eq.leo
@@ -13,5 +13,5 @@ inputs:
 */
 
 function main(a: i64, b: i64, c: bool) -> bool {
-    return a == b == c;
+    return (a == b) == c;
 }
diff --git a/tests/compiler/integers/i64/ne.leo b/tests/compiler/integers/i64/ne.leo
index 162ef26bea..18bbce98ac 100644
--- a/tests/compiler/integers/i64/ne.leo
+++ b/tests/compiler/integers/i64/ne.leo
@@ -21,5 +21,5 @@ inputs:
 */
 
 function main(a: i64, b: i64, c: bool) -> bool{
-    return a != b == c;
+    return (a != b) == c;
 }
diff --git a/tests/compiler/integers/i64/negate_min.leo b/tests/compiler/integers/i64/negate_min.leo
index 547ede816e..df00c0b9b1 100644
--- a/tests/compiler/integers/i64/negate_min.leo
+++ b/tests/compiler/integers/i64/negate_min.leo
@@ -7,5 +7,5 @@ input_file: ../input/dummy.in
 function main(y: bool) -> bool{
     const a = -128i64;
     const b = -a;
-    return b == -a == y;
+    return (b == -a) == y;
 }
diff --git a/tests/compiler/integers/i64/negate_zero.leo b/tests/compiler/integers/i64/negate_zero.leo
index 1af21e8f01..dbc6bceaf1 100644
--- a/tests/compiler/integers/i64/negate_zero.leo
+++ b/tests/compiler/integers/i64/negate_zero.leo
@@ -13,5 +13,5 @@ inputs:
 function main(y: bool) -> bool {
     const a = 0i64;
 
-    return -a == 0i64 == y;
+    return (-a == 0i64) == y;
 }
diff --git a/tests/compiler/integers/i8/eq.leo b/tests/compiler/integers/i8/eq.leo
index 29a2e0492b..bf39311411 100644
--- a/tests/compiler/integers/i8/eq.leo
+++ b/tests/compiler/integers/i8/eq.leo
@@ -13,5 +13,5 @@ inputs:
 */
 
 function main(a: i8, b: i8, c: bool) -> bool {
-    return a == b == c;
+    return (a == b) == c;
 }
diff --git a/tests/compiler/integers/i8/ne.leo b/tests/compiler/integers/i8/ne.leo
index 8f6349ace2..cdb4519a43 100644
--- a/tests/compiler/integers/i8/ne.leo
+++ b/tests/compiler/integers/i8/ne.leo
@@ -21,5 +21,5 @@ inputs:
 */
 
 function main(a: i8, b: i8, c: bool) -> bool{
-    return a != b == c;
+    return (a != b) == c;
 }
diff --git a/tests/compiler/integers/i8/negate_zero.leo b/tests/compiler/integers/i8/negate_zero.leo
index a115c2cb09..6e027462af 100644
--- a/tests/compiler/integers/i8/negate_zero.leo
+++ b/tests/compiler/integers/i8/negate_zero.leo
@@ -13,5 +13,5 @@ inputs:
 function main(y: bool) -> bool {
     const a = 0i8;
 
-    return -a == 0i8 == y;
+    return (-a == 0i8) == y;
 }
diff --git a/tests/compiler/integers/u128/eq.leo b/tests/compiler/integers/u128/eq.leo
index c850875671..60a2e3bc10 100644
--- a/tests/compiler/integers/u128/eq.leo
+++ b/tests/compiler/integers/u128/eq.leo
@@ -21,5 +21,5 @@ inputs:
 */
 
 function main(a: u128, b: u128, c: bool) -> bool {
-    return a == b == c;
+    return (a == b) == c;
 }
diff --git a/tests/compiler/integers/u128/ne.leo b/tests/compiler/integers/u128/ne.leo
index 4299cc3a3c..cc23e5c55b 100644
--- a/tests/compiler/integers/u128/ne.leo
+++ b/tests/compiler/integers/u128/ne.leo
@@ -13,5 +13,5 @@ inputs:
 */
 
 function main(a: u128, b: u128, c: bool) -> bool {
-    return a != b == c;
+    return (a != b) == c;
 }
diff --git a/tests/compiler/integers/u16/eq.leo b/tests/compiler/integers/u16/eq.leo
index 4f715a1706..ab93a79602 100644
--- a/tests/compiler/integers/u16/eq.leo
+++ b/tests/compiler/integers/u16/eq.leo
@@ -21,5 +21,5 @@ inputs:
 */
 
 function main(a: u16, b: u16, c: bool) -> bool {
-    return a == b == c;
+    return (a == b) == c;
 }
diff --git a/tests/compiler/integers/u16/ne.leo b/tests/compiler/integers/u16/ne.leo
index 823557a6e5..93a762eaf1 100644
--- a/tests/compiler/integers/u16/ne.leo
+++ b/tests/compiler/integers/u16/ne.leo
@@ -13,5 +13,5 @@ inputs:
 */
 
 function main(a: u16, b: u16, c: bool) -> bool {
-    return a != b == c;
+    return (a != b) == c;
 }
diff --git a/tests/compiler/integers/u32/eq.leo b/tests/compiler/integers/u32/eq.leo
index 38720ddd9d..8edb4b5093 100644
--- a/tests/compiler/integers/u32/eq.leo
+++ b/tests/compiler/integers/u32/eq.leo
@@ -21,5 +21,5 @@ inputs:
 */
 
 function main(a: u32, b: u32, c: bool) -> bool {
-    return a == b == c;
+    return (a == b) == c;
 }
diff --git a/tests/compiler/integers/u32/ne.leo b/tests/compiler/integers/u32/ne.leo
index 2df0af5399..e8b2124338 100644
--- a/tests/compiler/integers/u32/ne.leo
+++ b/tests/compiler/integers/u32/ne.leo
@@ -13,5 +13,5 @@ inputs:
 */
 
 function main(a: u32, b: u32, c: bool) -> bool {
-    return a != b == c;
+    return (a != b) == c;
 }
diff --git a/tests/compiler/integers/u64/eq.leo b/tests/compiler/integers/u64/eq.leo
index 0a6f49d7d6..ea7747beaa 100644
--- a/tests/compiler/integers/u64/eq.leo
+++ b/tests/compiler/integers/u64/eq.leo
@@ -21,5 +21,5 @@ inputs:
 */
 
 function main(a: u64, b: u64, c: bool) -> bool {
-    return a == b == c;
+    return (a == b) == c;
 }
diff --git a/tests/compiler/integers/u64/ne.leo b/tests/compiler/integers/u64/ne.leo
index 1cd832ce16..3c3d8c7461 100644
--- a/tests/compiler/integers/u64/ne.leo
+++ b/tests/compiler/integers/u64/ne.leo
@@ -13,5 +13,5 @@ inputs:
 */
 
 function main(a: u64, b: u64, c: bool) -> bool {
-    return a != b == c;
+    return (a != b) == c;
 }
diff --git a/tests/compiler/integers/u8/eq.leo b/tests/compiler/integers/u8/eq.leo
index d73a7ab9ca..71a24c69fc 100644
--- a/tests/compiler/integers/u8/eq.leo
+++ b/tests/compiler/integers/u8/eq.leo
@@ -21,5 +21,5 @@ inputs:
 */
 
 function main(a: u8, b: u8, c: bool) -> bool {
-    return a == b == c;
+    return (a == b) == c;
 }
diff --git a/tests/compiler/integers/u8/ne.leo b/tests/compiler/integers/u8/ne.leo
index d8c635ba6a..8c41c9a611 100644
--- a/tests/compiler/integers/u8/ne.leo
+++ b/tests/compiler/integers/u8/ne.leo
@@ -13,5 +13,5 @@ inputs:
 */
 
 function main(a: u8, b: u8, c: bool) -> bool {
-    return a != b == c;
+    return (a != b) == c;
 }
diff --git a/tests/compiler/mutability/array_mut.leo b/tests/compiler/mutability/array_mut.leo
index fb69d10acc..3d2fc40d6f 100644
--- a/tests/compiler/mutability/array_mut.leo
+++ b/tests/compiler/mutability/array_mut.leo
@@ -8,5 +8,5 @@ function main(y: bool) -> bool {
     let a = [1u32];
     a[0] = 0;
 
-    return a[0] == 0u32 == y;
+    return (a[0] == 0u32) == y;
 }
diff --git a/tests/compiler/mutability/circuit_function_mut.leo b/tests/compiler/mutability/circuit_function_mut.leo
index 466d44fb7f..5321f0203c 100644
--- a/tests/compiler/mutability/circuit_function_mut.leo
+++ b/tests/compiler/mutability/circuit_function_mut.leo
@@ -18,5 +18,5 @@ function main(y: bool) -> bool {
     
     a.foo();
 
-    return a.x == 10u32 == y;
+    return (a.x == 10u32) == y;
 }
diff --git a/tests/compiler/mutability/circuit_variable_mut.leo b/tests/compiler/mutability/circuit_variable_mut.leo
index 1933bd4746..20ebe63f9e 100644
--- a/tests/compiler/mutability/circuit_variable_mut.leo
+++ b/tests/compiler/mutability/circuit_variable_mut.leo
@@ -13,5 +13,5 @@ function main(y: bool) -> bool{
     let a = Foo { x: 1 };
     a.x = 0;
 
-    return a.x == 0u32 == y;
+    return (a.x == 0u32) == y;
 }
diff --git a/tests/compiler/mutability/cond_mut.leo b/tests/compiler/mutability/cond_mut.leo
index 5626cafccc..1cac6d876d 100644
--- a/tests/compiler/mutability/cond_mut.leo
+++ b/tests/compiler/mutability/cond_mut.leo
@@ -11,5 +11,5 @@ function main(y: bool) -> bool {
         x *= 100i8;
     }
     
-    return x == 100i8 == y;
+    return (x == 100i8) == y;
 }
diff --git a/tests/compiler/mutability/let_mut.leo b/tests/compiler/mutability/let_mut.leo
index 804ca80f71..7c303ac0d4 100644
--- a/tests/compiler/mutability/let_mut.leo
+++ b/tests/compiler/mutability/let_mut.leo
@@ -9,5 +9,5 @@ function main(y: bool) -> bool {
     let a = 1u32;
     a = 0;
 
-    return a == 0u32 == y;
+    return (a == 0u32) == y;
 }
diff --git a/tests/compiler/mutability/let_mut_nested.leo b/tests/compiler/mutability/let_mut_nested.leo
index d73c3b4505..98d1293211 100644
--- a/tests/compiler/mutability/let_mut_nested.leo
+++ b/tests/compiler/mutability/let_mut_nested.leo
@@ -8,5 +8,5 @@ function main(a: bool) -> bool {
     let x = 2u8;
     let y = x;
     let z = y / 2u8;
-    return z == 1 == a;
+    return (z == 1) == a;
 }
diff --git a/tests/compiler/tuples/access.leo b/tests/compiler/tuples/access.leo
index c75498876d..07b6585022 100644
--- a/tests/compiler/tuples/access.leo
+++ b/tests/compiler/tuples/access.leo
@@ -9,5 +9,5 @@ function main(a: (bool, bool)) -> (bool, bool) {
     let c = (false, true, false).1;
     let d = (true, false, false).2;
 
-    return (a.0 == b, a.1 == c == d);
+    return (a.0 == b, (a.1 == c) == d);
 }
diff --git a/tests/expectations/parser/parser/expression/binary/eq.leo.out b/tests/expectations/parser/parser/expression/binary/eq.leo.out
index 56f7df069c..a467a62bae 100644
--- a/tests/expectations/parser/parser/expression/binary/eq.leo.out
+++ b/tests/expectations/parser/parser/expression/binary/eq.leo.out
@@ -60,55 +60,6 @@ outputs:
         col_stop: 5
         path: test
         content: 2==3
-  - Binary:
-      left:
-        Binary:
-          left:
-            Value:
-              Implicit:
-                - "1"
-                - line_start: 1
-                  line_stop: 1
-                  col_start: 1
-                  col_stop: 2
-                  path: test
-                  content: 1 == 2 == 3
-          right:
-            Value:
-              Implicit:
-                - "2"
-                - line_start: 1
-                  line_stop: 1
-                  col_start: 6
-                  col_stop: 7
-                  path: test
-                  content: 1 == 2 == 3
-          op: Eq
-          span:
-            line_start: 1
-            line_stop: 1
-            col_start: 1
-            col_stop: 7
-            path: test
-            content: 1 == 2 == 3
-      right:
-        Value:
-          Implicit:
-            - "3"
-            - line_start: 1
-              line_stop: 1
-              col_start: 11
-              col_stop: 12
-              path: test
-              content: 1 == 2 == 3
-      op: Eq
-      span:
-        line_start: 1
-        line_stop: 1
-        col_start: 1
-        col_stop: 12
-        path: test
-        content: 1 == 2 == 3
   - Binary:
       left:
         Binary:
@@ -178,55 +129,6 @@ outputs:
         col_stop: 15
         path: test
         content: 1 < 2 == 3 < 4
-  - Binary:
-      left:
-        Binary:
-          left:
-            Value:
-              Implicit:
-                - "1"
-                - line_start: 1
-                  line_stop: 1
-                  col_start: 1
-                  col_stop: 2
-                  path: test
-                  content: 1 == 2 == 3
-          right:
-            Value:
-              Implicit:
-                - "2"
-                - line_start: 1
-                  line_stop: 1
-                  col_start: 6
-                  col_stop: 7
-                  path: test
-                  content: 1 == 2 == 3
-          op: Eq
-          span:
-            line_start: 1
-            line_stop: 1
-            col_start: 1
-            col_stop: 7
-            path: test
-            content: 1 == 2 == 3
-      right:
-        Value:
-          Implicit:
-            - "3"
-            - line_start: 1
-              line_stop: 1
-              col_start: 11
-              col_stop: 12
-              path: test
-              content: 1 == 2 == 3
-      op: Eq
-      span:
-        line_start: 1
-        line_stop: 1
-        col_start: 1
-        col_stop: 12
-        path: test
-        content: 1 == 2 == 3
   - Binary:
       left:
         Binary:
@@ -241,7 +143,7 @@ outputs:
                       col_start: 1
                       col_stop: 2
                       path: test
-                      content: 1 < 2 == 3 < 4 == 5 < 6
+                      content: 1 < 2 == 3 < 4 && 3 < 4 == 5 < 6
               right:
                 Value:
                   Implicit:
@@ -251,7 +153,7 @@ outputs:
                       col_start: 5
                       col_stop: 6
                       path: test
-                      content: 1 < 2 == 3 < 4 == 5 < 6
+                      content: 1 < 2 == 3 < 4 && 3 < 4 == 5 < 6
               op: Lt
               span:
                 line_start: 1
@@ -259,7 +161,7 @@ outputs:
                 col_start: 1
                 col_stop: 6
                 path: test
-                content: 1 < 2 == 3 < 4 == 5 < 6
+                content: 1 < 2 == 3 < 4 && 3 < 4 == 5 < 6
           right:
             Binary:
               left:
@@ -271,7 +173,7 @@ outputs:
                       col_start: 10
                       col_stop: 11
                       path: test
-                      content: 1 < 2 == 3 < 4 == 5 < 6
+                      content: 1 < 2 == 3 < 4 && 3 < 4 == 5 < 6
               right:
                 Value:
                   Implicit:
@@ -281,7 +183,7 @@ outputs:
                       col_start: 14
                       col_stop: 15
                       path: test
-                      content: 1 < 2 == 3 < 4 == 5 < 6
+                      content: 1 < 2 == 3 < 4 && 3 < 4 == 5 < 6
               op: Lt
               span:
                 line_start: 1
@@ -289,7 +191,7 @@ outputs:
                 col_start: 10
                 col_stop: 15
                 path: test
-                content: 1 < 2 == 3 < 4 == 5 < 6
+                content: 1 < 2 == 3 < 4 && 3 < 4 == 5 < 6
           op: Eq
           span:
             line_start: 1
@@ -297,42 +199,82 @@ outputs:
             col_start: 1
             col_stop: 15
             path: test
-            content: 1 < 2 == 3 < 4 == 5 < 6
+            content: 1 < 2 == 3 < 4 && 3 < 4 == 5 < 6
       right:
         Binary:
           left:
-            Value:
-              Implicit:
-                - "5"
-                - line_start: 1
-                  line_stop: 1
-                  col_start: 19
-                  col_stop: 20
-                  path: test
-                  content: 1 < 2 == 3 < 4 == 5 < 6
+            Binary:
+              left:
+                Value:
+                  Implicit:
+                    - "3"
+                    - line_start: 1
+                      line_stop: 1
+                      col_start: 19
+                      col_stop: 20
+                      path: test
+                      content: 1 < 2 == 3 < 4 && 3 < 4 == 5 < 6
+              right:
+                Value:
+                  Implicit:
+                    - "4"
+                    - line_start: 1
+                      line_stop: 1
+                      col_start: 23
+                      col_stop: 24
+                      path: test
+                      content: 1 < 2 == 3 < 4 && 3 < 4 == 5 < 6
+              op: Lt
+              span:
+                line_start: 1
+                line_stop: 1
+                col_start: 19
+                col_stop: 24
+                path: test
+                content: 1 < 2 == 3 < 4 && 3 < 4 == 5 < 6
           right:
-            Value:
-              Implicit:
-                - "6"
-                - line_start: 1
-                  line_stop: 1
-                  col_start: 23
-                  col_stop: 24
-                  path: test
-                  content: 1 < 2 == 3 < 4 == 5 < 6
-          op: Lt
+            Binary:
+              left:
+                Value:
+                  Implicit:
+                    - "5"
+                    - line_start: 1
+                      line_stop: 1
+                      col_start: 28
+                      col_stop: 29
+                      path: test
+                      content: 1 < 2 == 3 < 4 && 3 < 4 == 5 < 6
+              right:
+                Value:
+                  Implicit:
+                    - "6"
+                    - line_start: 1
+                      line_stop: 1
+                      col_start: 32
+                      col_stop: 33
+                      path: test
+                      content: 1 < 2 == 3 < 4 && 3 < 4 == 5 < 6
+              op: Lt
+              span:
+                line_start: 1
+                line_stop: 1
+                col_start: 28
+                col_stop: 33
+                path: test
+                content: 1 < 2 == 3 < 4 && 3 < 4 == 5 < 6
+          op: Eq
           span:
             line_start: 1
             line_stop: 1
             col_start: 19
-            col_stop: 24
+            col_stop: 33
             path: test
-            content: 1 < 2 == 3 < 4 == 5 < 6
-      op: Eq
+            content: 1 < 2 == 3 < 4 && 3 < 4 == 5 < 6
+      op: And
       span:
         line_start: 1
         line_stop: 1
         col_start: 1
-        col_stop: 24
+        col_stop: 33
         path: test
-        content: 1 < 2 == 3 < 4 == 5 < 6
+        content: 1 < 2 == 3 < 4 && 3 < 4 == 5 < 6
diff --git a/tests/expectations/parser/parser/expression/binary/eq_fail.leo.out b/tests/expectations/parser/parser/expression/binary/eq_fail.leo.out
new file mode 100644
index 0000000000..36645239b3
--- /dev/null
+++ b/tests/expectations/parser/parser/expression/binary/eq_fail.leo.out
@@ -0,0 +1,5 @@
+---
+namespace: ParseExpression
+expectation: Fail
+outputs:
+  - "did not consume all input: '==' @ 1:8-10\n'3' @ 1:11-12\n"
diff --git a/tests/expectations/parser/parser/expression/binary/ne.leo.out b/tests/expectations/parser/parser/expression/binary/ne.leo.out
index 65fb6e2298..9e1db5c244 100644
--- a/tests/expectations/parser/parser/expression/binary/ne.leo.out
+++ b/tests/expectations/parser/parser/expression/binary/ne.leo.out
@@ -60,55 +60,6 @@ outputs:
         col_stop: 5
         path: test
         content: 2!=3
-  - Binary:
-      left:
-        Binary:
-          left:
-            Value:
-              Implicit:
-                - "1"
-                - line_start: 1
-                  line_stop: 1
-                  col_start: 1
-                  col_stop: 2
-                  path: test
-                  content: 1 != 2 != 3
-          right:
-            Value:
-              Implicit:
-                - "2"
-                - line_start: 1
-                  line_stop: 1
-                  col_start: 6
-                  col_stop: 7
-                  path: test
-                  content: 1 != 2 != 3
-          op: Ne
-          span:
-            line_start: 1
-            line_stop: 1
-            col_start: 1
-            col_stop: 7
-            path: test
-            content: 1 != 2 != 3
-      right:
-        Value:
-          Implicit:
-            - "3"
-            - line_start: 1
-              line_stop: 1
-              col_start: 11
-              col_stop: 12
-              path: test
-              content: 1 != 2 != 3
-      op: Ne
-      span:
-        line_start: 1
-        line_stop: 1
-        col_start: 1
-        col_stop: 12
-        path: test
-        content: 1 != 2 != 3
   - Binary:
       left:
         Binary:
@@ -178,55 +129,6 @@ outputs:
         col_stop: 15
         path: test
         content: 1 < 2 != 3 < 4
-  - Binary:
-      left:
-        Binary:
-          left:
-            Value:
-              Implicit:
-                - "1"
-                - line_start: 1
-                  line_stop: 1
-                  col_start: 1
-                  col_stop: 2
-                  path: test
-                  content: 1 != 2 != 3
-          right:
-            Value:
-              Implicit:
-                - "2"
-                - line_start: 1
-                  line_stop: 1
-                  col_start: 6
-                  col_stop: 7
-                  path: test
-                  content: 1 != 2 != 3
-          op: Ne
-          span:
-            line_start: 1
-            line_stop: 1
-            col_start: 1
-            col_stop: 7
-            path: test
-            content: 1 != 2 != 3
-      right:
-        Value:
-          Implicit:
-            - "3"
-            - line_start: 1
-              line_stop: 1
-              col_start: 11
-              col_stop: 12
-              path: test
-              content: 1 != 2 != 3
-      op: Ne
-      span:
-        line_start: 1
-        line_stop: 1
-        col_start: 1
-        col_stop: 12
-        path: test
-        content: 1 != 2 != 3
   - Binary:
       left:
         Binary:
@@ -241,7 +143,7 @@ outputs:
                       col_start: 1
                       col_stop: 2
                       path: test
-                      content: 1 < 2 != 3 < 4 != 5 < 6
+                      content: 1 < 2 != 3 < 4 || 3 < 4 != 5 < 6
               right:
                 Value:
                   Implicit:
@@ -251,7 +153,7 @@ outputs:
                       col_start: 5
                       col_stop: 6
                       path: test
-                      content: 1 < 2 != 3 < 4 != 5 < 6
+                      content: 1 < 2 != 3 < 4 || 3 < 4 != 5 < 6
               op: Lt
               span:
                 line_start: 1
@@ -259,7 +161,7 @@ outputs:
                 col_start: 1
                 col_stop: 6
                 path: test
-                content: 1 < 2 != 3 < 4 != 5 < 6
+                content: 1 < 2 != 3 < 4 || 3 < 4 != 5 < 6
           right:
             Binary:
               left:
@@ -271,7 +173,7 @@ outputs:
                       col_start: 10
                       col_stop: 11
                       path: test
-                      content: 1 < 2 != 3 < 4 != 5 < 6
+                      content: 1 < 2 != 3 < 4 || 3 < 4 != 5 < 6
               right:
                 Value:
                   Implicit:
@@ -281,7 +183,7 @@ outputs:
                       col_start: 14
                       col_stop: 15
                       path: test
-                      content: 1 < 2 != 3 < 4 != 5 < 6
+                      content: 1 < 2 != 3 < 4 || 3 < 4 != 5 < 6
               op: Lt
               span:
                 line_start: 1
@@ -289,7 +191,7 @@ outputs:
                 col_start: 10
                 col_stop: 15
                 path: test
-                content: 1 < 2 != 3 < 4 != 5 < 6
+                content: 1 < 2 != 3 < 4 || 3 < 4 != 5 < 6
           op: Ne
           span:
             line_start: 1
@@ -297,42 +199,82 @@ outputs:
             col_start: 1
             col_stop: 15
             path: test
-            content: 1 < 2 != 3 < 4 != 5 < 6
+            content: 1 < 2 != 3 < 4 || 3 < 4 != 5 < 6
       right:
         Binary:
           left:
-            Value:
-              Implicit:
-                - "5"
-                - line_start: 1
-                  line_stop: 1
-                  col_start: 19
-                  col_stop: 20
-                  path: test
-                  content: 1 < 2 != 3 < 4 != 5 < 6
+            Binary:
+              left:
+                Value:
+                  Implicit:
+                    - "3"
+                    - line_start: 1
+                      line_stop: 1
+                      col_start: 19
+                      col_stop: 20
+                      path: test
+                      content: 1 < 2 != 3 < 4 || 3 < 4 != 5 < 6
+              right:
+                Value:
+                  Implicit:
+                    - "4"
+                    - line_start: 1
+                      line_stop: 1
+                      col_start: 23
+                      col_stop: 24
+                      path: test
+                      content: 1 < 2 != 3 < 4 || 3 < 4 != 5 < 6
+              op: Lt
+              span:
+                line_start: 1
+                line_stop: 1
+                col_start: 19
+                col_stop: 24
+                path: test
+                content: 1 < 2 != 3 < 4 || 3 < 4 != 5 < 6
           right:
-            Value:
-              Implicit:
-                - "6"
-                - line_start: 1
-                  line_stop: 1
-                  col_start: 23
-                  col_stop: 24
-                  path: test
-                  content: 1 < 2 != 3 < 4 != 5 < 6
-          op: Lt
+            Binary:
+              left:
+                Value:
+                  Implicit:
+                    - "5"
+                    - line_start: 1
+                      line_stop: 1
+                      col_start: 28
+                      col_stop: 29
+                      path: test
+                      content: 1 < 2 != 3 < 4 || 3 < 4 != 5 < 6
+              right:
+                Value:
+                  Implicit:
+                    - "6"
+                    - line_start: 1
+                      line_stop: 1
+                      col_start: 32
+                      col_stop: 33
+                      path: test
+                      content: 1 < 2 != 3 < 4 || 3 < 4 != 5 < 6
+              op: Lt
+              span:
+                line_start: 1
+                line_stop: 1
+                col_start: 28
+                col_stop: 33
+                path: test
+                content: 1 < 2 != 3 < 4 || 3 < 4 != 5 < 6
+          op: Ne
           span:
             line_start: 1
             line_stop: 1
             col_start: 19
-            col_stop: 24
+            col_stop: 33
             path: test
-            content: 1 < 2 != 3 < 4 != 5 < 6
-      op: Ne
+            content: 1 < 2 != 3 < 4 || 3 < 4 != 5 < 6
+      op: Or
       span:
         line_start: 1
         line_stop: 1
         col_start: 1
-        col_stop: 24
+        col_stop: 33
         path: test
-        content: 1 < 2 != 3 < 4 != 5 < 6
+        content: 1 < 2 != 3 < 4 || 3 < 4 != 5 < 6
diff --git a/tests/expectations/parser/parser/expression/binary/ne_fail.leo.out b/tests/expectations/parser/parser/expression/binary/ne_fail.leo.out
new file mode 100644
index 0000000000..14fbe7278e
--- /dev/null
+++ b/tests/expectations/parser/parser/expression/binary/ne_fail.leo.out
@@ -0,0 +1,5 @@
+---
+namespace: ParseExpression
+expectation: Fail
+outputs:
+  - "did not consume all input: '!=' @ 1:8-10\n'3' @ 1:11-12\n"
diff --git a/tests/expectations/parser/parser/functions/annotated_fail.leo.out b/tests/expectations/parser/parser/functions/annotated_fail.leo.out
new file mode 100644
index 0000000000..b4b2bc739d
--- /dev/null
+++ b/tests/expectations/parser/parser/functions/annotated_fail.leo.out
@@ -0,0 +1,5 @@
+---
+namespace: Parse
+expectation: Fail
+outputs:
+  - "    --> test:3:12\n     |\n   3 | @test(test,)\n     |            ^\n     |\n     = expected 'identifier', 'number' -- got ')'"
diff --git a/tests/expectations/parser/parser/import/invalid.leo.out b/tests/expectations/parser/parser/import/invalid.leo.out
new file mode 100644
index 0000000000..941b1887cd
--- /dev/null
+++ b/tests/expectations/parser/parser/import/invalid.leo.out
@@ -0,0 +1,5 @@
+---
+namespace: Parse
+expectation: Fail
+outputs:
+  - "    --> test:3:18\n     |\n   3 | import foo as bar;\n     |                  ^\n     |\n     = expected '.' -- got ';'"
diff --git a/tests/expectations/parser/parser/statement/conditional.leo.out b/tests/expectations/parser/parser/statement/conditional.leo.out
index 337dcace01..c29e4db887 100644
--- a/tests/expectations/parser/parser/statement/conditional.leo.out
+++ b/tests/expectations/parser/parser/statement/conditional.leo.out
@@ -118,7 +118,7 @@ outputs:
         content: "if (x) {\n...\n}"
   - Conditional:
       condition:
-        Identifier: "{\"name\":\"x\",\"span\":\"{\\\"line_start\\\":1,\\\"line_stop\\\":1,\\\"col_start\\\":5,\\\"col_stop\\\":6,\\\"path\\\":\\\"test\\\",\\\"content\\\":\\\"if (x) {}\\\"}\"}"
+        Identifier: "{\"name\":\"x\",\"span\":\"{\\\"line_start\\\":1,\\\"line_stop\\\":1,\\\"col_start\\\":5,\\\"col_stop\\\":6,\\\"path\\\":\\\"test\\\",\\\"content\\\":\\\"if (x) {} else {}\\\"}\"}"
       block:
         statements: []
         span:
@@ -127,22 +127,31 @@ outputs:
           col_start: 8
           col_stop: 10
           path: test
-          content: "if (x) {}"
-      next: ~
+          content: "if (x) {} else {}"
+      next:
+        Block:
+          statements: []
+          span:
+            line_start: 1
+            line_stop: 1
+            col_start: 16
+            col_stop: 18
+            path: test
+            content: "if (x) {} else {}"
       span:
         line_start: 1
         line_stop: 1
         col_start: 1
-        col_stop: 10
+        col_stop: 18
         path: test
-        content: "if (x) {}"
+        content: "if (x) {} else {}"
   - Conditional:
       condition:
         Binary:
           left:
-            Identifier: "{\"name\":\"x\",\"span\":\"{\\\"line_start\\\":1,\\\"line_stop\\\":1,\\\"col_start\\\":4,\\\"col_stop\\\":5,\\\"path\\\":\\\"test\\\",\\\"content\\\":\\\"if x+y {}\\\"}\"}"
+            Identifier: "{\"name\":\"x\",\"span\":\"{\\\"line_start\\\":1,\\\"line_stop\\\":1,\\\"col_start\\\":4,\\\"col_stop\\\":5,\\\"path\\\":\\\"test\\\",\\\"content\\\":\\\"if x+y {} else if x+z {} else {}\\\"}\"}"
           right:
-            Identifier: "{\"name\":\"y\",\"span\":\"{\\\"line_start\\\":1,\\\"line_stop\\\":1,\\\"col_start\\\":6,\\\"col_stop\\\":7,\\\"path\\\":\\\"test\\\",\\\"content\\\":\\\"if x+y {}\\\"}\"}"
+            Identifier: "{\"name\":\"y\",\"span\":\"{\\\"line_start\\\":1,\\\"line_stop\\\":1,\\\"col_start\\\":6,\\\"col_stop\\\":7,\\\"path\\\":\\\"test\\\",\\\"content\\\":\\\"if x+y {} else if x+z {} else {}\\\"}\"}"
           op: Add
           span:
             line_start: 1
@@ -150,7 +159,7 @@ outputs:
             col_start: 4
             col_stop: 7
             path: test
-            content: "if x+y {}"
+            content: "if x+y {} else if x+z {} else {}"
       block:
         statements: []
         span:
@@ -159,15 +168,56 @@ outputs:
           col_start: 8
           col_stop: 10
           path: test
-          content: "if x+y {}"
-      next: ~
+          content: "if x+y {} else if x+z {} else {}"
+      next:
+        Conditional:
+          condition:
+            Binary:
+              left:
+                Identifier: "{\"name\":\"x\",\"span\":\"{\\\"line_start\\\":1,\\\"line_stop\\\":1,\\\"col_start\\\":19,\\\"col_stop\\\":20,\\\"path\\\":\\\"test\\\",\\\"content\\\":\\\"if x+y {} else if x+z {} else {}\\\"}\"}"
+              right:
+                Identifier: "{\"name\":\"z\",\"span\":\"{\\\"line_start\\\":1,\\\"line_stop\\\":1,\\\"col_start\\\":21,\\\"col_stop\\\":22,\\\"path\\\":\\\"test\\\",\\\"content\\\":\\\"if x+y {} else if x+z {} else {}\\\"}\"}"
+              op: Add
+              span:
+                line_start: 1
+                line_stop: 1
+                col_start: 19
+                col_stop: 22
+                path: test
+                content: "if x+y {} else if x+z {} else {}"
+          block:
+            statements: []
+            span:
+              line_start: 1
+              line_stop: 1
+              col_start: 23
+              col_stop: 25
+              path: test
+              content: "if x+y {} else if x+z {} else {}"
+          next:
+            Block:
+              statements: []
+              span:
+                line_start: 1
+                line_stop: 1
+                col_start: 31
+                col_stop: 33
+                path: test
+                content: "if x+y {} else if x+z {} else {}"
+          span:
+            line_start: 1
+            line_stop: 1
+            col_start: 16
+            col_stop: 33
+            path: test
+            content: "if x+y {} else if x+z {} else {}"
       span:
         line_start: 1
         line_stop: 1
         col_start: 1
-        col_stop: 10
+        col_stop: 33
         path: test
-        content: "if x+y {}"
+        content: "if x+y {} else if x+z {} else {}"
   - Conditional:
       condition:
         Binary:
diff --git a/tests/expectations/parser/parser/statement/conditional_fail.leo.out b/tests/expectations/parser/parser/statement/conditional_fail.leo.out
new file mode 100644
index 0000000000..c54b2c1841
--- /dev/null
+++ b/tests/expectations/parser/parser/statement/conditional_fail.leo.out
@@ -0,0 +1,5 @@
+---
+namespace: ParseStatement
+expectation: Fail
+outputs:
+  - "    --> test:1:17\n     |\n   1 | if true {} else let x = 2;\n     |                 ^^^^^^^^^\n     |\n     = expected 'Block or Conditional', got 'let mut x = 2;'"
diff --git a/tests/parser/expression/binary/eq.leo b/tests/parser/expression/binary/eq.leo
index d84b5f0738..a7fb0bd341 100644
--- a/tests/parser/expression/binary/eq.leo
+++ b/tests/parser/expression/binary/eq.leo
@@ -7,10 +7,6 @@ expectation: Pass
 
 2==3
 
-1 == 2 == 3
-
 1 < 2 == 3 < 4
 
-1 == 2 == 3
-
-1 < 2 == 3 < 4 == 5 < 6
\ No newline at end of file
+1 < 2 == 3 < 4 && 3 < 4 == 5 < 6
\ No newline at end of file
diff --git a/tests/parser/expression/binary/eq_fail.leo b/tests/parser/expression/binary/eq_fail.leo
new file mode 100644
index 0000000000..d51df10f42
--- /dev/null
+++ b/tests/parser/expression/binary/eq_fail.leo
@@ -0,0 +1,6 @@
+/*
+namespace: ParseExpression
+expectation: Fail
+*/
+
+1 == 2 == 3
\ No newline at end of file
diff --git a/tests/parser/expression/binary/ne.leo b/tests/parser/expression/binary/ne.leo
index 46b4e8cdf6..bde0899778 100644
--- a/tests/parser/expression/binary/ne.leo
+++ b/tests/parser/expression/binary/ne.leo
@@ -7,10 +7,6 @@ expectation: Pass
 
 2!=3
 
-1 != 2 != 3
-
 1 < 2 != 3 < 4
 
-1 != 2 != 3
-
-1 < 2 != 3 < 4 != 5 < 6
\ No newline at end of file
+1 < 2 != 3 < 4 || 3 < 4 != 5 < 6
\ No newline at end of file
diff --git a/tests/parser/expression/binary/ne_fail.leo b/tests/parser/expression/binary/ne_fail.leo
new file mode 100644
index 0000000000..a96d955cb2
--- /dev/null
+++ b/tests/parser/expression/binary/ne_fail.leo
@@ -0,0 +1,6 @@
+/*
+namespace: ParseExpression
+expectation: Fail
+*/
+
+1 != 2 != 3
\ No newline at end of file
diff --git a/tests/parser/functions/annotated_fail.leo b/tests/parser/functions/annotated_fail.leo
new file mode 100644
index 0000000000..2f21d6171c
--- /dev/null
+++ b/tests/parser/functions/annotated_fail.leo
@@ -0,0 +1,9 @@
+/*
+namespace: Parse
+expectation: Fail
+*/
+
+@test(test,)
+function x() {
+    return ();
+}
\ No newline at end of file
diff --git a/tests/parser/import/invalid.leo b/tests/parser/import/invalid.leo
new file mode 100644
index 0000000000..245ef87b12
--- /dev/null
+++ b/tests/parser/import/invalid.leo
@@ -0,0 +1,7 @@
+/*
+namespace: Parse
+expectation: Fail
+*/
+
+import foo as bar;
+import *;
\ No newline at end of file
diff --git a/tests/parser/statement/conditional.leo b/tests/parser/statement/conditional.leo
index 2605521ae1..62a9e7f54e 100644
--- a/tests/parser/statement/conditional.leo
+++ b/tests/parser/statement/conditional.leo
@@ -15,9 +15,9 @@ if (x) {
     return ();
 }
 
-if (x) {}
+if (x) {} else {}
 
-if x+y {}
+if x+y {} else if x+z {} else {}
 
 if x+y {
     expr;
diff --git a/tests/parser/statement/conditional_fail.leo b/tests/parser/statement/conditional_fail.leo
new file mode 100644
index 0000000000..b1081556d5
--- /dev/null
+++ b/tests/parser/statement/conditional_fail.leo
@@ -0,0 +1,6 @@
+/*
+namespace: ParseStatement
+expectation: Fail
+*/
+
+if true {} else let x = 2;
\ No newline at end of file
