diff --git a/leo/commands/build.rs b/leo/commands/build.rs
index 66eb086895..cbf3fa09e1 100644
--- a/leo/commands/build.rs
+++ b/leo/commands/build.rs
@@ -27,7 +27,6 @@ use leo_span::symbol::with_session_globals;
 use aleo::commands::Build as AleoBuild;
 
 use clap::StructOpt;
-use colored::Colorize;
 use indexmap::IndexMap;
 use std::io::Write;
 use std::path::{Path, PathBuf};
@@ -183,6 +182,7 @@ impl Command for Build {
         if self.compiler_options.offline {
             args.push("--offline");
         }
+        println!();
         let command = AleoBuild::try_parse_from(&args).map_err(CliError::failed_to_execute_aleo_build)?;
         let result = command.parse().map_err(CliError::failed_to_execute_aleo_build)?;
 
@@ -275,14 +275,10 @@ fn compile_leo_file(
         .map_err(CliError::failed_to_load_instructions)?;
 
     // Prepare the path string.
-    let path_string = format!("(in \"{}\")", aleo_file_path.display());
+    let _path_string = format!("(in \"{}\")", aleo_file_path.display());
 
     // Log the build as successful.
-    tracing::info!(
-        "compiled '{}' into Aleo instructions {}",
-        file_name,
-        path_string.dimmed()
-    );
+    tracing::info!("Compiled '{}' into Aleo instructions", file_name,);
 
     Ok(symbol_table.circuits)
 }
diff --git a/leo/commands/deploy.rs b/leo/commands/deploy.rs
index a5a30d1373..5adabaa96f 100644
--- a/leo/commands/deploy.rs
+++ b/leo/commands/deploy.rs
@@ -50,6 +50,7 @@ impl Command for Deploy {
             .map_err(|err| PackageError::failed_to_set_cwd(build_directory.display(), err))?;
 
         // Call the `aleo node` command from the Aleo SDK.
+        println!();
         let command = AleoDeploy::try_parse_from([ALEO_CLI_COMMAND]).map_err(CliError::failed_to_parse_aleo_node)?;
         let res = command.parse().map_err(CliError::failed_to_execute_aleo_node)?;
 
diff --git a/leo/commands/node.rs b/leo/commands/node.rs
index 484ba7dfa7..146bdc965c 100644
--- a/leo/commands/node.rs
+++ b/leo/commands/node.rs
@@ -71,6 +71,7 @@ impl Command for Node {
         }
 
         // Call the `aleo node` command from the Aleo SDK.
+        println!();
         let command = AleoNode::try_parse_from(&arguments).map_err(CliError::failed_to_parse_aleo_node)?;
         let res = command.parse().map_err(CliError::failed_to_execute_aleo_node)?;
 
