diff --git a/.resources/license_header b/.resources/license_header
new file mode 100644
index 0000000000..6de9f3a6f4
--- /dev/null
+++ b/.resources/license_header
@@ -0,0 +1,15 @@
+// Copyright (C) 2019-2020 Aleo Systems Inc.
+// This file is part of the Leo library.
+
+// The Leo library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// The Leo library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with the Leo library. If not, see <https://www.gnu.org/licenses/>.
\ No newline at end of file
diff --git a/.rustfmt.toml b/.rustfmt.toml
index 18930656a9..e86fa4f991 100644
--- a/.rustfmt.toml
+++ b/.rustfmt.toml
@@ -9,6 +9,7 @@ use_try_shorthand = true
 
 # Nightly configurations
 imports_layout = "HorizontalVertical"
+license_template_path = ".resources/license_header"
 merge_imports = true
 overflow_delimited_expr = true
 reorder_impl_items = true
diff --git a/Cargo.lock b/Cargo.lock
index 72415a2769..4fdc01c88b 100644
--- a/Cargo.lock
+++ b/Cargo.lock
@@ -140,7 +140,7 @@ dependencies = [
  "cfg-if",
  "clang-sys",
  "clap",
- "env_logger",
+ "env_logger 0.7.1",
  "lazy_static",
  "lazycell",
  "log",
@@ -381,16 +381,15 @@ dependencies = [
 
 [[package]]
 name = "console"
-version = "0.12.0"
+version = "0.13.0"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "c0b1aacfaffdbff75be81c15a399b4bedf78aaefe840e8af1d299ac2ade885d2"
+checksum = "a50aab2529019abfabfa93f1e6c41ef392f91fbf179b347a7e96abb524884a08"
 dependencies = [
  "encode_unicode",
  "lazy_static",
  "libc",
  "regex",
  "terminal_size",
- "termios",
  "unicode-width",
  "winapi 0.3.9",
  "winapi-util",
@@ -667,7 +666,20 @@ source = "registry+https://github.com/rust-lang/crates.io-index"
 checksum = "44533bbbb3bb3c1fa17d9f2e4e38bbbaf8396ba82193c4cb1b6445d711445d36"
 dependencies = [
  "atty",
- "humantime",
+ "humantime 1.3.0",
+ "log",
+ "regex",
+ "termcolor",
+]
+
+[[package]]
+name = "env_logger"
+version = "0.8.1"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "54532e3223c5af90a6a757c90b5c5521564b07e5e7a958681bcd2afad421cdcd"
+dependencies = [
+ "atty",
+ "humantime 2.0.1",
  "log",
  "regex",
  "termcolor",
@@ -1000,6 +1012,12 @@ dependencies = [
  "quick-error",
 ]
 
+[[package]]
+name = "humantime"
+version = "2.0.1"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "3c1ad908cc71012b7bea4d0c53ba96a8cba9962f048fa68d143376143d863b7a"
+
 [[package]]
 name = "hyper"
 version = "0.13.7"
@@ -1279,6 +1297,7 @@ dependencies = [
 name = "leo-gadgets"
 version = "1.0.3"
 dependencies = [
+ "criterion",
  "rand",
  "rand_xorshift",
  "snarkos-errors",
@@ -1313,7 +1332,7 @@ dependencies = [
  "console",
  "dirs",
  "dotenv",
- "env_logger",
+ "env_logger 0.8.1",
  "from-pest",
  "lazy_static",
  "leo-compiler",
@@ -2266,9 +2285,9 @@ checksum = "388a1df253eca08550bef6c72392cfe7c30914bf41df5269b68cbd6ff8f570a3"
 
 [[package]]
 name = "serde"
-version = "1.0.116"
+version = "1.0.117"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "96fe57af81d28386a513cbc6858332abc6117cfdb5999647c6444b8f43a370a5"
+checksum = "b88fa983de7720629c9387e9f517353ed404164b1e482c970a90c1a4aaf7dc1a"
 dependencies = [
  "serde_derive",
 ]
@@ -2285,9 +2304,9 @@ dependencies = [
 
 [[package]]
 name = "serde_derive"
-version = "1.0.116"
+version = "1.0.117"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "f630a6370fd8e457873b4bd2ffdae75408bc291ba72be773772a4c2a065d9ae8"
+checksum = "cbd1ae72adb44aab48f325a02444a5fc079349a8d804c1fc922aed3f7454c74e"
 dependencies = [
  "proc-macro2 1.0.19",
  "quote 1.0.7",
@@ -2296,9 +2315,9 @@ dependencies = [
 
 [[package]]
 name = "serde_json"
-version = "1.0.57"
+version = "1.0.59"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "164eacbdb13512ec2745fb09d51fd5b22b0d65ed294a1dcf7285a360c80a675c"
+checksum = "dcac07dbffa1c65e7f816ab9eba78eb142c6d44410f4eeba1e26e4f5dfa56b95"
 dependencies = [
  "itoa",
  "ryu",
@@ -2663,15 +2682,6 @@ dependencies = [
  "winapi 0.3.9",
 ]
 
-[[package]]
-name = "termios"
-version = "0.3.2"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "6f0fcee7b24a25675de40d5bb4de6e41b0df07bc9856295e7e2b3a3600c400c2"
-dependencies = [
- "libc",
-]
-
 [[package]]
 name = "textwrap"
 version = "0.11.0"
@@ -2683,18 +2693,18 @@ dependencies = [
 
 [[package]]
 name = "thiserror"
-version = "1.0.20"
+version = "1.0.21"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "7dfdd070ccd8ccb78f4ad66bf1982dc37f620ef696c6b5028fe2ed83dd3d0d08"
+checksum = "318234ffa22e0920fe9a40d7b8369b5f649d490980cf7aadcf1eb91594869b42"
 dependencies = [
  "thiserror-impl",
 ]
 
 [[package]]
 name = "thiserror-impl"
-version = "1.0.20"
+version = "1.0.21"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "bd80fc12f73063ac132ac92aceea36734f04a1d93c1240c6944e23a3b8841793"
+checksum = "cae2447b6282786c3493999f40a9be2a6ad20cb8bd268b0a0dbf5a065535c0ab"
 dependencies = [
  "proc-macro2 1.0.19",
  "quote 1.0.7",
@@ -2781,9 +2791,9 @@ dependencies = [
 
 [[package]]
 name = "toml"
-version = "0.5.6"
+version = "0.5.7"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "ffc92d160b1eef40665be3a05630d003936a3bc7da7421277846c2613e92c71a"
+checksum = "75cf45bb0bef80604d001caaec0d09da99611b3c0fd39d3080468875cdb65645"
 dependencies = [
  "serde",
 ]
@@ -2850,9 +2860,9 @@ dependencies = [
 
 [[package]]
 name = "tracing-subscriber"
-version = "0.2.12"
+version = "0.2.13"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "82bb5079aa76438620837198db8a5c529fb9878c730bc2b28179b0241cf04c10"
+checksum = "4ef0a5e15477aa303afbfac3a44cba9b6430fdaad52423b1e6c0dbbe28c3eedd"
 dependencies = [
  "ansi_term 0.12.1",
  "chrono",
@@ -2864,6 +2874,7 @@ dependencies = [
  "sharded-slab",
  "smallvec",
  "thread_local",
+ "tracing",
  "tracing-core",
  "tracing-log",
  "tracing-serde",
diff --git a/Cargo.toml b/Cargo.toml
index 20a0585858..3f100a928f 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -94,13 +94,13 @@ version = "2.0"
 version = "3.0.1"
 
 [dependencies.console]
-version = "0.12.0"
+version = "0.13.0"
 
 [dependencies.dotenv]
 version = "0.15.0"
 
 [dependencies.env_logger]
-version = "0.7"
+version = "0.8"
 
 [dependencies.from-pest]
 version = "0.3.1"
diff --git a/ast/src/ast.rs b/ast/src/ast.rs
index 8506abf29d..a41bf13806 100644
--- a/ast/src/ast.rs
+++ b/ast/src/ast.rs
@@ -159,12 +159,11 @@ impl<'ast> FromPest<'ast> for Expression<'ast> {
     type Rule = Rule;
 
     fn from_pest(pest: &mut Pairs<'ast, Rule>) -> Result<Self, ConversionError<Void>> {
-        let mut clone = pest.clone();
-        let pair = clone.next().ok_or(::from_pest::ConversionError::NoMatch)?;
+        let pair = pest.peek().ok_or(::from_pest::ConversionError::NoMatch)?;
         match pair.as_rule() {
             Rule::expression => {
-                // Transfer iterated state to pest.
-                *pest = clone;
+                // advance the iterator
+                pest.next();
                 Ok(PRECEDENCE_CLIMBER.climb(pair.into_inner(), parse_term, binary_expression))
             }
             _ => Err(ConversionError::NoMatch),
diff --git a/ast/src/leo.pest b/ast/src/leo.pest
index 71024b6439..e695e3bb9f 100644
--- a/ast/src/leo.pest
+++ b/ast/src/leo.pest
@@ -114,7 +114,7 @@ operation_gt = { ">" }
 operation_le = { "<=" }
 operation_lt = { "<" }
 operation_add = { "+" }
-operation_sub = { "- " }
+operation_sub = { "-" }
 operation_mul = { "*" }
 operation_div = { "/" }
 operation_pow = { "**" }
diff --git a/ast/tests/expression.rs b/ast/tests/expression.rs
index c168d34750..99081e0004 100644
--- a/ast/tests/expression.rs
+++ b/ast/tests/expression.rs
@@ -1,3 +1,19 @@
+// Copyright (C) 2019-2020 Aleo Systems Inc.
+// This file is part of the Leo library.
+
+// The Leo library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// The Leo library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with the Leo library. If not, see <https://www.gnu.org/licenses/>.
+
 use leo_ast::ast::{LanguageParser, Rule};
 
 use pest::*;
diff --git a/ast/tests/function.rs b/ast/tests/function.rs
index f35ed1f47e..942976fb10 100644
--- a/ast/tests/function.rs
+++ b/ast/tests/function.rs
@@ -1,3 +1,19 @@
+// Copyright (C) 2019-2020 Aleo Systems Inc.
+// This file is part of the Leo library.
+
+// The Leo library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// The Leo library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with the Leo library. If not, see <https://www.gnu.org/licenses/>.
+
 use leo_ast::ast::{LanguageParser, Rule};
 
 use pest::*;
diff --git a/ast/tests/serialization/json.rs b/ast/tests/serialization/json.rs
index 1fec73e112..30f411a5d9 100644
--- a/ast/tests/serialization/json.rs
+++ b/ast/tests/serialization/json.rs
@@ -14,13 +14,12 @@
 // You should have received a copy of the GNU General Public License
 // along with the Leo library. If not, see <https://www.gnu.org/licenses/>.
 
-use leo_ast::LeoAst;
-
-use std::path::PathBuf;
-
 #[test]
 #[cfg(not(feature = "ci_skip"))]
 fn test_serialize() {
+    use leo_ast::LeoAst;
+    use std::path::PathBuf;
+
     let mut program_filepath = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
     program_filepath.push("tests/serialization/main.leo");
 
diff --git a/ast/tests/tuple.rs b/ast/tests/tuple.rs
index 4b550214fe..3b2825981c 100644
--- a/ast/tests/tuple.rs
+++ b/ast/tests/tuple.rs
@@ -1,3 +1,19 @@
+// Copyright (C) 2019-2020 Aleo Systems Inc.
+// This file is part of the Leo library.
+
+// The Leo library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// The Leo library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with the Leo library. If not, see <https://www.gnu.org/licenses/>.
+
 use leo_ast::ast::{LanguageParser, Rule};
 
 use pest::*;
diff --git a/compiler/src/expression/array/access.rs b/compiler/src/expression/array/access.rs
index f1a388d3c2..7f7b64d7be 100644
--- a/compiler/src/expression/array/access.rs
+++ b/compiler/src/expression/array/access.rs
@@ -36,14 +36,7 @@ impl<F: Field + PrimeField, G: GroupType<F>> ConstrainedProgram<F, G> {
         index: RangeOrExpression,
         span: &Span,
     ) -> Result<ConstrainedValue<F, G>, ExpressionError> {
-        let array = match self.enforce_operand(
-            cs,
-            file_scope.clone(),
-            function_scope.clone(),
-            expected_type,
-            array,
-            span,
-        )? {
+        let array = match self.enforce_operand(cs, file_scope, function_scope, expected_type, array, span)? {
             ConstrainedValue::Array(array) => array,
             value => return Err(ExpressionError::undefined_array(value.to_string(), span.to_owned())),
         };
diff --git a/compiler/src/expression/circuit/access.rs b/compiler/src/expression/circuit/access.rs
index 6668bc6eeb..d0c2f806c2 100644
--- a/compiler/src/expression/circuit/access.rs
+++ b/compiler/src/expression/circuit/access.rs
@@ -39,12 +39,12 @@ impl<F: Field + PrimeField, G: GroupType<F>> ConstrainedProgram<F, G> {
         file_scope: &str,
         function_scope: &str,
         expected_type: Option<Type>,
-        circuit_identifier: Box<Expression>,
+        circuit_identifier: Expression,
         circuit_member: Identifier,
         span: Span,
     ) -> Result<ConstrainedValue<F, G>, ExpressionError> {
         // access a circuit member using the `self` keyword
-        if let Expression::Identifier(ref identifier) = *circuit_identifier {
+        if let Expression::Identifier(ref identifier) = circuit_identifier {
             if identifier.is_self() {
                 let self_file_scope = new_scope(&file_scope, &identifier.name);
                 let self_function_scope = new_scope(&self_file_scope, &identifier.name);
@@ -56,17 +56,11 @@ impl<F: Field + PrimeField, G: GroupType<F>> ConstrainedProgram<F, G> {
             }
         }
 
-        let (circuit_name, members) = match self.enforce_operand(
-            cs,
-            file_scope,
-            function_scope,
-            expected_type,
-            *circuit_identifier,
-            &span,
-        )? {
-            ConstrainedValue::CircuitExpression(name, members) => (name, members),
-            value => return Err(ExpressionError::undefined_circuit(value.to_string(), span)),
-        };
+        let (circuit_name, members) =
+            match self.enforce_operand(cs, file_scope, function_scope, expected_type, circuit_identifier, &span)? {
+                ConstrainedValue::CircuitExpression(name, members) => (name, members),
+                value => return Err(ExpressionError::undefined_circuit(value.to_string(), span)),
+            };
 
         let matched_member = members.clone().into_iter().find(|member| member.0 == circuit_member);
 
diff --git a/compiler/src/expression/circuit/static_access.rs b/compiler/src/expression/circuit/static_access.rs
index 90fae8ca4f..1f01fce5dd 100644
--- a/compiler/src/expression/circuit/static_access.rs
+++ b/compiler/src/expression/circuit/static_access.rs
@@ -32,12 +32,12 @@ impl<F: Field + PrimeField, G: GroupType<F>> ConstrainedProgram<F, G> {
         file_scope: &str,
         function_scope: &str,
         expected_type: Option<Type>,
-        circuit_identifier: Box<Expression>,
+        circuit_identifier: Expression,
         circuit_member: Identifier,
         span: Span,
     ) -> Result<ConstrainedValue<F, G>, ExpressionError> {
         // Get defined circuit
-        let circuit = match *circuit_identifier {
+        let circuit = match circuit_identifier {
             Expression::Identifier(identifier) => {
                 // Use the "Self" keyword to access a static circuit function
                 if identifier.is_self() {
diff --git a/compiler/src/expression/expression.rs b/compiler/src/expression/expression.rs
index 5298c3eb50..9c0d98c41f 100644
--- a/compiler/src/expression/expression.rs
+++ b/compiler/src/expression/expression.rs
@@ -261,7 +261,7 @@ impl<F: Field + PrimeField, G: GroupType<F>> ConstrainedProgram<F, G> {
                 self.enforce_tuple(cs, file_scope, function_scope, expected_type, tuple, span)
             }
             Expression::TupleAccess(tuple, index, span) => {
-                self.enforce_tuple_access(cs, file_scope, function_scope, expected_type, tuple, index, &span)
+                self.enforce_tuple_access(cs, file_scope, function_scope, expected_type, *tuple, index, &span)
             }
 
             // Circuits
@@ -273,7 +273,7 @@ impl<F: Field + PrimeField, G: GroupType<F>> ConstrainedProgram<F, G> {
                 file_scope,
                 function_scope,
                 expected_type,
-                circuit_variable,
+                *circuit_variable,
                 circuit_member,
                 span,
             ),
@@ -283,7 +283,7 @@ impl<F: Field + PrimeField, G: GroupType<F>> ConstrainedProgram<F, G> {
                     file_scope,
                     function_scope,
                     expected_type,
-                    circuit_identifier,
+                    *circuit_identifier,
                     circuit_member,
                     span,
                 ),
@@ -294,7 +294,7 @@ impl<F: Field + PrimeField, G: GroupType<F>> ConstrainedProgram<F, G> {
                 file_scope,
                 function_scope,
                 expected_type,
-                function,
+                *function,
                 arguments,
                 span,
             ),
diff --git a/compiler/src/expression/function/function.rs b/compiler/src/expression/function/function.rs
index 5915c32084..f24bcc84c6 100644
--- a/compiler/src/expression/function/function.rs
+++ b/compiler/src/expression/function/function.rs
@@ -32,17 +32,17 @@ impl<F: Field + PrimeField, G: GroupType<F>> ConstrainedProgram<F, G> {
         file_scope: &str,
         function_scope: &str,
         expected_type: Option<Type>,
-        function: Box<Expression>,
+        function: Expression,
         arguments: Vec<Expression>,
         span: Span,
     ) -> Result<ConstrainedValue<F, G>, ExpressionError> {
-        let (declared_circuit_reference, function_value) = match *function {
+        let (declared_circuit_reference, function_value) = match function {
             Expression::CircuitMemberAccess(circuit_identifier, circuit_member, span) => {
                 // Call a circuit function that can mutate self.
 
                 // Save a reference to the circuit we are mutating.
                 let circuit_id_string = circuit_identifier.to_string();
-                let declared_circuit_reference = new_scope(function_scope.clone(), &circuit_id_string);
+                let declared_circuit_reference = new_scope(function_scope, &circuit_id_string);
 
                 (
                     declared_circuit_reference,
@@ -51,7 +51,7 @@ impl<F: Field + PrimeField, G: GroupType<F>> ConstrainedProgram<F, G> {
                         file_scope,
                         function_scope,
                         expected_type,
-                        circuit_identifier,
+                        *circuit_identifier,
                         circuit_member,
                         span,
                     )?,
@@ -65,15 +65,17 @@ impl<F: Field + PrimeField, G: GroupType<F>> ConstrainedProgram<F, G> {
 
         let (outer_scope, function_call) = function_value.extract_function(file_scope, &span)?;
 
-        let name_unique = format!(
-            "function call {} {}:{}",
-            function_call.get_name(),
-            span.line,
-            span.start,
-        );
+        let name_unique = || {
+            format!(
+                "function call {} {}:{}",
+                function_call.get_name(),
+                span.line,
+                span.start,
+            )
+        };
 
         self.enforce_function(
-            &mut cs.ns(|| name_unique),
+            &mut cs.ns(name_unique),
             &outer_scope,
             function_scope,
             function_call,
diff --git a/compiler/src/expression/logical/and.rs b/compiler/src/expression/logical/and.rs
index 1f901d0e19..60d6f58eae 100644
--- a/compiler/src/expression/logical/and.rs
+++ b/compiler/src/expression/logical/and.rs
@@ -33,9 +33,12 @@ pub fn enforce_and<F: Field + PrimeField, G: GroupType<F>, CS: ConstraintSystem<
     let name = format!("{} && {}", left, right);
 
     if let (ConstrainedValue::Boolean(left_bool), ConstrainedValue::Boolean(right_bool)) = (left, right) {
-        let name_unique = format!("{} {}:{}", name, span.line, span.start);
-        let result = Boolean::and(cs.ns(|| name_unique), &left_bool, &right_bool)
-            .map_err(|e| BooleanError::cannot_enforce("&&".to_string(), e, span.to_owned()))?;
+        let result = Boolean::and(
+            cs.ns(|| format!("{} {}:{}", name, span.line, span.start)),
+            &left_bool,
+            &right_bool,
+        )
+        .map_err(|e| BooleanError::cannot_enforce("&&".to_string(), e, span.to_owned()))?;
 
         return Ok(ConstrainedValue::Boolean(result));
     }
diff --git a/compiler/src/expression/logical/or.rs b/compiler/src/expression/logical/or.rs
index 644c90fa21..6ba45a3050 100644
--- a/compiler/src/expression/logical/or.rs
+++ b/compiler/src/expression/logical/or.rs
@@ -33,9 +33,12 @@ pub fn enforce_or<F: Field + PrimeField, G: GroupType<F>, CS: ConstraintSystem<F
     let name = format!("{} || {}", left, right);
 
     if let (ConstrainedValue::Boolean(left_bool), ConstrainedValue::Boolean(right_bool)) = (left, right) {
-        let name_unique = format!("{} {}:{}", name, span.line, span.start);
-        let result = Boolean::or(cs.ns(|| name_unique), &left_bool, &right_bool)
-            .map_err(|e| BooleanError::cannot_enforce("||".to_string(), e, span.to_owned()))?;
+        let result = Boolean::or(
+            cs.ns(|| format!("{} {}:{}", name, span.line, span.start)),
+            &left_bool,
+            &right_bool,
+        )
+        .map_err(|e| BooleanError::cannot_enforce("||".to_string(), e, span.to_owned()))?;
 
         return Ok(ConstrainedValue::Boolean(result));
     }
diff --git a/compiler/src/expression/tuple/access.rs b/compiler/src/expression/tuple/access.rs
index b489fb9a81..c628ac4243 100644
--- a/compiler/src/expression/tuple/access.rs
+++ b/compiler/src/expression/tuple/access.rs
@@ -32,11 +32,11 @@ impl<F: Field + PrimeField, G: GroupType<F>> ConstrainedProgram<F, G> {
         file_scope: &str,
         function_scope: &str,
         expected_type: Option<Type>,
-        tuple: Box<Expression>,
+        tuple: Expression,
         index: usize,
         span: &Span,
     ) -> Result<ConstrainedValue<F, G>, ExpressionError> {
-        let tuple = match self.enforce_operand(cs, file_scope, function_scope, expected_type, *tuple, &span)? {
+        let tuple = match self.enforce_operand(cs, file_scope, function_scope, expected_type, tuple, &span)? {
             ConstrainedValue::Tuple(tuple) => tuple,
             value => return Err(ExpressionError::undefined_array(value.to_string(), span.to_owned())),
         };
diff --git a/compiler/src/function/result/result.rs b/compiler/src/function/result/result.rs
index f4e84ae4c0..cceafc612a 100644
--- a/compiler/src/function/result/result.rs
+++ b/compiler/src/function/result/result.rs
@@ -61,12 +61,13 @@ impl<F: Field + PrimeField, G: GroupType<F>> ConstrainedProgram<F, G> {
             }
 
             let condition = indicator.unwrap_or(Boolean::Constant(true));
-            let name_unique = format!("select {} {}:{}", result, span.line, span.start);
-            let selected_value =
-                ConstrainedValue::conditionally_select(cs.ns(|| name_unique), &condition, &result, return_value)
-                    .map_err(|_| {
-                        StatementError::select_fail(result.to_string(), return_value.to_string(), span.to_owned())
-                    })?;
+            let selected_value = ConstrainedValue::conditionally_select(
+                cs.ns(|| format!("select {} {}:{}", result, span.line, span.start)),
+                &condition,
+                &result,
+                return_value,
+            )
+            .map_err(|_| StatementError::select_fail(result.to_string(), return_value.to_string(), span.to_owned()))?;
 
             *return_value = selected_value;
         }
diff --git a/compiler/src/statement/assign/array.rs b/compiler/src/statement/assign/array.rs
index dde5ac528b..c282ca0fee 100644
--- a/compiler/src/statement/assign/array.rs
+++ b/compiler/src/statement/assign/array.rs
@@ -52,9 +52,8 @@ impl<F: Field + PrimeField, G: GroupType<F>> ConstrainedProgram<F, G> {
                     ConstrainedValue::Array(old) => {
                         new_value.resolve_type(Some(old[index].to_type(&span)?), &span)?;
 
-                        let name_unique = format!("select {} {}:{}", new_value, span.line, span.start);
                         let selected_value = ConstrainedValue::conditionally_select(
-                            cs.ns(|| name_unique),
+                            cs.ns(|| format!("select {} {}:{}", new_value, span.line, span.start)),
                             &condition,
                             &new_value,
                             &old[index],
@@ -89,12 +88,15 @@ impl<F: Field + PrimeField, G: GroupType<F>> ConstrainedProgram<F, G> {
                     }
                     _ => return Err(StatementError::array_assign_range(span.to_owned())),
                 };
-                let name_unique = format!("select {} {}:{}", new_array, span.line, span.start);
-                let selected_array =
-                    ConstrainedValue::conditionally_select(cs.ns(|| name_unique), &condition, &new_array, old_array)
-                        .map_err(|_| {
-                            StatementError::select_fail(new_array.to_string(), old_array.to_string(), span.to_owned())
-                        })?;
+                let selected_array = ConstrainedValue::conditionally_select(
+                    cs.ns(|| format!("select {} {}:{}", new_array, span.line, span.start)),
+                    &condition,
+                    &new_array,
+                    old_array,
+                )
+                .map_err(|_| {
+                    StatementError::select_fail(new_array.to_string(), old_array.to_string(), span.to_owned())
+                })?;
 
                 *old_array = selected_array;
             }
diff --git a/compiler/src/statement/assign/assign.rs b/compiler/src/statement/assign/assign.rs
index 0536775975..2a317040a9 100644
--- a/compiler/src/statement/assign/assign.rs
+++ b/compiler/src/statement/assign/assign.rs
@@ -61,12 +61,15 @@ impl<F: Field + PrimeField, G: GroupType<F>> ConstrainedProgram<F, G> {
 
                 new_value.resolve_type(Some(old_value.to_type(&span)?), &span)?;
 
-                let name_unique = format!("select {} {}:{}", new_value, span.line, span.start);
-                let selected_value =
-                    ConstrainedValue::conditionally_select(cs.ns(|| name_unique), &condition, &new_value, old_value)
-                        .map_err(|_| {
-                            StatementError::select_fail(new_value.to_string(), old_value.to_string(), span.to_owned())
-                        })?;
+                let selected_value = ConstrainedValue::conditionally_select(
+                    cs.ns(|| format!("select {} {}:{}", new_value, span.line, span.start)),
+                    &condition,
+                    &new_value,
+                    old_value,
+                )
+                .map_err(|_| {
+                    StatementError::select_fail(new_value.to_string(), old_value.to_string(), span.to_owned())
+                })?;
 
                 *old_value = selected_value;
 
diff --git a/compiler/src/statement/assign/circuit_variable.rs b/compiler/src/statement/assign/circuit_variable.rs
index 099c24adf2..2aea133662 100644
--- a/compiler/src/statement/assign/circuit_variable.rs
+++ b/compiler/src/statement/assign/circuit_variable.rs
@@ -68,9 +68,8 @@ impl<F: Field + PrimeField, G: GroupType<F>> ConstrainedProgram<F, G> {
                             new_value.resolve_type(Some(value.to_type(span)?), span)?;
 
                             // Conditionally select the value if this branch is executed.
-                            let name_unique = format!("select {} {}:{}", new_value, span.line, span.start);
                             let mut selected_value = ConstrainedValue::conditionally_select(
-                                cs.ns(|| name_unique),
+                                cs.ns(|| format!("select {} {}:{}", new_value, span.line, span.start)),
                                 &condition,
                                 &new_value,
                                 &member.1,
diff --git a/compiler/src/statement/assign/tuple.rs b/compiler/src/statement/assign/tuple.rs
index 24d7d80f1f..a8a8a0a7fb 100644
--- a/compiler/src/statement/assign/tuple.rs
+++ b/compiler/src/statement/assign/tuple.rs
@@ -44,12 +44,15 @@ impl<F: Field + PrimeField, G: GroupType<F>> ConstrainedProgram<F, G> {
             ConstrainedValue::Tuple(old) => {
                 new_value.resolve_type(Some(old[index].to_type(&span)?), &span)?;
 
-                let name_unique = format!("select {} {}:{}", new_value, span.line, span.start);
-                let selected_value =
-                    ConstrainedValue::conditionally_select(cs.ns(|| name_unique), &condition, &new_value, &old[index])
-                        .map_err(|_| {
-                            StatementError::select_fail(new_value.to_string(), old[index].to_string(), span.to_owned())
-                        })?;
+                let selected_value = ConstrainedValue::conditionally_select(
+                    cs.ns(|| format!("select {} {}:{}", new_value, span.line, span.start)),
+                    &condition,
+                    &new_value,
+                    &old[index],
+                )
+                .map_err(|_| {
+                    StatementError::select_fail(new_value.to_string(), old[index].to_string(), span.to_owned())
+                })?;
 
                 old[index] = selected_value;
             }
diff --git a/compiler/src/statement/iteration/iteration.rs b/compiler/src/statement/iteration/iteration.rs
index 9bc96ca4be..f9c658ad1e 100644
--- a/compiler/src/statement/iteration/iteration.rs
+++ b/compiler/src/statement/iteration/iteration.rs
@@ -67,9 +67,8 @@ impl<F: Field + PrimeField, G: GroupType<F>> ConstrainedProgram<F, G> {
             );
 
             // Evaluate statements and possibly return early
-            let name_unique = format!("for loop iteration {} {}:{}", i, span.line, span.start);
             let mut result = self.evaluate_branch(
-                &mut cs.ns(|| name_unique),
+                &mut cs.ns(|| format!("for loop iteration {} {}:{}", i, span.line, span.start)),
                 file_scope,
                 function_scope,
                 indicator,
diff --git a/compiler/src/value/address/address.rs b/compiler/src/value/address/address.rs
index b98803d238..e385138402 100644
--- a/compiler/src/value/address/address.rs
+++ b/compiler/src/value/address/address.rs
@@ -60,13 +60,7 @@ impl Address {
     }
 
     pub(crate) fn is_constant(&self) -> bool {
-        let mut result = true;
-
-        for byte in self.bytes.iter() {
-            result = result && byte.is_constant()
-        }
-
-        result
+        self.bytes.iter().all(|byte| byte.is_constant())
     }
 
     pub(crate) fn from_input<F: Field + PrimeField, G: GroupType<F>, CS: ConstraintSystem<F>>(
@@ -87,12 +81,10 @@ impl Address {
             None => None,
         };
 
-        let address_name = format!("{}: address", name);
-        let address_namespace = format!("`{}` {}:{}", address_name, span.line, span.start);
-
-        let address = Address::alloc(cs.ns(|| address_namespace), || {
-            address_value.ok_or(SynthesisError::AssignmentMissing)
-        })
+        let address = Address::alloc(
+            cs.ns(|| format!("`{}: address` {}:{}", name, span.line, span.start)),
+            || address_value.ok_or(SynthesisError::AssignmentMissing),
+        )
         .map_err(|_| AddressError::missing_address(span.to_owned()))?;
 
         Ok(ConstrainedValue::Address(address))
diff --git a/compiler/src/value/boolean/input.rs b/compiler/src/value/boolean/input.rs
index c627e96bcb..9b712b95d6 100644
--- a/compiler/src/value/boolean/input.rs
+++ b/compiler/src/value/boolean/input.rs
@@ -42,13 +42,11 @@ pub(crate) fn allocate_bool<F: Field + PrimeField, CS: ConstraintSystem<F>>(
     option: Option<bool>,
     span: &Span,
 ) -> Result<Boolean, BooleanError> {
-    let boolean_name = format!("{}: bool", name);
-    let boolean_name_unique = format!("`{}` {}:{}", boolean_name, span.line, span.start);
-
-    Boolean::alloc(cs.ns(|| boolean_name_unique), || {
-        option.ok_or(SynthesisError::AssignmentMissing)
-    })
-    .map_err(|_| BooleanError::missing_boolean(boolean_name.to_owned(), span.to_owned()))
+    Boolean::alloc(
+        cs.ns(|| format!("`{}: bool` {}:{}", name, span.line, span.start)),
+        || option.ok_or(SynthesisError::AssignmentMissing),
+    )
+    .map_err(|_| BooleanError::missing_boolean(format!("{}: bool", name), span.to_owned()))
 }
 
 pub(crate) fn bool_from_input<F: Field + PrimeField, G: GroupType<F>, CS: ConstraintSystem<F>>(
diff --git a/compiler/src/value/field/input.rs b/compiler/src/value/field/input.rs
index ae1f092669..34d14d0f75 100644
--- a/compiler/src/value/field/input.rs
+++ b/compiler/src/value/field/input.rs
@@ -31,13 +31,11 @@ pub(crate) fn allocate_field<F: Field + PrimeField, CS: ConstraintSystem<F>>(
     option: Option<String>,
     span: &Span,
 ) -> Result<FieldType<F>, FieldError> {
-    let field_name = format!("{}: field", name);
-    let field_name_unique = format!("`{}` {}:{}", field_name, span.line, span.start);
-
-    FieldType::alloc(cs.ns(|| field_name_unique), || {
-        option.ok_or(SynthesisError::AssignmentMissing)
-    })
-    .map_err(|_| FieldError::missing_field(field_name, span.to_owned()))
+    FieldType::alloc(
+        cs.ns(|| format!("`{}: field` {}:{}", name, span.line, span.start)),
+        || option.ok_or(SynthesisError::AssignmentMissing),
+    )
+    .map_err(|_| FieldError::missing_field(format!("{}: field", name), span.to_owned()))
 }
 
 pub(crate) fn field_from_input<F: Field + PrimeField, G: GroupType<F>, CS: ConstraintSystem<F>>(
diff --git a/compiler/src/value/group/input.rs b/compiler/src/value/group/input.rs
index 1fa9492463..293ffd92a0 100644
--- a/compiler/src/value/group/input.rs
+++ b/compiler/src/value/group/input.rs
@@ -31,13 +31,11 @@ pub(crate) fn allocate_group<F: Field + PrimeField, G: GroupType<F>, CS: Constra
     option: Option<GroupValue>,
     span: &Span,
 ) -> Result<G, GroupError> {
-    let group_name = format!("{}: group", name);
-    let group_name_unique = format!("`{}` {}:{}", group_name, span.line, span.start);
-
-    G::alloc(cs.ns(|| group_name_unique), || {
-        option.ok_or(SynthesisError::AssignmentMissing)
-    })
-    .map_err(|_| GroupError::missing_group(group_name, span.to_owned()))
+    G::alloc(
+        cs.ns(|| format!("`{}: group` {}:{}", name, span.line, span.start)),
+        || option.ok_or(SynthesisError::AssignmentMissing),
+    )
+    .map_err(|_| GroupError::missing_group(format!("{}: group", name), span.to_owned()))
 }
 
 pub(crate) fn group_from_input<F: Field + PrimeField, G: GroupType<F>, CS: ConstraintSystem<F>>(
diff --git a/compiler/src/value/integer/integer.rs b/compiler/src/value/integer/integer.rs
index 3d79ac8bdf..0384248624 100644
--- a/compiler/src/value/integer/integer.rs
+++ b/compiler/src/value/integer/integer.rs
@@ -189,155 +189,142 @@ impl Integer {
     ) -> Result<Self, IntegerError> {
         Ok(match integer_type {
             IntegerType::U8 => {
-                let u8_name = format!("{}: u8", name);
-                let u8_name_unique = format!("`{}` {}:{}", u8_name, span.line, span.start);
-
                 let u8_option = option.map(|s| {
                     s.parse::<u8>()
                         .map_err(|_| IntegerError::invalid_integer(s, span.to_owned()))
                         .unwrap()
                 });
 
-                let u8_result = UInt8::alloc(cs.ns(|| u8_name_unique), || {
+                let u8_result = UInt8::alloc(cs.ns(|| format!("`{}: u8` {}:{}", name, span.line, span.start)), || {
                     u8_option.ok_or(SynthesisError::AssignmentMissing)
                 })
-                .map_err(|_| IntegerError::missing_integer(u8_name, span.to_owned()))?;
+                .map_err(|_| IntegerError::missing_integer(format!("{}: u8", name), span.to_owned()))?;
 
                 Integer::U8(u8_result)
             }
             IntegerType::U16 => {
-                let u16_name = format!("{}: u16", name);
-                let u16_name_unique = format!("`{}` {}:{}", u16_name, span.line, span.start);
                 let u16_option = option.map(|s| {
                     s.parse::<u16>()
                         .map_err(|_| IntegerError::invalid_integer(s, span.to_owned()))
                         .unwrap()
                 });
-                let u16_result = UInt16::alloc(cs.ns(|| u16_name_unique), || {
-                    u16_option.ok_or(SynthesisError::AssignmentMissing)
-                })
-                .map_err(|_| IntegerError::missing_integer(u16_name, span.to_owned()))?;
+                let u16_result = UInt16::alloc(
+                    cs.ns(|| format!("`{}: u16` {}:{}", name, span.line, span.start)),
+                    || u16_option.ok_or(SynthesisError::AssignmentMissing),
+                )
+                .map_err(|_| IntegerError::missing_integer(format!("{}: u16", name), span.to_owned()))?;
 
                 Integer::U16(u16_result)
             }
             IntegerType::U32 => {
-                let u32_name = format!("{}: u32", name);
-                let u32_name_unique = format!("`{}` {}:{}", u32_name, span.line, span.start);
                 let u32_option = option.map(|s| {
                     s.parse::<u32>()
                         .map_err(|_| IntegerError::invalid_integer(s, span.to_owned()))
                         .unwrap()
                 });
-                let u32_result = UInt32::alloc(cs.ns(|| u32_name_unique), || {
-                    u32_option.ok_or(SynthesisError::AssignmentMissing)
-                })
-                .map_err(|_| IntegerError::missing_integer(u32_name, span.to_owned()))?;
+                let u32_result = UInt32::alloc(
+                    cs.ns(|| format!("`{}: u32` {}:{}", name, span.line, span.start)),
+                    || u32_option.ok_or(SynthesisError::AssignmentMissing),
+                )
+                .map_err(|_| IntegerError::missing_integer(format!("{}: u32", name), span.to_owned()))?;
 
                 Integer::U32(u32_result)
             }
             IntegerType::U64 => {
-                let u64_name = format!("{}: u64", name);
-                let u64_name_unique = format!("`{}` {}:{}", u64_name, span.line, span.start);
                 let u64_option = option.map(|s| {
                     s.parse::<u64>()
                         .map_err(|_| IntegerError::invalid_integer(s, span.to_owned()))
                         .unwrap()
                 });
-                let u64_result = UInt64::alloc(cs.ns(|| u64_name_unique), || {
-                    u64_option.ok_or(SynthesisError::AssignmentMissing)
-                })
-                .map_err(|_| IntegerError::missing_integer(u64_name, span.to_owned()))?;
+                let u64_result = UInt64::alloc(
+                    cs.ns(|| format!("`{}: u64` {}:{}", name, span.line, span.start)),
+                    || u64_option.ok_or(SynthesisError::AssignmentMissing),
+                )
+                .map_err(|_| IntegerError::missing_integer(format!("{}: u64", name), span.to_owned()))?;
 
                 Integer::U64(u64_result)
             }
             IntegerType::U128 => {
-                let u128_name = format!("{}: u128", name);
-                let u128_name_unique = format!("`{}` {}:{}", u128_name, span.line, span.start);
                 let u128_option = option.map(|s| {
                     s.parse::<u128>()
                         .map_err(|_| IntegerError::invalid_integer(s, span.to_owned()))
                         .unwrap()
                 });
-                let u128_result = UInt128::alloc(cs.ns(|| u128_name_unique), || {
-                    u128_option.ok_or(SynthesisError::AssignmentMissing)
-                })
-                .map_err(|_| IntegerError::missing_integer(u128_name, span.to_owned()))?;
+                let u128_result = UInt128::alloc(
+                    cs.ns(|| format!("`{}: u128` {}:{}", name, span.line, span.start)),
+                    || u128_option.ok_or(SynthesisError::AssignmentMissing),
+                )
+                .map_err(|_| IntegerError::missing_integer(format!("{}: u128", name), span.to_owned()))?;
 
                 Integer::U128(u128_result)
             }
 
             IntegerType::I8 => {
-                let i8_name = format!("{}: i8", name);
-                let i8_name_unique = format!("`{}` {}:{}", i8_name, span.line, span.start);
                 let i8_option = option.map(|s| {
                     s.parse::<i8>()
                         .map_err(|_| IntegerError::invalid_integer(s, span.to_owned()))
                         .unwrap()
                 });
-                let i8_result = Int8::alloc(cs.ns(|| i8_name_unique), || {
+                let i8_result = Int8::alloc(cs.ns(|| format!("`{}: i8` {}:{}", name, span.line, span.start)), || {
                     i8_option.ok_or(SynthesisError::AssignmentMissing)
                 })
-                .map_err(|_| IntegerError::missing_integer(i8_name, span.to_owned()))?;
+                .map_err(|_| IntegerError::missing_integer(format!("{}: i8", name), span.to_owned()))?;
 
                 Integer::I8(i8_result)
             }
             IntegerType::I16 => {
-                let i16_name = format!("{}: i16", name);
-                let i16_name_unique = format!("`{}` {}:{}", i16_name, span.line, span.start);
                 let i16_option = option.map(|s| {
                     s.parse::<i16>()
                         .map_err(|_| IntegerError::invalid_integer(s, span.to_owned()))
                         .unwrap()
                 });
-                let i16_result = Int16::alloc(cs.ns(|| i16_name_unique), || {
-                    i16_option.ok_or(SynthesisError::AssignmentMissing)
-                })
-                .map_err(|_| IntegerError::missing_integer(i16_name, span.to_owned()))?;
+                let i16_result = Int16::alloc(
+                    cs.ns(|| format!("`{}: i16` {}:{}", name, span.line, span.start)),
+                    || i16_option.ok_or(SynthesisError::AssignmentMissing),
+                )
+                .map_err(|_| IntegerError::missing_integer(format!("{}: i16", name), span.to_owned()))?;
 
                 Integer::I16(i16_result)
             }
             IntegerType::I32 => {
-                let i32_name = format!("{}: i32", name);
-                let i32_name_unique = format!("`{}` {}:{}", i32_name, span.line, span.start);
                 let i32_option = option.map(|s| {
                     s.parse::<i32>()
                         .map_err(|_| IntegerError::invalid_integer(s, span.to_owned()))
                         .unwrap()
                 });
-                let i32_result = Int32::alloc(cs.ns(|| i32_name_unique), || {
-                    i32_option.ok_or(SynthesisError::AssignmentMissing)
-                })
-                .map_err(|_| IntegerError::missing_integer(i32_name, span.to_owned()))?;
+                let i32_result = Int32::alloc(
+                    cs.ns(|| format!("`{}: i32` {}:{}", name, span.line, span.start)),
+                    || i32_option.ok_or(SynthesisError::AssignmentMissing),
+                )
+                .map_err(|_| IntegerError::missing_integer(format!("{}: i32", name), span.to_owned()))?;
 
                 Integer::I32(i32_result)
             }
             IntegerType::I64 => {
-                let i64_name = format!("{}: i64", name);
-                let i64_name_unique = format!("`{}` {}:{}", i64_name, span.line, span.start);
                 let i64_option = option.map(|s| {
                     s.parse::<i64>()
                         .map_err(|_| IntegerError::invalid_integer(s, span.to_owned()))
                         .unwrap()
                 });
-                let i64_result = Int64::alloc(cs.ns(|| i64_name_unique), || {
-                    i64_option.ok_or(SynthesisError::AssignmentMissing)
-                })
-                .map_err(|_| IntegerError::missing_integer(i64_name, span.to_owned()))?;
+                let i64_result = Int64::alloc(
+                    cs.ns(|| format!("`{}: i64` {}:{}", name, span.line, span.start)),
+                    || i64_option.ok_or(SynthesisError::AssignmentMissing),
+                )
+                .map_err(|_| IntegerError::missing_integer(format!("{}: i64", name), span.to_owned()))?;
 
                 Integer::I64(i64_result)
             }
             IntegerType::I128 => {
-                let i128_name = format!("{}: i128", name);
-                let i128_name_unique = format!("`{}` {}:{}", i128_name, span.line, span.start);
                 let i128_option = option.map(|s| {
                     s.parse::<i128>()
                         .map_err(|_| IntegerError::invalid_integer(s, span.to_owned()))
                         .unwrap()
                 });
-                let i128_result = Int128::alloc(cs.ns(|| i128_name_unique), || {
-                    i128_option.ok_or(SynthesisError::AssignmentMissing)
-                })
-                .map_err(|_| IntegerError::missing_integer(i128_name, span.to_owned()))?;
+                let i128_result = Int128::alloc(
+                    cs.ns(|| format!("`{}: i128` {}:{}", name, span.line, span.start)),
+                    || i128_option.ok_or(SynthesisError::AssignmentMissing),
+                )
+                .map_err(|_| IntegerError::missing_integer(format!("{}: i128", name), span.to_owned()))?;
 
                 Integer::I128(i128_result)
             }
diff --git a/examples/fibonacci/.gitignore b/examples/fibonacci/.gitignore
deleted file mode 100644
index ef3a390676..0000000000
--- a/examples/fibonacci/.gitignore
+++ /dev/null
@@ -1,2 +0,0 @@
-outputs/
-/.leo
diff --git a/examples/fibonacci/Leo.toml b/examples/fibonacci/Leo.toml
deleted file mode 100644
index 74409f8b14..0000000000
--- a/examples/fibonacci/Leo.toml
+++ /dev/null
@@ -1,8 +0,0 @@
-[project]
-name = "fibonacci"
-version = "0.1.0"
-description = "Returns a fibonnaci sequence"
-license = "LICENSE-MIT"
-
-[remote]
-author = "aleo"
diff --git a/examples/fibonacci/README.md b/examples/fibonacci/README.md
deleted file mode 100644
index be8c859831..0000000000
--- a/examples/fibonacci/README.md
+++ /dev/null
@@ -1 +0,0 @@
-# fibonacci
diff --git a/examples/fibonacci/inputs/fibonacci.in b/examples/fibonacci/inputs/fibonacci.in
deleted file mode 100644
index 494a224d5d..0000000000
--- a/examples/fibonacci/inputs/fibonacci.in
+++ /dev/null
@@ -1,2 +0,0 @@
-[registers]
-r0: u32 = 0;
\ No newline at end of file
diff --git a/examples/fibonacci/inputs/fibonacci.state b/examples/fibonacci/inputs/fibonacci.state
deleted file mode 100644
index e69de29bb2..0000000000
diff --git a/examples/fibonacci/src/main.leo b/examples/fibonacci/src/main.leo
deleted file mode 100644
index a759e3d010..0000000000
--- a/examples/fibonacci/src/main.leo
+++ /dev/null
@@ -1,16 +0,0 @@
-function fibonacci() -> u32 {
-    let mut a = 0u32;
-    let mut b = 1u32;
-
-    for i in 0..10 {
-      a = b;
-      b = a + b;
-    }
-
-    return a
-}
-
-// The 'fibonacci' main function.
-function main() -> u32 {
-    return fibonacci()
-}
\ No newline at end of file
diff --git a/examples/square-root/.gitignore b/examples/square-root/.gitignore
deleted file mode 100644
index ad22736663..0000000000
--- a/examples/square-root/.gitignore
+++ /dev/null
@@ -1,2 +0,0 @@
-outputs/
-/.leo
\ No newline at end of file
diff --git a/examples/square-root/Leo.toml b/examples/square-root/Leo.toml
deleted file mode 100644
index 9717d86062..0000000000
--- a/examples/square-root/Leo.toml
+++ /dev/null
@@ -1,8 +0,0 @@
-[project]
-name = "square-root"
-version = "0.1.0"
-description = "prove knowledge of the square root `a` of a number `b`"
-license = "LICENSE-MIT"
-
-[remote]
-author = "aleo"
diff --git a/examples/square-root/README.md b/examples/square-root/README.md
deleted file mode 100644
index 8536c68b02..0000000000
--- a/examples/square-root/README.md
+++ /dev/null
@@ -1 +0,0 @@
-# square-root
diff --git a/examples/square-root/inputs/square-root.in b/examples/square-root/inputs/square-root.in
deleted file mode 100644
index 1d6a8cf5d3..0000000000
--- a/examples/square-root/inputs/square-root.in
+++ /dev/null
@@ -1,7 +0,0 @@
-// The program input for square_root/src/main.leo
-[main]
-a: u32 = 337;
-b: u32 = 113569;
-
-[registers]
-r0: bool = false;
\ No newline at end of file
diff --git a/examples/square-root/inputs/square-root.state b/examples/square-root/inputs/square-root.state
deleted file mode 100644
index e69de29bb2..0000000000
diff --git a/examples/square-root/src/main.leo b/examples/square-root/src/main.leo
deleted file mode 100644
index 5d69e2e375..0000000000
--- a/examples/square-root/src/main.leo
+++ /dev/null
@@ -1,5 +0,0 @@
-// The 'square-root' main function.
-// prove knowledge of the square root `a` of a number `b`
-function main(a: u32, b: u32) -> bool {
-    return a * a == b
-}
diff --git a/gadgets/Cargo.toml b/gadgets/Cargo.toml
index b7a0e42181..823847acae 100644
--- a/gadgets/Cargo.toml
+++ b/gadgets/Cargo.toml
@@ -41,3 +41,11 @@ version = "1.0"
 
 [dev-dependencies.snarkos-utilities]
 version = "1.1.3"
+
+[dev-dependencies.criterion]
+version = "0.3"
+
+[[bench]]
+name = "integer_arithmetic"
+path = "benches/integer_arithmetic.rs"
+harness = false
diff --git a/gadgets/benches/integer_arithmetic.rs b/gadgets/benches/integer_arithmetic.rs
new file mode 100644
index 0000000000..e0f4d3e151
--- /dev/null
+++ b/gadgets/benches/integer_arithmetic.rs
@@ -0,0 +1,429 @@
+// Copyright (C) 2019-2020 Aleo Systems Inc.
+// This file is part of the Leo library.
+
+// The Leo library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// The Leo library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with the Leo library. If not, see <https://www.gnu.org/licenses/>.
+
+use leo_gadgets::{arithmetic::*, Int128, Int16, Int32, Int64, Int8};
+
+use snarkos_models::gadgets::{
+    r1cs::{ConstraintSystem, Fr, TestConstraintSystem},
+    utilities::alloc::AllocGadget,
+};
+
+use rand::{Rng, SeedableRng};
+use rand_xorshift::XorShiftRng;
+use std::i128;
+
+use criterion::{criterion_group, criterion_main, Criterion};
+
+macro_rules! create_add_bench {
+    ($bench_name:ident, $bench_id:expr, $foo_name:ident, $std_type:ty, $bit_type:ty) => {
+        fn $bench_name(c: &mut Criterion) {
+            fn $foo_name(cs: &mut TestConstraintSystem<Fr>, rng: &mut XorShiftRng) {
+                let a: $std_type = rng.gen();
+                let b: $std_type = rng.gen();
+
+                if a.checked_add(b).is_none() {
+                    return;
+                }
+
+                let bench_run_id: u64 = rng.gen();
+
+                let a_bit = <$bit_type>::alloc(cs.ns(|| format!("{}: a (add)", bench_run_id)), || Ok(a)).unwrap();
+                let b_bit = <$bit_type>::alloc(cs.ns(|| format!("{}: b (add)", bench_run_id)), || Ok(b)).unwrap();
+
+                a_bit
+                    .add(cs.ns(|| format!("{}: a add b", bench_run_id)), &b_bit)
+                    .unwrap();
+            }
+
+            let mut cs = TestConstraintSystem::<Fr>::new();
+
+            let mut rng = XorShiftRng::seed_from_u64(1231275789u64);
+
+            c.bench_function(&format!("integer_arithmetic::{}", $bench_id), |b| {
+                b.iter(|| $foo_name(&mut cs, &mut rng))
+            });
+        }
+    };
+}
+
+macro_rules! create_add_bench_const {
+    ($bench_name:ident, $bench_id:expr, $foo_name:ident, $std_type:ty, $bit_type:ty) => {
+        fn $bench_name(c: &mut Criterion) {
+            fn $foo_name(cs: &mut TestConstraintSystem<Fr>, rng: &mut XorShiftRng) {
+                let a: $std_type = rng.gen();
+                let b: $std_type = rng.gen();
+
+                if a.checked_add(b).is_none() {
+                    return;
+                }
+
+                let bench_run_id: u64 = rng.gen();
+
+                let a_bit_const = <$bit_type>::constant(a);
+                let b_bit_const = <$bit_type>::constant(b);
+                a_bit_const
+                    .add(
+                        cs.ns(|| format!("{}: a add b: const", bench_run_id)),
+                        &b_bit_const,
+                    )
+                    .unwrap();
+            }
+
+            let mut cs = TestConstraintSystem::<Fr>::new();
+
+            let mut rng = XorShiftRng::seed_from_u64(1231275789u64);
+
+            c.bench_function(&format!("integer_arithmetic::{}", $bench_id), |b| {
+                b.iter(|| $foo_name(&mut cs, &mut rng))
+            });
+        }
+    };
+}
+
+macro_rules! create_sub_bench {
+    ($bench_name:ident, $bench_id:expr, $foo_name:ident, $std_type:ty, $bit_type:ty) => {
+        fn $bench_name(c: &mut Criterion) {
+            fn $foo_name(cs: &mut TestConstraintSystem<Fr>, rng: &mut XorShiftRng) {
+                let a: $std_type = rng.gen();
+                let b: $std_type = rng.gen();
+
+                if b.checked_neg().is_none() || a.checked_sub(b).is_none() {
+                    return;
+                }
+
+                let bench_run_id: u64 = rng.gen();
+
+                let a_bit = <$bit_type>::alloc(cs.ns(|| format!("{}: a (sub)", bench_run_id)), || Ok(a)).unwrap();
+                let b_bit = <$bit_type>::alloc(cs.ns(|| format!("{}: b (sub)", bench_run_id)), || Ok(b)).unwrap();
+
+                a_bit
+                    .sub(cs.ns(|| format!("{}: a sub b", bench_run_id)), &b_bit)
+                    .unwrap();
+            }
+
+            let mut cs = TestConstraintSystem::<Fr>::new();
+
+            let mut rng = XorShiftRng::seed_from_u64(1231275789u64);
+
+            c.bench_function(&format!("integer_arithmetic::{}", $bench_id), |b| {
+                b.iter(|| $foo_name(&mut cs, &mut rng))
+            });
+        }
+    };
+}
+
+macro_rules! create_sub_bench_const {
+    ($bench_name:ident, $bench_id:expr, $foo_name:ident, $std_type:ty, $bit_type:ty) => {
+        fn $bench_name(c: &mut Criterion) {
+            fn $foo_name(cs: &mut TestConstraintSystem<Fr>, rng: &mut XorShiftRng) {
+                let a: $std_type = rng.gen();
+                let b: $std_type = rng.gen();
+
+                if b.checked_neg().is_none() || a.checked_sub(b).is_none() {
+                    return;
+                }
+
+                let bench_run_id: u64 = rng.gen();
+
+                let a_bit_const = <$bit_type>::constant(a);
+                let b_bit_const = <$bit_type>::constant(b);
+                a_bit_const
+                    .sub(
+                        cs.ns(|| format!("{}: a sub b: const", bench_run_id)),
+                        &b_bit_const,
+                    )
+                    .unwrap();
+            }
+
+            let mut cs = TestConstraintSystem::<Fr>::new();
+
+            let mut rng = XorShiftRng::seed_from_u64(1231275789u64);
+
+            c.bench_function(&format!("integer_arithmetic::{}", $bench_id), |b| {
+                b.iter(|| $foo_name(&mut cs, &mut rng))
+            });
+        }
+    };
+}
+
+macro_rules! create_mul_bench {
+    ($bench_name:ident, $bench_id:expr, $foo_name:ident, $std_type:ty, $bit_type:ty) => {
+        fn $bench_name(c: &mut Criterion) {
+            fn $foo_name(cs: &mut TestConstraintSystem<Fr>, rng: &mut XorShiftRng) {
+                let a: $std_type = rng.gen();
+                let b: $std_type = rng.gen();
+
+                if a.checked_mul(b).is_none() {
+                    return;
+                }
+
+                let bench_run_id: u64 = rng.gen();
+
+                let a_bit = <$bit_type>::alloc(cs.ns(|| format!("{}: a (mul)", bench_run_id)), || Ok(a)).unwrap();
+                let b_bit = <$bit_type>::alloc(cs.ns(|| format!("{}: b (mul)", bench_run_id)), || Ok(b)).unwrap();
+                a_bit
+                    .mul(cs.ns(|| format!("{}: a mul b", bench_run_id)), &b_bit)
+                    .unwrap();
+            }
+
+            let mut cs = TestConstraintSystem::<Fr>::new();
+
+            let mut rng = XorShiftRng::seed_from_u64(1231275789u64);
+
+            c.bench_function(&format!("integer_arithmetic::{}", $bench_id), |b| {
+                b.iter(|| $foo_name(&mut cs, &mut rng))
+            });
+        }
+    };
+}
+
+macro_rules! create_mul_bench_const {
+    ($bench_name:ident, $bench_id:expr, $foo_name:ident, $std_type:ty, $bit_type:ty) => {
+        fn $bench_name(c: &mut Criterion) {
+            fn $foo_name(cs: &mut TestConstraintSystem<Fr>, rng: &mut XorShiftRng) {
+                let a: $std_type = rng.gen();
+                let b: $std_type = rng.gen();
+
+                if a.checked_mul(b).is_none() {
+                    return;
+                }
+
+                let bench_run_id: u64 = rng.gen();
+
+                let a_bit_const = <$bit_type>::constant(a);
+                let b_bit_const = <$bit_type>::constant(b);
+                a_bit_const
+                    .mul(
+                        cs.ns(|| format!("{}: a mul b: const", bench_run_id)),
+                        &b_bit_const,
+                    )
+                    .unwrap();
+            }
+
+            let mut cs = TestConstraintSystem::<Fr>::new();
+
+            let mut rng = XorShiftRng::seed_from_u64(1231275789u64);
+
+            c.bench_function(&format!("integer_arithmetic::{}", $bench_id), |b| {
+                b.iter(|| $foo_name(&mut cs, &mut rng))
+            });
+        }
+    };
+}
+
+macro_rules! create_div_bench {
+    ($bench_name:ident, $bench_id:expr, $foo_name:ident, $std_type:ty, $bit_type:ty) => {
+        fn $bench_name(c: &mut Criterion) {
+            fn $foo_name(cs: &mut TestConstraintSystem<Fr>, rng: &mut XorShiftRng) {
+                let a: $std_type = rng.gen();
+                let b: $std_type = rng.gen();
+
+                if a.checked_neg().is_none() || a.checked_div(b).is_none() {
+                    return;
+                }
+
+                let bench_run_id: u64 = rng.gen();
+
+                let a_bit = <$bit_type>::alloc(cs.ns(|| format!("{}: a (div)", bench_run_id)), || Ok(a)).unwrap();
+                let b_bit = <$bit_type>::alloc(cs.ns(|| format!("{}: b (div)", bench_run_id)), || Ok(b)).unwrap();
+                a_bit
+                    .div(cs.ns(|| format!("{}: a div b", bench_run_id)), &b_bit)
+                    .unwrap();
+            }
+
+            let mut cs = TestConstraintSystem::<Fr>::new();
+
+            let mut rng = XorShiftRng::seed_from_u64(1231275789u64);
+
+            c.bench_function(&format!("integer_arithmetic::{}", $bench_id), |b| {
+                b.iter(|| $foo_name(&mut cs, &mut rng))
+            });
+        }
+    };
+}
+
+macro_rules! create_div_bench_const {
+    ($bench_name:ident, $bench_id:expr, $foo_name:ident, $std_type:ty, $bit_type:ty) => {
+        fn $bench_name(c: &mut Criterion) {
+            fn $foo_name(cs: &mut TestConstraintSystem<Fr>, rng: &mut XorShiftRng) {
+                let a: $std_type = rng.gen();
+                let b: $std_type = rng.gen();
+
+                if a.checked_neg().is_none() || a.checked_div(b).is_none() {
+                    return;
+                }
+
+                let bench_run_id: u64 = rng.gen();
+
+                let a_bit_const = <$bit_type>::constant(a);
+                let b_bit_const = <$bit_type>::constant(b);
+                a_bit_const
+                    .div(
+                        cs.ns(|| format!("{}: a div b: const", bench_run_id)),
+                        &b_bit_const,
+                    )
+                    .unwrap();
+            }
+
+            let mut cs = TestConstraintSystem::<Fr>::new();
+
+            let mut rng = XorShiftRng::seed_from_u64(1231275789u64);
+
+            c.bench_function(&format!("integer_arithmetic::{}", $bench_id), |b| {
+                b.iter(|| $foo_name(&mut cs, &mut rng))
+            });
+        }
+    };
+}
+
+create_add_bench!(bench_i8_add, "i8_add", i8_add, i8, Int8);
+create_add_bench!(bench_i16_add, "i16_add", i16_add, i16, Int16);
+create_add_bench!(bench_i32_add, "i32_add", i32_add, i32, Int32);
+create_add_bench!(bench_i64_add, "i64_add", i64_add, i64, Int64);
+create_add_bench!(bench_i128_add, "i128_add", i128_add, i128, Int128);
+
+create_add_bench_const!(bench_i8_add_const, "i8_add_const", i8_add, i8, Int8);
+create_add_bench_const!(bench_i16_add_const, "i16_add_const", i16_add, i16, Int16);
+create_add_bench_const!(bench_i32_add_const, "i32_add_const", i32_add, i32, Int32);
+create_add_bench_const!(bench_i64_add_const, "i64_add_const", i64_add, i64, Int64);
+create_add_bench_const!(bench_i128_add_const, "i128_add_const", i128_add, i128, Int128);
+
+create_sub_bench!(bench_i8_sub, "i8_sub", i8_sub, i8, Int8);
+create_sub_bench!(bench_i16_sub, "i16_sub", i16_sub, i16, Int16);
+create_sub_bench!(bench_i32_sub, "i32_sub", i32_sub, i32, Int32);
+create_sub_bench!(bench_i64_sub, "i64_sub", i64_sub, i64, Int64);
+create_sub_bench!(bench_i128_sub, "i128_sub", i128_sub, i128, Int128);
+
+create_sub_bench_const!(bench_i8_sub_const, "i8_sub_const", i8_sub, i8, Int8);
+create_sub_bench_const!(bench_i16_sub_const, "i16_sub_const", i16_sub, i16, Int16);
+create_sub_bench_const!(bench_i32_sub_const, "i32_sub_const", i32_sub, i32, Int32);
+create_sub_bench_const!(bench_i64_sub_const, "i64_sub_const", i64_sub, i64, Int64);
+create_sub_bench_const!(bench_i128_sub_const, "i128_sub_const", i128_sub, i128, Int128);
+
+create_mul_bench!(bench_i8_mul, "i8_mul", i8_mul, i8, Int8);
+create_mul_bench!(bench_i16_mul, "i16_mul", i16_mul, i16, Int16);
+create_mul_bench!(bench_i32_mul, "i32_mul", i32_mul, i32, Int32);
+create_mul_bench!(bench_i64_mul, "i64_mul", i64_mul, i64, Int64);
+create_mul_bench!(bench_i128_mul, "i128_mul", i128_mul, i128, Int128);
+
+create_mul_bench_const!(bench_i8_mul_const, "i8_mul_const", i8_mul, i8, Int8);
+create_mul_bench_const!(bench_i16_mul_const, "i16_mul_const", i16_mul, i16, Int16);
+create_mul_bench_const!(bench_i32_mul_const, "i32_mul_const", i32_mul, i32, Int32);
+create_mul_bench_const!(bench_i64_mul_const, "i64_mul_const", i64_mul, i64, Int64);
+create_mul_bench_const!(bench_i128_mul_const, "i128_mul_const", i128_mul, i128, Int128);
+
+create_div_bench!(bench_i8_div, "i8_div", i8_div, i8, Int8);
+create_div_bench!(bench_i16_div, "i16_div", i16_div, i16, Int16);
+create_div_bench!(bench_i32_div, "i32_div", i32_div, i32, Int32);
+// create_div_bench!(bench_i64_div, "i64_div", i64_div, i64, Int64);
+// create_div_bench!(bench_i128_div, "i128_div", i128_div, i128, Int128);
+
+create_div_bench_const!(bench_i8_div_const, "i8_div_const", i8_div, i8, Int8);
+create_div_bench_const!(bench_i16_div_const, "i16_div_const", i16_div, i16, Int16);
+create_div_bench_const!(bench_i32_div_const, "i32_div_const", i32_div, i32, Int32);
+// create_div_bench_const!(bench_i64_div_const, "i64_div_const", i64_div, i64, Int64);
+// create_div_bench_const!(bench_i128_div_const, "i128_div_const", i128_div, i128, Int128);
+
+criterion_group!(
+    name = benches_add;
+    config = Criterion::default();
+    targets = bench_i8_add,
+    bench_i16_add,
+    bench_i32_add,
+    bench_i64_add,
+    bench_i128_add,
+);
+
+criterion_group!(
+    name = benches_add_const;
+    config = Criterion::default();
+    targets = bench_i8_add_const,
+    bench_i16_add_const,
+    bench_i32_add_const,
+    bench_i64_add_const,
+    bench_i128_add_const,
+);
+
+criterion_group!(
+    name = benches_sub;
+    config = Criterion::default();
+    targets = bench_i8_sub,
+    bench_i16_sub,
+    bench_i32_sub,
+    bench_i64_sub,
+    bench_i128_sub,
+);
+
+criterion_group!(
+    name = benches_sub_const;
+    config = Criterion::default();
+    targets = bench_i8_sub_const,
+    bench_i16_sub_const,
+    bench_i32_sub_const,
+    bench_i64_sub_const,
+    bench_i128_sub_const,
+);
+
+criterion_group!(
+    name = benches_mul;
+    config = Criterion::default();
+    targets = bench_i8_mul,
+    bench_i16_mul,
+    bench_i32_mul,
+    bench_i64_mul,
+    bench_i128_mul,
+);
+
+criterion_group!(
+    name = benches_mul_const;
+    config = Criterion::default();
+    targets = bench_i8_mul_const,
+    bench_i16_mul_const,
+    bench_i32_mul_const,
+    bench_i64_mul_const,
+    bench_i128_mul_const,
+);
+
+criterion_group!(
+    name = benches_div;
+    config = Criterion::default();
+    targets = bench_i8_div,
+    bench_i16_div,
+    bench_i32_div,
+    // bench_i64_div,
+    // bench_i128_div,
+);
+
+criterion_group!(
+    name = benches_div_const;
+    config = Criterion::default();
+    targets = bench_i8_div_const,
+    bench_i16_div_const,
+    bench_i32_div_const,
+    // bench_i64_div_const,
+    // bench_i128_div_const,
+);
+
+criterion_main!(
+    benches_add,
+    benches_add_const,
+    benches_sub,
+    benches_sub_const,
+    benches_mul,
+    benches_mul_const,
+    benches_div,
+    benches_div_const
+);
diff --git a/gadgets/src/arithmetic/neg.rs b/gadgets/src/arithmetic/neg.rs
index ba199c8eda..a35f474f46 100644
--- a/gadgets/src/arithmetic/neg.rs
+++ b/gadgets/src/arithmetic/neg.rs
@@ -22,6 +22,8 @@ use snarkos_models::{
     gadgets::{r1cs::ConstraintSystem, utilities::boolean::Boolean},
 };
 
+use std::iter;
+
 /// Returns a negated representation of `self` in the constraint system.
 pub trait Neg<F: Field>
 where
@@ -40,8 +42,9 @@ impl<F: Field> Neg<F> for Vec<Boolean> {
         let flipped: Self = self.iter().map(|bit| bit.not()).collect();
 
         // add one
-        let mut one = vec![Boolean::constant(true)];
-        one.append(&mut vec![Boolean::Constant(false); self.len() - 1]);
+        let mut one = Vec::with_capacity(self.len());
+        one.push(Boolean::constant(true));
+        one.extend(iter::repeat(Boolean::Constant(false)).take(self.len() - 1));
 
         let mut bits = flipped.add_bits(cs.ns(|| "add one"), &one)?;
         let _carry = bits.pop(); // we already accounted for overflow above
diff --git a/gadgets/src/bits/sign_extend.rs b/gadgets/src/bits/sign_extend.rs
index af0e5c2892..4bdede76d8 100644
--- a/gadgets/src/bits/sign_extend.rs
+++ b/gadgets/src/bits/sign_extend.rs
@@ -16,6 +16,8 @@
 
 use snarkos_models::gadgets::utilities::boolean::Boolean;
 
+use std::iter;
+
 /// Sign extends an array of bits to the desired length.
 /// Expects least significant bit first
 pub trait SignExtend
@@ -30,10 +32,10 @@ impl SignExtend for Boolean {
     fn sign_extend(bits: &[Boolean], length: usize) -> Vec<Boolean> {
         let msb = bits.last().expect("empty bit list");
         let bits_needed = length - bits.len();
-        let mut extension = vec![*msb; bits_needed];
 
-        let mut result = Vec::from(bits);
-        result.append(&mut extension);
+        let mut result = Vec::with_capacity(length);
+        result.extend_from_slice(bits);
+        result.extend(iter::repeat(*msb).take(bits_needed));
 
         result
     }
diff --git a/gadgets/src/signed_integer/arithmetic/div.rs b/gadgets/src/signed_integer/arithmetic/div.rs
index 774eae8bc0..2d12584bd6 100644
--- a/gadgets/src/signed_integer/arithmetic/div.rs
+++ b/gadgets/src/signed_integer/arithmetic/div.rs
@@ -163,7 +163,7 @@ macro_rules! div_int_impl {
                 )?;
 
                 let mut q = zero.clone();
-                let mut r = zero.clone();
+                let mut r = zero;
 
                 let mut index = <$gadget as Int>::SIZE - 1 as usize;
                 let mut bit_value = (1 as <$gadget as Int>::IntegerType) << ((index - 1) as <$gadget as Int>::IntegerType);
@@ -179,7 +179,7 @@ macro_rules! div_int_impl {
                     // Set the least-significant bit of remainder to bit i of the numerator
                     let r_new = r.add(
                         &mut cs.ns(|| format!("set_remainder_bit_{}", i)),
-                        &one.clone(),
+                        &one,
                     )?;
 
                     r = Self::conditionally_select(
@@ -197,20 +197,22 @@ macro_rules! div_int_impl {
                     let sub = r.sub(
                         &mut cs.ns(|| format!("subtract_divisor_{}", i)),
                         &b
-                    );
+                    )?;
 
                     r = Self::conditionally_select(
                         &mut cs.ns(|| format!("subtract_or_same_{}", i)),
                         &can_sub,
-                        &sub?,
+                        &sub,
                         &r
                     )?;
 
                     index -= 1;
 
                     let mut q_new = q.clone();
-                    q_new.bits[index] = true_bit.clone();
-                    q_new.value = Some(q_new.value.unwrap() + bit_value);
+                    q_new.bits[index] = true_bit;
+                    if let Some(ref mut value) = q_new.value {
+                        *value += bit_value;
+                    }
 
                     bit_value >>= 1;
 
diff --git a/gadgets/src/signed_integer/int_impl.rs b/gadgets/src/signed_integer/int_impl.rs
index 074723cfee..cb90ea2099 100644
--- a/gadgets/src/signed_integer/int_impl.rs
+++ b/gadgets/src/signed_integer/int_impl.rs
@@ -82,18 +82,8 @@ macro_rules! int_impl {
             }
 
             fn is_constant(&self) -> bool {
-                let mut constant = true;
-
                 // If any bits of self are allocated bits, return false
-                for bit in &self.bits {
-                    match *bit {
-                        Boolean::Is(ref _bit) => constant = false,
-                        Boolean::Not(ref _bit) => constant = false,
-                        Boolean::Constant(_bit) => {}
-                    }
-                }
-
-                constant
+                self.bits.iter().all(|bit| matches!(bit, Boolean::Constant(_)))
             }
         }
     };
diff --git a/gadgets/src/signed_integer/relational/eq.rs b/gadgets/src/signed_integer/relational/eq.rs
index 0deb39d145..d783deb340 100644
--- a/gadgets/src/signed_integer/relational/eq.rs
+++ b/gadgets/src/signed_integer/relational/eq.rs
@@ -53,7 +53,8 @@ macro_rules! eq_gadget_impl {
 
         impl PartialEq for $gadget {
             fn eq(&self, other: &Self) -> bool {
-                !self.value.is_none() && !other.value.is_none() && self.value == other.value
+                // self.value == other.value means that other.value.is_some() too
+                self.value.is_some() && self.value == other.value
             }
         }
 
diff --git a/gadgets/src/signed_integer/utilities/alloc.rs b/gadgets/src/signed_integer/utilities/alloc.rs
index 442abd7b6e..b47016a10b 100644
--- a/gadgets/src/signed_integer/utilities/alloc.rs
+++ b/gadgets/src/signed_integer/utilities/alloc.rs
@@ -16,7 +16,7 @@
 
 use crate::{Int, Int128, Int16, Int32, Int64, Int8};
 
-use core::borrow::Borrow;
+use core::{borrow::Borrow, iter};
 use snarkos_errors::gadgets::SynthesisError;
 use snarkos_models::{
     curves::Field,
@@ -29,6 +29,21 @@ use snarkos_models::{
     },
 };
 
+fn create_value<T: Borrow<bool>, I: IntoIterator<Item = Option<T>>, F: Field, CS: ConstraintSystem<F>>(
+    cs: &mut CS,
+    iter: I,
+) -> Result<Vec<Boolean>, SynthesisError> {
+    iter.into_iter()
+        .enumerate()
+        .map(|(i, v)| {
+            Ok(Boolean::from(AllocatedBit::alloc(
+                &mut cs.ns(|| format!("allocated bit_gadget {}", i)),
+                || v.ok_or(SynthesisError::AssignmentMissing),
+            )?))
+        })
+        .collect()
+}
+
 macro_rules! alloc_int_impl {
     ($($gadget: ident)*) => ($(
         impl<F: Field> AllocGadget<<$gadget as Int>::IntegerType, F> for $gadget {
@@ -41,30 +56,21 @@ macro_rules! alloc_int_impl {
                 value_gen: Fn,
             ) -> Result<Self, SynthesisError> {
                 let value = value_gen().map(|val| *val.borrow());
-                let values = match value {
+
+                let bits = match value {
                     Ok(mut val) => {
                         let mut v = Vec::with_capacity(<$gadget as Int>::SIZE);
-
                         for _ in 0..<$gadget as Int>::SIZE {
                             v.push(Some(val & 1 == 1));
                             val >>= 1;
                         }
-
-                        v
+                        create_value(&mut cs, v)
                     }
-                    _ => vec![None; <$gadget as Int>::SIZE],
-                };
-
-                let bits = values
-                    .into_iter()
-                    .enumerate()
-                    .map(|(i, v)| {
-                        Ok(Boolean::from(AllocatedBit::alloc(
-                            &mut cs.ns(|| format!("allocated bit_gadget {}", i)),
-                            || v.ok_or(SynthesisError::AssignmentMissing),
-                        )?))
-                    })
-                    .collect::<Result<Vec<_>, SynthesisError>>()?;
+                    Err(_) => {
+                        let i = iter::repeat(None::<bool>).take(<$gadget as Int>::SIZE);
+                        create_value(&mut cs, i)
+                    },
+                }?;
 
                 Ok(Self {
                     bits,
@@ -81,30 +87,21 @@ macro_rules! alloc_int_impl {
                 value_gen: Fn,
             ) -> Result<Self, SynthesisError> {
                 let value = value_gen().map(|val| *val.borrow());
-                let values = match value {
+
+                let bits = match value {
                     Ok(mut val) => {
                         let mut v = Vec::with_capacity(<$gadget as Int>::SIZE);
-
                         for _ in 0..<$gadget as Int>::SIZE {
                             v.push(Some(val & 1 == 1));
                             val >>= 1;
                         }
-
-                        v
+                        create_value(&mut cs, v)
                     }
-                    _ => vec![None; <$gadget as Int>::SIZE],
-                };
-
-                let bits = values
-                    .into_iter()
-                    .enumerate()
-                    .map(|(i, v)| {
-                        Ok(Boolean::from(AllocatedBit::alloc(
-                            &mut cs.ns(|| format!("allocated bit_gadget {}", i)),
-                            || v.ok_or(SynthesisError::AssignmentMissing),
-                        )?))
-                    })
-                    .collect::<Result<Vec<_>, SynthesisError>>()?;
+                    Err(_) => {
+                        let i = iter::repeat(None::<bool>).take(<$gadget as Int>::SIZE);
+                        create_value(&mut cs, i)
+                    },
+                }?;
 
                 Ok(Self {
                     bits,
diff --git a/gadgets/src/signed_integer/utilities/select.rs b/gadgets/src/signed_integer/utilities/select.rs
index 81d36531fb..7a5f8932a0 100644
--- a/gadgets/src/signed_integer/utilities/select.rs
+++ b/gadgets/src/signed_integer/utilities/select.rs
@@ -49,25 +49,17 @@ macro_rules! select_int_impl {
                         }
                     });
 
-                    let result = Self::alloc(cs.ns(|| "cond_select_result"), || result_val.get().map(|v| v))?;
+                    let result = Self::alloc(cs.ns(|| "cond_select_result"), || result_val.get())?;
 
-                    let expected_bits = first
-                        .bits
-                        .iter()
-                        .zip(&second.bits)
-                        .enumerate()
-                        .map(|(i, (a, b))| {
-                            Boolean::conditionally_select(
-                                &mut cs.ns(|| format!("{}_cond_select_{}", <$gadget as Int>::SIZE, i)),
-                                cond,
-                                a,
-                                b,
-                            ).unwrap()
-                        })
-                        .collect::<Vec<Boolean>>();
+                    for (i, ((bit1, bit2), actual)) in first.bits.iter().zip(second.bits.iter()).zip(result.bits.iter()).enumerate() {
+                        let expected = Boolean::conditionally_select(
+                            &mut cs.ns(|| format!("{}_cond_select_{}", <$gadget as Int>::SIZE, i)),
+                            cond,
+                            bit1,
+                            bit2,
+                        ).unwrap();
 
-                    for (i, (actual, expected)) in result.bits.iter().zip(expected_bits.iter()).enumerate() {
-                        actual.enforce_equal(&mut cs.ns(|| format!("selected_result_bit_{}", i)), expected)?;
+                        actual.enforce_equal(&mut cs.ns(|| format!("selected_result_bit_{}", i)), &expected)?;
                     }
 
                     Ok(result)
diff --git a/leo/updater.rs b/leo/updater.rs
index ec730963fa..84f088f68f 100644
--- a/leo/updater.rs
+++ b/leo/updater.rs
@@ -1,3 +1,18 @@
+// Copyright (C) 2019-2020 Aleo Systems Inc.
+// This file is part of the Leo library.
+
+// The Leo library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// The Leo library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with the Leo library. If not, see <https://www.gnu.org/licenses/>.
 use crate::config::Config;
 
 use colored::Colorize;
diff --git a/package/src/errors/package.rs b/package/src/errors/package.rs
index c2d89f0219..73694d3259 100644
--- a/package/src/errors/package.rs
+++ b/package/src/errors/package.rs
@@ -1,3 +1,18 @@
+// Copyright (C) 2019-2020 Aleo Systems Inc.
+// This file is part of the Leo library.
+
+// The Leo library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// The Leo library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with the Leo library. If not, see <https://www.gnu.org/licenses/>.
 use std::{ffi::OsString, io};
 
 #[derive(Debug, Error)]
diff --git a/typed/benches/typed_ast.rs b/typed/benches/typed_ast.rs
index 6fbda56b7a..cee9a756d4 100644
--- a/typed/benches/typed_ast.rs
+++ b/typed/benches/typed_ast.rs
@@ -18,10 +18,7 @@ use leo_ast::LeoAst;
 use leo_typed::LeoTypedAst;
 
 use criterion::{criterion_group, criterion_main, Criterion};
-use std::{
-    path::{Path, PathBuf},
-    time::Duration,
-};
+use std::{path::Path, time::Duration};
 
 fn leo_typed_ast<'ast>(ast: &LeoAst<'ast>) -> LeoTypedAst {
     LeoTypedAst::new("leo_typed_tree", &ast)
diff --git a/typed/src/expression.rs b/typed/src/expression.rs
index 2c6e7163ad..1a8734448c 100644
--- a/typed/src/expression.rs
+++ b/typed/src/expression.rs
@@ -118,39 +118,39 @@ pub enum Expression {
 }
 
 impl Expression {
-    pub fn set_span(&mut self, new_span: &Span) {
+    pub fn set_span(&mut self, new_span: Span) {
         match self {
-            Expression::Field(_, old_span) => *old_span = new_span.clone(),
+            Expression::Field(_, old_span) => *old_span = new_span,
             Expression::Group(value) => value.set_span(new_span),
 
-            Expression::Add(_, old_span) => *old_span = new_span.clone(),
-            Expression::Sub(_, old_span) => *old_span = new_span.clone(),
-            Expression::Mul(_, old_span) => *old_span = new_span.clone(),
-            Expression::Div(_, old_span) => *old_span = new_span.clone(),
-            Expression::Pow(_, old_span) => *old_span = new_span.clone(),
-
-            Expression::Not(_, old_span) => *old_span = new_span.clone(),
-            Expression::Or(_, old_span) => *old_span = new_span.clone(),
-            Expression::And(_, old_span) => *old_span = new_span.clone(),
-            Expression::Eq(_, old_span) => *old_span = new_span.clone(),
-            Expression::Ge(_, old_span) => *old_span = new_span.clone(),
-            Expression::Gt(_, old_span) => *old_span = new_span.clone(),
-            Expression::Le(_, old_span) => *old_span = new_span.clone(),
-            Expression::Lt(_, old_span) => *old_span = new_span.clone(),
-
-            Expression::IfElse(_, old_span) => *old_span = new_span.clone(),
-            Expression::Array(_, old_span) => *old_span = new_span.clone(),
-            Expression::ArrayAccess(_, old_span) => *old_span = new_span.clone(),
-
-            Expression::Tuple(_, old_span) => *old_span = new_span.clone(),
-            Expression::TupleAccess(_, _, old_span) => *old_span = new_span.clone(),
-
-            Expression::Circuit(_, _, old_span) => *old_span = new_span.clone(),
-            Expression::CircuitMemberAccess(_, _, old_span) => *old_span = new_span.clone(),
-            Expression::CircuitStaticFunctionAccess(_, _, old_span) => *old_span = new_span.clone(),
-
-            Expression::FunctionCall(_, _, old_span) => *old_span = new_span.clone(),
-            Expression::CoreFunctionCall(_, _, old_span) => *old_span = new_span.clone(),
+            Expression::Add(_, old_span) => *old_span = new_span,
+            Expression::Sub(_, old_span) => *old_span = new_span,
+            Expression::Mul(_, old_span) => *old_span = new_span,
+            Expression::Div(_, old_span) => *old_span = new_span,
+            Expression::Pow(_, old_span) => *old_span = new_span,
+
+            Expression::Not(_, old_span) => *old_span = new_span,
+            Expression::Or(_, old_span) => *old_span = new_span,
+            Expression::And(_, old_span) => *old_span = new_span,
+            Expression::Eq(_, old_span) => *old_span = new_span,
+            Expression::Ge(_, old_span) => *old_span = new_span,
+            Expression::Gt(_, old_span) => *old_span = new_span,
+            Expression::Le(_, old_span) => *old_span = new_span,
+            Expression::Lt(_, old_span) => *old_span = new_span,
+
+            Expression::IfElse(_, old_span) => *old_span = new_span,
+            Expression::Array(_, old_span) => *old_span = new_span,
+            Expression::ArrayAccess(_, old_span) => *old_span = new_span,
+
+            Expression::Tuple(_, old_span) => *old_span = new_span,
+            Expression::TupleAccess(_, _, old_span) => *old_span = new_span,
+
+            Expression::Circuit(_, _, old_span) => *old_span = new_span,
+            Expression::CircuitMemberAccess(_, _, old_span) => *old_span = new_span,
+            Expression::CircuitStaticFunctionAccess(_, _, old_span) => *old_span = new_span,
+
+            Expression::FunctionCall(_, _, old_span) => *old_span = new_span,
+            Expression::CoreFunctionCall(_, _, old_span) => *old_span = new_span,
             _ => {}
         }
     }
diff --git a/typed/src/groups/group_value.rs b/typed/src/groups/group_value.rs
index 01ef67253d..26f902d615 100644
--- a/typed/src/groups/group_value.rs
+++ b/typed/src/groups/group_value.rs
@@ -36,10 +36,10 @@ pub enum GroupValue {
 }
 
 impl GroupValue {
-    pub fn set_span(&mut self, new_span: &Span) {
+    pub fn set_span(&mut self, new_span: Span) {
         match self {
-            GroupValue::Single(_, old_span) => *old_span = new_span.clone(),
-            GroupValue::Tuple(tuple) => tuple.span = new_span.clone(),
+            GroupValue::Single(_, old_span) => *old_span = new_span,
+            GroupValue::Tuple(tuple) => tuple.span = new_span,
         }
     }
 }
diff --git a/typed/src/statements/statement.rs b/typed/src/statements/statement.rs
index 841e954c93..36da315e43 100644
--- a/typed/src/statements/statement.rs
+++ b/typed/src/statements/statement.rs
@@ -58,7 +58,7 @@ impl<'ast> From<DefinitionStatement<'ast>> for Statement {
             .into_iter()
             .map(|e| {
                 let mut expression = Expression::from(e);
-                expression.set_span(&span);
+                expression.set_span(span.clone());
 
                 expression
             })
@@ -155,7 +155,7 @@ impl<'ast> From<ExpressionStatement<'ast>> for Statement {
         let span = Span::from(statement.span);
         let mut expression = Expression::from(statement.expression);
 
-        expression.set_span(&span);
+        expression.set_span(span.clone());
 
         Statement::Expression(expression, span)
     }
diff --git a/typed/src/types/type_.rs b/typed/src/types/type_.rs
index 9c0e5b2a87..e27918c020 100644
--- a/typed/src/types/type_.rs
+++ b/typed/src/types/type_.rs
@@ -45,17 +45,11 @@ pub enum Type {
 
 impl Type {
     pub fn is_self(&self) -> bool {
-        if let Type::SelfType = self {
-            return true;
-        }
-        false
+        matches!(self, Type::SelfType)
     }
 
     pub fn is_circuit(&self) -> bool {
-        if let Type::Circuit(_) = self {
-            return true;
-        }
-        false
+        matches!(self, Type::Circuit(_))
     }
 
     pub fn match_array_types(&self, other: &Type) -> bool {
diff --git a/typed/tests/serialization/json.rs b/typed/tests/serialization/json.rs
index 732747553f..c952a8b992 100644
--- a/typed/tests/serialization/json.rs
+++ b/typed/tests/serialization/json.rs
@@ -15,7 +15,9 @@
 // along with the Leo library. If not, see <https://www.gnu.org/licenses/>.
 
 use leo_ast::LeoAst;
-use leo_typed::{LeoTypedAst, Program};
+use leo_typed::LeoTypedAst;
+#[cfg(not(feature = "ci_skip"))]
+use leo_typed::Program;
 
 use std::path::{Path, PathBuf};
 
