diff --git a/compiler/src/value/value.rs b/compiler/src/value/value.rs
index 94f47d7295..d50f93f2fb 100644
--- a/compiler/src/value/value.rs
+++ b/compiler/src/value/value.rs
@@ -457,7 +457,7 @@ impl<F: Field + PrimeField, G: GroupType<F>> CondSelectGadget<F> for Constrained
 
                 ConstrainedValue::Array(array)
             }
-            (ConstrainedValue::Tuple(tuple_1), ConstrainedValue::Array(tuple_2)) => {
+            (ConstrainedValue::Tuple(tuple_1), ConstrainedValue::Tuple(tuple_2)) => {
                 let mut array = Vec::with_capacity(tuple_1.len());
 
                 for (i, (first, second)) in tuple_1.iter().zip(tuple_2.iter()).enumerate() {
diff --git a/compiler/tests/function/mod.rs b/compiler/tests/function/mod.rs
index 8d8d33e65a..dc31094f76 100644
--- a/compiler/tests/function/mod.rs
+++ b/compiler/tests/function/mod.rs
@@ -184,3 +184,11 @@ fn test_return_tuple() {
 
     assert_satisfied(program);
 }
+
+#[test]
+fn test_return_tuple_conditional() {
+    let bytes = include_bytes!("return_tuple_conditional.leo");
+    let program = parse_program(bytes).unwrap();
+
+    assert_satisfied(program);
+}
diff --git a/compiler/tests/function/return_tuple_conditional.leo b/compiler/tests/function/return_tuple_conditional.leo
new file mode 100644
index 0000000000..c63967f548
--- /dev/null
+++ b/compiler/tests/function/return_tuple_conditional.leo
@@ -0,0 +1,15 @@
+// Returns a tuple using a conditional "if" statement.
+function tuple_conditional () -> (
+    i64,
+    i64
+) {
+    if true {
+        return (1, 1)
+    } else {
+        return (2, 2)
+    }
+}
+
+function main() {
+    let t = tuple_conditional();
+}
\ No newline at end of file
