diff --git a/.circleci/config.yml b/.circleci/config.yml
index e840bc91a4..4e6dc7bbcb 100755
--- a/.circleci/config.yml
+++ b/.circleci/config.yml
@@ -6,7 +6,7 @@ version: 2.1
 # - Incremental builds should be disabled in CI, since they don't provide much benefit.
 # - Enabling `--only_testnet` feature flag to reduce the amount of time spent building leo-lang, since it's not needed for the test suite.
 
-# Rust Version: 1.82.0
+# Rust Version: 1.85.1
 # Ensure that this matches the `rust-version` in `Cargo.toml`.
 # If this is changed, propogate the changes to all places in this file, including the `install-rust` command.
 
@@ -32,7 +32,7 @@ executors:
 
   linux-executor:
     docker:
-      - image: "cimg/rust:1.82.0"  # Ensure that this matches the `rust-version` in `Cargo.toml`.
+      - image: "cimg/rust:1.85"  # Ensure that this matches the `rust-version` in `Cargo.toml`.
     resource_class: xlarge
 
 commands:
@@ -43,14 +43,15 @@ commands:
           name: Install Rust
           command: |
             # If Rust is not installed on the machine, install it
-            if ! command -v rustc &> /dev/null; then
+            # (temporarily install it regardless, as we want 1.85.1, not 1.85.0)
+            # if ! command -v rustc &> /dev/null; then
               curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh -s -- -y
               source "$HOME/.cargo/env"
-              rustup install 1.82.0
-              rustup override set 1.82.0
+              rustup install 1.85.1
+              rustup override set 1.85.1
               cargo --version --verbose
               rustc --version
-            fi
+            # fi
             if [ ! -f "Cargo.lock" ]; then
               cargo generate-lockfile
             fi
@@ -69,8 +70,8 @@ commands:
             Invoke-WebRequest -Uri "https://win.rustup.rs/" -OutFile "C:\\rustup-init.exe"
             Start-Process "C:\\rustup-init.exe" -ArgumentList '-y' -Wait
             $Env:Path += ";$Env:USERPROFILE\.cargo\bin"
-            rustup install 1.82.0
-            rustup default 1.82.0
+            rustup install 1.85.1
+            rustup default 1.85.1
             cargo --version --verbose
             rustc --version | Out-File -FilePath "rust-version"
             if (!(Test-Path "Cargo.lock" -PathType Leaf)) {
@@ -250,7 +251,7 @@ jobs:
 
   leo-new:
     docker:
-      - image: cimg/rust:1.82
+      - image: cimg/rust:1.85
     resource_class: xlarge
     steps:
       - attach_workspace:
@@ -263,7 +264,7 @@ jobs:
 
   leo-clean:
     docker:
-      - image: cimg/rust:1.82
+      - image: cimg/rust:1.85
     resource_class: xlarge
     steps:
       - attach_workspace:
@@ -276,7 +277,7 @@ jobs:
 
   test-examples:
     docker:
-      - image: cimg/rust:1.82
+      - image: cimg/rust:1.85
     resource_class: xlarge
     steps:
       - attach_workspace:
diff --git a/Cargo.toml b/Cargo.toml
index ae87e8b212..faa8c3e448 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -26,8 +26,8 @@ include = [
   "examples/token/run.sh"
 ]
 license = "GPL-3.0"
-edition = "2021"
-rust-version = "1.82.0"
+edition = "2024"
+rust-version = "1.85.1"
 
 [workspace]
 members = [
diff --git a/compiler/ast/Cargo.toml b/compiler/ast/Cargo.toml
index 45a043a8d9..f1d3f094af 100644
--- a/compiler/ast/Cargo.toml
+++ b/compiler/ast/Cargo.toml
@@ -15,8 +15,8 @@ keywords = [
 categories = [ "compilers", "cryptography", "web-programming" ]
 include = [ "Cargo.toml", "src", "README.md", "LICENSE.md" ]
 license = "GPL-3.0"
-edition = "2021"
-rust-version = "1.82.0"
+edition = "2024"
+rust-version = "1.85.1"
 
 [dependencies.snarkvm]
 workspace = true
diff --git a/compiler/ast/src/lib.rs b/compiler/ast/src/lib.rs
index 102cb85e94..d26252b5bf 100644
--- a/compiler/ast/src/lib.rs
+++ b/compiler/ast/src/lib.rs
@@ -55,12 +55,11 @@ pub use self::statement::*;
 pub mod types;
 pub use self::types::*;
 
-pub mod value;
-
 pub mod stub;
 pub use self::stub::*;
 
-pub use self::value::*;
+mod value;
+pub use value::*;
 
 pub use common::node::*;
 
diff --git a/compiler/ast/src/value/mod.rs b/compiler/ast/src/value/mod.rs
index 5ce262ba8e..3e4d8f6756 100644
--- a/compiler/ast/src/value/mod.rs
+++ b/compiler/ast/src/value/mod.rs
@@ -26,11 +26,6 @@ use std::{
     ops::{BitAnd, BitOr, BitXor, Not},
 };
 
-// TODO: Consider refactoring this module to use the console implementations from snarkVM.
-
-// This is temporary since the currently unused code is used in constant folding.
-#[allow(dead_code)]
-
 // Macro for making implementing unary operations over appropriate types easier.
 macro_rules! implement_const_unary {
     (
diff --git a/compiler/compiler/Cargo.toml b/compiler/compiler/Cargo.toml
index c29555421c..1083f1bbe8 100644
--- a/compiler/compiler/Cargo.toml
+++ b/compiler/compiler/Cargo.toml
@@ -15,8 +15,8 @@ keywords = [
 categories = [ "compilers", "cryptography", "web-programming" ]
 include = [ "Cargo.toml", "src", "README.md", "LICENSE.md" ]
 license = "GPL-3.0"
-edition = "2021"
-rust-version = "1.82.0"
+edition = "2024"
+rust-version = "1.85.1"
 
 [dependencies.leo-ast]
 workspace = true
diff --git a/compiler/compiler/tests/integration/compile.rs b/compiler/compiler/tests/integration/compile.rs
index 3a453a74b8..e155ab0073 100644
--- a/compiler/compiler/tests/integration/compile.rs
+++ b/compiler/compiler/tests/integration/compile.rs
@@ -68,6 +68,8 @@ fn run_test(test: Test, handler: &Handler, buf: &BufferEmitter) -> Result<Value,
 
     let mut all_outputs = Vec::with_capacity(build_options.len());
 
+    let re = Regex::new(r"program\s+([^\s.]+)\.aleo").unwrap();
+
     for build in build_options {
         let compiler_options = CompilerOptions {
             build,
@@ -98,7 +100,6 @@ fn run_test(test: Test, handler: &Handler, buf: &BufferEmitter) -> Result<Value,
         // Compile each program string separately.
         for program_string in program_strings {
             // Parse the program name from the program string.
-            let re = Regex::new(r"program\s+([^\s.]+)\.aleo").unwrap();
             let program_name = re.captures(program_string).unwrap().get(1).unwrap().as_str();
 
             // Parse the program.
diff --git a/compiler/compiler/tests/integration/execute.rs b/compiler/compiler/tests/integration/execute.rs
index 170b2e3080..fe67f79906 100644
--- a/compiler/compiler/tests/integration/execute.rs
+++ b/compiler/compiler/tests/integration/execute.rs
@@ -85,6 +85,8 @@ fn run_test(test: Test, handler: &Handler, buf: &BufferEmitter) -> Result<Value,
 
     let mut outputs = Vec::with_capacity(build_options.len());
 
+    let re = Regex::new(r"program\s+([^\s.]+)\.aleo").unwrap();
+
     for build in build_options {
         let compiler_options = CompilerOptions {
             build,
@@ -127,7 +129,6 @@ fn run_test(test: Test, handler: &Handler, buf: &BufferEmitter) -> Result<Value,
         // Compile each program string separately.
         for program_string in program_strings {
             // Parse the program name from the program string.
-            let re = Regex::new(r"program\s+([^\s.]+)\.aleo").unwrap();
             let program_name = re.captures(program_string).unwrap().get(1).unwrap().as_str();
 
             // Parse the program.
diff --git a/compiler/parser/Cargo.toml b/compiler/parser/Cargo.toml
index 3f5b2e1fd1..a477f7f9c2 100644
--- a/compiler/parser/Cargo.toml
+++ b/compiler/parser/Cargo.toml
@@ -15,8 +15,8 @@ keywords = [
 categories = [ "compilers", "cryptography", "web-programming" ]
 include = [ "Cargo.toml", "src", "README.md", "LICENSE.md" ]
 license = "GPL-3.0"
-edition = "2021"
-rust-version = "1.82.0"
+edition = "2024"
+rust-version = "1.85.1"
 
 [dependencies.leo-ast]
 workspace = true
diff --git a/compiler/passes/Cargo.toml b/compiler/passes/Cargo.toml
index bb36b99aec..314652b33e 100644
--- a/compiler/passes/Cargo.toml
+++ b/compiler/passes/Cargo.toml
@@ -15,8 +15,8 @@ keywords = [
 categories = [ "compilers", "cryptography", "web-programming" ]
 include = [ "Cargo.toml", "src", "README.md", "LICENSE.md" ]
 license = "GPL-3.0"
-edition = "2021"
-rust-version = "1.82.0"
+edition = "2024"
+rust-version = "1.85.1"
 
 [lib]
 path = "src/lib.rs"
diff --git a/compiler/passes/src/function_inlining/inline_expression.rs b/compiler/passes/src/function_inlining/inline_expression.rs
index d1c1ec7e2e..01cdf7ec27 100644
--- a/compiler/passes/src/function_inlining/inline_expression.rs
+++ b/compiler/passes/src/function_inlining/inline_expression.rs
@@ -64,10 +64,7 @@ impl ExpressionReconstructor for FunctionInliner<'_> {
 
                 // Replace each input variable with the appropriate parameter.
                 let replace = |identifier: &Identifier| {
-                    parameter_to_argument
-                        .get(&identifier.name)
-                        .cloned()
-                        .unwrap_or_else(|| Expression::Identifier(*identifier))
+                    parameter_to_argument.get(&identifier.name).cloned().unwrap_or(Expression::Identifier(*identifier))
                 };
 
                 let mut inlined_statements =
diff --git a/compiler/passes/src/static_analysis/future_checker.rs b/compiler/passes/src/static_analysis/future_checker.rs
index be0ccba6f7..c7259bf301 100644
--- a/compiler/passes/src/static_analysis/future_checker.rs
+++ b/compiler/passes/src/static_analysis/future_checker.rs
@@ -44,7 +44,7 @@ struct FutureChecker<'a> {
     handler: &'a Handler,
 }
 
-impl<'a> FutureChecker<'a> {
+impl FutureChecker<'_> {
     fn emit_err(&self, err: StaticAnalyzerError) {
         self.handler.emit_err(err);
     }
diff --git a/compiler/passes/src/type_checking/checker.rs b/compiler/passes/src/type_checking/checker.rs
index 2302a07114..2715e64f07 100644
--- a/compiler/passes/src/type_checking/checker.rs
+++ b/compiler/passes/src/type_checking/checker.rs
@@ -288,7 +288,7 @@ impl<'a> TypeChecker<'a> {
 
         // Make sure the input is no bigger than 64 bits.
         // Due to overhead in the bitwise representations of types in SnarkVM, 64 bit integers
-        // input more than 64 bits to a hash function.
+        // input more than 64 bits to a hash function, as do all structs and arrays.
         let assert_pedersen_64_bit_input = |type_: &Type, span: Span| {
             if !matches!(
                 type_,
@@ -310,8 +310,12 @@ impl<'a> TypeChecker<'a> {
         };
 
         // Make sure the input is no bigger than 128 bits.
+        //
+        // Due to overhead in the bitwise representations of types in SnarkVM, 128 bit integers
+        // input more than 128 bits to a hash function, as do most structs and arrays. We could
+        // actually allow arrays with a single element of type smaller than 64 bits, but it
+        // seems most understandable to the user to simply disallow composite types entirely.
         let assert_pedersen_128_bit_input = |type_: &Type, span: Span| {
-            // This is presumably a little more conservative than necessary.
             if !matches!(
                 type_,
                 Type::Integer(IntegerType::U8)
@@ -861,7 +865,7 @@ impl<'a> TypeChecker<'a> {
             Type::Composite(struct_)
                 if self
                     .lookup_struct(struct_.program.or(self.scope_state.program_name), struct_.id.name)
-                    .map_or(false, |struct_| struct_.is_record) =>
+                    .is_some_and(|struct_| struct_.is_record) =>
             {
                 self.emit_err(TypeCheckerError::struct_or_record_cannot_contain_record(parent, struct_.id.name, span))
             }
diff --git a/compiler/span/Cargo.toml b/compiler/span/Cargo.toml
index dc0b1d0fd7..0ff4acea22 100644
--- a/compiler/span/Cargo.toml
+++ b/compiler/span/Cargo.toml
@@ -15,8 +15,8 @@ keywords = [
 categories = [ "compilers", "cryptography", "web-programming" ]
 include = [ "Cargo.toml", "src", "README.md", "LICENSE.md" ]
 license = "GPL-3.0"
-edition = "2021"
-rust-version = "1.82.0"
+edition = "2024"
+rust-version = "1.85.1"
 
 [dependencies.indexmap]
 workspace = true
diff --git a/docs/grammar/Cargo.toml b/docs/grammar/Cargo.toml
index 81389b3863..a656f17ab9 100644
--- a/docs/grammar/Cargo.toml
+++ b/docs/grammar/Cargo.toml
@@ -15,8 +15,8 @@ keywords = [
 categories = [ "compilers", "cryptography", "web-programming" ]
 include = [ "Cargo.toml", "src", "README.md", "LICENSE.md" ]
 license = "GPL-3.0"
-edition = "2021"
-rust-version = "1.82.0"
+edition = "2024"
+rust-version = "1.85.1"
 
 [dependencies.anyhow]
 version = "1.0"
diff --git a/errors/Cargo.toml b/errors/Cargo.toml
index a75f1668a3..25a19a5d28 100644
--- a/errors/Cargo.toml
+++ b/errors/Cargo.toml
@@ -15,8 +15,8 @@ keywords = [
 categories = [ "compilers", "cryptography", "web-programming" ]
 include = [ "Cargo.toml", "src", "README.md", "LICENSE.md" ]
 license = "GPL-3.0"
-edition = "2021"
-rust-version = "1.82.0"
+edition = "2024"
+rust-version = "1.85.1"
 
 [dependencies.leo-span]
 workspace = true
diff --git a/interpreter/Cargo.toml b/interpreter/Cargo.toml
index e73ca49f9d..a1572d3fa0 100644
--- a/interpreter/Cargo.toml
+++ b/interpreter/Cargo.toml
@@ -15,8 +15,8 @@ keywords = [
 categories = [ "compilers", "cryptography", "web-programming" ]
 include = [ "Cargo.toml", "src", "README.md", "LICENSE.md" ]
 license = "GPL-3.0"
-edition = "2021"
-rust-version = "1.82.0"
+edition = "2024"
+rust-version = "1.85.1"
 
 [dependencies.snarkvm]
 workspace = true
diff --git a/interpreter/src/core_functions.rs b/interpreter/src/core_functions.rs
index 266e75bd5a..b9071c0f0f 100644
--- a/interpreter/src/core_functions.rs
+++ b/interpreter/src/core_functions.rs
@@ -124,7 +124,7 @@ pub fn evaluate_core_function(
     macro_rules! maybe_gen {
         () => {
             if let Some(rng) = helper.rng() {
-                rng.gen()
+                rng.r#gen()
             } else {
                 return Ok(None);
             }
diff --git a/interpreter/src/cursor_aleo.rs b/interpreter/src/cursor_aleo.rs
index fd6fcd028b..c98b487b4c 100644
--- a/interpreter/src/cursor_aleo.rs
+++ b/interpreter/src/cursor_aleo.rs
@@ -867,28 +867,28 @@ impl Cursor<'_> {
                 for value in rand.operands().iter().map(|op| self.operand_value(op)) {
                     value.write_bits_le(&mut bits);
                 }
-                let field: Field<TestnetV0> = self.rng.gen();
+                let field: Field<TestnetV0> = self.rng.r#gen();
                 field.write_bits_le(&mut bits);
                 let seed_vec = TestnetV0::hash_bhp1024(&bits)?.to_bytes_le()?;
                 let mut seed = [0u8; 32];
                 seed.copy_from_slice(&seed_vec[..32]);
                 let mut rng = ChaCha20Rng::from_seed(seed);
                 let value = match rand.destination_type() {
-                    LiteralType::Address => Value::Address(rng.gen()),
-                    LiteralType::Boolean => Value::Bool(rng.gen()),
-                    LiteralType::Field => Value::Field(rng.gen()),
-                    LiteralType::Group => Value::Group(rng.gen()),
-                    LiteralType::I8 => Value::I8(rng.gen()),
-                    LiteralType::I16 => Value::I16(rng.gen()),
-                    LiteralType::I32 => Value::I32(rng.gen()),
-                    LiteralType::I64 => Value::I64(rng.gen()),
-                    LiteralType::I128 => Value::I128(rng.gen()),
-                    LiteralType::U8 => Value::U8(rng.gen()),
-                    LiteralType::U16 => Value::U16(rng.gen()),
-                    LiteralType::U32 => Value::U32(rng.gen()),
-                    LiteralType::U64 => Value::U64(rng.gen()),
-                    LiteralType::U128 => Value::U128(rng.gen()),
-                    LiteralType::Scalar => Value::Scalar(rng.gen()),
+                    LiteralType::Address => Value::Address(rng.r#gen()),
+                    LiteralType::Boolean => Value::Bool(rng.r#gen()),
+                    LiteralType::Field => Value::Field(rng.r#gen()),
+                    LiteralType::Group => Value::Group(rng.r#gen()),
+                    LiteralType::I8 => Value::I8(rng.r#gen()),
+                    LiteralType::I16 => Value::I16(rng.r#gen()),
+                    LiteralType::I32 => Value::I32(rng.r#gen()),
+                    LiteralType::I64 => Value::I64(rng.r#gen()),
+                    LiteralType::I128 => Value::I128(rng.r#gen()),
+                    LiteralType::U8 => Value::U8(rng.r#gen()),
+                    LiteralType::U16 => Value::U16(rng.r#gen()),
+                    LiteralType::U32 => Value::U32(rng.r#gen()),
+                    LiteralType::U64 => Value::U64(rng.r#gen()),
+                    LiteralType::U128 => Value::U128(rng.r#gen()),
+                    LiteralType::Scalar => Value::Scalar(rng.r#gen()),
                     LiteralType::Signature => halt_no_span!("Cannot create a random signature"),
                     LiteralType::String => halt_no_span!("Cannot create a random string"),
                 };
diff --git a/interpreter/src/ratatui_ui.rs b/interpreter/src/ratatui_ui.rs
index eef057823e..8a804a144f 100644
--- a/interpreter/src/ratatui_ui.rs
+++ b/interpreter/src/ratatui_ui.rs
@@ -154,7 +154,7 @@ struct Prompt {
     cursor: usize,
 }
 
-impl<'a> Widget for &'a Prompt {
+impl Widget for &Prompt {
     fn render(self, area: Rect, buf: &mut Buffer) {
         let mut plain = || {
             Text::raw(&self.current).render(area, buf);
diff --git a/leo/cli/cli.rs b/leo/cli/cli.rs
index ada4024fa1..d176e2b8a5 100644
--- a/leo/cli/cli.rs
+++ b/leo/cli/cli.rs
@@ -65,7 +65,7 @@ enum Commands {
     #[clap(about = "Deploy a program")]
     Deploy {
         #[clap(flatten)]
-        command: Deploy,
+        command: LeoDeploy,
     },
     #[clap(about = "Query live data from the Aleo network")]
     Query {
diff --git a/leo/cli/commands/add.rs b/leo/cli/commands/add.rs
index 416d5c0a1c..1123de2874 100755
--- a/leo/cli/commands/add.rs
+++ b/leo/cli/commands/add.rs
@@ -68,7 +68,7 @@ impl Command for LeoAdd {
 
         // Add dependency section to manifest if it doesn't exist.
         let mut dependencies = match (self.clear, manifest.dependencies()) {
-            (false, Some(ref dependencies)) => dependencies
+            (false, Some(dependencies)) => dependencies
                 .iter()
                 .filter_map(|dependency| {
                     // Overwrite old dependencies of the same name.
diff --git a/leo/cli/commands/deploy.rs b/leo/cli/commands/deploy.rs
index 71990969f8..917ff84794 100644
--- a/leo/cli/commands/deploy.rs
+++ b/leo/cli/commands/deploy.rs
@@ -41,7 +41,7 @@ use text_tables;
 
 /// Deploys an Aleo program.
 #[derive(Parser, Debug)]
-pub struct Deploy {
+pub struct LeoDeploy {
     #[clap(flatten)]
     pub(crate) fee_options: FeeOptions,
     #[clap(long, help = "Disables building of the project before deployment.", default_value = "false")]
@@ -58,7 +58,7 @@ pub struct Deploy {
     pub(crate) options: BuildOptions,
 }
 
-impl Command for Deploy {
+impl Command for LeoDeploy {
     type Input = ();
     type Output = ();
 
@@ -97,11 +97,11 @@ impl Command for Deploy {
 
 // A helper function to handle deployment logic.
 fn handle_deploy<A: Aleo<Network = N, BaseField = N::Field>, N: Network>(
-    command: &Deploy,
+    command: &LeoDeploy,
     context: Context,
     network: NetworkName,
     endpoint: &str,
-) -> Result<<Deploy as Command>::Output> {
+) -> Result<<LeoDeploy as Command>::Output> {
     // Get the program name.
     let project_name = context.open_manifest::<N>()?.program_id().to_string();
 
@@ -257,7 +257,7 @@ fn handle_deploy<A: Aleo<Network = N, BaseField = N::Field>, N: Network>(
 
 // A helper function to display a cost breakdown of the deployment.
 fn deploy_cost_breakdown(
-    name: &String,
+    name: &str,
     base_fee: f64,
     storage_cost: f64,
     synthesis_cost: f64,
diff --git a/leo/cli/commands/execute.rs b/leo/cli/commands/execute.rs
index d7072862f1..989dd538c9 100644
--- a/leo/cli/commands/execute.rs
+++ b/leo/cli/commands/execute.rs
@@ -460,7 +460,7 @@ fn load_program_from_network<N: Network>(
 
 // A helper function to display a cost breakdown of the execution.
 fn execution_cost_breakdown(
-    name: &String,
+    name: &str,
     base_fee: f64,
     storage_cost: f64,
     finalize_cost: f64,
diff --git a/leo/cli/commands/mod.rs b/leo/cli/commands/mod.rs
index d6c51ec3cd..0c2c14bf5b 100644
--- a/leo/cli/commands/mod.rs
+++ b/leo/cli/commands/mod.rs
@@ -30,7 +30,7 @@ pub mod debug;
 pub use debug::LeoDebug;
 
 pub mod deploy;
-pub use deploy::Deploy;
+pub use deploy::LeoDeploy;
 
 pub mod execute;
 pub use execute::LeoExecute;
@@ -293,8 +293,8 @@ fn get_latest_block_height(endpoint: &str, network: &str, context: &Context) ->
 }
 
 /// Determine if the transaction should be broadcast or displayed to user.
-fn handle_broadcast<N: Network>(endpoint: &String, transaction: Transaction<N>, operation: &String) -> Result<()> {
-    println!("Broadcasting transaction to {}...\n", endpoint.clone());
+fn handle_broadcast<N: Network>(endpoint: &str, transaction: Transaction<N>, operation: &str) -> Result<()> {
+    println!("Broadcasting transaction to {}...\n", endpoint);
     // Get the transaction id.
     let transaction_id = transaction.id();
 
diff --git a/leo/cli/commands/remove.rs b/leo/cli/commands/remove.rs
index 95333f97be..011e48fc41 100644
--- a/leo/cli/commands/remove.rs
+++ b/leo/cli/commands/remove.rs
@@ -60,7 +60,7 @@ impl Command for LeoRemove {
 
             let mut found_match = false;
             let dep = match manifest.dependencies() {
-                Some(ref dependencies) => dependencies
+                Some(dependencies) => dependencies
                     .iter()
                     .filter_map(|dependency| {
                         if dependency.name() == &name {
diff --git a/leo/cli/commands/run.rs b/leo/cli/commands/run.rs
index 552fc3531d..339d89cc5b 100644
--- a/leo/cli/commands/run.rs
+++ b/leo/cli/commands/run.rs
@@ -66,7 +66,9 @@ fn handle_run<N: Network>(command: LeoRun, context: Context) -> Result<<LeoRun a
     let mut inputs = command.inputs;
 
     // Compose the `run` command.
-    let mut arguments = vec![SNARKVM_COMMAND.to_string(), command.name];
+    // The argument "--" is the separator, used to make sure snarkvm doesn't try to parse negative
+    // values as CLI flags.
+    let mut arguments = vec![SNARKVM_COMMAND.to_string(), command.name, "--".to_string()];
 
     // Add the inputs to the arguments.
     match command.file {
diff --git a/leo/cli/helpers/context.rs b/leo/cli/helpers/context.rs
index c37a9bfa60..ada097b4cf 100644
--- a/leo/cli/helpers/context.rs
+++ b/leo/cli/helpers/context.rs
@@ -53,7 +53,7 @@ impl Context {
     /// Returns the path of the parent directory to the Leo package.
     pub fn parent_dir(&self) -> Result<PathBuf> {
         match &self.path {
-            Some(ref path) => {
+            Some(path) => {
                 let mut path = path.clone();
                 path.pop();
                 Ok(path)
diff --git a/leo/package/Cargo.toml b/leo/package/Cargo.toml
index 782d4b18a1..6426abd592 100644
--- a/leo/package/Cargo.toml
+++ b/leo/package/Cargo.toml
@@ -15,8 +15,8 @@ keywords = [
 categories = [ "compilers", "cryptography", "web-programming" ]
 include = [ "Cargo.toml", "src", "README.md", "LICENSE.md" ]
 license = "GPL-3.0"
-edition = "2021"
-rust-version = "1.82.0"
+edition = "2024"
+rust-version = "1.85.1"
 
 [dependencies.leo-errors]
 workspace = true
diff --git a/rust-toolchain b/rust-toolchain
index 71fae54fb2..8510ffad03 100644
--- a/rust-toolchain
+++ b/rust-toolchain
@@ -1 +1 @@
-1.82.0
+1.85.1
diff --git a/.rustfmt.toml b/rustfmt.toml
similarity index 95%
rename from .rustfmt.toml
rename to rustfmt.toml
index b5dd5e6730..9d717fc697 100644
--- a/.rustfmt.toml
+++ b/rustfmt.toml
@@ -1,16 +1,16 @@
 # https://github.com/rust-lang/rustfmt/blob/master/Configurations.md
 
 # Stable configurations
-edition = "2021"
+edition = "2024"
 max_width = 120
 merge_derives = true
 use_field_init_shorthand = true
 use_small_heuristics = "Max"
 use_try_shorthand = true
-
-# Nightly configurations
 imports_layout = "HorizontalVertical"
 imports_granularity = "Crate"
 overflow_delimited_expr = true
+
+# Nightly configurations
 reorder_impl_items = true
 style_edition = "2024"
diff --git a/tests/test-framework/Cargo.toml b/tests/test-framework/Cargo.toml
index 0c4f90eac9..863c4e42fc 100644
--- a/tests/test-framework/Cargo.toml
+++ b/tests/test-framework/Cargo.toml
@@ -16,7 +16,7 @@ categories = [ "compilers", "cryptography", "web-programming" ]
 include = [ "Cargo.toml", "src", "benches", "README.md", "LICENSE.md" ]
 license = "GPL-3.0"
 edition = "2018"
-rust-version = "1.82.0"
+rust-version = "1.85.1"
 
 [[bench]]
 name = "leo_compiler"
diff --git a/utils/disassembler/Cargo.toml b/utils/disassembler/Cargo.toml
index 712806d782..6f7509c4dd 100644
--- a/utils/disassembler/Cargo.toml
+++ b/utils/disassembler/Cargo.toml
@@ -15,8 +15,8 @@ keywords = [
 categories = [ "compilers", "cryptography", "web-programming" ]
 include = [ "Cargo.toml", "src", "README.md", "LICENSE.md" ]
 license = "GPL-3.0"
-edition = "2021"
-rust-version = "1.82.0"
+edition = "2024"
+rust-version = "1.85.1"
 
 [lib]
 path = "src/lib.rs"
diff --git a/utils/retriever/Cargo.toml b/utils/retriever/Cargo.toml
index e5d23c8824..3bc3a81d54 100644
--- a/utils/retriever/Cargo.toml
+++ b/utils/retriever/Cargo.toml
@@ -15,8 +15,8 @@ keywords = [
 categories = [ "compilers", "cryptography", "web-programming" ]
 include = [ "Cargo.toml", "src", "README.md", "LICENSE.md" ]
 license = "GPL-3.0"
-edition = "2021"
-rust-version = "1.82.0"
+edition = "2024"
+rust-version = "1.85.1"
 
 [dependencies.snarkvm]
 workspace = true
