diff --git a/compiler/passes/src/flattening/flatten_expression.rs b/compiler/passes/src/flattening/flatten_expression.rs
index cd9c192105..12f0790443 100644
--- a/compiler/passes/src/flattening/flatten_expression.rs
+++ b/compiler/passes/src/flattening/flatten_expression.rs
@@ -22,6 +22,8 @@ use leo_ast::{
     ExpressionReconstructor, MemberAccess, Statement, TernaryExpression, TupleExpression,
 };
 
+// TODO: Clean up logic. To be done in a follow-up PR (feat/tuples)
+
 impl ExpressionReconstructor for Flattener<'_> {
     type AdditionalOutput = Vec<Statement>;
 
@@ -72,18 +74,118 @@ impl ExpressionReconstructor for Flattener<'_> {
                             statements.extend(stmts);
 
                             // Create and accumulate an intermediate assignment statement for the ternary expression corresponding to the tuple element.
-                            let (identifier, statement) = self.assigner.unique_simple_assign_statement(ternary);
+                            let (identifier, statement) = self.unique_simple_assign_statement(ternary);
                             statements.push(statement);
 
                             // Return the identifier associated with the folded tuple element.
-                            identifier
+                            Expression::Identifier(identifier)
                         })
                         .collect(),
                     span: Default::default(),
                 });
                 (tuple, statements)
             }
-            // If both expressions are circuits, construct ternary expression for each of the members and a circuit expression for the result.
+            // If both expressions are access expressions which themselves are circuits, construct ternary expression for nested circuit member.
+            (
+                Expression::Access(AccessExpression::Member(first)),
+                Expression::Access(AccessExpression::Member(second)),
+            ) => {
+                // Lookup the circuit symbols associated with the expressions.
+                // TODO: Remove clones
+                let first_circuit_symbol =
+                    self.lookup_circuit_symbol(&Expression::Access(AccessExpression::Member(first.clone())));
+                let second_circuit_symbol =
+                    self.lookup_circuit_symbol(&Expression::Access(AccessExpression::Member(second.clone())));
+
+                match (first_circuit_symbol, second_circuit_symbol) {
+                    (Some(first_circuit_symbol), Some(second_circuit_symbol)) => {
+                        let first_member_circuit = self.symbol_table.lookup_circuit(first_circuit_symbol).unwrap();
+                        let second_member_circuit = self.symbol_table.lookup_circuit(second_circuit_symbol).unwrap();
+                        // Note that type checking guarantees that both expressions have the same same type. This is a sanity check.
+                        assert_eq!(first_member_circuit, second_member_circuit);
+
+                        // For each circuit member, construct a new ternary expression.
+                        let members = first_member_circuit
+                            .members
+                            .iter()
+                            .map(|CircuitMember::CircuitVariable(id, _)| {
+                                // Construct a new ternary expression for the circuit member.
+                                let (expression, stmts) = self.reconstruct_ternary(TernaryExpression {
+                                    condition: input.condition.clone(),
+                                    if_true: Box::new(Expression::Access(AccessExpression::Member(MemberAccess {
+                                        inner: Box::new(Expression::Access(AccessExpression::Member(first.clone()))),
+                                        name: *id,
+                                        span: Default::default(),
+                                    }))),
+                                    if_false: Box::new(Expression::Access(AccessExpression::Member(MemberAccess {
+                                        inner: Box::new(Expression::Access(AccessExpression::Member(second.clone()))),
+                                        name: *id,
+                                        span: Default::default(),
+                                    }))),
+                                    span: Default::default(),
+                                });
+
+                                // Accumulate any statements generated.
+                                statements.extend(stmts);
+
+                                // Create and accumulate an intermediate assignment statement for the ternary expression corresponding to the circuit member.
+                                let (identifier, statement) = self.unique_simple_assign_statement(expression);
+                                statements.push(statement);
+
+                                CircuitVariableInitializer {
+                                    identifier: *id,
+                                    expression: Some(Expression::Identifier(identifier)),
+                                }
+                            })
+                            .collect();
+
+                        let (expr, stmts) = self.reconstruct_circuit_init(CircuitExpression {
+                            name: first_member_circuit.identifier,
+                            members,
+                            span: Default::default(),
+                        });
+
+                        // Accumulate any statements generated.
+                        statements.extend(stmts);
+
+                        // Create a new assignment statement for the circuit expression.
+                        let (identifier, statement) = self.unique_simple_assign_statement(expr);
+
+                        // Mark the lhs of the assignment as a circuit.
+                        self.circuits
+                            .insert(identifier.name, first_member_circuit.identifier.name);
+
+                        statements.push(statement);
+
+                        (Expression::Identifier(identifier), statements)
+                    }
+                    _ => {
+                        let if_true = Expression::Access(AccessExpression::Member(first));
+                        let if_false = Expression::Access(AccessExpression::Member(second));
+                        // Reconstruct the true case.
+                        let (if_true, stmts) = self.reconstruct_expression(if_true);
+                        statements.extend(stmts);
+
+                        // Reconstruct the false case.
+                        let (if_false, stmts) = self.reconstruct_expression(if_false);
+                        statements.extend(stmts);
+
+                        let (identifier, statement) =
+                            self.unique_simple_assign_statement(Expression::Ternary(TernaryExpression {
+                                condition: input.condition,
+                                if_true: Box::new(if_true),
+                                if_false: Box::new(if_false),
+                                span: input.span,
+                            }));
+
+                        // Accumulate the new assignment statement.
+                        statements.push(statement);
+
+                        (Expression::Identifier(identifier), statements)
+                    }
+                }
+            }
+            // If both expressions are identifiers which are circuits, construct ternary expression for each of the members and a circuit expression for the result.
             (Expression::Identifier(first), Expression::Identifier(second))
                 if self.circuits.contains_key(&first.name) && self.circuits.contains_key(&second.name) =>
             {
@@ -123,12 +225,12 @@ impl ExpressionReconstructor for Flattener<'_> {
                         statements.extend(stmts);
 
                         // Create and accumulate an intermediate assignment statement for the ternary expression corresponding to the circuit member.
-                        let (identifier, statement) = self.assigner.unique_simple_assign_statement(expression);
+                        let (identifier, statement) = self.unique_simple_assign_statement(expression);
                         statements.push(statement);
 
                         CircuitVariableInitializer {
                             identifier: *id,
-                            expression: Some(identifier),
+                            expression: Some(Expression::Identifier(identifier)),
                         }
                     })
                     .collect();
@@ -143,21 +245,14 @@ impl ExpressionReconstructor for Flattener<'_> {
                 statements.extend(stmts);
 
                 // Create a new assignment statement for the circuit expression.
-                let (identifier, statement) = self.assigner.unique_simple_assign_statement(expr);
+                let (identifier, statement) = self.unique_simple_assign_statement(expr);
 
                 // Mark the lhs of the assignment as a circuit.
-                match identifier {
-                    Expression::Identifier(identifier) => {
-                        self.circuits.insert(identifier.name, first_circuit.identifier.name)
-                    }
-                    _ => unreachable!(
-                        "`unique_simple_assign_statement` always produces an identifier on the left hand size."
-                    ),
-                };
+                self.circuits.insert(identifier.name, first_circuit.identifier.name);
 
                 statements.push(statement);
 
-                (identifier, statements)
+                (Expression::Identifier(identifier), statements)
             }
             // Otherwise, create a new intermediate assignment for the ternary expression are return the assigned variable.
             // Note that a new assignment must be created to flattened nested ternary expressions.
@@ -171,18 +266,17 @@ impl ExpressionReconstructor for Flattener<'_> {
                 statements.extend(stmts);
 
                 let (identifier, statement) =
-                    self.assigner
-                        .unique_simple_assign_statement(Expression::Ternary(TernaryExpression {
-                            condition: input.condition,
-                            if_true: Box::new(if_true),
-                            if_false: Box::new(if_false),
-                            span: input.span,
-                        }));
+                    self.unique_simple_assign_statement(Expression::Ternary(TernaryExpression {
+                        condition: input.condition,
+                        if_true: Box::new(if_true),
+                        if_false: Box::new(if_false),
+                        span: input.span,
+                    }));
 
                 // Accumulate the new assignment statement.
                 statements.push(statement);
 
-                (identifier, statements)
+                (Expression::Identifier(identifier), statements)
             }
         }
     }
diff --git a/compiler/passes/src/flattening/flatten_program.rs b/compiler/passes/src/flattening/flatten_program.rs
index 9764571db6..44af8ee31f 100644
--- a/compiler/passes/src/flattening/flatten_program.rs
+++ b/compiler/passes/src/flattening/flatten_program.rs
@@ -18,6 +18,7 @@ use crate::Flattener;
 
 use leo_ast::{
     Finalize, FinalizeStatement, Function, ProgramReconstructor, ReturnStatement, Statement, StatementReconstructor,
+    Type,
 };
 
 impl ProgramReconstructor for Flattener<'_> {
@@ -26,6 +27,14 @@ impl ProgramReconstructor for Flattener<'_> {
         // First, flatten the finalize block. This allows us to initialize self.finalizes correctly.
         // Note that this is safe since the finalize block is independent of the function body.
         let finalize = function.finalize.map(|finalize| {
+            // Initialize `self.circuits` with the finalize's input as necessary.
+            self.circuits = Default::default();
+            for input in &finalize.input {
+                if let Type::Identifier(circuit_name) = input.type_ {
+                    self.circuits.insert(input.identifier.name, circuit_name.name);
+                }
+            }
+
             // Flatten the finalize block.
             let mut block = self.reconstruct_block(finalize.block).0;
 
@@ -59,6 +68,14 @@ impl ProgramReconstructor for Flattener<'_> {
             }
         });
 
+        // Initialize `self.circuits` with the function's input as necessary.
+        self.circuits = Default::default();
+        for input in &function.input {
+            if let Type::Identifier(circuit_name) = input.type_ {
+                self.circuits.insert(input.identifier.name, circuit_name.name);
+            }
+        }
+
         // Flatten the function body.
         let mut block = self.reconstruct_block(function.block).0;
 
diff --git a/compiler/passes/src/flattening/flatten_statement.rs b/compiler/passes/src/flattening/flatten_statement.rs
index f61f5dea3c..7eb97c7891 100644
--- a/compiler/passes/src/flattening/flatten_statement.rs
+++ b/compiler/passes/src/flattening/flatten_statement.rs
@@ -34,23 +34,15 @@ impl StatementReconstructor for Flattener<'_> {
         };
 
         let (value, statements) = match assign.value {
-            // If the rhs of the assignment is a circuit, add it to `self.circuits`.
-            Expression::Circuit(rhs) => {
-                self.circuits.insert(lhs.name, rhs.name.name);
-                (Expression::Circuit(rhs), Default::default())
-            }
-            // If the rhs of the assignment is an identifier that is a circuit, add it to `self.circuits`.
-            Expression::Identifier(rhs) if self.circuits.contains_key(&rhs.name) => {
-                // Note that this unwrap is safe because we just checked that the key exists.
-                self.circuits.insert(lhs.name, *self.circuits.get(&rhs.name).unwrap());
-                (Expression::Identifier(rhs), Default::default())
-            }
             // If the rhs of the assignment is ternary expression, reconstruct it.
             Expression::Ternary(ternary) => self.reconstruct_ternary(ternary),
             // Otherwise return the original statement.
             value => (value, Default::default()),
         };
 
+        // Update the `self.circuits` if the rhs is a circuit.
+        self.update_circuits(&lhs, &value);
+
         (
             Statement::Assign(Box::new(AssignStatement {
                 place: Expression::Identifier(lhs),
diff --git a/compiler/passes/src/flattening/flattener.rs b/compiler/passes/src/flattening/flattener.rs
index 6e16ccce69..12d5fd9b83 100644
--- a/compiler/passes/src/flattening/flattener.rs
+++ b/compiler/passes/src/flattening/flattener.rs
@@ -16,7 +16,10 @@
 
 use crate::{Assigner, SymbolTable};
 
-use leo_ast::{Expression, ExpressionReconstructor, Identifier, Statement, TernaryExpression};
+use leo_ast::{
+    AccessExpression, CircuitMember, Expression, ExpressionReconstructor, Identifier, Statement, TernaryExpression,
+    Type,
+};
 use leo_span::Symbol;
 
 use indexmap::IndexMap;
@@ -92,19 +95,84 @@ impl<'a> Flattener<'a> {
             });
             statements.extend(stmts);
 
-            statements.push(self.assigner.simple_assign_statement(place, value));
-            Expression::Identifier(place)
+            match &value {
+                // If the expression is a tuple, then use it directly.
+                // This must be done to ensure that intermediate tuple assignments are not created.
+                Expression::Tuple(_) => value,
+                // Otherwise, assign the expression to a variable and return the variable.
+                _ => {
+                    statements.push(self.simple_assign_statement(place, value));
+                    Expression::Identifier(place)
+                }
+            }
         };
 
         let expression = guards
             .into_iter()
             .rev()
             .fold(last_expression, |acc, (guard, expr)| match guard {
-                None => unreachable!("All expression except for the last one must have a guard."),
+                None => unreachable!("All expressions except for the last one must have a guard."),
                 // Note that type checking guarantees that all expressions have the same type.
                 Some(guard) => construct_ternary_assignment(guard, expr, acc),
             });
 
         (expression, statements)
     }
+
+    /// Looks up the name of the circuit associated with an identifier or access expression, if it exists.
+    pub(crate) fn lookup_circuit_symbol(&self, expression: &Expression) -> Option<Symbol> {
+        match expression {
+            Expression::Identifier(identifier) => self.circuits.get(&identifier.name).copied(),
+            Expression::Access(AccessExpression::Member(access)) => {
+                // The inner expression of an access expression is either an identifier or another access expression.
+                let name = self.lookup_circuit_symbol(&access.inner).unwrap();
+                let circuit = self.symbol_table.lookup_circuit(name).unwrap();
+                let CircuitMember::CircuitVariable(_, member_type) = circuit
+                    .members
+                    .iter()
+                    .find(|member| member.name() == access.name.name)
+                    .unwrap();
+                match member_type {
+                    Type::Identifier(identifier) => Some(identifier.name),
+                    _ => None,
+                }
+            }
+            _ => None,
+        }
+    }
+
+    /// Updates `self.circuits` for new assignment statements.
+    /// Expects the left hand side of the assignment to be an identifier.
+    pub(crate) fn update_circuits(&mut self, lhs: &Identifier, rhs: &Expression) {
+        match rhs {
+            Expression::Circuit(rhs) => {
+                self.circuits.insert(lhs.name, rhs.name.name);
+            }
+            // If the rhs of the assignment is an identifier that is a circuit, add it to `self.circuits`.
+            Expression::Identifier(rhs) if self.circuits.contains_key(&rhs.name) => {
+                // Note that this unwrap is safe because we just checked that the key exists.
+                self.circuits.insert(lhs.name, *self.circuits.get(&rhs.name).unwrap());
+            }
+            // Otherwise, do nothing.
+            _ => (),
+        }
+    }
+
+    /// A wrapper around `assigner.unique_simple_assign_statement` that updates `self.circuits`.
+    pub(crate) fn unique_simple_assign_statement(&mut self, expr: Expression) -> (Identifier, Statement) {
+        let (place, statement) = self.assigner.unique_simple_assign_statement(expr);
+        match &statement {
+            Statement::Assign(assign) => {
+                self.update_circuits(&place, &assign.value);
+            }
+            _ => unreachable!("`assigner.unique_simple_assign_statement` always returns an assignment statement."),
+        }
+        (place, statement)
+    }
+
+    /// A wrapper around `assigner.simple_assign_statement` that updates `self.circuits`.
+    pub(crate) fn simple_assign_statement(&mut self, lhs: Identifier, rhs: Expression) -> Statement {
+        self.update_circuits(&lhs, &rhs);
+        self.assigner.simple_assign_statement(lhs, rhs)
+    }
 }
diff --git a/compiler/passes/src/static_single_assignment/assigner.rs b/compiler/passes/src/static_single_assignment/assigner.rs
index c4370df904..7ef7520411 100644
--- a/compiler/passes/src/static_single_assignment/assigner.rs
+++ b/compiler/passes/src/static_single_assignment/assigner.rs
@@ -44,7 +44,7 @@ impl Assigner {
 
     /// Constructs a simple assign statement for `expr` with a unique name.
     /// For example, `expr` is transformed into `$var$0 = expr;`.
-    pub(crate) fn unique_simple_assign_statement(&mut self, expr: Expression) -> (Expression, Statement) {
+    pub(crate) fn unique_simple_assign_statement(&mut self, expr: Expression) -> (Identifier, Statement) {
         // Create a new variable for the expression.
         let name = self.unique_symbol("$var");
 
@@ -53,6 +53,6 @@ impl Assigner {
             span: Default::default(),
         };
 
-        (Expression::Identifier(place), self.simple_assign_statement(place, expr))
+        (place, self.simple_assign_statement(place, expr))
     }
 }
diff --git a/compiler/passes/src/static_single_assignment/rename_expression.rs b/compiler/passes/src/static_single_assignment/rename_expression.rs
index 53ca1f34a9..89f7aede36 100644
--- a/compiler/passes/src/static_single_assignment/rename_expression.rs
+++ b/compiler/passes/src/static_single_assignment/rename_expression.rs
@@ -84,7 +84,7 @@ impl ExpressionConsumer for StaticSingleAssigner {
         let (place, statement) = self.assigner.unique_simple_assign_statement(Expression::Access(expr));
         statements.push(statement);
 
-        (place, statements)
+        (Expression::Identifier(place), statements)
     }
 
     /// Consumes a binary expression, accumulating any statements that are generated.
@@ -107,7 +107,7 @@ impl ExpressionConsumer for StaticSingleAssigner {
             }));
         statements.push(statement);
 
-        (place, statements)
+        (Expression::Identifier(place), statements)
     }
 
     /// Consumes a call expression without visiting the function name, accumulating any statements that are generated.
@@ -137,7 +137,7 @@ impl ExpressionConsumer for StaticSingleAssigner {
             }));
         statements.push(statement);
 
-        (place, statements)
+        (Expression::Identifier(place), statements)
     }
 
     /// Consumes a circuit initialization expression with renamed variables, accumulating any statements that are generated.
@@ -178,7 +178,7 @@ impl ExpressionConsumer for StaticSingleAssigner {
             }));
         statements.push(statement);
 
-        (place, statements)
+        (Expression::Identifier(place), statements)
     }
 
     /// Produces a new `Identifier` with a unique name.
@@ -235,7 +235,7 @@ impl ExpressionConsumer for StaticSingleAssigner {
             }));
         statements.push(statement);
 
-        (place, statements)
+        (Expression::Identifier(place), statements)
     }
 
     /// Consumes a tuple expression, accumulating any statements that are generated
@@ -279,6 +279,6 @@ impl ExpressionConsumer for StaticSingleAssigner {
             }));
         statements.push(statement);
 
-        (place, statements)
+        (Expression::Identifier(place), statements)
     }
 }
diff --git a/examples/interest/src/main.leo b/examples/interest/src/main.leo
index b1d992e658..4905e3d2c6 100644
--- a/examples/interest/src/main.leo
+++ b/examples/interest/src/main.leo
@@ -27,7 +27,11 @@ function bounded_iteration_interest(capital: u32,
         if i < iterations {
             // Note that the added amount is rounded down.
             amount += (amount * rate) / 100u32;
+
         } // Skip the remaining iterations.
+        if i == 40u8 {
+            return amount;
+        }
     }
     return amount;
 }
diff --git a/examples/tictactoe/src/main.leo b/examples/tictactoe/src/main.leo
index 25743ae18d..5b68fee6db 100644
--- a/examples/tictactoe/src/main.leo
+++ b/examples/tictactoe/src/main.leo
@@ -4,17 +4,26 @@
 // The state of the board is captured by a value of this circuit type.
 // This will work better with arrays, but for now we just expand the 9 cells.
 
+circuit Row {
+    e1: u8,
+    e2: u8,
+    e3: u8
+}
+
 circuit Board {
-    // c(ell){row}{column} (0 for empty, 1 for X, 2 for O):
-    c11: u8,
-    c12: u8,
-    c13: u8,
-    c21: u8,
-    c22: u8,
-    c23: u8,
-    c31: u8,
-    c32: u8,
-    c33: u8
+    r1: Row,
+    r2: Row,
+    r3: Row,
+}
+
+// This function creates a new game board, with all cells empty.
+@program
+function new() -> Board {
+    return Board {
+        r1: Row { e1: 0u8, e2: 0u8, e3: 0u8 },
+        r2: Row { e1: 0u8, e2: 0u8, e3: 0u8 },
+        r3: Row { e1: 0u8, e2: 0u8, e3: 0u8 },
+    };
 }
 
 // This function checks whether a player is a winner on a board.
@@ -23,141 +32,123 @@ circuit Board {
 
 function win(b: Board, p: u8) -> bool {
     return
-        (b.c11 == p && b.c12 == p && b.c13 == p) || // row 1
-        (b.c21 == p && b.c22 == p && b.c23 == p) || // row 2
-        (b.c31 == p && b.c32 == p && b.c33 == p) || // row 3
-        (b.c11 == p && b.c21 == p && b.c31 == p) || // column 1
-        (b.c12 == p && b.c22 == p && b.c32 == p) || // column 2
-        (b.c13 == p && b.c23 == p && b.c33 == p) || // column 3
-        (b.c11 == p && b.c22 == p && b.c33 == p) || // diagonal
-        (b.c13 == p && b.c22 == p && b.c31 == p); // other diagonal
+        (b.r1.e1 == p && b.r1.e2 == p && b.r1.e3 == p) || // row 1
+        (b.r2.e1 == p && b.r2.e2 == p && b.r2.e3 == p) || // row 2
+        (b.r3.e1 == p && b.r3.e3 == p && b.r3.e3 == p) || // row 3
+        (b.r1.e1 == p && b.r2.e1 == p && b.r3.e1 == p) || // column 1
+        (b.r1.e2 == p && b.r2.e3 == p && b.r3.e2 == p) || // column 2
+        (b.r1.e3 == p && b.r2.e3 == p && b.r3.e3 == p) || // column 3
+        (b.r1.e1 == p && b.r2.e2 == p && b.r3.e3 == p) || // diagonal
+        (b.r1.e3 == p && b.r2.e2 == p && b.r3.e1 == p); // other diagonal
 }
 
 // This carries out a single move: a player (1 for X, 2 for O) marking a cell.
 // The cell must be empty (i.e. contain 0), otherwise this is a no-op.
 // The u8 result is the winning player, of 0 if there is no winner (yet).
-
+//
 @program
 function main(public player: u8, row: u8, col: u8, board: Board) -> (Board, u8) {
     console.assert(player == 1u8 || player == 2u8); // 1 for X, 2 for O
     console.assert(1u8 <= row && row <= 3u8);
     console.assert(1u8 <= col && col <= 3u8);
-    let new: Board = board;
+    let updated: Board = board;
     // This will work better when circuit assignment is supported.
-    if row == 1u8 && col == 1u8 && new.c11 == 0u8 {
-        new = Board {
-            c11: player,
-            c12: new.c12,
-            c13: new.c13,
-            c21: new.c21,
-            c22: new.c22,
-            c23: new.c23,
-            c31: new.c31,
-            c32: new.c32,
-            c33: new.c33,
+    if row == 1u8 && col == 1u8 && updated.r1.e1 == 0u8 {
+        updated = Board {
+            r1: Row {
+                e1: player,
+                e2: updated.r1.e2,
+                e3: updated.r1.e3
+            },
+            r2: updated.r2,
+            r3: updated.r3
         };
-    } else if row == 1u8 && col == 2u8 && new.c12 == 0u8 {
-        new = Board {
-            c11: new.c11,
-            c12: player,
-            c13: new.c13,
-            c21: new.c21,
-            c22: new.c22,
-            c23: new.c23,
-            c31: new.c31,
-            c32: new.c32,
-            c33: new.c33,
+    } else if row == 1u8 && col == 2u8 && updated.r1.e2 == 0u8 {
+        updated = Board {
+            r1: Row {
+                e1: updated.r1.e1,
+                e2: player,
+                e3: updated.r1.e3
+            },
+            r2: updated.r2,
+            r3: updated.r3
         };
-    } else if row == 1u8 && col == 3u8 && new.c13 == 0u8 {
-        new = Board {
-            c11: new.c11,
-            c12: new.c12,
-            c13: player,
-            c21: new.c21,
-            c22: new.c22,
-            c23: new.c23,
-            c31: new.c31,
-            c32: new.c32,
-            c33: new.c33,
+    } else if row == 1u8 && col == 3u8 && updated.r1.e3 == 0u8 {
+        updated = Board {
+            r1: Row {
+                e1: updated.r1.e1,
+                e2: updated.r1.e2,
+                e3: player
+            },
+            r2: updated.r2,
+            r3: updated.r3
         };
-    } else if row == 2u8 && col == 1u8 && new.c21 == 0u8 {
-        new = Board {
-            c11: new.c11,
-            c12: new.c12,
-            c13: new.c13,
-            c21: player,
-            c22: new.c22,
-            c23: new.c23,
-            c31: new.c31,
-            c32: new.c32,
-            c33: new.c33,
+    } else if row == 2u8 && col == 1u8 && updated.r2.e1 == 0u8 {
+        updated = Board {
+            r1: updated.r1,
+            r2: Row {
+                e1: player,
+                e2: updated.r2.e2,
+                e3: updated.r2.e3
+            },
+            r3: updated.r3
         };
-    } else if row == 2u8 && col == 2u8 && new.c22 == 0u8 {
-        new = Board {
-            c11: new.c11,
-            c12: new.c12,
-            c13: new.c13,
-            c21: new.c21,
-            c22: player,
-            c23: new.c23,
-            c31: new.c31,
-            c32: new.c32,
-            c33: new.c33,
+    } else if row == 2u8 && col == 2u8 && updated.r2.e2 == 0u8 {
+        updated = Board {
+            r1: updated.r1,
+            r2: Row {
+                e1: updated.r2.e1,
+                e2: player,
+                e3: updated.r2.e3
+            },
+            r3: updated.r3
         };
-    } else if row == 2u8 && col == 3u8 && new.c23 == 0u8 {
-        new = Board {
-            c11: new.c11,
-            c12: new.c12,
-            c13: new.c13,
-            c21: new.c21,
-            c22: new.c22,
-            c23: player,
-            c31: new.c31,
-            c32: new.c32,
-            c33: new.c33,
+    } else if row == 2u8 && col == 3u8 && updated.r2.e3 == 0u8 {
+        updated = Board {
+            r1: updated.r1,
+            r2: Row {
+                e1: updated.r2.e1,
+                e2: updated.r2.e2,
+                e3: player
+            },
+            r3: updated.r3
         };
-    } else if row == 3u8 && col == 1u8 && new.c31 == 0u8 {
-        new = Board {
-            c11: new.c11,
-            c12: new.c12,
-            c13: new.c13,
-            c21: new.c21,
-            c22: new.c22,
-            c23: new.c23,
-            c31: player,
-            c32: new.c32,
-            c33: new.c33,
+    } else if row == 3u8 && col == 1u8 && updated.r3.e1 == 0u8 {
+        updated = Board {
+            r1: updated.r1,
+            r2: updated.r2,
+            r3: Row {
+                e1: player,
+                e2: updated.r3.e2,
+                e3: updated.r3.e3
+            }
         };
-    } else if row == 3u8 && col == 2u8 && new.c32 == 0u8 {
-        new = Board {
-            c11: new.c11,
-            c12: new.c12,
-            c13: new.c13,
-            c21: new.c21,
-            c22: new.c22,
-            c23: new.c23,
-            c31: new.c31,
-            c32: player,
-            c33: new.c33,
+    } else if row == 3u8 && col == 2u8 && updated.r3.e2 == 0u8 {
+        updated = Board {
+            r1: updated.r1,
+            r2: updated.r2,
+            r3: Row {
+                e1: updated.r3.e1,
+                e2: player,
+                e3: updated.r3.e3
+            }
         };
-    } else if row == 3u8 && col == 3u8 && new.c33 == 0u8 {
-        new = Board {
-            c11: new.c11,
-            c12: new.c12,
-            c13: new.c13,
-            c21: new.c21,
-            c22: new.c22,
-            c23: new.c23,
-            c31: new.c31,
-            c32: new.c32,
-            c33: player,
+    } else if row == 3u8 && col == 3u8 && updated.r3.e3 == 0u8 {
+        updated = Board {
+            r1: updated.r1,
+            r2: updated.r2,
+            r3: Row {
+                e1: updated.r3.e1,
+                e2: updated.r3.e2,
+                e3: player
+            }
         };
     }
-    if win(new, 1u8) {
-        return (new, 1u8);
-    } else if win(new, 2u8) {
-        return (new, 2u8);
+    if win(updated, 1u8) {
+        return (updated, 1u8);
+    } else if win(updated, 2u8) {
+        return (updated, 2u8);
     } else {
-        return (new, 0u8);
+        return (updated, 0u8);
     }
 }
diff --git a/tests/compiler/function/flatten_test.leo b/tests/compiler/function/flatten_test.leo
new file mode 100644
index 0000000000..2605e8975d
--- /dev/null
+++ b/tests/compiler/function/flatten_test.leo
@@ -0,0 +1,131 @@
+/*
+namespace: Compile
+expectation: Pass
+*/
+
+circuit Row {
+    e1: u8,
+    e2: u8,
+    e3: u8
+}
+
+circuit Board {
+    r1: Row,
+    r2: Row,
+    r3: Row,
+}
+
+function win(b: Board, p: u8) -> bool {
+    return
+        (b.r1.e1 == p && b.r1.e2 == p && b.r1.e3 == p) || // row 1
+        (b.r2.e1 == p && b.r2.e2 == p && b.r2.e3 == p) || // row 2
+        (b.r3.e1 == p && b.r3.e3 == p && b.r3.e3 == p) || // row 3
+        (b.r1.e1 == p && b.r2.e1 == p && b.r3.e1 == p) || // column 1
+        (b.r1.e2 == p && b.r2.e3 == p && b.r3.e2 == p) || // column 2
+        (b.r1.e3 == p && b.r2.e3 == p && b.r3.e3 == p) || // column 3
+        (b.r1.e1 == p && b.r2.e2 == p && b.r3.e3 == p) || // diagonal
+        (b.r1.e3 == p && b.r2.e2 == p && b.r3.e1 == p); // other diagonal
+}
+
+@program
+function main(public player: u8, row: u8, col: u8, board: Board) -> (Board, u8) {
+    let updated: Board = board;
+    if row == 1u8 && col == 1u8 && updated.r1.e1 == 0u8 {
+        updated = Board {
+            r1: Row {
+                e1: player,
+                e2: updated.r1.e2,
+                e3: updated.r1.e3
+            },
+            r2: updated.r2,
+            r3: updated.r3
+        };
+    } else if row == 1u8 && col == 2u8 && updated.r1.e2 == 0u8 {
+        updated = Board {
+            r1: Row {
+                e1: updated.r1.e1,
+                e2: player,
+                e3: updated.r1.e3
+            },
+            r2: updated.r2,
+            r3: updated.r3
+        };
+    } else if row == 1u8 && col == 3u8 && updated.r1.e3 == 0u8 {
+        updated = Board {
+            r1: Row {
+                e1: updated.r1.e1,
+                e2: updated.r1.e2,
+                e3: player
+            },
+            r2: updated.r2,
+            r3: updated.r3
+        };
+    } else if row == 2u8 && col == 1u8 && updated.r2.e1 == 0u8 {
+        updated = Board {
+            r1: updated.r1,
+            r2: Row {
+                e1: player,
+                e2: updated.r2.e2,
+                e3: updated.r2.e3
+            },
+            r3: updated.r3
+        };
+    } else if row == 2u8 && col == 2u8 && updated.r2.e2 == 0u8 {
+        updated = Board {
+            r1: updated.r1,
+            r2: Row {
+                e1: updated.r2.e1,
+                e2: player,
+                e3: updated.r2.e3
+            },
+            r3: updated.r3
+        };
+    } else if row == 2u8 && col == 3u8 && updated.r2.e3 == 0u8 {
+        updated = Board {
+            r1: updated.r1,
+            r2: Row {
+                e1: updated.r2.e1,
+                e2: updated.r2.e2,
+                e3: player
+            },
+            r3: updated.r3
+        };
+    } else if row == 3u8 && col == 1u8 && updated.r3.e1 == 0u8 {
+        updated = Board {
+            r1: updated.r1,
+            r2: updated.r2,
+            r3: Row {
+                e1: player,
+                e2: updated.r3.e2,
+                e3: updated.r3.e3
+            }
+        };
+    } else if row == 3u8 && col == 2u8 && updated.r3.e2 == 0u8 {
+        updated = Board {
+            r1: updated.r1,
+            r2: updated.r2,
+            r3: Row {
+                e1: updated.r3.e1,
+                e2: player,
+                e3: updated.r3.e3
+            }
+        };
+    } else if row == 3u8 && col == 3u8 && updated.r3.e3 == 0u8 {
+        updated = Board {
+            r1: updated.r1,
+            r2: updated.r2,
+            r3: Row {
+                e1: updated.r3.e1,
+                e2: updated.r3.e2,
+                e3: player
+            }
+        };
+    }
+    if win(updated, 1u8) {
+        return (updated, 1u8);
+    } else if win(updated, 2u8) {
+        return (updated, 2u8);
+    } else {
+        return (updated, 0u8);
+    }
+}
\ No newline at end of file
diff --git a/tests/compiler/strings/string.leo b/tests/compiler/strings/string.leo
new file mode 100644
index 0000000000..20e94e87d5
--- /dev/null
+++ b/tests/compiler/strings/string.leo
@@ -0,0 +1,16 @@
+/*
+namespace: Compile
+expectation: Pass
+*/
+
+@program
+function main(public a: u32, b: u32) -> u32 {
+    let c: u32 = a + b * 1u32; // c = 3
+    let str:string = "a a a";
+    if(str == "b b  b") {
+        c = 123u32;
+    } else {
+        c = 321u32;
+    }
+    return c;
+}
\ No newline at end of file
diff --git a/tests/expectations/compiler/function/flatten_test.out b/tests/expectations/compiler/function/flatten_test.out
new file mode 100644
index 0000000000..c87bfd1d99
--- /dev/null
+++ b/tests/expectations/compiler/function/flatten_test.out
@@ -0,0 +1,10 @@
+---
+namespace: Compile
+expectation: Pass
+outputs:
+  - output:
+      - initial_input_ast: no input
+    initial_ast: e0516243c09ed592d62da2f5db28b9c23d62a65a2d1d2f36994af7a5200ef832
+    unrolled_ast: e0516243c09ed592d62da2f5db28b9c23d62a65a2d1d2f36994af7a5200ef832
+    ssa_ast: dd25b1abb6b78fa6fac53f5d2c27d9b183728441b9a796a5b24e4d8be238e604
+    flattened_ast: 6a0e4a111f150fd817ba4ba9e4b507395fb7f79993c816ba0d0a8207b6b8ce9c
diff --git a/tests/expectations/compiler/strings/string.out b/tests/expectations/compiler/strings/string.out
new file mode 100644
index 0000000000..a9583a2e4f
--- /dev/null
+++ b/tests/expectations/compiler/strings/string.out
@@ -0,0 +1,10 @@
+---
+namespace: Compile
+expectation: Pass
+outputs:
+  - output:
+      - initial_input_ast: no input
+    initial_ast: eba57d5b70b00976a6f79c9d44cdd4e118accd2d58994f1945a48ca9e992337a
+    unrolled_ast: eba57d5b70b00976a6f79c9d44cdd4e118accd2d58994f1945a48ca9e992337a
+    ssa_ast: 3d21091cc901751cf7dadc488a3c476e09907d4cb246b81ba784c950ac9d5d8b
+    flattened_ast: e1132a0b9e0921130364730da6a1b78208d65f9da1e58aa41513b91616ccb623
diff --git a/tests/expectations/compiler/tuple/function_early_return.out b/tests/expectations/compiler/tuple/function_early_return.out
index 3b4d16839d..fa5cdaba64 100644
--- a/tests/expectations/compiler/tuple/function_early_return.out
+++ b/tests/expectations/compiler/tuple/function_early_return.out
@@ -7,4 +7,4 @@ outputs:
     initial_ast: 6e72ebbc53bc0d5fb75c4c71da657696d7a78596026a0df530338ea4c9f18312
     unrolled_ast: 6e72ebbc53bc0d5fb75c4c71da657696d7a78596026a0df530338ea4c9f18312
     ssa_ast: 7d24d447c2c19bb26fbfdb02ab7d4d46014c73cdc930eb7d6cf48772f60a9715
-    flattened_ast: f73ab695290e1fe161f6438a1fbcc8c31dbb44260b2bd34b9e9912f25201e035
+    flattened_ast: 84ba6ba2c0b2add7ee930629028ebaf1e09b2cfb7515fa65f29b67d261999d20
