diff --git a/compiler/ast/src/expressions/call.rs b/compiler/ast/src/expressions/call.rs
index 12486c1535..95dc172c8d 100644
--- a/compiler/ast/src/expressions/call.rs
+++ b/compiler/ast/src/expressions/call.rs
@@ -24,7 +24,7 @@ use itertools::Itertools as _;
 pub struct CallExpression {
     /// An expression evaluating to a callable function,
     /// either a member of a structure or a free function.
-    pub function: Expression, // todo: make this identifier?
+    pub function: Identifier,
     /// Expressions for the const arguments passed to the function's const parameters.
     pub const_arguments: Vec<Expression>,
     /// Expressions for the arguments passed to the function's parameters.
diff --git a/compiler/parser/src/parser/expression.rs b/compiler/parser/src/parser/expression.rs
index f096c59b53..2af7959411 100644
--- a/compiler/parser/src/parser/expression.rs
+++ b/compiler/parser/src/parser/expression.rs
@@ -441,7 +441,7 @@ impl<N: Network> ParserContext<'_, N> {
 
         Ok(CallExpression {
             span: expr.span() + span,
-            function: name.into(),
+            function: name,
             program: Some(program.name),
             const_arguments: vec![], // we do not expect const arguments for external calls at this time
             arguments,
@@ -522,9 +522,13 @@ impl<N: Network> ParserContext<'_, N> {
                 // square brackets.
                 if self.check(&Token::LeftSquare) {
                     // Check that the expression is an identifier.
-                    if !matches!(expr, Expression::Identifier(_)) {
-                        self.emit_err(ParserError::unexpected(expr.to_string(), "an identifier", expr.span()))
-                    }
+                    let Expression::Identifier(ident) = &expr else {
+                        return Err(leo_errors::LeoError::ParserError(ParserError::unexpected(
+                            expr.to_string(),
+                            "an identifier",
+                            expr.span(),
+                        )));
+                    };
 
                     // Parse a list of const arguments in between `[..]`
                     let const_arguments = self.parse_bracket_comma_list(|p| p.parse_expression().map(Some))?.0;
@@ -535,7 +539,7 @@ impl<N: Network> ParserContext<'_, N> {
                     // Now form a `CallExpression`
                     expr = CallExpression {
                         span: expr.span() + span,
-                        function: expr,
+                        function: *ident,
                         program: self.program_name,
                         const_arguments,
                         arguments,
@@ -556,14 +560,19 @@ impl<N: Network> ParserContext<'_, N> {
                     ArrayAccess { array: expr, index, span: expr_span + span, id: self.node_builder.next_id() }.into();
             } else if self.check(&Token::LeftParen) {
                 // Check that the expression is an identifier.
-                if !matches!(expr, Expression::Identifier(_)) {
-                    self.emit_err(ParserError::unexpected(expr.to_string(), "an identifier", expr.span()))
-                }
+                let Expression::Identifier(ident) = &expr else {
+                    return Err(leo_errors::LeoError::ParserError(ParserError::unexpected(
+                        expr.to_string(),
+                        "an identifier",
+                        expr.span(),
+                    )));
+                };
+
                 // Parse a function call that's by itself.
                 let (arguments, _, span) = self.parse_paren_comma_list(|p| p.parse_expression().map(Some))?;
                 expr = CallExpression {
                     span: expr.span() + span,
-                    function: expr,
+                    function: *ident,
                     program: self.program_name,
                     const_arguments: vec![],
                     arguments,
diff --git a/compiler/passes/src/code_generation/expression.rs b/compiler/passes/src/code_generation/expression.rs
index 6b76068317..f2ae5ab09c 100644
--- a/compiler/passes/src/code_generation/expression.rs
+++ b/compiler/passes/src/code_generation/expression.rs
@@ -45,7 +45,7 @@ use leo_ast::{
 };
 use leo_span::sym;
 
-use std::{borrow::Borrow, fmt::Write as _};
+use std::fmt::Write as _;
 
 /// Implement the necessary methods to visit nodes in the AST.
 impl CodeGeneratingVisitor<'_> {
@@ -519,19 +519,13 @@ impl CodeGeneratingVisitor<'_> {
     }
 
     fn visit_call(&mut self, input: &CallExpression) -> (String, String) {
-        // Lookup the function return type.
-        let function_name = match input.function.borrow() {
-            Expression::Identifier(identifier) => identifier.name,
-            _ => panic!("Parsing guarantees that a function name is always an identifier."),
-        };
-
         let caller_program = self.program_id.expect("Calls only appear within programs.").name.name;
         let callee_program = input.program.unwrap_or(caller_program);
 
         let func_symbol = self
             .state
             .symbol_table
-            .lookup_function(Location::new(callee_program, function_name))
+            .lookup_function(Location::new(callee_program, input.function.name))
             .expect("Type checking guarantees functions exist");
 
         // Need to determine the program the function originated from as well as if the function has a finalize block.
diff --git a/compiler/passes/src/const_propagation_and_unrolling.rs b/compiler/passes/src/const_propagation_and_unrolling.rs
index d1c15d96c1..3cc0500ea4 100644
--- a/compiler/passes/src/const_propagation_and_unrolling.rs
+++ b/compiler/passes/src/const_propagation_and_unrolling.rs
@@ -52,16 +52,11 @@ impl Pass for ConstPropagationAndUnrolling {
 
                 // Emit errors for all problematic calls
                 for call in &monomorphization_output.unresolved_calls {
-                    let callee_name = match call.function {
-                        leo_ast::Expression::Identifier(ref ident) => ident.name,
-                        _ => panic!("Parser ensures `function` is always an identifier."),
-                    };
-
                     if let Some(arg) =
                         call.const_arguments.iter().find(|arg| !matches!(arg, leo_ast::Expression::Literal(_)))
                     {
                         state.handler.emit_err(CompilerError::call_to_generic_function_not_resolved(
-                            callee_name,
+                            call.function.name,
                             arg,
                             call.span,
                         ));
diff --git a/compiler/passes/src/function_inlining/expression.rs b/compiler/passes/src/function_inlining/expression.rs
index d6bca615c7..780af29b37 100644
--- a/compiler/passes/src/function_inlining/expression.rs
+++ b/compiler/passes/src/function_inlining/expression.rs
@@ -42,15 +42,10 @@ impl ExpressionReconstructor for FunctionInliningVisitor<'_> {
             return (input.into(), Default::default());
         }
 
-        // Get the name of the callee function.
-        let function_name = match input.function {
-            Expression::Identifier(identifier) => identifier.name,
-            _ => panic!("Parser guarantees that `input.function` is always an identifier."),
-        };
-
         // Lookup the reconstructed callee function.
         // Since this pass processes functions in post-order, the callee function is guaranteed to exist in `self.reconstructed_functions`
-        let (_, callee) = self.reconstructed_functions.iter().find(|(symbol, _)| *symbol == function_name).unwrap();
+        let (_, callee) =
+            self.reconstructed_functions.iter().find(|(symbol, _)| *symbol == input.function.name).unwrap();
 
         // Inline the callee function, if required, otherwise, return the call expression.
         match callee.variant {
diff --git a/compiler/passes/src/monomorphization/expression.rs b/compiler/passes/src/monomorphization/expression.rs
index c1de0d5400..9e545a6d99 100644
--- a/compiler/passes/src/monomorphization/expression.rs
+++ b/compiler/passes/src/monomorphization/expression.rs
@@ -39,15 +39,10 @@ impl ExpressionReconstructor for MonomorphizationVisitor<'_> {
             return (input_call.into(), Default::default());
         }
 
-        // Extract the function name from the call expression.
-        let Expression::Identifier(Identifier { name: callee_name, .. }) = &input_call.function else {
-            panic!("Parser ensures `function` is always an identifier.")
-        };
-
         // Look up the already reconstructed function by name.
         let callee_fn = self
             .reconstructed_functions
-            .get(callee_name)
+            .get(&input_call.function.name)
             .expect("Callee should already be reconstructed (post-order traversal).");
 
         // Proceed only if the function variant is `inline` and if there are some const arguments.
@@ -70,8 +65,8 @@ impl ExpressionReconstructor for MonomorphizationVisitor<'_> {
         // identifier in the user code.
         let new_callee_name = leo_span::Symbol::intern(&format!(
             "{}::[{}]",
-            callee_name,
-            input_call.const_arguments.iter().map(|arg| arg.to_string()).format(", ")
+            input_call.function.name,
+            input_call.const_arguments.iter().format(", ")
         ));
 
         // Check if the new callee name is not already present in `reconstructed_functions`. This ensures that we do not
@@ -109,14 +104,14 @@ impl ExpressionReconstructor for MonomorphizationVisitor<'_> {
             });
 
             // Now keep track of the function we just monomorphized
-            self.monomorphized_functions.insert(*callee_name);
+            self.monomorphized_functions.insert(input_call.function.name);
         }
 
         // Update call graph with edges for the monomorphized function. We do this by basically cloning the edges in
         // and out of `callee_name` and replicating them for a new node that contains `new_callee_name`.
-        if let Some(neighbors) = self.state.call_graph.neighbors(callee_name) {
+        if let Some(neighbors) = self.state.call_graph.neighbors(&input_call.function.name) {
             for neighbor in neighbors {
-                if neighbor != *callee_name {
+                if neighbor != input_call.function.name {
                     self.state.call_graph.add_edge(new_callee_name, neighbor);
                 }
             }
@@ -126,11 +121,11 @@ impl ExpressionReconstructor for MonomorphizationVisitor<'_> {
         // Finally, construct the updated call expression that points to the monomorphized version and return it.
         (
             CallExpression {
-                function: Expression::Identifier(Identifier {
+                function: Identifier {
                     name: new_callee_name, // use the new name
                     span: leo_span::Span::default(),
                     id: self.state.node_builder.next_id(),
-                }),
+                },
                 const_arguments: vec![], // remove const arguments
                 arguments: input_call.arguments,
                 program: input_call.program,
diff --git a/compiler/passes/src/monomorphization/program.rs b/compiler/passes/src/monomorphization/program.rs
index 2dc4e4aa58..fc386ed99b 100644
--- a/compiler/passes/src/monomorphization/program.rs
+++ b/compiler/passes/src/monomorphization/program.rs
@@ -57,10 +57,7 @@ impl ProgramReconstructor for MonomorphizationVisitor<'_> {
         self.reconstructed_functions.retain(|f, _| {
             let is_monomorphized = self.monomorphized_functions.contains(f);
 
-            let is_still_called = self.unresolved_calls.iter().any(|c| match &c.function {
-                leo_ast::Expression::Identifier(ident) => ident.name == *f,
-                _ => panic!("Parser guarantees `function` is always an identifier."),
-            });
+            let is_still_called = self.unresolved_calls.iter().any(|c| c.function.name == *f);
 
             if is_monomorphized && !is_still_called {
                 // It's monomorphized and there are no unresolved calls to it - remove it.
diff --git a/compiler/passes/src/processing_script/expression.rs b/compiler/passes/src/processing_script/expression.rs
index 3eefa6ddbf..cc06ddd3f2 100644
--- a/compiler/passes/src/processing_script/expression.rs
+++ b/compiler/passes/src/processing_script/expression.rs
@@ -24,20 +24,16 @@ impl ExpressionReconstructor for ProcessingScriptVisitor<'_> {
 
     fn reconstruct_call(&mut self, input: CallExpression) -> (Expression, Self::AdditionalOutput) {
         if !matches!(self.current_variant, Variant::Script) {
-            // Get the function symbol.
-            let Expression::Identifier(ident) = &input.function else {
-                panic!("Parsing guarantees that a function name is always an identifier.");
-            };
-
             let callee_program = input.program.unwrap_or(self.program_name);
 
-            let Some(func_symbol) = self.state.symbol_table.lookup_function(Location::new(callee_program, ident.name))
+            let Some(func_symbol) =
+                self.state.symbol_table.lookup_function(Location::new(callee_program, input.function.name))
             else {
                 panic!("Type checking should have prevented this.");
             };
 
             if matches!(func_symbol.function.variant, Variant::Script) {
-                self.state.handler.emit_err(TypeCheckerError::non_script_calls_script(ident, input.span))
+                self.state.handler.emit_err(TypeCheckerError::non_script_calls_script(input.function, input.span))
             }
         }
         (input.into(), Default::default())
diff --git a/compiler/passes/src/static_analysis/expression.rs b/compiler/passes/src/static_analysis/expression.rs
index 72cc4f0e0a..018481c7eb 100644
--- a/compiler/passes/src/static_analysis/expression.rs
+++ b/compiler/passes/src/static_analysis/expression.rs
@@ -39,10 +39,6 @@ impl ExpressionVisitor for StaticAnalyzingVisitor<'_> {
     }
 
     fn visit_call(&mut self, input: &CallExpression, _: &Self::AdditionalInput) -> Self::Output {
-        let Expression::Identifier(ident) = &input.function else {
-            panic!("Parsing guarantees that a function name is always an identifier.");
-        };
-
         let caller_program = self.current_program;
         let callee_program = input.program.unwrap_or(caller_program);
 
@@ -52,7 +48,7 @@ impl ExpressionVisitor for StaticAnalyzingVisitor<'_> {
             && self.variant == Some(Variant::AsyncTransition)
             && callee_program != caller_program
         {
-            self.assert_simple_async_transition_call(callee_program, ident.name, input.span());
+            self.assert_simple_async_transition_call(callee_program, input.function.name, input.span());
         }
 
         // Look up the function and check if it is a non-async call.
@@ -61,7 +57,7 @@ impl ExpressionVisitor for StaticAnalyzingVisitor<'_> {
         let func_symbol = self
             .state
             .symbol_table
-            .lookup_function(Location::new(function_program, ident.name))
+            .lookup_function(Location::new(function_program, input.function.name))
             .expect("Type checking guarantees functions exist.");
 
         if func_symbol.function.variant == Variant::Transition {
diff --git a/compiler/passes/src/type_checking/expression.rs b/compiler/passes/src/type_checking/expression.rs
index 8760fccde1..ad252015ea 100644
--- a/compiler/passes/src/type_checking/expression.rs
+++ b/compiler/passes/src/type_checking/expression.rs
@@ -813,16 +813,12 @@ impl ExpressionVisitor for TypeCheckingVisitor<'_> {
     }
 
     fn visit_call(&mut self, input: &CallExpression, expected: &Self::AdditionalInput) -> Self::Output {
-        // Get the function symbol.
-        let Expression::Identifier(ident) = &input.function else {
-            panic!("Parsing guarantees that a function name is always an identifier.");
-        };
-
         let callee_program = input.program.or(self.scope_state.program_name).unwrap();
 
-        let Some(func_symbol) = self.state.symbol_table.lookup_function(Location::new(callee_program, ident.name))
+        let Some(func_symbol) =
+            self.state.symbol_table.lookup_function(Location::new(callee_program, input.function.name))
         else {
-            self.emit_err(TypeCheckerError::unknown_sym("function", ident.name, ident.span()));
+            self.emit_err(TypeCheckerError::unknown_sym("function", input.function.name, input.function.span()));
             return Type::Err;
         };
 
@@ -851,8 +847,11 @@ impl ExpressionVisitor for TypeCheckingVisitor<'_> {
         // Async functions return a single future.
         let mut ret = if func.variant == Variant::AsyncFunction {
             // Type check after resolving the input types.
-            let actual =
-                Type::Future(FutureType::new(Vec::new(), Some(Location::new(callee_program, ident.name)), false));
+            let actual = Type::Future(FutureType::new(
+                Vec::new(),
+                Some(Location::new(callee_program, input.function.name)),
+                false,
+            ));
             match expected {
                 Some(Type::Future(_)) | None => {
                     // If the expected type is a `Future` or if it's not set, then just return the
@@ -869,18 +868,21 @@ impl ExpressionVisitor for TypeCheckingVisitor<'_> {
             // Fully infer future type.
             let Some(inputs) = self
                 .async_function_input_types
-                .get(&Location::new(callee_program, Symbol::intern(&format!("finalize/{}", ident.name))))
+                .get(&Location::new(callee_program, Symbol::intern(&format!("finalize/{}", input.function.name))))
             else {
-                self.emit_err(TypeCheckerError::async_function_not_found(ident.name, input.span));
+                self.emit_err(TypeCheckerError::async_function_not_found(input.function.name, input.span));
                 return Type::Future(FutureType::new(
                     Vec::new(),
-                    Some(Location::new(callee_program, ident.name)),
+                    Some(Location::new(callee_program, input.function.name)),
                     false,
                 ));
             };
 
-            let future_type =
-                Type::Future(FutureType::new(inputs.clone(), Some(Location::new(callee_program, ident.name)), true));
+            let future_type = Type::Future(FutureType::new(
+                inputs.clone(),
+                Some(Location::new(callee_program, input.function.name)),
+                true,
+            ));
             let fully_inferred_type = match &func.output_type {
                 Type::Tuple(tup) => Type::Tuple(TupleType::new(
                     tup.elements()
@@ -981,7 +983,7 @@ impl ExpressionVisitor for TypeCheckingVisitor<'_> {
 
         // Don't add external functions to call graph. Since imports are acyclic, these can never produce a cycle.
         if input.program.unwrap() == self.scope_state.program_name.unwrap() {
-            self.state.call_graph.add_edge(caller_name, ident.name);
+            self.state.call_graph.add_edge(caller_name, input.function.name);
         }
 
         if func.variant.is_transition()
@@ -1020,14 +1022,14 @@ impl ExpressionVisitor for TypeCheckingVisitor<'_> {
                 .symbol_table
                 .attach_finalizer(
                     Location::new(callee_program, caller_name),
-                    Location::new(callee_program, ident.name),
+                    Location::new(callee_program, input.function.name),
                     input_futures,
                     inferred_finalize_inputs.clone(),
                 )
                 .expect("Failed to attach finalizer");
             // Create expectation for finalize inputs that will be checked when checking corresponding finalize function signature.
             self.async_function_callers
-                .entry(Location::new(self.scope_state.program_name.unwrap(), ident.name))
+                .entry(Location::new(self.scope_state.program_name.unwrap(), input.function.name))
                 .or_default()
                 .insert(self.scope_state.location());
 
@@ -1037,7 +1039,7 @@ impl ExpressionVisitor for TypeCheckingVisitor<'_> {
             // Update ret to reflect fully inferred future type.
             ret = Type::Future(FutureType::new(
                 inferred_finalize_inputs,
-                Some(Location::new(callee_program, ident.name)),
+                Some(Location::new(callee_program, input.function.name)),
                 true,
             ));
 
@@ -1046,7 +1048,7 @@ impl ExpressionVisitor for TypeCheckingVisitor<'_> {
         }
 
         // Set call location so that definition statement knows where future comes from.
-        self.scope_state.call_location = Some(Location::new(callee_program, ident.name));
+        self.scope_state.call_location = Some(Location::new(callee_program, input.function.name));
 
         ret
     }
diff --git a/interpreter/src/cursor.rs b/interpreter/src/cursor.rs
index 6e1a03ebf8..340d460520 100644
--- a/interpreter/src/cursor.rs
+++ b/interpreter/src/cursor.rs
@@ -815,17 +815,13 @@ impl Cursor {
             }
             Expression::Call(call) if step == 1 => {
                 let len = self.values.len();
-                let (program, name) = match &call.function {
-                    Expression::Identifier(id) => {
-                        let maybe_program = call.program.or_else(|| self.current_program());
-                        if let Some(program) = maybe_program {
-                            (program, id.name)
-                        } else {
-                            halt!(call.span, "No current program");
-                        }
+                let (program, name) = {
+                    let maybe_program = call.program.or_else(|| self.current_program());
+                    if let Some(program) = maybe_program {
+                        (program, call.function.name)
+                    } else {
+                        halt!(call.span, "No current program");
                     }
-                    Expression::Locator(locator) => (locator.program.name.name, locator.name),
-                    _ => tc_fail!(),
                 };
                 // It's a bit cheesy to collect the arguments into a Vec first, but it's the easiest way
                 // to handle lifetimes here.
diff --git a/interpreter/src/lib.rs b/interpreter/src/lib.rs
index 9a4de679b5..9800f3a0a3 100644
--- a/interpreter/src/lib.rs
+++ b/interpreter/src/lib.rs
@@ -191,7 +191,7 @@ pub fn find_and_run_tests(
         assert!(function.variant.is_script(), "Type checking should ensure test functions are transitions or scripts.");
 
         let call = CallExpression {
-            function: Identifier::new(function.identifier.name, interpreter.node_builder.next_id()).into(),
+            function: Identifier::new(function.identifier.name, interpreter.node_builder.next_id()),
             const_arguments: vec![], // scripts don't have const parameters for now
             arguments: Vec::new(),
             program: Some(id.program),
diff --git a/tests/expectations/parser-expression/expression/access/call.out b/tests/expectations/parser-expression/expression/access/call.out
index febb63fd59..f3a65ce160 100644
--- a/tests/expectations/parser-expression/expression/access/call.out
+++ b/tests/expectations/parser-expression/expression/access/call.out
@@ -1,14 +1,12 @@
 {
   "Call": {
     "function": {
-      "Identifier": {
-        "name": "x",
-        "span": {
-          "lo": 0,
-          "hi": 1
-        },
-        "id": 0
-      }
+      "name": "x",
+      "span": {
+        "lo": 0,
+        "hi": 1
+      },
+      "id": 0
     },
     "const_arguments": [],
     "arguments": [],
@@ -24,14 +22,12 @@
 {
   "Call": {
     "function": {
-      "Identifier": {
-        "name": "X",
-        "span": {
-          "lo": 4,
-          "hi": 5
-        },
-        "id": 0
-      }
+      "name": "X",
+      "span": {
+        "lo": 4,
+        "hi": 5
+      },
+      "id": 0
     },
     "const_arguments": [],
     "arguments": [],
@@ -47,14 +43,12 @@
 {
   "Call": {
     "function": {
-      "Identifier": {
-        "name": "x",
-        "span": {
-          "lo": 8,
-          "hi": 9
-        },
-        "id": 0
-      }
+      "name": "x",
+      "span": {
+        "lo": 8,
+        "hi": 9
+      },
+      "id": 0
     },
     "const_arguments": [],
     "arguments": [
@@ -81,14 +75,12 @@
 {
   "Call": {
     "function": {
-      "Identifier": {
-        "name": "x",
-        "span": {
-          "lo": 13,
-          "hi": 14
-        },
-        "id": 0
-      }
+      "name": "x",
+      "span": {
+        "lo": 13,
+        "hi": 14
+      },
+      "id": 0
     },
     "const_arguments": [],
     "arguments": [
@@ -125,14 +117,12 @@
 {
   "Call": {
     "function": {
-      "Identifier": {
-        "name": "x",
-        "span": {
-          "lo": 21,
-          "hi": 22
-        },
-        "id": 0
-      }
+      "name": "x",
+      "span": {
+        "lo": 21,
+        "hi": 22
+      },
+      "id": 0
     },
     "const_arguments": [],
     "arguments": [
@@ -179,14 +169,12 @@
 {
   "Call": {
     "function": {
-      "Identifier": {
-        "name": "x",
-        "span": {
-          "lo": 32,
-          "hi": 33
-        },
-        "id": 0
-      }
+      "name": "x",
+      "span": {
+        "lo": 32,
+        "hi": 33
+      },
+      "id": 0
     },
     "const_arguments": [],
     "arguments": [
@@ -233,14 +221,12 @@
 {
   "Call": {
     "function": {
-      "Identifier": {
-        "name": "x",
-        "span": {
-          "lo": 47,
-          "hi": 48
-        },
-        "id": 0
-      }
+      "name": "x",
+      "span": {
+        "lo": 47,
+        "hi": 48
+      },
+      "id": 0
     },
     "const_arguments": [
       {
@@ -298,14 +284,12 @@
 {
   "Call": {
     "function": {
-      "Identifier": {
-        "name": "x",
-        "span": {
-          "lo": 63,
-          "hi": 64
-        },
-        "id": 0
-      }
+      "name": "x",
+      "span": {
+        "lo": 63,
+        "hi": 64
+      },
+      "id": 0
     },
     "const_arguments": [
       {
diff --git a/tests/expectations/parser-expression/expression/unary/neg.out b/tests/expectations/parser-expression/expression/unary/neg.out
index 85c3bc9c89..6c4b17879f 100644
--- a/tests/expectations/parser-expression/expression/unary/neg.out
+++ b/tests/expectations/parser-expression/expression/unary/neg.out
@@ -24,14 +24,12 @@
     "receiver": {
       "Call": {
         "function": {
-          "Identifier": {
-            "name": "x",
-            "span": {
-              "lo": 4,
-              "hi": 5
-            },
-            "id": 0
-          }
+          "name": "x",
+          "span": {
+            "lo": 4,
+            "hi": 5
+          },
+          "id": 0
         },
         "const_arguments": [],
         "arguments": [],
diff --git a/tests/expectations/parser-expression/expression/unary/not.out b/tests/expectations/parser-expression/expression/unary/not.out
index 2988e0a755..8fca15dedd 100644
--- a/tests/expectations/parser-expression/expression/unary/not.out
+++ b/tests/expectations/parser-expression/expression/unary/not.out
@@ -24,14 +24,12 @@
     "receiver": {
       "Call": {
         "function": {
-          "Identifier": {
-            "name": "x",
-            "span": {
-              "lo": 4,
-              "hi": 5
-            },
-            "id": 0
-          }
+          "name": "x",
+          "span": {
+            "lo": 4,
+            "hi": 5
+          },
+          "id": 0
         },
         "const_arguments": [],
         "arguments": [],
diff --git a/tests/expectations/parser-statement/finalize/decrement.out b/tests/expectations/parser-statement/finalize/decrement.out
index ba9e4cb4e7..2f9197ed6d 100644
--- a/tests/expectations/parser-statement/finalize/decrement.out
+++ b/tests/expectations/parser-statement/finalize/decrement.out
@@ -3,14 +3,12 @@
     "expression": {
       "Call": {
         "function": {
-          "Identifier": {
-            "name": "decrement",
-            "span": {
-              "lo": 0,
-              "hi": 9
-            },
-            "id": 0
-          }
+          "name": "decrement",
+          "span": {
+            "lo": 0,
+            "hi": 9
+          },
+          "id": 0
         },
         "const_arguments": [],
         "arguments": [
@@ -66,14 +64,12 @@
     "expression": {
       "Call": {
         "function": {
-          "Identifier": {
-            "name": "decrement",
-            "span": {
-              "lo": 26,
-              "hi": 35
-            },
-            "id": 0
-          }
+          "name": "decrement",
+          "span": {
+            "lo": 26,
+            "hi": 35
+          },
+          "id": 0
         },
         "const_arguments": [],
         "arguments": [
@@ -139,14 +135,12 @@
     "expression": {
       "Call": {
         "function": {
-          "Identifier": {
-            "name": "decrement",
-            "span": {
-              "lo": 52,
-              "hi": 61
-            },
-            "id": 0
-          }
+          "name": "decrement",
+          "span": {
+            "lo": 52,
+            "hi": 61
+          },
+          "id": 0
         },
         "const_arguments": [],
         "arguments": [
@@ -192,14 +186,12 @@
     "expression": {
       "Call": {
         "function": {
-          "Identifier": {
-            "name": "decrement",
-            "span": {
-              "lo": 73,
-              "hi": 82
-            },
-            "id": 0
-          }
+          "name": "decrement",
+          "span": {
+            "lo": 73,
+            "hi": 82
+          },
+          "id": 0
         },
         "const_arguments": [],
         "arguments": [],
diff --git a/tests/expectations/parser-statement/finalize/increment.out b/tests/expectations/parser-statement/finalize/increment.out
index 35dd6a035d..b42587c095 100644
--- a/tests/expectations/parser-statement/finalize/increment.out
+++ b/tests/expectations/parser-statement/finalize/increment.out
@@ -3,14 +3,12 @@
     "expression": {
       "Call": {
         "function": {
-          "Identifier": {
-            "name": "increment",
-            "span": {
-              "lo": 0,
-              "hi": 9
-            },
-            "id": 0
-          }
+          "name": "increment",
+          "span": {
+            "lo": 0,
+            "hi": 9
+          },
+          "id": 0
         },
         "const_arguments": [],
         "arguments": [
@@ -66,14 +64,12 @@
     "expression": {
       "Call": {
         "function": {
-          "Identifier": {
-            "name": "increment",
-            "span": {
-              "lo": 26,
-              "hi": 35
-            },
-            "id": 0
-          }
+          "name": "increment",
+          "span": {
+            "lo": 26,
+            "hi": 35
+          },
+          "id": 0
         },
         "const_arguments": [],
         "arguments": [
@@ -139,14 +135,12 @@
     "expression": {
       "Call": {
         "function": {
-          "Identifier": {
-            "name": "increment",
-            "span": {
-              "lo": 52,
-              "hi": 61
-            },
-            "id": 0
-          }
+          "name": "increment",
+          "span": {
+            "lo": 52,
+            "hi": 61
+          },
+          "id": 0
         },
         "const_arguments": [],
         "arguments": [
@@ -192,14 +186,12 @@
     "expression": {
       "Call": {
         "function": {
-          "Identifier": {
-            "name": "increment",
-            "span": {
-              "lo": 73,
-              "hi": 82
-            },
-            "id": 0
-          }
+          "name": "increment",
+          "span": {
+            "lo": 73,
+            "hi": 82
+          },
+          "id": 0
         },
         "const_arguments": [],
         "arguments": [],
diff --git a/tests/expectations/parser-statement/statement/assign.out b/tests/expectations/parser-statement/statement/assign.out
index fa5e790873..a6bdd895a7 100644
--- a/tests/expectations/parser-statement/statement/assign.out
+++ b/tests/expectations/parser-statement/statement/assign.out
@@ -93,14 +93,12 @@
     "value": {
       "Call": {
         "function": {
-          "Identifier": {
-            "name": "x",
-            "span": {
-              "lo": 23,
-              "hi": 24
-            },
-            "id": 1
-          }
+          "name": "x",
+          "span": {
+            "lo": 23,
+            "hi": 24
+          },
+          "id": 1
         },
         "const_arguments": [],
         "arguments": [],
diff --git a/tests/expectations/parser-statement/statement/definition.out b/tests/expectations/parser-statement/statement/definition.out
index ba4b739bd5..df2ba9d95b 100644
--- a/tests/expectations/parser-statement/statement/definition.out
+++ b/tests/expectations/parser-statement/statement/definition.out
@@ -102,14 +102,12 @@
     "value": {
       "Call": {
         "function": {
-          "Identifier": {
-            "name": "x",
-            "span": {
-              "lo": 48,
-              "hi": 49
-            },
-            "id": 1
-          }
+          "name": "x",
+          "span": {
+            "lo": 48,
+            "hi": 49
+          },
+          "id": 1
         },
         "const_arguments": [],
         "arguments": [],
@@ -264,14 +262,12 @@
     "value": {
       "Call": {
         "function": {
-          "Identifier": {
-            "name": "x",
-            "span": {
-              "lo": 123,
-              "hi": 124
-            },
-            "id": 1
-          }
+          "name": "x",
+          "span": {
+            "lo": 123,
+            "hi": 124
+          },
+          "id": 1
         },
         "const_arguments": [],
         "arguments": [],
@@ -350,14 +346,12 @@
     "value": {
       "Call": {
         "function": {
-          "Identifier": {
-            "name": "mint_private",
-            "span": {
-              "lo": 180,
-              "hi": 192
-            },
-            "id": 3
-          }
+          "name": "mint_private",
+          "span": {
+            "lo": 180,
+            "hi": 192
+          },
+          "id": 3
         },
         "const_arguments": [],
         "arguments": [
@@ -421,14 +415,12 @@
     "value": {
       "Call": {
         "function": {
-          "Identifier": {
-            "name": "mint_private",
-            "span": {
-              "lo": 245,
-              "hi": 257
-            },
-            "id": 4
-          }
+          "name": "mint_private",
+          "span": {
+            "lo": 245,
+            "hi": 257
+          },
+          "id": 4
         },
         "const_arguments": [],
         "arguments": [
@@ -568,14 +560,12 @@
     "value": {
       "Call": {
         "function": {
-          "Identifier": {
-            "name": "x",
-            "span": {
-              "lo": 317,
-              "hi": 318
-            },
-            "id": 1
-          }
+          "name": "x",
+          "span": {
+            "lo": 317,
+            "hi": 318
+          },
+          "id": 1
         },
         "const_arguments": [],
         "arguments": [],
@@ -699,14 +689,12 @@
     "value": {
       "Call": {
         "function": {
-          "Identifier": {
-            "name": "x",
-            "span": {
-              "lo": 372,
-              "hi": 373
-            },
-            "id": 1
-          }
+          "name": "x",
+          "span": {
+            "lo": 372,
+            "hi": 373
+          },
+          "id": 1
         },
         "const_arguments": [],
         "arguments": [],
diff --git a/tests/expectations/parser-statement/statement/expression.out b/tests/expectations/parser-statement/statement/expression.out
index 976fb0bced..99d4a72eda 100644
--- a/tests/expectations/parser-statement/statement/expression.out
+++ b/tests/expectations/parser-statement/statement/expression.out
@@ -63,14 +63,12 @@
     "expression": {
       "Call": {
         "function": {
-          "Identifier": {
-            "name": "x",
-            "span": {
-              "lo": 11,
-              "hi": 12
-            },
-            "id": 0
-          }
+          "name": "x",
+          "span": {
+            "lo": 11,
+            "hi": 12
+          },
+          "id": 0
         },
         "const_arguments": [],
         "arguments": [],
diff --git a/tests/expectations/parser-statement/statement/finalize_fail.out b/tests/expectations/parser-statement/statement/finalize_fail.out
index d7406a4e1a..6c29665cde 100644
--- a/tests/expectations/parser-statement/statement/finalize_fail.out
+++ b/tests/expectations/parser-statement/statement/finalize_fail.out
@@ -3,14 +3,12 @@
     "expression": {
       "Call": {
         "function": {
-          "Identifier": {
-            "name": "finalize",
-            "span": {
-              "lo": 0,
-              "hi": 8
-            },
-            "id": 0
-          }
+          "name": "finalize",
+          "span": {
+            "lo": 0,
+            "hi": 8
+          },
+          "id": 0
         },
         "const_arguments": [],
         "arguments": [
diff --git a/tests/expectations/parser/functions/bounded_recursion.out b/tests/expectations/parser/functions/bounded_recursion.out
index 2b7b9ce6b8..6ca869b759 100644
--- a/tests/expectations/parser/functions/bounded_recursion.out
+++ b/tests/expectations/parser/functions/bounded_recursion.out
@@ -122,14 +122,12 @@
                             "expression": {
                               "Call": {
                                 "function": {
-                                  "Identifier": {
-                                    "name": "x",
-                                    "span": {
-                                      "lo": 95,
-                                      "hi": 96
-                                    },
-                                    "id": 9
-                                  }
+                                  "name": "x",
+                                  "span": {
+                                    "lo": 95,
+                                    "hi": 96
+                                  },
+                                  "id": 9
                                 },
                                 "const_arguments": [],
                                 "arguments": [
@@ -265,14 +263,12 @@
                     "expression": {
                       "Call": {
                         "function": {
-                          "Identifier": {
-                            "name": "x",
-                            "span": {
-                              "lo": 168,
-                              "hi": 169
-                            },
-                            "id": 23
-                          }
+                          "name": "x",
+                          "span": {
+                            "lo": 168,
+                            "hi": 169
+                          },
+                          "id": 23
                         },
                         "const_arguments": [],
                         "arguments": [
diff --git a/tests/expectations/parser/functions/infinite_recursion.out b/tests/expectations/parser/functions/infinite_recursion.out
index be49404c39..3f045e1099 100644
--- a/tests/expectations/parser/functions/infinite_recursion.out
+++ b/tests/expectations/parser/functions/infinite_recursion.out
@@ -63,14 +63,12 @@
                     "expression": {
                       "Call": {
                         "function": {
-                          "Identifier": {
-                            "name": "inf",
-                            "span": {
-                              "lo": 56,
-                              "hi": 59
-                            },
-                            "id": 4
-                          }
+                          "name": "inf",
+                          "span": {
+                            "lo": 56,
+                            "hi": 59
+                          },
+                          "id": 4
                         },
                         "const_arguments": [],
                         "arguments": [],
@@ -155,14 +153,12 @@
                     "expression": {
                       "Call": {
                         "function": {
-                          "Identifier": {
-                            "name": "inf",
-                            "span": {
-                              "lo": 115,
-                              "hi": 118
-                            },
-                            "id": 13
-                          }
+                          "name": "inf",
+                          "span": {
+                            "lo": 115,
+                            "hi": 118
+                          },
+                          "id": 13
                         },
                         "const_arguments": [],
                         "arguments": [],
diff --git a/tests/expectations/parser/future/explicit_future_typing.out b/tests/expectations/parser/future/explicit_future_typing.out
index 2b208e0506..810f4ce700 100644
--- a/tests/expectations/parser/future/explicit_future_typing.out
+++ b/tests/expectations/parser/future/explicit_future_typing.out
@@ -166,14 +166,12 @@
                     "value": {
                       "Call": {
                         "function": {
-                          "Identifier": {
-                            "name": "add",
-                            "span": {
-                              "lo": 167,
-                              "hi": 170
-                            },
-                            "id": 6
-                          }
+                          "name": "add",
+                          "span": {
+                            "lo": 167,
+                            "hi": 170
+                          },
+                          "id": 6
                         },
                         "const_arguments": [],
                         "arguments": [],
@@ -232,14 +230,12 @@
                     "value": {
                       "Call": {
                         "function": {
-                          "Identifier": {
-                            "name": "nested_fn",
-                            "span": {
-                              "lo": 241,
-                              "hi": 250
-                            },
-                            "id": 11
-                          }
+                          "name": "nested_fn",
+                          "span": {
+                            "lo": 241,
+                            "hi": 250
+                          },
+                          "id": 11
                         },
                         "const_arguments": [],
                         "arguments": [],
@@ -319,14 +315,12 @@
                     "value": {
                       "Call": {
                         "function": {
-                          "Identifier": {
-                            "name": "finalize_main",
-                            "span": {
-                              "lo": 346,
-                              "hi": 359
-                            },
-                            "id": 15
-                          }
+                          "name": "finalize_main",
+                          "span": {
+                            "lo": 346,
+                            "hi": 359
+                          },
+                          "id": 15
                         },
                         "const_arguments": [],
                         "arguments": [
diff --git a/tests/expectations/parser/program/async_basic.out b/tests/expectations/parser/program/async_basic.out
index c841c93790..0960c2d4aa 100644
--- a/tests/expectations/parser/program/async_basic.out
+++ b/tests/expectations/parser/program/async_basic.out
@@ -113,14 +113,12 @@
                     "value": {
                       "Call": {
                         "function": {
-                          "Identifier": {
-                            "name": "finalize_main",
-                            "span": {
-                              "lo": 114,
-                              "hi": 127
-                            },
-                            "id": 7
-                          }
+                          "name": "finalize_main",
+                          "span": {
+                            "lo": 114,
+                            "hi": 127
+                          },
+                          "id": 7
                         },
                         "const_arguments": [],
                         "arguments": [
