diff --git a/ast/src/errors/syntax.rs b/ast/src/errors/syntax.rs
index 253a416149..fa255b07f7 100644
--- a/ast/src/errors/syntax.rs
+++ b/ast/src/errors/syntax.rs
@@ -15,6 +15,7 @@ impl From<Error<Rule>> for SyntaxError {
             Rule::type_integer => "`u32`".to_owned(),
             Rule::type_field => "`field`".to_owned(),
             Rule::type_group => "`group`".to_owned(),
+            Rule::address => "an aleo address: `aleo1...`".to_owned(),
             Rule::file => "an import, circuit, or function".to_owned(),
             Rule::identifier => "a variable name".to_owned(),
             Rule::type_ => "a type".to_owned(),
diff --git a/ast/src/expressions/array_initializer_expression.rs b/ast/src/expressions/array_initializer_expression.rs
index 10b8b2bf1b..53308977a3 100644
--- a/ast/src/expressions/array_initializer_expression.rs
+++ b/ast/src/expressions/array_initializer_expression.rs
@@ -1,4 +1,4 @@
-use crate::{ast::Rule, common::SpreadOrExpression, values::Value, SpanDef};
+use crate::{ast::Rule, common::SpreadOrExpression, values::PositiveNumber, SpanDef};
 
 use pest::Span;
 use pest_ast::FromPest;
@@ -8,7 +8,7 @@ use serde::Serialize;
 #[pest_ast(rule(Rule::expression_array_initializer))]
 pub struct ArrayInitializerExpression<'ast> {
     pub expression: Box<SpreadOrExpression<'ast>>,
-    pub count: Value<'ast>,
+    pub count: PositiveNumber<'ast>,
     #[pest_ast(outer())]
     #[serde(with = "SpanDef")]
     pub span: Span<'ast>,
diff --git a/ast/src/leo.pest b/ast/src/leo.pest
index b92ae1762e..b0aaf9603d 100644
--- a/ast/src/leo.pest
+++ b/ast/src/leo.pest
@@ -169,7 +169,7 @@ type_self = { "Self" }
 type_circuit = { identifier }
 
 // Declared in types/array_type.rs
-type_array = { type_data ~ ("[" ~ value ~ "]")+ }
+type_array = { type_data ~ ("[" ~ positive_number ~ "]")+ }
 
 type_list = _{ (type_ ~ ("," ~ type_)*)? }
 
@@ -188,6 +188,9 @@ value = {
 // Declared in values/number_value.rs
 value_number = @{ (("-" ~ ASCII_NONZERO_DIGIT) | "0" | ASCII_NONZERO_DIGIT) ~ ASCII_DIGIT* }
 
+// Declared in values/number_positive.rs
+positive_number = @{ ("0" | ASCII_NONZERO_DIGIT) ~ ASCII_DIGIT*}
+
 // Declared in values/number_implicit_value.rs
 value_implicit = { value_number }
 
@@ -206,7 +209,7 @@ group_tuple = {"(" ~ value_number ~ "," ~ value_number ~ ")"}
 group_single_or_tuple = {value_number | group_tuple}
 
 // Declared in values/address.rs
-address = @{ (LOWERCASE_LETTER | ASCII_DIGIT)* }
+address = @{ "aleo" ~ ASCII_DIGIT ~ (LOWERCASE_LETTER | ASCII_DIGIT){58} }
 
 // Declared in values/address_value.rs
 value_address = { type_address ~ "(" ~ address ~ ")" }
@@ -271,7 +274,7 @@ expression_tuple = _{ (expression ~ ("," ~ expression)*)? }
 expression = { expression_term ~ (operation_binary ~ expression_term)* }
 
 // Declared in expressions/array_initializer_expression.rs
-expression_array_initializer = { "[" ~ spread_or_expression ~ ";" ~ value ~ "]" }
+expression_array_initializer = { "[" ~ spread_or_expression ~ ";" ~ positive_number ~ "]" }
 
 // Declared in expressions/array_inline_expression.rs
 expression_array_inline = { "[" ~ NEWLINE* ~ inline_array_inner ~ NEWLINE* ~ "]"}
@@ -294,12 +297,11 @@ statement = {
     (statement_return
     | statement_conditional
     | statement_for
-      | (statement_multiple_assignment
-      | statement_macro
-      | statement_definition
-      | statement_assign
-      | statement_expression
-      )
+    | statement_multiple_assignment
+    | statement_macro
+    | statement_definition
+    | statement_assign
+    | statement_expression
     ) ~ NEWLINE*
 }
 
diff --git a/ast/src/types/array_type.rs b/ast/src/types/array_type.rs
index bc5b04dd7b..2795020e7c 100644
--- a/ast/src/types/array_type.rs
+++ b/ast/src/types/array_type.rs
@@ -1,4 +1,4 @@
-use crate::{ast::Rule, types::DataType, values::Value, SpanDef};
+use crate::{ast::Rule, types::DataType, values::PositiveNumber, SpanDef};
 
 use pest::Span;
 use pest_ast::FromPest;
@@ -8,7 +8,7 @@ use serde::Serialize;
 #[pest_ast(rule(Rule::type_array))]
 pub struct ArrayType<'ast> {
     pub _type: DataType,
-    pub dimensions: Vec<Value<'ast>>,
+    pub dimensions: Vec<PositiveNumber<'ast>>,
     #[pest_ast(outer())]
     #[serde(with = "SpanDef")]
     pub span: Span<'ast>,
diff --git a/ast/src/values/mod.rs b/ast/src/values/mod.rs
index d3c957c48e..ab289284f1 100644
--- a/ast/src/values/mod.rs
+++ b/ast/src/values/mod.rs
@@ -22,5 +22,8 @@ pub use number_implicit_value::*;
 pub mod number_value;
 pub use number_value::*;
 
+pub mod positive_number;
+pub use positive_number::*;
+
 pub mod value;
 pub use value::*;
diff --git a/ast/src/values/positive_number.rs b/ast/src/values/positive_number.rs
new file mode 100644
index 0000000000..2ebd822454
--- /dev/null
+++ b/ast/src/values/positive_number.rs
@@ -0,0 +1,25 @@
+use crate::{
+    ast::{span_into_string, Rule},
+    span::SpanDef,
+};
+
+use pest::Span;
+use pest_ast::FromPest;
+use serde::Serialize;
+use std::fmt;
+
+#[derive(Clone, Debug, FromPest, PartialEq, Serialize)]
+#[pest_ast(rule(Rule::positive_number))]
+pub struct PositiveNumber<'ast> {
+    #[pest_ast(outer(with(span_into_string)))]
+    pub value: String,
+    #[pest_ast(outer())]
+    #[serde(with = "SpanDef")]
+    pub span: Span<'ast>,
+}
+
+impl<'ast> fmt::Display for PositiveNumber<'ast> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{}", self.value)
+    }
+}
diff --git a/compiler/tests/address/empty.leo b/compiler/tests/address/empty.leo
new file mode 100644
index 0000000000..e5c06d0def
--- /dev/null
+++ b/compiler/tests/address/empty.leo
@@ -0,0 +1,3 @@
+function main() {
+    let owner = address();
+}
\ No newline at end of file
diff --git a/compiler/tests/address/invalid_length.leo b/compiler/tests/address/invalid_length.leo
new file mode 100644
index 0000000000..c1b7f719be
--- /dev/null
+++ b/compiler/tests/address/invalid_length.leo
@@ -0,0 +1,3 @@
+function main() {
+    let public_key_string = address(aleo1qnr4dkkvkgfqph0vzc3y6z2eu975wnpz2925ntjccd5cfqxtyu8sta57j88);
+}
\ No newline at end of file
diff --git a/compiler/tests/address/invalid.leo b/compiler/tests/address/invalid_prefix.leo
similarity index 100%
rename from compiler/tests/address/invalid.leo
rename to compiler/tests/address/invalid_prefix.leo
diff --git a/compiler/tests/address/mod.rs b/compiler/tests/address/mod.rs
index b905056baa..6c3996c30e 100644
--- a/compiler/tests/address/mod.rs
+++ b/compiler/tests/address/mod.rs
@@ -13,11 +13,27 @@ fn test_valid() {
 }
 
 #[test]
-fn test_invalid() {
-    let bytes = include_bytes!("invalid.leo");
-    let program = parse_program(bytes).unwrap();
+fn test_invalid_prefix() {
+    let bytes = include_bytes!("invalid_prefix.leo");
+    let syntax_error = parse_program(bytes).is_err();
 
-    let _output = expect_compiler_error(program);
+    assert!(syntax_error);
+}
+
+#[test]
+fn test_invalid_length() {
+    let bytes = include_bytes!("invalid_length.leo");
+    let syntax_error = parse_program(bytes).is_err();
+
+    assert!(syntax_error);
+}
+
+#[test]
+fn test_empty() {
+    let bytes = include_bytes!("empty.leo");
+    let syntax_error = parse_program(bytes).is_err();
+
+    assert!(syntax_error);
 }
 
 #[test]
diff --git a/compiler/tests/array/initializer_fail.leo b/compiler/tests/array/initializer_fail.leo
new file mode 100644
index 0000000000..a912075c44
--- /dev/null
+++ b/compiler/tests/array/initializer_fail.leo
@@ -0,0 +1,3 @@
+function main(a: u8[3]) {
+    assert_eq!(a, [1u8; -3]);
+}
\ No newline at end of file
diff --git a/compiler/tests/array/mod.rs b/compiler/tests/array/mod.rs
index 2667402334..63090e2bb7 100644
--- a/compiler/tests/array/mod.rs
+++ b/compiler/tests/array/mod.rs
@@ -42,6 +42,14 @@ fn test_registers() {
 
 // Expressions
 
+#[test]
+fn test_type_fail() {
+    let program_bytes = include_bytes!("type_fail.leo");
+    let syntax_error = parse_program(program_bytes).is_err();
+
+    assert!(syntax_error);
+}
+
 #[test]
 fn test_inline() {
     let program_bytes = include_bytes!("inline.leo");
@@ -68,6 +76,15 @@ fn test_initializer() {
     assert_satisfied(program);
 }
 
+#[test]
+fn test_initializer_fail() {
+    let program_bytes = include_bytes!("initializer_fail.leo");
+    let input_bytes = include_bytes!("input/three_ones.in");
+    let syntax_error = parse_program_with_input(program_bytes, input_bytes).is_err();
+
+    assert!(syntax_error);
+}
+
 #[test]
 fn test_spread() {
     let program_bytes = include_bytes!("spread.leo");
diff --git a/compiler/tests/array/type_fail.leo b/compiler/tests/array/type_fail.leo
new file mode 100644
index 0000000000..c21a6d7b13
--- /dev/null
+++ b/compiler/tests/array/type_fail.leo
@@ -0,0 +1,3 @@
+let main() {
+    let a: u32[-2] = [0u32; 2];
+}
\ No newline at end of file
diff --git a/typed/src/expression.rs b/typed/src/expression.rs
index c13c19bff9..95ad64d0bf 100644
--- a/typed/src/expression.rs
+++ b/typed/src/expression.rs
@@ -13,11 +13,19 @@ use leo_ast::{
         TernaryExpression,
     },
     operations::BinaryOperation,
-    values::{BooleanValue, FieldValue, GroupValue, IntegerValue, NumberImplicitValue, Value},
+    values::{
+        AddressValue,
+        BooleanValue,
+        FieldValue,
+        GroupValue,
+        IntegerValue,
+        NumberImplicitValue,
+        PositiveNumber as LeoPositiveNumber,
+        Value,
+    },
 };
-
-use leo_ast::values::AddressValue;
 use leo_input::values::NumberValue;
+
 use serde::{Deserialize, Serialize};
 use std::fmt;
 
@@ -104,21 +112,16 @@ impl Expression {
 }
 
 impl<'ast> Expression {
-    pub(crate) fn get_count_from_value(count: Value<'ast>) -> usize {
-        match count {
-            Value::Integer(integer) => integer
-                .number
-                .value
-                .parse::<usize>()
-                .expect("Unable to read array size"),
-            Value::Implicit(number) => number.number.value.parse::<usize>().expect("Unable to read array size"),
-            size => unimplemented!("Array size should be an integer {}", size),
-        }
-    }
-
     pub(crate) fn get_count_from_number(number: NumberValue<'ast>) -> usize {
         number.value.parse::<usize>().expect("Unable to read array size")
     }
+
+    pub(crate) fn get_count_from_positive_number(number: LeoPositiveNumber<'ast>) -> usize {
+        number
+            .value
+            .parse::<usize>()
+            .expect("Array size should be a positive number")
+    }
 }
 
 impl<'ast> fmt::Display for Expression {
@@ -403,7 +406,7 @@ impl<'ast> From<ArrayInlineExpression<'ast>> for Expression {
 
 impl<'ast> From<ArrayInitializerExpression<'ast>> for Expression {
     fn from(array: ArrayInitializerExpression<'ast>) -> Self {
-        let count = Expression::get_count_from_value(array.count);
+        let count = Expression::get_count_from_positive_number(array.count);
         let expression = Box::new(SpreadOrExpression::from(*array.expression));
 
         Expression::Array(vec![expression; count], Span::from(array.span))
diff --git a/typed/src/types/type_.rs b/typed/src/types/type_.rs
index b34b1d60a2..87ca87557c 100644
--- a/typed/src/types/type_.rs
+++ b/typed/src/types/type_.rs
@@ -101,7 +101,7 @@ impl<'ast> From<ArrayType<'ast>> for Type {
         let dimensions = array_type
             .dimensions
             .into_iter()
-            .map(|row| Expression::get_count_from_value(row))
+            .map(|row| Expression::get_count_from_positive_number(row))
             .collect();
 
         Type::Array(element_type, dimensions)
