diff --git a/leo/commands/build.rs b/leo/commands/build.rs
index cbf3fa09e1..fcd5ae0bd4 100644
--- a/leo/commands/build.rs
+++ b/leo/commands/build.rs
@@ -182,7 +182,6 @@ impl Command for Build {
         if self.compiler_options.offline {
             args.push("--offline");
         }
-        println!();
         let command = AleoBuild::try_parse_from(&args).map_err(CliError::failed_to_execute_aleo_build)?;
         let result = command.parse().map_err(CliError::failed_to_execute_aleo_build)?;
 
diff --git a/leo/commands/run.rs b/leo/commands/run.rs
index cb2763a13d..de641013c5 100644
--- a/leo/commands/run.rs
+++ b/leo/commands/run.rs
@@ -86,6 +86,7 @@ impl Command for Run {
         if self.compiler_options.offline {
             arguments.push(String::from("--offline"));
         }
+        println!();
         let command = AleoRun::try_parse_from(&arguments).map_err(CliError::failed_to_parse_aleo_run)?;
         let res = command.parse().map_err(CliError::failed_to_execute_aleo_run)?;
 
