diff --git a/compiler/passes/src/static_single_assignment/expression.rs b/compiler/passes/src/static_single_assignment/expression.rs
index 674299d59c..36d580b41f 100644
--- a/compiler/passes/src/static_single_assignment/expression.rs
+++ b/compiler/passes/src/static_single_assignment/expression.rs
@@ -211,9 +211,13 @@ impl ExpressionConsumer for SsaFormingVisitor<'_> {
     /// Note that this shouldn't be used for `Identifier`s on the lhs of definitions or
     /// assignments.
     fn consume_path(&mut self, path: Path) -> Self::Output {
-        // If lookup fails, either it's the name of a mapping or we didn't rename it.
-        let name = *self.rename_table.lookup(path.identifier().name).unwrap_or(&path.identifier().name);
-        (path.with_updated_last_symbol(name).into(), Default::default())
+        if let Some(name) = path.try_local_symbol() {
+            // If lookup fails, then we didn't rename it.
+            let name = *self.rename_table.lookup(name).unwrap_or(&name);
+            (path.with_updated_last_symbol(name).into(), Default::default())
+        } else {
+            (path.into(), Default::default())
+        }
     }
 
     /// Consumes and returns the literal without making any modifications.
diff --git a/compiler/passes/src/type_checking/visitor.rs b/compiler/passes/src/type_checking/visitor.rs
index d71bbba98f..72e263417e 100644
--- a/compiler/passes/src/type_checking/visitor.rs
+++ b/compiler/passes/src/type_checking/visitor.rs
@@ -309,62 +309,103 @@ impl TypeCheckingVisitor<'_> {
             }
 
             Intrinsic::MappingGet => {
-                let [container, key_or_index] = arguments else { panic!("number of arguments is already checked") };
+                let [container, key] = arguments else { panic!("number of arguments is already checked") };
 
                 let container_ty = self.visit_expression(container, &None);
 
-                // Key type depends on container type
-                let key_or_index_ty = if let Type::Mapping(MappingType { ref key, .. }) = container_ty {
-                    self.visit_expression(key_or_index, &Some(*key.clone()))
+                // Key type is obtained from the mapping type. Otherwise, don't do anything. Error
+                // emission is deferred.
+                let key_ty = if let Type::Mapping(MappingType { key: ref key_ty, .. }) = container_ty {
+                    self.visit_expression(key, &Some(*key_ty.clone()))
                 } else {
-                    self.visit_expression(key_or_index, &None)
+                    self.visit_expression(key, &None)
                 };
 
-                vec![(container_ty, container), (key_or_index_ty, key_or_index)]
+                vec![(container_ty, container), (key_ty, key)]
             }
 
-            Intrinsic::VectorGet => {
-                let [container, key_or_index] = arguments else { panic!("number of arguments is already checked") };
+            Intrinsic::MappingSet => {
+                let [container, key, val] = arguments else { panic!("number of arguments is already checked") };
 
                 let container_ty = self.visit_expression(container, &None);
 
-                // Key type depends on container type
-                let key_or_index_ty = self.visit_expression_infer_default_u32(key_or_index);
+                let (key_ty, val_ty) =
+                    if let Type::Mapping(MappingType { key: ref key_ty, value: ref value_ty, .. }) = container_ty {
+                        (
+                            self.visit_expression(key, &Some(*key_ty.clone())),
+                            self.visit_expression(val, &Some(*value_ty.clone())),
+                        )
+                    } else {
+                        (self.visit_expression(key, &None), self.visit_expression(val, &None))
+                    };
 
-                vec![(container_ty, container), (key_or_index_ty, key_or_index)]
+                vec![(container_ty, container), (key_ty, key), (val_ty, val)]
             }
 
-            Intrinsic::MappingSet => {
-                let [container, key_or_index, val] = arguments else {
-                    panic!("number of arguments is already checked")
-                };
+            Intrinsic::MappingGetOrUse => {
+                let [container, key, default] = arguments else { panic!("number of arguments is already checked") };
 
                 let container_ty = self.visit_expression(container, &None);
 
-                let key_or_index_ty = match container_ty {
-                    Type::Vector(_) => self.visit_expression_infer_default_u32(key_or_index),
-                    Type::Mapping(MappingType { ref key, .. }) => {
-                        self.visit_expression(key_or_index, &Some(*key.clone()))
-                    }
-                    _ => self.visit_expression(key_or_index, &None),
+                let (key_ty, default_ty) =
+                    if let Type::Mapping(MappingType { key: ref key_ty, value: ref value_ty, .. }) = container_ty {
+                        (
+                            self.visit_expression(key, &Some(*key_ty.clone())),
+                            self.visit_expression(default, &Some(*value_ty.clone())),
+                        )
+                    } else {
+                        (self.visit_expression(key, &None), self.visit_expression(default, &None))
+                    };
+
+                vec![(container_ty, container), (key_ty, key), (default_ty, default)]
+            }
+
+            Intrinsic::MappingRemove => {
+                let [container, key] = arguments else { panic!("number of arguments is already checked") };
+
+                let container_ty = self.visit_expression(container, &None);
+
+                let key_ty = if let Type::Mapping(MappingType { key: ref key_ty, .. }) = container_ty {
+                    self.visit_expression(key, &Some(*key_ty.clone()))
+                } else {
+                    self.visit_expression(key, &None)
                 };
 
-                let val_ty = if let Type::Mapping(MappingType { ref value, .. }) = container_ty {
-                    self.visit_expression(val, &Some(*value.clone()))
+                vec![(container_ty, container), (key_ty, key)]
+            }
+
+            Intrinsic::MappingContains => {
+                let [container, key] = arguments else { panic!("number of arguments is already checked") };
+
+                let container_ty = self.visit_expression(container, &None);
+
+                let key_ty = if let Type::Mapping(MappingType { key: ref key_ty, .. }) = container_ty {
+                    self.visit_expression(key, &Some(*key_ty.clone()))
                 } else {
-                    self.visit_expression(val, &None)
+                    self.visit_expression(key, &None)
                 };
 
-                vec![(container_ty, container), (key_or_index_ty, key_or_index), (val_ty, val)]
+                vec![(container_ty, container), (key_ty, key)]
+            }
+
+            Intrinsic::VectorGet => {
+                let [container, index] = arguments else { panic!("number of arguments is already checked") };
+
+                let container_ty = self.visit_expression(container, &None);
+
+                // Indices default to `u32` when numeric.
+                let index_ty = self.visit_expression_infer_default_u32(index);
+
+                vec![(container_ty, container), (index_ty, index)]
             }
+
             Intrinsic::VectorSet => {
-                let [container, key_or_index, val] = arguments else {
-                    panic!("number of arguments is already checked")
-                };
+                let [container, index, val] = arguments else { panic!("number of arguments is already checked") };
 
                 let container_ty = self.visit_expression(container, &None);
 
-                let key_or_index_ty = self.visit_expression_infer_default_u32(key_or_index);
+                // Indices default to `u32` when numeric.
+                let index_ty = self.visit_expression_infer_default_u32(index);
 
                 let val_ty = if let Type::Vector(VectorType { ref element_type }) = container_ty {
                     self.visit_expression(val, &Some(*element_type.clone()))
@@ -372,7 +413,7 @@ impl TypeCheckingVisitor<'_> {
                     self.visit_expression(val, &None)
                 };
 
-                vec![(container_ty, container), (key_or_index_ty, key_or_index), (val_ty, val)]
+                vec![(container_ty, container), (index_ty, index), (val_ty, val)]
             }
 
             Intrinsic::VectorPush => {
@@ -382,7 +423,7 @@ impl TypeCheckingVisitor<'_> {
                 let vec_ty = self.visit_expression(vec, &None);
 
                 // Type-check value against vector element type
-                let val_ty = if let Type::Vector(VectorType { element_type }) = &vec_ty {
+                let val_ty = if let Type::Vector(VectorType { ref element_type }) = vec_ty {
                     self.visit_expression(val, &Some(*element_type.clone()))
                 } else {
                     self.visit_expression(val, &None)
@@ -694,7 +735,7 @@ impl TypeCheckingVisitor<'_> {
                 let (map_ty, map_expr) = &arguments[0];
 
                 let Type::Mapping(MappingType { value, .. }) = map_ty else {
-                    self.assert_vector_or_mapping_type(map_ty, map_expr.span());
+                    self.assert_mapping_type(map_ty, map_expr.span());
                     return Type::Err;
                 };
 
@@ -707,7 +748,7 @@ impl TypeCheckingVisitor<'_> {
                 let (map_ty, map_expr) = &arguments[0];
 
                 let Type::Mapping(_) = map_ty else {
-                    self.assert_vector_or_mapping_type(map_ty, map_expr.span());
+                    self.assert_mapping_type(map_ty, map_expr.span());
                     return Type::Err;
                 };
 
@@ -735,17 +776,11 @@ impl TypeCheckingVisitor<'_> {
                 // Check that the first argument is a mapping.
                 self.assert_mapping_type(map_ty, map_expr.span());
 
-                let Type::Mapping(MappingType { key, value, .. }) = map_ty else {
+                let Type::Mapping(MappingType { value, .. }) = map_ty else {
                     // We already handled the error in the assertion.
                     return Type::Err;
                 };
 
-                // Check that the second argument matches the key type of the mapping.
-                self.assert_type(&arguments[1].0, key, arguments[1].1.span());
-
-                // Check that the third argument matches the value type of the mapping.
-                self.assert_type(&arguments[2].0, value, arguments[2].1.span());
-
                 value.deref().clone()
             }
             Intrinsic::MappingRemove => {
@@ -757,7 +792,7 @@ impl TypeCheckingVisitor<'_> {
                 // Check that the first argument is a mapping.
                 self.assert_mapping_type(map_ty, map_expr.span());
 
-                let Type::Mapping(MappingType { key, .. }) = map_ty else {
+                let Type::Mapping(_) = map_ty else {
                     // We already handled the error in the assertion.
                     return Type::Err;
                 };
@@ -772,9 +807,6 @@ impl TypeCheckingVisitor<'_> {
                     return Type::Err;
                 }
 
-                // Check that the second argument matches the key type of the mapping.
-                self.assert_type(&arguments[1].0, key, arguments[1].1.span());
-
                 Type::Unit
             }
             Intrinsic::MappingContains => {
@@ -786,14 +818,11 @@ impl TypeCheckingVisitor<'_> {
                 // Check that the first argument is a mapping.
                 self.assert_mapping_type(map_ty, map_expr.span());
 
-                let Type::Mapping(MappingType { key, .. }) = map_ty else {
+                let Type::Mapping(_) = map_ty else {
                     // We already handled the error in the assertion.
                     return Type::Err;
                 };
 
-                // Check that the second argument matches the key type of the mapping.
-                self.assert_type(&arguments[1].0, key, arguments[1].1.span());
-
                 Type::Boolean
             }
             Intrinsic::OptionalUnwrap => {
diff --git a/tests/expectations/compiler/statements/b29134.out b/tests/expectations/compiler/statements/b29134.out
new file mode 100644
index 0000000000..f801f42f44
--- /dev/null
+++ b/tests/expectations/compiler/statements/b29134.out
@@ -0,0 +1,46 @@
+program child.aleo;
+
+mapping foo:
+    key as u32.public;
+    value as u32.public;
+
+function bar:
+// --- Next Program --- //
+import child.aleo;
+program parent.aleo;
+
+mapping f:
+    key as u32.public;
+    value as u32.public;
+
+function set_test:
+    async set_test into r0;
+    output r0 as parent.aleo/set_test.future;
+
+finalize set_test:
+    set 0u32 into f[0u32];
+    set 0u32 into f[1u32];
+
+function get_test:
+    async get_test into r0;
+    output r0 as parent.aleo/get_test.future;
+
+finalize get_test:
+    get child.aleo/foo[0u32] into r0;
+    is.eq r0 42u32 into r1;
+    assert.eq r1 true;
+    get child.aleo/foo[1u32] into r2;
+    is.eq r2 42u32 into r3;
+    assert.eq r3 true;
+
+function get_or_use_test:
+    async get_or_use_test into r0;
+    output r0 as parent.aleo/get_or_use_test.future;
+
+finalize get_or_use_test:
+    get.or_use child.aleo/foo[0u32] 42u32 into r0;
+    is.eq r0 42u32 into r1;
+    assert.eq r1 true;
+    get.or_use child.aleo/foo[1u32] 42u32 into r2;
+    is.eq r2 42u32 into r3;
+    assert.eq r3 true;
diff --git a/tests/tests/compiler/statements/b29134.leo b/tests/tests/compiler/statements/b29134.leo
new file mode 100644
index 0000000000..5d47da3e5f
--- /dev/null
+++ b/tests/tests/compiler/statements/b29134.leo
@@ -0,0 +1,39 @@
+program child.aleo {
+    mapping foo: u32 => u32; 
+
+    fn bar() { }
+}
+
+// --- Next Program --- //
+
+import child.aleo;
+program parent.aleo {
+    mapping f: u32 => u32;
+
+    fn set_test() -> Final {
+        return final {
+            for i in 0u32..2u32 {
+                f.set(i, 0);
+            }
+        };
+    }
+
+    fn get_test() -> Final {
+        return final {
+            for i in 0u32..2u32 {
+                let foo = child.aleo/foo.get(i);
+                assert(foo == 42);
+            }
+        };
+    }
+
+    fn get_or_use_test() -> Final {
+        return final {
+            for i in 0u32..2u32 {
+                let foo = child.aleo/foo.get_or_use(i, 42);
+                assert(foo == 42);
+            }
+        };
+    }
+}
+
