diff --git a/Cargo.lock b/Cargo.lock
index 3fa513cb86..336d80bd3d 100644
--- a/Cargo.lock
+++ b/Cargo.lock
@@ -149,6 +149,33 @@ version = "1.3.4"
 source = "registry+https://github.com/rust-lang/crates.io-index"
 checksum = "08c48aae112d48ed9f069b33538ea9e3e90aa263cfa3d1c24309612b1f7472de"
 
+[[package]]
+name = "bzip2"
+version = "0.3.3"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "42b7c3cbf0fa9c1b82308d57191728ca0256cb821220f4e2fd410a72ade26e3b"
+dependencies = [
+ "bzip2-sys",
+ "libc",
+]
+
+[[package]]
+name = "bzip2-sys"
+version = "0.1.9+1.0.8"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "ad3b39a260062fca31f7b0b12f207e8f2590a67d32ec7d59c20484b07ea7285e"
+dependencies = [
+ "cc",
+ "libc",
+ "pkg-config",
+]
+
+[[package]]
+name = "cc"
+version = "1.0.55"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "b1be3409f94d7bdceeb5f5fac551039d9b3f00e25da7a74fc4d33400a0d96368"
+
 [[package]]
 name = "cfg-if"
 version = "0.1.10"
@@ -190,6 +217,15 @@ dependencies = [
  "winapi",
 ]
 
+[[package]]
+name = "crc32fast"
+version = "1.2.0"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "ba125de2af0df55319f41944744ad91c71113bf74a4646efff39afe1f6842db1"
+dependencies = [
+ "cfg-if",
+]
+
 [[package]]
 name = "crossbeam-deque"
 version = "0.7.3"
@@ -334,6 +370,18 @@ version = "0.1.2"
 source = "registry+https://github.com/rust-lang/crates.io-index"
 checksum = "e88a8acf291dafb59c2d96e8f59828f3838bb1a70398823ade51a84de6a6deed"
 
+[[package]]
+name = "flate2"
+version = "1.0.14"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "2cfff41391129e0a856d6d822600b8d71179d46879e310417eb9c762eb178b42"
+dependencies = [
+ "cfg-if",
+ "crc32fast",
+ "libc",
+ "miniz_oxide",
+]
+
 [[package]]
 name = "from-pest"
 version = "0.3.1"
@@ -492,6 +540,8 @@ dependencies = [
  "snarkos-utilities",
  "thiserror",
  "toml",
+ "walkdir",
+ "zip",
 ]
 
 [[package]]
@@ -511,6 +561,7 @@ dependencies = [
 name = "leo-compiler"
 version = "0.1.0"
 dependencies = [
+ "bincode",
  "hex",
  "leo-ast",
  "leo-inputs",
@@ -550,6 +601,7 @@ dependencies = [
  "leo-ast",
  "leo-inputs",
  "pest",
+ "serde",
  "snarkos-errors",
  "snarkos-models",
  "thiserror",
@@ -690,6 +742,18 @@ dependencies = [
  "sha-1",
 ]
 
+[[package]]
+name = "pkg-config"
+version = "0.3.17"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "05da548ad6865900e60eaba7f589cc0783590a92e940c26953ff81ddbab2d677"
+
+[[package]]
+name = "podio"
+version = "0.1.7"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "b18befed8bc2b61abc79a457295e7e838417326da1586050b919414073977f19"
+
 [[package]]
 name = "ppv-lite86"
 version = "0.2.8"
@@ -855,6 +919,15 @@ version = "1.0.5"
 source = "registry+https://github.com/rust-lang/crates.io-index"
 checksum = "71d301d4193d031abdd79ff7e3dd721168a9572ef3fe51a1517aba235bd8f86e"
 
+[[package]]
+name = "same-file"
+version = "1.0.6"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "93fc1dc3aaa9bfed95e02e6eadabb4baf7e3078b0bd1b4d7b6b0b68378900502"
+dependencies = [
+ "winapi-util",
+]
+
 [[package]]
 name = "scopeguard"
 version = "1.1.0"
@@ -1137,6 +1210,16 @@ dependencies = [
  "lazy_static",
 ]
 
+[[package]]
+name = "time"
+version = "0.1.43"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "ca8a50ef2360fbd1eeb0ecd46795a87a19024eb4b53c5dc916ca1fd95fe62438"
+dependencies = [
+ "libc",
+ "winapi",
+]
+
 [[package]]
 name = "toml"
 version = "0.5.6"
@@ -1194,6 +1277,17 @@ version = "1.0.2"
 source = "registry+https://github.com/rust-lang/crates.io-index"
 checksum = "6a02e4885ed3bc0f2de90ea6dd45ebcbb66dacffe03547fadbb0eeae2770887d"
 
+[[package]]
+name = "walkdir"
+version = "2.3.1"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "777182bc735b6424e1a57516d35ed72cb8019d85c8c9bf536dccb3445c1a2f7d"
+dependencies = [
+ "same-file",
+ "winapi",
+ "winapi-util",
+]
+
 [[package]]
 name = "wasi"
 version = "0.9.0+wasi-snapshot-preview1"
@@ -1230,3 +1324,16 @@ name = "winapi-x86_64-pc-windows-gnu"
 version = "0.4.0"
 source = "registry+https://github.com/rust-lang/crates.io-index"
 checksum = "712e227841d057c1ee1cd2fb22fa7e5a5461ae8e48fa2ca79ec42cfc1931183f"
+
+[[package]]
+name = "zip"
+version = "0.5.6"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "58287c28d78507f5f91f2a4cf1e8310e2c76fd4c6932f93ac60fd1ceb402db7d"
+dependencies = [
+ "bzip2",
+ "crc32fast",
+ "flate2",
+ "podio",
+ "time",
+]
diff --git a/Cargo.toml b/Cargo.toml
index 17cc6e4e69..c41f2fb8aa 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -37,6 +37,8 @@ serde = { version = "1.0", features = ["derive"] }
 serde_json = { version = "1.0" }
 toml = { version = "0.5" }
 thiserror = { version = "1.0" }
+walkdir = { version = "2" }
+zip = { version = "0.5" }
 
 [dev-dependencies]
 rusty-hook = { version = "0.11.1" }
diff --git a/ast/src/common/range.rs b/ast/src/common/range.rs
index d872f78d1e..0c272ef26d 100644
--- a/ast/src/common/range.rs
+++ b/ast/src/common/range.rs
@@ -6,16 +6,8 @@ use pest_ast::FromPest;
 #[derive(Clone, Debug, FromPest, PartialEq)]
 #[pest_ast(rule(Rule::range))]
 pub struct Range<'ast> {
-    pub from: Option<FromExpression<'ast>>,
-    pub to: Option<ToExpression<'ast>>,
+    pub from: Option<Expression<'ast>>,
+    pub to: Option<Expression<'ast>>,
     #[pest_ast(outer())]
     pub span: Span<'ast>,
 }
-
-#[derive(Clone, Debug, FromPest, PartialEq)]
-#[pest_ast(rule(Rule::from_expression))]
-pub struct FromExpression<'ast>(pub Expression<'ast>);
-
-#[derive(Clone, Debug, FromPest, PartialEq)]
-#[pest_ast(rule(Rule::to_expression))]
-pub struct ToExpression<'ast>(pub Expression<'ast>);
diff --git a/ast/src/common/range_or_expression.rs b/ast/src/common/range_or_expression.rs
index 03f773af76..f41dc4a776 100644
--- a/ast/src/common/range_or_expression.rs
+++ b/ast/src/common/range_or_expression.rs
@@ -17,8 +17,8 @@ impl<'ast> fmt::Display for RangeOrExpression<'ast> {
             RangeOrExpression::Range(ref range) => write!(
                 f,
                 "{}..{}",
-                range.from.as_ref().map(|e| e.0.to_string()).unwrap_or("".to_string()),
-                range.to.as_ref().map(|e| e.0.to_string()).unwrap_or("".to_string())
+                range.from.as_ref().map(|v| format!("{}", v)).unwrap_or(format!("")),
+                range.to.as_ref().map(|v| format!("{}", v)).unwrap_or(format!("")),
             ),
         }
     }
diff --git a/ast/src/imports/import.rs b/ast/src/imports/import.rs
index 269b04b005..dbecc1b8fc 100644
--- a/ast/src/imports/import.rs
+++ b/ast/src/imports/import.rs
@@ -1,8 +1,4 @@
-use crate::{
-    ast::Rule,
-    common::LineEnd,
-    imports::{ImportSource, ImportSymbol},
-};
+use crate::{ast::Rule, common::LineEnd, imports::Package};
 
 use pest::Span;
 use pest_ast::FromPest;
@@ -10,8 +6,7 @@ use pest_ast::FromPest;
 #[derive(Clone, Debug, FromPest, PartialEq)]
 #[pest_ast(rule(Rule::import))]
 pub struct Import<'ast> {
-    pub source: ImportSource<'ast>,
-    pub symbols: Vec<ImportSymbol<'ast>>,
+    pub package: Package<'ast>,
     pub line_end: LineEnd,
     #[pest_ast(outer())]
     pub span: Span<'ast>,
diff --git a/ast/src/imports/import_source.rs b/ast/src/imports/import_source.rs
deleted file mode 100644
index 221fb1022c..0000000000
--- a/ast/src/imports/import_source.rs
+++ /dev/null
@@ -1,13 +0,0 @@
-use crate::ast::{span_into_string, Rule};
-
-use pest::Span;
-use pest_ast::FromPest;
-
-#[derive(Clone, Debug, FromPest, PartialEq)]
-#[pest_ast(rule(Rule::import_source))]
-pub struct ImportSource<'ast> {
-    #[pest_ast(outer(with(span_into_string)))]
-    pub value: String,
-    #[pest_ast(outer())]
-    pub span: Span<'ast>,
-}
diff --git a/ast/src/imports/mod.rs b/ast/src/imports/mod.rs
index 20a794eeda..c2b4b01765 100644
--- a/ast/src/imports/mod.rs
+++ b/ast/src/imports/mod.rs
@@ -1,8 +1,14 @@
 pub mod import;
 pub use import::*;
 
-pub mod import_source;
-pub use import_source::*;
-
 pub mod import_symbol;
 pub use import_symbol::*;
+
+pub mod package;
+pub use package::*;
+
+pub mod package_access;
+pub use package_access::*;
+
+pub mod star;
+pub use star::*;
diff --git a/ast/src/imports/package.rs b/ast/src/imports/package.rs
new file mode 100644
index 0000000000..42ac528961
--- /dev/null
+++ b/ast/src/imports/package.rs
@@ -0,0 +1,13 @@
+use crate::{ast::Rule, common::Identifier, imports::PackageAccess};
+
+use pest::Span;
+use pest_ast::FromPest;
+
+#[derive(Clone, Debug, FromPest, PartialEq)]
+#[pest_ast(rule(Rule::package))]
+pub struct Package<'ast> {
+    pub name: Identifier<'ast>,
+    pub access: PackageAccess<'ast>,
+    #[pest_ast(outer())]
+    pub span: Span<'ast>,
+}
diff --git a/ast/src/imports/package_access.rs b/ast/src/imports/package_access.rs
new file mode 100644
index 0000000000..0574dc94a4
--- /dev/null
+++ b/ast/src/imports/package_access.rs
@@ -0,0 +1,15 @@
+use crate::{
+    ast::Rule,
+    imports::{ImportSymbol, Package, Star},
+};
+
+use pest_ast::FromPest;
+
+#[derive(Clone, Debug, FromPest, PartialEq)]
+#[pest_ast(rule(Rule::package_access))]
+pub enum PackageAccess<'ast> {
+    Star(Star<'ast>),
+    SubPackage(Box<Package<'ast>>),
+    Symbol(ImportSymbol<'ast>),
+    Multiple(Vec<PackageAccess<'ast>>),
+}
diff --git a/ast/src/imports/star.rs b/ast/src/imports/star.rs
new file mode 100644
index 0000000000..4f4882801e
--- /dev/null
+++ b/ast/src/imports/star.rs
@@ -0,0 +1,11 @@
+use crate::ast::Rule;
+
+use pest::Span;
+use pest_ast::FromPest;
+
+#[derive(Clone, Debug, FromPest, PartialEq)]
+#[pest_ast(rule(Rule::star))]
+pub struct Star<'ast> {
+    #[pest_ast(outer())]
+    pub span: Span<'ast>,
+}
diff --git a/ast/src/leo.pest b/ast/src/leo.pest
index 40886ba32a..18f5ce31ca 100644
--- a/ast/src/leo.pest
+++ b/ast/src/leo.pest
@@ -18,12 +18,10 @@ protected_name = {
 LINE_END = { ";" ~ NEWLINE* }
 
 // Declared in common/mutable.rs
-mutable = { "mut" }
+mutable = { "mut " }
 
 // Declared in common/range.rs
-range = { from_expression? ~ ".." ~ to_expression? }
-from_expression = { expression }
-to_expression = { expression }
+range = { expression? ~ ".." ~ expression? }
 
 // Declared in common/range_or_expression.rs
 range_or_expression = { range | expression }
@@ -35,15 +33,15 @@ spread = { "..." ~ expression }
 spread_or_expression = { spread | expression }
 
 // Declared in common/static_.rs
-static_ = { "static" }
+static_ = { "static " }
 
 // Declared in common/variable.rs
 variable = { mutable? ~ identifier ~ (":" ~ type_)? }
 
 // Declared in common/declare.rs
 declare = { let_ | const_ }
-const_ = { "const" }
-let_ = { "let" }
+const_ = { "const " }
+let_ = { "let " }
 
 /// Operations
 
@@ -177,23 +175,23 @@ access_static_member = { "::" ~ identifier }
 /// Circuits
 
 // Declared in circuits/circuit_definition.rs
-circuit = { "circuit" ~ identifier ~ "{" ~ NEWLINE* ~ circuit_member* ~ NEWLINE* ~ "}" ~ NEWLINE* }
+circuit = { "circuit " ~ identifier ~ "{" ~ NEWLINE* ~ circuit_member* ~ NEWLINE* ~ "}" ~ NEWLINE* }
 
 // Declared in circuits/circuit_field.rs
 circuit_field = { identifier ~ ":" ~ expression }
 
 // Declared in circuits/circuit_field_definition.rs
-circuit_field_definition = { identifier ~ ":" ~ type_ ~ NEWLINE* }
+circuit_field_definition = { identifier ~ ":" ~ type_ ~ ","?}
 
 // Declared in circuits/circuit_function.rs
 circuit_function = { static_? ~ function_definition }
 
 // Declared in circuits/circuit_member.rs
-circuit_member = { circuit_function | circuit_field_definition }
+circuit_member = { circuit_function | circuit_field_definition ~ NEWLINE*}
 
 /// Conditionals
 
-expression_conditional = { "if" ~ expression ~ "?" ~ expression ~ ":" ~ expression}
+expression_conditional = { "if " ~ expression ~ "? " ~ expression ~ ": " ~ expression}
 
 /// Expressions
 
@@ -254,7 +252,7 @@ assert_eq = {"assert_eq!" ~ "(" ~ NEWLINE* ~ expression ~ "," ~ NEWLINE* ~ expre
 statement_assign = { assignee ~ operation_assign ~ expression ~ LINE_END }
 
 // Declared in statements/conditional_statement.rs
-statement_conditional = {"if" ~ (expression | "(" ~ expression ~ ")") ~ "{" ~ NEWLINE* ~ statement+ ~ "}" ~ ("else" ~ conditional_nested_or_end_statement)?}
+statement_conditional = {"if " ~ (expression | "(" ~ expression ~ ")") ~ "{" ~ NEWLINE* ~ statement+ ~ "}" ~ ("else" ~ conditional_nested_or_end_statement)?}
 conditional_nested_or_end_statement = { statement_conditional | "{" ~ NEWLINE* ~ statement+ ~ "}"}
 
 // Declared in statements/definition_statement.rs
@@ -264,40 +262,52 @@ statement_definition = { declare ~ variable ~ "=" ~ expression ~ LINE_END}
 statement_expression = { expression ~ LINE_END }
 
 // Declared in statements/for_statement.rs
-statement_for = { "for" ~ identifier ~ "in" ~ expression ~ ".." ~ expression ~ "{" ~ NEWLINE* ~ statement+ ~ "}"}
+statement_for = { "for " ~ identifier ~ "in " ~ expression ~ ".." ~ expression ~ "{" ~ NEWLINE* ~ statement+ ~ "}"}
 
 // Declared in statements/multiple_assignment_statement.rs
 statement_multiple_assignment = { declare ~ "(" ~ variable_tuple ~ ")" ~ "=" ~  identifier ~ "(" ~ expression_tuple ~ ")" ~ LINE_END}
 variable_tuple = _{ variable ~ ("," ~ variable)* }
 
 // Declared in statements/return_statement.rs
-statement_return = { "return" ~ expression_tuple }
+statement_return = { "return " ~ expression_tuple }
 
 /// Functions
 
 // Declared in functions/function.rs
-function_definition = { "function" ~ identifier ~ "(" ~ input_model_list ~ ")" ~ ("->" ~ (type_ | "(" ~ type_list ~ ")"))? ~ "{" ~ NEWLINE* ~ statement* ~ NEWLINE* ~ "}" ~ NEWLINE* }
+function_definition = { "function " ~ identifier ~ "(" ~ input_model_list ~ ")" ~ ("->" ~ (type_ | "(" ~ type_list ~ ")"))? ~ "{" ~ NEWLINE* ~ statement* ~ NEWLINE* ~ "}" ~ NEWLINE* }
 
 // Declared in functions/function_input.rs
 function_input = { mutable? ~ identifier ~ ":" ~ type_ }
 input_model_list = _{ (function_input ~ ("," ~ function_input)*)? }
 
 // Declared in functions/test_function.rs
-test_function = { "test" ~ function_definition }
+test_function = { "test " ~ function_definition }
 
 /// Imports
 
 // Declared in imports/import.rs
-import = { "from" ~ "\"" ~ import_source ~ "\"" ~ "import" ~ ("*" | ("{" ~ NEWLINE* ~ import_symbol_tuple ~ NEWLINE* ~ "}") | import_symbol) ~ LINE_END}
+import = { "import" ~ package ~ LINE_END}
 
-// Declared in imports/import_source.rs
-import_source = @{ (!"\"" ~ ANY)* }
+// Declared in imports/package.rs
+package = { identifier ~ "." ~ package_access }
+
+// Declared in imports/package_access
+package_access = {
+    star
+    | package // subpackage
+    | multiple_package_access
+    | import_symbol
+}
+
+multiple_package_access = _{ "(" ~ NEWLINE* ~ package_access ~ ("," ~ NEWLINE* ~ package_access)* ~ NEWLINE* ~ ")"}
+
+// Declared in imports/star.rs
+star = {"*"}
 
 // Declared in imports/import_symbol.rs
 import_symbol = { identifier ~ ("as" ~ identifier)? }
-import_symbol_tuple = _{ import_symbol ~ ("," ~ NEWLINE* ~ import_symbol)* }
 
 /// Utilities
 
 COMMENT = _{ ("/*" ~ (!"*/" ~ ANY)* ~ "*/") | ("//" ~ (!NEWLINE ~ ANY)*) }
-WHITESPACE = _{ " " | "\t" ~ (NEWLINE)* }
+WHITESPACE = _{ " " | "\t" ~ (NEWLINE)* } // pest implicit whitespace keyword
diff --git a/compiler/Cargo.toml b/compiler/Cargo.toml
index cbf4c1d846..f1ce29aab0 100644
--- a/compiler/Cargo.toml
+++ b/compiler/Cargo.toml
@@ -14,6 +14,7 @@ snarkos-errors = { git = "ssh://git@github.com/AleoHQ/snarkOS.git", rev = "c7a56
 snarkos-gadgets = { git = "ssh://git@github.com/AleoHQ/snarkOS.git", rev = "c7a56d9", default-features = false }
 snarkos-models =  { git = "ssh://git@github.com/AleoHQ/snarkOS.git", rev = "c7a56d9", default-features = false }
 
+bincode = { version = "1.0" }
 hex = { version = "0.4.2" }
 log = { version = "0.4" }
 pest = { version = "2.0" }
diff --git a/compiler/src/compiler.rs b/compiler/src/compiler.rs
index 4b5ea8be63..8391446535 100644
--- a/compiler/src/compiler.rs
+++ b/compiler/src/compiler.rs
@@ -40,15 +40,15 @@ impl<F: Field + PrimeField, G: GroupType<F>> Compiler<F, G> {
         }
     }
 
-    pub fn init(package_name: String, main_file_path: PathBuf) -> Result<Self, CompilerError> {
-        let mut program = Self::new(package_name);
-        program.set_path(main_file_path);
+    pub fn new_from_path(package_name: String, main_file_path: PathBuf) -> Result<Self, CompilerError> {
+        let mut compiler = Self::new(package_name);
+        compiler.set_path(main_file_path);
 
         // Generate the abstract syntax tree and assemble the program
-        let program_string = program.load_program()?;
-        program.parse_program(&program_string)?;
+        let program_string = compiler.load_program()?;
+        compiler.parse_program(&program_string)?;
 
-        Ok(program)
+        Ok(compiler)
     }
 
     pub fn set_path(&mut self, main_file_path: PathBuf) {
@@ -116,6 +116,26 @@ impl<F: Field + PrimeField, G: GroupType<F>> Compiler<F, G> {
 
         Ok(())
     }
+
+    pub fn to_bytes(&self) -> Result<Vec<u8>, CompilerError> {
+        Ok(bincode::serialize(&self.program)?)
+    }
+
+    pub fn from_bytes(bytes: &[u8]) -> Result<Self, CompilerError> {
+        let program: Program = bincode::deserialize(bytes)?;
+        let mut program_inputs = Inputs::new();
+
+        program_inputs.set_inputs_size(program.expected_inputs.len());
+
+        Ok(Self {
+            package_name: program.name.clone(),
+            main_file_path: PathBuf::new(),
+            program,
+            program_inputs,
+            output: None,
+            _engine: PhantomData,
+        })
+    }
 }
 
 impl<F: Field + PrimeField, G: GroupType<F>> ConstraintSynthesizer<F> for Compiler<F, G> {
diff --git a/compiler/src/constraints/expression.rs b/compiler/src/constraints/expression.rs
index 8a1ee60d0b..c41656dbb5 100644
--- a/compiler/src/constraints/expression.rs
+++ b/compiler/src/constraints/expression.rs
@@ -6,6 +6,7 @@ use crate::{
     enforce_or,
     errors::ExpressionError,
     evaluate_not,
+    new_bool_constant,
     new_scope,
     FieldType,
     GroupType,
@@ -573,11 +574,15 @@ impl<F: Field + PrimeField, G: GroupType<F>> ConstrainedProgram<F, G> {
         match index {
             RangeOrExpression::Range(from, to) => {
                 let from_resolved = match from {
-                    Some(from_index) => from_index.to_usize(span.clone())?,
+                    Some(from_index) => {
+                        self.enforce_index(cs, file_scope.clone(), function_scope.clone(), from_index, span.clone())?
+                    }
                     None => 0usize, // Array slice starts at index 0
                 };
                 let to_resolved = match to {
-                    Some(to_index) => to_index.to_usize(span.clone())?,
+                    Some(to_index) => {
+                        self.enforce_index(cs, file_scope.clone(), function_scope.clone(), to_index, span.clone())?
+                    }
                     None => array.len(), // Array slice ends at array length
                 };
                 Ok(ConstrainedValue::Array(array[from_resolved..to_resolved].to_owned()))
@@ -899,10 +904,12 @@ impl<F: Field + PrimeField, G: GroupType<F>> ConstrainedProgram<F, G> {
             }
 
             // Values
-            Expression::Integer(integer) => Ok(ConstrainedValue::Integer(integer)),
+            Expression::Integer(type_, integer, span) => {
+                Ok(ConstrainedValue::Integer(Integer::new_constant(&type_, integer, span)?))
+            }
             Expression::Field(field, span) => Ok(ConstrainedValue::Field(FieldType::constant(field, span)?)),
             Expression::Group(group_affine, span) => Ok(ConstrainedValue::Group(G::constant(group_affine, span)?)),
-            Expression::Boolean(bool) => Ok(ConstrainedValue::Boolean(bool)),
+            Expression::Boolean(boolean, span) => Ok(ConstrainedValue::Boolean(new_bool_constant(boolean, span)?)),
             Expression::Implicit(value, span) => Self::enforce_number_implicit(expected_types, value, span),
 
             // Binary operations
diff --git a/compiler/src/constraints/import.rs b/compiler/src/constraints/import.rs
deleted file mode 100644
index 0c351a64d5..0000000000
--- a/compiler/src/constraints/import.rs
+++ /dev/null
@@ -1,91 +0,0 @@
-use crate::{
-    constraints::{ConstrainedProgram, ConstrainedValue},
-    errors::constraints::ImportError,
-    new_scope,
-    GroupType,
-};
-use leo_ast::LeoParser;
-use leo_types::{Import, Program};
-
-use snarkos_models::curves::{Field, PrimeField};
-use std::env::current_dir;
-
-impl<F: Field + PrimeField, G: GroupType<F>> ConstrainedProgram<F, G> {
-    pub fn enforce_import(&mut self, scope: String, import: Import) -> Result<(), ImportError> {
-        let path = current_dir().map_err(|error| ImportError::directory_error(error, import.span.clone()))?;
-
-        // Sanitize the package path to the imports directory
-        let mut package_path = path.clone();
-        if package_path.is_file() {
-            package_path.pop();
-        }
-
-        // Construct the path to the import file in the import directory
-        let mut main_file_path = package_path.clone();
-        main_file_path.push(import.path_string_full());
-
-        println!("Compiling import - {:?}", main_file_path);
-
-        // Build the abstract syntax tree
-        let file_path = &main_file_path;
-        let input_file = &LeoParser::load_file(file_path)?;
-        let syntax_tree = LeoParser::parse_file(file_path, input_file)?;
-
-        // Generate aleo program from file
-        let mut program = Program::from(syntax_tree, import.path_string.clone());
-
-        // Use same namespace as calling function for imported symbols
-        program = program.name(scope);
-
-        // * -> import all imports, circuits, functions in the current scope
-        if import.is_star() {
-            // recursively evaluate program statements
-            self.resolve_definitions(program)
-        } else {
-            let program_name = program.name.clone();
-
-            // match each import symbol to a symbol in the imported file
-            for symbol in import.symbols.into_iter() {
-                // see if the imported symbol is a circuit
-                let matched_circuit = program
-                    .circuits
-                    .clone()
-                    .into_iter()
-                    .find(|(circuit_name, _circuit_def)| symbol.symbol == *circuit_name);
-
-                let value = match matched_circuit {
-                    Some((_circuit_name, circuit_def)) => ConstrainedValue::CircuitDefinition(circuit_def),
-                    None => {
-                        // see if the imported symbol is a function
-                        let matched_function = program
-                            .functions
-                            .clone()
-                            .into_iter()
-                            .find(|(function_name, _function)| symbol.symbol.name == *function_name.name);
-
-                        match matched_function {
-                            Some((_function_name, function)) => ConstrainedValue::Function(None, function),
-                            None => return Err(ImportError::unknown_symbol(symbol, program_name, file_path)),
-                        }
-                    }
-                };
-
-                // take the alias if it is present
-                let resolved_name = symbol.alias.unwrap_or(symbol.symbol);
-                let resolved_circuit_name = new_scope(program_name.clone(), resolved_name.to_string());
-
-                // store imported circuit under resolved name
-                self.store(resolved_circuit_name, value);
-            }
-
-            // evaluate all import statements in imported file
-            program
-                .imports
-                .into_iter()
-                .map(|nested_import| self.enforce_import(program_name.clone(), nested_import))
-                .collect::<Result<Vec<_>, ImportError>>()?;
-
-            Ok(())
-        }
-    }
-}
diff --git a/compiler/src/constraints/import/import.rs b/compiler/src/constraints/import/import.rs
new file mode 100644
index 0000000000..04f0bd4e90
--- /dev/null
+++ b/compiler/src/constraints/import/import.rs
@@ -0,0 +1,13 @@
+use crate::{constraints::ConstrainedProgram, errors::constraints::ImportError, GroupType};
+use leo_types::Import;
+
+use snarkos_models::curves::{Field, PrimeField};
+use std::env::current_dir;
+
+impl<F: Field + PrimeField, G: GroupType<F>> ConstrainedProgram<F, G> {
+    pub fn enforce_import(&mut self, scope: String, import: Import) -> Result<(), ImportError> {
+        let path = current_dir().map_err(|error| ImportError::directory_error(error, import.span.clone()))?;
+
+        self.enforce_package(scope, path, import.package)
+    }
+}
diff --git a/compiler/src/constraints/import/import_symbol.rs b/compiler/src/constraints/import/import_symbol.rs
new file mode 100644
index 0000000000..4011df354c
--- /dev/null
+++ b/compiler/src/constraints/import/import_symbol.rs
@@ -0,0 +1,102 @@
+use crate::{
+    constraints::{ConstrainedProgram, ConstrainedValue},
+    errors::constraints::ImportError,
+    new_scope,
+    GroupType,
+};
+use leo_ast::LeoParser;
+use leo_types::{ImportSymbol, Program, Span};
+
+use snarkos_models::curves::{Field, PrimeField};
+use std::fs::DirEntry;
+
+fn parse_import_file(entry: &DirEntry, span: &Span) -> Result<Program, ImportError> {
+    // make sure the given entry is file
+    let file_type = entry
+        .file_type()
+        .map_err(|error| ImportError::directory_error(error, span.clone()))?;
+    let file_name = entry
+        .file_name()
+        .into_string()
+        .map_err(|_| ImportError::convert_os_string(span.clone()))?;
+
+    if file_type.is_dir() {
+        return Err(ImportError::expected_file(file_name, span.clone()));
+    }
+
+    // Build the abstract syntax tree
+    let file_path = &entry.path();
+    let input_file = &LeoParser::load_file(file_path)?;
+    let syntax_tree = LeoParser::parse_file(file_path, input_file)?;
+
+    // Generate aleo program from file
+    Ok(Program::from(syntax_tree, file_name.clone()))
+}
+
+impl<F: Field + PrimeField, G: GroupType<F>> ConstrainedProgram<F, G> {
+    pub fn enforce_import_star(&mut self, scope: String, entry: &DirEntry, span: Span) -> Result<(), ImportError> {
+        let mut program = parse_import_file(entry, &span)?;
+
+        // Use same namespace as calling function for imported symbols
+        program = program.name(scope);
+
+        // * -> import all imports, circuits, functions in the current scope
+        self.resolve_definitions(program)
+    }
+
+    pub fn enforce_import_symbol(
+        &mut self,
+        scope: String,
+        entry: &DirEntry,
+        symbol: ImportSymbol,
+    ) -> Result<(), ImportError> {
+        // Generate aleo program from file
+        let mut program = parse_import_file(entry, &symbol.span)?;
+
+        // Use same namespace as calling function for imported symbols
+        program = program.name(scope);
+
+        let program_name = program.name.clone();
+
+        // see if the imported symbol is a circuit
+        let matched_circuit = program
+            .circuits
+            .clone()
+            .into_iter()
+            .find(|(circuit_name, _circuit_def)| symbol.symbol == *circuit_name);
+
+        let value = match matched_circuit {
+            Some((_circuit_name, circuit_def)) => ConstrainedValue::CircuitDefinition(circuit_def),
+            None => {
+                // see if the imported symbol is a function
+                let matched_function = program
+                    .functions
+                    .clone()
+                    .into_iter()
+                    .find(|(function_name, _function)| symbol.symbol == *function_name);
+
+                match matched_function {
+                    Some((_function_name, function)) => ConstrainedValue::Function(None, function),
+                    None => return Err(ImportError::unknown_symbol(symbol, program_name, &entry.path())),
+                }
+            }
+        };
+
+        // take the alias if it is present
+        let name = symbol.alias.unwrap_or(symbol.symbol);
+        let resolved_name = new_scope(program_name.clone(), name.to_string());
+
+        // store imported circuit under resolved name
+        self.store(resolved_name, value);
+
+        // evaluate all import statements in imported file
+        // todo: add logic to detect import loops
+        program
+            .imports
+            .into_iter()
+            .map(|nested_import| self.enforce_import(program_name.clone(), nested_import))
+            .collect::<Result<Vec<_>, ImportError>>()?;
+
+        Ok(())
+    }
+}
diff --git a/compiler/src/constraints/import/mod.rs b/compiler/src/constraints/import/mod.rs
new file mode 100644
index 0000000000..a007859c04
--- /dev/null
+++ b/compiler/src/constraints/import/mod.rs
@@ -0,0 +1,8 @@
+pub mod import;
+pub use import::*;
+
+pub mod import_symbol;
+pub use import_symbol::*;
+
+pub mod package;
+pub use package::*;
diff --git a/compiler/src/constraints/import/package.rs b/compiler/src/constraints/import/package.rs
new file mode 100644
index 0000000000..67dab822b3
--- /dev/null
+++ b/compiler/src/constraints/import/package.rs
@@ -0,0 +1,79 @@
+use crate::{constraints::ConstrainedProgram, errors::constraints::ImportError, GroupType};
+use leo_types::{Package, PackageAccess};
+
+use snarkos_models::curves::{Field, PrimeField};
+use std::{fs, fs::DirEntry, path::PathBuf};
+
+static SOURCE_FILE_EXTENSION: &str = ".leo";
+static SOURCE_DIRECTORY_NAME: &str = "src/";
+// pub(crate) static IMPORTS_DIRECTORY_NAME: &str = "imports/";
+
+impl<F: Field + PrimeField, G: GroupType<F>> ConstrainedProgram<F, G> {
+    pub fn enforce_package_access(
+        &mut self,
+        scope: String,
+        entry: &DirEntry,
+        access: PackageAccess,
+    ) -> Result<(), ImportError> {
+        // bring one or more import symbols into scope for the current constrained program
+        // we will recursively traverse sub packages here until we find the desired symbol
+        match access {
+            PackageAccess::Star(span) => self.enforce_import_star(scope, entry, span),
+            PackageAccess::Symbol(symbol) => self.enforce_import_symbol(scope, entry, symbol),
+            PackageAccess::SubPackage(package) => self.enforce_package(scope, entry.path(), *package),
+            PackageAccess::Multiple(accesses) => {
+                for access in accesses {
+                    self.enforce_package_access(scope.clone(), entry, access)?;
+                }
+
+                Ok(())
+            }
+        }
+    }
+
+    pub fn enforce_package(&mut self, scope: String, path: PathBuf, package: Package) -> Result<(), ImportError> {
+        let package_name = package.name;
+
+        // search for package name in local src directory
+        let mut source_directory = path.clone();
+        source_directory.push(SOURCE_DIRECTORY_NAME);
+
+        let entries = fs::read_dir(source_directory)
+            .map_err(|error| ImportError::directory_error(error, package_name.span.clone()))?
+            .into_iter()
+            .collect::<Result<Vec<_>, std::io::Error>>()
+            .map_err(|error| ImportError::directory_error(error, package_name.span.clone()))?;
+
+        let matched_source_entry = entries.into_iter().find(|entry| {
+            entry
+                .file_name()
+                .into_string()
+                .unwrap()
+                .trim_end_matches(SOURCE_FILE_EXTENSION)
+                .eq(&package_name.name)
+        });
+
+        // search for package name in imports directory
+        // let mut source_directory = path.clone();
+        // source_directory.push(IMPORTS_DIRECTORY_NAME);
+        //
+        // let entries = fs::read_dir(source_directory)
+        //     .map_err(|error| ImportError::directory_error(error, package_name.span.clone()))?
+        //     .into_iter()
+        //     .collect::<Result<Vec<_>, std::io::Error>>()
+        //     .map_err(|error| ImportError::directory_error(error, package_name.span.clone()))?;
+        //
+        // let matched_import_entry = entries.into_iter().find(|entry| {
+        //     entry.file_name().eq(&package_name.name)
+        // });
+
+        // todo: return error if package name is present in both directories
+
+        // Enforce package access
+        if let Some(entry) = matched_source_entry {
+            self.enforce_package_access(scope, &entry, package.access)
+        } else {
+            Err(ImportError::unknown_package(package_name))
+        }
+    }
+}
diff --git a/compiler/src/constraints/mod.rs b/compiler/src/constraints/mod.rs
index 2045087aff..42a3ce22f4 100644
--- a/compiler/src/constraints/mod.rs
+++ b/compiler/src/constraints/mod.rs
@@ -53,7 +53,6 @@ pub fn generate_constraints<F: Field + PrimeField, G: GroupType<F>, CS: Constrai
     match main.clone() {
         ConstrainedValue::Function(_circuit_identifier, function) => {
             let result = resolved_program.enforce_main_function(cs, program_name, function, parameters)?;
-            log::debug!("{}", result);
             Ok(result)
         }
         _ => Err(CompilerError::NoMainFunction),
diff --git a/compiler/src/constraints/statement.rs b/compiler/src/constraints/statement.rs
index 22ad5a9211..deda3e04a5 100644
--- a/compiler/src/constraints/statement.rs
+++ b/compiler/src/constraints/statement.rs
@@ -2,7 +2,7 @@
 
 use crate::{
     constraints::{ConstrainedProgram, ConstrainedValue},
-    errors::StatementError,
+    errors::{StatementError, ValueError},
     new_scope,
     GroupType,
 };
@@ -21,7 +21,6 @@ use leo_types::{
     Variable,
 };
 
-use crate::errors::ValueError;
 use snarkos_models::{
     curves::{Field, PrimeField},
     gadgets::{
@@ -95,11 +94,19 @@ impl<F: Field + PrimeField, G: GroupType<F>> ConstrainedProgram<F, G> {
             }
             RangeOrExpression::Range(from, to) => {
                 let from_index = match from {
-                    Some(integer) => integer.to_usize(span.clone())?,
+                    Some(integer) => {
+                        self.enforce_index(cs, file_scope.clone(), function_scope.clone(), integer, span.clone())?
+                    }
                     None => 0usize,
                 };
                 let to_index_option = match to {
-                    Some(integer) => Some(integer.to_usize(span.clone())?),
+                    Some(integer) => Some(self.enforce_index(
+                        cs,
+                        file_scope.clone(),
+                        function_scope.clone(),
+                        integer,
+                        span.clone(),
+                    )?),
                     None => None,
                 };
 
@@ -525,15 +532,16 @@ impl<F: Field + PrimeField, G: GroupType<F>> ConstrainedProgram<F, G> {
         function_scope: String,
         indicator: Option<Boolean>,
         index: Identifier,
-        start: Integer,
-        stop: Integer,
+        start: Expression,
+        stop: Expression,
         statements: Vec<Statement>,
         return_types: Vec<Type>,
         span: Span,
     ) -> Result<Vec<(Option<Boolean>, ConstrainedValue<F, G>)>, StatementError> {
         let mut results = vec![];
-        let from = start.to_usize(span.clone())?;
-        let to = stop.to_usize(span.clone())?;
+
+        let from = self.enforce_index(cs, file_scope.clone(), function_scope.clone(), start, span.clone())?;
+        let to = self.enforce_index(cs, file_scope.clone(), function_scope.clone(), stop, span.clone())?;
 
         for i in from..to {
             // Store index in current function scope.
diff --git a/compiler/src/errors/compiler.rs b/compiler/src/errors/compiler.rs
index 90ba339810..9f086b6b01 100644
--- a/compiler/src/errors/compiler.rs
+++ b/compiler/src/errors/compiler.rs
@@ -2,6 +2,7 @@ use crate::errors::{FunctionError, ImportError};
 use leo_ast::ParserError;
 use leo_inputs::InputParserError;
 
+use bincode::Error as SerdeError;
 use std::path::PathBuf;
 
 #[derive(Debug, Error)]
@@ -26,6 +27,9 @@ pub enum CompilerError {
 
     #[error("{}", _0)]
     ParserError(#[from] ParserError),
+
+    #[error("{}", _0)]
+    SerdeError(#[from] SerdeError),
 }
 
 impl CompilerError {
diff --git a/compiler/src/errors/constraints/import.rs b/compiler/src/errors/constraints/import.rs
index c5d3c238d9..4b5bf200f3 100644
--- a/compiler/src/errors/constraints/import.rs
+++ b/compiler/src/errors/constraints/import.rs
@@ -1,5 +1,5 @@
 use leo_ast::ParserError;
-use leo_types::{Error as FormattedError, ImportSymbol, Span};
+use leo_types::{Error as FormattedError, Identifier, ImportSymbol, Span};
 
 use std::{io, path::PathBuf};
 
@@ -17,12 +17,33 @@ impl ImportError {
         ImportError::Error(FormattedError::new_from_span(message, span))
     }
 
+    pub fn convert_os_string(span: Span) -> Self {
+        let message = format!("failed to convert file string name, maybe an illegal character?");
+
+        Self::new_from_span(message, span)
+    }
+
     pub fn directory_error(error: io::Error, span: Span) -> Self {
-        let message = format!("attempt to access current directory failed - {:?}", error);
+        let message = format!("compilation failed due to directory error - {:?}", error);
 
         Self::new_from_span(message, span)
     }
 
+    pub fn expected_file(entry: String, span: Span) -> Self {
+        let message = format!("cannot import symbol `{}` from directory `{}`", span.text, entry);
+
+        Self::new_from_span(message, span)
+    }
+
+    pub fn unknown_package(identifier: Identifier) -> Self {
+        let message = format!(
+            "cannot find imported package `{}` in source files or import directory",
+            identifier.name
+        );
+
+        Self::new_from_span(message, identifier.span)
+    }
+
     pub fn unknown_symbol(symbol: ImportSymbol, file: String, file_path: &PathBuf) -> Self {
         let message = format!("cannot find imported symbol `{}` in imported file `{}`", symbol, file);
         let mut error = FormattedError::new_from_span(message, symbol.span);
diff --git a/compiler/tests/circuits/member_field.leo b/compiler/tests/circuits/member_field.leo
index fe29638c0e..b9e7b7d88e 100644
--- a/compiler/tests/circuits/member_field.leo
+++ b/compiler/tests/circuits/member_field.leo
@@ -1,5 +1,5 @@
 circuit Circ {
-  x: u32
+  x: u32,
 }
 
 function main() -> u32 {
diff --git a/compiler/tests/circuits/member_field_and_function.leo b/compiler/tests/circuits/member_field_and_function.leo
index d020942c5c..207ae068e5 100644
--- a/compiler/tests/circuits/member_field_and_function.leo
+++ b/compiler/tests/circuits/member_field_and_function.leo
@@ -1,5 +1,5 @@
 circuit Foo {
-    foo: u32
+    foo: u32,
 
     static function bar() -> u32 {
         return 0
diff --git a/compiler/tests/import/alias.leo b/compiler/tests/import/alias.leo
index 7584591ae3..6af2e64a8a 100644
--- a/compiler/tests/import/alias.leo
+++ b/compiler/tests/import/alias.leo
@@ -1,4 +1,4 @@
-from "tests/import/test_import" import foo as bar;
+import test_import.foo as bar;
 
 function main() -> u32 {
   return bar()
diff --git a/compiler/tests/import/basic.leo b/compiler/tests/import/basic.leo
index f63096a4db..8166710c7a 100644
--- a/compiler/tests/import/basic.leo
+++ b/compiler/tests/import/basic.leo
@@ -1,4 +1,4 @@
-from "tests/import/test_import" import foo;
+import test_import.foo;
 
 function main() -> u32 {
   return foo()
diff --git a/compiler/tests/import/mod.rs b/compiler/tests/import/mod.rs
index 1219dfd100..5c7ba19c0a 100644
--- a/compiler/tests/import/mod.rs
+++ b/compiler/tests/import/mod.rs
@@ -1,13 +1,26 @@
 use crate::{integers::u32::output_one, parse_program};
 
+use std::env::{current_dir, set_current_dir};
+
+static TEST_SOURCE_DIRECTORY: &str = "tests/import";
+
 // Import tests rely on knowledge of local directories. They should be run locally only.
 
+fn set_local_dir() {
+    let mut local = current_dir().unwrap();
+    local.push(TEST_SOURCE_DIRECTORY);
+
+    set_current_dir(local).unwrap();
+}
+
 #[test]
 #[ignore]
 fn test_basic() {
     let bytes = include_bytes!("basic.leo");
     let program = parse_program(bytes).unwrap();
 
+    set_local_dir();
+
     output_one(program);
 }
 
@@ -17,6 +30,8 @@ fn test_multiple() {
     let bytes = include_bytes!("multiple.leo");
     let program = parse_program(bytes).unwrap();
 
+    set_local_dir();
+
     output_one(program);
 }
 
@@ -26,6 +41,8 @@ fn test_star() {
     let bytes = include_bytes!("star.leo");
     let program = parse_program(bytes).unwrap();
 
+    set_local_dir();
+
     output_one(program);
 }
 
@@ -35,5 +52,7 @@ fn test_alias() {
     let bytes = include_bytes!("alias.leo");
     let program = parse_program(bytes).unwrap();
 
+    set_local_dir();
+
     output_one(program);
 }
diff --git a/compiler/tests/import/multiple.leo b/compiler/tests/import/multiple.leo
index 4dd9720db1..ddf9093347 100644
--- a/compiler/tests/import/multiple.leo
+++ b/compiler/tests/import/multiple.leo
@@ -1,9 +1,9 @@
-from "tests/import/test_import" import {
+import test_import.(
   Point,
   foo
-};
+);
 
 function main() -> u32 {
   let p = Point { x: 1u32, y: 0u32 };
-  return foo()
+  return p.x
 }
\ No newline at end of file
diff --git a/compiler/tests/import/test_import.leo b/compiler/tests/import/src/test_import.leo
similarity index 100%
rename from compiler/tests/import/test_import.leo
rename to compiler/tests/import/src/test_import.leo
diff --git a/compiler/tests/import/star.leo b/compiler/tests/import/star.leo
index 5dca6dead2..113adc7088 100644
--- a/compiler/tests/import/star.leo
+++ b/compiler/tests/import/star.leo
@@ -1,4 +1,4 @@
-from "tests/import/test_import" import *;
+import test_import.*;
 
 function main() -> u32 {
   let p = Point { x: 1u32, y: 0u32 };
diff --git a/examples/pedersen_hash/src/main.leo b/examples/pedersen_hash/src/main.leo
index ce7b714716..5b417f604f 100644
--- a/examples/pedersen_hash/src/main.leo
+++ b/examples/pedersen_hash/src/main.leo
@@ -15,8 +15,8 @@ circuit PedersenHash {
 
 // The 'pedersen_hash' main function.
 function main() -> u32 {
-    let parameters = [0u32; 512];
-    let pedersen = PedersenHash::new(parameters);
+    const parameters = [0u32; 512];
+    const pedersen = PedersenHash::new(parameters);
     let input: bool[512] = [true; 512];
     return pedersen.hash(input)
 }
diff --git a/leo/cli.rs b/leo/cli.rs
index 2e321ec02a..ac49bfbf75 100644
--- a/leo/cli.rs
+++ b/leo/cli.rs
@@ -1,4 +1,4 @@
-use crate::{cli_types::*, errors::CLIError};
+use crate::{cli_types::*, errors::CLIError, logger};
 
 use clap::{App, AppSettings, Arg, ArgMatches, SubCommand};
 
@@ -66,6 +66,19 @@ pub trait CLI {
             .subcommands(subcommands)
     }
 
+    #[cfg_attr(tarpaulin, skip)]
+    fn process(arguments: &ArgMatches) -> Result<(), CLIError> {
+        // Set logging environment
+        match arguments.is_present("debug") {
+            true => logger::init_logger("leo", 2),
+            false => logger::init_logger("leo", 1),
+        }
+
+        let options = Self::parse(arguments)?;
+        let _output = Self::output(options)?;
+        Ok(())
+    }
+
     #[cfg_attr(tarpaulin, skip)]
     fn parse(arguments: &ArgMatches) -> Result<Self::Options, CLIError>;
 
diff --git a/leo/commands/build.rs b/leo/commands/build.rs
index 6011e13f57..d36cda5365 100644
--- a/leo/commands/build.rs
+++ b/leo/commands/build.rs
@@ -9,6 +9,7 @@ use leo_compiler::{compiler::Compiler, group::edwards_bls12::EdwardsGroupType};
 
 use snarkos_algorithms::snark::KeypairAssembly;
 use snarkos_curves::{bls12_377::Bls12_377, edwards_bls12::Fq};
+use snarkos_models::gadgets::r1cs::ConstraintSystem;
 
 use clap::ArgMatches;
 use std::{convert::TryFrom, env::current_dir};
@@ -51,6 +52,8 @@ impl CLI for BuildCommand {
             return Err(BuildError::MainFileDoesNotExist(package_path.as_os_str().to_owned()).into());
         }
 
+        log::info!("Compiling...");
+
         // Create the outputs directory
         OutputsDirectory::create(&package_path)?;
 
@@ -59,8 +62,10 @@ impl CLI for BuildCommand {
         main_file_path.push(SOURCE_DIRECTORY_NAME);
         main_file_path.push(MAIN_FILE_NAME);
 
+        // Load the program at `main_file_path`
+        let program = Compiler::<Fq, EdwardsGroupType>::new_from_path(package_name.clone(), main_file_path.clone())?;
+
         // Compute the current program checksum
-        let program = Compiler::<Fq, EdwardsGroupType>::init(package_name.clone(), main_file_path.clone())?;
         let program_checksum = program.checksum()?;
 
         // Generate the program on the constraint system and verify correctness
@@ -76,6 +81,7 @@ impl CLI for BuildCommand {
             let temporary_program = program.clone();
             let output = temporary_program.compile_constraints(&mut cs)?;
             log::debug!("Compiled constraints - {:#?}", output);
+            log::debug!("Number of constraints - {:#?}", cs.num_constraints());
         }
 
         // If a checksum file exists, check if it differs from the new checksum
@@ -92,9 +98,11 @@ impl CLI for BuildCommand {
         if checksum_differs {
             // Write the new checksum to the outputs directory
             checksum_file.write_to(&path, program_checksum)?;
+
+            log::debug!("Checksum saved ({:?})", path);
         }
 
-        log::info!("Compiled program in {:?}", main_file_path);
+        log::info!("Compiled program ({:?})", main_file_path);
 
         Ok((program, checksum_differs))
     }
diff --git a/leo/commands/clean.rs b/leo/commands/clean.rs
new file mode 100644
index 0000000000..3c41690019
--- /dev/null
+++ b/leo/commands/clean.rs
@@ -0,0 +1,50 @@
+use crate::{
+    cli::*,
+    cli_types::*,
+    errors::CLIError,
+    files::{ChecksumFile, Manifest, ProofFile, ProvingKeyFile, VerificationKeyFile},
+};
+
+use clap::ArgMatches;
+use std::{convert::TryFrom, env::current_dir};
+
+#[derive(Debug)]
+pub struct CleanCommand;
+
+impl CLI for CleanCommand {
+    type Options = ();
+    type Output = ();
+
+    const ABOUT: AboutType = "Clean the outputs directory";
+    const ARGUMENTS: &'static [ArgumentType] = &[];
+    const FLAGS: &'static [FlagType] = &[];
+    const NAME: NameType = "clean";
+    const OPTIONS: &'static [OptionType] = &[];
+    const SUBCOMMANDS: &'static [SubCommandType] = &[];
+
+    #[cfg_attr(tarpaulin, skip)]
+    fn parse(_arguments: &ArgMatches) -> Result<Self::Options, CLIError> {
+        Ok(())
+    }
+
+    #[cfg_attr(tarpaulin, skip)]
+    fn output(_options: Self::Options) -> Result<Self::Output, CLIError> {
+        // Get the package name
+        let path = current_dir()?;
+        let package_name = Manifest::try_from(&path)?.get_package_name();
+
+        // Remove the checksum from the outputs directory
+        ChecksumFile::new(&package_name).remove(&path)?;
+
+        // Remove the proving key from the outputs directory
+        ProvingKeyFile::new(&package_name).remove(&path)?;
+
+        // Remove the verification key from the outputs directory
+        VerificationKeyFile::new(&package_name).remove(&path)?;
+
+        // Remove the proof from the outputs directory
+        ProofFile::new(&package_name).remove(&path)?;
+
+        Ok(())
+    }
+}
diff --git a/leo/commands/load.rs b/leo/commands/load.rs
index 8cef939d0b..8129b98179 100644
--- a/leo/commands/load.rs
+++ b/leo/commands/load.rs
@@ -30,7 +30,7 @@ impl CLI for LoadCommand {
         let path = current_dir()?;
         let _package_name = Manifest::try_from(&path)?.get_package_name();
 
-        log::info!("Unimplemented - `leo deploy`");
+        log::info!("Unimplemented - `leo load`");
 
         Ok(())
     }
diff --git a/leo/commands/mod.rs b/leo/commands/mod.rs
index 14bc6c664d..152ebb158a 100644
--- a/leo/commands/mod.rs
+++ b/leo/commands/mod.rs
@@ -1,6 +1,9 @@
 pub mod build;
 pub use self::build::*;
 
+pub mod clean;
+pub use self::clean::*;
+
 pub mod deploy;
 pub use self::deploy::*;
 
diff --git a/leo/commands/prove.rs b/leo/commands/prove.rs
index 14d66e450b..9fd9993e22 100644
--- a/leo/commands/prove.rs
+++ b/leo/commands/prove.rs
@@ -40,6 +40,8 @@ impl CLI for ProveCommand {
         let path = current_dir()?;
         let package_name = Manifest::try_from(&path)?.get_package_name();
 
+        log::info!("Proving...");
+
         // Fetch program inputs here
         let inputs_string = InputsFile::new(&package_name).read_from(&path)?;
         program.parse_inputs(&inputs_string)?;
diff --git a/leo/commands/publish.rs b/leo/commands/publish.rs
index e1755457db..073221904d 100644
--- a/leo/commands/publish.rs
+++ b/leo/commands/publish.rs
@@ -1,4 +1,10 @@
-use crate::{cli::*, cli_types::*, commands::BuildCommand, errors::CLIError, files::Manifest};
+use crate::{
+    cli::*,
+    cli_types::*,
+    commands::BuildCommand,
+    errors::CLIError,
+    files::{Manifest, ZipFile},
+};
 
 use clap::ArgMatches;
 use std::{convert::TryFrom, env::current_dir};
@@ -28,9 +34,15 @@ impl CLI for PublishCommand {
 
         // Get the package name
         let path = current_dir()?;
-        let _package_name = Manifest::try_from(&path)?.get_package_name();
-
-        log::info!("Unimplemented - `leo publish`");
+        let package_name = Manifest::try_from(&path)?.get_package_name();
+
+        // Create zip file
+        let zip_file = ZipFile::new(&package_name);
+        if zip_file.exists_at(&path) {
+            log::info!("Existing package zip file found. Skipping compression.")
+        } else {
+            zip_file.write(&path)?;
+        }
 
         Ok(())
     }
diff --git a/leo/commands/setup.rs b/leo/commands/setup.rs
index 8cd115bc3e..072d18c870 100644
--- a/leo/commands/setup.rs
+++ b/leo/commands/setup.rs
@@ -52,6 +52,8 @@ impl CLI for SetupCommand {
 
         // If keys do not exist or the checksum differs, run the program setup
         if !keys_exist || checksum_differs {
+            log::info!("Setup starting...");
+
             // Start the timer
             let start = Instant::now();
 
@@ -61,18 +63,25 @@ impl CLI for SetupCommand {
             let prepared_verifying_key = prepare_verifying_key::<Bls12_377>(&parameters.vk);
 
             // End the timer
-            log::info!("Setup completed in {:?} milliseconds", start.elapsed().as_millis());
+            let end = start.elapsed().as_millis();
 
             // TODO (howardwu): Convert parameters to a 'proving key' struct for serialization.
             // Write the proving key file to the outputs directory
             let mut proving_key = vec![];
             parameters.write(&mut proving_key)?;
             ProvingKeyFile::new(&package_name).write_to(&path, &proving_key)?;
+            log::info!("Saving proving key ({:?})", path);
 
-            // Write the proving key file to the outputs directory
+            // Write the verification key file to the outputs directory
             let mut verification_key = vec![];
             prepared_verifying_key.write(&mut verification_key)?;
             VerificationKeyFile::new(&package_name).write_to(&path, &verification_key)?;
+            log::info!("Saving verification key ({:?})", path);
+
+            // Output the setup time
+            log::info!("Setup completed in {:?} milliseconds", end);
+        } else {
+            log::info!("Setup complete");
         }
 
         // Read the proving key file from the outputs directory
@@ -98,7 +107,7 @@ impl CLI for SetupCommand {
             }
         }
 
-        log::info!("Completed program setup");
+        log::info!("Program setup complete");
 
         Ok((program, parameters, prepared_verifying_key))
     }
diff --git a/leo/commands/test.rs b/leo/commands/test.rs
index 835c1a3919..cbff24c38e 100644
--- a/leo/commands/test.rs
+++ b/leo/commands/test.rs
@@ -57,7 +57,7 @@ impl CLI for TestCommand {
         main_file_path.push(MAIN_FILE_NAME);
 
         // Compute the current program checksum
-        let program = Compiler::<Fq, EdwardsGroupType>::init(package_name.clone(), main_file_path.clone())?;
+        let program = Compiler::<Fq, EdwardsGroupType>::new_from_path(package_name.clone(), main_file_path.clone())?;
 
         // Generate the program on the constraint system and verify correctness
         {
diff --git a/leo/commands/unload.rs b/leo/commands/unload.rs
index 52fe84b2b0..614506fe55 100644
--- a/leo/commands/unload.rs
+++ b/leo/commands/unload.rs
@@ -30,7 +30,7 @@ impl CLI for UnloadCommand {
         let path = current_dir()?;
         let _package_name = Manifest::try_from(&path)?.get_package_name();
 
-        log::info!("Unimplemented - `leo deploy`");
+        log::info!("Unimplemented - `leo unload`");
 
         Ok(())
     }
diff --git a/leo/errors/cli.rs b/leo/errors/cli.rs
index a1e4cb1957..a11c2bcb18 100644
--- a/leo/errors/cli.rs
+++ b/leo/errors/cli.rs
@@ -5,6 +5,9 @@ pub enum CLIError {
     #[error("{}", _0)]
     BuildError(BuildError),
 
+    #[error("{}", _0)]
+    BytesFileError(ZipFileError),
+
     #[error("{}: {}", _0, _1)]
     Crate(&'static str, String),
 
@@ -54,6 +57,13 @@ pub enum CLIError {
     VerificationKeyFileError(VerificationKeyFileError),
 }
 
+impl From<ZipFileError> for CLIError {
+    fn from(error: ZipFileError) -> Self {
+        log::error!("{}\n", error);
+        CLIError::BytesFileError(error)
+    }
+}
+
 impl From<BuildError> for CLIError {
     fn from(error: BuildError) -> Self {
         log::error!("{}\n", error);
diff --git a/leo/errors/files/checksum.rs b/leo/errors/files/checksum.rs
index 4fed871e41..65fe78c9e6 100644
--- a/leo/errors/files/checksum.rs
+++ b/leo/errors/files/checksum.rs
@@ -11,6 +11,9 @@ pub enum ChecksumFileError {
     #[error("Cannot read from the provided file path - {:?}", _0)]
     FileReadError(PathBuf),
 
+    #[error("Cannot remove the provided file - {:?}", _0)]
+    FileRemovalError(PathBuf),
+
     #[error("writing: {}", _0)]
     Writing(io::Error),
 }
diff --git a/leo/errors/files/mod.rs b/leo/errors/files/mod.rs
index c33f941ef7..d7f66d149d 100644
--- a/leo/errors/files/mod.rs
+++ b/leo/errors/files/mod.rs
@@ -1,3 +1,6 @@
+pub mod zip;
+pub use self::zip::*;
+
 pub mod checksum;
 pub use self::checksum::*;
 
diff --git a/leo/errors/files/proof.rs b/leo/errors/files/proof.rs
index fe7785cc5a..e0e8b1b022 100644
--- a/leo/errors/files/proof.rs
+++ b/leo/errors/files/proof.rs
@@ -11,6 +11,9 @@ pub enum ProofFileError {
     #[error("Cannot read from the provided file path - {:?}", _0)]
     FileReadError(PathBuf),
 
+    #[error("Cannot remove the provided file - {:?}", _0)]
+    FileRemovalError(PathBuf),
+
     #[error("writing: {}", _0)]
     Writing(io::Error),
 }
diff --git a/leo/errors/files/proving_key.rs b/leo/errors/files/proving_key.rs
index e24bfc6fa8..ef9481551a 100644
--- a/leo/errors/files/proving_key.rs
+++ b/leo/errors/files/proving_key.rs
@@ -11,6 +11,9 @@ pub enum ProvingKeyFileError {
     #[error("Cannot read from the provided file path - {:?}", _0)]
     FileReadError(PathBuf),
 
+    #[error("Cannot remove the provided file - {:?}", _0)]
+    FileRemovalError(PathBuf),
+
     #[error("writing: {}", _0)]
     Writing(io::Error),
 }
diff --git a/leo/errors/files/verification_key.rs b/leo/errors/files/verification_key.rs
index 1003af52ab..ffa3d34a61 100644
--- a/leo/errors/files/verification_key.rs
+++ b/leo/errors/files/verification_key.rs
@@ -11,6 +11,9 @@ pub enum VerificationKeyFileError {
     #[error("Cannot read from the provided file path - {:?}", _0)]
     FileReadError(PathBuf),
 
+    #[error("Cannot remove the provided file - {:?}", _0)]
+    FileRemovalError(PathBuf),
+
     #[error("Verification key file was corrupted")]
     IncorrectVerificationKey,
 
diff --git a/leo/errors/files/zip.rs b/leo/errors/files/zip.rs
new file mode 100644
index 0000000000..ca477d8167
--- /dev/null
+++ b/leo/errors/files/zip.rs
@@ -0,0 +1,30 @@
+use std::{io, path::PathBuf};
+use walkdir::Error as WalkDirError;
+use zip::result::ZipError;
+
+#[derive(Debug, Error)]
+pub enum ZipFileError {
+    #[error("{}: {}", _0, _1)]
+    Crate(&'static str, String),
+
+    #[error("creating: {}", _0)]
+    Creating(io::Error),
+
+    #[error("Cannot read from the provided file path - {:?}", _0)]
+    FileReadError(PathBuf),
+
+    #[error("writing: {}", _0)]
+    Writing(io::Error),
+
+    #[error("{}", _0)]
+    WalkDirError(#[from] WalkDirError),
+
+    #[error("{}", _0)]
+    ZipError(#[from] ZipError),
+}
+
+impl From<std::io::Error> for ZipFileError {
+    fn from(error: std::io::Error) -> Self {
+        ZipFileError::Crate("std::io", format!("{}", error))
+    }
+}
diff --git a/leo/files/checksum.rs b/leo/files/checksum.rs
index b71c227808..a8d9053021 100644
--- a/leo/files/checksum.rs
+++ b/leo/files/checksum.rs
@@ -42,11 +42,21 @@ impl ChecksumFile {
         let mut file = File::create(&path)?;
         file.write_all(checksum.as_bytes())?;
 
-        log::info!("Checksum stored to {:?}", path);
-
         Ok(())
     }
 
+    /// Removes the checksum at the given path if it exists. Returns `true` on success,
+    /// `false` if the file doesn't exist, and `Error` if the file system fails during operation.
+    pub fn remove(&self, path: &PathBuf) -> Result<bool, ChecksumFileError> {
+        let path = self.setup_file_path(path);
+        if !path.exists() {
+            return Ok(false);
+        }
+
+        fs::remove_file(&path).map_err(|_| ChecksumFileError::FileRemovalError(path.clone()))?;
+        Ok(true)
+    }
+
     fn setup_file_path(&self, path: &PathBuf) -> PathBuf {
         let mut path = path.to_owned();
         if path.is_dir() {
diff --git a/leo/files/mod.rs b/leo/files/mod.rs
index 6ef50f807b..472330e132 100644
--- a/leo/files/mod.rs
+++ b/leo/files/mod.rs
@@ -1,3 +1,6 @@
+pub mod zip;
+pub use self::zip::*;
+
 pub mod checksum;
 pub use self::checksum::*;
 
diff --git a/leo/files/proof.rs b/leo/files/proof.rs
index 0e6fa5578e..2353699ef8 100644
--- a/leo/files/proof.rs
+++ b/leo/files/proof.rs
@@ -43,11 +43,23 @@ impl ProofFile {
         let mut file = File::create(&path)?;
         file.write_all(proof)?;
 
-        log::info!("Proof stored to {:?}", path);
+        log::info!("Proof stored ({:?})", path);
 
         Ok(())
     }
 
+    /// Removes the proof at the given path if it exists. Returns `true` on success,
+    /// `false` if the file doesn't exist, and `Error` if the file system fails during operation.
+    pub fn remove(&self, path: &PathBuf) -> Result<bool, ProofFileError> {
+        let path = self.setup_file_path(path);
+        if !path.exists() {
+            return Ok(false);
+        }
+
+        fs::remove_file(&path).map_err(|_| ProofFileError::FileRemovalError(path.clone()))?;
+        Ok(true)
+    }
+
     fn setup_file_path(&self, path: &PathBuf) -> PathBuf {
         let mut path = path.to_owned();
         if path.is_dir() {
diff --git a/leo/files/proving_key.rs b/leo/files/proving_key.rs
index 49b42ba037..4c62a5cea9 100644
--- a/leo/files/proving_key.rs
+++ b/leo/files/proving_key.rs
@@ -42,11 +42,21 @@ impl ProvingKeyFile {
         let mut file = File::create(&path)?;
         file.write_all(proving_key)?;
 
-        log::info!("Proving key stored to {:?}", path);
-
         Ok(())
     }
 
+    /// Removes the proving key at the given path if it exists. Returns `true` on success,
+    /// `false` if the file doesn't exist, and `Error` if the file system fails during operation.
+    pub fn remove(&self, path: &PathBuf) -> Result<bool, ProvingKeyFileError> {
+        let path = self.setup_file_path(path);
+        if !path.exists() {
+            return Ok(false);
+        }
+
+        fs::remove_file(&path).map_err(|_| ProvingKeyFileError::FileRemovalError(path.clone()))?;
+        Ok(true)
+    }
+
     fn setup_file_path(&self, path: &PathBuf) -> PathBuf {
         let mut path = path.to_owned();
         if path.is_dir() {
diff --git a/leo/files/verification_key.rs b/leo/files/verification_key.rs
index 07071c8b18..6087e2d999 100644
--- a/leo/files/verification_key.rs
+++ b/leo/files/verification_key.rs
@@ -42,11 +42,21 @@ impl VerificationKeyFile {
         let mut file = File::create(&path)?;
         file.write_all(verification_key)?;
 
-        log::info!("Verification key stored to {:?}", path);
-
         Ok(())
     }
 
+    /// Removes the verification key at the given path if it exists. Returns `true` on success,
+    /// `false` if the file doesn't exist, and `Error` if the file system fails during operation.
+    pub fn remove(&self, path: &PathBuf) -> Result<bool, VerificationKeyFileError> {
+        let path = self.setup_file_path(path);
+        if !path.exists() {
+            return Ok(false);
+        }
+
+        fs::remove_file(&path).map_err(|_| VerificationKeyFileError::FileRemovalError(path.clone()))?;
+        Ok(true)
+    }
+
     fn setup_file_path(&self, path: &PathBuf) -> PathBuf {
         let mut path = path.to_owned();
         if path.is_dir() {
diff --git a/leo/files/zip.rs b/leo/files/zip.rs
new file mode 100644
index 0000000000..d536c00838
--- /dev/null
+++ b/leo/files/zip.rs
@@ -0,0 +1,136 @@
+//! The program package zip file.
+
+use crate::{
+    directories::{INPUTS_DIRECTORY_NAME, OUTPUTS_DIRECTORY_NAME},
+    errors::ZipFileError,
+    files::{
+        CHECKSUM_FILE_EXTENSION,
+        INPUTS_FILE_EXTENSION,
+        PROOF_FILE_EXTENSION,
+        PROVING_KEY_FILE_EXTENSION,
+        VERIFICATION_KEY_FILE_EXTENSION,
+    },
+};
+
+use serde::Deserialize;
+use std::{
+    fs::File,
+    io::{Read, Write},
+    path::{Path, PathBuf},
+};
+use walkdir::WalkDir;
+use zip::write::{FileOptions, ZipWriter};
+
+pub static ZIP_FILE_EXTENSION: &str = ".zip";
+
+#[derive(Deserialize)]
+pub struct ZipFile {
+    pub package_name: String,
+}
+
+impl ZipFile {
+    pub fn new(package_name: &str) -> Self {
+        Self {
+            package_name: package_name.to_string(),
+        }
+    }
+
+    pub fn exists_at(&self, path: &PathBuf) -> bool {
+        let path = self.setup_file_path(path);
+        path.exists()
+    }
+
+    // /// Reads the program bytes from the given file path if it exists.
+    // pub fn read_from(&self, path: &PathBuf) -> Result<Vec<u8>, ZipFileError> {
+    //     let path = self.setup_file_path(path);
+    //
+    //     Ok(fs::read(&path).map_err(|_| ZipFileError::FileReadError(path.clone()))?)
+    // }
+
+    /// Writes the current package contents to a zip file.
+    pub fn write(&self, src_dir: &PathBuf) -> Result<(), ZipFileError> {
+        // Build walkdir iterator from current package
+        let walkdir = WalkDir::new(src_dir.clone());
+
+        // Create zip file
+        let path = self.setup_file_path(src_dir);
+
+        let file = &mut File::create(&path)?;
+        let mut zip = ZipWriter::new(file);
+        let options = FileOptions::default()
+            .compression_method(zip::CompressionMethod::Stored)
+            .unix_permissions(0o755);
+
+        // Walk through files in directory and write desired ones to the zip file
+        let mut buffer = Vec::new();
+        for entry in walkdir.into_iter().filter_map(|e| e.ok()) {
+            let path = entry.path();
+            let name = path.strip_prefix(src_dir.as_path()).unwrap();
+
+            // filter excluded paths
+            if is_excluded(name) {
+                continue;
+            }
+
+            // write file or directory
+            if path.is_file() {
+                log::info!("\tadding file {:?} as {:?}", path, name);
+                zip.start_file_from_path(name, options)?;
+                let mut f = File::open(path)?;
+
+                f.read_to_end(&mut buffer)?;
+                zip.write_all(&*buffer)?;
+                buffer.clear();
+            } else if name.as_os_str().len() != 0 {
+                // Only if not root Avoids path spec / warning
+                // and mapname conversion failed error on unzip
+                log::info!("\tadding dir {:?} as {:?}", path, name);
+                zip.add_directory_from_path(name, options)?;
+            }
+        }
+
+        zip.finish()?;
+
+        log::info!("Package zip file created successfully {:?}", path);
+
+        Ok(())
+    }
+
+    fn setup_file_path(&self, path: &PathBuf) -> PathBuf {
+        let mut path = path.to_owned();
+        if path.is_dir() {
+            if !path.ends_with(OUTPUTS_DIRECTORY_NAME) {
+                path.push(PathBuf::from(OUTPUTS_DIRECTORY_NAME));
+            }
+            path.push(PathBuf::from(format!("{}{}", self.package_name, ZIP_FILE_EXTENSION)));
+        }
+        path
+    }
+}
+
+fn is_excluded(path: &Path) -> bool {
+    // excluded directories: `/inputs`, `/outputs`
+    if path.ends_with(INPUTS_DIRECTORY_NAME.trim_end_matches("/"))
+        | path.ends_with(OUTPUTS_DIRECTORY_NAME.trim_end_matches("/"))
+    {
+        return true;
+    }
+
+    // excluded extensions: `.in`, `.bytes`, `lpk`, `lvk`, `.proof`, `.sum`
+    path.extension()
+        .map(|ext| {
+            if ext.eq(INPUTS_FILE_EXTENSION.trim_start_matches("."))
+                | ext.eq(ZIP_FILE_EXTENSION.trim_start_matches("."))
+                | ext.eq(PROVING_KEY_FILE_EXTENSION.trim_start_matches("."))
+                | ext.eq(VERIFICATION_KEY_FILE_EXTENSION.trim_start_matches("."))
+                | ext.eq(PROOF_FILE_EXTENSION.trim_start_matches("."))
+                | ext.eq(CHECKSUM_FILE_EXTENSION.trim_start_matches("."))
+                | ext.eq(ZIP_FILE_EXTENSION.trim_start_matches("."))
+            {
+                true
+            } else {
+                false
+            }
+        })
+        .unwrap_or(false)
+}
diff --git a/leo/logger.rs b/leo/logger.rs
index e4c12234f7..98c48c34d7 100644
--- a/leo/logger.rs
+++ b/leo/logger.rs
@@ -3,33 +3,18 @@ use std::io::Write;
 
 const LEVEL_NAME_LENGTH: usize = 10;
 
-#[allow(dead_code)]
-fn level_string(level: log::Level) -> colored::ColoredString {
-    match level {
-        log::Level::Error => "ERROR".bold().red(),
-        log::Level::Warn => "WARN".bold().yellow(),
-        log::Level::Info => "INFO".bold().blue(),
-        log::Level::Debug => "DEBUG".bold().magenta(),
-        log::Level::Trace => "TRACE".bold(),
-    }
-}
-
 #[allow(dead_code)]
 fn colored_string(level: log::Level, message: &str) -> colored::ColoredString {
     match level {
         log::Level::Error => message.bold().red(),
         log::Level::Warn => message.bold().yellow(),
-        log::Level::Info => message.bold().blue(),
+        log::Level::Info => message.bold().cyan(),
         log::Level::Debug => message.bold().magenta(),
         log::Level::Trace => message.bold(),
     }
 }
 
 /// Initialize logger with custom format and verbosity.
-///
-/// # Arguments
-///
-/// * `verbosity` - Verbosity level. 0 for `Warn`, 1 for `Info`, 2 for `Debug`, more for `Trace`
 pub fn init_logger(app_name: &'static str, verbosity: usize) {
     env_logger::builder()
         .filter_level(match verbosity {
@@ -46,8 +31,7 @@ pub fn init_logger(app_name: &'static str, verbosity: usize) {
 
             writeln!(
                 buf,
-                "{:>5}{:>5}  {}",
-                level_string(record.level()),
+                "{:>5}  {}",
                 colored_string(record.level(), app_name),
                 record.args().to_string().replace("\n", &padding)
             )
diff --git a/leo/main.rs b/leo/main.rs
index 6b542b3f41..f11932b0b4 100644
--- a/leo/main.rs
+++ b/leo/main.rs
@@ -1,11 +1,9 @@
-use leo::{cli::*, commands::*, errors::CLIError, logger};
+use leo::{cli::*, commands::*, errors::CLIError};
 
-use clap::{App, AppSettings};
+use clap::{App, AppSettings, Arg};
 
 #[cfg_attr(tarpaulin, skip)]
 fn main() -> Result<(), CLIError> {
-    logger::init_logger("leo", 1);
-
     let arguments = App::new("leo")
         .version("v0.1.0")
         .about("Leo compiler and package manager")
@@ -16,46 +14,41 @@ fn main() -> Result<(), CLIError> {
             AppSettings::DisableVersion,
             AppSettings::SubcommandRequiredElseHelp,
         ])
+        .args(&[Arg::with_name("debug")
+            .short("d")
+            .long("debug")
+            .help("Enables debugging mode")
+            .global(true)])
         .subcommands(vec![
             NewCommand::new().display_order(0),
             InitCommand::new().display_order(1),
             BuildCommand::new().display_order(2),
-            LoadCommand::new().display_order(3),
-            UnloadCommand::new().display_order(4),
-            SetupCommand::new().display_order(5),
-            ProveCommand::new().display_order(6),
-            RunCommand::new().display_order(7),
-            PublishCommand::new().display_order(8),
-            DeployCommand::new().display_order(9),
-            TestCommand::new().display_order(10),
+            TestCommand::new().display_order(3),
+            LoadCommand::new().display_order(4),
+            UnloadCommand::new().display_order(5),
+            SetupCommand::new().display_order(6),
+            ProveCommand::new().display_order(7),
+            RunCommand::new().display_order(8),
+            PublishCommand::new().display_order(9),
+            DeployCommand::new().display_order(10),
+            CleanCommand::new().display_order(11),
         ])
         .set_term_width(0)
         .get_matches();
 
     match arguments.subcommand() {
-        ("new", Some(arguments)) => NewCommand::output(NewCommand::parse(arguments)?),
-        ("init", Some(arguments)) => InitCommand::output(InitCommand::parse(arguments)?),
-        ("build", Some(arguments)) => {
-            BuildCommand::output(BuildCommand::parse(arguments)?)?;
-            Ok(())
-        }
-        ("load", Some(arguments)) => LoadCommand::output(LoadCommand::parse(arguments)?),
-        ("unload", Some(arguments)) => UnloadCommand::output(UnloadCommand::parse(arguments)?),
-        ("setup", Some(arguments)) => {
-            SetupCommand::output(SetupCommand::parse(arguments)?)?;
-            Ok(())
-        }
-        ("prove", Some(arguments)) => {
-            ProveCommand::output(ProveCommand::parse(arguments)?)?;
-            Ok(())
-        }
-        ("run", Some(arguments)) => RunCommand::output(RunCommand::parse(arguments)?),
-        ("publish", Some(arguments)) => PublishCommand::output(PublishCommand::parse(arguments)?),
-        ("deploy", Some(arguments)) => DeployCommand::output(DeployCommand::parse(arguments)?),
-        ("test", Some(arguments)) => {
-            TestCommand::output(TestCommand::parse(arguments)?)?;
-            Ok(())
-        }
+        ("new", Some(arguments)) => NewCommand::process(arguments),
+        ("init", Some(arguments)) => InitCommand::process(arguments),
+        ("build", Some(arguments)) => BuildCommand::process(arguments),
+        ("test", Some(arguments)) => TestCommand::process(arguments),
+        ("load", Some(arguments)) => LoadCommand::process(arguments),
+        ("unload", Some(arguments)) => UnloadCommand::process(arguments),
+        ("setup", Some(arguments)) => SetupCommand::process(arguments),
+        ("prove", Some(arguments)) => ProveCommand::process(arguments),
+        ("run", Some(arguments)) => RunCommand::process(arguments),
+        ("publish", Some(arguments)) => PublishCommand::process(arguments),
+        ("deploy", Some(arguments)) => DeployCommand::process(arguments),
+        ("clean", Some(arguments)) => CleanCommand::process(arguments),
         _ => unreachable!(),
     }
 }
diff --git a/types/Cargo.toml b/types/Cargo.toml
index 4925f2a017..d02059ca6e 100644
--- a/types/Cargo.toml
+++ b/types/Cargo.toml
@@ -12,4 +12,5 @@ snarkos-errors = { git = "ssh://git@github.com/AleoHQ/snarkOS.git", rev = "c7a56
 snarkos-models =  { git = "ssh://git@github.com/AleoHQ/snarkOS.git", rev = "c7a56d9", default-features = false }
 
 pest = { version = "2.0" }
+serde = { version = "1.0" }
 thiserror = { version = "1.0" }
diff --git a/types/src/circuits/circuit.rs b/types/src/circuits/circuit.rs
index f2067ff602..d49d7de5e4 100644
--- a/types/src/circuits/circuit.rs
+++ b/types/src/circuits/circuit.rs
@@ -1,9 +1,10 @@
 use crate::{CircuitMember, Identifier};
 use leo_ast::circuits::Circuit as AstCircuit;
 
+use serde::{Deserialize, Serialize};
 use std::fmt;
 
-#[derive(Clone, PartialEq, Eq)]
+#[derive(Clone, PartialEq, Eq, Serialize, Deserialize)]
 pub struct Circuit {
     pub circuit_name: Identifier,
     pub members: Vec<CircuitMember>,
diff --git a/types/src/circuits/circuit_field_definition.rs b/types/src/circuits/circuit_field_definition.rs
index 50e40c985d..38582efef2 100644
--- a/types/src/circuits/circuit_field_definition.rs
+++ b/types/src/circuits/circuit_field_definition.rs
@@ -1,7 +1,9 @@
 use crate::{Expression, Identifier};
 use leo_ast::circuits::CircuitField;
 
-#[derive(Clone, Debug, PartialEq, Eq)]
+use serde::{Deserialize, Serialize};
+
+#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
 pub struct CircuitFieldDefinition {
     pub identifier: Identifier,
     pub expression: Expression,
diff --git a/types/src/circuits/circuit_member.rs b/types/src/circuits/circuit_member.rs
index 534bf2fb48..dda39dd3ee 100644
--- a/types/src/circuits/circuit_member.rs
+++ b/types/src/circuits/circuit_member.rs
@@ -5,9 +5,10 @@ use leo_ast::circuits::{
     CircuitMember as AstCircuitMember,
 };
 
+use serde::{Deserialize, Serialize};
 use std::fmt;
 
-#[derive(Clone, PartialEq, Eq)]
+#[derive(Clone, PartialEq, Eq, Serialize, Deserialize)]
 pub enum CircuitMember {
     CircuitField(Identifier, Type),
     CircuitFunction(bool, Function),
diff --git a/types/src/common/assignee.rs b/types/src/common/assignee.rs
index ae4245027f..8ea03fa708 100644
--- a/types/src/common/assignee.rs
+++ b/types/src/common/assignee.rs
@@ -4,10 +4,11 @@ use leo_ast::{
     common::{Assignee as AstAssignee, Identifier as AstIdentifier},
 };
 
+use serde::{Deserialize, Serialize};
 use std::fmt;
 
 /// Definition assignee: v, arr[0..2], Point p.x
-#[derive(Debug, Clone, PartialEq, Eq)]
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
 pub enum Assignee {
     Identifier(Identifier),
     Array(Box<Assignee>, RangeOrExpression),
diff --git a/types/src/common/declare.rs b/types/src/common/declare.rs
index d543100fd1..12930d32d6 100644
--- a/types/src/common/declare.rs
+++ b/types/src/common/declare.rs
@@ -1,8 +1,9 @@
 use leo_ast::common::Declare as AstDeclare;
 
+use serde::{Deserialize, Serialize};
 use std::fmt;
 
-#[derive(Debug, Clone, PartialEq, Eq)]
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
 pub enum Declare {
     Const,
     Let,
diff --git a/types/src/common/identifier.rs b/types/src/common/identifier.rs
index 4c7e0dca18..0739dbe3d3 100644
--- a/types/src/common/identifier.rs
+++ b/types/src/common/identifier.rs
@@ -1,10 +1,11 @@
 use crate::Span;
 use leo_ast::common::Identifier as AstIdentifier;
 
+use serde::{Deserialize, Serialize};
 use std::fmt;
 
 /// An identifier in the constrained program.
-#[derive(Clone, Hash)]
+#[derive(Clone, Hash, Serialize, Deserialize)]
 pub struct Identifier {
     pub name: String,
     pub span: Span,
diff --git a/types/src/common/range_or_expression.rs b/types/src/common/range_or_expression.rs
index a006f28c26..b4d45f6326 100644
--- a/types/src/common/range_or_expression.rs
+++ b/types/src/common/range_or_expression.rs
@@ -1,32 +1,32 @@
-use crate::{Expression, Integer};
-use leo_ast::common::RangeOrExpression as AstRangeOrExpression;
+use crate::{Error as FormattedError, Expression, Span};
+use leo_ast::{common::RangeOrExpression as AstRangeOrExpression, values::NumberValue};
 
+use serde::{Deserialize, Serialize};
 use std::fmt;
 
 /// Range or expression enum
-#[derive(Debug, Clone, PartialEq, Eq)]
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
 pub enum RangeOrExpression {
-    Range(Option<Integer>, Option<Integer>),
+    Range(Option<Expression>, Option<Expression>),
     Expression(Expression),
 }
 
+pub fn unwrap_bound(bound: Option<NumberValue>) -> Option<usize> {
+    bound.map(|number| {
+        let message = format!("Range bounds should be integers");
+        let error = FormattedError::new_from_span(message, Span::from(number.span.clone()));
+
+        number.value.parse::<usize>().expect(&error.to_string())
+    })
+}
+
 impl<'ast> From<AstRangeOrExpression<'ast>> for RangeOrExpression {
     fn from(range_or_expression: AstRangeOrExpression<'ast>) -> Self {
         match range_or_expression {
-            AstRangeOrExpression::Range(range) => {
-                let from = range.from.map(|from| match Expression::from(from.0) {
-                    Expression::Integer(number) => number,
-                    Expression::Implicit(string, _span) => Integer::from_implicit(string),
-                    expression => unimplemented!("Range bounds should be integers, found {}", expression),
-                });
-                let to = range.to.map(|to| match Expression::from(to.0) {
-                    Expression::Integer(number) => number,
-                    Expression::Implicit(string, _span) => Integer::from_implicit(string),
-                    expression => unimplemented!("Range bounds should be integers, found {}", expression),
-                });
-
-                RangeOrExpression::Range(from, to)
-            }
+            AstRangeOrExpression::Range(range) => RangeOrExpression::Range(
+                range.from.map(|expression| Expression::from(expression)),
+                range.to.map(|expression| Expression::from(expression)),
+            ),
             AstRangeOrExpression::Expression(expression) => RangeOrExpression::Expression(Expression::from(expression)),
         }
     }
diff --git a/types/src/common/span.rs b/types/src/common/span.rs
index b9a0adc144..cc62e32796 100644
--- a/types/src/common/span.rs
+++ b/types/src/common/span.rs
@@ -1,6 +1,7 @@
 use pest::{Position, Span as AstSpan};
+use serde::{Deserialize, Serialize};
 
-#[derive(Clone, Debug, PartialEq, Eq, Hash)]
+#[derive(Clone, Debug, PartialEq, Eq, Hash, Serialize, Deserialize)]
 pub struct Span {
     /// text of input string
     pub text: String,
diff --git a/types/src/common/spread_or_expression.rs b/types/src/common/spread_or_expression.rs
index 2110194310..79dd1c2dda 100644
--- a/types/src/common/spread_or_expression.rs
+++ b/types/src/common/spread_or_expression.rs
@@ -1,10 +1,11 @@
 use crate::Expression;
 use leo_ast::common::SpreadOrExpression as AstSpreadOrExpression;
 
+use serde::{Deserialize, Serialize};
 use std::fmt;
 
 /// Spread or expression
-#[derive(Debug, Clone, PartialEq, Eq)]
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
 pub enum SpreadOrExpression {
     Spread(Expression),
     Expression(Expression),
diff --git a/types/src/common/variable.rs b/types/src/common/variable.rs
index eb38c1e03f..247d5255d0 100644
--- a/types/src/common/variable.rs
+++ b/types/src/common/variable.rs
@@ -1,10 +1,11 @@
 use crate::{Identifier, Type};
 use leo_ast::common::Variable as AstVariable;
 
+use serde::{Deserialize, Serialize};
 use std::fmt;
 
 /// A variable that is assigned to a value in the constrained program
-#[derive(Clone, PartialEq, Eq)]
+#[derive(Clone, PartialEq, Eq, Serialize, Deserialize)]
 pub struct Variable {
     pub identifier: Identifier,
     pub mutable: bool,
diff --git a/types/src/expression.rs b/types/src/expression.rs
index bc5c6b1ba4..08c89ee3ef 100644
--- a/types/src/expression.rs
+++ b/types/src/expression.rs
@@ -1,4 +1,4 @@
-use crate::{CircuitFieldDefinition, Identifier, Integer, RangeOrExpression, Span, SpreadOrExpression};
+use crate::{CircuitFieldDefinition, Identifier, IntegerType, RangeOrExpression, Span, SpreadOrExpression};
 use leo_ast::{
     access::{Access, AssigneeAccess},
     common::{Assignee, Identifier as AstIdentifier},
@@ -16,21 +16,20 @@ use leo_ast::{
     values::{BooleanValue, FieldValue, GroupValue, IntegerValue, NumberImplicitValue, Value},
 };
 
-use snarkos_models::gadgets::utilities::boolean::Boolean;
-
+use serde::{Deserialize, Serialize};
 use std::fmt;
 
 /// Expression that evaluates to a value
-#[derive(Debug, Clone, PartialEq, Eq)]
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
 pub enum Expression {
     // Identifier
     Identifier(Identifier),
 
     // Values
-    Integer(Integer),
+    Integer(IntegerType, String, Span),
     Field(String, Span),
     Group(String, Span),
-    Boolean(Boolean),
+    Boolean(String, Span),
     Implicit(String, Span),
 
     // Number operations
@@ -122,10 +121,10 @@ impl<'ast> fmt::Display for Expression {
             Expression::Identifier(ref variable) => write!(f, "{}", variable),
 
             // Values
-            Expression::Integer(ref integer) => write!(f, "{}", integer),
+            Expression::Integer(ref type_, ref integer, ref _span) => write!(f, "{}{}", integer, type_),
             Expression::Field(ref field, ref _span) => write!(f, "{}", field),
             Expression::Group(ref group, ref _span) => write!(f, "{}", group),
-            Expression::Boolean(ref bool) => write!(f, "{}", bool.get_value().unwrap()),
+            Expression::Boolean(ref bool, ref _span) => write!(f, "{}", bool),
             Expression::Implicit(ref value, ref _span) => write!(f, "{}", value),
 
             // Number operations
@@ -438,9 +437,7 @@ impl<'ast> From<GroupValue<'ast>> for Expression {
 
 impl<'ast> From<BooleanValue<'ast>> for Expression {
     fn from(boolean: BooleanValue<'ast>) -> Self {
-        Expression::Boolean(Boolean::Constant(
-            boolean.value.parse::<bool>().expect("unable to parse boolean"),
-        ))
+        Expression::Boolean(boolean.value, Span::from(boolean.span))
     }
 }
 
@@ -451,8 +448,12 @@ impl<'ast> From<NumberImplicitValue<'ast>> for Expression {
 }
 
 impl<'ast> From<IntegerValue<'ast>> for Expression {
-    fn from(field: IntegerValue<'ast>) -> Self {
-        Expression::Integer(Integer::from(field.number, field._type))
+    fn from(integer: IntegerValue<'ast>) -> Self {
+        Expression::Integer(
+            IntegerType::from(integer._type),
+            integer.number.value,
+            Span::from(integer.span),
+        )
     }
 }
 
diff --git a/types/src/functions/function.rs b/types/src/functions/function.rs
index 1f09329e14..8808234d61 100644
--- a/types/src/functions/function.rs
+++ b/types/src/functions/function.rs
@@ -1,9 +1,10 @@
 use crate::{FunctionInput, Identifier, Span, Statement, Type};
 use leo_ast::functions::Function as AstFunction;
 
+use serde::{Deserialize, Serialize};
 use std::fmt;
 
-#[derive(Clone, PartialEq, Eq)]
+#[derive(Clone, PartialEq, Eq, Serialize, Deserialize)]
 pub struct Function {
     pub function_name: Identifier,
     pub inputs: Vec<FunctionInput>,
diff --git a/types/src/functions/function_input.rs b/types/src/functions/function_input.rs
index e28b12a8e2..98c88071dc 100644
--- a/types/src/functions/function_input.rs
+++ b/types/src/functions/function_input.rs
@@ -1,9 +1,10 @@
 use crate::{Identifier, Span, Type};
 use leo_ast::functions::FunctionInput as AstFunctionInput;
 
+use serde::{Deserialize, Serialize};
 use std::fmt;
 
-#[derive(Clone, PartialEq, Eq)]
+#[derive(Clone, PartialEq, Eq, Serialize, Deserialize)]
 pub struct FunctionInput {
     pub identifier: Identifier,
     pub mutable: bool,
diff --git a/types/src/functions/test_function.rs b/types/src/functions/test_function.rs
index c4a318a75f..cc20c13ad7 100644
--- a/types/src/functions/test_function.rs
+++ b/types/src/functions/test_function.rs
@@ -1,7 +1,9 @@
 use crate::Function;
 use leo_ast::functions::TestFunction as AstTestFunction;
 
-#[derive(Clone, Debug, PartialEq, Eq)]
+use serde::{Deserialize, Serialize};
+
+#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
 pub struct TestFunction(pub Function);
 
 impl<'ast> From<AstTestFunction<'ast>> for TestFunction {
diff --git a/types/src/imports/import.rs b/types/src/imports/import.rs
index 5f0a1947cc..30c231f466 100644
--- a/types/src/imports/import.rs
+++ b/types/src/imports/import.rs
@@ -1,26 +1,21 @@
 //! The import type for a Leo program.
 
-use crate::{ImportSymbol, Span};
+use crate::{Package, Span};
 use leo_ast::imports::Import as AstImport;
 
+use serde::{Deserialize, Serialize};
 use std::fmt;
 
-#[derive(Clone)]
+#[derive(Clone, Serialize, Deserialize)]
 pub struct Import {
-    pub path_string: String,
-    pub symbols: Vec<ImportSymbol>,
+    pub package: Package,
     pub span: Span,
 }
 
 impl<'ast> From<AstImport<'ast>> for Import {
     fn from(import: AstImport<'ast>) -> Self {
         Import {
-            path_string: import.source.value,
-            symbols: import
-                .symbols
-                .into_iter()
-                .map(|symbol| ImportSymbol::from(symbol))
-                .collect(),
+            package: Package::from(import.package),
             span: Span::from(import.span),
         }
     }
@@ -28,28 +23,17 @@ impl<'ast> From<AstImport<'ast>> for Import {
 
 impl Import {
     pub fn path_string_full(&self) -> String {
-        format!("{}.leo", self.path_string)
+        format!("{}.leo", self.package.name)
     }
 
     // from "./import" import *;
     pub fn is_star(&self) -> bool {
-        self.symbols.is_empty()
+        // self.symbols.is_empty()
+        false
     }
 
     fn format(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        write!(f, "from {} import ", self.path_string)?;
-        if self.symbols.is_empty() {
-            write!(f, "*")
-        } else {
-            write!(f, "{{\n")?;
-            for (i, symbol) in self.symbols.iter().enumerate() {
-                write!(f, "{}", symbol)?;
-                if i < self.symbols.len() - 1 {
-                    write!(f, ",\n")?;
-                }
-            }
-            write!(f, "\n}}")
-        }
+        write!(f, "import {};", self.package)
     }
 }
 
diff --git a/types/src/imports/import_symbol.rs b/types/src/imports/import_symbol.rs
index 1c96198764..5865df3324 100644
--- a/types/src/imports/import_symbol.rs
+++ b/types/src/imports/import_symbol.rs
@@ -1,9 +1,10 @@
 use crate::{Identifier, Span};
 use leo_ast::imports::ImportSymbol as AstImportSymbol;
 
+use serde::{Deserialize, Serialize};
 use std::fmt;
 
-#[derive(Clone)]
+#[derive(Clone, Serialize, Deserialize)]
 pub struct ImportSymbol {
     pub symbol: Identifier,
     pub alias: Option<Identifier>,
diff --git a/types/src/imports/mod.rs b/types/src/imports/mod.rs
index 78250a4179..09c4d143f2 100644
--- a/types/src/imports/mod.rs
+++ b/types/src/imports/mod.rs
@@ -3,3 +3,9 @@ pub use import::*;
 
 pub mod import_symbol;
 pub use import_symbol::*;
+
+pub mod package;
+pub use package::*;
+
+pub mod package_access;
+pub use package_access::*;
diff --git a/types/src/imports/package.rs b/types/src/imports/package.rs
new file mode 100644
index 0000000000..666f92e5b9
--- /dev/null
+++ b/types/src/imports/package.rs
@@ -0,0 +1,40 @@
+use crate::{common::Identifier, PackageAccess, Span};
+use leo_ast::imports::Package as AstPackage;
+
+use serde::{Deserialize, Serialize};
+use std::fmt;
+
+#[derive(Clone, Serialize, Deserialize)]
+pub struct Package {
+    pub name: Identifier,
+    pub access: PackageAccess,
+    pub span: Span,
+}
+
+impl<'ast> From<AstPackage<'ast>> for Package {
+    fn from(package: AstPackage<'ast>) -> Self {
+        Package {
+            name: Identifier::from(package.name),
+            access: PackageAccess::from(package.access),
+            span: Span::from(package.span),
+        }
+    }
+}
+
+impl Package {
+    fn format(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{}.{}", self.name, self.access)
+    }
+}
+
+impl fmt::Display for Package {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        self.format(f)
+    }
+}
+
+impl fmt::Debug for Package {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        self.format(f)
+    }
+}
diff --git a/types/src/imports/package_access.rs b/types/src/imports/package_access.rs
new file mode 100644
index 0000000000..ba92029f9f
--- /dev/null
+++ b/types/src/imports/package_access.rs
@@ -0,0 +1,58 @@
+use crate::{ImportSymbol, Package, Span};
+use leo_ast::imports::PackageAccess as AstPackageAccess;
+
+use serde::{Deserialize, Serialize};
+use std::fmt;
+
+#[derive(Clone, Serialize, Deserialize)]
+pub enum PackageAccess {
+    Star(Span),
+    SubPackage(Box<Package>),
+    Symbol(ImportSymbol),
+    Multiple(Vec<PackageAccess>),
+}
+
+impl<'ast> From<AstPackageAccess<'ast>> for PackageAccess {
+    fn from(access: AstPackageAccess<'ast>) -> Self {
+        match access {
+            AstPackageAccess::Star(star) => PackageAccess::Star(Span::from(star.span)),
+            AstPackageAccess::SubPackage(package) => PackageAccess::SubPackage(Box::new(Package::from(*package))),
+            AstPackageAccess::Symbol(symbol) => PackageAccess::Symbol(ImportSymbol::from(symbol)),
+            AstPackageAccess::Multiple(accesses) => {
+                PackageAccess::Multiple(accesses.into_iter().map(|access| PackageAccess::from(access)).collect())
+            }
+        }
+    }
+}
+
+impl PackageAccess {
+    fn format(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            PackageAccess::Star(ref _span) => write!(f, ".*"),
+            PackageAccess::SubPackage(ref package) => write!(f, ".{}", package),
+            PackageAccess::Symbol(ref symbol) => write!(f, ".{}", symbol),
+            PackageAccess::Multiple(ref accesses) => {
+                write!(f, ".(")?;
+                for (i, access) in accesses.iter().enumerate() {
+                    write!(f, "{}", access)?;
+                    if i < accesses.len() - 1 {
+                        write!(f, ", ")?;
+                    }
+                }
+                write!(f, ")")
+            }
+        }
+    }
+}
+
+impl fmt::Debug for PackageAccess {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        self.format(f)
+    }
+}
+
+impl fmt::Display for PackageAccess {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        self.format(f)
+    }
+}
diff --git a/types/src/integer.rs b/types/src/integer.rs
index e83c5b8eab..68bb5f0b14 100644
--- a/types/src/integer.rs
+++ b/types/src/integer.rs
@@ -1,7 +1,6 @@
 //! Conversion of integer declarations to constraints in Leo.
 
 use crate::{errors::IntegerError, InputValue, IntegerType, Span};
-use leo_ast::{types::IntegerType as AstIntegerType, values::NumberValue};
 
 use snarkos_errors::gadgets::SynthesisError;
 use snarkos_models::{
@@ -11,17 +10,15 @@ use snarkos_models::{
         utilities::{
             alloc::AllocGadget,
             boolean::Boolean,
-            eq::{ConditionalEqGadget, EqGadget},
+            eq::{ConditionalEqGadget, EqGadget, EvaluateEqGadget},
             select::CondSelectGadget,
             uint::{UInt, UInt128, UInt16, UInt32, UInt64, UInt8},
         },
     },
 };
-
-use snarkos_models::gadgets::utilities::eq::EvaluateEqGadget;
 use std::fmt;
 
-/// An integer type enum wrapping the integer value. Used only in expressions.
+/// An integer type enum wrapping the integer value.
 #[derive(Debug, Clone, PartialEq, Eq, PartialOrd)]
 pub enum Integer {
     U8(UInt8),
@@ -31,33 +28,11 @@ pub enum Integer {
     U128(UInt128),
 }
 
-impl<'ast> Integer {
-    pub fn from(number: NumberValue<'ast>, _type: AstIntegerType) -> Self {
-        match _type {
-            AstIntegerType::U8Type(_u8) => {
-                Integer::U8(UInt8::constant(number.value.parse::<u8>().expect("unable to parse u8")))
-            }
-            AstIntegerType::U16Type(_u16) => Integer::U16(UInt16::constant(
-                number.value.parse::<u16>().expect("unable to parse u16"),
-            )),
-            AstIntegerType::U32Type(_u32) => Integer::U32(UInt32::constant(
-                number.value.parse::<u32>().expect("unable to parse integers.u32"),
-            )),
-            AstIntegerType::U64Type(_u64) => Integer::U64(UInt64::constant(
-                number.value.parse::<u64>().expect("unable to parse u64"),
-            )),
-            AstIntegerType::U128Type(_u128) => Integer::U128(UInt128::constant(
-                number.value.parse::<u128>().expect("unable to parse u128"),
-            )),
-        }
-    }
-
+impl Integer {
     pub fn from_implicit(number: String) -> Self {
         Integer::U128(UInt128::constant(number.parse::<u128>().expect("unable to parse u128")))
     }
-}
 
-impl Integer {
     pub fn new_constant(integer_type: &IntegerType, string: String, span: Span) -> Result<Self, IntegerError> {
         match integer_type {
             IntegerType::U8 => {
diff --git a/types/src/program.rs b/types/src/program.rs
index 94e8ad1c64..931766122d 100644
--- a/types/src/program.rs
+++ b/types/src/program.rs
@@ -4,10 +4,11 @@
 use crate::{Circuit, Function, FunctionInput, Identifier, Import, TestFunction};
 use leo_ast::files::File;
 
+use serde::{Deserialize, Serialize};
 use std::collections::HashMap;
 
 /// A simple program with statement expressions, program arguments and program returns.
-#[derive(Debug, Clone)]
+#[derive(Debug, Clone, Serialize, Deserialize)]
 pub struct Program {
     pub name: String,
     pub expected_inputs: Vec<FunctionInput>,
diff --git a/types/src/statements/conditional_nested_or_end_statement.rs b/types/src/statements/conditional_nested_or_end_statement.rs
index 2897e282dc..9b2f01a6db 100644
--- a/types/src/statements/conditional_nested_or_end_statement.rs
+++ b/types/src/statements/conditional_nested_or_end_statement.rs
@@ -1,9 +1,10 @@
 use crate::{ConditionalStatement, Statement};
 use leo_ast::statements::ConditionalNestedOrEndStatement as AstConditionalNestedOrEndStatement;
 
+use serde::{Deserialize, Serialize};
 use std::fmt;
 
-#[derive(Clone, PartialEq, Eq)]
+#[derive(Clone, PartialEq, Eq, Serialize, Deserialize)]
 pub enum ConditionalNestedOrEndStatement {
     Nested(Box<ConditionalStatement>),
     End(Vec<Statement>),
diff --git a/types/src/statements/conditional_statement.rs b/types/src/statements/conditional_statement.rs
index 9bde6e2ab5..1e3d82316f 100644
--- a/types/src/statements/conditional_statement.rs
+++ b/types/src/statements/conditional_statement.rs
@@ -1,9 +1,10 @@
 use crate::{ConditionalNestedOrEndStatement, Expression, Statement};
 use leo_ast::statements::ConditionalStatement as AstConditionalStatement;
 
+use serde::{Deserialize, Serialize};
 use std::fmt;
 
-#[derive(Clone, PartialEq, Eq)]
+#[derive(Clone, PartialEq, Eq, Serialize, Deserialize)]
 pub struct ConditionalStatement {
     pub condition: Expression,
     pub statements: Vec<Statement>,
diff --git a/types/src/statements/statement.rs b/types/src/statements/statement.rs
index 2929d548c4..172682a532 100644
--- a/types/src/statements/statement.rs
+++ b/types/src/statements/statement.rs
@@ -1,4 +1,4 @@
-use crate::{Assignee, ConditionalStatement, Declare, Expression, Identifier, Integer, Span, Variable};
+use crate::{Assignee, ConditionalStatement, Declare, Expression, Identifier, Span, Variable};
 use leo_ast::{
     operations::AssignOperation,
     statements::{
@@ -13,17 +13,18 @@ use leo_ast::{
     },
 };
 
+use serde::{Deserialize, Serialize};
 use std::fmt;
 
 /// Program statement that defines some action (or expression) to be carried out.
-#[derive(Clone, PartialEq, Eq)]
+#[derive(Clone, PartialEq, Eq, Serialize, Deserialize)]
 pub enum Statement {
     Return(Vec<Expression>, Span),
     Definition(Declare, Variable, Expression, Span),
     Assign(Assignee, Expression, Span),
     MultipleAssign(Vec<Variable>, Expression, Span),
     Conditional(ConditionalStatement, Span),
-    For(Identifier, Integer, Integer, Vec<Statement>, Span),
+    For(Identifier, Expression, Expression, Vec<Statement>, Span),
     AssertEq(Expression, Expression, Span),
     Expression(Expression, Span),
 }
@@ -149,21 +150,10 @@ impl<'ast> From<MultipleAssignmentStatement<'ast>> for Statement {
 
 impl<'ast> From<ForStatement<'ast>> for Statement {
     fn from(statement: ForStatement<'ast>) -> Self {
-        let from = match Expression::from(statement.start) {
-            Expression::Integer(number) => number,
-            Expression::Implicit(string, _span) => Integer::from_implicit(string),
-            expression => unimplemented!("Range bounds should be integers, found {}", expression),
-        };
-        let to = match Expression::from(statement.stop) {
-            Expression::Integer(number) => number,
-            Expression::Implicit(string, _span) => Integer::from_implicit(string),
-            expression => unimplemented!("Range bounds should be integers, found {}", expression),
-        };
-
         Statement::For(
             Identifier::from(statement.index),
-            from,
-            to,
+            Expression::from(statement.start),
+            Expression::from(statement.stop),
             statement
                 .statements
                 .into_iter()
diff --git a/types/src/types/integer_type.rs b/types/src/types/integer_type.rs
index ea1f4c95a8..0b6fdf6b84 100644
--- a/types/src/types/integer_type.rs
+++ b/types/src/types/integer_type.rs
@@ -1,9 +1,10 @@
 use leo_ast::types::IntegerType as AstIntegerType;
 
+use serde::{Deserialize, Serialize};
 use std::fmt;
 
 /// Explicit integer type
-#[derive(Clone, Debug, PartialEq, Eq)]
+#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
 pub enum IntegerType {
     U8,
     U16,
diff --git a/types/src/types/type_.rs b/types/src/types/type_.rs
index bed59bb394..710d1da7c0 100644
--- a/types/src/types/type_.rs
+++ b/types/src/types/type_.rs
@@ -1,10 +1,11 @@
 use crate::{Expression, Identifier, IntegerType};
 use leo_ast::types::{ArrayType, CircuitType, DataType, Type as AstType};
 
+use serde::{Deserialize, Serialize};
 use std::fmt;
 
 /// Explicit type used for defining a variable or expression type
-#[derive(Clone, Debug, PartialEq, Eq)]
+#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
 pub enum Type {
     IntegerType(IntegerType),
     Field,
