diff --git a/Cargo.toml b/Cargo.toml
index 2be47aacb0..287ec45584 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -112,6 +112,12 @@ version = "1.11.1"
 [workspace.dependencies.snarkvm]
 version = "=1.5.0"
 
+[workspace.dependencies.snarkvm-circuit]
+version = "=1.5.0"
+
+[workspace.dependencies.snarkvm-synthesizer-program]
+version = "=1.5.0"
+
 [workspace.dependencies.serde]
 version = "1.0.214"
 features = [ "derive", "rc" ]
@@ -275,6 +281,14 @@ opt-level = 2
 lto = "off"
 incremental = true
 
+# snarkVM has a lot of verbose output if debug assertions are on. They'll already be off
+# in release builds; let's turn them off in dev builds too. Theoretically we should be
+# able to turn them off only in snarkvm, like this:
+# [profile.dev.package.snarkvm]
+# debug-assertions = false
+# But that doesn't seem to work, so just do it here.
+debug-assertions = false
+
 [profile.test]
 opt-level = 2
 lto = "off"
diff --git a/interpreter/Cargo.toml b/interpreter/Cargo.toml
index f6392a6f06..3ac18815f6 100644
--- a/interpreter/Cargo.toml
+++ b/interpreter/Cargo.toml
@@ -22,10 +22,10 @@ rust-version = "1.85.1"
 workspace = true
 
 [dependencies.snarkvm-circuit]
-version = "1.0.0"
+workspace = true
 
 [dependencies.snarkvm-synthesizer-program]
-version = "1.0.0"
+workspace = true
 
 [dependencies.leo-ast]
 workspace = true
diff --git a/leo/cli/main.rs b/leo/cli/main.rs
index 5520e9af9d..4f9959e805 100644
--- a/leo/cli/main.rs
+++ b/leo/cli/main.rs
@@ -20,7 +20,6 @@ use leo_span::create_session_if_not_set_then;
 use clap::Parser;
 
 fn set_panic_hook() {
-    #[cfg(not(debug_assertions))]
     std::panic::set_hook({
         Box::new(move |e| {
             eprintln!("thread `{}` {}", std::thread::current().name().unwrap_or("<unnamed>"), e);
diff --git a/leo/package/src/lib.rs b/leo/package/src/lib.rs
index a827f129c4..f247718fd4 100644
--- a/leo/package/src/lib.rs
+++ b/leo/package/src/lib.rs
@@ -39,14 +39,16 @@
 //!
 //! Such a directory structure, together with a `.gitignore` file, may be created
 //! on the file system using `Package::initialize`.
-//! ```norun
-//! let path = Package::("my_package", "path/to/parent", NetworkName::TestnetV0, "http://localhost:3030")?;
+//! ```no_run
+//! # use leo_package::{NetworkName, Package};
+//! let path = Package::initialize("my_package", "path/to/parent", NetworkName::TestnetV0, "http://localhost:3030").unwrap();
 //! ```
 //!
 //! Given an existing directory with such a structure, a `Package` may be created from it with
 //! `Package::from_directory`:
-//! ```norun
-//! let package = Package::from_directory("path/to/package", "/home/me/.aleo")?;
+//! ```no_run
+//! # use leo_package::Package;
+//! let package = Package::from_directory("path/to/package", "/home/me/.aleo").unwrap();
 //! ```
 //! This will read the manifest and env file and keep their data in `package.manifest` and `package.env`.
 //! It will also process dependencies and store them in topological order in `package.programs`. This processing
