diff --git a/website/src/tabs/account/SignMessage.js b/website/src/tabs/account/SignMessage.js
index 8010405d..a29fecbd 100644
--- a/website/src/tabs/account/SignMessage.js
+++ b/website/src/tabs/account/SignMessage.js
@@ -2,13 +2,13 @@ import React, {useState} from "react";
 import {Card, Divider, Form, Input} from "antd";
 import {CopyButton} from "../../components/CopyButton";
 import {useAleoWASM} from "../../aleo-wasm-hook";
-import {stringToUint8Array} from "../../utils/Utils";
 
 export const SignMessage = () => {
     const [signingAccount, setSigningAccount] = useState(null);
     const [signingKey, setSigningKey] = useState(null);
     const [message, setMessage] = useState(null);
     const aleo = useAleoWASM();
+    const textEncoder = new TextEncoder();
 
     const onKeyChange = (event) => {
         setSigningAccount(null);
@@ -25,7 +25,7 @@ export const SignMessage = () => {
     }
     const signString = (str) => {
         if (str === "" | signingAccount === null) return;
-        return signingAccount.sign(stringToUint8Array(str)).to_string();
+        return signingAccount.sign(textEncoder.encode(str)).to_string();
     }
     const onMessageChange = (event) => {
         setMessage(event.target.value);
diff --git a/website/src/tabs/account/VerifyMessage.js b/website/src/tabs/account/VerifyMessage.js
index 2289598a..a048ffe0 100644
--- a/website/src/tabs/account/VerifyMessage.js
+++ b/website/src/tabs/account/VerifyMessage.js
@@ -1,7 +1,6 @@
 import React, {useEffect, useRef, useState} from "react";
-import {Alert, Card, Divider, Form, Input, Row} from "antd";
+import {Card, Form, Input} from "antd";
 import {useAleoWASM} from "../../aleo-wasm-hook";
-import {stringToUint8Array} from "../../utils/Utils";
 
 export const VerifyMessage = () => {
     const [inputAddress, setInputAddress] = useState(null);
@@ -10,11 +9,12 @@ export const VerifyMessage = () => {
     const [messageInput, setMessageInput] = useState(null);
     const didMount = useRef(false);
     const aleo = useAleoWASM();
+    const textEncoder = new TextEncoder();
 
     const attemptVerify = () => {
         if (messageInput !== null && signatureInput !== null && inputAddress !== null) {
             try {
-                let messageBytes = stringToUint8Array(messageInput);
+                let messageBytes = textEncoder.encode(messageInput);
                 let signature = aleo.Signature.from_string(signatureInput);
                 let isVerified = inputAddress.verify(messageBytes, signature);
                 setVerified(isVerified);
diff --git a/website/src/utils/Utils.js b/website/src/utils/Utils.js
deleted file mode 100644
index e94bc5c1..00000000
--- a/website/src/utils/Utils.js
+++ /dev/null
@@ -1,10 +0,0 @@
-function stringToUint8Array(str) {
-    let arr = [];
-    for (let i = 0, l = str.length; i < l; i ++) {
-        let hex = Number(str.charCodeAt(i)).toString(16);
-        arr.push(hex);
-    }
-    return new Uint8Array(arr);
-}
-
-export {stringToUint8Array}
\ No newline at end of file
