diff --git a/website/package.json b/website/package.json
index d5d328ca..fad0df66 100644
--- a/website/package.json
+++ b/website/package.json
@@ -5,9 +5,6 @@
   "dependencies": {
     "@aleohq/ui": "^0.0.9",
     "@ant-design/icons": "^4.4.0",
-    "@testing-library/jest-dom": "^5.11.4",
-    "@testing-library/react": "^11.1.0",
-    "@testing-library/user-event": "^12.1.10",
     "aleo-wasm": "file:../wasm/pkg",
     "antd": "^4.11.2",
     "copy-to-clipboard": "^3.3.1",
@@ -32,8 +29,7 @@
   },
   "eslintConfig": {
     "extends": [
-      "react-app",
-      "react-app/jest"
+      "react-app"
     ]
   },
   "browserslist": {
diff --git a/website/src/App.js b/website/src/App.js
index 728037f9..4d0424c1 100644
--- a/website/src/App.js
+++ b/website/src/App.js
@@ -18,11 +18,11 @@ function App() {
                 <Menu mode="horizontal" defaultSelectedKeys={['1']}>
                     <Menu.Item key="1" onClick={() => setMenuIndex(0)}>Account</Menu.Item>
                     <Menu.Item key="2" onClick={() => setMenuIndex(1)}>Record</Menu.Item>
-                    <Menu.Item key="3">
-                        <a href="./dev/bench" target="_blank" rel="noopener noreferrer">
-                            Benchmarks
-                        </a>
-                    </Menu.Item>
+                    {/*<Menu.Item key="3">*/}
+                    {/*    <a href="./dev/bench" target="_blank" rel="noopener noreferrer">*/}
+                    {/*        Benchmarks*/}
+                    {/*    </a>*/}
+                    {/*</Menu.Item>*/}
                 </Menu>
             </Header>
             <Content style={{padding: '50px 50px'}}>
