diff --git a/cli/helpers/updater.rs b/cli/helpers/updater.rs
index af7dd4ec..b0891738 100644
--- a/cli/helpers/updater.rs
+++ b/cli/helpers/updater.rs
@@ -80,11 +80,12 @@ impl Updater {
     }
 
     /// Display the CLI message.
+    #[allow(clippy::unnecessary_to_owned)]
     pub fn print_cli() -> String {
         if let Ok(latest_version) = Self::update_available() {
             let mut output = "🟢 A new version is available! Run".bold().green().to_string();
             output += &" `aleo update` ".bold().white();
-            output += &format!("to update to v{latest_version}.").bold().green();
+            output += &format!("to update to v{latest_version}.\n").bold().green().to_string();
             output
         } else {
             String::new()
diff --git a/cli/lib.rs b/cli/lib.rs
index 735cd16d..0560c833 100644
--- a/cli/lib.rs
+++ b/cli/lib.rs
@@ -23,8 +23,8 @@ pub mod commands;
 pub mod errors;
 pub mod helpers;
 
-pub(crate) type Network = snarkvm::prelude::Testnet3;
-pub(crate) type Aleo = snarkvm::circuit::AleoV0;
+pub type Network = snarkvm::prelude::Testnet3;
+pub type Aleo = snarkvm::circuit::AleoV0;
 
 #[cfg(feature = "account")]
 pub use aleo_account as account;
