diff --git a/.circleci/config.yml b/.circleci/config.yml
index f0e549dd..7bca8137 100644
--- a/.circleci/config.yml
+++ b/.circleci/config.yml
@@ -82,6 +82,15 @@ jobs:
           command: |
             yarn start
 
+  template-node-ts:
+    executor: rust-node
+    steps:
+      - setup-sdk
+      - run:
+          working_directory: create-aleo-app/template-node-ts
+          command: |
+            yarn start
+
   template-extension:
     executor: rust-node
     steps:
@@ -138,6 +147,9 @@ workflows:
       - template-node:
           requires:
             - sdk
+      - template-node-ts:
+          requires:
+            - sdk
       - template-extension:
           requires:
             - sdk
diff --git a/create-aleo-app/template-node-ts/README.md b/create-aleo-app/template-node-ts/README.md
new file mode 100644
index 00000000..e3c985d3
--- /dev/null
+++ b/create-aleo-app/template-node-ts/README.md
@@ -0,0 +1,5 @@
+# Aleo + Node.js + TypeScript
+
+`npm start`
+
+Recommend Node.js 20+ for best performance.
diff --git a/create-aleo-app/template-node-ts/_gitignore b/create-aleo-app/template-node-ts/_gitignore
new file mode 100644
index 00000000..a547bf36
--- /dev/null
+++ b/create-aleo-app/template-node-ts/_gitignore
@@ -0,0 +1,24 @@
+# Logs
+logs
+*.log
+npm-debug.log*
+yarn-debug.log*
+yarn-error.log*
+pnpm-debug.log*
+lerna-debug.log*
+
+node_modules
+dist
+dist-ssr
+*.local
+
+# Editor directories and files
+.vscode/*
+!.vscode/extensions.json
+.idea
+.DS_Store
+*.suo
+*.ntvs*
+*.njsproj
+*.sln
+*.sw?
diff --git a/create-aleo-app/template-node-ts/package.json b/create-aleo-app/template-node-ts/package.json
new file mode 100644
index 00000000..fd443e73
--- /dev/null
+++ b/create-aleo-app/template-node-ts/package.json
@@ -0,0 +1,19 @@
+{
+  "name": "node-ts-starter",
+  "private": true,
+  "version": "0.0.0",
+  "type": "module",
+  "scripts": {
+    "build": "rimraf dist/js && rollup --config",
+    "start": "npm run build && node dist/index.js"
+  },
+  "dependencies": {
+    "@aleohq/sdk": "^0.6.0"
+  },
+  "devDependencies": {
+    "rimraf": "^5.0.1",
+    "rollup": "^3.27.2",
+    "rollup-plugin-typescript2": "^0.36.0",
+    "typescript": "^5.2.2"
+  }
+}
diff --git a/create-aleo-app/template-node-ts/rollup.config.js b/create-aleo-app/template-node-ts/rollup.config.js
new file mode 100644
index 00000000..bebb2e39
--- /dev/null
+++ b/create-aleo-app/template-node-ts/rollup.config.js
@@ -0,0 +1,19 @@
+import typescript from "rollup-plugin-typescript2";
+
+export default {
+    input: {
+        index: "./src/index.ts",
+    },
+    output: {
+        dir: `dist`,
+        format: "es",
+        sourcemap: true,
+    },
+    external: ['@aleohq/sdk'],
+    plugins: [
+        typescript({
+            tsconfig: "tsconfig.json",
+            clean: true,
+        }),
+    ],
+};
diff --git a/create-aleo-app/template-node-ts/src/index.ts b/create-aleo-app/template-node-ts/src/index.ts
new file mode 100644
index 00000000..fe10b2fa
--- /dev/null
+++ b/create-aleo-app/template-node-ts/src/index.ts
@@ -0,0 +1,64 @@
+import {Account, initThreadPool, ProgramManager, AleoKeyProvider, AleoKeyProviderParams} from "@aleohq/sdk";
+
+await initThreadPool();
+
+const hello_hello_program =
+    "program hello_hello.aleo;\n" +
+    "\n" +
+    "function hello:\n" +
+    "    input r0 as u32.public;\n" +
+    "    input r1 as u32.private;\n" +
+    "    add r0 r1 into r2;\n" +
+    "    output r2 as u32.private;\n";
+
+async function localProgramExecution() {
+    const programManager = new ProgramManager(undefined, undefined, undefined);
+
+    // Create a temporary account for the execution of the program
+    const account = new Account();
+    programManager.setAccount(account);
+
+    // Create a key provider in order to re-use the same key for each execution
+    const keyProvider = new AleoKeyProvider();
+    keyProvider.useCache(true);
+    programManager.setKeyProvider(keyProvider);
+
+    // Pre-synthesize the program keys and then cache them in memory using key provider
+    const keyPair = await programManager.synthesizeKeys(hello_hello_program, "hello", ["1u32", "1u32"]);
+
+    if (keyPair instanceof Error) {
+        throw new Error(`Failed to synthesize keys: ${keyPair.message}`);
+    } else {
+        programManager.keyProvider.cacheKeys("hello_hello.aleo:hello", keyPair);
+    }
+
+    programManager.keyProvider.cacheKeys("hello_hello.aleo:hello", keyPair);
+
+    // Specify parameters for the key provider to use search for program keys. In particular specify the cache key
+    // that was used to cache the keys in the previous step.
+    const keyProviderParams = new AleoKeyProviderParams({cacheKey: "hello_hello.aleo:hello"});
+
+    // Execute once using the key provider params defined above. This will use the cached proving keys and make
+    // execution significantly faster.
+    let executionResponse = await programManager.executeOffline(
+        hello_hello_program,
+        "hello",
+        ["5u32", "5u32"],
+        true,
+        undefined,
+        keyProviderParams,
+    );
+    console.log("hello_hello/hello executed - result:", executionResponse.getOutputs());
+
+    // Verify the execution using the verifying key that was generated earlier.
+    if (programManager.verifyExecution(executionResponse)) {
+        console.log("hello_hello/hello execution verified!");
+    } else {
+        throw("Execution failed verification!");
+    }
+}
+
+const start = Date.now();
+console.log("Starting execute!");
+await localProgramExecution();
+console.log("Execute finished!", Date.now() - start);
diff --git a/create-aleo-app/template-node-ts/tsconfig.json b/create-aleo-app/template-node-ts/tsconfig.json
new file mode 100644
index 00000000..add1c57a
--- /dev/null
+++ b/create-aleo-app/template-node-ts/tsconfig.json
@@ -0,0 +1,15 @@
+{
+  "compilerOptions": {
+    /* Basic Options */
+    "target": "es2017",
+    "module": "esnext",
+
+    /* Module Resolution Options */
+    "moduleResolution": "node",
+    "esModuleInterop": true,
+
+    /* Advanced Options */
+    "skipLibCheck": true,
+    "forceConsistentCasingInFileNames": true
+  }
+}
\ No newline at end of file
diff --git a/sdk/src/network-client.ts b/sdk/src/network-client.ts
index 1420c9bf..bdc75d2b 100644
--- a/sdk/src/network-client.ts
+++ b/sdk/src/network-client.ts
@@ -8,7 +8,6 @@ import {
   PrivateKey,
   WasmTransaction,
   Transaction,
-  Transition,
   logAndThrow
 } from "./index";
 
@@ -603,15 +602,15 @@ class AleoNetworkClient {
   }
 
   /**
-   * Returns the transition id by its unique identifier
-   * @param {string} transition_id - The transition id to get
+   * Returns the transition ID of the transition corresponding to the ID of the input or output.
+   * @param {string} inputOrOutputID - ID of the input or output.
    *
    * @example
-   * const transition = networkClient.getTransitionId("2429232855236830926144356377868449890830704336664550203176918782554219952323field");
+   * const transitionId = networkClient.getTransitionId("2429232855236830926144356377868449890830704336664550203176918782554219952323field");
    */
-  async getTransitionId(transition_id: string): Promise<Transition | Error> {
+  async getTransitionId(inputOrOutputID: string): Promise<string | Error> {
     try {
-      return await this.fetchData<Transition>("/find/transitionID/" + transition_id);
+      return await this.fetchData<string>("/find/transitionID/" + inputOrOutputID);
     } catch (error) {
       throw new Error("Error fetching transition ID.");
     }
