diff --git a/Cargo.toml b/Cargo.toml
index 89d2a2ad..3f317624 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -44,6 +44,7 @@ thiserror = { version = "1.0" }
 wasm-bindgen = { version = "0.2.68" }
 
 [dev-dependencies]
+colored = { version = "2", features = ["no-color"] }
 criterion = { version = "0.3.1" }
 rusty-hook = { version = "0.11.2" }
 wasm-bindgen-test = { version = "0.3.18" }
diff --git a/tests/new.rs b/tests/new.rs
index fccaf6fe..7e1595ea 100644
--- a/tests/new.rs
+++ b/tests/new.rs
@@ -4,12 +4,10 @@ use aleo::{cli::Command, commands::parse};
 fn test_new() {
     let seed = Some(1231275789u64);
     let expected = r"
-        
   Private Key  APrivateKey1tvv5YV1dipNiku2My8jMkqpqCyYKvR5Jq4y2mtjw7s77Zpn
      View Key  AViewKey1m8gvywHKHKfUzZiLiLoHedcdHEjKwo5TWo6efz8gK7wF
       Address  aleo1faksgtpmculyzt6tgaq26fe4fgdjtwualyljjvfn2q6k42ydegzspfz9uh
-
-        ";
+";
 
     let actual = parse(Command::New { seed }).unwrap();
     assert_eq!(expected, actual);
