diff --git a/Cargo.toml b/Cargo.toml
index 3f317624..e14368ba 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -49,9 +49,6 @@ criterion = { version = "0.3.1" }
 rusty-hook = { version = "0.11.2" }
 wasm-bindgen-test = { version = "0.3.18" }
 
-[features]
-uncolored = ["colored/no-color"]
-
 [profile.release]
 opt-level = 3
 lto = "thin"
@@ -66,11 +63,13 @@ debug-assertions = false
 rpath = false
 
 [profile.dev]
-opt-level = 0
+opt-level = 3
+lto = "thin"
+incremental = true
 
 [profile.test]
 opt-level = 3
 lto = "thin"
 incremental = true
 debug = true
-debug-assertions = true
+debug-assertions = false
diff --git a/benches/account.rs b/benches/account.rs
index 23d42dd8..0f24c2f8 100644
--- a/benches/account.rs
+++ b/benches/account.rs
@@ -14,7 +14,7 @@
 // You should have received a copy of the GNU General Public License
 // along with the snarkOS library. If not, see <https://www.gnu.org/licenses/>.
 
-use snarkvm_toolkit::account::{Address, PrivateKey};
+use aleo::account::{Address, PrivateKey};
 
 use criterion::{criterion_group, criterion_main, Criterion};
 use rand::SeedableRng;
diff --git a/tests/mod.rs b/tests/mod.rs
deleted file mode 100644
index b8c25d56..00000000
--- a/tests/mod.rs
+++ /dev/null
@@ -1 +0,0 @@
-mod new;
diff --git a/tests/new.rs b/tests/new.rs
index 7e1595ea..0e6ba597 100644
--- a/tests/new.rs
+++ b/tests/new.rs
@@ -2,6 +2,13 @@ use aleo::{cli::Command, commands::parse};
 
 #[test]
 fn test_new() {
+    for _ in 0..3 {
+        assert!(parse(Command::New { seed: None }).is_ok());
+    }
+}
+
+#[test]
+fn test_new_seeded() {
     let seed = Some(1231275789u64);
     let expected = r"
   Private Key  APrivateKey1tvv5YV1dipNiku2My8jMkqpqCyYKvR5Jq4y2mtjw7s77Zpn
