diff --git a/website/src/index.jsx b/website/src/index.jsx
index f635882b..fd2eff28 100644
--- a/website/src/index.jsx
+++ b/website/src/index.jsx
@@ -8,7 +8,7 @@ import { init } from '@amplitude/analytics-browser';
 import { autocapturePlugin } from '@amplitude/plugin-autocapture-browser';
 
 // Initialize Amplitude with your API key
-const AMPLITUDE_API_KEY = import.meta.env.VITE_AMPLITUDE_API_KEY;
+const AMPLITUDE_API_KEY = process.env.VITE_AMPLITUDE_API_KEY;
 
 if (AMPLITUDE_API_KEY) {
   init(AMPLITUDE_API_KEY, {
diff --git a/website/webpack.config.js b/website/webpack.config.js
index ad29f70d..f2eec9a2 100644
--- a/website/webpack.config.js
+++ b/website/webpack.config.js
@@ -1,9 +1,17 @@
 import CopyPlugin from "copy-webpack-plugin";
 import TerserPlugin from "terser-webpack-plugin";
 import HtmlWebpackPlugin from "html-webpack-plugin";
-
+import webpack from "webpack";
+import dotenv from "dotenv";
 import path from "path";
 
+const env = dotenv.config().parsed || {};
+const envKeys = Object.keys(env).reduce((acc, key) => {
+  acc[`process.env.${key}`] = JSON.stringify(env[key]);
+  return acc;
+}, {});
+
+
 const appConfig = {
     mode: "production",
     entry: {
@@ -52,6 +60,7 @@ const appConfig = {
         new HtmlWebpackPlugin({
             template: "./index.html",
         }),
+        new webpack.DefinePlugin(envKeys),
     ],
     optimization: {
         minimize: true,
