diff --git a/sdk/rollup.test.js b/sdk/rollup.test.js
index 48770758..e282b31d 100644
--- a/sdk/rollup.test.js
+++ b/sdk/rollup.test.js
@@ -39,6 +39,7 @@ export default networks.map((network) => {
 
             // Used by the SDK
             "comlink",
+            `@provablehq/sdk/${network}.js`,
             `@provablehq/wasm/${network}.js`,
             "core-js/proposals/json-parse-with-source.js",
 
diff --git a/sdk/tests/account.test.ts b/sdk/tests/account.test.ts
index a0474758..8d9082ee 100644
--- a/sdk/tests/account.test.ts
+++ b/sdk/tests/account.test.ts
@@ -1,6 +1,6 @@
 import sinon from "sinon";
 import { expect } from "chai";
-import { Account, Address, PrivateKey, RecordCiphertext, ViewKey } from "../src/node";
+import { Account, Address, PrivateKey, RecordCiphertext, ViewKey } from "@provablehq/sdk/%%NETWORK%%.js";
 import { seed, message, beaconPrivateKeyString, beaconViewKeyString, beaconAddressString, recordCiphertextString, foreignCiphertextString, recordPlaintextString } from "./data/account-data";
 
 describe('Account', () => {
diff --git a/sdk/tests/arithmetic.test.ts b/sdk/tests/arithmetic.test.ts
index 34ccc1b3..57558e34 100644
--- a/sdk/tests/arithmetic.test.ts
+++ b/sdk/tests/arithmetic.test.ts
@@ -1,6 +1,6 @@
 import sinon from "sinon";
 import { expect } from "chai";
-import { Field, Scalar, Group} from "../src/node";
+import { Field, Scalar, Group} from "@provablehq/sdk/%%NETWORK%%.js";
 
 describe('Field and Group Arithmetic Tests', () => {
     afterEach(() => {
diff --git a/sdk/tests/key-provider.test.ts b/sdk/tests/key-provider.test.ts
index 50562abd..878b7860 100644
--- a/sdk/tests/key-provider.test.ts
+++ b/sdk/tests/key-provider.test.ts
@@ -1,5 +1,5 @@
 import { expect } from "chai";
-import {AleoKeyProvider, CachedKeyPair, CREDITS_PROGRAM_KEYS, FunctionKeyPair, OfflineKeyProvider, ProvingKey, VerifyingKey} from "../src/node";
+import {AleoKeyProvider, CachedKeyPair, CREDITS_PROGRAM_KEYS, FunctionKeyPair, OfflineKeyProvider, ProvingKey, VerifyingKey} from "@provablehq/sdk/%%NETWORK%%.js";
 
 describe('KeyProvider', () => {
     let keyProvider: AleoKeyProvider;
diff --git a/sdk/tests/network-client.integration.ts b/sdk/tests/network-client.integration.ts
index 1cacf3ad..3cd9b9e2 100644
--- a/sdk/tests/network-client.integration.ts
+++ b/sdk/tests/network-client.integration.ts
@@ -1,6 +1,6 @@
 import { expect } from "chai";
 import {beaconAddressString, beaconPrivateKeyString} from "./data/account-data";
-import {Account, AleoNetworkClient} from "../src/browser";
+import {Account, AleoNetworkClient} from "@provablehq/sdk/%%NETWORK%%.js";
 
 describe('NodeConnection', () => {
     let localApiClient: AleoNetworkClient;
diff --git a/sdk/tests/network-client.test.ts b/sdk/tests/network-client.test.ts
index 776b75e4..4a10a68e 100644
--- a/sdk/tests/network-client.test.ts
+++ b/sdk/tests/network-client.test.ts
@@ -7,10 +7,17 @@ import {
     TransactionObject,
     InputObject,
     OutputObject,
-    ExecutionObject
-} from "../src/node";
+    ExecutionObject,
+    DeploymentObject,
+    ExecutionJSON,
+    InputJSON,
+    OutputJSON,
+    Plaintext,
+    PlaintextObject,
+    Transition,
+    TransitionObject,
+} from "@provablehq/sdk/%%NETWORK%%.js";
 import { beaconPrivateKeyString } from "./data/account-data";
-import { DeploymentObject, ExecutionJSON, InputJSON, OutputJSON, Plaintext, PlaintextObject, Transition, TransitionObject } from "../src/node";
 
 async function catchError(f: () => Promise<any>): Promise<Error | null> {
     try {
diff --git a/sdk/tests/program-manager.test.ts b/sdk/tests/program-manager.test.ts
index f7033ab7..6ae4dc10 100644
--- a/sdk/tests/program-manager.test.ts
+++ b/sdk/tests/program-manager.test.ts
@@ -7,7 +7,7 @@ import {
     statePathRecordOwnerPrivateKey,
     stateRoot
 } from "./data/account-data";
-import { Account, ExecutionResponse, OfflineQuery, ProgramManager, RecordPlaintext } from "../src/node";
+import { Account, ExecutionResponse, OfflineQuery, ProgramManager, RecordPlaintext } from "@provablehq/sdk/%%NETWORK%%.js";
 
 describe('Program Manager', () => {
     const programManager = new ProgramManager("https://api.explorer.provable.com/v1", undefined, undefined);
diff --git a/sdk/tests/record-provider.integration.ts b/sdk/tests/record-provider.integration.ts
index 2ca23bae..571bb14c 100644
--- a/sdk/tests/record-provider.integration.ts
+++ b/sdk/tests/record-provider.integration.ts
@@ -1,6 +1,6 @@
 import { expect } from "chai";
 import { beaconPrivateKeyString } from "./data/account-data";
-import { Account, AleoNetworkClient, NetworkRecordProvider, RecordPlaintext } from "../src/node";
+import { Account, AleoNetworkClient, NetworkRecordProvider, RecordPlaintext } from "@provablehq/sdk/%%NETWORK%%.js";
 
 describe('RecordProvider', () => {
     let account: Account;
diff --git a/sdk/tests/record-provider.test.ts b/sdk/tests/record-provider.test.ts
index 8557ff96..f5e79ef3 100644
--- a/sdk/tests/record-provider.test.ts
+++ b/sdk/tests/record-provider.test.ts
@@ -1,5 +1,5 @@
 import { expect } from "chai";
-import {Account, AleoNetworkClient, BlockHeightSearch, NetworkRecordProvider} from "../src/node";
+import {Account, AleoNetworkClient, BlockHeightSearch, NetworkRecordProvider} from "@provablehq/sdk/%%NETWORK%%.js";
 import {beaconPrivateKeyString} from "./data/account-data";
 
 describe.skip('RecordProvider', () => {
diff --git a/sdk/tests/wasm.test.ts b/sdk/tests/wasm.test.ts
index d69ddbcf..c5aebd79 100644
--- a/sdk/tests/wasm.test.ts
+++ b/sdk/tests/wasm.test.ts
@@ -1,5 +1,5 @@
 import { expect } from "chai";
-import { Address, PrivateKey, ViewKey, Signature, RecordCiphertext, RecordPlaintext, PrivateKeyCiphertext } from "../src/node";
+import { Address, PrivateKey, ViewKey, Signature, RecordCiphertext, RecordPlaintext, PrivateKeyCiphertext } from "@provablehq/sdk/%%NETWORK%%.js";
 import {
     seed,
     message,
@@ -191,7 +191,7 @@ describe('WASM Objects', () => {
         });
 
         it('properly assesses equality and inequality', () => {
-            const viewKey1 = new ViewKey();
+            const viewKey1 = new (ViewKey as any)();
             const viewKey2 = ViewKey.from_string(viewKeyString);
             const viewKey3 = ViewKey.from_string(viewKeyString);
 
diff --git a/sdk/tsconfig.test.json b/sdk/tsconfig.test.json
index 45805ff8..ba3d43a0 100644
--- a/sdk/tsconfig.test.json
+++ b/sdk/tsconfig.test.json
@@ -2,5 +2,7 @@
   "extends": "./tsconfig.json",
   "compilerOptions": {
     "rootDir": "."
-  }
+  },
+  "include": ["tests"],
+  "exclude": ["node_modules", "src"],
 }
