diff --git a/.github/workflows/release.yml b/.github/workflows/release.yml
index 8f5bc7c2..76886267 100644
--- a/.github/workflows/release.yml
+++ b/.github/workflows/release.yml
@@ -124,7 +124,7 @@ jobs:
       - name: Zip
         run: |
           mkdir tempdir
-          mv target/aarch64-apple-darwin/release/leo tempdir
+          mv target/aarch64-apple-darwin/release/aleo tempdir
           cd tempdir
           zip -r aleo-${{ steps.get_version.outputs.version }}-aarch64-apple-darwin.zip aleo
           cd ..
diff --git a/cli/commands/mod.rs b/cli/commands/mod.rs
index 076c0dd4..2b2da0a6 100644
--- a/cli/commands/mod.rs
+++ b/cli/commands/mod.rs
@@ -68,7 +68,7 @@ pub enum Command {
     // Node(Node),
     #[clap(name = "run")]
     Run(Run),
-    #[clap(subcommand)]
+    #[clap(name = "update")]
     Update(Update),
 }
 
diff --git a/cli/commands/update.rs b/cli/commands/update.rs
index 065c7546..80e0d605 100644
--- a/cli/commands/update.rs
+++ b/cli/commands/update.rs
@@ -19,48 +19,43 @@ use crate::helpers::Updater;
 use anyhow::Result;
 use clap::Parser;
 
-/// Update Aleo.
+/// Update Aleo to the latest version
 #[derive(Debug, Parser)]
-pub enum Update {
-    /// Update Aleo to the latest version
-    Update {
-        /// Lists all available versions of Aleo
-        #[clap(short = 'l', long)]
-        list: bool,
-        /// Suppress outputs to terminal
-        #[clap(short = 'q', long)]
-        quiet: bool,
-    },
+pub struct Update {
+    /// Lists all available versions of Aleo
+    #[clap(short = 'l', long)]
+    list: bool,
+    /// Suppress outputs to terminal
+    #[clap(short = 'q', long)]
+    quiet: bool,
 }
 
 impl Update {
     pub fn parse(self) -> Result<String> {
-        match self {
-            Self::Update { list, quiet } => match list {
-                true => match Updater::show_available_releases() {
-                    Ok(output) => Ok(output),
-                    Err(error) => Ok(format!("Failed to list the available versions of Aleo\n{error}\n")),
-                },
-                false => {
-                    let result = Updater::update_to_latest_release(!quiet);
-                    if !quiet {
-                        match result {
-                            Ok(status) => {
-                                if status.uptodate() {
-                                    Ok("\nAleo is already on the latest version".to_string())
-                                } else if status.updated() {
-                                    Ok(format!("\nAleo has updated to version {}", status.version()))
-                                } else {
-                                    Ok(String::new())
-                                }
+        match self.list {
+            true => match Updater::show_available_releases() {
+                Ok(output) => Ok(output),
+                Err(error) => Ok(format!("Failed to list the available versions of Aleo\n{error}\n")),
+            },
+            false => {
+                let result = Updater::update_to_latest_release(!self.quiet);
+                if !self.quiet {
+                    match result {
+                        Ok(status) => {
+                            if status.uptodate() {
+                                Ok("\nAleo is already on the latest version".to_string())
+                            } else if status.updated() {
+                                Ok(format!("\nAleo has updated to version {}", status.version()))
+                            } else {
+                                Ok(String::new())
                             }
-                            Err(e) => Ok(format!("\nFailed to update Aleo to the latest version\n{e}\n")),
                         }
-                    } else {
-                        Ok(String::new())
+                        Err(e) => Ok(format!("\nFailed to update Aleo to the latest version\n{e}\n")),
                     }
+                } else {
+                    Ok(String::new())
                 }
-            },
+            }
         }
     }
 }
