diff --git a/.github/workflows/staging-website.yml b/.github/workflows/staging-website.yml
index 60ca848e..1877d4e1 100644
--- a/.github/workflows/staging-website.yml
+++ b/.github/workflows/staging-website.yml
@@ -22,7 +22,7 @@ jobs:
           override: true
           components: rustfmt, rust-src
 
-      - uses: actions/cache@v2
+      - uses: actions/cache@v4
         with:
           path: |
             ~/.cargo/registry
diff --git a/.github/workflows/test-website.yml b/.github/workflows/test-website.yml
index 5f05e4de..15890752 100644
--- a/.github/workflows/test-website.yml
+++ b/.github/workflows/test-website.yml
@@ -19,7 +19,7 @@ jobs:
           override: true
           components: rustfmt, rust-src
 
-      - uses: actions/cache@v2
+      - uses: actions/cache@v4
         with:
           path: |
             ~/.cargo/registry
diff --git a/.github/workflows/website.yml b/.github/workflows/website.yml
index ad4e5a62..54ac5dc2 100644
--- a/.github/workflows/website.yml
+++ b/.github/workflows/website.yml
@@ -22,7 +22,7 @@ jobs:
           override: true
           components: rustfmt, rust-src
 
-      - uses: actions/cache@v2
+      - uses: actions/cache@v4
         with:
           path: |
             ~/.cargo/registry
diff --git a/create-leo-app/README.md b/create-leo-app/README.md
index 268476ce..b9660d32 100644
--- a/create-leo-app/README.md
+++ b/create-leo-app/README.md
@@ -13,14 +13,14 @@ npm create leo-app@latest
 
 Then follow the prompts!
 
-You can also directly specify the project name and the template you want to use via additional command line options. For example, to scaffold an Aleo + React project, run:
+You can also directly specify the project name you want to use via additional command line options. For example, to scaffold a Leo project, run:
 
 ```bash
 # npm 6.x
-npm create leo-app@latest my-leo-app --template react
+npm create leo-app@latest my-leo-app
 
 # npm 7+, extra double-dash is needed:
-npm create leo-app@latest my-leo-app -- --template react
+npm create leo-app@latest my-leo-app
 ```
 
 Currently supported template presets include:
diff --git a/create-leo-app/template-react-leo/vite.config.js b/create-leo-app/template-react-leo/vite.config.js
index 24cd5e05..9c0aa495 100644
--- a/create-leo-app/template-react-leo/vite.config.js
+++ b/create-leo-app/template-react-leo/vite.config.js
@@ -4,8 +4,9 @@ import react from "@vitejs/plugin-react";
 // https://vitejs.dev/config/
 export default defineConfig({
   plugins: [react()],
+  assetsInclude: ['**/*.wasm'],
   optimizeDeps: {
-    exclude: ["@provablehq/wasm", "@provablehq/sdk"],
+    exclude: ["@provablehq/wasm"],
   },
   server: {
     headers: {
diff --git a/create-leo-app/template-react-managed-worker/vite.config.js b/create-leo-app/template-react-managed-worker/vite.config.js
index f7dfd3ca..ff78648f 100644
--- a/create-leo-app/template-react-managed-worker/vite.config.js
+++ b/create-leo-app/template-react-managed-worker/vite.config.js
@@ -3,9 +3,10 @@ import react from "@vitejs/plugin-react";
 
 // https://vitejs.dev/config/
 export default defineConfig({
+  assetsInclude: ['**/*.wasm'],
   plugins: [react()],
   optimizeDeps: {
-    exclude: ["@provablehq/wasm", "@provablehq/sdk"],
+    exclude: ["@provablehq/wasm"],
   },
   server: {
     // Needed if you are linking local packages for development
diff --git a/create-leo-app/template-react-ts/vite.config.ts b/create-leo-app/template-react-ts/vite.config.ts
index 24cd5e05..159a5c12 100644
--- a/create-leo-app/template-react-ts/vite.config.ts
+++ b/create-leo-app/template-react-ts/vite.config.ts
@@ -3,9 +3,10 @@ import react from "@vitejs/plugin-react";
 
 // https://vitejs.dev/config/
 export default defineConfig({
+  assetsInclude: ['**/*.wasm'],
   plugins: [react()],
   optimizeDeps: {
-    exclude: ["@provablehq/wasm", "@provablehq/sdk"],
+    exclude: ["@provablehq/wasm"],
   },
   server: {
     headers: {
diff --git a/create-leo-app/template-vanilla/vite.config.js b/create-leo-app/template-vanilla/vite.config.js
index 5e903a2a..f19a5503 100644
--- a/create-leo-app/template-vanilla/vite.config.js
+++ b/create-leo-app/template-vanilla/vite.config.js
@@ -1,7 +1,9 @@
 import { defineConfig } from "vite";
+
 export default defineConfig({
+    assetsInclude: ['**/*.wasm'],
     optimizeDeps: {
-        exclude: ["@provablehq/wasm", "@provablehq/sdk"],
+        exclude: ["@provablehq/wasm",],
     },
     server: {
         headers: {
diff --git a/website/vite.config.js b/website/vite.config.js
index cebc5c4b..8a5d8caa 100644
--- a/website/vite.config.js
+++ b/website/vite.config.js
@@ -3,6 +3,7 @@ import react from "@vitejs/plugin-react-swc";
 
 // https://vitejs.dev/config/
 export default defineConfig({
+assetsInclude: ['**/*.wasm'],
     worker: {
         format: "es",
     },
@@ -12,7 +13,7 @@ export default defineConfig({
         sourcemap: true,
     },
     optimizeDeps: {
-        exclude: ["@provablehq/wasm", "@provablehq/sdk"],
+        exclude: ["@provablehq/wasm"],
     },
     server: {
         fs: {
