diff --git a/sdk/src/record-provider.ts b/sdk/src/record-provider.ts
index dd17292a..acffadc7 100644
--- a/sdk/src/record-provider.ts
+++ b/sdk/src/record-provider.ts
@@ -193,7 +193,7 @@ class NetworkRecordProvider implements RecordProvider {
         let maxAmount = undefined;
 
         if (searchParameters) {
-            if ("startHeight" in searchParameters && typeof searchParameters["endHeight"] == "number") {
+            if ("startHeight" in searchParameters && typeof searchParameters["startHeight"] == "number") {
                 startHeight = searchParameters["startHeight"];
             }
 
@@ -290,7 +290,7 @@ class NetworkRecordProvider implements RecordProvider {
         let programs = undefined;
 
         if (searchParameters) {
-            if ("startHeight" in searchParameters && typeof searchParameters["endHeight"] == "number") {
+            if ("startHeight" in searchParameters && typeof searchParameters["startHeight"] == "number") {
                 startHeight = searchParameters["startHeight"];
             }
 
