diff --git a/static/app/views/alerts/rules/metric/eapMetricsField.tsx b/static/app/views/alerts/rules/metric/eapMetricsField.tsx
index 3acd7e9844e..51d779b2764 100644
--- a/static/app/views/alerts/rules/metric/eapMetricsField.tsx
+++ b/static/app/views/alerts/rules/metric/eapMetricsField.tsx
@@ -216,7 +216,7 @@ export default function EAPMetricsField({
           searchable
           options={isFetching ? previousOptions : (metricOptions ?? [])}
           value={traceMetricSelectValue}
-          loading={isFetching}
+          isLoading={isFetching}
           onInputChange={debouncedSetSearch}
           placeholder={t('Select a metric')}
           noOptionsMessage={() => t('No metrics found')}
