diff --git a/src/sentry/search/eap/trace_metrics/types.py b/src/sentry/search/eap/trace_metrics/types.py
index f28d4e00484eaf..4bc6ee1f68f8b7 100644
--- a/src/sentry/search/eap/trace_metrics/types.py
+++ b/src/sentry/search/eap/trace_metrics/types.py
@@ -7,6 +7,7 @@
     ComparisonFilter,
     ExistsFilter,
     NotFilter,
+    OrFilter,
     TraceItemFilter,
 )
 
@@ -45,12 +46,34 @@ def get_filter(self) -> TraceItemFilter:
 
         if self.metric_unit == NONE_UNIT:
             unit_key = AttributeKey(name="sentry.metric_unit", type=AttributeKey.Type.TYPE_STRING)
-            filters.append(
-                TraceItemFilter(
-                    not_filter=NotFilter(
-                        filters=[TraceItemFilter(exists_filter=ExistsFilter(key=unit_key))]
+
+            # In the case of NONE_UNIT, we want to select all items that do not have a unit, or
+            # have a unit explicitly set to "none"
+            filters.extend(
+                [
+                    TraceItemFilter(
+                        or_filter=OrFilter(
+                            filters=[
+                                TraceItemFilter(
+                                    not_filter=NotFilter(
+                                        filters=[
+                                            TraceItemFilter(
+                                                exists_filter=ExistsFilter(key=unit_key)
+                                            )
+                                        ]
+                                    )
+                                ),
+                                TraceItemFilter(
+                                    comparison_filter=ComparisonFilter(
+                                        key=unit_key,
+                                        op=ComparisonFilter.OP_EQUALS,
+                                        value=AttributeValue(val_str=self.metric_unit),
+                                    )
+                                ),
+                            ]
+                        )
                     )
-                )
+                ]
             )
         elif self.metric_unit is not None and self.metric_unit != ANY_UNIT:
             filters.append(
diff --git a/tests/snuba/api/endpoints/test_organization_events_trace_metrics.py b/tests/snuba/api/endpoints/test_organization_events_trace_metrics.py
index 1125304ac0146d..39f275f4b49523 100644
--- a/tests/snuba/api/endpoints/test_organization_events_trace_metrics.py
+++ b/tests/snuba/api/endpoints/test_organization_events_trace_metrics.py
@@ -726,3 +726,26 @@ def test_aggregate_with_none_unit_defined_only_selects_items_with_no_unit(self):
 
         assert len(data) == 1
         assert data[0]["count(value,request_duration,distribution,none)"] == 1
+
+    def test_aggregation_with_none_includes_items_with_none_explicitly_set(self):
+        trace_metrics = [
+            self.create_trace_metric(
+                "request_duration", 100.0, "distribution", metric_unit="millisecond"
+            ),
+            self.create_trace_metric("request_duration", 200.0, "distribution", metric_unit="none"),
+        ]
+        self.store_trace_metrics(trace_metrics)
+
+        response = self.do_request(
+            {
+                "field": ["count(value,request_duration,distribution,none)"],
+                "project": self.project.id,
+                "dataset": self.dataset,
+                "statsPeriod": "10m",
+            }
+        )
+        assert response.status_code == 200, response.content
+        data = response.data["data"]
+
+        assert len(data) == 1
+        assert data[0]["count(value,request_duration,distribution,none)"] == 1
