diff --git a/static/app/components/core/button/linkButton.tsx b/static/app/components/core/button/linkButton.tsx
index 407781bbac6..8e8bee79d8d 100644
--- a/static/app/components/core/button/linkButton.tsx
+++ b/static/app/components/core/button/linkButton.tsx
@@ -105,10 +105,6 @@ const StyledLinkButton = styled(
   }
 )<LinkButtonProps>`
   ${p => getLinkButtonStyles(p)}
-
-  &:focus-visible {
-    box-shadow: none;
-  }
 `;
 
 const getLinkButtonStyles = (p: LinkButtonProps) => {
