diff --git a/src/sentry/grouping/parameterization.py b/src/sentry/grouping/parameterization.py
index 7f2d2922f1e..563e2222aa8 100644
--- a/src/sentry/grouping/parameterization.py
+++ b/src/sentry/grouping/parameterization.py
@@ -57,6 +57,10 @@ DEFAULT_PARAMETERIZATION_REGEXES = [
     ParameterizationRegex(
         name="ip",
         raw_pattern=r"""
+            # This negative lookbehind ensures two things (depending on the pattern):
+            #     - We don't match starting in the middle of a valid set of initial characters
+            #     - We don't match things like `::` when they appear in expressions like `SomeClass::someMethod()`
+            (?<![0-9a-zA-Z_])
             (
                 ([0-9a-fA-F]{1,4}:){7,7}[0-9a-fA-F]{1,4}|
                 ([0-9a-fA-F]{1,4}:){1,7}:|
@@ -74,7 +78,12 @@ DEFAULT_PARAMETERIZATION_REGEXES = [
                 ([0-9a-fA-F]{1,4}:){1,4}:
                 ((25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9])\.){3,3}
                 (25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9])\b
-            ) |
+            )
+            # This negative lookahead works with the negative lookbehind above to block false
+            # positives on expressions of the form `SomeClass::someMethod()`, ensuring that even if
+            # the class name is valid hex, the method name being invalid will block the match
+            (?![0-9a-zA-Z])
+            |
             (
                 \b((25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9])\.){3,3}
                 (25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9])\b
diff --git a/tests/sentry/grouping/test_parameterization.py b/tests/sentry/grouping/test_parameterization.py
index 054f4c6365e..a379149f0df 100644
--- a/tests/sentry/grouping/test_parameterization.py
+++ b/tests/sentry/grouping/test_parameterization.py
@@ -24,6 +24,8 @@ standard_cases = [
     ("hostname - tld", "example.com", "<hostname>"),
     ("hostname - subdomain", "www.example.net", "<hostname>"),
     ("ip", "0.0.0.0", "<ip>"),
+    ("ip - double colon object property", "Option::unwrap()", "Option::unwrap()"),
+    ("ip - double colon object property including hex", "Bee::buzz()", "Bee::buzz()"),
     (
         "traceparent - header",
         "traceparent: 00-0af7651916cd43dd8448eb211c80319c-b7ad6b7169203331-01",
@@ -228,18 +230,6 @@ incorrect_cases = [
         "<int>",
         "<int>,<int>,<int>",
     ),
-    (
-        "ip - double colon object property",
-        "Option::unwrap()",
-        "Option::unwrap()",
-        "Option<ip>unwrap()",
-    ),
-    (
-        "ip - double colon object property including hex",
-        "Bee::buzz()",
-        "Bee::buzz()",
-        "<ip>buzz()",
-    ),
     (
         "json - double quotes",
         '{"dogs are great": true, "dog_id": "greatdog1231"}',
