diff --git a/static/app/components/modals/widgetViewerModal.tsx b/static/app/components/modals/widgetViewerModal.tsx
index 457295b43a2..57ca4bcefdf 100644
--- a/static/app/components/modals/widgetViewerModal.tsx
+++ b/static/app/components/modals/widgetViewerModal.tsx
@@ -136,6 +136,7 @@ export interface WidgetViewerModalOptions {
   seriesResultsType?: Record<string, AggregationOutputType>;
   tableData?: TableDataWithTitle[];
   totalIssuesCount?: string;
+  widgetInterval?: string;
 }
 
 interface Props extends ModalRenderProps, WidgetViewerModalOptions {
@@ -210,6 +211,7 @@ function WidgetViewerModal(props: Props) {
     dashboardCreator,
     confidence,
     sampleCount,
+    widgetInterval,
   } = props;
   const theme = useTheme();
   const location = useLocation();
@@ -560,6 +562,7 @@ function WidgetViewerModal(props: Props) {
             }
             cursor={cursor}
             dashboardFilters={dashboardFilters}
+            widgetInterval={widgetInterval}
           >
             {renderIssuesTable}
           </IssueWidgetQueries>
@@ -583,6 +586,7 @@ function WidgetViewerModal(props: Props) {
             }
             cursor={cursor}
             dashboardFilters={dashboardFilters}
+            widgetInterval={widgetInterval}
           >
             {renderTable}
           </ReleaseWidgetQueries>
@@ -607,6 +611,7 @@ function WidgetViewerModal(props: Props) {
             }
             cursor={cursor}
             dashboardFilters={dashboardFilters}
+            widgetInterval={widgetInterval}
           >
             {({tableResults, loading, pageLinks}) => {
               return renderTable({tableResults, loading, pageLinks});
@@ -676,6 +681,7 @@ function WidgetViewerModal(props: Props) {
                 noPadding
                 widgetLegendState={widgetLegendState}
                 showConfidenceWarning={widget.widgetType === WidgetType.SPANS}
+                widgetInterval={widgetInterval}
               />
             )}
           </Container>
diff --git a/static/app/views/dashboards/detail.tsx b/static/app/views/dashboards/detail.tsx
index a1661a3b771..48ac2d3f84d 100644
--- a/static/app/views/dashboards/detail.tsx
+++ b/static/app/views/dashboards/detail.tsx
@@ -308,6 +308,7 @@ class DashboardDetail extends Component<Props, State> {
           dashboardFilters: getDashboardFiltersFromURL(location) ?? dashboard.filters,
           dashboardPermissions: dashboard.permissions,
           dashboardCreator: dashboard.createdBy,
+          widgetInterval: this.props.widgetInterval,
           onClose: () => {
             // Filter out Widget Viewer Modal query params when exiting the Modal
             const query = omit(location.query, Object.values(WidgetViewerQueryField));
diff --git a/static/app/views/dashboards/widgetCard/widgetCardChartContainer.tsx b/static/app/views/dashboards/widgetCard/widgetCardChartContainer.tsx
index 88a49f64166..9bb8f35b8f6 100644
--- a/static/app/views/dashboards/widgetCard/widgetCardChartContainer.tsx
+++ b/static/app/views/dashboards/widgetCard/widgetCardChartContainer.tsx
@@ -57,6 +57,7 @@ type Props = {
   showConfidenceWarning?: boolean;
   showLoadingText?: boolean;
   tableItemLimit?: number;
+  widgetInterval?: string;
   windowWidth?: number;
 };
 
@@ -85,6 +86,7 @@ export function WidgetCardChartContainer({
   onWidgetTableSort,
   onWidgetTableResizeColumn,
   disableTableActions,
+  widgetInterval,
 }: Props) {
   const keepLegendState: EChartLegendSelectChangeHandler = ({selected}) => {
     widgetLegendState.setWidgetSelectionState(selected, widget);
@@ -118,6 +120,7 @@ export function WidgetCardChartContainer({
       onWidgetSplitDecision={onWidgetSplitDecision}
       onDataFetchStart={onDataFetchStart}
       tableItemLimit={tableItemLimit}
+      widgetInterval={widgetInterval}
     >
       {({
         tableResults,
