diff --git a/static/app/views/explore/constants.tsx b/static/app/views/explore/constants.tsx
index be6f97029b2ae9..2116ebe2911e96 100644
--- a/static/app/views/explore/constants.tsx
+++ b/static/app/views/explore/constants.tsx
@@ -93,11 +93,10 @@ export const SENTRY_PREPROD_STRING_TAGS: string[] = [
   'build_version',
   'git_base_ref',
   'git_head_ref',
-  'git_pr_number',
   'platform_name',
 ];
 
-export const SENTRY_PREPROD_NUMBER_TAGS: string[] = [];
+export const SENTRY_PREPROD_NUMBER_TAGS: string[] = ['git_pr_number'];
 
 export const SENTRY_PREPROD_BOOLEAN_TAGS: string[] = [];
 
