diff --git a/static/app/data/forms/projectIssueGrouping.tsx b/static/app/data/forms/projectIssueGrouping.tsx
index 9030dcf3380..ca6fd2c99bc 100644
--- a/static/app/data/forms/projectIssueGrouping.tsx
+++ b/static/app/data/forms/projectIssueGrouping.tsx
@@ -100,7 +100,7 @@ stack.function:mylibrary_* +app`}
   derivedGroupingEnhancements: {
     name: 'derivedGroupingEnhancements',
     type: 'string',
-    label: 'Derived Grouping Enhancements (super user only)',
+    label: 'Derived Grouping Enhancements',
     hideLabel: true,
     placeholder: '',
     multiline: true,
@@ -114,10 +114,9 @@ stack.function:mylibrary_* +app`}
     formatMessageValue: false,
     help: () => (
       <RuleDescription>
-        These rules are automatically derived for some languages for customers that have
-        the GitHub integration and the language has been marked to derive in-app rules.
-        These rules are not editable but they can be negated by adding their own rules in
-        the Stack Trace Rules section.
+        These rules are automatically derived for some languages for organizations that
+        have the GitHub integration. These rules are not editable but they can be negated
+        by adding you own rules in the Stack Trace Rules section.
       </RuleDescription>
     ),
     validate: () => [],
diff --git a/static/app/views/settings/projectIssueGrouping/index.spec.tsx b/static/app/views/settings/projectIssueGrouping/index.spec.tsx
index c7cdec65e54..9fcb7dd6626 100644
--- a/static/app/views/settings/projectIssueGrouping/index.spec.tsx
+++ b/static/app/views/settings/projectIssueGrouping/index.spec.tsx
@@ -60,9 +60,12 @@ describe('projectIssueGrouping', () => {
       outletContext: {project},
     });
 
-    // Verify the section is not visible for non-superuser
+    // Verify the section is visible for non-superuser
     expect(await screen.findByText('Issue Grouping')).toBeInTheDocument();
-    expect(screen.queryByText(/Derived Grouping Enhancements/)).not.toBeInTheDocument();
+    expect(screen.getByText(/Derived Grouping Enhancements/)).toBeInTheDocument();
+    expect(
+      screen.getByRole('textbox', {name: /Derived Grouping Enhancements/})
+    ).toBeDisabled();
 
     // Re-render for superuser
     jest.mocked(isActiveSuperuser).mockReturnValue(true);
diff --git a/static/app/views/settings/projectIssueGrouping/index.tsx b/static/app/views/settings/projectIssueGrouping/index.tsx
index aa00213cd99..08a56b0ae6b 100644
--- a/static/app/views/settings/projectIssueGrouping/index.tsx
+++ b/static/app/views/settings/projectIssueGrouping/index.tsx
@@ -12,7 +12,6 @@ import ProjectsStore from 'sentry/stores/projectsStore';
 import type {EventGroupingConfig} from 'sentry/types/event';
 import type {Project} from 'sentry/types/project';
 import getApiUrl from 'sentry/utils/api/getApiUrl';
-import {isActiveSuperuser} from 'sentry/utils/isActiveSuperuser';
 import {useApiQuery} from 'sentry/utils/queryClient';
 import routeTitleGen from 'sentry/utils/routeTitle';
 import useOrganization from 'sentry/utils/useOrganization';
@@ -56,7 +55,6 @@ export default function ProjectIssueGrouping() {
   const endpoint = `/projects/${organization.slug}/${project.slug}/`;
 
   const access = new Set(organization.access.concat(project.access));
-  const activeSuperUser = isActiveSuperuser();
   const hasAccess = hasEveryAccess(['project:write'], {organization, project});
 
   const jsonFormProps = {
@@ -106,14 +104,12 @@ export default function ProjectIssueGrouping() {
           fields={[fields.groupingEnhancements]}
         />
 
-        {activeSuperUser && (
-          <JsonForm
-            {...jsonFormProps}
-            title={t('Derived Grouping Enhancements')}
-            fields={[fields.derivedGroupingEnhancements]}
-            disabled
-          />
-        )}
+        <JsonForm
+          {...jsonFormProps}
+          title={t('Derived Grouping Enhancements')}
+          fields={[fields.derivedGroupingEnhancements]}
+          disabled
+        />
       </Form>
     </SentryDocumentTitle>
   );
