diff --git a/static/app/components/core/avatar/useAvatar.ts b/static/app/components/core/avatar/useAvatar.ts
index 1f23e14dbfcd5a..b88fbec0f7355d 100644
--- a/static/app/components/core/avatar/useAvatar.ts
+++ b/static/app/components/core/avatar/useAvatar.ts
@@ -158,7 +158,8 @@ async function hashGravatarId(gravatarId: string): Promise<string> {
  * the svg, etc) will also need to be changed there.
  */
 function getInitials(name: string | undefined): Tagged<string, '__avatar'> {
-  const sanitizedName = name?.trim();
+  const sanitizedName =
+    name === null || name === undefined ? undefined : String(name).trim();
 
   if (!sanitizedName) {
     return '?' as Tagged<string, '__avatar'>;
