diff --git a/static/app/views/dashboards/utils/usePopulateLinkedDashboards.tsx b/static/app/views/dashboards/utils/usePopulateLinkedDashboards.tsx
index bcbd717b425ffc..11dc5062a6ab47 100644
--- a/static/app/views/dashboards/utils/usePopulateLinkedDashboards.tsx
+++ b/static/app/views/dashboards/utils/usePopulateLinkedDashboards.tsx
@@ -45,7 +45,10 @@ const usePopulatePrebuiltIdsWithActualIds = (
         path: {organizationIdOrSlug: organization.slug},
       }),
       {
-        query: {prebuiltId: [...prebuiltIds.sort(), prebuiltId].filter(defined)},
+        query: {
+          prebuiltId: [...prebuiltIds.sort(), prebuiltId].filter(defined),
+          filter: 'showHidden',
+        },
       },
     ],
     {
