diff --git a/src/sentry/grouping/parameterization.py b/src/sentry/grouping/parameterization.py
index 10483064fa3..8c457ee30ef 100644
--- a/src/sentry/grouping/parameterization.py
+++ b/src/sentry/grouping/parameterization.py
@@ -142,7 +142,7 @@ DEFAULT_PARAMETERIZATION_REGEXES = [
             )
             |
             # Kitchen
-            ([1-9]\d?:\d{2}(:\d{2})?(?: [aApP][Mm])?)
+            ([1-9]\d?:\d{2}(:\d{2})?(?:\s?[aApP][Mm])?)
             |
             # Date
             (\d{4}-[01]\d-[0-3]\d)
diff --git a/tests/sentry/grouping/test_parameterization.py b/tests/sentry/grouping/test_parameterization.py
index 49157912bbd..75e20b57deb 100644
--- a/tests/sentry/grouping/test_parameterization.py
+++ b/tests/sentry/grouping/test_parameterization.py
@@ -76,7 +76,11 @@ standard_cases = [
     ("date - datetime compressed T-separated", "20060102T150405", "<date>"),
     ("date - datetime compressed T-separated UTC", "20060102T150405Z", "<date>"),
     ("date - datetime compressed T-separated w offset", "20060102T150405+0100", "<date>"),
-    ("date - kitchen", "3:04PM", "<date>"),
+    ("date - kitchen uppercase without space", "11:21PM", "<date>"),
+    ("date - kitchen uppercase with space", "12:31 PM", "<date>"),
+    ("date - kitchen lowercase without space", "11:21pm", "<date>"),
+    ("date - kitchen lowercase with space", "12:31 pm", "<date>"),
+    ("date - kitchen 24-hour", "23:21", "<date>"),
     ("date - time", "15:04:05", "<date>"),
     ("date - basic", "Mon Jan 02, 1999", "<date>"),
     ("date - datetime compressed date", "20240220 11:55:33.546593", "<date>"),
