diff --git a/static/app/views/dashboards/widgets/timeSeriesWidget/formatters/formatTimeSeriesLabel.spec.tsx b/static/app/views/dashboards/widgets/timeSeriesWidget/formatters/formatTimeSeriesLabel.spec.tsx
index 8d687b3b901..356eda1ece9 100644
--- a/static/app/views/dashboards/widgets/timeSeriesWidget/formatters/formatTimeSeriesLabel.spec.tsx
+++ b/static/app/views/dashboards/widgets/timeSeriesWidget/formatters/formatTimeSeriesLabel.spec.tsx
@@ -46,8 +46,8 @@ describe('formatSeriesName', () => {
 
   describe('aggregates of measurements', () => {
     it.each([
-      ['p75(measurements.lcp)', 'LCP'],
-      ['p50(measurements.lcp)', 'LCP'],
+      ['p75(measurements.lcp)', 'p75(measurements.lcp)'],
+      ['p50(measurements.lcp)', 'p50(measurements.lcp)'],
     ])('Formats %s as %s', (name, result) => {
       const timeSeries = TimeSeriesFixture({
         yAxis: name,
diff --git a/static/app/views/dashboards/widgets/timeSeriesWidget/formatters/formatTimeSeriesLabel.tsx b/static/app/views/dashboards/widgets/timeSeriesWidget/formatters/formatTimeSeriesLabel.tsx
index 4d09fc8c4a3..7075ab0cf83 100644
--- a/static/app/views/dashboards/widgets/timeSeriesWidget/formatters/formatTimeSeriesLabel.tsx
+++ b/static/app/views/dashboards/widgets/timeSeriesWidget/formatters/formatTimeSeriesLabel.tsx
@@ -1,10 +1,5 @@
 import {t} from 'sentry/locale';
-import {
-  getAggregateArg,
-  getMeasurementSlug,
-  maybeEquationAlias,
-  stripEquationPrefix,
-} from 'sentry/utils/discover/fields';
+import {maybeEquationAlias, stripEquationPrefix} from 'sentry/utils/discover/fields';
 import {formatVersion} from 'sentry/utils/versions/formatVersion';
 import WidgetLegendNameEncoderDecoder from 'sentry/views/dashboards/widgetLegendNameEncoderDecoder';
 import type {TimeSeries} from 'sentry/views/dashboards/widgets/common/types';
@@ -55,16 +50,6 @@ export function formatTimeSeriesLabel(timeSeries: TimeSeries): string {
   // `/events-timeseries` endpoint which does this automatically.
   seriesName = formatVersion(seriesName);
 
-  // Check for special-case measurement formatting
-  const arg = getAggregateArg(seriesName);
-  if (arg) {
-    const slug = getMeasurementSlug(arg);
-
-    if (slug) {
-      seriesName = slug.toUpperCase();
-    }
-  }
-
   // Strip equation prefix
   if (maybeEquationAlias(seriesName)) {
     seriesName = stripEquationPrefix(seriesName);
