diff --git a/node/router/tests/common/mod.rs b/node/router/tests/common/mod.rs
index 7aa86c054..8938e5b06 100644
--- a/node/router/tests/common/mod.rs
+++ b/node/router/tests/common/mod.rs
@@ -111,7 +111,7 @@ pub async fn validator(listening_port: u16, max_peers: u16) -> TestRouter<Curren
         sample_account(),
         &[],
         max_peers,
-        false,
+        true, // Router tests require validators to connect to peers.
         true,
     )
     .await
