diff --git a/.github/workflows/ci-branches.yml b/.github/workflows/ci-branches.yml
index 546e484c8..aed7da25b 100644
--- a/.github/workflows/ci-branches.yml
+++ b/.github/workflows/ci-branches.yml
@@ -39,7 +39,7 @@ jobs:
         uses: actions-rs/toolchain@v1
         with:
           profile: minimal
-          toolchain: nightly
+          toolchain: nightly-2020-07-13
           override: true
 
       - name: Test
diff --git a/algorithms/src/encryption/group.rs b/algorithms/src/encryption/group.rs
index d2e789e6f..2d370eb1f 100644
--- a/algorithms/src/encryption/group.rs
+++ b/algorithms/src/encryption/group.rs
@@ -29,10 +29,19 @@ impl<G: Group + ProjectiveCurve> FromBytes for GroupEncryptionPublicKey<G> {
     fn read<R: Read>(mut reader: R) -> IoResult<Self> {
         let x_coordinate = <G::Affine as AffineCurve>::BaseField::read(&mut reader)?;
 
-        match <G as ProjectiveCurve>::Affine::from_x_coordinate(x_coordinate, true) {
-            Some(element) => Ok(Self(element.into_projective())),
-            _ => Err(EncryptionError::Message("Failed to read encryption public key".into()).into()),
+        if let Some(element) = <G as ProjectiveCurve>::Affine::from_x_coordinate(x_coordinate, true) {
+            if element.is_in_correct_subgroup_assuming_on_curve() {
+                return Ok(Self(element.into_projective()));
+            }
         }
+
+        if let Some(element) = <G as ProjectiveCurve>::Affine::from_x_coordinate(x_coordinate, false) {
+            if element.is_in_correct_subgroup_assuming_on_curve() {
+                return Ok(Self(element.into_projective()));
+            }
+        }
+
+        Err(EncryptionError::Message("Failed to read encryption public key".into()).into())
     }
 }
 
diff --git a/algorithms/src/encryption/tests.rs b/algorithms/src/encryption/tests.rs
index 54da729f8..867f4f7a4 100644
--- a/algorithms/src/encryption/tests.rs
+++ b/algorithms/src/encryption/tests.rs
@@ -4,12 +4,15 @@ use snarkos_models::{
     algorithms::EncryptionScheme,
     curves::{Group, ProjectiveCurve},
 };
+use snarkos_utilities::{to_bytes, FromBytes, ToBytes};
 
 use rand::{Rng, SeedableRng};
 use rand_xorshift::XorShiftRng;
 
 type TestEncryptionScheme = GroupEncryption<EdwardsProjective>;
 
+pub const ITERATIONS: usize = 1000;
+
 fn generate_input<G: Group + ProjectiveCurve, R: Rng>(input_size: usize, rng: &mut R) -> Vec<G> {
     let mut input = vec![];
     for _ in 0..input_size {
@@ -36,3 +39,21 @@ fn simple_encryption() {
 
     assert_eq!(message, decrypted_message);
 }
+
+#[test]
+fn encryption_public_key_serialization() {
+    let rng = &mut XorShiftRng::seed_from_u64(1231275789u64);
+
+    let encryption_scheme = TestEncryptionScheme::setup(rng);
+
+    for _ in 0..ITERATIONS {
+        let private_key = encryption_scheme.generate_private_key(rng);
+        let public_key = encryption_scheme.generate_public_key(&private_key).unwrap();
+
+        let public_key_bytes = to_bytes![public_key].unwrap();
+        let recovered_public_key =
+            <TestEncryptionScheme as EncryptionScheme>::PublicKey::read(&public_key_bytes[..]).unwrap();
+
+        assert_eq!(public_key, recovered_public_key);
+    }
+}
