diff --git a/node/bft/events/src/challenge_response.rs b/node/bft/events/src/challenge_response.rs
index 0fc678a86..aad59f760 100644
--- a/node/bft/events/src/challenge_response.rs
+++ b/node/bft/events/src/challenge_response.rs
@@ -17,6 +17,7 @@ use super::*;
 #[derive(Clone, Debug, PartialEq, Eq)]
 pub struct ChallengeResponse<N: Network> {
     pub signature: Data<Signature<N>>,
+    pub nonce: u64,
 }
 
 impl<N: Network> EventTrait for ChallengeResponse<N> {
@@ -30,6 +31,7 @@ impl<N: Network> EventTrait for ChallengeResponse<N> {
 impl<N: Network> ToBytes for ChallengeResponse<N> {
     fn write_le<W: Write>(&self, mut writer: W) -> IoResult<()> {
         self.signature.write_le(&mut writer)?;
+        self.nonce.write_le(&mut writer)?;
         Ok(())
     }
 }
@@ -37,8 +39,9 @@ impl<N: Network> ToBytes for ChallengeResponse<N> {
 impl<N: Network> FromBytes for ChallengeResponse<N> {
     fn read_le<R: Read>(mut reader: R) -> IoResult<Self> {
         let signature = Data::read_le(&mut reader)?;
+        let nonce = u64::read_le(&mut reader)?;
 
-        Ok(Self { signature })
+        Ok(Self { signature, nonce })
     }
 }
 
@@ -53,7 +56,7 @@ pub mod prop_tests {
     };
 
     use bytes::{Buf, BufMut, BytesMut};
-    use proptest::prelude::{BoxedStrategy, Strategy};
+    use proptest::prelude::{any, BoxedStrategy, Strategy};
     use test_strategy::proptest;
 
     type CurrentNetwork = snarkvm::prelude::Testnet3;
@@ -70,7 +73,9 @@ pub mod prop_tests {
     }
 
     pub fn any_challenge_response() -> BoxedStrategy<ChallengeResponse<CurrentNetwork>> {
-        any_signature().prop_map(|sig| ChallengeResponse { signature: Data::Object(sig) }).boxed()
+        (any_signature(), any::<u64>())
+            .prop_map(|(sig, nonce)| ChallengeResponse { signature: Data::Object(sig), nonce })
+            .boxed()
     }
 
     #[proptest]
diff --git a/node/bft/events/src/lib.rs b/node/bft/events/src/lib.rs
index d49073b1c..7a3c6d30e 100644
--- a/node/bft/events/src/lib.rs
+++ b/node/bft/events/src/lib.rs
@@ -118,7 +118,7 @@ impl<N: Network> From<DisconnectReason> for Event<N> {
 
 impl<N: Network> Event<N> {
     /// The version of the event protocol; it can be incremented in order to force users to update.
-    pub const VERSION: u32 = 5;
+    pub const VERSION: u32 = 6;
 
     /// Returns the event name.
     #[inline]
diff --git a/node/bft/src/gateway.rs b/node/bft/src/gateway.rs
index 733b4ea3a..c635ee11a 100644
--- a/node/bft/src/gateway.rs
+++ b/node/bft/src/gateway.rs
@@ -1181,11 +1181,13 @@ impl<N: Network> Gateway<N> {
         /* Step 3: Send the challenge response. */
 
         // Sign the counterparty nonce.
-        let Ok(our_signature) = self.account.sign_bytes(&peer_request.nonce.to_le_bytes(), rng) else {
+        let response_nonce: u64 = rng.gen();
+        let data = [peer_request.nonce.to_le_bytes(), response_nonce.to_le_bytes()].concat();
+        let Ok(our_signature) = self.account.sign_bytes(&data, rng) else {
             return Err(error(format!("Failed to sign the challenge request nonce from '{peer_addr}'")));
         };
         // Send the challenge response.
-        let our_response = ChallengeResponse { signature: Data::Object(our_signature) };
+        let our_response = ChallengeResponse { signature: Data::Object(our_signature), nonce: response_nonce };
         send_event(&mut framed, peer_addr, Event::ChallengeResponse(our_response)).await?;
 
         // Add the peer to the gateway.
@@ -1234,11 +1236,13 @@ impl<N: Network> Gateway<N> {
         let rng = &mut rand::rngs::OsRng;
 
         // Sign the counterparty nonce.
-        let Ok(our_signature) = self.account.sign_bytes(&peer_request.nonce.to_le_bytes(), rng) else {
+        let response_nonce: u64 = rng.gen();
+        let data = [peer_request.nonce.to_le_bytes(), response_nonce.to_le_bytes()].concat();
+        let Ok(our_signature) = self.account.sign_bytes(&data, rng) else {
             return Err(error(format!("Failed to sign the challenge request nonce from '{peer_addr}'")));
         };
         // Send the challenge response.
-        let our_response = ChallengeResponse { signature: Data::Object(our_signature) };
+        let our_response = ChallengeResponse { signature: Data::Object(our_signature), nonce: response_nonce };
         send_event(&mut framed, peer_addr, Event::ChallengeResponse(our_response)).await?;
 
         // Sample a random nonce.
@@ -1295,14 +1299,14 @@ impl<N: Network> Gateway<N> {
         expected_nonce: u64,
     ) -> Option<DisconnectReason> {
         // Retrieve the components of the challenge response.
-        let ChallengeResponse { signature } = response;
+        let ChallengeResponse { signature, nonce } = response;
         // Perform the deferred non-blocking deserialization of the signature.
         let Ok(signature) = spawn_blocking!(signature.deserialize_blocking()) else {
             warn!("{CONTEXT} Gateway handshake with '{peer_addr}' failed (cannot deserialize the signature)");
             return Some(DisconnectReason::InvalidChallengeResponse);
         };
         // Verify the signature.
-        if !signature.verify_bytes(&peer_address, &expected_nonce.to_le_bytes()) {
+        if !signature.verify_bytes(&peer_address, &[expected_nonce.to_le_bytes(), nonce.to_le_bytes()].concat()) {
             warn!("{CONTEXT} Gateway handshake with '{peer_addr}' failed (invalid signature)");
             return Some(DisconnectReason::InvalidChallengeResponse);
         }
diff --git a/node/consensus/src/lib.rs b/node/consensus/src/lib.rs
index a5d8dd086..dd21abf31 100644
--- a/node/consensus/src/lib.rs
+++ b/node/consensus/src/lib.rs
@@ -93,10 +93,10 @@ impl<N: Network> Consensus<N> {
             bft,
             primary_sender: Default::default(),
             solutions_queue: Arc::new(Mutex::new(LruCache::new(
-                NonZeroUsize::new(MAX_TRANSMISSIONS_PER_BATCH).unwrap(),
+                NonZeroUsize::new(BatchHeader::<N>::MAX_TRANSMISSIONS_PER_BATCH).unwrap(),
             ))),
             transactions_queue: Arc::new(Mutex::new(LruCache::new(
-                NonZeroUsize::new(MAX_TRANSMISSIONS_PER_BATCH).unwrap(),
+                NonZeroUsize::new(BatchHeader::<N>::MAX_TRANSMISSIONS_PER_BATCH).unwrap(),
             ))),
             seen_solutions: Arc::new(Mutex::new(LruCache::new(NonZeroUsize::new(1 << 16).unwrap()))),
             seen_transactions: Arc::new(Mutex::new(LruCache::new(NonZeroUsize::new(1 << 16).unwrap()))),
diff --git a/node/router/messages/src/challenge_response.rs b/node/router/messages/src/challenge_response.rs
index 3c75d4db1..d51c5bb71 100644
--- a/node/router/messages/src/challenge_response.rs
+++ b/node/router/messages/src/challenge_response.rs
@@ -25,6 +25,7 @@ use std::borrow::Cow;
 pub struct ChallengeResponse<N: Network> {
     pub genesis_header: Header<N>,
     pub signature: Data<Signature<N>>,
+    pub nonce: u64,
 }
 
 impl<N: Network> MessageTrait for ChallengeResponse<N> {
@@ -38,13 +39,18 @@ impl<N: Network> MessageTrait for ChallengeResponse<N> {
 impl<N: Network> ToBytes for ChallengeResponse<N> {
     fn write_le<W: io::Write>(&self, mut writer: W) -> io::Result<()> {
         self.genesis_header.write_le(&mut writer)?;
-        self.signature.write_le(&mut writer)
+        self.signature.write_le(&mut writer)?;
+        self.nonce.write_le(&mut writer)
     }
 }
 
 impl<N: Network> FromBytes for ChallengeResponse<N> {
     fn read_le<R: io::Read>(mut reader: R) -> io::Result<Self> {
-        Ok(Self { genesis_header: Header::read_le(&mut reader)?, signature: Data::read_le(reader)? })
+        Ok(Self {
+            genesis_header: Header::read_le(&mut reader)?,
+            signature: Data::read_le(&mut reader)?,
+            nonce: u64::read_le(reader)?,
+        })
     }
 }
 
@@ -80,8 +86,12 @@ pub mod prop_tests {
     }
 
     pub fn any_challenge_response() -> BoxedStrategy<ChallengeResponse<CurrentNetwork>> {
-        (any_signature(), any_genesis_header())
-            .prop_map(|(sig, genesis_header)| ChallengeResponse { signature: Data::Object(sig), genesis_header })
+        (any_signature(), any_genesis_header(), any::<u64>())
+            .prop_map(|(sig, genesis_header, nonce)| ChallengeResponse {
+                signature: Data::Object(sig),
+                genesis_header,
+                nonce,
+            })
             .boxed()
     }
 
diff --git a/node/router/messages/src/lib.rs b/node/router/messages/src/lib.rs
index baa512b5b..09b065a49 100644
--- a/node/router/messages/src/lib.rs
+++ b/node/router/messages/src/lib.rs
@@ -111,7 +111,7 @@ impl<N: Network> From<DisconnectReason> for Message<N> {
 
 impl<N: Network> Message<N> {
     /// The version of the network protocol; it can be incremented in order to force users to update.
-    pub const VERSION: u32 = 13;
+    pub const VERSION: u32 = 14;
 
     /// Returns the message name.
     #[inline]
diff --git a/node/router/src/handshake.rs b/node/router/src/handshake.rs
index 7ccb67fa8..195298eae 100644
--- a/node/router/src/handshake.rs
+++ b/node/router/src/handshake.rs
@@ -164,12 +164,15 @@ impl<N: Network> Router<N> {
         }
         /* Step 3: Send the challenge response. */
 
+        let response_nonce: u64 = rng.gen();
+        let data = [peer_request.nonce.to_le_bytes(), response_nonce.to_le_bytes()].concat();
         // Sign the counterparty nonce.
-        let Ok(our_signature) = self.account.sign_bytes(&peer_request.nonce.to_le_bytes(), rng) else {
+        let Ok(our_signature) = self.account.sign_bytes(&data, rng) else {
             return Err(error(format!("Failed to sign the challenge request nonce from '{peer_addr}'")));
         };
         // Send the challenge response.
-        let our_response = ChallengeResponse { genesis_header, signature: Data::Object(our_signature) };
+        let our_response =
+            ChallengeResponse { genesis_header, signature: Data::Object(our_signature), nonce: response_nonce };
         send(&mut framed, peer_addr, Message::ChallengeResponse(our_response)).await?;
 
         // Add the peer to the router.
@@ -213,11 +216,14 @@ impl<N: Network> Router<N> {
         let rng = &mut OsRng;
 
         // Sign the counterparty nonce.
-        let Ok(our_signature) = self.account.sign_bytes(&peer_request.nonce.to_le_bytes(), rng) else {
+        let response_nonce: u64 = rng.gen();
+        let data = [peer_request.nonce.to_le_bytes(), response_nonce.to_le_bytes()].concat();
+        let Ok(our_signature) = self.account.sign_bytes(&data, rng) else {
             return Err(error(format!("Failed to sign the challenge request nonce from '{peer_addr}'")));
         };
         // Send the challenge response.
-        let our_response = ChallengeResponse { genesis_header, signature: Data::Object(our_signature) };
+        let our_response =
+            ChallengeResponse { genesis_header, signature: Data::Object(our_signature), nonce: response_nonce };
         send(&mut framed, peer_addr, Message::ChallengeResponse(our_response)).await?;
 
         // Sample a random nonce.
@@ -303,7 +309,7 @@ impl<N: Network> Router<N> {
         expected_nonce: u64,
     ) -> Option<DisconnectReason> {
         // Retrieve the components of the challenge response.
-        let ChallengeResponse { genesis_header, signature } = response;
+        let ChallengeResponse { genesis_header, signature, nonce } = response;
 
         // Verify the challenge response, by checking that the block header matches.
         if genesis_header != expected_genesis_header {
@@ -316,7 +322,7 @@ impl<N: Network> Router<N> {
             return Some(DisconnectReason::InvalidChallengeResponse);
         };
         // Verify the signature.
-        if !signature.verify_bytes(&peer_address, &expected_nonce.to_le_bytes()) {
+        if !signature.verify_bytes(&peer_address, &[expected_nonce.to_le_bytes(), nonce.to_le_bytes()].concat()) {
             warn!("Handshake with '{peer_addr}' failed (invalid signature)");
             return Some(DisconnectReason::InvalidChallengeResponse);
         }
diff --git a/node/tests/common/test_peer.rs b/node/tests/common/test_peer.rs
index d480c6a3e..8d9d5e39b 100644
--- a/node/tests/common/test_peer.rs
+++ b/node/tests/common/test_peer.rs
@@ -140,10 +140,13 @@ impl Handshake for TestPeer {
                 let peer_request = expect_message!(Message::ChallengeRequest, framed, peer_addr);
 
                 // Sign the nonce.
-                let signature = self.account().sign_bytes(&peer_request.nonce.to_le_bytes(), rng).unwrap();
+                let response_nonce: u64 = rng.gen();
+                let data = [peer_request.nonce.to_le_bytes(), response_nonce.to_le_bytes()].concat();
+                let signature = self.account().sign_bytes(&data, rng).unwrap();
 
                 // Send the challenge response.
-                let our_response = ChallengeResponse { genesis_header, signature: Data::Object(signature) };
+                let our_response =
+                    ChallengeResponse { genesis_header, signature: Data::Object(signature), nonce: response_nonce };
                 framed.send(Message::ChallengeResponse(our_response)).await?;
             }
             ConnectionSide::Responder => {
@@ -151,10 +154,13 @@ impl Handshake for TestPeer {
                 let peer_request = expect_message!(Message::ChallengeRequest, framed, peer_addr);
 
                 // Sign the nonce.
-                let signature = self.account().sign_bytes(&peer_request.nonce.to_le_bytes(), rng).unwrap();
+                let response_nonce: u64 = rng.gen();
+                let data = [peer_request.nonce.to_le_bytes(), response_nonce.to_le_bytes()].concat();
+                let signature = self.account().sign_bytes(&data, rng).unwrap();
 
                 // Send our challenge bundle.
-                let our_response = ChallengeResponse { genesis_header, signature: Data::Object(signature) };
+                let our_response =
+                    ChallengeResponse { genesis_header, signature: Data::Object(signature), nonce: response_nonce };
                 framed.send(Message::ChallengeResponse(our_response)).await?;
                 let our_request = ChallengeRequest::new(local_ip.port(), self.node_type(), self.address(), rng.gen());
                 framed.send(Message::ChallengeRequest(our_request)).await?;
