diff --git a/Cargo.lock b/Cargo.lock
index dbce87733..a49f2faca 100644
--- a/Cargo.lock
+++ b/Cargo.lock
@@ -1462,6 +1462,20 @@ dependencies = [
  "want",
 ]
 
+[[package]]
+name = "hyper-rustls"
+version = "0.24.2"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "ec3efd23720e2049821a693cbc7e65ea87c72f1c58ff2f9522ff332b1491e590"
+dependencies = [
+ "futures-util",
+ "http 0.2.12",
+ "hyper 0.14.30",
+ "rustls 0.21.12",
+ "tokio",
+ "tokio-rustls",
+]
+
 [[package]]
 name = "hyper-tls"
 version = "0.5.0"
@@ -2628,6 +2642,7 @@ dependencies = [
  "http 0.2.12",
  "http-body 0.4.6",
  "hyper 0.14.30",
+ "hyper-rustls",
  "hyper-tls 0.5.0",
  "ipnet",
  "js-sys",
@@ -2637,6 +2652,7 @@ dependencies = [
  "once_cell",
  "percent-encoding",
  "pin-project-lite",
+ "rustls 0.21.12",
  "rustls-pemfile 1.0.4",
  "serde",
  "serde_json",
@@ -2645,11 +2661,13 @@ dependencies = [
  "system-configuration",
  "tokio",
  "tokio-native-tls",
+ "tokio-rustls",
  "tower-service",
  "url",
  "wasm-bindgen",
  "wasm-bindgen-futures",
  "web-sys",
+ "webpki-roots 0.25.4",
  "winreg",
 ]
 
@@ -2752,6 +2770,18 @@ dependencies = [
  "windows-sys 0.52.0",
 ]
 
+[[package]]
+name = "rustls"
+version = "0.21.12"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "3f56a14d1f48b391359b22f731fd4bd7e43c97f3c50eee276f3aa09c94784d3e"
+dependencies = [
+ "log",
+ "ring",
+ "rustls-webpki 0.101.7",
+ "sct",
+]
+
 [[package]]
 name = "rustls"
 version = "0.23.12"
@@ -2762,7 +2792,7 @@ dependencies = [
  "once_cell",
  "ring",
  "rustls-pki-types",
- "rustls-webpki",
+ "rustls-webpki 0.102.8",
  "subtle",
  "zeroize",
 ]
@@ -2792,6 +2822,16 @@ version = "1.8.0"
 source = "registry+https://github.com/rust-lang/crates.io-index"
 checksum = "fc0a2ce646f8655401bb81e7927b812614bd5d91dbc968696be50603510fcaf0"
 
+[[package]]
+name = "rustls-webpki"
+version = "0.101.7"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "8b6275d1ee7a1cd780b64aca7726599a1dbc893b1e64144529e55c3c2f745765"
+dependencies = [
+ "ring",
+ "untrusted",
+]
+
 [[package]]
 name = "rustls-webpki"
 version = "0.102.8"
@@ -2863,6 +2903,16 @@ version = "1.2.0"
 source = "registry+https://github.com/rust-lang/crates.io-index"
 checksum = "94143f37725109f92c262ed2cf5e59bce7498c01bcc1502d7b9afe439a4e9f49"
 
+[[package]]
+name = "sct"
+version = "0.7.1"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "da046153aa2352493d6cb7da4b6e5c0c057d8a1d0a9aa8560baffdd945acd414"
+dependencies = [
+ "ring",
+ "untrusted",
+]
+
 [[package]]
 name = "security-framework"
 version = "2.11.1"
@@ -4794,6 +4844,16 @@ dependencies = [
  "tokio",
 ]
 
+[[package]]
+name = "tokio-rustls"
+version = "0.24.1"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "c28327cf380ac148141087fbfb9de9d7bd4e84ab5d2c28fbc911d753de8a7081"
+dependencies = [
+ "rustls 0.21.12",
+ "tokio",
+]
+
 [[package]]
 name = "tokio-stream"
 version = "0.1.16"
@@ -5140,12 +5200,12 @@ dependencies = [
  "flate2",
  "log",
  "once_cell",
- "rustls",
+ "rustls 0.23.12",
  "rustls-pki-types",
  "serde",
  "serde_json",
  "url",
- "webpki-roots",
+ "webpki-roots 0.26.5",
 ]
 
 [[package]]
@@ -5300,6 +5360,12 @@ dependencies = [
  "wasm-bindgen",
 ]
 
+[[package]]
+name = "webpki-roots"
+version = "0.25.4"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "5f20c57d8d7db6d3b86154206ae5d8fba62dd39573114de97c2cb0578251f8e1"
+
 [[package]]
 name = "webpki-roots"
 version = "0.26.5"
diff --git a/README.md b/README.md
index 075ef5901..c90b965b9 100644
--- a/README.md
+++ b/README.md
@@ -120,6 +120,15 @@ Please ensure ports `4130/tcp` and `3030/tcp` are open on your router and OS fir
 | 9000/tcp | TCP      | Allow      | Internal VPC or VPN          | Metrics export, should only be open within an internal VPC or VPN |
 | 9090/tcp | TCP      | Allow      | Internal VPC or VPN          | Prometheus metrics, should only be open within an internal VPC or VPN |
 
+**Note:** Ensure that your open file limit is set to 16,384 or above.
+For the recommended setting run:
+```
+# Increase the open file limit for the current user (replace <username> with your username)
+echo "<username> - nofile 65536" | sudo tee -a /etc/security/limits.conf
+# Increase the default system open file limit
+sudo bash -c 'echo "DefaultLimitNOFILE=65536" >> /etc/systemd/system.conf'
+```
+
 ## 3. Run an Aleo Node
 
 ## 3.1 Run an Aleo Client
diff --git a/cli/src/commands/developer/scan.rs b/cli/src/commands/developer/scan.rs
index 1b5ce0468..45a7af4c1 100644
--- a/cli/src/commands/developer/scan.rs
+++ b/cli/src/commands/developer/scan.rs
@@ -15,6 +15,7 @@
 
 #![allow(clippy::type_complexity)]
 
+use crate::commands::CDN_BASE_URL;
 use snarkvm::{
     console::network::{CanaryV0, MainnetV0, Network, TestnetV0},
     prelude::{Ciphertext, Field, FromBytes, Plaintext, PrivateKey, Record, ViewKey, block::Block},
@@ -31,8 +32,6 @@ use std::{
 use zeroize::Zeroize;
 
 const MAX_BLOCK_RANGE: u32 = 50;
-// TODO (raychu86): This should be configurable based on network.
-const CDN_ENDPOINT: &str = "https://s3.us-west-1.amazonaws.com/testnet3.blocks/phase3";
 
 /// Scan the snarkOS node for records.
 #[derive(Debug, Parser, Zeroize)]
@@ -172,6 +171,16 @@ impl Scan {
         }
     }
 
+    /// Returns the CDN to prefetch initial blocks from, from the given configurations.
+    fn parse_cdn<N: Network>() -> Result<String> {
+        match N::ID {
+            MainnetV0::ID => Ok(format!("{CDN_BASE_URL}/mainnet/v0")),
+            TestnetV0::ID => Ok(format!("{CDN_BASE_URL}/testnet/v0")),
+            CanaryV0::ID => Ok(format!("{CDN_BASE_URL}/canary/v0")),
+            _ => bail!("Unknown network ID ({})", N::ID),
+        }
+    }
+
     /// Fetch owned ciphertext records from the endpoint.
     fn fetch_records<N: Network>(
         private_key: Option<PrivateKey<N>>,
@@ -215,11 +224,13 @@ impl Scan {
         let mut request_start = match is_development_network {
             true => start_height,
             false => {
+                // Parse the CDN endpoint.
+                let cdn_endpoint = Self::parse_cdn::<N>()?;
                 // Scan the CDN first for records.
                 Self::scan_from_cdn(
                     start_height,
                     end_height,
-                    CDN_ENDPOINT.to_string(),
+                    cdn_endpoint,
                     endpoint.to_string(),
                     private_key,
                     *view_key,
diff --git a/cli/src/commands/start.rs b/cli/src/commands/start.rs
index b87a94dd2..b07a8c33a 100644
--- a/cli/src/commands/start.rs
+++ b/cli/src/commands/start.rs
@@ -59,7 +59,7 @@ const DEVELOPMENT_MODE_RNG_SEED: u64 = 1234567890u64;
 const DEVELOPMENT_MODE_NUM_GENESIS_COMMITTEE_MEMBERS: u16 = 4;
 
 /// The CDN base url.
-const CDN_BASE_URL: &str = "https://blocks.aleo.org";
+pub(crate) const CDN_BASE_URL: &str = "https://blocks.aleo.org";
 
 /// A mapping of `staker_address` to `(validator_address, withdrawal_address, amount)`.
 #[derive(Clone, Debug, PartialEq, Eq, Deserialize, Serialize)]
diff --git a/node/cdn/Cargo.toml b/node/cdn/Cargo.toml
index ccf1c3b47..7f8ebfffa 100644
--- a/node/cdn/Cargo.toml
+++ b/node/cdn/Cargo.toml
@@ -41,6 +41,7 @@ optional = true
 
 [dependencies.reqwest]
 version = "0.11"
+features = [ "rustls-tls" ]
 
 [dependencies.serde]
 version = "1"
diff --git a/node/cdn/src/blocks.rs b/node/cdn/src/blocks.rs
index bcd8e7a25..d147f3124 100644
--- a/node/cdn/src/blocks.rs
+++ b/node/cdn/src/blocks.rs
@@ -107,7 +107,7 @@ pub async fn load_blocks<N: Network>(
     process: impl FnMut(Block<N>) -> Result<()> + Clone + Send + Sync + 'static,
 ) -> Result<u32, (u32, anyhow::Error)> {
     // Create a Client to maintain a connection pool throughout the sync.
-    let client = match Client::builder().build() {
+    let client = match Client::builder().use_rustls_tls().build() {
         Ok(client) => client,
         Err(error) => {
             return Err((start_height.saturating_sub(1), anyhow!("Failed to create a CDN request client - {error}")));
@@ -439,7 +439,7 @@ mod tests {
 
     type CurrentNetwork = MainnetV0;
 
-    const TEST_BASE_URL: &str = "https://s3.us-west-1.amazonaws.com/testnet3.blocks/phase3";
+    const TEST_BASE_URL: &str = "https://blocks.aleo.org/mainnet/v0";
 
     fn check_load_blocks(start: u32, end: Option<u32>, expected: usize) {
         let blocks = Arc::new(RwLock::new(Vec::new()));
@@ -494,7 +494,7 @@ mod tests {
     #[test]
     fn test_cdn_height() {
         let rt = tokio::runtime::Runtime::new().unwrap();
-        let client = reqwest::Client::builder().build().unwrap();
+        let client = reqwest::Client::builder().use_rustls_tls().build().unwrap();
         rt.block_on(async {
             let height = cdn_height::<BLOCKS_PER_FILE>(&client, TEST_BASE_URL).await.unwrap();
             assert!(height > 0);
@@ -505,7 +505,7 @@ mod tests {
     fn test_cdn_get() {
         let rt = tokio::runtime::Runtime::new().unwrap();
         rt.block_on(async {
-            let client = reqwest::Client::new();
+            let client = reqwest::Client::builder().use_rustls_tls().build().unwrap();
             let height =
                 cdn_get::<u32>(client, &format!("{TEST_BASE_URL}/mainnet/latest/height"), "height").await.unwrap();
             assert!(height > 0);
diff --git a/node/rest/src/lib.rs b/node/rest/src/lib.rs
index 20314996a..17881a511 100644
--- a/node/rest/src/lib.rs
+++ b/node/rest/src/lib.rs
@@ -111,9 +111,18 @@ impl<N: Network, C: ConsensusStorage<N>, R: Routing<N>> Rest<N, C, R> {
         // Prepare the rate limiting setup.
         let governor_config = Box::new(
             GovernorConfigBuilder::default()
-                .per_second(1)
+                .per_nanosecond((1_000_000_000 / rest_rps) as u64)
                 .burst_size(rest_rps)
-                .error_handler(|error| Response::new(error.to_string().into()))
+                .error_handler(|error| {
+                    // Properly return a 429 Too Many Requests error
+                    let error_message = error.to_string();
+                    let mut response = Response::new(error_message.clone().into());
+                    *response.status_mut() = StatusCode::INTERNAL_SERVER_ERROR;
+                    if error_message.contains("Too Many Requests") {
+                        *response.status_mut() = StatusCode::TOO_MANY_REQUESTS;
+                    }
+                    response
+                })
                 .finish()
                 .expect("Couldn't set up rate limiting for the REST server!"),
         );
diff --git a/node/router/src/heartbeat.rs b/node/router/src/heartbeat.rs
index 63f0f0ddd..7b5477302 100644
--- a/node/router/src/heartbeat.rs
+++ b/node/router/src/heartbeat.rs
@@ -21,7 +21,7 @@ use crate::{
 use snarkvm::prelude::Network;
 
 use colored::Colorize;
-use rand::{prelude::IteratorRandom, rngs::OsRng, Rng};
+use rand::{Rng, prelude::IteratorRandom, rngs::OsRng};
 
 /// A helper function to compute the maximum of two numbers.
 /// See Rust issue 92391: https://github.com/rust-lang/rust/issues/92391.
