diff --git a/node/bft/events/src/block_response.rs b/node/bft/events/src/block_response.rs
index c2b5e1e99..57d7cd022 100644
--- a/node/bft/events/src/block_response.rs
+++ b/node/bft/events/src/block_response.rs
@@ -64,7 +64,7 @@ impl<N: Network> ToBytes for BlockResponse<N> {
         };
 
         // Send the consensus version starting with V12.
-        if latest_consensus_version > ConsensusVersion::V11 {
+        if latest_consensus_version >= ConsensusVersion::V12 {
             // Currently, we simply write four zero bytes as the version number,
             // because we know a valid request start height is always non-zero.
             // In the future we can encode the real version here.
@@ -216,7 +216,16 @@ pub mod prop_tests {
         let decoded = BlockResponse::<CurrentNetwork>::read_le(&mut bytes.into_inner().reader()).unwrap();
 
         assert_eq!(block_response.request, decoded.request);
-        assert_eq!(block_response.latest_consensus_version, decoded.latest_consensus_version);
+
+        // A block response will never contain a version below 12.
+        if let Some(vno) = block_response.latest_consensus_version
+            && vno < ConsensusVersion::V12
+        {
+            assert_eq!(decoded.latest_consensus_version, None)
+        } else {
+            assert_eq!(decoded.latest_consensus_version, block_response.latest_consensus_version);
+        }
+
         assert_eq!(
             block_response.blocks.deserialize_blocking().unwrap(),
             decoded.blocks.deserialize_blocking().unwrap(),
diff --git a/node/sync/src/block_sync.rs b/node/sync/src/block_sync.rs
index 2e1b35f35..c63f5a311 100644
--- a/node/sync/src/block_sync.rs
+++ b/node/sync/src/block_sync.rs
@@ -467,7 +467,7 @@ impl<N: Network> BlockSync<N> {
 
         // Perform consensus version check, if possible.
         // This check is only enabled after nodes have reached V12.
-        if expected_consensus_version > ConsensusVersion::V11 {
+        if expected_consensus_version >= ConsensusVersion::V12 {
             if let Some(latest_consensus_version) = latest_consensus_version {
                 ensure_equals!(
                     expected_consensus_version,
