diff --git a/Cargo.lock b/Cargo.lock
index bfd0ecdf4..84389bdd5 100644
--- a/Cargo.lock
+++ b/Cargo.lock
@@ -2186,9 +2186,9 @@ dependencies = [
 
 [[package]]
 name = "rustls-webpki"
-version = "0.100.1"
+version = "0.100.2"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "d6207cd5ed3d8dca7816f8f3725513a34609c0c765bf652b8c3cb4cfd87db46b"
+checksum = "e98ff011474fa39949b7e5c0428f9b4937eda7da7848bbb947786b7be0b27dab"
 dependencies = [
  "ring",
  "untrusted",
@@ -3789,7 +3789,7 @@ dependencies = [
  "log",
  "once_cell",
  "rustls",
- "rustls-webpki 0.100.1",
+ "rustls-webpki 0.100.2",
  "serde",
  "serde_json",
  "url",
@@ -3979,7 +3979,7 @@ version = "0.23.1"
 source = "registry+https://github.com/rust-lang/crates.io-index"
 checksum = "b03058f88386e5ff5310d9111d53f48b17d732b401aeb83a8d5190f2ac459338"
 dependencies = [
- "rustls-webpki 0.100.1",
+ "rustls-webpki 0.100.2",
 ]
 
 [[package]]
diff --git a/ledger/narwhal/batch-header/src/lib.rs b/ledger/narwhal/batch-header/src/lib.rs
index 334d4b3d1..abe394872 100644
--- a/ledger/narwhal/batch-header/src/lib.rs
+++ b/ledger/narwhal/batch-header/src/lib.rs
@@ -67,7 +67,7 @@ impl<N: Network> BatchHeader<N> {
         // Compute the batch ID.
         let batch_id = Self::compute_batch_id(author, round, timestamp, &transmission_ids, &previous_certificate_ids)?;
         // Sign the preimage.
-        let signature = private_key.sign(&[batch_id, Field::from_u64(timestamp as u64)], rng)?;
+        let signature = private_key.sign(&[batch_id], rng)?;
         // Return the batch header.
         Ok(Self { author, batch_id, round, timestamp, transmission_ids, previous_certificate_ids, signature })
     }
@@ -90,7 +90,7 @@ impl<N: Network> BatchHeader<N> {
         // Compute the batch ID.
         let batch_id = Self::compute_batch_id(author, round, timestamp, &transmission_ids, &previous_certificate_ids)?;
         // Verify the signature.
-        if !signature.verify(&author, &[batch_id, Field::from_u64(timestamp as u64)]) {
+        if !signature.verify(&author, &[batch_id]) {
             bail!("Invalid signature for the batch header");
         }
         // Return the batch header.
diff --git a/ledger/src/check_transaction_basic.rs b/ledger/src/check_transaction_basic.rs
index 10af079fb..e41880945 100644
--- a/ledger/src/check_transaction_basic.rs
+++ b/ledger/src/check_transaction_basic.rs
@@ -23,8 +23,8 @@ impl<N: Network, C: ConsensusStorage<N>> Ledger<N, C> {
 
         // If the transaction contains only 1 transition, and the transition is a split, then the fee can be skipped.
         let is_fee_required = match transaction.execution() {
-            Some(execution) => execution.len() == 1 && transaction.contains_split(),
-            None => false,
+            Some(execution) => !(execution.len() == 1 && transaction.contains_split()),
+            None => true,
         };
 
         if is_fee_required {
diff --git a/ledger/src/tests.rs b/ledger/src/tests.rs
index 74e0308a0..581fcc9d1 100644
--- a/ledger/src/tests.rs
+++ b/ledger/src/tests.rs
@@ -80,6 +80,114 @@ fn test_state_path() {
     let _state_path = ledger.get_state_path_for_commitment(commitment).unwrap();
 }
 
+#[test]
+fn test_insufficient_fees() {
+    let rng = &mut TestRng::default();
+
+    // Initialize the test environment.
+    let crate::test_helpers::TestEnv { ledger, private_key, view_key, address, .. } =
+        crate::test_helpers::sample_test_env(rng);
+
+    // A helper function to find records.
+    let find_records = || {
+        let microcredits = Identifier::from_str("microcredits").unwrap();
+        ledger
+            .find_records(&view_key, RecordsFilter::SlowUnspent(private_key))
+            .unwrap()
+            .filter(|(_, record)| match record.data().get(&microcredits) {
+                Some(Entry::Private(Plaintext::Literal(Literal::U64(amount), _))) => !amount.is_zero(),
+                _ => false,
+            })
+            .collect::<indexmap::IndexMap<_, _>>()
+    };
+
+    // Fetch the unspent records.
+    let records = find_records();
+    let record_1 = records[0].clone();
+    let record_2 = records[1].clone();
+
+    // Check fee amount requirements for `split` calls.
+    {
+        // Prepare a `split` execution without a fee.
+        let inputs = [Value::Record(record_1.clone()), Value::from_str("100u64").unwrap()];
+        let authorization =
+            ledger.vm.authorize(&private_key, "credits.aleo", "split", inputs.into_iter(), rng).unwrap();
+        let split_transaction_without_fee = ledger.vm.execute_authorization(authorization, None, None, rng).unwrap();
+        assert!(ledger.check_transaction_basic(&split_transaction_without_fee, None).is_ok());
+    }
+
+    // Check fee amount requirements for executions.
+    {
+        // Prepare an execution without a fee.
+        let inputs = [
+            Value::Record(record_1.clone()),
+            Value::from_str(&format!("{address}")).unwrap(),
+            Value::from_str("100u64").unwrap(),
+        ];
+        let authorization =
+            ledger.vm.authorize(&private_key, "credits.aleo", "transfer_private", inputs.into_iter(), rng).unwrap();
+        let transaction_without_fee = ledger.vm.execute_authorization(authorization, None, None, rng).unwrap();
+        let execution = transaction_without_fee.execution().unwrap();
+
+        // Check that a transaction with sufficient fee will succeed.
+        let fee_authorization = ledger
+            .vm
+            .authorize_fee_private(
+                &private_key,
+                record_2.clone(),
+                10_000_000,
+                execution.to_execution_id().unwrap(),
+                rng,
+            )
+            .unwrap();
+        let fee = ledger.vm.execute_fee_authorization(fee_authorization, None, rng).unwrap();
+        let sufficient_fee_transaction = Transaction::from_execution(execution.clone(), Some(fee)).unwrap();
+        assert!(ledger.check_transaction_basic(&sufficient_fee_transaction, None).is_ok());
+
+        // Check that a transaction with insufficient fee will fail.
+        let insufficient_fee_authorization = ledger
+            .vm
+            .authorize_fee_private(&private_key, record_2.clone(), 1, execution.to_execution_id().unwrap(), rng)
+            .unwrap();
+        let insufficient_fee = ledger.vm.execute_fee_authorization(insufficient_fee_authorization, None, rng).unwrap();
+        let insufficient_fee_transaction =
+            Transaction::from_execution(execution.clone(), Some(insufficient_fee)).unwrap();
+        assert!(ledger.check_transaction_basic(&insufficient_fee_transaction, None).is_err());
+    }
+
+    // Check fee amount requirements for deployment.
+
+    {
+        // Deploy a test program to the ledger.
+        let program = Program::<CurrentNetwork>::from_str(
+            r"
+program dummy.aleo;
+function foo:
+    input r0 as u8.private;
+    finalize r0;
+finalize foo:
+    input r0 as u8.public;
+    add r0 r0 into r1;",
+        )
+        .unwrap();
+
+        // Check that a deployment transaction with sufficient fee will succeed.
+        let transaction = ledger.vm.deploy(&private_key, &program, Some(record_2.clone()), 0, None, rng).unwrap();
+        assert!(ledger.check_transaction_basic(&transaction, None).is_ok());
+
+        // Check that a deployment transaction with insufficient fee will fail.
+        let deployment = transaction.deployment().unwrap();
+        let insufficient_fee_authorization = ledger
+            .vm
+            .authorize_fee_private(&private_key, record_2, 1, deployment.to_deployment_id().unwrap(), rng)
+            .unwrap();
+        let insufficient_fee = ledger.vm.execute_fee_authorization(insufficient_fee_authorization, None, rng).unwrap();
+        let insufficient_fee_transaction =
+            Transaction::from_deployment(*transaction.owner().unwrap(), deployment.clone(), insufficient_fee).unwrap();
+        assert!(ledger.check_transaction_basic(&insufficient_fee_transaction, None).is_err());
+    }
+}
+
 #[test]
 fn test_insufficient_finalize_fees() {
     let rng = &mut TestRng::default();
