diff --git a/console/network/src/canary_v0.rs b/console/network/src/canary_v0.rs
index ae7cae84e..2d7df0cdf 100644
--- a/console/network/src/canary_v0.rs
+++ b/console/network/src/canary_v0.rs
@@ -143,7 +143,7 @@ impl Network for CanaryV0 {
         (ConsensusVersion::V4, 5_730_000),
         (ConsensusVersion::V5, 5_780_000),
         (ConsensusVersion::V6, 6_240_000),
-        (ConsensusVersion::V7, 6_895_000),
+        (ConsensusVersion::V7, 6_880_000),
     ];
     /// A list of (consensus_version, block_height) pairs indicating when each consensus version takes effect.
     /// Documentation for what is changed at each version can be found in `ConsensusVersion`.
diff --git a/ledger/store/src/block/mod.rs b/ledger/store/src/block/mod.rs
index bb87f98e5..ca7f79160 100644
--- a/ledger/store/src/block/mod.rs
+++ b/ledger/store/src/block/mod.rs
@@ -1220,7 +1220,7 @@ impl<N: Network, B: BlockStorage<N>> BlockStore<N, B> {
 
     /// Returns the current block height.
     pub fn current_block_height(&self) -> u32 {
-        u32::try_from(self.tree.read().number_of_leaves()).unwrap() - 1
+        u32::try_from(self.tree.read().number_of_leaves()).unwrap().saturating_sub(1)
     }
 
     /// Returns the state root that contains the given `block height`.
@@ -1411,6 +1411,18 @@ mod tests {
 
     type CurrentNetwork = console::network::MainnetV0;
 
+    #[test]
+    fn test_current_block_height_empty() {
+        // Initialize a new block store.
+        let block_store = BlockStore::<CurrentNetwork, BlockMemory<_>>::open(StorageMode::new_test(None)).unwrap();
+
+        // Current_block_height shouldn't panic.
+        assert_eq!(block_store.current_block_height(), 0);
+
+        // Verify the equivalence of the alternative method.
+        assert_eq!(block_store.max_height().unwrap_or_default(), 0);
+    }
+
     #[test]
     fn test_insert_get_remove() {
         let rng = &mut TestRng::default();
