diff --git a/circuit/program/src/data/identifier/from_bits.rs b/circuit/program/src/data/identifier/from_bits.rs
index d321e7efa..a38d7756b 100644
--- a/circuit/program/src/data/identifier/from_bits.rs
+++ b/circuit/program/src/data/identifier/from_bits.rs
@@ -29,29 +29,17 @@ impl<A: Aleo> FromBits for Identifier<A> {
         // Recover the field element from the bits.
         let field = Field::from_bits_le(bits_le);
 
-        // Eject the bits in **little-endian** form, and convert the bits to bytes.
-        let bytes = bits_le
-            .to_bits_le()
-            .eject_value()
-            .chunks(8)
-            .map(|byte| match u8::from_bits_le(byte) {
-                Ok(byte) => byte,
-                Err(error) => A::halt(format!("Failed to recover an identifier from bits: {error}")),
-            })
-            .collect::<Vec<_>>();
-
-        // Recover the identifier length from the bits, by finding the first instance of a `0` byte,
-        // which is the null character '\0' in UTF-8, and an invalid character in an identifier.
-        let num_bytes = match bytes.iter().position(|&byte| byte == 0) {
-            Some(index) => index, // `index` is 0-indexed, and we exclude the null character.
-            None => bytes.len(),  // No null character found, so the identifier is the full length.
+        // Eject the bits in **little-endian** form, and determine the number of bytes.
+        let num_bytes = match console::Identifier::<A::Network>::from_bits_le(&bits_le.eject_value()) {
+            Ok(console_identifier) => console_identifier.size_in_bits() / 8,
+            Err(error) => A::halt(format!("Failed to recover an identifier from bits: {error}")),
         };
 
         // Ensure identifier fits within the data capacity of the base field.
         let max_bytes = A::BaseField::size_in_data_bits() / 8; // Note: This intentionally rounds down.
-        match num_bytes <= max_bytes {
+        match num_bytes as usize <= max_bytes {
             // Return the identifier.
-            true => Self(field, num_bytes as u8),
+            true => Self(field, num_bytes),
             false => A::halt("Identifier exceeds the maximum capacity allowed"),
         }
     }
diff --git a/circuit/program/src/data/identifier/mod.rs b/circuit/program/src/data/identifier/mod.rs
index 88b0194e9..8ad349d06 100644
--- a/circuit/program/src/data/identifier/mod.rs
+++ b/circuit/program/src/data/identifier/mod.rs
@@ -26,7 +26,7 @@ mod to_field;
 
 use snarkvm_circuit_network::Aleo;
 use snarkvm_circuit_types::{environment::prelude::*, Boolean, Field, U8};
-use snarkvm_utilities::{FromBits as FB, ToBits as TB};
+use snarkvm_utilities::ToBits as TB;
 
 /// An identifier is an **immutable** UTF-8 string,
 /// represented as a **constant** field element in the circuit.
diff --git a/console/program/src/data/identifier/from_bits.rs b/console/program/src/data/identifier/from_bits.rs
index 3f241fdf0..1f3485fc1 100644
--- a/console/program/src/data/identifier/from_bits.rs
+++ b/console/program/src/data/identifier/from_bits.rs
@@ -24,9 +24,6 @@ impl<N: Network> FromBits for Identifier<N> {
         // the final step checks the byte-aligned field element is within the data capacity.
         ensure!(bits_le.len() <= Field::<N>::size_in_bits(), "Identifier exceeds the maximum bits allowed");
 
-        // Recover the field element from the bits.
-        let field = Field::<N>::from_bits_le(bits_le)?;
-
         // Convert the bits to bytes, and parse the bytes as a UTF-8 string.
         let bytes = bits_le.chunks(8).map(u8::from_bits_le).collect::<Result<Vec<u8>>>()?;
 
@@ -37,13 +34,8 @@ impl<N: Network> FromBits for Identifier<N> {
             None => bytes.len(),  // No null character found, so the identifier is the full length.
         };
 
-        // Ensure identifier fits within the data capacity of the base field.
-        let max_bytes = Field::<N>::size_in_data_bits() / 8; // Note: This intentionally rounds down.
-        match num_bytes <= max_bytes {
-            // Return the identifier.
-            true => Ok(Self(field, u8::try_from(num_bytes).or_halt_with::<N>("Invalid identifier byte size"))),
-            false => bail!("Identifier exceeds the maximum capacity allowed"),
-        }
+        // Parse the bytes as a UTF-8 string.
+        Self::from_str(str::from_utf8(&bytes[0..num_bytes])?)
     }
 
     /// Initializes a new identifier from a list of big-endian bits *without* leading zeros.
