diff --git a/Cargo.lock b/Cargo.lock
index 93a76abe8..a0fd090f5 100644
--- a/Cargo.lock
+++ b/Cargo.lock
@@ -2231,6 +2231,7 @@ dependencies = [
  "csv",
  "derivative",
  "digest 0.8.1",
+ "itertools",
  "rand",
  "rand_chacha",
  "rand_xorshift",
@@ -2370,6 +2371,7 @@ dependencies = [
  "blake2",
  "derivative",
  "digest 0.8.1",
+ "itertools",
  "rand",
  "rand_xorshift",
  "snarkos-algorithms",
@@ -2405,6 +2407,7 @@ version = "0.8.0"
 dependencies = [
  "bincode",
  "derivative",
+ "itertools",
  "rand",
  "rand_xorshift",
  "serde",
diff --git a/algorithms/Cargo.toml b/algorithms/Cargo.toml
index ed91c37c2..3b09085f4 100644
--- a/algorithms/Cargo.toml
+++ b/algorithms/Cargo.toml
@@ -25,6 +25,7 @@ snarkos-utilities = { path = "../utilities", version = "0.8.0", default-features
 blake2 = { version = "0.8.1" }
 derivative = { version = "2" }
 digest = { version = "0.8.1" }
+itertools = { version = "0.9.0" }
 rand = { version = "0.7" }
 rand_chacha = { version = "0.2.2" }
 rayon = { version = "1" }
diff --git a/algorithms/src/signature/schnorr.rs b/algorithms/src/signature/schnorr.rs
index f39026be0..d17b92dbc 100644
--- a/algorithms/src/signature/schnorr.rs
+++ b/algorithms/src/signature/schnorr.rs
@@ -12,11 +12,11 @@ use snarkos_utilities::{
 };
 
 use digest::Digest;
+use itertools::Itertools;
 use rand::Rng;
 use std::{
     hash::Hash,
     io::{Read, Result as IoResult, Write},
-    marker::PhantomData,
 };
 
 pub fn bytes_to_bits(bytes: &[u8]) -> Vec<bool> {
@@ -116,19 +116,9 @@ where
 
     fn setup<R: Rng>(rng: &mut R) -> Result<Self, SignatureError> {
         let setup_time = start_timer!(|| "SchnorrSignature::setup");
-
-        let mut salt = [0u8; 32];
-        rng.fill_bytes(&mut salt);
-        let generator = G::rand(rng);
-
+        let parameters = Self::Parameters::setup(rng, Self::PrivateKey::size_in_bits());
         end_timer!(setup_time);
 
-        let parameters = SchnorrParameters {
-            _hash: PhantomData,
-            generator,
-            salt,
-        };
-
         Ok(Self { parameters })
     }
 
@@ -140,13 +130,21 @@ where
         let keygen_time = start_timer!(|| "SchnorrSignature::generate_private_key");
         let private_key = <G as Group>::ScalarField::rand(rng);
         end_timer!(keygen_time);
-
         Ok(private_key)
     }
 
     fn generate_public_key(&self, private_key: &Self::PrivateKey) -> Result<Self::PublicKey, SignatureError> {
         let keygen_time = start_timer!(|| "SchnorrSignature::generate_public_key");
-        let public_key = self.parameters.generator.mul(private_key);
+
+        let mut public_key = G::zero();
+        for (bit, base_power) in bytes_to_bits(&to_bytes![private_key]?)
+            .iter()
+            .zip_eq(&self.parameters.generator_powers)
+        {
+            if *bit {
+                public_key += &base_power;
+            }
+        }
         end_timer!(keygen_time);
 
         Ok(SchnorrPublicKey(public_key))
@@ -165,7 +163,15 @@ where
             let random_scalar: <G as Group>::ScalarField = <G as Group>::ScalarField::rand(rng);
             // Commit to the random scalar via r := k · g.
             // This is the prover's first msg in the Sigma protocol.
-            let prover_commitment: G = self.parameters.generator.mul(&random_scalar);
+            let mut prover_commitment = G::zero();
+            for (bit, base_power) in bytes_to_bits(&to_bytes![random_scalar]?)
+                .iter()
+                .zip_eq(&self.parameters.generator_powers)
+            {
+                if *bit {
+                    prover_commitment += &base_power;
+                }
+            }
 
             // Hash everything to get verifier challenge.
             let mut hash_input = Vec::new();
@@ -202,7 +208,17 @@ where
             prover_response,
             verifier_challenge,
         } = signature;
-        let mut claimed_prover_commitment = self.parameters.generator.mul(prover_response);
+
+        let mut claimed_prover_commitment = G::zero();
+        for (bit, base_power) in bytes_to_bits(&to_bytes![prover_response]?)
+            .iter()
+            .zip_eq(&self.parameters.generator_powers)
+        {
+            if *bit {
+                claimed_prover_commitment += &base_power;
+            }
+        }
+
         let public_key_times_verifier_challenge = public_key.0.mul(verifier_challenge);
         claimed_prover_commitment += &public_key_times_verifier_challenge;
 
@@ -230,13 +246,15 @@ where
         let rand_pk_time = start_timer!(|| "SchnorrSignature::randomize_public_key");
 
         let mut randomized_pk = public_key.0.clone();
-        let mut base = self.parameters.generator;
+
         let mut encoded = G::zero();
-        for bit in bytes_to_bits(randomness) {
-            if bit {
-                encoded += &base;
+        for (bit, base_power) in bytes_to_bits(&to_bytes![randomness]?)
+            .iter()
+            .zip_eq(&self.parameters.generator_powers)
+        {
+            if *bit {
+                encoded += &base_power;
             }
-            <G as Group>::double_in_place(&mut base);
         }
         randomized_pk += &encoded;
 
diff --git a/algorithms/src/signature/schnorr_parameters.rs b/algorithms/src/signature/schnorr_parameters.rs
index 73a021d9f..f8c340633 100644
--- a/algorithms/src/signature/schnorr_parameters.rs
+++ b/algorithms/src/signature/schnorr_parameters.rs
@@ -3,6 +3,7 @@ use snarkos_models::curves::{to_field_vec::ToConstraintField, Field, Group};
 use snarkos_utilities::bytes::{FromBytes, ToBytes};
 
 use digest::Digest;
+use rand::Rng;
 use std::{
     io::{Read, Result as IoResult, Write},
     marker::PhantomData,
@@ -16,14 +17,40 @@ use std::{
     Eq(bound = "G: Group, D: Digest")
 )]
 pub struct SchnorrParameters<G: Group, D: Digest> {
-    pub generator: G,
+    pub generator_powers: Vec<G>,
     pub salt: [u8; 32],
     pub _hash: PhantomData<D>,
 }
 
+impl<G: Group, D: Digest> SchnorrParameters<G, D> {
+    pub fn setup<R: Rng>(rng: &mut R, private_key_size_in_bits: usize) -> Self {
+        // Round to the closest multiple of 64 to factor bit and byte encoding differences.
+        assert!(private_key_size_in_bits < usize::MAX - 63);
+        let num_powers = (private_key_size_in_bits + 63) & !63usize;
+        Self {
+            generator_powers: Self::generator(num_powers, rng),
+            salt: rng.gen(),
+            _hash: PhantomData,
+        }
+    }
+
+    fn generator<R: Rng>(num_powers: usize, rng: &mut R) -> Vec<G> {
+        let mut generator_powers = Vec::with_capacity(num_powers);
+        let mut generator = G::rand(rng);
+        for _ in 0..num_powers {
+            generator_powers.push(generator);
+            generator.double_in_place();
+        }
+        generator_powers
+    }
+}
+
 impl<G: Group, D: Digest> ToBytes for SchnorrParameters<G, D> {
     fn write<W: Write>(&self, mut writer: W) -> IoResult<()> {
-        self.generator.write(&mut writer)?;
+        (self.generator_powers.len() as u32).write(&mut writer)?;
+        for g in &self.generator_powers {
+            g.write(&mut writer)?;
+        }
         self.salt.write(&mut writer)
     }
 }
@@ -31,11 +58,17 @@ impl<G: Group, D: Digest> ToBytes for SchnorrParameters<G, D> {
 impl<G: Group, D: Digest> FromBytes for SchnorrParameters<G, D> {
     #[inline]
     fn read<R: Read>(mut reader: R) -> IoResult<Self> {
-        let generator: G = FromBytes::read(&mut reader)?;
+        let mut generator_powers = vec![];
+        let generator_powers_length: u32 = FromBytes::read(&mut reader)?;
+        for _ in 0..generator_powers_length {
+            let g: G = FromBytes::read(&mut reader)?;
+            generator_powers.push(g);
+        }
+
         let salt: [u8; 32] = FromBytes::read(&mut reader)?;
 
         Ok(Self {
-            generator,
+            generator_powers,
             salt,
             _hash: PhantomData,
         })
@@ -45,6 +78,6 @@ impl<G: Group, D: Digest> FromBytes for SchnorrParameters<G, D> {
 impl<F: Field, G: Group + ToConstraintField<F>, D: Digest> ToConstraintField<F> for SchnorrParameters<G, D> {
     #[inline]
     fn to_field_elements(&self) -> Result<Vec<F>, ConstraintFieldError> {
-        self.generator.to_field_elements()
+        Ok(Vec::new())
     }
 }
diff --git a/gadgets/Cargo.toml b/gadgets/Cargo.toml
index 0d079af1c..5f3a73382 100644
--- a/gadgets/Cargo.toml
+++ b/gadgets/Cargo.toml
@@ -20,6 +20,7 @@ snarkos-utilities = { path = "../utilities", version = "0.8.0", default-features
 
 derivative = { version = "2" }
 digest = { version = "0.8.1" }
+itertools = { version = "0.9.0" }
 
 [dev-dependencies]
 blake2 = { version = "0.8.1" }
diff --git a/gadgets/src/algorithms/signature/schnorr.rs b/gadgets/src/algorithms/signature/schnorr.rs
index 733b1a7d5..03cddd1ab 100644
--- a/gadgets/src/algorithms/signature/schnorr.rs
+++ b/gadgets/src/algorithms/signature/schnorr.rs
@@ -18,23 +18,25 @@ use snarkos_models::{
 use snarkos_utilities::serialize::{CanonicalDeserialize, CanonicalSerialize};
 
 use digest::Digest;
+use itertools::Itertools;
 use std::{borrow::Borrow, marker::PhantomData};
 
-pub struct SchnorrParametersGadget<G: Group, F: Field, GG: GroupGadget<G, F>> {
-    generator: GG,
+#[derive(Clone)]
+pub struct SchnorrParametersGadget<G: Group, F: Field, D: Digest> {
+    parameters: SchnorrParameters<G, D>,
     _group: PhantomData<*const G>,
     _engine: PhantomData<*const F>,
 }
 
-impl<G: Group, F: Field, GG: GroupGadget<G, F>, D: Digest> AllocGadget<SchnorrParameters<G, D>, F>
-    for SchnorrParametersGadget<G, F, GG>
-{
+impl<G: Group, F: Field, D: Digest> AllocGadget<SchnorrParameters<G, D>, F> for SchnorrParametersGadget<G, F, D> {
     fn alloc<Fn: FnOnce() -> Result<T, SynthesisError>, T: Borrow<SchnorrParameters<G, D>>, CS: ConstraintSystem<F>>(
-        cs: CS,
-        f: Fn,
+        _cs: CS,
+        value_gen: Fn,
     ) -> Result<Self, SynthesisError> {
+        let value = value_gen()?;
+        let parameters = value.borrow().clone();
         Ok(Self {
-            generator: GG::alloc_checked(cs, || f().map(|pp| pp.borrow().generator))?,
+            parameters,
             _engine: PhantomData,
             _group: PhantomData,
         })
@@ -45,27 +47,19 @@ impl<G: Group, F: Field, GG: GroupGadget<G, F>, D: Digest> AllocGadget<SchnorrPa
         T: Borrow<SchnorrParameters<G, D>>,
         CS: ConstraintSystem<F>,
     >(
-        cs: CS,
-        f: Fn,
+        _cs: CS,
+        value_gen: Fn,
     ) -> Result<Self, SynthesisError> {
+        let value = value_gen()?;
+        let parameters = value.borrow().clone();
         Ok(Self {
-            generator: GG::alloc_input(cs, || f().map(|pp| pp.borrow().generator))?,
+            parameters,
             _engine: PhantomData,
             _group: PhantomData,
         })
     }
 }
 
-impl<G: Group, F: Field, GG: GroupGadget<G, F>> Clone for SchnorrParametersGadget<G, F, GG> {
-    fn clone(&self) -> Self {
-        Self {
-            generator: self.generator.clone(),
-            _group: PhantomData,
-            _engine: PhantomData,
-        }
-    }
-}
-
 #[derive(Debug, Clone, PartialEq, Eq)]
 pub struct SchnorrPublicKeyGadget<G: Group, F: Field, GG: GroupGadget<G, F>> {
     public_key: GG,
@@ -145,7 +139,7 @@ pub struct SchnorrPublicKeyRandomizationGadget<G: Group, F: Field, GG: GroupGadg
 impl<G: Group + CanonicalSerialize + CanonicalDeserialize, GG: GroupGadget<G, F>, D: Digest + Send + Sync, F: Field>
     SignaturePublicKeyRandomizationGadget<SchnorrSignature<G, D>, F> for SchnorrPublicKeyRandomizationGadget<G, F, GG>
 {
-    type ParametersGadget = SchnorrParametersGadget<G, F, GG>;
+    type ParametersGadget = SchnorrParametersGadget<G, F, D>;
     type PublicKeyGadget = SchnorrPublicKeyGadget<G, F, GG>;
 
     fn check_randomization_gadget<CS: ConstraintSystem<F>>(
@@ -154,13 +148,13 @@ impl<G: Group + CanonicalSerialize + CanonicalDeserialize, GG: GroupGadget<G, F>
         public_key: &Self::PublicKeyGadget,
         randomness: &[UInt8],
     ) -> Result<Self::PublicKeyGadget, SynthesisError> {
-        let base = parameters.generator.clone();
         let randomness = randomness.iter().flat_map(|b| b.to_bits_le()).collect::<Vec<_>>();
-        let rand_pk = base.mul_bits(
-            &mut cs.ns(|| "check_randomization_gadget"),
-            &public_key.public_key,
-            randomness.iter(),
+        let mut rand_pk = public_key.public_key.clone();
+        rand_pk.precomputed_base_scalar_mul(
+            cs.ns(|| "check_randomization_gadget"),
+            randomness.iter().zip_eq(&parameters.parameters.generator_powers),
         )?;
+
         Ok(SchnorrPublicKeyGadget {
             public_key: rand_pk,
             _group: PhantomData,
diff --git a/gadgets/src/algorithms/signature/tests.rs b/gadgets/src/algorithms/signature/tests.rs
index dcfa8b314..bcee2b9f1 100644
--- a/gadgets/src/algorithms/signature/tests.rs
+++ b/gadgets/src/algorithms/signature/tests.rs
@@ -57,7 +57,7 @@ fn test_schnorr_signature_randomize_public_key_gadget() {
 
     // Circuit Schnorr randomized public key (candidate)
 
-    let candidate_parameters_gadget = SchnorrParametersGadget::<EdwardsAffine, Fr, EdwardsBlsGadget>::alloc_input(
+    let candidate_parameters_gadget = SchnorrParametersGadget::<EdwardsAffine, Fr, Blake2s>::alloc_input(
         &mut cs.ns(|| "candidate_parameters"),
         || Ok(schnorr_signature.parameters()),
     )
diff --git a/models/Cargo.toml b/models/Cargo.toml
index af503870a..b5ca33d97 100644
--- a/models/Cargo.toml
+++ b/models/Cargo.toml
@@ -17,6 +17,7 @@ snarkos-utilities = { path = "../utilities", version = "0.8.0", default-features
 
 bincode = { version = "1.3.1" }
 derivative = { version = "2" }
+itertools = { version = "0.9.0" }
 rand = { version = "0.7", default-features = false }
 rand_xorshift = { version = "0.2", default-features = false }
 serde = { version = "1.0.114", default-features = false, features = ["derive"] }
diff --git a/models/src/gadgets/algorithms/signature.rs b/models/src/gadgets/algorithms/signature.rs
index ab7fe07ab..86967221e 100644
--- a/models/src/gadgets/algorithms/signature.rs
+++ b/models/src/gadgets/algorithms/signature.rs
@@ -9,7 +9,7 @@ use crate::{
 use snarkos_errors::gadgets::SynthesisError;
 
 pub trait SignaturePublicKeyRandomizationGadget<S: SignatureScheme, F: Field> {
-    type ParametersGadget: AllocGadget<S::Parameters, F> + Clone;
+    type ParametersGadget: AllocGadget<S::Parameters, F>;
     type PublicKeyGadget: ToBytesGadget<F> + EqGadget<F> + AllocGadget<S::PublicKey, F> + Clone;
 
     fn check_randomization_gadget<CS: ConstraintSystem<F>>(
diff --git a/models/src/gadgets/curves/group.rs b/models/src/gadgets/curves/group.rs
index 27dc03825..b4c79e481 100644
--- a/models/src/gadgets/curves/group.rs
+++ b/models/src/gadgets/curves/group.rs
@@ -14,6 +14,7 @@ use crate::{
 };
 use snarkos_errors::gadgets::SynthesisError;
 
+use itertools::Itertools;
 use std::{borrow::Borrow, fmt::Debug};
 
 pub trait GroupGadget<G: Group, F: Field>:
@@ -166,10 +167,10 @@ pub trait GroupGadget<G: Group, F: Field>:
     {
         let mut result = Self::zero(&mut cs.ns(|| "Declare Result"))?;
         // Compute ∏(h_i^{m_i}) for all i.
-        for (i, (bits, base_powers)) in scalars.zip(bases).enumerate() {
+        for (i, (bits, base_powers)) in scalars.zip_eq(bases).enumerate() {
             let base_powers = base_powers.borrow();
             let bits = bits.to_bits(&mut cs.ns(|| format!("Convert Scalar {} to bits", i)))?;
-            result.precomputed_base_scalar_mul(cs.ns(|| format!("Chunk {}", i)), bits.iter().zip(base_powers))?;
+            result.precomputed_base_scalar_mul(cs.ns(|| format!("Chunk {}", i)), bits.iter().zip_eq(base_powers))?;
         }
         Ok(result)
     }
@@ -187,12 +188,13 @@ pub trait GroupGadget<G: Group, F: Field>:
     {
         let mut result = Self::zero(&mut cs.ns(|| "Declare Result"))?;
         // Compute ∏(h_i^{1  - 2*m_i}) for all i.
-        for (i, (bits, base_powers)) in scalars.zip(bases).enumerate() {
+        for (i, (bits, base_powers)) in scalars.zip_eq(bases).enumerate() {
             let base_powers = base_powers.borrow();
             let bits = bits.to_bits(&mut cs.ns(|| format!("Convert Scalar {} to bits", i)))?;
+
             result.precomputed_base_symmetric_scalar_mul(
                 cs.ns(|| format!("Chunk {}", i)),
-                bits.iter().zip(base_powers),
+                bits.iter().zip_eq(base_powers),
             )?;
         }
         Ok(result)
@@ -217,14 +219,16 @@ pub trait GroupGadget<G: Group, F: Field>:
     {
         let mut result = Self::zero(&mut cs.ns(|| "Declare Result"))?;
         for (i, (((scalar, mask), base_powers), mask_powers)) in
-            scalars.zip(masks).zip(bases).zip(mask_bases).enumerate()
+            scalars.zip_eq(masks).zip_eq(bases).zip_eq(mask_bases).enumerate()
         {
             let base_powers = base_powers.borrow();
             let mask_powers = mask_powers.borrow();
             let scalar_bits = scalar.to_bits(&mut cs.ns(|| format!("Convert scalar {} to bits", i)))?;
             let mask_bits = mask.to_bits(&mut cs.ns(|| format!("Convert mask {} to bits", i)))?;
-            let scalar_bits_with_base_powers = scalar_bits.into_iter().zip(base_powers);
-            let mask_bits_with_mask_powers = mask_bits.into_iter().zip(mask_powers);
+
+            let scalar_bits_with_base_powers = scalar_bits.into_iter().zip_eq(base_powers);
+            let mask_bits_with_mask_powers = mask_bits.into_iter().zip_eq(mask_powers);
+
             result.precomputed_base_scalar_mul_masked(
                 cs.ns(|| format!("Chunk {}", i)),
                 scalar_bits_with_base_powers,
diff --git a/parameters/src/genesis/block_header/block_header.genesis b/parameters/src/genesis/block_header/block_header.genesis
index 6bb0791dc..d7e70350e 100644
Binary files a/parameters/src/genesis/block_header/block_header.genesis and b/parameters/src/genesis/block_header/block_header.genesis differ
diff --git a/parameters/src/genesis/transaction_1/transaction_1.genesis b/parameters/src/genesis/transaction_1/transaction_1.genesis
index 261b2697a..739b45d02 100644
Binary files a/parameters/src/genesis/transaction_1/transaction_1.genesis and b/parameters/src/genesis/transaction_1/transaction_1.genesis differ
diff --git a/parameters/src/params.rs b/parameters/src/params.rs
index 170a6031a..927df57fe 100644
--- a/parameters/src/params.rs
+++ b/parameters/src/params.rs
@@ -193,7 +193,7 @@ impl_params!(
     AccountSignatureParameters,
     account_signature_test,
     "account_signature",
-    96
+    16420
 );
 impl_params!(
     LedgerMerkleTreeParameters,
@@ -255,8 +255,8 @@ impl_params!(
 
 // Inner SNARK
 impl_params_remote!(InnerSNARKPKParameters, "inner_snark_pk", 210199697);
-impl_params!(InnerSNARKVKParameters, inner_snark_vk_test, "inner_snark_vk", 2426);
+impl_params!(InnerSNARKVKParameters, inner_snark_vk_test, "inner_snark_vk", 2232);
 
 // Outer SNARK
 impl_params_remote!(OuterSNARKPKParameters, "outer_snark_pk", 467030109);
-impl_params!(OuterSNARKVKParameters, outer_snark_vk_test, "outer_snark_vk", 4443);
+impl_params!(OuterSNARKVKParameters, outer_snark_vk_test, "outer_snark_vk", 4057);
diff --git a/parameters/src/params/account_commitment.checksum b/parameters/src/params/account_commitment.checksum
index 29a404664..a7f5c6963 100644
--- a/parameters/src/params/account_commitment.checksum
+++ b/parameters/src/params/account_commitment.checksum
@@ -1 +1 @@
-2fc188eaedf667adbf0b159430c4705617299ee531de3c33b49a69c99608434a
\ No newline at end of file
+b7e6f56f632fcba59178901a0d148b556deb80d513939ab9c5ee94edb1a3e89e
\ No newline at end of file
diff --git a/parameters/src/params/account_commitment.params b/parameters/src/params/account_commitment.params
index df370e4eb..6a86132c4 100644
Binary files a/parameters/src/params/account_commitment.params and b/parameters/src/params/account_commitment.params differ
diff --git a/parameters/src/params/account_signature.checksum b/parameters/src/params/account_signature.checksum
index d9835dad7..9727af8ff 100644
--- a/parameters/src/params/account_signature.checksum
+++ b/parameters/src/params/account_signature.checksum
@@ -1 +1 @@
-a423244ed06e83b51a84a94aa8e60e6c62ac7eec9541516a4d66ae7526d7ef2d
\ No newline at end of file
+972288153f5a2d0a8c561d67f08cf3ab813a5265d87952114e3e87cb00fc1eb6
\ No newline at end of file
diff --git a/parameters/src/params/account_signature.params b/parameters/src/params/account_signature.params
index 835cafd26..c7ca04aea 100644
Binary files a/parameters/src/params/account_signature.params and b/parameters/src/params/account_signature.params differ
diff --git a/parameters/src/params/inner_snark_pk.checksum b/parameters/src/params/inner_snark_pk.checksum
index f05dcacf4..309037bdc 100644
--- a/parameters/src/params/inner_snark_pk.checksum
+++ b/parameters/src/params/inner_snark_pk.checksum
@@ -1 +1 @@
-8becc2396d06a1a41577bfcacfe794cbb0128f975acacd498471cc506952bc89
\ No newline at end of file
+ec5c5431499f00a1a980327dcca1f80c05637faa1ea71c34c1bc69483c352555
\ No newline at end of file
diff --git a/parameters/src/params/inner_snark_vk.checksum b/parameters/src/params/inner_snark_vk.checksum
index 3cd4b1874..d18150695 100644
--- a/parameters/src/params/inner_snark_vk.checksum
+++ b/parameters/src/params/inner_snark_vk.checksum
@@ -1 +1 @@
-24dc14658412320bf93914342698577f53d63932017e5bd0cf3bb8d48e415920
\ No newline at end of file
+7390f6d62ea423bd1cbadc6daabe54d6f4a05dccb7a7d719729f9172c0c6fdd0
\ No newline at end of file
diff --git a/parameters/src/params/inner_snark_vk.params b/parameters/src/params/inner_snark_vk.params
index 141db3ee1..a8da4ca5a 100644
Binary files a/parameters/src/params/inner_snark_vk.params and b/parameters/src/params/inner_snark_vk.params differ
diff --git a/parameters/src/params/ledger_merkle_tree.checksum b/parameters/src/params/ledger_merkle_tree.checksum
index 475350e9e..9ec794722 100644
--- a/parameters/src/params/ledger_merkle_tree.checksum
+++ b/parameters/src/params/ledger_merkle_tree.checksum
@@ -1 +1 @@
-c527b0eaf696c429eb57e45008fdb4a2c179d0f47af32f7caee7396ba3b51068
\ No newline at end of file
+89afeede57ccdcba8b0bd12684c8992cfe464ebe2a8167ff14636bf6728bcad0
\ No newline at end of file
diff --git a/parameters/src/params/ledger_merkle_tree.params b/parameters/src/params/ledger_merkle_tree.params
index 4a4914969..5f441a629 100644
Binary files a/parameters/src/params/ledger_merkle_tree.params and b/parameters/src/params/ledger_merkle_tree.params differ
diff --git a/parameters/src/params/local_data_commitment.checksum b/parameters/src/params/local_data_commitment.checksum
index b1560e6b8..5327142fa 100644
--- a/parameters/src/params/local_data_commitment.checksum
+++ b/parameters/src/params/local_data_commitment.checksum
@@ -1 +1 @@
-3d271e9e63de5c9d80f43e3a599e37352d84506476cf32fe3306111d75b3e37a
\ No newline at end of file
+b000076f99efe2e53ac7ddd181d37063f0606679ea31dabdb18de466ac06eba3
\ No newline at end of file
diff --git a/parameters/src/params/local_data_commitment.params b/parameters/src/params/local_data_commitment.params
index 8851a1cd3..fa4280959 100644
Binary files a/parameters/src/params/local_data_commitment.params and b/parameters/src/params/local_data_commitment.params differ
diff --git a/parameters/src/params/local_data_crh.checksum b/parameters/src/params/local_data_crh.checksum
index cbcb53c0a..c49ef320f 100644
--- a/parameters/src/params/local_data_crh.checksum
+++ b/parameters/src/params/local_data_crh.checksum
@@ -1 +1 @@
-b826eb94c5bdea770ce2491be612b184a97e6a8cac88c1830b26ee97e1f06823
\ No newline at end of file
+3013938f1b9eb8dad2ef90b18223e9f5da8ab228b5edd59947b4f997c2f9de20
\ No newline at end of file
diff --git a/parameters/src/params/local_data_crh.params b/parameters/src/params/local_data_crh.params
index f812f0941..9839d49cc 100644
Binary files a/parameters/src/params/local_data_crh.params and b/parameters/src/params/local_data_crh.params differ
diff --git a/parameters/src/params/outer_snark_pk.checksum b/parameters/src/params/outer_snark_pk.checksum
index 63bf90d81..940b7a514 100644
--- a/parameters/src/params/outer_snark_pk.checksum
+++ b/parameters/src/params/outer_snark_pk.checksum
@@ -1 +1 @@
-5628e99bd5bd0f105fa2c473f3aeabe63b356474f4036afebb812ca24e19ff7a
\ No newline at end of file
+78b7ba3a396162bf22ae95d15b4c5b12433cca384eee9822273b99f28ba81414
\ No newline at end of file
diff --git a/parameters/src/params/outer_snark_vk.checksum b/parameters/src/params/outer_snark_vk.checksum
index 187bd0935..e9a2b77bf 100644
--- a/parameters/src/params/outer_snark_vk.checksum
+++ b/parameters/src/params/outer_snark_vk.checksum
@@ -1 +1 @@
-d6a01d9ab7e017cc911bcba914205a975ca0928b1110a2c3cece97e8049d56ff
\ No newline at end of file
+03e3c89c7f80348bb1fb010ca01862388a828ac6425e5bfa6aed892cac39d8cb
\ No newline at end of file
diff --git a/parameters/src/params/outer_snark_vk.params b/parameters/src/params/outer_snark_vk.params
index 9ea8dfbe7..818d65c0a 100644
Binary files a/parameters/src/params/outer_snark_vk.params and b/parameters/src/params/outer_snark_vk.params differ
diff --git a/parameters/src/params/posw_snark_pk.checksum b/parameters/src/params/posw_snark_pk.checksum
index 6961cf81f..dcb8ff38b 100644
--- a/parameters/src/params/posw_snark_pk.checksum
+++ b/parameters/src/params/posw_snark_pk.checksum
@@ -1 +1 @@
-0fb68903c1b48ebead5ff8179eadbe6368a9c1180677a0f1ae655f05cc9e2a41
\ No newline at end of file
+fce49940bd2de466bbd663774e48fef364b96596c902e55e5508d0a24cf75613
\ No newline at end of file
diff --git a/parameters/src/params/posw_snark_pk.params b/parameters/src/params/posw_snark_pk.params
index 9adf4effb..697316ede 100644
Binary files a/parameters/src/params/posw_snark_pk.params and b/parameters/src/params/posw_snark_pk.params differ
diff --git a/parameters/src/params/posw_snark_vk.checksum b/parameters/src/params/posw_snark_vk.checksum
index 4b11150af..ca9c4bd48 100644
--- a/parameters/src/params/posw_snark_vk.checksum
+++ b/parameters/src/params/posw_snark_vk.checksum
@@ -1 +1 @@
-435f6cb75f8d691aec4e293c550b3023f0d843785bba166e5dcff24d008b0fac
\ No newline at end of file
+f1d217666352ff684e5a45416774c24ac089c1800da163c3ed0ad346f40bdd80
\ No newline at end of file
diff --git a/parameters/src/params/posw_snark_vk.params b/parameters/src/params/posw_snark_vk.params
index bf7fcd2ed..9f2bb2d92 100644
Binary files a/parameters/src/params/posw_snark_vk.params and b/parameters/src/params/posw_snark_vk.params differ
diff --git a/parameters/src/params/predicate_snark_pk.checksum b/parameters/src/params/predicate_snark_pk.checksum
index 42e60796f..c0831fe08 100644
--- a/parameters/src/params/predicate_snark_pk.checksum
+++ b/parameters/src/params/predicate_snark_pk.checksum
@@ -1 +1 @@
-4a19420c96dc5314bcda0e516c3305b4a5af22f440ebd05a9401588b2dc033da
\ No newline at end of file
+c5d3c4ceec05a5b5d727b32261a68eb24e0de1008da45a06b7c892fc984f7307
\ No newline at end of file
diff --git a/parameters/src/params/predicate_snark_pk.params b/parameters/src/params/predicate_snark_pk.params
index ff6a05051..7bb251caf 100644
Binary files a/parameters/src/params/predicate_snark_pk.params and b/parameters/src/params/predicate_snark_pk.params differ
diff --git a/parameters/src/params/predicate_snark_vk.checksum b/parameters/src/params/predicate_snark_vk.checksum
index 2b940d770..e118c0ee6 100644
--- a/parameters/src/params/predicate_snark_vk.checksum
+++ b/parameters/src/params/predicate_snark_vk.checksum
@@ -1 +1 @@
-cd3b382627eba954f073494c12919c3625a64b9d6e090da05eda28e2ce11ab1f
\ No newline at end of file
+d2396fd002571d07b3b373866f8ca922279172d1a62a8729da84c9de238f7986
\ No newline at end of file
diff --git a/parameters/src/params/predicate_snark_vk.params b/parameters/src/params/predicate_snark_vk.params
index 20bb7bf5b..a0781c07c 100644
Binary files a/parameters/src/params/predicate_snark_vk.params and b/parameters/src/params/predicate_snark_vk.params differ
diff --git a/parameters/src/params/predicate_vk_crh.checksum b/parameters/src/params/predicate_vk_crh.checksum
index d53cd4548..13901ac26 100644
--- a/parameters/src/params/predicate_vk_crh.checksum
+++ b/parameters/src/params/predicate_vk_crh.checksum
@@ -1 +1 @@
-20aebd1304338d4025da593dd3dbeaa80e5143a628871fee833fe2d5555f568f
\ No newline at end of file
+874ec0d2b06f7a0bb90c025144e274c2395d6f55d2cedd393ac5016b07ecab12
\ No newline at end of file
diff --git a/parameters/src/params/predicate_vk_crh.params b/parameters/src/params/predicate_vk_crh.params
index 9d7776c87..3a6eeeafe 100644
Binary files a/parameters/src/params/predicate_vk_crh.params and b/parameters/src/params/predicate_vk_crh.params differ
diff --git a/parameters/src/params/record_commitment.checksum b/parameters/src/params/record_commitment.checksum
index 8a3da5418..2fed38914 100644
--- a/parameters/src/params/record_commitment.checksum
+++ b/parameters/src/params/record_commitment.checksum
@@ -1 +1 @@
-d09deb79eac3b44cdebb8f8a9d5a0e1c0c9999c93cb56ea6af39d6544ed61654
\ No newline at end of file
+86bf3415b7a943d16130f0ff209a2a131767bada1e1ca91a389f08c9bab70ac8
\ No newline at end of file
diff --git a/parameters/src/params/record_commitment.params b/parameters/src/params/record_commitment.params
index 7555ac307..ee35203a2 100644
Binary files a/parameters/src/params/record_commitment.params and b/parameters/src/params/record_commitment.params differ
diff --git a/parameters/src/params/serial_number_nonce_crh.checksum b/parameters/src/params/serial_number_nonce_crh.checksum
index 50652ccb3..e898a784d 100644
--- a/parameters/src/params/serial_number_nonce_crh.checksum
+++ b/parameters/src/params/serial_number_nonce_crh.checksum
@@ -1 +1 @@
-842cab1b6fddb18830994461b4e311ff8ac7e23060b29ef4f92c384f2540008a
\ No newline at end of file
+acf74d68b8dc3630dc67a058e5b1f4081720562395467386775ca42e1b1b0ae6
\ No newline at end of file
diff --git a/parameters/src/params/serial_number_nonce_crh.params b/parameters/src/params/serial_number_nonce_crh.params
index 583f9e652..24ea75174 100644
Binary files a/parameters/src/params/serial_number_nonce_crh.params and b/parameters/src/params/serial_number_nonce_crh.params differ
diff --git a/parameters/src/params/value_commitment.checksum b/parameters/src/params/value_commitment.checksum
index 7e27f616e..c82955910 100644
--- a/parameters/src/params/value_commitment.checksum
+++ b/parameters/src/params/value_commitment.checksum
@@ -1 +1 @@
-cf8685211398bb9287121f01c5f92d0457c66bbb0058b87b66705bce273651e2
\ No newline at end of file
+989897430cc2edfddce88d78fa3b494a75997de739c35d8d5ea42d47ea567610
\ No newline at end of file
diff --git a/parameters/src/params/value_commitment.params b/parameters/src/params/value_commitment.params
index 03c37ed27..7104914c4 100644
Binary files a/parameters/src/params/value_commitment.params and b/parameters/src/params/value_commitment.params differ
diff --git a/testing/src/consensus/test_data b/testing/src/consensus/test_data
index aad6a3887..d609ea826 100644
Binary files a/testing/src/consensus/test_data and b/testing/src/consensus/test_data differ
