diff --git a/circuit/environment/src/helpers/updatable_count.rs b/circuit/environment/src/helpers/updatable_count.rs
index 5c6842e5b..122db5c86 100644
--- a/circuit/environment/src/helpers/updatable_count.rs
+++ b/circuit/environment/src/helpers/updatable_count.rs
@@ -499,7 +499,7 @@ mod test {
     fn check_position() {
         let count = count_is!(0, 0, 0, 0);
         assert_eq!(count.file, "circuit/environment/src/helpers/updatable_count.rs");
-        assert_eq!(count.line, 499);
+        assert_eq!(count.line, 500);
         assert_eq!(count.column, 21);
     }
 
