diff --git a/algorithms/src/crh/pedersen.rs b/algorithms/src/crh/pedersen.rs
index bee7429bb..48d37550d 100644
--- a/algorithms/src/crh/pedersen.rs
+++ b/algorithms/src/crh/pedersen.rs
@@ -47,9 +47,9 @@ impl<G: Group, S: PedersenSize> CRH for PedersenCRH<G, S> {
             ));
         }
 
-        let mut padded_input = vec![];
-        let mut input = input;
         // Pad the input if it is not the current length.
+        let mut input = input;
+        let mut padded_input = vec![];
         if (input.len() * 8) < S::WINDOW_SIZE * S::NUM_WINDOWS {
             padded_input.extend_from_slice(input);
             for _ in input.len()..((S::WINDOW_SIZE * S::NUM_WINDOWS) / 8) {
diff --git a/algorithms/src/lib.rs b/algorithms/src/lib.rs
index 0099a8f2c..85b520a35 100644
--- a/algorithms/src/lib.rs
+++ b/algorithms/src/lib.rs
@@ -3,7 +3,7 @@
 #![deny(non_shorthand_field_patterns, unused_attributes)]
 #![deny(renamed_and_removed_lints, unused_allocation, unused_comparisons)]
 #![deny(const_err, unused_must_use, unused_mut, private_in_public)]
-#![deny(unreachable_pub, unused_extern_crates, trivial_numeric_casts)]
+#![deny(unused_extern_crates, trivial_numeric_casts)]
 #![forbid(unsafe_code)]
 
 #[macro_use]
diff --git a/algorithms/src/merkle_tree/merkle_path.rs b/algorithms/src/merkle_tree/merkle_path.rs
index a1ca70201..be2b3a666 100644
--- a/algorithms/src/merkle_tree/merkle_path.rs
+++ b/algorithms/src/merkle_tree/merkle_path.rs
@@ -1,6 +1,6 @@
 use snarkos_errors::algorithms::MerkleError;
 use snarkos_models::algorithms::{MerkleParameters, CRH};
-use snarkos_utilities::bytes::ToBytes;
+use snarkos_utilities::ToBytes;
 
 pub type MerkleTreeDigest<P> = <<P as MerkleParameters>::H as CRH>::Output;
 
@@ -14,14 +14,15 @@ pub struct MerklePath<P: MerkleParameters> {
 
 impl<P: MerkleParameters> MerklePath<P> {
     pub fn verify<L: ToBytes>(&self, root_hash: &<P::H as CRH>::Output, leaf: &L) -> Result<bool, MerkleError> {
-        if self.path.len() != P::HEIGHT - 1 {
+        if self.path.len() != P::DEPTH {
             return Ok(false);
         }
 
         // Check that the given leaf matches the leaf in the membership proof.
-        let mut buffer = [0u8; 128];
-
         if !self.path.is_empty() {
+            let hash_input_size_in_bytes = (P::H::INPUT_SIZE_BITS / 8) * 2;
+            let mut buffer = vec![0u8; hash_input_size_in_bytes];
+
             let claimed_leaf_hash = self.parameters.hash_leaf::<L>(leaf, &mut buffer)?;
 
             // Check if leaf is one of the bottom-most siblings.
@@ -29,19 +30,21 @@ impl<P: MerkleParameters> MerklePath<P> {
                 return Ok(false);
             };
 
-            let mut prev = claimed_leaf_hash;
             // Check levels between leaf level and root.
+            let mut previous_hash = claimed_leaf_hash;
+            let mut buffer = vec![0u8; hash_input_size_in_bytes];
             for &(ref hash, ref sibling_hash) in &self.path {
                 // Check if the previous hash matches the correct current hash.
-                if &prev != hash && &prev != sibling_hash {
+                if &previous_hash != hash && &previous_hash != sibling_hash {
                     return Ok(false);
                 };
-                prev = self.parameters.hash_inner_node(hash, sibling_hash, &mut buffer)?;
+                previous_hash = self.parameters.hash_inner_node(hash, sibling_hash, &mut buffer)?;
             }
 
-            if root_hash != &prev {
+            if root_hash != &previous_hash {
                 return Ok(false);
             }
+
             Ok(true)
         } else {
             Ok(false)
@@ -51,8 +54,8 @@ impl<P: MerkleParameters> MerklePath<P> {
 
 impl<P: MerkleParameters> Default for MerklePath<P> {
     fn default() -> Self {
-        let mut path = Vec::with_capacity(P::HEIGHT);
-        for _i in 1..P::HEIGHT {
+        let mut path = Vec::with_capacity(P::DEPTH);
+        for _i in 0..P::DEPTH {
             path.push((<P::H as CRH>::Output::default(), <P::H as CRH>::Output::default()));
         }
         Self {
diff --git a/algorithms/src/merkle_tree/merkle_tree.rs b/algorithms/src/merkle_tree/merkle_tree.rs
index 64c95cbab..27aaebd7c 100644
--- a/algorithms/src/merkle_tree/merkle_tree.rs
+++ b/algorithms/src/merkle_tree/merkle_tree.rs
@@ -1,55 +1,66 @@
 use crate::merkle_tree::MerklePath;
 use snarkos_errors::algorithms::MerkleError;
 use snarkos_models::algorithms::{MerkleParameters, CRH};
-use snarkos_utilities::bytes::ToBytes;
+use snarkos_utilities::ToBytes;
 
 pub struct MerkleTree<P: MerkleParameters> {
+    /// The computed root of the full Merkle tree.
     root: Option<<P::H as CRH>::Output>,
+
+    /// The internal hashes, from root to hashed leaves, of the full Merkle tree.
     tree: Vec<<P::H as CRH>::Output>,
-    leaves_hashed: Vec<<P::H as CRH>::Output>,
-    padding_tree: Vec<(<P::H as CRH>::Output, <P::H as CRH>::Output)>, // For each level after a full tree has been built from the leaves, keeps both the roots the siblings that are used to get to the desired height.
+
+    /// The hash of each non-empty leaf in the Merkle tree.
+    hashed_leaves: Vec<<P::H as CRH>::Output>,
+
+    /// For each level after a full tree has been built from the leaves,
+    /// keeps both the roots the siblings that are used to get to the desired depth.
+    padding_tree: Vec<(<P::H as CRH>::Output, <P::H as CRH>::Output)>,
+
+    /// The Merkle tree parameters (e.g. the hash function).
     parameters: P,
 }
 
 impl<P: MerkleParameters> MerkleTree<P> {
-    pub const HEIGHT: u8 = P::HEIGHT as u8;
+    pub const DEPTH: u8 = P::DEPTH as u8;
 
     pub fn new<L: ToBytes>(parameters: P, leaves: &[L]) -> Result<Self, MerkleError> {
-        let new_time = start_timer!(|| "MerkleTree::New");
+        let new_time = start_timer!(|| "MerkleTree::new");
 
         let last_level_size = leaves.len().next_power_of_two();
         let tree_size = 2 * last_level_size - 1;
-        let tree_height = tree_height(tree_size);
+        let tree_depth = tree_depth(tree_size);
 
-        if tree_height > Self::HEIGHT as usize {
-            return Err(MerkleError::InvalidTreeHeight(tree_height, Self::HEIGHT as usize));
+        if tree_depth > Self::DEPTH as usize {
+            return Err(MerkleError::InvalidTreeDepth(tree_depth, Self::DEPTH as usize));
         }
 
-        // Initialize the merkle tree.
+        // Initialize the Merkle tree.
         let mut tree = Vec::with_capacity(tree_size);
         let empty_hash = parameters.hash_empty()?;
         for _ in 0..tree_size {
             tree.push(empty_hash.clone());
         }
 
-        // Compute the starting indices for each level of the tree.
+        // Compute the starting index (on the left) for each level of the tree.
         let mut index = 0;
-        let mut level_indices = Vec::with_capacity(tree_height);
-        for _ in 0..tree_height {
+        let mut level_indices = Vec::with_capacity(tree_depth);
+        for _ in 0..=tree_depth {
             level_indices.push(index);
             index = left_child(index);
         }
 
         // Compute and store the hash values for each leaf.
-        let last_level_index = level_indices.pop().unwrap();
-        let mut buffer = [0u8; 128];
+        let hash_input_size_in_bytes = (P::H::INPUT_SIZE_BITS / 8) * 2;
+        let last_level_index = level_indices.pop().unwrap_or(0);
+        let mut buffer = vec![0u8; hash_input_size_in_bytes];
         for (i, leaf) in leaves.iter().enumerate() {
             tree[last_level_index + i] = parameters.hash_leaf(leaf, &mut buffer)?;
         }
 
         // Compute the hash values for every node in the tree.
         let mut upper_bound = last_level_index;
-        let mut buffer = [0u8; 128];
+        let mut buffer = vec![0u8; hash_input_size_in_bytes];
         level_indices.reverse();
         for &start_index in &level_indices {
             // Iterate over the current level.
@@ -62,29 +73,30 @@ impl<P: MerkleParameters> MerkleTree<P> {
             }
             upper_bound = start_index;
         }
+
         // Finished computing actual tree.
-        // Now, we compute the dummy nodes until we hit our HEIGHT goal.
-        let mut cur_height = tree_height;
+        // Now, we compute the dummy nodes until we hit our DEPTH goal.
+        let mut current_depth = tree_depth;
         let mut padding_tree = vec![];
-        let mut cur_hash = tree[0].clone();
-        while cur_height < Self::HEIGHT as usize {
-            cur_hash = parameters.hash_inner_node(&cur_hash, &empty_hash, &mut buffer)?;
+        let mut current_hash = tree[0].clone();
+        while current_depth < Self::DEPTH as usize {
+            current_hash = parameters.hash_inner_node(&current_hash, &empty_hash, &mut buffer)?;
 
             // do not pad at the top-level of the tree
-            if cur_height < Self::HEIGHT as usize - 1 {
-                padding_tree.push((cur_hash.clone(), empty_hash.clone()));
+            if current_depth < Self::DEPTH as usize - 1 {
+                padding_tree.push((current_hash.clone(), empty_hash.clone()));
             }
-            cur_height += 1;
+            current_depth += 1;
         }
-        let root_hash = cur_hash;
+        let root_hash = current_hash;
 
         end_timer!(new_time);
 
-        let leaves_hashed = tree[last_level_index..].to_vec();
+        let hashed_leaves = tree[last_level_index..].to_vec();
         Ok(MerkleTree {
             tree,
             padding_tree,
-            leaves_hashed,
+            hashed_leaves,
             parameters,
             root: Some(root_hash),
         })
@@ -96,19 +108,21 @@ impl<P: MerkleParameters> MerkleTree<P> {
     }
 
     #[inline]
-    pub fn leaves_hashed(&self) -> Vec<<P::H as CRH>::Output> {
-        self.leaves_hashed.clone()
+    pub fn hashed_leaves(&self) -> Vec<<P::H as CRH>::Output> {
+        self.hashed_leaves.clone()
     }
 
     pub fn generate_proof<L: ToBytes>(&self, index: usize, leaf: &L) -> Result<MerklePath<P>, MerkleError> {
-        let prove_time = start_timer!(|| "MerkleTree::GenProof");
+        let prove_time = start_timer!(|| "MerkleTree::generate_proof");
         let mut path = vec![];
 
-        let mut buffer = [0u8; 128];
+        let hash_input_size_in_bytes = (P::H::INPUT_SIZE_BITS / 8) * 2;
+        let mut buffer = vec![0u8; hash_input_size_in_bytes];
+
         let leaf_hash = self.parameters.hash_leaf(leaf, &mut buffer)?;
-        let tree_height = tree_height(self.tree.len());
-        let tree_index = convert_index_to_last_level(index, tree_height);
-        let empty_hash = self.parameters.hash_empty()?;
+
+        let tree_depth = tree_depth(self.tree.len());
+        let tree_index = convert_index_to_last_level(index, tree_depth);
 
         // Check that the given index corresponds to the correct leaf.
         if leaf_hash != self.tree[tree_index] {
@@ -129,19 +143,21 @@ impl<P: MerkleParameters> MerkleTree<P> {
         }
 
         // Store the root node. Set boolean as true for consistency with digest location.
-        if path.len() >= Self::HEIGHT as usize {
-            return Err(MerkleError::InvalidPathLength(path.len(), Self::HEIGHT as usize));
+        if path.len() > Self::DEPTH as usize {
+            return Err(MerkleError::InvalidPathLength(path.len(), Self::DEPTH as usize));
         }
 
-        if path.len() != (Self::HEIGHT - 1) as usize {
+        if path.len() != Self::DEPTH as usize {
+            let empty_hash = self.parameters.hash_empty()?;
             path.push((self.tree[0].clone(), empty_hash));
+
             for &(ref hash, ref sibling_hash) in &self.padding_tree {
                 path.push((hash.clone(), sibling_hash.clone()));
             }
         }
         end_timer!(prove_time);
 
-        if path.len() != (Self::HEIGHT - 1) as usize {
+        if path.len() != Self::DEPTH as usize {
             Err(MerkleError::IncorrectPathLength(path.len()))
         } else {
             Ok(MerklePath {
@@ -157,22 +173,22 @@ impl<P: MerkleParameters> Default for MerkleTree<P> {
         MerkleTree {
             tree: vec![],
             padding_tree: vec![],
-            leaves_hashed: vec![],
+            hashed_leaves: vec![],
             root: None,
             parameters: P::default(),
         }
     }
 }
 
-/// Returns the height of the tree, given the size of the tree.
+/// Returns the depth of the tree, given the size of the tree.
 #[inline]
-fn tree_height(tree_size: usize) -> usize {
+fn tree_depth(tree_size: usize) -> usize {
     // Returns the log2 value of the given number.
     fn log2(number: usize) -> usize {
         (number as f64).log2() as usize
     }
 
-    log2(tree_size + 1)
+    log2(tree_size)
 }
 
 /// Returns true iff the index represents the root.
@@ -218,6 +234,6 @@ fn parent(index: usize) -> Option<usize> {
 }
 
 #[inline]
-fn convert_index_to_last_level(index: usize, tree_height: usize) -> usize {
-    index + (1 << (tree_height - 1)) - 1
+fn convert_index_to_last_level(index: usize, tree_depth: usize) -> usize {
+    index + (1 << tree_depth) - 1
 }
diff --git a/algorithms/src/merkle_tree/mod.rs b/algorithms/src/merkle_tree/mod.rs
index 616213270..714b22152 100644
--- a/algorithms/src/merkle_tree/mod.rs
+++ b/algorithms/src/merkle_tree/mod.rs
@@ -20,12 +20,15 @@ pub fn prng() -> impl Rng {
 }
 
 #[macro_export]
-/// Defines a Merkle Tree using the provided hash and height.
+/// Defines a Merkle tree using the provided hash and depth.
 macro_rules! define_merkle_tree_parameters {
-    ($struct_name:ident, $hash:ty, $height:expr) => {
+    ($struct_name:ident, $hash:ty, $depth:expr) => {
+        #[allow(unused_imports)]
         use snarkos_models::algorithms::{MerkleParameters, CRH};
+        #[allow(unused_imports)]
         use $crate::merkle_tree::MerkleTree;
 
+        #[allow(unused_imports)]
         use rand::Rng;
 
         #[derive(Clone, PartialEq, Eq, Debug)]
@@ -34,7 +37,7 @@ macro_rules! define_merkle_tree_parameters {
         impl MerkleParameters for $struct_name {
             type H = $hash;
 
-            const HEIGHT: usize = $height;
+            const DEPTH: usize = $depth;
 
             fn setup<R: Rng>(rng: &mut R) -> Self {
                 Self(Self::H::setup(rng))
diff --git a/algorithms/src/merkle_tree/tests.rs b/algorithms/src/merkle_tree/tests.rs
index d3394e080..6cdfd0f0e 100644
--- a/algorithms/src/merkle_tree/tests.rs
+++ b/algorithms/src/merkle_tree/tests.rs
@@ -1,67 +1,279 @@
 use crate::{
-    crh::{PedersenCRH, PedersenSize},
+    crh::{PedersenCRH, PedersenCompressedCRH, PedersenSize},
     define_merkle_tree_parameters,
+    merkle_tree::MerkleTree,
 };
-use snarkos_curves::edwards_bls12::EdwardsAffine as Edwards;
-use snarkos_models::curves::Zero;
-use snarkos_utilities::bytes::ToBytes;
-
-#[derive(Clone, Debug, PartialEq, Eq)]
-pub struct Size;
-impl PedersenSize for Size {
-    const NUM_WINDOWS: usize = 256;
-    const WINDOW_SIZE: usize = 4;
-}
-
-define_merkle_tree_parameters!(MTParameters, PedersenCRH<Edwards, Size>, 32);
-
-type EdwardsMerkleTree = MerkleTree<MTParameters>;
+use snarkos_models::algorithms::{crh::CRH, merkle_tree::MerkleParameters};
+use snarkos_utilities::{to_bytes, ToBytes};
 
-fn generate_merkle_tree<L: ToBytes + Clone + Eq>(leaves: &[L]) -> () {
-    let parameters = MTParameters::default();
-    let tree = EdwardsMerkleTree::new(parameters.clone(), leaves).unwrap();
+/// Generates a valid Merkle tree and verifies the Merkle path witness for each leaf.
+fn generate_merkle_tree<P: MerkleParameters, L: ToBytes + Clone + Eq>(leaves: &[L], parameters: &P) -> MerkleTree<P> {
+    let tree = MerkleTree::<P>::new(parameters.clone(), leaves).unwrap();
     for (i, leaf) in leaves.iter().enumerate() {
         let proof = tree.generate_proof(i, &leaf).unwrap();
+        assert_eq!(P::DEPTH, proof.path.len());
         assert!(proof.verify(&tree.root(), &leaf).unwrap());
     }
+    tree
 }
 
-fn bad_merkle_tree_verify<L: ToBytes + Clone + Eq>(leaves: &[L]) -> () {
-    let parameters = MTParameters::default();
-    let tree = EdwardsMerkleTree::new(parameters, leaves).unwrap();
+/// Generates a valid Merkle tree and verifies the Merkle path witness for each leaf does not verify to an invalid root hash.
+fn bad_merkle_tree_verify<P: MerkleParameters, L: ToBytes + Clone + Eq>(leaves: &[L], parameters: &P) -> () {
+    let tree = MerkleTree::<P>::new(parameters.clone(), leaves).unwrap();
     for (i, leaf) in leaves.iter().enumerate() {
         let proof = tree.generate_proof(i, &leaf).unwrap();
-        assert!(proof.verify(&Edwards::zero(), &leaf).unwrap());
+        assert!(proof.verify(&<P::H as CRH>::Output::default(), &leaf).unwrap());
     }
 }
 
-#[test]
-fn good_root_test() {
+fn run_empty_merkle_tree_test<P: MerkleParameters>() {
+    let parameters = &P::default();
+    generate_merkle_tree::<P, Vec<u8>>(&vec![], parameters);
+}
+
+fn run_good_root_test<P: MerkleParameters>() {
+    let parameters = &P::default();
+
     let mut leaves = vec![];
     for i in 0..4u8 {
         leaves.push([i, i, i, i, i, i, i, i]);
     }
-    generate_merkle_tree(&leaves);
+    generate_merkle_tree::<P, _>(&leaves, parameters);
 
     let mut leaves = vec![];
     for i in 0..15u8 {
         leaves.push([i, i, i, i, i, i, i, i]);
     }
-    generate_merkle_tree(&leaves);
+    generate_merkle_tree::<P, _>(&leaves, parameters);
 }
 
-#[should_panic]
-#[test]
-fn bad_root_test() {
+fn run_bad_root_test<P: MerkleParameters>() {
+    let parameters = &P::default();
+
     let mut leaves = vec![];
     for i in 0..4u8 {
         leaves.push([i, i, i, i, i, i, i, i]);
     }
-    generate_merkle_tree(&leaves);
+    generate_merkle_tree::<P, _>(&leaves, parameters);
 
     let mut leaves = vec![];
     for i in 0..15u8 {
         leaves.push([i, i, i, i, i, i, i, i]);
     }
-    bad_merkle_tree_verify(&leaves);
+    bad_merkle_tree_verify::<P, _>(&leaves, parameters);
+}
+
+fn run_merkle_tree_matches_hashing_test<P: MerkleParameters>() {
+    let parameters = &P::default();
+
+    // Evaluate the Merkle tree root
+
+    let mut leaves = Vec::new();
+    for i in 0..4u8 {
+        let input = [i; 64];
+        leaves.push(input.to_vec());
+    }
+    let merkle_tree = generate_merkle_tree(&leaves, parameters);
+    let merkle_tree_root = merkle_tree.root();
+
+    // Evaluate the root by direct hashing
+
+    let pedersen = &P::crh(parameters);
+
+    // depth 2
+    let leaf1 = pedersen.hash(&leaves[0]).unwrap();
+    let leaf2 = pedersen.hash(&leaves[1]).unwrap();
+    let leaf3 = pedersen.hash(&leaves[2]).unwrap();
+    let leaf4 = pedersen.hash(&leaves[3]).unwrap();
+
+    // depth 1
+    let left = pedersen.hash(&to_bytes![leaf1, leaf2].unwrap()).unwrap();
+    let right = pedersen.hash(&to_bytes![leaf3, leaf4].unwrap()).unwrap();
+
+    // depth 0
+    let expected_root = pedersen.hash(&to_bytes![left, right].unwrap()).unwrap();
+
+    println!(
+        "merkle_root == expected_root\n\t{} == {}",
+        merkle_tree.root(),
+        expected_root
+    );
+    assert_eq!(merkle_tree_root, expected_root);
+}
+
+fn run_padded_merkle_tree_matches_hashing_test<P: MerkleParameters>() {
+    let parameters = &P::default();
+
+    // Evaluate the Merkle tree root
+
+    let mut leaves = Vec::new();
+    for i in 0..4u8 {
+        let input = [i; 64];
+        leaves.push(input.to_vec());
+    }
+    let merkle_tree = generate_merkle_tree(&leaves, parameters);
+    let merkle_tree_root = merkle_tree.root();
+
+    // Evaluate the root by direct hashing
+
+    let pedersen = &P::crh(parameters);
+
+    // depth 3
+    let leaf1 = pedersen.hash(&leaves[0]).unwrap();
+    let leaf2 = pedersen.hash(&leaves[1]).unwrap();
+    let leaf3 = pedersen.hash(&leaves[2]).unwrap();
+    let leaf4 = pedersen.hash(&leaves[3]).unwrap();
+
+    // depth 2
+    let left = pedersen.hash(&to_bytes![leaf1, leaf2].unwrap()).unwrap();
+    let right = pedersen.hash(&to_bytes![leaf3, leaf4].unwrap()).unwrap();
+
+    // depth 1
+    let penultimate_left = pedersen.hash(&to_bytes![left, right].unwrap()).unwrap();
+    let penultimate_right = parameters.hash_empty().unwrap();
+
+    // depth 0
+    let expected_root = pedersen
+        .hash(&to_bytes![penultimate_left, penultimate_right].unwrap())
+        .unwrap();
+
+    println!(
+        "merkle_root == expected_root\n\t{} == {}",
+        merkle_tree.root(),
+        expected_root
+    );
+    assert_eq!(merkle_tree_root, expected_root);
+}
+
+mod pedersen_crh_on_affine {
+    use super::*;
+    use snarkos_curves::edwards_bls12::EdwardsAffine as Edwards;
+
+    #[derive(Clone, Debug, PartialEq, Eq)]
+    pub struct Size;
+    impl PedersenSize for Size {
+        const NUM_WINDOWS: usize = 512;
+        const WINDOW_SIZE: usize = 4;
+    }
+
+    #[test]
+    fn empty_merkle_tree_test() {
+        define_merkle_tree_parameters!(MTParameters, PedersenCRH<Edwards, Size>, 32);
+        run_empty_merkle_tree_test::<MTParameters>();
+    }
+
+    #[test]
+    fn good_root_test() {
+        define_merkle_tree_parameters!(MTParameters, PedersenCRH<Edwards, Size>, 32);
+        run_good_root_test::<MTParameters>();
+    }
+
+    #[should_panic]
+    #[test]
+    fn bad_root_test() {
+        define_merkle_tree_parameters!(MTParameters, PedersenCRH<Edwards, Size>, 32);
+        run_bad_root_test::<MTParameters>();
+    }
+
+    #[test]
+    fn depth2_merkle_tree_matches_hashing_test() {
+        define_merkle_tree_parameters!(MTParameters, PedersenCRH<Edwards, Size>, 2);
+        run_merkle_tree_matches_hashing_test::<MTParameters>();
+    }
+
+    #[test]
+    fn depth3_padded_merkle_tree_matches_hashing_test() {
+        define_merkle_tree_parameters!(MTParameters, PedersenCRH<Edwards, Size>, 3);
+        run_padded_merkle_tree_matches_hashing_test::<MTParameters>();
+    }
+}
+
+mod pedersen_crh_on_projective {
+    use super::*;
+    use snarkos_curves::edwards_bls12::EdwardsProjective as Edwards;
+
+    #[derive(Clone, Debug, PartialEq, Eq)]
+    pub struct Size;
+    impl PedersenSize for Size {
+        const NUM_WINDOWS: usize = 512;
+        const WINDOW_SIZE: usize = 4;
+    }
+
+    #[test]
+    fn empty_merkle_tree_test() {
+        define_merkle_tree_parameters!(MTParameters, PedersenCRH<Edwards, Size>, 32);
+        run_empty_merkle_tree_test::<MTParameters>();
+    }
+
+    #[test]
+    fn good_root_test() {
+        define_merkle_tree_parameters!(MTParameters, PedersenCRH<Edwards, Size>, 32);
+        run_good_root_test::<MTParameters>();
+    }
+
+    #[should_panic]
+    #[test]
+    fn bad_root_test() {
+        define_merkle_tree_parameters!(MTParameters, PedersenCRH<Edwards, Size>, 32);
+        run_bad_root_test::<MTParameters>();
+    }
+
+    // TODO (howardwu): Debug why PedersenCRH fails and make this test pass.
+    #[ignore]
+    #[test]
+    fn depth2_merkle_tree_matches_hashing_test() {
+        define_merkle_tree_parameters!(MTParameters, PedersenCRH<Edwards, Size>, 2);
+        run_merkle_tree_matches_hashing_test::<MTParameters>();
+    }
+
+    // TODO (howardwu): Debug why PedersenCRH fails and make this test pass.
+    #[ignore]
+    #[test]
+    fn depth3_padded_merkle_tree_matches_hashing_test() {
+        define_merkle_tree_parameters!(MTParameters, PedersenCRH<Edwards, Size>, 3);
+        run_padded_merkle_tree_matches_hashing_test::<MTParameters>();
+    }
+}
+
+mod pedersen_compressed_crh_on_projective {
+    use super::*;
+    use snarkos_curves::edwards_bls12::EdwardsProjective as Edwards;
+
+    #[derive(Clone, Debug, PartialEq, Eq)]
+    pub struct Size;
+    impl PedersenSize for Size {
+        const NUM_WINDOWS: usize = 512;
+        const WINDOW_SIZE: usize = 4;
+    }
+
+    #[test]
+    fn empty_merkle_tree_test() {
+        define_merkle_tree_parameters!(MTParameters, PedersenCompressedCRH<Edwards, Size>, 32);
+        run_empty_merkle_tree_test::<MTParameters>();
+    }
+
+    #[test]
+    fn good_root_test() {
+        define_merkle_tree_parameters!(MTParameters, PedersenCompressedCRH<Edwards, Size>, 32);
+        run_good_root_test::<MTParameters>();
+    }
+
+    #[should_panic]
+    #[test]
+    fn bad_root_test() {
+        define_merkle_tree_parameters!(MTParameters, PedersenCompressedCRH<Edwards, Size>, 32);
+        run_bad_root_test::<MTParameters>();
+    }
+
+    #[test]
+    fn depth2_merkle_tree_matches_hashing_test() {
+        define_merkle_tree_parameters!(MTParameters, PedersenCompressedCRH<Edwards, Size>, 2);
+        run_merkle_tree_matches_hashing_test::<MTParameters>();
+    }
+
+    #[test]
+    fn depth3_padded_merkle_tree_matches_hashing_test() {
+        define_merkle_tree_parameters!(MTParameters, PedersenCompressedCRH<Edwards, Size>, 3);
+        run_padded_merkle_tree_matches_hashing_test::<MTParameters>();
+    }
 }
diff --git a/errors/src/algorithms/crh.rs b/errors/src/algorithms/crh.rs
index faac58889..f06c7288b 100644
--- a/errors/src/algorithms/crh.rs
+++ b/errors/src/algorithms/crh.rs
@@ -8,7 +8,7 @@ pub enum CRHError {
     #[error("incorrect input length {} x 8 for window params {}x{}", _0, _1, _2)]
     IncorrectInputLength(usize, usize, usize),
 
-    #[error("incorrect pp of size {}x{} for window params {}x{}", _0, _1, _2, _3)]
+    #[error("incorrect parameter size {}x{} for window params {}x{}", _0, _1, _2, _3)]
     IncorrectParameterSize(usize, usize, usize, usize),
 
     #[error("{}", _0)]
diff --git a/errors/src/algorithms/merkle.rs b/errors/src/algorithms/merkle.rs
index 8d4fddce2..973f911f0 100644
--- a/errors/src/algorithms/merkle.rs
+++ b/errors/src/algorithms/merkle.rs
@@ -14,11 +14,11 @@ pub enum MerkleError {
     #[error("Incorrect path length: {}", _0)]
     IncorrectPathLength(usize),
 
-    #[error("Invalid path length: {}. Must be less than: {}", _0, _1)]
+    #[error("Invalid path length: {}. Must be less than or equal to: {}", _0, _1)]
     InvalidPathLength(usize, usize),
 
-    #[error("Invalid tree height: {}. Must be less than or equal to: {}", _0, _1)]
-    InvalidTreeHeight(usize, usize),
+    #[error("Invalid tree depth: {}. Must be less than or equal to: {}", _0, _1)]
+    InvalidTreeDepth(usize, usize),
 
     #[error("{}", _0)]
     Message(String),
diff --git a/gadgets/src/algorithms/merkle_tree/merkle_path.rs b/gadgets/src/algorithms/merkle_tree/merkle_path.rs
index d906cee81..e31a49760 100644
--- a/gadgets/src/algorithms/merkle_tree/merkle_path.rs
+++ b/gadgets/src/algorithms/merkle_tree/merkle_path.rs
@@ -40,7 +40,7 @@ impl<P: MerkleParameters, HG: CRHGadget<P::H, F>, F: Field> MerklePathGadget<P,
         leaf: impl ToBytesGadget<F>,
         should_enforce: &Boolean,
     ) -> Result<(), SynthesisError> {
-        assert_eq!(self.path.len(), P::HEIGHT - 1);
+        assert_eq!(self.path.len(), P::DEPTH);
         // Check that the hash of the given leaf matches the leaf hash in the membership
         // proof.
         let leaf_bits = leaf.to_bytes(&mut cs.ns(|| "leaf_to_bytes"))?;
diff --git a/gadgets/src/algorithms/merkle_tree/tests.rs b/gadgets/src/algorithms/merkle_tree/tests.rs
index a34c6c45e..bb4012cb7 100644
--- a/gadgets/src/algorithms/merkle_tree/tests.rs
+++ b/gadgets/src/algorithms/merkle_tree/tests.rs
@@ -24,7 +24,7 @@ impl PedersenSize for Size {
     const WINDOW_SIZE: usize = 4;
 }
 
-define_merkle_tree_parameters!(EdwardsMerkleParameters, H, 4);
+define_merkle_tree_parameters!(EdwardsMerkleParameters, H, 3);
 
 type EdwardsMerkleTree = MerkleTree<EdwardsMerkleParameters>;
 type H = PedersenCompressedCRH<Edwards, Size>;
@@ -103,7 +103,7 @@ fn generate_masked_merkle_tree(leaves: &[[u8; 30]], use_bad_root: bool) -> () {
 
     let mut cs = TestConstraintSystem::<Fq>::new();
     let leaf_gadgets = tree
-        .leaves_hashed()
+        .hashed_leaves()
         .iter()
         .enumerate()
         .map(|(i, l)| <HG as CRHGadget<H, Fq>>::OutputGadget::alloc(cs.ns(|| format!("leaf {}", i)), || Ok(l)))
@@ -169,7 +169,7 @@ fn bad_root_test() {
 #[test]
 fn good_masked_root_test() {
     let mut leaves = Vec::new();
-    for i in 0..1 << EdwardsMerkleParameters::HEIGHT - 1 {
+    for i in 0..1 << EdwardsMerkleParameters::DEPTH {
         let input = [i; 30];
         leaves.push(input);
     }
@@ -180,7 +180,7 @@ fn good_masked_root_test() {
 #[test]
 fn bad_masked_root_test() {
     let mut leaves = Vec::new();
-    for i in 0..1 << EdwardsMerkleParameters::HEIGHT - 1 {
+    for i in 0..1 << EdwardsMerkleParameters::DEPTH {
         let input = [i; 30];
         leaves.push(input);
     }
diff --git a/models/src/algorithms/crh.rs b/models/src/algorithms/crh.rs
index 9d54939d9..692d8a87a 100644
--- a/models/src/algorithms/crh.rs
+++ b/models/src/algorithms/crh.rs
@@ -2,10 +2,13 @@ use snarkos_errors::algorithms::CRHError;
 use snarkos_utilities::bytes::{FromBytes, ToBytes};
 
 use rand::Rng;
-use std::{fmt::Debug, hash::Hash};
+use std::{
+    fmt::{Debug, Display},
+    hash::Hash,
+};
 
 pub trait CRH: Clone + From<<Self as CRH>::Parameters> {
-    type Output: Clone + Debug + ToBytes + FromBytes + Eq + Hash + Default;
+    type Output: Clone + Debug + Display + ToBytes + FromBytes + Eq + Hash + Default;
     type Parameters: Clone + Debug + ToBytes + FromBytes + Eq;
 
     const INPUT_SIZE_BITS: usize;
diff --git a/models/src/algorithms/merkle_parameters.rs b/models/src/algorithms/merkle_tree.rs
similarity index 98%
rename from models/src/algorithms/merkle_parameters.rs
rename to models/src/algorithms/merkle_tree.rs
index 096fd0e40..b58a3b287 100644
--- a/models/src/algorithms/merkle_parameters.rs
+++ b/models/src/algorithms/merkle_tree.rs
@@ -9,7 +9,7 @@ use std::io::Cursor;
 pub trait MerkleParameters: Clone + Default + From<<Self as MerkleParameters>::H> {
     type H: CRH;
 
-    const HEIGHT: usize;
+    const DEPTH: usize;
 
     /// Setup the MerkleParameters
     fn setup<R: Rng>(rng: &mut R) -> Self;
@@ -40,6 +40,7 @@ pub trait MerkleParameters: Clone + Default + From<<Self as MerkleParameters>::H
 
         // Construct left input.
         left.write(&mut writer)?;
+
         // Construct right input.
         right.write(&mut writer)?;
 
diff --git a/models/src/algorithms/mod.rs b/models/src/algorithms/mod.rs
index 784d799d7..aa8772c03 100644
--- a/models/src/algorithms/mod.rs
+++ b/models/src/algorithms/mod.rs
@@ -4,8 +4,8 @@ pub use self::commitment::*;
 pub mod crh;
 pub use self::crh::*;
 
-pub mod merkle_parameters;
-pub use self::merkle_parameters::*;
+pub mod merkle_tree;
+pub use self::merkle_tree::*;
 
 pub mod prf;
 pub use self::prf::*;
diff --git a/objects/src/pedersen_merkle_tree.rs b/objects/src/pedersen_merkle_tree.rs
index 96656721a..8ba08a76a 100644
--- a/objects/src/pedersen_merkle_tree.rs
+++ b/objects/src/pedersen_merkle_tree.rs
@@ -21,10 +21,10 @@ mod window {
 
 pub type MerkleTreeCRH = PedersenCompressedCRH<EdwardsBls, window::TwoToOneWindow>;
 
-// We instantiate the tree here with height = 3. This may change in the future.
-pub const MASKED_TREE_HEIGHT: usize = 3;
+// We instantiate the tree here with depth = 2. This may change in the future.
+pub const MASKED_TREE_DEPTH: usize = 2;
 
-define_merkle_tree_parameters!(MaskedMerkleTreeParameters, MerkleTreeCRH, MASKED_TREE_HEIGHT);
+define_merkle_tree_parameters!(MaskedMerkleTreeParameters, MerkleTreeCRH, MASKED_TREE_DEPTH);
 
 /// A Merkle Tree instantiated with the Masked Pedersen hasher over BLS12-377
 pub type EdwardsMaskedMerkleTree = MerkleTree<MaskedMerkleTreeParameters>;
@@ -64,7 +64,7 @@ pub fn pedersen_merkle_root_hash(hashes: &[Vec<u8>]) -> Fr {
 /// base layer hashes leaved
 pub fn pedersen_merkle_root_hash_with_leaves(hashes: &[Vec<u8>]) -> (Fr, Vec<Fr>) {
     let tree = EdwardsMaskedMerkleTree::new(PARAMS.clone(), hashes).expect("could not create merkle tree");
-    (tree.root(), tree.leaves_hashed())
+    (tree.root(), tree.hashed_leaves())
 }
 
 impl From<Fr> for PedersenMerkleRootHash {
diff --git a/objects/src/posw.rs b/objects/src/posw.rs
index e2134de58..f8fbf185b 100644
--- a/objects/src/posw.rs
+++ b/objects/src/posw.rs
@@ -15,7 +15,7 @@ use std::{
 };
 
 // 2 * G1 + 1 * G2 assuming Bls12-377 and GM17.
-// Marlin requires 13 * G1 + 21 * Fq = 1296 btyes.
+// Marlin requires 13 * G1 + 21 * Fq = 1296 bytes.
 // Points are encoded with an extra byte for the infinity bit.
 const PROOF_SIZE: usize = 387;
 
diff --git a/parameters/src/genesis/block_header/block_header.genesis b/parameters/src/genesis/block_header/block_header.genesis
index a91b8bed2..8a77811f6 100644
Binary files a/parameters/src/genesis/block_header/block_header.genesis and b/parameters/src/genesis/block_header/block_header.genesis differ
diff --git a/parameters/src/genesis/transaction_1/transaction_1.genesis b/parameters/src/genesis/transaction_1/transaction_1.genesis
index f00984485..de2564eac 100644
Binary files a/parameters/src/genesis/transaction_1/transaction_1.genesis and b/parameters/src/genesis/transaction_1/transaction_1.genesis differ
diff --git a/parameters/src/params.rs b/parameters/src/params.rs
index edf954bc5..719d82981 100644
--- a/parameters/src/params.rs
+++ b/parameters/src/params.rs
@@ -45,138 +45,139 @@ macro_rules! impl_params {
 macro_rules! impl_params_remote {
     ($name: ident, $fname: tt, $size: tt) => {
 
-    pub struct $name;
+        pub struct $name;
 
-    impl Parameters for $name {
-        const CHECKSUM: &'static str = include_str!(concat!("params/", $fname, ".checksum"));
-        const SIZE: u64 = $size;
+        impl Parameters for $name {
+            const CHECKSUM: &'static str = include_str!(concat!("params/", $fname, ".checksum"));
+            const SIZE: u64 = $size;
 
-        fn load_bytes() -> Result<Vec<u8>, ParametersError> {
-            // Compose the correct file path for the parameter file.
-            let filename = Self::versioned_filename();
-            let mut file_path = PathBuf::from(file!());
-            file_path.pop();
-            file_path.push("params/");
-            file_path.push(&filename);
+            fn load_bytes() -> Result<Vec<u8>, ParametersError> {
+                // Compose the correct file path for the parameter file.
+                let filename = Self::versioned_filename();
+                let mut file_path = PathBuf::from(file!());
+                file_path.pop();
+                file_path.push("params/");
+                file_path.push(&filename);
 
-            // Compute the relative path.
-            let relative_path = file_path.strip_prefix("parameters")?.to_path_buf();
+                // Compute the relative path.
+                let relative_path = file_path.strip_prefix("parameters")?.to_path_buf();
 
-            // Compute the absolute path.
-            let mut absolute_path = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
-            absolute_path.push(&relative_path);
+                // Compute the absolute path.
+                let mut absolute_path = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
+                absolute_path.push(&relative_path);
 
-            let buffer = if relative_path.exists() {
-                // Attempts to load the parameter file locally with a relative path.
-                fs::read(relative_path)?.to_vec()
-            } else if absolute_path.exists() {
-                // Attempts to load the parameter file locally with an absolute path.
-                fs::read(absolute_path)?.to_vec()
-            } else {
-                // Downloads the missing parameters and stores it in the local directory for use.
-                eprintln!(
-                    "\nWARNING - \"{}\" does not exist. snarkOS will download this file remotely and store it locally. Please ensure \"{}\" is stored in {:?}.\n",
-                    filename, filename, file_path
-                );
-                let output = Self::load_remote()?;
-                match Self::store_bytes(&output, &relative_path, &absolute_path, &file_path) {
-                    Ok(()) => output,
-                    Err(_) => {
-                        eprintln!(
-                            "\nWARNING - Failed to store \"{}\" locally. Please download this file manually and ensure it is stored in {:?}.\n",
-                            filename, file_path
-                        );
-                        output
+                let buffer = if relative_path.exists() {
+                    // Attempts to load the parameter file locally with a relative path.
+                    fs::read(relative_path)?.to_vec()
+                } else if absolute_path.exists() {
+                    // Attempts to load the parameter file locally with an absolute path.
+                    fs::read(absolute_path)?.to_vec()
+                } else {
+                    // Downloads the missing parameters and stores it in the local directory for use.
+                    eprintln!(
+                        "\nWARNING - \"{}\" does not exist. snarkOS will download this file remotely and store it locally. Please ensure \"{}\" is stored in {:?}.\n",
+                        filename, filename, file_path
+                    );
+                    let output = Self::load_remote()?;
+                    match Self::store_bytes(&output, &relative_path, &absolute_path, &file_path) {
+                        Ok(()) => output,
+                        Err(_) => {
+                            eprintln!(
+                                "\nWARNING - Failed to store \"{}\" locally. Please download this file manually and ensure it is stored in {:?}.\n",
+                                filename, file_path
+                            );
+                            output
+                        }
                     }
-                }
-            };
+                };
 
-            let checksum = hex::encode(sha256(&buffer));
-            match Self::CHECKSUM == checksum {
-                true => Ok(buffer),
-                false => Err(ParametersError::ChecksumMismatch(Self::CHECKSUM.into(), checksum)),
+                let checksum = hex::encode(sha256(&buffer));
+                match Self::CHECKSUM == checksum {
+                    true => Ok(buffer),
+                    false => Err(ParametersError::ChecksumMismatch(Self::CHECKSUM.into(), checksum)),
+                }
             }
         }
-    }
 
-    impl $name {
-        #[cfg(any(test, feature = "remote"))]
-        pub fn load_remote() -> Result<Vec<u8>, ParametersError> {
-            println!("{} - Downloading parameters...", module_path!());
-            let mut buffer = vec![];
-            let url = Self::remote_url();
-            Self::remote_fetch(&mut buffer, &url)?;
-            println!("\n{} - Download complete", module_path!());
+        impl $name {
+            #[cfg(any(test, feature = "remote"))]
+            pub fn load_remote() -> Result<Vec<u8>, ParametersError> {
+                println!("{} - Downloading parameters...", module_path!());
+                let mut buffer = vec![];
+                let url = Self::remote_url();
+                Self::remote_fetch(&mut buffer, &url)?;
+                println!("\n{} - Download complete", module_path!());
 
-            // Verify the checksum of the remote data before returning
-            let checksum = hex::encode(sha256(&buffer));
-            match Self::CHECKSUM == checksum {
-                true => Ok(buffer),
-                false => Err(ParametersError::ChecksumMismatch(Self::CHECKSUM.into(), checksum)),
+                // Verify the checksum of the remote data before returning
+                let checksum = hex::encode(sha256(&buffer));
+                match Self::CHECKSUM == checksum {
+                    true => Ok(buffer),
+                    false => Err(ParametersError::ChecksumMismatch(Self::CHECKSUM.into(), checksum)),
+                }
             }
-        }
 
-        #[cfg(not(any(test, feature = "remote")))]
-        pub fn load_remote() -> Result<Vec<u8>, ParametersError> {
-            Err(ParametersError::RemoteFetchDisabled)
-        }
+            #[cfg(not(any(test, feature = "remote")))]
+            pub fn load_remote() -> Result<Vec<u8>, ParametersError> {
+                Err(ParametersError::RemoteFetchDisabled)
+            }
 
-        fn versioned_filename() -> String {
-            match Self::CHECKSUM.get(0..7) {
-                Some(sum) => format!("{}-{}.params", $fname, sum),
-                _ => concat!($fname, ".params",).to_string()
+            fn versioned_filename() -> String {
+                match Self::CHECKSUM.get(0..7) {
+                    Some(sum) => format!("{}-{}.params", $fname, sum),
+                    _ => concat!($fname, ".params",).to_string()
+                }
             }
-        }
 
-        #[cfg(any(test, feature = "remote"))]
-        fn remote_url() -> String {
-            format!("{}/{}", REMOTE_URL, Self::versioned_filename())
-        }
+            #[cfg(any(test, feature = "remote"))]
+            fn remote_url() -> String {
+                format!("{}/{}", REMOTE_URL, Self::versioned_filename())
+            }
 
-        fn store_bytes(
-            buffer: &Vec<u8>,
-            relative_path: &PathBuf,
-            absolute_path: &PathBuf,
-            file_path: &PathBuf,
-        ) -> Result<(), ParametersError> {
-            println!("{} - Storing parameters ({:?})", module_path!(), file_path);
-            // Attempt to write the parameter buffer to a file.
-            if let Ok(mut file) = File::create(relative_path) {
-                file.write_all(&buffer)?;
-                drop(file);
-            } else if let Ok(mut file) = File::create(absolute_path) {
-                file.write_all(&buffer)?;
-                drop(file);
+            fn store_bytes(
+                buffer: &Vec<u8>,
+                relative_path: &PathBuf,
+                absolute_path: &PathBuf,
+                file_path: &PathBuf,
+            ) -> Result<(), ParametersError> {
+                println!("{} - Storing parameters ({:?})", module_path!(), file_path);
+                // Attempt to write the parameter buffer to a file.
+                if let Ok(mut file) = File::create(relative_path) {
+                    file.write_all(&buffer)?;
+                    drop(file);
+                } else if let Ok(mut file) = File::create(absolute_path) {
+                    file.write_all(&buffer)?;
+                    drop(file);
+                }
+                Ok(())
             }
-            Ok(())
-        }
 
-        #[cfg(any(test, feature = "remote"))]
-        fn remote_fetch(buffer: &mut Vec<u8>, url: &str) -> Result<(), ParametersError> {
-            let mut easy = Easy::new();
-            easy.url(url)?;
-            easy.progress(true)?;
-            easy.progress_function(|total_download, current_download, _, _| {
-                let percent = (current_download / total_download) * 100.0;
-                let size_in_megabytes = total_download as u64 / 1_048_576;
-                print!(
-                    "\r{} - {:.2}% complete ({:#} MB total)",
-                    module_path!(),
-                    percent,
-                    size_in_megabytes
-                );
-                true
-            })?;
+            #[cfg(any(test, feature = "remote"))]
+            fn remote_fetch(buffer: &mut Vec<u8>, url: &str) -> Result<(), ParametersError> {
+                let mut easy = Easy::new();
+                easy.url(url)?;
+                easy.progress(true)?;
+                easy.progress_function(|total_download, current_download, _, _| {
+                    let percent = (current_download / total_download) * 100.0;
+                    let size_in_megabytes = total_download as u64 / 1_048_576;
+                    print!(
+                        "\r{} - {:.2}% complete ({:#} MB total)",
+                        module_path!(),
+                        percent,
+                        size_in_megabytes
+                    );
+                    true
+                })?;
 
-            let mut transfer = easy.transfer();
-            transfer.write_function(|data| {
-                buffer.extend_from_slice(data);
-                Ok(data.len())
-            })?;
-            Ok(transfer.perform()?)
+                let mut transfer = easy.transfer();
+                transfer.write_function(|data| {
+                    buffer.extend_from_slice(data);
+                    Ok(data.len())
+                })?;
+                Ok(transfer.perform()?)
+            }
         }
     }
-}}
+}
 
 // Commitments
 impl_params!(
@@ -250,7 +251,7 @@ impl_params!(
 );
 
 // Inner SNARK
-impl_params_remote!(InnerSNARKPKParameters, "inner_snark_pk", 545102262);
+impl_params_remote!(InnerSNARKPKParameters, "inner_snark_pk", 539664038);
 impl_params!(InnerSNARKVKParameters, inner_snark_vk_test, "inner_snark_vk", 2523);
 
 // Outer SNARK
diff --git a/parameters/src/params/account_commitment.checksum b/parameters/src/params/account_commitment.checksum
index 001e2f2e1..bd03f2bbb 100644
--- a/parameters/src/params/account_commitment.checksum
+++ b/parameters/src/params/account_commitment.checksum
@@ -1 +1 @@
-389bed6840af0699a55edef05c6a2b67e237b5abf0bec4ecd13a6e7f3c87ce37
\ No newline at end of file
+723ee2319e1a907912dda153108ca06f46fa2704fec945a681a86890d2b69e55
\ No newline at end of file
diff --git a/parameters/src/params/account_commitment.params b/parameters/src/params/account_commitment.params
index 63e5db919..7265ed14e 100644
Binary files a/parameters/src/params/account_commitment.params and b/parameters/src/params/account_commitment.params differ
diff --git a/parameters/src/params/account_signature.checksum b/parameters/src/params/account_signature.checksum
index 25cc19ad5..8a809c82e 100644
--- a/parameters/src/params/account_signature.checksum
+++ b/parameters/src/params/account_signature.checksum
@@ -1 +1 @@
-352b2ded0bb3b690f5afbe7245719b958e1caf4371e7bdb461196bb0994cece6
\ No newline at end of file
+f098ffef542b08b8b6236e7aad50ba05c98fdc29f80515c890bfc5ce68fa0b09
\ No newline at end of file
diff --git a/parameters/src/params/account_signature.params b/parameters/src/params/account_signature.params
index 80539f96b..d57cdc10e 100644
Binary files a/parameters/src/params/account_signature.params and b/parameters/src/params/account_signature.params differ
diff --git a/parameters/src/params/inner_snark_pk.checksum b/parameters/src/params/inner_snark_pk.checksum
index 0710b31ae..87f22b927 100644
--- a/parameters/src/params/inner_snark_pk.checksum
+++ b/parameters/src/params/inner_snark_pk.checksum
@@ -1 +1 @@
-bad6b67850848ef6a72208455310058b5e906f1aabca56eff2675782b2eae474
\ No newline at end of file
+c035370a1190499e65b649507da149e38642c6477f78eb2123c3ededd100d3c7
\ No newline at end of file
diff --git a/parameters/src/params/inner_snark_vk.checksum b/parameters/src/params/inner_snark_vk.checksum
index 9a59486e6..f7f750079 100644
--- a/parameters/src/params/inner_snark_vk.checksum
+++ b/parameters/src/params/inner_snark_vk.checksum
@@ -1 +1 @@
-a05e8955504c330a2fb974a83d4e4c37f9d0c1be1a8e16d6a2a06832857754d7
\ No newline at end of file
+8c5e84a3078f5eccb314f91a1e2eddb8ceb80521c94030f781e1a8063ace2974
\ No newline at end of file
diff --git a/parameters/src/params/inner_snark_vk.params b/parameters/src/params/inner_snark_vk.params
index cd40478ab..4f67a1b9d 100644
Binary files a/parameters/src/params/inner_snark_vk.params and b/parameters/src/params/inner_snark_vk.params differ
diff --git a/parameters/src/params/ledger_merkle_tree.checksum b/parameters/src/params/ledger_merkle_tree.checksum
index 6abf3ca0c..4d5fc6ff0 100644
--- a/parameters/src/params/ledger_merkle_tree.checksum
+++ b/parameters/src/params/ledger_merkle_tree.checksum
@@ -1 +1 @@
-5f68420ab07c07e47cf40d3c80dad67d4c6cccfe1d252cbe691bf2a88b96ae20
\ No newline at end of file
+769495da039c33b61eb8b22ceff830d4e6c7a10e302846430eecc0410eeba83d
\ No newline at end of file
diff --git a/parameters/src/params/ledger_merkle_tree.params b/parameters/src/params/ledger_merkle_tree.params
index 857ba3f73..8d71a797b 100644
Binary files a/parameters/src/params/ledger_merkle_tree.params and b/parameters/src/params/ledger_merkle_tree.params differ
diff --git a/parameters/src/params/local_data_commitment.checksum b/parameters/src/params/local_data_commitment.checksum
index b21a99023..e551bb80b 100644
--- a/parameters/src/params/local_data_commitment.checksum
+++ b/parameters/src/params/local_data_commitment.checksum
@@ -1 +1 @@
-aa7f686b962f9860ac81b625d97d2dbbc89bde67b7f0a1b3177f21a27f08adbf
\ No newline at end of file
+e583f5840f45e0bf376e71c994c89a3a649c46c331b0ef6faa66b1af3cfcdf9a
\ No newline at end of file
diff --git a/parameters/src/params/local_data_commitment.params b/parameters/src/params/local_data_commitment.params
index f8ca658ff..698996a39 100644
Binary files a/parameters/src/params/local_data_commitment.params and b/parameters/src/params/local_data_commitment.params differ
diff --git a/parameters/src/params/local_data_crh.checksum b/parameters/src/params/local_data_crh.checksum
index f884206e6..0701e04c4 100644
--- a/parameters/src/params/local_data_crh.checksum
+++ b/parameters/src/params/local_data_crh.checksum
@@ -1 +1 @@
-0f443934334b49ba9e5bd0a6a0481cd6841d124c3de6b0272792e6b272c287ff
\ No newline at end of file
+993e1834f2f78c1eb8ecb2a0337254d8d3625c17a6d31d64a1218d98b2102bf8
\ No newline at end of file
diff --git a/parameters/src/params/local_data_crh.params b/parameters/src/params/local_data_crh.params
index ac5a54422..db939c772 100644
Binary files a/parameters/src/params/local_data_crh.params and b/parameters/src/params/local_data_crh.params differ
diff --git a/parameters/src/params/outer_snark_pk.checksum b/parameters/src/params/outer_snark_pk.checksum
index 3363dc6f8..95d1ba102 100644
--- a/parameters/src/params/outer_snark_pk.checksum
+++ b/parameters/src/params/outer_snark_pk.checksum
@@ -1 +1 @@
-eaff518e730e73a93f0dc3a5f19ac3a4e28ad000f08031d207239757cc9fcb54
\ No newline at end of file
+7f092711604c73b0093c950a47915f279a0fa010a3075866c1514cc740d64ce7
\ No newline at end of file
diff --git a/parameters/src/params/outer_snark_vk.checksum b/parameters/src/params/outer_snark_vk.checksum
index ae0047638..f7a979ace 100644
--- a/parameters/src/params/outer_snark_vk.checksum
+++ b/parameters/src/params/outer_snark_vk.checksum
@@ -1 +1 @@
-a755d0a10563ff4f7a4585a95f4bc93e2828058c6a28ff7c3ed85371ddeb2e4a
\ No newline at end of file
+ae38d33ddac51aeace587fbf88ad5a9578676a5f978e820bd1cb251372826cf8
\ No newline at end of file
diff --git a/parameters/src/params/outer_snark_vk.params b/parameters/src/params/outer_snark_vk.params
index 8edf957d0..1c3c8eb95 100644
Binary files a/parameters/src/params/outer_snark_vk.params and b/parameters/src/params/outer_snark_vk.params differ
diff --git a/parameters/src/params/posw_snark_pk.checksum b/parameters/src/params/posw_snark_pk.checksum
index b8357283b..a49fc755e 100644
--- a/parameters/src/params/posw_snark_pk.checksum
+++ b/parameters/src/params/posw_snark_pk.checksum
@@ -1 +1 @@
-9328c6d6a61c056881eac0f80e64b42eca738b5efb534d3b76728dcb5699b7fb
\ No newline at end of file
+031f7d597785aa041903a1fab48d6f2ec6e13da84d10bc76c4dc00431d75d63a
\ No newline at end of file
diff --git a/parameters/src/params/posw_snark_pk.params b/parameters/src/params/posw_snark_pk.params
index 1d5297bd7..3802a64cb 100644
Binary files a/parameters/src/params/posw_snark_pk.params and b/parameters/src/params/posw_snark_pk.params differ
diff --git a/parameters/src/params/posw_snark_vk.checksum b/parameters/src/params/posw_snark_vk.checksum
index fbd9cf003..bbd2fad27 100644
--- a/parameters/src/params/posw_snark_vk.checksum
+++ b/parameters/src/params/posw_snark_vk.checksum
@@ -1 +1 @@
-e7cef5a8a2270c12c21e20994dc5ac1ff0b65435e68d2049ba5a61e7aa1fb0a4
\ No newline at end of file
+f00d86991c8763cddf671d15c4b7fc87b4949f11ce9bda3f5e756f3fc0e0f536
\ No newline at end of file
diff --git a/parameters/src/params/posw_snark_vk.params b/parameters/src/params/posw_snark_vk.params
index 362e033ef..ef2d1273d 100644
Binary files a/parameters/src/params/posw_snark_vk.params and b/parameters/src/params/posw_snark_vk.params differ
diff --git a/parameters/src/params/predicate_snark_pk.checksum b/parameters/src/params/predicate_snark_pk.checksum
index 273943012..a4d0ba1a5 100644
--- a/parameters/src/params/predicate_snark_pk.checksum
+++ b/parameters/src/params/predicate_snark_pk.checksum
@@ -1 +1 @@
-85055f243df823ccc3dc01ff0b4f2166a9fcff2137fd2b2bc16dcff2e96aedde
\ No newline at end of file
+626c3a078eae1f4587360b55c7dbc46e805a7a2531bcb13fe9be4610d44b3e34
\ No newline at end of file
diff --git a/parameters/src/params/predicate_snark_pk.params b/parameters/src/params/predicate_snark_pk.params
index 92f555284..c215f5daf 100644
Binary files a/parameters/src/params/predicate_snark_pk.params and b/parameters/src/params/predicate_snark_pk.params differ
diff --git a/parameters/src/params/predicate_snark_vk.checksum b/parameters/src/params/predicate_snark_vk.checksum
index d84563ccb..865061947 100644
--- a/parameters/src/params/predicate_snark_vk.checksum
+++ b/parameters/src/params/predicate_snark_vk.checksum
@@ -1 +1 @@
-3377c2ddee2aa03bd4fa96bb41c8f3f32bcad1fabff1d33f96adacc8fc543e4f
\ No newline at end of file
+254064975a32cdc0921acb7c3e10befeeb39771df6bdc73efe22c42fe1e9f7fa
\ No newline at end of file
diff --git a/parameters/src/params/predicate_snark_vk.params b/parameters/src/params/predicate_snark_vk.params
index 09a2b0394..23927cd62 100644
Binary files a/parameters/src/params/predicate_snark_vk.params and b/parameters/src/params/predicate_snark_vk.params differ
diff --git a/parameters/src/params/predicate_vk_crh.checksum b/parameters/src/params/predicate_vk_crh.checksum
index 68048ac23..0aa7f9a07 100644
--- a/parameters/src/params/predicate_vk_crh.checksum
+++ b/parameters/src/params/predicate_vk_crh.checksum
@@ -1 +1 @@
-4fe54935316982911a4c27f32617d120243e048c494c2b0944d6c52a044201b7
\ No newline at end of file
+4a3c71748a43699ad3c0279431af7d8734b3155880ab840339f4f88944daf8eb
\ No newline at end of file
diff --git a/parameters/src/params/predicate_vk_crh.params b/parameters/src/params/predicate_vk_crh.params
index f7e391c14..b0d2332f4 100644
Binary files a/parameters/src/params/predicate_vk_crh.params and b/parameters/src/params/predicate_vk_crh.params differ
diff --git a/parameters/src/params/record_commitment.checksum b/parameters/src/params/record_commitment.checksum
index 5a1b1ec30..f3e52e7b5 100644
--- a/parameters/src/params/record_commitment.checksum
+++ b/parameters/src/params/record_commitment.checksum
@@ -1 +1 @@
-e108690eee222d921f55055cd5a936cf4a38902a1c58b9abd1d0859746de5e4c
\ No newline at end of file
+94c9b86728ad11b9eb7c385542371819bf1cd1a04e916de76a1ee8395daa126a
\ No newline at end of file
diff --git a/parameters/src/params/record_commitment.params b/parameters/src/params/record_commitment.params
index 522d19142..b23909670 100644
Binary files a/parameters/src/params/record_commitment.params and b/parameters/src/params/record_commitment.params differ
diff --git a/parameters/src/params/serial_number_nonce_crh.checksum b/parameters/src/params/serial_number_nonce_crh.checksum
index 79617f370..8195ea72d 100644
--- a/parameters/src/params/serial_number_nonce_crh.checksum
+++ b/parameters/src/params/serial_number_nonce_crh.checksum
@@ -1 +1 @@
-5b5c07cf0b3b884a6f38d95637a2c806084d417c0b37bef21a6b81b9eccac12d
\ No newline at end of file
+9287a2f414b872b343e9196b328b2c8db81f0d00f280203eda32d06b3d6754d6
\ No newline at end of file
diff --git a/parameters/src/params/serial_number_nonce_crh.params b/parameters/src/params/serial_number_nonce_crh.params
index ff941685c..caf82a805 100644
Binary files a/parameters/src/params/serial_number_nonce_crh.params and b/parameters/src/params/serial_number_nonce_crh.params differ
diff --git a/parameters/src/params/value_commitment.checksum b/parameters/src/params/value_commitment.checksum
index 5e93101d7..421f368e7 100644
--- a/parameters/src/params/value_commitment.checksum
+++ b/parameters/src/params/value_commitment.checksum
@@ -1 +1 @@
-b680f07c49a2cacf14a0c49081ffb2794943e82bb692d9dca9a046a229bf90d7
\ No newline at end of file
+2018cae394bb940ca1647b4d7510bd0a624bb5bb881f9e409243b7789c89bc93
\ No newline at end of file
diff --git a/parameters/src/params/value_commitment.params b/parameters/src/params/value_commitment.params
index e19537a13..6f2e43189 100644
Binary files a/parameters/src/params/value_commitment.params and b/parameters/src/params/value_commitment.params differ
diff --git a/posw/src/circuit.rs b/posw/src/circuit.rs
index 3306fe7e1..d29ebef0a 100644
--- a/posw/src/circuit.rs
+++ b/posw/src/circuit.rs
@@ -49,9 +49,7 @@ impl<F: PrimeField, M: MerkleParameters, HG: MaskedCRHGadget<M::H, F>, CP: POSWC
                 Ok(crh_parameters)
             })?;
 
-        // According to the native tree in https://github.com/AleoHQ/snarkOS/blob/master/algorithms/src/merkle_tree/merkle_tree.rs,
-        // the tree height is calculated as ceil(log2(num_leaves)) + 1
-        let leaves_number = 2u32.pow(M::HEIGHT as u32 - 1) as usize;
+        let leaves_number = 2u32.pow(M::DEPTH as u32) as usize;
         assert!(self.leaves.len() <= leaves_number);
 
         // Initialize the leaves.
@@ -122,7 +120,7 @@ mod test {
     }
 
     // We use a small tree in this test
-    define_merkle_tree_parameters!(EdwardsMaskedMerkleParameters, PedersenCompressedCRH<Edwards, Size>, 5);
+    define_merkle_tree_parameters!(EdwardsMaskedMerkleParameters, PedersenCompressedCRH<Edwards, Size>, 4);
 
     type HashGadget = PedersenCompressedCRHGadget<Edwards, Fq, EdwardsBlsGadget>;
     type EdwardsMaskedMerkleTree = MerkleTree<EdwardsMaskedMerkleParameters>;
@@ -158,7 +156,7 @@ mod test {
         h.input(root_bytes.as_ref());
         let mask = h.result().to_vec();
 
-        let snark_leaves = tree.leaves_hashed().into_iter().map(Some).collect();
+        let snark_leaves = tree.hashed_leaves().into_iter().map(Some).collect();
         let proof = create_random_proof(
             POSWCircuit::<_, EdwardsMaskedMerkleParameters, HashGadget, TestPOSWCircuitParameters> {
                 leaves: snark_leaves,
diff --git a/posw/src/lib.rs b/posw/src/lib.rs
index 81642aaa5..7bd3cc534 100644
--- a/posw/src/lib.rs
+++ b/posw/src/lib.rs
@@ -12,7 +12,7 @@ use snarkos_objects::{
     pedersen_merkle_root,
     MerkleRootHash,
     PedersenMerkleRootHash,
-    MASKED_TREE_HEIGHT,
+    MASKED_TREE_DEPTH,
 };
 
 /// PoSW instantiated over BLS12-377 with GM17.
@@ -39,7 +39,7 @@ mod params {
 
 /// Subtree calculation
 pub fn txids_to_roots(transaction_ids: &[Vec<u8>]) -> (MerkleRootHash, PedersenMerkleRootHash, Vec<Vec<u8>>) {
-    let (root, subroots) = merkle_root_with_subroots(transaction_ids, MASKED_TREE_HEIGHT);
+    let (root, subroots) = merkle_root_with_subroots(transaction_ids, MASKED_TREE_DEPTH);
     let mut merkle_root_bytes = [0u8; 32];
     merkle_root_bytes[..].copy_from_slice(&root);
 
diff --git a/testing/src/consensus/test_data b/testing/src/consensus/test_data
index d1f12ec82..efb8d31a9 100644
Binary files a/testing/src/consensus/test_data and b/testing/src/consensus/test_data differ
