diff --git a/console/program/src/request/bytes.rs b/console/program/src/request/bytes.rs
index 6c800c989..2cc6f0615 100644
--- a/console/program/src/request/bytes.rs
+++ b/console/program/src/request/bytes.rs
@@ -23,7 +23,7 @@ impl<N: Network> FromBytes for Request<N> {
         let version = u8::read_le(&mut reader)?;
         // Ensure the version is valid.
         if version != 0 {
-            return Err(error("Invalid verifying key version"));
+            return Err(error("Invalid request version"));
         }
 
         // Read the caller.
diff --git a/console/program/src/state_path/bytes.rs b/console/program/src/state_path/bytes.rs
index 8934832dc..14356595f 100644
--- a/console/program/src/state_path/bytes.rs
+++ b/console/program/src/state_path/bytes.rs
@@ -23,7 +23,7 @@ impl<N: Network> FromBytes for StatePath<N> {
         let version = u8::read_le(&mut reader)?;
         // Ensure the version is valid.
         if version != 0 {
-            return Err(error("Invalid block version"));
+            return Err(error("Invalid state path version"));
         }
 
         // Read the state path.
diff --git a/parameters/examples/genesis.rs b/parameters/examples/genesis.rs
deleted file mode 100644
index bb736babf..000000000
--- a/parameters/examples/genesis.rs
+++ /dev/null
@@ -1,70 +0,0 @@
-// Copyright (C) 2019-2023 Aleo Systems Inc.
-// This file is part of the snarkVM library.
-
-// The snarkVM library is free software: you can redistribute it and/or modify
-// it under the terms of the GNU General Public License as published by
-// the Free Software Foundation, either version 3 of the License, or
-// (at your option) any later version.
-
-// The snarkVM library is distributed in the hope that it will be useful,
-// but WITHOUT ANY WARRANTY; without even the implied warranty of
-// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
-// GNU General Public License for more details.
-
-// You should have received a copy of the GNU General Public License
-// along with the snarkVM library. If not, see <https://www.gnu.org/licenses/>.
-
-use snarkvm_console::{account::PrivateKey, network::Testnet3, prelude::*};
-use snarkvm_synthesizer::{store::helpers::memory::ConsensusMemory, Block, ConsensusStore, VM};
-
-use rand::thread_rng;
-use std::{
-    fs::File,
-    io::{Result as IoResult, Write},
-    path::Path,
-};
-
-pub fn generate<N: Network>(private_key: PrivateKey<N>) -> Result<Vec<u8>> {
-    // Initialize the consensus store.
-    let store = ConsensusStore::<N, ConsensusMemory<N>>::open(None)?;
-    // Initialize the VM.
-    let vm = VM::from(store)?;
-    // Create a genesis block.
-    let genesis_block = Block::genesis(&vm, &private_key, &mut thread_rng())?;
-    // assert!(genesis_block.verify(&VM::<N>::new()?));
-    assert!(genesis_block.is_genesis());
-    assert!(genesis_block.header().is_genesis());
-
-    println!("\n{}\n", serde_json::to_string_pretty(&genesis_block)?);
-
-    println!("Genesis block size - {}\n", genesis_block.to_bytes_le()?.len());
-    println!("Genesis block header size - {}\n", genesis_block.header().to_bytes_le()?.len());
-
-    genesis_block.to_bytes_le()
-}
-
-pub fn store<P: AsRef<Path>>(path: P, bytes: &[u8]) -> IoResult<()> {
-    let mut file = File::create(path)?;
-    file.write_all(bytes)?;
-    Ok(())
-}
-
-#[allow(clippy::match_single_binding)]
-pub fn main() {
-    let args: Vec<String> = std::env::args().collect();
-    if args.len() < 4 {
-        println!("Invalid number of arguments. Given: {} - Required: 3", args.len() - 1);
-        return;
-    }
-
-    match args[1].as_str() {
-        "testnet3" => {
-            let private_key = PrivateKey::from_str(&args[2]).unwrap();
-            let genesis_file = &args[3];
-
-            let genesis_block = generate::<Testnet3>(private_key).unwrap();
-            store(genesis_file, &genesis_block).unwrap();
-        }
-        _ => panic!("Invalid network"),
-    };
-}
diff --git a/parameters/scripts/README.md b/parameters/scripts/README.md
index 42171b7d0..12185721d 100644
--- a/parameters/scripts/README.md
+++ b/parameters/scripts/README.md
@@ -4,8 +4,3 @@ These scripts will run the parameter setup programs in the `examples` folder and
 and `.checksum` files to `testnet*` folder under the `src` directory.
 
 If the parameter size has changed, you will need to manually update these in each corresponding struct.
-
-To perform the genesis block generation only, run:
-```$xslt
-./genesis.sh
-```
diff --git a/parameters/scripts/testnet3/genesis.sh b/parameters/scripts/testnet3/genesis.sh
deleted file mode 100755
index bc5e51092..000000000
--- a/parameters/scripts/testnet3/genesis.sh
+++ /dev/null
@@ -1,11 +0,0 @@
-# This script will run the transaction and block generation programs in the `examples` folder and move the resulting `.genesis` files
-# to their respective folders under the `src` directory.
-# If the transaction size or checksum has changed, you will need to manually update these in each corresponding struct.
-
-# Generate transactions
-
-# Inputs: network, recipient private key, genesis filepath, transaction filepath
-
-cargo run --release --example genesis testnet3 $1 block.genesis -- --nocapture || exit
-
-mv block.genesis ../../src/testnet3/resources || exit
diff --git a/synthesizer/src/block/transactions/bytes.rs b/synthesizer/src/block/transactions/bytes.rs
index 29034f4fc..5cba2d7c2 100644
--- a/synthesizer/src/block/transactions/bytes.rs
+++ b/synthesizer/src/block/transactions/bytes.rs
@@ -24,7 +24,7 @@ impl<N: Network> FromBytes for Transactions<N> {
         let version = u8::read_le(&mut reader)?;
         // Ensure the version is valid.
         if version != 0 {
-            return Err(error("Invalid header version"));
+            return Err(error("Invalid transactions version"));
         }
         // Read the number of transactions.
         let num_txs: u32 = FromBytes::read_le(&mut reader)?;
diff --git a/synthesizer/src/snark/proving_key/bytes.rs b/synthesizer/src/snark/proving_key/bytes.rs
index c2b22733c..bd0d6dfd7 100644
--- a/synthesizer/src/snark/proving_key/bytes.rs
+++ b/synthesizer/src/snark/proving_key/bytes.rs
@@ -23,7 +23,7 @@ impl<N: Network> FromBytes for ProvingKey<N> {
         let version = u8::read_le(&mut reader)?;
         // Ensure the version is valid.
         if version != 0 {
-            return Err(error("Invalid verifying key version"));
+            return Err(error("Invalid proving key version"));
         }
         // Read the proving key.
         let proving_key = Arc::new(FromBytes::read_le(&mut reader)?);
