diff --git a/Cargo.lock b/Cargo.lock
index 084610cae..4b7eee5c2 100644
--- a/Cargo.lock
+++ b/Cargo.lock
@@ -4130,7 +4130,7 @@ dependencies = [
 [[package]]
 name = "snarkvm"
 version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=0da94881fe6794659835d0616c070ac2a742148e#0da94881fe6794659835d0616c070ac2a742148e"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=8c0524f46209db1dd87d840376292f95406f1df7#8c0524f46209db1dd87d840376292f95406f1df7"
 dependencies = [
  "anstyle",
  "anyhow",
@@ -4155,7 +4155,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-algorithms"
 version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=0da94881fe6794659835d0616c070ac2a742148e#0da94881fe6794659835d0616c070ac2a742148e"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=8c0524f46209db1dd87d840376292f95406f1df7#8c0524f46209db1dd87d840376292f95406f1df7"
 dependencies = [
  "aleo-std",
  "anyhow",
@@ -4183,7 +4183,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-algorithms-cuda"
 version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=0da94881fe6794659835d0616c070ac2a742148e#0da94881fe6794659835d0616c070ac2a742148e"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=8c0524f46209db1dd87d840376292f95406f1df7#8c0524f46209db1dd87d840376292f95406f1df7"
 dependencies = [
  "blst",
  "cc",
@@ -4194,7 +4194,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-circuit"
 version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=0da94881fe6794659835d0616c070ac2a742148e#0da94881fe6794659835d0616c070ac2a742148e"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=8c0524f46209db1dd87d840376292f95406f1df7#8c0524f46209db1dd87d840376292f95406f1df7"
 dependencies = [
  "snarkvm-circuit-account",
  "snarkvm-circuit-algorithms",
@@ -4208,7 +4208,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-circuit-account"
 version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=0da94881fe6794659835d0616c070ac2a742148e#0da94881fe6794659835d0616c070ac2a742148e"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=8c0524f46209db1dd87d840376292f95406f1df7#8c0524f46209db1dd87d840376292f95406f1df7"
 dependencies = [
  "snarkvm-circuit-network",
  "snarkvm-circuit-types",
@@ -4218,7 +4218,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-circuit-algorithms"
 version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=0da94881fe6794659835d0616c070ac2a742148e#0da94881fe6794659835d0616c070ac2a742148e"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=8c0524f46209db1dd87d840376292f95406f1df7#8c0524f46209db1dd87d840376292f95406f1df7"
 dependencies = [
  "snarkvm-circuit-types",
  "snarkvm-console-algorithms",
@@ -4228,7 +4228,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-circuit-collections"
 version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=0da94881fe6794659835d0616c070ac2a742148e#0da94881fe6794659835d0616c070ac2a742148e"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=8c0524f46209db1dd87d840376292f95406f1df7#8c0524f46209db1dd87d840376292f95406f1df7"
 dependencies = [
  "snarkvm-circuit-algorithms",
  "snarkvm-circuit-types",
@@ -4238,7 +4238,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-circuit-environment"
 version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=0da94881fe6794659835d0616c070ac2a742148e#0da94881fe6794659835d0616c070ac2a742148e"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=8c0524f46209db1dd87d840376292f95406f1df7#8c0524f46209db1dd87d840376292f95406f1df7"
 dependencies = [
  "indexmap 2.11.0",
  "itertools 0.14.0",
@@ -4256,12 +4256,12 @@ dependencies = [
 [[package]]
 name = "snarkvm-circuit-environment-witness"
 version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=0da94881fe6794659835d0616c070ac2a742148e#0da94881fe6794659835d0616c070ac2a742148e"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=8c0524f46209db1dd87d840376292f95406f1df7#8c0524f46209db1dd87d840376292f95406f1df7"
 
 [[package]]
 name = "snarkvm-circuit-network"
 version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=0da94881fe6794659835d0616c070ac2a742148e#0da94881fe6794659835d0616c070ac2a742148e"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=8c0524f46209db1dd87d840376292f95406f1df7#8c0524f46209db1dd87d840376292f95406f1df7"
 dependencies = [
  "snarkvm-circuit-algorithms",
  "snarkvm-circuit-collections",
@@ -4272,7 +4272,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-circuit-program"
 version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=0da94881fe6794659835d0616c070ac2a742148e#0da94881fe6794659835d0616c070ac2a742148e"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=8c0524f46209db1dd87d840376292f95406f1df7#8c0524f46209db1dd87d840376292f95406f1df7"
 dependencies = [
  "snarkvm-circuit-account",
  "snarkvm-circuit-algorithms",
@@ -4286,7 +4286,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-circuit-types"
 version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=0da94881fe6794659835d0616c070ac2a742148e#0da94881fe6794659835d0616c070ac2a742148e"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=8c0524f46209db1dd87d840376292f95406f1df7#8c0524f46209db1dd87d840376292f95406f1df7"
 dependencies = [
  "snarkvm-circuit-environment",
  "snarkvm-circuit-types-address",
@@ -4301,7 +4301,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-circuit-types-address"
 version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=0da94881fe6794659835d0616c070ac2a742148e#0da94881fe6794659835d0616c070ac2a742148e"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=8c0524f46209db1dd87d840376292f95406f1df7#8c0524f46209db1dd87d840376292f95406f1df7"
 dependencies = [
  "snarkvm-circuit-environment",
  "snarkvm-circuit-types-boolean",
@@ -4314,7 +4314,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-circuit-types-boolean"
 version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=0da94881fe6794659835d0616c070ac2a742148e#0da94881fe6794659835d0616c070ac2a742148e"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=8c0524f46209db1dd87d840376292f95406f1df7#8c0524f46209db1dd87d840376292f95406f1df7"
 dependencies = [
  "snarkvm-circuit-environment",
  "snarkvm-console-types-boolean",
@@ -4323,7 +4323,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-circuit-types-field"
 version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=0da94881fe6794659835d0616c070ac2a742148e#0da94881fe6794659835d0616c070ac2a742148e"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=8c0524f46209db1dd87d840376292f95406f1df7#8c0524f46209db1dd87d840376292f95406f1df7"
 dependencies = [
  "snarkvm-circuit-environment",
  "snarkvm-circuit-types-boolean",
@@ -4333,7 +4333,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-circuit-types-group"
 version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=0da94881fe6794659835d0616c070ac2a742148e#0da94881fe6794659835d0616c070ac2a742148e"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=8c0524f46209db1dd87d840376292f95406f1df7#8c0524f46209db1dd87d840376292f95406f1df7"
 dependencies = [
  "snarkvm-circuit-environment",
  "snarkvm-circuit-types-boolean",
@@ -4345,7 +4345,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-circuit-types-integers"
 version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=0da94881fe6794659835d0616c070ac2a742148e#0da94881fe6794659835d0616c070ac2a742148e"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=8c0524f46209db1dd87d840376292f95406f1df7#8c0524f46209db1dd87d840376292f95406f1df7"
 dependencies = [
  "snarkvm-circuit-environment",
  "snarkvm-circuit-types-boolean",
@@ -4357,7 +4357,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-circuit-types-scalar"
 version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=0da94881fe6794659835d0616c070ac2a742148e#0da94881fe6794659835d0616c070ac2a742148e"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=8c0524f46209db1dd87d840376292f95406f1df7#8c0524f46209db1dd87d840376292f95406f1df7"
 dependencies = [
  "snarkvm-circuit-environment",
  "snarkvm-circuit-types-boolean",
@@ -4368,7 +4368,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-circuit-types-string"
 version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=0da94881fe6794659835d0616c070ac2a742148e#0da94881fe6794659835d0616c070ac2a742148e"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=8c0524f46209db1dd87d840376292f95406f1df7#8c0524f46209db1dd87d840376292f95406f1df7"
 dependencies = [
  "snarkvm-circuit-environment",
  "snarkvm-circuit-types-boolean",
@@ -4380,7 +4380,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-console"
 version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=0da94881fe6794659835d0616c070ac2a742148e#0da94881fe6794659835d0616c070ac2a742148e"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=8c0524f46209db1dd87d840376292f95406f1df7#8c0524f46209db1dd87d840376292f95406f1df7"
 dependencies = [
  "snarkvm-console-account",
  "snarkvm-console-algorithms",
@@ -4393,7 +4393,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-console-account"
 version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=0da94881fe6794659835d0616c070ac2a742148e#0da94881fe6794659835d0616c070ac2a742148e"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=8c0524f46209db1dd87d840376292f95406f1df7#8c0524f46209db1dd87d840376292f95406f1df7"
 dependencies = [
  "bs58",
  "snarkvm-console-network",
@@ -4404,7 +4404,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-console-algorithms"
 version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=0da94881fe6794659835d0616c070ac2a742148e#0da94881fe6794659835d0616c070ac2a742148e"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=8c0524f46209db1dd87d840376292f95406f1df7#8c0524f46209db1dd87d840376292f95406f1df7"
 dependencies = [
  "blake2s_simd",
  "smallvec",
@@ -4417,7 +4417,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-console-collections"
 version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=0da94881fe6794659835d0616c070ac2a742148e#0da94881fe6794659835d0616c070ac2a742148e"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=8c0524f46209db1dd87d840376292f95406f1df7#8c0524f46209db1dd87d840376292f95406f1df7"
 dependencies = [
  "aleo-std",
  "rayon",
@@ -4428,7 +4428,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-console-network"
 version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=0da94881fe6794659835d0616c070ac2a742148e#0da94881fe6794659835d0616c070ac2a742148e"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=8c0524f46209db1dd87d840376292f95406f1df7#8c0524f46209db1dd87d840376292f95406f1df7"
 dependencies = [
  "anyhow",
  "enum-iterator",
@@ -4448,7 +4448,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-console-network-environment"
 version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=0da94881fe6794659835d0616c070ac2a742148e#0da94881fe6794659835d0616c070ac2a742148e"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=8c0524f46209db1dd87d840376292f95406f1df7#8c0524f46209db1dd87d840376292f95406f1df7"
 dependencies = [
  "anyhow",
  "bech32",
@@ -4466,7 +4466,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-console-program"
 version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=0da94881fe6794659835d0616c070ac2a742148e#0da94881fe6794659835d0616c070ac2a742148e"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=8c0524f46209db1dd87d840376292f95406f1df7#8c0524f46209db1dd87d840376292f95406f1df7"
 dependencies = [
  "enum-iterator",
  "enum_index",
@@ -4486,7 +4486,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-console-types"
 version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=0da94881fe6794659835d0616c070ac2a742148e#0da94881fe6794659835d0616c070ac2a742148e"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=8c0524f46209db1dd87d840376292f95406f1df7#8c0524f46209db1dd87d840376292f95406f1df7"
 dependencies = [
  "snarkvm-console-network-environment",
  "snarkvm-console-types-address",
@@ -4501,7 +4501,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-console-types-address"
 version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=0da94881fe6794659835d0616c070ac2a742148e#0da94881fe6794659835d0616c070ac2a742148e"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=8c0524f46209db1dd87d840376292f95406f1df7#8c0524f46209db1dd87d840376292f95406f1df7"
 dependencies = [
  "snarkvm-console-network-environment",
  "snarkvm-console-types-boolean",
@@ -4512,7 +4512,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-console-types-boolean"
 version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=0da94881fe6794659835d0616c070ac2a742148e#0da94881fe6794659835d0616c070ac2a742148e"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=8c0524f46209db1dd87d840376292f95406f1df7#8c0524f46209db1dd87d840376292f95406f1df7"
 dependencies = [
  "snarkvm-console-network-environment",
 ]
@@ -4520,7 +4520,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-console-types-field"
 version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=0da94881fe6794659835d0616c070ac2a742148e#0da94881fe6794659835d0616c070ac2a742148e"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=8c0524f46209db1dd87d840376292f95406f1df7#8c0524f46209db1dd87d840376292f95406f1df7"
 dependencies = [
  "snarkvm-console-network-environment",
  "snarkvm-console-types-boolean",
@@ -4530,7 +4530,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-console-types-group"
 version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=0da94881fe6794659835d0616c070ac2a742148e#0da94881fe6794659835d0616c070ac2a742148e"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=8c0524f46209db1dd87d840376292f95406f1df7#8c0524f46209db1dd87d840376292f95406f1df7"
 dependencies = [
  "snarkvm-console-network-environment",
  "snarkvm-console-types-boolean",
@@ -4541,7 +4541,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-console-types-integers"
 version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=0da94881fe6794659835d0616c070ac2a742148e#0da94881fe6794659835d0616c070ac2a742148e"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=8c0524f46209db1dd87d840376292f95406f1df7#8c0524f46209db1dd87d840376292f95406f1df7"
 dependencies = [
  "snarkvm-console-network-environment",
  "snarkvm-console-types-boolean",
@@ -4552,7 +4552,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-console-types-scalar"
 version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=0da94881fe6794659835d0616c070ac2a742148e#0da94881fe6794659835d0616c070ac2a742148e"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=8c0524f46209db1dd87d840376292f95406f1df7#8c0524f46209db1dd87d840376292f95406f1df7"
 dependencies = [
  "snarkvm-console-network-environment",
  "snarkvm-console-types-boolean",
@@ -4563,7 +4563,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-console-types-string"
 version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=0da94881fe6794659835d0616c070ac2a742148e#0da94881fe6794659835d0616c070ac2a742148e"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=8c0524f46209db1dd87d840376292f95406f1df7#8c0524f46209db1dd87d840376292f95406f1df7"
 dependencies = [
  "snarkvm-console-network-environment",
  "snarkvm-console-types-boolean",
@@ -4574,7 +4574,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-curves"
 version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=0da94881fe6794659835d0616c070ac2a742148e#0da94881fe6794659835d0616c070ac2a742148e"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=8c0524f46209db1dd87d840376292f95406f1df7#8c0524f46209db1dd87d840376292f95406f1df7"
 dependencies = [
  "rand 0.8.5",
  "rayon",
@@ -4588,7 +4588,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-fields"
 version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=0da94881fe6794659835d0616c070ac2a742148e#0da94881fe6794659835d0616c070ac2a742148e"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=8c0524f46209db1dd87d840376292f95406f1df7#8c0524f46209db1dd87d840376292f95406f1df7"
 dependencies = [
  "aleo-std",
  "anyhow",
@@ -4605,7 +4605,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-ledger"
 version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=0da94881fe6794659835d0616c070ac2a742148e#0da94881fe6794659835d0616c070ac2a742148e"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=8c0524f46209db1dd87d840376292f95406f1df7#8c0524f46209db1dd87d840376292f95406f1df7"
 dependencies = [
  "aleo-std",
  "anyhow",
@@ -4632,7 +4632,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-ledger-authority"
 version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=0da94881fe6794659835d0616c070ac2a742148e#0da94881fe6794659835d0616c070ac2a742148e"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=8c0524f46209db1dd87d840376292f95406f1df7#8c0524f46209db1dd87d840376292f95406f1df7"
 dependencies = [
  "anyhow",
  "rand 0.8.5",
@@ -4644,7 +4644,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-ledger-block"
 version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=0da94881fe6794659835d0616c070ac2a742148e#0da94881fe6794659835d0616c070ac2a742148e"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=8c0524f46209db1dd87d840376292f95406f1df7#8c0524f46209db1dd87d840376292f95406f1df7"
 dependencies = [
  "indexmap 2.11.0",
  "rayon",
@@ -4664,7 +4664,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-ledger-committee"
 version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=0da94881fe6794659835d0616c070ac2a742148e#0da94881fe6794659835d0616c070ac2a742148e"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=8c0524f46209db1dd87d840376292f95406f1df7#8c0524f46209db1dd87d840376292f95406f1df7"
 dependencies = [
  "anyhow",
  "indexmap 2.11.0",
@@ -4683,7 +4683,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-ledger-narwhal"
 version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=0da94881fe6794659835d0616c070ac2a742148e#0da94881fe6794659835d0616c070ac2a742148e"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=8c0524f46209db1dd87d840376292f95406f1df7#8c0524f46209db1dd87d840376292f95406f1df7"
 dependencies = [
  "snarkvm-ledger-narwhal-batch-certificate",
  "snarkvm-ledger-narwhal-batch-header",
@@ -4696,7 +4696,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-ledger-narwhal-batch-certificate"
 version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=0da94881fe6794659835d0616c070ac2a742148e#0da94881fe6794659835d0616c070ac2a742148e"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=8c0524f46209db1dd87d840376292f95406f1df7#8c0524f46209db1dd87d840376292f95406f1df7"
 dependencies = [
  "indexmap 2.11.0",
  "rayon",
@@ -4709,7 +4709,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-ledger-narwhal-batch-header"
 version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=0da94881fe6794659835d0616c070ac2a742148e#0da94881fe6794659835d0616c070ac2a742148e"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=8c0524f46209db1dd87d840376292f95406f1df7#8c0524f46209db1dd87d840376292f95406f1df7"
 dependencies = [
  "indexmap 2.11.0",
  "rayon",
@@ -4722,7 +4722,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-ledger-narwhal-data"
 version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=0da94881fe6794659835d0616c070ac2a742148e#0da94881fe6794659835d0616c070ac2a742148e"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=8c0524f46209db1dd87d840376292f95406f1df7#8c0524f46209db1dd87d840376292f95406f1df7"
 dependencies = [
  "bytes",
  "serde_json",
@@ -4733,7 +4733,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-ledger-narwhal-subdag"
 version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=0da94881fe6794659835d0616c070ac2a742148e#0da94881fe6794659835d0616c070ac2a742148e"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=8c0524f46209db1dd87d840376292f95406f1df7#8c0524f46209db1dd87d840376292f95406f1df7"
 dependencies = [
  "indexmap 2.11.0",
  "rayon",
@@ -4748,7 +4748,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-ledger-narwhal-transmission"
 version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=0da94881fe6794659835d0616c070ac2a742148e#0da94881fe6794659835d0616c070ac2a742148e"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=8c0524f46209db1dd87d840376292f95406f1df7#8c0524f46209db1dd87d840376292f95406f1df7"
 dependencies = [
  "bytes",
  "serde_json",
@@ -4761,7 +4761,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-ledger-narwhal-transmission-id"
 version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=0da94881fe6794659835d0616c070ac2a742148e#0da94881fe6794659835d0616c070ac2a742148e"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=8c0524f46209db1dd87d840376292f95406f1df7#8c0524f46209db1dd87d840376292f95406f1df7"
 dependencies = [
  "snarkvm-console",
  "snarkvm-ledger-puzzle",
@@ -4770,7 +4770,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-ledger-puzzle"
 version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=0da94881fe6794659835d0616c070ac2a742148e#0da94881fe6794659835d0616c070ac2a742148e"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=8c0524f46209db1dd87d840376292f95406f1df7#8c0524f46209db1dd87d840376292f95406f1df7"
 dependencies = [
  "aleo-std",
  "anyhow",
@@ -4790,7 +4790,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-ledger-puzzle-epoch"
 version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=0da94881fe6794659835d0616c070ac2a742148e#0da94881fe6794659835d0616c070ac2a742148e"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=8c0524f46209db1dd87d840376292f95406f1df7#8c0524f46209db1dd87d840376292f95406f1df7"
 dependencies = [
  "aleo-std",
  "anyhow",
@@ -4813,7 +4813,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-ledger-query"
 version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=0da94881fe6794659835d0616c070ac2a742148e#0da94881fe6794659835d0616c070ac2a742148e"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=8c0524f46209db1dd87d840376292f95406f1df7#8c0524f46209db1dd87d840376292f95406f1df7"
 dependencies = [
  "anyhow",
  "async-trait",
@@ -4830,7 +4830,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-ledger-store"
 version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=0da94881fe6794659835d0616c070ac2a742148e#0da94881fe6794659835d0616c070ac2a742148e"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=8c0524f46209db1dd87d840376292f95406f1df7#8c0524f46209db1dd87d840376292f95406f1df7"
 dependencies = [
  "aleo-std-storage",
  "anyhow",
@@ -4857,7 +4857,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-ledger-test-helpers"
 version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=0da94881fe6794659835d0616c070ac2a742148e#0da94881fe6794659835d0616c070ac2a742148e"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=8c0524f46209db1dd87d840376292f95406f1df7#8c0524f46209db1dd87d840376292f95406f1df7"
 dependencies = [
  "aleo-std",
  "snarkvm-algorithms",
@@ -4873,7 +4873,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-metrics"
 version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=0da94881fe6794659835d0616c070ac2a742148e#0da94881fe6794659835d0616c070ac2a742148e"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=8c0524f46209db1dd87d840376292f95406f1df7#8c0524f46209db1dd87d840376292f95406f1df7"
 dependencies = [
  "metrics",
 ]
@@ -4881,7 +4881,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-parameters"
 version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=0da94881fe6794659835d0616c070ac2a742148e#0da94881fe6794659835d0616c070ac2a742148e"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=8c0524f46209db1dd87d840376292f95406f1df7#8c0524f46209db1dd87d840376292f95406f1df7"
 dependencies = [
  "aleo-std",
  "anyhow",
@@ -4904,7 +4904,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-synthesizer"
 version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=0da94881fe6794659835d0616c070ac2a742148e#0da94881fe6794659835d0616c070ac2a742148e"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=8c0524f46209db1dd87d840376292f95406f1df7#8c0524f46209db1dd87d840376292f95406f1df7"
 dependencies = [
  "aleo-std",
  "anyhow",
@@ -4937,7 +4937,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-synthesizer-process"
 version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=0da94881fe6794659835d0616c070ac2a742148e#0da94881fe6794659835d0616c070ac2a742148e"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=8c0524f46209db1dd87d840376292f95406f1df7#8c0524f46209db1dd87d840376292f95406f1df7"
 dependencies = [
  "aleo-std",
  "colored 3.0.0",
@@ -4962,7 +4962,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-synthesizer-program"
 version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=0da94881fe6794659835d0616c070ac2a742148e#0da94881fe6794659835d0616c070ac2a742148e"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=8c0524f46209db1dd87d840376292f95406f1df7#8c0524f46209db1dd87d840376292f95406f1df7"
 dependencies = [
  "indexmap 2.11.0",
  "paste",
@@ -4980,7 +4980,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-synthesizer-snark"
 version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=0da94881fe6794659835d0616c070ac2a742148e#0da94881fe6794659835d0616c070ac2a742148e"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=8c0524f46209db1dd87d840376292f95406f1df7#8c0524f46209db1dd87d840376292f95406f1df7"
 dependencies = [
  "bincode",
  "serde_json",
@@ -4993,7 +4993,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-utilities"
 version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=0da94881fe6794659835d0616c070ac2a742148e#0da94881fe6794659835d0616c070ac2a742148e"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=8c0524f46209db1dd87d840376292f95406f1df7#8c0524f46209db1dd87d840376292f95406f1df7"
 dependencies = [
  "aleo-std",
  "anyhow",
@@ -5014,7 +5014,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-utilities-derives"
 version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=0da94881fe6794659835d0616c070ac2a742148e#0da94881fe6794659835d0616c070ac2a742148e"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=8c0524f46209db1dd87d840376292f95406f1df7#8c0524f46209db1dd87d840376292f95406f1df7"
 dependencies = [
  "proc-macro2",
  "quote 1.0.40",
diff --git a/Cargo.toml b/Cargo.toml
index c6d2450e0..03a11a920 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -44,12 +44,11 @@ version = "1.0.1"
 default-features = false
 
 [workspace.dependencies.snarkvm]
-#path = "../snarkVM"
+# path = "../snarkVM"
 git = "https://github.com/ProvableHQ/snarkVM.git"
-rev = "0da94881fe6794659835d0616c070ac2a742148e"
+rev = "8c0524f46209db1dd87d840376292f95406f1df7"
 #version = "=4.0.1"
 default-features = false
-#features = [ "circuit", "console", "rocks" ]
 
 [workspace.dependencies.anyhow]
 version = "1.0"
diff --git a/cli/src/commands/developer/deploy.rs b/cli/src/commands/developer/deploy.rs
index ed75472a8..fcb2009fc 100644
--- a/cli/src/commands/developer/deploy.rs
+++ b/cli/src/commands/developer/deploy.rs
@@ -147,8 +147,12 @@ impl Deploy {
 
         println!("📦 Creating deployment transaction for '{}'...\n", &program_id.to_string().bold());
 
+        // Generate the process with the appropriate imports.
+        let process = package.get_process()?;
+
         // Generate the deployment
-        let mut deployment = package.deploy::<A>(None).with_context(|| "Failed to generate the deployment")?;
+        let mut deployment =
+            package.deploy::<A>(&process, None).with_context(|| "Failed to generate the deployment")?;
 
         // Get the consensus version.
         let consensus_version =
@@ -167,6 +171,10 @@ impl Deploy {
         // Compute the deployment ID.
         let deployment_id = deployment.to_deployment_id().with_context(|| "Failed to compute deployment ID")?;
 
+        // Compute the minimum deployment cost.
+        let (minimum_deployment_cost, (_, _, _, _)) = deployment_cost(&process, &deployment, consensus_version)
+            .with_context(|| "Failed to compute the minimum deployment cost")?;
+
         // Generate the deployment transaction.
         let transaction = {
             // Initialize an RNG.
@@ -177,15 +185,12 @@ impl Deploy {
                 Some(path) => StorageMode::Custom(path.clone()),
                 None => StorageMode::Production,
             };
-            let store = ConsensusStore::<N, ConsensusMemory<N>>::open(storage_mode)?;
+            let store = ConsensusStore::<N, ConsensusMemory<N>>::open(storage_mode)
+                .with_context(|| "Failed to open the consensus store")?;
 
             // Initialize the VM.
             let vm = VM::from(store).with_context(|| "Failed to initialize the virtual machine")?;
 
-            // Compute the minimum deployment cost.
-            let (minimum_deployment_cost, (_, _, _, _)) =
-                deployment_cost(&vm.process().read(), &deployment, consensus_version)?;
-
             // Prepare the fees.
             let fee = match &self.record {
                 Some(record) => {
diff --git a/cli/src/helpers/logger.rs b/cli/src/helpers/logger.rs
index 4c4c652fb..006cadc42 100644
--- a/cli/src/helpers/logger.rs
+++ b/cli/src/helpers/logger.rs
@@ -107,7 +107,9 @@ fn parse_log_filter(filter_str: &str) -> Result<EnvFilter> {
 }
 
 /// Sets the log filter based on the given verbosity level.
+/// Initializes the logger with the specified verbosity level, where 0 is the lowest verbosity and 6 the highest.
 ///
+/// The following shows what messages are enabled at each level.
 /// ```ignore
 /// 0 => info
 /// 1 => info, debug
diff --git a/node/bft/ledger-service/src/ledger.rs b/node/bft/ledger-service/src/ledger.rs
index bdf506e32..b01e13998 100644
--- a/node/bft/ledger-service/src/ledger.rs
+++ b/node/bft/ledger-service/src/ledger.rs
@@ -16,8 +16,10 @@
 use crate::{LedgerService, fmt_id, spawn_blocking};
 use snarkvm::{
     ledger::{
+        Block,
         Ledger,
-        block::{Block, Transaction},
+        PendingBlock,
+        Transaction,
         committee::Committee,
         narwhal::{BatchCertificate, Data, Subdag, Transmission, TransmissionID},
         puzzle::{Solution, SolutionID},
@@ -352,6 +354,14 @@ impl<N: Network, C: ConsensusStorage<N>> LedgerService<N> for CoreLedgerService<
         spawn_blocking!(ledger.check_transaction_basic(&transaction, None, &mut rand::thread_rng()))
     }
 
+    fn check_block_subdag(&self, block: Block<N>, prefix: &[PendingBlock<N>]) -> Result<PendingBlock<N>> {
+        self.ledger.check_block_subdag(block, prefix)
+    }
+
+    fn check_block_content(&self, block: PendingBlock<N>) -> Result<Block<N>> {
+        self.ledger.check_block_content(block, &mut rand::thread_rng())
+    }
+
     /// Checks the given block is valid next block.
     fn check_next_block(&self, block: &Block<N>) -> Result<()> {
         self.ledger.check_next_block(block, &mut rand::thread_rng())
diff --git a/node/bft/ledger-service/src/mock.rs b/node/bft/ledger-service/src/mock.rs
index 94154d6c3..ec0288dab 100644
--- a/node/bft/ledger-service/src/mock.rs
+++ b/node/bft/ledger-service/src/mock.rs
@@ -16,7 +16,9 @@
 use crate::{LedgerService, fmt_id};
 use snarkvm::{
     ledger::{
-        block::{Block, Transaction},
+        Block,
+        PendingBlock,
+        Transaction,
         committee::Committee,
         narwhal::{BatchCertificate, Data, Subdag, Transmission, TransmissionID},
         puzzle::{Solution, SolutionID},
@@ -211,6 +213,14 @@ impl<N: Network> LedgerService<N> for MockLedgerService<N> {
         Ok(())
     }
 
+    fn check_block_subdag(&self, _block: Block<N>, _prefix: &[PendingBlock<N>]) -> Result<PendingBlock<N>> {
+        unimplemented!();
+    }
+
+    fn check_block_content(&self, _block: PendingBlock<N>) -> Result<Block<N>> {
+        unimplemented!();
+    }
+
     /// Checks the given block is valid next block.
     fn check_next_block(&self, _block: &Block<N>) -> Result<()> {
         Ok(())
diff --git a/node/bft/ledger-service/src/prover.rs b/node/bft/ledger-service/src/prover.rs
index e5f2523e6..17bee1499 100644
--- a/node/bft/ledger-service/src/prover.rs
+++ b/node/bft/ledger-service/src/prover.rs
@@ -16,7 +16,9 @@
 use crate::LedgerService;
 use snarkvm::{
     ledger::{
-        block::{Block, Transaction},
+        Block,
+        PendingBlock,
+        Transaction,
         committee::Committee,
         narwhal::{BatchCertificate, Data, Subdag, Transmission, TransmissionID},
         puzzle::{Solution, SolutionID},
@@ -166,6 +168,14 @@ impl<N: Network> LedgerService<N> for ProverLedgerService<N> {
         Ok(())
     }
 
+    fn check_block_subdag(&self, _block: Block<N>, _prefix: &[PendingBlock<N>]) -> Result<PendingBlock<N>> {
+        bail!("Cannot check block subDAG in prover")
+    }
+
+    fn check_block_content(&self, _pending_block: PendingBlock<N>) -> Result<Block<N>> {
+        bail!("Cannot check block content in prover")
+    }
+
     /// Checks the given block is valid next block.
     fn check_next_block(&self, _block: &Block<N>) -> Result<()> {
         Ok(())
diff --git a/node/bft/ledger-service/src/traits.rs b/node/bft/ledger-service/src/traits.rs
index 02b1f2d7e..2471194bc 100644
--- a/node/bft/ledger-service/src/traits.rs
+++ b/node/bft/ledger-service/src/traits.rs
@@ -15,7 +15,9 @@
 
 use snarkvm::{
     ledger::{
-        block::{Block, Transaction},
+        Block,
+        PendingBlock,
+        Transaction,
         committee::Committee,
         narwhal::{BatchCertificate, Data, Subdag, Transmission, TransmissionID},
         puzzle::{Solution, SolutionID},
@@ -106,6 +108,12 @@ pub trait LedgerService<N: Network>: Debug + Send + Sync {
         transaction: Transaction<N>,
     ) -> Result<()>;
 
+    /// Checks that the subDAG in a given block is valid, but does not fully verify the block.
+    fn check_block_subdag(&self, block: Block<N>, prefix: &[PendingBlock<N>]) -> Result<PendingBlock<N>>;
+
+    /// Takes a pending block and performs the remaining checks to full verify it.
+    fn check_block_content(&self, _block: PendingBlock<N>) -> Result<Block<N>>;
+
     /// Checks the given block is valid next block.
     fn check_next_block(&self, block: &Block<N>) -> Result<()>;
 
diff --git a/node/bft/ledger-service/src/translucent.rs b/node/bft/ledger-service/src/translucent.rs
index c6f48a41e..33eb4afda 100644
--- a/node/bft/ledger-service/src/translucent.rs
+++ b/node/bft/ledger-service/src/translucent.rs
@@ -18,8 +18,10 @@ use async_trait::async_trait;
 use indexmap::IndexMap;
 use snarkvm::{
     ledger::{
+        Block,
         Ledger,
-        block::{Block, Transaction},
+        PendingBlock,
+        Transaction,
         committee::Committee,
         narwhal::{Data, Subdag, Transmission, TransmissionID},
         puzzle::{Solution, SolutionID},
@@ -177,6 +179,14 @@ impl<N: Network, C: ConsensusStorage<N>> LedgerService<N> for TranslucentLedgerS
         Ok(())
     }
 
+    fn check_block_subdag(&self, _block: Block<N>, _prefix: &[PendingBlock<N>]) -> Result<PendingBlock<N>> {
+        unimplemented!();
+    }
+
+    fn check_block_content(&self, _block: PendingBlock<N>) -> Result<Block<N>> {
+        unimplemented!();
+    }
+
     /// Always succeeds.
     fn check_next_block(&self, _block: &Block<N>) -> Result<()> {
         Ok(())
diff --git a/node/bft/src/sync/mod.rs b/node/bft/src/sync/mod.rs
index ec3177fb7..be6aac984 100644
--- a/node/bft/src/sync/mod.rs
+++ b/node/bft/src/sync/mod.rs
@@ -27,11 +27,11 @@ use snarkos_node_bft_ledger_service::LedgerService;
 use snarkos_node_sync::{BLOCK_REQUEST_BATCH_DELAY, BlockSync, Ping, PrepareSyncRequest, locators::BlockLocators};
 use snarkvm::{
     console::{network::Network, types::Field},
-    ledger::{authority::Authority, block::Block, narwhal::BatchCertificate},
+    ledger::{PendingBlock, authority::Authority, block::Block, narwhal::BatchCertificate},
     prelude::{cfg_into_iter, cfg_iter},
 };
 
-use anyhow::{Result, anyhow, bail};
+use anyhow::{Result, anyhow, bail, ensure};
 use indexmap::IndexMap;
 #[cfg(feature = "locktick")]
 use locktick::{parking_lot::Mutex, tokio::Mutex as TMutex};
@@ -40,7 +40,7 @@ use parking_lot::Mutex;
 #[cfg(not(feature = "serial"))]
 use rayon::prelude::*;
 use std::{
-    collections::{BTreeMap, HashMap},
+    collections::{HashMap, VecDeque},
     future::Future,
     net::SocketAddr,
     sync::Arc,
@@ -87,12 +87,12 @@ pub struct Sync<N: Network> {
     sync_lock: Arc<TMutex<()>>,
     /// The latest block responses.
     ///
-    /// This is used in [`Sync::sync_storage_with_block()`] to accumulate blocks  whose addition to the ledger is
-    /// deferred until certain checks pass.
+    /// This is used in [`Sync::sync_storage_with_block()`] to accumulate blocks
+    /// whose addition to the ledger is deferred until certain checks pass.
     /// Blocks need to be processed in order, hence a BTree map.
     ///
     /// Whenever a new block is added to this map, BlockSync::set_sync_height needs to be called.
-    latest_block_responses: Arc<TMutex<BTreeMap<u32, Block<N>>>>,
+    pending_blocks: Arc<TMutex<VecDeque<PendingBlock<N>>>>,
 }
 
 impl<N: Network> Sync<N> {
@@ -114,7 +114,7 @@ impl<N: Network> Sync<N> {
             handles: Default::default(),
             response_lock: Default::default(),
             sync_lock: Default::default(),
-            latest_block_responses: Default::default(),
+            pending_blocks: Default::default(),
         }
     }
 
@@ -446,13 +446,17 @@ impl<N: Network> Sync<N> {
     /// If there are queued block responses, this might be higher than the latest block in the ledger.
     async fn compute_sync_height(&self) -> u32 {
         let ledger_height = self.ledger.latest_block_height();
-        let mut responses = self.latest_block_responses.lock().await;
+        let mut pending_blocks = self.pending_blocks.lock().await;
 
         // Remove any old responses.
-        responses.retain(|height, _| *height > ledger_height);
+        while let Some(b) = pending_blocks.front()
+            && b.height() <= ledger_height
+        {
+            pending_blocks.pop_front();
+        }
 
         // Ensure the returned value is always greater or equal than ledger height.
-        responses.last_key_value().map(|(height, _)| *height).unwrap_or(0).max(ledger_height)
+        pending_blocks.back().map(|b| b.height()).unwrap_or(0).max(ledger_height)
     }
 
     /// Aims to advance synchronization using any recent block responses received from peers.
@@ -463,7 +467,7 @@ impl<N: Network> Sync<N> {
     /// This returns Ok(true) if we successfully advanced the ledger by at least one new block.
     ///
     /// A key difference to `BlockSync`'s versions is that it will only add blocks to the ledger once they have been confirmed by the network.
-    /// If blocks are not confirmed yet, they will be kept in [`Self::latest_block_responses`].
+    /// If blocks are not confirmed yet, they will be kept in [`Self::pending_blocks`].
     /// It will also pass certificates from synced blocks to the BFT module so that consensus can progress as expected
     /// (see [`Self::sync_storage_with_block`] for more details).
     ///
@@ -616,216 +620,191 @@ impl<N: Network> Sync<N> {
         .await?
     }
 
+    /// Helper function for [`Self::sync_storage_with_block`].
+    /// It syncs the batch certificates with the BFT, if the block's authority is a sub-DAG.
+    ///
+    /// Note that the block authority is always a sub-DAG in production; beacon signatures are only used for testing,
+    /// and as placeholder (irrelevant) block authority in the genesis block.i
+    async fn add_block_subdag_to_bft(&self, block: &Block<N>) -> Result<()> {
+        // Nothing to do if this is a beacon block
+        let Authority::Quorum(subdag) = block.authority() else {
+            return Ok(());
+        };
+
+        // Reconstruct the unconfirmed transactions.
+        let unconfirmed_transactions = cfg_iter!(block.transactions())
+            .filter_map(|tx| tx.to_unconfirmed_transaction().map(|unconfirmed| (unconfirmed.id(), unconfirmed)).ok())
+            .collect::<HashMap<_, _>>();
+
+        // Iterate over the certificates.
+        for certificates in subdag.values().cloned() {
+            cfg_into_iter!(certificates.clone()).for_each(|certificate| {
+                // Sync the batch certificate with the block.
+                self.storage.sync_certificate_with_block(block, certificate.clone(), &unconfirmed_transactions);
+            });
+
+            // Sync the BFT DAG with the certificates.
+            for certificate in certificates {
+                // If a BFT sender was provided, send the certificate to the BFT.
+                // For validators, BFT spawns a receiver task in `BFT::start_handlers`.
+                if let Some(bft_sender) = self.bft_sender.get() {
+                    // Await the callback to continue.
+                    if let Err(err) = bft_sender.send_sync_bft(certificate).await {
+                        bail!("Failed to sync certificate - {err}");
+                    };
+                }
+            }
+        }
+        Ok(())
+    }
+
+    /// Helper function for [`Self::sync_storage_with_block`].
+    ///
+    /// It checks that successor of a given block contains enough votes to commit it.
+    /// This can only return `Ok(true)` if the certificates of the block's successor were added to the storage.
+    fn is_block_availability_threshold_reached(&self, block: &PendingBlock<N>) -> Result<bool> {
+        // Fetch the leader certificate and the relevant rounds.
+        let leader_certificate = match block.authority() {
+            Authority::Quorum(subdag) => subdag.leader_certificate().clone(),
+            _ => bail!("Received a block with an unexpected authority type."),
+        };
+        let commit_round = leader_certificate.round();
+        let certificate_round =
+            commit_round.checked_add(1).ok_or_else(|| anyhow!("Integer overflow on round number"))?;
+
+        // Get the committee lookback for the round just after the leader.
+        let certificate_committee_lookback = self.ledger.get_committee_lookback_for_round(certificate_round)?;
+        // Retrieve all of the certificates for the round just after the leader.
+        let certificates = self.storage.get_certificates_for_round(certificate_round);
+        // Construct a set over the authors, at the round just after the leader,
+        // who included the leader's certificate in their previous certificate IDs.
+        let authors = certificates
+            .iter()
+            .filter_map(|c| match c.previous_certificate_ids().contains(&leader_certificate.id()) {
+                true => Some(c.author()),
+                false => None,
+            })
+            .collect();
+
+        // Check if the leader is ready to be committed.
+        if certificate_committee_lookback.is_availability_threshold_reached(&authors) {
+            trace!(
+                "Block {hash} at height {height} has reached availability threshold",
+                hash = block.hash(),
+                height = block.height()
+            );
+            Ok(true)
+        } else {
+            Ok(false)
+        }
+    }
+
     /// Advances the ledger by the given block and updates the storage accordingly.
     ///
     /// This also updates the DAG, and uses the DAG to ensure that the block's leader certificate
     /// meets the voter availability threshold (i.e. > f voting stake)
     /// or is reachable via a DAG path from a later leader certificate that does.
     /// Since performing this check requires DAG certificates from later blocks,
-    /// the block is stored in `Sync::latest_block_responses`,
+    /// the block is stored in `Sync::pending_blocks`,
     /// and its addition to the ledger is deferred until the check passes.
-    /// Several blocks may be stored in `Sync::latest_block_responses`
+    /// Several blocks may be stored in `Sync::pending_blocks`
     /// before they can be all checked and added to the ledger.
-    async fn sync_storage_with_block(&self, block: Block<N>) -> Result<()> {
+    ///
+    /// # Usage
+    /// This function assumes that blocks are passed in order, i.e.,
+    /// that the given block is a direct successor of the block that was last passed to this function.
+    async fn sync_storage_with_block(&self, new_block: Block<N>) -> Result<()> {
         // Acquire the sync lock.
         let _lock = self.sync_lock.lock().await;
 
         // If this block has already been processed, return early.
         // TODO(kaimast): Should we remove the response here?
-        if self.ledger.contains_block_height(block.height()) {
-            debug!("Ledger is already synced with block at height {}. Will not sync.", block.height());
-            return Ok(());
-        }
-
-        // Acquire the latest block responses lock.
-        let mut latest_block_responses = self.latest_block_responses.lock().await;
-
-        if latest_block_responses.contains_key(&block.height()) {
-            debug!("An unconfirmed block is queued already for height {}. Will not sync.", block.height());
+        if self.ledger.contains_block_height(new_block.height()) {
+            debug!(
+                "Ledger is already synced with block at height {height}. Will not sync.",
+                height = new_block.height()
+            );
             return Ok(());
         }
 
-        // If the block authority is a sub-DAG, then sync the batch certificates with the block.
-        // Note that the block authority is always a sub-DAG in production;
-        // beacon signatures are only used for testing,
-        // and as placeholder (irrelevant) block authority in the genesis block.
-        if let Authority::Quorum(subdag) = block.authority() {
-            // Reconstruct the unconfirmed transactions.
-            let unconfirmed_transactions = cfg_iter!(block.transactions())
-                .filter_map(|tx| {
-                    tx.to_unconfirmed_transaction().map(|unconfirmed| (unconfirmed.id(), unconfirmed)).ok()
-                })
-                .collect::<HashMap<_, _>>();
-
-            // Iterate over the certificates.
-            for certificates in subdag.values().cloned() {
-                cfg_into_iter!(certificates.clone()).for_each(|certificate| {
-                    // Sync the batch certificate with the block.
-                    self.storage.sync_certificate_with_block(&block, certificate.clone(), &unconfirmed_transactions);
-                });
+        // Acquire the pending blocks lock.
+        let mut pending_blocks = self.pending_blocks.lock().await;
 
-                // Sync the BFT DAG with the certificates.
-                for certificate in certificates {
-                    // If a BFT sender was provided, send the certificate to the BFT.
-                    // For validators, BFT spawns a receiver task in `BFT::start_handlers`.
-                    if let Some(bft_sender) = self.bft_sender.get() {
-                        // Await the callback to continue.
-                        if let Err(err) = bft_sender.send_sync_bft(certificate).await {
-                            bail!("Failed to sync certificate - {err}");
-                        };
-                    }
-                }
+        if let Some(tail) = pending_blocks.back() {
+            if tail.height() >= new_block.height() {
+                debug!(
+                    "A unconfirmed block is queued already for height {height}. \
+                    Will not sync.",
+                    height = new_block.height()
+                );
+                return Ok(());
             }
+
+            ensure!(tail.height() + 1 == new_block.height(), "Got an out-of-order block");
         }
 
         // Fetch the latest block height.
         let ledger_block_height = self.ledger.latest_block_height();
 
-        // Insert the latest block response.
-        latest_block_responses.insert(block.height(), block);
-        // Clear the latest block responses of older blocks.
-        latest_block_responses.retain(|height, _| *height > ledger_block_height);
+        // Clear any older pending blocks.
+        // TODO(kaimast): ensure there are no dangling block requests
+        while let Some(pending_block) = pending_blocks.front() {
+            if pending_block.height() > ledger_block_height {
+                break;
+            }
+
+            pending_blocks.pop_front();
+        }
 
         // Get a list of contiguous blocks from the latest block responses.
-        let contiguous_blocks: Vec<Block<N>> = (ledger_block_height.saturating_add(1)..)
-            .take_while(|&k| latest_block_responses.contains_key(&k))
-            .filter_map(|k| latest_block_responses.get(&k).cloned())
-            .collect();
+        let new_block = self.ledger.check_block_subdag(new_block, pending_blocks.make_contiguous())?;
+
+        // Append the new block to the set of pending blocks and add its certificates to the storage.
+        self.add_block_subdag_to_bft(&new_block).await?;
+        pending_blocks.push_back(new_block);
+
+        // Now, figure out if and which pending block we can commit.
+        // To do this effectively and because commits are transitive,
+        // we iterate in reverse so that we can stop at the first successful check.
+        //
+        // Note, that if the storage already contains certificates for the round after new block,
+        // the availability threshold for the new block could also be reached.
+        let mut commit_height = None;
+        for block in pending_blocks.iter().rev() {
+            if self.is_block_availability_threshold_reached(block)? {
+                commit_height = Some(block.height());
+                break;
+            }
+        }
 
-        // Check if each block response, from the contiguous sequence just constructed,
-        // is ready to be added to the ledger.
-        // Ensure that the block's leader certificate meets the availability threshold
-        // based on the certificates in the DAG just after the block's round.
-        // If the availability threshold is not met,
-        // process the next block and check if it is linked to the current block,
-        // in the sense that there is a path in the DAG
-        // from the next block's leader certificate
-        // to the current block's leader certificate.
-        // Note: We do not advance to the most recent block response because we would be unable to
-        // validate if the leader certificate in the block has been certified properly.
-        for next_block in contiguous_blocks.into_iter() {
-            // Retrieve the height of the next block.
-            let next_block_height = next_block.height();
-
-            // Fetch the leader certificate and the relevant rounds.
-            let leader_certificate = match next_block.authority() {
-                Authority::Quorum(subdag) => subdag.leader_certificate().clone(),
-                _ => bail!("Received a block with an unexpected authority type."),
-            };
-            let commit_round = leader_certificate.round();
-            let certificate_round =
-                commit_round.checked_add(1).ok_or_else(|| anyhow!("Integer overflow on round number"))?;
-
-            // Get the committee lookback for the round just after the leader.
-            let certificate_committee_lookback = self.ledger.get_committee_lookback_for_round(certificate_round)?;
-            // Retrieve all of the certificates for the round just after the leader.
-            let certificates = self.storage.get_certificates_for_round(certificate_round);
-            // Construct a set over the authors, at the round just after the leader,
-            // who included the leader's certificate in their previous certificate IDs.
-            let authors = certificates
-                .iter()
-                .filter_map(|c| match c.previous_certificate_ids().contains(&leader_certificate.id()) {
-                    true => Some(c.author()),
-                    false => None,
-                })
-                .collect();
-
-            debug!("Validating sync block {next_block_height} at round {commit_round}...");
-            // Check if the leader is ready to be committed.
-            if certificate_committee_lookback.is_availability_threshold_reached(&authors) {
-                // Initialize the current certificate.
-                let mut current_certificate = leader_certificate;
-                // Check if there are any linked blocks that need to be added.
-                let mut blocks_to_add = vec![next_block];
-
-                // Check if there are other blocks to process based on `is_linked`.
-                for height in (self.ledger.latest_block_height().saturating_add(1)..next_block_height).rev() {
-                    // Retrieve the previous block.
-                    let Some(previous_block) = latest_block_responses.get(&height) else {
-                        bail!("Block {height} is missing from the latest block responses.");
-                    };
-                    // Retrieve the previous block's leader certificate.
-                    let previous_certificate = match previous_block.authority() {
-                        Authority::Quorum(subdag) => subdag.leader_certificate().clone(),
-                        _ => bail!("Received a block with an unexpected authority type."),
-                    };
-                    // Determine if there is a path between the previous certificate and the current certificate.
-                    if self.is_linked(previous_certificate.clone(), current_certificate.clone())? {
-                        debug!("Previous sync block {height} is linked to the current block {next_block_height}");
-                        // Add the previous leader certificate to the list of certificates to commit.
-                        blocks_to_add.insert(0, previous_block.clone());
-                        // Update the current certificate to the previous leader certificate.
-                        current_certificate = previous_certificate;
-                    }
-                }
+        if let Some(commit_height) = commit_height {
+            let start_height = ledger_block_height + 1;
+            ensure!(commit_height >= start_height, "Invalid commit height");
+            let num_blocks = (commit_height - start_height + 1) as usize;
 
-                // Add the blocks to the ledger.
-                for block in blocks_to_add {
-                    // Check that the blocks are sequential and can be added to the ledger.
-                    let block_height = block.height();
-                    if block_height != self.ledger.latest_block_height().saturating_add(1) {
-                        warn!("Skipping block {block_height} from the latest block responses - not sequential.");
-                        continue;
-                    }
-                    #[cfg(feature = "telemetry")]
-                    let block_authority = block.authority().clone();
-
-                    let self_ = self.clone();
-                    tokio::task::spawn_blocking(move || {
-                        // Check the next block.
-                        self_.ledger.check_next_block(&block)?;
-                        // Attempt to advance to the next block.
-                        self_.ledger.advance_to_next_block(&block)?;
-
-                        // Sync the height with the block.
-                        self_.storage.sync_height_with_block(block.height());
-                        // Sync the round with the block.
-                        self_.storage.sync_round_with_block(block.round());
-
-                        Ok::<(), anyhow::Error>(())
-                    })
-                    .await??;
-                    // Remove the block height from the latest block responses.
-                    latest_block_responses.remove(&block_height);
-
-                    // Update the validator telemetry.
-                    #[cfg(feature = "telemetry")]
-                    if let Authority::Quorum(subdag) = block_authority {
-                        self_.gateway.validator_telemetry().insert_subdag(&subdag);
-                    }
-                }
-            } else {
-                debug!(
-                    "Availability threshold was not reached for block {next_block_height} at round {commit_round}. Checking next block..."
+            // Create a more detailed log message if we are committing more than one block at a time.
+            if num_blocks > 1 {
+                trace!(
+                    "Attempting to commit {chain_length} pending block(s) starting at height {start_height}.",
+                    chain_length = pending_blocks.len(),
                 );
             }
 
-            // Don't remove the response from BlockSync yet as we might fall back to non-BFT sync.
+            for pending_block in pending_blocks.drain(0..num_blocks) {
+                let hash = pending_block.hash();
+                let height = pending_block.height();
+                match self.ledger.check_block_content(pending_block) {
+                    Ok(block) => {
+                        trace!("Adding pending block {hash} at height {height} to the ledger");
+                        self.ledger.advance_to_next_block(&block)?;
+                    }
+                    Err(err) => bail!("Failed to check contents of pending block {hash} at height {height}: {err}"),
+                }
+            }
         }
 
         Ok(())
     }
-
-    /// Returns `true` if there is a path from the previous certificate to the current certificate.
-    fn is_linked(
-        &self,
-        previous_certificate: BatchCertificate<N>,
-        current_certificate: BatchCertificate<N>,
-    ) -> Result<bool> {
-        // Initialize the list containing the traversal.
-        let mut traversal = vec![current_certificate.clone()];
-        // Iterate over the rounds from the current certificate to the previous certificate.
-        for round in (previous_certificate.round()..current_certificate.round()).rev() {
-            // Retrieve all of the certificates for this past round.
-            let certificates = self.storage.get_certificates_for_round(round);
-            // Filter the certificates to only include those that are in the traversal.
-            traversal = certificates
-                .into_iter()
-                .filter(|p| traversal.iter().any(|c| c.previous_certificate_ids().contains(&p.id())))
-                .collect();
-        }
-        Ok(traversal.contains(&previous_certificate))
-    }
 }
 
 // Methods to assist with the block sync module.
@@ -972,13 +951,24 @@ mod tests {
     type CurrentLedger = Ledger<CurrentNetwork, ConsensusMemory<CurrentNetwork>>;
     type CurrentConsensusStore = ConsensusStore<CurrentNetwork, ConsensusMemory<CurrentNetwork>>;
 
+    /// Tests that commits work as expected when some anchors are not committed immediately.
     #[tokio::test]
     #[tracing_test::traced_test]
-    async fn test_commit_via_is_linked() -> anyhow::Result<()> {
+    async fn test_commit_chain() -> anyhow::Result<()> {
         let rng = &mut TestRng::default();
         // Initialize the round parameters.
         let max_gc_rounds = BatchHeader::<CurrentNetwork>::MAX_GC_ROUNDS as u64;
-        let commit_round = 2;
+
+        // The first round of the first block.
+        let first_round = 1;
+        // The total number of blocks we test
+        let num_blocks = 3;
+        // The number of certificate rounds needed.
+        // There is one additional round to provide availability for the inal block.
+        let num_rounds = first_round + num_blocks * 2 + 1;
+        // The first round that has at least N-f certificates referencing the anchor from the previous round.
+        // This is also the last round we use in the test.
+        let first_committed_round = num_rounds - 1;
 
         // Initialize the store.
         let store = CurrentConsensusStore::open(StorageMode::new_test(None)).unwrap();
@@ -1019,77 +1009,70 @@ mod tests {
                 HashMap::new();
             let mut previous_certificates: IndexSet<BatchCertificate<CurrentNetwork>> = IndexSet::with_capacity(4);
 
-            for round in 0..=commit_round + 8 {
+            for round in first_round..=first_committed_round {
                 let mut current_certificates = IndexSet::new();
                 let previous_certificate_ids: IndexSet<_> = if round == 0 || round == 1 {
                     IndexSet::new()
                 } else {
                     previous_certificates.iter().map(|c| c.id()).collect()
                 };
+
                 let committee_id = committee.id();
+                let prev_leader = committee.get_leader(round - 1).unwrap();
+
+                // For the first two blocks non-leaders will not reference the leader certificate.
+                // This means, while there is an anchor, it is isn't committed
+                // until later.
+                for (i, private_key) in private_keys.iter().enumerate() {
+                    let leader_index = addresses.iter().position(|&address| address == prev_leader).unwrap();
+                    let is_certificate_round = round % 2 == 1;
+                    let is_leader = i == leader_index;
+
+                    let previous_certs = if round < first_committed_round && is_certificate_round && !is_leader {
+                        previous_certificate_ids
+                            .iter()
+                            .cloned()
+                            .enumerate()
+                            .filter(|(idx, _)| *idx != leader_index)
+                            .map(|(_, id)| id)
+                            .collect()
+                    } else {
+                        previous_certificate_ids.clone()
+                    };
 
-                // Create a certificate for the leader.
-                if round <= 5 {
-                    let leader = committee.get_leader(round).unwrap();
-                    let leader_index = addresses.iter().position(|&address| address == leader).unwrap();
-                    let non_leader_index = addresses.iter().position(|&address| address != leader).unwrap();
-                    for i in [leader_index, non_leader_index].into_iter() {
-                        let batch_header = BatchHeader::new(
-                            &private_keys[i],
-                            round,
-                            now(),
-                            committee_id,
-                            Default::default(),
-                            previous_certificate_ids.clone(),
-                            rng,
-                        )
-                        .unwrap();
-                        // Sign the batch header.
-                        let mut signatures = IndexSet::with_capacity(4);
-                        for (j, private_key_2) in private_keys.iter().enumerate() {
-                            if i != j {
-                                signatures.insert(private_key_2.sign(&[batch_header.batch_id()], rng).unwrap());
-                            }
-                        }
-                        current_certificates.insert(BatchCertificate::from(batch_header, signatures).unwrap());
-                    }
-                }
+                    let batch_header = BatchHeader::new(
+                        private_key,
+                        round,
+                        now(),
+                        committee_id,
+                        Default::default(),
+                        previous_certs,
+                        rng,
+                    )
+                    .unwrap();
 
-                // Create a certificate for each validator.
-                if round > 5 {
-                    for (i, private_key_1) in private_keys.iter().enumerate() {
-                        let batch_header = BatchHeader::new(
-                            private_key_1,
-                            round,
-                            now(),
-                            committee_id,
-                            Default::default(),
-                            previous_certificate_ids.clone(),
-                            rng,
-                        )
-                        .unwrap();
-                        // Sign the batch header.
-                        let mut signatures = IndexSet::with_capacity(4);
-                        for (j, private_key_2) in private_keys.iter().enumerate() {
-                            if i != j {
-                                signatures.insert(private_key_2.sign(&[batch_header.batch_id()], rng).unwrap());
-                            }
+                    // Sign the batch header.
+                    let mut signatures = IndexSet::with_capacity(4);
+                    for (j, private_key_2) in private_keys.iter().enumerate() {
+                        if i != j {
+                            signatures.insert(private_key_2.sign(&[batch_header.batch_id()], rng).unwrap());
                         }
-                        current_certificates.insert(BatchCertificate::from(batch_header, signatures).unwrap());
                     }
+                    current_certificates.insert(BatchCertificate::from(batch_header, signatures).unwrap());
                 }
+
                 // Update the map of certificates.
                 round_to_certificates_map.insert(round, current_certificates.clone());
-                previous_certificates = current_certificates.clone();
+                previous_certificates = current_certificates;
             }
             (round_to_certificates_map, committee)
         };
 
         // Initialize the storage.
         let storage = Storage::new(core_ledger.clone(), Arc::new(BFTMemoryService::new()), max_gc_rounds);
-        // Insert certificates into storage.
+        // Insert all certificates into storage.
         let mut certificates: Vec<BatchCertificate<CurrentNetwork>> = Vec::new();
-        for i in 1..=commit_round + 8 {
+        for i in first_round..=first_committed_round {
             let c = (*round_to_certificates_map.get(&i).unwrap()).clone();
             certificates.extend(c);
         }
@@ -1097,100 +1080,79 @@ mod tests {
             storage.testing_only_insert_certificate_testing_only(certificate.clone());
         }
 
-        // Create block 1.
-        let leader_round_1 = commit_round;
-        let leader_1 = committee.get_leader(leader_round_1).unwrap();
-        let leader_certificate = storage.get_certificate_for_round_with_author(commit_round, leader_1).unwrap();
-        let block_1 = {
+        // Create the blocks
+        let mut previous_leader_cert = None;
+        let mut blocks = vec![];
+
+        for block_height in 1..=num_blocks {
+            let leader_round = block_height * 2;
+
+            let leader = committee.get_leader(leader_round).unwrap();
+            let leader_certificate = storage.get_certificate_for_round_with_author(leader_round, leader).unwrap();
+
             let mut subdag_map: BTreeMap<u64, IndexSet<BatchCertificate<CurrentNetwork>>> = BTreeMap::new();
             let mut leader_cert_map = IndexSet::new();
             leader_cert_map.insert(leader_certificate.clone());
-            let mut previous_cert_map = IndexSet::new();
-            for cert in storage.get_certificates_for_round(commit_round - 1) {
-                previous_cert_map.insert(cert);
+
+            let previous_cert_map = storage.get_certificates_for_round(leader_round - 1);
+
+            subdag_map.insert(leader_round, leader_cert_map.clone());
+            subdag_map.insert(leader_round - 1, previous_cert_map.clone());
+
+            if leader_round > 2 {
+                let previous_commit_cert_map: IndexSet<_> = storage
+                    .get_certificates_for_round(leader_round - 2)
+                    .into_iter()
+                    .filter(|cert| {
+                        if let Some(previous_leader_cert) = &previous_leader_cert {
+                            cert != previous_leader_cert
+                        } else {
+                            true
+                        }
+                    })
+                    .collect();
+                subdag_map.insert(leader_round - 2, previous_commit_cert_map);
             }
-            subdag_map.insert(commit_round, leader_cert_map.clone());
-            subdag_map.insert(commit_round - 1, previous_cert_map.clone());
+
             let subdag = Subdag::from(subdag_map.clone())?;
-            core_ledger.prepare_advance_to_next_quorum_block(subdag, Default::default())?
-        };
-        // Insert block 1.
-        core_ledger.advance_to_next_block(&block_1)?;
+            let block = core_ledger.prepare_advance_to_next_quorum_block(subdag, Default::default())?;
 
-        // Create block 2.
-        let leader_round_2 = commit_round + 2;
-        let leader_2 = committee.get_leader(leader_round_2).unwrap();
-        let leader_certificate_2 = storage.get_certificate_for_round_with_author(leader_round_2, leader_2).unwrap();
-        let block_2 = {
-            let mut subdag_map_2: BTreeMap<u64, IndexSet<BatchCertificate<CurrentNetwork>>> = BTreeMap::new();
-            let mut leader_cert_map_2 = IndexSet::new();
-            leader_cert_map_2.insert(leader_certificate_2.clone());
-            let mut previous_cert_map_2 = IndexSet::new();
-            for cert in storage.get_certificates_for_round(leader_round_2 - 1) {
-                previous_cert_map_2.insert(cert);
-            }
-            let mut prev_commit_cert_map_2 = IndexSet::new();
-            for cert in storage.get_certificates_for_round(leader_round_2 - 2) {
-                if cert != leader_certificate {
-                    prev_commit_cert_map_2.insert(cert);
-                }
-            }
-            subdag_map_2.insert(leader_round_2, leader_cert_map_2.clone());
-            subdag_map_2.insert(leader_round_2 - 1, previous_cert_map_2.clone());
-            subdag_map_2.insert(leader_round_2 - 2, prev_commit_cert_map_2.clone());
-            let subdag_2 = Subdag::from(subdag_map_2.clone())?;
-            core_ledger.prepare_advance_to_next_quorum_block(subdag_2, Default::default())?
-        };
-        // Insert block 2.
-        core_ledger.advance_to_next_block(&block_2)?;
+            previous_leader_cert = Some(leader_certificate);
 
-        // Create block 3
-        let leader_round_3 = commit_round + 4;
-        let leader_3 = committee.get_leader(leader_round_3).unwrap();
-        let leader_certificate_3 = storage.get_certificate_for_round_with_author(leader_round_3, leader_3).unwrap();
-        let block_3 = {
-            let mut subdag_map_3: BTreeMap<u64, IndexSet<BatchCertificate<CurrentNetwork>>> = BTreeMap::new();
-            let mut leader_cert_map_3 = IndexSet::new();
-            leader_cert_map_3.insert(leader_certificate_3.clone());
-            let mut previous_cert_map_3 = IndexSet::new();
-            for cert in storage.get_certificates_for_round(leader_round_3 - 1) {
-                previous_cert_map_3.insert(cert);
-            }
-            let mut prev_commit_cert_map_3 = IndexSet::new();
-            for cert in storage.get_certificates_for_round(leader_round_3 - 2) {
-                if cert != leader_certificate_2 {
-                    prev_commit_cert_map_3.insert(cert);
-                }
-            }
-            subdag_map_3.insert(leader_round_3, leader_cert_map_3.clone());
-            subdag_map_3.insert(leader_round_3 - 1, previous_cert_map_3.clone());
-            subdag_map_3.insert(leader_round_3 - 2, prev_commit_cert_map_3.clone());
-            let subdag_3 = Subdag::from(subdag_map_3.clone())?;
-            core_ledger.prepare_advance_to_next_quorum_block(subdag_3, Default::default())?
-        };
-        // Insert block 3.
-        core_ledger.advance_to_next_block(&block_3)?;
+            core_ledger.advance_to_next_block(&block)?;
+            blocks.push(block);
+        }
+
+        // ### Test that sync works as expected ###
 
-        // Initialize the syncing ledger.
+        // Create a new ledger to test with, but use the existing storage
+        // so that the certificates exist.
         let syncing_ledger = Arc::new(CoreLedgerService::new(
             CurrentLedger::load(genesis, StorageMode::new_test(None)).unwrap(),
             Default::default(),
         ));
-        // Initialize the gateway.
+
+        // Set up sync and its dependencies.
         let gateway = Gateway::new(account.clone(), storage.clone(), syncing_ledger.clone(), None, &[], None)?;
-        // Initialize the block synchronization logic.
         let block_sync = Arc::new(BlockSync::new(syncing_ledger.clone()));
-        // Initialize the sync module.
         let sync = Sync::new(gateway.clone(), storage.clone(), syncing_ledger.clone(), block_sync);
-        // Try to sync block 1.
-        sync.sync_storage_with_block(block_1).await?;
-        assert_eq!(syncing_ledger.latest_block_height(), 1);
-        // Try to sync block 2.
-        sync.sync_storage_with_block(block_2).await?;
-        assert_eq!(syncing_ledger.latest_block_height(), 2);
-        // Try to sync block 3.
-        sync.sync_storage_with_block(block_3).await?;
+
+        let mut block_iter = blocks.into_iter();
+
+        // Insert the blocks into the new sync module
+        for _ in 0..num_blocks - 1 {
+            let block = block_iter.next().unwrap();
+            sync.sync_storage_with_block(block).await?;
+
+            // Availability threshold is not met, so we should not advance yet.
+            assert_eq!(syncing_ledger.latest_block_height(), 0);
+        }
+
+        // Only for the final block, the availability threshold is met,
+        // because certificates for the subsequent round are already in storage.
+        sync.sync_storage_with_block(block_iter.next().unwrap()).await?;
         assert_eq!(syncing_ledger.latest_block_height(), 3);
+
         // Ensure blocks 1 and 2 were added to the ledger.
         assert!(syncing_ledger.contains_block_height(1));
         assert!(syncing_ledger.contains_block_height(2));
diff --git a/node/bft/src/worker.rs b/node/bft/src/worker.rs
index 04c62ab6c..e7756952c 100644
--- a/node/bft/src/worker.rs
+++ b/node/bft/src/worker.rs
@@ -26,7 +26,7 @@ use snarkos_node_bft_ledger_service::LedgerService;
 use snarkvm::{
     console::prelude::*,
     ledger::{
-        block::Transaction,
+        Transaction,
         narwhal::{BatchHeader, Data, Transmission, TransmissionID},
         puzzle::{Solution, SolutionID},
     },
@@ -560,7 +560,8 @@ mod tests {
     use snarkvm::{
         console::{network::Network, types::Field},
         ledger::{
-            block::Block,
+            Block,
+            PendingBlock,
             committee::Committee,
             narwhal::{BatchCertificate, Subdag, Transmission, TransmissionID},
             snarkvm_ledger_test_helpers::sample_execution_transaction_with_fee,
@@ -626,6 +627,8 @@ mod tests {
                 transaction_id: N::TransactionID,
                 transaction: Transaction<N>,
             ) -> Result<()>;
+            fn check_block_subdag(&self, _block: Block<N>, _prefix: &[PendingBlock<N>]) -> Result<PendingBlock<N>>;
+            fn check_block_content(&self, _block: PendingBlock<N>) -> Result<Block<N>>;
             fn check_next_block(&self, block: &Block<N>) -> Result<()>;
             fn prepare_advance_to_next_quorum_block(
                 &self,
diff --git a/node/router/src/lib.rs b/node/router/src/lib.rs
index 679613283..d73853d3c 100644
--- a/node/router/src/lib.rs
+++ b/node/router/src/lib.rs
@@ -118,7 +118,7 @@ impl<N: Network> Router<N> {
     const CONNECTION_ATTEMPTS_SINCE_SECS: i64 = 10;
     /// The maximum number of candidate peers permitted to be stored in the node.
     const MAXIMUM_CANDIDATE_PEERS: usize = 10_000;
-    /// The maximum amount of connection attempts within a 10 second threshold
+    /// The maximum amount of connection attempts within a 10 second threshold.
     #[cfg(not(test))]
     const MAX_CONNECTION_ATTEMPTS: usize = 10;
     /// The duration after which a connected peer is considered inactive or
