diff --git a/cli/src/commands/clean.rs b/cli/src/commands/clean.rs
index a8b342079..b6519a7df 100644
--- a/cli/src/commands/clean.rs
+++ b/cli/src/commands/clean.rs
@@ -38,27 +38,30 @@ pub struct Clean {
 impl Clean {
     /// Cleans the snarkOS node storage.
     pub fn parse(self) -> Result<String> {
+        // Initialize the storage mode.
+        let storage_mode = match self.path {
+            Some(path) => StorageMode::Custom(path),
+            None => match self.dev {
+                Some(id) => StorageMode::Development(id),
+                None => StorageMode::Production,
+            },
+        };
+
         // Remove the current proposal cache file, if it exists.
-        let proposal_cache_path = proposal_cache_path(self.network, self.dev);
+        let proposal_cache_path = proposal_cache_path(self.network, &storage_mode);
         if proposal_cache_path.exists() {
             if let Err(err) = std::fs::remove_file(&proposal_cache_path) {
                 bail!("Failed to remove the current proposal cache file at {}: {err}", proposal_cache_path.display());
             }
         }
         // Remove the specified ledger from storage.
-        Self::remove_ledger(self.network, match self.path {
-            Some(path) => StorageMode::Custom(path),
-            None => match self.dev {
-                Some(id) => StorageMode::Development(id),
-                None => StorageMode::Production,
-            },
-        })
+        Self::remove_ledger(self.network, &storage_mode)
     }
 
     /// Removes the specified ledger from storage.
-    pub(crate) fn remove_ledger(network: u16, mode: StorageMode) -> Result<String> {
+    pub(crate) fn remove_ledger(network: u16, mode: &StorageMode) -> Result<String> {
         // Construct the path to the ledger in storage.
-        let path = aleo_std::aleo_ledger_dir(network, &mode);
+        let path = aleo_std::aleo_ledger_dir(network, mode);
 
         // Prepare the path string.
         let path_string = format!("(in \"{}\")", path.display()).dimmed();
diff --git a/node/bft/examples/simple_node.rs b/node/bft/examples/simple_node.rs
index a0eef2abc..5f8561d9d 100644
--- a/node/bft/examples/simple_node.rs
+++ b/node/bft/examples/simple_node.rs
@@ -56,7 +56,7 @@ use indexmap::IndexMap;
 use rand::{CryptoRng, Rng, SeedableRng};
 use std::{
     collections::HashMap,
-    net::SocketAddr,
+    net::{IpAddr, Ipv4Addr, SocketAddr},
     path::PathBuf,
     str::FromStr,
     sync::{Arc, Mutex, OnceLock, atomic::AtomicBool},
@@ -129,11 +129,12 @@ pub async fn start_bft(
         Arc::new(BFTMemoryService::new()),
         BatchHeader::<CurrentNetwork>::MAX_GC_ROUNDS as u64,
     );
-    // Initialize the gateway IP and dev mode.
-    let (ip, dev) = match peers.get(&node_id) {
-        Some(ip) => (Some(*ip), None),
-        None => (None, Some(node_id)),
+    // Initialize the gateway IP and storage mode.
+    let ip = match peers.get(&node_id) {
+        Some(ip) => Some(*ip),
+        None => Some(SocketAddr::new(IpAddr::V4(Ipv4Addr::LOCALHOST), MEMORY_POOL_PORT + node_id)),
     };
+    let storage_mode = StorageMode::new_test(None);
     // Initialize the trusted validators.
     let trusted_validators = trusted_validators(node_id, num_nodes, peers);
     // Initialize the consensus channels.
@@ -141,7 +142,7 @@ pub async fn start_bft(
     // Initialize the consensus receiver handler.
     consensus_handler(consensus_receiver);
     // Initialize the BFT instance.
-    let mut bft = BFT::<CurrentNetwork>::new(account, storage, ledger, ip, &trusted_validators, dev)?;
+    let mut bft = BFT::<CurrentNetwork>::new(account, storage, ledger, ip, &trusted_validators, storage_mode)?;
     // Run the BFT instance.
     bft.run(Some(consensus_sender), sender.clone(), receiver).await?;
     // Retrieve the BFT's primary.
@@ -170,15 +171,16 @@ pub async fn start_primary(
         Arc::new(BFTMemoryService::new()),
         BatchHeader::<CurrentNetwork>::MAX_GC_ROUNDS as u64,
     );
-    // Initialize the gateway IP and dev mode.
-    let (ip, dev) = match peers.get(&node_id) {
-        Some(ip) => (Some(*ip), None),
-        None => (None, Some(node_id)),
+    // Initialize the gateway IP and storage mode.
+    let ip = match peers.get(&node_id) {
+        Some(ip) => Some(*ip),
+        None => Some(SocketAddr::new(IpAddr::V4(Ipv4Addr::LOCALHOST), MEMORY_POOL_PORT + node_id)),
     };
+    let storage_mode = StorageMode::new_test(None);
     // Initialize the trusted validators.
     let trusted_validators = trusted_validators(node_id, num_nodes, peers);
     // Initialize the primary instance.
-    let mut primary = Primary::<CurrentNetwork>::new(account, storage, ledger, ip, &trusted_validators, dev)?;
+    let mut primary = Primary::<CurrentNetwork>::new(account, storage, ledger, ip, &trusted_validators, storage_mode)?;
     // Run the primary instance.
     primary.run(None, sender.clone(), receiver).await?;
     // Handle OS signals.
diff --git a/node/bft/src/bft.rs b/node/bft/src/bft.rs
index d871e584e..c051692d9 100644
--- a/node/bft/src/bft.rs
+++ b/node/bft/src/bft.rs
@@ -41,6 +41,7 @@ use snarkvm::{
     prelude::{Field, Network, Result, bail, ensure},
 };
 
+use aleo_std::StorageMode;
 use colored::Colorize;
 use indexmap::{IndexMap, IndexSet};
 #[cfg(feature = "locktick")]
@@ -92,10 +93,10 @@ impl<N: Network> BFT<N> {
         ledger: Arc<dyn LedgerService<N>>,
         ip: Option<SocketAddr>,
         trusted_validators: &[SocketAddr],
-        dev: Option<u16>,
+        storage_mode: StorageMode,
     ) -> Result<Self> {
         Ok(Self {
-            primary: Primary::new(account, storage, ledger, ip, trusted_validators, dev)?,
+            primary: Primary::new(account, storage, ledger, ip, trusted_validators, storage_mode)?,
             dag: Default::default(),
             leader_certificate: Default::default(),
             leader_certificate_timer: Default::default(),
@@ -918,6 +919,7 @@ mod tests {
         utilities::TestRng,
     };
 
+    use aleo_std::StorageMode;
     use anyhow::Result;
     use indexmap::{IndexMap, IndexSet};
     use std::sync::Arc;
@@ -978,7 +980,7 @@ mod tests {
         // Initialize the account.
         let account = Account::new(rng)?;
         // Initialize the BFT.
-        let bft = BFT::new(account.clone(), storage.clone(), ledger.clone(), None, &[], None)?;
+        let bft = BFT::new(account.clone(), storage.clone(), ledger.clone(), None, &[], StorageMode::new_test(None))?;
         assert!(bft.is_timer_expired());
         // Ensure this call succeeds on an odd round.
         let result = bft.is_leader_quorum_or_nonleaders_available(1);
@@ -1013,7 +1015,7 @@ mod tests {
         assert_eq!(storage.max_gc_rounds(), 10);
 
         // Initialize the BFT.
-        let bft = BFT::new(account, storage, ledger, None, &[], None)?;
+        let bft = BFT::new(account, storage, ledger, None, &[], StorageMode::new_test(None))?;
         assert!(bft.is_timer_expired()); // 0 + 5 < now()
 
         // Store is at round 1, and we are checking for round 2.
@@ -1035,7 +1037,7 @@ mod tests {
         assert_eq!(storage.max_gc_rounds(), 10);
 
         // Initialize the BFT.
-        let bft = BFT::new(account, storage, ledger, None, &[], None)?;
+        let bft = BFT::new(account, storage, ledger, None, &[], StorageMode::new_test(None))?;
         assert!(bft.is_timer_expired()); // 0 + 5 < now()
 
         // Ensure this call fails on an even round.
@@ -1075,7 +1077,7 @@ mod tests {
         // Initialize the account.
         let account = Account::new(rng)?;
         // Initialize the BFT.
-        let bft = BFT::new(account.clone(), storage.clone(), ledger.clone(), None, &[], None)?;
+        let bft = BFT::new(account.clone(), storage.clone(), ledger.clone(), None, &[], StorageMode::new_test(None))?;
         // Set the leader certificate.
         let leader_certificate = sample_batch_certificate_for_round(2, rng);
         *bft.leader_certificate.write() = Some(leader_certificate);
@@ -1087,7 +1089,8 @@ mod tests {
         assert!(result);
 
         // Initialize a new BFT.
-        let bft_timer = BFT::new(account.clone(), storage.clone(), ledger.clone(), None, &[], None)?;
+        let bft_timer =
+            BFT::new(account.clone(), storage.clone(), ledger.clone(), None, &[], StorageMode::new_test(None))?;
         // If the leader certificate is not set and the timer has not expired, we are not ready for the next round.
         let result = bft_timer.is_even_round_ready_for_next_round(certificates.clone(), committee.clone(), 2);
         if !bft_timer.is_timer_expired() {
@@ -1118,7 +1121,7 @@ mod tests {
         assert_eq!(storage.max_gc_rounds(), 10);
 
         // Initialize the BFT.
-        let bft = BFT::new(account, storage, ledger, None, &[], None)?;
+        let bft = BFT::new(account, storage, ledger, None, &[], StorageMode::new_test(None))?;
 
         // Ensure this call fails on an odd round.
         let result = bft.update_leader_certificate_to_even_round(1);
@@ -1136,7 +1139,7 @@ mod tests {
         assert_eq!(storage.max_gc_rounds(), 10);
 
         // Initialize the BFT.
-        let bft = BFT::new(account, storage, ledger, None, &[], None)?;
+        let bft = BFT::new(account, storage, ledger, None, &[], StorageMode::new_test(None))?;
 
         // Ensure this call succeeds on an even round.
         let result = bft.update_leader_certificate_to_even_round(6);
@@ -1188,7 +1191,7 @@ mod tests {
 
         // Initialize the BFT.
         let account = Account::new(rng)?;
-        let bft = BFT::new(account, storage.clone(), ledger, None, &[], None)?;
+        let bft = BFT::new(account, storage.clone(), ledger, None, &[], StorageMode::new_test(None))?;
 
         // Set the leader certificate.
         *bft.leader_certificate.write() = Some(leader_certificate);
@@ -1226,7 +1229,7 @@ mod tests {
             // Initialize the storage.
             let storage = Storage::new(ledger.clone(), Arc::new(BFTMemoryService::new()), 1);
             // Initialize the BFT.
-            let bft = BFT::new(account.clone(), storage, ledger.clone(), None, &[], None)?;
+            let bft = BFT::new(account.clone(), storage, ledger.clone(), None, &[], StorageMode::new_test(None))?;
 
             // Insert a mock DAG in the BFT.
             *bft.dag.write() = crate::helpers::dag::test_helpers::mock_dag_with_modified_last_committed_round(3);
@@ -1256,7 +1259,7 @@ mod tests {
             // Initialize the storage.
             let storage = Storage::new(ledger.clone(), Arc::new(BFTMemoryService::new()), 1);
             // Initialize the BFT.
-            let bft = BFT::new(account, storage, ledger, None, &[], None)?;
+            let bft = BFT::new(account, storage, ledger, None, &[], StorageMode::new_test(None))?;
 
             // Insert a mock DAG in the BFT.
             *bft.dag.write() = crate::helpers::dag::test_helpers::mock_dag_with_modified_last_committed_round(2);
@@ -1314,7 +1317,7 @@ mod tests {
         /* Test missing previous certificate. */
 
         // Initialize the BFT.
-        let bft = BFT::new(account, storage, ledger, None, &[], None)?;
+        let bft = BFT::new(account, storage, ledger, None, &[], StorageMode::new_test(None))?;
 
         // The expected error message.
         let error_msg = format!(
@@ -1375,7 +1378,7 @@ mod tests {
 
         // Initialize the BFT.
         let account = Account::new(rng)?;
-        let bft = BFT::new(account, storage.clone(), ledger, None, &[], None)?;
+        let bft = BFT::new(account, storage.clone(), ledger, None, &[], StorageMode::new_test(None))?;
         // Insert a mock DAG in the BFT.
         *bft.dag.write() = crate::helpers::dag::test_helpers::mock_dag_with_modified_last_committed_round(commit_round);
 
@@ -1441,7 +1444,7 @@ mod tests {
 
         // Initialize the BFT.
         let account = Account::new(rng)?;
-        let bft = BFT::new(account.clone(), storage, ledger.clone(), None, &[], None)?;
+        let bft = BFT::new(account.clone(), storage, ledger.clone(), None, &[], StorageMode::new_test(None))?;
 
         // Insert a mock DAG in the BFT.
         *bft.dag.write() = crate::helpers::dag::test_helpers::mock_dag_with_modified_last_committed_round(commit_round);
@@ -1459,7 +1462,7 @@ mod tests {
         // Initialize a new instance of storage.
         let storage_2 = Storage::new(ledger.clone(), Arc::new(BFTMemoryService::new()), max_gc_rounds);
         // Initialize a new instance of BFT.
-        let bootup_bft = BFT::new(account, storage_2, ledger, None, &[], None)?;
+        let bootup_bft = BFT::new(account, storage_2, ledger, None, &[], StorageMode::new_test(None))?;
 
         // Sync the BFT DAG at bootup.
         bootup_bft.sync_bft_dag_at_bootup(certificates.clone()).await;
@@ -1613,7 +1616,7 @@ mod tests {
 
         // Initialize the BFT without bootup.
         let account = Account::new(rng)?;
-        let bft = BFT::new(account.clone(), storage, ledger.clone(), None, &[], None)?;
+        let bft = BFT::new(account.clone(), storage, ledger.clone(), None, &[], StorageMode::new_test(None))?;
 
         // Insert a mock DAG in the BFT without bootup.
         *bft.dag.write() = crate::helpers::dag::test_helpers::mock_dag_with_modified_last_committed_round(0);
@@ -1638,7 +1641,8 @@ mod tests {
         let bootup_storage = Storage::new(ledger.clone(), Arc::new(BFTMemoryService::new()), max_gc_rounds);
 
         // Initialize a new instance of BFT with bootup.
-        let bootup_bft = BFT::new(account, bootup_storage.clone(), ledger.clone(), None, &[], None)?;
+        let bootup_bft =
+            BFT::new(account, bootup_storage.clone(), ledger.clone(), None, &[], StorageMode::new_test(None))?;
 
         // Sync the BFT DAG at bootup.
         bootup_bft.sync_bft_dag_at_bootup(pre_shutdown_certificates.clone()).await;
@@ -1816,7 +1820,8 @@ mod tests {
         }
         // Initialize the bootup BFT.
         let account = Account::new(rng)?;
-        let bootup_bft = BFT::new(account.clone(), storage.clone(), ledger.clone(), None, &[], None)?;
+        let bootup_bft =
+            BFT::new(account.clone(), storage.clone(), ledger.clone(), None, &[], StorageMode::new_test(None))?;
         // Insert a mock DAG in the BFT without bootup.
         *bootup_bft.dag.write() = crate::helpers::dag::test_helpers::mock_dag_with_modified_last_committed_round(0);
         // Sync the BFT DAG at bootup.
diff --git a/node/bft/src/helpers/proposal_cache.rs b/node/bft/src/helpers/proposal_cache.rs
index 4664c2b42..2af7b9872 100644
--- a/node/bft/src/helpers/proposal_cache.rs
+++ b/node/bft/src/helpers/proposal_cache.rs
@@ -26,16 +26,14 @@ use indexmap::IndexSet;
 use std::{fs, path::PathBuf};
 
 /// Returns the path where a proposal cache file may be stored.
-pub fn proposal_cache_path(network: u16, dev: Option<u16>) -> PathBuf {
+pub fn proposal_cache_path(network: u16, storage_mode: &StorageMode) -> PathBuf {
     const PROPOSAL_CACHE_FILE_NAME: &str = "current-proposal-cache";
-
-    let storage_mode = if let Some(id) = dev { StorageMode::Development(id) } else { StorageMode::Production };
     // Obtain the path to the ledger.
-    let mut path = aleo_ledger_dir(network, &storage_mode);
+    let mut path = aleo_ledger_dir(network, storage_mode);
     // Go to the folder right above the ledger.
     path.pop();
     // Append the proposal store's file name.
-    match dev {
+    match storage_mode.dev() {
         Some(id) => path.push(format!(".{PROPOSAL_CACHE_FILE_NAME}-{}-{}", network, id)),
         None => path.push(format!("{PROPOSAL_CACHE_FILE_NAME}-{}", network)),
     }
@@ -79,14 +77,14 @@ impl<N: Network> ProposalCache<N> {
     }
 
     /// Returns `true` if a proposal cache exists for the given network and `dev`.
-    pub fn exists(dev: Option<u16>) -> bool {
-        proposal_cache_path(N::ID, dev).exists()
+    pub fn exists(storage_mode: &StorageMode) -> bool {
+        proposal_cache_path(N::ID, storage_mode).exists()
     }
 
     /// Load the proposal cache from the file system and ensure that the proposal cache is valid.
-    pub fn load(expected_signer: Address<N>, dev: Option<u16>) -> Result<Self> {
+    pub fn load(expected_signer: Address<N>, storage_mode: &StorageMode) -> Result<Self> {
         // Construct the proposal cache file system path.
-        let path = proposal_cache_path(N::ID, dev);
+        let path = proposal_cache_path(N::ID, storage_mode);
 
         // Deserialize the proposal cache from the file system.
         let proposal_cache = match fs::read(&path) {
@@ -108,8 +106,8 @@ impl<N: Network> ProposalCache<N> {
     }
 
     /// Store the proposal cache to the file system.
-    pub fn store(&self, dev: Option<u16>) -> Result<()> {
-        let path = proposal_cache_path(N::ID, dev);
+    pub fn store(&self, storage_mode: &StorageMode) -> Result<()> {
+        let path = proposal_cache_path(N::ID, storage_mode);
         info!("Storing the proposal cache to {}...", path.display());
 
         // Serialize the proposal cache.
diff --git a/node/bft/src/primary.rs b/node/bft/src/primary.rs
index 6a37b1b0c..2bd557ff1 100644
--- a/node/bft/src/primary.rs
+++ b/node/bft/src/primary.rs
@@ -58,6 +58,7 @@ use snarkvm::{
     prelude::committee::Committee,
 };
 
+use aleo_std::StorageMode;
 use colored::Colorize;
 use futures::stream::{FuturesUnordered, StreamExt};
 use indexmap::{IndexMap, IndexSet};
@@ -107,6 +108,8 @@ pub struct Primary<N: Network> {
     handles: Arc<Mutex<Vec<JoinHandle<()>>>>,
     /// The lock for propose_batch.
     propose_lock: Arc<TMutex<u64>>,
+    /// The storage mode of the node.
+    storage_mode: StorageMode,
 }
 
 impl<N: Network> Primary<N> {
@@ -120,10 +123,11 @@ impl<N: Network> Primary<N> {
         ledger: Arc<dyn LedgerService<N>>,
         ip: Option<SocketAddr>,
         trusted_validators: &[SocketAddr],
-        dev: Option<u16>,
+        storage_mode: StorageMode,
     ) -> Result<Self> {
         // Initialize the gateway.
-        let gateway = Gateway::new(account, storage.clone(), ledger.clone(), ip, trusted_validators, dev)?;
+        let gateway =
+            Gateway::new(account, storage.clone(), ledger.clone(), ip, trusted_validators, storage_mode.dev())?;
         // Initialize the sync module.
         let sync = Sync::new(gateway.clone(), storage.clone(), ledger.clone());
 
@@ -140,15 +144,16 @@ impl<N: Network> Primary<N> {
             signed_proposals: Default::default(),
             handles: Default::default(),
             propose_lock: Default::default(),
+            storage_mode,
         })
     }
 
     /// Load the proposal cache file and update the Primary state with the stored data.
     async fn load_proposal_cache(&self) -> Result<()> {
         // Fetch the signed proposals from the file system if it exists.
-        match ProposalCache::<N>::exists(self.gateway.dev()) {
+        match ProposalCache::<N>::exists(&self.storage_mode) {
             // If the proposal cache exists, then process the proposal cache.
-            true => match ProposalCache::<N>::load(self.gateway.account().address(), self.gateway.dev()) {
+            true => match ProposalCache::<N>::load(self.gateway.account().address(), &self.storage_mode) {
                 Ok(proposal_cache) => {
                     // Extract the proposal and signed proposals.
                     let (latest_certificate_round, proposed_batch, signed_proposals, pending_certificates) =
@@ -1878,7 +1883,7 @@ impl<N: Network> Primary<N> {
             let pending_certificates = self.storage.get_pending_certificates();
             ProposalCache::new(latest_round, proposal, signed_proposals, pending_certificates)
         };
-        if let Err(err) = proposal_cache.store(self.gateway.dev()) {
+        if let Err(err) = proposal_cache.store(&self.storage_mode) {
             error!("Failed to store the current proposal cache: {err}");
         }
         // Close the gateway.
@@ -1934,7 +1939,7 @@ mod tests {
 
         // Initialize the primary.
         let account = accounts[account_index].1.clone();
-        let mut primary = Primary::new(account, storage, ledger, None, &[], None).unwrap();
+        let mut primary = Primary::new(account, storage, ledger, None, &[], StorageMode::new_test(None)).unwrap();
 
         // Construct a worker instance.
         primary.workers = Arc::from([Worker::new(
diff --git a/node/bft/src/sync/mod.rs b/node/bft/src/sync/mod.rs
index a88aaf289..8c2223584 100644
--- a/node/bft/src/sync/mod.rs
+++ b/node/bft/src/sync/mod.rs
@@ -793,7 +793,7 @@ mod tests {
         ];
 
         // Initialize the ledger with the genesis block.
-        let ledger = CurrentLedger::load(genesis.clone(), StorageMode::Production).unwrap();
+        let ledger = CurrentLedger::load(genesis.clone(), StorageMode::new_test(None)).unwrap();
         // Initialize the ledger.
         let core_ledger = Arc::new(CoreLedgerService::new(ledger.clone(), Default::default()));
 
@@ -967,7 +967,7 @@ mod tests {
 
         // Initialize the syncing ledger.
         let syncing_ledger = Arc::new(CoreLedgerService::new(
-            CurrentLedger::load(genesis, StorageMode::Production).unwrap(),
+            CurrentLedger::load(genesis, StorageMode::new_test(None)).unwrap(),
             Default::default(),
         ));
         // Initialize the gateway.
@@ -1016,7 +1016,7 @@ mod tests {
             PrivateKey::new(genesis_rng)?,
         ];
         // Initialize the ledger with the genesis block.
-        let ledger = CurrentLedger::load(genesis.clone(), StorageMode::Production).unwrap();
+        let ledger = CurrentLedger::load(genesis.clone(), StorageMode::new_test(None)).unwrap();
         // Initialize the ledger.
         let core_ledger = Arc::new(CoreLedgerService::new(ledger.clone(), Default::default()));
         // Sample rounds of batch certificates starting at the genesis round from a static set of 4 authors.
diff --git a/node/bft/tests/common/primary.rs b/node/bft/tests/common/primary.rs
index 196143cc9..29aaceb78 100644
--- a/node/bft/tests/common/primary.rs
+++ b/node/bft/tests/common/primary.rs
@@ -22,6 +22,7 @@ use snarkos_account::Account;
 use snarkos_node_bft::{
     BFT,
     MAX_BATCH_DELAY_IN_MS,
+    MEMORY_POOL_PORT,
     Primary,
     helpers::{PrimarySender, Storage, init_primary_channels},
 };
@@ -52,6 +53,7 @@ use locktick::parking_lot::Mutex;
 use parking_lot::Mutex;
 use std::{
     collections::HashMap,
+    net::{IpAddr, Ipv4Addr, SocketAddr},
     ops::RangeBounds,
     sync::{Arc, OnceLock},
     time::Duration,
@@ -164,11 +166,26 @@ impl TestNetwork {
             );
 
             let (primary, bft) = if config.bft {
-                let bft = BFT::<CurrentNetwork>::new(account, storage, ledger, None, &[], Some(id as u16)).unwrap();
+                let bft = BFT::<CurrentNetwork>::new(
+                    account,
+                    storage,
+                    ledger,
+                    Some(SocketAddr::new(IpAddr::V4(Ipv4Addr::LOCALHOST), MEMORY_POOL_PORT + id as u16)),
+                    &[],
+                    StorageMode::new_test(None),
+                )
+                .unwrap();
                 (bft.primary().clone(), Some(bft))
             } else {
-                let primary =
-                    Primary::<CurrentNetwork>::new(account, storage, ledger, None, &[], Some(id as u16)).unwrap();
+                let primary = Primary::<CurrentNetwork>::new(
+                    account,
+                    storage,
+                    ledger,
+                    Some(SocketAddr::new(IpAddr::V4(Ipv4Addr::LOCALHOST), MEMORY_POOL_PORT + id as u16)),
+                    &[],
+                    StorageMode::new_test(None),
+                )
+                .unwrap();
                 (primary, None)
             };
 
@@ -402,5 +419,5 @@ pub fn genesis_ledger(
         })
         .clone();
     // Initialize the ledger with the genesis block.
-    CurrentLedger::load(block, StorageMode::Production).unwrap()
+    CurrentLedger::load(block, StorageMode::new_test(None)).unwrap()
 }
diff --git a/node/consensus/src/lib.rs b/node/consensus/src/lib.rs
index 8f4d3abb2..b8bf93ae3 100644
--- a/node/consensus/src/lib.rs
+++ b/node/consensus/src/lib.rs
@@ -121,17 +121,12 @@ impl<N: Network> Consensus<N> {
         trusted_validators: &[SocketAddr],
         storage_mode: StorageMode,
     ) -> Result<Self> {
-        // Recover the development ID, if it is present.
-        let dev = match storage_mode {
-            StorageMode::Development(id) => Some(id),
-            StorageMode::Production | StorageMode::Custom(..) | StorageMode::Test(_) => None,
-        };
         // Initialize the Narwhal transmissions.
-        let transmissions = Arc::new(BFTPersistentStorage::open(storage_mode)?);
+        let transmissions = Arc::new(BFTPersistentStorage::open(storage_mode.clone())?);
         // Initialize the Narwhal storage.
         let storage = NarwhalStorage::new(ledger.clone(), transmissions, BatchHeader::<N>::MAX_GC_ROUNDS as u64);
         // Initialize the BFT.
-        let bft = BFT::new(account, storage, ledger.clone(), ip, trusted_validators, dev)?;
+        let bft = BFT::new(account, storage, ledger.clone(), ip, trusted_validators, storage_mode)?;
         // Return the consensus.
         Ok(Self {
             ledger,
diff --git a/node/tests/common/node.rs b/node/tests/common/node.rs
index 4d54fe87c..930210dfc 100644
--- a/node/tests/common/node.rs
+++ b/node/tests/common/node.rs
@@ -30,7 +30,7 @@ pub async fn client() -> Client<CurrentNetwork, ConsensusMemory<CurrentNetwork>>
         &[],
         sample_genesis_block(),
         None, // No CDN.
-        StorageMode::Production,
+        StorageMode::new_test(None),
         false, // No extra peer rotation.
         Default::default(),
     )
@@ -44,7 +44,7 @@ pub async fn prover() -> Prover<CurrentNetwork, ConsensusMemory<CurrentNetwork>>
         Account::<CurrentNetwork>::from_str("APrivateKey1zkp2oVPTci9kKcUprnbzMwq95Di1MQERpYBhEeqvkrDirK1").unwrap(),
         &[],
         sample_genesis_block(),
-        StorageMode::Production,
+        StorageMode::new_test(None),
         Default::default(),
     )
     .await
@@ -62,7 +62,7 @@ pub async fn validator() -> Validator<CurrentNetwork, ConsensusMemory<CurrentNet
         &[],
         sample_genesis_block(), // Should load the current network's genesis block.
         None,                   // No CDN.
-        StorageMode::Production,
+        StorageMode::new_test(None),
         true,  // This test requires validators to connect to peers.
         false, // No dev traffic in production mode.
         Default::default(),
