diff --git a/algorithms/src/snark/varuna/ahp/prover/oracles.rs b/algorithms/src/snark/varuna/ahp/prover/oracles.rs
index b4b645346..725b6a291 100644
--- a/algorithms/src/snark/varuna/ahp/prover/oracles.rs
+++ b/algorithms/src/snark/varuna/ahp/prover/oracles.rs
@@ -112,17 +112,17 @@ impl<F: PrimeField> ThirdOracles<F> {
 /// The fourth set of prover oracles.
 #[derive(Debug)]
 pub struct FourthOracles<F: PrimeField> {
-    pub(super) gs: BTreeMap<CircuitId, MatrixGs<F>>,
+    pub(in crate::snark::varuna) gs: BTreeMap<CircuitId, MatrixGs<F>>,
 }
 
 #[derive(Debug)]
-pub(super) struct MatrixGs<F: PrimeField> {
+pub(in crate::snark::varuna) struct MatrixGs<F: PrimeField> {
     /// The polynomial `g_a` resulting from the second sumcheck.
-    pub(super) g_a: LabeledPolynomial<F>,
+    pub(in crate::snark::varuna) g_a: LabeledPolynomial<F>,
     /// The polynomial `g_b` resulting from the second sumcheck.
-    pub(super) g_b: LabeledPolynomial<F>,
+    pub(in crate::snark::varuna) g_b: LabeledPolynomial<F>,
     /// The polynomial `g_c` resulting from the second sumcheck.
-    pub(super) g_c: LabeledPolynomial<F>,
+    pub(in crate::snark::varuna) g_c: LabeledPolynomial<F>,
 }
 
 impl<F: PrimeField> MatrixGs<F> {
diff --git a/algorithms/src/snark/varuna/ahp/prover/round_functions/third.rs b/algorithms/src/snark/varuna/ahp/prover/round_functions/third.rs
index 4e994d498..bccfedfaa 100644
--- a/algorithms/src/snark/varuna/ahp/prover/round_functions/third.rs
+++ b/algorithms/src/snark/varuna/ahp/prover/round_functions/third.rs
@@ -218,7 +218,9 @@ impl<F: PrimeField, SM: SNARKMode> AHPForR1CS<F, SM> {
         Ok((h_1_sum, xg_1_sum, msg))
     }
 
-    fn calculate_assignments(state: &mut prover::State<F, SM>) -> Result<BTreeMap<CircuitId, Vec<DensePolynomial<F>>>> {
+    pub(in crate::snark::varuna) fn calculate_assignments(
+        state: &mut prover::State<F, SM>,
+    ) -> Result<BTreeMap<CircuitId, Vec<DensePolynomial<F>>>> {
         let assignments_time = start_timer!(|| "Calculate assignments");
         let assignments: BTreeMap<_, _> = state
             .circuit_specific_states
diff --git a/algorithms/src/snark/varuna/data_structures/test_circuit.rs b/algorithms/src/snark/varuna/data_structures/test_circuit.rs
index 8e66d71f0..e269ef7d4 100644
--- a/algorithms/src/snark/varuna/data_structures/test_circuit.rs
+++ b/algorithms/src/snark/varuna/data_structures/test_circuit.rs
@@ -84,6 +84,7 @@ impl<ConstraintF: Field> ConstraintSynthesizer<ConstraintF> for TestCircuit<Cons
 }
 
 impl<F: Field> TestCircuit<F> {
+    // Generate a test circuit with a random witness.
     pub fn gen_rand<R: Rng + CryptoRng>(
         mul_depth: usize,
         num_constraints: usize,
@@ -104,4 +105,26 @@ impl<F: Field> TestCircuit<F> {
 
         (TestCircuit { a: Some(a), b: Some(b), num_constraints, num_variables, mul_depth }, public_inputs)
     }
+
+    // Generate a test circuit with a fixed witness.
+    pub fn generate_circuit_with_fixed_witness(
+        a: u128,
+        b: u128,
+        mul_depth: usize,
+        num_constraints: usize,
+        num_variables: usize,
+    ) -> (Self, Vec<F>) {
+        let mut public_inputs: Vec<F> = Vec::with_capacity(mul_depth);
+        let a = F::from(a);
+        let b = F::from(b);
+        for j in 1..(mul_depth + 1) {
+            let mut new_var = a;
+            for _ in 0..j {
+                new_var.mul_assign(&b);
+            }
+            public_inputs.push(new_var);
+        }
+
+        (TestCircuit { a: Some(a), b: Some(b), num_constraints, num_variables, mul_depth }, public_inputs)
+    }
 }
diff --git a/algorithms/src/snark/varuna/resources/circuit_0/challenges.input b/algorithms/src/snark/varuna/resources/circuit_0/challenges.input
new file mode 100644
index 000000000..c452bb9ba
--- /dev/null
+++ b/algorithms/src/snark/varuna/resources/circuit_0/challenges.input
@@ -0,0 +1,9 @@
+3848747268438146429751199396409351181775389242768022193485885831738448017147
+1
+8197944265508088395536605774074305135172727109973647025295290482999689956740
+969057258436037177120044092706484307847087860293738150232755543560372962965
+1261454636320080423466301508402274008580035865105120100172548996301504688503
+1
+4987583518937978349829618221849930643691403053432331091973338029344238378359
+5292820491592383411924896857610185298390500160570506754003580089093852949536
+672738024541753390172108082983901395072703770443783662610485842877496432861
\ No newline at end of file
diff --git a/algorithms/src/snark/varuna/resources/circuit_0/domain/C.txt b/algorithms/src/snark/varuna/resources/circuit_0/domain/C.txt
new file mode 100644
index 000000000..4fee833f1
--- /dev/null
+++ b/algorithms/src/snark/varuna/resources/circuit_0/domain/C.txt
@@ -0,0 +1 @@
+[1, 3279917132858342911831074864712036382710139745724269329239664300762234227201, 880904806456922042258150504921383618666682042621506879489, 1973030855696769125460623085327505793054673234941098473458474059731617992635, 8444461749428370424248824938781546531375899335154063827935233455917409239040, 5164544616570027512417750074069510148665759589429794498695569155155175011840, 8444461749428370423367920132324624489117748830232680209268551413295902359552, 6471430893731601298788201853454040738321226100212965354476759396185791246406]
\ No newline at end of file
diff --git a/algorithms/src/snark/varuna/resources/circuit_0/domain/K.txt b/algorithms/src/snark/varuna/resources/circuit_0/domain/K.txt
new file mode 100644
index 000000000..4fee833f1
--- /dev/null
+++ b/algorithms/src/snark/varuna/resources/circuit_0/domain/K.txt
@@ -0,0 +1 @@
+[1, 3279917132858342911831074864712036382710139745724269329239664300762234227201, 880904806456922042258150504921383618666682042621506879489, 1973030855696769125460623085327505793054673234941098473458474059731617992635, 8444461749428370424248824938781546531375899335154063827935233455917409239040, 5164544616570027512417750074069510148665759589429794498695569155155175011840, 8444461749428370423367920132324624489117748830232680209268551413295902359552, 6471430893731601298788201853454040738321226100212965354476759396185791246406]
\ No newline at end of file
diff --git a/algorithms/src/snark/varuna/resources/circuit_0/domain/R.txt b/algorithms/src/snark/varuna/resources/circuit_0/domain/R.txt
new file mode 100644
index 000000000..4fee833f1
--- /dev/null
+++ b/algorithms/src/snark/varuna/resources/circuit_0/domain/R.txt
@@ -0,0 +1 @@
+[1, 3279917132858342911831074864712036382710139745724269329239664300762234227201, 880904806456922042258150504921383618666682042621506879489, 1973030855696769125460623085327505793054673234941098473458474059731617992635, 8444461749428370424248824938781546531375899335154063827935233455917409239040, 5164544616570027512417750074069510148665759589429794498695569155155175011840, 8444461749428370423367920132324624489117748830232680209268551413295902359552, 6471430893731601298788201853454040738321226100212965354476759396185791246406]
\ No newline at end of file
diff --git a/algorithms/src/snark/varuna/resources/circuit_0/instance.input b/algorithms/src/snark/varuna/resources/circuit_0/instance.input
new file mode 100644
index 000000000..57534290a
--- /dev/null
+++ b/algorithms/src/snark/varuna/resources/circuit_0/instance.input
@@ -0,0 +1,24 @@
+A
+0, 0, 0, 0, 1, 0, 0,
+0, 0, 0, 0, 1, 0, 0,
+0, 0, 0, 0, 1, 0, 0,
+0, 0, 0, 0, 1, 0, 0,
+0, 0, 0, 0, 1, 0, 0,
+0, 1, 0, 0, 0, 0, 0,
+0, 0, 1, 0, 0, 0, 0,
+B
+0, 0, 0, 0, 0, 1, 0,
+0, 0, 0, 0, 0, 1, 0,
+0, 0, 0, 0, 0, 1, 0,
+0, 0, 0, 0, 0, 1, 0,
+0, 0, 0, 0, 0, 1, 0,
+0, 0, 0, 0, 0, 1, 0,
+0, 0, 0, 0, 0, 1, 0,
+C
+0, 1, 0, 0, 0, 0, 0,
+0, 1, 0, 0, 0, 0, 0,
+0, 1, 0, 0, 0, 0, 0,
+0, 1, 0, 0, 0, 0, 0,
+0, 1, 0, 0, 0, 0, 0,
+0, 0, 1, 0, 0, 0, 0,
+0, 0, 0, 1, 0, 0, 0,
\ No newline at end of file
diff --git a/algorithms/src/snark/varuna/resources/circuit_0/polynomials/g_1.txt b/algorithms/src/snark/varuna/resources/circuit_0/polynomials/g_1.txt
new file mode 100644
index 000000000..66c9b89ec
--- /dev/null
+++ b/algorithms/src/snark/varuna/resources/circuit_0/polynomials/g_1.txt
@@ -0,0 +1 @@
+[6619044933395879708506708233785160900848417926433900807753717609232615537930, 3821893577243645065467553327033258263806878911070937029265276300425404027087, 1076457615332399821357918023630894203218737643598338498309569170592118610105, 4055793519271390007054111087031631384766037777614854573354385584678647572635, 5151025438151686500811984829825775386132972581217168951169837107565191699213, 5451596424008550546888850265776737232625677836482903207071209344567449341848, 3448452754104072214340119903505567632967740660166690880356165129542564717469]
\ No newline at end of file
diff --git a/algorithms/src/snark/varuna/resources/circuit_0/polynomials/g_a.txt b/algorithms/src/snark/varuna/resources/circuit_0/polynomials/g_a.txt
new file mode 100644
index 000000000..406dfcb42
--- /dev/null
+++ b/algorithms/src/snark/varuna/resources/circuit_0/polynomials/g_a.txt
@@ -0,0 +1 @@
+[7298426509888201269891881417058316917443386443130089954044569917277532451124, 4513665418379767804370095812278055346820156274660773052289744107232412707613, 4475996609146606675867680993589252805148646238137939055092225188380378102122, 1164867027976035030197277870836287693729511372979478632421748407217397685680, 241156138388132028780023732378805911361134319610119168887483568462409314180, 5388924537180330478887903617303497159805558377912056989902396473207071789448, 4874868009316021445675218576593739719745472140358591754846406994220385278800]
\ No newline at end of file
diff --git a/algorithms/src/snark/varuna/resources/circuit_0/polynomials/g_b.txt b/algorithms/src/snark/varuna/resources/circuit_0/polynomials/g_b.txt
new file mode 100644
index 000000000..5bd1204a5
--- /dev/null
+++ b/algorithms/src/snark/varuna/resources/circuit_0/polynomials/g_b.txt
@@ -0,0 +1 @@
+[5156102875137215468349822857341787486070810620190600082313993541539039694040, 642243162042132842519243997058576263060505644548708158364282235038668235582, 4232914989795410755789500032792636846277428417456597259068424201075935215723, 5294805659491479092801660967840299543626893059046679165413663075966109761338, 4428888977080738940451109661040008630918481794519438117648033775895551925110, 4353953527706377329056342142048774285909447424786294362569762658050561552035, 3147342381590763655483856523901415730605847626281467108012225760335596205826]
\ No newline at end of file
diff --git a/algorithms/src/snark/varuna/resources/circuit_0/polynomials/g_c.txt b/algorithms/src/snark/varuna/resources/circuit_0/polynomials/g_c.txt
new file mode 100644
index 000000000..5bd1204a5
--- /dev/null
+++ b/algorithms/src/snark/varuna/resources/circuit_0/polynomials/g_c.txt
@@ -0,0 +1 @@
+[5156102875137215468349822857341787486070810620190600082313993541539039694040, 642243162042132842519243997058576263060505644548708158364282235038668235582, 4232914989795410755789500032792636846277428417456597259068424201075935215723, 5294805659491479092801660967840299543626893059046679165413663075966109761338, 4428888977080738940451109661040008630918481794519438117648033775895551925110, 4353953527706377329056342142048774285909447424786294362569762658050561552035, 3147342381590763655483856523901415730605847626281467108012225760335596205826]
\ No newline at end of file
diff --git a/algorithms/src/snark/varuna/resources/circuit_0/polynomials/h_0.txt b/algorithms/src/snark/varuna/resources/circuit_0/polynomials/h_0.txt
new file mode 100644
index 000000000..5b5fb1391
--- /dev/null
+++ b/algorithms/src/snark/varuna/resources/circuit_0/polynomials/h_0.txt
@@ -0,0 +1 @@
+[7388904030749824121217721821433853214953911918259805849443329273927733084164, 2135176220032946734943829824459106118384054561849250372330716048683975588842, 4754990573109988415472033418479071784044875994712276662893248150540227089791, 2758334532884618923758430413601802477377254380601541511107093225976782776983, 8095203738994452079955122826479915970380729551727226719303332837983196321727, 6177424990788360784192413675740409023496996883035955845846493361503576983134, 5094286905430846656542049153180259645877255793223574061499131126611824661186]
\ No newline at end of file
diff --git a/algorithms/src/snark/varuna/resources/circuit_0/polynomials/h_1.txt b/algorithms/src/snark/varuna/resources/circuit_0/polynomials/h_1.txt
new file mode 100644
index 000000000..b635e6289
--- /dev/null
+++ b/algorithms/src/snark/varuna/resources/circuit_0/polynomials/h_1.txt
@@ -0,0 +1 @@
+[165474815053007573527603177207806062734264344258107184951357675349729517680, 5801432311648519240757166384710257525281134619669773204356367264314316460174, 5779294849724998706280834835495248447327865373185082571164164495999431007313, 6195154650395302844568965061674617811933630360198179606653410538752273578806, 1424571001001798030720515357589755476564998467791760803818317281629675760920, 644304059262549681272611563429424161652824970479283239535226429205883512483, 1985534495267414125495515420354543072112034540966649759533850695585151887015]
\ No newline at end of file
diff --git a/algorithms/src/snark/varuna/resources/circuit_0/polynomials/h_2.txt b/algorithms/src/snark/varuna/resources/circuit_0/polynomials/h_2.txt
new file mode 100644
index 000000000..ffe41016e
--- /dev/null
+++ b/algorithms/src/snark/varuna/resources/circuit_0/polynomials/h_2.txt
@@ -0,0 +1 @@
+[8290113549992594314646287997621385217684845052221443536793209549506488920846, 4219277537425088937784490235792966206782906842687518132171996608740394894532, 590462121801222593314525094454279217199506852881783863984329676220226405050, 2657750100182903992679485191896472785743422123052387651366434608567501075462, 591064579761258100262799594047186364485022182696204209512789652131850886891, 3322216460403101876331086740648778039637725754343916626810493733358099221767, 8341638249347740785844637855796311718756490590925437333299426275491585624411]
\ No newline at end of file
diff --git a/algorithms/src/snark/varuna/resources/circuit_0/polynomials/w_lde.txt b/algorithms/src/snark/varuna/resources/circuit_0/polynomials/w_lde.txt
new file mode 100644
index 000000000..33e90cb35
--- /dev/null
+++ b/algorithms/src/snark/varuna/resources/circuit_0/polynomials/w_lde.txt
@@ -0,0 +1 @@
+[7388904030749824121217721821433853214953911918259805849443329273927733084181, 4806631722464810378222418881252928774493289697365662880095775335672748770572, 7388904030749824121217721821433853214953911918259805849443329273927733084148, 4153188583884023458610048797853002211921041294332568892204718936512234268619]
\ No newline at end of file
diff --git a/algorithms/src/snark/varuna/resources/circuit_0/polynomials/z_lde.txt b/algorithms/src/snark/varuna/resources/circuit_0/polynomials/z_lde.txt
new file mode 100644
index 000000000..89b0c81ec
--- /dev/null
+++ b/algorithms/src/snark/varuna/resources/circuit_0/polynomials/z_lde.txt
@@ -0,0 +1 @@
+[7388904030749824121217721821433853214953911918259805849443329273927733084183, 1526714589606467466391344016540892391783149951641393550856111034910514543371, 7388904030749824121217721821433853214953911918259805849443329273927733084148, 2180157728187254333149425712525496418866368059391470418746244876780616275984, 7388904030749824121217721821433853214953911918259805849443329273927733084181, 4806631722464810378222418881252928774493289697365662880095775335672748770572, 7388904030749824121217721821433853214953911918259805849443329273927733084148, 4153188583884023458610048797853002211921041294332568892204718936512234268619]
\ No newline at end of file
diff --git a/algorithms/src/snark/varuna/resources/circuit_0/witness.input b/algorithms/src/snark/varuna/resources/circuit_0/witness.input
new file mode 100644
index 000000000..4e9e62114
--- /dev/null
+++ b/algorithms/src/snark/varuna/resources/circuit_0/witness.input
@@ -0,0 +1,2 @@
+[2, 4]
+[1, 8, 32, 128, 2, 4, 2]
\ No newline at end of file
diff --git a/algorithms/src/snark/varuna/tests.rs b/algorithms/src/snark/varuna/tests.rs
index 8c7edaa3b..70eeb548e 100644
--- a/algorithms/src/snark/varuna/tests.rs
+++ b/algorithms/src/snark/varuna/tests.rs
@@ -531,3 +531,281 @@ mod varuna_hiding {
         assert!(VarunaInst::verify(universal_verifier, &fs_parameters, &vk1, public_inputs1, &proof1).unwrap());
     }
 }
+
+mod varuna_test_vectors {
+    use crate::{
+        fft::EvaluationDomain,
+        snark::varuna::{ahp::verifier, AHPForR1CS, TestCircuit, VarunaNonHidingMode, VarunaSNARK},
+        traits::snark::SNARK,
+    };
+    use snarkvm_curves::bls12_377::{Bls12_377, Fq, Fr};
+    use snarkvm_fields::One;
+    use std::{collections::BTreeMap, fs, ops::Deref, path::PathBuf, str::FromStr, sync::Arc};
+
+    type FS = crate::crypto_hash::PoseidonSponge<Fq, 2, 1>;
+    type MM = VarunaNonHidingMode;
+    type VarunaSonicInst = VarunaSNARK<Bls12_377, FS, MM>;
+
+    // Create the path for the `resources` folder.
+    fn resources_path(create_dir: bool) -> PathBuf {
+        let mut path = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
+        path.push("src");
+        path.push("snark");
+        path.push("varuna");
+        path.push("resources");
+
+        // Create the `resources` folder, if it does not exist.
+        if !path.exists() {
+            if create_dir {
+                fs::create_dir(&path).unwrap_or_else(|_| panic!("Failed to create resources folder: {path:?}"));
+            } else {
+                panic!("Resources folder does not exist: {path:?}");
+            }
+        }
+
+        path
+    }
+
+    // Create the file path.
+    fn test_vector_path(folder: &str, file: &str, circuit: &str, create_dir: bool) -> PathBuf {
+        let mut path = resources_path(create_dir);
+
+        // Construct the path where the test data lives.
+        path.push(circuit);
+        path.push(folder);
+
+        // Create the test folder if it does not exist if specified, otherwise panic.
+        if !path.exists() {
+            if create_dir {
+                fs::create_dir(&path).unwrap_or_else(|_| panic!("Failed to create resources folder: {path:?}"));
+            } else {
+                panic!("Resources folder does not exist: {path:?}");
+            }
+        }
+
+        // Construct the path for the test file.
+        path.push(file);
+        path.set_extension("txt");
+
+        path
+    }
+
+    // Loads the given `test_folder/test_file` and asserts the given `candidate` matches the expected values.
+    #[track_caller]
+    fn assert_test_vector_equality(test_folder: &str, test_file: &str, candidate: &str, circuit: &str) {
+        // Get the path to the test file.
+        let path = test_vector_path(test_folder, test_file, circuit, false);
+
+        // Assert the test file is equal to the expected value.
+        expect_test::expect_file![path].assert_eq(candidate);
+    }
+
+    // Create a test vector from a trusted revision of Varuna.
+    fn create_test_vector(folder: &str, file: &str, data: &str, circuit: &str) {
+        // Get the path to the test file.
+        let path = test_vector_path(folder, file, circuit, true);
+
+        // Write the test vector to file.
+        fs::write(&path, data).unwrap_or_else(|_| panic!("Failed to write to file: {:?}", path));
+    }
+
+    // Tests varuna against the test vectors in all circuits in the resources folder.
+    fn test_varuna_with_all_circuits(create_test_vectors: bool) {
+        let entries = fs::read_dir(resources_path(create_test_vectors)).expect("Failed to read resources folder");
+        entries.into_iter().for_each(|entry| {
+            let path = entry.unwrap().path();
+            if path.is_dir() {
+                let circuit = path.file_name().unwrap().to_str().unwrap();
+                test_circuit_with_test_vectors(create_test_vectors, circuit);
+            }
+        });
+    }
+
+    // Test Varuna against test vectors for a specific circuit.
+    fn test_circuit_with_test_vectors(create_test_vectors: bool, circuit: &str) {
+        // Initialize the parts of the witness used in the multiplicative constraints.
+        let witness_path = format!("src/snark/varuna/resources/{}/witness.input", circuit);
+        let instance_file = fs::read_to_string(witness_path).expect("Could not read the file");
+        let witness: Vec<u128> = serde_json::from_str(instance_file.lines().next().unwrap()).unwrap();
+        let (a, b) = (witness[0], witness[1]);
+
+        // Initialize challenges from file.
+        let challenges_path = format!("src/snark/varuna/resources/{}/challenges.input", circuit);
+        let challenges_file = fs::read_to_string(challenges_path).expect("Could not read the file");
+        let mut challenges = Vec::new();
+        for line in challenges_file.lines() {
+            challenges.push(line)
+        }
+        let (alpha, _eta_a, eta_b, eta_c, beta, delta_a, delta_b, delta_c, _gamma) = (
+            Fr::from_str(challenges[0]).unwrap(),
+            Fr::from_str(challenges[1]).unwrap(),
+            Fr::from_str(challenges[2]).unwrap(),
+            Fr::from_str(challenges[3]).unwrap(),
+            Fr::from_str(challenges[4]).unwrap(),
+            vec![Fr::from_str(challenges[5]).unwrap()],
+            vec![Fr::from_str(challenges[6]).unwrap()],
+            vec![Fr::from_str(challenges[7]).unwrap()],
+            Fr::from_str(challenges[8]).unwrap(),
+        );
+
+        let circuit_combiner = Fr::one();
+        let instance_combiners = vec![Fr::one()];
+
+        // Create sample circuit which corresponds to instance.input file.
+        let mul_depth = 3;
+        let num_constraints = 7;
+        let num_variables = 7;
+
+        // Create a fixed seed rng that matches those the test vectors were generated with.
+        let rng = &mut snarkvm_utilities::rand::TestRng::fixed(4730);
+        let max_degree =
+            AHPForR1CS::<Fr, MM>::max_degree(num_constraints, num_variables, num_variables * num_constraints).unwrap();
+        let universal_srs = VarunaSonicInst::universal_setup(max_degree).unwrap();
+        let (circ, _) =
+            TestCircuit::generate_circuit_with_fixed_witness(a, b, mul_depth, num_constraints, num_variables);
+        println!("Circuit: {:?}", circ);
+        let (index_pk, _index_vk) = VarunaSonicInst::circuit_setup(&universal_srs, &circ).unwrap();
+        let mut keys_to_constraints = BTreeMap::new();
+        keys_to_constraints.insert(index_pk.circuit.deref(), std::slice::from_ref(&circ));
+
+        // Begin the Varuna protocol execution.
+        let prover_state = AHPForR1CS::<_, MM>::init_prover(&keys_to_constraints, rng).unwrap();
+        let mut prover_state = AHPForR1CS::<_, MM>::prover_first_round(prover_state, rng).unwrap();
+        let first_round_oracles = Arc::new(prover_state.first_round_oracles.as_ref().unwrap());
+
+        // Get private witness polynomial coefficients.
+        let (_, w_poly) = first_round_oracles.batches.iter().next().unwrap();
+        let w_lde = format!("{:?}", w_poly[0].0.coeffs().map(|(_, coeff)| coeff).collect::<Vec<_>>());
+        if create_test_vectors {
+            create_test_vector("polynomials", "w_lde", &w_lde, circuit);
+        }
+
+        // Generate test vectors from assignments.
+        let assignments = AHPForR1CS::<_, MM>::calculate_assignments(&mut prover_state).unwrap();
+
+        // Get full witness polynomial coefficients.
+        let (_, z_poly) = assignments.iter().next().unwrap();
+        let z_lde = format!("{:?}", z_poly[0].coeffs().iter().collect::<Vec<_>>());
+        if create_test_vectors {
+            create_test_vector("polynomials", "z_lde", &z_lde, circuit);
+        }
+
+        let combiners = verifier::BatchCombiners::<Fr> { circuit_combiner, instance_combiners };
+        let batch_combiners = BTreeMap::from_iter([(index_pk.circuit.id, combiners)]);
+        let verifier_first_msg = verifier::FirstMessage::<Fr> { batch_combiners };
+
+        let (second_oracles, prover_state) =
+            AHPForR1CS::<_, MM>::prover_second_round::<_>(&verifier_first_msg, prover_state, rng).unwrap();
+
+        // Get round 2 rowcheck polynomial oracle coefficients.
+        let h_0 = format!("{:?}", second_oracles.h_0.coeffs().map(|(_, coeff)| coeff).collect::<Vec<_>>());
+        if create_test_vectors {
+            create_test_vector("polynomials", "h_0", &h_0, circuit);
+        }
+
+        let verifier_second_msg = verifier::SecondMessage::<Fr> { alpha, eta_b, eta_c };
+        let (_prover_third_message, third_oracles, prover_state) =
+            AHPForR1CS::<_, MM>::prover_third_round(&verifier_first_msg, &verifier_second_msg, prover_state, rng)
+                .unwrap();
+
+        // Get coefficients round 3 univariate rowcheck polynomial oracles.
+        let g_1 = format!("{:?}", third_oracles.g_1.coeffs().map(|(_, coeff)| coeff).collect::<Vec<_>>());
+        if create_test_vectors {
+            create_test_vector("polynomials", "g_1", &g_1, circuit);
+        }
+        let h_1 = format!("{:?}", third_oracles.h_1.coeffs().map(|(_, coeff)| coeff).collect::<Vec<_>>());
+        if create_test_vectors {
+            create_test_vector("polynomials", "h_1", &h_1, circuit);
+        }
+
+        let verifier_third_msg = verifier::ThirdMessage::<Fr> { beta };
+        let (_prover_fourth_message, fourth_oracles, prover_state) =
+            AHPForR1CS::<_, MM>::prover_fourth_round(&verifier_second_msg, &verifier_third_msg, prover_state, rng)
+                .unwrap();
+
+        // Create round 4 rational sumcheck oracle polynomials.
+        let (_, gm_polys) = fourth_oracles.gs.iter().next().unwrap();
+        let g_a = format!("{:?}", gm_polys.g_a.coeffs().map(|(_, coeff)| coeff).collect::<Vec<_>>());
+        let g_b = format!("{:?}", gm_polys.g_b.coeffs().map(|(_, coeff)| coeff).collect::<Vec<_>>());
+        let g_c = format!("{:?}", gm_polys.g_b.coeffs().map(|(_, coeff)| coeff).collect::<Vec<_>>());
+        if create_test_vectors {
+            create_test_vector("polynomials", "g_a", &g_a, circuit);
+            create_test_vector("polynomials", "g_b", &g_b, circuit);
+            create_test_vector("polynomials", "g_c", &g_c, circuit);
+        }
+
+        // Create the verifier's fourth message.
+        let verifier_fourth_msg = verifier::FourthMessage::<Fr> { delta_a, delta_b, delta_c };
+
+        let mut public_inputs = BTreeMap::new();
+        let public_input = prover_state.public_inputs(&index_pk.circuit).unwrap();
+        public_inputs.insert(index_pk.circuit.id, public_input);
+        let non_zero_a_domain = EvaluationDomain::<Fr>::new(index_pk.circuit.index_info.num_non_zero_a).unwrap();
+        let non_zero_b_domain = EvaluationDomain::<Fr>::new(index_pk.circuit.index_info.num_non_zero_b).unwrap();
+        let non_zero_c_domain = EvaluationDomain::<Fr>::new(index_pk.circuit.index_info.num_non_zero_c).unwrap();
+        let variable_domain = EvaluationDomain::<Fr>::new(index_pk.circuit.index_info.num_variables).unwrap();
+        let constraint_domain = EvaluationDomain::<Fr>::new(index_pk.circuit.index_info.num_constraints).unwrap();
+        let input_domain = EvaluationDomain::<Fr>::new(index_pk.circuit.index_info.num_public_inputs).unwrap();
+
+        // Get constraint domain elements.
+        let mut constraint_domain_elements = Vec::with_capacity(constraint_domain.size());
+        for el in constraint_domain.elements() {
+            constraint_domain_elements.push(el);
+        }
+        if create_test_vectors {
+            create_test_vector("domain", "R", &format!("{:?}", constraint_domain_elements), circuit);
+        }
+
+        // Get non_zero_domain elements.
+        let non_zero_domain = *[&non_zero_a_domain, &non_zero_b_domain, &non_zero_c_domain]
+            .iter()
+            .max_by_key(|domain| domain.size)
+            .unwrap();
+        let mut non_zero_domain_elements = Vec::with_capacity(non_zero_domain.size());
+        for el in non_zero_domain.elements() {
+            non_zero_domain_elements.push(el);
+        }
+        if create_test_vectors {
+            create_test_vector("domain", "K", &format!("{:?}", non_zero_domain_elements), circuit);
+        }
+
+        // Get variable domain elements.
+        let mut variable_domain_elements = Vec::with_capacity(input_domain.size());
+        for el in variable_domain.elements() {
+            variable_domain_elements.push(el);
+        }
+        if create_test_vectors {
+            create_test_vector("domain", "C", &format!("{:?}", variable_domain_elements), circuit);
+        }
+
+        let fifth_oracles =
+            AHPForR1CS::<_, MM>::prover_fifth_round(verifier_fourth_msg.clone(), prover_state, rng).unwrap();
+
+        // Get coefficients of final oracle polynomial from round 5.
+        let h_2 = format!("{:?}", fifth_oracles.h_2.coeffs().map(|(_, coeff)| coeff).collect::<Vec<_>>());
+        if create_test_vectors {
+            create_test_vector("polynomials", "h_2", &h_2, circuit);
+        }
+
+        // Check the intermediate oracle polynomials against the test vectors.
+        assert_test_vector_equality("polynomials", "w_lde", &w_lde, circuit);
+        assert_test_vector_equality("polynomials", "z_lde", &z_lde, circuit);
+        assert_test_vector_equality("polynomials", "h_0", &h_0, circuit);
+        assert_test_vector_equality("polynomials", "h_1", &h_1, circuit);
+        assert_test_vector_equality("polynomials", "g_1", &g_1, circuit);
+        assert_test_vector_equality("polynomials", "h_2", &h_2, circuit);
+        assert_test_vector_equality("polynomials", "g_a", &g_a, circuit);
+        assert_test_vector_equality("polynomials", "g_b", &g_b, circuit);
+        assert_test_vector_equality("polynomials", "g_c", &g_c, circuit);
+
+        // Check that the domains match the test vectors.
+        assert_test_vector_equality("domain", "R", &format!("{:?}", constraint_domain_elements), circuit);
+        assert_test_vector_equality("domain", "K", &format!("{:?}", non_zero_domain_elements), circuit);
+        assert_test_vector_equality("domain", "C", &format!("{:?}", variable_domain_elements), circuit);
+    }
+
+    #[test]
+    fn test_varuna_with_prover_test_vectors() {
+        test_varuna_with_all_circuits(false);
+    }
+}
diff --git a/console/program/src/request/serialize.rs b/console/program/src/request/serialize.rs
index 650279e49..905b67d99 100644
--- a/console/program/src/request/serialize.rs
+++ b/console/program/src/request/serialize.rs
@@ -21,7 +21,7 @@ impl<N: Network> Serialize for Request<N> {
     fn serialize<S: Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
         match serializer.is_human_readable() {
             true => {
-                let mut transition = serializer.serialize_struct("Request", 9)?;
+                let mut transition = serializer.serialize_struct("Request", 10)?;
                 transition.serialize_field("signer", &self.signer)?;
                 transition.serialize_field("network", &self.network_id)?;
                 transition.serialize_field("program", &self.program_id)?;
diff --git a/ledger/block/src/helpers/target.rs b/ledger/block/src/helpers/target.rs
index b39623425..ffb6265bd 100644
--- a/ledger/block/src/helpers/target.rs
+++ b/ledger/block/src/helpers/target.rs
@@ -27,7 +27,7 @@ pub const fn block_reward(total_supply: u64, block_time: u16, coinbase_reward: u
     // Compute the expected block height at year 1.
     let block_height_at_year_1 = block_height_at_year(block_time, 1);
     // Compute the annual reward: (0.05 * S).
-    let annual_reward = (total_supply / 1000) * 50;
+    let annual_reward = total_supply / 20;
     // Compute the block reward: (0.05 * S) / H_Y1.
     let block_reward = annual_reward / block_height_at_year_1 as u64;
     // Return the sum of the block reward, coinbase reward, and transaction fees.
diff --git a/ledger/block/src/serialize.rs b/ledger/block/src/serialize.rs
index b50a82c80..24edfa868 100644
--- a/ledger/block/src/serialize.rs
+++ b/ledger/block/src/serialize.rs
@@ -19,7 +19,7 @@ impl<N: Network> Serialize for Block<N> {
     fn serialize<S: Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
         match serializer.is_human_readable() {
             true => {
-                let mut block = serializer.serialize_struct("Block", 6)?;
+                let mut block = serializer.serialize_struct("Block", 7 + self.solutions.is_some() as usize)?;
                 block.serialize_field("block_hash", &self.block_hash)?;
                 block.serialize_field("previous_hash", &self.previous_hash)?;
                 block.serialize_field("header", &self.header)?;
diff --git a/ledger/block/src/transaction/execution/serialize.rs b/ledger/block/src/transaction/execution/serialize.rs
index 08724acbb..b432dd8e7 100644
--- a/ledger/block/src/transaction/execution/serialize.rs
+++ b/ledger/block/src/transaction/execution/serialize.rs
@@ -19,7 +19,7 @@ impl<N: Network> Serialize for Execution<N> {
     fn serialize<S: Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
         match serializer.is_human_readable() {
             true => {
-                let mut execution = serializer.serialize_struct("Execution", 3)?;
+                let mut execution = serializer.serialize_struct("Execution", 2 + self.proof.is_some() as usize)?;
                 execution
                     .serialize_field("transitions", &self.transitions.values().collect::<Vec<&Transition<N>>>())?;
                 execution.serialize_field("global_state_root", &self.global_state_root)?;
diff --git a/ledger/block/src/transaction/fee/serialize.rs b/ledger/block/src/transaction/fee/serialize.rs
index 97c1f08f6..b73ec5e03 100644
--- a/ledger/block/src/transaction/fee/serialize.rs
+++ b/ledger/block/src/transaction/fee/serialize.rs
@@ -19,7 +19,7 @@ impl<N: Network> Serialize for Fee<N> {
     fn serialize<S: Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
         match serializer.is_human_readable() {
             true => {
-                let mut fee = serializer.serialize_struct("Fee", 3)?;
+                let mut fee = serializer.serialize_struct("Fee", 2 + self.proof.is_some() as usize)?;
                 fee.serialize_field("transition", &self.transition)?;
                 fee.serialize_field("global_state_root", &self.global_state_root)?;
                 if let Some(proof) = &self.proof {
diff --git a/ledger/block/src/transaction/serialize.rs b/ledger/block/src/transaction/serialize.rs
index 9e9b7f941..ec6aae144 100644
--- a/ledger/block/src/transaction/serialize.rs
+++ b/ledger/block/src/transaction/serialize.rs
@@ -29,7 +29,7 @@ impl<N: Network> Serialize for Transaction<N> {
                     transaction.end()
                 }
                 Self::Execute(id, execution, fee) => {
-                    let mut transaction = serializer.serialize_struct("Transaction", 4)?;
+                    let mut transaction = serializer.serialize_struct("Transaction", 3 + fee.is_some() as usize)?;
                     transaction.serialize_field("type", "execute")?;
                     transaction.serialize_field("id", &id)?;
                     transaction.serialize_field("execution", &execution)?;
diff --git a/ledger/block/src/transition/input/serialize.rs b/ledger/block/src/transition/input/serialize.rs
index 281d47cff..3a959b7c4 100644
--- a/ledger/block/src/transition/input/serialize.rs
+++ b/ledger/block/src/transition/input/serialize.rs
@@ -20,7 +20,7 @@ impl<N: Network> Serialize for Input<N> {
         match serializer.is_human_readable() {
             true => match self {
                 Self::Constant(id, value) => {
-                    let mut input = serializer.serialize_struct("Input", 3)?;
+                    let mut input = serializer.serialize_struct("Input", 2 + value.is_some() as usize)?;
                     input.serialize_field("type", "constant")?;
                     input.serialize_field("id", &id)?;
                     if let Some(value) = value {
@@ -29,7 +29,7 @@ impl<N: Network> Serialize for Input<N> {
                     input.end()
                 }
                 Self::Public(id, value) => {
-                    let mut input = serializer.serialize_struct("Input", 3)?;
+                    let mut input = serializer.serialize_struct("Input", 2 + value.is_some() as usize)?;
                     input.serialize_field("type", "public")?;
                     input.serialize_field("id", &id)?;
                     if let Some(value) = value {
@@ -38,7 +38,7 @@ impl<N: Network> Serialize for Input<N> {
                     input.end()
                 }
                 Self::Private(id, value) => {
-                    let mut input = serializer.serialize_struct("Input", 3)?;
+                    let mut input = serializer.serialize_struct("Input", 2 + value.is_some() as usize)?;
                     input.serialize_field("type", "private")?;
                     input.serialize_field("id", &id)?;
                     if let Some(value) = value {
diff --git a/ledger/block/src/transition/output/serialize.rs b/ledger/block/src/transition/output/serialize.rs
index bdf49c4ba..47041735a 100644
--- a/ledger/block/src/transition/output/serialize.rs
+++ b/ledger/block/src/transition/output/serialize.rs
@@ -20,7 +20,7 @@ impl<N: Network> Serialize for Output<N> {
         match serializer.is_human_readable() {
             true => match self {
                 Self::Constant(id, value) => {
-                    let mut output = serializer.serialize_struct("Output", 3)?;
+                    let mut output = serializer.serialize_struct("Output", 2 + value.is_some() as usize)?;
                     output.serialize_field("type", "constant")?;
                     output.serialize_field("id", &id)?;
                     if let Some(value) = value {
@@ -29,7 +29,7 @@ impl<N: Network> Serialize for Output<N> {
                     output.end()
                 }
                 Self::Public(id, value) => {
-                    let mut output = serializer.serialize_struct("Output", 3)?;
+                    let mut output = serializer.serialize_struct("Output", 2 + value.is_some() as usize)?;
                     output.serialize_field("type", "public")?;
                     output.serialize_field("id", &id)?;
                     if let Some(value) = value {
@@ -38,7 +38,7 @@ impl<N: Network> Serialize for Output<N> {
                     output.end()
                 }
                 Self::Private(id, value) => {
-                    let mut output = serializer.serialize_struct("Output", 3)?;
+                    let mut output = serializer.serialize_struct("Output", 2 + value.is_some() as usize)?;
                     output.serialize_field("type", "private")?;
                     output.serialize_field("id", &id)?;
                     if let Some(value) = value {
@@ -47,7 +47,7 @@ impl<N: Network> Serialize for Output<N> {
                     output.end()
                 }
                 Self::Record(id, checksum, value) => {
-                    let mut output = serializer.serialize_struct("Output", 5)?;
+                    let mut output = serializer.serialize_struct("Output", 3 + value.is_some() as usize)?;
                     output.serialize_field("type", "record")?;
                     output.serialize_field("id", &id)?;
                     output.serialize_field("checksum", &checksum)?;
@@ -63,7 +63,7 @@ impl<N: Network> Serialize for Output<N> {
                     output.end()
                 }
                 Self::Future(id, value) => {
-                    let mut output = serializer.serialize_struct("Output", 3)?;
+                    let mut output = serializer.serialize_struct("Output", 2 + value.is_some() as usize)?;
                     output.serialize_field("type", "future")?;
                     output.serialize_field("id", &id)?;
                     if let Some(value) = value {
diff --git a/ledger/block/src/verify.rs b/ledger/block/src/verify.rs
index 057ff2b56..7e2afa985 100644
--- a/ledger/block/src/verify.rs
+++ b/ledger/block/src/verify.rs
@@ -137,16 +137,11 @@ impl<N: Network> Block<N> {
         previous_height: u32,
         current_committee: &Committee<N>,
     ) -> Result<(u64, u32, i64)> {
+        #[cfg(not(any(test, feature = "test")))]
+        ensure!(self.authority.is_quorum(), "The next block must be a quorum block");
+
         // Determine the expected height.
         let expected_height = previous_height.saturating_add(1);
-        // Ensure the block type is correct.
-        match expected_height == 0 {
-            true => ensure!(self.authority.is_beacon(), "The genesis block must be a beacon block"),
-            false => {
-                #[cfg(not(any(test, feature = "test")))]
-                ensure!(self.authority.is_quorum(), "The next block must be a quorum block");
-            }
-        }
 
         // Determine the expected round.
         let expected_round = match &self.authority {
diff --git a/ledger/coinbase/src/helpers/prover_solution/serialize.rs b/ledger/coinbase/src/helpers/prover_solution/serialize.rs
index 32ce1b22d..564cba8be 100644
--- a/ledger/coinbase/src/helpers/prover_solution/serialize.rs
+++ b/ledger/coinbase/src/helpers/prover_solution/serialize.rs
@@ -21,7 +21,8 @@ impl<N: Network> Serialize for ProverSolution<N> {
     fn serialize<S: Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
         match serializer.is_human_readable() {
             true => {
-                let mut prover_solution = serializer.serialize_struct("ProverSolution", 3)?;
+                let mut prover_solution =
+                    serializer.serialize_struct("ProverSolution", 2 + self.proof.random_v.is_some() as usize)?;
                 prover_solution.serialize_field("partial_solution", &self.partial_solution)?;
                 prover_solution.serialize_field("proof.w", &self.proof.w)?;
                 if let Some(random_v) = &self.proof.random_v {
diff --git a/ledger/narwhal/transmission/src/serialize.rs b/ledger/narwhal/transmission/src/serialize.rs
index 2acd03750..4657a7ff7 100644
--- a/ledger/narwhal/transmission/src/serialize.rs
+++ b/ledger/narwhal/transmission/src/serialize.rs
@@ -18,23 +18,25 @@ impl<N: Network> Serialize for Transmission<N> {
     #[inline]
     fn serialize<S: Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
         match serializer.is_human_readable() {
-            true => {
-                let mut transmission = serializer.serialize_struct("Transmission", 2)?;
-                match self {
-                    Self::Ratification => {
-                        transmission.serialize_field("type", "ratification")?;
-                    }
-                    Self::Solution(solution) => {
-                        transmission.serialize_field("type", "solution")?;
-                        transmission.serialize_field("transmission", solution)?;
-                    }
-                    Self::Transaction(transaction) => {
-                        transmission.serialize_field("type", "transaction")?;
-                        transmission.serialize_field("transmission", transaction)?;
-                    }
+            true => match self {
+                Self::Ratification => {
+                    let mut transmission = serializer.serialize_struct("Transmission", 1)?;
+                    transmission.serialize_field("type", "ratification")?;
+                    transmission.end()
                 }
-                transmission.end()
-            }
+                Self::Solution(solution) => {
+                    let mut transmission = serializer.serialize_struct("Transmission", 2)?;
+                    transmission.serialize_field("type", "solution")?;
+                    transmission.serialize_field("transmission", solution)?;
+                    transmission.end()
+                }
+                Self::Transaction(transaction) => {
+                    let mut transmission = serializer.serialize_struct("Transmission", 2)?;
+                    transmission.serialize_field("type", "transaction")?;
+                    transmission.serialize_field("transmission", transaction)?;
+                    transmission.end()
+                }
+            },
             false => ToBytesSerializer::serialize_with_size_encoding(self, serializer),
         }
     }
diff --git a/ledger/src/lib.rs b/ledger/src/lib.rs
index 0e5511b79..0259e567b 100644
--- a/ledger/src/lib.rs
+++ b/ledger/src/lib.rs
@@ -145,8 +145,9 @@ impl<N: Network, C: ConsensusStorage<N>> Ledger<N, C> {
         let timer = timer!("Ledger::load_unchecked");
 
         // Initialize the consensus store.
-        let Ok(store) = ConsensusStore::<N, C>::open(dev) else {
-            bail!("Failed to load ledger (run 'snarkos clean' and try again)");
+        let store = match ConsensusStore::<N, C>::open(dev) {
+            Ok(store) => store,
+            Err(e) => bail!("Failed to load ledger (run 'snarkos clean' and try again)\n\n{e}\n"),
         };
         lap!(timer, "Load consensus store");
 
diff --git a/synthesizer/process/src/stack/register_types/initialize.rs b/synthesizer/process/src/stack/register_types/initialize.rs
index 173e71d31..ee76d5f31 100644
--- a/synthesizer/process/src/stack/register_types/initialize.rs
+++ b/synthesizer/process/src/stack/register_types/initialize.rs
@@ -305,6 +305,7 @@ impl<N: Network> RegisterTypes<N> {
         operand: &Operand<N>,
         register_type: &RegisterType<N>,
     ) -> Result<()> {
+        #[cfg(feature = "aleo-cli")]
         match operand {
             // Inform the user the output operand is an input register, to ensure this is intended behavior.
             Operand::Register(register) if self.is_input(register) => {
diff --git a/synthesizer/process/src/stack/register_types/matches.rs b/synthesizer/process/src/stack/register_types/matches.rs
index 2082fa995..395772684 100644
--- a/synthesizer/process/src/stack/register_types/matches.rs
+++ b/synthesizer/process/src/stack/register_types/matches.rs
@@ -168,8 +168,6 @@ impl<N: Network> RegisterTypes<N> {
     }
 
     /// Checks that the given record matches the layout of the record type.
-    /// Note: Ordering for `owner` **does** matter, however ordering
-    /// for record data does **not** matter, as long as all defined members are present.
     pub fn matches_record(
         &self,
         stack: &(impl StackMatches<N> + StackProgram<N>),
diff --git a/synthesizer/program/src/logic/finalize_operation/serialize.rs b/synthesizer/program/src/logic/finalize_operation/serialize.rs
index f9f666e5d..1a128d3e9 100644
--- a/synthesizer/program/src/logic/finalize_operation/serialize.rs
+++ b/synthesizer/program/src/logic/finalize_operation/serialize.rs
@@ -19,41 +19,51 @@ impl<N: Network> Serialize for FinalizeOperation<N> {
     fn serialize<S: Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
         match serializer.is_human_readable() {
             true => {
-                let mut operation = serializer.serialize_struct("FinalizeOperation", 5)?;
                 // Serialize the components.
                 match self {
                     Self::InitializeMapping(mapping_id) => {
+                        let mut operation = serializer.serialize_struct("FinalizeOperation", 2)?;
                         operation.serialize_field("type", "initialize_mapping")?;
                         operation.serialize_field("mapping_id", mapping_id)?;
+                        operation.end()
                     }
                     Self::InsertKeyValue(mapping_id, key_id, value_id) => {
+                        let mut operation = serializer.serialize_struct("FinalizeOperation", 4)?;
                         operation.serialize_field("type", "insert_key_value")?;
                         operation.serialize_field("mapping_id", mapping_id)?;
                         operation.serialize_field("key_id", key_id)?;
                         operation.serialize_field("value_id", value_id)?;
+                        operation.end()
                     }
                     Self::UpdateKeyValue(mapping_id, index, key_id, value_id) => {
+                        let mut operation = serializer.serialize_struct("FinalizeOperation", 5)?;
                         operation.serialize_field("type", "update_key_value")?;
                         operation.serialize_field("mapping_id", mapping_id)?;
                         operation.serialize_field("index", index)?;
                         operation.serialize_field("key_id", key_id)?;
                         operation.serialize_field("value_id", value_id)?;
+                        operation.end()
                     }
                     Self::RemoveKeyValue(mapping_id, index) => {
+                        let mut operation = serializer.serialize_struct("FinalizeOperation", 3)?;
                         operation.serialize_field("type", "remove_key_value")?;
                         operation.serialize_field("mapping_id", mapping_id)?;
                         operation.serialize_field("index", index)?;
+                        operation.end()
                     }
                     Self::ReplaceMapping(mapping_id) => {
+                        let mut operation = serializer.serialize_struct("FinalizeOperation", 2)?;
                         operation.serialize_field("type", "replace_mapping")?;
                         operation.serialize_field("mapping_id", mapping_id)?;
+                        operation.end()
                     }
                     Self::RemoveMapping(mapping_id) => {
+                        let mut operation = serializer.serialize_struct("FinalizeOperation", 2)?;
                         operation.serialize_field("type", "remove_mapping")?;
                         operation.serialize_field("mapping_id", mapping_id)?;
+                        operation.end()
                     }
                 }
-                operation.end()
             }
             false => ToBytesSerializer::serialize_with_size_encoding(self, serializer),
         }
diff --git a/synthesizer/src/vm/verify.rs b/synthesizer/src/vm/verify.rs
index 99c921dec..f6aeaceaa 100644
--- a/synthesizer/src/vm/verify.rs
+++ b/synthesizer/src/vm/verify.rs
@@ -255,7 +255,7 @@ impl<N: Network, C: ConsensusStorage<N>> VM<N, C> {
 
         // Ensure the fee does not exceed the limit.
         let fee_amount = fee.amount()?;
-        ensure!(*fee_amount < N::MAX_FEE, "Fee verification failed: fee exceeds the maximum limit");
+        ensure!(*fee_amount <= N::MAX_FEE, "Fee verification failed: fee exceeds the maximum limit");
 
         // Verify the fee.
         let verification = self.process.read().verify_fee(fee, deployment_or_execution_id);
