diff --git a/.cargo/release-version b/.cargo/release-version
index ab664c2b5..ea535dc59 100644
--- a/.cargo/release-version
+++ b/.cargo/release-version
@@ -1 +1 @@
-v3.7.1
\ No newline at end of file
+v3.8.0
\ No newline at end of file
diff --git a/.circleci/config.yml b/.circleci/config.yml
index e75e69292..9999a47f7 100644
--- a/.circleci/config.yml
+++ b/.circleci/config.yml
@@ -721,7 +721,7 @@ jobs:
     resource_class: << pipeline.parameters.twoxlarge >>
     steps:
       - run_serial:
-          flags: --test '*' --features test -- --test-threads=8
+          flags: --test '*' --features test -- --test-threads=4
           workspace_member: synthesizer
           cache_key: v1.3.0-rust-1.83.0-snarkvm-synthesizer-integration-cache
 
diff --git a/Cargo.lock b/Cargo.lock
index 9a1c5e889..642eac837 100644
--- a/Cargo.lock
+++ b/Cargo.lock
@@ -2673,7 +2673,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm"
-version = "3.7.1"
+version = "3.8.0"
 dependencies = [
  "anstyle",
  "anyhow",
@@ -2705,7 +2705,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-algorithms"
-version = "3.7.1"
+version = "3.8.0"
 dependencies = [
  "aleo-std",
  "anyhow",
@@ -2738,7 +2738,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-algorithms-cuda"
-version = "3.7.1"
+version = "3.8.0"
 dependencies = [
  "blst",
  "cc",
@@ -2748,7 +2748,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-circuit"
-version = "3.7.1"
+version = "3.8.0"
 dependencies = [
  "snarkvm-circuit-account",
  "snarkvm-circuit-algorithms",
@@ -2761,7 +2761,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-circuit-account"
-version = "3.7.1"
+version = "3.8.0"
 dependencies = [
  "anyhow",
  "snarkvm-circuit-network",
@@ -2772,7 +2772,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-circuit-algorithms"
-version = "3.7.1"
+version = "3.8.0"
 dependencies = [
  "anyhow",
  "snarkvm-circuit-types",
@@ -2784,7 +2784,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-circuit-collections"
-version = "3.7.1"
+version = "3.8.0"
 dependencies = [
  "anyhow",
  "snarkvm-circuit-algorithms",
@@ -2798,7 +2798,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-circuit-environment"
-version = "3.7.1"
+version = "3.8.0"
 dependencies = [
  "criterion",
  "indexmap 2.9.0",
@@ -2821,11 +2821,11 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-circuit-environment-witness"
-version = "3.7.1"
+version = "3.8.0"
 
 [[package]]
 name = "snarkvm-circuit-network"
-version = "3.7.1"
+version = "3.8.0"
 dependencies = [
  "snarkvm-circuit-algorithms",
  "snarkvm-circuit-collections",
@@ -2836,7 +2836,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-circuit-program"
-version = "3.7.1"
+version = "3.8.0"
 dependencies = [
  "anyhow",
  "rand",
@@ -2853,7 +2853,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-circuit-types"
-version = "3.7.1"
+version = "3.8.0"
 dependencies = [
  "snarkvm-circuit-environment",
  "snarkvm-circuit-types-address",
@@ -2868,7 +2868,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-circuit-types-address"
-version = "3.7.1"
+version = "3.8.0"
 dependencies = [
  "snarkvm-circuit-environment",
  "snarkvm-circuit-types-boolean",
@@ -2880,7 +2880,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-circuit-types-boolean"
-version = "3.7.1"
+version = "3.8.0"
 dependencies = [
  "criterion",
  "snarkvm-circuit-environment",
@@ -2889,7 +2889,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-circuit-types-field"
-version = "3.7.1"
+version = "3.8.0"
 dependencies = [
  "snarkvm-circuit-environment",
  "snarkvm-circuit-types-boolean",
@@ -2898,7 +2898,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-circuit-types-group"
-version = "3.7.1"
+version = "3.8.0"
 dependencies = [
  "snarkvm-circuit-environment",
  "snarkvm-circuit-types-boolean",
@@ -2910,7 +2910,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-circuit-types-integers"
-version = "3.7.1"
+version = "3.8.0"
 dependencies = [
  "paste",
  "snarkvm-circuit-environment",
@@ -2923,7 +2923,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-circuit-types-scalar"
-version = "3.7.1"
+version = "3.8.0"
 dependencies = [
  "snarkvm-circuit-environment",
  "snarkvm-circuit-types-boolean",
@@ -2933,7 +2933,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-circuit-types-string"
-version = "3.7.1"
+version = "3.8.0"
 dependencies = [
  "rand",
  "snarkvm-circuit-environment",
@@ -2946,7 +2946,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-console"
-version = "3.7.1"
+version = "3.8.0"
 dependencies = [
  "snarkvm-console-account",
  "snarkvm-console-algorithms",
@@ -2958,7 +2958,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-console-account"
-version = "3.7.1"
+version = "3.8.0"
 dependencies = [
  "bincode",
  "bs58",
@@ -2971,7 +2971,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-console-algorithms"
-version = "3.7.1"
+version = "3.8.0"
 dependencies = [
  "blake2s_simd",
  "criterion",
@@ -2989,7 +2989,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-console-collections"
-version = "3.7.1"
+version = "3.8.0"
 dependencies = [
  "aleo-std",
  "criterion",
@@ -3002,7 +3002,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-console-network"
-version = "3.7.1"
+version = "3.8.0"
 dependencies = [
  "anyhow",
  "indexmap 2.9.0",
@@ -3021,7 +3021,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-console-network-environment"
-version = "3.7.1"
+version = "3.8.0"
 dependencies = [
  "anyhow",
  "bech32",
@@ -3038,7 +3038,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-console-program"
-version = "3.7.1"
+version = "3.8.0"
 dependencies = [
  "bincode",
  "enum-iterator",
@@ -3059,7 +3059,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-console-types"
-version = "3.7.1"
+version = "3.8.0"
 dependencies = [
  "criterion",
  "snarkvm-console-network",
@@ -3075,7 +3075,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-console-types-address"
-version = "3.7.1"
+version = "3.8.0"
 dependencies = [
  "bincode",
  "serde_json",
@@ -3087,7 +3087,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-console-types-boolean"
-version = "3.7.1"
+version = "3.8.0"
 dependencies = [
  "bincode",
  "serde_json",
@@ -3096,7 +3096,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-console-types-field"
-version = "3.7.1"
+version = "3.8.0"
 dependencies = [
  "bincode",
  "serde_json",
@@ -3107,7 +3107,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-console-types-group"
-version = "3.7.1"
+version = "3.8.0"
 dependencies = [
  "bincode",
  "serde_json",
@@ -3119,7 +3119,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-console-types-integers"
-version = "3.7.1"
+version = "3.8.0"
 dependencies = [
  "bincode",
  "serde_json",
@@ -3131,7 +3131,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-console-types-scalar"
-version = "3.7.1"
+version = "3.8.0"
 dependencies = [
  "bincode",
  "serde_json",
@@ -3143,7 +3143,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-console-types-string"
-version = "3.7.1"
+version = "3.8.0"
 dependencies = [
  "bincode",
  "serde_json",
@@ -3155,7 +3155,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-curves"
-version = "3.7.1"
+version = "3.8.0"
 dependencies = [
  "bincode",
  "criterion",
@@ -3170,7 +3170,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-fields"
-version = "3.7.1"
+version = "3.8.0"
 dependencies = [
  "aleo-std",
  "anyhow",
@@ -3188,7 +3188,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-ledger"
-version = "3.7.1"
+version = "3.8.0"
 dependencies = [
  "aleo-std",
  "anyhow",
@@ -3219,7 +3219,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-ledger-authority"
-version = "3.7.1"
+version = "3.8.0"
 dependencies = [
  "anyhow",
  "bincode",
@@ -3232,7 +3232,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-ledger-block"
-version = "3.7.1"
+version = "3.8.0"
 dependencies = [
  "aleo-std",
  "bincode",
@@ -3259,7 +3259,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-ledger-committee"
-version = "3.7.1"
+version = "3.8.0"
 dependencies = [
  "anyhow",
  "bincode",
@@ -3280,7 +3280,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-ledger-narwhal"
-version = "3.7.1"
+version = "3.8.0"
 dependencies = [
  "snarkvm-ledger-narwhal",
  "snarkvm-ledger-narwhal-batch-certificate",
@@ -3293,7 +3293,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-ledger-narwhal-batch-certificate"
-version = "3.7.1"
+version = "3.8.0"
 dependencies = [
  "bincode",
  "indexmap 2.9.0",
@@ -3307,7 +3307,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-ledger-narwhal-batch-header"
-version = "3.7.1"
+version = "3.8.0"
 dependencies = [
  "bincode",
  "indexmap 2.9.0",
@@ -3321,7 +3321,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-ledger-narwhal-data"
-version = "3.7.1"
+version = "3.8.0"
 dependencies = [
  "bytes",
  "serde_json",
@@ -3333,7 +3333,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-ledger-narwhal-subdag"
-version = "3.7.1"
+version = "3.8.0"
 dependencies = [
  "bincode",
  "indexmap 2.9.0",
@@ -3349,7 +3349,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-ledger-narwhal-transmission"
-version = "3.7.1"
+version = "3.8.0"
 dependencies = [
  "bincode",
  "bytes",
@@ -3362,7 +3362,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-ledger-narwhal-transmission-id"
-version = "3.7.1"
+version = "3.8.0"
 dependencies = [
  "bincode",
  "serde_json",
@@ -3372,7 +3372,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-ledger-puzzle"
-version = "3.7.1"
+version = "3.8.0"
 dependencies = [
  "aleo-std",
  "anyhow",
@@ -3393,7 +3393,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-ledger-puzzle-epoch"
-version = "3.7.1"
+version = "3.8.0"
 dependencies = [
  "aleo-std",
  "anyhow",
@@ -3414,7 +3414,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-ledger-query"
-version = "3.7.1"
+version = "3.8.0"
 dependencies = [
  "async-trait",
  "http",
@@ -3427,7 +3427,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-ledger-store"
-version = "3.7.1"
+version = "3.8.0"
 dependencies = [
  "aleo-std",
  "aleo-std-storage",
@@ -3458,7 +3458,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-ledger-test-helpers"
-version = "3.7.1"
+version = "3.8.0"
 dependencies = [
  "aleo-std",
  "once_cell",
@@ -3474,14 +3474,14 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-metrics"
-version = "3.7.1"
+version = "3.8.0"
 dependencies = [
  "metrics",
 ]
 
 [[package]]
 name = "snarkvm-parameters"
-version = "3.7.1"
+version = "3.8.0"
 dependencies = [
  "aleo-std",
  "anyhow",
@@ -3513,7 +3513,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-synthesizer"
-version = "3.7.1"
+version = "3.8.0"
 dependencies = [
  "aleo-std",
  "anyhow",
@@ -3553,7 +3553,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-synthesizer-process"
-version = "3.7.1"
+version = "3.8.0"
 dependencies = [
  "aleo-std",
  "bincode",
@@ -3583,7 +3583,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-synthesizer-program"
-version = "3.7.1"
+version = "3.8.0"
 dependencies = [
  "bincode",
  "criterion",
@@ -3600,7 +3600,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-synthesizer-snark"
-version = "3.7.1"
+version = "3.8.0"
 dependencies = [
  "bincode",
  "colored",
@@ -3613,7 +3613,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-utilities"
-version = "3.7.1"
+version = "3.8.0"
 dependencies = [
  "aleo-std",
  "anyhow",
@@ -3633,7 +3633,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-utilities-derives"
-version = "3.7.1"
+version = "3.8.0"
 dependencies = [
  "proc-macro2",
  "quote 1.0.40",
@@ -3642,7 +3642,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-wasm"
-version = "3.7.1"
+version = "3.8.0"
 dependencies = [
  "getrandom 0.2.16",
  "snarkvm-console",
diff --git a/Cargo.toml b/Cargo.toml
index 1883392a9..a9f4769a4 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm"
-version = "3.7.1"
+version = "3.8.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "A decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -163,58 +163,58 @@ test_consensus_heights = [ "snarkvm-console/test_consensus_heights" ]
 
 [dependencies.snarkvm-algorithms]
 path = "./algorithms"
-version = "=3.7.1"
+version = "=3.8.0"
 optional = true
 
 [dependencies.snarkvm-circuit]
 path = "./circuit"
-version = "=3.7.1"
+version = "=3.8.0"
 optional = true
 
 [dependencies.snarkvm-console]
 path = "./console"
-version = "=3.7.1"
+version = "=3.8.0"
 optional = true
 
 [dependencies.snarkvm-curves]
 path = "./curves"
-version = "=3.7.1"
+version = "=3.8.0"
 optional = true
 
 [dependencies.snarkvm-fields]
 path = "./fields"
-version = "=3.7.1"
+version = "=3.8.0"
 optional = true
 
 [dependencies.snarkvm-ledger]
 path = "./ledger"
-version = "=3.7.1"
+version = "=3.8.0"
 optional = true
 
 [dependencies.snarkvm-metrics]
 path = "./metrics"
-version = "=3.7.1"
+version = "=3.8.0"
 optional = true
 
 [dependencies.snarkvm-parameters]
 path = "./parameters"
-version = "=3.7.1"
+version = "=3.8.0"
 optional = true
 
 [dependencies.snarkvm-synthesizer]
 path = "./synthesizer"
-version = "=3.7.1"
+version = "=3.8.0"
 default-features = false
 optional = true
 
 [dependencies.snarkvm-utilities]
 path = "./utilities"
-version = "=3.7.1"
+version = "=3.8.0"
 optional = true
 
 [dependencies.snarkvm-wasm]
 path = "./wasm"
-version = "=3.7.1"
+version = "=3.8.0"
 optional = true
 
 [dependencies.anstyle]
diff --git a/algorithms/Cargo.toml b/algorithms/Cargo.toml
index aca60693e..91865cac7 100644
--- a/algorithms/Cargo.toml
+++ b/algorithms/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-algorithms"
-version = "3.7.1"
+version = "3.8.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Algorithms for a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -47,27 +47,27 @@ required-features = [ "test" ]
 
 [dependencies.snarkvm-curves]
 path = "../curves"
-version = "=3.7.1"
+version = "=3.8.0"
 default-features = false
 
 [dependencies.snarkvm-fields]
 path = "../fields"
-version = "=3.7.1"
+version = "=3.8.0"
 default-features = false
 
 [dependencies.snarkvm-parameters]
 path = "../parameters"
-version = "=3.7.1"
+version = "=3.8.0"
 optional = true
 
 [dependencies.snarkvm-utilities]
 path = "../utilities"
-version = "=3.7.1"
+version = "=3.8.0"
 default-features = false
 
 [dependencies.snarkvm-algorithms-cuda]
 path = "./cuda"
-version = "=3.7.1"
+version = "=3.8.0"
 optional = true
 
 [dependencies.aleo-std]
diff --git a/algorithms/cuda/Cargo.toml b/algorithms/cuda/Cargo.toml
index c6e5100e1..23491f28d 100644
--- a/algorithms/cuda/Cargo.toml
+++ b/algorithms/cuda/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-algorithms-cuda"
-version = "3.7.1"
+version = "3.8.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Cuda optimizations for a decentralized virtual machine"
 homepage = "https://aleo.org"
diff --git a/algorithms/src/snark/varuna/varuna.rs b/algorithms/src/snark/varuna/varuna.rs
index eb98a5717..5029a8c14 100644
--- a/algorithms/src/snark/varuna/varuna.rs
+++ b/algorithms/src/snark/varuna/varuna.rs
@@ -304,6 +304,11 @@ where
             bail!(SNARKError::CircuitNotFound);
         }
 
+        // Make sure certificate is not hiding
+        if certificate.pc_proof.is_hiding() {
+            bail!("Certificate should not be hiding");
+        }
+
         // Initialize sponge.
         let mut sponge = Self::init_sponge_for_certificate(fs_parameters, verifying_key)?;
 
diff --git a/circuit/Cargo.toml b/circuit/Cargo.toml
index d20f77379..02a8202fb 100644
--- a/circuit/Cargo.toml
+++ b/circuit/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-circuit"
-version = "3.7.1"
+version = "3.8.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Circuits for a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -28,28 +28,28 @@ save_r1cs_hashes = [ "snarkvm-circuit-environment/save_r1cs_hashes" ]
 
 [dependencies.snarkvm-circuit-account]
 path = "./account"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.snarkvm-circuit-algorithms]
 path = "./algorithms"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.snarkvm-circuit-collections]
 path = "./collections"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.snarkvm-circuit-environment]
 path = "./environment"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.snarkvm-circuit-network]
 path = "./network"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.snarkvm-circuit-program]
 path = "./program"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.snarkvm-circuit-types]
 path = "./types"
-version = "=3.7.1"
+version = "=3.8.0"
diff --git a/circuit/account/Cargo.toml b/circuit/account/Cargo.toml
index 394dbefa0..975672b83 100644
--- a/circuit/account/Cargo.toml
+++ b/circuit/account/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-circuit-account"
-version = "3.7.1"
+version = "3.8.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Account circuit library for a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -11,16 +11,16 @@ edition = "2021"
 [dependencies.console]
 package = "snarkvm-console-account"
 path = "../../console/account"
-version = "=3.7.1"
+version = "=3.8.0"
 optional = true
 
 [dependencies.snarkvm-circuit-network]
 path = "../network"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.snarkvm-circuit-types]
 path = "../types"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dev-dependencies.snarkvm-utilities]
 path = "../../utilities"
diff --git a/circuit/algorithms/Cargo.toml b/circuit/algorithms/Cargo.toml
index 2c42e78a7..231f5d714 100644
--- a/circuit/algorithms/Cargo.toml
+++ b/circuit/algorithms/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-circuit-algorithms"
-version = "3.7.1"
+version = "3.8.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Algorithm circuit library for a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -11,16 +11,16 @@ edition = "2021"
 [dependencies.console]
 package = "snarkvm-console-algorithms"
 path = "../../console/algorithms"
-version = "=3.7.1"
+version = "=3.8.0"
 optional = true
 
 [dependencies.snarkvm-circuit-types]
 path = "../types"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.snarkvm-fields]
 path = "../../fields"
-version = "=3.7.1"
+version = "=3.8.0"
 default-features = false
 
 [dev-dependencies.anyhow]
diff --git a/circuit/collections/Cargo.toml b/circuit/collections/Cargo.toml
index eda34ec1a..50fbbcba7 100644
--- a/circuit/collections/Cargo.toml
+++ b/circuit/collections/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-circuit-collections"
-version = "3.7.1"
+version = "3.8.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Collections circuit library for a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -11,16 +11,16 @@ edition = "2021"
 [dependencies.console]
 package = "snarkvm-console-collections"
 path = "../../console/collections"
-version = "=3.7.1"
+version = "=3.8.0"
 optional = true
 
 [dependencies.snarkvm-circuit-algorithms]
 path = "../algorithms"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.snarkvm-circuit-types]
 path = "../types"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dev-dependencies.snarkvm-circuit-network]
 path = "../network"
diff --git a/circuit/environment/Cargo.toml b/circuit/environment/Cargo.toml
index cbcd255df..637806072 100644
--- a/circuit/environment/Cargo.toml
+++ b/circuit/environment/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-circuit-environment"
-version = "3.7.1"
+version = "3.8.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Circuit environment for a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -16,32 +16,32 @@ harness = false
 [dependencies.console]
 package = "snarkvm-console-network"
 path = "../../console/network"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.snarkvm-algorithms]
 path = "../../algorithms"
-version = "=3.7.1"
+version = "=3.8.0"
 default-features = false
 features = [ "r1cs" ]
 
 [dependencies.snarkvm-circuit-environment-witness]
 path = "./witness"
-version = "=3.7.1"
+version = "=3.8.0"
 default-features = false
 
 [dependencies.snarkvm-curves]
 path = "../../curves"
-version = "=3.7.1"
+version = "=3.8.0"
 default-features = false
 
 [dependencies.snarkvm-fields]
 path = "../../fields"
-version = "=3.7.1"
+version = "=3.8.0"
 default-features = false
 
 [dependencies.snarkvm-utilities]
 path = "../../utilities"
-version = "=3.7.1"
+version = "=3.8.0"
 default-features = false
 
 [dependencies.indexmap]
diff --git a/circuit/environment/witness/Cargo.toml b/circuit/environment/witness/Cargo.toml
index 7051c9e1f..2a1f305db 100644
--- a/circuit/environment/witness/Cargo.toml
+++ b/circuit/environment/witness/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-circuit-environment-witness"
-version = "3.7.1"
+version = "3.8.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "A procedural macro to construct a witness in an environment"
 homepage = "https://aleo.org"
diff --git a/circuit/network/Cargo.toml b/circuit/network/Cargo.toml
index 3a14f4e1a..5f769fbc5 100644
--- a/circuit/network/Cargo.toml
+++ b/circuit/network/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-circuit-network"
-version = "3.7.1"
+version = "3.8.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Network circuit library for a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -11,20 +11,20 @@ edition = "2021"
 [dependencies.console]
 package = "snarkvm-console-network"
 path = "../../console/network"
-version = "=3.7.1"
+version = "=3.8.0"
 optional = true
 
 [dependencies.snarkvm-circuit-algorithms]
 path = "../algorithms"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.snarkvm-circuit-collections]
 path = "../collections"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.snarkvm-circuit-types]
 path = "../types"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dev-dependencies.snarkvm-console-types]
 path = "../../console/types"
diff --git a/circuit/program/Cargo.toml b/circuit/program/Cargo.toml
index 2064cdb57..48045ba7d 100644
--- a/circuit/program/Cargo.toml
+++ b/circuit/program/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-circuit-program"
-version = "3.7.1"
+version = "3.8.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Program circuit library for a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -11,32 +11,32 @@ edition = "2021"
 [dependencies.console]
 package = "snarkvm-console-program"
 path = "../../console/program"
-version = "=3.7.1"
+version = "=3.8.0"
 optional = true
 
 [dependencies.snarkvm-circuit-account]
 path = "../account"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.snarkvm-circuit-algorithms]
 path = "../algorithms"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.snarkvm-circuit-collections]
 path = "../collections"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.snarkvm-circuit-network]
 path = "../network"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.snarkvm-circuit-types]
 path = "../types"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.snarkvm-utilities]
 path = "../../utilities"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dev-dependencies.console_root]
 package = "snarkvm-console"
diff --git a/circuit/types/Cargo.toml b/circuit/types/Cargo.toml
index 6a4b98c11..4e8e041ff 100644
--- a/circuit/types/Cargo.toml
+++ b/circuit/types/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-circuit-types"
-version = "3.7.1"
+version = "3.8.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Primitive circuit for a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -10,35 +10,35 @@ edition = "2021"
 
 [dependencies.snarkvm-circuit-environment]
 path = "../environment"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.snarkvm-circuit-types-address]
 path = "./address"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.snarkvm-circuit-types-boolean]
 path = "./boolean"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.snarkvm-circuit-types-field]
 path = "./field"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.snarkvm-circuit-types-group]
 path = "./group"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.snarkvm-circuit-types-integers]
 path = "./integers"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.snarkvm-circuit-types-scalar]
 path = "./scalar"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.snarkvm-circuit-types-string]
 path = "./string"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dev-dependencies.console]
 package = "snarkvm-console"
diff --git a/circuit/types/address/Cargo.toml b/circuit/types/address/Cargo.toml
index f0e0ff8d5..0fffb55d0 100644
--- a/circuit/types/address/Cargo.toml
+++ b/circuit/types/address/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-circuit-types-address"
-version = "3.7.1"
+version = "3.8.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Address circuit for a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -11,28 +11,28 @@ edition = "2021"
 [dependencies.console]
 package = "snarkvm-console-types-address"
 path = "../../../console/types/address"
-version = "=3.7.1"
+version = "=3.8.0"
 optional = true
 
 [dependencies.snarkvm-circuit-environment]
 path = "../../environment"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.snarkvm-circuit-types-boolean]
 path = "../boolean"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.snarkvm-circuit-types-field]
 path = "../field"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.snarkvm-circuit-types-group]
 path = "../group"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.snarkvm-circuit-types-scalar]
 path = "../scalar"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [features]
 default = [ "enable_console" ]
diff --git a/circuit/types/boolean/Cargo.toml b/circuit/types/boolean/Cargo.toml
index 1aac5cf6f..931211d7e 100644
--- a/circuit/types/boolean/Cargo.toml
+++ b/circuit/types/boolean/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-circuit-types-boolean"
-version = "3.7.1"
+version = "3.8.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Boolean circuit for a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -16,12 +16,12 @@ harness = false
 [dependencies.console]
 package = "snarkvm-console-types-boolean"
 path = "../../../console/types/boolean"
-version = "=3.7.1"
+version = "=3.8.0"
 optional = true
 
 [dependencies.snarkvm-circuit-environment]
 path = "../../environment"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dev-dependencies.criterion]
 version = "0.5"
diff --git a/circuit/types/field/Cargo.toml b/circuit/types/field/Cargo.toml
index b321702c5..224e91ab6 100644
--- a/circuit/types/field/Cargo.toml
+++ b/circuit/types/field/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-circuit-types-field"
-version = "3.7.1"
+version = "3.8.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Field circuit for a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -11,16 +11,16 @@ edition = "2021"
 [dependencies.console]
 package = "snarkvm-console-types-field"
 path = "../../../console/types/field"
-version = "=3.7.1"
+version = "=3.8.0"
 optional = true
 
 [dependencies.snarkvm-circuit-environment]
 path = "../../environment"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.snarkvm-circuit-types-boolean]
 path = "../boolean"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [features]
 default = [ "enable_console" ]
diff --git a/circuit/types/group/Cargo.toml b/circuit/types/group/Cargo.toml
index 060d92159..70d40b6a2 100644
--- a/circuit/types/group/Cargo.toml
+++ b/circuit/types/group/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-circuit-types-group"
-version = "3.7.1"
+version = "3.8.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Group circuit for a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -11,24 +11,24 @@ edition = "2021"
 [dependencies.console]
 package = "snarkvm-console-types-group"
 path = "../../../console/types/group"
-version = "=3.7.1"
+version = "=3.8.0"
 optional = true
 
 [dependencies.snarkvm-circuit-environment]
 path = "../../environment"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.snarkvm-circuit-types-boolean]
 path = "../boolean"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.snarkvm-circuit-types-field]
 path = "../field"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.snarkvm-circuit-types-scalar]
 path = "../scalar"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dev-dependencies.snarkvm-utilities]
 path = "../../../utilities"
diff --git a/circuit/types/integers/Cargo.toml b/circuit/types/integers/Cargo.toml
index ad095a6ba..9cf412610 100644
--- a/circuit/types/integers/Cargo.toml
+++ b/circuit/types/integers/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-circuit-types-integers"
-version = "3.7.1"
+version = "3.8.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Integer circuit for a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -11,24 +11,24 @@ edition = "2021"
 [dependencies.console]
 package = "snarkvm-console-types-integers"
 path = "../../../console/types/integers"
-version = "=3.7.1"
+version = "=3.8.0"
 optional = true
 
 [dependencies.snarkvm-circuit-environment]
 path = "../../environment"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.snarkvm-circuit-types-boolean]
 path = "../boolean"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.snarkvm-circuit-types-field]
 path = "../field"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.snarkvm-circuit-types-scalar]
 path = "../scalar"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dev-dependencies.snarkvm-utilities]
 path = "../../../utilities"
diff --git a/circuit/types/scalar/Cargo.toml b/circuit/types/scalar/Cargo.toml
index 4cfe3548c..5214aa94f 100644
--- a/circuit/types/scalar/Cargo.toml
+++ b/circuit/types/scalar/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-circuit-types-scalar"
-version = "3.7.1"
+version = "3.8.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Scalar circuit for a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -11,20 +11,20 @@ edition = "2021"
 [dependencies.console]
 package = "snarkvm-console-types-scalar"
 path = "../../../console/types/scalar"
-version = "=3.7.1"
+version = "=3.8.0"
 optional = true
 
 [dependencies.snarkvm-circuit-environment]
 path = "../../environment"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.snarkvm-circuit-types-boolean]
 path = "../boolean"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.snarkvm-circuit-types-field]
 path = "../field"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [features]
 default = [ "enable_console" ]
diff --git a/circuit/types/string/Cargo.toml b/circuit/types/string/Cargo.toml
index 18f4ba614..22013fc30 100644
--- a/circuit/types/string/Cargo.toml
+++ b/circuit/types/string/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-circuit-types-string"
-version = "3.7.1"
+version = "3.8.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "String circuit for a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -11,24 +11,24 @@ edition = "2021"
 [dependencies.console]
 package = "snarkvm-console-types-string"
 path = "../../../console/types/string"
-version = "=3.7.1"
+version = "=3.8.0"
 optional = true
 
 [dependencies.snarkvm-circuit-environment]
 path = "../../environment"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.snarkvm-circuit-types-boolean]
 path = "../boolean"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.snarkvm-circuit-types-field]
 path = "../field"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.snarkvm-circuit-types-integers]
 path = "../integers"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dev-dependencies.snarkvm-utilities]
 path = "../../../utilities"
diff --git a/console/Cargo.toml b/console/Cargo.toml
index 22cadc590..ae02cd714 100644
--- a/console/Cargo.toml
+++ b/console/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-console"
-version = "3.7.1"
+version = "3.8.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Console environment for a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -10,32 +10,32 @@ edition = "2021"
 
 [dependencies.snarkvm-console-account]
 path = "./account"
-version = "=3.7.1"
+version = "=3.8.0"
 optional = true
 
 [dependencies.snarkvm-console-algorithms]
 path = "./algorithms"
-version = "=3.7.1"
+version = "=3.8.0"
 optional = true
 
 [dependencies.snarkvm-console-collections]
 path = "./collections"
-version = "=3.7.1"
+version = "=3.8.0"
 optional = true
 
 [dependencies.snarkvm-console-network]
 path = "./network"
-version = "=3.7.1"
+version = "=3.8.0"
 optional = true
 
 [dependencies.snarkvm-console-program]
 path = "./program"
-version = "=3.7.1"
+version = "=3.8.0"
 optional = true
 
 [dependencies.snarkvm-console-types]
 path = "./types"
-version = "=3.7.1"
+version = "=3.8.0"
 optional = true
 
 [features]
diff --git a/console/account/Cargo.toml b/console/account/Cargo.toml
index d27dc78fa..60326a4ad 100644
--- a/console/account/Cargo.toml
+++ b/console/account/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-console-account"
-version = "3.7.1"
+version = "3.8.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Account operations for a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -15,11 +15,11 @@ harness = false
 
 [dependencies.snarkvm-console-network]
 path = "../network"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.snarkvm-console-types]
 path = "../types"
-version = "=3.7.1"
+version = "=3.8.0"
 default-features = false
 features = [ "address", "boolean", "field", "group", "scalar" ]
 
diff --git a/console/algorithms/Cargo.toml b/console/algorithms/Cargo.toml
index 45db10e82..aacb65bd2 100644
--- a/console/algorithms/Cargo.toml
+++ b/console/algorithms/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-console-algorithms"
-version = "3.7.1"
+version = "3.8.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Console algorithms for a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -25,18 +25,18 @@ harness = false
 
 [dependencies.snarkvm-console-types]
 path = "../types"
-version = "=3.7.1"
+version = "=3.8.0"
 default-features = false
 features = [ "field", "group", "integers", "scalar" ]
 
 [dependencies.snarkvm-fields]
 path = "../../fields"
-version = "=3.7.1"
+version = "=3.8.0"
 default-features = false
 
 [dependencies.snarkvm-utilities]
 path = "../../utilities"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.blake2s_simd]
 version = "1.0"
diff --git a/console/collections/Cargo.toml b/console/collections/Cargo.toml
index b3336fd6d..5ea532f37 100644
--- a/console/collections/Cargo.toml
+++ b/console/collections/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-console-collections"
-version = "3.7.1"
+version = "3.8.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Collections for a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -20,11 +20,11 @@ harness = false
 
 [dependencies.snarkvm-console-algorithms]
 path = "../algorithms"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.snarkvm-console-types]
 path = "../types"
-version = "=3.7.1"
+version = "=3.8.0"
 default-features = false
 features = [ "field", "integers" ]
 
diff --git a/console/network/Cargo.toml b/console/network/Cargo.toml
index 5753e401b..1f7a130fb 100644
--- a/console/network/Cargo.toml
+++ b/console/network/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-console-network"
-version = "3.7.1"
+version = "3.8.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Network console library for a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -20,36 +20,36 @@ test_consensus_heights = [ ]
 
 [dependencies.snarkvm-algorithms]
 path = "../../algorithms"
-version = "=3.7.1"
+version = "=3.8.0"
 default-features = false
 features = [ "snark" ]
 
 [dependencies.snarkvm-console-algorithms]
 path = "../algorithms"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.snarkvm-console-collections]
 path = "../collections"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.snarkvm-console-network-environment]
 path = "./environment"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.snarkvm-console-types]
 path = "../types"
-version = "=3.7.1"
+version = "=3.8.0"
 default-features = false
 features = [ "field", "group", "scalar" ]
 
 [dependencies.snarkvm-curves]
 path = "../../curves"
-version = "=3.7.1"
+version = "=3.8.0"
 default-features = false
 
 [dependencies.snarkvm-parameters]
 path = "../../parameters"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.anyhow]
 version = "1.0.73"
diff --git a/console/network/environment/Cargo.toml b/console/network/environment/Cargo.toml
index d1e3270b7..a606c9384 100644
--- a/console/network/environment/Cargo.toml
+++ b/console/network/environment/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-console-network-environment"
-version = "3.7.1"
+version = "3.8.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Environment console library for a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -10,17 +10,17 @@ edition = "2021"
 
 [dependencies.snarkvm-curves]
 path = "../../../curves"
-version = "=3.7.1"
+version = "=3.8.0"
 default-features = false
 
 [dependencies.snarkvm-fields]
 path = "../../../fields"
-version = "=3.7.1"
+version = "=3.8.0"
 default-features = false
 
 [dependencies.snarkvm-utilities]
 path = "../../../utilities"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.anyhow]
 version = "1.0.73"
diff --git a/console/program/Cargo.toml b/console/program/Cargo.toml
index ec03e8909..40c5b03b2 100644
--- a/console/program/Cargo.toml
+++ b/console/program/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-console-program"
-version = "3.7.1"
+version = "3.8.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Program operations for a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -14,27 +14,27 @@ test = [ ]
 
 [dependencies.snarkvm-console-account]
 path = "../account"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.snarkvm-console-algorithms]
 path = "../algorithms"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.snarkvm-console-collections]
 path = "../collections"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.snarkvm-console-network]
 path = "../network"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.snarkvm-console-types]
 path = "../types"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.snarkvm-utilities]
 path = "../../utilities"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.enum_index]
 version = "0.2"
diff --git a/console/types/Cargo.toml b/console/types/Cargo.toml
index 1dabcc0e6..a5049d463 100644
--- a/console/types/Cargo.toml
+++ b/console/types/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-console-types"
-version = "3.7.1"
+version = "3.8.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Console types for a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -15,41 +15,41 @@ harness = false
 
 [dependencies.snarkvm-console-network-environment]
 path = "../network/environment"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.snarkvm-console-types-address]
 path = "./address"
-version = "=3.7.1"
+version = "=3.8.0"
 optional = true
 
 [dependencies.snarkvm-console-types-boolean]
 path = "./boolean"
-version = "=3.7.1"
+version = "=3.8.0"
 optional = true
 
 [dependencies.snarkvm-console-types-field]
 path = "./field"
-version = "=3.7.1"
+version = "=3.8.0"
 optional = true
 
 [dependencies.snarkvm-console-types-group]
 path = "./group"
-version = "=3.7.1"
+version = "=3.8.0"
 optional = true
 
 [dependencies.snarkvm-console-types-integers]
 path = "./integers"
-version = "=3.7.1"
+version = "=3.8.0"
 optional = true
 
 [dependencies.snarkvm-console-types-scalar]
 path = "./scalar"
-version = "=3.7.1"
+version = "=3.8.0"
 optional = true
 
 [dependencies.snarkvm-console-types-string]
 path = "./string"
-version = "=3.7.1"
+version = "=3.8.0"
 optional = true
 
 [dev-dependencies.criterion]
diff --git a/console/types/address/Cargo.toml b/console/types/address/Cargo.toml
index db9bb4977..dd0c5bbcc 100644
--- a/console/types/address/Cargo.toml
+++ b/console/types/address/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-console-types-address"
-version = "3.7.1"
+version = "3.8.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Type operations for a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -10,19 +10,19 @@ edition = "2021"
 
 [dependencies.snarkvm-console-network-environment]
 path = "../../network/environment"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.snarkvm-console-types-boolean]
 path = "../boolean"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.snarkvm-console-types-field]
 path = "../field"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.snarkvm-console-types-group]
 path = "../group"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dev-dependencies.bincode]
 version = "1.3"
diff --git a/console/types/boolean/Cargo.toml b/console/types/boolean/Cargo.toml
index 2ebcec7ee..b96804da9 100644
--- a/console/types/boolean/Cargo.toml
+++ b/console/types/boolean/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-console-types-boolean"
-version = "3.7.1"
+version = "3.8.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Type operations for a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -10,7 +10,7 @@ edition = "2021"
 
 [dependencies.snarkvm-console-network-environment]
 path = "../../network/environment"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dev-dependencies.bincode]
 version = "1.3"
diff --git a/console/types/field/Cargo.toml b/console/types/field/Cargo.toml
index 947841d1c..bd873ba95 100644
--- a/console/types/field/Cargo.toml
+++ b/console/types/field/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-console-types-field"
-version = "3.7.1"
+version = "3.8.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Type operations for a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -10,11 +10,11 @@ edition = "2021"
 
 [dependencies.snarkvm-console-network-environment]
 path = "../../network/environment"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.snarkvm-console-types-boolean]
 path = "../boolean"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.zeroize]
 version = "1"
diff --git a/console/types/group/Cargo.toml b/console/types/group/Cargo.toml
index 2f83eb7b9..a548f78cb 100644
--- a/console/types/group/Cargo.toml
+++ b/console/types/group/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-console-types-group"
-version = "3.7.1"
+version = "3.8.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Type operations for a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -10,19 +10,19 @@ edition = "2021"
 
 [dependencies.snarkvm-console-network-environment]
 path = "../../network/environment"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.snarkvm-console-types-boolean]
 path = "../boolean"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.snarkvm-console-types-field]
 path = "../field"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.snarkvm-console-types-scalar]
 path = "../scalar"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dev-dependencies.bincode]
 version = "1.3"
diff --git a/console/types/integers/Cargo.toml b/console/types/integers/Cargo.toml
index a4e10eed9..59d209673 100644
--- a/console/types/integers/Cargo.toml
+++ b/console/types/integers/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-console-types-integers"
-version = "3.7.1"
+version = "3.8.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Type operations for a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -10,19 +10,19 @@ edition = "2021"
 
 [dependencies.snarkvm-console-network-environment]
 path = "../../network/environment"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.snarkvm-console-types-boolean]
 path = "../boolean"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.snarkvm-console-types-field]
 path = "../field"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.snarkvm-console-types-scalar]
 path = "../scalar"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dev-dependencies.bincode]
 version = "1.3"
diff --git a/console/types/scalar/Cargo.toml b/console/types/scalar/Cargo.toml
index 14a2400f6..38734a46c 100644
--- a/console/types/scalar/Cargo.toml
+++ b/console/types/scalar/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-console-types-scalar"
-version = "3.7.1"
+version = "3.8.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Type operations for a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -10,15 +10,15 @@ edition = "2021"
 
 [dependencies.snarkvm-console-network-environment]
 path = "../../network/environment"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.snarkvm-console-types-boolean]
 path = "../boolean"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.snarkvm-console-types-field]
 path = "../field"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.zeroize]
 version = "1"
diff --git a/console/types/string/Cargo.toml b/console/types/string/Cargo.toml
index 7bf806c22..d04f4f340 100644
--- a/console/types/string/Cargo.toml
+++ b/console/types/string/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-console-types-string"
-version = "3.7.1"
+version = "3.8.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Type operations for a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -10,19 +10,19 @@ edition = "2021"
 
 [dependencies.snarkvm-console-network-environment]
 path = "../../network/environment"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.snarkvm-console-types-boolean]
 path = "../boolean"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.snarkvm-console-types-field]
 path = "../field"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.snarkvm-console-types-integers]
 path = "../integers"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dev-dependencies.bincode]
 version = "1.3"
diff --git a/curves/Cargo.toml b/curves/Cargo.toml
index 5a01153ab..5b59463b1 100644
--- a/curves/Cargo.toml
+++ b/curves/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-curves"
-version = "3.7.1"
+version = "3.8.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Curves for a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -36,12 +36,12 @@ harness = false
 
 [dependencies.snarkvm-fields]
 path = "../fields"
-version = "=3.7.1"
+version = "=3.8.0"
 default-features = false
 
 [dependencies.snarkvm-utilities]
 path = "../utilities"
-version = "=3.7.1"
+version = "=3.8.0"
 default-features = false
 
 [dependencies.rand]
diff --git a/fields/Cargo.toml b/fields/Cargo.toml
index b2bbd474f..2a582797d 100644
--- a/fields/Cargo.toml
+++ b/fields/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-fields"
-version = "3.7.1"
+version = "3.8.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Fields for a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -30,7 +30,7 @@ harness = false
 
 [dependencies.snarkvm-utilities]
 path = "../utilities"
-version = "=3.7.1"
+version = "=3.8.0"
 default-features = false
 
 [dependencies.aleo-std]
diff --git a/ledger/Cargo.toml b/ledger/Cargo.toml
index 0c5484f9b..796a1a212 100644
--- a/ledger/Cargo.toml
+++ b/ledger/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-ledger"
-version = "3.7.1"
+version = "3.8.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "A node ledger for a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -68,54 +68,54 @@ timer = [ "aleo-std/timer" ]
 [dependencies.console]
 package = "snarkvm-console"
 path = "../console"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.ledger-authority]
 package = "snarkvm-ledger-authority"
 path = "./authority"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.ledger-block]
 package = "snarkvm-ledger-block"
 path = "./block"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.ledger-committee]
 package = "snarkvm-ledger-committee"
 path = "./committee"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.ledger-narwhal]
 package = "snarkvm-ledger-narwhal"
 path = "./narwhal"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.ledger-puzzle]
 package = "snarkvm-ledger-puzzle"
 path = "puzzle"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.ledger-query]
 package = "snarkvm-ledger-query"
 path = "./query"
-version = "=3.7.1"
+version = "=3.8.0"
 features = [ "query" ]
 
 [dependencies.ledger-store]
 package = "snarkvm-ledger-store"
 path = "./store"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.ledger-test-helpers]
 package = "snarkvm-ledger-test-helpers"
 path = "./test-helpers"
-version = "=3.7.1"
+version = "=3.8.0"
 optional = true
 
 [dependencies.synthesizer]
 package = "snarkvm-synthesizer"
 path = "../synthesizer"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.aleo-std]
 version = "1.0.1"
diff --git a/ledger/authority/Cargo.toml b/ledger/authority/Cargo.toml
index f18fc371d..b15acb1e1 100644
--- a/ledger/authority/Cargo.toml
+++ b/ledger/authority/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-ledger-authority"
-version = "3.7.1"
+version = "3.8.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Data structures for a block authority in a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -32,12 +32,12 @@ test-helpers = [ "narwhal-subdag/test-helpers" ]
 [dependencies.console]
 package = "snarkvm-console"
 path = "../../console"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.narwhal-subdag]
 package = "snarkvm-ledger-narwhal-subdag"
 path = "../narwhal/subdag"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.anyhow]
 version = "1"
diff --git a/ledger/block/Cargo.toml b/ledger/block/Cargo.toml
index b3c071ed5..9b201625e 100644
--- a/ledger/block/Cargo.toml
+++ b/ledger/block/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-ledger-block"
-version = "3.7.1"
+version = "3.8.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "A block for a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -39,52 +39,52 @@ test = [ ]
 [dependencies.console]
 package = "snarkvm-console"
 path = "../../console"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.ledger-authority]
 package = "snarkvm-ledger-authority"
 path = "../authority"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.ledger-committee]
 package = "snarkvm-ledger-committee"
 path = "../../ledger/committee"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.ledger-narwhal-batch-header]
 package = "snarkvm-ledger-narwhal-batch-header"
 path = "../narwhal/batch-header"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.ledger-narwhal-data]
 package = "snarkvm-ledger-narwhal-data"
 path = "../narwhal/data"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.ledger-narwhal-subdag]
 package = "snarkvm-ledger-narwhal-subdag"
 path = "../narwhal/subdag"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.ledger-narwhal-transmission-id]
 package = "snarkvm-ledger-narwhal-transmission-id"
 path = "../narwhal/transmission-id"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.ledger-puzzle]
 package = "snarkvm-ledger-puzzle"
 path = "../puzzle"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.synthesizer-program]
 package = "snarkvm-synthesizer-program"
 path = "../../synthesizer/program"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.synthesizer-snark]
 package = "snarkvm-synthesizer-snark"
 path = "../../synthesizer/snark"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.indexmap]
 version = "2.0"
diff --git a/ledger/committee/Cargo.toml b/ledger/committee/Cargo.toml
index 84ffd0afe..84d19a97e 100644
--- a/ledger/committee/Cargo.toml
+++ b/ledger/committee/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-ledger-committee"
-version = "3.7.1"
+version = "3.8.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "A committee for a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -34,12 +34,12 @@ test-helpers = [ "prop-tests", "rand_distr" ]
 [dependencies.console]
 package = "snarkvm-console"
 path = "../../console"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.ledger-narwhal-batch-header]
 package = "snarkvm-ledger-narwhal-batch-header"
 path = "../narwhal/batch-header"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.indexmap]
 version = "2.0"
@@ -48,7 +48,7 @@ features = [ "serde", "rayon" ]
 [dependencies.metrics]
 package = "snarkvm-metrics"
 path = "../../metrics"
-version = "=3.7.1"
+version = "=3.8.0"
 optional = true
 
 [dependencies.serde_json]
diff --git a/ledger/narwhal/Cargo.toml b/ledger/narwhal/Cargo.toml
index 25c08ac62..13bb2bd22 100644
--- a/ledger/narwhal/Cargo.toml
+++ b/ledger/narwhal/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-ledger-narwhal"
-version = "3.7.1"
+version = "3.8.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Data structures for a Narwhal-style memory pool in a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -64,37 +64,37 @@ transmission-id = [ "narwhal-transmission-id" ]
 [dependencies.narwhal-batch-certificate]
 package = "snarkvm-ledger-narwhal-batch-certificate"
 path = "./batch-certificate"
-version = "=3.7.1"
+version = "=3.8.0"
 optional = true
 
 [dependencies.narwhal-batch-header]
 package = "snarkvm-ledger-narwhal-batch-header"
 path = "./batch-header"
-version = "=3.7.1"
+version = "=3.8.0"
 optional = true
 
 [dependencies.narwhal-data]
 package = "snarkvm-ledger-narwhal-data"
 path = "./data"
-version = "=3.7.1"
+version = "=3.8.0"
 optional = true
 
 [dependencies.narwhal-subdag]
 package = "snarkvm-ledger-narwhal-subdag"
 path = "./subdag"
-version = "=3.7.1"
+version = "=3.8.0"
 optional = true
 
 [dependencies.narwhal-transmission]
 package = "snarkvm-ledger-narwhal-transmission"
 path = "./transmission"
-version = "=3.7.1"
+version = "=3.8.0"
 optional = true
 
 [dependencies.narwhal-transmission-id]
 package = "snarkvm-ledger-narwhal-transmission-id"
 path = "./transmission-id"
-version = "=3.7.1"
+version = "=3.8.0"
 optional = true
 
 [dev-dependencies.snarkvm-ledger-narwhal]
diff --git a/ledger/narwhal/batch-certificate/Cargo.toml b/ledger/narwhal/batch-certificate/Cargo.toml
index 62874a889..5049b9e38 100644
--- a/ledger/narwhal/batch-certificate/Cargo.toml
+++ b/ledger/narwhal/batch-certificate/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-ledger-narwhal-batch-certificate"
-version = "3.7.1"
+version = "3.8.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "A batch certificate for a Narwhal-style memory pool in a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -32,17 +32,17 @@ test-helpers = [ "narwhal-batch-header/test-helpers" ]
 [dependencies.console]
 package = "snarkvm-console"
 path = "../../../console"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.narwhal-batch-header]
 package = "snarkvm-ledger-narwhal-batch-header"
 path = "../batch-header"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.narwhal-transmission-id]
 package = "snarkvm-ledger-narwhal-transmission-id"
 path = "../transmission-id"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.indexmap]
 version = "2.0"
diff --git a/ledger/narwhal/batch-header/Cargo.toml b/ledger/narwhal/batch-header/Cargo.toml
index 02a0eb2bd..06c679d17 100644
--- a/ledger/narwhal/batch-header/Cargo.toml
+++ b/ledger/narwhal/batch-header/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-ledger-narwhal-batch-header"
-version = "3.7.1"
+version = "3.8.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "A batch header for a Narwhal-style memory pool in a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -32,12 +32,12 @@ test-helpers = [ "narwhal-transmission-id/test-helpers", "time" ]
 [dependencies.console]
 package = "snarkvm-console"
 path = "../../../console"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.narwhal-transmission-id]
 package = "snarkvm-ledger-narwhal-transmission-id"
 path = "../transmission-id"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.indexmap]
 version = "2.0"
diff --git a/ledger/narwhal/data/Cargo.toml b/ledger/narwhal/data/Cargo.toml
index 32a27c684..8df6b7396 100644
--- a/ledger/narwhal/data/Cargo.toml
+++ b/ledger/narwhal/data/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-ledger-narwhal-data"
-version = "3.7.1"
+version = "3.8.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "A batch certificate for a Narwhal-style memory pool in a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -29,7 +29,7 @@ async = [ "tokio" ]
 [dependencies.console]
 package = "snarkvm-console"
 path = "../../../console"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.bytes]
 version = "1"
diff --git a/ledger/narwhal/subdag/Cargo.toml b/ledger/narwhal/subdag/Cargo.toml
index bdcd61e76..127f912b6 100644
--- a/ledger/narwhal/subdag/Cargo.toml
+++ b/ledger/narwhal/subdag/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-ledger-narwhal-subdag"
-version = "3.7.1"
+version = "3.8.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "A subdag for a Narwhal-style memory pool in a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -32,27 +32,27 @@ test-helpers = [ "narwhal-batch-certificate/test-helpers" ]
 [dependencies.console]
 package = "snarkvm-console"
 path = "../../../console"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.narwhal-batch-certificate]
 package = "snarkvm-ledger-narwhal-batch-certificate"
 path = "../batch-certificate"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.narwhal-batch-header]
 package = "snarkvm-ledger-narwhal-batch-header"
 path = "../batch-header"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.ledger-committee]
 package = "snarkvm-ledger-committee"
 path = "../../committee"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.narwhal-transmission-id]
 package = "snarkvm-ledger-narwhal-transmission-id"
 path = "../transmission-id"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.indexmap]
 version = "2.0"
diff --git a/ledger/narwhal/transmission-id/Cargo.toml b/ledger/narwhal/transmission-id/Cargo.toml
index 216aaef1b..a0158e30a 100644
--- a/ledger/narwhal/transmission-id/Cargo.toml
+++ b/ledger/narwhal/transmission-id/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-ledger-narwhal-transmission-id"
-version = "3.7.1"
+version = "3.8.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "A transmission ID for a Narwhal-style memory pool in a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -32,12 +32,12 @@ test-helpers = [ ]
 [dependencies.console]
 package = "snarkvm-console"
 path = "../../../console"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.ledger-puzzle]
 package = "snarkvm-ledger-puzzle"
 path = "../../puzzle"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dev-dependencies.bincode]
 version = "1.3"
diff --git a/ledger/narwhal/transmission/Cargo.toml b/ledger/narwhal/transmission/Cargo.toml
index b22d231bf..a8a7d280d 100644
--- a/ledger/narwhal/transmission/Cargo.toml
+++ b/ledger/narwhal/transmission/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-ledger-narwhal-transmission"
-version = "3.7.1"
+version = "3.8.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "A transmission for a Narwhal-style memory pool in a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -32,22 +32,22 @@ test-helpers = [ ]
 [dependencies.console]
 package = "snarkvm-console"
 path = "../../../console"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.ledger-block]
 package = "snarkvm-ledger-block"
 path = "../../block"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.ledger-narwhal-data]
 package = "snarkvm-ledger-narwhal-data"
 path = "../data"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.ledger-puzzle]
 package = "snarkvm-ledger-puzzle"
 path = "../../puzzle"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.bytes]
 version = "1"
diff --git a/ledger/puzzle/Cargo.toml b/ledger/puzzle/Cargo.toml
index 542752d7a..b81fa5b2c 100644
--- a/ledger/puzzle/Cargo.toml
+++ b/ledger/puzzle/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-ledger-puzzle"
-version = "3.7.1"
+version = "3.8.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Puzzle for a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -41,11 +41,11 @@ wasm = [ "console/wasm", "snarkvm-algorithms/wasm" ]
 [dependencies.console]
 package = "snarkvm-console"
 path = "../../console"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.snarkvm-algorithms]
 path = "../../algorithms"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.aleo-std]
 version = "1.0.1"
diff --git a/ledger/puzzle/epoch/Cargo.toml b/ledger/puzzle/epoch/Cargo.toml
index 50af6a9b0..a9f9536f6 100644
--- a/ledger/puzzle/epoch/Cargo.toml
+++ b/ledger/puzzle/epoch/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-ledger-puzzle-epoch"
-version = "3.7.1"
+version = "3.8.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Epoch puzzle for a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -46,26 +46,26 @@ wasm = [ "console/wasm" ]
 [dependencies.circuit]
 package = "snarkvm-circuit"
 path = "../../../circuit"
-version = "=3.7.1"
+version = "=3.8.0"
 optional = true
 
 [dependencies.console]
 package = "snarkvm-console"
 path = "../../../console"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.snarkvm-ledger-puzzle]
 path = "../."
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.snarkvm-synthesizer-process]
 path = "../../../synthesizer/process"
-version = "=3.7.1"
+version = "=3.8.0"
 optional = true
 
 [dependencies.snarkvm-synthesizer-program]
 path = "../../../synthesizer/program"
-version = "=3.7.1"
+version = "=3.8.0"
 optional = true
 
 [dependencies.aleo-std]
diff --git a/ledger/query/Cargo.toml b/ledger/query/Cargo.toml
index 03b6bbb31..a1c80be57 100644
--- a/ledger/query/Cargo.toml
+++ b/ledger/query/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-ledger-query"
-version = "3.7.1"
+version = "3.8.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "A query for a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -39,18 +39,18 @@ query = [
 [dependencies.console]
 package = "snarkvm-console"
 path = "../../console"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.ledger-store]
 package = "snarkvm-ledger-store"
 path = "../store"
-version = "=3.7.1"
+version = "=3.8.0"
 optional = true
 
 [dependencies.synthesizer-program]
 package = "snarkvm-synthesizer-program"
 path = "../../synthesizer/program"
-version = "=3.7.1"
+version = "=3.8.0"
 optional = true
 
 [dependencies.async-trait]
diff --git a/ledger/src/tests.rs b/ledger/src/tests.rs
index bcd56d212..a17f1f932 100644
--- a/ledger/src/tests.rs
+++ b/ledger/src/tests.rs
@@ -2936,7 +2936,7 @@ mod valid_solutions {
         let validator_address = Address::try_from(&private_key).unwrap();
 
         // Initialize the store.
-        let store = ConsensusStore::<_, LedgerType<_>>::open(StorageMode::new_test(None)).unwrap();
+        let store = ConsensusStore::<_, LedgerType>::open(StorageMode::new_test(None)).unwrap();
         // Create a genesis block with a seeded RNG to reproduce the same genesis private keys.
         let seed: u64 = rng.gen();
         let genesis_rng = &mut TestRng::from_seed(seed);
@@ -2955,8 +2955,7 @@ mod valid_solutions {
         let mut chain_builder = TestChainBuilder::new(private_keys.to_vec(), genesis.clone());
 
         // Construct the ledger.
-        let ledger =
-            Ledger::<CurrentNetwork, LedgerType<CurrentNetwork>>::load(genesis, StorageMode::new_test(None)).unwrap();
+        let ledger = Ledger::<CurrentNetwork, LedgerType>::load(genesis, StorageMode::new_test(None)).unwrap();
 
         // Retrieve the puzzle parameters.
         let puzzle = ledger.puzzle();
diff --git a/ledger/store/Cargo.toml b/ledger/store/Cargo.toml
index 1c531a193..d7b80037b 100644
--- a/ledger/store/Cargo.toml
+++ b/ledger/store/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-ledger-store"
-version = "3.7.1"
+version = "3.8.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "A data store for a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -43,42 +43,42 @@ test = [ ]
 [dependencies.console]
 package = "snarkvm-console"
 path = "../../console"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.ledger-authority]
 package = "snarkvm-ledger-authority"
 path = "../authority"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.ledger-block]
 package = "snarkvm-ledger-block"
 path = "../block"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.ledger-committee]
 package = "snarkvm-ledger-committee"
 path = "../committee"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.ledger-narwhal-batch-certificate]
 package = "snarkvm-ledger-narwhal-batch-certificate"
 path = "../narwhal/batch-certificate"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.ledger-puzzle]
 package = "snarkvm-ledger-puzzle"
 path = "../puzzle"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.synthesizer-program]
 package = "snarkvm-synthesizer-program"
 path = "../../synthesizer/program"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.synthesizer-snark]
 package = "snarkvm-synthesizer-snark"
 path = "../../synthesizer/snark"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.aleo-std-storage]
 version = "1.0.1"
diff --git a/ledger/test-helpers/Cargo.toml b/ledger/test-helpers/Cargo.toml
index 3358624b0..b8738dcc4 100644
--- a/ledger/test-helpers/Cargo.toml
+++ b/ledger/test-helpers/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-ledger-test-helpers"
-version = "3.7.1"
+version = "3.8.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Test helpers for a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -19,44 +19,44 @@ edition = "2021"
 [dependencies.algorithms]
 package = "snarkvm-algorithms"
 path = "../../algorithms"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.console]
 package = "snarkvm-console"
 path = "../../console"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.circuit]
 package = "snarkvm-circuit"
 path = "../../circuit"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.ledger-block]
 package = "snarkvm-ledger-block"
 path = "../block"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.ledger-query]
 package = "snarkvm-ledger-query"
 path = "../query"
-version = "=3.7.1"
+version = "=3.8.0"
 default-features = false
 features = [ "query" ]
 
 [dependencies.ledger-store]
 package = "snarkvm-ledger-store"
 path = "../store"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.synthesizer-program]
 package = "snarkvm-synthesizer-program"
 path = "../../synthesizer/program"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.synthesizer-process]
 package = "snarkvm-synthesizer-process"
 path = "../../synthesizer/process"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.aleo-std]
 version = "1.0.1"
diff --git a/metrics/Cargo.toml b/metrics/Cargo.toml
index a68d3912c..8c345f23a 100644
--- a/metrics/Cargo.toml
+++ b/metrics/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-metrics"
-version = "3.7.1"
+version = "3.8.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Metrics for a decentralized virtual machine"
 homepage = "https://aleo.org"
diff --git a/parameters/Cargo.toml b/parameters/Cargo.toml
index 06b85dcb1..91c62b1f0 100644
--- a/parameters/Cargo.toml
+++ b/parameters/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-parameters"
-version = "3.7.1"
+version = "3.8.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Parameters for a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -34,12 +34,12 @@ wasm = [ "encoding", "js-sys", "web-sys" ]
 
 [dependencies.snarkvm-curves]
 path = "../curves"
-version = "=3.7.1"
+version = "=3.8.0"
 default-features = false
 
 [dependencies.snarkvm-utilities]
 path = "../utilities"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.aleo-std]
 version = "1.0.1"
diff --git a/synthesizer/Cargo.toml b/synthesizer/Cargo.toml
index c4508d763..665b5ab66 100644
--- a/synthesizer/Cargo.toml
+++ b/synthesizer/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-synthesizer"
-version = "3.7.1"
+version = "3.8.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Synthesizer for a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -76,78 +76,78 @@ harness = false
 [dependencies.algorithms]
 package = "snarkvm-algorithms"
 path = "../algorithms"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.circuit]
 package = "snarkvm-circuit"
 path = "../circuit"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.console]
 package = "snarkvm-console"
 path = "../console"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.ledger-block]
 package = "snarkvm-ledger-block"
 path = "../ledger/block"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.ledger-committee]
 package = "snarkvm-ledger-committee"
 path = "../ledger/committee"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.ledger-narwhal-data]
 package = "snarkvm-ledger-narwhal-data"
 path = "../ledger/narwhal/data"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.ledger-puzzle]
 package = "snarkvm-ledger-puzzle"
 path = "../ledger/puzzle"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.ledger-puzzle-epoch]
 package = "snarkvm-ledger-puzzle-epoch"
 path = "../ledger/puzzle/epoch"
-version = "=3.7.1"
+version = "=3.8.0"
 features = [ "synthesis" ]
 
 [dependencies.ledger-query]
 package = "snarkvm-ledger-query"
 path = "../ledger/query"
-version = "=3.7.1"
+version = "=3.8.0"
 default-features = false
 features = [ "query" ]
 
 [dependencies.ledger-store]
 package = "snarkvm-ledger-store"
 path = "../ledger/store"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.synthesizer-process]
 package = "snarkvm-synthesizer-process"
 path = "./process"
-version = "=3.7.1"
+version = "=3.8.0"
 optional = true
 
 [dependencies.synthesizer-program]
 package = "snarkvm-synthesizer-program"
 path = "./program"
-version = "=3.7.1"
+version = "=3.8.0"
 optional = true
 
 [dependencies.synthesizer-snark]
 package = "snarkvm-synthesizer-snark"
 path = "./snark"
-version = "=3.7.1"
+version = "=3.8.0"
 optional = true
 
 [dependencies.utilities]
 package = "snarkvm-utilities"
 path = "../utilities"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.aleo-std]
 version = "1.0.1"
diff --git a/synthesizer/process/Cargo.toml b/synthesizer/process/Cargo.toml
index 1c0e39498..0af7fc408 100644
--- a/synthesizer/process/Cargo.toml
+++ b/synthesizer/process/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-synthesizer-process"
-version = "3.7.1"
+version = "3.8.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "A process for a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -59,50 +59,50 @@ harness = false
 [dependencies.algorithms]
 package = "snarkvm-algorithms"
 path = "../../algorithms"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.console]
 package = "snarkvm-console"
 path = "../../console"
-version = "=3.7.1"
+version = "=3.8.0"
 default-features = false
 features = [ "network", "program", "types" ]
 
 [dependencies.circuit]
 package = "snarkvm-circuit"
 path = "../../circuit"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.ledger-block]
 package = "snarkvm-ledger-block"
 path = "../../ledger/block"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.ledger-query]
 package = "snarkvm-ledger-query"
 path = "../../ledger/query"
-version = "=3.7.1"
+version = "=3.8.0"
 default-features = false
 
 [dependencies.ledger-store]
 package = "snarkvm-ledger-store"
 path = "../../ledger/store"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.synthesizer-program]
 package = "snarkvm-synthesizer-program"
 path = "../../synthesizer/program"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.synthesizer-snark]
 package = "snarkvm-synthesizer-snark"
 path = "../../synthesizer/snark"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.utilities]
 package = "snarkvm-utilities"
 path = "../../utilities"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.aleo-std]
 version = "1.0.1"
diff --git a/synthesizer/program/Cargo.toml b/synthesizer/program/Cargo.toml
index 94ae8ca8f..b3115fe51 100644
--- a/synthesizer/program/Cargo.toml
+++ b/synthesizer/program/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-synthesizer-program"
-version = "3.7.1"
+version = "3.8.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Program for a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -31,12 +31,12 @@ wasm = [ "console/wasm" ]
 [dependencies.circuit]
 package = "snarkvm-circuit"
 path = "../../circuit"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.console]
 package = "snarkvm-console"
 path = "../../console"
-version = "=3.7.1"
+version = "=3.8.0"
 default-features = false
 features = [ "account", "network", "program", "types" ]
 
@@ -60,7 +60,7 @@ features = [ "preserve_order" ]
 [dependencies.synthesizer-snark]
 package = "snarkvm-synthesizer-snark"
 path = "../snark"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dev-dependencies.bincode]
 version = "1"
diff --git a/synthesizer/snark/Cargo.toml b/synthesizer/snark/Cargo.toml
index 2ac903752..3fefdcba7 100644
--- a/synthesizer/snark/Cargo.toml
+++ b/synthesizer/snark/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-synthesizer-snark"
-version = "3.7.1"
+version = "3.8.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "SNARK wrappers for a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -33,18 +33,18 @@ wasm = [ "console/wasm", "snarkvm-algorithms/wasm" ]
 [dependencies.circuit]
 package = "snarkvm-circuit"
 path = "../../circuit"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.console]
 package = "snarkvm-console"
 path = "../../console"
-version = "=3.7.1"
+version = "=3.8.0"
 default-features = false
 features = [ "network" ]
 
 [dependencies.snarkvm-algorithms]
 path = "../../algorithms"
-version = "=3.7.1"
+version = "=3.8.0"
 
 [dependencies.bincode]
 version = "1"
diff --git a/utilities/Cargo.toml b/utilities/Cargo.toml
index 351dbeb90..712b3b9fa 100644
--- a/utilities/Cargo.toml
+++ b/utilities/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-utilities"
-version = "3.7.1"
+version = "3.8.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Utilities for a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -25,7 +25,7 @@ edition = "2021"
 
 [dependencies.snarkvm-utilities-derives]
 path = "./derives"
-version = "=3.7.1"
+version = "=3.8.0"
 optional = true
 
 [dependencies.aleo-std]
diff --git a/utilities/derives/Cargo.toml b/utilities/derives/Cargo.toml
index 852f8c6c7..ab1047eb0 100644
--- a/utilities/derives/Cargo.toml
+++ b/utilities/derives/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-utilities-derives"
-version = "3.7.1"
+version = "3.8.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Canonical serialization for a decentralized virtual machine"
 homepage = "https://aleo.org"
diff --git a/wasm/Cargo.toml b/wasm/Cargo.toml
index 37e70fbe3..1c1d2f0f3 100644
--- a/wasm/Cargo.toml
+++ b/wasm/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-wasm"
-version = "3.7.1"
+version = "3.8.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "WASM for a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -51,48 +51,48 @@ utilities = [ "snarkvm-utilities" ]
 
 [dependencies.snarkvm-console]
 path = "../console"
-version = "=3.7.1"
+version = "=3.8.0"
 features = [ "wasm" ]
 optional = true
 
 [dependencies.snarkvm-curves]
 path = "../curves"
-version = "=3.7.1"
+version = "=3.8.0"
 optional = true
 
 [dependencies.snarkvm-fields]
 path = "../fields"
-version = "=3.7.1"
+version = "=3.8.0"
 optional = true
 
 [dependencies.snarkvm-ledger-block]
 path = "../ledger/block"
-version = "=3.7.1"
+version = "=3.8.0"
 features = [ "wasm" ]
 optional = true
 
 [dependencies.snarkvm-ledger-query]
 path = "../ledger/query"
-version = "=3.7.1"
+version = "=3.8.0"
 features = [ "async", "wasm" ]
 optional = true
 
 [dependencies.snarkvm-ledger-store]
 path = "../ledger/store"
-version = "=3.7.1"
+version = "=3.8.0"
 features = [ "wasm" ]
 optional = true
 
 [dependencies.snarkvm-synthesizer]
 path = "../synthesizer"
-version = "=3.7.1"
+version = "=3.8.0"
 default-features = false
 features = [ "async", "wasm" ]
 optional = true
 
 [dependencies.snarkvm-utilities]
 path = "../utilities"
-version = "=3.7.1"
+version = "=3.8.0"
 features = [ "wasm" ]
 optional = true
 
