diff --git a/Cargo.lock b/Cargo.lock
index 8365a7f06..d57e13b77 100644
--- a/Cargo.lock
+++ b/Cargo.lock
@@ -536,9 +536,9 @@ dependencies = [
 
 [[package]]
 name = "cc"
-version = "1.2.36"
+version = "1.2.37"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "5252b3d2648e5eedbc1a6f501e3c795e07025c1e93bbf8bbdd6eef7f447a6d54"
+checksum = "65193589c6404eb80b450d618eaf9a2cafaaafd57ecce47370519ef674a7bd44"
 dependencies = [
  "find-msvc-tools",
  "jobserver",
@@ -1454,7 +1454,7 @@ dependencies = [
  "js-sys",
  "libc",
  "r-efi",
- "wasi 0.14.5+wasi-0.2.4",
+ "wasi 0.14.6+wasi-0.2.4",
  "wasm-bindgen",
 ]
 
@@ -1770,9 +1770,9 @@ dependencies = [
 
 [[package]]
 name = "hyper-util"
-version = "0.1.16"
+version = "0.1.17"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "8d9b05277c7e8da2c93a568989bb6207bef0112e8d17df7a6eda4a3cf143bc5e"
+checksum = "3c6995591a8f1380fcb4ba966a252a4b29188d51d2b89e3a252f5305be65aea8"
 dependencies = [
  "base64 0.22.1",
  "bytes",
@@ -1796,9 +1796,9 @@ dependencies = [
 
 [[package]]
 name = "iana-time-zone"
-version = "0.1.63"
+version = "0.1.64"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "b0c919e5debc312ad217002b8048a17b7d83f80703865bbfcfebb0458b0b27d8"
+checksum = "33e57f83510bb73707521ebaffa789ec8caf86f9657cad665b092b581d40e9fb"
 dependencies = [
  "android_system_properties",
  "core-foundation-sys",
@@ -2154,9 +2154,9 @@ checksum = "f9fbbcab51052fe104eb5e5d351cf728d30a5be1fe14d9be8a3b097481fb97de"
 
 [[package]]
 name = "libredox"
-version = "0.1.9"
+version = "0.1.10"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "391290121bad3d37fbddad76d8f5d1c1c314cfc646d143d7e07a3086ddff0ce3"
+checksum = "416f7e718bdb06000964960ffa43b4335ad4012ae8b99060261aa4a8088d5ccb"
 dependencies = [
  "bitflags 2.9.4",
  "libc",
@@ -3321,9 +3321,9 @@ dependencies = [
 
 [[package]]
 name = "rustls-webpki"
-version = "0.103.5"
+version = "0.103.6"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "b5a37813727b78798e53c2bec3f5e8fe12a6d6f8389bf9ca7802add4c9905ad8"
+checksum = "8572f3c2cb9934231157b45499fc41e1f58c589fdfb81a844ba873265e80f8eb"
 dependencies = [
  "ring",
  "rustls-pki-types",
@@ -3471,24 +3471,34 @@ dependencies = [
 
 [[package]]
 name = "semver"
-version = "1.0.26"
+version = "1.0.27"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "56e6fa9c48d24d85fb3de5ad847117517440f6beceb7798af16b4a87d616b8d0"
+checksum = "d767eb0aabc880b29956c35734170f26ed551a859dbd361d140cdbeca61ab1e2"
 
 [[package]]
 name = "serde"
-version = "1.0.219"
+version = "1.0.224"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "6aaeb1e94f53b16384af593c71e20b095e958dab1d26939c1b70645c5cfbcc0b"
+dependencies = [
+ "serde_core",
+ "serde_derive",
+]
+
+[[package]]
+name = "serde_core"
+version = "1.0.224"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "5f0e2c6ed6606019b4e29e69dbaba95b11854410e5347d525002456dbbb786b6"
+checksum = "32f39390fa6346e24defbcdd3d9544ba8a19985d0af74df8501fbfe9a64341ab"
 dependencies = [
  "serde_derive",
 ]
 
 [[package]]
 name = "serde_derive"
-version = "1.0.219"
+version = "1.0.224"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "5b0276cf7f2c73365f7157c8123c21cd9a50fbbd844757af28ca1f5925fc2a00"
+checksum = "87ff78ab5e8561c9a675bfc1785cb07ae721f0ee53329a595cefd8c04c2ac4e0"
 dependencies = [
  "proc-macro2",
  "quote 1.0.40",
@@ -3497,34 +3507,36 @@ dependencies = [
 
 [[package]]
 name = "serde_json"
-version = "1.0.143"
+version = "1.0.145"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "d401abef1d108fbd9cbaebc3e46611f4b1021f714a0597a71f41ee463f5f4a5a"
+checksum = "402a6f66d8c709116cf22f558eab210f5a50187f702eb4d7e5ef38d9a7f1c79c"
 dependencies = [
  "indexmap 2.11.1",
  "itoa",
  "memchr",
  "ryu",
  "serde",
+ "serde_core",
 ]
 
 [[package]]
 name = "serde_path_to_error"
-version = "0.1.17"
+version = "0.1.20"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "59fab13f937fa393d08645bf3a84bdfe86e296747b506ada67bb15f10f218b2a"
+checksum = "10a9ff822e371bb5403e391ecd83e182e0e77ba7f6fe0160b795797109d1b457"
 dependencies = [
  "itoa",
  "serde",
+ "serde_core",
 ]
 
 [[package]]
 name = "serde_spanned"
-version = "1.0.0"
+version = "1.0.1"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "40734c41988f7306bb04f0ecf60ec0f3f1caa34290e4e8ea471dcd3346483b83"
+checksum = "2789234a13a53fc4be1b51ea1bab45a3c338bdb884862a257d10e5a74ae009e6"
 dependencies = [
- "serde",
+ "serde_core",
 ]
 
 [[package]]
@@ -3723,7 +3735,7 @@ dependencies = [
  "snarkos-node-tcp",
  "snarkvm",
  "tikv-jemallocator",
- "toml 0.9.5",
+ "toml 0.9.6",
  "tracing",
  "walkdir",
 ]
@@ -4126,8 +4138,8 @@ dependencies = [
 
 [[package]]
 name = "snarkvm"
-version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=595af97169519#595af971695196fe7587341c8e347122422acb04"
+version = "4.2.1"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=35c82646eeb2e9561be#35c82646eeb2e9561be9b8d549d1f7532dfe2b22"
 dependencies = [
  "anstyle",
  "anyhow",
@@ -4151,8 +4163,8 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-algorithms"
-version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=595af97169519#595af971695196fe7587341c8e347122422acb04"
+version = "4.2.1"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=35c82646eeb2e9561be#35c82646eeb2e9561be9b8d549d1f7532dfe2b22"
 dependencies = [
  "aleo-std",
  "anyhow",
@@ -4179,8 +4191,8 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-algorithms-cuda"
-version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=595af97169519#595af971695196fe7587341c8e347122422acb04"
+version = "4.2.1"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=35c82646eeb2e9561be#35c82646eeb2e9561be9b8d549d1f7532dfe2b22"
 dependencies = [
  "blst",
  "cc",
@@ -4190,8 +4202,8 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-circuit"
-version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=595af97169519#595af971695196fe7587341c8e347122422acb04"
+version = "4.2.1"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=35c82646eeb2e9561be#35c82646eeb2e9561be9b8d549d1f7532dfe2b22"
 dependencies = [
  "snarkvm-circuit-account",
  "snarkvm-circuit-algorithms",
@@ -4204,8 +4216,8 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-circuit-account"
-version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=595af97169519#595af971695196fe7587341c8e347122422acb04"
+version = "4.2.1"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=35c82646eeb2e9561be#35c82646eeb2e9561be9b8d549d1f7532dfe2b22"
 dependencies = [
  "snarkvm-circuit-network",
  "snarkvm-circuit-types",
@@ -4214,8 +4226,8 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-circuit-algorithms"
-version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=595af97169519#595af971695196fe7587341c8e347122422acb04"
+version = "4.2.1"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=35c82646eeb2e9561be#35c82646eeb2e9561be9b8d549d1f7532dfe2b22"
 dependencies = [
  "snarkvm-circuit-types",
  "snarkvm-console-algorithms",
@@ -4224,8 +4236,8 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-circuit-collections"
-version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=595af97169519#595af971695196fe7587341c8e347122422acb04"
+version = "4.2.1"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=35c82646eeb2e9561be#35c82646eeb2e9561be9b8d549d1f7532dfe2b22"
 dependencies = [
  "snarkvm-circuit-algorithms",
  "snarkvm-circuit-types",
@@ -4234,8 +4246,8 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-circuit-environment"
-version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=595af97169519#595af971695196fe7587341c8e347122422acb04"
+version = "4.2.1"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=35c82646eeb2e9561be#35c82646eeb2e9561be9b8d549d1f7532dfe2b22"
 dependencies = [
  "indexmap 2.11.1",
  "itertools 0.14.0",
@@ -4252,13 +4264,13 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-circuit-environment-witness"
-version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=595af97169519#595af971695196fe7587341c8e347122422acb04"
+version = "4.2.1"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=35c82646eeb2e9561be#35c82646eeb2e9561be9b8d549d1f7532dfe2b22"
 
 [[package]]
 name = "snarkvm-circuit-network"
-version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=595af97169519#595af971695196fe7587341c8e347122422acb04"
+version = "4.2.1"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=35c82646eeb2e9561be#35c82646eeb2e9561be9b8d549d1f7532dfe2b22"
 dependencies = [
  "snarkvm-circuit-algorithms",
  "snarkvm-circuit-collections",
@@ -4268,8 +4280,8 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-circuit-program"
-version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=595af97169519#595af971695196fe7587341c8e347122422acb04"
+version = "4.2.1"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=35c82646eeb2e9561be#35c82646eeb2e9561be9b8d549d1f7532dfe2b22"
 dependencies = [
  "snarkvm-circuit-account",
  "snarkvm-circuit-algorithms",
@@ -4282,8 +4294,8 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-circuit-types"
-version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=595af97169519#595af971695196fe7587341c8e347122422acb04"
+version = "4.2.1"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=35c82646eeb2e9561be#35c82646eeb2e9561be9b8d549d1f7532dfe2b22"
 dependencies = [
  "snarkvm-circuit-environment",
  "snarkvm-circuit-types-address",
@@ -4297,8 +4309,8 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-circuit-types-address"
-version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=595af97169519#595af971695196fe7587341c8e347122422acb04"
+version = "4.2.1"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=35c82646eeb2e9561be#35c82646eeb2e9561be9b8d549d1f7532dfe2b22"
 dependencies = [
  "snarkvm-circuit-environment",
  "snarkvm-circuit-types-boolean",
@@ -4310,8 +4322,8 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-circuit-types-boolean"
-version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=595af97169519#595af971695196fe7587341c8e347122422acb04"
+version = "4.2.1"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=35c82646eeb2e9561be#35c82646eeb2e9561be9b8d549d1f7532dfe2b22"
 dependencies = [
  "snarkvm-circuit-environment",
  "snarkvm-console-types-boolean",
@@ -4319,8 +4331,8 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-circuit-types-field"
-version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=595af97169519#595af971695196fe7587341c8e347122422acb04"
+version = "4.2.1"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=35c82646eeb2e9561be#35c82646eeb2e9561be9b8d549d1f7532dfe2b22"
 dependencies = [
  "snarkvm-circuit-environment",
  "snarkvm-circuit-types-boolean",
@@ -4329,8 +4341,8 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-circuit-types-group"
-version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=595af97169519#595af971695196fe7587341c8e347122422acb04"
+version = "4.2.1"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=35c82646eeb2e9561be#35c82646eeb2e9561be9b8d549d1f7532dfe2b22"
 dependencies = [
  "snarkvm-circuit-environment",
  "snarkvm-circuit-types-boolean",
@@ -4341,8 +4353,8 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-circuit-types-integers"
-version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=595af97169519#595af971695196fe7587341c8e347122422acb04"
+version = "4.2.1"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=35c82646eeb2e9561be#35c82646eeb2e9561be9b8d549d1f7532dfe2b22"
 dependencies = [
  "snarkvm-circuit-environment",
  "snarkvm-circuit-types-boolean",
@@ -4353,8 +4365,8 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-circuit-types-scalar"
-version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=595af97169519#595af971695196fe7587341c8e347122422acb04"
+version = "4.2.1"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=35c82646eeb2e9561be#35c82646eeb2e9561be9b8d549d1f7532dfe2b22"
 dependencies = [
  "snarkvm-circuit-environment",
  "snarkvm-circuit-types-boolean",
@@ -4364,8 +4376,8 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-circuit-types-string"
-version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=595af97169519#595af971695196fe7587341c8e347122422acb04"
+version = "4.2.1"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=35c82646eeb2e9561be#35c82646eeb2e9561be9b8d549d1f7532dfe2b22"
 dependencies = [
  "snarkvm-circuit-environment",
  "snarkvm-circuit-types-boolean",
@@ -4376,8 +4388,8 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-console"
-version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=595af97169519#595af971695196fe7587341c8e347122422acb04"
+version = "4.2.1"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=35c82646eeb2e9561be#35c82646eeb2e9561be9b8d549d1f7532dfe2b22"
 dependencies = [
  "snarkvm-console-account",
  "snarkvm-console-algorithms",
@@ -4389,8 +4401,8 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-console-account"
-version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=595af97169519#595af971695196fe7587341c8e347122422acb04"
+version = "4.2.1"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=35c82646eeb2e9561be#35c82646eeb2e9561be9b8d549d1f7532dfe2b22"
 dependencies = [
  "bs58",
  "snarkvm-console-network",
@@ -4400,8 +4412,8 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-console-algorithms"
-version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=595af97169519#595af971695196fe7587341c8e347122422acb04"
+version = "4.2.1"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=35c82646eeb2e9561be#35c82646eeb2e9561be9b8d549d1f7532dfe2b22"
 dependencies = [
  "blake2s_simd",
  "smallvec",
@@ -4413,8 +4425,8 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-console-collections"
-version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=595af97169519#595af971695196fe7587341c8e347122422acb04"
+version = "4.2.1"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=35c82646eeb2e9561be#35c82646eeb2e9561be9b8d549d1f7532dfe2b22"
 dependencies = [
  "aleo-std",
  "rayon",
@@ -4424,8 +4436,8 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-console-network"
-version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=595af97169519#595af971695196fe7587341c8e347122422acb04"
+version = "4.2.1"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=35c82646eeb2e9561be#35c82646eeb2e9561be9b8d549d1f7532dfe2b22"
 dependencies = [
  "anyhow",
  "enum-iterator",
@@ -4444,8 +4456,8 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-console-network-environment"
-version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=595af97169519#595af971695196fe7587341c8e347122422acb04"
+version = "4.2.1"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=35c82646eeb2e9561be#35c82646eeb2e9561be9b8d549d1f7532dfe2b22"
 dependencies = [
  "anyhow",
  "bech32",
@@ -4462,8 +4474,8 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-console-program"
-version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=595af97169519#595af971695196fe7587341c8e347122422acb04"
+version = "4.2.1"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=35c82646eeb2e9561be#35c82646eeb2e9561be9b8d549d1f7532dfe2b22"
 dependencies = [
  "enum-iterator",
  "enum_index",
@@ -4482,8 +4494,8 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-console-types"
-version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=595af97169519#595af971695196fe7587341c8e347122422acb04"
+version = "4.2.1"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=35c82646eeb2e9561be#35c82646eeb2e9561be9b8d549d1f7532dfe2b22"
 dependencies = [
  "snarkvm-console-network-environment",
  "snarkvm-console-types-address",
@@ -4497,8 +4509,8 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-console-types-address"
-version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=595af97169519#595af971695196fe7587341c8e347122422acb04"
+version = "4.2.1"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=35c82646eeb2e9561be#35c82646eeb2e9561be9b8d549d1f7532dfe2b22"
 dependencies = [
  "snarkvm-console-network-environment",
  "snarkvm-console-types-boolean",
@@ -4508,16 +4520,16 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-console-types-boolean"
-version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=595af97169519#595af971695196fe7587341c8e347122422acb04"
+version = "4.2.1"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=35c82646eeb2e9561be#35c82646eeb2e9561be9b8d549d1f7532dfe2b22"
 dependencies = [
  "snarkvm-console-network-environment",
 ]
 
 [[package]]
 name = "snarkvm-console-types-field"
-version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=595af97169519#595af971695196fe7587341c8e347122422acb04"
+version = "4.2.1"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=35c82646eeb2e9561be#35c82646eeb2e9561be9b8d549d1f7532dfe2b22"
 dependencies = [
  "snarkvm-console-network-environment",
  "snarkvm-console-types-boolean",
@@ -4526,8 +4538,8 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-console-types-group"
-version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=595af97169519#595af971695196fe7587341c8e347122422acb04"
+version = "4.2.1"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=35c82646eeb2e9561be#35c82646eeb2e9561be9b8d549d1f7532dfe2b22"
 dependencies = [
  "snarkvm-console-network-environment",
  "snarkvm-console-types-boolean",
@@ -4537,8 +4549,8 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-console-types-integers"
-version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=595af97169519#595af971695196fe7587341c8e347122422acb04"
+version = "4.2.1"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=35c82646eeb2e9561be#35c82646eeb2e9561be9b8d549d1f7532dfe2b22"
 dependencies = [
  "snarkvm-console-network-environment",
  "snarkvm-console-types-boolean",
@@ -4548,8 +4560,8 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-console-types-scalar"
-version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=595af97169519#595af971695196fe7587341c8e347122422acb04"
+version = "4.2.1"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=35c82646eeb2e9561be#35c82646eeb2e9561be9b8d549d1f7532dfe2b22"
 dependencies = [
  "snarkvm-console-network-environment",
  "snarkvm-console-types-boolean",
@@ -4559,8 +4571,8 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-console-types-string"
-version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=595af97169519#595af971695196fe7587341c8e347122422acb04"
+version = "4.2.1"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=35c82646eeb2e9561be#35c82646eeb2e9561be9b8d549d1f7532dfe2b22"
 dependencies = [
  "snarkvm-console-network-environment",
  "snarkvm-console-types-boolean",
@@ -4570,8 +4582,8 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-curves"
-version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=595af97169519#595af971695196fe7587341c8e347122422acb04"
+version = "4.2.1"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=35c82646eeb2e9561be#35c82646eeb2e9561be9b8d549d1f7532dfe2b22"
 dependencies = [
  "rand 0.8.5",
  "rayon",
@@ -4584,8 +4596,8 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-fields"
-version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=595af97169519#595af971695196fe7587341c8e347122422acb04"
+version = "4.2.1"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=35c82646eeb2e9561be#35c82646eeb2e9561be9b8d549d1f7532dfe2b22"
 dependencies = [
  "aleo-std",
  "anyhow",
@@ -4601,8 +4613,8 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-ledger"
-version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=595af97169519#595af971695196fe7587341c8e347122422acb04"
+version = "4.2.1"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=35c82646eeb2e9561be#35c82646eeb2e9561be9b8d549d1f7532dfe2b22"
 dependencies = [
  "aleo-std",
  "anyhow",
@@ -4628,8 +4640,8 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-ledger-authority"
-version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=595af97169519#595af971695196fe7587341c8e347122422acb04"
+version = "4.2.1"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=35c82646eeb2e9561be#35c82646eeb2e9561be9b8d549d1f7532dfe2b22"
 dependencies = [
  "anyhow",
  "rand 0.8.5",
@@ -4640,8 +4652,8 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-ledger-block"
-version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=595af97169519#595af971695196fe7587341c8e347122422acb04"
+version = "4.2.1"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=35c82646eeb2e9561be#35c82646eeb2e9561be9b8d549d1f7532dfe2b22"
 dependencies = [
  "anyhow",
  "indexmap 2.11.1",
@@ -4662,8 +4674,8 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-ledger-committee"
-version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=595af97169519#595af971695196fe7587341c8e347122422acb04"
+version = "4.2.1"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=35c82646eeb2e9561be#35c82646eeb2e9561be9b8d549d1f7532dfe2b22"
 dependencies = [
  "anyhow",
  "indexmap 2.11.1",
@@ -4681,8 +4693,8 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-ledger-narwhal"
-version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=595af97169519#595af971695196fe7587341c8e347122422acb04"
+version = "4.2.1"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=35c82646eeb2e9561be#35c82646eeb2e9561be9b8d549d1f7532dfe2b22"
 dependencies = [
  "snarkvm-ledger-narwhal-batch-certificate",
  "snarkvm-ledger-narwhal-batch-header",
@@ -4694,8 +4706,8 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-ledger-narwhal-batch-certificate"
-version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=595af97169519#595af971695196fe7587341c8e347122422acb04"
+version = "4.2.1"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=35c82646eeb2e9561be#35c82646eeb2e9561be9b8d549d1f7532dfe2b22"
 dependencies = [
  "indexmap 2.11.1",
  "rayon",
@@ -4707,8 +4719,8 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-ledger-narwhal-batch-header"
-version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=595af97169519#595af971695196fe7587341c8e347122422acb04"
+version = "4.2.1"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=35c82646eeb2e9561be#35c82646eeb2e9561be9b8d549d1f7532dfe2b22"
 dependencies = [
  "indexmap 2.11.1",
  "rayon",
@@ -4720,8 +4732,8 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-ledger-narwhal-data"
-version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=595af97169519#595af971695196fe7587341c8e347122422acb04"
+version = "4.2.1"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=35c82646eeb2e9561be#35c82646eeb2e9561be9b8d549d1f7532dfe2b22"
 dependencies = [
  "bytes",
  "serde_json",
@@ -4731,8 +4743,8 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-ledger-narwhal-subdag"
-version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=595af97169519#595af971695196fe7587341c8e347122422acb04"
+version = "4.2.1"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=35c82646eeb2e9561be#35c82646eeb2e9561be9b8d549d1f7532dfe2b22"
 dependencies = [
  "indexmap 2.11.1",
  "rayon",
@@ -4746,8 +4758,8 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-ledger-narwhal-transmission"
-version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=595af97169519#595af971695196fe7587341c8e347122422acb04"
+version = "4.2.1"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=35c82646eeb2e9561be#35c82646eeb2e9561be9b8d549d1f7532dfe2b22"
 dependencies = [
  "bytes",
  "serde_json",
@@ -4759,8 +4771,8 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-ledger-narwhal-transmission-id"
-version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=595af97169519#595af971695196fe7587341c8e347122422acb04"
+version = "4.2.1"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=35c82646eeb2e9561be#35c82646eeb2e9561be9b8d549d1f7532dfe2b22"
 dependencies = [
  "snarkvm-console",
  "snarkvm-ledger-puzzle",
@@ -4768,8 +4780,8 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-ledger-puzzle"
-version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=595af97169519#595af971695196fe7587341c8e347122422acb04"
+version = "4.2.1"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=35c82646eeb2e9561be#35c82646eeb2e9561be9b8d549d1f7532dfe2b22"
 dependencies = [
  "aleo-std",
  "anyhow",
@@ -4788,8 +4800,8 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-ledger-puzzle-epoch"
-version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=595af97169519#595af971695196fe7587341c8e347122422acb04"
+version = "4.2.1"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=35c82646eeb2e9561be#35c82646eeb2e9561be9b8d549d1f7532dfe2b22"
 dependencies = [
  "aleo-std",
  "anyhow",
@@ -4811,8 +4823,8 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-ledger-query"
-version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=595af97169519#595af971695196fe7587341c8e347122422acb04"
+version = "4.2.1"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=35c82646eeb2e9561be#35c82646eeb2e9561be9b8d549d1f7532dfe2b22"
 dependencies = [
  "anyhow",
  "async-trait",
@@ -4828,8 +4840,8 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-ledger-store"
-version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=595af97169519#595af971695196fe7587341c8e347122422acb04"
+version = "4.2.1"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=35c82646eeb2e9561be#35c82646eeb2e9561be9b8d549d1f7532dfe2b22"
 dependencies = [
  "aleo-std-storage",
  "anyhow",
@@ -4855,8 +4867,8 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-ledger-test-helpers"
-version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=595af97169519#595af971695196fe7587341c8e347122422acb04"
+version = "4.2.1"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=35c82646eeb2e9561be#35c82646eeb2e9561be9b8d549d1f7532dfe2b22"
 dependencies = [
  "aleo-std",
  "anyhow",
@@ -4873,16 +4885,16 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-metrics"
-version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=595af97169519#595af971695196fe7587341c8e347122422acb04"
+version = "4.2.1"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=35c82646eeb2e9561be#35c82646eeb2e9561be9b8d549d1f7532dfe2b22"
 dependencies = [
  "metrics",
 ]
 
 [[package]]
 name = "snarkvm-parameters"
-version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=595af97169519#595af971695196fe7587341c8e347122422acb04"
+version = "4.2.1"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=35c82646eeb2e9561be#35c82646eeb2e9561be9b8d549d1f7532dfe2b22"
 dependencies = [
  "aleo-std",
  "anyhow",
@@ -4904,8 +4916,8 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-synthesizer"
-version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=595af97169519#595af971695196fe7587341c8e347122422acb04"
+version = "4.2.1"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=35c82646eeb2e9561be#35c82646eeb2e9561be9b8d549d1f7532dfe2b22"
 dependencies = [
  "aleo-std",
  "anyhow",
@@ -4937,8 +4949,8 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-synthesizer-process"
-version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=595af97169519#595af971695196fe7587341c8e347122422acb04"
+version = "4.2.1"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=35c82646eeb2e9561be#35c82646eeb2e9561be9b8d549d1f7532dfe2b22"
 dependencies = [
  "aleo-std",
  "colored 3.0.0",
@@ -4962,8 +4974,8 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-synthesizer-program"
-version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=595af97169519#595af971695196fe7587341c8e347122422acb04"
+version = "4.2.1"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=35c82646eeb2e9561be#35c82646eeb2e9561be9b8d549d1f7532dfe2b22"
 dependencies = [
  "indexmap 2.11.1",
  "paste",
@@ -4980,8 +4992,8 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-synthesizer-snark"
-version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=595af97169519#595af971695196fe7587341c8e347122422acb04"
+version = "4.2.1"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=35c82646eeb2e9561be#35c82646eeb2e9561be9b8d549d1f7532dfe2b22"
 dependencies = [
  "bincode",
  "serde_json",
@@ -4993,8 +5005,8 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-utilities"
-version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=595af97169519#595af971695196fe7587341c8e347122422acb04"
+version = "4.2.1"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=35c82646eeb2e9561be#35c82646eeb2e9561be9b8d549d1f7532dfe2b22"
 dependencies = [
  "aleo-std",
  "anyhow",
@@ -5015,8 +5027,8 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-utilities-derives"
-version = "4.1.0"
-source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=595af97169519#595af971695196fe7587341c8e347122422acb04"
+version = "4.2.1"
+source = "git+https://github.com/ProvableHQ/snarkVM.git?rev=35c82646eeb2e9561be#35c82646eeb2e9561be9b8d549d1f7532dfe2b22"
 dependencies = [
  "proc-macro2",
  "quote 1.0.40",
@@ -5542,12 +5554,12 @@ dependencies = [
 
 [[package]]
 name = "toml"
-version = "0.9.5"
+version = "0.9.6"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "75129e1dc5000bfbaa9fee9d1b21f974f9fbad9daec557a521ee6e080825f6e8"
+checksum = "ae2a4cf385da23d1d53bc15cdfa5c2109e93d8d362393c801e87da2f72f0e201"
 dependencies = [
  "indexmap 2.11.1",
- "serde",
+ "serde_core",
  "serde_spanned",
  "toml_datetime",
  "toml_parser",
@@ -5557,11 +5569,11 @@ dependencies = [
 
 [[package]]
 name = "toml_datetime"
-version = "0.7.0"
+version = "0.7.1"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "bade1c3e902f58d73d3f294cd7f20391c1cb2fbcb643b73566bc773971df91e3"
+checksum = "a197c0ec7d131bfc6f7e82c8442ba1595aeab35da7adbf05b6b73cd06a16b6be"
 dependencies = [
- "serde",
+ "serde_core",
 ]
 
 [[package]]
@@ -5953,18 +5965,18 @@ checksum = "ccf3ec651a847eb01de73ccad15eb7d99f80485de043efb2f370cd654f4ea44b"
 
 [[package]]
 name = "wasi"
-version = "0.14.5+wasi-0.2.4"
+version = "0.14.6+wasi-0.2.4"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "a4494f6290a82f5fe584817a676a34b9d6763e8d9d18204009fb31dceca98fd4"
+checksum = "7f71243a3f320c00a8459e455c046ce571229c2f31fd11645d9dc095e3068ca0"
 dependencies = [
  "wasip2",
 ]
 
 [[package]]
 name = "wasip2"
-version = "1.0.0+wasi-0.2.4"
+version = "1.0.1+wasi-0.2.4"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "03fa2761397e5bd52002cd7e73110c71af2109aca4e521a9f40473fe685b0a24"
+checksum = "0562428422c63773dad2c345a1882263bbf4d65cf3f42e90921f787ef5ad58e7"
 dependencies = [
  "wit-bindgen",
 ]
@@ -6126,15 +6138,15 @@ checksum = "712e227841d057c1ee1cd2fb22fa7e5a5461ae8e48fa2ca79ec42cfc1931183f"
 
 [[package]]
 name = "windows-core"
-version = "0.61.2"
+version = "0.62.0"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "c0fdd3ddb90610c7638aa2b3a3ab2904fb9e5cdbecc643ddb3647212781c4ae3"
+checksum = "57fe7168f7de578d2d8a05b07fd61870d2e73b4020e9f49aa00da8471723497c"
 dependencies = [
  "windows-implement",
  "windows-interface",
- "windows-link 0.1.3",
- "windows-result",
- "windows-strings",
+ "windows-link 0.2.0",
+ "windows-result 0.4.0",
+ "windows-strings 0.5.0",
 ]
 
 [[package]]
@@ -6178,8 +6190,8 @@ source = "registry+https://github.com/rust-lang/crates.io-index"
 checksum = "5b8a9ed28765efc97bbc954883f4e6796c33a06546ebafacbabee9696967499e"
 dependencies = [
  "windows-link 0.1.3",
- "windows-result",
- "windows-strings",
+ "windows-result 0.3.4",
+ "windows-strings 0.4.2",
 ]
 
 [[package]]
@@ -6191,6 +6203,15 @@ dependencies = [
  "windows-link 0.1.3",
 ]
 
+[[package]]
+name = "windows-result"
+version = "0.4.0"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "7084dcc306f89883455a206237404d3eaf961e5bd7e0f312f7c91f57eb44167f"
+dependencies = [
+ "windows-link 0.2.0",
+]
+
 [[package]]
 name = "windows-strings"
 version = "0.4.2"
@@ -6200,6 +6221,15 @@ dependencies = [
  "windows-link 0.1.3",
 ]
 
+[[package]]
+name = "windows-strings"
+version = "0.5.0"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "7218c655a553b0bed4426cf54b20d7ba363ef543b52d515b3e48d7fd55318dda"
+dependencies = [
+ "windows-link 0.2.0",
+]
+
 [[package]]
 name = "windows-sys"
 version = "0.52.0"
@@ -6379,9 +6409,9 @@ checksum = "d135d17ab770252ad95e9a872d365cf3090e3be864a34ab46f48555993efc904"
 
 [[package]]
 name = "wit-bindgen"
-version = "0.45.1"
+version = "0.46.0"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "5c573471f125075647d03df72e026074b7203790d41351cd6edc96f46bcccd36"
+checksum = "f17a85883d4e6d00e8a97c586de764dabcc06133f7f1d55dce5cdc070ad7fe59"
 
 [[package]]
 name = "writeable"
diff --git a/Cargo.toml b/Cargo.toml
index c60159df6..e922558d3 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -46,9 +46,10 @@ default-features = false
 [workspace.dependencies.snarkvm]
 #path = "../snarkVM"
 git = "https://github.com/ProvableHQ/snarkVM.git"
-rev = "595af97169519"
+rev = "35c82646eeb2e9561be"
 #version = "=4.2.1"
 default-features = false
+#features = [ "circuit", "console", "rocks" ]
 
 [workspace.dependencies.anyhow]
 version = "1.0"
diff --git a/cli/src/helpers/logger.rs b/cli/src/helpers/logger.rs
index 006cadc42..4c4c652fb 100644
--- a/cli/src/helpers/logger.rs
+++ b/cli/src/helpers/logger.rs
@@ -107,9 +107,7 @@ fn parse_log_filter(filter_str: &str) -> Result<EnvFilter> {
 }
 
 /// Sets the log filter based on the given verbosity level.
-/// Initializes the logger with the specified verbosity level, where 0 is the lowest verbosity and 6 the highest.
 ///
-/// The following shows what messages are enabled at each level.
 /// ```ignore
 /// 0 => info
 /// 1 => info, debug
diff --git a/node/bft/ledger-service/src/ledger.rs b/node/bft/ledger-service/src/ledger.rs
index b01e13998..bdf506e32 100644
--- a/node/bft/ledger-service/src/ledger.rs
+++ b/node/bft/ledger-service/src/ledger.rs
@@ -16,10 +16,8 @@
 use crate::{LedgerService, fmt_id, spawn_blocking};
 use snarkvm::{
     ledger::{
-        Block,
         Ledger,
-        PendingBlock,
-        Transaction,
+        block::{Block, Transaction},
         committee::Committee,
         narwhal::{BatchCertificate, Data, Subdag, Transmission, TransmissionID},
         puzzle::{Solution, SolutionID},
@@ -354,14 +352,6 @@ impl<N: Network, C: ConsensusStorage<N>> LedgerService<N> for CoreLedgerService<
         spawn_blocking!(ledger.check_transaction_basic(&transaction, None, &mut rand::thread_rng()))
     }
 
-    fn check_block_subdag(&self, block: Block<N>, prefix: &[PendingBlock<N>]) -> Result<PendingBlock<N>> {
-        self.ledger.check_block_subdag(block, prefix)
-    }
-
-    fn check_block_content(&self, block: PendingBlock<N>) -> Result<Block<N>> {
-        self.ledger.check_block_content(block, &mut rand::thread_rng())
-    }
-
     /// Checks the given block is valid next block.
     fn check_next_block(&self, block: &Block<N>) -> Result<()> {
         self.ledger.check_next_block(block, &mut rand::thread_rng())
diff --git a/node/bft/ledger-service/src/mock.rs b/node/bft/ledger-service/src/mock.rs
index ec0288dab..94154d6c3 100644
--- a/node/bft/ledger-service/src/mock.rs
+++ b/node/bft/ledger-service/src/mock.rs
@@ -16,9 +16,7 @@
 use crate::{LedgerService, fmt_id};
 use snarkvm::{
     ledger::{
-        Block,
-        PendingBlock,
-        Transaction,
+        block::{Block, Transaction},
         committee::Committee,
         narwhal::{BatchCertificate, Data, Subdag, Transmission, TransmissionID},
         puzzle::{Solution, SolutionID},
@@ -213,14 +211,6 @@ impl<N: Network> LedgerService<N> for MockLedgerService<N> {
         Ok(())
     }
 
-    fn check_block_subdag(&self, _block: Block<N>, _prefix: &[PendingBlock<N>]) -> Result<PendingBlock<N>> {
-        unimplemented!();
-    }
-
-    fn check_block_content(&self, _block: PendingBlock<N>) -> Result<Block<N>> {
-        unimplemented!();
-    }
-
     /// Checks the given block is valid next block.
     fn check_next_block(&self, _block: &Block<N>) -> Result<()> {
         Ok(())
diff --git a/node/bft/ledger-service/src/prover.rs b/node/bft/ledger-service/src/prover.rs
index 17bee1499..e5f2523e6 100644
--- a/node/bft/ledger-service/src/prover.rs
+++ b/node/bft/ledger-service/src/prover.rs
@@ -16,9 +16,7 @@
 use crate::LedgerService;
 use snarkvm::{
     ledger::{
-        Block,
-        PendingBlock,
-        Transaction,
+        block::{Block, Transaction},
         committee::Committee,
         narwhal::{BatchCertificate, Data, Subdag, Transmission, TransmissionID},
         puzzle::{Solution, SolutionID},
@@ -168,14 +166,6 @@ impl<N: Network> LedgerService<N> for ProverLedgerService<N> {
         Ok(())
     }
 
-    fn check_block_subdag(&self, _block: Block<N>, _prefix: &[PendingBlock<N>]) -> Result<PendingBlock<N>> {
-        bail!("Cannot check block subDAG in prover")
-    }
-
-    fn check_block_content(&self, _pending_block: PendingBlock<N>) -> Result<Block<N>> {
-        bail!("Cannot check block content in prover")
-    }
-
     /// Checks the given block is valid next block.
     fn check_next_block(&self, _block: &Block<N>) -> Result<()> {
         Ok(())
diff --git a/node/bft/ledger-service/src/traits.rs b/node/bft/ledger-service/src/traits.rs
index 2471194bc..02b1f2d7e 100644
--- a/node/bft/ledger-service/src/traits.rs
+++ b/node/bft/ledger-service/src/traits.rs
@@ -15,9 +15,7 @@
 
 use snarkvm::{
     ledger::{
-        Block,
-        PendingBlock,
-        Transaction,
+        block::{Block, Transaction},
         committee::Committee,
         narwhal::{BatchCertificate, Data, Subdag, Transmission, TransmissionID},
         puzzle::{Solution, SolutionID},
@@ -108,12 +106,6 @@ pub trait LedgerService<N: Network>: Debug + Send + Sync {
         transaction: Transaction<N>,
     ) -> Result<()>;
 
-    /// Checks that the subDAG in a given block is valid, but does not fully verify the block.
-    fn check_block_subdag(&self, block: Block<N>, prefix: &[PendingBlock<N>]) -> Result<PendingBlock<N>>;
-
-    /// Takes a pending block and performs the remaining checks to full verify it.
-    fn check_block_content(&self, _block: PendingBlock<N>) -> Result<Block<N>>;
-
     /// Checks the given block is valid next block.
     fn check_next_block(&self, block: &Block<N>) -> Result<()>;
 
diff --git a/node/bft/ledger-service/src/translucent.rs b/node/bft/ledger-service/src/translucent.rs
index 33eb4afda..c6f48a41e 100644
--- a/node/bft/ledger-service/src/translucent.rs
+++ b/node/bft/ledger-service/src/translucent.rs
@@ -18,10 +18,8 @@ use async_trait::async_trait;
 use indexmap::IndexMap;
 use snarkvm::{
     ledger::{
-        Block,
         Ledger,
-        PendingBlock,
-        Transaction,
+        block::{Block, Transaction},
         committee::Committee,
         narwhal::{Data, Subdag, Transmission, TransmissionID},
         puzzle::{Solution, SolutionID},
@@ -179,14 +177,6 @@ impl<N: Network, C: ConsensusStorage<N>> LedgerService<N> for TranslucentLedgerS
         Ok(())
     }
 
-    fn check_block_subdag(&self, _block: Block<N>, _prefix: &[PendingBlock<N>]) -> Result<PendingBlock<N>> {
-        unimplemented!();
-    }
-
-    fn check_block_content(&self, _block: PendingBlock<N>) -> Result<Block<N>> {
-        unimplemented!();
-    }
-
     /// Always succeeds.
     fn check_next_block(&self, _block: &Block<N>) -> Result<()> {
         Ok(())
diff --git a/node/bft/src/sync/mod.rs b/node/bft/src/sync/mod.rs
index be6aac984..ec3177fb7 100644
--- a/node/bft/src/sync/mod.rs
+++ b/node/bft/src/sync/mod.rs
@@ -27,11 +27,11 @@ use snarkos_node_bft_ledger_service::LedgerService;
 use snarkos_node_sync::{BLOCK_REQUEST_BATCH_DELAY, BlockSync, Ping, PrepareSyncRequest, locators::BlockLocators};
 use snarkvm::{
     console::{network::Network, types::Field},
-    ledger::{PendingBlock, authority::Authority, block::Block, narwhal::BatchCertificate},
+    ledger::{authority::Authority, block::Block, narwhal::BatchCertificate},
     prelude::{cfg_into_iter, cfg_iter},
 };
 
-use anyhow::{Result, anyhow, bail, ensure};
+use anyhow::{Result, anyhow, bail};
 use indexmap::IndexMap;
 #[cfg(feature = "locktick")]
 use locktick::{parking_lot::Mutex, tokio::Mutex as TMutex};
@@ -40,7 +40,7 @@ use parking_lot::Mutex;
 #[cfg(not(feature = "serial"))]
 use rayon::prelude::*;
 use std::{
-    collections::{HashMap, VecDeque},
+    collections::{BTreeMap, HashMap},
     future::Future,
     net::SocketAddr,
     sync::Arc,
@@ -87,12 +87,12 @@ pub struct Sync<N: Network> {
     sync_lock: Arc<TMutex<()>>,
     /// The latest block responses.
     ///
-    /// This is used in [`Sync::sync_storage_with_block()`] to accumulate blocks
-    /// whose addition to the ledger is deferred until certain checks pass.
+    /// This is used in [`Sync::sync_storage_with_block()`] to accumulate blocks  whose addition to the ledger is
+    /// deferred until certain checks pass.
     /// Blocks need to be processed in order, hence a BTree map.
     ///
     /// Whenever a new block is added to this map, BlockSync::set_sync_height needs to be called.
-    pending_blocks: Arc<TMutex<VecDeque<PendingBlock<N>>>>,
+    latest_block_responses: Arc<TMutex<BTreeMap<u32, Block<N>>>>,
 }
 
 impl<N: Network> Sync<N> {
@@ -114,7 +114,7 @@ impl<N: Network> Sync<N> {
             handles: Default::default(),
             response_lock: Default::default(),
             sync_lock: Default::default(),
-            pending_blocks: Default::default(),
+            latest_block_responses: Default::default(),
         }
     }
 
@@ -446,17 +446,13 @@ impl<N: Network> Sync<N> {
     /// If there are queued block responses, this might be higher than the latest block in the ledger.
     async fn compute_sync_height(&self) -> u32 {
         let ledger_height = self.ledger.latest_block_height();
-        let mut pending_blocks = self.pending_blocks.lock().await;
+        let mut responses = self.latest_block_responses.lock().await;
 
         // Remove any old responses.
-        while let Some(b) = pending_blocks.front()
-            && b.height() <= ledger_height
-        {
-            pending_blocks.pop_front();
-        }
+        responses.retain(|height, _| *height > ledger_height);
 
         // Ensure the returned value is always greater or equal than ledger height.
-        pending_blocks.back().map(|b| b.height()).unwrap_or(0).max(ledger_height)
+        responses.last_key_value().map(|(height, _)| *height).unwrap_or(0).max(ledger_height)
     }
 
     /// Aims to advance synchronization using any recent block responses received from peers.
@@ -467,7 +463,7 @@ impl<N: Network> Sync<N> {
     /// This returns Ok(true) if we successfully advanced the ledger by at least one new block.
     ///
     /// A key difference to `BlockSync`'s versions is that it will only add blocks to the ledger once they have been confirmed by the network.
-    /// If blocks are not confirmed yet, they will be kept in [`Self::pending_blocks`].
+    /// If blocks are not confirmed yet, they will be kept in [`Self::latest_block_responses`].
     /// It will also pass certificates from synced blocks to the BFT module so that consensus can progress as expected
     /// (see [`Self::sync_storage_with_block`] for more details).
     ///
@@ -620,191 +616,216 @@ impl<N: Network> Sync<N> {
         .await?
     }
 
-    /// Helper function for [`Self::sync_storage_with_block`].
-    /// It syncs the batch certificates with the BFT, if the block's authority is a sub-DAG.
-    ///
-    /// Note that the block authority is always a sub-DAG in production; beacon signatures are only used for testing,
-    /// and as placeholder (irrelevant) block authority in the genesis block.i
-    async fn add_block_subdag_to_bft(&self, block: &Block<N>) -> Result<()> {
-        // Nothing to do if this is a beacon block
-        let Authority::Quorum(subdag) = block.authority() else {
-            return Ok(());
-        };
-
-        // Reconstruct the unconfirmed transactions.
-        let unconfirmed_transactions = cfg_iter!(block.transactions())
-            .filter_map(|tx| tx.to_unconfirmed_transaction().map(|unconfirmed| (unconfirmed.id(), unconfirmed)).ok())
-            .collect::<HashMap<_, _>>();
-
-        // Iterate over the certificates.
-        for certificates in subdag.values().cloned() {
-            cfg_into_iter!(certificates.clone()).for_each(|certificate| {
-                // Sync the batch certificate with the block.
-                self.storage.sync_certificate_with_block(block, certificate.clone(), &unconfirmed_transactions);
-            });
-
-            // Sync the BFT DAG with the certificates.
-            for certificate in certificates {
-                // If a BFT sender was provided, send the certificate to the BFT.
-                // For validators, BFT spawns a receiver task in `BFT::start_handlers`.
-                if let Some(bft_sender) = self.bft_sender.get() {
-                    // Await the callback to continue.
-                    if let Err(err) = bft_sender.send_sync_bft(certificate).await {
-                        bail!("Failed to sync certificate - {err}");
-                    };
-                }
-            }
-        }
-        Ok(())
-    }
-
-    /// Helper function for [`Self::sync_storage_with_block`].
-    ///
-    /// It checks that successor of a given block contains enough votes to commit it.
-    /// This can only return `Ok(true)` if the certificates of the block's successor were added to the storage.
-    fn is_block_availability_threshold_reached(&self, block: &PendingBlock<N>) -> Result<bool> {
-        // Fetch the leader certificate and the relevant rounds.
-        let leader_certificate = match block.authority() {
-            Authority::Quorum(subdag) => subdag.leader_certificate().clone(),
-            _ => bail!("Received a block with an unexpected authority type."),
-        };
-        let commit_round = leader_certificate.round();
-        let certificate_round =
-            commit_round.checked_add(1).ok_or_else(|| anyhow!("Integer overflow on round number"))?;
-
-        // Get the committee lookback for the round just after the leader.
-        let certificate_committee_lookback = self.ledger.get_committee_lookback_for_round(certificate_round)?;
-        // Retrieve all of the certificates for the round just after the leader.
-        let certificates = self.storage.get_certificates_for_round(certificate_round);
-        // Construct a set over the authors, at the round just after the leader,
-        // who included the leader's certificate in their previous certificate IDs.
-        let authors = certificates
-            .iter()
-            .filter_map(|c| match c.previous_certificate_ids().contains(&leader_certificate.id()) {
-                true => Some(c.author()),
-                false => None,
-            })
-            .collect();
-
-        // Check if the leader is ready to be committed.
-        if certificate_committee_lookback.is_availability_threshold_reached(&authors) {
-            trace!(
-                "Block {hash} at height {height} has reached availability threshold",
-                hash = block.hash(),
-                height = block.height()
-            );
-            Ok(true)
-        } else {
-            Ok(false)
-        }
-    }
-
     /// Advances the ledger by the given block and updates the storage accordingly.
     ///
     /// This also updates the DAG, and uses the DAG to ensure that the block's leader certificate
     /// meets the voter availability threshold (i.e. > f voting stake)
     /// or is reachable via a DAG path from a later leader certificate that does.
     /// Since performing this check requires DAG certificates from later blocks,
-    /// the block is stored in `Sync::pending_blocks`,
+    /// the block is stored in `Sync::latest_block_responses`,
     /// and its addition to the ledger is deferred until the check passes.
-    /// Several blocks may be stored in `Sync::pending_blocks`
+    /// Several blocks may be stored in `Sync::latest_block_responses`
     /// before they can be all checked and added to the ledger.
-    ///
-    /// # Usage
-    /// This function assumes that blocks are passed in order, i.e.,
-    /// that the given block is a direct successor of the block that was last passed to this function.
-    async fn sync_storage_with_block(&self, new_block: Block<N>) -> Result<()> {
+    async fn sync_storage_with_block(&self, block: Block<N>) -> Result<()> {
         // Acquire the sync lock.
         let _lock = self.sync_lock.lock().await;
 
         // If this block has already been processed, return early.
         // TODO(kaimast): Should we remove the response here?
-        if self.ledger.contains_block_height(new_block.height()) {
-            debug!(
-                "Ledger is already synced with block at height {height}. Will not sync.",
-                height = new_block.height()
-            );
+        if self.ledger.contains_block_height(block.height()) {
+            debug!("Ledger is already synced with block at height {}. Will not sync.", block.height());
             return Ok(());
         }
 
-        // Acquire the pending blocks lock.
-        let mut pending_blocks = self.pending_blocks.lock().await;
+        // Acquire the latest block responses lock.
+        let mut latest_block_responses = self.latest_block_responses.lock().await;
 
-        if let Some(tail) = pending_blocks.back() {
-            if tail.height() >= new_block.height() {
-                debug!(
-                    "A unconfirmed block is queued already for height {height}. \
-                    Will not sync.",
-                    height = new_block.height()
-                );
-                return Ok(());
-            }
+        if latest_block_responses.contains_key(&block.height()) {
+            debug!("An unconfirmed block is queued already for height {}. Will not sync.", block.height());
+            return Ok(());
+        }
+
+        // If the block authority is a sub-DAG, then sync the batch certificates with the block.
+        // Note that the block authority is always a sub-DAG in production;
+        // beacon signatures are only used for testing,
+        // and as placeholder (irrelevant) block authority in the genesis block.
+        if let Authority::Quorum(subdag) = block.authority() {
+            // Reconstruct the unconfirmed transactions.
+            let unconfirmed_transactions = cfg_iter!(block.transactions())
+                .filter_map(|tx| {
+                    tx.to_unconfirmed_transaction().map(|unconfirmed| (unconfirmed.id(), unconfirmed)).ok()
+                })
+                .collect::<HashMap<_, _>>();
+
+            // Iterate over the certificates.
+            for certificates in subdag.values().cloned() {
+                cfg_into_iter!(certificates.clone()).for_each(|certificate| {
+                    // Sync the batch certificate with the block.
+                    self.storage.sync_certificate_with_block(&block, certificate.clone(), &unconfirmed_transactions);
+                });
 
-            ensure!(tail.height() + 1 == new_block.height(), "Got an out-of-order block");
+                // Sync the BFT DAG with the certificates.
+                for certificate in certificates {
+                    // If a BFT sender was provided, send the certificate to the BFT.
+                    // For validators, BFT spawns a receiver task in `BFT::start_handlers`.
+                    if let Some(bft_sender) = self.bft_sender.get() {
+                        // Await the callback to continue.
+                        if let Err(err) = bft_sender.send_sync_bft(certificate).await {
+                            bail!("Failed to sync certificate - {err}");
+                        };
+                    }
+                }
+            }
         }
 
         // Fetch the latest block height.
         let ledger_block_height = self.ledger.latest_block_height();
 
-        // Clear any older pending blocks.
-        // TODO(kaimast): ensure there are no dangling block requests
-        while let Some(pending_block) = pending_blocks.front() {
-            if pending_block.height() > ledger_block_height {
-                break;
-            }
-
-            pending_blocks.pop_front();
-        }
+        // Insert the latest block response.
+        latest_block_responses.insert(block.height(), block);
+        // Clear the latest block responses of older blocks.
+        latest_block_responses.retain(|height, _| *height > ledger_block_height);
 
         // Get a list of contiguous blocks from the latest block responses.
-        let new_block = self.ledger.check_block_subdag(new_block, pending_blocks.make_contiguous())?;
-
-        // Append the new block to the set of pending blocks and add its certificates to the storage.
-        self.add_block_subdag_to_bft(&new_block).await?;
-        pending_blocks.push_back(new_block);
-
-        // Now, figure out if and which pending block we can commit.
-        // To do this effectively and because commits are transitive,
-        // we iterate in reverse so that we can stop at the first successful check.
-        //
-        // Note, that if the storage already contains certificates for the round after new block,
-        // the availability threshold for the new block could also be reached.
-        let mut commit_height = None;
-        for block in pending_blocks.iter().rev() {
-            if self.is_block_availability_threshold_reached(block)? {
-                commit_height = Some(block.height());
-                break;
-            }
-        }
-
-        if let Some(commit_height) = commit_height {
-            let start_height = ledger_block_height + 1;
-            ensure!(commit_height >= start_height, "Invalid commit height");
-            let num_blocks = (commit_height - start_height + 1) as usize;
+        let contiguous_blocks: Vec<Block<N>> = (ledger_block_height.saturating_add(1)..)
+            .take_while(|&k| latest_block_responses.contains_key(&k))
+            .filter_map(|k| latest_block_responses.get(&k).cloned())
+            .collect();
 
-            // Create a more detailed log message if we are committing more than one block at a time.
-            if num_blocks > 1 {
-                trace!(
-                    "Attempting to commit {chain_length} pending block(s) starting at height {start_height}.",
-                    chain_length = pending_blocks.len(),
-                );
-            }
+        // Check if each block response, from the contiguous sequence just constructed,
+        // is ready to be added to the ledger.
+        // Ensure that the block's leader certificate meets the availability threshold
+        // based on the certificates in the DAG just after the block's round.
+        // If the availability threshold is not met,
+        // process the next block and check if it is linked to the current block,
+        // in the sense that there is a path in the DAG
+        // from the next block's leader certificate
+        // to the current block's leader certificate.
+        // Note: We do not advance to the most recent block response because we would be unable to
+        // validate if the leader certificate in the block has been certified properly.
+        for next_block in contiguous_blocks.into_iter() {
+            // Retrieve the height of the next block.
+            let next_block_height = next_block.height();
+
+            // Fetch the leader certificate and the relevant rounds.
+            let leader_certificate = match next_block.authority() {
+                Authority::Quorum(subdag) => subdag.leader_certificate().clone(),
+                _ => bail!("Received a block with an unexpected authority type."),
+            };
+            let commit_round = leader_certificate.round();
+            let certificate_round =
+                commit_round.checked_add(1).ok_or_else(|| anyhow!("Integer overflow on round number"))?;
+
+            // Get the committee lookback for the round just after the leader.
+            let certificate_committee_lookback = self.ledger.get_committee_lookback_for_round(certificate_round)?;
+            // Retrieve all of the certificates for the round just after the leader.
+            let certificates = self.storage.get_certificates_for_round(certificate_round);
+            // Construct a set over the authors, at the round just after the leader,
+            // who included the leader's certificate in their previous certificate IDs.
+            let authors = certificates
+                .iter()
+                .filter_map(|c| match c.previous_certificate_ids().contains(&leader_certificate.id()) {
+                    true => Some(c.author()),
+                    false => None,
+                })
+                .collect();
+
+            debug!("Validating sync block {next_block_height} at round {commit_round}...");
+            // Check if the leader is ready to be committed.
+            if certificate_committee_lookback.is_availability_threshold_reached(&authors) {
+                // Initialize the current certificate.
+                let mut current_certificate = leader_certificate;
+                // Check if there are any linked blocks that need to be added.
+                let mut blocks_to_add = vec![next_block];
+
+                // Check if there are other blocks to process based on `is_linked`.
+                for height in (self.ledger.latest_block_height().saturating_add(1)..next_block_height).rev() {
+                    // Retrieve the previous block.
+                    let Some(previous_block) = latest_block_responses.get(&height) else {
+                        bail!("Block {height} is missing from the latest block responses.");
+                    };
+                    // Retrieve the previous block's leader certificate.
+                    let previous_certificate = match previous_block.authority() {
+                        Authority::Quorum(subdag) => subdag.leader_certificate().clone(),
+                        _ => bail!("Received a block with an unexpected authority type."),
+                    };
+                    // Determine if there is a path between the previous certificate and the current certificate.
+                    if self.is_linked(previous_certificate.clone(), current_certificate.clone())? {
+                        debug!("Previous sync block {height} is linked to the current block {next_block_height}");
+                        // Add the previous leader certificate to the list of certificates to commit.
+                        blocks_to_add.insert(0, previous_block.clone());
+                        // Update the current certificate to the previous leader certificate.
+                        current_certificate = previous_certificate;
+                    }
+                }
 
-            for pending_block in pending_blocks.drain(0..num_blocks) {
-                let hash = pending_block.hash();
-                let height = pending_block.height();
-                match self.ledger.check_block_content(pending_block) {
-                    Ok(block) => {
-                        trace!("Adding pending block {hash} at height {height} to the ledger");
-                        self.ledger.advance_to_next_block(&block)?;
+                // Add the blocks to the ledger.
+                for block in blocks_to_add {
+                    // Check that the blocks are sequential and can be added to the ledger.
+                    let block_height = block.height();
+                    if block_height != self.ledger.latest_block_height().saturating_add(1) {
+                        warn!("Skipping block {block_height} from the latest block responses - not sequential.");
+                        continue;
+                    }
+                    #[cfg(feature = "telemetry")]
+                    let block_authority = block.authority().clone();
+
+                    let self_ = self.clone();
+                    tokio::task::spawn_blocking(move || {
+                        // Check the next block.
+                        self_.ledger.check_next_block(&block)?;
+                        // Attempt to advance to the next block.
+                        self_.ledger.advance_to_next_block(&block)?;
+
+                        // Sync the height with the block.
+                        self_.storage.sync_height_with_block(block.height());
+                        // Sync the round with the block.
+                        self_.storage.sync_round_with_block(block.round());
+
+                        Ok::<(), anyhow::Error>(())
+                    })
+                    .await??;
+                    // Remove the block height from the latest block responses.
+                    latest_block_responses.remove(&block_height);
+
+                    // Update the validator telemetry.
+                    #[cfg(feature = "telemetry")]
+                    if let Authority::Quorum(subdag) = block_authority {
+                        self_.gateway.validator_telemetry().insert_subdag(&subdag);
                     }
-                    Err(err) => bail!("Failed to check contents of pending block {hash} at height {height}: {err}"),
                 }
+            } else {
+                debug!(
+                    "Availability threshold was not reached for block {next_block_height} at round {commit_round}. Checking next block..."
+                );
             }
+
+            // Don't remove the response from BlockSync yet as we might fall back to non-BFT sync.
         }
 
         Ok(())
     }
+
+    /// Returns `true` if there is a path from the previous certificate to the current certificate.
+    fn is_linked(
+        &self,
+        previous_certificate: BatchCertificate<N>,
+        current_certificate: BatchCertificate<N>,
+    ) -> Result<bool> {
+        // Initialize the list containing the traversal.
+        let mut traversal = vec![current_certificate.clone()];
+        // Iterate over the rounds from the current certificate to the previous certificate.
+        for round in (previous_certificate.round()..current_certificate.round()).rev() {
+            // Retrieve all of the certificates for this past round.
+            let certificates = self.storage.get_certificates_for_round(round);
+            // Filter the certificates to only include those that are in the traversal.
+            traversal = certificates
+                .into_iter()
+                .filter(|p| traversal.iter().any(|c| c.previous_certificate_ids().contains(&p.id())))
+                .collect();
+        }
+        Ok(traversal.contains(&previous_certificate))
+    }
 }
 
 // Methods to assist with the block sync module.
@@ -951,24 +972,13 @@ mod tests {
     type CurrentLedger = Ledger<CurrentNetwork, ConsensusMemory<CurrentNetwork>>;
     type CurrentConsensusStore = ConsensusStore<CurrentNetwork, ConsensusMemory<CurrentNetwork>>;
 
-    /// Tests that commits work as expected when some anchors are not committed immediately.
     #[tokio::test]
     #[tracing_test::traced_test]
-    async fn test_commit_chain() -> anyhow::Result<()> {
+    async fn test_commit_via_is_linked() -> anyhow::Result<()> {
         let rng = &mut TestRng::default();
         // Initialize the round parameters.
         let max_gc_rounds = BatchHeader::<CurrentNetwork>::MAX_GC_ROUNDS as u64;
-
-        // The first round of the first block.
-        let first_round = 1;
-        // The total number of blocks we test
-        let num_blocks = 3;
-        // The number of certificate rounds needed.
-        // There is one additional round to provide availability for the inal block.
-        let num_rounds = first_round + num_blocks * 2 + 1;
-        // The first round that has at least N-f certificates referencing the anchor from the previous round.
-        // This is also the last round we use in the test.
-        let first_committed_round = num_rounds - 1;
+        let commit_round = 2;
 
         // Initialize the store.
         let store = CurrentConsensusStore::open(StorageMode::new_test(None)).unwrap();
@@ -1009,70 +1019,77 @@ mod tests {
                 HashMap::new();
             let mut previous_certificates: IndexSet<BatchCertificate<CurrentNetwork>> = IndexSet::with_capacity(4);
 
-            for round in first_round..=first_committed_round {
+            for round in 0..=commit_round + 8 {
                 let mut current_certificates = IndexSet::new();
                 let previous_certificate_ids: IndexSet<_> = if round == 0 || round == 1 {
                     IndexSet::new()
                 } else {
                     previous_certificates.iter().map(|c| c.id()).collect()
                 };
-
                 let committee_id = committee.id();
-                let prev_leader = committee.get_leader(round - 1).unwrap();
-
-                // For the first two blocks non-leaders will not reference the leader certificate.
-                // This means, while there is an anchor, it is isn't committed
-                // until later.
-                for (i, private_key) in private_keys.iter().enumerate() {
-                    let leader_index = addresses.iter().position(|&address| address == prev_leader).unwrap();
-                    let is_certificate_round = round % 2 == 1;
-                    let is_leader = i == leader_index;
-
-                    let previous_certs = if round < first_committed_round && is_certificate_round && !is_leader {
-                        previous_certificate_ids
-                            .iter()
-                            .cloned()
-                            .enumerate()
-                            .filter(|(idx, _)| *idx != leader_index)
-                            .map(|(_, id)| id)
-                            .collect()
-                    } else {
-                        previous_certificate_ids.clone()
-                    };
-
-                    let batch_header = BatchHeader::new(
-                        private_key,
-                        round,
-                        now(),
-                        committee_id,
-                        Default::default(),
-                        previous_certs,
-                        rng,
-                    )
-                    .unwrap();
 
-                    // Sign the batch header.
-                    let mut signatures = IndexSet::with_capacity(4);
-                    for (j, private_key_2) in private_keys.iter().enumerate() {
-                        if i != j {
-                            signatures.insert(private_key_2.sign(&[batch_header.batch_id()], rng).unwrap());
+                // Create a certificate for the leader.
+                if round <= 5 {
+                    let leader = committee.get_leader(round).unwrap();
+                    let leader_index = addresses.iter().position(|&address| address == leader).unwrap();
+                    let non_leader_index = addresses.iter().position(|&address| address != leader).unwrap();
+                    for i in [leader_index, non_leader_index].into_iter() {
+                        let batch_header = BatchHeader::new(
+                            &private_keys[i],
+                            round,
+                            now(),
+                            committee_id,
+                            Default::default(),
+                            previous_certificate_ids.clone(),
+                            rng,
+                        )
+                        .unwrap();
+                        // Sign the batch header.
+                        let mut signatures = IndexSet::with_capacity(4);
+                        for (j, private_key_2) in private_keys.iter().enumerate() {
+                            if i != j {
+                                signatures.insert(private_key_2.sign(&[batch_header.batch_id()], rng).unwrap());
+                            }
                         }
+                        current_certificates.insert(BatchCertificate::from(batch_header, signatures).unwrap());
                     }
-                    current_certificates.insert(BatchCertificate::from(batch_header, signatures).unwrap());
                 }
 
+                // Create a certificate for each validator.
+                if round > 5 {
+                    for (i, private_key_1) in private_keys.iter().enumerate() {
+                        let batch_header = BatchHeader::new(
+                            private_key_1,
+                            round,
+                            now(),
+                            committee_id,
+                            Default::default(),
+                            previous_certificate_ids.clone(),
+                            rng,
+                        )
+                        .unwrap();
+                        // Sign the batch header.
+                        let mut signatures = IndexSet::with_capacity(4);
+                        for (j, private_key_2) in private_keys.iter().enumerate() {
+                            if i != j {
+                                signatures.insert(private_key_2.sign(&[batch_header.batch_id()], rng).unwrap());
+                            }
+                        }
+                        current_certificates.insert(BatchCertificate::from(batch_header, signatures).unwrap());
+                    }
+                }
                 // Update the map of certificates.
                 round_to_certificates_map.insert(round, current_certificates.clone());
-                previous_certificates = current_certificates;
+                previous_certificates = current_certificates.clone();
             }
             (round_to_certificates_map, committee)
         };
 
         // Initialize the storage.
         let storage = Storage::new(core_ledger.clone(), Arc::new(BFTMemoryService::new()), max_gc_rounds);
-        // Insert all certificates into storage.
+        // Insert certificates into storage.
         let mut certificates: Vec<BatchCertificate<CurrentNetwork>> = Vec::new();
-        for i in first_round..=first_committed_round {
+        for i in 1..=commit_round + 8 {
             let c = (*round_to_certificates_map.get(&i).unwrap()).clone();
             certificates.extend(c);
         }
@@ -1080,79 +1097,100 @@ mod tests {
             storage.testing_only_insert_certificate_testing_only(certificate.clone());
         }
 
-        // Create the blocks
-        let mut previous_leader_cert = None;
-        let mut blocks = vec![];
-
-        for block_height in 1..=num_blocks {
-            let leader_round = block_height * 2;
-
-            let leader = committee.get_leader(leader_round).unwrap();
-            let leader_certificate = storage.get_certificate_for_round_with_author(leader_round, leader).unwrap();
-
+        // Create block 1.
+        let leader_round_1 = commit_round;
+        let leader_1 = committee.get_leader(leader_round_1).unwrap();
+        let leader_certificate = storage.get_certificate_for_round_with_author(commit_round, leader_1).unwrap();
+        let block_1 = {
             let mut subdag_map: BTreeMap<u64, IndexSet<BatchCertificate<CurrentNetwork>>> = BTreeMap::new();
             let mut leader_cert_map = IndexSet::new();
             leader_cert_map.insert(leader_certificate.clone());
-
-            let previous_cert_map = storage.get_certificates_for_round(leader_round - 1);
-
-            subdag_map.insert(leader_round, leader_cert_map.clone());
-            subdag_map.insert(leader_round - 1, previous_cert_map.clone());
-
-            if leader_round > 2 {
-                let previous_commit_cert_map: IndexSet<_> = storage
-                    .get_certificates_for_round(leader_round - 2)
-                    .into_iter()
-                    .filter(|cert| {
-                        if let Some(previous_leader_cert) = &previous_leader_cert {
-                            cert != previous_leader_cert
-                        } else {
-                            true
-                        }
-                    })
-                    .collect();
-                subdag_map.insert(leader_round - 2, previous_commit_cert_map);
+            let mut previous_cert_map = IndexSet::new();
+            for cert in storage.get_certificates_for_round(commit_round - 1) {
+                previous_cert_map.insert(cert);
             }
-
+            subdag_map.insert(commit_round, leader_cert_map.clone());
+            subdag_map.insert(commit_round - 1, previous_cert_map.clone());
             let subdag = Subdag::from(subdag_map.clone())?;
-            let block = core_ledger.prepare_advance_to_next_quorum_block(subdag, Default::default())?;
-
-            previous_leader_cert = Some(leader_certificate);
+            core_ledger.prepare_advance_to_next_quorum_block(subdag, Default::default())?
+        };
+        // Insert block 1.
+        core_ledger.advance_to_next_block(&block_1)?;
 
-            core_ledger.advance_to_next_block(&block)?;
-            blocks.push(block);
-        }
+        // Create block 2.
+        let leader_round_2 = commit_round + 2;
+        let leader_2 = committee.get_leader(leader_round_2).unwrap();
+        let leader_certificate_2 = storage.get_certificate_for_round_with_author(leader_round_2, leader_2).unwrap();
+        let block_2 = {
+            let mut subdag_map_2: BTreeMap<u64, IndexSet<BatchCertificate<CurrentNetwork>>> = BTreeMap::new();
+            let mut leader_cert_map_2 = IndexSet::new();
+            leader_cert_map_2.insert(leader_certificate_2.clone());
+            let mut previous_cert_map_2 = IndexSet::new();
+            for cert in storage.get_certificates_for_round(leader_round_2 - 1) {
+                previous_cert_map_2.insert(cert);
+            }
+            let mut prev_commit_cert_map_2 = IndexSet::new();
+            for cert in storage.get_certificates_for_round(leader_round_2 - 2) {
+                if cert != leader_certificate {
+                    prev_commit_cert_map_2.insert(cert);
+                }
+            }
+            subdag_map_2.insert(leader_round_2, leader_cert_map_2.clone());
+            subdag_map_2.insert(leader_round_2 - 1, previous_cert_map_2.clone());
+            subdag_map_2.insert(leader_round_2 - 2, prev_commit_cert_map_2.clone());
+            let subdag_2 = Subdag::from(subdag_map_2.clone())?;
+            core_ledger.prepare_advance_to_next_quorum_block(subdag_2, Default::default())?
+        };
+        // Insert block 2.
+        core_ledger.advance_to_next_block(&block_2)?;
 
-        // ### Test that sync works as expected ###
+        // Create block 3
+        let leader_round_3 = commit_round + 4;
+        let leader_3 = committee.get_leader(leader_round_3).unwrap();
+        let leader_certificate_3 = storage.get_certificate_for_round_with_author(leader_round_3, leader_3).unwrap();
+        let block_3 = {
+            let mut subdag_map_3: BTreeMap<u64, IndexSet<BatchCertificate<CurrentNetwork>>> = BTreeMap::new();
+            let mut leader_cert_map_3 = IndexSet::new();
+            leader_cert_map_3.insert(leader_certificate_3.clone());
+            let mut previous_cert_map_3 = IndexSet::new();
+            for cert in storage.get_certificates_for_round(leader_round_3 - 1) {
+                previous_cert_map_3.insert(cert);
+            }
+            let mut prev_commit_cert_map_3 = IndexSet::new();
+            for cert in storage.get_certificates_for_round(leader_round_3 - 2) {
+                if cert != leader_certificate_2 {
+                    prev_commit_cert_map_3.insert(cert);
+                }
+            }
+            subdag_map_3.insert(leader_round_3, leader_cert_map_3.clone());
+            subdag_map_3.insert(leader_round_3 - 1, previous_cert_map_3.clone());
+            subdag_map_3.insert(leader_round_3 - 2, prev_commit_cert_map_3.clone());
+            let subdag_3 = Subdag::from(subdag_map_3.clone())?;
+            core_ledger.prepare_advance_to_next_quorum_block(subdag_3, Default::default())?
+        };
+        // Insert block 3.
+        core_ledger.advance_to_next_block(&block_3)?;
 
-        // Create a new ledger to test with, but use the existing storage
-        // so that the certificates exist.
+        // Initialize the syncing ledger.
         let syncing_ledger = Arc::new(CoreLedgerService::new(
             CurrentLedger::load(genesis, StorageMode::new_test(None)).unwrap(),
             Default::default(),
         ));
-
-        // Set up sync and its dependencies.
+        // Initialize the gateway.
         let gateway = Gateway::new(account.clone(), storage.clone(), syncing_ledger.clone(), None, &[], None)?;
+        // Initialize the block synchronization logic.
         let block_sync = Arc::new(BlockSync::new(syncing_ledger.clone()));
+        // Initialize the sync module.
         let sync = Sync::new(gateway.clone(), storage.clone(), syncing_ledger.clone(), block_sync);
-
-        let mut block_iter = blocks.into_iter();
-
-        // Insert the blocks into the new sync module
-        for _ in 0..num_blocks - 1 {
-            let block = block_iter.next().unwrap();
-            sync.sync_storage_with_block(block).await?;
-
-            // Availability threshold is not met, so we should not advance yet.
-            assert_eq!(syncing_ledger.latest_block_height(), 0);
-        }
-
-        // Only for the final block, the availability threshold is met,
-        // because certificates for the subsequent round are already in storage.
-        sync.sync_storage_with_block(block_iter.next().unwrap()).await?;
+        // Try to sync block 1.
+        sync.sync_storage_with_block(block_1).await?;
+        assert_eq!(syncing_ledger.latest_block_height(), 1);
+        // Try to sync block 2.
+        sync.sync_storage_with_block(block_2).await?;
+        assert_eq!(syncing_ledger.latest_block_height(), 2);
+        // Try to sync block 3.
+        sync.sync_storage_with_block(block_3).await?;
         assert_eq!(syncing_ledger.latest_block_height(), 3);
-
         // Ensure blocks 1 and 2 were added to the ledger.
         assert!(syncing_ledger.contains_block_height(1));
         assert!(syncing_ledger.contains_block_height(2));
diff --git a/node/bft/src/worker.rs b/node/bft/src/worker.rs
index e7756952c..04c62ab6c 100644
--- a/node/bft/src/worker.rs
+++ b/node/bft/src/worker.rs
@@ -26,7 +26,7 @@ use snarkos_node_bft_ledger_service::LedgerService;
 use snarkvm::{
     console::prelude::*,
     ledger::{
-        Transaction,
+        block::Transaction,
         narwhal::{BatchHeader, Data, Transmission, TransmissionID},
         puzzle::{Solution, SolutionID},
     },
@@ -560,8 +560,7 @@ mod tests {
     use snarkvm::{
         console::{network::Network, types::Field},
         ledger::{
-            Block,
-            PendingBlock,
+            block::Block,
             committee::Committee,
             narwhal::{BatchCertificate, Subdag, Transmission, TransmissionID},
             snarkvm_ledger_test_helpers::sample_execution_transaction_with_fee,
@@ -627,8 +626,6 @@ mod tests {
                 transaction_id: N::TransactionID,
                 transaction: Transaction<N>,
             ) -> Result<()>;
-            fn check_block_subdag(&self, _block: Block<N>, _prefix: &[PendingBlock<N>]) -> Result<PendingBlock<N>>;
-            fn check_block_content(&self, _block: PendingBlock<N>) -> Result<Block<N>>;
             fn check_next_block(&self, block: &Block<N>) -> Result<()>;
             fn prepare_advance_to_next_quorum_block(
                 &self,
diff --git a/node/router/src/lib.rs b/node/router/src/lib.rs
index ad18d9f7b..161266b5e 100644
--- a/node/router/src/lib.rs
+++ b/node/router/src/lib.rs
@@ -126,7 +126,7 @@ impl<N: Network> Router<N> {
     const CONNECTION_ATTEMPTS_SINCE_SECS: i64 = 10;
     /// The maximum number of candidate peers permitted to be stored in the node.
     const MAXIMUM_CANDIDATE_PEERS: usize = 10_000;
-    /// The maximum amount of connection attempts within a 10 second threshold.
+    /// The maximum amount of connection attempts within a 10 second threshold
     #[cfg(not(test))]
     const MAX_CONNECTION_ATTEMPTS: usize = 10;
     /// The duration after which a connected peer is considered inactive or
