diff --git a/.integration/Cargo.toml b/.integration/Cargo.toml
index b58fa4f2b..634efb25c 100644
--- a/.integration/Cargo.toml
+++ b/.integration/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkos-integration"
-version = "2.2.4"
+version = "2.2.2"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "A integration testing suite for a decentralized operating system"
 homepage = "https://aleo.org"
diff --git a/.resources/release-version b/.resources/release-version
index 15274788c..a690e5d45 100644
--- a/.resources/release-version
+++ b/.resources/release-version
@@ -1 +1 @@
-v2.2.4
\ No newline at end of file
+v2.2.2
\ No newline at end of file
diff --git a/Cargo.lock b/Cargo.lock
index 6e68aa807..c964ef38f 100644
--- a/Cargo.lock
+++ b/Cargo.lock
@@ -1466,9 +1466,9 @@ dependencies = [
 
 [[package]]
 name = "indexmap"
-version = "2.1.0"
+version = "2.0.2"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "d530e1a18b1cb4c484e6e34556a0d948706958449fca0cab753d649f2bce3d1f"
+checksum = "8adf3ddd720272c6ea8bf59463c04e0f93d0bbf7c5439b691bca2987e0270897"
 dependencies = [
  "equivalent",
  "hashbrown 0.14.1",
@@ -2844,11 +2844,11 @@ dependencies = [
 
 [[package]]
 name = "serde_json"
-version = "1.0.108"
+version = "1.0.107"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "3d1c7e3eac408d115102c4c24ad393e0821bb3a5df4d506a80f85f7a742a526b"
+checksum = "6b420ce6e3d8bd882e9b243c6eed35dbc9a6110c9769e74b584e0d68d1f20c65"
 dependencies = [
- "indexmap 2.1.0",
+ "indexmap 2.0.2",
  "itoa",
  "ryu",
  "serde",
@@ -2987,7 +2987,7 @@ dependencies = [
 
 [[package]]
 name = "snarkos"
-version = "2.2.4"
+version = "2.2.2"
 dependencies = [
  "anyhow",
  "clap",
@@ -3010,7 +3010,7 @@ dependencies = [
 
 [[package]]
 name = "snarkos-account"
-version = "2.2.4"
+version = "2.2.2"
 dependencies = [
  "anyhow",
  "colored",
@@ -3020,7 +3020,7 @@ dependencies = [
 
 [[package]]
 name = "snarkos-cli"
-version = "2.2.4"
+version = "2.2.2"
 dependencies = [
  "aleo-std",
  "anstyle",
@@ -3029,7 +3029,7 @@ dependencies = [
  "clap",
  "colored",
  "crossterm 0.27.0",
- "indexmap 2.1.0",
+ "indexmap 2.0.2",
  "nix",
  "num_cpus",
  "parking_lot",
@@ -3054,7 +3054,7 @@ dependencies = [
 
 [[package]]
 name = "snarkos-display"
-version = "2.2.4"
+version = "2.2.2"
 dependencies = [
  "anyhow",
  "crossterm 0.27.0",
@@ -3066,7 +3066,7 @@ dependencies = [
 
 [[package]]
 name = "snarkos-integration"
-version = "2.2.4"
+version = "2.2.2"
 dependencies = [
  "snarkos-node-cdn",
  "snarkvm",
@@ -3077,7 +3077,7 @@ dependencies = [
 
 [[package]]
 name = "snarkos-node"
-version = "2.2.4"
+version = "2.2.2"
 dependencies = [
  "aleo-std",
  "anyhow",
@@ -3085,7 +3085,7 @@ dependencies = [
  "colored",
  "deadline",
  "futures-util",
- "indexmap 2.1.0",
+ "indexmap 2.0.2",
  "num_cpus",
  "once_cell",
  "parking_lot",
@@ -3112,7 +3112,7 @@ dependencies = [
 
 [[package]]
 name = "snarkos-node-bft"
-version = "2.2.4"
+version = "2.2.2"
 dependencies = [
  "anyhow",
  "async-recursion",
@@ -3124,7 +3124,7 @@ dependencies = [
  "colored",
  "deadline",
  "futures",
- "indexmap 2.1.0",
+ "indexmap 2.0.2",
  "itertools 0.11.0",
  "mockall",
  "open",
@@ -3157,11 +3157,11 @@ dependencies = [
 
 [[package]]
 name = "snarkos-node-bft-events"
-version = "2.2.4"
+version = "2.2.2"
 dependencies = [
  "anyhow",
  "bytes",
- "indexmap 2.1.0",
+ "indexmap 2.0.2",
  "proptest",
  "rayon",
  "serde",
@@ -3176,10 +3176,10 @@ dependencies = [
 
 [[package]]
 name = "snarkos-node-bft-ledger-service"
-version = "2.2.4"
+version = "2.2.2"
 dependencies = [
  "async-trait",
- "indexmap 2.1.0",
+ "indexmap 2.0.2",
  "parking_lot",
  "snarkvm",
  "tokio",
@@ -3188,7 +3188,7 @@ dependencies = [
 
 [[package]]
 name = "snarkos-node-cdn"
-version = "2.2.4"
+version = "2.2.2"
 dependencies = [
  "anyhow",
  "backoff",
@@ -3208,11 +3208,11 @@ dependencies = [
 
 [[package]]
 name = "snarkos-node-consensus"
-version = "2.2.4"
+version = "2.2.2"
 dependencies = [
  "anyhow",
  "colored",
- "indexmap 2.1.0",
+ "indexmap 2.0.2",
  "itertools 0.11.0",
  "lru",
  "once_cell",
@@ -3229,7 +3229,7 @@ dependencies = [
 
 [[package]]
 name = "snarkos-node-metrics"
-version = "2.2.4"
+version = "2.2.2"
 dependencies = [
  "metrics",
  "metrics-exporter-prometheus",
@@ -3238,13 +3238,13 @@ dependencies = [
 
 [[package]]
 name = "snarkos-node-rest"
-version = "2.2.4"
+version = "2.2.2"
 dependencies = [
  "anyhow",
  "axum",
  "axum-extra",
  "http",
- "indexmap 2.1.0",
+ "indexmap 2.0.2",
  "jsonwebtoken",
  "once_cell",
  "parking_lot",
@@ -3262,7 +3262,7 @@ dependencies = [
 
 [[package]]
 name = "snarkos-node-router"
-version = "2.2.4"
+version = "2.2.2"
 dependencies = [
  "anyhow",
  "async-trait",
@@ -3272,7 +3272,7 @@ dependencies = [
  "deadline",
  "futures",
  "futures-util",
- "indexmap 2.1.0",
+ "indexmap 2.0.2",
  "linked-hash-map",
  "parking_lot",
  "peak_alloc",
@@ -3296,11 +3296,11 @@ dependencies = [
 
 [[package]]
 name = "snarkos-node-router-messages"
-version = "2.2.4"
+version = "2.2.2"
 dependencies = [
  "anyhow",
  "bytes",
- "indexmap 2.1.0",
+ "indexmap 2.0.2",
  "proptest",
  "rayon",
  "serde",
@@ -3316,10 +3316,10 @@ dependencies = [
 
 [[package]]
 name = "snarkos-node-sync"
-version = "2.2.4"
+version = "2.2.2"
 dependencies = [
  "anyhow",
- "indexmap 2.1.0",
+ "indexmap 2.0.2",
  "itertools 0.11.0",
  "once_cell",
  "parking_lot",
@@ -3335,7 +3335,7 @@ dependencies = [
 
 [[package]]
 name = "snarkos-node-sync-communication-service"
-version = "2.2.4"
+version = "2.2.2"
 dependencies = [
  "async-trait",
  "tokio",
@@ -3343,10 +3343,10 @@ dependencies = [
 
 [[package]]
 name = "snarkos-node-sync-locators"
-version = "2.2.4"
+version = "2.2.2"
 dependencies = [
  "anyhow",
- "indexmap 2.1.0",
+ "indexmap 2.0.2",
  "serde",
  "snarkvm",
  "tracing",
@@ -3354,7 +3354,7 @@ dependencies = [
 
 [[package]]
 name = "snarkos-node-tcp"
-version = "2.2.4"
+version = "2.2.2"
 dependencies = [
  "async-trait",
  "bytes",
@@ -3368,16 +3368,15 @@ dependencies = [
 
 [[package]]
 name = "snarkvm"
-version = "0.16.8"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "ac2de3b2fcf7239acac5eea5703b48ace0a33b082175148832889fd65754f389"
+version = "0.16.6"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
  "anstyle",
  "anyhow",
  "clap",
  "colored",
  "dotenvy",
- "indexmap 2.1.0",
+ "indexmap 2.0.2",
  "num-format",
  "once_cell",
  "parking_lot",
@@ -3399,9 +3398,8 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-algorithms"
-version = "0.16.8"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "e27bd74136c52ac9c745d7b5fbd86febd8254992ddfad364630149bc32834220"
+version = "0.16.6"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
  "aleo-std",
  "anyhow",
@@ -3410,7 +3408,7 @@ dependencies = [
  "fxhash",
  "hashbrown 0.14.1",
  "hex",
- "indexmap 2.1.0",
+ "indexmap 2.0.2",
  "itertools 0.11.0",
  "num-traits",
  "parking_lot",
@@ -3430,9 +3428,8 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-circuit"
-version = "0.16.8"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "ab179b2dc61b734f929a0dbdf4498c6df14ded238fd4b359636e8d24ba93254e"
+version = "0.16.6"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
  "snarkvm-circuit-account",
  "snarkvm-circuit-algorithms",
@@ -3445,9 +3442,8 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-circuit-account"
-version = "0.16.8"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "d4867a808f44a14430ab62d16437da8ba0629faae7af147ecc2e205f9dd1c5af"
+version = "0.16.6"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
  "snarkvm-circuit-algorithms",
  "snarkvm-circuit-network",
@@ -3457,9 +3453,8 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-circuit-algorithms"
-version = "0.16.8"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "91bb88497b1a0c9dd7ba2063b833fbb4a933e78621a6c99d7e240f096ffb0cb2"
+version = "0.16.6"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
  "snarkvm-circuit-types",
  "snarkvm-console-algorithms",
@@ -3468,9 +3463,8 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-circuit-collections"
-version = "0.16.8"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "525950b4fa312ea53e441f07c0a2fc417cc733ef904950e4769aaa957ccedbf3"
+version = "0.16.6"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
  "snarkvm-circuit-algorithms",
  "snarkvm-circuit-types",
@@ -3479,11 +3473,10 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-circuit-environment"
-version = "0.16.8"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "3676fb9628c839402fb976f4ef78b5ecc6de28d690de8634691a8a330e7d1749"
+version = "0.16.6"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
- "indexmap 2.1.0",
+ "indexmap 2.0.2",
  "itertools 0.11.0",
  "nom",
  "num-traits",
@@ -3498,15 +3491,13 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-circuit-environment-witness"
-version = "0.16.8"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "96967cbe71293f982df0317b7d6f4eaf6cff7b53c23c894b08c9d69c196f6550"
+version = "0.16.6"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 
 [[package]]
 name = "snarkvm-circuit-network"
-version = "0.16.8"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "264c4f1af7ebb69ebac1c8228daf6d3446bb04129147af719db1e792bd71de31"
+version = "0.16.6"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
  "snarkvm-circuit-algorithms",
  "snarkvm-circuit-collections",
@@ -3516,9 +3507,8 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-circuit-program"
-version = "0.16.8"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "9febb7006a520f2b46badcaab112e88d47bdea592eb8b13eeb37d2da049448b3"
+version = "0.16.6"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
  "paste",
  "snarkvm-circuit-account",
@@ -3532,9 +3522,8 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-circuit-types"
-version = "0.16.8"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "bf57c6f84ac480e2de20024cf5438a4f1c9229ee8633c7284d62559df43c3ee2"
+version = "0.16.6"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
  "snarkvm-circuit-environment",
  "snarkvm-circuit-types-address",
@@ -3548,9 +3537,8 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-circuit-types-address"
-version = "0.16.8"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "d5f47a7a9ce921935b5f835c891285ee57d48be131981b906a88edf37506e94f"
+version = "0.16.6"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
  "snarkvm-circuit-environment",
  "snarkvm-circuit-types-boolean",
@@ -3562,9 +3550,8 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-circuit-types-boolean"
-version = "0.16.8"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "9dea05fec3fc17f2b27f57df78824403601a9940dcbd4d2ebbd9d3971cbd9d94"
+version = "0.16.6"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
  "snarkvm-circuit-environment",
  "snarkvm-console-types-boolean",
@@ -3572,9 +3559,8 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-circuit-types-field"
-version = "0.16.8"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "b91170b4c957ebfe05cd5cf19c2d12df087334d09b45f7de3ca3db636a293401"
+version = "0.16.6"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
  "snarkvm-circuit-environment",
  "snarkvm-circuit-types-boolean",
@@ -3583,9 +3569,8 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-circuit-types-group"
-version = "0.16.8"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "0773e37037d2cb97b8321a71aa14b197979d937214cf891eb3303a9f8b16e282"
+version = "0.16.6"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
  "snarkvm-circuit-environment",
  "snarkvm-circuit-types-boolean",
@@ -3596,9 +3581,8 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-circuit-types-integers"
-version = "0.16.8"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "d038afe1dab25a721d0f2e5a96c8ea3f50fa93fef2921d1fa74f3d62e01f5376"
+version = "0.16.6"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
  "snarkvm-circuit-environment",
  "snarkvm-circuit-types-boolean",
@@ -3609,9 +3593,8 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-circuit-types-scalar"
-version = "0.16.8"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "c4c6a3d8c413efc2a8c3e2799b76cfa5174c2642c0178327832c2ba0332c24e7"
+version = "0.16.6"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
  "snarkvm-circuit-environment",
  "snarkvm-circuit-types-boolean",
@@ -3621,9 +3604,8 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-circuit-types-string"
-version = "0.16.8"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "4502fa8576e6cbc45c43b479e2e260876f0bcaf3a2fecbbc68ab4fbe72967c12"
+version = "0.16.6"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
  "snarkvm-circuit-environment",
  "snarkvm-circuit-types-boolean",
@@ -3634,9 +3616,8 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-console"
-version = "0.16.8"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "c2481b30ac987f92f257f3932785869e069293b31f31d23c456be4da091f4026"
+version = "0.16.6"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
  "snarkvm-console-account",
  "snarkvm-console-algorithms",
@@ -3648,9 +3629,8 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-console-account"
-version = "0.16.8"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "5ed8108908f3a2b03ac5cc70d793660c67ea29f26b927c0d761bb9b13b8183ef"
+version = "0.16.6"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
  "bs58",
  "snarkvm-console-network",
@@ -3660,9 +3640,8 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-console-algorithms"
-version = "0.16.8"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "4e89f3c4d74e9a1875cd5b2bef3fd9294ea03d5e62372540a6297f04ae950860"
+version = "0.16.6"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
  "blake2s_simd",
  "smallvec",
@@ -3674,9 +3653,8 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-console-collections"
-version = "0.16.8"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "4d97faaf74a8f5de5aa0b2a9c5b2cd7735202bd41467c89b45b7af52b98c54fd"
+version = "0.16.6"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
  "aleo-std",
  "rayon",
@@ -3686,12 +3664,11 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-console-network"
-version = "0.16.8"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "023ac3b6df60f839fcd96b3e2e17a4acf217d631d19f58e48762894b3aa29b9e"
+version = "0.16.6"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
  "anyhow",
- "indexmap 2.1.0",
+ "indexmap 2.0.2",
  "itertools 0.11.0",
  "lazy_static",
  "once_cell",
@@ -3710,9 +3687,8 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-console-network-environment"
-version = "0.16.8"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "645c1bea05aca3cf4ebf1c3530a7f8fc03b625c6aa1165a7fe09f25fa0611db9"
+version = "0.16.6"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
  "anyhow",
  "bech32",
@@ -3729,13 +3705,12 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-console-program"
-version = "0.16.8"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "41c05f06e0f65d08441603e3305fff51fa01307fefaf397cf9f3d78f0014c964"
+version = "0.16.6"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
  "enum_index",
  "enum_index_derive",
- "indexmap 2.1.0",
+ "indexmap 2.0.2",
  "num-derive",
  "num-traits",
  "once_cell",
@@ -3751,9 +3726,8 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-console-types"
-version = "0.16.8"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "36fd411863dbb47e139a7def1efe9e7b1eab665a535ee6e35630701f4a418007"
+version = "0.16.6"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
  "snarkvm-console-network-environment",
  "snarkvm-console-types-address",
@@ -3767,9 +3741,8 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-console-types-address"
-version = "0.16.8"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "b7ce0e5ccbcf7969ac0be0cc3136078ba253709f518892fa341c69235dac1683"
+version = "0.16.6"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
  "snarkvm-console-network-environment",
  "snarkvm-console-types-boolean",
@@ -3779,18 +3752,16 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-console-types-boolean"
-version = "0.16.8"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "5967d6243b8799a516b6aa4b9865092ccc0050f8adf4e72caeca67ce730ba2e9"
+version = "0.16.6"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
  "snarkvm-console-network-environment",
 ]
 
 [[package]]
 name = "snarkvm-console-types-field"
-version = "0.16.8"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "b839fc4a13df8f0d484f23505fa6d4121ff038fc1a10175ae6978ec16789f042"
+version = "0.16.6"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
  "snarkvm-console-network-environment",
  "snarkvm-console-types-boolean",
@@ -3799,9 +3770,8 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-console-types-group"
-version = "0.16.8"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "a81ccb898f0a13fc847e0c8f0593e1786f763158bf31094a0f3a09c4fbd13c8a"
+version = "0.16.6"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
  "snarkvm-console-network-environment",
  "snarkvm-console-types-boolean",
@@ -3811,9 +3781,8 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-console-types-integers"
-version = "0.16.8"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "660049d3d27b81351c54d358f43cd192bd75dfc0f2b51c80c9938b988191c901"
+version = "0.16.6"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
  "snarkvm-console-network-environment",
  "snarkvm-console-types-boolean",
@@ -3823,9 +3792,8 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-console-types-scalar"
-version = "0.16.8"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "1ae08843c01cefcafb80054bda31f964c39b0e9abe2beb62a15b34fbf11399ad"
+version = "0.16.6"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
  "snarkvm-console-network-environment",
  "snarkvm-console-types-boolean",
@@ -3835,9 +3803,8 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-console-types-string"
-version = "0.16.8"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "f938eba30b145f179836d41b579dac76dd111e24de3c17b52b24ccccfdca1796"
+version = "0.16.6"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
  "snarkvm-console-network-environment",
  "snarkvm-console-types-boolean",
@@ -3847,9 +3814,8 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-curves"
-version = "0.16.8"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "e85c03008a787b204873f7cdec7a948b6564dadcda7b4fbe0deeaae4d6466567"
+version = "0.16.6"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
  "rand",
  "rayon",
@@ -3862,9 +3828,8 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-fields"
-version = "0.16.8"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "fa11bc763ad6f19492d98e68bde012f599ed64f85bb7bb3151edb9d3029ad714"
+version = "0.16.6"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
  "aleo-std",
  "anyhow",
@@ -3881,13 +3846,12 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-ledger"
-version = "0.16.8"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "17c14e9741a1c96243ccd53ab2f6c1e80271f8998eeafa4c07dcd82d5db7710e"
+version = "0.16.6"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
  "aleo-std",
  "anyhow",
- "indexmap 2.1.0",
+ "indexmap 2.0.2",
  "parking_lot",
  "rand",
  "rayon",
@@ -3907,9 +3871,8 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-ledger-authority"
-version = "0.16.8"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "af3eac5c27b2eb46e2d2cc33bbe594f3db065820c8d3d1a6d2417df9f3b38d75"
+version = "0.16.6"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
  "anyhow",
  "rand",
@@ -3920,11 +3883,10 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-ledger-block"
-version = "0.16.8"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "ebe99a8c352be148b17d6b574b169edfeb6595542edeafe6cb69376a27d45a74"
+version = "0.16.6"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
- "indexmap 2.1.0",
+ "indexmap 2.0.2",
  "rayon",
  "serde_json",
  "snarkvm-console",
@@ -3939,15 +3901,14 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-ledger-coinbase"
-version = "0.16.8"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "6fb7b2acd75abdc7b2f1e1f1cae815c3d710b10ec8dc865aff0257d87ab7908b"
+version = "0.16.6"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
  "aleo-std",
  "anyhow",
  "bincode",
  "blake2",
- "indexmap 2.1.0",
+ "indexmap 2.0.2",
  "rayon",
  "serde_json",
  "snarkvm-algorithms",
@@ -3960,12 +3921,11 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-ledger-committee"
-version = "0.16.8"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "1f4071647a5392a456f40b75c1355221b0863118e84a2600aaecfa67beeba59e"
+version = "0.16.6"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
  "anyhow",
- "indexmap 2.1.0",
+ "indexmap 2.0.2",
  "proptest",
  "rand",
  "rand_chacha",
@@ -3977,9 +3937,8 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-ledger-narwhal"
-version = "0.16.8"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "90cc583a293ecd9137ade263088add52362bec366e3c0cb0b803b5ce4ea4db88"
+version = "0.16.6"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
  "snarkvm-ledger-narwhal-batch-certificate",
  "snarkvm-ledger-narwhal-batch-header",
@@ -3991,11 +3950,10 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-ledger-narwhal-batch-certificate"
-version = "0.16.8"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "e01b3071ae8fdcd198170fe816f2e11b7fda0c7f664ae21bccec3dcb86fc5337"
+version = "0.16.6"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
- "indexmap 2.1.0",
+ "indexmap 2.0.2",
  "serde_json",
  "snarkvm-console",
  "snarkvm-ledger-narwhal-batch-header",
@@ -4004,11 +3962,10 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-ledger-narwhal-batch-header"
-version = "0.16.8"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "05a58fcf7beb74075b5423e7a3d5e6a87228845053df363960309a8cbcb104b0"
+version = "0.16.6"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
- "indexmap 2.1.0",
+ "indexmap 2.0.2",
  "serde_json",
  "snarkvm-console",
  "snarkvm-ledger-narwhal-transmission-id",
@@ -4017,9 +3974,8 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-ledger-narwhal-data"
-version = "0.16.8"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "f7026ee0e8c2c55a8eb1d76e20a8e7845c2b4b833e66ecbddd93af15b1f6a29f"
+version = "0.16.6"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
  "bytes",
  "serde_json",
@@ -4029,11 +3985,10 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-ledger-narwhal-subdag"
-version = "0.16.8"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "fd84dc1e8db4959fd0438d9c9d0d5426de7ce8962fcb6b6838cb85a768b34c36"
+version = "0.16.6"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
- "indexmap 2.1.0",
+ "indexmap 2.0.2",
  "rayon",
  "serde_json",
  "snarkvm-console",
@@ -4043,9 +3998,8 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-ledger-narwhal-transmission"
-version = "0.16.8"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "504e4baf8a07d3464708e2818c02946c63c9259d24bfcc3f039f0c64fab90eca"
+version = "0.16.6"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
  "bytes",
  "serde_json",
@@ -4057,9 +4011,8 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-ledger-narwhal-transmission-id"
-version = "0.16.8"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "0d5773ea758cb790c2c5e7b766590def32ec8bb296e405f316045107148c18ed"
+version = "0.16.6"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
  "snarkvm-console",
  "snarkvm-ledger-coinbase",
@@ -4067,9 +4020,8 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-ledger-query"
-version = "0.16.8"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "280a11c5c2935ac37bce5df2418bc3460d7404696becdb1bf6212ba3a9b5841d"
+version = "0.16.6"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
  "async-trait",
  "reqwest",
@@ -4081,14 +4033,13 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-ledger-store"
-version = "0.16.8"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "d35bda21dfb9436183e0ba33fd5da346badb7d3e3dd9ebd0cc682f1e91e1cdfe"
+version = "0.16.6"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
  "aleo-std",
  "anyhow",
  "bincode",
- "indexmap 2.1.0",
+ "indexmap 2.0.2",
  "once_cell",
  "parking_lot",
  "rayon",
@@ -4107,9 +4058,8 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-ledger-test-helpers"
-version = "0.16.8"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "28440dd7e354d0919598dd59553d7dbc28d3fd7954d731db73c89ff43cbe1e62"
+version = "0.16.6"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
  "once_cell",
  "snarkvm-circuit",
@@ -4123,9 +4073,8 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-parameters"
-version = "0.16.8"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "67404c246abe860aaab5ce85e46ae05cad9851e2a1fa178ad56a9b16cdaeb2d1"
+version = "0.16.6"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
  "aleo-std",
  "anyhow",
@@ -4134,7 +4083,7 @@ dependencies = [
  "colored",
  "curl",
  "hex",
- "indexmap 2.1.0",
+ "indexmap 2.0.2",
  "itertools 0.11.0",
  "lazy_static",
  "paste",
@@ -4148,13 +4097,12 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-synthesizer"
-version = "0.16.8"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "56d1134844e0a46893b39a5985e744ba6bdd21f6829737fb16f529984830f2e7"
+version = "0.16.6"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
  "aleo-std",
  "anyhow",
- "indexmap 2.1.0",
+ "indexmap 2.0.2",
  "parking_lot",
  "rand",
  "rayon",
@@ -4174,13 +4122,12 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-synthesizer-process"
-version = "0.16.8"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "c0d0237b5c12a990a626952467adb4724570fe7cfc876724b00d26f0e201219c"
+version = "0.16.6"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
  "aleo-std",
  "colored",
- "indexmap 2.1.0",
+ "indexmap 2.0.2",
  "once_cell",
  "parking_lot",
  "rand",
@@ -4197,11 +4144,10 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-synthesizer-program"
-version = "0.16.8"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "2c100528631557457c84ddcfa7981e1de5c1574c916b1a36388faa9a859b02a3"
+version = "0.16.6"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
- "indexmap 2.1.0",
+ "indexmap 2.0.2",
  "paste",
  "rand",
  "rand_chacha",
@@ -4212,9 +4158,8 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-synthesizer-snark"
-version = "0.16.8"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "e38719ab2925cff5002e8447f62327e89be64c939d6849556224fca473918c7c"
+version = "0.16.6"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
  "bincode",
  "once_cell",
@@ -4226,9 +4171,8 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-utilities"
-version = "0.16.8"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "4d4b78dc59206e7006ec535fce12e7b9a5ee3dfbdeac823b7a1dde1cfc9314af"
+version = "0.16.6"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
  "aleo-std",
  "anyhow",
@@ -4248,9 +4192,8 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-utilities-derives"
-version = "0.16.8"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "dd421915b54ec20eed02007c1304d6fc10fa600606674e5f7a2a6358d84a618a"
+version = "0.16.6"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
  "proc-macro2",
  "quote 1.0.33",
diff --git a/Cargo.toml b/Cargo.toml
index e1c0d99bf..16023389c 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkos"
-version = "2.2.4"
+version = "2.2.2"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "A decentralized operating system"
 homepage = "https://aleo.org"
@@ -40,7 +40,9 @@ members = [
 ]
 
 [workspace.dependencies.snarkvm]
-version = "=0.16.8"
+#version = "=0.16.6"
+git = "https://github.com/AleoHQ/snarkVM.git"
+rev = "2e245e3"
 features = [ "circuit", "console", "rocks" ]
 
 [[bin]]
@@ -62,47 +64,47 @@ version = "1"
 
 [dependencies.snarkos-account]
 path = "./account"
-version = "=2.2.4"
+version = "=2.2.2"
 
 [dependencies.snarkos-cli]
 path = "./cli"
-version = "=2.2.4"
+version = "=2.2.2"
 
 [dependencies.snarkos-display]
 path = "./display"
-version = "=2.2.4"
+version = "=2.2.2"
 
 [dependencies.snarkos-node]
 path = "./node"
-version = "=2.2.4"
+version = "=2.2.2"
 
 [dependencies.snarkos-node-bft]
 path = "./node/bft"
-version = "=2.2.4"
+version = "=2.2.2"
 
 [dependencies.snarkos-node-cdn]
 path = "./node/cdn"
-version = "=2.2.4"
+version = "=2.2.2"
 
 [dependencies.snarkos-node-consensus]
 path = "./node/consensus"
-version = "=2.2.4"
+version = "=2.2.2"
 
 [dependencies.snarkos-node-rest]
 path = "./node/rest"
-version = "=2.2.4"
+version = "=2.2.2"
 
 [dependencies.snarkos-node-router]
 path = "./node/router"
-version = "=2.2.4"
+version = "=2.2.2"
 
 [dependencies.snarkos-node-sync]
 path = "./node/sync"
-version = "=2.2.4"
+version = "=2.2.2"
 
 [dependencies.snarkos-node-tcp]
 path = "./node/tcp"
-version = "=2.2.4"
+version = "=2.2.2"
 
 [dependencies.tikv-jemallocator]
 version = "0.5"
diff --git a/account/Cargo.toml b/account/Cargo.toml
index 432069754..66fb7ff0a 100644
--- a/account/Cargo.toml
+++ b/account/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkos-account"
-version = "2.2.4"
+version = "2.2.2"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Account for a decentralized operating system"
 homepage = "https://aleo.org"
diff --git a/cli/Cargo.toml b/cli/Cargo.toml
index 6e8350382..f2b2667c6 100644
--- a/cli/Cargo.toml
+++ b/cli/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkos-cli"
-version = "2.2.4"
+version = "2.2.2"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "A CLI for a decentralized operating system"
 homepage = "https://aleo.org"
@@ -40,7 +40,7 @@ version = "2"
 version = "0.27"
 
 [dependencies.indexmap]
-version = "2.1"
+version = "2.0"
 features = [ "serde", "rayon" ]
 
 [dependencies.num_cpus]
@@ -72,23 +72,23 @@ features = [ "preserve_order" ]
 
 [dependencies.snarkos-account]
 path = "../account"
-version = "=2.2.4"
+version = "=2.2.2"
 
 [dependencies.snarkos-display]
 path = "../display"
-version = "=2.2.4"
+version = "=2.2.2"
 
 [dependencies.snarkos-node]
 path = "../node"
-version = "=2.2.4"
+version = "=2.2.2"
 
 [dependencies.snarkos-node-cdn]
 path = "../node/cdn"
-version = "=2.2.4"
+version = "=2.2.2"
 
 [dependencies.snarkos-node-rest]
 path = "../node/rest"
-version = "=2.2.4"
+version = "=2.2.2"
 
 [dependencies.snarkvm]
 workspace = true
diff --git a/display/Cargo.toml b/display/Cargo.toml
index 7a72f9481..6ae9e629d 100644
--- a/display/Cargo.toml
+++ b/display/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkos-display"
-version = "2.2.4"
+version = "2.2.2"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "A display for a decentralized operating system"
 homepage = "https://aleo.org"
@@ -24,7 +24,7 @@ version = "0.27"
 
 [dependencies.snarkos-node]
 path = "../node"
-version = "=2.2.4"
+version = "=2.2.2"
 
 [dependencies.snarkvm]
 workspace = true
diff --git a/node/Cargo.toml b/node/Cargo.toml
index 206be5e8b..6af84ca94 100644
--- a/node/Cargo.toml
+++ b/node/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkos-node"
-version = "2.2.4"
+version = "2.2.2"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "A node for a decentralized operating system"
 homepage = "https://aleo.org"
@@ -39,7 +39,7 @@ version = "0.3"
 features = [ "sink" ]
 
 [dependencies.indexmap]
-version = "2.1"
+version = "2.0"
 
 [dependencies.num_cpus]
 version = "1"
@@ -64,35 +64,35 @@ features = [ "preserve_order" ]
 
 [dependencies.snarkos-account]
 path = "../account"
-version = "=2.2.4"
+version = "=2.2.2"
 
 [dependencies.snarkos-node-bft]
 path = "./bft"
-version = "=2.2.4"
+version = "=2.2.2"
 
 [dependencies.snarkos-node-cdn]
 path = "./cdn"
-version = "=2.2.4"
+version = "=2.2.2"
 
 [dependencies.snarkos-node-consensus]
 path = "./consensus"
-version = "=2.2.4"
+version = "=2.2.2"
 
 [dependencies.snarkos-node-rest]
 path = "./rest"
-version = "=2.2.4"
+version = "=2.2.2"
 
 [dependencies.snarkos-node-router]
 path = "./router"
-version = "=2.2.4"
+version = "=2.2.2"
 
 [dependencies.snarkos-node-sync]
 path = "./sync"
-version = "=2.2.4"
+version = "=2.2.2"
 
 [dependencies.snarkos-node-tcp]
 path = "./tcp"
-version = "=2.2.4"
+version = "=2.2.2"
 
 [dependencies.snarkvm]
 workspace = true
diff --git a/node/bft/Cargo.toml b/node/bft/Cargo.toml
index 9e1b3e1cc..8c782afcf 100644
--- a/node/bft/Cargo.toml
+++ b/node/bft/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkos-node-bft"
-version = "2.2.4"
+version = "2.2.2"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "A memory pool for a decentralized operating system"
 homepage = "https://aleo.org"
@@ -36,7 +36,7 @@ version = "0.3.29"
 features = [ "thread-pool" ]
 
 [dependencies.indexmap]
-version = "2.1"
+version = "2.0"
 features = [ "serde", "rayon" ]
 
 [dependencies.parking_lot]
@@ -57,24 +57,24 @@ default-features = false
 
 [dependencies.snarkos-account]
 path = "../../account"
-version = "=2.2.4"
+version = "=2.2.2"
 
 [dependencies.snarkos-node-bft-events]
 path = "./events"
-version = "=2.2.4"
+version = "=2.2.2"
 
 [dependencies.snarkos-node-bft-ledger-service]
 path = "./ledger-service"
-version = "=2.2.4"
+version = "=2.2.2"
 features = [ "ledger", "prover" ]
 
 [dependencies.snarkos-node-sync]
 path = "../sync"
-version = "=2.2.4"
+version = "=2.2.2"
 
 [dependencies.snarkos-node-tcp]
 path = "../tcp"
-version = "=2.2.4"
+version = "=2.2.2"
 
 [dependencies.snarkvm]
 workspace = true
diff --git a/node/bft/events/Cargo.toml b/node/bft/events/Cargo.toml
index b0b45ad1b..0e3eb673f 100644
--- a/node/bft/events/Cargo.toml
+++ b/node/bft/events/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkos-node-bft-events"
-version = "2.2.4"
+version = "2.2.2"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Events for the gateway in a decentralized operating system"
 homepage = "https://aleo.org"
@@ -23,7 +23,7 @@ version = "1.0"
 version = "1"
 
 [dependencies.indexmap]
-version = "2.1"
+version = "2.0"
 features = [ "serde", "rayon" ]
 
 [dependencies.rayon]
@@ -34,7 +34,7 @@ version = "1"
 
 [dependencies.snarkos-node-sync-locators]
 path = "../../sync/locators"
-version = "=2.2.4"
+version = "=2.2.2"
 
 [dependencies.snarkvm]
 workspace = true
diff --git a/node/bft/ledger-service/Cargo.toml b/node/bft/ledger-service/Cargo.toml
index a715efbbd..cf9bdaee6 100644
--- a/node/bft/ledger-service/Cargo.toml
+++ b/node/bft/ledger-service/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkos-node-bft-ledger-service"
-version = "2.2.4"
+version = "2.2.2"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "A ledger service for the memory pool in a decentralized operating system"
 homepage = "https://aleo.org"
@@ -29,7 +29,7 @@ translucent = [ "ledger" ]
 version = "0.1"
 
 [dependencies.indexmap]
-version = "2.1"
+version = "2.0"
 features = [ "serde", "rayon" ]
 
 [dependencies.parking_lot]
diff --git a/node/bft/src/bft.rs b/node/bft/src/bft.rs
index 593130899..7f5ec2b00 100644
--- a/node/bft/src/bft.rs
+++ b/node/bft/src/bft.rs
@@ -25,7 +25,7 @@ use crate::{
         DAG,
     },
     Primary,
-    MAX_LEADER_CERTIFICATE_DELAY_IN_SECS,
+    MAX_LEADER_CERTIFICATE_DELAY,
 };
 use snarkos_account::Account;
 use snarkos_node_bft_ledger_service::LedgerService;
@@ -321,7 +321,7 @@ impl<N: Network> BFT<N> {
 
     /// Returns `true` if the timer for the leader certificate has expired.
     fn is_timer_expired(&self) -> bool {
-        self.leader_certificate_timer.load(Ordering::SeqCst) + MAX_LEADER_CERTIFICATE_DELAY_IN_SECS <= now()
+        self.leader_certificate_timer.load(Ordering::SeqCst) + MAX_LEADER_CERTIFICATE_DELAY <= now()
     }
 
     /// Returns 'true' if any of the following conditions hold:
diff --git a/node/bft/src/gateway.rs b/node/bft/src/gateway.rs
index ba69cf548..cd5d4056a 100644
--- a/node/bft/src/gateway.rs
+++ b/node/bft/src/gateway.rs
@@ -17,7 +17,7 @@ use crate::{
     helpers::{assign_to_worker, Cache, PrimarySender, Resolver, SyncSender, WorkerSender},
     spawn_blocking,
     CONTEXT,
-    MAX_BATCH_DELAY_IN_MS,
+    MAX_BATCH_DELAY,
     MAX_GC_ROUNDS,
     MAX_TRANSMISSIONS_PER_BATCH,
     MAX_TRANSMISSIONS_PER_WORKER_PING,
@@ -71,14 +71,14 @@ use tokio_stream::StreamExt;
 use tokio_util::codec::Framed;
 
 /// The maximum interval of events to cache.
-const CACHE_EVENTS_INTERVAL: i64 = (MAX_BATCH_DELAY_IN_MS / 1000) as i64; // seconds
+const CACHE_EVENTS_INTERVAL: i64 = (MAX_BATCH_DELAY / 1000) as i64; // seconds
 /// The maximum interval of requests to cache.
-const CACHE_REQUESTS_INTERVAL: i64 = (MAX_BATCH_DELAY_IN_MS / 1000) as i64; // seconds
+const CACHE_REQUESTS_INTERVAL: i64 = (MAX_BATCH_DELAY / 1000) as i64; // seconds
 
 /// The maximum number of connection attempts in an interval.
 const MAX_CONNECTION_ATTEMPTS: usize = 10;
 /// The maximum interval to restrict a peer.
-const RESTRICTED_INTERVAL: i64 = (MAX_CONNECTION_ATTEMPTS as u64 * MAX_BATCH_DELAY_IN_MS / 1000) as i64; // seconds
+const RESTRICTED_INTERVAL: i64 = (MAX_CONNECTION_ATTEMPTS as u64 * MAX_BATCH_DELAY / 1000) as i64; // seconds
 
 /// The minimum number of validators to maintain a connection to.
 const MIN_CONNECTED_VALIDATORS: usize = 175;
diff --git a/node/bft/src/lib.rs b/node/bft/src/lib.rs
index 1a580600f..45d24815d 100644
--- a/node/bft/src/lib.rs
+++ b/node/bft/src/lib.rs
@@ -46,11 +46,11 @@ pub const CONTEXT: &str = "[MemoryPool]";
 pub const MEMORY_POOL_PORT: u16 = 5000; // port
 
 /// The maximum number of milliseconds to wait before proposing a batch.
-pub const MAX_BATCH_DELAY_IN_MS: u64 = 2500; // ms
+pub const MAX_BATCH_DELAY: u64 = 2500; // ms
 /// The maximum number of rounds to store before garbage collecting.
 pub const MAX_GC_ROUNDS: u64 = 50; // rounds
 /// The maximum number of seconds allowed for the leader to send their certificate.
-pub const MAX_LEADER_CERTIFICATE_DELAY_IN_SECS: i64 = 2 * MAX_BATCH_DELAY_IN_MS as i64 / 1000; // seconds
+pub const MAX_LEADER_CERTIFICATE_DELAY: i64 = 2 * MAX_BATCH_DELAY as i64 / 1000; // seconds
 /// The maximum number of seconds before the timestamp is considered expired.
 pub const MAX_TIMESTAMP_DELTA_IN_SECS: i64 = 10; // seconds
 /// The maximum number of transmissions allowed in a batch.
@@ -61,9 +61,9 @@ pub const MAX_TRANSMISSIONS_PER_WORKER_PING: usize = MAX_TRANSMISSIONS_PER_BATCH
 pub const MAX_WORKERS: u8 = 1; // workers
 
 /// The frequency at which each primary broadcasts a ping to every other node.
-pub const PRIMARY_PING_IN_MS: u64 = 2 * MAX_BATCH_DELAY_IN_MS; // ms
+pub const PRIMARY_PING_INTERVAL: u64 = 2 * MAX_BATCH_DELAY; // ms
 /// The frequency at which each worker broadcasts a ping to every other node.
-pub const WORKER_PING_IN_MS: u64 = 4 * MAX_BATCH_DELAY_IN_MS; // ms
+pub const WORKER_PING_INTERVAL: u64 = 4 * MAX_BATCH_DELAY; // ms
 
 /// A helper macro to spawn a blocking task.
 #[macro_export]
diff --git a/node/bft/src/primary.rs b/node/bft/src/primary.rs
index ba139799e..a47ecc777 100644
--- a/node/bft/src/primary.rs
+++ b/node/bft/src/primary.rs
@@ -32,18 +32,17 @@ use crate::{
     Sync,
     Transport,
     Worker,
-    MAX_BATCH_DELAY_IN_MS,
+    MAX_BATCH_DELAY,
     MAX_TRANSMISSIONS_PER_BATCH,
     MAX_WORKERS,
-    PRIMARY_PING_IN_MS,
-    WORKER_PING_IN_MS,
+    PRIMARY_PING_INTERVAL,
+    WORKER_PING_INTERVAL,
 };
 use snarkos_account::Account;
 use snarkos_node_bft_events::PrimaryPing;
 use snarkos_node_bft_ledger_service::LedgerService;
 use snarkvm::{
     console::{
-        account::Signature,
         prelude::*,
         types::{Address, Field},
     },
@@ -90,12 +89,12 @@ pub struct Primary<N: Network> {
     bft_sender: Arc<OnceCell<BFTSender<N>>>,
     /// The batch proposal, if the primary is currently proposing a batch.
     proposed_batch: Arc<ProposedBatch<N>>,
-    /// The recently-signed batch proposals (a map from the address to the round, batch ID, and signature).
-    signed_proposals: Arc<RwLock<HashMap<Address<N>, (u64, Field<N>, Signature<N>)>>>,
+    /// The recently-signed batch proposals (a map from the address to the round and batch ID).
+    signed_proposals: Arc<RwLock<HashMap<Address<N>, (u64, Field<N>)>>>,
     /// The spawned handles.
     handles: Arc<Mutex<Vec<JoinHandle<()>>>>,
-    /// The lock for propose_batch.
-    propose_lock: Arc<TMutex<u64>>,
+    /// The primary lock.
+    lock: Arc<TMutex<()>>,
 }
 
 impl<N: Network> Primary<N> {
@@ -123,7 +122,7 @@ impl<N: Network> Primary<N> {
             proposed_batch: Default::default(),
             signed_proposals: Default::default(),
             handles: Default::default(),
-            propose_lock: Default::default(),
+            lock: Default::default(),
         })
     }
 
@@ -270,9 +269,6 @@ impl<N: Network> Primary<N> {
     /// 3. Set the batch proposal in the primary.
     /// 4. Broadcast the batch header to all validators for signing.
     pub async fn propose_batch(&self) -> Result<()> {
-        // This function isn't re-entrant.
-        let mut lock_guard = self.propose_lock.lock().await;
-
         // Check if the proposed batch has expired, and clear it if it has expired.
         if let Err(e) = self.check_proposed_batch_for_expiration().await {
             warn!("Failed to check the proposed batch for expiration - {e}");
@@ -289,15 +285,14 @@ impl<N: Network> Primary<N> {
             for address in proposal.nonsigners(&self.ledger.get_previous_committee_for_round(proposal.round())?) {
                 // Resolve the address to the peer IP.
                 match self.gateway.resolver().get_peer_ip_for_address(address) {
-                    // Resend the batch proposal to the validator for signing.
+                    // Broadcast the batch to all validators for signing.
                     Some(peer_ip) => {
-                        let (gateway, event_, round) = (self.gateway.clone(), event.clone(), proposal.round());
+                        debug!("Resending batch proposal for round {} to peer '{peer_ip}'", proposal.round());
+                        // Broadcast the event.
+                        let self_ = self.clone();
+                        let event_ = event.clone();
                         tokio::spawn(async move {
-                            debug!("Resending batch proposal for round {round} to peer '{peer_ip}'");
-                            // Resend the batch proposal to the peer.
-                            if gateway.send(peer_ip, event_).await.is_none() {
-                                warn!("Failed to resend batch proposal for round {round} to peer '{peer_ip}'");
-                            }
+                            let _ = self_.gateway.send(peer_ip, event_).await;
                         });
                     }
                     None => continue,
@@ -430,14 +425,6 @@ impl<N: Network> Primary<N> {
             debug!("Primary is safely skipping a batch proposal {}", "(no unconfirmed transactions)".dimmed());
             return Ok(());
         }
-        // Ditto if the batch had already been proposed.
-        ensure!(round > 0, "Round 0 cannot have transaction batches");
-        if *lock_guard == round {
-            warn!("Primary is safely skipping a batch proposal - round {round} already proposed");
-            return Ok(());
-        }
-
-        *lock_guard = round;
 
         /* Proceeding to sign & propose the batch. */
         info!("Proposing a batch with {} transmissions for round {round}...", transmissions.len());
@@ -488,35 +475,28 @@ impl<N: Network> Primary<N> {
             bail!("Malicious peer - proposed round {batch_round}, but sent batch for round {}", batch_header.round());
         }
 
-        // Retrieve the batch author.
-        let batch_author = batch_header.author();
+        // // Acquire the lock.
+        // let _lock = self.lock.lock().await;
 
         // Ensure the batch proposal is from the validator.
-        match self.gateway.resolver().get_address(peer_ip) {
-            // If the peer is a validator, then ensure the batch proposal is from the validator.
-            Some(address) => {
-                if address != batch_author {
-                    // Proceed to disconnect the validator.
-                    self.gateway.disconnect(peer_ip);
-                    bail!("Malicious peer - proposed batch from a different validator ({batch_author})");
-                }
-            }
-            None => bail!("Batch proposal from a disconnected validator"),
+        if self.gateway.resolver().get_address(peer_ip).map_or(true, |address| address != batch_header.author()) {
+            // Proceed to disconnect the validator.
+            self.gateway.disconnect(peer_ip);
+            bail!("Malicious peer - proposed batch from a different validator ({})", batch_header.author());
         }
         // Ensure the batch author is a current committee member.
-        if !self.gateway.is_authorized_validator_address(batch_author) {
+        if !self.gateway.is_authorized_validator_address(batch_header.author()) {
             // Proceed to disconnect the validator.
             self.gateway.disconnect(peer_ip);
-            bail!("Malicious peer - proposed batch from a non-committee member ({batch_author})");
+            bail!("Malicious peer - proposed batch from a non-committee member ({})", batch_header.author());
         }
         // Ensure the batch proposal is not from the current primary.
-        if self.gateway.account().address() == batch_author {
-            bail!("Invalid peer - proposed batch from myself ({batch_author})");
+        if self.gateway.account().address() == batch_header.author() {
+            bail!("Invalid peer - proposed batch from myself ({})", batch_header.author());
         }
 
         // Retrieve the cached round and batch ID for this validator.
-        if let Some((signed_round, signed_batch_id, signature)) =
-            self.signed_proposals.read().get(&batch_author).copied()
+        if let Some((signed_round, signed_batch_id)) = self.signed_proposals.read().get(&batch_header.author()).copied()
         {
             // If the round matches and the batch ID differs, then the validator is malicious.
             if signed_round == batch_header.round() && signed_batch_id != batch_header.batch_id() {
@@ -525,18 +505,8 @@ impl<N: Network> Primary<N> {
                 bail!("Malicious peer - proposed another batch for the same round ({signed_round})");
             }
             // If the round and batch ID matches, then skip signing the batch a second time.
-            // Instead, rebroadcast the cached signature to the peer.
             if signed_round == batch_header.round() && signed_batch_id == batch_header.batch_id() {
-                let gateway = self.gateway.clone();
-                tokio::spawn(async move {
-                    debug!("Resending a signature for a batch in round {batch_round} from '{peer_ip}'");
-                    let event = Event::BatchSignature(BatchSignature::new(batch_header.batch_id(), signature));
-                    // Resend the batch signature to the peer.
-                    if gateway.send(peer_ip, event).await.is_none() {
-                        warn!("Failed to resend a signature for a batch in round {batch_round} to '{peer_ip}'");
-                    }
-                });
-                // Return early.
+                debug!("Skipping a proposal for round {signed_round} from '{peer_ip}' {}", "(already signed)".dimmed());
                 return Ok(());
             }
         }
@@ -561,31 +531,8 @@ impl<N: Network> Primary<N> {
         // Sign the batch ID.
         let account = self.gateway.account().clone();
         let signature = spawn_blocking!(account.sign(&[batch_id], &mut rand::thread_rng()))?;
-
-        // Ensure the proposal has not already been signed.
-        //
-        // Note: Due to the need to sync the batch header with the peer, it is possible
-        // for the primary to receive the same 'BatchPropose' event again, whereby only
-        // one instance of this handler should sign the batch. This check guarantees this.
-        match self.signed_proposals.write().entry(batch_author) {
-            std::collections::hash_map::Entry::Occupied(mut entry) => {
-                // If the validator has already signed a batch for this round, then return early,
-                // since, if the peer still has not received the signature, they will request it again,
-                // and the logic at the start of this function will resend the (now cached) signature
-                // to the peer if asked to sign this batch proposal again.
-                if entry.get().0 == batch_round {
-                    return Ok(());
-                }
-                // Otherwise, cache the round, batch ID, and signature for this validator.
-                entry.insert((batch_round, batch_id, signature));
-            }
-            // If the validator has not signed a batch before, then continue.
-            std::collections::hash_map::Entry::Vacant(entry) => {
-                // Cache the round, batch ID, and signature for this validator.
-                entry.insert((batch_round, batch_id, signature));
-            }
-        };
-
+        // Cache the round and batch ID for this validator.
+        self.signed_proposals.write().insert(batch_header.author(), (batch_header.round(), batch_id));
         // Broadcast the signature back to the validator.
         let self_ = self.clone();
         tokio::spawn(async move {
@@ -631,6 +578,9 @@ impl<N: Network> Primary<N> {
             bail!("Invalid peer - received a batch signature from myself ({signer})");
         }
 
+        // // Acquire the lock.
+        // let _lock = self.lock.lock().await;
+
         let proposal = {
             // Acquire the write lock.
             let mut proposed_batch = self.proposed_batch.write();
@@ -705,20 +655,21 @@ impl<N: Network> Primary<N> {
             return Ok(());
         }
 
+        // // Acquire the lock.
+        // let _lock = self.lock.lock().await;
+
         // Retrieve the batch certificate author.
         let author = certificate.author();
 
         // Ensure the batch certificate is from the validator.
         match self.gateway.resolver().get_address(peer_ip) {
-            // If the peer is a validator, then ensure the batch certificate is from the validator.
-            Some(address) => {
-                if address != author {
-                    // Proceed to disconnect the validator.
-                    self.gateway.disconnect(peer_ip);
-                    bail!("Malicious peer - batch certificate from a different validator ({author})");
-                }
+            Some(address) if address != author => {
+                // Proceed to disconnect the validator.
+                self.gateway.disconnect(peer_ip);
+                bail!("Malicious peer - batch certificate from a different validator ({author})");
             }
             None => bail!("Batch certificate from a disconnected validator"),
+            _ => (),
         }
         // Ensure the batch certificate is authored by a current committee member.
         if !self.gateway.is_authorized_validator_address(author) {
@@ -777,6 +728,9 @@ impl<N: Network> Primary<N> {
             return Ok(());
         }
 
+        // // Acquire the lock.
+        // let _lock = self.lock.lock().await;
+
         // Retrieve the batch certificate author.
         let author = certificate.author();
 
@@ -811,7 +765,7 @@ impl<N: Network> Primary<N> {
             self.spawn(async move {
                 loop {
                     // Sleep briefly.
-                    tokio::time::sleep(Duration::from_millis(PRIMARY_PING_IN_MS)).await;
+                    tokio::time::sleep(Duration::from_millis(PRIMARY_PING_INTERVAL)).await;
 
                     // Retrieve the block locators.
                     let block_locators = match self_.sync.get_block_locators() {
@@ -946,7 +900,7 @@ impl<N: Network> Primary<N> {
             let self_ = self.clone();
             self.spawn(async move {
                 loop {
-                    tokio::time::sleep(Duration::from_millis(WORKER_PING_IN_MS)).await;
+                    tokio::time::sleep(Duration::from_millis(WORKER_PING_INTERVAL)).await;
                     // If the primary is not synced, then do not broadcast the worker ping(s).
                     if !self_.sync.is_synced() {
                         trace!("Skipping worker ping(s) {}", "(node is syncing)".dimmed());
@@ -965,15 +919,13 @@ impl<N: Network> Primary<N> {
         self.spawn(async move {
             loop {
                 // Sleep briefly, but longer than if there were no batch.
-                tokio::time::sleep(Duration::from_millis(MAX_BATCH_DELAY_IN_MS)).await;
+                tokio::time::sleep(Duration::from_millis(MAX_BATCH_DELAY)).await;
                 // If the primary is not synced, then do not propose a batch.
                 if !self_.sync.is_synced() {
                     debug!("Skipping batch proposal {}", "(node is syncing)".dimmed());
                     continue;
                 }
                 // If there is no proposed batch, attempt to propose a batch.
-                // Note: Do NOT spawn a task around this function call. Proposing a batch is a critical path,
-                // and only one batch needs be proposed at a time.
                 if let Err(e) = self_.propose_batch().await {
                     warn!("Cannot propose a batch - {e}");
                 }
@@ -989,14 +941,10 @@ impl<N: Network> Primary<N> {
                     trace!("Skipping a batch proposal from '{peer_ip}' {}", "(node is syncing)".dimmed());
                     continue;
                 }
-                // Spawn a task to process the proposed batch.
-                let self_ = self_.clone();
-                tokio::spawn(async move {
-                    // Process the batch proposal.
-                    if let Err(e) = self_.process_batch_propose_from_peer(peer_ip, batch_propose).await {
-                        warn!("Cannot sign a batch from '{peer_ip}' - {e}");
-                    }
-                });
+                // Process the proposed batch.
+                if let Err(e) = self_.process_batch_propose_from_peer(peer_ip, batch_propose).await {
+                    warn!("Cannot sign a batch from '{peer_ip}' - {e}");
+                }
             }
         });
 
@@ -1010,10 +958,6 @@ impl<N: Network> Primary<N> {
                     continue;
                 }
                 // Process the batch signature.
-                // Note: Do NOT spawn a task around this function call. Processing signatures from peers
-                // is a critical path, and we should only store the minimum required number of signatures.
-                // In addition, spawning a task can cause concurrent processing of signatures (even with a lock),
-                // which means the RwLock for the proposed batch must become a 'tokio::sync' to be safe.
                 if let Err(e) = self_.process_batch_signature_from_peer(peer_ip, batch_signature).await {
                     warn!("Cannot store a signature from '{peer_ip}' - {e}");
                 }
@@ -1029,19 +973,17 @@ impl<N: Network> Primary<N> {
                     trace!("Skipping a certified batch from '{peer_ip}' {}", "(node is syncing)".dimmed());
                     continue;
                 }
-                // Spawn a task to process the batch certificate.
-                let self_ = self_.clone();
-                tokio::spawn(async move {
-                    // Deserialize the batch certificate.
-                    let Ok(batch_certificate) = spawn_blocking!(batch_certificate.deserialize_blocking()) else {
-                        warn!("Failed to deserialize the batch certificate from '{peer_ip}'");
-                        return;
-                    };
-                    // Process the batch certificate.
-                    if let Err(e) = self_.process_batch_certificate_from_peer(peer_ip, batch_certificate).await {
-                        warn!("Cannot store a certificate from '{peer_ip}' - {e}");
-                    }
-                });
+
+                // Deserialize the batch certificate.
+                let Ok(batch_certificate) = spawn_blocking!(batch_certificate.deserialize_blocking()) else {
+                    warn!("Failed to deserialize the batch certificate from '{peer_ip}'");
+                    continue;
+                };
+
+                // Process the batch certificate.
+                if let Err(e) = self_.process_batch_certificate_from_peer(peer_ip, batch_certificate).await {
+                    warn!("Cannot store a certificate from '{peer_ip}' - {e}");
+                }
             }
         });
 
@@ -1298,12 +1240,12 @@ impl<N: Network> Primary<N> {
         }
 
         // Check if our primary should move to the next round.
-        // TODO (howardwu): Re-evaluate whether we need to guard this to increment after quorum threshold is reached.
-        let is_behind_schedule = batch_round > self.current_round();
+        let is_behind_schedule = batch_round > self.current_round(); // TODO: Check if threshold is reached.
         // Check if our primary is far behind the peer.
         let is_peer_far_in_future = batch_round > self.current_round() + self.storage.max_gc_rounds();
         // If our primary is far behind the peer, update our committee to the batch round.
         if is_behind_schedule || is_peer_far_in_future {
+            // TODO (howardwu): Guard this to increment after quorum threshold is reached.
             // If the batch round is greater than the current committee round, update the committee.
             self.try_increment_to_the_next_round(batch_round).await?;
         }
@@ -1438,6 +1380,8 @@ impl<N: Network> Primary<N> {
     /// Shuts down the primary.
     pub async fn shut_down(&self) {
         info!("Shutting down the primary...");
+        // Acquire the lock.
+        let _lock = self.lock.lock().await;
         // Shut down the workers.
         self.workers.iter().for_each(|worker| worker.shut_down());
         // Abort the tasks.
diff --git a/node/bft/src/sync/mod.rs b/node/bft/src/sync/mod.rs
index 1f93bdf5c..9db3bee33 100644
--- a/node/bft/src/sync/mod.rs
+++ b/node/bft/src/sync/mod.rs
@@ -16,8 +16,8 @@ use crate::{
     helpers::{BFTSender, Pending, Storage, SyncReceiver},
     Gateway,
     Transport,
-    MAX_BATCH_DELAY_IN_MS,
-    PRIMARY_PING_IN_MS,
+    MAX_BATCH_DELAY,
+    PRIMARY_PING_INTERVAL,
 };
 use snarkos_node_bft_events::{CertificateRequest, CertificateResponse, Event};
 use snarkos_node_bft_ledger_service::LedgerService;
@@ -92,7 +92,7 @@ impl<N: Network> Sync<N> {
         self.handles.lock().push(tokio::spawn(async move {
             loop {
                 // Sleep briefly to avoid triggering spam detection.
-                tokio::time::sleep(std::time::Duration::from_millis(PRIMARY_PING_IN_MS)).await;
+                tokio::time::sleep(std::time::Duration::from_secs(PRIMARY_PING_INTERVAL)).await;
                 // Perform the sync routine.
                 let communication = &self_.gateway;
                 // let communication = &node.router;
@@ -339,7 +339,7 @@ impl<N: Network> Sync<N> {
             }
         }
         // Wait for the certificate to be fetched.
-        match tokio::time::timeout(core::time::Duration::from_millis(MAX_BATCH_DELAY_IN_MS), callback_receiver).await {
+        match tokio::time::timeout(core::time::Duration::from_millis(MAX_BATCH_DELAY), callback_receiver).await {
             // If the certificate was fetched, return it.
             Ok(result) => Ok(result?),
             // If the certificate was not fetched, return an error.
diff --git a/node/bft/src/worker.rs b/node/bft/src/worker.rs
index 685fde8f2..7f3da6204 100644
--- a/node/bft/src/worker.rs
+++ b/node/bft/src/worker.rs
@@ -17,7 +17,7 @@ use crate::{
     helpers::{fmt_id, Pending, Ready, Storage, WorkerReceiver},
     ProposedBatch,
     Transport,
-    MAX_BATCH_DELAY_IN_MS,
+    MAX_BATCH_DELAY,
     MAX_TRANSMISSIONS_PER_BATCH,
     MAX_TRANSMISSIONS_PER_WORKER_PING,
     MAX_WORKERS,
@@ -383,7 +383,7 @@ impl<N: Network> Worker<N> {
             bail!("Unable to fetch transmission - failed to send request")
         }
         // Wait for the transmission to be fetched.
-        match timeout(Duration::from_millis(MAX_BATCH_DELAY_IN_MS), callback_receiver).await {
+        match timeout(Duration::from_millis(MAX_BATCH_DELAY), callback_receiver).await {
             // If the transmission was fetched, return it.
             Ok(result) => Ok((transmission_id, result?)),
             // If the transmission was not fetched, return an error.
diff --git a/node/bft/tests/bft_e2e.rs b/node/bft/tests/bft_e2e.rs
index 796cd50d4..70192af13 100644
--- a/node/bft/tests/bft_e2e.rs
+++ b/node/bft/tests/bft_e2e.rs
@@ -18,7 +18,7 @@ mod common;
 use crate::common::primary::{TestNetwork, TestNetworkConfig};
 use deadline::deadline;
 use itertools::Itertools;
-use snarkos_node_bft::MAX_BATCH_DELAY_IN_MS;
+use snarkos_node_bft::MAX_BATCH_DELAY;
 use std::time::Duration;
 use tokio::time::sleep;
 
@@ -114,7 +114,7 @@ async fn test_quorum_threshold() {
     // Start the cannons for node 0.
     network.fire_transmissions_at(0, TRANSMISSION_INTERVAL_MS);
 
-    sleep(Duration::from_millis(MAX_BATCH_DELAY_IN_MS * 2)).await;
+    sleep(Duration::from_millis(MAX_BATCH_DELAY * 2)).await;
 
     // Check each node is still at round 1.
     for validator in network.validators.values() {
@@ -125,7 +125,7 @@ async fn test_quorum_threshold() {
     network.connect_validators(0, 1).await;
     network.fire_transmissions_at(1, TRANSMISSION_INTERVAL_MS);
 
-    sleep(Duration::from_millis(MAX_BATCH_DELAY_IN_MS * 2)).await;
+    sleep(Duration::from_millis(MAX_BATCH_DELAY * 2)).await;
 
     // Check each node is still at round 1.
     for validator in network.validators.values() {
diff --git a/node/bft/tests/common/primary.rs b/node/bft/tests/common/primary.rs
index 344847291..9e0350405 100644
--- a/node/bft/tests/common/primary.rs
+++ b/node/bft/tests/common/primary.rs
@@ -22,7 +22,7 @@ use snarkos_node_bft::{
     helpers::{init_primary_channels, PrimarySender, Storage},
     Primary,
     BFT,
-    MAX_BATCH_DELAY_IN_MS,
+    MAX_BATCH_DELAY,
     MAX_GC_ROUNDS,
 };
 use snarkvm::{
@@ -281,7 +281,7 @@ impl TestNetwork {
     // Checks if all the nodes have stopped progressing.
     pub async fn is_halted(&self) -> bool {
         let halt_round = self.validators.values().map(|v| v.primary.current_round()).max().unwrap();
-        sleep(Duration::from_millis(MAX_BATCH_DELAY_IN_MS * 2)).await;
+        sleep(Duration::from_millis(MAX_BATCH_DELAY * 2)).await;
         self.validators.values().all(|v| v.primary.current_round() <= halt_round)
     }
 
diff --git a/node/bft/tests/narwhal_e2e.rs b/node/bft/tests/narwhal_e2e.rs
index b2783b1d9..2c4d90e83 100644
--- a/node/bft/tests/narwhal_e2e.rs
+++ b/node/bft/tests/narwhal_e2e.rs
@@ -16,7 +16,7 @@
 mod common;
 
 use crate::common::primary::{TestNetwork, TestNetworkConfig};
-use snarkos_node_bft::MAX_BATCH_DELAY_IN_MS;
+use snarkos_node_bft::MAX_BATCH_DELAY;
 
 use std::time::Duration;
 
@@ -72,7 +72,7 @@ async fn test_quorum_threshold() {
     // Start the cannons for node 0.
     network.fire_transmissions_at(0, TRANSMISSION_INTERVAL_MS);
 
-    sleep(Duration::from_millis(MAX_BATCH_DELAY_IN_MS * 2)).await;
+    sleep(Duration::from_millis(MAX_BATCH_DELAY * 2)).await;
 
     // Check each node is still at round 1.
     for validator in network.validators.values() {
@@ -83,7 +83,7 @@ async fn test_quorum_threshold() {
     network.connect_validators(0, 1).await;
     network.fire_transmissions_at(1, TRANSMISSION_INTERVAL_MS);
 
-    sleep(Duration::from_millis(MAX_BATCH_DELAY_IN_MS * 2)).await;
+    sleep(Duration::from_millis(MAX_BATCH_DELAY * 2)).await;
 
     // Check each node is still at round 1.
     for validator in network.validators.values() {
diff --git a/node/cdn/Cargo.toml b/node/cdn/Cargo.toml
index 455abc300..5c1379eb4 100644
--- a/node/cdn/Cargo.toml
+++ b/node/cdn/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkos-node-cdn"
-version = "2.2.4"
+version = "2.2.2"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "A CDN client for a decentralized operating system"
 homepage = "https://aleo.org"
diff --git a/node/consensus/Cargo.toml b/node/consensus/Cargo.toml
index 21002b9a2..15114fd30 100644
--- a/node/consensus/Cargo.toml
+++ b/node/consensus/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkos-node-consensus"
-version = "2.2.4"
+version = "2.2.2"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "A node consensus for a decentralized operating system"
 homepage = "https://aleo.org"
@@ -26,7 +26,7 @@ version = "1.0.75"
 version = "2"
 
 [dependencies.indexmap]
-version = "2.1"
+version = "2.0"
 features = [ "serde", "rayon" ]
 
 [dependencies.lru]
@@ -40,15 +40,15 @@ version = "0.8"
 
 [dependencies.snarkos-account]
 path = "../../account"
-version = "=2.2.4"
+version = "=2.2.2"
 
 [dependencies.snarkos-node-bft]
 path = "../bft"
-version = "=2.2.4"
+version = "=2.2.2"
 
 [dependencies.snarkos-node-bft-ledger-service]
 path = "../bft/ledger-service"
-version = "=2.2.4"
+version = "=2.2.2"
 default-features = false
 features = [ "ledger", "ledger-write" ]
 
diff --git a/node/consensus/src/lib.rs b/node/consensus/src/lib.rs
index 5f6a004f5..332891073 100644
--- a/node/consensus/src/lib.rs
+++ b/node/consensus/src/lib.rs
@@ -192,10 +192,9 @@ impl<N: Network> Consensus<N> {
             }
             // Add the solution to the memory pool.
             trace!("Received unconfirmed solution '{}' in the queue", fmt_id(solution_id));
-            // TODO (howardwu) - Attention, this is being disabled temporarily until transmission storage is updated.
-            // if self.solutions_queue.lock().insert(solution_id, solution).is_some() {
-            //     bail!("Solution '{}' exists in the memory pool", fmt_id(solution_id));
-            // }
+            if self.solutions_queue.lock().insert(solution_id, solution).is_some() {
+                bail!("Solution '{}' exists in the memory pool", fmt_id(solution_id));
+            }
         }
 
         // If the memory pool of this node is full, return early.
diff --git a/node/metrics/Cargo.toml b/node/metrics/Cargo.toml
index 72969e74a..980292263 100644
--- a/node/metrics/Cargo.toml
+++ b/node/metrics/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkos-node-metrics"
-version = "2.2.4"
+version = "2.2.2"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "A node for a decentralized operating system"
 homepage = "https://aleo.org"
diff --git a/node/rest/Cargo.toml b/node/rest/Cargo.toml
index 7d5be7199..17a430653 100644
--- a/node/rest/Cargo.toml
+++ b/node/rest/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkos-node-rest"
-version = "2.2.4"
+version = "2.2.2"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "A REST API server for a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -35,7 +35,7 @@ features = [ "erased-json" ]
 version = "0.2"
 
 [dependencies.indexmap]
-version = "2.1"
+version = "2.0"
 features = [ "serde", "rayon" ]
 
 [dependencies.jsonwebtoken]
@@ -54,11 +54,11 @@ features = [ "derive" ]
 
 [dependencies.snarkos-node-consensus]
 path = "../consensus"
-version = "=2.2.4"
+version = "=2.2.2"
 
 [dependencies.snarkos-node-router]
 path = "../router"
-version = "=2.2.4"
+version = "=2.2.2"
 
 [dependencies.rand]
 version = "0.8"
diff --git a/node/router/Cargo.toml b/node/router/Cargo.toml
index ed5fe5bf7..475bce799 100644
--- a/node/router/Cargo.toml
+++ b/node/router/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkos-node-router"
-version = "2.2.4"
+version = "2.2.2"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "A node router for a decentralized operating system"
 homepage = "https://aleo.org"
@@ -39,7 +39,7 @@ version = "0.3.29"
 features = [ "thread-pool" ]
 
 [dependencies.indexmap]
-version = "2.1"
+version = "2.0"
 features = [ "serde", "rayon" ]
 
 [dependencies.linked-hash-map]
@@ -59,19 +59,19 @@ version = "1"
 
 [dependencies.snarkos-account]
 path = "../../account"
-version = "=2.2.4"
+version = "=2.2.2"
 
 [dependencies.snarkos-node-sync-locators]
 path = "../sync/locators"
-version = "=2.2.4"
+version = "=2.2.2"
 
 [dependencies.snarkos-node-router-messages]
 path = "messages"
-version = "=2.2.4"
+version = "=2.2.2"
 
 [dependencies.snarkos-node-tcp]
 path = "../tcp"
-version = "=2.2.4"
+version = "=2.2.2"
 
 [dependencies.snarkvm]
 workspace = true
diff --git a/node/router/messages/Cargo.toml b/node/router/messages/Cargo.toml
index 88b8f6af9..7488d5011 100644
--- a/node/router/messages/Cargo.toml
+++ b/node/router/messages/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkos-node-router-messages"
-version = "2.2.4"
+version = "2.2.2"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Node messages for a decentralized operating system"
 homepage = "https://aleo.org"
@@ -27,7 +27,7 @@ version = "1.0"
 version = "1"
 
 [dependencies.indexmap]
-version = "2.1"
+version = "2.0"
 features = [ "serde", "rayon" ]
 
 [dependencies.rayon]
@@ -38,11 +38,11 @@ version = "1"
 
 [dependencies.snarkos-node-bft-events]
 path = "../../bft/events"
-version = "=2.2.4"
+version = "=2.2.2"
 
 [dependencies.snarkos-node-sync-locators]
 path = "../../sync/locators"
-version = "=2.2.4"
+version = "=2.2.2"
 
 [dependencies.snarkvm]
 workspace = true
diff --git a/node/src/validator/router.rs b/node/src/validator/router.rs
index a9b415ee3..3a01e6046 100644
--- a/node/src/validator/router.rs
+++ b/node/src/validator/router.rs
@@ -235,19 +235,18 @@ impl<N: Network, C: ConsensusStorage<N>> Inbound<N> for Validator<N, C> {
     /// Propagates the unconfirmed solution to all connected validators.
     async fn unconfirmed_solution(
         &self,
-        _peer_ip: SocketAddr,
-        _serialized: UnconfirmedSolution<N>,
-        _solution: ProverSolution<N>,
+        peer_ip: SocketAddr,
+        serialized: UnconfirmedSolution<N>,
+        solution: ProverSolution<N>,
     ) -> bool {
-        // TODO (howardwu) - Attention, this is being disabled temporarily until transmission storage is updated.
-        // // Add the unconfirmed solution to the memory pool.
-        // if let Err(error) = self.consensus.add_unconfirmed_solution(solution).await {
-        //     trace!("[UnconfirmedSolution] {error}");
-        //     return true; // Maintain the connection.
-        // }
-        // let message = Message::UnconfirmedSolution(serialized);
-        // // Propagate the "UnconfirmedSolution" to the connected validators.
-        // self.propagate_to_validators(message, &[peer_ip]);
+        // Add the unconfirmed solution to the memory pool.
+        if let Err(error) = self.consensus.add_unconfirmed_solution(solution).await {
+            trace!("[UnconfirmedSolution] {error}");
+            return true; // Maintain the connection.
+        }
+        let message = Message::UnconfirmedSolution(serialized);
+        // Propagate the "UnconfirmedSolution" to the connected validators.
+        self.propagate_to_validators(message, &[peer_ip]);
         true
     }
 
diff --git a/node/sync/Cargo.toml b/node/sync/Cargo.toml
index 002f0d6ec..04196617b 100644
--- a/node/sync/Cargo.toml
+++ b/node/sync/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkos-node-sync"
-version = "2.2.4"
+version = "2.2.2"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "A synchronization module for a decentralized operating system"
 homepage = "https://aleo.org"
@@ -24,7 +24,7 @@ test = [ "snarkos-node-sync-locators/test" ]
 version = "1.0"
 
 [dependencies.indexmap]
-version = "2.1"
+version = "2.0"
 features = [ "serde", "rayon" ]
 
 [dependencies.itertools]
@@ -44,16 +44,16 @@ version = "1"
 
 [dependencies.snarkos-node-bft-ledger-service]
 path = "../bft/ledger-service"
-version = "=2.2.4"
+version = "=2.2.2"
 features = [ "ledger-write" ]
 
 [dependencies.snarkos-node-sync-communication-service]
 path = "communication-service"
-version = "=2.2.4"
+version = "=2.2.2"
 
 [dependencies.snarkos-node-sync-locators]
 path = "locators"
-version = "=2.2.4"
+version = "=2.2.2"
 
 [dependencies.snarkvm]
 workspace = true
diff --git a/node/sync/communication-service/Cargo.toml b/node/sync/communication-service/Cargo.toml
index 04330e26b..30a23955f 100644
--- a/node/sync/communication-service/Cargo.toml
+++ b/node/sync/communication-service/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkos-node-sync-communication-service"
-version = "2.2.4"
+version = "2.2.2"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "A communication service for the sync module in a decentralized operating system"
 homepage = "https://aleo.org"
diff --git a/node/sync/locators/Cargo.toml b/node/sync/locators/Cargo.toml
index 729171e81..cf99b884b 100644
--- a/node/sync/locators/Cargo.toml
+++ b/node/sync/locators/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkos-node-sync-locators"
-version = "2.2.4"
+version = "2.2.2"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Locators to synchronize a decentralized operating system"
 homepage = "https://aleo.org"
@@ -24,7 +24,7 @@ test = [ ]
 version = "1.0"
 
 [dependencies.indexmap]
-version = "2.1"
+version = "2.0"
 features = [ "serde", "rayon" ]
 
 [dependencies.serde]
diff --git a/node/tcp/Cargo.toml b/node/tcp/Cargo.toml
index bc44e1a1e..ae8c2dc5b 100644
--- a/node/tcp/Cargo.toml
+++ b/node/tcp/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkos-node-tcp"
-version = "2.2.4"
+version = "2.2.2"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "A TCP stack for a decentralized operating system"
 homepage = "https://aleo.org"
diff --git a/snarkos/main.rs b/snarkos/main.rs
index 64cef2f18..a456e330a 100644
--- a/snarkos/main.rs
+++ b/snarkos/main.rs
@@ -15,8 +15,6 @@
 use snarkos_cli::{commands::CLI, helpers::Updater};
 
 use clap::Parser;
-use std::process::exit;
-
 #[cfg(feature = "jemalloc")]
 use tikv_jemallocator::Jemalloc;
 
@@ -32,10 +30,7 @@ fn main() -> anyhow::Result<()> {
     // Run the CLI.
     match cli.command.parse() {
         Ok(output) => println!("{output}\n"),
-        Err(error) => {
-            println!("⚠️  {error}\n");
-            exit(1);
-        }
+        Err(error) => println!("⚠️  {error}\n"),
     }
     Ok(())
 }
