diff --git a/.circleci/config.yml b/.circleci/config.yml
index b2cc70664..240b01f97 100644
--- a/.circleci/config.yml
+++ b/.circleci/config.yml
@@ -1,5 +1,16 @@
 version: 2.1
 
+parameters:
+  large:
+    type: string
+    default: aleonet/large
+  xlarge:
+    type: string
+    default: aleonet/xlarge
+  twoxlarge:
+    type: string
+    default: aleonet/2xlarge
+
 orbs:
   windows: circleci/windows@5.0
 
@@ -80,7 +91,6 @@ commands:
           paths:
             - .cache/sccache
             - .cargo
-
   run_serial:
     description: "Build and run tests"
     parameters:
@@ -137,7 +147,7 @@ jobs:
   snarkvm:
     docker:
       - image: cimg/rust:1.76.0 # Attention - Change the MSRV in Cargo.toml and rust-toolchain as well
-    resource_class: anf/2xlarge
+    resource_class: << pipeline.parameters.twoxlarge >>
     steps:
       - checkout
       - run:
@@ -147,7 +157,7 @@ jobs:
   algorithms:
     docker:
       - image: cimg/rust:1.76.0 # Attention - Change the MSRV in Cargo.toml and rust-toolchain as well
-    resource_class: anf/2xlarge
+    resource_class: << pipeline.parameters.twoxlarge >>
     steps:
       - run_serial:
           workspace_member: algorithms
@@ -156,7 +166,7 @@ jobs:
   algorithms-profiler:
     docker:
       - image: cimg/rust:1.76.0 # Attention - Change the MSRV in Cargo.toml and rust-toolchain as well
-    resource_class: anf/xlarge
+    resource_class: << pipeline.parameters.xlarge >>
     steps:
       - run_serial: # This runs a single test with profiler enabled
           workspace_member: algorithms
@@ -166,7 +176,7 @@ jobs:
   circuit:
     docker:
       - image: cimg/rust:1.76.0 # Attention - Change the MSRV in Cargo.toml and rust-toolchain as well
-    resource_class: anf/xlarge
+    resource_class: << pipeline.parameters.xlarge >>
     steps:
       - run_serial:
           workspace_member: circuit
@@ -175,7 +185,7 @@ jobs:
   circuit-account:
     docker:
       - image: cimg/rust:1.76.0 # Attention - Change the MSRV in Cargo.toml and rust-toolchain as well
-    resource_class: anf/xlarge
+    resource_class: << pipeline.parameters.xlarge >>
     steps:
       - run_serial:
           workspace_member: circuit/account
@@ -185,7 +195,7 @@ jobs:
   circuit-account-noconsole:
     docker:
       - image: cimg/rust:1.76.0 # Attention - Change the MSRV in Cargo.toml and rust-toolchain as well
-    resource_class: anf/xlarge
+    resource_class: << pipeline.parameters.xlarge >>
     steps:
       - run_serial:
           workspace_member: circuit/account
@@ -195,7 +205,7 @@ jobs:
   circuit-algorithms:
     docker:
       - image: cimg/rust:1.76.0 # Attention - Change the MSRV in Cargo.toml and rust-toolchain as well
-    resource_class: anf/xlarge
+    resource_class: << pipeline.parameters.xlarge >>
     steps:
       - run_serial:
           workspace_member: circuit/algorithms
@@ -204,7 +214,7 @@ jobs:
   circuit-collections:
     docker:
       - image: cimg/rust:1.76.0 # Attention - Change the MSRV in Cargo.toml and rust-toolchain as well
-    resource_class: anf/xlarge
+    resource_class: << pipeline.parameters.xlarge >>
     steps:
       - run_serial:
           workspace_member: circuit/collections
@@ -214,7 +224,7 @@ jobs:
   circuit-collections-noconsole:
     docker:
       - image: cimg/rust:1.76.0 # Attention - Change the MSRV in Cargo.toml and rust-toolchain as well
-    resource_class: anf/xlarge
+    resource_class: << pipeline.parameters.xlarge >>
     steps:
       - run_serial:
           workspace_member: circuit/collections
@@ -224,7 +234,7 @@ jobs:
   circuit-environment:
     docker:
       - image: cimg/rust:1.76.0 # Attention - Change the MSRV in Cargo.toml and rust-toolchain as well
-    resource_class: anf/xlarge
+    resource_class: << pipeline.parameters.xlarge >>
     steps:
       - run_serial:
           workspace_member: circuit/environment
@@ -233,7 +243,7 @@ jobs:
   circuit-network:
     docker:
       - image: cimg/rust:1.76.0 # Attention - Change the MSRV in Cargo.toml and rust-toolchain as well
-    resource_class: anf/xlarge
+    resource_class: << pipeline.parameters.xlarge >>
     steps:
       - run_serial:
           workspace_member: circuit/network
@@ -242,7 +252,7 @@ jobs:
   circuit-program:
     docker:
       - image: cimg/rust:1.76.0 # Attention - Change the MSRV in Cargo.toml and rust-toolchain as well
-    resource_class: anf/xlarge
+    resource_class: << pipeline.parameters.xlarge >>
     steps:
       - run_serial:
           workspace_member: circuit/program
@@ -251,7 +261,7 @@ jobs:
   circuit-types:
     docker:
       - image: cimg/rust:1.76.0 # Attention - Change the MSRV in Cargo.toml and rust-toolchain as well
-    resource_class: anf/xlarge
+    resource_class: << pipeline.parameters.xlarge >>
     steps:
       - run_serial:
           workspace_member: circuit/types
@@ -260,7 +270,7 @@ jobs:
   circuit-types-address:
     docker:
       - image: cimg/rust:1.76.0 # Attention - Change the MSRV in Cargo.toml and rust-toolchain as well
-    resource_class: anf/xlarge
+    resource_class: << pipeline.parameters.xlarge >>
     steps:
       - run_serial:
           workspace_member: circuit/types/address
@@ -269,7 +279,7 @@ jobs:
   circuit-types-boolean:
     docker:
       - image: cimg/rust:1.76.0 # Attention - Change the MSRV in Cargo.toml and rust-toolchain as well
-    resource_class: anf/xlarge
+    resource_class: << pipeline.parameters.xlarge >>
     steps:
       - run_serial:
           workspace_member: circuit/types/boolean
@@ -278,7 +288,7 @@ jobs:
   circuit-types-field:
     docker:
       - image: cimg/rust:1.76.0 # Attention - Change the MSRV in Cargo.toml and rust-toolchain as well
-    resource_class: anf/xlarge
+    resource_class: << pipeline.parameters.xlarge >>
     steps:
       - run_serial:
           workspace_member: circuit/types/field
@@ -287,7 +297,7 @@ jobs:
   circuit-types-group:
     docker:
       - image: cimg/rust:1.76.0 # Attention - Change the MSRV in Cargo.toml and rust-toolchain as well
-    resource_class: anf/xlarge
+    resource_class: << pipeline.parameters.xlarge >>
     steps:
       - run_serial:
           workspace_member: circuit/types/group
@@ -296,7 +306,7 @@ jobs:
   circuit-types-integers:
     docker:
       - image: cimg/rust:1.76.0 # Attention - Change the MSRV in Cargo.toml and rust-toolchain as well
-    resource_class: anf/2xlarge
+    resource_class: << pipeline.parameters.twoxlarge >>
     steps:
       - run_serial:
           workspace_member: circuit/types/integers
@@ -306,7 +316,7 @@ jobs:
   circuit-types-scalar:
     docker:
       - image: cimg/rust:1.76.0 # Attention - Change the MSRV in Cargo.toml and rust-toolchain as well
-    resource_class: anf/xlarge
+    resource_class: << pipeline.parameters.xlarge >>
     steps:
       - run_serial:
           workspace_member: circuit/types/scalar
@@ -315,7 +325,7 @@ jobs:
   circuit-types-string:
     docker:
       - image: cimg/rust:1.76.0 # Attention - Change the MSRV in Cargo.toml and rust-toolchain as well
-    resource_class: anf/xlarge
+    resource_class: << pipeline.parameters.xlarge >>
     steps:
       - run_serial:
           workspace_member: circuit/types/string
@@ -323,7 +333,7 @@ jobs:
   console:
     docker:
       - image: cimg/rust:1.76.0 # Attention - Change the MSRV in Cargo.toml and rust-toolchain as well
-    resource_class: anf/xlarge
+    resource_class: << pipeline.parameters.xlarge >>
     steps:
       - run_serial:
           workspace_member: console
@@ -332,7 +342,7 @@ jobs:
   console-account:
     docker:
       - image: cimg/rust:1.76.0 # Attention - Change the MSRV in Cargo.toml and rust-toolchain as well
-    resource_class: anf/xlarge
+    resource_class: << pipeline.parameters.xlarge >>
     steps:
       - run_serial:
           workspace_member: console/account
@@ -341,7 +351,7 @@ jobs:
   console-algorithms:
     docker:
       - image: cimg/rust:1.76.0 # Attention - Change the MSRV in Cargo.toml and rust-toolchain as well
-    resource_class: anf/xlarge
+    resource_class: << pipeline.parameters.xlarge >>
     steps:
       - run_serial:
           workspace_member: console/algorithms
@@ -350,7 +360,7 @@ jobs:
   console-collections:
     docker:
       - image: cimg/rust:1.76.0 # Attention - Change the MSRV in Cargo.toml and rust-toolchain as well
-    resource_class: anf/xlarge
+    resource_class: << pipeline.parameters.xlarge >>
     steps:
       - run_serial:
           workspace_member: console/collections
@@ -359,7 +369,7 @@ jobs:
   console-network:
     docker:
       - image: cimg/rust:1.76.0 # Attention - Change the MSRV in Cargo.toml and rust-toolchain as well
-    resource_class: anf/xlarge
+    resource_class: << pipeline.parameters.xlarge >>
     steps:
       - run_serial:
           workspace_member: console/network
@@ -368,7 +378,7 @@ jobs:
   console-network-environment:
     docker:
       - image: cimg/rust:1.76.0 # Attention - Change the MSRV in Cargo.toml and rust-toolchain as well
-    resource_class: anf/xlarge
+    resource_class: << pipeline.parameters.xlarge >>
     steps:
       - run_serial:
           workspace_member: console/network/environment
@@ -377,7 +387,7 @@ jobs:
   console-program:
     docker:
       - image: cimg/rust:1.76.0 # Attention - Change the MSRV in Cargo.toml and rust-toolchain as well
-    resource_class: anf/xlarge
+    resource_class: << pipeline.parameters.xlarge >>
     steps:
       - run_serial:
           workspace_member: console/program
@@ -386,7 +396,7 @@ jobs:
   console-types:
     docker:
       - image: cimg/rust:1.76.0 # Attention - Change the MSRV in Cargo.toml and rust-toolchain as well
-    resource_class: anf/xlarge
+    resource_class: << pipeline.parameters.xlarge >>
     steps:
       - run_serial:
           workspace_member: console/types
@@ -395,7 +405,7 @@ jobs:
   console-types-address:
     docker:
       - image: cimg/rust:1.76.0 # Attention - Change the MSRV in Cargo.toml and rust-toolchain as well
-    resource_class: anf/xlarge
+    resource_class: << pipeline.parameters.xlarge >>
     steps:
       - run_serial:
           workspace_member: console/types/address
@@ -404,7 +414,7 @@ jobs:
   console-types-boolean:
     docker:
       - image: cimg/rust:1.76.0 # Attention - Change the MSRV in Cargo.toml and rust-toolchain as well
-    resource_class: anf/xlarge
+    resource_class: << pipeline.parameters.xlarge >>
     steps:
       - run_serial:
           workspace_member: console/types/boolean
@@ -413,7 +423,7 @@ jobs:
   console-types-field:
     docker:
       - image: cimg/rust:1.76.0 # Attention - Change the MSRV in Cargo.toml and rust-toolchain as well
-    resource_class: anf/xlarge
+    resource_class: << pipeline.parameters.xlarge >>
     steps:
       - run_serial:
           workspace_member: console/types/field
@@ -422,7 +432,7 @@ jobs:
   console-types-group:
     docker:
       - image: cimg/rust:1.76.0 # Attention - Change the MSRV in Cargo.toml and rust-toolchain as well
-    resource_class: anf/xlarge
+    resource_class: << pipeline.parameters.xlarge >>
     steps:
       - run_serial:
           workspace_member: console/types/group
@@ -431,7 +441,7 @@ jobs:
   console-types-integers:
     docker:
       - image: cimg/rust:1.76.0 # Attention - Change the MSRV in Cargo.toml and rust-toolchain as well
-    resource_class: anf/xlarge
+    resource_class: << pipeline.parameters.xlarge >>
     steps:
       - run_serial:
           workspace_member: console/types/integers
@@ -440,7 +450,7 @@ jobs:
   console-types-scalar:
     docker:
       - image: cimg/rust:1.76.0 # Attention - Change the MSRV in Cargo.toml and rust-toolchain as well
-    resource_class: anf/xlarge
+    resource_class: << pipeline.parameters.xlarge >>
     steps:
       - run_serial:
           workspace_member: console/types/scalar
@@ -449,7 +459,7 @@ jobs:
   console-types-string:
     docker:
       - image: cimg/rust:1.76.0 # Attention - Change the MSRV in Cargo.toml and rust-toolchain as well
-    resource_class: anf/xlarge
+    resource_class: << pipeline.parameters.xlarge >>
     steps:
       - run_serial:
           workspace_member: console/types/string
@@ -458,7 +468,7 @@ jobs:
   curves:
     docker:
       - image: cimg/rust:1.76.0 # Attention - Change the MSRV in Cargo.toml and rust-toolchain as well
-    resource_class: anf/xlarge
+    resource_class: << pipeline.parameters.xlarge >>
     steps:
       - run_serial:
           workspace_member: curves
@@ -467,7 +477,7 @@ jobs:
   fields:
     docker:
       - image: cimg/rust:1.76.0 # Attention - Change the MSRV in Cargo.toml and rust-toolchain as well
-    resource_class: anf/xlarge
+    resource_class: << pipeline.parameters.xlarge >>
     steps:
       - run_serial:
           workspace_member: fields
@@ -476,7 +486,7 @@ jobs:
   ledger:
     docker:
       - image: cimg/rust:1.76.0 # Attention - Change the MSRV in Cargo.toml and rust-toolchain as well
-    resource_class: anf/2xlarge
+    resource_class: << pipeline.parameters.twoxlarge >>
     steps:
       - run_serial:
           workspace_member: ledger
@@ -485,7 +495,7 @@ jobs:
   ledger-with-rocksdb:
     docker:
       - image: cimg/rust:1.76.0 # Attention - Change the MSRV in Cargo.toml and rust-toolchain as well
-    resource_class: anf/2xlarge
+    resource_class: << pipeline.parameters.twoxlarge >>
     steps:
       - run_serial:
           flags: --features=rocks
@@ -495,7 +505,7 @@ jobs:
   ledger-with-valid-solutions:
     docker:
       - image: cimg/rust:1.76.0 # Attention - Change the MSRV in Cargo.toml and rust-toolchain as well
-    resource_class: anf/2xlarge
+    resource_class: << pipeline.parameters.twoxlarge >>
     steps:
       - run_serial:
           flags: valid_solutions --features=test
@@ -505,7 +515,7 @@ jobs:
   ledger-authority:
     docker:
       - image: cimg/rust:1.76.0 # Attention - Change the MSRV in Cargo.toml and rust-toolchain as well
-    resource_class: anf/2xlarge
+    resource_class: << pipeline.parameters.twoxlarge >>
     steps:
       - run_serial:
           workspace_member: ledger/authority
@@ -514,7 +524,7 @@ jobs:
   ledger-block:
     docker:
       - image: cimg/rust:1.76.0 # Attention - Change the MSRV in Cargo.toml and rust-toolchain as well
-    resource_class: anf/2xlarge
+    resource_class: << pipeline.parameters.twoxlarge >>
     steps:
       - run_serial:
           workspace_member: ledger/block
@@ -523,7 +533,7 @@ jobs:
   ledger-committee:
     docker:
       - image: cimg/rust:1.76.0 # Attention - Change the MSRV in Cargo.toml and rust-toolchain as well
-    resource_class: anf/xlarge
+    resource_class: << pipeline.parameters.xlarge >>
     steps:
       - run_serial:
           workspace_member: ledger/committee
@@ -532,7 +542,7 @@ jobs:
   ledger-narwhal:
     docker:
       - image: cimg/rust:1.76.0 # Attention - Change the MSRV in Cargo.toml and rust-toolchain as well
-    resource_class: anf/2xlarge
+    resource_class: << pipeline.parameters.twoxlarge >>
     steps:
       - run_serial:
           workspace_member: ledger/narwhal
@@ -541,7 +551,7 @@ jobs:
   ledger-narwhal-batch-certificate:
     docker:
       - image: cimg/rust:1.76.0 # Attention - Change the MSRV in Cargo.toml and rust-toolchain as well
-    resource_class: anf/2xlarge
+    resource_class: << pipeline.parameters.twoxlarge >>
     steps:
       - run_serial:
           workspace_member: ledger/narwhal/batch-certificate
@@ -550,7 +560,7 @@ jobs:
   ledger-narwhal-batch-header:
     docker:
       - image: cimg/rust:1.76.0 # Attention - Change the MSRV in Cargo.toml and rust-toolchain as well
-    resource_class: anf/2xlarge
+    resource_class: << pipeline.parameters.twoxlarge >>
     steps:
       - run_serial:
           workspace_member: ledger/narwhal/batch-header
@@ -559,7 +569,7 @@ jobs:
   ledger-narwhal-data:
     docker:
       - image: cimg/rust:1.76.0 # Attention - Change the MSRV in Cargo.toml and rust-toolchain as well
-    resource_class: anf/2xlarge
+    resource_class: << pipeline.parameters.twoxlarge >>
     steps:
       - run_serial:
           workspace_member: ledger/narwhal/data
@@ -568,7 +578,7 @@ jobs:
   ledger-narwhal-subdag:
     docker:
       - image: cimg/rust:1.76.0 # Attention - Change the MSRV in Cargo.toml and rust-toolchain as well
-    resource_class: anf/2xlarge
+    resource_class: << pipeline.parameters.twoxlarge >>
     steps:
       - run_serial:
           workspace_member: ledger/narwhal/subdag
@@ -577,7 +587,7 @@ jobs:
   ledger-narwhal-transmission:
     docker:
       - image: cimg/rust:1.76.0 # Attention - Change the MSRV in Cargo.toml and rust-toolchain as well
-    resource_class: anf/2xlarge
+    resource_class: << pipeline.parameters.twoxlarge >>
     steps:
       - run_serial:
           workspace_member: ledger/narwhal/transmission
@@ -586,7 +596,7 @@ jobs:
   ledger-narwhal-transmission-id:
     docker:
       - image: cimg/rust:1.76.0 # Attention - Change the MSRV in Cargo.toml and rust-toolchain as well
-    resource_class: anf/2xlarge
+    resource_class: << pipeline.parameters.twoxlarge >>
     steps:
       - run_serial:
           workspace_member: ledger/narwhal/transmission-id
@@ -595,7 +605,7 @@ jobs:
   ledger-puzzle:
     docker:
       - image: cimg/rust:1.76.0 # Attention - Change the MSRV in Cargo.toml and rust-toolchain as well
-    resource_class: anf/2xlarge
+    resource_class: << pipeline.parameters.twoxlarge >>
     steps:
       - run_serial:
           workspace_member: ledger/puzzle
@@ -604,7 +614,7 @@ jobs:
   ledger-puzzle-epoch:
     docker:
       - image: cimg/rust:1.76.0 # Attention - Change the MSRV in Cargo.toml and rust-toolchain as well
-    resource_class: anf/2xlarge
+    resource_class: << pipeline.parameters.twoxlarge >>
     steps:
       - run_serial:
           workspace_member: ledger/puzzle/epoch
@@ -613,7 +623,7 @@ jobs:
   ledger-query:
     docker:
       - image: cimg/rust:1.76.0 # Attention - Change the MSRV in Cargo.toml and rust-toolchain as well
-    resource_class: anf/2xlarge
+    resource_class: << pipeline.parameters.twoxlarge >>
     steps:
       - run_serial:
           workspace_member: ledger/query
@@ -622,7 +632,7 @@ jobs:
   ledger-store:
     docker:
       - image: cimg/rust:1.76.0 # Attention - Change the MSRV in Cargo.toml and rust-toolchain as well
-    resource_class: anf/2xlarge
+    resource_class: << pipeline.parameters.twoxlarge >>
     steps:
       - run_serial:
           flags: --features=rocks
@@ -632,7 +642,7 @@ jobs:
   ledger-test-helpers:
     docker:
       - image: cimg/rust:1.76.0 # Attention - Change the MSRV in Cargo.toml and rust-toolchain as well
-    resource_class: anf/2xlarge
+    resource_class: << pipeline.parameters.twoxlarge >>
     steps:
       - run_serial:
           workspace_member: ledger/test-helpers
@@ -641,7 +651,7 @@ jobs:
   parameters:
     docker:
       - image: cimg/rust:1.76.0 # Attention - Change the MSRV in Cargo.toml and rust-toolchain as well
-    resource_class: anf/xlarge
+    resource_class: << pipeline.parameters.xlarge >>
     steps:
       - run_serial:
           workspace_member: parameters
@@ -650,7 +660,7 @@ jobs:
   synthesizer:
     docker:
       - image: cimg/rust:1.76.0 # Attention - Change the MSRV in Cargo.toml and rust-toolchain as well
-    resource_class: anf/2xlarge
+    resource_class: << pipeline.parameters.twoxlarge >>
     steps:
       - run_serial:
           flags: --lib --bins
@@ -660,7 +670,7 @@ jobs:
   synthesizer-integration:
     docker:
       - image: cimg/rust:1.76.0 # Attention - Change the MSRV in Cargo.toml and rust-toolchain as well
-    resource_class: anf/2xlarge
+    resource_class: << pipeline.parameters.twoxlarge >>
     steps:
       - run_serial:
           flags: --test '*'
@@ -670,7 +680,7 @@ jobs:
   synthesizer-process:
     docker:
       - image: cimg/rust:1.76.0 # Attention - Change the MSRV in Cargo.toml and rust-toolchain as well
-    resource_class: anf/2xlarge
+    resource_class: << pipeline.parameters.twoxlarge >>
     steps:
       - run_serial:
           workspace_member: synthesizer/process
@@ -679,7 +689,7 @@ jobs:
   synthesizer-process-with-rocksdb:
     docker:
       - image: cimg/rust:1.76.0 # Attention - Change the MSRV in Cargo.toml and rust-toolchain as well
-    resource_class: anf/2xlarge
+    resource_class: << pipeline.parameters.twoxlarge >>
     steps:
       - run_serial:
           flags: --features=rocks
@@ -689,7 +699,7 @@ jobs:
   synthesizer-program:
     docker:
       - image: cimg/rust:1.76.0 # Attention - Change the MSRV in Cargo.toml and rust-toolchain as well
-    resource_class: anf/2xlarge
+    resource_class: << pipeline.parameters.twoxlarge >>
     steps:
       - run_serial:
           flags: --lib --bins
@@ -699,17 +709,17 @@ jobs:
   synthesizer-program-integration:
     docker:
       - image: cimg/rust:1.76.0 # Attention - Change the MSRV in Cargo.toml and rust-toolchain as well
-    resource_class: anf/2xlarge
+    resource_class: << pipeline.parameters.twoxlarge >>
     steps:
       - run_serial:
-          flags: --test '*' -- --skip keccak --skip psd --skip sha
+          flags: --test '*' -- --skip keccak --skip psd --skip sha --skip instruction::is --skip instruction::equal --skip instruction::commit
           workspace_member: synthesizer/program
           cache_key: snarkvm-synthesizer-program-cache
 
   synthesizer-program-integration-keccak:
     docker:
       - image: cimg/rust:1.76.0 # Attention - Change the MSRV in Cargo.toml and rust-toolchain as well
-    resource_class: anf/2xlarge
+    resource_class: << pipeline.parameters.twoxlarge >>
     steps:
       - run_serial:
           flags: keccak --test '*'
@@ -719,7 +729,7 @@ jobs:
   synthesizer-program-integration-psd:
     docker:
       - image: cimg/rust:1.76.0 # Attention - Change the MSRV in Cargo.toml and rust-toolchain as well
-    resource_class: anf/2xlarge
+    resource_class: << pipeline.parameters.twoxlarge >>
     steps:
       - run_serial:
           flags: psd --test '*'
@@ -729,17 +739,47 @@ jobs:
   synthesizer-program-integration-sha:
     docker:
       - image: cimg/rust:1.76.0 # Attention - Change the MSRV in Cargo.toml and rust-toolchain as well
-    resource_class: anf/2xlarge
+    resource_class: << pipeline.parameters.twoxlarge >>
     steps:
       - run_serial:
           flags: sha --test '*'
           workspace_member: synthesizer/program
           cache_key: snarkvm-synthesizer-program-cache
 
+  synthesizer-program-integration-instruction-is:
+    docker:
+      - image: cimg/rust:1.76.0 # Attention - Change the MSRV in Cargo.toml and rust-toolchain as well
+    resource_class: << pipeline.parameters.twoxlarge >>
+    steps:
+      - run_serial:
+          flags: instruction::is --test '*'
+          workspace_member: synthesizer/program
+          cache_key: snarkvm-synthesizer-program-cache
+
+  synthesizer-program-integration-instruction-equal:
+    docker:
+      - image: cimg/rust:1.76.0 # Attention - Change the MSRV in Cargo.toml and rust-toolchain as well
+    resource_class: << pipeline.parameters.twoxlarge >>
+    steps:
+      - run_serial:
+          flags: instruction::equal --test '*'
+          workspace_member: synthesizer/program
+          cache_key: snarkvm-synthesizer-program-cache
+
+  synthesizer-program-integration-instruction-commit:
+    docker:
+      - image: cimg/rust:1.76.0 # Attention - Change the MSRV in Cargo.toml and rust-toolchain as well
+    resource_class: << pipeline.parameters.twoxlarge >>
+    steps:
+      - run_serial:
+          flags: instruction::commit --test '*'
+          workspace_member: synthesizer/program
+          cache_key: snarkvm-synthesizer-program-cache
+
   synthesizer-snark:
     docker:
       - image: cimg/rust:1.76.0 # Attention - Change the MSRV in Cargo.toml and rust-toolchain as well
-    resource_class: anf/2xlarge
+    resource_class: << pipeline.parameters.twoxlarge >>
     steps:
       - run_serial:
           workspace_member: synthesizer/snark
@@ -748,7 +788,7 @@ jobs:
   utilities:
     docker:
       - image: cimg/rust:1.76.0 # Attention - Change the MSRV in Cargo.toml and rust-toolchain as well
-    resource_class: anf/xlarge
+    resource_class: << pipeline.parameters.xlarge >>
     steps:
       - run_serial:
           workspace_member: utilities
@@ -757,7 +797,7 @@ jobs:
   utilities-derives:
     docker:
       - image: cimg/rust:1.76.0 # Attention - Change the MSRV in Cargo.toml and rust-toolchain as well
-    resource_class: anf/xlarge
+    resource_class: << pipeline.parameters.xlarge >>
     steps:
       - run_serial:
           workspace_member: utilities/derives
@@ -766,7 +806,7 @@ jobs:
   wasm:
     docker:
       - image: cimg/rust:1.76.0 # Attention - Change the MSRV in Cargo.toml and rust-toolchain as well
-    resource_class: anf/2xlarge
+    resource_class: << pipeline.parameters.twoxlarge >>
     steps:
       - checkout
       - setup_environment:
@@ -784,7 +824,7 @@ jobs:
   check-fmt:
     docker:
       - image: cimg/rust:1.76.0 # Attention - Change the MSRV in Cargo.toml and rust-toolchain as well
-    resource_class: anf/xlarge
+    resource_class: << pipeline.parameters.xlarge >>
     steps:
       - checkout
       - install_rust_nightly
@@ -800,7 +840,7 @@ jobs:
   check-clippy:
     docker:
       - image: cimg/rust:1.76.0 # Attention - Change the MSRV in Cargo.toml and rust-toolchain as well
-    resource_class: anf/2xlarge
+    resource_class: << pipeline.parameters.twoxlarge >>
     steps:
       - checkout
       - setup_environment:
@@ -817,7 +857,7 @@ jobs:
   check-all-targets:
     docker:
       - image: cimg/rust:1.76.0 # Attention - Change the MSRV in Cargo.toml and rust-toolchain as well
-    resource_class: anf/xlarge
+    resource_class: << pipeline.parameters.xlarge >>
     steps:
       - checkout
       - setup_environment:
@@ -913,6 +953,9 @@ workflows:
       - synthesizer-program-integration-keccak
       - synthesizer-program-integration-psd
       - synthesizer-program-integration-sha
+      - synthesizer-program-integration-instruction-is
+      - synthesizer-program-integration-instruction-equal
+      - synthesizer-program-integration-instruction-commit
       - synthesizer-snark
       - utilities
       - utilities-derives
diff --git a/Cargo.lock b/Cargo.lock
index 3697583df..103dd9f88 100644
--- a/Cargo.lock
+++ b/Cargo.lock
@@ -857,6 +857,26 @@ dependencies = [
  "cfg-if",
 ]
 
+[[package]]
+name = "enum-iterator"
+version = "2.1.0"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "c280b9e6b3ae19e152d8e31cf47f18389781e119d4013a2a2bb0180e5facc635"
+dependencies = [
+ "enum-iterator-derive",
+]
+
+[[package]]
+name = "enum-iterator-derive"
+version = "1.4.0"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "a1ab991c1362ac86c61ab6f556cff143daa22e5a15e4e189df818b2fd19fe65b"
+dependencies = [
+ "proc-macro2",
+ "quote 1.0.35",
+ "syn 2.0.48",
+]
+
 [[package]]
 name = "enum_index"
 version = "0.2.0"
@@ -2912,6 +2932,7 @@ name = "snarkvm-console-program"
 version = "0.16.19"
 dependencies = [
  "bincode",
+ "enum-iterator",
  "enum_index",
  "enum_index_derive",
  "indexmap 2.1.0",
@@ -3395,6 +3416,7 @@ dependencies = [
  "snarkvm-synthesizer-program",
  "snarkvm-synthesizer-snark",
  "snarkvm-utilities",
+ "tempfile",
  "tracing",
  "walkdir",
 ]
diff --git a/console/network/src/mainnet_v0.rs b/console/network/src/mainnet_v0.rs
index 1724e084c..baf63b7e2 100644
--- a/console/network/src/mainnet_v0.rs
+++ b/console/network/src/mainnet_v0.rs
@@ -140,7 +140,7 @@ impl Network for MainnetV0 {
     /// The function name for the inclusion circuit.
     const INCLUSION_FUNCTION_NAME: &'static str = snarkvm_parameters::mainnet::NETWORK_INCLUSION_FUNCTION_NAME;
     /// The maximum number of certificates in a batch.
-    const MAX_CERTIFICATES: u16 = 10;
+    const MAX_CERTIFICATES: u16 = 15;
     /// The network name.
     const NAME: &'static str = "Aleo Mainnet (v0)";
 
diff --git a/console/program/Cargo.toml b/console/program/Cargo.toml
index 23df63249..8f20c2f8e 100644
--- a/console/program/Cargo.toml
+++ b/console/program/Cargo.toml
@@ -40,6 +40,9 @@ version = "0.2"
 [dependencies.enum_index_derive]
 version = "0.2"
 
+[dependencies.enum-iterator]
+version = "2.1"
+
 [dependencies.indexmap]
 version = "2.0"
 
diff --git a/console/program/src/data/identifier/parse.rs b/console/program/src/data/identifier/parse.rs
index daf02185f..1ab087141 100644
--- a/console/program/src/data/identifier/parse.rs
+++ b/console/program/src/data/identifier/parse.rs
@@ -54,7 +54,7 @@ impl<N: Network> FromStr for Identifier<N> {
 
         // Ensure that the identifier is not a literal.
         ensure!(
-            crate::LiteralType::from_str(identifier).is_err(),
+            !enum_iterator::all::<crate::LiteralType>().any(|lt| lt.type_name() == identifier),
             "Identifier '{identifier}' is a reserved literal type"
         );
 
diff --git a/console/program/src/data_types/literal_type/mod.rs b/console/program/src/data_types/literal_type/mod.rs
index d6ddf2f5a..0f0dea6e7 100644
--- a/console/program/src/data_types/literal_type/mod.rs
+++ b/console/program/src/data_types/literal_type/mod.rs
@@ -23,10 +23,11 @@ use snarkvm_console_network::prelude::*;
 use snarkvm_console_types::{prelude::*, Boolean};
 
 use core::fmt::{self, Debug, Display};
+use enum_iterator::Sequence;
 use num_derive::FromPrimitive;
 use num_traits::FromPrimitive;
 
-#[derive(Copy, Clone, PartialEq, Eq, Hash, FromPrimitive)]
+#[derive(Copy, Clone, PartialEq, Eq, Hash, FromPrimitive, Sequence)]
 pub enum LiteralType {
     /// The Aleo address type.
     Address,
diff --git a/ledger/Cargo.toml b/ledger/Cargo.toml
index e75f892a8..709a32abe 100644
--- a/ledger/Cargo.toml
+++ b/ledger/Cargo.toml
@@ -149,6 +149,10 @@ package = "snarkvm-ledger-block"
 path = "./block"
 features = [ "test" ]
 
+[dev-dependencies.ledger-test-helpers]
+package = "snarkvm-ledger-test-helpers"
+path = "./test-helpers"
+
 [dev-dependencies.serde_json]
 version = "1.0"
 features = [ "preserve_order" ]
diff --git a/ledger/block/src/solutions/serialize.rs b/ledger/block/src/solutions/serialize.rs
index 8418b3e49..09aa2667b 100644
--- a/ledger/block/src/solutions/serialize.rs
+++ b/ledger/block/src/solutions/serialize.rs
@@ -63,7 +63,7 @@ impl<'de, N: Network> Deserialize<'de> for Solutions<N> {
 pub(super) mod tests {
     use super::*;
     use console::account::{Address, PrivateKey};
-    use ledger_puzzle::Solution;
+    use ledger_puzzle::{PartialSolution, Solution};
 
     type CurrentNetwork = console::network::MainnetV0;
 
@@ -73,7 +73,8 @@ pub(super) mod tests {
         for _ in 0..rng.gen_range(1..=CurrentNetwork::MAX_SOLUTIONS) {
             let private_key = PrivateKey::<CurrentNetwork>::new(rng).unwrap();
             let address = Address::try_from(private_key).unwrap();
-            solutions.push(Solution::new(rng.gen(), address, u64::rand(rng)).unwrap());
+            let partial_solution = PartialSolution::new(rng.gen(), address, u64::rand(rng)).unwrap();
+            solutions.push(Solution::new(partial_solution, u64::rand(rng)));
         }
         Solutions::new(PuzzleSolutions::new(solutions).unwrap()).unwrap()
     }
diff --git a/ledger/puzzle/src/lib.rs b/ledger/puzzle/src/lib.rs
index c4fd8d7a4..fe6433ced 100644
--- a/ledger/puzzle/src/lib.rs
+++ b/ledger/puzzle/src/lib.rs
@@ -16,6 +16,9 @@
 #![allow(clippy::too_many_arguments)]
 #![warn(clippy::cast_possible_truncation)]
 
+mod partial_solution;
+pub use partial_solution::*;
+
 mod solution;
 pub use solution::*;
 
@@ -97,7 +100,7 @@ impl<N: Network> Puzzle<N> {
     }
 
     /// Returns the Merkle leaves for the puzzle, given the solution.
-    pub fn get_leaves(&self, solution: &Solution<N>) -> Result<Vec<Vec<bool>>> {
+    pub fn get_leaves(&self, solution: &PartialSolution<N>) -> Result<Vec<Vec<bool>>> {
         // Initialize a seeded random number generator.
         let mut rng = ChaChaRng::seed_from_u64(*solution.id());
         // Output the leaves.
@@ -119,18 +122,36 @@ impl<N: Network> Puzzle<N> {
 
     /// Returns the proof target given the solution.
     pub fn get_proof_target(&self, solution: &Solution<N>) -> Result<u64> {
+        // Calculate the proof target.
+        let proof_target = self.get_proof_target_unchecked(solution)?;
+        // Ensure the proof target matches the expected proof target.
+        ensure!(solution.target() == proof_target, "The proof target does not match the expected proof target");
+        // Return the proof target.
+        Ok(proof_target)
+    }
+
+    /// Returns the proof target given the solution.
+    ///
+    /// Note: This method does **not** check the proof target against the expected proof target.
+    pub fn get_proof_target_unchecked(&self, solution: &Solution<N>) -> Result<u64> {
+        // Calculate the proof target.
+        self.get_proof_target_from_partial_solution(solution.partial_solution())
+    }
+
+    /// Returns the proof target given the partial solution.
+    pub fn get_proof_target_from_partial_solution(&self, partial_solution: &PartialSolution<N>) -> Result<u64> {
         // If the proof target is in the cache, then return it.
-        if let Some(proof_target) = self.proof_target_cache.write().get(&solution.id()) {
+        if let Some(proof_target) = self.proof_target_cache.write().get(&partial_solution.id()) {
             return Ok(*proof_target);
         }
 
         // Construct the leaves of the Merkle tree.
-        let leaves = self.get_leaves(solution)?;
+        let leaves = self.get_leaves(partial_solution)?;
         // Get the proof target.
         let proof_target = Self::leaves_to_proof_target(&leaves)?;
 
         // Insert the proof target into the cache.
-        self.proof_target_cache.write().put(solution.id(), proof_target);
+        self.proof_target_cache.write().put(partial_solution.id(), proof_target);
         // Return the proof target.
         Ok(proof_target)
     }
@@ -147,7 +168,14 @@ impl<N: Network> Puzzle<N> {
             // Check if the proof target is in the cache.
             match self.proof_target_cache.write().get(id) {
                 // If the proof target is in the cache, then store it.
-                Some(proof_target) => targets[i] = *proof_target,
+                Some(proof_target) => {
+                    // Ensure that the proof target matches the expected proof target.
+                    ensure!(
+                        solution.target() == *proof_target,
+                        "The proof target does not match the expected proof target"
+                    );
+                    targets[i] = *proof_target
+                }
                 // Otherwise, add it to the list of solutions that need to be computed.
                 None => to_compute.push((i, id, *solution)),
             }
@@ -160,10 +188,15 @@ impl<N: Network> Puzzle<N> {
             let leaves = self.get_all_leaves(&solutions_subset)?;
             // Construct the Merkle roots and truncate them to a u64.
             let targets_subset = cfg_iter!(leaves)
-                .zip(cfg_keys!(solutions_subset))
-                .map(|(leaves, solution_id)| {
+                .zip(cfg_iter!(solutions_subset))
+                .map(|(leaves, (solution_id, solution))| {
                     // Get the proof target.
                     let proof_target = Self::leaves_to_proof_target(leaves)?;
+                    // Ensure that the proof target matches the expected proof target.
+                    ensure!(
+                        solution.target() == proof_target,
+                        "The proof target does not match the expected proof target"
+                    );
                     // Insert the proof target into the cache.
                     self.proof_target_cache.write().put(*solution_id, proof_target);
                     // Return the proof target.
@@ -196,18 +229,19 @@ impl<N: Network> Puzzle<N> {
         counter: u64,
         minimum_proof_target: Option<u64>,
     ) -> Result<Solution<N>> {
-        // Construct the solution.
-        let solution = Solution::new(epoch_hash, address, counter)?;
+        // Construct the partial solution.
+        let partial_solution = PartialSolution::new(epoch_hash, address, counter)?;
         // Compute the proof target.
-        let proof_target = self.get_proof_target(&solution)?;
+        let proof_target = self.get_proof_target_from_partial_solution(&partial_solution)?;
         // Check that the minimum proof target is met.
         if let Some(minimum_proof_target) = minimum_proof_target {
             if proof_target < minimum_proof_target {
                 bail!("Solution was below the minimum proof target ({proof_target} < {minimum_proof_target})")
             }
         }
-        // Return the solution.
-        Ok(solution)
+
+        // Construct the solution.
+        Ok(Solution::new(partial_solution, proof_target))
     }
 
     /// Returns `Ok(())` if the solution is valid.
@@ -445,6 +479,71 @@ mod tests {
         assert!(puzzle.check_solutions(&solutions, epoch_hash, 0u64).is_ok());
     }
 
+    #[test]
+    fn test_check_solution_with_incorrect_target_fails() {
+        let mut rng = rand::thread_rng();
+
+        // Initialize a new puzzle.
+        let puzzle = sample_puzzle();
+
+        // Initialize an epoch hash.
+        let epoch_hash = rng.gen();
+
+        // Generate inputs.
+        let private_key = PrivateKey::<CurrentNetwork>::new(&mut rng).unwrap();
+        let address = Address::try_from(private_key).unwrap();
+
+        // Generate a solution.
+        let solution = puzzle.prove(epoch_hash, address, rng.gen(), None).unwrap();
+
+        // Generate a solution with an incorrect target.
+        let incorrect_solution = Solution::new(*solution.partial_solution(), solution.target().saturating_add(1));
+
+        // Ensure the incorrect solution is invalid.
+        assert!(puzzle.check_solution(&incorrect_solution, epoch_hash, 0u64).is_err());
+
+        // Ensure the invalid solution is invalid on a fresh puzzle instance.
+        let new_puzzle = sample_puzzle();
+        assert!(new_puzzle.check_solution(&incorrect_solution, epoch_hash, 0u64).is_err());
+
+        // Ensure the incorrect solutions are invalid.
+        let incorrect_solutions = PuzzleSolutions::new(vec![incorrect_solution]).unwrap();
+        assert!(puzzle.check_solutions(&incorrect_solutions, epoch_hash, 0u64).is_err());
+
+        // Ensure the incorrect solutions are invalid on a fresh puzzle instance.
+        let new_puzzle = sample_puzzle();
+        assert!(new_puzzle.check_solutions(&incorrect_solutions, epoch_hash, 0u64).is_err());
+    }
+
+    #[test]
+    fn test_check_solutions_with_incorrect_target_fails() {
+        let mut rng = TestRng::default();
+
+        // Initialize a new puzzle.
+        let puzzle = sample_puzzle();
+
+        // Initialize an epoch hash.
+        let epoch_hash = rng.gen();
+
+        for batch_size in 1..=CurrentNetwork::MAX_SOLUTIONS {
+            // Initialize the incorrect solutions.
+            let incorrect_solutions = (0..batch_size)
+                .map(|_| {
+                    let solution = puzzle.prove(epoch_hash, rng.gen(), rng.gen(), None).unwrap();
+                    Solution::new(*solution.partial_solution(), solution.target().saturating_add(1))
+                })
+                .collect::<Vec<_>>();
+            let incorrect_solutions = PuzzleSolutions::new(incorrect_solutions).unwrap();
+
+            // Ensure the incorrect solutions are invalid.
+            assert!(puzzle.check_solutions(&incorrect_solutions, epoch_hash, 0u64).is_err());
+
+            // Ensure the incorrect solutions are invalid on a fresh puzzle instance.
+            let new_puzzle = sample_puzzle();
+            assert!(new_puzzle.check_solutions(&incorrect_solutions, epoch_hash, 0u64).is_err());
+        }
+    }
+
     #[test]
     fn test_check_solutions_with_duplicate_nonces() {
         let mut rng = TestRng::default();
diff --git a/ledger/puzzle/src/partial_solution/bytes.rs b/ledger/puzzle/src/partial_solution/bytes.rs
new file mode 100644
index 000000000..5434a25fc
--- /dev/null
+++ b/ledger/puzzle/src/partial_solution/bytes.rs
@@ -0,0 +1,60 @@
+// Copyright (C) 2019-2023 Aleo Systems Inc.
+// This file is part of the snarkVM library.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+// http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use super::*;
+
+impl<N: Network> FromBytes for PartialSolution<N> {
+    /// Reads the partial solution from the buffer.
+    fn read_le<R: Read>(mut reader: R) -> IoResult<Self> {
+        let epoch_hash = N::BlockHash::read_le(&mut reader)?;
+        let address = Address::<N>::read_le(&mut reader)?;
+        let counter = u64::read_le(&mut reader)?;
+
+        Self::new(epoch_hash, address, counter).map_err(error)
+    }
+}
+
+impl<N: Network> ToBytes for PartialSolution<N> {
+    /// Writes the parital solution to the buffer.
+    fn write_le<W: Write>(&self, mut writer: W) -> IoResult<()> {
+        self.epoch_hash.write_le(&mut writer)?;
+        self.address.write_le(&mut writer)?;
+        self.counter.write_le(&mut writer)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use console::{account::PrivateKey, network::MainnetV0};
+
+    type CurrentNetwork = MainnetV0;
+
+    #[test]
+    fn test_bytes() -> Result<()> {
+        let mut rng = TestRng::default();
+        let private_key = PrivateKey::<CurrentNetwork>::new(&mut rng)?;
+        let address = Address::try_from(private_key)?;
+
+        // Sample a new partial solution.
+        let expected = PartialSolution::new(rng.gen(), address, u64::rand(&mut rng)).unwrap();
+
+        // Check the byte representation.
+        let expected_bytes = expected.to_bytes_le()?;
+        assert_eq!(expected, PartialSolution::read_le(&expected_bytes[..])?);
+        assert!(PartialSolution::<CurrentNetwork>::read_le(&expected_bytes[1..]).is_err());
+
+        Ok(())
+    }
+}
diff --git a/ledger/puzzle/src/partial_solution/mod.rs b/ledger/puzzle/src/partial_solution/mod.rs
new file mode 100644
index 000000000..37ceb38a5
--- /dev/null
+++ b/ledger/puzzle/src/partial_solution/mod.rs
@@ -0,0 +1,63 @@
+// Copyright (C) 2019-2023 Aleo Systems Inc.
+// This file is part of the snarkVM library.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+// http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+mod bytes;
+mod serialize;
+mod string;
+
+use crate::SolutionID;
+use console::{account::Address, network::prelude::*, prelude::DeserializeExt};
+
+/// The partial solution for the puzzle from a prover.
+#[derive(Copy, Clone, Eq, PartialEq, Hash)]
+pub struct PartialSolution<N: Network> {
+    /// The solution ID.
+    solution_id: SolutionID<N>,
+    /// The epoch hash.
+    epoch_hash: N::BlockHash,
+    /// The address of the prover.
+    address: Address<N>,
+    /// The counter for the solution.
+    counter: u64,
+}
+
+impl<N: Network> PartialSolution<N> {
+    /// Initializes a new instance of the partial solution.
+    pub fn new(epoch_hash: N::BlockHash, address: Address<N>, counter: u64) -> Result<Self> {
+        // Compute the solution ID.
+        let solution_id = SolutionID::new(epoch_hash, address, counter)?;
+        // Return the partial solution.
+        Ok(Self { solution_id, epoch_hash, address, counter })
+    }
+
+    /// Returns the solution ID.
+    pub const fn id(&self) -> SolutionID<N> {
+        self.solution_id
+    }
+
+    /// Returns the epoch hash of the solution.
+    pub const fn epoch_hash(&self) -> N::BlockHash {
+        self.epoch_hash
+    }
+
+    /// Returns the address of the prover.
+    pub const fn address(&self) -> Address<N> {
+        self.address
+    }
+
+    /// Returns the counter for the solution.
+    pub const fn counter(&self) -> u64 {
+        self.counter
+    }
+}
diff --git a/ledger/puzzle/src/partial_solution/serialize.rs b/ledger/puzzle/src/partial_solution/serialize.rs
new file mode 100644
index 000000000..944b056d7
--- /dev/null
+++ b/ledger/puzzle/src/partial_solution/serialize.rs
@@ -0,0 +1,110 @@
+// Copyright (C) 2019-2023 Aleo Systems Inc.
+// This file is part of the snarkVM library.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+// http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use super::*;
+
+impl<N: Network> Serialize for PartialSolution<N> {
+    /// Serializes the partial solution to a JSON-string or buffer.
+    fn serialize<S: Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
+        match serializer.is_human_readable() {
+            true => {
+                let mut partial_solution = serializer.serialize_struct("PartialSolution", 4)?;
+                partial_solution.serialize_field("solution_id", &self.solution_id)?;
+                partial_solution.serialize_field("epoch_hash", &self.epoch_hash)?;
+                partial_solution.serialize_field("address", &self.address)?;
+                partial_solution.serialize_field("counter", &self.counter)?;
+                partial_solution.end()
+            }
+            false => ToBytesSerializer::serialize_with_size_encoding(self, serializer),
+        }
+    }
+}
+
+impl<'de, N: Network> Deserialize<'de> for PartialSolution<N> {
+    /// Deserializes the partial solution from a JSON-string or buffer.
+    fn deserialize<D: Deserializer<'de>>(deserializer: D) -> Result<Self, D::Error> {
+        match deserializer.is_human_readable() {
+            true => {
+                let mut partial_solution = serde_json::Value::deserialize(deserializer)?;
+                let solution_id: SolutionID<N> =
+                    DeserializeExt::take_from_value::<D>(&mut partial_solution, "solution_id")?;
+
+                // Recover the partial solution.
+                let partial_solution = Self::new(
+                    DeserializeExt::take_from_value::<D>(&mut partial_solution, "epoch_hash")?,
+                    DeserializeExt::take_from_value::<D>(&mut partial_solution, "address")?,
+                    DeserializeExt::take_from_value::<D>(&mut partial_solution, "counter")?,
+                )
+                .map_err(de::Error::custom)?;
+
+                // Ensure the solution ID matches.
+                match solution_id == partial_solution.id() {
+                    true => Ok(partial_solution),
+                    false => Err(de::Error::custom(error("Mismatching solution ID, possible data corruption"))),
+                }
+            }
+            false => FromBytesDeserializer::<Self>::deserialize_with_size_encoding(deserializer, "solution"),
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use console::{account::PrivateKey, network::MainnetV0};
+
+    type CurrentNetwork = MainnetV0;
+
+    #[test]
+    fn test_serde_json() -> Result<()> {
+        let mut rng = TestRng::default();
+        let private_key = PrivateKey::<CurrentNetwork>::new(&mut rng)?;
+        let address = Address::try_from(private_key)?;
+
+        // Sample a new partial solution.
+        let expected = PartialSolution::new(rng.gen(), address, u64::rand(&mut rng)).unwrap();
+
+        // Serialize
+        let expected_string = &expected.to_string();
+        let candidate_string = serde_json::to_string(&expected)?;
+        assert_eq!(expected, serde_json::from_str(&candidate_string)?);
+
+        // Deserialize
+        assert_eq!(expected, PartialSolution::from_str(expected_string)?);
+        assert_eq!(expected, serde_json::from_str(&candidate_string)?);
+
+        Ok(())
+    }
+
+    #[test]
+    fn test_bincode() -> Result<()> {
+        let mut rng = TestRng::default();
+        let private_key = PrivateKey::<CurrentNetwork>::new(&mut rng)?;
+        let address = Address::try_from(private_key)?;
+
+        // Sample a new partial solution.
+        let expected = PartialSolution::new(rng.gen(), address, u64::rand(&mut rng)).unwrap();
+
+        // Serialize
+        let expected_bytes = expected.to_bytes_le()?;
+        let expected_bytes_with_size_encoding = bincode::serialize(&expected)?;
+        assert_eq!(&expected_bytes[..], &expected_bytes_with_size_encoding[8..]);
+
+        // Deserialize
+        assert_eq!(expected, PartialSolution::read_le(&expected_bytes[..])?);
+        assert_eq!(expected, bincode::deserialize(&expected_bytes_with_size_encoding[..])?);
+
+        Ok(())
+    }
+}
diff --git a/ledger/puzzle/src/partial_solution/string.rs b/ledger/puzzle/src/partial_solution/string.rs
new file mode 100644
index 000000000..d75703341
--- /dev/null
+++ b/ledger/puzzle/src/partial_solution/string.rs
@@ -0,0 +1,62 @@
+// Copyright (C) 2019-2023 Aleo Systems Inc.
+// This file is part of the snarkVM library.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+// http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use super::*;
+
+impl<N: Network> FromStr for PartialSolution<N> {
+    type Err = Error;
+
+    /// Initializes the partial solution from a JSON-string.
+    fn from_str(solution: &str) -> Result<Self, Self::Err> {
+        Ok(serde_json::from_str(solution)?)
+    }
+}
+
+impl<N: Network> Debug for PartialSolution<N> {
+    /// Prints the partial solution as a JSON-string.
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        Display::fmt(self, f)
+    }
+}
+
+impl<N: Network> Display for PartialSolution<N> {
+    /// Displays the partial solution as a JSON-string.
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        write!(f, "{}", serde_json::to_string(self).map_err::<fmt::Error, _>(ser::Error::custom)?)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use console::{account::PrivateKey, network::MainnetV0};
+
+    type CurrentNetwork = MainnetV0;
+
+    #[test]
+    fn test_string() -> Result<()> {
+        let mut rng = TestRng::default();
+        let private_key = PrivateKey::<CurrentNetwork>::new(&mut rng)?;
+        let address = Address::try_from(private_key)?;
+
+        // Sample a new partial solution.
+        let expected = PartialSolution::new(rng.gen(), address, u64::rand(&mut rng)).unwrap();
+
+        // Check the string representation.
+        let candidate = expected.to_string();
+        assert_eq!(expected, PartialSolution::from_str(&candidate)?);
+
+        Ok(())
+    }
+}
diff --git a/ledger/puzzle/src/solution/bytes.rs b/ledger/puzzle/src/solution/bytes.rs
index b8022389c..66d0ed9c6 100644
--- a/ledger/puzzle/src/solution/bytes.rs
+++ b/ledger/puzzle/src/solution/bytes.rs
@@ -17,20 +17,18 @@ use super::*;
 impl<N: Network> FromBytes for Solution<N> {
     /// Reads the solution from the buffer.
     fn read_le<R: Read>(mut reader: R) -> IoResult<Self> {
-        let epoch_hash = N::BlockHash::read_le(&mut reader)?;
-        let address = Address::<N>::read_le(&mut reader)?;
-        let counter = u64::read_le(&mut reader)?;
+        let partial_solution = PartialSolution::read_le(&mut reader)?;
+        let target = u64::read_le(&mut reader)?;
 
-        Self::new(epoch_hash, address, counter).map_err(error)
+        Ok(Self::new(partial_solution, target))
     }
 }
 
 impl<N: Network> ToBytes for Solution<N> {
     /// Writes the solution to the buffer.
     fn write_le<W: Write>(&self, mut writer: W) -> IoResult<()> {
-        self.epoch_hash.write_le(&mut writer)?;
-        self.address.write_le(&mut writer)?;
-        self.counter.write_le(&mut writer)
+        self.partial_solution.write_le(&mut writer)?;
+        self.target.write_le(&mut writer)
     }
 }
 
@@ -48,7 +46,9 @@ mod tests {
         let address = Address::try_from(private_key)?;
 
         // Sample a new solution.
-        let expected = Solution::new(rng.gen(), address, u64::rand(&mut rng)).unwrap();
+        let partial_solution = PartialSolution::new(rng.gen(), address, u64::rand(&mut rng)).unwrap();
+        let target = u64::rand(&mut rng);
+        let expected = Solution::new(partial_solution, target);
 
         // Check the byte representation.
         let expected_bytes = expected.to_bytes_le()?;
diff --git a/ledger/puzzle/src/solution/mod.rs b/ledger/puzzle/src/solution/mod.rs
index e8a2ecce4..e163d4d73 100644
--- a/ledger/puzzle/src/solution/mod.rs
+++ b/ledger/puzzle/src/solution/mod.rs
@@ -16,48 +16,51 @@ mod bytes;
 mod serialize;
 mod string;
 
-use crate::SolutionID;
+use crate::{PartialSolution, SolutionID};
 use console::{account::Address, network::prelude::*, prelude::DeserializeExt};
 
-/// The solution for the puzzle from a prover.
+/// A helper struct around a puzzle solution.
 #[derive(Copy, Clone, Eq, PartialEq, Hash)]
 pub struct Solution<N: Network> {
-    /// The solution ID.
-    solution_id: SolutionID<N>,
-    /// The epoch hash.
-    epoch_hash: N::BlockHash,
-    /// The address of the prover.
-    address: Address<N>,
-    /// The counter for the solution.
-    counter: u64,
+    /// The partial solution.
+    partial_solution: PartialSolution<N>,
+    /// The solution target.
+    target: u64,
 }
 
 impl<N: Network> Solution<N> {
     /// Initializes a new instance of the solution.
-    pub fn new(epoch_hash: N::BlockHash, address: Address<N>, counter: u64) -> Result<Self> {
-        // Compute the solution ID.
-        let solution_id = SolutionID::new(epoch_hash, address, counter)?;
-        // Return the partial solution.
-        Ok(Self { solution_id, epoch_hash, address, counter })
+    pub fn new(partial_solution: PartialSolution<N>, target: u64) -> Self {
+        Self { partial_solution, target }
+    }
+
+    /// Returns the partial solution..
+    pub const fn partial_solution(&self) -> &PartialSolution<N> {
+        &self.partial_solution
     }
 
     /// Returns the solution ID.
     pub const fn id(&self) -> SolutionID<N> {
-        self.solution_id
+        self.partial_solution.id()
     }
 
     /// Returns the epoch hash of the solution.
     pub const fn epoch_hash(&self) -> N::BlockHash {
-        self.epoch_hash
+        self.partial_solution.epoch_hash()
     }
 
     /// Returns the address of the prover.
     pub const fn address(&self) -> Address<N> {
-        self.address
+        self.partial_solution.address()
     }
 
     /// Returns the counter for the solution.
     pub const fn counter(&self) -> u64 {
-        self.counter
+        self.partial_solution.counter()
+    }
+
+    /// Returns the target for the solution.
+    pub const fn target(&self) -> u64 {
+        self.target
     }
 }
diff --git a/ledger/puzzle/src/solution/serialize.rs b/ledger/puzzle/src/solution/serialize.rs
index 12ebd1b84..3678b7140 100644
--- a/ledger/puzzle/src/solution/serialize.rs
+++ b/ledger/puzzle/src/solution/serialize.rs
@@ -19,12 +19,10 @@ impl<N: Network> Serialize for Solution<N> {
     fn serialize<S: Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
         match serializer.is_human_readable() {
             true => {
-                let mut partial_solution = serializer.serialize_struct("Solution", 4)?;
-                partial_solution.serialize_field("solution_id", &self.solution_id)?;
-                partial_solution.serialize_field("epoch_hash", &self.epoch_hash)?;
-                partial_solution.serialize_field("address", &self.address)?;
-                partial_solution.serialize_field("counter", &self.counter)?;
-                partial_solution.end()
+                let mut solution = serializer.serialize_struct("Solution", 2)?;
+                solution.serialize_field("partial_solution", &self.partial_solution)?;
+                solution.serialize_field("target", &self.target)?;
+                solution.end()
             }
             false => ToBytesSerializer::serialize_with_size_encoding(self, serializer),
         }
@@ -36,23 +34,12 @@ impl<'de, N: Network> Deserialize<'de> for Solution<N> {
     fn deserialize<D: Deserializer<'de>>(deserializer: D) -> Result<Self, D::Error> {
         match deserializer.is_human_readable() {
             true => {
-                let mut partial_solution = serde_json::Value::deserialize(deserializer)?;
-                let solution_id: SolutionID<N> =
-                    DeserializeExt::take_from_value::<D>(&mut partial_solution, "solution_id")?;
-
-                // Recover the partial solution.
-                let solution = Self::new(
-                    DeserializeExt::take_from_value::<D>(&mut partial_solution, "epoch_hash")?,
-                    DeserializeExt::take_from_value::<D>(&mut partial_solution, "address")?,
-                    DeserializeExt::take_from_value::<D>(&mut partial_solution, "counter")?,
-                )
-                .map_err(de::Error::custom)?;
-
-                // Ensure the solution ID matches.
-                match solution_id == solution.id() {
-                    true => Ok(solution),
-                    false => Err(de::Error::custom(error("Mismatching solution ID, possible data corruption"))),
-                }
+                let mut solution = serde_json::Value::deserialize(deserializer)?;
+
+                Ok(Self::new(
+                    DeserializeExt::take_from_value::<D>(&mut solution, "partial_solution")?,
+                    DeserializeExt::take_from_value::<D>(&mut solution, "target")?,
+                ))
             }
             false => FromBytesDeserializer::<Self>::deserialize_with_size_encoding(deserializer, "solution"),
         }
@@ -73,7 +60,9 @@ mod tests {
         let address = Address::try_from(private_key)?;
 
         // Sample a new solution.
-        let expected = Solution::new(rng.gen(), address, u64::rand(&mut rng)).unwrap();
+        let partial_solution = PartialSolution::new(rng.gen(), address, u64::rand(&mut rng)).unwrap();
+        let target = u64::rand(&mut rng);
+        let expected = Solution::new(partial_solution, target);
 
         // Serialize
         let expected_string = &expected.to_string();
@@ -94,7 +83,9 @@ mod tests {
         let address = Address::try_from(private_key)?;
 
         // Sample a new solution.
-        let expected = Solution::new(rng.gen(), address, u64::rand(&mut rng)).unwrap();
+        let partial_solution = PartialSolution::new(rng.gen(), address, u64::rand(&mut rng)).unwrap();
+        let target = u64::rand(&mut rng);
+        let expected = Solution::new(partial_solution, target);
 
         // Serialize
         let expected_bytes = expected.to_bytes_le()?;
diff --git a/ledger/puzzle/src/solution/string.rs b/ledger/puzzle/src/solution/string.rs
index 187fd4870..79cb9ddfe 100644
--- a/ledger/puzzle/src/solution/string.rs
+++ b/ledger/puzzle/src/solution/string.rs
@@ -51,7 +51,9 @@ mod tests {
         let address = Address::try_from(private_key)?;
 
         // Sample a new solution.
-        let expected = Solution::new(rng.gen(), address, u64::rand(&mut rng)).unwrap();
+        let partial_solution = PartialSolution::new(rng.gen(), address, u64::rand(&mut rng)).unwrap();
+        let target = u64::rand(&mut rng);
+        let expected = Solution::new(partial_solution, target);
 
         // Check the string representation.
         let candidate = expected.to_string();
diff --git a/ledger/puzzle/src/solutions/mod.rs b/ledger/puzzle/src/solutions/mod.rs
index 906a2f606..6d618283e 100644
--- a/ledger/puzzle/src/solutions/mod.rs
+++ b/ledger/puzzle/src/solutions/mod.rs
@@ -86,6 +86,7 @@ impl<N: Network> Deref for PuzzleSolutions<N> {
 #[cfg(test)]
 mod tests {
     use super::*;
+    use crate::PartialSolution;
     use console::account::{Address, PrivateKey};
 
     use std::collections::HashSet;
@@ -103,7 +104,8 @@ mod tests {
             let private_key = PrivateKey::<CurrentNetwork>::new(rng).unwrap();
             let address = Address::try_from(private_key).unwrap();
 
-            let solution = Solution::new(rng.gen(), address, u64::rand(rng)).unwrap();
+            let partial_solution = PartialSolution::new(rng.gen(), address, u64::rand(rng)).unwrap();
+            let solution = Solution::new(partial_solution, u64::rand(rng));
             solutions.push(solution);
         }
         PuzzleSolutions::new(solutions).unwrap()
diff --git a/ledger/puzzle/src/solutions/serialize.rs b/ledger/puzzle/src/solutions/serialize.rs
index 0c429c157..b85d14461 100644
--- a/ledger/puzzle/src/solutions/serialize.rs
+++ b/ledger/puzzle/src/solutions/serialize.rs
@@ -44,6 +44,7 @@ impl<'de, N: Network> Deserialize<'de> for PuzzleSolutions<N> {
 #[cfg(test)]
 pub(super) mod tests {
     use super::*;
+    use crate::PartialSolution;
     use console::account::{Address, PrivateKey};
 
     type CurrentNetwork = console::network::MainnetV0;
@@ -55,7 +56,8 @@ pub(super) mod tests {
             let private_key = PrivateKey::<CurrentNetwork>::new(rng).unwrap();
             let address = Address::try_from(private_key).unwrap();
 
-            let solution = Solution::new(rng.gen(), address, u64::rand(rng)).unwrap();
+            let partial_solution = PartialSolution::new(rng.gen(), address, u64::rand(rng)).unwrap();
+            let solution = Solution::new(partial_solution, u64::rand(rng));
             solutions.push(solution);
         }
         PuzzleSolutions::new(solutions).unwrap()
diff --git a/ledger/src/tests.rs b/ledger/src/tests.rs
index 4089c7ba4..ce56a2efe 100644
--- a/ledger/src/tests.rs
+++ b/ledger/src/tests.rs
@@ -25,7 +25,7 @@ use console::{
     program::{Entry, Identifier, Literal, Plaintext, ProgramID, Value},
     types::U16,
 };
-use ledger_block::{ConfirmedTransaction, Ratify, Rejected, Transaction};
+use ledger_block::{ConfirmedTransaction, Execution, Ratify, Rejected, Transaction};
 use ledger_committee::{Committee, MIN_VALIDATOR_STAKE};
 use ledger_store::{helpers::memory::ConsensusMemory, ConsensusStore};
 use synthesizer::{program::Program, vm::VM, Stack};
@@ -779,56 +779,107 @@ fn test_execute_duplicate_input_ids() {
 
     // Fetch the unspent records.
     let records = find_records();
-    let record_1 = records[0].clone();
+    let record_execution = records[0].clone();
+    let record_deployment = records[1].clone();
 
-    // Prepare a transfer that spends the record.
+    // Prepare a transfer that spends a record.
     let inputs = [
-        Value::Record(record_1.clone()),
+        Value::Record(record_execution.clone()),
         Value::from_str(&format!("{address}")).unwrap(),
         Value::from_str("100u64").unwrap(),
     ];
-    let transfer_1 = ledger
-        .vm
-        .execute(&private_key, ("credits.aleo", "transfer_private"), inputs.into_iter(), None, 0, None, rng)
-        .unwrap();
-    let transfer_1_id = transfer_1.id();
 
-    // Prepare a transfer that attempts to spend the same record.
-    let inputs = [
-        Value::Record(record_1.clone()),
-        Value::from_str(&format!("{address}")).unwrap(),
-        Value::from_str("1000u64").unwrap(),
-    ];
-    let transfer_2 = ledger
-        .vm
-        .execute(&private_key, ("credits.aleo", "transfer_private"), inputs.into_iter(), None, 0, None, rng)
+    let num_duplicate_deployments = 3;
+    let mut executions = Vec::with_capacity(num_duplicate_deployments + 1);
+    let mut execution_ids = Vec::with_capacity(num_duplicate_deployments + 1);
+    let mut deployments = Vec::with_capacity(num_duplicate_deployments);
+    let mut deployment_ids = Vec::with_capacity(num_duplicate_deployments);
+
+    // Create Executions and Deployments, spending the same record.
+    for i in 0..num_duplicate_deployments {
+        // Execute.
+        let execution = ledger
+            .vm
+            .execute(&private_key, ("credits.aleo", "transfer_private"), inputs.clone().iter(), None, 0, None, rng)
+            .unwrap();
+        execution_ids.push(execution.id());
+        executions.push(execution);
+        // Deploy.
+        let program_id = ProgramID::<CurrentNetwork>::from_str(&format!("dummy_program_{i}.aleo")).unwrap();
+        let program = Program::<CurrentNetwork>::from_str(&format!(
+            "
+program {program_id};
+function foo:
+    input r0 as u8.private;
+    async foo r0 into r1;
+    output r1 as {program_id}/foo.future;
+finalize foo:
+    input r0 as u8.public;
+    add r0 r0 into r1;",
+        ))
         .unwrap();
-    let transfer_2_id = transfer_2.id();
+        let deployment =
+            ledger.vm.deploy(&private_key, &program, Some(record_deployment.clone()), 0, None, rng).unwrap();
+        deployment_ids.push(deployment.id());
+        deployments.push(deployment);
+    }
 
-    // Prepare a transfer that attempts to spend the same record in the fee.
+    // Create one more execution which spends the record as a fee.
     let inputs = [Value::from_str(&format!("{address}")).unwrap(), Value::from_str("100u64").unwrap()];
-    let transfer_3 = ledger
+    let execution = ledger
         .vm
         .execute(
             &private_key,
             ("credits.aleo", "transfer_public"),
-            inputs.into_iter(),
-            Some(record_1.clone()),
+            inputs.clone().iter(),
+            Some(record_execution.clone()),
             0,
             None,
             rng,
         )
         .unwrap();
-    let transfer_3_id = transfer_3.id();
-
-    // Prepare a transfer that attempts to spend the same record for the subsequent block.
-    let inputs =
-        [Value::Record(record_1), Value::from_str(&format!("{address}")).unwrap(), Value::from_str("1000u64").unwrap()];
-    let transfer_4 = ledger
+    execution_ids.push(execution.id());
+    executions.push(execution);
+
+    // Select a transaction to mutate by a malicious validator.
+    let transaction_to_mutate = executions.last().unwrap().clone();
+
+    // Create a mutated execution which adds one transition, resulting in a different transaction id.
+    // This simulates a malicious validator re-using execution content.
+    let execution_to_mutate = transaction_to_mutate.execution().unwrap();
+    // Sample a transition.
+    let sample = ledger_test_helpers::sample_transition(rng);
+    // Extend the transitions.
+    let mutated_transitions = std::iter::once(sample).chain(execution_to_mutate.transitions().cloned());
+    // Create a mutated execution.
+    let mutated_execution = Execution::from(
+        mutated_transitions,
+        execution_to_mutate.global_state_root(),
+        execution_to_mutate.proof().cloned(),
+    )
+    .unwrap();
+    // Create a new fee for the execution.
+    let fee_authorization = ledger
         .vm
-        .execute(&private_key, ("credits.aleo", "transfer_private"), inputs.into_iter(), None, 0, None, rng)
+        .authorize_fee_public(
+            &private_key,
+            *executions.last().unwrap().fee_amount().unwrap(),
+            0,
+            mutated_execution.to_execution_id().unwrap(),
+            rng,
+        )
         .unwrap();
-    let transfer_4_id = transfer_4.id();
+    let fee = ledger.vm.execute_fee_authorization(fee_authorization, None, rng).unwrap();
+    // Create a mutated transaction.
+    let mutated_transaction = Transaction::from_execution(mutated_execution, Some(fee)).unwrap();
+    execution_ids.push(mutated_transaction.id());
+    executions.push(mutated_transaction);
+
+    // Create a mutated execution which just takes the fee transition, resulting in a different transaction id.
+    // This simulates a malicious validator transforming a transaction to a fee transaction.
+    let mutated_transaction = Transaction::from_fee(transaction_to_mutate.fee_transition().unwrap()).unwrap();
+    execution_ids.push(mutated_transaction.id());
+    executions.push(mutated_transaction);
 
     // Create a block.
     let block = ledger
@@ -836,7 +887,15 @@ fn test_execute_duplicate_input_ids() {
             &private_key,
             vec![],
             vec![],
-            vec![transfer_1, transfer_2, transfer_3],
+            vec![
+                executions.pop().unwrap(),
+                executions.pop().unwrap(),
+                executions.pop().unwrap(),
+                executions.pop().unwrap(),
+                executions.pop().unwrap(),
+                deployments.pop().unwrap(),
+                deployments.pop().unwrap(),
+            ],
             rng,
         )
         .unwrap();
@@ -848,27 +907,45 @@ fn test_execute_duplicate_input_ids() {
     ledger.advance_to_next_block(&block).unwrap();
 
     // Enforce that the block transactions were correct.
-    assert_eq!(block.transactions().num_accepted(), 1);
-    assert_eq!(block.transactions().transaction_ids().collect::<Vec<_>>(), vec![&transfer_1_id]);
-    assert_eq!(block.aborted_transaction_ids(), &vec![transfer_2_id, transfer_3_id]);
-
-    // Ensure that verification was not run on aborted transactions.
+    assert_eq!(block.transactions().num_accepted(), 2);
+    println!("execution_ids: {:?}", execution_ids);
+    assert_eq!(block.transactions().transaction_ids().collect::<Vec<_>>(), vec![&execution_ids[2], &deployment_ids[2]]);
+    assert_eq!(block.aborted_transaction_ids(), &vec![
+        execution_ids[5],
+        execution_ids[4],
+        execution_ids[3],
+        execution_ids[1],
+        deployment_ids[1]
+    ]);
+
+    // Ensure that verification was not run on aborted deployments.
     let partially_verified_transaction = ledger.vm().partially_verified_transactions().read().clone();
-    assert!(partially_verified_transaction.contains(&transfer_1_id));
-    assert!(!partially_verified_transaction.contains(&transfer_2_id));
-    assert!(!partially_verified_transaction.contains(&transfer_3_id));
+
+    assert!(partially_verified_transaction.contains(&execution_ids[2]));
+    assert!(partially_verified_transaction.contains(&deployment_ids[2]));
+    assert!(!partially_verified_transaction.contains(&execution_ids[1]));
+    assert!(!partially_verified_transaction.contains(&deployment_ids[1]));
+    assert!(!partially_verified_transaction.contains(&execution_ids[3]));
+    assert!(!partially_verified_transaction.contains(&execution_ids[4])); // Verification was run, but the execution was invalid.
+    assert!(!partially_verified_transaction.contains(&execution_ids[5]));
 
     // Prepare a transfer that will succeed for the subsequent block.
     let inputs = [Value::from_str(&format!("{address}")).unwrap(), Value::from_str("1000u64").unwrap()];
-    let transfer_5 = ledger
+    let transfer = ledger
         .vm
         .execute(&private_key, ("credits.aleo", "transfer_public"), inputs.into_iter(), None, 0, None, rng)
         .unwrap();
-    let transfer_5_id = transfer_5.id();
+    let transfer_id = transfer.id();
 
     // Create a block.
     let block = ledger
-        .prepare_advance_to_next_beacon_block(&private_key, vec![], vec![], vec![transfer_4, transfer_5], rng)
+        .prepare_advance_to_next_beacon_block(
+            &private_key,
+            vec![],
+            vec![],
+            vec![executions.pop().unwrap(), deployments.pop().unwrap(), transfer],
+            rng,
+        )
         .unwrap();
 
     // Check that the next block is valid.
@@ -879,13 +956,14 @@ fn test_execute_duplicate_input_ids() {
 
     // Enforce that the block transactions were correct.
     assert_eq!(block.transactions().num_accepted(), 1);
-    assert_eq!(block.transactions().transaction_ids().collect::<Vec<_>>(), vec![&transfer_5_id]);
-    assert_eq!(block.aborted_transaction_ids(), &vec![transfer_4_id]);
+    assert_eq!(block.transactions().transaction_ids().collect::<Vec<_>>(), vec![&transfer_id]);
+    assert_eq!(block.aborted_transaction_ids(), &vec![execution_ids[0], deployment_ids[0]]);
 
-    // Ensure that verification was not run on aborted transactions.
+    // Ensure that verification was not run on transactions aborted in a previous block.
     let partially_verified_transaction = ledger.vm().partially_verified_transactions().read().clone();
-    assert!(partially_verified_transaction.contains(&transfer_5_id));
-    assert!(!partially_verified_transaction.contains(&transfer_4_id));
+    assert!(partially_verified_transaction.contains(&transfer_id));
+    assert!(!partially_verified_transaction.contains(&execution_ids[0]));
+    assert!(!partially_verified_transaction.contains(&deployment_ids[0]));
 }
 
 #[test]
@@ -893,7 +971,8 @@ fn test_execute_duplicate_output_ids() {
     let rng = &mut TestRng::default();
 
     // Initialize the test environment.
-    let crate::test_helpers::TestEnv { ledger, private_key, address, .. } = crate::test_helpers::sample_test_env(rng);
+    let crate::test_helpers::TestEnv { ledger, private_key, view_key, address, .. } =
+        crate::test_helpers::sample_test_env(rng);
 
     // Deploy a test program to the ledger.
     let program = Program::<CurrentNetwork>::from_str(
@@ -926,8 +1005,25 @@ function create_duplicate_record:
     // Add the block to the ledger.
     ledger.advance_to_next_block(&block).unwrap();
 
-    // Create a transaction with different transition ids, but with a fixed output record (output ID).
-    let mut create_transaction_with_duplicate_output_id = |x: u64| -> Transaction<CurrentNetwork> {
+    // A helper function to find records.
+    let find_records = || {
+        let microcredits = Identifier::from_str("microcredits").unwrap();
+        ledger
+            .find_records(&view_key, RecordsFilter::SlowUnspent(private_key))
+            .unwrap()
+            .filter(|(_, record)| match record.data().get(&microcredits) {
+                Some(Entry::Private(Plaintext::Literal(Literal::U64(amount), _))) => !amount.is_zero(),
+                _ => false,
+            })
+            .collect::<indexmap::IndexMap<_, _>>()
+    };
+
+    // Fetch the unspent records.
+    let records = find_records();
+    let record_1 = records[0].clone();
+
+    // Create an execution with different transition ids, but with a fixed output record (output ID).
+    let mut create_execution_with_duplicate_output_id = |x: u64| -> Transaction<CurrentNetwork> {
         // Use a fixed seed RNG.
         let fixed_rng = &mut TestRng::from_seed(1);
 
@@ -964,16 +1060,61 @@ function create_duplicate_record:
         Transaction::from_execution(execution, Some(fee)).unwrap()
     };
 
+    // Create an deployment with different transition ids, but with a fixed output record (output ID).
+    let create_deployment_with_duplicate_output_id = |x: u64| -> Transaction<CurrentNetwork> {
+        // Use a fixed seed RNG.
+        let fixed_rng = &mut TestRng::from_seed(1);
+
+        // Deploy a test program to the ledger.
+        let program = Program::<CurrentNetwork>::from_str(&format!(
+            "
+program dummy_program_{x}.aleo;
+
+record dummy_program:
+    owner as address.private;
+    rand_var as u64.private;
+
+function create_duplicate_record:
+    input r0 as u64.private;
+    cast self.caller 1u64 into r1 as dummy_program.record;
+    output r1 as dummy_program.record;"
+        ))
+        .unwrap();
+
+        // Create a transaction with a fixed rng.
+        let transaction = ledger.vm.deploy(&private_key, &program, None, 0, None, fixed_rng).unwrap();
+
+        // Extract the deployment and owner.
+        let deployment = transaction.deployment().unwrap().clone();
+        let owner = *transaction.owner().unwrap();
+
+        // Create a new fee for the execution.
+        let fee_authorization = ledger
+            .vm
+            .authorize_fee_private(
+                &private_key,
+                record_1.clone(),
+                *transaction.fee_amount().unwrap(),
+                0,
+                deployment.to_deployment_id().unwrap(),
+                fixed_rng,
+            )
+            .unwrap();
+        let fee = ledger.vm.execute_fee_authorization(fee_authorization, None, fixed_rng).unwrap();
+
+        Transaction::from_deployment(owner, deployment, fee).unwrap()
+    };
+
     // Create the first transfer.
-    let transfer_1 = create_transaction_with_duplicate_output_id(1);
+    let transfer_1 = create_execution_with_duplicate_output_id(1);
     let transfer_1_id = transfer_1.id();
 
     // Create a second transfer with the same output id.
-    let transfer_2 = create_transaction_with_duplicate_output_id(2);
+    let transfer_2 = create_execution_with_duplicate_output_id(2);
     let transfer_2_id = transfer_2.id();
 
     // Create a third transfer with the same output id.
-    let transfer_3 = create_transaction_with_duplicate_output_id(3);
+    let transfer_3 = create_execution_with_duplicate_output_id(3);
     let transfer_3_id = transfer_3.id();
 
     // Ensure that each transaction has a duplicate output id.
@@ -983,9 +1124,34 @@ function create_duplicate_record:
     assert_eq!(tx_1_output_id, tx_2_output_id);
     assert_eq!(tx_1_output_id, tx_3_output_id);
 
+    // Create the first deployment.
+    let deployment_1 = create_deployment_with_duplicate_output_id(1);
+    let deployment_1_id = deployment_1.id();
+
+    // Create a second deployment with the same output id.
+    let deployment_2 = create_deployment_with_duplicate_output_id(2);
+    let deployment_2_id = deployment_2.id();
+
+    // Create a third deployment with the same output id.
+    let deployment_3 = create_deployment_with_duplicate_output_id(3);
+    let deployment_3_id = deployment_3.id();
+
+    // Ensure that each transaction has a duplicate output id.
+    let deployment_1_output_id = deployment_1.output_ids().next().unwrap();
+    let deployment_2_output_id = deployment_2.output_ids().next().unwrap();
+    let deployment_3_output_id = deployment_3.output_ids().next().unwrap();
+    assert_eq!(deployment_1_output_id, deployment_2_output_id);
+    assert_eq!(deployment_1_output_id, deployment_3_output_id);
+
     // Create a block.
     let block = ledger
-        .prepare_advance_to_next_beacon_block(&private_key, vec![], vec![], vec![transfer_1, transfer_2], rng)
+        .prepare_advance_to_next_beacon_block(
+            &private_key,
+            vec![],
+            vec![],
+            vec![transfer_1, transfer_2, deployment_1, deployment_2],
+            rng,
+        )
         .unwrap();
 
     // Check that the next block is valid.
@@ -995,14 +1161,16 @@ function create_duplicate_record:
     ledger.advance_to_next_block(&block).unwrap();
 
     // Enforce that the block transactions were correct.
-    assert_eq!(block.transactions().num_accepted(), 1);
-    assert_eq!(block.transactions().transaction_ids().collect::<Vec<_>>(), vec![&transfer_1_id]);
-    assert_eq!(block.aborted_transaction_ids(), &vec![transfer_2_id]);
+    assert_eq!(block.transactions().num_accepted(), 2);
+    assert_eq!(block.transactions().transaction_ids().collect::<Vec<_>>(), vec![&transfer_1_id, &deployment_1_id]);
+    assert_eq!(block.aborted_transaction_ids(), &vec![transfer_2_id, deployment_2_id]);
 
-    // Ensure that verification was not run on aborted transactions.
+    // Ensure that verification was not run on aborted deployments.
     let partially_verified_transaction = ledger.vm().partially_verified_transactions().read().clone();
     assert!(partially_verified_transaction.contains(&transfer_1_id));
+    assert!(partially_verified_transaction.contains(&deployment_1_id));
     assert!(!partially_verified_transaction.contains(&transfer_2_id));
+    assert!(!partially_verified_transaction.contains(&deployment_2_id));
 
     // Prepare a transfer that will succeed for the subsequent block.
     let inputs = [Value::from_str(&format!("{address}")).unwrap(), Value::from_str("1000u64").unwrap()];
@@ -1014,7 +1182,13 @@ function create_duplicate_record:
 
     // Create a block.
     let block = ledger
-        .prepare_advance_to_next_beacon_block(&private_key, vec![], vec![], vec![transfer_3, transfer_4], rng)
+        .prepare_advance_to_next_beacon_block(
+            &private_key,
+            vec![],
+            vec![],
+            vec![transfer_3, transfer_4, deployment_3],
+            rng,
+        )
         .unwrap();
 
     // Check that the next block is valid.
@@ -1026,12 +1200,13 @@ function create_duplicate_record:
     // Enforce that the block transactions were correct.
     assert_eq!(block.transactions().num_accepted(), 1);
     assert_eq!(block.transactions().transaction_ids().collect::<Vec<_>>(), vec![&transfer_4_id]);
-    assert_eq!(block.aborted_transaction_ids(), &vec![transfer_3_id]);
+    assert_eq!(block.aborted_transaction_ids(), &vec![transfer_3_id, deployment_3_id]);
 
-    // Ensure that verification was not run on aborted transactions.
+    // Ensure that verification was not run on transactions aborted in a previous block.
     let partially_verified_transaction = ledger.vm().partially_verified_transactions().read().clone();
     assert!(partially_verified_transaction.contains(&transfer_4_id));
     assert!(!partially_verified_transaction.contains(&transfer_3_id));
+    assert!(!partially_verified_transaction.contains(&deployment_3_id));
 }
 
 #[test]
@@ -1067,6 +1242,9 @@ function empty_function:
     ledger.advance_to_next_block(&block).unwrap();
 
     // Create a transaction with different transaction IDs, but with a fixed transition ID.
+    // NOTE: there's no use creating deployments with duplicate (fee) transition ids,
+    //       as this is only possible if they have duplicate programs, duplicate transaction_ids,
+    //       which will not abort but fail on check_next_block.
     let mut create_transaction_with_duplicate_transition_id = || -> Transaction<CurrentNetwork> {
         // Use a fixed seed RNG.
         let fixed_rng = &mut TestRng::from_seed(1);
@@ -1174,7 +1352,7 @@ function empty_function:
     assert_eq!(block.transactions().transaction_ids().collect::<Vec<_>>(), vec![&transfer_transaction_id]);
     assert_eq!(block.aborted_transaction_ids(), &vec![transaction_3_id]);
 
-    // Ensure that verification was not run on aborted transactions.
+    // Ensure that verification was not run on transactions aborted in a previous block.
     let partially_verified_transaction = ledger.vm().partially_verified_transactions().read().clone();
     assert!(partially_verified_transaction.contains(&transfer_transaction_id));
     assert!(!partially_verified_transaction.contains(&transaction_3_id));
@@ -1215,7 +1393,10 @@ function simple_output:
     // Add the block to the ledger.
     ledger.advance_to_next_block(&block).unwrap();
 
-    // Create a transaction with different transaction ids, but with a TPK.
+    // Create a transaction with different transaction ids, but with a duplicate TPK.
+    // NOTE: there's no use creating deployments with duplicate (fee) TPKs,
+    //       as this is only possible if they have duplicate programs, duplicate transaction_ids,
+    //       which will not abort but fail on check_next_block.
     let mut create_transaction_with_duplicate_tpk = |function: &str| -> Transaction<CurrentNetwork> {
         // Use a fixed seed RNG.
         let fixed_rng = &mut TestRng::from_seed(1);
@@ -1321,7 +1502,7 @@ function simple_output:
     assert_eq!(block.transactions().transaction_ids().collect::<Vec<_>>(), vec![&transfer_transaction_id]);
     assert_eq!(block.aborted_transaction_ids(), &vec![transaction_3_id]);
 
-    // Ensure that verification was not run on aborted transactions.
+    // Ensure that verification was not run on transactions aborted in a previous block.
     let partially_verified_transaction = ledger.vm().partially_verified_transactions().read().clone();
     assert!(partially_verified_transaction.contains(&transfer_transaction_id));
     assert!(!partially_verified_transaction.contains(&transaction_3_id));
diff --git a/parameters/src/canary/mod.rs b/parameters/src/canary/mod.rs
index 46f7a1bff..08bb2c1e4 100644
--- a/parameters/src/canary/mod.rs
+++ b/parameters/src/canary/mod.rs
@@ -15,7 +15,7 @@
 pub mod genesis;
 pub use genesis::*;
 
-const REMOTE_URL: &str = "https://s3-us-west-1.amazonaws.com/canary.parameters";
+const REMOTE_URL: &str = "https://parameters.aleo.org/canary";
 
 // BondPublic
 impl_remote!(BondPublicProver, REMOTE_URL, "resources/", "bond_public", "prover");
diff --git a/parameters/src/mainnet/mod.rs b/parameters/src/mainnet/mod.rs
index 735a2b474..2499dd1d9 100644
--- a/parameters/src/mainnet/mod.rs
+++ b/parameters/src/mainnet/mod.rs
@@ -18,7 +18,7 @@ pub use genesis::*;
 pub mod powers;
 pub use powers::*;
 
-const REMOTE_URL: &str = "https://s3-us-west-1.amazonaws.com/mainnet.parameters";
+const REMOTE_URL: &str = "https://parameters.aleo.org/mainnet";
 
 // Degrees
 #[cfg(not(feature = "wasm"))]
diff --git a/parameters/src/testnet/mod.rs b/parameters/src/testnet/mod.rs
index edc6da8b3..17a23fdaf 100644
--- a/parameters/src/testnet/mod.rs
+++ b/parameters/src/testnet/mod.rs
@@ -15,7 +15,7 @@
 pub mod genesis;
 pub use genesis::*;
 
-const REMOTE_URL: &str = "https://s3-us-west-1.amazonaws.com/testnet.parameters";
+const REMOTE_URL: &str = "https://parameters.aleo.org/testnet";
 
 // BondPublic
 impl_remote!(BondPublicProver, REMOTE_URL, "resources/", "bond_public", "prover");
diff --git a/synthesizer/Cargo.toml b/synthesizer/Cargo.toml
index 13167e27d..8b8248eec 100644
--- a/synthesizer/Cargo.toml
+++ b/synthesizer/Cargo.toml
@@ -198,5 +198,8 @@ features = [ "preserve_order" ]
 [dev-dependencies.serde_yaml]
 version = "0.9"
 
+[dev-dependencies.tempfile]
+version = "3"
+
 [dev-dependencies.walkdir]
 version = "2"
diff --git a/synthesizer/program/src/resources/credits.aleo b/synthesizer/program/src/resources/credits.aleo
index 4ac0a7f37..fb370ef29 100644
--- a/synthesizer/program/src/resources/credits.aleo
+++ b/synthesizer/program/src/resources/credits.aleo
@@ -43,7 +43,7 @@
 // 10. count committee[] == metadata[aleo1qqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqq3ljyzc]
 // 11. 100,000 + count committee[] >= count bonded[] >= count delegated[] >= count committee[]
 // 12. bond[*].microcredits > 100
-// 13. delegated[*] >= 10,000
+
 
 /**********************************************************************************************************************/
 
@@ -284,7 +284,7 @@ finalize bond_validator:
 // This function allows any delegator to bond their microcredits to a validator and specify a withdrawal address.
 //
 // The declared validator is **not** required to be in the committee yet, however it may **not** be unbonding.
-// This function enforces a minimum of 10,000 microcredits for each delegator.
+// This function enforces a minimum of 10,000 credits for each delegator.
 // The maximum number of delegators allowed in the committee is 100,000 delegator addresses.
 //
 // The corresponding functions for 'bond_public' are 'bond_validator', 'unbond_public', and 'claim_unbond_public'.
diff --git a/synthesizer/program/tests/instruction/assert.rs b/synthesizer/program/tests/instruction/assert.rs
index 61811eaef..e35b8ac3f 100644
--- a/synthesizer/program/tests/instruction/assert.rs
+++ b/synthesizer/program/tests/instruction/assert.rs
@@ -28,7 +28,7 @@ use synthesizer_process::{Process, Stack};
 type CurrentNetwork = MainnetV0;
 type CurrentAleo = AleoV0;
 
-const ITERATIONS: usize = 100;
+const ITERATIONS: usize = 25;
 
 /// Samples the stack. Note: Do not replicate this for real program use, it is insecure.
 fn sample_stack(
diff --git a/synthesizer/program/tests/instruction/commit.rs b/synthesizer/program/tests/instruction/commit.rs
index abdc6f6cb..de594e80a 100644
--- a/synthesizer/program/tests/instruction/commit.rs
+++ b/synthesizer/program/tests/instruction/commit.rs
@@ -41,7 +41,7 @@ use synthesizer_process::{Process, Stack};
 type CurrentNetwork = MainnetV0;
 type CurrentAleo = AleoV0;
 
-const ITERATIONS: usize = 50;
+const ITERATIONS: usize = 25;
 
 /// **Attention**: When changing this, also update in `src/logic/instruction/commit.rs`.
 fn valid_destination_types() -> &'static [LiteralType] {
diff --git a/synthesizer/program/tests/instruction/hash.rs b/synthesizer/program/tests/instruction/hash.rs
index 46f0caf73..eb5577231 100644
--- a/synthesizer/program/tests/instruction/hash.rs
+++ b/synthesizer/program/tests/instruction/hash.rs
@@ -50,7 +50,7 @@ use synthesizer_process::{Process, Stack};
 type CurrentNetwork = MainnetV0;
 type CurrentAleo = AleoV0;
 
-const ITERATIONS: usize = 50;
+const ITERATIONS: usize = 25;
 
 /// **Attention**: When changing this, also update in `src/logic/instruction/hash.rs`.
 fn valid_destination_types<N: Network>() -> &'static [PlaintextType<N>] {
diff --git a/synthesizer/program/tests/instruction/is.rs b/synthesizer/program/tests/instruction/is.rs
index a60121da9..c281085d6 100644
--- a/synthesizer/program/tests/instruction/is.rs
+++ b/synthesizer/program/tests/instruction/is.rs
@@ -37,7 +37,7 @@ use synthesizer_process::{Process, Stack};
 type CurrentNetwork = MainnetV0;
 type CurrentAleo = AleoV0;
 
-const ITERATIONS: usize = 100;
+const ITERATIONS: usize = 25;
 
 /// Samples the stack. Note: Do not replicate this for real program use, it is insecure.
 #[allow(clippy::type_complexity)]
diff --git a/synthesizer/src/vm/finalize.rs b/synthesizer/src/vm/finalize.rs
index 78e3ab8a6..c8182aabd 100644
--- a/synthesizer/src/vm/finalize.rs
+++ b/synthesizer/src/vm/finalize.rs
@@ -855,6 +855,12 @@ impl<N: Network, C: ConsensusStorage<N>> VM<N, C> {
         // Abort the transactions that are have duplicates or are invalid. This will prevent the VM from performing
         // verification on transactions that would have been aborted in `VM::atomic_speculate`.
         for transaction in transactions.iter() {
+            // Abort the transaction early if it is a fee transaction.
+            if transaction.is_fee() {
+                aborted_transactions.push((*transaction, "Fee transactions are not allowed in speculate".to_string()));
+                continue;
+            }
+
             // Determine if the transaction should be aborted.
             match self.should_abort_transaction(
                 transaction,
@@ -900,14 +906,6 @@ impl<N: Network, C: ConsensusStorage<N>> VM<N, C> {
             // Verify the transactions and collect the error message if there is one.
             let (valid, invalid): (Vec<_>, Vec<_>) =
                 cfg_into_iter!(transactions).zip(rngs).partition_map(|(transaction, mut rng)| {
-                    // Abort the transaction if it is a fee transaction.
-                    if transaction.is_fee() {
-                        return Either::Right((
-                            *transaction,
-                            "Fee transactions are not allowed in speculate".to_string(),
-                        ));
-                    }
-
                     // Verify the transaction.
                     match self.check_transaction(transaction, None, &mut rng) {
                         // If the transaction is valid, add it to the list of valid transactions.
diff --git a/synthesizer/src/vm/helpers/rewards.rs b/synthesizer/src/vm/helpers/rewards.rs
index 6dcc5af96..3ffc3a740 100644
--- a/synthesizer/src/vm/helpers/rewards.rs
+++ b/synthesizer/src/vm/helpers/rewards.rs
@@ -34,7 +34,7 @@ const MAX_COINBASE_REWARD: u64 = ledger_block::MAX_COINBASE_REWARD; // Coinbase
 ///
 /// This method ensures that stakers who are bonded to validators with more than **25%**
 /// of the total stake will not receive a staking reward. In addition, this method
-/// ensures stakers who have less than 10 credit are not eligible for a staking reward.
+/// ensures delegators who have less than 10,000 credits are not eligible for a staking reward.
 ///
 /// The choice of 25% is to ensure at least 4 validators are operational at any given time,
 /// since our security model adheres to 3f+1, where f=1. As such, we tolerate Byzantine behavior
diff --git a/synthesizer/src/vm/mod.rs b/synthesizer/src/vm/mod.rs
index 669970efc..1649cc7cc 100644
--- a/synthesizer/src/vm/mod.rs
+++ b/synthesizer/src/vm/mod.rs
@@ -398,7 +398,19 @@ impl<N: Network, C: ConsensusStorage<N>> VM<N, C> {
         self.block_store().pause_atomic_writes()?;
 
         // First, insert the block.
-        self.block_store().insert(block)?;
+        if let Err(insert_error) = self.block_store().insert(block) {
+            if cfg!(feature = "rocks") {
+                // Clear all pending atomic operations so that unpausing the atomic writes
+                // doesn't execute any of the queued storage operations.
+                self.block_store().abort_atomic();
+                // Disable the atomic batch override.
+                // Note: This call is guaranteed to succeed (without error), because `DISCARD_BATCH == true`.
+                self.block_store().unpause_atomic_writes::<true>()?;
+            }
+
+            return Err(insert_error);
+        };
+
         // Next, finalize the transactions.
         match self.finalize(state, block.ratifications(), block.solutions(), block.transactions()) {
             Ok(_ratified_finalize_operations) => {
@@ -450,12 +462,16 @@ pub(crate) mod test_helpers {
     };
     use ledger_block::{Block, Header, Metadata, Transition};
     use ledger_store::helpers::memory::ConsensusMemory;
+    #[cfg(feature = "rocks")]
+    use ledger_store::helpers::rocksdb::ConsensusDB;
     use ledger_test_helpers::{large_transaction_program, small_transaction_program};
     use synthesizer_program::Program;
 
     use indexmap::IndexMap;
     use once_cell::sync::OnceCell;
     use std::borrow::Borrow;
+    #[cfg(feature = "rocks")]
+    use std::path::Path;
     use synthesizer_snark::VerifyingKey;
 
     pub(crate) type CurrentNetwork = MainnetV0;
@@ -470,6 +486,12 @@ pub(crate) mod test_helpers {
         VM::from(ConsensusStore::open(None).unwrap()).unwrap()
     }
 
+    #[cfg(feature = "rocks")]
+    pub(crate) fn sample_vm_rocks(path: &Path) -> VM<CurrentNetwork, ConsensusDB<CurrentNetwork>> {
+        // Initialize a new VM.
+        VM::from(ConsensusStore::open(path.to_owned()).unwrap()).unwrap()
+    }
+
     pub(crate) fn sample_genesis_private_key(rng: &mut TestRng) -> PrivateKey<CurrentNetwork> {
         static INSTANCE: OnceCell<PrivateKey<CurrentNetwork>> = OnceCell::new();
         *INSTANCE.get_or_init(|| {
@@ -2475,4 +2497,33 @@ finalize transfer_public_to_private:
         // Ensure this call succeeds.
         vm.puzzle.prove(rng.gen(), rng.gen(), rng.gen(), None).unwrap();
     }
+
+    #[cfg(feature = "rocks")]
+    #[test]
+    fn test_atomic_unpause_on_error() {
+        let rng = &mut TestRng::default();
+
+        // Initialize a genesis private key..
+        let genesis_private_key = sample_genesis_private_key(rng);
+
+        // Initialize the genesis block.
+        let genesis = sample_genesis_block(rng);
+
+        // Initialize a VM and sample 2 blocks using it.
+        let vm = sample_vm();
+        vm.add_next_block(&genesis).unwrap();
+        let block1 = sample_next_block(&vm, &genesis_private_key, &[], rng).unwrap();
+        vm.add_next_block(&block1).unwrap();
+        let block2 = sample_next_block(&vm, &genesis_private_key, &[], rng).unwrap();
+
+        // Create a new, rocks-based VM shadowing the 1st one.
+        let tempdir = tempfile::tempdir().unwrap();
+        let vm = sample_vm_rocks(tempdir.path());
+        vm.add_next_block(&genesis).unwrap();
+        // This time, however, try to insert the 2nd block first, which fails due to height.
+        assert!(vm.add_next_block(&block2).is_err());
+
+        // It should still be possible to insert the 1st block afterwards.
+        vm.add_next_block(&block1).unwrap();
+    }
 }
