diff --git a/cli/src/commands/start.rs b/cli/src/commands/start.rs
index 562e00073..809242433 100644
--- a/cli/src/commands/start.rs
+++ b/cli/src/commands/start.rs
@@ -55,6 +55,7 @@ const RECOMMENDED_MIN_NOFILES_LIMIT: u64 = 2048;
 
 /// The development mode RNG seed.
 const DEVELOPMENT_MODE_RNG_SEED: u64 = 1234567890u64;
+
 /// The development mode number of genesis committee members.
 const DEVELOPMENT_MODE_NUM_GENESIS_COMMITTEE_MEMBERS: u16 = 4;
 
@@ -161,7 +162,7 @@ pub struct Start {
     /// If development mode is enabled, specify the number of genesis validators (default: 4)
     #[clap(long)]
     pub dev_num_validators: Option<u16>,
-    /// If developtment mode is enabled, specify whether node 0 should generate traffic to drive the network
+    /// If development mode is enabled, specify whether node 0 should generate traffic to drive the network
     #[clap(default_value = "false", long = "no-dev-txs")]
     pub no_dev_txs: bool,
     /// If development mode is enabled, specify the custom bonded balances as a JSON object (default: None)
@@ -742,8 +743,8 @@ fn load_or_compute_genesis<N: Network>(
     // Initialize the hasher.
     let hasher = snarkvm::console::algorithms::BHP256::<N>::setup("aleo.dev.block")?;
     // Compute the hash.
-    // NOTE: this is a fast-to-compute but *IMPERFECT* identifier for the genesis block.
-    //       to know the actualy genesis block hash, you need to compute the block itself.
+    // NOTE: this is a fast-to-compute but *IMPERFECT* identifier for the genesis block;
+    //       to know the actual genesis block hash, you need to compute the block itself.
     let hash = hasher.hash(&preimage.to_bits_le())?.to_string();
 
     // A closure to load the block.
diff --git a/node/bft/src/helpers/storage.rs b/node/bft/src/helpers/storage.rs
index ddb6bdcd9..cf8bd7152 100644
--- a/node/bft/src/helpers/storage.rs
+++ b/node/bft/src/helpers/storage.rs
@@ -331,7 +331,7 @@ impl<N: Network> Storage<N> {
     }
 
     /// Returns the certificates for the given `round`.
-    /// If the round does not exist in storage, `None` is returned.
+    /// If the round does not exist in storage, an empty set is returned.
     pub fn get_certificates_for_round(&self, round: u64) -> IndexSet<BatchCertificate<N>> {
         // The genesis round does not have batch certificates.
         if round == 0 {
@@ -347,7 +347,7 @@ impl<N: Network> Storage<N> {
     }
 
     /// Returns the certificate IDs for the given `round`.
-    /// If the round does not exist in storage, `None` is returned.
+    /// If the round does not exist in storage, an empty set is returned.
     pub fn get_certificate_ids_for_round(&self, round: u64) -> IndexSet<Field<N>> {
         // The genesis round does not have batch certificates.
         if round == 0 {
@@ -362,7 +362,7 @@ impl<N: Network> Storage<N> {
     }
 
     /// Returns the certificate authors for the given `round`.
-    /// If the round does not exist in storage, `None` is returned.
+    /// If the round does not exist in storage, an empty set is returned.
     pub fn get_certificate_authors_for_round(&self, round: u64) -> HashSet<Address<N>> {
         // The genesis round does not have batch certificates.
         if round == 0 {
