diff --git a/circuit/types/integers/src/shl_checked.rs b/circuit/types/integers/src/shl_checked.rs
index 74faf5cfa..5b556751e 100644
--- a/circuit/types/integers/src/shl_checked.rs
+++ b/circuit/types/integers/src/shl_checked.rs
@@ -69,16 +69,33 @@ impl<E: Environment, I: IntegerType, M: Magnitude> ShlChecked<Integer<E, M>> for
     #[inline]
     fn shl_checked(&self, rhs: &Integer<E, M>) -> Self::Output {
         let two = Self::one() + Self::one();
-        let base = match I::is_signed() {
-            // If `self` is a signed integer and is negative, then negate the base.
-            // Note that if `self.msb()` is 1 then `self` is negative.
-            true => Self::ternary(self.msb(), &(Self::zero() - &two), &two),
-            // Otherwise return two.
-            false => two,
-        };
-
-        // Compute `self` * 2 ^ `rhs`.
-        self.mul_checked(&base.pow_checked(rhs))
+        match I::is_signed() {
+            true => {
+                // Compute 2 ^ `lhs` as unsigned integer of the size I::BITS.
+                // This is necessary to avoid a spurious overflow when `rhs` is I::BITS - 1.
+                // For example, 2i8 ^ 7i8 overflows, however -1i8 << 7i8 ==> -1i8 * 2i8 ^ 7i8 ==> -128i8, which is a valid i8 value.
+                let unsigned_two = two.cast_as_dual();
+                // Note that `pow_checked` is used to enforce that `rhs` < I::BITS.
+                let unsigned_factor = unsigned_two.pow_checked(rhs);
+                // For all values of `rhs` such that `rhs` < I::BITS,
+                //  - if `rhs` == I::BITS - 1, `signed_factor` == I::MIN,
+                //  - otherwise, `signed_factor` is the same as `unsigned_factor`.
+                let signed_factor = Self { bits_le: unsigned_factor.bits_le, phantom: Default::default() };
+
+                // If `signed_factor` is I::MIN, then negate `self` in order to balance the sign of I::MIN.
+                let signed_factor_is_min = &signed_factor.is_equal(&Self::constant(console::Integer::MIN));
+                // - If `signed_factor` is I::MIN,
+                //     - and `self` is zero or I::MIN, then `lhs` is equal to `self`.
+                //     - otherwise, `lhs` is equal to `-self`.
+                // - Otherwise, `lhs` is equal to `self`.
+                let lhs = Self::ternary(signed_factor_is_min, &Self::zero().sub_wrapped(self), self);
+
+                // Compute `lhs` * `factor`, which is equivalent to `lhs` * 2 ^ `rhs`.
+                lhs.mul_checked(&signed_factor)
+            }
+            // Compute `lhs` * 2 ^ `rhs`.
+            false => self.mul_checked(&two.pow_checked(rhs)),
+        }
     }
 }
 
@@ -146,7 +163,7 @@ mod tests {
 
     const ITERATIONS: u64 = 32;
 
-    fn check_shl<I: IntegerType + RefUnwindSafe, M: Magnitude + RefUnwindSafe>(
+    fn check_shl<I: IntegerType + RefUnwindSafe, M: Magnitude + RefUnwindSafe + TryFrom<u64>>(
         name: &str,
         first: console::Integer<<Circuit as Environment>::Network, I>,
         second: console::Integer<<Circuit as Environment>::Network, M>,
@@ -156,14 +173,7 @@ mod tests {
         let a = Integer::<Circuit, I>::new(mode_a, first);
         let b = Integer::<Circuit, M>::new(mode_b, second);
 
-        let two = I::one().wrapping_add(&I::one());
-        let base = if I::is_signed() {
-            if (*first).lt(&I::zero()) { I::zero().wrapping_sub(&two) } else { two }
-        } else {
-            two
-        };
-
-        match base.checked_pow(&second.to_u32().unwrap()).and_then(|result| first.checked_mul(&result)) {
+        match first.checked_shl(&second.to_u32().unwrap()) {
             Some(expected) => Circuit::scope(name, || {
                 let candidate = a.shl_checked(&b);
                 assert_eq!(expected, *candidate.eject_value());
@@ -174,9 +184,12 @@ mod tests {
             }),
             None => match (mode_a, mode_b) {
                 (Mode::Constant, Mode::Constant) => check_operation_halts(&a, &b, Integer::shl_checked),
-                // TODO: This can either halt or be unsatisfied. Make the expectation more precise once output_mode is completely defined.
                 (_, Mode::Constant) => {
-                    if std::panic::catch_unwind(|| Integer::shl_checked(&a, &b)).is_ok() {
+                    // If `second` >= I::BITS, then the invocation to `pow_checked` will halt.
+                    // Otherwise, the invocation to `mul_checked` will not be satisfied.
+                    if *second >= M::try_from(I::BITS).unwrap_or_default() {
+                        check_operation_halts(&a, &b, Integer::shl_checked);
+                    } else {
                         Circuit::scope(name, || {
                             let _candidate = a.shl_checked(&b);
                             // assert_count_fails!(ShlChecked(Integer<I>, Integer<M>) => Integer<I>, &(mode_a, mode_b));
@@ -194,7 +207,10 @@ mod tests {
         Circuit::reset();
     }
 
-    fn run_test<I: IntegerType + RefUnwindSafe, M: Magnitude + RefUnwindSafe>(mode_a: Mode, mode_b: Mode) {
+    fn run_test<I: IntegerType + RefUnwindSafe, M: Magnitude + RefUnwindSafe + TryFrom<u64>>(
+        mode_a: Mode,
+        mode_b: Mode,
+    ) {
         for i in 0..ITERATIONS {
             let first = Uniform::rand(&mut test_rng());
             let second = Uniform::rand(&mut test_rng());
@@ -202,6 +218,10 @@ mod tests {
             let name = format!("Shl: {} << {} {}", mode_a, mode_b, i);
             check_shl::<I, M>(&name, first, second, mode_a, mode_b);
 
+            // Check that shift left by zero is computed correctly.
+            let name = format!("Identity: {} << {} {}", mode_a, mode_b, i);
+            check_shl::<I, M>(&name, first, console::Integer::zero(), mode_a, mode_b);
+
             // Check that shift left by one is computed correctly.
             let name = format!("Double: {} << {} {}", mode_a, mode_b, i);
             check_shl::<I, M>(&name, first, console::Integer::one(), mode_a, mode_b);
@@ -212,8 +232,10 @@ mod tests {
         }
     }
 
-    fn run_exhaustive_test<I: IntegerType + RefUnwindSafe, M: Magnitude + RefUnwindSafe>(mode_a: Mode, mode_b: Mode)
-    where
+    fn run_exhaustive_test<I: IntegerType + RefUnwindSafe, M: Magnitude + RefUnwindSafe + TryFrom<u64>>(
+        mode_a: Mode,
+        mode_b: Mode,
+    ) where
         RangeInclusive<I>: Iterator<Item = I>,
         RangeInclusive<M>: Iterator<Item = M>,
     {
diff --git a/console/network/environment/src/traits/types.rs b/console/network/environment/src/traits/types.rs
index c299e07cd..711244b78 100644
--- a/console/network/environment/src/traits/types.rs
+++ b/console/network/environment/src/traits/types.rs
@@ -312,7 +312,6 @@ pub(super) mod integer_type {
     use num_traits::{
         CheckedNeg,
         CheckedRem,
-        CheckedShl,
         CheckedShr,
         One as NumOne,
         PrimInt,
@@ -403,6 +402,31 @@ pub(super) mod integer_type {
     binary_impl!(CheckedPow, i64, checked_pow, self, v, u32, Option<i64>, i64::checked_pow(*self, *v));
     binary_impl!(CheckedPow, i128, checked_pow, self, v, u32, Option<i128>, i128::checked_pow(*self, *v));
 
+    pub trait CheckedShl: Sized {
+        fn checked_shl(&self, v: &u32) -> Option<Self>;
+    }
+
+    #[rustfmt::skip]
+    binary_impl!(CheckedShl, u8, checked_shl, self, v, u32, Option<u8>, u8::checked_pow(2u8, *v).and_then(|x| u8::checked_mul(*self, x)));
+    #[rustfmt::skip]
+    binary_impl!(CheckedShl, u16, checked_shl, self, v, u32, Option<u16>, u16::checked_pow(2u16, *v).and_then(|x| u16::checked_mul(*self, x)));
+    #[rustfmt::skip]
+    binary_impl!(CheckedShl, u32, checked_shl, self, v, u32, Option<u32>, u32::checked_pow(2u32, *v).and_then(|x| u32::checked_mul(*self, x)));
+    #[rustfmt::skip]
+    binary_impl!(CheckedShl, u64, checked_shl, self, v, u32, Option<u64>, u64::checked_pow(2u64, *v).and_then(|x| u64::checked_mul(*self, x)));
+    #[rustfmt::skip]
+    binary_impl!(CheckedShl, u128, checked_shl, self, v, u32, Option<u128>, u128::checked_pow(2u128, *v).and_then(|x| u128::checked_mul(*self, x)));
+    #[rustfmt::skip]
+    binary_impl!(CheckedShl, i8, checked_shl, self, v, u32, Option<i8>, u8::checked_pow(2u8, *v).and_then(|x| i8::checked_mul(if (x as i8) == i8::MIN { self.wrapping_neg() } else { *self }, x as i8)));
+    #[rustfmt::skip]
+    binary_impl!(CheckedShl, i16, checked_shl, self, v, u32, Option<i16>, u16::checked_pow(2u16, *v).and_then(|x| i16::checked_mul(if (x as i16) == i16::MIN { self.wrapping_neg() } else { *self }, x as i16)));
+    #[rustfmt::skip]
+    binary_impl!(CheckedShl, i32, checked_shl, self, v, u32, Option<i32>, u32::checked_pow(2u32, *v).and_then(|x| i32::checked_mul(if (x as i32) == i32::MIN { self.wrapping_neg() } else { *self }, x as i32)));
+    #[rustfmt::skip]
+    binary_impl!(CheckedShl, i64, checked_shl, self, v, u32, Option<i64>, u64::checked_pow(2u64, *v).and_then(|x| i64::checked_mul(if (x as i64) == i64::MIN { self.wrapping_neg() } else { *self }, x as i64)));
+    #[rustfmt::skip]
+    binary_impl!(CheckedShl, i128, checked_shl, self, v, u32, Option<i128>, u128::checked_pow(2u128, *v).and_then(|x| i128::checked_mul(if (x as i128) == i128::MIN { self.wrapping_neg() } else { *self }, x as i128)));
+
     pub trait WrappingDiv: Sized + Div<Self, Output = Self> {
         fn wrapping_div(&self, v: &Self) -> Self;
     }
diff --git a/console/types/integers/src/bitwise.rs b/console/types/integers/src/bitwise.rs
index 6d76ec48f..5fe079223 100644
--- a/console/types/integers/src/bitwise.rs
+++ b/console/types/integers/src/bitwise.rs
@@ -154,7 +154,7 @@ impl<E: Environment, I: IntegerType, M: Magnitude> Shl<Integer<E, M>> for Intege
     /// Shifts `self` to the left by `n` bits.
     #[inline]
     fn shl(self, n: Integer<E, M>) -> Self::Output {
-        match self.integer.checked_shl(n.integer.to_u32().unwrap()) {
+        match self.integer.checked_shl(&n.integer.to_u32().unwrap()) {
             // Unwrap is safe as we only cast up.
             Some(shifted) => Integer::new(shifted),
             None => E::halt(format!("Failed to shift {self} left by {n} bits")),
@@ -168,7 +168,7 @@ impl<E: Environment, I: IntegerType, M: Magnitude> Shl<&Integer<E, M>> for Integ
     /// Shifts `self` to the left by `n` bits.
     #[inline]
     fn shl(self, n: &Integer<E, M>) -> Self::Output {
-        match self.integer.checked_shl(n.integer.to_u32().unwrap()) {
+        match self.integer.checked_shl(&n.integer.to_u32().unwrap()) {
             // Unwrap is safe as we only cast up.
             Some(shifted) => Integer::new(shifted),
             None => E::halt(format!("Failed to shift {self} left by {n} bits")),
@@ -182,13 +182,11 @@ impl<E: Environment, I: IntegerType, M: Magnitude> ShlChecked<Integer<E, M>> for
     /// Shifts `self` to the left by `n` bits.
     #[inline]
     fn shl_checked(&self, n: &Integer<E, M>) -> Self::Output {
-        let two = Self::one() + Self::one();
-        let base = match I::is_signed() {
-            true => Self::ternary(&self.is_less_than(&Self::zero()), &(Self::zero() - two), &two),
-            false => two,
-        };
-
-        self.mul(&base.pow(n))
+        match self.integer.checked_shl(&n.integer.to_u32().unwrap()) {
+            // Unwrap is safe as we only cast up.
+            Some(shifted) => Integer::new(shifted),
+            None => E::halt(format!("Failed to shift {self} left by {n} bits")),
+        }
     }
 }
 
@@ -206,7 +204,7 @@ impl<E: Environment, I: IntegerType, M: Magnitude> ShlAssign<Integer<E, M>> for
     /// Shifts `self` to the left by `n` bits and assigns the result to `self`.
     #[inline]
     fn shl_assign(&mut self, n: Integer<E, M>) {
-        match self.integer.checked_shl(n.integer.to_u32().unwrap()) {
+        match self.integer.checked_shl(&n.integer.to_u32().unwrap()) {
             // Unwrap is safe as we only cast up.
             Some(shifted) => {
                 self.integer = shifted;
diff --git a/vm/compiler/src/program/instruction/operation/macros.rs b/vm/compiler/src/program/instruction/operation/macros.rs
index 7bd5d367d..b88bdb718 100644
--- a/vm/compiler/src/program/instruction/operation/macros.rs
+++ b/vm/compiler/src/program/instruction/operation/macros.rs
@@ -707,6 +707,8 @@ mod tests {
                         #[allow(unused_mut)]
                         let mut is_shift_operator = false;
                         #[allow(unused_mut)]
+                        let mut shift_exceeds_bitwidth = false;
+                        #[allow(unused_mut)]
                         let mut is_division_operator = false;
                         /// A helper macro to check the conditions.
                         #[allow(unused_macros)]
@@ -726,12 +728,36 @@ mod tests {
                             };
                             ("ensure shifting past boundary halts") => {
                                 match *<$operation as $crate::Operation<_, _, _, 2>>::OPCODE {
-                                    "shl" => should_succeed &= (*a).checked_shl(*b as u32).is_some(),
+                                    // Note that this case needs special handling, since the desired behavior of `checked_shl` deviates from Rust semantics.
+                                    "shl" => {
+                                        // Instantiate two as a constant.
+                                        let zero = $input_a::<CurrentNetwork>::zero();
+                                        let two = $input_a::<CurrentNetwork>::one() + $input_a::one();
+                                        should_succeed &= match *a < *zero {
+                                            // For positive values, perform a checked exponentiation and multiplication.
+                                            false => (*two).checked_pow(*b as u32).and_then(|x| (*a).checked_mul(x)).is_some(),
+                                            true => {
+                                                // Compute a wrapping exponentiation.
+                                                let factor = (*two).wrapping_pow(*b as u32);
+                                                match factor.cmp(&0) {
+                                                    // If `factor` is equal to zero, the exponentiation above overflowed.
+                                                    Ordering::Equal => false,
+                                                    // If `factor` is less than zero, then `b` is equal to 1 - bitwidth(a).
+                                                    // In this case, negate `a` to balance signs.
+                                                    Ordering::Less => (a.wrapping_neg()).checked_mul(factor).is_some(),
+                                                    // If `factor` is greater than zero, then the above exponentiation did not overflow
+                                                    // Then, perform a checked multiplication.
+                                                    Ordering::Greater => (*a).checked_mul(factor).is_some(),
+                                                }
+                                            }
+                                        };
+                                    }
                                     "shr" => should_succeed &= (*a).checked_shr(*b as u32).is_some(),
                                     _ => panic!("Unsupported test enforcement for '{}'", <$operation as $crate::Operation<_, _, _, 2>>::OPCODE),
                                 }
-                                // This indicator is later used in the for-loops below.
+                                // These indicators are later used in the for-loops below.
                                 is_shift_operator |= true;
+                                shift_exceeds_bitwidth |= ((*b as u32) >= $input_a::<CurrentNetwork>::size_in_bits() as u32);
                             };
                             ("ensure divide by zero halts") => {
                                 should_succeed &= (*b) != 0;
@@ -761,8 +787,8 @@ mod tests {
                                 // This indicator bit is used to check that a case panics on halt,
                                 // instead of checking that the circuit is not satisfied (i.e. for `Public|Private && Constant`).
                                 let mut should_panic_on_halt = false;
-                                // If the operation is a shift operator, check if the mode of the RHS is a constant.
-                                should_panic_on_halt |= is_shift_operator && mode_b.is_constant();
+                                // If the operation is a shift operator, check if the mode of the RHS is a constant and if the shift amount exceeds the bitwidth.
+                                should_panic_on_halt |= is_shift_operator && shift_exceeds_bitwidth && mode_b.is_constant();
                                 // If the operation is a division operator, check if the mode of the RHS is a constant.
                                 should_panic_on_halt |= is_division_operator && mode_b.is_constant();
 
