diff --git a/node/bft/ledger-service/src/ledger.rs b/node/bft/ledger-service/src/ledger.rs
index b39763f61..c5a500c50 100644
--- a/node/bft/ledger-service/src/ledger.rs
+++ b/node/bft/ledger-service/src/ledger.rs
@@ -148,7 +148,6 @@ impl<N: Network, C: ConsensusStorage<N>> LedgerService<N> for CoreLedgerService<
     }
 
     /// Returns the committee for the given round.
-    /// If the given round is in the future, then the current committee is returned.
     fn get_committee_for_round(&self, round: u64) -> Result<Committee<N>> {
         // Check if the committee is already in the cache.
         if let Some(committee) = self.committee_cache.lock().get(&round) {
@@ -163,12 +162,12 @@ impl<N: Network, C: ConsensusStorage<N>> LedgerService<N> for CoreLedgerService<
                 // Return the committee.
                 Ok(committee)
             }
-            // Return the current committee if the round is in the future.
+            // Return the current committee if the round is equivalent.
             None => {
                 // Retrieve the current committee.
                 let current_committee = self.current_committee()?;
-                // Return the current committee if the round is in the future.
-                match current_committee.starting_round() <= round {
+                // Return the current committee if the round is equivalent.
+                match current_committee.starting_round() == round {
                     true => Ok(current_committee),
                     false => bail!("No committee found for round {round} in the ledger"),
                 }
@@ -177,7 +176,6 @@ impl<N: Network, C: ConsensusStorage<N>> LedgerService<N> for CoreLedgerService<
     }
 
     /// Returns the committee lookback for the given round.
-    /// If the committee lookback round is in the future, then the current committee is returned.
     fn get_committee_lookback_for_round(&self, round: u64) -> Result<Committee<N>> {
         // Get the round number for the previous committee. Note, we subtract 2 from odd rounds,
         // because committees are updated in even rounds.
diff --git a/node/bft/ledger-service/src/mock.rs b/node/bft/ledger-service/src/mock.rs
index 681c8cd89..7a182446b 100644
--- a/node/bft/ledger-service/src/mock.rs
+++ b/node/bft/ledger-service/src/mock.rs
@@ -147,7 +147,6 @@ impl<N: Network> LedgerService<N> for MockLedgerService<N> {
     }
 
     /// Returns the committee for the given round.
-    /// If the given round is in the future, then the current committee is returned.
     fn get_committee_for_round(&self, _round: u64) -> Result<Committee<N>> {
         Ok(self.committee.clone())
     }
diff --git a/node/bft/ledger-service/src/prover.rs b/node/bft/ledger-service/src/prover.rs
index 2d3839a19..0038e7403 100644
--- a/node/bft/ledger-service/src/prover.rs
+++ b/node/bft/ledger-service/src/prover.rs
@@ -118,13 +118,11 @@ impl<N: Network> LedgerService<N> for ProverLedgerService<N> {
     }
 
     /// Returns the committee for the given round.
-    /// If the given round is in the future, then the current committee is returned.
     fn get_committee_for_round(&self, round: u64) -> Result<Committee<N>> {
         bail!("Committee for round {round} does not exist in prover")
     }
 
     /// Returns the committee lookback for the given round.
-    /// If the committee lookback round is in the future, then the current committee is returned.
     fn get_committee_lookback_for_round(&self, round: u64) -> Result<Committee<N>> {
         bail!("Previous committee for round {round} does not exist in prover")
     }
diff --git a/node/bft/ledger-service/src/traits.rs b/node/bft/ledger-service/src/traits.rs
index 2bbad3387..d8477c3a2 100644
--- a/node/bft/ledger-service/src/traits.rs
+++ b/node/bft/ledger-service/src/traits.rs
@@ -74,11 +74,9 @@ pub trait LedgerService<N: Network>: Debug + Send + Sync {
     fn current_committee(&self) -> Result<Committee<N>>;
 
     /// Returns the committee for the given round.
-    /// If the given round is in the future, then the current committee is returned.
     fn get_committee_for_round(&self, round: u64) -> Result<Committee<N>>;
 
     /// Returns the committee lookback for the given round.
-    /// If the committee lookback round is in the future, then the current committee is returned.
     fn get_committee_lookback_for_round(&self, round: u64) -> Result<Committee<N>>;
 
     /// Returns `true` if the ledger contains the given certificate ID.
diff --git a/node/bft/ledger-service/src/translucent.rs b/node/bft/ledger-service/src/translucent.rs
index 044cd95ee..8a5baf642 100644
--- a/node/bft/ledger-service/src/translucent.rs
+++ b/node/bft/ledger-service/src/translucent.rs
@@ -129,7 +129,6 @@ impl<N: Network, C: ConsensusStorage<N>> LedgerService<N> for TranslucentLedgerS
     }
 
     /// Returns the committee for the given round.
-    /// If the given round is in the future, then the current committee is returned.
     fn get_committee_for_round(&self, round: u64) -> Result<Committee<N>> {
         self.inner.get_committee_for_round(round)
     }
diff --git a/node/bft/src/sync/mod.rs b/node/bft/src/sync/mod.rs
index 545c477b2..e5dc84731 100644
--- a/node/bft/src/sync/mod.rs
+++ b/node/bft/src/sync/mod.rs
@@ -102,6 +102,11 @@ impl<N: Network> Sync<N> {
         // Start the block sync loop.
         let self_ = self.clone();
         self.handles.lock().push(tokio::spawn(async move {
+            // Sleep briefly to allow an initial primary ping to come in prior to entering the loop.
+            // Ideally, a node does not consider itself synced when it has not received
+            // any block locators from peer. However, in the initial bootup of validators,
+            // this needs to happen, so we use this additional sleep as a grace period.
+            tokio::time::sleep(Duration::from_millis(PRIMARY_PING_IN_MS)).await;
             loop {
                 // Sleep briefly to avoid triggering spam detection.
                 tokio::time::sleep(Duration::from_millis(PRIMARY_PING_IN_MS)).await;
