diff --git a/Cargo.lock b/Cargo.lock
index 36247275d..180b986aa 100644
--- a/Cargo.lock
+++ b/Cargo.lock
@@ -2975,6 +2975,7 @@ dependencies = [
  "aleo-std",
  "anstyle",
  "anyhow",
+ "bincode",
  "clap",
  "colored",
  "crossterm 0.27.0",
@@ -2991,6 +2992,7 @@ dependencies = [
  "snarkos-account",
  "snarkos-display",
  "snarkos-node",
+ "snarkos-node-cdn",
  "snarkos-node-rest",
  "snarkvm",
  "sys-info",
diff --git a/cli/Cargo.toml b/cli/Cargo.toml
index e062f7d95..094a221b2 100644
--- a/cli/Cargo.toml
+++ b/cli/Cargo.toml
@@ -26,6 +26,9 @@ version = "1"
 [dependencies.anyhow]
 version = "1.0.75"
 
+[dependencies.bincode]
+version = "1.0"
+
 [dependencies.clap]
 version = "4.4"
 features = [ "derive", "color", "unstable-styles" ]
@@ -79,6 +82,10 @@ version = "=2.1.7"
 path = "../node"
 version = "=2.1.7"
 
+[dependencies.snarkos-node-cdn]
+path = "../node/cdn"
+version = "=2.1.7"
+
 [dependencies.snarkos-node-rest]
 path = "../node/rest"
 version = "=2.1.7"
diff --git a/cli/src/commands/developer/scan.rs b/cli/src/commands/developer/scan.rs
index 942266625..d65fb103f 100644
--- a/cli/src/commands/developer/scan.rs
+++ b/cli/src/commands/developer/scan.rs
@@ -12,17 +12,23 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+#![allow(clippy::type_complexity)]
+
 use super::CurrentNetwork;
 
 use snarkvm::prelude::{block::Block, Ciphertext, Field, Network, Plaintext, PrivateKey, Record, ViewKey};
 
 use anyhow::{bail, ensure, Result};
 use clap::Parser;
+use parking_lot::RwLock;
 use std::{
     io::{stdout, Write},
     str::FromStr,
+    sync::Arc,
 };
 
+const MAX_BLOCK_RANGE: u32 = 50;
+
 /// Scan the snarkOS node for records.
 #[derive(Debug, Parser)]
 pub struct Scan {
@@ -49,6 +55,10 @@ pub struct Scan {
     /// The endpoint to scan blocks from.
     #[clap(long)]
     endpoint: String,
+
+    /// Enables the node to prefetch initial blocks from a CDN
+    #[clap(default_value = "https://s3.us-west-1.amazonaws.com/testnet3.blocks/phase3", long = "cdn")]
+    cdn: String,
 }
 
 impl Scan {
@@ -60,7 +70,7 @@ impl Scan {
         let (start_height, end_height) = self.parse_block_range()?;
 
         // Fetch the records from the network.
-        let records = Self::fetch_records(private_key, &view_key, &self.endpoint, start_height, end_height)?;
+        let records = Self::fetch_records(private_key, &view_key, &self.cdn, &self.endpoint, start_height, end_height)?;
 
         // Output the decrypted records associated with the view key.
         if records.is_empty() {
@@ -142,6 +152,7 @@ impl Scan {
     fn fetch_records(
         private_key: Option<PrivateKey<CurrentNetwork>>,
         view_key: &ViewKey<CurrentNetwork>,
+        cdn: &str,
         endpoint: &str,
         start_height: u32,
         end_height: u32,
@@ -154,15 +165,30 @@ impl Scan {
         // Derive the x-coordinate of the address corresponding to the given view key.
         let address_x_coordinate = view_key.to_address().to_x_coordinate();
 
-        const MAX_BLOCK_RANGE: u32 = 50;
-
-        let mut records = Vec::new();
+        // Initialize a vector to store the records.
+        let records = Arc::new(RwLock::new(Vec::new()));
 
         // Calculate the number of blocks to scan.
         let total_blocks = end_height.saturating_sub(start_height);
 
-        // Scan the endpoint starting from the start height
-        let mut request_start = start_height;
+        // Log the initial progress.
+        print!("\rScanning {total_blocks} blocks for records (0% complete)...");
+        stdout().flush()?;
+
+        // Scan the CDN first for records.
+        Self::scan_from_cdn(
+            start_height,
+            end_height,
+            cdn.to_string(),
+            endpoint.to_string(),
+            private_key,
+            *view_key,
+            address_x_coordinate,
+            records.clone(),
+        )?;
+
+        // Scan the endpoint for the remaining blocks.
+        let mut request_start = end_height.saturating_sub(start_height % MAX_BLOCK_RANGE);
         while request_start <= end_height {
             // Log the progress.
             let percentage_complete = request_start.saturating_sub(start_height) as f64 * 100.0 / total_blocks as f64;
@@ -175,23 +201,12 @@ impl Scan {
 
             // Establish the endpoint.
             let blocks_endpoint = format!("{endpoint}/testnet3/blocks?start={request_start}&end={request_end}");
-
             // Fetch blocks
             let blocks: Vec<Block<CurrentNetwork>> = ureq::get(&blocks_endpoint).call()?.into_json()?;
 
             // Scan the blocks for owned records.
             for block in &blocks {
-                for (commitment, ciphertext_record) in block.records() {
-                    // Check if the record is owned by the given view key.
-                    if ciphertext_record.is_owner_with_address_x_coordinate(view_key, &address_x_coordinate) {
-                        // Decrypt and optionally filter the records.
-                        if let Some(record) =
-                            Self::decrypt_record(private_key, view_key, endpoint, *commitment, ciphertext_record)?
-                        {
-                            records.push(record);
-                        }
-                    }
-                }
+                Self::scan_block(block, endpoint, private_key, view_key, &address_x_coordinate, records.clone())?;
             }
 
             request_start = request_start.saturating_add(num_blocks_to_request);
@@ -201,7 +216,79 @@ impl Scan {
         println!("\rScanning {total_blocks} blocks for records (100% complete)...   \n");
         stdout().flush()?;
 
-        Ok(records)
+        let result = records.read().clone();
+        Ok(result)
+    }
+
+    /// Scan the blocks from the CDN.
+    #[allow(clippy::too_many_arguments)]
+    fn scan_from_cdn(
+        start_height: u32,
+        end_height: u32,
+        cdn: String,
+        endpoint: String,
+        private_key: Option<PrivateKey<CurrentNetwork>>,
+        view_key: ViewKey<CurrentNetwork>,
+        address_x_coordinate: Field<CurrentNetwork>,
+        records: Arc<RwLock<Vec<Record<CurrentNetwork, Plaintext<CurrentNetwork>>>>>,
+    ) -> Result<()> {
+        // Calculate the number of blocks to scan.
+        let total_blocks = end_height.saturating_sub(start_height);
+
+        // Get the start_height with
+        let cdn_request_start = start_height.saturating_sub(start_height % MAX_BLOCK_RANGE);
+        let cdn_request_end = end_height.saturating_sub(start_height % MAX_BLOCK_RANGE);
+
+        // Construct the runtime.
+        let rt = tokio::runtime::Runtime::new()?;
+
+        // Scan the blocks via the CDN.
+        rt.block_on(async move {
+            let _ = snarkos_node_cdn::load_blocks(&cdn, cdn_request_start, Some(cdn_request_end), move |block| {
+                // Check if the block is within the requested range.
+                if block.height() < start_height || block.height() > end_height {
+                    return Ok(());
+                }
+
+                // Log the progress.
+                let percentage_complete =
+                    block.height().saturating_sub(start_height) as f64 * 100.0 / total_blocks as f64;
+                print!("\rScanning {total_blocks} blocks for records ({percentage_complete:.2}% complete)...");
+                stdout().flush()?;
+
+                // Scan the block for records.
+                Self::scan_block(&block, &endpoint, private_key, &view_key, &address_x_coordinate, records.clone())?;
+
+                Ok(())
+            })
+            .await;
+        });
+
+        Ok(())
+    }
+
+    /// Scan a block for owned records.
+    fn scan_block(
+        block: &Block<CurrentNetwork>,
+        endpoint: &str,
+        private_key: Option<PrivateKey<CurrentNetwork>>,
+        view_key: &ViewKey<CurrentNetwork>,
+        address_x_coordinate: &Field<CurrentNetwork>,
+        records: Arc<RwLock<Vec<Record<CurrentNetwork, Plaintext<CurrentNetwork>>>>>,
+    ) -> Result<()> {
+        for (commitment, ciphertext_record) in block.records() {
+            // Check if the record is owned by the given view key.
+            if ciphertext_record.is_owner_with_address_x_coordinate(view_key, address_x_coordinate) {
+                // Decrypt and optionally filter the records.
+                if let Some(record) =
+                    Self::decrypt_record(private_key, view_key, endpoint, *commitment, ciphertext_record)?
+                {
+                    records.write().push(record);
+                }
+            }
+        }
+
+        Ok(())
     }
 
     /// Decrypts the ciphertext record and filters spend record if a private key was provided.
diff --git a/node/narwhal/events/Cargo.toml b/node/narwhal/events/Cargo.toml
index 6e4ae5973..27958240a 100644
--- a/node/narwhal/events/Cargo.toml
+++ b/node/narwhal/events/Cargo.toml
@@ -52,6 +52,9 @@ version = "0.1"
 [dev-dependencies.proptest]
 version = "1.3.1"
 
+[dev-dependencies.snarkvm-ledger-test-helpers]
+version = "0.15.4"
+
 [dev-dependencies.snarkvm]
 workspace = true
 features = [ "test-helpers" ]
diff --git a/node/narwhal/events/src/block_response.rs b/node/narwhal/events/src/block_response.rs
index e13744676..2a5667ef1 100644
--- a/node/narwhal/events/src/block_response.rs
+++ b/node/narwhal/events/src/block_response.rs
@@ -140,20 +140,19 @@ pub mod prop_tests {
     use bytes::{Buf, BufMut, BytesMut};
     use proptest::{
         collection::vec,
-        prelude::{BoxedStrategy, Strategy},
-        strategy::Just,
+        prelude::{any, BoxedStrategy, Strategy},
     };
     use snarkvm::{
-        prelude::{block::Block, narwhal::Data, Network},
-        utilities::{FromBytes, ToBytes},
+        prelude::{block::Block, narwhal::Data},
+        utilities::{FromBytes, TestRng, ToBytes},
     };
+    use snarkvm_ledger_test_helpers::sample_genesis_block;
     use test_strategy::proptest;
 
     type CurrentNetwork = snarkvm::prelude::Testnet3;
 
-    // TODO: create a random block, not always the same one.
     pub fn any_block() -> BoxedStrategy<Block<CurrentNetwork>> {
-        Just(Block::<CurrentNetwork>::from_bytes_le(CurrentNetwork::genesis_bytes()).unwrap()).boxed()
+        any::<u64>().prop_map(|seed| sample_genesis_block(&mut TestRng::fixed(seed))).boxed()
     }
 
     pub fn any_data_blocks() -> BoxedStrategy<DataBlocks<CurrentNetwork>> {
diff --git a/run-client.sh b/run-client.sh
index d5fcf117f..ba2301f16 100755
--- a/run-client.sh
+++ b/run-client.sh
@@ -3,7 +3,7 @@
 # USAGE examples: 
   # CLI :  ./run-client.sh
 
-COMMAND='cargo run --release -- start --nodisplay'
+COMMAND='cargo run --release -- start --nodisplay --client'
 
 for word in $*;
 do
