diff --git a/.devnet/reinstall.sh b/.devnet/reinstall.sh
index a1b6e606a..3db5b4f53 100755
--- a/.devnet/reinstall.sh
+++ b/.devnet/reinstall.sh
@@ -31,6 +31,7 @@ run_installation() {
 #      rm -rf \$WORKSPACE
 #      git clone https://github.com/AleoHQ/snarkOS.git \$WORKSPACE
       cd \$WORKSPACE
+      git pull # If we are switching branches, this will find the new branch
       git checkout $BRANCH  # Checkout the specified branch
       git pull origin $BRANCH
     else
diff --git a/Cargo.lock b/Cargo.lock
index 7ff14c3d4..b9ecee995 100644
--- a/Cargo.lock
+++ b/Cargo.lock
@@ -2844,9 +2844,9 @@ dependencies = [
 
 [[package]]
 name = "serde_json"
-version = "1.0.107"
+version = "1.0.108"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "6b420ce6e3d8bd882e9b243c6eed35dbc9a6110c9769e74b584e0d68d1f20c65"
+checksum = "3d1c7e3eac408d115102c4c24ad393e0821bb3a5df4d506a80f85f7a742a526b"
 dependencies = [
  "indexmap 2.0.2",
  "itoa",
@@ -3369,8 +3369,7 @@ dependencies = [
 [[package]]
 name = "snarkvm"
 version = "0.16.6"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "12e33a17dd727deeb0309daca9b3d48d2c103dcaeb55ab87d63c534a03cc2ea4"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
  "anstyle",
  "anyhow",
@@ -3400,8 +3399,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-algorithms"
 version = "0.16.6"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "2a3cd0ead41ef019e92b5bbcfa8226572291afb67fdee503239f8e9f7309fbb4"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
  "aleo-std",
  "anyhow",
@@ -3431,8 +3429,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-circuit"
 version = "0.16.6"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "2077a5e63c449fe3983050292f5e53f8865186b97f0543f1e7b1e6af8cdfac82"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
  "snarkvm-circuit-account",
  "snarkvm-circuit-algorithms",
@@ -3446,8 +3443,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-circuit-account"
 version = "0.16.6"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "573576d631120789f288a354b4a1756f1a71d9b3108ae5d5c35a74e77772e546"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
  "snarkvm-circuit-algorithms",
  "snarkvm-circuit-network",
@@ -3458,8 +3454,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-circuit-algorithms"
 version = "0.16.6"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "c99806eeadf6895c031cb180362c834829e6acc17d2af9b3a0d52a61f0d02485"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
  "snarkvm-circuit-types",
  "snarkvm-console-algorithms",
@@ -3469,8 +3464,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-circuit-collections"
 version = "0.16.6"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "8e77b30e1e7996a403029bdab6f3cc29c0f25c3b1e7fcfc3b3aa15535367d3b0"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
  "snarkvm-circuit-algorithms",
  "snarkvm-circuit-types",
@@ -3480,8 +3474,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-circuit-environment"
 version = "0.16.6"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "223cc48dd71f4bfd0057e2495f20e29a37e56927292747b34253d77dba9242fa"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
  "indexmap 2.0.2",
  "itertools 0.11.0",
@@ -3499,14 +3492,12 @@ dependencies = [
 [[package]]
 name = "snarkvm-circuit-environment-witness"
 version = "0.16.6"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "f71c2eeab0b10c64a495bee0afbc1abf27c6f6c3b43ec7ee56c828cb052a946b"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 
 [[package]]
 name = "snarkvm-circuit-network"
 version = "0.16.6"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "b8771980ab198c49d46c7317e0745fcb5230094b612d9edf837057e1891c60af"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
  "snarkvm-circuit-algorithms",
  "snarkvm-circuit-collections",
@@ -3517,8 +3508,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-circuit-program"
 version = "0.16.6"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "c6f5cf618d21e4086817e2c6877096612e5507cb8174886ed91399d44eecc584"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
  "paste",
  "snarkvm-circuit-account",
@@ -3533,8 +3523,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-circuit-types"
 version = "0.16.6"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "6eb78d6fa5fc1d1f1deed7cca1ef3b77dceedfe764a30d54d111769d8dd8806d"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
  "snarkvm-circuit-environment",
  "snarkvm-circuit-types-address",
@@ -3549,8 +3538,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-circuit-types-address"
 version = "0.16.6"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "902177b15907ff6599d82fb8d432aa7cbbf8d84ab8dcad0598b3d73c6202f1eb"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
  "snarkvm-circuit-environment",
  "snarkvm-circuit-types-boolean",
@@ -3563,8 +3551,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-circuit-types-boolean"
 version = "0.16.6"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "e58f2f52369b70af7dfd316bc89485aa0d9561b9824cdea89ea24acec40c644f"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
  "snarkvm-circuit-environment",
  "snarkvm-console-types-boolean",
@@ -3573,8 +3560,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-circuit-types-field"
 version = "0.16.6"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "74eaae0b7624bd73dde9def62ffdb6ee30aaf64f211c8d2155fca3bc441bbe51"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
  "snarkvm-circuit-environment",
  "snarkvm-circuit-types-boolean",
@@ -3584,8 +3570,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-circuit-types-group"
 version = "0.16.6"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "de9098c2f9ff6714b5628f1210dab2f27019519213d4f5d6329b640875433b89"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
  "snarkvm-circuit-environment",
  "snarkvm-circuit-types-boolean",
@@ -3597,8 +3582,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-circuit-types-integers"
 version = "0.16.6"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "2f210df5327b3c4f77fd854dc0093d637f406cecf7ab055086b2c4344a85dd80"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
  "snarkvm-circuit-environment",
  "snarkvm-circuit-types-boolean",
@@ -3610,8 +3594,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-circuit-types-scalar"
 version = "0.16.6"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "096791aca4d8deedc27f8f7c960eb9a235d3376d0920fab5fa8f50c53d6c0210"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
  "snarkvm-circuit-environment",
  "snarkvm-circuit-types-boolean",
@@ -3622,8 +3605,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-circuit-types-string"
 version = "0.16.6"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "e58efa0189da37ae6e2b2d7680a31fe7ea89d8ffa19452f8e2710ada9e25bacc"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
  "snarkvm-circuit-environment",
  "snarkvm-circuit-types-boolean",
@@ -3635,8 +3617,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-console"
 version = "0.16.6"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "b7113a092a6bcf026823e93c17797c30c5cb94ded77becfed6af9c7840113d87"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
  "snarkvm-console-account",
  "snarkvm-console-algorithms",
@@ -3649,8 +3630,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-console-account"
 version = "0.16.6"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "673448b3679feb3c70a533fe04a75567a39fb8209c14d48e092cce11cfbddcbf"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
  "bs58",
  "snarkvm-console-network",
@@ -3661,8 +3641,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-console-algorithms"
 version = "0.16.6"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "2f47b413a30d3f7117b8ee8e2be935820606fc3c59725588b301976824ae2b66"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
  "blake2s_simd",
  "smallvec",
@@ -3675,8 +3654,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-console-collections"
 version = "0.16.6"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "ac544da5cf5c11fde6774de793649528ce73c92964b6d102f94662faae156b26"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
  "aleo-std",
  "rayon",
@@ -3687,8 +3665,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-console-network"
 version = "0.16.6"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "b175dc188dd13ab30f381ba3041e2b8a1623d34027b2c416620899e9ec42d097"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
  "anyhow",
  "indexmap 2.0.2",
@@ -3711,8 +3688,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-console-network-environment"
 version = "0.16.6"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "b870845575c2683018590279742f17e2ea604fb2f0065e77b787f920b90b771f"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
  "anyhow",
  "bech32",
@@ -3730,8 +3706,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-console-program"
 version = "0.16.6"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "1a5c9316868b05568df87099034d6b49cd7214d6a57f13de169bfe5be86ba04d"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
  "enum_index",
  "enum_index_derive",
@@ -3752,8 +3727,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-console-types"
 version = "0.16.6"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "0ad39ef257f8f313a6f3e45eb74fbde35d85571810e31594e389b091270a7ebe"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
  "snarkvm-console-network-environment",
  "snarkvm-console-types-address",
@@ -3768,8 +3742,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-console-types-address"
 version = "0.16.6"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "4185b4743912dbd46971e15b3d9af78a9a897a9bbc78f8735380f97bd6bc4aa9"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
  "snarkvm-console-network-environment",
  "snarkvm-console-types-boolean",
@@ -3780,8 +3753,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-console-types-boolean"
 version = "0.16.6"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "19e802cbc98655d27cebe26bfd32c9a5de809f4846c6a87f13dfd430117e9b34"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
  "snarkvm-console-network-environment",
 ]
@@ -3789,8 +3761,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-console-types-field"
 version = "0.16.6"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "3d549071da6521f5fd852ce9635efd86973b7f039713a4264244e006ea2c613f"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
  "snarkvm-console-network-environment",
  "snarkvm-console-types-boolean",
@@ -3800,8 +3771,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-console-types-group"
 version = "0.16.6"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "f3215e7a711aecf339e84657d0d3c0cc504cd1e2194bbfd1fcbf5fac5670dcce"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
  "snarkvm-console-network-environment",
  "snarkvm-console-types-boolean",
@@ -3812,8 +3782,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-console-types-integers"
 version = "0.16.6"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "0bd543873da9ba4964878ef859cca25e4c590eb23c5caee21c231532e1f42056"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
  "snarkvm-console-network-environment",
  "snarkvm-console-types-boolean",
@@ -3824,8 +3793,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-console-types-scalar"
 version = "0.16.6"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "7fa556a8563fb0442344b7f3a123a53dc9bd74ab63f2d2a3c689cf9b8798bc06"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
  "snarkvm-console-network-environment",
  "snarkvm-console-types-boolean",
@@ -3836,8 +3804,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-console-types-string"
 version = "0.16.6"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "afb9c2151d6a99f760bee50df25aa6b32e5bbed9f9d677dc21d262842932cdcb"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
  "snarkvm-console-network-environment",
  "snarkvm-console-types-boolean",
@@ -3848,8 +3815,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-curves"
 version = "0.16.6"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "c0676eba867fd9c0a77919506e382bc4afe96f2699668f0e1b58386a688886ff"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
  "rand",
  "rayon",
@@ -3863,8 +3829,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-fields"
 version = "0.16.6"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "770735a0ba7fb573bf4d7a5e36c3babdaecb3f25ca1dae79b348d76c9df4161c"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
  "aleo-std",
  "anyhow",
@@ -3882,8 +3847,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-ledger"
 version = "0.16.6"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "044a399c9cfe43b121dad85adaa7137a1ed18d50096d640f16a5a0c0221298c0"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
  "aleo-std",
  "anyhow",
@@ -3908,8 +3872,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-ledger-authority"
 version = "0.16.6"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "4e3263f84bb48f50ef7422eaacab347c931b46242676def1cf36bf8fd5a23e54"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
  "anyhow",
  "rand",
@@ -3921,8 +3884,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-ledger-block"
 version = "0.16.6"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "60d35da2edc62a32b0d522d3283fddd6556326c9fd6a378456fe1334f13f42fc"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
  "indexmap 2.0.2",
  "rayon",
@@ -3940,8 +3902,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-ledger-coinbase"
 version = "0.16.6"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "0db2eb7c2e2723241ac9302aaefe4c852e13d6a00cbba8feb781c48ecb801ae0"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
  "aleo-std",
  "anyhow",
@@ -3961,8 +3922,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-ledger-committee"
 version = "0.16.6"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "b8d0178ee395f0c6816fe728cf3b04c28ff1654a00ed5b21bceabfc975e15c53"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
  "anyhow",
  "indexmap 2.0.2",
@@ -3978,8 +3938,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-ledger-narwhal"
 version = "0.16.6"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "9b423625396052f6dde8c2d6bd98ffadb39d26d7db72f70a44dc807af438bde0"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
  "snarkvm-ledger-narwhal-batch-certificate",
  "snarkvm-ledger-narwhal-batch-header",
@@ -3992,22 +3951,19 @@ dependencies = [
 [[package]]
 name = "snarkvm-ledger-narwhal-batch-certificate"
 version = "0.16.6"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "d5435688eb5db7be8f18501fc76dd52e77d51eff25ce934925e3b881e0966431"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
  "indexmap 2.0.2",
  "serde_json",
  "snarkvm-console",
  "snarkvm-ledger-narwhal-batch-header",
  "snarkvm-ledger-narwhal-transmission-id",
- "time",
 ]
 
 [[package]]
 name = "snarkvm-ledger-narwhal-batch-header"
 version = "0.16.6"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "648861a8c25712b955f3d5946db958f5c740596962732087c34fbe465b1e78b1"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
  "indexmap 2.0.2",
  "serde_json",
@@ -4019,8 +3975,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-ledger-narwhal-data"
 version = "0.16.6"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "3425151ba63e48b5fc1b5474f0d5634e2408c105ecf5e7e359555d8493cb49f1"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
  "bytes",
  "serde_json",
@@ -4031,8 +3986,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-ledger-narwhal-subdag"
 version = "0.16.6"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "1a334dadac03e278117a233ed4d036c5239d25a0d87a90c6f73638620210b3b9"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
  "indexmap 2.0.2",
  "rayon",
@@ -4045,8 +3999,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-ledger-narwhal-transmission"
 version = "0.16.6"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "a4e30079402ac160228545ce57a2ac20b44942b67dc154213243984ea0a6ae4e"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
  "bytes",
  "serde_json",
@@ -4059,8 +4012,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-ledger-narwhal-transmission-id"
 version = "0.16.6"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "c9128f7a2f60853f9ed88df2ced9c288cb9c55e4a6fdfc70414ac98c136ee793"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
  "snarkvm-console",
  "snarkvm-ledger-coinbase",
@@ -4069,8 +4021,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-ledger-query"
 version = "0.16.6"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "a8e2236c4a459d33d793b0214f2f36e86be1402b50b8764925792292d2655e2a"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
  "async-trait",
  "reqwest",
@@ -4083,8 +4034,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-ledger-store"
 version = "0.16.6"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "0aadc60f9aa02273689bfa8d3dc758a2b1243e78ded3a98ee32b8785f5ab0927"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
  "aleo-std",
  "anyhow",
@@ -4109,8 +4059,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-ledger-test-helpers"
 version = "0.16.6"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "dabd347f51b437bba202f5ff75ebe60d2bf024b737db72b22f992c7bdca1eeee"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
  "once_cell",
  "snarkvm-circuit",
@@ -4125,8 +4074,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-parameters"
 version = "0.16.6"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "3f6fa068af0b84484d9d45973973c81797b23af7aaf2a9e94d37fa6aabc95899"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
  "aleo-std",
  "anyhow",
@@ -4150,8 +4098,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-synthesizer"
 version = "0.16.6"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "d5c9b356d418f942bc84e758d8feb5422862f3ff4f75d7146cb015c3336cd7ae"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
  "aleo-std",
  "anyhow",
@@ -4176,8 +4123,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-synthesizer-process"
 version = "0.16.6"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "9594461d2c1e230329234bdd02f32227f5386885391e6dffb5ec708cf4f9bd9f"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
  "aleo-std",
  "colored",
@@ -4199,8 +4145,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-synthesizer-program"
 version = "0.16.6"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "bf10d9b11062d76ba99a9313900395378543d6545a3cc1025afb70f9f6890e89"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
  "indexmap 2.0.2",
  "paste",
@@ -4214,8 +4159,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-synthesizer-snark"
 version = "0.16.6"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "7f9cf76c1a25a1ce0a64fb587c8aa5e5949a6166c03a4e4cc079380f07ea6512"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
  "bincode",
  "once_cell",
@@ -4228,8 +4172,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-utilities"
 version = "0.16.6"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "7e4808f52a6f0c2a13df5b65706dda2ca64ca6c0fb785cf48dc2d246853d724a"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
  "aleo-std",
  "anyhow",
@@ -4250,8 +4193,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-utilities-derives"
 version = "0.16.6"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "b89d440d3c45df5de552f1e4f84549204f67a39104e638f5552fa1572cc7901a"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=2e245e3#2e245e30b9b53c1dd6ca27aa9f8efca209815bec"
 dependencies = [
  "proc-macro2",
  "quote 1.0.33",
diff --git a/Cargo.toml b/Cargo.toml
index 05568206e..16023389c 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -40,7 +40,9 @@ members = [
 ]
 
 [workspace.dependencies.snarkvm]
-version = "=0.16.6"
+#version = "=0.16.6"
+git = "https://github.com/AleoHQ/snarkVM.git"
+rev = "2e245e3"
 features = [ "circuit", "console", "rocks" ]
 
 [[bin]]
diff --git a/node/bft/events/src/batch_signature.rs b/node/bft/events/src/batch_signature.rs
index 91905a3a3..0785c96a7 100644
--- a/node/bft/events/src/batch_signature.rs
+++ b/node/bft/events/src/batch_signature.rs
@@ -18,13 +18,12 @@ use super::*;
 pub struct BatchSignature<N: Network> {
     pub batch_id: Field<N>,
     pub signature: Signature<N>,
-    pub timestamp: i64,
 }
 
 impl<N: Network> BatchSignature<N> {
     /// Initializes a new batch signature event.
-    pub fn new(batch_id: Field<N>, signature: Signature<N>, timestamp: i64) -> Self {
-        Self { batch_id, signature, timestamp }
+    pub fn new(batch_id: Field<N>, signature: Signature<N>) -> Self {
+        Self { batch_id, signature }
     }
 }
 
@@ -40,7 +39,6 @@ impl<N: Network> ToBytes for BatchSignature<N> {
     fn write_le<W: Write>(&self, mut writer: W) -> IoResult<()> {
         self.batch_id.write_le(&mut writer)?;
         self.signature.write_le(&mut writer)?;
-        self.timestamp.write_le(&mut writer)?;
         Ok(())
     }
 }
@@ -49,9 +47,8 @@ impl<N: Network> FromBytes for BatchSignature<N> {
     fn read_le<R: Read>(mut reader: R) -> IoResult<Self> {
         let batch_id = Field::read_le(&mut reader)?;
         let signature = Signature::read_le(&mut reader)?;
-        let timestamp = i64::read_le(&mut reader)?;
 
-        Ok(Self { batch_id, signature, timestamp })
+        Ok(Self { batch_id, signature })
     }
 }
 
@@ -60,22 +57,19 @@ pub mod prop_tests {
     use crate::{
         certificate_request::prop_tests::any_field,
         challenge_response::prop_tests::any_signature,
-        prop_tests::now,
         BatchSignature,
     };
     use snarkvm::console::prelude::{FromBytes, ToBytes};
 
     use bytes::{Buf, BufMut, BytesMut};
-    use proptest::prelude::{BoxedStrategy, Just, Strategy};
+    use proptest::prelude::{BoxedStrategy, Strategy};
     use test_strategy::proptest;
 
     type CurrentNetwork = snarkvm::prelude::Testnet3;
 
     pub fn any_batch_signature() -> BoxedStrategy<BatchSignature<CurrentNetwork>> {
-        (any_field(), any_signature(), Just(now()), -10..10i64)
-            .prop_map(|(certificate_id, signature, timestamp, drift)| {
-                BatchSignature::new(certificate_id, signature, timestamp + drift)
-            })
+        (any_field(), any_signature())
+            .prop_map(|(certificate_id, signature)| BatchSignature::new(certificate_id, signature))
             .boxed()
     }
 
diff --git a/node/bft/events/src/certificate_response.rs b/node/bft/events/src/certificate_response.rs
index 1b8098faa..20977dce1 100644
--- a/node/bft/events/src/certificate_response.rs
+++ b/node/bft/events/src/certificate_response.rs
@@ -63,7 +63,6 @@ pub mod prop_tests {
         console::{
             account::Signature,
             prelude::{FromBytes, ToBytes},
-            types::Field,
         },
         ledger::{
             committee::prop_tests::{CommitteeContext, ValidatorSet},
@@ -73,7 +72,7 @@ pub mod prop_tests {
     };
 
     use bytes::{Buf, BufMut, BytesMut};
-    use indexmap::IndexMap;
+    use indexmap::IndexSet;
     use proptest::{
         collection::vec,
         prelude::{any, BoxedStrategy, Just, Strategy},
@@ -102,7 +101,7 @@ pub mod prop_tests {
             .prop_map(|(committee, batch_header)| {
                 let CommitteeContext(_, validator_set) = committee;
                 let mut rng = TestRng::default();
-                BatchCertificate::new(batch_header.clone(), sign_batch_header(&validator_set, &batch_header, &mut rng))
+                BatchCertificate::from(batch_header.clone(), sign_batch_header(&validator_set, &batch_header, &mut rng))
                     .unwrap()
             })
             .boxed()
@@ -112,13 +111,11 @@ pub mod prop_tests {
         validator_set: &ValidatorSet,
         batch_header: &BatchHeader<CurrentNetwork>,
         rng: &mut TestRng,
-    ) -> IndexMap<Signature<CurrentNetwork>, i64> {
-        let mut signatures = IndexMap::with_capacity(validator_set.0.len());
+    ) -> IndexSet<Signature<CurrentNetwork>> {
+        let mut signatures = IndexSet::with_capacity(validator_set.0.len());
         for validator in validator_set.0.iter() {
             let private_key = validator.private_key;
-            let timestamp = time::OffsetDateTime::now_utc().unix_timestamp();
-            let timestamp_field = Field::from_u64(timestamp as u64);
-            signatures.insert(private_key.sign(&[batch_header.batch_id(), timestamp_field], rng).unwrap(), timestamp);
+            signatures.insert(private_key.sign(&[batch_header.batch_id()], rng).unwrap());
         }
         signatures
     }
diff --git a/node/bft/ledger-service/src/ledger.rs b/node/bft/ledger-service/src/ledger.rs
index 2b34321c2..3c17a2aba 100644
--- a/node/bft/ledger-service/src/ledger.rs
+++ b/node/bft/ledger-service/src/ledger.rs
@@ -98,9 +98,9 @@ impl<N: Network, C: ConsensusStorage<N>> LedgerService<N> for CoreLedgerService<
         self.ledger.get_solution(solution_id)
     }
 
-    /// Returns the transaction for the given transaction ID.
-    fn get_transaction(&self, transaction_id: N::TransactionID) -> Result<Transaction<N>> {
-        self.ledger.get_transaction(transaction_id)
+    /// Returns the unconfirmed transaction for the given transaction ID.
+    fn get_unconfirmed_transaction(&self, transaction_id: N::TransactionID) -> Result<Transaction<N>> {
+        self.ledger.get_unconfirmed_transaction(&transaction_id)
     }
 
     /// Returns the batch certificate for the given batch certificate ID.
diff --git a/node/bft/ledger-service/src/mock.rs b/node/bft/ledger-service/src/mock.rs
index a580c2b4d..e2d95b412 100644
--- a/node/bft/ledger-service/src/mock.rs
+++ b/node/bft/ledger-service/src/mock.rs
@@ -105,9 +105,9 @@ impl<N: Network> LedgerService<N> for MockLedgerService<N> {
         unreachable!("MockLedgerService does not support get_solution")
     }
 
-    /// Returns the transaction for the given transaction ID.
-    fn get_transaction(&self, _transaction_id: N::TransactionID) -> Result<Transaction<N>> {
-        unreachable!("MockLedgerService does not support get_transaction")
+    /// Returns the unconfirmed transaction for the given transaction ID.
+    fn get_unconfirmed_transaction(&self, _transaction_id: N::TransactionID) -> Result<Transaction<N>> {
+        unreachable!("MockLedgerService does not support get_unconfirmed_transaction")
     }
 
     /// Returns the batch certificate for the given batch certificate ID.
diff --git a/node/bft/ledger-service/src/prover.rs b/node/bft/ledger-service/src/prover.rs
index 8f98f06ec..1c0e3eee1 100644
--- a/node/bft/ledger-service/src/prover.rs
+++ b/node/bft/ledger-service/src/prover.rs
@@ -87,8 +87,8 @@ impl<N: Network> LedgerService<N> for ProverLedgerService<N> {
         bail!("Solution '{solution_id}' does not exist in prover")
     }
 
-    /// Returns the transaction for the given transaction ID.
-    fn get_transaction(&self, transaction_id: N::TransactionID) -> Result<Transaction<N>> {
+    /// Returns the unconfirmed transaction for the given transaction ID.
+    fn get_unconfirmed_transaction(&self, transaction_id: N::TransactionID) -> Result<Transaction<N>> {
         bail!("Transaction '{transaction_id}' does not exist in prover")
     }
 
diff --git a/node/bft/ledger-service/src/traits.rs b/node/bft/ledger-service/src/traits.rs
index 05e55b5d4..4daf0f3b5 100644
--- a/node/bft/ledger-service/src/traits.rs
+++ b/node/bft/ledger-service/src/traits.rs
@@ -55,8 +55,8 @@ pub trait LedgerService<N: Network>: Debug + Send + Sync {
     /// Returns the solution for the given solution ID.
     fn get_solution(&self, solution_id: &PuzzleCommitment<N>) -> Result<ProverSolution<N>>;
 
-    /// Returns the transaction for the given transaction ID.
-    fn get_transaction(&self, transaction_id: N::TransactionID) -> Result<Transaction<N>>;
+    /// Returns the unconfirmed transaction for the given transaction ID.
+    fn get_unconfirmed_transaction(&self, transaction_id: N::TransactionID) -> Result<Transaction<N>>;
 
     /// Returns the batch certificate for the given batch certificate ID.
     fn get_batch_certificate(&self, certificate_id: &Field<N>) -> Result<BatchCertificate<N>>;
diff --git a/node/bft/ledger-service/src/translucent.rs b/node/bft/ledger-service/src/translucent.rs
index 58ed5a0e8..9b9ce91e1 100644
--- a/node/bft/ledger-service/src/translucent.rs
+++ b/node/bft/ledger-service/src/translucent.rs
@@ -94,9 +94,9 @@ impl<N: Network, C: ConsensusStorage<N>> LedgerService<N> for TranslucentLedgerS
         self.inner.get_solution(solution_id)
     }
 
-    /// Returns the transaction for the given transaction ID.
-    fn get_transaction(&self, transaction_id: N::TransactionID) -> Result<Transaction<N>> {
-        self.inner.get_transaction(transaction_id)
+    /// Returns the unconfirmed transaction for the given transaction ID.
+    fn get_unconfirmed_transaction(&self, transaction_id: N::TransactionID) -> Result<Transaction<N>> {
+        self.inner.get_unconfirmed_transaction(transaction_id)
     }
 
     /// Returns the batch certificate for the given batch certificate ID.
diff --git a/node/bft/src/bft.rs b/node/bft/src/bft.rs
index aa826d843..7f5ec2b00 100644
--- a/node/bft/src/bft.rs
+++ b/node/bft/src/bft.rs
@@ -349,7 +349,7 @@ impl<N: Network> BFT<N> {
             return true;
         };
         // Retrieve the leader certificate ID.
-        let leader_certificate_id = leader_certificate.certificate_id();
+        let leader_certificate_id = leader_certificate.id();
         // Retrieve the certificates for the current round.
         let current_certificates = self.storage().get_certificates_for_round(current_round);
         // Retrieve the previous committee of the current round.
@@ -447,7 +447,7 @@ impl<N: Network> BFT<N> {
         // Construct a set over the authors who included the leader's certificate in the certificate round.
         let authors = certificates
             .values()
-            .filter_map(|c| match c.previous_certificate_ids().contains(&leader_certificate.certificate_id()) {
+            .filter_map(|c| match c.previous_certificate_ids().contains(&leader_certificate.id()) {
                 true => Some(c.author()),
                 false => None,
             })
@@ -498,7 +498,11 @@ impl<N: Network> BFT<N> {
                 }
                 // Retrieve the transmission.
                 let Some(transmission) = self.storage().get_transmission(*transmission_id) else {
-                    bail!("BFT failed to retrieve transmission {}", fmt_id(transmission_id));
+                    bail!(
+                        "BFT failed to retrieve transmission '{}' from round {}",
+                        fmt_id(transmission_id),
+                        certificate.round()
+                    );
                 };
                 // Add the transmission to the set.
                 transmissions.insert(*transmission_id, transmission);
@@ -622,7 +626,7 @@ impl<N: Network> BFT<N> {
                     }
                 };
                 // Insert the previous certificate into the set of already ordered certificates.
-                already_ordered.insert(previous_certificate.certificate_id());
+                already_ordered.insert(previous_certificate.id());
                 // Insert the previous certificate into the buffer.
                 buffer.push(previous_certificate);
             }
@@ -715,7 +719,7 @@ impl<N: Network> BFT<N> {
             // Iterate over the certificates.
             for certificate in certificates {
                 // If the certificate is not the latest leader certificate, insert it.
-                if leader_certificate.certificate_id() != certificate.certificate_id() {
+                if leader_certificate.id() != certificate.id() {
                     // Insert the certificate into the DAG.
                     dag.insert(certificate);
                 }
@@ -1012,8 +1016,8 @@ mod tests {
             assert_eq!(candidate_certificates.len(), 1);
             let expected_certificates = vec![certificate.clone()];
             assert_eq!(
-                candidate_certificates.iter().map(|c| c.certificate_id()).collect::<Vec<_>>(),
-                expected_certificates.iter().map(|c| c.certificate_id()).collect::<Vec<_>>()
+                candidate_certificates.iter().map(|c| c.id()).collect::<Vec<_>>(),
+                expected_certificates.iter().map(|c| c.id()).collect::<Vec<_>>()
             );
             assert_eq!(candidate_certificates, expected_certificates);
         }
@@ -1048,8 +1052,8 @@ mod tests {
                 certificate,
             ];
             assert_eq!(
-                candidate_certificates.iter().map(|c| c.certificate_id()).collect::<Vec<_>>(),
-                expected_certificates.iter().map(|c| c.certificate_id()).collect::<Vec<_>>()
+                candidate_certificates.iter().map(|c| c.id()).collect::<Vec<_>>(),
+                expected_certificates.iter().map(|c| c.id()).collect::<Vec<_>>()
             );
             assert_eq!(candidate_certificates, expected_certificates);
         }
@@ -1076,7 +1080,7 @@ mod tests {
             rng,
         );
         // Construct the previous certificate IDs.
-        let previous_certificate_ids: IndexSet<_> = previous_certificates.iter().map(|c| c.certificate_id()).collect();
+        let previous_certificate_ids: IndexSet<_> = previous_certificates.iter().map(|c| c.id()).collect();
 
         /* Test missing previous certificate. */
 
diff --git a/node/bft/src/gateway.rs b/node/bft/src/gateway.rs
index ef7888b16..cd5d4056a 100644
--- a/node/bft/src/gateway.rs
+++ b/node/bft/src/gateway.rs
@@ -500,7 +500,7 @@ impl<N: Network> Gateway<N> {
             // Retrieve the certificate ID.
             let certificate_id = match &event {
                 Event::CertificateRequest(CertificateRequest { certificate_id }) => *certificate_id,
-                Event::CertificateResponse(CertificateResponse { certificate }) => certificate.certificate_id(),
+                Event::CertificateResponse(CertificateResponse { certificate }) => certificate.id(),
                 _ => unreachable!(),
             };
             // Skip processing this certificate if the rate limit was exceed (i.e. someone is spamming a specific certificate).
diff --git a/node/bft/src/helpers/dag.rs b/node/bft/src/helpers/dag.rs
index 72aaace74..a97b31689 100644
--- a/node/bft/src/helpers/dag.rs
+++ b/node/bft/src/helpers/dag.rs
@@ -61,9 +61,7 @@ impl<N: Network> DAG<N> {
 
     /// Returns `true` if the given certificate ID exists in the given round.
     pub fn contains_certificate_in_round(&self, round: u64, certificate_id: Field<N>) -> bool {
-        self.graph
-            .get(&round)
-            .map_or(false, |map| map.values().any(|certificate| certificate.certificate_id() == certificate_id))
+        self.graph.get(&round).map_or(false, |map| map.values().any(|certificate| certificate.id() == certificate_id))
     }
 
     /// Returns the batch certificate for the given round and author.
@@ -79,7 +77,7 @@ impl<N: Network> DAG<N> {
     ) -> Option<BatchCertificate<N>> {
         self.graph
             .get(&round)
-            .and_then(|map| map.values().find(|certificate| certificate.certificate_id() == certificate_id))
+            .and_then(|map| map.values().find(|certificate| certificate.id() == certificate_id))
             .cloned()
     }
 
@@ -94,7 +92,7 @@ impl<N: Network> DAG<N> {
         let author = certificate.author();
 
         // If the certificate was not recently committed, insert it into the DAG.
-        if !self.is_recently_committed(round, certificate.certificate_id()) {
+        if !self.is_recently_committed(round, certificate.id()) {
             // Insert the certificate into the DAG.
             let _previous = self.graph.entry(round).or_default().insert(author, certificate);
             // If a previous certificate existed for the author, log it.
@@ -107,7 +105,7 @@ impl<N: Network> DAG<N> {
 
     /// Commits a certificate, removing all certificates for this author at or before this round from the DAG.
     pub fn commit(&mut self, certificate: &BatchCertificate<N>, max_gc_rounds: u64) {
-        let certificate_id = certificate.certificate_id();
+        let certificate_id = certificate.id();
         let certificate_round = certificate.round();
         let author = certificate.author();
 
@@ -176,12 +174,9 @@ mod tests {
         // Sample a certificate for round 2.
         let certificate = sample_batch_certificate_for_round(ROUND, rng);
         dag.insert(certificate.clone());
-        assert!(dag.contains_certificate_in_round(ROUND, certificate.certificate_id()));
+        assert!(dag.contains_certificate_in_round(ROUND, certificate.id()));
         assert_eq!(dag.get_certificate_for_round_with_author(ROUND, certificate.author()), Some(certificate.clone()));
-        assert_eq!(
-            dag.get_certificate_for_round_with_id(ROUND, certificate.certificate_id()),
-            Some(certificate.clone())
-        );
+        assert_eq!(dag.get_certificate_for_round_with_id(ROUND, certificate.id()), Some(certificate.clone()));
         assert_eq!(
             dag.get_certificates_for_round(ROUND),
             Some(vec![(certificate.author(), certificate)].into_iter().collect())
@@ -200,12 +195,9 @@ mod tests {
 
         // Insert the certificate for round 2.
         dag.insert(certificate_2.clone());
-        assert!(dag.contains_certificate_in_round(2, certificate_2.certificate_id()));
+        assert!(dag.contains_certificate_in_round(2, certificate_2.id()));
         assert_eq!(dag.get_certificate_for_round_with_author(2, certificate_2.author()), Some(certificate_2.clone()));
-        assert_eq!(
-            dag.get_certificate_for_round_with_id(2, certificate_2.certificate_id()),
-            Some(certificate_2.clone())
-        );
+        assert_eq!(dag.get_certificate_for_round_with_id(2, certificate_2.id()), Some(certificate_2.clone()));
         assert_eq!(
             dag.get_certificates_for_round(2),
             Some(vec![(certificate_2.author(), certificate_2.clone())].into_iter().collect())
@@ -214,12 +206,9 @@ mod tests {
 
         // Insert the certificate for round 3.
         dag.insert(certificate_3.clone());
-        assert!(dag.contains_certificate_in_round(3, certificate_3.certificate_id()));
+        assert!(dag.contains_certificate_in_round(3, certificate_3.id()));
         assert_eq!(dag.get_certificate_for_round_with_author(3, certificate_3.author()), Some(certificate_3.clone()));
-        assert_eq!(
-            dag.get_certificate_for_round_with_id(3, certificate_3.certificate_id()),
-            Some(certificate_3.clone())
-        );
+        assert_eq!(dag.get_certificate_for_round_with_id(3, certificate_3.id()), Some(certificate_3.clone()));
         assert_eq!(
             dag.get_certificates_for_round(3),
             Some(vec![(certificate_3.author(), certificate_3.clone())].into_iter().collect())
@@ -236,10 +225,10 @@ mod tests {
 
         // Now commit the certificate for round 3, this will trigger GC.
         dag.commit(&certificate_3, 10);
-        assert!(!dag.contains_certificate_in_round(2, certificate_2.certificate_id()));
-        assert!(!dag.contains_certificate_in_round(3, certificate_3.certificate_id()));
-        assert!(dag.contains_certificate_in_round(2, lower.certificate_id()));
-        assert!(dag.contains_certificate_in_round(4, higher.certificate_id()));
+        assert!(!dag.contains_certificate_in_round(2, certificate_2.id()));
+        assert!(!dag.contains_certificate_in_round(3, certificate_3.id()));
+        assert!(dag.contains_certificate_in_round(2, lower.id()));
+        assert!(dag.contains_certificate_in_round(4, higher.id()));
         assert_eq!(dag.last_committed_round(), 3);
     }
 
@@ -256,36 +245,36 @@ mod tests {
 
         // Insert the certificate for round 2.
         dag.insert(certificate_2.clone());
-        assert!(!dag.is_recently_committed(2, certificate_2.certificate_id()));
-        assert!(!dag.is_recently_committed(3, certificate_3.certificate_id()));
+        assert!(!dag.is_recently_committed(2, certificate_2.id()));
+        assert!(!dag.is_recently_committed(3, certificate_3.id()));
 
         // Now commit the certificate for round 2.
         dag.commit(&certificate_2, 2);
-        assert!(dag.is_recently_committed(2, certificate_2.certificate_id()));
-        assert!(!dag.is_recently_committed(3, certificate_3.certificate_id()));
+        assert!(dag.is_recently_committed(2, certificate_2.id()));
+        assert!(!dag.is_recently_committed(3, certificate_3.id()));
 
         // Insert the certificate for round 3.
         dag.insert(certificate_3.clone());
-        assert!(dag.is_recently_committed(2, certificate_2.certificate_id()));
-        assert!(!dag.is_recently_committed(3, certificate_3.certificate_id()));
+        assert!(dag.is_recently_committed(2, certificate_2.id()));
+        assert!(!dag.is_recently_committed(3, certificate_3.id()));
 
         // Now commit the certificate for round 3.
         dag.commit(&certificate_3, 2);
-        assert!(dag.is_recently_committed(2, certificate_2.certificate_id()));
-        assert!(dag.is_recently_committed(3, certificate_3.certificate_id()));
+        assert!(dag.is_recently_committed(2, certificate_2.id()));
+        assert!(dag.is_recently_committed(3, certificate_3.id()));
 
         /* Test GC case */
 
         // Insert the certificate for round 4.
         dag.insert(certificate_4.clone());
-        assert!(dag.is_recently_committed(2, certificate_2.certificate_id()));
-        assert!(dag.is_recently_committed(3, certificate_3.certificate_id()));
-        assert!(!dag.is_recently_committed(4, certificate_4.certificate_id()));
+        assert!(dag.is_recently_committed(2, certificate_2.id()));
+        assert!(dag.is_recently_committed(3, certificate_3.id()));
+        assert!(!dag.is_recently_committed(4, certificate_4.id()));
 
         // Now commit the certificate for round 4.
         dag.commit(&certificate_4, 2);
-        assert!(!dag.is_recently_committed(2, certificate_2.certificate_id()));
-        assert!(dag.is_recently_committed(3, certificate_3.certificate_id()));
-        assert!(dag.is_recently_committed(4, certificate_4.certificate_id()));
+        assert!(!dag.is_recently_committed(2, certificate_2.id()));
+        assert!(dag.is_recently_committed(3, certificate_3.id()));
+        assert!(dag.is_recently_committed(4, certificate_4.id()));
     }
 }
diff --git a/node/bft/src/helpers/proposal.rs b/node/bft/src/helpers/proposal.rs
index 1ffc80b3f..7b26e0f9f 100644
--- a/node/bft/src/helpers/proposal.rs
+++ b/node/bft/src/helpers/proposal.rs
@@ -12,7 +12,6 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-use crate::helpers::check_timestamp_for_liveness;
 use snarkvm::{
     console::{
         account::{Address, Signature},
@@ -26,7 +25,7 @@ use snarkvm::{
     prelude::{bail, ensure, Itertools, Result},
 };
 
-use indexmap::IndexMap;
+use indexmap::{IndexMap, IndexSet};
 use std::collections::HashSet;
 
 pub struct Proposal<N: Network> {
@@ -34,8 +33,8 @@ pub struct Proposal<N: Network> {
     batch_header: BatchHeader<N>,
     /// The proposed transmissions.
     transmissions: IndexMap<TransmissionID<N>, Transmission<N>>,
-    /// The map of `(signature, timestamp)` entries.
-    signatures: IndexMap<Signature<N>, i64>,
+    /// The set of signatures.
+    signatures: IndexSet<Signature<N>>,
 }
 
 impl<N: Network> Proposal<N> {
@@ -95,7 +94,7 @@ impl<N: Network> Proposal<N> {
 
     /// Returns the signers.
     pub fn signers(&self) -> HashSet<Address<N>> {
-        self.signatures.keys().chain(Some(self.batch_header.signature())).map(Signature::to_address).collect()
+        self.signatures.iter().chain(Some(self.batch_header.signature())).map(Signature::to_address).collect()
     }
 
     /// Returns the nonsigners.
@@ -136,7 +135,6 @@ impl<N: Network> Proposal<N> {
         &mut self,
         signer: Address<N>,
         signature: Signature<N>,
-        timestamp: i64,
         committee: &Committee<N>,
     ) -> Result<()> {
         // Ensure the signer is in the committee.
@@ -149,13 +147,11 @@ impl<N: Network> Proposal<N> {
         }
         // Verify the signature. If the signature is not valid, return an error.
         // Note: This check ensures the peer's address matches the address of the signature.
-        if !signature.verify(&signer, &[self.batch_id(), Field::from_u64(timestamp as u64)]) {
+        if !signature.verify(&signer, &[self.batch_id()]) {
             bail!("Signature verification failed")
         }
-        // Check the timestamp for liveness.
-        check_timestamp_for_liveness(timestamp)?;
         // Insert the signature.
-        self.signatures.insert(signature, timestamp);
+        self.signatures.insert(signature);
         Ok(())
     }
 
@@ -167,7 +163,7 @@ impl<N: Network> Proposal<N> {
         // Ensure the quorum threshold has been reached.
         ensure!(self.is_quorum_threshold_reached(committee), "The quorum threshold has not been reached");
         // Create the batch certificate.
-        let certificate = BatchCertificate::new(self.batch_header.clone(), self.signatures.clone())?;
+        let certificate = BatchCertificate::from(self.batch_header.clone(), self.signatures.clone())?;
         // Return the certificate and transmissions.
         Ok((certificate, self.transmissions.clone()))
     }
diff --git a/node/bft/src/helpers/storage.rs b/node/bft/src/helpers/storage.rs
index 9ce18d253..107664e19 100644
--- a/node/bft/src/helpers/storage.rs
+++ b/node/bft/src/helpers/storage.rs
@@ -200,7 +200,7 @@ impl<N: Network> Storage<N> {
                 // Iterate over the certificates for the GC round.
                 for certificate in self.get_certificates_for_round(gc_round).iter() {
                     // Remove the certificate from storage.
-                    self.remove_certificate(certificate.certificate_id());
+                    self.remove_certificate(certificate.id());
                 }
             }
             // Update the GC round.
@@ -437,7 +437,7 @@ impl<N: Network> Storage<N> {
         let gc_log = format!("(gc = {gc_round})");
 
         // Ensure the certificate ID does not already exist in storage.
-        if self.contains_certificate(certificate.certificate_id()) {
+        if self.contains_certificate(certificate.id()) {
             bail!("Certificate for round {round} already exists in storage {gc_log}")
         }
 
@@ -449,11 +449,8 @@ impl<N: Network> Storage<N> {
         // Ensure the batch header is well-formed.
         let missing_transmissions = self.check_batch_header(certificate.batch_header(), transmissions)?;
 
-        // Iterate over the timestamps.
-        for timestamp in certificate.timestamps() {
-            // Check the timestamp for liveness.
-            check_timestamp_for_liveness(timestamp)?;
-        }
+        // Check the timestamp for liveness.
+        check_timestamp_for_liveness(certificate.timestamp())?;
 
         // Retrieve the previous committee for the batch round.
         let Ok(previous_committee) = self.ledger.get_previous_committee_for_round(round) else {
@@ -522,7 +519,7 @@ impl<N: Network> Storage<N> {
         // Retrieve the round.
         let round = certificate.round();
         // Retrieve the certificate ID.
-        let certificate_id = certificate.certificate_id();
+        let certificate_id = certificate.id();
         // Retrieve the batch ID.
         let batch_id = certificate.batch_id();
         // Retrieve the author of the batch.
@@ -644,7 +641,7 @@ impl<N: Network> Storage<N> {
             return;
         }
         // If the certificate ID already exists in storage, skip it.
-        if self.contains_certificate(certificate.certificate_id()) {
+        if self.contains_certificate(certificate.id()) {
             return;
         }
         // Retrieve the transmissions for the certificate.
@@ -689,8 +686,8 @@ impl<N: Network> Storage<N> {
                     match unconfirmed_transactions.remove(transaction_id) {
                         // Insert the transaction.
                         Some(transaction) => missing_transmissions.insert(*transmission_id, transaction.into()),
-                        // Otherwise, try to load the transaction from the ledger.
-                        None => match self.ledger.get_transaction(*transaction_id) {
+                        // Otherwise, try to load the unconfirmed transaction from the ledger.
+                        None => match self.ledger.get_unconfirmed_transaction(*transaction_id) {
                             // Insert the transaction.
                             Ok(transaction) => missing_transmissions.insert(*transmission_id, transaction.into()),
                             Err(_) => {
@@ -703,7 +700,7 @@ impl<N: Network> Storage<N> {
             }
         }
         // Insert the batch certificate into storage.
-        let certificate_id = fmt_id(certificate.certificate_id());
+        let certificate_id = fmt_id(certificate.id());
         debug!(
             "Syncing certificate '{certificate_id}' for round {} with {} transmissions",
             certificate.round(),
@@ -752,7 +749,7 @@ impl<N: Network> Storage<N> {
         // Retrieve the round.
         let round = certificate.round();
         // Retrieve the certificate ID.
-        let certificate_id = certificate.certificate_id();
+        let certificate_id = certificate.id();
         // Retrieve the batch ID.
         let batch_id = certificate.batch_id();
         // Retrieve the author of the batch.
@@ -840,7 +837,7 @@ mod tests {
         HashMap<TransmissionID<CurrentNetwork>, (Transmission<CurrentNetwork>, IndexSet<Field<CurrentNetwork>>)>,
     ) {
         // Retrieve the certificate ID.
-        let certificate_id = certificate.certificate_id();
+        let certificate_id = certificate.id();
 
         let mut missing_transmissions = HashMap::new();
         let mut transmissions = HashMap::<_, (_, IndexSet<Field<CurrentNetwork>>)>::new();
@@ -878,7 +875,7 @@ mod tests {
         // Create a new certificate.
         let certificate = snarkvm::ledger::narwhal::batch_certificate::test_helpers::sample_batch_certificate(rng);
         // Retrieve the certificate ID.
-        let certificate_id = certificate.certificate_id();
+        let certificate_id = certificate.id();
         // Retrieve the round.
         let round = certificate.round();
         // Retrieve the batch ID.
@@ -944,7 +941,7 @@ mod tests {
         // Create a new certificate.
         let certificate = snarkvm::ledger::narwhal::batch_certificate::test_helpers::sample_batch_certificate(rng);
         // Retrieve the certificate ID.
-        let certificate_id = certificate.certificate_id();
+        let certificate_id = certificate.id();
         // Retrieve the round.
         let round = certificate.round();
         // Retrieve the batch ID.
@@ -1129,13 +1126,11 @@ pub mod prop_tests {
         validator_set: &ValidatorSet,
         batch_header: &BatchHeader<CurrentNetwork>,
         rng: &mut R,
-    ) -> IndexMap<Signature<CurrentNetwork>, i64> {
-        let mut signatures = IndexMap::with_capacity(validator_set.0.len());
+    ) -> IndexSet<Signature<CurrentNetwork>> {
+        let mut signatures = IndexSet::with_capacity(validator_set.0.len());
         for validator in validator_set.0.iter() {
             let private_key = validator.private_key;
-            let timestamp = time::OffsetDateTime::now_utc().unix_timestamp();
-            let timestamp_field = Field::from_u64(timestamp as u64);
-            signatures.insert(private_key.sign(&[batch_header.batch_id(), timestamp_field], rng).unwrap(), timestamp);
+            signatures.insert(private_key.sign(&[batch_header.batch_id()], rng).unwrap());
         }
         signatures
     }
@@ -1174,14 +1169,14 @@ pub mod prop_tests {
             &mut rng,
         )
         .unwrap();
-        let certificate = BatchCertificate::new(
+        let certificate = BatchCertificate::from(
             batch_header.clone(),
             sign_batch_header(&ValidatorSet(validators), &batch_header, &mut rng),
         )
         .unwrap();
 
         // Retrieve the certificate ID.
-        let certificate_id = certificate.certificate_id();
+        let certificate_id = certificate.id();
         let mut internal_transmissions = HashMap::<_, (_, IndexSet<Field<CurrentNetwork>>)>::new();
         for (AnyTransmissionID(id), AnyTransmission(t)) in transmissions.iter().cloned() {
             internal_transmissions.entry(id).or_insert((t, Default::default())).1.insert(certificate_id);
diff --git a/node/bft/src/helpers/timestamp.rs b/node/bft/src/helpers/timestamp.rs
index 291f9e5f5..0c546bd26 100644
--- a/node/bft/src/helpers/timestamp.rs
+++ b/node/bft/src/helpers/timestamp.rs
@@ -38,7 +38,7 @@ pub fn check_timestamp_for_liveness(timestamp: i64) -> Result<()> {
 
 #[cfg(test)]
 mod prop_tests {
-    use super::{check_timestamp_for_liveness, now};
+    use super::*;
     use crate::MAX_TIMESTAMP_DELTA_IN_SECS;
 
     use proptest::prelude::*;
@@ -54,13 +54,11 @@ mod prop_tests {
 
     #[proptest]
     fn test_check_timestamp_for_liveness(#[strategy(any_valid_timestamp())] timestamp: i64) {
-        let result = check_timestamp_for_liveness(timestamp);
-        assert!(result.is_ok());
+        check_timestamp_for_liveness(timestamp).unwrap();
     }
 
     #[proptest]
     fn test_check_timestamp_for_liveness_too_far_in_future(#[strategy(any_invalid_timestamp())] timestamp: i64) {
-        let result = check_timestamp_for_liveness(timestamp);
-        assert!(result.is_err());
+        assert!(check_timestamp_for_liveness(timestamp).is_err());
     }
 }
diff --git a/node/bft/src/primary.rs b/node/bft/src/primary.rs
index 1c3cf3aed..21c087258 100644
--- a/node/bft/src/primary.rs
+++ b/node/bft/src/primary.rs
@@ -431,17 +431,15 @@ impl<N: Network> Primary<N> {
 
         // Retrieve the private key.
         let private_key = *self.gateway.account().private_key();
-        // Generate the local timestamp for batch
-        let timestamp = now();
         // Prepare the transmission IDs.
         let transmission_ids = transmissions.keys().copied().collect();
         // Prepare the certificate IDs.
-        let certificate_ids = previous_certificates.into_iter().map(|c| c.certificate_id()).collect();
+        let certificate_ids = previous_certificates.into_iter().map(|c| c.id()).collect();
         // Sign the batch header.
         let batch_header = spawn_blocking!(BatchHeader::new(
             &private_key,
             round,
-            timestamp,
+            now(),
             transmission_ids,
             certificate_ids,
             &mut rand::thread_rng()
@@ -528,18 +526,15 @@ impl<N: Network> Primary<N> {
 
         // Retrieve the batch ID.
         let batch_id = batch_header.batch_id();
-        // Generate a timestamp.
-        let timestamp = now();
         // Sign the batch ID.
         let account = self.gateway.account().clone();
-        let signature =
-            spawn_blocking!(account.sign(&[batch_id, Field::from_u64(now() as u64)], &mut rand::thread_rng()))?;
+        let signature = spawn_blocking!(account.sign(&[batch_id], &mut rand::thread_rng()))?;
         // Cache the round and batch ID for this validator.
         self.signed_proposals.write().insert(batch_header.author(), (batch_header.round(), batch_id));
         // Broadcast the signature back to the validator.
         let self_ = self.clone();
         tokio::spawn(async move {
-            let event = Event::BatchSignature(BatchSignature::new(batch_id, signature, timestamp));
+            let event = Event::BatchSignature(BatchSignature::new(batch_id, signature));
             // Send the batch signature to the peer.
             if self_.gateway.send(peer_ip, event).await.is_some() {
                 debug!("Signed a batch for round {batch_round} from '{peer_ip}'");
@@ -565,7 +560,7 @@ impl<N: Network> Primary<N> {
         self.check_proposed_batch_for_expiration().await?;
 
         // Retrieve the signature and timestamp.
-        let BatchSignature { batch_id, signature, timestamp } = batch_signature;
+        let BatchSignature { batch_id, signature } = batch_signature;
 
         // Retrieve the signer.
         let signer = spawn_blocking!(Ok(signature.to_address()))?;
@@ -594,7 +589,11 @@ impl<N: Network> Primary<N> {
                     if proposal.batch_id() != batch_id {
                         match self.storage.contains_batch(batch_id) {
                             true => bail!("This batch was already certified"),
-                            false => bail!("Unknown batch ID '{batch_id}'"),
+                            false => bail!(
+                                "Unknown batch ID '{batch_id}', expected '{}' for round {}",
+                                proposal.batch_id(),
+                                proposal.round()
+                            ),
                         }
                     }
                     // Retrieve the previous committee for the round.
@@ -604,7 +603,7 @@ impl<N: Network> Primary<N> {
                         bail!("Signature is from a disconnected validator");
                     };
                     // Add the signature to the batch.
-                    proposal.add_signature(signer, signature, timestamp, &previous_committee)?;
+                    proposal.add_signature(signer, signature, &previous_committee)?;
                     info!("Received a batch signature for round {} from '{peer_ip}'", proposal.round());
                     // Check if the batch is ready to be certified.
                     if !proposal.is_quorum_threshold_reached(&previous_committee) {
@@ -650,7 +649,7 @@ impl<N: Network> Primary<N> {
         certificate: BatchCertificate<N>,
     ) -> Result<()> {
         // Ensure storage does not already contain the certificate.
-        if self.storage.contains_certificate(certificate.certificate_id()) {
+        if self.storage.contains_certificate(certificate.id()) {
             return Ok(());
         }
 
@@ -661,10 +660,14 @@ impl<N: Network> Primary<N> {
         let author = certificate.author();
 
         // Ensure the batch certificate is from the validator.
-        if self.gateway.resolver().get_address(peer_ip).map_or(true, |address| address != author) {
-            // Proceed to disconnect the validator.
-            self.gateway.disconnect(peer_ip);
-            bail!("Malicious peer - batch certificate from a different validator ({author})");
+        match self.gateway.resolver().get_address(peer_ip) {
+            Some(address) if address != author => {
+                // Proceed to disconnect the validator.
+                self.gateway.disconnect(peer_ip);
+                bail!("Malicious peer - batch certificate from a different validator ({author})");
+            }
+            None => bail!("Batch certificate from a disconnected validator"),
+            _ => (),
         }
         // Ensure the batch certificate is authored by a current committee member.
         if !self.gateway.is_authorized_validator_address(author) {
@@ -719,7 +722,7 @@ impl<N: Network> Primary<N> {
         certificate: BatchCertificate<N>,
     ) -> Result<()> {
         // Ensure storage does not already contain the certificate.
-        if self.storage.contains_certificate(certificate.certificate_id()) {
+        if self.storage.contains_certificate(certificate.id()) {
             return Ok(());
         }
 
@@ -1164,7 +1167,7 @@ impl<N: Network> Primary<N> {
             return Ok(());
         }
         // If the certificate already exists in storage, return early.
-        if self.storage.contains_certificate(certificate.certificate_id()) {
+        if self.storage.contains_certificate(certificate.id()) {
             return Ok(());
         }
 
@@ -1172,7 +1175,7 @@ impl<N: Network> Primary<N> {
         let missing_transmissions = self.sync_with_batch_header_from_peer(peer_ip, batch_header).await?;
 
         // Check if the certificate needs to be stored.
-        if !self.storage.contains_certificate(certificate.certificate_id()) {
+        if !self.storage.contains_certificate(certificate.id()) {
             // Store the batch certificate.
             self.storage.insert_certificate(certificate.clone(), missing_transmissions)?;
             debug!("Stored a batch certificate for round {batch_round} from '{peer_ip}'");
@@ -1486,35 +1489,29 @@ mod tests {
             if account.address() == primary.gateway.account().address() {
                 continue;
             }
-
             let batch_id = primary.proposed_batch.read().as_ref().unwrap().batch_id();
-            let timestamp = now();
-            let signature = account.sign(&[batch_id, Field::from_u64(timestamp as u64)], rng).unwrap();
-            signatures.push((*socket_addr, BatchSignature::new(batch_id, signature, timestamp)));
+            let signature = account.sign(&[batch_id], rng).unwrap();
+            signatures.push((*socket_addr, BatchSignature::new(batch_id, signature)));
         }
 
         signatures
     }
 
-    // Creates a signature of the batch ID for each committe member
-    // (excluding the primary).
+    /// Creates a signature of the batch ID for each committee member (excluding the primary).
     fn peer_signatures_for_batch(
         primary_address: Address<CurrentNetwork>,
         accounts: &[(SocketAddr, Account<CurrentNetwork>)],
         batch_id: Field<CurrentNetwork>,
         rng: &mut TestRng,
-    ) -> IndexMap<Signature<CurrentNetwork>, i64> {
-        let mut signatures = IndexMap::new();
+    ) -> IndexSet<Signature<CurrentNetwork>> {
+        let mut signatures = IndexSet::new();
         for (_, account) in accounts {
             if account.address() == primary_address {
                 continue;
             }
-
-            let timestamp = now();
-            let signature = account.sign(&[batch_id, Field::from_u64(timestamp as u64)], rng).unwrap();
-            signatures.insert(signature, timestamp);
+            let signature = account.sign(&[batch_id], rng).unwrap();
+            signatures.insert(signature);
         }
-
         signatures
     }
 
@@ -1544,7 +1541,7 @@ mod tests {
         let batch_header =
             BatchHeader::new(private_key, round, timestamp, transmission_ids, previous_certificate_ids, rng).unwrap();
         let signatures = peer_signatures_for_batch(primary_address, accounts, batch_header.batch_id(), rng);
-        let certificate = BatchCertificate::<CurrentNetwork>::new(batch_header, signatures).unwrap();
+        let certificate = BatchCertificate::<CurrentNetwork>::from(batch_header, signatures).unwrap();
         (certificate, transmissions)
     }
 
@@ -1566,7 +1563,7 @@ mod tests {
                     previous_certificates.clone(),
                     rng,
                 );
-                next_certificates.insert(certificate.certificate_id());
+                next_certificates.insert(certificate.id());
                 assert!(primary.storage.insert_certificate(certificate, transmissions).is_ok());
             }
 
diff --git a/node/bft/src/sync/mod.rs b/node/bft/src/sync/mod.rs
index a43e7963f..f755dfd6c 100644
--- a/node/bft/src/sync/mod.rs
+++ b/node/bft/src/sync/mod.rs
@@ -360,12 +360,12 @@ impl<N: Network> Sync<N> {
     fn finish_certificate_request(&self, peer_ip: SocketAddr, response: CertificateResponse<N>) {
         let certificate = response.certificate;
         // Check if the peer IP exists in the pending queue for the given certificate ID.
-        let exists = self.pending.get(certificate.certificate_id()).unwrap_or_default().contains(&peer_ip);
+        let exists = self.pending.get(certificate.id()).unwrap_or_default().contains(&peer_ip);
         // If the peer IP exists, finish the pending request.
         if exists {
             // TODO: Validate the certificate.
             // Remove the certificate ID from the pending queue.
-            self.pending.remove(certificate.certificate_id(), Some(certificate));
+            self.pending.remove(certificate.id(), Some(certificate));
         }
     }
 }
diff --git a/node/bft/src/worker.rs b/node/bft/src/worker.rs
index 6812cf025..7f3da6204 100644
--- a/node/bft/src/worker.rs
+++ b/node/bft/src/worker.rs
@@ -476,7 +476,7 @@ mod tests {
             fn get_block(&self, height: u32) -> Result<Block<N>>;
             fn get_blocks(&self, heights: Range<u32>) -> Result<Vec<Block<N>>>;
             fn get_solution(&self, solution_id: &PuzzleCommitment<N>) -> Result<ProverSolution<N>>;
-            fn get_transaction(&self, transaction_id: N::TransactionID) -> Result<Transaction<N>>;
+            fn get_unconfirmed_transaction(&self, transaction_id: N::TransactionID) -> Result<Transaction<N>>;
             fn get_batch_certificate(&self, certificate_id: &Field<N>) -> Result<BatchCertificate<N>>;
             fn current_committee(&self) -> Result<Committee<N>>;
             fn get_committee_for_round(&self, round: u64) -> Result<Committee<N>>;
diff --git a/snarkos/main.rs b/snarkos/main.rs
index a456e330a..64cef2f18 100644
--- a/snarkos/main.rs
+++ b/snarkos/main.rs
@@ -15,6 +15,8 @@
 use snarkos_cli::{commands::CLI, helpers::Updater};
 
 use clap::Parser;
+use std::process::exit;
+
 #[cfg(feature = "jemalloc")]
 use tikv_jemallocator::Jemalloc;
 
@@ -30,7 +32,10 @@ fn main() -> anyhow::Result<()> {
     // Run the CLI.
     match cli.command.parse() {
         Ok(output) => println!("{output}\n"),
-        Err(error) => println!("⚠️  {error}\n"),
+        Err(error) => {
+            println!("⚠️  {error}\n");
+            exit(1);
+        }
     }
     Ok(())
 }
