diff --git a/network/src/sync/miner.rs b/network/src/sync/miner.rs
index e478d3e6b..33f7b0274 100644
--- a/network/src/sync/miner.rs
+++ b/network/src/sync/miner.rs
@@ -23,8 +23,10 @@ use snarkvm_dpc::{testnet1::instantiated::*, Address};
 use tokio::{task, time::sleep};
 use tracing::*;
 
-use snarkos_consensus::MineContext;
+use snarkos_consensus::{error::ConsensusError, MineContext};
 use snarkos_metrics::{self as metrics};
+use snarkvm_algorithms::SNARKError;
+use snarkvm_posw::error::PoswError;
 
 use crate::{Node, State};
 
@@ -95,6 +97,11 @@ impl MinerInstance {
                             self.node.set_state(State::Idle);
                         }
 
+                        // Miner termination shouldn't count as a failure, as it's expected to occur regularly.
+                        if let ConsensusError::PoswError(PoswError::SnarkError(SNARKError::Terminated)) = error {
+                            continue;
+                        }
+
                         warn!(
                             "Miner failed to mine a block {} time(s). (error message: {}).",
                             mining_failure_count, error
