diff --git a/Cargo.lock b/Cargo.lock
index 14da1d80f..69801fdd5 100644
--- a/Cargo.lock
+++ b/Cargo.lock
@@ -2386,6 +2386,7 @@ dependencies = [
  "indexmap",
  "once_cell",
  "rand",
+ "reqwest",
  "serde",
  "snarkos-node-executor",
  "snarkos-node-messages",
diff --git a/account/src/lib.rs b/account/src/lib.rs
index b74acbe8d..9538b6f3f 100644
--- a/account/src/lib.rs
+++ b/account/src/lib.rs
@@ -74,8 +74,8 @@ impl<N: Network> Account<N> {
     }
 
     /// Returns the account address.
-    pub const fn address(&self) -> &Address<N> {
-        &self.address
+    pub const fn address(&self) -> Address<N> {
+        self.address
     }
 }
 
diff --git a/node/ledger/src/lib.rs b/node/ledger/src/lib.rs
index f903f1a89..69877251c 100644
--- a/node/ledger/src/lib.rs
+++ b/node/ledger/src/lib.rs
@@ -248,7 +248,7 @@ impl<N: Network, C: ConsensusStorage<N>> Ledger<N, C> {
     }
 
     /// Creates a transfer transaction.
-    pub fn create_transfer(&self, private_key: &PrivateKey<N>, to: &Address<N>, amount: u64) -> Result<Transaction<N>> {
+    pub fn create_transfer(&self, private_key: &PrivateKey<N>, to: Address<N>, amount: u64) -> Result<Transaction<N>> {
         // Fetch the unspent records.
         let records = self.find_unspent_records(&ViewKey::try_from(private_key)?)?;
         ensure!(!records.len().is_zero(), "The Aleo account has no records to spend.");
diff --git a/node/rest/src/routes.rs b/node/rest/src/routes.rs
index c4f22ee4e..791e542ea 100644
--- a/node/rest/src/routes.rs
+++ b/node/rest/src/routes.rs
@@ -27,6 +27,7 @@ struct BlockRange {
 
 impl<N: Network, C: ConsensusStorage<N>> Rest<N, C> {
     /// Initializes the routes, given the ledger and ledger sender.
+    #[allow(opaque_hidden_inferred_bound)]
     pub fn routes(&self) -> impl Filter<Extract = impl Reply, Error = Rejection> + Clone {
         // GET /testnet3/latest/height
         let latest_height = warp::get()
diff --git a/node/router/Cargo.toml b/node/router/Cargo.toml
index c281f4752..ef5730188 100644
--- a/node/router/Cargo.toml
+++ b/node/router/Cargo.toml
@@ -42,6 +42,9 @@ version = "1"
 [dependencies.rand]
 version = "0.8"
 
+[dependencies.reqwest]
+version = "0.11"
+
 [dependencies.serde]
 version = "1"
 
diff --git a/node/router/src/lib.rs b/node/router/src/lib.rs
index 6e526c87c..4cb890d76 100644
--- a/node/router/src/lib.rs
+++ b/node/router/src/lib.rs
@@ -35,12 +35,13 @@ pub use peer::*;
 
 use snarkos_node_executor::{spawn_task, spawn_task_loop, Executor};
 use snarkos_node_messages::*;
-use snarkvm::prelude::{Network, PuzzleCommitment};
+use snarkvm::prelude::{Address, Network, PuzzleCommitment};
 
 use anyhow::Result;
 use indexmap::{IndexMap, IndexSet};
 use rand::{prelude::IteratorRandom, rngs::OsRng, Rng};
 use std::{
+    collections::HashMap,
     net::SocketAddr,
     sync::Arc,
     time::{Duration, Instant, SystemTime},
@@ -94,6 +95,8 @@ pub struct Router<N: Network> {
     router_sender: RouterSender<N>,
     /// The local IP of the node.
     local_ip: SocketAddr,
+    /// The address of the node.
+    address: Address<N>,
     /// The set of trusted peers.
     trusted_peers: Arc<IndexSet<SocketAddr>>,
     /// The map of connected peer IPs to their peer handlers.
@@ -145,6 +148,7 @@ impl<N: Network> Router<N> {
     /// Initializes a new `Router` instance.
     pub async fn new<E: Handshake + Inbound<N> + Outbound>(
         node_ip: SocketAddr,
+        address: Address<N>,
         trusted_peers: &[SocketAddr],
     ) -> Result<(Self, RouterReceiver<N>)> {
         // Initialize a new TCP listener at the given IP.
@@ -160,6 +164,7 @@ impl<N: Network> Router<N> {
         let router = Self {
             router_sender,
             local_ip,
+            address,
             trusted_peers: Arc::new(trusted_peers.iter().copied().collect()),
             connected_peers: Default::default(),
             candidate_peers: Default::default(),
@@ -178,6 +183,8 @@ impl<N: Network> Router<N> {
         router.initialize_listener::<E>(listener).await;
         // Initialize the heartbeat.
         router.initialize_heartbeat::<E>().await;
+        // Initialize the report.
+        router.initialize_report::<E>().await;
         // Initialize the GC.
         router.initialize_gc::<E>().await;
 
@@ -318,6 +325,26 @@ impl<N: Network> Router<N> {
         });
     }
 
+    /// Initialize a new instance of the report.
+    async fn initialize_report<E: Executor>(&self) {
+        let router = self.clone();
+        spawn_task_loop!(E, {
+            let url = "https://vm.aleo.org/testnet3/report";
+            loop {
+                // Prepare the report.
+                let mut report = HashMap::new();
+                report.insert("node_address".to_string(), router.address.to_string());
+                report.insert("node_type".to_string(), E::node_type().to_string());
+                // Transmit the report.
+                if reqwest::Client::new().post(url).json(&report).send().await.is_err() {
+                    warn!("Failed to send report");
+                }
+                // Sleep for a fixed duration in seconds.
+                tokio::time::sleep(Duration::from_secs(3600 * 6)).await;
+            }
+        });
+    }
+
     /// Initialize a new instance of the garbage collector.
     async fn initialize_gc<E: Executor>(&self) {
         let router = self.clone();
diff --git a/node/src/beacon/mod.rs b/node/src/beacon/mod.rs
index 93cb77fcf..8a5ad9c4e 100644
--- a/node/src/beacon/mod.rs
+++ b/node/src/beacon/mod.rs
@@ -75,12 +75,12 @@ impl<N: Network> Beacon<N> {
         // Initialize the consensus.
         let consensus = Consensus::new(ledger.clone())?;
         // Initialize the node router.
-        let (router, router_receiver) = Router::new::<Self>(node_ip, trusted_peers).await?;
+        let (router, router_receiver) = Router::new::<Self>(node_ip, account.address(), trusted_peers).await?;
         // Initialize the REST server.
         let rest = match rest_ip {
             Some(rest_ip) => Some(Arc::new(Rest::start(
                 rest_ip,
-                *account.address(),
+                account.address(),
                 Some(consensus.clone()),
                 ledger.clone(),
                 router.clone(),
@@ -164,7 +164,7 @@ impl<N: Network> NodeInterface<N> for Beacon<N> {
     }
 
     /// Returns the account address of the node.
-    fn address(&self) -> &Address<N> {
+    fn address(&self) -> Address<N> {
         self.account.address()
     }
 }
diff --git a/node/src/client/mod.rs b/node/src/client/mod.rs
index d9e412598..642f1ee79 100644
--- a/node/src/client/mod.rs
+++ b/node/src/client/mod.rs
@@ -47,7 +47,7 @@ impl<N: Network> Client<N> {
         // Initialize the node account.
         let account = Account::from(private_key)?;
         // Initialize the node router.
-        let (router, router_receiver) = Router::new::<Self>(node_ip, trusted_peers).await?;
+        let (router, router_receiver) = Router::new::<Self>(node_ip, account.address(), trusted_peers).await?;
         // Initialize the node.
         let node = Self {
             account,
@@ -94,7 +94,7 @@ impl<N: Network> NodeInterface<N> for Client<N> {
     }
 
     /// Returns the account address of the node.
-    fn address(&self) -> &Address<N> {
+    fn address(&self) -> Address<N> {
         self.account.address()
     }
 }
diff --git a/node/src/lib.rs b/node/src/lib.rs
index 52e779a04..d72008eaa 100644
--- a/node/src/lib.rs
+++ b/node/src/lib.rs
@@ -127,7 +127,7 @@ impl<N: Network> Node<N> {
     }
 
     /// Returns the account address of the node.
-    pub fn address(&self) -> &Address<N> {
+    pub fn address(&self) -> Address<N> {
         match self {
             Self::Beacon(node) => node.address(),
             Self::Validator(node) => node.address(),
diff --git a/node/src/prover/mod.rs b/node/src/prover/mod.rs
index 2ce696c45..6dde36318 100644
--- a/node/src/prover/mod.rs
+++ b/node/src/prover/mod.rs
@@ -56,7 +56,7 @@ impl<N: Network> Prover<N> {
         // Initialize the node account.
         let account = Account::from(private_key)?;
         // Initialize the node router.
-        let (router, router_receiver) = Router::new::<Self>(node_ip, trusted_peers).await?;
+        let (router, router_receiver) = Router::new::<Self>(node_ip, account.address(), trusted_peers).await?;
         // Load the coinbase puzzle.
         let coinbase_puzzle = CoinbasePuzzle::<N>::load()?;
         // Initialize the node.
@@ -109,7 +109,7 @@ impl<N: Network> NodeInterface<N> for Prover<N> {
     }
 
     /// Returns the account address of the node.
-    fn address(&self) -> &Address<N> {
+    fn address(&self) -> Address<N> {
         self.account.address()
     }
 }
@@ -208,7 +208,7 @@ impl<N: Network> Prover<N> {
                         // Construct a prover solution.
                         let prover_solution = match prover.coinbase_puzzle.prove(
                             &epoch_challenge,
-                            *prover.address(),
+                            prover.address(),
                             rand::thread_rng().gen(),
                         ) {
                             Ok(proof) => proof,
diff --git a/node/src/traits.rs b/node/src/traits.rs
index fef743d5d..0b2e8e2fe 100644
--- a/node/src/traits.rs
+++ b/node/src/traits.rs
@@ -32,5 +32,5 @@ pub trait NodeInterface<N: Network>: Executor {
     fn view_key(&self) -> &ViewKey<N>;
 
     /// Returns the account address of the node.
-    fn address(&self) -> &Address<N>;
+    fn address(&self) -> Address<N>;
 }
diff --git a/node/src/validator/mod.rs b/node/src/validator/mod.rs
index afd83fd09..7268e28cb 100644
--- a/node/src/validator/mod.rs
+++ b/node/src/validator/mod.rs
@@ -56,11 +56,11 @@ impl<N: Network> Validator<N> {
         // Initialize the ledger.
         let ledger = Ledger::load(genesis, dev)?;
         // Initialize the node router.
-        let (router, router_receiver) = Router::new::<Self>(node_ip, trusted_peers).await?;
+        let (router, router_receiver) = Router::new::<Self>(node_ip, account.address(), trusted_peers).await?;
         // Initialize the REST server.
         let rest = match rest_ip {
             Some(rest_ip) => {
-                Some(Arc::new(Rest::start(rest_ip, *account.address(), None, ledger.clone(), router.clone())?))
+                Some(Arc::new(Rest::start(rest_ip, account.address(), None, ledger.clone(), router.clone())?))
             }
             None => None,
         };
@@ -128,7 +128,7 @@ impl<N: Network> NodeInterface<N> for Validator<N> {
     }
 
     /// Returns the account address of the node.
-    fn address(&self) -> &Address<N> {
+    fn address(&self) -> Address<N> {
         self.account.address()
     }
 }
