diff --git a/.circleci/config.yml b/.circleci/config.yml
index 7f1fec8ad..ae3b2e113 100644
--- a/.circleci/config.yml
+++ b/.circleci/config.yml
@@ -112,7 +112,7 @@ commands:
       - setup_environment:
           cache_key: << parameters.cache_key >>
       - run:
-          no_output_timeout: 30m
+          no_output_timeout: 45m
           command: |
             cd << parameters.workspace_member >>
             cargo test -- --list --format terse | sed 's/: test//' > test_names.txt
diff --git a/Cargo.lock b/Cargo.lock
index e522d1ee7..89e1f3b1f 100644
--- a/Cargo.lock
+++ b/Cargo.lock
@@ -2293,6 +2293,7 @@ dependencies = [
  "regex",
  "serde",
  "serde_json",
+ "serial_test",
  "snarkvm-algorithms",
  "snarkvm-circuit",
  "snarkvm-console",
diff --git a/vm/compiler/Cargo.toml b/vm/compiler/Cargo.toml
index 26e6f26b8..878daf103 100644
--- a/vm/compiler/Cargo.toml
+++ b/vm/compiler/Cargo.toml
@@ -93,6 +93,9 @@ version = "1.3"
 [dev-dependencies.regex]
 version = "1.6"
 
+[dev-dependencies.serial_test]
+version = "0.9"
+
 [dev-dependencies.tracing-test]
 version = "0.2"
 
diff --git a/vm/compiler/src/ledger/mod.rs b/vm/compiler/src/ledger/mod.rs
index 2690f21e6..1381aeb0b 100644
--- a/vm/compiler/src/ledger/mod.rs
+++ b/vm/compiler/src/ledger/mod.rs
@@ -287,6 +287,16 @@ impl<N: Network, B: BlockStorage<N>, P: ProgramStorage<N>> Ledger<N, B, P> {
             }
         }
 
+        /* Program */
+
+        // Ensure that the ledger does not already contain the given program ID.
+        if let Transaction::Deploy(_, deployment, _) = &transaction {
+            let program_id = deployment.program_id();
+            if self.contains_program_id(program_id)? {
+                bail!("Program ID '{program_id}' already exists in the ledger")
+            }
+        }
+
         /* Metadata */
 
         // Ensure the ledger does not already contain a given transition public keys.
@@ -310,6 +320,7 @@ impl<N: Network, B: BlockStorage<N>, P: ProgramStorage<N>> Ledger<N, B, P> {
 
     /// Returns a candidate for the next block in the ledger.
     pub fn propose_next_block<R: Rng + CryptoRng>(&self, private_key: &PrivateKey<N>, rng: &mut R) -> Result<Block<N>> {
+        // TODO (raychu86): Add logic to transaction selection from the mempool.
         // Construct the transactions for the block.
         let transactions = self.memory_pool.values().collect::<Transactions<N>>();
 
diff --git a/vm/compiler/src/program/instruction/operation/assert.rs b/vm/compiler/src/program/instruction/operation/assert.rs
index e31dca199..c3c653dc4 100644
--- a/vm/compiler/src/program/instruction/operation/assert.rs
+++ b/vm/compiler/src/program/instruction/operation/assert.rs
@@ -505,6 +505,7 @@ mod tests {
     }
 
     #[test]
+    #[serial_test::serial]
     fn test_assert_eq_fails() {
         use rayon::prelude::*;
 
@@ -522,9 +523,9 @@ mod tests {
 
         literals_a.par_iter().for_each(|literal_a| {
             for literal_b in &literals_b {
-                for mode_a in &modes_a {
-                    for mode_b in &modes_b {
-                        if literal_a.to_type() != literal_b.to_type() {
+                if literal_a.to_type() != literal_b.to_type() {
+                    for mode_a in &modes_a {
+                        for mode_b in &modes_b {
                             // Check the operation fails.
                             check_assert_fails(opcode, literal_a, literal_b, mode_a, mode_b);
                         }
@@ -561,6 +562,7 @@ mod tests {
     }
 
     #[test]
+    #[serial_test::serial]
     fn test_assert_neq_fails() {
         use rayon::prelude::*;
 
@@ -578,9 +580,9 @@ mod tests {
 
         literals_a.par_iter().for_each(|literal_a| {
             for literal_b in &literals_b {
-                for mode_a in &modes_a {
-                    for mode_b in &modes_b {
-                        if literal_a.to_type() != literal_b.to_type() {
+                if literal_a.to_type() != literal_b.to_type() {
+                    for mode_a in &modes_a {
+                        for mode_b in &modes_b {
                             // Check the operation fails.
                             check_assert_fails(opcode, literal_a, literal_b, mode_a, mode_b);
                         }
diff --git a/vm/compiler/src/program/instruction/operation/is.rs b/vm/compiler/src/program/instruction/operation/is.rs
index 2815acc4b..259b7b4dd 100644
--- a/vm/compiler/src/program/instruction/operation/is.rs
+++ b/vm/compiler/src/program/instruction/operation/is.rs
@@ -551,6 +551,7 @@ mod tests {
     }
 
     #[test]
+    #[serial_test::serial]
     fn test_is_eq_fails() {
         use rayon::prelude::*;
 
@@ -566,17 +567,19 @@ mod tests {
         let modes_a = [/* circuit::Mode::Constant, */ circuit::Mode::Public, circuit::Mode::Private];
         let modes_b = [/* circuit::Mode::Constant, */ circuit::Mode::Public, circuit::Mode::Private];
 
-        literals_a.par_iter().for_each(|literal_a| {
-            for literal_b in &literals_b {
-                for mode_a in &modes_a {
-                    for mode_b in &modes_b {
-                        if literal_a.to_type() != literal_b.to_type() {
-                            // Check the operation fails.
-                            check_is_fails(opcode, literal_a, literal_b, mode_a, mode_b);
+        literals_a.par_iter().chunks(8).for_each(|literals_a| {
+            literals_a.iter().for_each(|literal_a| {
+                for literal_b in &literals_b {
+                    if literal_a.to_type() != literal_b.to_type() {
+                        for mode_a in &modes_a {
+                            for mode_b in &modes_b {
+                                // Check the operation fails.
+                                check_is_fails(opcode, literal_a, literal_b, mode_a, mode_b);
+                            }
                         }
                     }
                 }
-            }
+            })
         });
     }
 
@@ -607,6 +610,7 @@ mod tests {
     }
 
     #[test]
+    #[serial_test::serial]
     fn test_is_neq_fails() {
         use rayon::prelude::*;
 
@@ -622,17 +626,19 @@ mod tests {
         let modes_a = [/* circuit::Mode::Constant, */ circuit::Mode::Public, circuit::Mode::Private];
         let modes_b = [/* circuit::Mode::Constant, */ circuit::Mode::Public, circuit::Mode::Private];
 
-        literals_a.par_iter().for_each(|literal_a| {
-            for literal_b in &literals_b {
-                for mode_a in &modes_a {
-                    for mode_b in &modes_b {
-                        if literal_a.to_type() != literal_b.to_type() {
-                            // Check the operation fails.
-                            check_is_fails(opcode, literal_a, literal_b, mode_a, mode_b);
+        literals_a.par_iter().chunks(8).for_each(|literals_a| {
+            literals_a.iter().for_each(|literal_a| {
+                for literal_b in &literals_b {
+                    if literal_a.to_type() != literal_b.to_type() {
+                        for mode_a in &modes_a {
+                            for mode_b in &modes_b {
+                                // Check the operation fails.
+                                check_is_fails(opcode, literal_a, literal_b, mode_a, mode_b);
+                            }
                         }
                     }
                 }
-            }
+            })
         });
     }
 
diff --git a/vm/package/deploy.rs b/vm/package/deploy.rs
index 2274e7c6c..abd2b8589 100644
--- a/vm/package/deploy.rs
+++ b/vm/package/deploy.rs
@@ -26,17 +26,6 @@ pub struct DeployRequest<N: Network> {
     program_id: ProgramID<N>,
 }
 
-impl<N: Network> TryFrom<Deployment<N>> for DeployRequest<N> {
-    type Error = anyhow::Error;
-
-    fn try_from(deployment: Deployment<N>) -> Result<Self, Self::Error> {
-        let program_id = *deployment.program_id();
-        let address = program_id.to_address()?;
-
-        Ok(Self { deployment, address, program_id })
-    }
-}
-
 impl<N: Network> DeployRequest<N> {
     /// Sends the request to the given endpoint.
     pub fn new(deployment: Deployment<N>, address: Address<N>, program_id: ProgramID<N>) -> Self {
@@ -67,7 +56,7 @@ impl<N: Network> DeployRequest<N> {
 impl<N: Network> Serialize for DeployRequest<N> {
     /// Serializes the deploy request into string or bytes.
     fn serialize<S: Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
-        let mut request = serializer.serialize_struct("DeployRequest", 1)?;
+        let mut request = serializer.serialize_struct("DeployRequest", 3)?;
         // Serialize the deployment.
         request.serialize_field("deployment", &self.deployment)?;
         // Serialize the address.
@@ -140,22 +129,24 @@ impl<N: Network> Package<N> {
     ) -> Result<Deployment<N>> {
         // Retrieve the main program.
         let program = self.program();
-
-        // Retrieve the program ID.
+        // Retrieve the main program ID.
         let program_id = program.id();
 
+        // Retrieve the Aleo address of the deployment caller.
+        let caller = self.manifest_file().development_address();
+
         // Construct the process.
         let process = Process::<N>::load()?;
         let rng = &mut test_crypto_rng();
 
-        // Make the deploy
+        // Compute the deployment.
         let deployment = process.deploy::<A, _>(program, rng).unwrap();
 
         match endpoint {
             Some(ref endpoint) => {
-                // Prepare the request
-                let request = DeployRequest::try_from(deployment.clone())?;
-                // Load the proving and verifying keys.
+                // Construct the deploy request.
+                let request = DeployRequest::new(deployment.clone(), *caller, *program_id);
+                // Send the deploy request.
                 let response = request.send(endpoint)?;
                 // Ensure the program ID matches.
                 ensure!(
