diff --git a/Cargo.lock b/Cargo.lock
index f3a9804ea..72d338014 100644
--- a/Cargo.lock
+++ b/Cargo.lock
@@ -295,9 +295,9 @@ dependencies = [
 
 [[package]]
 name = "clap"
-version = "2.33.2"
+version = "2.33.3"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "10040cdf04294b565d9e0319955430099ec3813a64c952b86a41200ad714ae48"
+checksum = "37e58ac78573c40708d45522f0d80fa2f01cc4f9b4e2bf749807255454312002"
 dependencies = [
  "ansi_term",
  "atty",
@@ -1450,7 +1450,7 @@ dependencies = [
  "cloudabi 0.0.3",
  "libc",
  "redox_syscall",
- "smallvec 1.4.1",
+ "smallvec 1.4.2",
  "winapi 0.3.9",
 ]
 
@@ -1465,7 +1465,7 @@ dependencies = [
  "instant",
  "libc",
  "redox_syscall",
- "smallvec 1.4.1",
+ "smallvec 1.4.2",
  "winapi 0.3.9",
 ]
 
@@ -1911,9 +1911,9 @@ dependencies = [
 
 [[package]]
 name = "smallvec"
-version = "1.4.1"
+version = "1.4.2"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "3757cb9d89161a2f24e1cf78efa0c1fcff485d18e3f55e0aa3480824ddaa0f3f"
+checksum = "fbee7696b84bbf3d89a1c2eccff0850e3047ed46bfcd2e92c29a2d074d57e252"
 
 [[package]]
 name = "snarkos"
@@ -1956,7 +1956,7 @@ dependencies = [
  "rand_xorshift",
  "rayon",
  "sha2",
- "smallvec 1.4.1",
+ "smallvec 1.4.2",
  "snarkos-curves",
  "snarkos-errors",
  "snarkos-models",
@@ -1976,7 +1976,7 @@ dependencies = [
  "rand_xorshift",
  "rayon",
  "sha2",
- "smallvec 1.4.1",
+ "smallvec 1.4.2",
  "snarkos-algorithms",
  "snarkos-curves",
  "snarkos-errors",
@@ -2120,7 +2120,7 @@ dependencies = [
  "rand",
  "rand_xorshift",
  "serde",
- "smallvec 1.4.1",
+ "smallvec 1.4.2",
  "snarkos-errors",
  "snarkos-utilities",
 ]
diff --git a/Cargo.toml b/Cargo.toml
index e6c822017..79fced194 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -52,7 +52,7 @@ snarkos-rpc = { path = "./rpc", version = "0.8.0" }
 snarkos-storage = { path = "storage", version = "0.8.0" }
 snarkos-utilities = { path = "./utilities", version = "0.8.0" }
 
-clap = { version = "2.33.2" }
+clap = { version = "2.33.3" }
 colored = { version = "2.0" }
 dirs = { version = "3.0.1" }
 env_logger = { version = "0.7.1" }
diff --git a/SECURITY.md b/SECURITY.md
new file mode 100644
index 000000000..eb19bfb3c
--- /dev/null
+++ b/SECURITY.md
@@ -0,0 +1,17 @@
+# Security
+
+The following describes our procedure for addressing major and minor security concerns in snarkOS.
+
+## Testnet I
+
+As Aleo is currently in the prototype stage and does not operate a platform intended for production use,
+our security procedures are designed to promote public disclosure and quick security resolution.
+
+In preparation for the production stage, we will release new security guidelines and
+issue new procedures for addressing the disclosure of sensitive security vulnerabilities.
+
+### Reporting a Bug
+
+During Testnet I, all software bugs should be reported by filing a Github issue.
+
+If you are unsure and would like to reach out to us directly, please email security \_at\_ aleo.org to elaborate on the issue.
diff --git a/dpc/src/base_dpc/inner_circuit/inner_circuit_gadget.rs b/dpc/src/base_dpc/inner_circuit/inner_circuit_gadget.rs
index 3ad1a1a26..420271cf9 100644
--- a/dpc/src/base_dpc/inner_circuit/inner_circuit_gadget.rs
+++ b/dpc/src/base_dpc/inner_circuit/inner_circuit_gadget.rs
@@ -22,13 +22,13 @@ use snarkos_models::{
     dpc::Record,
     gadgets::{
         algorithms::{CRHGadget, CommitmentGadget, EncryptionGadget, PRFGadget, SignaturePublicKeyRandomizationGadget},
-        curves::FieldGadget,
+        curves::{FieldGadget, FpGadget},
         r1cs::ConstraintSystem,
         utilities::{
             alloc::AllocGadget,
             arithmetic::{add::Add, sub::Sub},
             boolean::Boolean,
-            eq::{ConditionalEqGadget, EqGadget, EvaluateEqGadget},
+            eq::{ConditionalEqGadget, EqGadget},
             int::{Int, Int64},
             uint::UInt8,
             ToBitsGadget,
@@ -43,6 +43,7 @@ use snarkos_utilities::{
     to_bytes,
 };
 
+use snarkos_models::gadgets::utilities::eq::NEqGadget;
 use std::ops::Mul;
 
 pub fn execute_inner_proof_gadget<C: BaseDPCComponents, CS: ConstraintSystem<C::InnerField>>(
@@ -1060,6 +1061,8 @@ where
             let u = C::InnerField::read(&to_bytes![u]?[..])?;
             let ua = C::InnerField::read(&to_bytes![ua]?[..])?;
 
+            let fp_zero = FpGadget::zero(encryption_cs.ns(|| "fpg zero"))?;
+
             for (i, (element, y_gadget)) in record_field_elements_gadgets
                 .iter()
                 .skip(1)
@@ -1105,20 +1108,31 @@ where
                     .0
                     .mul(encryption_cs.ns(|| format!("element_{} ^ 2", i)), &element.0)?;
 
-                let a_i_is_correct =
-                    element_squared.evaluate_equal(encryption_cs.ns(|| format!("element_squared == a_{}", i)), &a_i)?;
-                let b_i_is_correct =
-                    element_squared.evaluate_equal(encryption_cs.ns(|| format!("element_squared == b_{}", i)), &b_i)?;
+                // Enforce that either a_i == element_squared or b_i == element_squared
 
-                // Enforce that either a_i or b_i was valid
-                let single_valid_recovery = a_i_is_correct.evaluate_equal(
-                    encryption_cs.ns(|| format!("(element_squared == a_{}) == (element_squared == b_{})", i, i)),
-                    &b_i_is_correct,
+                let a_i_minus_element_squared = a_i.sub(
+                    encryption_cs.ns(|| format!("a_{} - element_{}^2", i, i)),
+                    &element_squared,
+                )?;
+                let b_i_minus_element_squared = b_i.sub(
+                    encryption_cs.ns(|| format!("b_{} - element_{}^2", i, i)),
+                    &element_squared,
                 )?;
-                single_valid_recovery.enforce_equal(
-                    encryption_cs.ns(|| format!("single_valid_recovery_{} == false", i)),
-                    &Boolean::Constant(false),
+
+                let expected_zero = a_i_minus_element_squared.mul(
+                    encryption_cs.ns(|| format!("(-a_{} - element_{}^2) * (-b_{} - element_{}^2)", i, i, i, i)),
+                    &b_i_minus_element_squared,
                 )?;
+
+                expected_zero.enforce_equal(
+                    encryption_cs.ns(|| format!("(-a_{} - element_{}^2) * (-b_{} - element_{}^2) == 0", i, i, i, i)),
+                    &fp_zero,
+                )?;
+
+                a_i.enforce_not_equal(encryption_cs.ns(|| format!("a_{} != b_{}", i, i)), &b_i)?;
+
+                element_squared
+                    .enforce_not_equal(encryption_cs.ns(|| format!("element_squared_{} != 0", i)), &fp_zero)?;
             }
 
             // *******************************************************************
diff --git a/dpc/src/base_dpc/record/record_encryption.rs b/dpc/src/base_dpc/record/record_encryption.rs
index 1272930f1..0fc6372d2 100644
--- a/dpc/src/base_dpc/record/record_encryption.rs
+++ b/dpc/src/base_dpc/record/record_encryption.rs
@@ -8,7 +8,7 @@ use snarkos_algorithms::encoding::Elligator2;
 use snarkos_errors::dpc::DPCError;
 use snarkos_models::{
     algorithms::{CommitmentScheme, EncryptionScheme, CRH},
-    curves::{AffineCurve, ModelParameters, ProjectiveCurve},
+    curves::{AffineCurve, ModelParameters, One, ProjectiveCurve},
     dpc::{DPCComponents, Record, RecordSerializerScheme},
 };
 use snarkos_objects::{AccountAddress, AccountViewKey};
@@ -44,9 +44,10 @@ impl<C: BaseDPCComponents> Default for RecordEncryptionGadgetComponents<C> {
     fn default() -> Self {
         // TODO (raychu86) Fix the lengths to be generic
         let record_encoding_length = 7;
+        let base_field_one = <C::EncryptionModelParameters as ModelParameters>::BaseField::one();
         let base_field_default = <C::EncryptionModelParameters as ModelParameters>::BaseField::default();
 
-        let record_field_elements = vec![base_field_default; record_encoding_length];
+        let record_field_elements = vec![base_field_one; record_encoding_length];
         let record_group_encoding = vec![(base_field_default, base_field_default); record_encoding_length];
 
         let ciphertext_selectors = vec![false; record_encoding_length + 1];
diff --git a/models/src/gadgets/curves/fp.rs b/models/src/gadgets/curves/fp.rs
index ce136420d..85af18731 100644
--- a/models/src/gadgets/curves/fp.rs
+++ b/models/src/gadgets/curves/fp.rs
@@ -11,7 +11,7 @@ use crate::{
         utilities::{
             alloc::AllocGadget,
             boolean::{AllocatedBit, Boolean},
-            eq::{ConditionalEqGadget, EqGadget, EvaluateEqGadget, NEqGadget},
+            eq::{ConditionalEqGadget, EqGadget, NEqGadget},
             select::{CondSelectGadget, ThreeBitCondNegLookupGadget, TwoBitLookupGadget},
             uint::unsigned_integer::{UInt, UInt8},
             ToBitsGadget,
@@ -260,16 +260,6 @@ impl<F: PrimeField> PartialEq for FpGadget<F> {
 
 impl<F: PrimeField> Eq for FpGadget<F> {}
 
-impl<F: PrimeField> EvaluateEqGadget<F> for FpGadget<F> {
-    fn evaluate_equal<CS: ConstraintSystem<F>>(&self, mut cs: CS, other: &Self) -> Result<Boolean, SynthesisError> {
-        let bool_option = self.value.and_then(|a| other.value.map(|b| a.eq(&b)));
-
-        Boolean::alloc(&mut cs.ns(|| "evaluate_equal"), || {
-            bool_option.ok_or(SynthesisError::AssignmentMissing)
-        })
-    }
-}
-
 impl<F: PrimeField> EqGadget<F> for FpGadget<F> {}
 
 impl<F: PrimeField> ConditionalEqGadget<F> for FpGadget<F> {
diff --git a/models/src/gadgets/utilities/boolean.rs b/models/src/gadgets/utilities/boolean.rs
index 620250590..ace8f8966 100644
--- a/models/src/gadgets/utilities/boolean.rs
+++ b/models/src/gadgets/utilities/boolean.rs
@@ -4,7 +4,7 @@ use crate::{
         r1cs::{Assignment, ConstraintSystem, ConstraintVar, LinearCombination, Variable},
         utilities::{
             alloc::AllocGadget,
-            eq::{ConditionalEqGadget, EqGadget, EvaluateEqGadget},
+            eq::{ConditionalEqGadget, EqGadget},
             select::CondSelectGadget,
             uint::UInt8,
             ToBytesGadget,
@@ -626,14 +626,6 @@ impl From<AllocatedBit> for Boolean {
     }
 }
 
-/// a == b = !(a XOR b)
-impl<F: PrimeField> EvaluateEqGadget<F> for Boolean {
-    fn evaluate_equal<CS: ConstraintSystem<F>>(&self, cs: CS, other: &Self) -> Result<Boolean, SynthesisError> {
-        let xor = Boolean::xor(cs, self, other)?;
-        Ok(xor.not())
-    }
-}
-
 impl<F: Field> AllocGadget<bool, F> for Boolean {
     fn alloc<Fn, T, CS: ConstraintSystem<F>>(cs: CS, value_gen: Fn) -> Result<Self, SynthesisError>
     where
diff --git a/models/src/gadgets/utilities/eq.rs b/models/src/gadgets/utilities/eq.rs
index e3b82043a..9c28c0504 100644
--- a/models/src/gadgets/utilities/eq.rs
+++ b/models/src/gadgets/utilities/eq.rs
@@ -7,10 +7,6 @@ use crate::{
 };
 use snarkos_errors::gadgets::SynthesisError;
 
-pub trait EvaluateEqGadget<F: Field> {
-    fn evaluate_equal<CS: ConstraintSystem<F>>(&self, cs: CS, other: &Self) -> Result<Boolean, SynthesisError>;
-}
-
 /// If `condition == 1`, then enforces that `self` and `other` are equal;
 /// otherwise, it doesn't enforce anything.
 pub trait ConditionalEqGadget<F: Field>: Eq {
diff --git a/models/src/gadgets/utilities/int/arithmetic/div.rs b/models/src/gadgets/utilities/int/arithmetic/div.rs
deleted file mode 100644
index 657aa7d91..000000000
--- a/models/src/gadgets/utilities/int/arithmetic/div.rs
+++ /dev/null
@@ -1,188 +0,0 @@
-use crate::{
-    curves::PrimeField,
-    gadgets::{
-        r1cs::ConstraintSystem,
-        utilities::{
-            alloc::AllocGadget,
-            arithmetic::{Add, Div, Neg, Sub},
-            bits::ComparatorGadget,
-            boolean::{AllocatedBit, Boolean},
-            eq::EvaluateEqGadget,
-            int::{Int, Int128, Int16, Int32, Int64, Int8},
-            select::CondSelectGadget,
-        },
-    },
-};
-use snarkos_errors::gadgets::SignedIntegerError;
-
-macro_rules! div_int_impl {
-    ($($gadget:ident)*) => ($(
-        impl<F: PrimeField> Div<F> for $gadget {
-            type ErrorType = SignedIntegerError;
-
-            fn div<CS: ConstraintSystem<F>>(
-                &self,
-                mut cs: CS,
-                other: &Self
-            ) -> Result<Self, Self::ErrorType> {
-                // N / D pseudocode:
-                //
-                // if D = 0 then error(DivisionByZeroException) end
-                //
-                // positive = msb(N) == msb(D) -- if msb's equal, return positive result
-                //
-                // Q := 0                  -- Initialize quotient and remainder to zero
-                // R := 0
-                //
-                // for i := n − 1 .. 0 do  -- Where n is number of bits in N
-                //   R := R << 1           -- Left-shift R by 1 bit
-                //   R(0) := N(i)          -- Set the least-significant bit of R equal to bit i of the numerator
-                //   if R ≥ D then
-                //     R := R − D
-                //     Q(i) := 1
-                //   end
-                // end
-                //
-                // if positive then           -- positive result
-                //    Q
-                // else
-                //    !Q                      -- negative result
-
-                if other.eq(&Self::constant(0 as <$gadget as Int>::IntegerType)) {
-                    return Err(SignedIntegerError::DivisionByZero);
-                }
-
-                let is_constant = Boolean::constant(Self::result_is_constant(&self, &other));
-
-                let allocated_true = Boolean::from(AllocatedBit::alloc(&mut cs.ns(|| "true"), || Ok(true)).unwrap());
-                let true_bit = Boolean::conditionally_select(
-                    &mut cs.ns(|| "constant_or_allocated_true"),
-                    &is_constant,
-                    &Boolean::constant(true),
-                    &allocated_true,
-                )?;
-
-                let allocated_one = Self::alloc(&mut cs.ns(|| "one"), || Ok(1 as <$gadget as Int>::IntegerType))?;
-                let one = Self::conditionally_select(
-                    &mut cs.ns(|| "constant_or_allocated_1"),
-                    &is_constant,
-                    &Self::constant(1 as <$gadget as Int>::IntegerType),
-                    &allocated_one,
-                )?;
-
-                let allocated_zero = Self::alloc(&mut cs.ns(|| "zero"), || Ok(0 as <$gadget as Int>::IntegerType))?;
-                let zero = Self::conditionally_select(
-                    &mut cs.ns(|| "constant_or_allocated_0"),
-                    &is_constant,
-                    &Self::constant(0 as <$gadget as Int>::IntegerType),
-                    &allocated_zero,
-                )?;
-
-                let self_is_zero = Boolean::Constant(self.eq(&Self::constant(0 as <$gadget as Int>::IntegerType)));
-
-                // If the most significant bits of both numbers are equal, the quotient will be positive
-                let a_msb = self.bits.last().unwrap();
-                let b_msb = other.bits.last().unwrap();
-                let positive = a_msb.evaluate_equal(cs.ns(|| "compare_msb"), &b_msb)?;
-
-                // Get the absolute value of each number
-                let a_comp = self.neg(&mut cs.ns(|| "a_neg"))?;
-                let a = Self::conditionally_select(
-                    &mut cs.ns(|| "a_abs"),
-                    &a_msb,
-                    &a_comp,
-                    &self
-                )?;
-
-                let b_comp = other.neg(&mut cs.ns(|| "b_neg"))?;
-                let b = Self::conditionally_select(
-                    &mut cs.ns(|| "b_abs"),
-                    &b_msb,
-                    &b_comp,
-                    &other,
-                )?;
-
-                let mut q = zero.clone();
-                let mut r = zero.clone();
-
-                let mut index = <$gadget as Int>::SIZE - 1 as usize;
-                let mut bit_value = (1 as <$gadget as Int>::IntegerType) << ((index - 1) as <$gadget as Int>::IntegerType);
-
-                for (i, bit) in a.bits.iter().rev().enumerate().skip(1) {
-
-                    // Left shift remainder by 1
-                    r = r.add(
-                        &mut cs.ns(|| format!("shift_left_{}", i)),
-                        &r
-                    )?;
-
-                    // Set the least-significant bit of remainder to bit i of the numerator
-                    let r_new = r.add(
-                        &mut cs.ns(|| format!("set_remainder_bit_{}", i)),
-                        &one.clone(),
-                    )?;
-
-                    r = Self::conditionally_select(
-                        &mut cs.ns(|| format!("increment_or_remainder_{}", i)),
-                        &bit,
-                        &r_new,
-                        &r
-                    )?;
-
-                    let can_sub = r.greater_than_or_equal(
-                        &mut cs.ns(|| format!("compare_remainder_{}", i)),
-                        &b
-                    )?;
-
-
-                    let sub = r.sub(
-                        &mut cs.ns(|| format!("subtract_divisor_{}", i)),
-                        &b
-                    );
-
-
-                    r = Self::conditionally_select(
-                        &mut cs.ns(|| format!("subtract_or_same_{}", i)),
-                        &can_sub,
-                        &sub?,
-                        &r
-                    )?;
-
-                    index = index - 1;
-
-                    let mut q_new = q.clone();
-                    q_new.bits[index] = true_bit.clone();
-                    q_new.value = Some(q_new.value.unwrap() + bit_value);
-
-                    bit_value = (bit_value >> 1);
-
-                    q = Self::conditionally_select(
-                        &mut cs.ns(|| format!("set_bit_or_same_{}", i)),
-                        &can_sub,
-                        &q_new,
-                        &q,
-                    )?;
-
-                }
-
-                let q_neg = q.neg(&mut cs.ns(|| "negate"))?;
-
-                q = Self::conditionally_select(
-                    &mut cs.ns(|| "positive or negative"),
-                    &positive,
-                    &q,
-                    &q_neg,
-                )?;
-
-                Ok(Self::conditionally_select(
-                    &mut cs.ns(|| "self_or_quotient"),
-                    &self_is_zero,
-                    self,
-                    &q
-                )?)
-            }
-        }
-    )*)
-}
-
-div_int_impl!(Int8 Int16 Int32 Int64 Int128);
diff --git a/models/src/gadgets/utilities/int/arithmetic/mod.rs b/models/src/gadgets/utilities/int/arithmetic/mod.rs
index 8ddd2e086..9306ff04d 100644
--- a/models/src/gadgets/utilities/int/arithmetic/mod.rs
+++ b/models/src/gadgets/utilities/int/arithmetic/mod.rs
@@ -2,9 +2,6 @@
 pub mod add;
 pub use self::add::*;
 
-pub mod div;
-pub use self::div::*;
-
 pub mod mul;
 pub use self::mul::*;
 
diff --git a/models/src/gadgets/utilities/int/relational/eq.rs b/models/src/gadgets/utilities/int/relational/eq.rs
index 39c3bf0bf..b792c14d0 100644
--- a/models/src/gadgets/utilities/int/relational/eq.rs
+++ b/models/src/gadgets/utilities/int/relational/eq.rs
@@ -1,42 +1,7 @@
-use crate::{
-    curves::PrimeField,
-    gadgets::{
-        r1cs::ConstraintSystem,
-        utilities::{
-            boolean::Boolean,
-            eq::EvaluateEqGadget,
-            int::{Int, Int128, Int16, Int32, Int64, Int8},
-        },
-    },
-};
-use snarkos_errors::gadgets::SynthesisError;
+use crate::gadgets::utilities::int::{Int128, Int16, Int32, Int64, Int8};
 
 macro_rules! eq_gadget_impl {
     ($($gadget: ident)*) => ($(
-        impl<F: PrimeField> EvaluateEqGadget<F> for $gadget {
-            fn evaluate_equal<CS: ConstraintSystem<F>>(
-                &self,
-                mut cs: CS,
-                other: &Self
-            ) -> Result<Boolean, SynthesisError> {
-                let mut result = Boolean::constant(true);
-                for (i, (a, b)) in self.bits.iter().zip(&other.bits).enumerate() {
-                    let equal = a.evaluate_equal(
-                        &mut cs.ns(|| format!("{} evaluate equality for {}-th bit", <$gadget as Int>::SIZE, i)),
-                        b,
-                    )?;
-
-                    result = Boolean::and(
-                        &mut cs.ns(|| format!("{} and result for {}-th bit", <$gadget as Int>::SIZE, i)),
-                        &equal,
-                        &result,
-                    )?;
-                }
-
-                Ok(result)
-            }
-        }
-
         impl PartialEq for $gadget {
             fn eq(&self, other: &Self) -> bool {
                 !self.value.is_none() && !other.value.is_none() && self.value == other.value
diff --git a/models/src/gadgets/utilities/int/tests/int128.rs b/models/src/gadgets/utilities/int/tests/int128.rs
index 5d3879929..65d31a279 100644
--- a/models/src/gadgets/utilities/int/tests/int128.rs
+++ b/models/src/gadgets/utilities/int/tests/int128.rs
@@ -280,60 +280,6 @@ fn test_int128_mul() {
     }
 }
 
-#[test]
-fn test_int128_div_constants() {
-    let mut rng = XorShiftRng::seed_from_u64(1231275789u64);
-
-    for _ in 0..10 {
-        let mut cs = TestConstraintSystem::<Fr>::new();
-
-        let a: i128 = rng.gen_range(-i128::MAX, i128::MAX);
-        let b: i128 = rng.gen_range(-i128::MAX, i128::MAX);
-
-        let expected = match a.checked_div(b) {
-            Some(valid) => valid,
-            None => continue,
-        };
-
-        let a_bit = Int128::constant(a);
-        let b_bit = Int128::constant(b);
-
-        let r = a_bit.div(cs.ns(|| "division"), &b_bit).unwrap();
-
-        assert!(r.value == Some(expected));
-
-        check_all_constant_bits(expected, r);
-    }
-}
-
-#[test]
-fn test_int128_div() {
-    let mut rng = XorShiftRng::seed_from_u64(1231275789u64);
-
-    for _ in 0..10 {
-        let mut cs = TestConstraintSystem::<Fr>::new();
-
-        let a: i128 = rng.gen_range(-i128::MAX, i128::MAX);
-        let b: i128 = rng.gen_range(-i128::MAX, i128::MAX);
-
-        let expected = match a.checked_div(b) {
-            Some(valid) => valid,
-            None => continue,
-        };
-
-        let a_bit = Int128::alloc(cs.ns(|| "a_bit"), || Ok(a)).unwrap();
-        let b_bit = Int128::alloc(cs.ns(|| "b_bit"), || Ok(b)).unwrap();
-
-        let r = a_bit.div(cs.ns(|| "division"), &b_bit).unwrap();
-
-        assert!(cs.is_satisfied());
-
-        assert!(r.value == Some(expected));
-
-        check_all_allocated_bits(expected, r);
-    }
-}
-
 #[test]
 fn test_int128_pow_constants() {
     let mut rng = XorShiftRng::seed_from_u64(1231275789u64);
diff --git a/models/src/gadgets/utilities/int/tests/int16.rs b/models/src/gadgets/utilities/int/tests/int16.rs
index f2e2506f5..d9d1552fa 100644
--- a/models/src/gadgets/utilities/int/tests/int16.rs
+++ b/models/src/gadgets/utilities/int/tests/int16.rs
@@ -280,60 +280,6 @@ fn test_int16_mul() {
     }
 }
 
-#[test]
-fn test_int16_div_constants() {
-    let mut rng = XorShiftRng::seed_from_u64(1231275789u64);
-
-    for _ in 0..1000 {
-        let mut cs = TestConstraintSystem::<Fr>::new();
-
-        let a: i16 = rng.gen_range(-i16::MAX, i16::MAX);
-        let b: i16 = rng.gen_range(-i16::MAX, i16::MAX);
-
-        let expected = match a.checked_div(b) {
-            Some(valid) => valid,
-            None => continue,
-        };
-
-        let a_bit = Int16::constant(a);
-        let b_bit = Int16::constant(b);
-
-        let r = a_bit.div(cs.ns(|| "division"), &b_bit).unwrap();
-
-        assert!(r.value == Some(expected));
-
-        check_all_constant_bits(expected, r);
-    }
-}
-
-#[test]
-fn test_int16_div() {
-    let mut rng = XorShiftRng::seed_from_u64(1231275789u64);
-
-    for _ in 0..100 {
-        let mut cs = TestConstraintSystem::<Fr>::new();
-
-        let a: i16 = rng.gen_range(-i16::MAX, i16::MAX);
-        let b: i16 = rng.gen_range(-i16::MAX, i16::MAX);
-
-        let expected = match a.checked_div(b) {
-            Some(valid) => valid,
-            None => continue,
-        };
-
-        let a_bit = Int16::alloc(cs.ns(|| "a_bit"), || Ok(a)).unwrap();
-        let b_bit = Int16::alloc(cs.ns(|| "b_bit"), || Ok(b)).unwrap();
-
-        let r = a_bit.div(cs.ns(|| "division"), &b_bit).unwrap();
-
-        assert!(cs.is_satisfied());
-
-        assert!(r.value == Some(expected));
-
-        check_all_allocated_bits(expected, r);
-    }
-}
-
 #[test]
 fn test_int16_pow_constants() {
     let mut rng = XorShiftRng::seed_from_u64(1231275789u64);
diff --git a/models/src/gadgets/utilities/int/tests/int32.rs b/models/src/gadgets/utilities/int/tests/int32.rs
index 7b96c677f..357fb4134 100644
--- a/models/src/gadgets/utilities/int/tests/int32.rs
+++ b/models/src/gadgets/utilities/int/tests/int32.rs
@@ -280,60 +280,6 @@ fn test_int32_mul() {
     }
 }
 
-#[test]
-fn test_int32_div_constants() {
-    let mut rng = XorShiftRng::seed_from_u64(1231275789u64);
-
-    for _ in 0..1000 {
-        let mut cs = TestConstraintSystem::<Fr>::new();
-
-        let a: i32 = rng.gen_range(-i32::MAX, i32::MAX);
-        let b: i32 = rng.gen_range(-i32::MAX, i32::MAX);
-
-        let expected = match a.checked_div(b) {
-            Some(valid) => valid,
-            None => continue,
-        };
-
-        let a_bit = Int32::constant(a);
-        let b_bit = Int32::constant(b);
-
-        let r = a_bit.div(cs.ns(|| "division"), &b_bit).unwrap();
-
-        assert!(r.value == Some(expected));
-
-        check_all_constant_bits(expected, r);
-    }
-}
-
-#[test]
-fn test_int32_div() {
-    let mut rng = XorShiftRng::seed_from_u64(1231275789u64);
-
-    for _ in 0..100 {
-        let mut cs = TestConstraintSystem::<Fr>::new();
-
-        let a: i32 = rng.gen_range(-i32::MAX, i32::MAX);
-        let b: i32 = rng.gen_range(-i32::MAX, i32::MAX);
-
-        let expected = match a.checked_div(b) {
-            Some(valid) => valid,
-            None => continue,
-        };
-
-        let a_bit = Int32::alloc(cs.ns(|| "a_bit"), || Ok(a)).unwrap();
-        let b_bit = Int32::alloc(cs.ns(|| "b_bit"), || Ok(b)).unwrap();
-
-        let r = a_bit.div(cs.ns(|| "division"), &b_bit).unwrap();
-
-        assert!(cs.is_satisfied());
-
-        assert!(r.value == Some(expected));
-
-        check_all_allocated_bits(expected, r);
-    }
-}
-
 #[test]
 fn test_int32_pow_constants() {
     let mut rng = XorShiftRng::seed_from_u64(1231275789u64);
diff --git a/models/src/gadgets/utilities/int/tests/int64.rs b/models/src/gadgets/utilities/int/tests/int64.rs
index 06741c3dd..8a89d144e 100644
--- a/models/src/gadgets/utilities/int/tests/int64.rs
+++ b/models/src/gadgets/utilities/int/tests/int64.rs
@@ -280,60 +280,6 @@ fn test_int64_mul() {
     }
 }
 
-#[test]
-fn test_int64_div_constants() {
-    let mut rng = XorShiftRng::seed_from_u64(1231275789u64);
-
-    for _ in 0..10 {
-        let mut cs = TestConstraintSystem::<Fr>::new();
-
-        let a: i64 = rng.gen_range(-i64::MAX, i64::MAX);
-        let b: i64 = rng.gen_range(-i64::MAX, i64::MAX);
-
-        let expected = match a.checked_div(b) {
-            Some(valid) => valid,
-            None => continue,
-        };
-
-        let a_bit = Int64::constant(a);
-        let b_bit = Int64::constant(b);
-
-        let r = a_bit.div(cs.ns(|| "division"), &b_bit).unwrap();
-
-        assert!(r.value == Some(expected));
-
-        check_all_constant_bits(expected, r);
-    }
-}
-
-#[test]
-fn test_int64_div() {
-    let mut rng = XorShiftRng::seed_from_u64(1231275789u64);
-
-    for _ in 0..10 {
-        let mut cs = TestConstraintSystem::<Fr>::new();
-
-        let a: i64 = rng.gen_range(-i64::MAX, i64::MAX);
-        let b: i64 = rng.gen_range(-i64::MAX, i64::MAX);
-
-        let expected = match a.checked_div(b) {
-            Some(valid) => valid,
-            None => continue,
-        };
-
-        let a_bit = Int64::alloc(cs.ns(|| "a_bit"), || Ok(a)).unwrap();
-        let b_bit = Int64::alloc(cs.ns(|| "b_bit"), || Ok(b)).unwrap();
-
-        let r = a_bit.div(cs.ns(|| "division"), &b_bit).unwrap();
-
-        assert!(cs.is_satisfied());
-
-        assert!(r.value == Some(expected));
-
-        check_all_allocated_bits(expected, r);
-    }
-}
-
 #[test]
 fn test_int64_pow_constants() {
     let mut rng = XorShiftRng::seed_from_u64(1231275789u64);
diff --git a/models/src/gadgets/utilities/int/tests/int8.rs b/models/src/gadgets/utilities/int/tests/int8.rs
index 523e37a2b..f8c057da9 100644
--- a/models/src/gadgets/utilities/int/tests/int8.rs
+++ b/models/src/gadgets/utilities/int/tests/int8.rs
@@ -280,60 +280,6 @@ fn test_int8_mul() {
     }
 }
 
-#[test]
-fn test_int8_div_constants() {
-    let mut rng = XorShiftRng::seed_from_u64(1231275789u64);
-
-    for _ in 0..1000 {
-        let mut cs = TestConstraintSystem::<Fr>::new();
-
-        let a: i8 = rng.gen_range(-i8::MAX, i8::MAX);
-        let b: i8 = rng.gen_range(-i8::MAX, i8::MAX);
-
-        let expected = match a.checked_div(b) {
-            Some(valid) => valid,
-            None => continue,
-        };
-
-        let a_bit = Int8::constant(a);
-        let b_bit = Int8::constant(b);
-
-        let r = a_bit.div(cs.ns(|| "division"), &b_bit).unwrap();
-
-        assert!(r.value == Some(expected));
-
-        check_all_constant_bits(expected, r);
-    }
-}
-
-#[test]
-fn test_int8_div() {
-    let mut rng = XorShiftRng::seed_from_u64(1231275789u64);
-
-    for _ in 0..100 {
-        let mut cs = TestConstraintSystem::<Fr>::new();
-
-        let a: i8 = rng.gen_range(-i8::MAX, i8::MAX);
-        let b: i8 = rng.gen_range(-i8::MAX, i8::MAX);
-
-        let expected = match a.checked_div(b) {
-            Some(valid) => valid,
-            None => continue,
-        };
-
-        let a_bit = Int8::alloc(cs.ns(|| "a_bit"), || Ok(a)).unwrap();
-        let b_bit = Int8::alloc(cs.ns(|| "b_bit"), || Ok(b)).unwrap();
-
-        let r = a_bit.div(cs.ns(|| "division"), &b_bit).unwrap();
-
-        assert!(cs.is_satisfied());
-
-        assert!(r.value == Some(expected));
-
-        check_all_allocated_bits(expected, r);
-    }
-}
-
 #[test]
 fn test_int8_pow_constants() {
     let mut rng = XorShiftRng::seed_from_u64(1231275789u64);
diff --git a/models/src/gadgets/utilities/uint/macros.rs b/models/src/gadgets/utilities/uint/macros.rs
index a85b19f2a..63482a018 100644
--- a/models/src/gadgets/utilities/uint/macros.rs
+++ b/models/src/gadgets/utilities/uint/macros.rs
@@ -595,30 +595,6 @@ macro_rules! uint_impl {
             }
         }
 
-        impl<F: PrimeField> EvaluateEqGadget<F> for $name {
-            fn evaluate_equal<CS: ConstraintSystem<F>>(
-                &self,
-                mut cs: CS,
-                other: &Self,
-            ) -> Result<Boolean, SynthesisError> {
-                let mut result = Boolean::constant(true);
-                for (i, (a, b)) in self.bits.iter().zip(&other.bits).enumerate() {
-                    let equal = a.evaluate_equal(
-                        &mut cs.ns(|| format!("{} evaluate equality for {}-th bit", $size, i)),
-                        b,
-                    )?;
-
-                    result = Boolean::and(
-                        &mut cs.ns(|| format!("{} and result for {}-th bit", $size, i)),
-                        &equal,
-                        &result,
-                    )?;
-                }
-
-                Ok(result)
-            }
-        }
-
         impl<F: Field> EqGadget<F> for $name {}
 
         impl<F: Field> ConditionalEqGadget<F> for $name {
diff --git a/models/src/gadgets/utilities/uint/uint128.rs b/models/src/gadgets/utilities/uint/uint128.rs
index 14a532525..56aac6353 100644
--- a/models/src/gadgets/utilities/uint/uint128.rs
+++ b/models/src/gadgets/utilities/uint/uint128.rs
@@ -5,7 +5,7 @@ use crate::{
         utilities::{
             alloc::AllocGadget,
             boolean::{AllocatedBit, Boolean},
-            eq::{ConditionalEqGadget, EqGadget, EvaluateEqGadget},
+            eq::{ConditionalEqGadget, EqGadget},
             select::CondSelectGadget,
             uint::unsigned_integer::{UInt, UInt8},
             ToBytesGadget,
@@ -621,23 +621,6 @@ impl PartialOrd for UInt128 {
     }
 }
 
-impl<F: PrimeField> EvaluateEqGadget<F> for UInt128 {
-    fn evaluate_equal<CS: ConstraintSystem<F>>(&self, mut cs: CS, other: &Self) -> Result<Boolean, SynthesisError> {
-        let mut result = Boolean::constant(true);
-        for (i, (a, b)) in self.bits.iter().zip(&other.bits).enumerate() {
-            let equal = a.evaluate_equal(&mut cs.ns(|| format!("u128 evaluate equality for {}-th bit", i)), b)?;
-
-            result = Boolean::and(
-                &mut cs.ns(|| format!("u128 and result for {}-th bit", i)),
-                &equal,
-                &result,
-            )?;
-        }
-
-        Ok(result)
-    }
-}
-
 impl<F: Field> EqGadget<F> for UInt128 {}
 
 impl<F: Field> ConditionalEqGadget<F> for UInt128 {
diff --git a/models/src/gadgets/utilities/uint/unsigned_integer.rs b/models/src/gadgets/utilities/uint/unsigned_integer.rs
index a1ffb2ca0..ff7439a7f 100644
--- a/models/src/gadgets/utilities/uint/unsigned_integer.rs
+++ b/models/src/gadgets/utilities/uint/unsigned_integer.rs
@@ -6,7 +6,7 @@ use crate::{
         utilities::{
             alloc::AllocGadget,
             boolean::{AllocatedBit, Boolean},
-            eq::{ConditionalEqGadget, EqGadget, EvaluateEqGadget},
+            eq::{ConditionalEqGadget, EqGadget},
             select::CondSelectGadget,
             ToBitsGadget,
             ToBytesGadget,
diff --git a/parameters/src/genesis/block_header/block_header.genesis b/parameters/src/genesis/block_header/block_header.genesis
index 6115e4b52..3bc2d14cd 100644
Binary files a/parameters/src/genesis/block_header/block_header.genesis and b/parameters/src/genesis/block_header/block_header.genesis differ
diff --git a/parameters/src/genesis/transaction_1/transaction_1.genesis b/parameters/src/genesis/transaction_1/transaction_1.genesis
index 8f5761df8..a909e565e 100644
Binary files a/parameters/src/genesis/transaction_1/transaction_1.genesis and b/parameters/src/genesis/transaction_1/transaction_1.genesis differ
diff --git a/parameters/src/params/encrypted_record_crh.checksum b/parameters/src/params/encrypted_record_crh.checksum
index db652874e..8343067e1 100644
--- a/parameters/src/params/encrypted_record_crh.checksum
+++ b/parameters/src/params/encrypted_record_crh.checksum
@@ -1 +1 @@
-aedf2dd53f3d035e48ca2777075cd46a30677fa89c0919dbcc45dc6c7edd931e
\ No newline at end of file
+7f59e71cd8b56e24ad838e3c08eec6416e2fbd77e82d845cbfdbd61956ce9c56
\ No newline at end of file
diff --git a/parameters/src/params/encrypted_record_crh.params b/parameters/src/params/encrypted_record_crh.params
index ae878ef9d..b3bcebe3e 100644
Binary files a/parameters/src/params/encrypted_record_crh.params and b/parameters/src/params/encrypted_record_crh.params differ
diff --git a/parameters/src/params/inner_snark_pk.checksum b/parameters/src/params/inner_snark_pk.checksum
index 4612973fb..b725496d8 100644
--- a/parameters/src/params/inner_snark_pk.checksum
+++ b/parameters/src/params/inner_snark_pk.checksum
@@ -1 +1 @@
-54e627e21fa9e33e87148763811e1026722b4366080cc04bf2e28d587cdc61e3
\ No newline at end of file
+246e4488bfa2e5a2fc8abf39f17fc3868fad30357df7adb5913f55afbd36ec52
\ No newline at end of file
diff --git a/parameters/src/params/inner_snark_vk.checksum b/parameters/src/params/inner_snark_vk.checksum
index ea305c3a4..16d60315b 100644
--- a/parameters/src/params/inner_snark_vk.checksum
+++ b/parameters/src/params/inner_snark_vk.checksum
@@ -1 +1 @@
-8e346167c1201844c6ba7f694a8e4f268780cf24709bb9ca73330303de21f741
\ No newline at end of file
+91627ce464a5220ab355b0ee42de20980c1861a05f4ad73b5fa5c48e8a8d7bc8
\ No newline at end of file
diff --git a/parameters/src/params/inner_snark_vk.params b/parameters/src/params/inner_snark_vk.params
index 1f4513d9b..944554111 100644
Binary files a/parameters/src/params/inner_snark_vk.params and b/parameters/src/params/inner_snark_vk.params differ
diff --git a/parameters/src/params/noop_program_snark_pk.checksum b/parameters/src/params/noop_program_snark_pk.checksum
index 3960d73c9..cc11b2252 100644
--- a/parameters/src/params/noop_program_snark_pk.checksum
+++ b/parameters/src/params/noop_program_snark_pk.checksum
@@ -1 +1 @@
-5cf77889c9c2a505a17525fdcf88c8f01cda2793874ebfbd3176d177c83173b8
\ No newline at end of file
+93483af1eef72df28f932ac044a9964f676985e4d45a9847ecc15abd2d56775c
\ No newline at end of file
diff --git a/parameters/src/params/noop_program_snark_pk.params b/parameters/src/params/noop_program_snark_pk.params
index ad0b363c8..f930e32e5 100644
Binary files a/parameters/src/params/noop_program_snark_pk.params and b/parameters/src/params/noop_program_snark_pk.params differ
diff --git a/parameters/src/params/noop_program_snark_vk.checksum b/parameters/src/params/noop_program_snark_vk.checksum
index be7b04cf3..cfb73de6c 100644
--- a/parameters/src/params/noop_program_snark_vk.checksum
+++ b/parameters/src/params/noop_program_snark_vk.checksum
@@ -1 +1 @@
-cc6ce55a37b4afbdcdb9caddd88cea0d2ade92e61f5e82ea2668e416e1b8ddcd
\ No newline at end of file
+19f6e9b4381528675b64c2cec248b733afefe30fab70a8e5686c59bc641932e2
\ No newline at end of file
diff --git a/parameters/src/params/noop_program_snark_vk.params b/parameters/src/params/noop_program_snark_vk.params
index b3f63ebdf..aa15c627f 100644
Binary files a/parameters/src/params/noop_program_snark_vk.params and b/parameters/src/params/noop_program_snark_vk.params differ
diff --git a/parameters/src/params/outer_snark_pk.checksum b/parameters/src/params/outer_snark_pk.checksum
index 0aab1297d..b54166522 100644
--- a/parameters/src/params/outer_snark_pk.checksum
+++ b/parameters/src/params/outer_snark_pk.checksum
@@ -1 +1 @@
-c71402d27a0c2b2dd09b7cfbae8e66922a9a7032cad6d5bfdae92490bc230b45
\ No newline at end of file
+4bb86f4316c804801ac6957a99e276f2869445fb25f0a14ade043ffb83c9276d
\ No newline at end of file
diff --git a/parameters/src/params/outer_snark_vk.checksum b/parameters/src/params/outer_snark_vk.checksum
index 3679e9f5a..96fd143e3 100644
--- a/parameters/src/params/outer_snark_vk.checksum
+++ b/parameters/src/params/outer_snark_vk.checksum
@@ -1 +1 @@
-de5d85464e38247f0f7e2cd3f9f95dbe21dbc31d9caa3b7ff1abf2a57fc257e0
\ No newline at end of file
+fc277bea57a9143de2e84260938691ed698d63486e69070c785c7f364d2c3550
\ No newline at end of file
diff --git a/parameters/src/params/outer_snark_vk.params b/parameters/src/params/outer_snark_vk.params
index 3cd19db75..368c524b5 100644
Binary files a/parameters/src/params/outer_snark_vk.params and b/parameters/src/params/outer_snark_vk.params differ
diff --git a/parameters/src/params/program_vk_crh.checksum b/parameters/src/params/program_vk_crh.checksum
index 1d0a6c2ac..76211147f 100644
--- a/parameters/src/params/program_vk_crh.checksum
+++ b/parameters/src/params/program_vk_crh.checksum
@@ -1 +1 @@
-e79e063c5e24878d19942e0a8a55afa32250ad4d437cb7e8c1edc172b7bf941f
\ No newline at end of file
+79809e7962789d7d7a95c2b4a8f3cc3f3318195d61249f204cb220f49cfaffb6
\ No newline at end of file
diff --git a/parameters/src/params/program_vk_crh.params b/parameters/src/params/program_vk_crh.params
index 2648e13d1..88d32fb65 100644
Binary files a/parameters/src/params/program_vk_crh.params and b/parameters/src/params/program_vk_crh.params differ
diff --git a/testing/src/consensus/test_data b/testing/src/consensus/test_data
index a0b1fa712..f8d372d8c 100644
Binary files a/testing/src/consensus/test_data and b/testing/src/consensus/test_data differ
