diff --git a/.circleci/config.yml b/.circleci/config.yml
index 21641255c..2d1df5321 100644
--- a/.circleci/config.yml
+++ b/.circleci/config.yml
@@ -1,5 +1,16 @@
 version: 2.1
 
+parameters:
+  large:
+    type: string
+    default: aleonet/large
+  xlarge:
+    type: string
+    default: aleonet/xlarge
+  twoxlarge:
+    type: string
+    default: aleonet/2xlarge
+
 orbs:
   windows: circleci/windows@5.0
 
@@ -160,7 +171,7 @@ jobs:
   integration:
     docker:
       - image: cimg/rust:1.76.0 # Attention - Change the MSRV in Cargo.toml and rust-toolchain as well
-    resource_class: anf/2xlarge
+    resource_class: << pipeline.parameters.twoxlarge >>
     steps:
       - run_serial_long:
           workspace_member: .integration
@@ -169,7 +180,7 @@ jobs:
   snarkos:
     docker:
       - image: cimg/rust:1.76.0 # Attention - Change the MSRV in Cargo.toml and rust-toolchain as well
-    resource_class: anf/2xlarge
+    resource_class: << pipeline.parameters.twoxlarge >>
     steps:
       - run_serial:
           workspace_member: .
@@ -178,7 +189,7 @@ jobs:
   account:
     docker:
       - image: cimg/rust:1.76.0 # Attention - Change the MSRV in Cargo.toml and rust-toolchain as well
-    resource_class: anf/xlarge
+    resource_class: << pipeline.parameters.xlarge >>
     steps:
       - run_serial:
           workspace_member: account
@@ -187,7 +198,7 @@ jobs:
   cli:
     docker:
       - image: cimg/rust:1.76.0 # Attention - Change the MSRV in Cargo.toml and rust-toolchain as well
-    resource_class: anf/xlarge
+    resource_class: << pipeline.parameters.xlarge >>
     steps:
       - run_serial:
           workspace_member: cli
@@ -196,7 +207,7 @@ jobs:
   display:
     docker:
       - image: cimg/rust:1.76.0 # Attention - Change the MSRV in Cargo.toml and rust-toolchain as well
-    resource_class: anf/xlarge
+    resource_class: << pipeline.parameters.xlarge >>
     steps:
       - run_serial:
           workspace_member: display
@@ -205,7 +216,7 @@ jobs:
   node:
     docker:
       - image: cimg/rust:1.76.0 # Attention - Change the MSRV in Cargo.toml and rust-toolchain as well
-    resource_class: anf/2xlarge
+    resource_class: << pipeline.parameters.twoxlarge >>
     steps:
       - run_serial:
           workspace_member: node
@@ -214,7 +225,7 @@ jobs:
   node-bft:
     docker:
       - image: cimg/rust:1.76.0 # Attention - Change the MSRV in Cargo.toml and rust-toolchain as well
-    resource_class: anf/2xlarge
+    resource_class: << pipeline.parameters.twoxlarge >>
     steps:
       - run_serial:
           workspace_member: node/bft
@@ -223,7 +234,7 @@ jobs:
   node-bft-events:
     docker:
       - image: cimg/rust:1.76.0 # Attention - Change the MSRV in Cargo.toml and rust-toolchain as well
-    resource_class: anf/2xlarge
+    resource_class: << pipeline.parameters.twoxlarge >>
     steps:
       - run_serial:
           workspace_member: node/bft/events
@@ -232,7 +243,7 @@ jobs:
   node-bft-ledger-service:
     docker:
       - image: cimg/rust:1.76.0 # Attention - Change the MSRV in Cargo.toml and rust-toolchain as well
-    resource_class: anf/xlarge
+    resource_class: << pipeline.parameters.xlarge >>
     steps:
       - run_serial:
           workspace_member: node/bft/ledger-service
@@ -241,7 +252,7 @@ jobs:
   node-bft-storage-service:
     docker:
       - image: cimg/rust:1.76.0 # Attention - Change the MSRV in Cargo.toml and rust-toolchain as well
-    resource_class: anf/xlarge
+    resource_class: << pipeline.parameters.xlarge >>
     steps:
       - run_serial:
           workspace_member: node/bft/storage-service
@@ -250,7 +261,7 @@ jobs:
   node-cdn:
     docker:
       - image: cimg/rust:1.76.0 # Attention - Change the MSRV in Cargo.toml and rust-toolchain as well
-    resource_class: anf/xlarge
+    resource_class: << pipeline.parameters.xlarge >>
     steps:
       - run_serial:
           workspace_member: node/cdn
@@ -259,7 +270,7 @@ jobs:
   node-consensus:
     docker:
       - image: cimg/rust:1.76.0 # Attention - Change the MSRV in Cargo.toml and rust-toolchain as well
-    resource_class: anf/2xlarge
+    resource_class: << pipeline.parameters.twoxlarge >>
     steps:
       - run_serial:
           workspace_member: node/consensus
@@ -268,7 +279,7 @@ jobs:
   node-rest:
     docker:
       - image: cimg/rust:1.76.0 # Attention - Change the MSRV in Cargo.toml and rust-toolchain as well
-    resource_class: anf/xlarge
+    resource_class: << pipeline.parameters.xlarge >>
     steps:
       - run_serial:
           workspace_member: node/rest
@@ -277,7 +288,7 @@ jobs:
   node-router:
     docker:
       - image: cimg/rust:1.76.0 # Attention - Change the MSRV in Cargo.toml and rust-toolchain as well
-    resource_class: anf/xlarge
+    resource_class: << pipeline.parameters.xlarge >>
     steps:
       - run_serial:
           workspace_member: node/router
@@ -286,7 +297,7 @@ jobs:
   node-router-messages:
     docker:
       - image: cimg/rust:1.76.0 # Attention - Change the MSRV in Cargo.toml and rust-toolchain as well
-    resource_class: anf/xlarge
+    resource_class: << pipeline.parameters.xlarge >>
     steps:
       - run_serial:
           workspace_member: node/router/messages
@@ -295,7 +306,7 @@ jobs:
   node-sync:
     docker:
       - image: cimg/rust:1.76.0 # Attention - Change the MSRV in Cargo.toml and rust-toolchain as well
-    resource_class: anf/xlarge
+    resource_class: << pipeline.parameters.xlarge >>
     steps:
       - run_serial:
           workspace_member: node/sync
@@ -304,7 +315,7 @@ jobs:
   node-sync-communication-service:
     docker:
       - image: cimg/rust:1.76.0 # Attention - Change the MSRV in Cargo.toml and rust-toolchain as well
-    resource_class: anf/xlarge
+    resource_class: << pipeline.parameters.xlarge >>
     steps:
       - run_serial:
           workspace_member: node/sync/communication-service
@@ -313,7 +324,7 @@ jobs:
   node-sync-locators:
     docker:
       - image: cimg/rust:1.76.0 # Attention - Change the MSRV in Cargo.toml and rust-toolchain as well
-    resource_class: anf/xlarge
+    resource_class: << pipeline.parameters.xlarge >>
     steps:
       - run_serial:
           workspace_member: node/sync/locators
@@ -322,7 +333,7 @@ jobs:
   node-tcp:
     docker:
       - image: cimg/rust:1.76.0 # Attention - Change the MSRV in Cargo.toml and rust-toolchain as well
-    resource_class: anf/xlarge
+    resource_class: << pipeline.parameters.xlarge >>
     steps:
       - run_serial:
           workspace_member: node/tcp
@@ -331,7 +342,7 @@ jobs:
   check-fmt:
     docker:
       - image: cimg/rust:1.76.0 # Attention - Change the MSRV in Cargo.toml and rust-toolchain as well
-    resource_class: anf/xlarge
+    resource_class: << pipeline.parameters.xlarge >>
     steps:
       - checkout
       - install_rust_nightly
@@ -347,7 +358,7 @@ jobs:
   check-clippy:
     docker:
       - image: cimg/rust:1.76.0 # Attention - Change the MSRV in Cargo.toml and rust-toolchain as well
-    resource_class: anf/2xlarge
+    resource_class: << pipeline.parameters.twoxlarge >>
     steps:
       - checkout
       - setup_environment:
@@ -421,4 +432,4 @@ workflows:
               only:
                 - mainnet
     jobs:
-      - integration
\ No newline at end of file
+      - integration
diff --git a/.devnet/.analytics/analytics.js b/.devnet/.analytics/analytics.js
index 601be7a19..02fcbb0d1 100644
--- a/.devnet/.analytics/analytics.js
+++ b/.devnet/.analytics/analytics.js
@@ -115,7 +115,7 @@ async function calculateRoundsInBlocks(baseUrl, latestHeight) {
     }
 }
 
-async function checkBlockHash(blockHeight) {
+async function checkBlockHash(networkName, blockHeight) {
     const numNodes = await getAWSNodeCount();
     console.log(`Detected ${numNodes} AWS nodes... \n`);
 
@@ -125,7 +125,7 @@ async function checkBlockHash(blockHeight) {
         // Get the IP address of the AWS node
         const ipAddress = await getIPAddress(awsNodeName);
         // Define the base URL for the node
-        const baseUrl = `http://${ipAddress}:3030/mainnet/block`;
+        const baseUrl = `http://${ipAddress}:3030/${networkName}/block`;
 
         // Fetch the block data
         const blockData = await fetchBlockData(baseUrl, blockHeight);
@@ -138,7 +138,23 @@ async function checkBlockHash(blockHeight) {
 }
 
 // Main function to fetch block metrics
-async function fetchBlockMetrics(metricType, optionalBlockHeight) {
+async function fetchBlockMetrics(metricType, optionalBlockHeight, networkID) {
+    // Derive the network name based on the network ID.
+    let networkName;
+    switch (networkID) {
+        case 0:
+            networkName = "mainnet";
+            break;
+        case 1:
+            networkName = "testnet";
+            break;
+        case 2:
+            networkName = "canary";
+            break;
+        default:
+            throw new Error(`Unknown network ID (${networkID})`);
+    }
+
     // Function to get the latest block height
     async function getLatestBlockHeight(baseUrl) {
         try {
@@ -157,7 +173,7 @@ async function fetchBlockMetrics(metricType, optionalBlockHeight) {
     // Get the IP address of the AWS node
     const ipAddress = await getIPAddress(awsNodeName);
     // Define the base URL for the node.
-    const baseUrl = `http://${ipAddress}:3030/mainnet/block`;
+    const baseUrl = `http://${ipAddress}:3030/${networkName}/block`;
 
     console.log(`${dimStart}IP Address: ${ipAddress}${dimEnd}`);
     console.log(`${dimStart}Base URL: ${baseUrl}${dimEnd}`);
@@ -175,7 +191,7 @@ async function fetchBlockMetrics(metricType, optionalBlockHeight) {
     } else if (metricType === 'roundsInBlocks') {
         calculateRoundsInBlocks(baseUrl, latestHeight);
     } else if (metricType === 'checkBlockHash' && optionalBlockHeight) {
-        checkBlockHash(optionalBlockHeight);
+        checkBlockHash(networkName, optionalBlockHeight);
     } else {
         console.error('Invalid metric type. Supported types: "averageBlockTime" or "roundsInBlocks".');
     }
@@ -196,6 +212,13 @@ async function main() {
                 describe: 'Block height to examine for checkBlockHash metric',
                 type: 'number',
             },
+            'network-id': {
+                alias: 'n',
+                describe: 'Network ID to fetch block metrics from',
+                demandOption: true,
+                type: 'number',
+                choices: [0, 1, 2],
+            }
         })
         .check((argv) => {
             // Check if metric-type is checkBlockHash and block-height is provided
@@ -207,7 +230,7 @@ async function main() {
         .argv;
 
     // Fetch and output the specified block metric
-    fetchBlockMetrics(argv['metric-type'], argv['block-height']);
+    fetchBlockMetrics(argv['metric-type'], argv['block-height'], argv['network-id']);
 }
 
 // Run the main function
diff --git a/.devnet/analytics.sh b/.devnet/analytics.sh
index 9284f3a7e..8e12f2988 100755
--- a/.devnet/analytics.sh
+++ b/.devnet/analytics.sh
@@ -14,6 +14,17 @@ else
   echo "Node.js dependencies already installed."
 fi
 
+# Prompt the user to specify a network ID
+while true; do
+  echo "Please specify a network ID (0 for mainnet, 1 for testnet, 2 for canary):"
+  read networkID
+  if [[ $networkID == 0 || $networkID == 1 || $networkID == 2 ]]; then
+    break
+  else
+    echo "Invalid network ID. Please enter 0 or 1."
+  fi
+done
+
 # Prompt the user to select a metric type
 PS3="Select a metric type: "
 options=("Average Block Time" "Rounds in Blocks" "Check Block Hash" "Quit")
@@ -22,12 +33,12 @@ do
   case $opt in
     "Average Block Time")
       echo ""
-      node analytics.js --metric-type averageBlockTime
+      node analytics.js --metric-type averageBlockTime --network-id $networkID
       break
       ;;
     "Rounds in Blocks")
       echo ""
-      node analytics.js --metric-type roundsInBlocks
+      node analytics.js --metric-type roundsInBlocks --network-id $networkID
       break
       ;;
     "Check Block Hash")
@@ -39,7 +50,7 @@ do
         echo "Error: Block height must be a positive integer."
         exit 1
       fi
-      node analytics.js --metric-type checkBlockHash --block-height "$blockHeight"
+      node analytics.js --metric-type checkBlockHash --block-height "$blockHeight" --network-id $networkID
       break
       ;;
     "Quit")
diff --git a/.devnet/clean.sh b/.devnet/clean.sh
index d06f8d596..888356e7a 100755
--- a/.devnet/clean.sh
+++ b/.devnet/clean.sh
@@ -12,6 +12,12 @@ NUM_INSTANCES="${NUM_INSTANCES:-$NODE_ID}"
 
 echo "Using $NUM_INSTANCES AWS EC2 instances for querying."
 
+# Read the network ID from user or use a default value of 1
+read -p "Enter the network ID (mainnet = 0, testnet = 1, canary = 2) (default: 1): " NETWORK_ID
+NETWORK_ID=${NETWORK_ID:-1}
+
+echo "Using network ID $NETWORK_ID."
+
 # Define a function to terminate the tmux session on a node
 terminate_tmux_session() {
   local NODE_ID=$1
@@ -24,7 +30,7 @@ terminate_tmux_session() {
     cd \$WORKSPACE
 
     tmux kill-session -t snarkos-session
-    snarkos clean --dev $NODE_ID
+    snarkos clean --dev $NODE_ID --network $NETWORK_ID
 
     exit  # Exit root user
 EOF
diff --git a/.devnet/start.sh b/.devnet/start.sh
index 77294f080..a155b4104 100755
--- a/.devnet/start.sh
+++ b/.devnet/start.sh
@@ -12,6 +12,12 @@ NUM_INSTANCES="${NUM_INSTANCES:-$NODE_ID}"
 
 echo "Using $NUM_INSTANCES AWS EC2 instances for querying."
 
+# Read the network ID from user or use a default value of 1
+read -p "Enter the network ID (mainnet = 0, testnet = 1, canary = 2) (default: 1): " NETWORK_ID
+NETWORK_ID=${NETWORK_ID:-1}
+
+echo "Using network ID $NETWORK_ID."
+
 # Read the verbosity level from the user (default: 1)
 read -p "Enter the verbosity level (default: 1): " VERBOSITY
 VERBOSITY="${VERBOSITY:-1}"
@@ -37,7 +43,7 @@ start_snarkos_in_tmux() {
     tmux new-session -d -s snarkos-session
 
     # Send the snarkOS start command to the tmux session with the NODE_ID
-    tmux send-keys -t "snarkos-session" "snarkos start --nodisplay --bft 0.0.0.0:5000 --rest 0.0.0.0:3030 --allow-external-peers --peers $NODE_IP:4130 --validators $NODE_IP:5000 --rest-rps 1000 --verbosity $VERBOSITY --dev $NODE_ID --dev-num-validators $NUM_INSTANCES --validator --metrics" C-m
+    tmux send-keys -t "snarkos-session" "snarkos start --nodisplay --bft 0.0.0.0:5000 --rest 0.0.0.0:3030 --allow-external-peers --peers $NODE_IP:4130 --validators $NODE_IP:5000 --rest-rps 1000 --verbosity $VERBOSITY --network $NETWORK_ID --dev $NODE_ID --dev-num-validators $NUM_INSTANCES --validator --metrics" C-m
 
     exit  # Exit root user
 EOF
diff --git a/.devnet/start_sync_test.sh b/.devnet/start_sync_test.sh
index 6ca5fbb77..41c35dacf 100755
--- a/.devnet/start_sync_test.sh
+++ b/.devnet/start_sync_test.sh
@@ -12,6 +12,12 @@ NUM_INSTANCES="${NUM_INSTANCES:-$NODE_ID}"
 
 echo "Using $NUM_INSTANCES AWS EC2 instances for querying."
 
+# Read the network ID from user or use a default value of 1
+read -p "Enter the network ID (mainnet = 0, testnet = 1, canary = 2) (default: 1): " NETWORK_ID
+NETWORK_ID=${NETWORK_ID:-1}
+
+echo "Using network ID $NETWORK_ID."
+
 # Read the verbosity level from the user (default: 1)
 read -p "Enter the verbosity level (default: 1): " VERBOSITY
 VERBOSITY="${VERBOSITY:-1}"
@@ -37,7 +43,7 @@ start_snarkos_in_tmux() {
     tmux new-session -d -s snarkos-session
 
     # Send the snarkOS start command to the tmux session with the NODE_ID
-    tmux send-keys -t "snarkos-session" "snarkos start --client --nocdn --nodisplay --rest 0.0.0.0:3030 --node 0.0.0.0:4130 --verbosity 4 --metrics --logfile "/tmp/snarkos-syncing-range-3.log" --peers 167.71.249.65:4130,157.245.218.195:4130,167.71.249.55:4130" C-m
+    tmux send-keys -t "snarkos-session" "snarkos start --client --nocdn --nodisplay --rest 0.0.0.0:3030 --node 0.0.0.0:4130 --verbosity 4 --network $NETWORK_ID --metrics --logfile "/tmp/snarkos-syncing-range-3.log" --peers 167.71.249.65:4130,157.245.218.195:4130,167.71.249.55:4130" C-m
 
     exit  # Exit root user
 EOF
diff --git a/.gitignore b/.gitignore
index f58c794d3..98b96026a 100644
--- a/.gitignore
+++ b/.gitignore
@@ -4,6 +4,7 @@
 wasm/Cargo.lock
 **/build
 **.ledger-*
+**.current-proposal-cache-*
 **.logs-*
 validator-*
 **.bft-storage-*/
diff --git a/Cargo.lock b/Cargo.lock
index d069cedea..02707b96d 100644
--- a/Cargo.lock
+++ b/Cargo.lock
@@ -3205,6 +3205,7 @@ dependencies = [
  "parking_lot",
  "peak_alloc",
  "rand",
+ "rayon",
  "reqwest",
  "serde",
  "snarkos-account",
diff --git a/node/bft/src/helpers/pending.rs b/node/bft/src/helpers/pending.rs
index cf68a39e5..ca07ee090 100644
--- a/node/bft/src/helpers/pending.rs
+++ b/node/bft/src/helpers/pending.rs
@@ -85,6 +85,16 @@ impl<T: Copy + Clone + PartialEq + Eq + Hash, V: Clone> Pending<T, V> {
         self.pending.read().get(&item.into()).map_or(false, |peer_ips| peer_ips.contains_key(&peer_ip))
     }
 
+    /// Returns `true` if the pending queue contains the specified `item` for the specified `peer IP` with a sent request.
+    pub fn contains_peer_with_sent_request(&self, item: impl Into<T>, peer_ip: SocketAddr) -> bool {
+        self.pending.read().get(&item.into()).map_or(false, |peer_ips| {
+            peer_ips
+                .get(&peer_ip)
+                .map(|callbacks| callbacks.iter().any(|(_, _, request_sent)| *request_sent))
+                .unwrap_or(false)
+        })
+    }
+
     /// Returns the peer IPs for the specified `item`.
     pub fn get_peers(&self, item: impl Into<T>) -> Option<HashSet<SocketAddr>> {
         self.pending.read().get(&item.into()).map(|map| map.keys().cloned().collect())
@@ -254,24 +264,29 @@ mod tests {
         let solution_id_1 = TransmissionID::Solution(rng.gen::<u64>().into());
         let solution_id_2 = TransmissionID::Solution(rng.gen::<u64>().into());
         let solution_id_3 = TransmissionID::Solution(rng.gen::<u64>().into());
+        let solution_id_4 = TransmissionID::Solution(rng.gen::<u64>().into());
 
         // Initialize the SocketAddrs.
         let addr_1 = SocketAddr::from(([127, 0, 0, 1], 1234));
         let addr_2 = SocketAddr::from(([127, 0, 0, 1], 2345));
         let addr_3 = SocketAddr::from(([127, 0, 0, 1], 3456));
+        let addr_4 = SocketAddr::from(([127, 0, 0, 1], 4567));
 
         // Initialize the callbacks.
         let (callback_sender_1, _) = oneshot::channel();
         let (callback_sender_2, _) = oneshot::channel();
         let (callback_sender_3, _) = oneshot::channel();
+        let (callback_sender_4, _) = oneshot::channel();
 
         // Insert the solution IDs.
         assert!(pending.insert(solution_id_1, addr_1, Some((callback_sender_1, true))));
         assert!(pending.insert(solution_id_2, addr_2, Some((callback_sender_2, true))));
         assert!(pending.insert(solution_id_3, addr_3, Some((callback_sender_3, true))));
+        // Add a callback without a sent request.
+        assert!(pending.insert(solution_id_4, addr_4, Some((callback_sender_4, false))));
 
         // Check the number of SocketAddrs.
-        assert_eq!(pending.len(), 3);
+        assert_eq!(pending.len(), 4);
         assert!(!pending.is_empty());
 
         // Check the items.
@@ -282,7 +297,12 @@ mod tests {
             let id = ids[i];
             assert!(pending.contains(id));
             assert!(pending.contains_peer(id, peers[i]));
+            assert!(pending.contains_peer_with_sent_request(id, peers[i]));
         }
+        // Ensure the last item does not have a sent request.
+        assert!(pending.contains_peer(solution_id_4, addr_4));
+        assert!(!pending.contains_peer_with_sent_request(solution_id_4, addr_4));
+
         let unknown_id = TransmissionID::Solution(rng.gen::<u64>().into());
         assert!(!pending.contains(unknown_id));
 
@@ -290,12 +310,14 @@ mod tests {
         assert_eq!(pending.get_peers(solution_id_1), Some(HashSet::from([addr_1])));
         assert_eq!(pending.get_peers(solution_id_2), Some(HashSet::from([addr_2])));
         assert_eq!(pending.get_peers(solution_id_3), Some(HashSet::from([addr_3])));
+        assert_eq!(pending.get_peers(solution_id_4), Some(HashSet::from([addr_4])));
         assert_eq!(pending.get_peers(unknown_id), None);
 
         // Check remove.
         assert!(pending.remove(solution_id_1, None).is_some());
         assert!(pending.remove(solution_id_2, None).is_some());
         assert!(pending.remove(solution_id_3, None).is_some());
+        assert!(pending.remove(solution_id_4, None).is_some());
         assert!(pending.remove(unknown_id, None).is_none());
 
         // Check empty again.
diff --git a/node/bft/src/helpers/ready.rs b/node/bft/src/helpers/ready.rs
index 6fa581ccd..7a366dff8 100644
--- a/node/bft/src/helpers/ready.rs
+++ b/node/bft/src/helpers/ready.rs
@@ -126,6 +126,14 @@ impl<N: Network> Ready<N> {
         // Drain the transmission IDs.
         transmissions.drain(range).collect::<IndexMap<_, _>>()
     }
+
+    /// Clears all solutions from the ready queue.
+    pub(crate) fn clear_solutions(&self) {
+        // Acquire the write lock.
+        let mut transmissions = self.transmissions.write();
+        // Remove all solutions.
+        transmissions.retain(|id, _| !matches!(id, TransmissionID::Solution(..)));
+    }
 }
 
 #[cfg(test)]
diff --git a/node/bft/src/helpers/storage.rs b/node/bft/src/helpers/storage.rs
index bda8bce76..abb93310d 100644
--- a/node/bft/src/helpers/storage.rs
+++ b/node/bft/src/helpers/storage.rs
@@ -788,6 +788,7 @@ impl<N: Network> Storage<N> {
     /// Inserts the given `certificate` into storage.
     ///
     /// Note: Do NOT use this in production. This is for **testing only**.
+    #[cfg(test)]
     #[doc(hidden)]
     pub(crate) fn testing_only_insert_certificate_testing_only(&self, certificate: BatchCertificate<N>) {
         // Retrieve the round.
diff --git a/node/bft/src/primary.rs b/node/bft/src/primary.rs
index 330e7422c..cb7e289d6 100644
--- a/node/bft/src/primary.rs
+++ b/node/bft/src/primary.rs
@@ -161,7 +161,8 @@ impl<N: Network> Primary<N> {
                         // We use a dummy IP because the node should not need to request from any peers.
                         // The storage should have stored all the transmissions. If not, we simply
                         // skip the certificate.
-                        if let Err(err) = self.sync_with_certificate_from_peer(DUMMY_SELF_IP, certificate).await {
+                        if let Err(err) = self.sync_with_certificate_from_peer::<true>(DUMMY_SELF_IP, certificate).await
+                        {
                             warn!("Failed to load stored certificate {} from proposal cache - {err}", fmt_id(batch_id));
                         }
                     }
@@ -319,6 +320,13 @@ impl<N: Network> Primary<N> {
     }
 }
 
+impl<N: Network> Primary<N> {
+    /// Clears the worker solutions.
+    pub fn clear_worker_solutions(&self) {
+        self.workers.iter().for_each(Worker::clear_solutions);
+    }
+}
+
 impl<N: Network> Primary<N> {
     /// Proposes the batch for the current round.
     ///
@@ -695,7 +703,7 @@ impl<N: Network> Primary<N> {
         }
 
         // If the peer is ahead, use the batch header to sync up to the peer.
-        let mut transmissions = self.sync_with_batch_header_from_peer(peer_ip, &batch_header).await?;
+        let mut transmissions = self.sync_with_batch_header_from_peer::<false>(peer_ip, &batch_header).await?;
 
         // Check that the transmission ids match and are not fee transactions.
         if let Err(err) = cfg_iter_mut!(transmissions).try_for_each(|(transmission_id, transmission)| {
@@ -907,7 +915,7 @@ impl<N: Network> Primary<N> {
         }
 
         // Store the certificate, after ensuring it is valid.
-        self.sync_with_certificate_from_peer(peer_ip, certificate).await?;
+        self.sync_with_certificate_from_peer::<false>(peer_ip, certificate).await?;
 
         // If there are enough certificates to reach quorum threshold for the certificate round,
         // then proceed to advance to the next round.
@@ -1426,7 +1434,7 @@ impl<N: Network> Primary<N> {
     ///   - Ensure the previous certificates have reached the quorum threshold.
     ///   - Ensure we have not already signed the batch ID.
     #[async_recursion::async_recursion]
-    async fn sync_with_certificate_from_peer(
+    async fn sync_with_certificate_from_peer<const IS_SYNCING: bool>(
         &self,
         peer_ip: SocketAddr,
         certificate: BatchCertificate<N>,
@@ -1445,8 +1453,16 @@ impl<N: Network> Primary<N> {
             return Ok(());
         }
 
+        // If node is not in sync mode and the node is not synced. Then return an error.
+        if !IS_SYNCING && !self.is_synced() {
+            bail!(
+                "Failed to process certificate `{}` at round {batch_round} from '{peer_ip}' (node is syncing)",
+                fmt_id(certificate.id())
+            );
+        }
+
         // If the peer is ahead, use the batch header to sync up to the peer.
-        let missing_transmissions = self.sync_with_batch_header_from_peer(peer_ip, batch_header).await?;
+        let missing_transmissions = self.sync_with_batch_header_from_peer::<IS_SYNCING>(peer_ip, batch_header).await?;
 
         // Check if the certificate needs to be stored.
         if !self.storage.contains_certificate(certificate.id()) {
@@ -1467,7 +1483,7 @@ impl<N: Network> Primary<N> {
     }
 
     /// Recursively syncs using the given batch header.
-    async fn sync_with_batch_header_from_peer(
+    async fn sync_with_batch_header_from_peer<const IS_SYNCING: bool>(
         &self,
         peer_ip: SocketAddr,
         batch_header: &BatchHeader<N>,
@@ -1480,6 +1496,14 @@ impl<N: Network> Primary<N> {
             bail!("Round {batch_round} is too far in the past")
         }
 
+        // If node is not in sync mode and the node is not synced. Then return an error.
+        if !IS_SYNCING && !self.is_synced() {
+            bail!(
+                "Failed to process batch header `{}` at round {batch_round} from '{peer_ip}' (node is syncing)",
+                fmt_id(batch_header.batch_id())
+            );
+        }
+
         // Determine if quorum threshold is reached on the batch round.
         let is_quorum_threshold_reached = {
             let certificates = self.storage.get_certificates_for_round(batch_round);
@@ -1515,7 +1539,7 @@ impl<N: Network> Primary<N> {
         // Iterate through the missing previous certificates.
         for batch_certificate in missing_previous_certificates {
             // Store the batch certificate (recursively fetching any missing previous certificates).
-            self.sync_with_certificate_from_peer(peer_ip, batch_certificate).await?;
+            self.sync_with_certificate_from_peer::<IS_SYNCING>(peer_ip, batch_certificate).await?;
         }
         Ok(missing_transmissions)
     }
@@ -2070,6 +2094,8 @@ mod tests {
 
         // The author must be known to resolver to pass propose checks.
         primary.gateway.resolver().insert_peer(peer_ip, peer_ip, peer_account.1.address());
+        // The primary must be considered synced.
+        primary.sync.block_sync().try_block_sync(&primary.gateway.clone()).await;
 
         // Try to process the batch proposal from the peer, should succeed.
         assert!(
@@ -2077,6 +2103,39 @@ mod tests {
         );
     }
 
+    #[tokio::test]
+    async fn test_batch_propose_from_peer_when_not_synced() {
+        let mut rng = TestRng::default();
+        let (primary, accounts) = primary_without_handlers(&mut rng).await;
+
+        // Create a valid proposal with an author that isn't the primary.
+        let round = 1;
+        let peer_account = &accounts[1];
+        let peer_ip = peer_account.0;
+        let timestamp = now() + MIN_BATCH_DELAY_IN_SECS as i64;
+        let proposal = create_test_proposal(
+            &peer_account.1,
+            primary.ledger.current_committee().unwrap(),
+            round,
+            Default::default(),
+            timestamp,
+            &mut rng,
+        );
+
+        // Make sure the primary is aware of the transmissions in the proposal.
+        for (transmission_id, transmission) in proposal.transmissions() {
+            primary.workers[0].process_transmission_from_peer(peer_ip, *transmission_id, transmission.clone())
+        }
+
+        // The author must be known to resolver to pass propose checks.
+        primary.gateway.resolver().insert_peer(peer_ip, peer_ip, peer_account.1.address());
+
+        // Try to process the batch proposal from the peer, should fail.
+        assert!(
+            primary.process_batch_propose_from_peer(peer_ip, (*proposal.batch_header()).clone().into()).await.is_err()
+        );
+    }
+
     #[tokio::test]
     async fn test_batch_propose_from_peer_in_round() {
         let round = 2;
@@ -2106,6 +2165,8 @@ mod tests {
 
         // The author must be known to resolver to pass propose checks.
         primary.gateway.resolver().insert_peer(peer_ip, peer_ip, peer_account.1.address());
+        // The primary must be considered synced.
+        primary.sync.block_sync().try_block_sync(&primary.gateway.clone()).await;
 
         // Try to process the batch proposal from the peer, should succeed.
         primary.process_batch_propose_from_peer(peer_ip, (*proposal.batch_header()).clone().into()).await.unwrap();
@@ -2137,6 +2198,8 @@ mod tests {
 
         // The author must be known to resolver to pass propose checks.
         primary.gateway.resolver().insert_peer(peer_ip, peer_ip, peer_account.1.address());
+        // The primary must be considered synced.
+        primary.sync.block_sync().try_block_sync(&primary.gateway.clone()).await;
 
         // Try to process the batch proposal from the peer, should error.
         assert!(
@@ -2179,6 +2242,8 @@ mod tests {
 
         // The author must be known to resolver to pass propose checks.
         primary.gateway.resolver().insert_peer(peer_ip, peer_ip, peer_account.1.address());
+        // The primary must be considered synced.
+        primary.sync.block_sync().try_block_sync(&primary.gateway.clone()).await;
 
         // Try to process the batch proposal from the peer, should error.
         assert!(
@@ -2221,6 +2286,8 @@ mod tests {
 
         // The author must be known to resolver to pass propose checks.
         primary.gateway.resolver().insert_peer(peer_ip, peer_ip, peer_account.1.address());
+        // The primary must be considered synced.
+        primary.sync.block_sync().try_block_sync(&primary.gateway.clone()).await;
 
         // Try to process the batch proposal from the peer, should error.
         assert!(
@@ -2257,6 +2324,8 @@ mod tests {
 
         // The author must be known to resolver to pass propose checks.
         primary.gateway.resolver().insert_peer(peer_ip, peer_ip, peer_account.1.address());
+        // The primary must be considered synced.
+        primary.sync.block_sync().try_block_sync(&primary.gateway.clone()).await;
 
         // Try to process the batch proposal from the peer, should error.
         assert!(
diff --git a/node/bft/src/sync/mod.rs b/node/bft/src/sync/mod.rs
index e5dc84731..ade20eccc 100644
--- a/node/bft/src/sync/mod.rs
+++ b/node/bft/src/sync/mod.rs
@@ -557,6 +557,13 @@ impl<N: Network> Sync<N> {
     pub fn get_block_locators(&self) -> Result<BlockLocators<N>> {
         self.block_sync.get_block_locators()
     }
+
+    /// Returns the block sync module.
+    #[cfg(test)]
+    #[doc(hidden)]
+    pub(super) fn block_sync(&self) -> &BlockSync<N> {
+        &self.block_sync
+    }
 }
 
 // Methods to assist with fetching batch certificates from peers.
@@ -571,10 +578,13 @@ impl<N: Network> Sync<N> {
         let (callback_sender, callback_receiver) = oneshot::channel();
         // Determine how many sent requests are pending.
         let num_sent_requests = self.pending.num_sent_requests(certificate_id);
+        // Determine if we've already sent a request to the peer.
+        let contains_peer_with_sent_request = self.pending.contains_peer_with_sent_request(certificate_id, peer_ip);
         // Determine the maximum number of redundant requests.
         let num_redundant_requests = max_redundant_requests(self.ledger.clone(), self.storage.current_round());
         // Determine if we should send a certificate request to the peer.
-        let should_send_request = num_sent_requests < num_redundant_requests;
+        // We send at most `num_redundant_requests` requests and each peer can only receive one request at a time.
+        let should_send_request = num_sent_requests < num_redundant_requests && !contains_peer_with_sent_request;
 
         // Insert the certificate ID into the pending queue.
         self.pending.insert(certificate_id, peer_ip, Some((callback_sender, should_send_request)));
diff --git a/node/bft/src/worker.rs b/node/bft/src/worker.rs
index fa9f837c7..a2e7f85e3 100644
--- a/node/bft/src/worker.rs
+++ b/node/bft/src/worker.rs
@@ -151,6 +151,13 @@ impl<N: Network> Worker<N> {
     }
 }
 
+impl<N: Network> Worker<N> {
+    /// Clears the solutions from the ready queue.
+    pub(super) fn clear_solutions(&self) {
+        self.ready.clear_solutions()
+    }
+}
+
 impl<N: Network> Worker<N> {
     /// Returns `true` if the transmission ID exists in the ready queue, proposed batch, storage, or ledger.
     pub fn contains_transmission(&self, transmission_id: impl Into<TransmissionID<N>>) -> bool {
@@ -410,10 +417,13 @@ impl<N: Network> Worker<N> {
         let (callback_sender, callback_receiver) = oneshot::channel();
         // Determine how many sent requests are pending.
         let num_sent_requests = self.pending.num_sent_requests(transmission_id);
+        // Determine if we've already sent a request to the peer.
+        let contains_peer_with_sent_request = self.pending.contains_peer_with_sent_request(transmission_id, peer_ip);
         // Determine the maximum number of redundant requests.
         let num_redundant_requests = max_redundant_requests(self.ledger.clone(), self.storage.current_round());
         // Determine if we should send a transmission request to the peer.
-        let should_send_request = num_sent_requests < num_redundant_requests;
+        // We send at most `num_redundant_requests` requests and each peer can only receive one request at a time.
+        let should_send_request = num_sent_requests < num_redundant_requests && !contains_peer_with_sent_request;
 
         // Insert the transmission ID into the pending queue.
         self.pending.insert(transmission_id, peer_ip, Some((callback_sender, should_send_request)));
@@ -845,14 +855,18 @@ mod tests {
         let worker = Worker::new(0, Arc::new(gateway), storage, ledger, Default::default()).unwrap();
         let transaction_id: <CurrentNetwork as Network>::TransactionID = Field::<CurrentNetwork>::rand(&mut rng).into();
         let transmission_id = TransmissionID::Transaction(transaction_id);
-        let peer_ip = SocketAddr::from(([127, 0, 0, 1], 1234));
 
         // Determine the number of redundant requests are sent.
         let num_redundant_requests = max_redundant_requests(worker.ledger.clone(), worker.storage.current_round());
         let num_flood_requests = num_redundant_requests * 10;
+        let mut peer_ips =
+            (0..num_flood_requests).map(|i| SocketAddr::from(([127, 0, 0, 1], 1234 + i as u16))).collect_vec();
+        let first_peer_ip = peer_ips[0];
+
         // Flood the pending queue with transmission requests.
         for i in 1..=num_flood_requests {
             let worker_ = worker.clone();
+            let peer_ip = peer_ips.pop().unwrap();
             tokio::spawn(async move {
                 let _ = worker_.send_transmission_request(peer_ip, transmission_id).await;
             });
@@ -870,18 +884,18 @@ mod tests {
         assert_eq!(worker.pending.num_sent_requests(transmission_id), 0);
         assert_eq!(worker.pending.num_callbacks(transmission_id), 0);
 
-        // Flood the pending queue with transmission requests again.
+        // Flood the pending queue with transmission requests again, this time to a single peer
         for i in 1..=num_flood_requests {
             let worker_ = worker.clone();
             tokio::spawn(async move {
-                let _ = worker_.send_transmission_request(peer_ip, transmission_id).await;
+                let _ = worker_.send_transmission_request(first_peer_ip, transmission_id).await;
             });
             tokio::time::sleep(Duration::from_millis(10)).await;
             assert!(worker.pending.num_sent_requests(transmission_id) <= num_redundant_requests);
             assert_eq!(worker.pending.num_callbacks(transmission_id), i);
         }
         // Check that the number of sent requests does not exceed the maximum number of redundant requests.
-        assert_eq!(worker.pending.num_sent_requests(transmission_id), num_redundant_requests);
+        assert_eq!(worker.pending.num_sent_requests(transmission_id), 1);
         assert_eq!(worker.pending.num_callbacks(transmission_id), num_flood_requests);
 
         // Check that fulfilling a transmission request clears the pending queue.
diff --git a/node/bft/storage-service/src/memory.rs b/node/bft/storage-service/src/memory.rs
index 38cd430d6..25648742e 100644
--- a/node/bft/storage-service/src/memory.rs
+++ b/node/bft/storage-service/src/memory.rs
@@ -120,7 +120,9 @@ impl<N: Network> StorageService<N> for BFTMemoryService<N> {
                 Entry::Vacant(vacant_entry) => {
                     // Retrieve the missing transmission.
                     let Some(transmission) = missing_transmissions.remove(&transmission_id) else {
-                        if !aborted_transmission_ids.contains(&transmission_id) {
+                        if !aborted_transmission_ids.contains(&transmission_id)
+                            && !self.contains_transmission(transmission_id)
+                        {
                             error!("Failed to provide a missing transmission {transmission_id}");
                         }
                         continue 'outer;
diff --git a/node/bft/storage-service/src/persistent.rs b/node/bft/storage-service/src/persistent.rs
index b18b3771d..96a9d0c8e 100644
--- a/node/bft/storage-service/src/persistent.rs
+++ b/node/bft/storage-service/src/persistent.rs
@@ -171,7 +171,9 @@ impl<N: Network> StorageService<N> for BFTPersistentStorage<N> {
                 Ok(None) => {
                     // Retrieve the missing transmission.
                     let Some(transmission) = missing_transmissions.remove(&transmission_id) else {
-                        if !aborted_transmission_ids.contains(&transmission_id) {
+                        if !aborted_transmission_ids.contains(&transmission_id)
+                            && !self.contains_transmission(transmission_id)
+                        {
                             error!("Failed to provide a missing transmission {transmission_id}");
                         }
                         continue 'outer;
diff --git a/node/bft/tests/components/worker.rs b/node/bft/tests/components/worker.rs
index a439a8bdf..63a74c913 100644
--- a/node/bft/tests/components/worker.rs
+++ b/node/bft/tests/components/worker.rs
@@ -43,8 +43,12 @@ async fn test_resend_transmission_request() {
     let max_redundancy = max_redundant_requests(ledger.clone(), 0);
     assert_eq!(max_redundancy, 34, "Update me if the formula changes");
 
+    // Prepare peer ips.
+    let num_test_requests = 11;
+    let mut peer_ips = (0..num_test_requests).map(|i| SocketAddr::from(([127, 0, 0, 1], 1234 + i as u16))).collect::<Vec<_>>();
+    let initial_peer_ip = peer_ips.pop().unwrap();
+
     // Prepare a dummy transmission ID.
-    let peer_ip = SocketAddr::from(([127, 0, 0, 1], 1234));
     let transmission_id = TransmissionID::Transaction(<CurrentNetwork as Network>::TransactionID::default());
 
     // Ensure the worker does not have the dummy transmission ID.
@@ -52,7 +56,7 @@ async fn test_resend_transmission_request() {
 
     // Send a request to fetch the dummy transmission.
     let worker_ = worker.clone();
-    tokio::spawn(async move { worker_.get_or_fetch_transmission(peer_ip, transmission_id).await });
+    tokio::spawn(async move { worker_.get_or_fetch_transmission(initial_peer_ip, transmission_id).await });
 
     tokio::time::sleep(std::time::Duration::from_millis(10)).await;
 
@@ -60,16 +64,17 @@ async fn test_resend_transmission_request() {
     // Ensure the transmission ID exists in the pending queue.
     assert!(pending.contains(transmission_id), "Missing a transmission in the pending queue");
     // Ensure the peer IP is in the pending queue for the transmission ID.
-    assert!(pending.contains_peer(transmission_id, peer_ip), "Missing a peer IP for transmission in the pending queue");
-    assert_eq!(pending.get_peers(transmission_id), Some([peer_ip].into_iter().collect()), "Missing a peer IP for transmission in the pending queue");
+    assert!(pending.contains_peer(transmission_id, initial_peer_ip), "Missing a peer IP for transmission in the pending queue");
+    assert_eq!(pending.get_peers(transmission_id), Some([initial_peer_ip].into_iter().collect()), "Missing a peer IP for transmission in the pending queue");
     // Ensure the number of callbacks is correct.
     assert_eq!(pending.num_callbacks(transmission_id), 1, "Incorrect number of callbacks for transmission");
     // Ensure the number of sent requests is correct.
     assert_eq!(pending.num_sent_requests(transmission_id), 1, "Incorrect number of sent requests for transmission");
 
-    // Rebroadcast the same request to fetch the dummy transmission.
-    for i in 1..=10 {
+    // Rebroadcast the same request to the same peer to fetch the dummy transmission.
+    for i in 1..num_test_requests {
         let worker_ = worker.clone();
+        let peer_ip = initial_peer_ip;
         tokio::spawn(async move { worker_.get_or_fetch_transmission(peer_ip, transmission_id).await });
 
         tokio::time::sleep(std::time::Duration::from_millis(10)).await;
@@ -82,6 +87,22 @@ async fn test_resend_transmission_request() {
         // Ensure the number of callbacks is correct.
         assert_eq!(pending.num_callbacks(transmission_id), 1 + i, "Incorrect number of callbacks for transmission");
         // Ensure the number of sent requests is correct.
+        assert_eq!(pending.num_sent_requests(transmission_id), 1, "Incorrect number of sent requests for transmission");
+    }
+
+    // Rebroadcast the same request to new peers fetch the dummy transmission.
+    for i in 1..num_test_requests {
+        let peer_ip = peer_ips.pop().unwrap();
+        let worker_ = worker.clone();
+        tokio::spawn(async move { worker_.get_or_fetch_transmission(peer_ip, transmission_id).await });
+
+        tokio::time::sleep(std::time::Duration::from_millis(10)).await;
+
+        // Ensure the transmission ID exists in the pending queue.
+        assert!(pending.contains(transmission_id), "Missing a transmission in the pending queue");
+        // Ensure the peer IP is in the pending queue for the transmission ID.
+        assert!(pending.contains_peer(transmission_id, peer_ip), "Missing a peer IP for transmission in the pending queue");
+        // Ensure the number of sent requests is correct.
         assert_eq!(pending.num_sent_requests(transmission_id), (1 + i).min(max_redundancy), "Incorrect number of sent requests for transmission");
     }
 }
@@ -104,15 +125,20 @@ async fn test_flood_transmission_requests() {
     let max_redundancy = max_redundant_requests(ledger.clone(), 0);
     assert_eq!(max_redundancy, 34, "Update me if the formula changes");
 
+    // Prepare peer ips.
+    let mut peer_ips = (0..max_redundancy + 1).map(|i| SocketAddr::from(([127, 0, 0, 1], 1234 + i as u16))).collect::<Vec<_>>();
+    let all_peer_ips = peer_ips.clone();
+    let initial_peer_ip = peer_ips.pop().unwrap();
+    let mut remaining_peer_ips = peer_ips;
+
     // Prepare a dummy transmission ID.
-    let peer_ip = SocketAddr::from(([127, 0, 0, 1], 1234));
     let transmission_id = TransmissionID::Transaction(<CurrentNetwork as Network>::TransactionID::default());
 
     // Ensure the worker does not have the dummy transmission ID.
     assert!(!worker.contains_transmission(transmission_id), "Transmission should not exist");
 
     // Send the maximum number of redundant requests to fetch the dummy transmission.
-    for _ in 0..max_redundancy {
+    for peer_ip in remaining_peer_ips.clone() {
         let worker_ = worker.clone();
         tokio::spawn(async move { worker_.get_or_fetch_transmission(peer_ip, transmission_id).await });
     }
@@ -123,16 +149,16 @@ async fn test_flood_transmission_requests() {
     // Ensure the transmission ID exists in the pending queue.
     assert!(pending.contains(transmission_id), "Missing a transmission in the pending queue");
     // Ensure the peer IP is in the pending queue for the transmission ID.
-    assert!(pending.contains_peer(transmission_id, peer_ip), "Missing a peer IP for transmission in the pending queue");
-    assert_eq!(pending.get_peers(transmission_id), Some([peer_ip].into_iter().collect()), "Missing a peer IP for transmission in the pending queue");
+    assert_eq!(pending.get_peers(transmission_id), Some(remaining_peer_ips.clone().into_iter().collect()), "Missing a peer IP for transmission in the pending queue");
     // Ensure the number of callbacks is correct.
     assert_eq!(pending.num_callbacks(transmission_id), max_redundancy, "Incorrect number of callbacks for transmission");
     // Ensure the number of sent requests is correct.
     assert_eq!(pending.num_sent_requests(transmission_id), max_redundancy, "Incorrect number of sent requests for transmission");
 
-    // Ensure any further redundant requests are not sent.
+    // Ensure any further redundant requests are not sent when sending to the same peer.
     for i in 1..=20 {
         let worker_ = worker.clone();
+        let peer_ip = initial_peer_ip;
         tokio::spawn(async move { worker_.get_or_fetch_transmission(peer_ip, transmission_id).await });
 
         tokio::time::sleep(std::time::Duration::from_millis(10)).await;
@@ -141,10 +167,29 @@ async fn test_flood_transmission_requests() {
         assert!(pending.contains(transmission_id), "Missing a transmission in the pending queue");
         // Ensure the peer IP is in the pending queue for the transmission ID.
         assert!(pending.contains_peer(transmission_id, peer_ip), "Missing a peer IP for transmission in the pending queue");
-        assert_eq!(pending.get_peers(transmission_id), Some([peer_ip].into_iter().collect()), "Missing a peer IP for transmission in the pending queue");
+        assert_eq!(pending.get_peers(transmission_id), Some(all_peer_ips.clone().into_iter().collect()), "Missing a peer IP for transmission in the pending queue");
         // Ensure the number of callbacks is correct.
         assert_eq!(pending.num_callbacks(transmission_id), max_redundancy + i, "Incorrect number of callbacks for transmission");
         // Ensure the number of sent requests is correct.
         assert_eq!(pending.num_sent_requests(transmission_id), max_redundancy, "Incorrect number of sent requests for transmission");
     }
+
+    // Ensure any further redundant requests are not sent when sending to new peers.
+    for i in 1..=20 {
+        let worker_ = worker.clone();
+        let peer_ip = remaining_peer_ips.pop().unwrap();
+        tokio::spawn(async move { worker_.get_or_fetch_transmission(peer_ip, transmission_id).await });
+
+        tokio::time::sleep(std::time::Duration::from_millis(10)).await;
+
+        // Ensure the transmission ID exists in the pending queue.
+        assert!(pending.contains(transmission_id), "Missing a transmission in the pending queue");
+        // Ensure the peer IP is in the pending queue for the transmission ID.
+        assert!(pending.contains_peer(transmission_id, peer_ip), "Missing a peer IP for transmission in the pending queue");
+        assert_eq!(pending.get_peers(transmission_id), Some(all_peer_ips.clone().into_iter().collect()), "Missing a peer IP for transmission in the pending queue");
+        // Ensure the number of callbacks is correct.
+        assert_eq!(pending.num_callbacks(transmission_id), max_redundancy + 20 + i, "Incorrect number of callbacks for transmission");
+        // Ensure the number of sent requests is correct.
+        assert_eq!(pending.num_sent_requests(transmission_id), max_redundancy, "Incorrect number of sent requests for transmission");
+    }
 }
diff --git a/node/consensus/src/lib.rs b/node/consensus/src/lib.rs
index 530893893..fdf247dfd 100644
--- a/node/consensus/src/lib.rs
+++ b/node/consensus/src/lib.rs
@@ -484,6 +484,14 @@ impl<N: Network> Consensus<N> {
         // Advance to the next block.
         self.ledger.advance_to_next_block(&next_block)?;
 
+        // If the next block starts a new epoch, clear the existing solutions.
+        if next_block.height() % N::NUM_BLOCKS_PER_EPOCH == 0 {
+            // Clear the solutions queue.
+            self.solutions_queue.lock().clear();
+            // Clear the worker solutions.
+            self.bft.primary().clear_worker_solutions();
+        }
+
         #[cfg(feature = "metrics")]
         {
             let elapsed = std::time::Duration::from_secs((snarkos_node_bft::helpers::now() - start) as u64);
diff --git a/node/rest/src/routes.rs b/node/rest/src/routes.rs
index 20ef76291..422a65701 100644
--- a/node/rest/src/routes.rs
+++ b/node/rest/src/routes.rs
@@ -135,16 +135,18 @@ impl<N: Network, C: ConsensusStorage<N>, R: Routing<N>> Rest<N, C, R> {
             )));
         }
 
-        // Fetch the blocks from ledger.
-        match tokio::task::spawn_blocking(move || {
-            cfg_into_iter!((start_height..end_height))
+        // Prepare a closure for the blocking work.
+        let get_json_blocks = move || -> Result<ErasedJson, RestError> {
+            let blocks = cfg_into_iter!((start_height..end_height))
                 .map(|height| rest.ledger.get_block(height))
-                .collect::<Result<Vec<_>, _>>()
-        })
-        .await
-        {
-            Ok(Ok(blocks)) => Ok(ErasedJson::pretty(blocks)),
-            Ok(Err(err)) => Err(RestError(format!("Failed to get blocks '{start_height}..{end_height}' - {err}"))),
+                .collect::<Result<Vec<_>, _>>()?;
+
+            Ok(ErasedJson::pretty(blocks))
+        };
+
+        // Fetch the blocks from ledger and serialize to json.
+        match tokio::task::spawn_blocking(get_json_blocks).await {
+            Ok(json) => json,
             Err(err) => Err(RestError(format!("Failed to get blocks '{start_height}..{end_height}' - {err}"))),
         }
     }
diff --git a/node/router/Cargo.toml b/node/router/Cargo.toml
index c0483487c..a796cf267 100644
--- a/node/router/Cargo.toml
+++ b/node/router/Cargo.toml
@@ -58,6 +58,9 @@ version = "0.12"
 [dependencies.rand]
 version = "0.8"
 
+[dependencies.rayon]
+version = "1.10"
+
 [dependencies.reqwest]
 version = "0.11"
 
diff --git a/node/router/messages/src/unconfirmed_solution.rs b/node/router/messages/src/unconfirmed_solution.rs
index 9caf066ab..0b34060c6 100644
--- a/node/router/messages/src/unconfirmed_solution.rs
+++ b/node/router/messages/src/unconfirmed_solution.rs
@@ -52,7 +52,7 @@ impl<N: Network> FromBytes for UnconfirmedSolution<N> {
 pub mod prop_tests {
     use crate::{Solution, SolutionID, UnconfirmedSolution};
     use snarkvm::{
-        ledger::narwhal::Data,
+        ledger::{narwhal::Data, puzzle::PartialSolution},
         prelude::{Address, FromBytes, PrivateKey, Rng, TestRng, ToBytes},
     };
 
@@ -72,7 +72,8 @@ pub mod prop_tests {
                 let mut rng = TestRng::fixed(seed);
                 let private_key = PrivateKey::<CurrentNetwork>::new(&mut rng).unwrap();
                 let address = Address::try_from(private_key).unwrap();
-                Solution::new(rng.gen(), address, rng.gen()).unwrap()
+                let partial_solution = PartialSolution::new(rng.gen(), address, rng.gen()).unwrap();
+                Solution::new(partial_solution, rng.gen())
             })
             .boxed()
     }
diff --git a/node/router/src/inbound.rs b/node/router/src/inbound.rs
index b8d985040..7ebbac2b7 100644
--- a/node/router/src/inbound.rs
+++ b/node/router/src/inbound.rs
@@ -36,7 +36,7 @@ use snarkvm::prelude::{
 
 use anyhow::{anyhow, bail, Result};
 use snarkos_node_tcp::is_bogon_ip;
-use std::{net::SocketAddr, time::Instant};
+use std::net::SocketAddr;
 use tokio::task::spawn_blocking;
 
 /// The max number of peers to send in a `PeerResponse` message.
@@ -76,6 +76,9 @@ pub trait Inbound<N: Network>: Reading + Outbound<N> {
 
         trace!("Received '{}' from '{peer_ip}'", message.name());
 
+        // Update the last seen timestamp of the peer.
+        self.router().update_last_seen_for_connected_peer(peer_ip);
+
         // This match statement handles the inbound message by deserializing the message,
         // checking that the message is valid, and then calling the appropriate (trait) handler.
         match message {
@@ -110,7 +113,19 @@ pub trait Inbound<N: Network>: Reading + Outbound<N> {
                     bail!("Peer '{peer_ip}' is not following the protocol (unexpected block response)")
                 }
                 // Perform the deferred non-blocking deserialization of the blocks.
-                let blocks = blocks.deserialize().await.map_err(|error| anyhow!("[BlockResponse] {error}"))?;
+                // The deserialization can take a long time (minutes). We should not be running
+                // this on a blocking task, but on a rayon thread pool.
+                let (send, recv) = tokio::sync::oneshot::channel();
+                rayon::spawn_fifo(move || {
+                    let blocks = blocks.deserialize_blocking().map_err(|error| anyhow!("[BlockResponse] {error}"));
+                    let _ = send.send(blocks);
+                });
+                let blocks = match recv.await {
+                    Ok(Ok(blocks)) => blocks,
+                    Ok(Err(error)) => bail!("Peer '{peer_ip}' sent an invalid block response - {error}"),
+                    Err(error) => bail!("Peer '{peer_ip}' sent an invalid block response - {error}"),
+                };
+
                 // Ensure the block response is well-formed.
                 blocks.ensure_response_is_well_formed(peer_ip, request.start_height, request.end_height)?;
 
@@ -169,8 +184,6 @@ pub trait Inbound<N: Network>: Reading + Outbound<N> {
                         peer.set_version(message.version);
                         // Update the node type of the peer.
                         peer.set_node_type(message.node_type);
-                        // Update the last seen timestamp of the peer.
-                        peer.set_last_seen(Instant::now());
                     })
                 {
                     bail!("[Ping] {error}");
diff --git a/node/router/src/lib.rs b/node/router/src/lib.rs
index e957a214e..45e6470e8 100644
--- a/node/router/src/lib.rs
+++ b/node/router/src/lib.rs
@@ -500,6 +500,12 @@ impl<N: Network> Router<N> {
         Ok(())
     }
 
+    pub fn update_last_seen_for_connected_peer(&self, peer_ip: SocketAddr) {
+        if let Some(peer) = self.connected_peers.write().get_mut(&peer_ip) {
+            peer.set_last_seen(Instant::now());
+        }
+    }
+
     /// Removes the connected peer and adds them to the candidate peers.
     pub fn remove_connected_peer(&self, peer_ip: SocketAddr) {
         // Removes the bidirectional map between the listener address and (ambiguous) peer address.
diff --git a/node/src/client/router.rs b/node/src/client/router.rs
index 53f7356f8..e621a2a35 100644
--- a/node/src/client/router.rs
+++ b/node/src/client/router.rs
@@ -115,6 +115,26 @@ impl<N: Network, C: ConsensusStorage<N>> Reading for Client<N, C> {
 
     /// Processes a message received from the network.
     async fn process_message(&self, peer_addr: SocketAddr, message: Self::Message) -> io::Result<()> {
+        let clone = self.clone();
+        if matches!(message, Message::BlockRequest(_) | Message::BlockResponse(_)) {
+            // Handle BlockRequest and BlockResponse messages in a separate task to not block the
+            // inbound queue.
+            tokio::spawn(async move {
+                clone.process_message_inner(peer_addr, message).await;
+            });
+        } else {
+            self.process_message_inner(peer_addr, message).await;
+        }
+        Ok(())
+    }
+}
+
+impl<N: Network, C: ConsensusStorage<N>> Client<N, C> {
+    async fn process_message_inner(
+        &self,
+        peer_addr: SocketAddr,
+        message: <Client<N, C> as snarkos_node_tcp::protocols::Reading>::Message,
+    ) {
         // Process the message. Disconnect if the peer violated the protocol.
         if let Err(error) = self.inbound(peer_addr, message).await {
             if let Some(peer_ip) = self.router().resolve_to_listener(&peer_addr) {
@@ -124,7 +144,6 @@ impl<N: Network, C: ConsensusStorage<N>> Reading for Client<N, C> {
                 self.router().disconnect(peer_ip);
             }
         }
-        Ok(())
     }
 }
 
diff --git a/node/src/validator/router.rs b/node/src/validator/router.rs
index 0ac3743ae..550171d3b 100644
--- a/node/src/validator/router.rs
+++ b/node/src/validator/router.rs
@@ -111,6 +111,26 @@ impl<N: Network, C: ConsensusStorage<N>> Reading for Validator<N, C> {
 
     /// Processes a message received from the network.
     async fn process_message(&self, peer_addr: SocketAddr, message: Self::Message) -> io::Result<()> {
+        let clone = self.clone();
+        if matches!(message, Message::BlockRequest(_) | Message::BlockResponse(_)) {
+            // Handle BlockRequest and BlockResponse messages in a separate task to not block the
+            // inbound queue.
+            tokio::spawn(async move {
+                clone.process_message_inner(peer_addr, message).await;
+            });
+        } else {
+            self.process_message_inner(peer_addr, message).await;
+        }
+        Ok(())
+    }
+}
+
+impl<N: Network, C: ConsensusStorage<N>> Validator<N, C> {
+    async fn process_message_inner(
+        &self,
+        peer_addr: SocketAddr,
+        message: <Validator<N, C> as snarkos_node_tcp::protocols::Reading>::Message,
+    ) {
         // Process the message. Disconnect if the peer violated the protocol.
         if let Err(error) = self.inbound(peer_addr, message).await {
             if let Some(peer_ip) = self.router().resolve_to_listener(&peer_addr) {
@@ -120,7 +140,6 @@ impl<N: Network, C: ConsensusStorage<N>> Reading for Validator<N, C> {
                 self.router().disconnect(peer_ip);
             }
         }
-        Ok(())
     }
 }
 
diff --git a/node/sync/src/block_sync.rs b/node/sync/src/block_sync.rs
index 525962ed0..54a21e898 100644
--- a/node/sync/src/block_sync.rs
+++ b/node/sync/src/block_sync.rs
@@ -453,8 +453,11 @@ impl<N: Network> BlockSync<N> {
             // Return the list of block requests.
             (self.construct_requests(&sync_peers, min_common_ancestor), sync_peers)
         } else {
-            // Update the state of `is_block_synced` for the sync module.
-            self.update_is_block_synced(0, MAX_BLOCKS_BEHIND);
+            // Update `is_block_synced` if there are no pending requests or responses.
+            if self.requests.read().is_empty() && self.responses.read().is_empty() {
+                // Update the state of `is_block_synced` for the sync module.
+                self.update_is_block_synced(0, MAX_BLOCKS_BEHIND);
+            }
             // Return an empty list of block requests.
             (Default::default(), Default::default())
         }
@@ -646,6 +649,7 @@ impl<N: Network> BlockSync<N> {
 
             let retain = !peer_ips.is_empty() || responses.get(height).is_some();
             if !retain {
+                trace!("Removed block request timestamp for {peer_ip} at height {height}");
                 self.request_timestamps.write().remove(height);
             }
             retain
@@ -665,11 +669,15 @@ impl<N: Network> BlockSync<N> {
         // Retrieve the current time.
         let now = Instant::now();
 
+        // Retrieve the current block height
+        let current_height = self.canon.latest_block_height();
+
         // Track the number of timed out block requests.
         let mut num_timed_out_block_requests = 0;
 
         // Remove timed out block requests.
         request_timestamps.retain(|height, timestamp| {
+            let is_obsolete = *height < current_height;
             // Determine if the duration since the request timestamp has exceeded the request timeout.
             let is_time_passed = now.duration_since(*timestamp).as_secs() > BLOCK_REQUEST_TIMEOUT_IN_SECS;
             // Determine if the request is incomplete.
@@ -678,8 +686,9 @@ impl<N: Network> BlockSync<N> {
             // Determine if the request has timed out.
             let is_timeout = is_time_passed && is_request_incomplete;
 
-            // If the request has timed out, then remove it.
-            if is_timeout {
+            // If the request has timed out, or is obsolete, then remove it.
+            if is_timeout || is_obsolete {
+                trace!("Block request {height} has timed out: is_time_passed = {is_time_passed}, is_request_incomplete = {is_request_incomplete}, is_obsolete = {is_obsolete}");
                 // Remove the request entry for the given height.
                 requests.remove(height);
                 // Remove the response entry for the given height.
@@ -687,8 +696,8 @@ impl<N: Network> BlockSync<N> {
                 // Increment the number of timed out block requests.
                 num_timed_out_block_requests += 1;
             }
-            // Retain if this is not a timeout.
-            !is_timeout
+            // Retain if this is not a timeout and is not obsolete.
+            !is_timeout && !is_obsolete
         });
 
         num_timed_out_block_requests
