diff --git a/.cargo/release-version b/.cargo/release-version
index 05f629f1b..ab664c2b5 100644
--- a/.cargo/release-version
+++ b/.cargo/release-version
@@ -1 +1 @@
-v1.6.0
\ No newline at end of file
+v3.7.1
\ No newline at end of file
diff --git a/Cargo.lock b/Cargo.lock
index 151cdca9a..f5b5329eb 100644
--- a/Cargo.lock
+++ b/Cargo.lock
@@ -2789,7 +2789,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm"
-version = "1.6.0"
+version = "3.7.1"
 dependencies = [
  "anstyle",
  "anyhow",
@@ -2826,7 +2826,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-algorithms"
-version = "1.6.0"
+version = "3.7.1"
 dependencies = [
  "aleo-std",
  "anyhow",
@@ -2866,7 +2866,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-algorithms-cuda"
-version = "1.6.0"
+version = "3.7.1"
 dependencies = [
  "blst",
  "cc",
@@ -2876,7 +2876,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-circuit"
-version = "1.6.0"
+version = "3.7.1"
 dependencies = [
  "snarkvm-circuit-account",
  "snarkvm-circuit-algorithms",
@@ -2889,7 +2889,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-circuit-account"
-version = "1.6.0"
+version = "3.7.1"
 dependencies = [
  "anyhow",
  "snarkvm-circuit-algorithms",
@@ -2901,7 +2901,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-circuit-algorithms"
-version = "1.6.0"
+version = "3.7.1"
 dependencies = [
  "anyhow",
  "snarkvm-circuit-types",
@@ -2913,7 +2913,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-circuit-collections"
-version = "1.6.0"
+version = "3.7.1"
 dependencies = [
  "anyhow",
  "snarkvm-circuit-algorithms",
@@ -2927,7 +2927,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-circuit-environment"
-version = "1.6.0"
+version = "3.7.1"
 dependencies = [
  "criterion",
  "indexmap 2.9.0",
@@ -2950,11 +2950,11 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-circuit-environment-witness"
-version = "1.6.0"
+version = "3.7.1"
 
 [[package]]
 name = "snarkvm-circuit-network"
-version = "1.6.0"
+version = "3.7.1"
 dependencies = [
  "snarkvm-circuit-algorithms",
  "snarkvm-circuit-collections",
@@ -2965,7 +2965,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-circuit-program"
-version = "1.6.0"
+version = "3.7.1"
 dependencies = [
  "anyhow",
  "paste",
@@ -2983,7 +2983,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-circuit-types"
-version = "1.6.0"
+version = "3.7.1"
 dependencies = [
  "snarkvm-circuit-environment",
  "snarkvm-circuit-types-address",
@@ -2998,7 +2998,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-circuit-types-address"
-version = "1.6.0"
+version = "3.7.1"
 dependencies = [
  "snarkvm-circuit-environment",
  "snarkvm-circuit-types-boolean",
@@ -3010,7 +3010,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-circuit-types-boolean"
-version = "1.6.0"
+version = "3.7.1"
 dependencies = [
  "criterion",
  "snarkvm-circuit-environment",
@@ -3019,7 +3019,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-circuit-types-field"
-version = "1.6.0"
+version = "3.7.1"
 dependencies = [
  "snarkvm-circuit-environment",
  "snarkvm-circuit-types-boolean",
@@ -3028,7 +3028,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-circuit-types-group"
-version = "1.6.0"
+version = "3.7.1"
 dependencies = [
  "snarkvm-circuit-environment",
  "snarkvm-circuit-types-boolean",
@@ -3040,7 +3040,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-circuit-types-integers"
-version = "1.6.0"
+version = "3.7.1"
 dependencies = [
  "paste",
  "snarkvm-circuit-environment",
@@ -3053,7 +3053,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-circuit-types-scalar"
-version = "1.6.0"
+version = "3.7.1"
 dependencies = [
  "snarkvm-circuit-environment",
  "snarkvm-circuit-types-boolean",
@@ -3063,7 +3063,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-circuit-types-string"
-version = "1.6.0"
+version = "3.7.1"
 dependencies = [
  "rand",
  "snarkvm-circuit-environment",
@@ -3076,7 +3076,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-console"
-version = "1.6.0"
+version = "3.7.1"
 dependencies = [
  "snarkvm-console-account",
  "snarkvm-console-algorithms",
@@ -3088,7 +3088,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-console-account"
-version = "1.6.0"
+version = "3.7.1"
 dependencies = [
  "bincode",
  "bs58",
@@ -3101,7 +3101,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-console-algorithms"
-version = "1.6.0"
+version = "3.7.1"
 dependencies = [
  "blake2s_simd",
  "criterion",
@@ -3119,7 +3119,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-console-collections"
-version = "1.6.0"
+version = "3.7.1"
 dependencies = [
  "aleo-std",
  "criterion",
@@ -3132,7 +3132,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-console-network"
-version = "1.6.0"
+version = "3.7.1"
 dependencies = [
  "anyhow",
  "indexmap 2.9.0",
@@ -3154,7 +3154,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-console-network-environment"
-version = "1.6.0"
+version = "3.7.1"
 dependencies = [
  "anyhow",
  "bech32",
@@ -3171,7 +3171,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-console-program"
-version = "1.6.0"
+version = "3.7.1"
 dependencies = [
  "bincode",
  "enum-iterator",
@@ -3193,7 +3193,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-console-types"
-version = "1.6.0"
+version = "3.7.1"
 dependencies = [
  "criterion",
  "snarkvm-console-network",
@@ -3209,7 +3209,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-console-types-address"
-version = "1.6.0"
+version = "3.7.1"
 dependencies = [
  "bincode",
  "serde_json",
@@ -3221,7 +3221,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-console-types-boolean"
-version = "1.6.0"
+version = "3.7.1"
 dependencies = [
  "bincode",
  "serde_json",
@@ -3230,7 +3230,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-console-types-field"
-version = "1.6.0"
+version = "3.7.1"
 dependencies = [
  "bincode",
  "serde_json",
@@ -3241,7 +3241,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-console-types-group"
-version = "1.6.0"
+version = "3.7.1"
 dependencies = [
  "bincode",
  "serde_json",
@@ -3253,7 +3253,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-console-types-integers"
-version = "1.6.0"
+version = "3.7.1"
 dependencies = [
  "bincode",
  "serde_json",
@@ -3265,7 +3265,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-console-types-scalar"
-version = "1.6.0"
+version = "3.7.1"
 dependencies = [
  "bincode",
  "serde_json",
@@ -3277,7 +3277,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-console-types-string"
-version = "1.6.0"
+version = "3.7.1"
 dependencies = [
  "bincode",
  "serde_json",
@@ -3289,7 +3289,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-curves"
-version = "1.6.0"
+version = "3.7.1"
 dependencies = [
  "bincode",
  "criterion",
@@ -3304,7 +3304,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-fields"
-version = "1.6.0"
+version = "3.7.1"
 dependencies = [
  "aleo-std",
  "anyhow",
@@ -3322,7 +3322,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-ledger"
-version = "1.6.0"
+version = "3.7.1"
 dependencies = [
  "aleo-std",
  "anyhow",
@@ -3353,7 +3353,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-ledger-authority"
-version = "1.6.0"
+version = "3.7.1"
 dependencies = [
  "anyhow",
  "bincode",
@@ -3366,7 +3366,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-ledger-block"
-version = "1.6.0"
+version = "3.7.1"
 dependencies = [
  "aleo-std",
  "bincode",
@@ -3393,7 +3393,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-ledger-committee"
-version = "1.6.0"
+version = "3.7.1"
 dependencies = [
  "anyhow",
  "bincode",
@@ -3414,7 +3414,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-ledger-narwhal"
-version = "1.6.0"
+version = "3.7.1"
 dependencies = [
  "snarkvm-ledger-narwhal",
  "snarkvm-ledger-narwhal-batch-certificate",
@@ -3427,7 +3427,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-ledger-narwhal-batch-certificate"
-version = "1.6.0"
+version = "3.7.1"
 dependencies = [
  "bincode",
  "indexmap 2.9.0",
@@ -3441,7 +3441,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-ledger-narwhal-batch-header"
-version = "1.6.0"
+version = "3.7.1"
 dependencies = [
  "bincode",
  "indexmap 2.9.0",
@@ -3455,7 +3455,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-ledger-narwhal-data"
-version = "1.6.0"
+version = "3.7.1"
 dependencies = [
  "bytes",
  "serde_json",
@@ -3467,7 +3467,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-ledger-narwhal-subdag"
-version = "1.6.0"
+version = "3.7.1"
 dependencies = [
  "bincode",
  "indexmap 2.9.0",
@@ -3483,7 +3483,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-ledger-narwhal-transmission"
-version = "1.6.0"
+version = "3.7.1"
 dependencies = [
  "bincode",
  "bytes",
@@ -3496,7 +3496,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-ledger-narwhal-transmission-id"
-version = "1.6.0"
+version = "3.7.1"
 dependencies = [
  "bincode",
  "serde_json",
@@ -3506,7 +3506,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-ledger-puzzle"
-version = "1.6.0"
+version = "3.7.1"
 dependencies = [
  "aleo-std",
  "anyhow",
@@ -3528,7 +3528,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-ledger-puzzle-epoch"
-version = "1.6.0"
+version = "3.7.1"
 dependencies = [
  "aleo-std",
  "anyhow",
@@ -3549,7 +3549,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-ledger-query"
-version = "1.6.0"
+version = "3.7.1"
 dependencies = [
  "async-trait",
  "reqwest",
@@ -3561,7 +3561,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-ledger-store"
-version = "1.6.0"
+version = "3.7.1"
 dependencies = [
  "aleo-std",
  "aleo-std-storage",
@@ -3592,7 +3592,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-ledger-test-helpers"
-version = "1.6.0"
+version = "3.7.1"
 dependencies = [
  "aleo-std",
  "once_cell",
@@ -3608,7 +3608,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-metrics"
-version = "1.6.0"
+version = "3.7.1"
 dependencies = [
  "metrics",
  "metrics-exporter-prometheus",
@@ -3616,7 +3616,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-parameters"
-version = "1.6.0"
+version = "3.7.1"
 dependencies = [
  "aleo-std",
  "anyhow",
@@ -3651,7 +3651,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-synthesizer"
-version = "1.6.0"
+version = "3.7.1"
 dependencies = [
  "aleo-std",
  "anyhow",
@@ -3691,7 +3691,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-synthesizer-process"
-version = "1.6.0"
+version = "3.7.1"
 dependencies = [
  "aleo-std",
  "bincode",
@@ -3721,7 +3721,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-synthesizer-program"
-version = "1.6.0"
+version = "3.7.1"
 dependencies = [
  "bincode",
  "criterion",
@@ -3738,7 +3738,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-synthesizer-snark"
-version = "1.6.0"
+version = "3.7.1"
 dependencies = [
  "bincode",
  "colored",
@@ -3751,7 +3751,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-utilities"
-version = "1.6.0"
+version = "3.7.1"
 dependencies = [
  "aleo-std",
  "anyhow",
@@ -3771,7 +3771,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-utilities-derives"
-version = "1.6.0"
+version = "3.7.1"
 dependencies = [
  "proc-macro2",
  "quote 1.0.40",
@@ -3780,7 +3780,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-wasm"
-version = "1.6.0"
+version = "3.7.1"
 dependencies = [
  "getrandom 0.2.15",
  "snarkvm-circuit-network",
diff --git a/Cargo.toml b/Cargo.toml
index d80671701..9999acbb2 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm"
-version = "1.6.0"
+version = "3.7.1"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "A decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -164,58 +164,58 @@ test_consensus_heights = [ "snarkvm-console/test_consensus_heights" ]
 
 [dependencies.snarkvm-algorithms]
 path = "./algorithms"
-version = "=1.6.0"
+version = "=3.7.1"
 optional = true
 
 [dependencies.snarkvm-circuit]
 path = "./circuit"
-version = "=1.6.0"
+version = "=3.7.1"
 optional = true
 
 [dependencies.snarkvm-console]
 path = "./console"
-version = "=1.6.0"
+version = "=3.7.1"
 optional = true
 
 [dependencies.snarkvm-curves]
 path = "./curves"
-version = "=1.6.0"
+version = "=3.7.1"
 optional = true
 
 [dependencies.snarkvm-fields]
 path = "./fields"
-version = "=1.6.0"
+version = "=3.7.1"
 optional = true
 
 [dependencies.snarkvm-ledger]
 path = "./ledger"
-version = "=1.6.0"
+version = "=3.7.1"
 optional = true
 
 [dependencies.snarkvm-metrics]
 path = "./metrics"
-version = "=1.6.0"
+version = "=3.7.1"
 optional = true
 
 [dependencies.snarkvm-parameters]
 path = "./parameters"
-version = "=1.6.0"
+version = "=3.7.1"
 optional = true
 
 [dependencies.snarkvm-synthesizer]
 path = "./synthesizer"
-version = "=1.6.0"
+version = "=3.7.1"
 default-features = false
 optional = true
 
 [dependencies.snarkvm-utilities]
 path = "./utilities"
-version = "=1.6.0"
+version = "=3.7.1"
 optional = true
 
 [dependencies.snarkvm-wasm]
 path = "./wasm"
-version = "=1.6.0"
+version = "=3.7.1"
 optional = true
 
 [dependencies.anstyle]
diff --git a/algorithms/Cargo.toml b/algorithms/Cargo.toml
index 55079ce18..a64b2bf18 100644
--- a/algorithms/Cargo.toml
+++ b/algorithms/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-algorithms"
-version = "1.6.0"
+version = "3.7.1"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Algorithms for a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -47,27 +47,27 @@ required-features = [ "test" ]
 
 [dependencies.snarkvm-curves]
 path = "../curves"
-version = "=1.6.0"
+version = "=3.7.1"
 default-features = false
 
 [dependencies.snarkvm-fields]
 path = "../fields"
-version = "=1.6.0"
+version = "=3.7.1"
 default-features = false
 
 [dependencies.snarkvm-parameters]
 path = "../parameters"
-version = "=1.6.0"
+version = "=3.7.1"
 optional = true
 
 [dependencies.snarkvm-utilities]
 path = "../utilities"
-version = "=1.6.0"
+version = "=3.7.1"
 default-features = false
 
 [dependencies.snarkvm-algorithms-cuda]
 path = "./cuda"
-version = "=1.6.0"
+version = "=3.7.1"
 optional = true
 
 [dependencies.aleo-std]
diff --git a/algorithms/cuda/Cargo.toml b/algorithms/cuda/Cargo.toml
index 6dec68e53..c6e5100e1 100644
--- a/algorithms/cuda/Cargo.toml
+++ b/algorithms/cuda/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-algorithms-cuda"
-version = "1.6.0"
+version = "3.7.1"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Cuda optimizations for a decentralized virtual machine"
 homepage = "https://aleo.org"
diff --git a/algorithms/src/crypto_hash/poseidon.rs b/algorithms/src/crypto_hash/poseidon.rs
index df52c8222..49b56e93b 100644
--- a/algorithms/src/crypto_hash/poseidon.rs
+++ b/algorithms/src/crypto_hash/poseidon.rs
@@ -483,7 +483,7 @@ impl<F: PrimeField, const RATE: usize> PoseidonSponge<F, RATE, 1> {
         let dest_elements = bits
             .chunks_exact(num_bits_per_nonnative)
             .map(|per_nonnative_bits| {
-                // technically, this can be done via BigInterger::from_bits; here, we use this
+                // technically, this can be done via BigInteger::from_bits; here, we use this
                 // method for consistency with the gadget counterpart
                 let mut res = TargetField::zero();
 
diff --git a/circuit/Cargo.toml b/circuit/Cargo.toml
index c6b328d01..d20f77379 100644
--- a/circuit/Cargo.toml
+++ b/circuit/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-circuit"
-version = "1.6.0"
+version = "3.7.1"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Circuits for a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -28,28 +28,28 @@ save_r1cs_hashes = [ "snarkvm-circuit-environment/save_r1cs_hashes" ]
 
 [dependencies.snarkvm-circuit-account]
 path = "./account"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.snarkvm-circuit-algorithms]
 path = "./algorithms"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.snarkvm-circuit-collections]
 path = "./collections"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.snarkvm-circuit-environment]
 path = "./environment"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.snarkvm-circuit-network]
 path = "./network"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.snarkvm-circuit-program]
 path = "./program"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.snarkvm-circuit-types]
 path = "./types"
-version = "=1.6.0"
+version = "=3.7.1"
diff --git a/circuit/account/Cargo.toml b/circuit/account/Cargo.toml
index d5d7bd29a..989b019d1 100644
--- a/circuit/account/Cargo.toml
+++ b/circuit/account/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-circuit-account"
-version = "1.6.0"
+version = "3.7.1"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Account circuit library for a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -11,20 +11,20 @@ edition = "2021"
 [dependencies.console]
 package = "snarkvm-console-account"
 path = "../../console/account"
-version = "=1.6.0"
+version = "=3.7.1"
 optional = true
 
 [dependencies.snarkvm-circuit-algorithms]
 path = "../algorithms"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.snarkvm-circuit-network]
 path = "../network"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.snarkvm-circuit-types]
 path = "../types"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dev-dependencies.snarkvm-utilities]
 path = "../../utilities"
diff --git a/circuit/algorithms/Cargo.toml b/circuit/algorithms/Cargo.toml
index a17e984c9..2c42e78a7 100644
--- a/circuit/algorithms/Cargo.toml
+++ b/circuit/algorithms/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-circuit-algorithms"
-version = "1.6.0"
+version = "3.7.1"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Algorithm circuit library for a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -11,16 +11,16 @@ edition = "2021"
 [dependencies.console]
 package = "snarkvm-console-algorithms"
 path = "../../console/algorithms"
-version = "=1.6.0"
+version = "=3.7.1"
 optional = true
 
 [dependencies.snarkvm-circuit-types]
 path = "../types"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.snarkvm-fields]
 path = "../../fields"
-version = "=1.6.0"
+version = "=3.7.1"
 default-features = false
 
 [dev-dependencies.anyhow]
diff --git a/circuit/collections/Cargo.toml b/circuit/collections/Cargo.toml
index 7058ff914..eda34ec1a 100644
--- a/circuit/collections/Cargo.toml
+++ b/circuit/collections/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-circuit-collections"
-version = "1.6.0"
+version = "3.7.1"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Collections circuit library for a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -11,16 +11,16 @@ edition = "2021"
 [dependencies.console]
 package = "snarkvm-console-collections"
 path = "../../console/collections"
-version = "=1.6.0"
+version = "=3.7.1"
 optional = true
 
 [dependencies.snarkvm-circuit-algorithms]
 path = "../algorithms"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.snarkvm-circuit-types]
 path = "../types"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dev-dependencies.snarkvm-circuit-network]
 path = "../network"
diff --git a/circuit/environment/Cargo.toml b/circuit/environment/Cargo.toml
index 054dc6d82..cbcd255df 100644
--- a/circuit/environment/Cargo.toml
+++ b/circuit/environment/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-circuit-environment"
-version = "1.6.0"
+version = "3.7.1"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Circuit environment for a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -16,32 +16,32 @@ harness = false
 [dependencies.console]
 package = "snarkvm-console-network"
 path = "../../console/network"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.snarkvm-algorithms]
 path = "../../algorithms"
-version = "=1.6.0"
+version = "=3.7.1"
 default-features = false
 features = [ "r1cs" ]
 
 [dependencies.snarkvm-circuit-environment-witness]
 path = "./witness"
-version = "=1.6.0"
+version = "=3.7.1"
 default-features = false
 
 [dependencies.snarkvm-curves]
 path = "../../curves"
-version = "=1.6.0"
+version = "=3.7.1"
 default-features = false
 
 [dependencies.snarkvm-fields]
 path = "../../fields"
-version = "=1.6.0"
+version = "=3.7.1"
 default-features = false
 
 [dependencies.snarkvm-utilities]
 path = "../../utilities"
-version = "=1.6.0"
+version = "=3.7.1"
 default-features = false
 
 [dependencies.indexmap]
diff --git a/circuit/environment/src/helpers/r1cs.rs b/circuit/environment/src/helpers/r1cs.rs
index fd5a9db41..8a4a6f18e 100644
--- a/circuit/environment/src/helpers/r1cs.rs
+++ b/circuit/environment/src/helpers/r1cs.rs
@@ -52,7 +52,7 @@ fn hash_to_sha256<T: Hash>(t: &T) -> [u8; 32] {
 pub type Scope = String;
 
 /// A list of hashes of all the R1CS objects that have reached the
-/// coversion to Assignment stage. It's a vector in case there are
+/// conversion to Assignment stage. It's a vector in case there are
 /// any duplicates (which could indicate no change or redundant
 /// work) and since they need to eventually be sorted in order to
 /// have deterministic order, as they may be created in parallel.
diff --git a/circuit/environment/witness/Cargo.toml b/circuit/environment/witness/Cargo.toml
index 08211b08d..7051c9e1f 100644
--- a/circuit/environment/witness/Cargo.toml
+++ b/circuit/environment/witness/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-circuit-environment-witness"
-version = "1.6.0"
+version = "3.7.1"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "A procedural macro to construct a witness in an environment"
 homepage = "https://aleo.org"
diff --git a/circuit/network/Cargo.toml b/circuit/network/Cargo.toml
index f89c420d9..3a14f4e1a 100644
--- a/circuit/network/Cargo.toml
+++ b/circuit/network/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-circuit-network"
-version = "1.6.0"
+version = "3.7.1"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Network circuit library for a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -11,20 +11,20 @@ edition = "2021"
 [dependencies.console]
 package = "snarkvm-console-network"
 path = "../../console/network"
-version = "=1.6.0"
+version = "=3.7.1"
 optional = true
 
 [dependencies.snarkvm-circuit-algorithms]
 path = "../algorithms"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.snarkvm-circuit-collections]
 path = "../collections"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.snarkvm-circuit-types]
 path = "../types"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dev-dependencies.snarkvm-console-types]
 path = "../../console/types"
diff --git a/circuit/program/Cargo.toml b/circuit/program/Cargo.toml
index 54ca65053..d289c48ad 100644
--- a/circuit/program/Cargo.toml
+++ b/circuit/program/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-circuit-program"
-version = "1.6.0"
+version = "3.7.1"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Program circuit library for a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -11,32 +11,32 @@ edition = "2021"
 [dependencies.console]
 package = "snarkvm-console-program"
 path = "../../console/program"
-version = "=1.6.0"
+version = "=3.7.1"
 optional = true
 
 [dependencies.snarkvm-circuit-account]
 path = "../account"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.snarkvm-circuit-algorithms]
 path = "../algorithms"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.snarkvm-circuit-collections]
 path = "../collections"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.snarkvm-circuit-network]
 path = "../network"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.snarkvm-circuit-types]
 path = "../types"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.snarkvm-utilities]
 path = "../../utilities"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.paste]
 version = "1.0"
diff --git a/circuit/types/Cargo.toml b/circuit/types/Cargo.toml
index 8252627af..6a4b98c11 100644
--- a/circuit/types/Cargo.toml
+++ b/circuit/types/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-circuit-types"
-version = "1.6.0"
+version = "3.7.1"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Primitive circuit for a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -10,35 +10,35 @@ edition = "2021"
 
 [dependencies.snarkvm-circuit-environment]
 path = "../environment"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.snarkvm-circuit-types-address]
 path = "./address"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.snarkvm-circuit-types-boolean]
 path = "./boolean"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.snarkvm-circuit-types-field]
 path = "./field"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.snarkvm-circuit-types-group]
 path = "./group"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.snarkvm-circuit-types-integers]
 path = "./integers"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.snarkvm-circuit-types-scalar]
 path = "./scalar"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.snarkvm-circuit-types-string]
 path = "./string"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dev-dependencies.console]
 package = "snarkvm-console"
diff --git a/circuit/types/address/Cargo.toml b/circuit/types/address/Cargo.toml
index 9e3653ae9..f0e0ff8d5 100644
--- a/circuit/types/address/Cargo.toml
+++ b/circuit/types/address/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-circuit-types-address"
-version = "1.6.0"
+version = "3.7.1"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Address circuit for a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -11,28 +11,28 @@ edition = "2021"
 [dependencies.console]
 package = "snarkvm-console-types-address"
 path = "../../../console/types/address"
-version = "=1.6.0"
+version = "=3.7.1"
 optional = true
 
 [dependencies.snarkvm-circuit-environment]
 path = "../../environment"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.snarkvm-circuit-types-boolean]
 path = "../boolean"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.snarkvm-circuit-types-field]
 path = "../field"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.snarkvm-circuit-types-group]
 path = "../group"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.snarkvm-circuit-types-scalar]
 path = "../scalar"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [features]
 default = [ "enable_console" ]
diff --git a/circuit/types/boolean/Cargo.toml b/circuit/types/boolean/Cargo.toml
index ad21be4be..1aac5cf6f 100644
--- a/circuit/types/boolean/Cargo.toml
+++ b/circuit/types/boolean/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-circuit-types-boolean"
-version = "1.6.0"
+version = "3.7.1"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Boolean circuit for a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -16,12 +16,12 @@ harness = false
 [dependencies.console]
 package = "snarkvm-console-types-boolean"
 path = "../../../console/types/boolean"
-version = "=1.6.0"
+version = "=3.7.1"
 optional = true
 
 [dependencies.snarkvm-circuit-environment]
 path = "../../environment"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dev-dependencies.criterion]
 version = "0.5"
diff --git a/circuit/types/field/Cargo.toml b/circuit/types/field/Cargo.toml
index 408bd7631..b321702c5 100644
--- a/circuit/types/field/Cargo.toml
+++ b/circuit/types/field/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-circuit-types-field"
-version = "1.6.0"
+version = "3.7.1"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Field circuit for a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -11,16 +11,16 @@ edition = "2021"
 [dependencies.console]
 package = "snarkvm-console-types-field"
 path = "../../../console/types/field"
-version = "=1.6.0"
+version = "=3.7.1"
 optional = true
 
 [dependencies.snarkvm-circuit-environment]
 path = "../../environment"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.snarkvm-circuit-types-boolean]
 path = "../boolean"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [features]
 default = [ "enable_console" ]
diff --git a/circuit/types/group/Cargo.toml b/circuit/types/group/Cargo.toml
index de9cc4557..060d92159 100644
--- a/circuit/types/group/Cargo.toml
+++ b/circuit/types/group/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-circuit-types-group"
-version = "1.6.0"
+version = "3.7.1"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Group circuit for a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -11,24 +11,24 @@ edition = "2021"
 [dependencies.console]
 package = "snarkvm-console-types-group"
 path = "../../../console/types/group"
-version = "=1.6.0"
+version = "=3.7.1"
 optional = true
 
 [dependencies.snarkvm-circuit-environment]
 path = "../../environment"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.snarkvm-circuit-types-boolean]
 path = "../boolean"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.snarkvm-circuit-types-field]
 path = "../field"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.snarkvm-circuit-types-scalar]
 path = "../scalar"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dev-dependencies.snarkvm-utilities]
 path = "../../../utilities"
diff --git a/circuit/types/integers/Cargo.toml b/circuit/types/integers/Cargo.toml
index 3f4ecd89d..ad095a6ba 100644
--- a/circuit/types/integers/Cargo.toml
+++ b/circuit/types/integers/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-circuit-types-integers"
-version = "1.6.0"
+version = "3.7.1"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Integer circuit for a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -11,24 +11,24 @@ edition = "2021"
 [dependencies.console]
 package = "snarkvm-console-types-integers"
 path = "../../../console/types/integers"
-version = "=1.6.0"
+version = "=3.7.1"
 optional = true
 
 [dependencies.snarkvm-circuit-environment]
 path = "../../environment"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.snarkvm-circuit-types-boolean]
 path = "../boolean"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.snarkvm-circuit-types-field]
 path = "../field"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.snarkvm-circuit-types-scalar]
 path = "../scalar"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dev-dependencies.snarkvm-utilities]
 path = "../../../utilities"
diff --git a/circuit/types/scalar/Cargo.toml b/circuit/types/scalar/Cargo.toml
index 588e314b4..4cfe3548c 100644
--- a/circuit/types/scalar/Cargo.toml
+++ b/circuit/types/scalar/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-circuit-types-scalar"
-version = "1.6.0"
+version = "3.7.1"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Scalar circuit for a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -11,20 +11,20 @@ edition = "2021"
 [dependencies.console]
 package = "snarkvm-console-types-scalar"
 path = "../../../console/types/scalar"
-version = "=1.6.0"
+version = "=3.7.1"
 optional = true
 
 [dependencies.snarkvm-circuit-environment]
 path = "../../environment"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.snarkvm-circuit-types-boolean]
 path = "../boolean"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.snarkvm-circuit-types-field]
 path = "../field"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [features]
 default = [ "enable_console" ]
diff --git a/circuit/types/string/Cargo.toml b/circuit/types/string/Cargo.toml
index ad588edd1..18f4ba614 100644
--- a/circuit/types/string/Cargo.toml
+++ b/circuit/types/string/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-circuit-types-string"
-version = "1.6.0"
+version = "3.7.1"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "String circuit for a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -11,24 +11,24 @@ edition = "2021"
 [dependencies.console]
 package = "snarkvm-console-types-string"
 path = "../../../console/types/string"
-version = "=1.6.0"
+version = "=3.7.1"
 optional = true
 
 [dependencies.snarkvm-circuit-environment]
 path = "../../environment"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.snarkvm-circuit-types-boolean]
 path = "../boolean"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.snarkvm-circuit-types-field]
 path = "../field"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.snarkvm-circuit-types-integers]
 path = "../integers"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dev-dependencies.snarkvm-utilities]
 path = "../../../utilities"
diff --git a/console/Cargo.toml b/console/Cargo.toml
index f04ffd8a5..22cadc590 100644
--- a/console/Cargo.toml
+++ b/console/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-console"
-version = "1.6.0"
+version = "3.7.1"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Console environment for a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -10,32 +10,32 @@ edition = "2021"
 
 [dependencies.snarkvm-console-account]
 path = "./account"
-version = "=1.6.0"
+version = "=3.7.1"
 optional = true
 
 [dependencies.snarkvm-console-algorithms]
 path = "./algorithms"
-version = "=1.6.0"
+version = "=3.7.1"
 optional = true
 
 [dependencies.snarkvm-console-collections]
 path = "./collections"
-version = "=1.6.0"
+version = "=3.7.1"
 optional = true
 
 [dependencies.snarkvm-console-network]
 path = "./network"
-version = "=1.6.0"
+version = "=3.7.1"
 optional = true
 
 [dependencies.snarkvm-console-program]
 path = "./program"
-version = "=1.6.0"
+version = "=3.7.1"
 optional = true
 
 [dependencies.snarkvm-console-types]
 path = "./types"
-version = "=1.6.0"
+version = "=3.7.1"
 optional = true
 
 [features]
diff --git a/console/account/Cargo.toml b/console/account/Cargo.toml
index 1250ad2b9..d27dc78fa 100644
--- a/console/account/Cargo.toml
+++ b/console/account/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-console-account"
-version = "1.6.0"
+version = "3.7.1"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Account operations for a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -15,11 +15,11 @@ harness = false
 
 [dependencies.snarkvm-console-network]
 path = "../network"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.snarkvm-console-types]
 path = "../types"
-version = "=1.6.0"
+version = "=3.7.1"
 default-features = false
 features = [ "address", "boolean", "field", "group", "scalar" ]
 
diff --git a/console/algorithms/Cargo.toml b/console/algorithms/Cargo.toml
index 43eb36320..45db10e82 100644
--- a/console/algorithms/Cargo.toml
+++ b/console/algorithms/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-console-algorithms"
-version = "1.6.0"
+version = "3.7.1"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Console algorithms for a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -25,18 +25,18 @@ harness = false
 
 [dependencies.snarkvm-console-types]
 path = "../types"
-version = "=1.6.0"
+version = "=3.7.1"
 default-features = false
 features = [ "field", "group", "integers", "scalar" ]
 
 [dependencies.snarkvm-fields]
 path = "../../fields"
-version = "=1.6.0"
+version = "=3.7.1"
 default-features = false
 
 [dependencies.snarkvm-utilities]
 path = "../../utilities"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.blake2s_simd]
 version = "1.0"
diff --git a/console/collections/Cargo.toml b/console/collections/Cargo.toml
index 428e0de9a..b3336fd6d 100644
--- a/console/collections/Cargo.toml
+++ b/console/collections/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-console-collections"
-version = "1.6.0"
+version = "3.7.1"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Collections for a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -20,11 +20,11 @@ harness = false
 
 [dependencies.snarkvm-console-algorithms]
 path = "../algorithms"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.snarkvm-console-types]
 path = "../types"
-version = "=1.6.0"
+version = "=3.7.1"
 default-features = false
 features = [ "field", "integers" ]
 
diff --git a/console/network/Cargo.toml b/console/network/Cargo.toml
index b2e0b55cb..ea0b36dc2 100644
--- a/console/network/Cargo.toml
+++ b/console/network/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-console-network"
-version = "1.6.0"
+version = "3.7.1"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Network console library for a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -20,45 +20,45 @@ test_consensus_heights = [ ]
 
 [dependencies.snarkvm-algorithms]
 path = "../../algorithms"
-version = "=1.6.0"
+version = "=3.7.1"
 default-features = false
 features = [ "snark" ]
 
 [dependencies.snarkvm-console-algorithms]
 path = "../algorithms"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.snarkvm-console-collections]
 path = "../collections"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.snarkvm-console-network-environment]
 path = "./environment"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.snarkvm-console-types]
 path = "../types"
-version = "=1.6.0"
+version = "=3.7.1"
 default-features = false
 features = [ "field", "group", "scalar" ]
 
 [dependencies.snarkvm-curves]
 path = "../../curves"
-version = "=1.6.0"
+version = "=3.7.1"
 default-features = false
 
 [dependencies.snarkvm-fields]
 path = "../../fields"
-version = "=1.6.0"
+version = "=3.7.1"
 default-features = false
 
 [dependencies.snarkvm-parameters]
 path = "../../parameters"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.snarkvm-utilities]
 path = "../../utilities"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.anyhow]
 version = "1.0.73"
diff --git a/console/network/environment/Cargo.toml b/console/network/environment/Cargo.toml
index d38723cef..d1e3270b7 100644
--- a/console/network/environment/Cargo.toml
+++ b/console/network/environment/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-console-network-environment"
-version = "1.6.0"
+version = "3.7.1"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Environment console library for a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -10,17 +10,17 @@ edition = "2021"
 
 [dependencies.snarkvm-curves]
 path = "../../../curves"
-version = "=1.6.0"
+version = "=3.7.1"
 default-features = false
 
 [dependencies.snarkvm-fields]
 path = "../../../fields"
-version = "=1.6.0"
+version = "=3.7.1"
 default-features = false
 
 [dependencies.snarkvm-utilities]
 path = "../../../utilities"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.anyhow]
 version = "1.0.73"
diff --git a/console/network/src/canary_v0.rs b/console/network/src/canary_v0.rs
index 3427328c6..ae7cae84e 100644
--- a/console/network/src/canary_v0.rs
+++ b/console/network/src/canary_v0.rs
@@ -136,24 +136,26 @@ impl Network for CanaryV0 {
     /// A list of (consensus_version, block_height) pairs indicating when each consensus version takes effect.
     /// Documentation for what is changed at each version can be found in `ConsensusVersion`.
     #[cfg(not(any(test, feature = "test", feature = "test_consensus_heights")))]
-    const CONSENSUS_VERSION_HEIGHTS: [(ConsensusVersion, u32); 6] = [
+    const CONSENSUS_VERSION_HEIGHTS: [(ConsensusVersion, u32); 7] = [
         (ConsensusVersion::V1, 0),
         (ConsensusVersion::V2, 2_900_000),
         (ConsensusVersion::V3, 4_560_000),
         (ConsensusVersion::V4, 5_730_000),
         (ConsensusVersion::V5, 5_780_000),
         (ConsensusVersion::V6, 6_240_000),
+        (ConsensusVersion::V7, 6_895_000),
     ];
     /// A list of (consensus_version, block_height) pairs indicating when each consensus version takes effect.
     /// Documentation for what is changed at each version can be found in `ConsensusVersion`.
     #[cfg(any(test, feature = "test", feature = "test_consensus_heights"))]
-    const CONSENSUS_VERSION_HEIGHTS: [(ConsensusVersion, u32); 6] = [
+    const CONSENSUS_VERSION_HEIGHTS: [(ConsensusVersion, u32); 7] = [
         (ConsensusVersion::V1, 0),
         (ConsensusVersion::V2, 10),
         (ConsensusVersion::V3, 11),
         (ConsensusVersion::V4, 12),
         (ConsensusVersion::V5, 13),
         (ConsensusVersion::V6, 14),
+        (ConsensusVersion::V7, 15),
     ];
     /// The network edition.
     const EDITION: u16 = 0;
diff --git a/console/network/src/lib.rs b/console/network/src/lib.rs
index 2a99eea3a..eea3d7dc8 100644
--- a/console/network/src/lib.rs
+++ b/console/network/src/lib.rs
@@ -85,6 +85,8 @@ pub enum ConsensusVersion {
     V5 = 5,
     /// V6: Update to the number of validators.
     V6 = 6,
+    /// V7: Update to program rules.
+    V7 = 7,
 }
 
 pub trait Network:
@@ -224,7 +226,7 @@ pub trait Network:
 
     /// A list of (consensus_version, block_height) pairs indicating when each consensus version takes effect.
     /// Documentation for what is changed at each version can be found in `N::CONSENSUS_VERSION`
-    const CONSENSUS_VERSION_HEIGHTS: [(ConsensusVersion, u32); 6];
+    const CONSENSUS_VERSION_HEIGHTS: [(ConsensusVersion, u32); 7];
     ///  A list of (consensus_version, size) pairs indicating the maximum number of validators in a committee.
     //  Note: This value must **not** decrease without considering the impact on serialization.
     //  Decreasing this value will break backwards compatibility of serialization without explicit
diff --git a/console/network/src/mainnet_v0.rs b/console/network/src/mainnet_v0.rs
index 11fff1ea1..4e5229046 100644
--- a/console/network/src/mainnet_v0.rs
+++ b/console/network/src/mainnet_v0.rs
@@ -137,24 +137,26 @@ impl Network for MainnetV0 {
     /// A list of (consensus_version, block_height) pairs indicating when each consensus version takes effect.
     /// Documentation for what is changed at each version can be found in `ConsensusVersion`.
     #[cfg(not(any(test, feature = "test")))]
-    const CONSENSUS_VERSION_HEIGHTS: [(ConsensusVersion, u32); 6] = [
+    const CONSENSUS_VERSION_HEIGHTS: [(ConsensusVersion, u32); 7] = [
         (ConsensusVersion::V1, 0),
         (ConsensusVersion::V2, 2_800_000),
         (ConsensusVersion::V3, 4_900_000),
         (ConsensusVersion::V4, 6_135_000),
         (ConsensusVersion::V5, 7_060_000),
-        (ConsensusVersion::V6, 7_900_000),
+        (ConsensusVersion::V6, 7_560_000),
+        (ConsensusVersion::V7, 7_570_000),
     ];
     /// A list of (consensus_version, block_height) pairs indicating when each consensus version takes effect.
     /// Documentation for what is changed at each version can be found in `ConsensusVersion`.
     #[cfg(any(test, feature = "test"))]
-    const CONSENSUS_VERSION_HEIGHTS: [(ConsensusVersion, u32); 6] = [
+    const CONSENSUS_VERSION_HEIGHTS: [(ConsensusVersion, u32); 7] = [
         (ConsensusVersion::V1, 0),
         (ConsensusVersion::V2, 10),
         (ConsensusVersion::V3, 11),
         (ConsensusVersion::V4, 12),
         (ConsensusVersion::V5, 13),
         (ConsensusVersion::V6, 14),
+        (ConsensusVersion::V7, 15),
     ];
     /// The network edition.
     const EDITION: u16 = 0;
diff --git a/console/network/src/testnet_v0.rs b/console/network/src/testnet_v0.rs
index 010e60938..6bec3d4b4 100644
--- a/console/network/src/testnet_v0.rs
+++ b/console/network/src/testnet_v0.rs
@@ -136,24 +136,26 @@ impl Network for TestnetV0 {
     /// A list of (consensus_version, block_height) pairs indicating when each consensus version takes effect.
     /// Documentation for what is changed at each version can be found in `ConsensusVersion`.
     #[cfg(not(any(test, feature = "test", feature = "test_consensus_heights")))]
-    const CONSENSUS_VERSION_HEIGHTS: [(ConsensusVersion, u32); 6] = [
+    const CONSENSUS_VERSION_HEIGHTS: [(ConsensusVersion, u32); 7] = [
         (ConsensusVersion::V1, 0),
         (ConsensusVersion::V2, 2_950_000),
         (ConsensusVersion::V3, 4_800_000),
         (ConsensusVersion::V4, 6_625_000),
         (ConsensusVersion::V5, 6_765_000),
         (ConsensusVersion::V6, 7_600_000),
+        (ConsensusVersion::V7, 8_365_000),
     ];
     /// A list of (consensus_version, block_height) pairs indicating when each consensus version takes effect.
     /// Documentation for what is changed at each version can be found in `ConsensusVersion`.
     #[cfg(any(test, feature = "test", feature = "test_consensus_heights"))]
-    const CONSENSUS_VERSION_HEIGHTS: [(ConsensusVersion, u32); 6] = [
+    const CONSENSUS_VERSION_HEIGHTS: [(ConsensusVersion, u32); 7] = [
         (ConsensusVersion::V1, 0),
         (ConsensusVersion::V2, 10),
         (ConsensusVersion::V3, 11),
         (ConsensusVersion::V4, 12),
         (ConsensusVersion::V5, 13),
         (ConsensusVersion::V6, 14),
+        (ConsensusVersion::V7, 15),
     ];
     /// The network edition.
     const EDITION: u16 = 0;
diff --git a/console/program/Cargo.toml b/console/program/Cargo.toml
index b8958b3c8..2f9fb5c8f 100644
--- a/console/program/Cargo.toml
+++ b/console/program/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-console-program"
-version = "1.6.0"
+version = "3.7.1"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Program operations for a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -14,27 +14,27 @@ test = [ ]
 
 [dependencies.snarkvm-console-account]
 path = "../account"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.snarkvm-console-algorithms]
 path = "../algorithms"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.snarkvm-console-collections]
 path = "../collections"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.snarkvm-console-network]
 path = "../network"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.snarkvm-console-types]
 path = "../types"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.snarkvm-utilities]
 path = "../../utilities"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.enum_index]
 version = "0.2"
diff --git a/console/types/Cargo.toml b/console/types/Cargo.toml
index 8ffa02d8c..1dabcc0e6 100644
--- a/console/types/Cargo.toml
+++ b/console/types/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-console-types"
-version = "1.6.0"
+version = "3.7.1"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Console types for a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -15,41 +15,41 @@ harness = false
 
 [dependencies.snarkvm-console-network-environment]
 path = "../network/environment"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.snarkvm-console-types-address]
 path = "./address"
-version = "=1.6.0"
+version = "=3.7.1"
 optional = true
 
 [dependencies.snarkvm-console-types-boolean]
 path = "./boolean"
-version = "=1.6.0"
+version = "=3.7.1"
 optional = true
 
 [dependencies.snarkvm-console-types-field]
 path = "./field"
-version = "=1.6.0"
+version = "=3.7.1"
 optional = true
 
 [dependencies.snarkvm-console-types-group]
 path = "./group"
-version = "=1.6.0"
+version = "=3.7.1"
 optional = true
 
 [dependencies.snarkvm-console-types-integers]
 path = "./integers"
-version = "=1.6.0"
+version = "=3.7.1"
 optional = true
 
 [dependencies.snarkvm-console-types-scalar]
 path = "./scalar"
-version = "=1.6.0"
+version = "=3.7.1"
 optional = true
 
 [dependencies.snarkvm-console-types-string]
 path = "./string"
-version = "=1.6.0"
+version = "=3.7.1"
 optional = true
 
 [dev-dependencies.criterion]
diff --git a/console/types/address/Cargo.toml b/console/types/address/Cargo.toml
index d6c751c55..db9bb4977 100644
--- a/console/types/address/Cargo.toml
+++ b/console/types/address/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-console-types-address"
-version = "1.6.0"
+version = "3.7.1"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Type operations for a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -10,19 +10,19 @@ edition = "2021"
 
 [dependencies.snarkvm-console-network-environment]
 path = "../../network/environment"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.snarkvm-console-types-boolean]
 path = "../boolean"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.snarkvm-console-types-field]
 path = "../field"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.snarkvm-console-types-group]
 path = "../group"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dev-dependencies.bincode]
 version = "1.3"
diff --git a/console/types/boolean/Cargo.toml b/console/types/boolean/Cargo.toml
index 10298924b..2ebcec7ee 100644
--- a/console/types/boolean/Cargo.toml
+++ b/console/types/boolean/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-console-types-boolean"
-version = "1.6.0"
+version = "3.7.1"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Type operations for a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -10,7 +10,7 @@ edition = "2021"
 
 [dependencies.snarkvm-console-network-environment]
 path = "../../network/environment"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dev-dependencies.bincode]
 version = "1.3"
diff --git a/console/types/field/Cargo.toml b/console/types/field/Cargo.toml
index c54073b98..947841d1c 100644
--- a/console/types/field/Cargo.toml
+++ b/console/types/field/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-console-types-field"
-version = "1.6.0"
+version = "3.7.1"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Type operations for a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -10,11 +10,11 @@ edition = "2021"
 
 [dependencies.snarkvm-console-network-environment]
 path = "../../network/environment"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.snarkvm-console-types-boolean]
 path = "../boolean"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.zeroize]
 version = "1"
diff --git a/console/types/group/Cargo.toml b/console/types/group/Cargo.toml
index 277912161..2f83eb7b9 100644
--- a/console/types/group/Cargo.toml
+++ b/console/types/group/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-console-types-group"
-version = "1.6.0"
+version = "3.7.1"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Type operations for a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -10,19 +10,19 @@ edition = "2021"
 
 [dependencies.snarkvm-console-network-environment]
 path = "../../network/environment"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.snarkvm-console-types-boolean]
 path = "../boolean"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.snarkvm-console-types-field]
 path = "../field"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.snarkvm-console-types-scalar]
 path = "../scalar"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dev-dependencies.bincode]
 version = "1.3"
diff --git a/console/types/integers/Cargo.toml b/console/types/integers/Cargo.toml
index a6d98d13c..a4e10eed9 100644
--- a/console/types/integers/Cargo.toml
+++ b/console/types/integers/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-console-types-integers"
-version = "1.6.0"
+version = "3.7.1"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Type operations for a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -10,19 +10,19 @@ edition = "2021"
 
 [dependencies.snarkvm-console-network-environment]
 path = "../../network/environment"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.snarkvm-console-types-boolean]
 path = "../boolean"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.snarkvm-console-types-field]
 path = "../field"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.snarkvm-console-types-scalar]
 path = "../scalar"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dev-dependencies.bincode]
 version = "1.3"
diff --git a/console/types/scalar/Cargo.toml b/console/types/scalar/Cargo.toml
index 6c8126047..14a2400f6 100644
--- a/console/types/scalar/Cargo.toml
+++ b/console/types/scalar/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-console-types-scalar"
-version = "1.6.0"
+version = "3.7.1"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Type operations for a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -10,15 +10,15 @@ edition = "2021"
 
 [dependencies.snarkvm-console-network-environment]
 path = "../../network/environment"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.snarkvm-console-types-boolean]
 path = "../boolean"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.snarkvm-console-types-field]
 path = "../field"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.zeroize]
 version = "1"
diff --git a/console/types/string/Cargo.toml b/console/types/string/Cargo.toml
index 94d81b69c..7bf806c22 100644
--- a/console/types/string/Cargo.toml
+++ b/console/types/string/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-console-types-string"
-version = "1.6.0"
+version = "3.7.1"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Type operations for a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -10,19 +10,19 @@ edition = "2021"
 
 [dependencies.snarkvm-console-network-environment]
 path = "../../network/environment"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.snarkvm-console-types-boolean]
 path = "../boolean"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.snarkvm-console-types-field]
 path = "../field"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.snarkvm-console-types-integers]
 path = "../integers"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dev-dependencies.bincode]
 version = "1.3"
diff --git a/curves/Cargo.toml b/curves/Cargo.toml
index 825653871..5a01153ab 100644
--- a/curves/Cargo.toml
+++ b/curves/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-curves"
-version = "1.6.0"
+version = "3.7.1"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Curves for a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -36,12 +36,12 @@ harness = false
 
 [dependencies.snarkvm-fields]
 path = "../fields"
-version = "=1.6.0"
+version = "=3.7.1"
 default-features = false
 
 [dependencies.snarkvm-utilities]
 path = "../utilities"
-version = "=1.6.0"
+version = "=3.7.1"
 default-features = false
 
 [dependencies.rand]
diff --git a/curves/src/traits/tests_field.rs b/curves/src/traits/tests_field.rs
index ae0a5d556..03eb7e393 100644
--- a/curves/src/traits/tests_field.rs
+++ b/curves/src/traits/tests_field.rs
@@ -103,7 +103,7 @@ fn random_inversion_tests<F: Field, R: Rng>(rng: &mut R) {
 
     for _ in 0..ITERATIONS {
         let mut a = F::rand(rng);
-        let b = a.inverse().unwrap(); // probablistically nonzero
+        let b = a.inverse().unwrap(); // probabilistically nonzero
         a *= &b;
 
         assert_eq!(a, F::one());
diff --git a/fields/Cargo.toml b/fields/Cargo.toml
index cbd5e13bc..b2bbd474f 100644
--- a/fields/Cargo.toml
+++ b/fields/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-fields"
-version = "1.6.0"
+version = "3.7.1"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Fields for a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -30,7 +30,7 @@ harness = false
 
 [dependencies.snarkvm-utilities]
 path = "../utilities"
-version = "=1.6.0"
+version = "=3.7.1"
 default-features = false
 
 [dependencies.aleo-std]
diff --git a/ledger/Cargo.toml b/ledger/Cargo.toml
index c80e02191..0c5484f9b 100644
--- a/ledger/Cargo.toml
+++ b/ledger/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-ledger"
-version = "1.6.0"
+version = "3.7.1"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "A node ledger for a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -68,54 +68,54 @@ timer = [ "aleo-std/timer" ]
 [dependencies.console]
 package = "snarkvm-console"
 path = "../console"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.ledger-authority]
 package = "snarkvm-ledger-authority"
 path = "./authority"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.ledger-block]
 package = "snarkvm-ledger-block"
 path = "./block"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.ledger-committee]
 package = "snarkvm-ledger-committee"
 path = "./committee"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.ledger-narwhal]
 package = "snarkvm-ledger-narwhal"
 path = "./narwhal"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.ledger-puzzle]
 package = "snarkvm-ledger-puzzle"
 path = "puzzle"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.ledger-query]
 package = "snarkvm-ledger-query"
 path = "./query"
-version = "=1.6.0"
+version = "=3.7.1"
 features = [ "query" ]
 
 [dependencies.ledger-store]
 package = "snarkvm-ledger-store"
 path = "./store"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.ledger-test-helpers]
 package = "snarkvm-ledger-test-helpers"
 path = "./test-helpers"
-version = "=1.6.0"
+version = "=3.7.1"
 optional = true
 
 [dependencies.synthesizer]
 package = "snarkvm-synthesizer"
 path = "../synthesizer"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.aleo-std]
 version = "1.0.1"
diff --git a/ledger/authority/Cargo.toml b/ledger/authority/Cargo.toml
index 0854d5995..f18fc371d 100644
--- a/ledger/authority/Cargo.toml
+++ b/ledger/authority/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-ledger-authority"
-version = "1.6.0"
+version = "3.7.1"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Data structures for a block authority in a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -32,12 +32,12 @@ test-helpers = [ "narwhal-subdag/test-helpers" ]
 [dependencies.console]
 package = "snarkvm-console"
 path = "../../console"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.narwhal-subdag]
 package = "snarkvm-ledger-narwhal-subdag"
 path = "../narwhal/subdag"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.anyhow]
 version = "1"
diff --git a/ledger/block/Cargo.toml b/ledger/block/Cargo.toml
index fdce126b8..b3c071ed5 100644
--- a/ledger/block/Cargo.toml
+++ b/ledger/block/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-ledger-block"
-version = "1.6.0"
+version = "3.7.1"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "A block for a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -39,52 +39,52 @@ test = [ ]
 [dependencies.console]
 package = "snarkvm-console"
 path = "../../console"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.ledger-authority]
 package = "snarkvm-ledger-authority"
 path = "../authority"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.ledger-committee]
 package = "snarkvm-ledger-committee"
 path = "../../ledger/committee"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.ledger-narwhal-batch-header]
 package = "snarkvm-ledger-narwhal-batch-header"
 path = "../narwhal/batch-header"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.ledger-narwhal-data]
 package = "snarkvm-ledger-narwhal-data"
 path = "../narwhal/data"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.ledger-narwhal-subdag]
 package = "snarkvm-ledger-narwhal-subdag"
 path = "../narwhal/subdag"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.ledger-narwhal-transmission-id]
 package = "snarkvm-ledger-narwhal-transmission-id"
 path = "../narwhal/transmission-id"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.ledger-puzzle]
 package = "snarkvm-ledger-puzzle"
 path = "../puzzle"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.synthesizer-program]
 package = "snarkvm-synthesizer-program"
 path = "../../synthesizer/program"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.synthesizer-snark]
 package = "snarkvm-synthesizer-snark"
 path = "../../synthesizer/snark"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.indexmap]
 version = "2.0"
diff --git a/ledger/committee/Cargo.toml b/ledger/committee/Cargo.toml
index f0a72fa08..84ffd0afe 100644
--- a/ledger/committee/Cargo.toml
+++ b/ledger/committee/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-ledger-committee"
-version = "1.6.0"
+version = "3.7.1"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "A committee for a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -34,12 +34,12 @@ test-helpers = [ "prop-tests", "rand_distr" ]
 [dependencies.console]
 package = "snarkvm-console"
 path = "../../console"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.ledger-narwhal-batch-header]
 package = "snarkvm-ledger-narwhal-batch-header"
 path = "../narwhal/batch-header"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.indexmap]
 version = "2.0"
@@ -48,7 +48,7 @@ features = [ "serde", "rayon" ]
 [dependencies.metrics]
 package = "snarkvm-metrics"
 path = "../../metrics"
-version = "=1.6.0"
+version = "=3.7.1"
 optional = true
 
 [dependencies.serde_json]
diff --git a/ledger/narwhal/Cargo.toml b/ledger/narwhal/Cargo.toml
index 7799451d1..25c08ac62 100644
--- a/ledger/narwhal/Cargo.toml
+++ b/ledger/narwhal/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-ledger-narwhal"
-version = "1.6.0"
+version = "3.7.1"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Data structures for a Narwhal-style memory pool in a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -64,37 +64,37 @@ transmission-id = [ "narwhal-transmission-id" ]
 [dependencies.narwhal-batch-certificate]
 package = "snarkvm-ledger-narwhal-batch-certificate"
 path = "./batch-certificate"
-version = "=1.6.0"
+version = "=3.7.1"
 optional = true
 
 [dependencies.narwhal-batch-header]
 package = "snarkvm-ledger-narwhal-batch-header"
 path = "./batch-header"
-version = "=1.6.0"
+version = "=3.7.1"
 optional = true
 
 [dependencies.narwhal-data]
 package = "snarkvm-ledger-narwhal-data"
 path = "./data"
-version = "=1.6.0"
+version = "=3.7.1"
 optional = true
 
 [dependencies.narwhal-subdag]
 package = "snarkvm-ledger-narwhal-subdag"
 path = "./subdag"
-version = "=1.6.0"
+version = "=3.7.1"
 optional = true
 
 [dependencies.narwhal-transmission]
 package = "snarkvm-ledger-narwhal-transmission"
 path = "./transmission"
-version = "=1.6.0"
+version = "=3.7.1"
 optional = true
 
 [dependencies.narwhal-transmission-id]
 package = "snarkvm-ledger-narwhal-transmission-id"
 path = "./transmission-id"
-version = "=1.6.0"
+version = "=3.7.1"
 optional = true
 
 [dev-dependencies.snarkvm-ledger-narwhal]
diff --git a/ledger/narwhal/batch-certificate/Cargo.toml b/ledger/narwhal/batch-certificate/Cargo.toml
index 50768d045..62874a889 100644
--- a/ledger/narwhal/batch-certificate/Cargo.toml
+++ b/ledger/narwhal/batch-certificate/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-ledger-narwhal-batch-certificate"
-version = "1.6.0"
+version = "3.7.1"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "A batch certificate for a Narwhal-style memory pool in a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -32,17 +32,17 @@ test-helpers = [ "narwhal-batch-header/test-helpers" ]
 [dependencies.console]
 package = "snarkvm-console"
 path = "../../../console"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.narwhal-batch-header]
 package = "snarkvm-ledger-narwhal-batch-header"
 path = "../batch-header"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.narwhal-transmission-id]
 package = "snarkvm-ledger-narwhal-transmission-id"
 path = "../transmission-id"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.indexmap]
 version = "2.0"
diff --git a/ledger/narwhal/batch-header/Cargo.toml b/ledger/narwhal/batch-header/Cargo.toml
index 78f04f90d..02a0eb2bd 100644
--- a/ledger/narwhal/batch-header/Cargo.toml
+++ b/ledger/narwhal/batch-header/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-ledger-narwhal-batch-header"
-version = "1.6.0"
+version = "3.7.1"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "A batch header for a Narwhal-style memory pool in a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -32,12 +32,12 @@ test-helpers = [ "narwhal-transmission-id/test-helpers", "time" ]
 [dependencies.console]
 package = "snarkvm-console"
 path = "../../../console"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.narwhal-transmission-id]
 package = "snarkvm-ledger-narwhal-transmission-id"
 path = "../transmission-id"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.indexmap]
 version = "2.0"
diff --git a/ledger/narwhal/data/Cargo.toml b/ledger/narwhal/data/Cargo.toml
index dfe9c8b22..32a27c684 100644
--- a/ledger/narwhal/data/Cargo.toml
+++ b/ledger/narwhal/data/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-ledger-narwhal-data"
-version = "1.6.0"
+version = "3.7.1"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "A batch certificate for a Narwhal-style memory pool in a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -29,7 +29,7 @@ async = [ "tokio" ]
 [dependencies.console]
 package = "snarkvm-console"
 path = "../../../console"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.bytes]
 version = "1"
diff --git a/ledger/narwhal/subdag/Cargo.toml b/ledger/narwhal/subdag/Cargo.toml
index e806037aa..bdcd61e76 100644
--- a/ledger/narwhal/subdag/Cargo.toml
+++ b/ledger/narwhal/subdag/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-ledger-narwhal-subdag"
-version = "1.6.0"
+version = "3.7.1"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "A subdag for a Narwhal-style memory pool in a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -32,27 +32,27 @@ test-helpers = [ "narwhal-batch-certificate/test-helpers" ]
 [dependencies.console]
 package = "snarkvm-console"
 path = "../../../console"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.narwhal-batch-certificate]
 package = "snarkvm-ledger-narwhal-batch-certificate"
 path = "../batch-certificate"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.narwhal-batch-header]
 package = "snarkvm-ledger-narwhal-batch-header"
 path = "../batch-header"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.ledger-committee]
 package = "snarkvm-ledger-committee"
 path = "../../committee"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.narwhal-transmission-id]
 package = "snarkvm-ledger-narwhal-transmission-id"
 path = "../transmission-id"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.indexmap]
 version = "2.0"
diff --git a/ledger/narwhal/transmission-id/Cargo.toml b/ledger/narwhal/transmission-id/Cargo.toml
index 8b9dc31ad..216aaef1b 100644
--- a/ledger/narwhal/transmission-id/Cargo.toml
+++ b/ledger/narwhal/transmission-id/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-ledger-narwhal-transmission-id"
-version = "1.6.0"
+version = "3.7.1"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "A transmission ID for a Narwhal-style memory pool in a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -32,12 +32,12 @@ test-helpers = [ ]
 [dependencies.console]
 package = "snarkvm-console"
 path = "../../../console"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.ledger-puzzle]
 package = "snarkvm-ledger-puzzle"
 path = "../../puzzle"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dev-dependencies.bincode]
 version = "1.3"
diff --git a/ledger/narwhal/transmission/Cargo.toml b/ledger/narwhal/transmission/Cargo.toml
index 5bff73cd9..b22d231bf 100644
--- a/ledger/narwhal/transmission/Cargo.toml
+++ b/ledger/narwhal/transmission/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-ledger-narwhal-transmission"
-version = "1.6.0"
+version = "3.7.1"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "A transmission for a Narwhal-style memory pool in a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -32,22 +32,22 @@ test-helpers = [ ]
 [dependencies.console]
 package = "snarkvm-console"
 path = "../../../console"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.ledger-block]
 package = "snarkvm-ledger-block"
 path = "../../block"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.ledger-narwhal-data]
 package = "snarkvm-ledger-narwhal-data"
 path = "../data"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.ledger-puzzle]
 package = "snarkvm-ledger-puzzle"
 path = "../../puzzle"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.bytes]
 version = "1"
diff --git a/ledger/puzzle/Cargo.toml b/ledger/puzzle/Cargo.toml
index 54f121717..e9bbbaf0b 100644
--- a/ledger/puzzle/Cargo.toml
+++ b/ledger/puzzle/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-ledger-puzzle"
-version = "1.6.0"
+version = "3.7.1"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Puzzle for a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -41,11 +41,11 @@ wasm = [ "console/wasm", "snarkvm-algorithms/wasm" ]
 [dependencies.console]
 package = "snarkvm-console"
 path = "../../console"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.snarkvm-algorithms]
 path = "../../algorithms"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.aleo-std]
 version = "1.0.1"
diff --git a/ledger/puzzle/epoch/Cargo.toml b/ledger/puzzle/epoch/Cargo.toml
index 7320ea849..50af6a9b0 100644
--- a/ledger/puzzle/epoch/Cargo.toml
+++ b/ledger/puzzle/epoch/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-ledger-puzzle-epoch"
-version = "1.6.0"
+version = "3.7.1"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Epoch puzzle for a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -46,26 +46,26 @@ wasm = [ "console/wasm" ]
 [dependencies.circuit]
 package = "snarkvm-circuit"
 path = "../../../circuit"
-version = "=1.6.0"
+version = "=3.7.1"
 optional = true
 
 [dependencies.console]
 package = "snarkvm-console"
 path = "../../../console"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.snarkvm-ledger-puzzle]
 path = "../."
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.snarkvm-synthesizer-process]
 path = "../../../synthesizer/process"
-version = "=1.6.0"
+version = "=3.7.1"
 optional = true
 
 [dependencies.snarkvm-synthesizer-program]
 path = "../../../synthesizer/program"
-version = "=1.6.0"
+version = "=3.7.1"
 optional = true
 
 [dependencies.aleo-std]
diff --git a/ledger/query/Cargo.toml b/ledger/query/Cargo.toml
index 1da0a12b3..6692c6c30 100644
--- a/ledger/query/Cargo.toml
+++ b/ledger/query/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-ledger-query"
-version = "1.6.0"
+version = "3.7.1"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "A query for a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -34,18 +34,18 @@ query = [ "ledger-store", "synthesizer-program", "ureq" ]
 [dependencies.console]
 package = "snarkvm-console"
 path = "../../console"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.ledger-store]
 package = "snarkvm-ledger-store"
 path = "../store"
-version = "=1.6.0"
+version = "=3.7.1"
 optional = true
 
 [dependencies.synthesizer-program]
 package = "snarkvm-synthesizer-program"
 path = "../../synthesizer/program"
-version = "=1.6.0"
+version = "=3.7.1"
 optional = true
 
 [dependencies.async-trait]
diff --git a/ledger/src/check_next_block.rs b/ledger/src/check_next_block.rs
index 04679d66a..67154e4bc 100644
--- a/ledger/src/check_next_block.rs
+++ b/ledger/src/check_next_block.rs
@@ -137,7 +137,7 @@ impl<N: Network, C: ConsensusStorage<N>> Ledger<N, C> {
     /// Check that leaves in the subdag point to batches in other blocks that are valid.
     ///
     /// This does not verify that the batches are signed correctly or that the edges are valid
-    /// (only point to the previous round), as those checks already happend when the node received the batch.
+    /// (only point to the previous round), as those checks already happened when the node received the batch.
     fn check_block_subdag_leaves(&self, block: &Block<N>) -> Result<()> {
         // Check if the block has a subdag.
         let Authority::Quorum(subdag) = block.authority() else {
diff --git a/ledger/src/tests.rs b/ledger/src/tests.rs
index e7524def2..7c5e5efd8 100644
--- a/ledger/src/tests.rs
+++ b/ledger/src/tests.rs
@@ -68,7 +68,7 @@ fn extract_transmissions(
     transmissions
 }
 
-/// Helper to build chains with custom sturctures for testing
+/// Helper to build chains with custom structures for testing
 struct TestChainBuilder {
     /// The keys of all validators.
     private_keys: Vec<PrivateKey<CurrentNetwork>>,
@@ -91,7 +91,7 @@ struct TestChainBuilder {
 }
 
 impl TestChainBuilder {
-    /// Initialize the builder with the specified commitee and gensis block
+    /// Initialize the builder with the specified committee and genesis block
     pub fn new(private_keys: Vec<PrivateKey<CurrentNetwork>>, genesis: Block<CurrentNetwork>) -> Self {
         // Initialize the ledger with the genesis block.
         let ledger =
@@ -147,7 +147,7 @@ impl TestChainBuilder {
 
         let block_height = self.current_height + 1;
 
-        // SubDAGs can be at most GC roudns long.
+        // SubDAGs can be at most GC rounds long.
         let mut round = if self.last_block_round < BatchHeader::<CurrentNetwork>::MAX_GC_ROUNDS as u64 {
             1
         } else {
@@ -773,7 +773,7 @@ fn test_bond_and_unbond_validator() {
     let new_member_withdrawal_private_key = PrivateKey::<CurrentNetwork>::new(rng).unwrap();
     let new_member_withdrawal_address = Address::try_from(&new_member_withdrawal_private_key).unwrap();
 
-    // Fund the new committee member and their withdrawwal address.
+    // Fund the new committee member and their withdrawal address.
     let inputs = [
         Value::from_str(&format!("{new_member_address}")).unwrap(),
         Value::from_str("20000000000000u64").unwrap(), // 20 million credits.
@@ -840,7 +840,7 @@ fn test_bond_and_unbond_validator() {
     let committee = ledger.latest_committee().unwrap();
     assert!(committee.is_committee_member(new_member_address));
 
-    // Check that number of validators in the `metadata` mapping in `credtis.aleo` is updated.
+    // Check that number of validators in the `metadata` mapping in `credits.aleo` is updated.
     let program_id = ProgramID::<CurrentNetwork>::from_str("credits.aleo").unwrap();
     let metadata_mapping_name = Identifier::from_str("metadata").unwrap();
     let key = Plaintext::<CurrentNetwork>::from_str("aleo1qqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqq3ljyzc")
diff --git a/ledger/store/Cargo.toml b/ledger/store/Cargo.toml
index 86803caa0..1c531a193 100644
--- a/ledger/store/Cargo.toml
+++ b/ledger/store/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-ledger-store"
-version = "1.6.0"
+version = "3.7.1"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "A data store for a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -43,42 +43,42 @@ test = [ ]
 [dependencies.console]
 package = "snarkvm-console"
 path = "../../console"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.ledger-authority]
 package = "snarkvm-ledger-authority"
 path = "../authority"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.ledger-block]
 package = "snarkvm-ledger-block"
 path = "../block"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.ledger-committee]
 package = "snarkvm-ledger-committee"
 path = "../committee"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.ledger-narwhal-batch-certificate]
 package = "snarkvm-ledger-narwhal-batch-certificate"
 path = "../narwhal/batch-certificate"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.ledger-puzzle]
 package = "snarkvm-ledger-puzzle"
 path = "../puzzle"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.synthesizer-program]
 package = "snarkvm-synthesizer-program"
 path = "../../synthesizer/program"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.synthesizer-snark]
 package = "snarkvm-synthesizer-snark"
 path = "../../synthesizer/snark"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.aleo-std-storage]
 version = "1.0.1"
diff --git a/ledger/test-helpers/Cargo.toml b/ledger/test-helpers/Cargo.toml
index 31ad09566..3358624b0 100644
--- a/ledger/test-helpers/Cargo.toml
+++ b/ledger/test-helpers/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-ledger-test-helpers"
-version = "1.6.0"
+version = "3.7.1"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Test helpers for a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -19,44 +19,44 @@ edition = "2021"
 [dependencies.algorithms]
 package = "snarkvm-algorithms"
 path = "../../algorithms"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.console]
 package = "snarkvm-console"
 path = "../../console"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.circuit]
 package = "snarkvm-circuit"
 path = "../../circuit"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.ledger-block]
 package = "snarkvm-ledger-block"
 path = "../block"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.ledger-query]
 package = "snarkvm-ledger-query"
 path = "../query"
-version = "=1.6.0"
+version = "=3.7.1"
 default-features = false
 features = [ "query" ]
 
 [dependencies.ledger-store]
 package = "snarkvm-ledger-store"
 path = "../store"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.synthesizer-program]
 package = "snarkvm-synthesizer-program"
 path = "../../synthesizer/program"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.synthesizer-process]
 package = "snarkvm-synthesizer-process"
 path = "../../synthesizer/process"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.aleo-std]
 version = "1.0.1"
diff --git a/metrics/Cargo.toml b/metrics/Cargo.toml
index 78fa002dd..39a63d4e0 100644
--- a/metrics/Cargo.toml
+++ b/metrics/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-metrics"
-version = "1.6.0"
+version = "3.7.1"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Metrics for a decentralized virtual machine"
 homepage = "https://aleo.org"
diff --git a/parameters/Cargo.toml b/parameters/Cargo.toml
index c71f1b19f..e2a6fb10f 100644
--- a/parameters/Cargo.toml
+++ b/parameters/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-parameters"
-version = "1.6.0"
+version = "3.7.1"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Parameters for a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -34,12 +34,12 @@ wasm = [ "encoding", "js-sys", "web-sys" ]
 
 [dependencies.snarkvm-curves]
 path = "../curves"
-version = "=1.6.0"
+version = "=3.7.1"
 default-features = false
 
 [dependencies.snarkvm-utilities]
 path = "../utilities"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.aleo-std]
 version = "1.0.1"
diff --git a/synthesizer/Cargo.toml b/synthesizer/Cargo.toml
index b839a1ed1..c4508d763 100644
--- a/synthesizer/Cargo.toml
+++ b/synthesizer/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-synthesizer"
-version = "1.6.0"
+version = "3.7.1"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Synthesizer for a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -76,78 +76,78 @@ harness = false
 [dependencies.algorithms]
 package = "snarkvm-algorithms"
 path = "../algorithms"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.circuit]
 package = "snarkvm-circuit"
 path = "../circuit"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.console]
 package = "snarkvm-console"
 path = "../console"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.ledger-block]
 package = "snarkvm-ledger-block"
 path = "../ledger/block"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.ledger-committee]
 package = "snarkvm-ledger-committee"
 path = "../ledger/committee"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.ledger-narwhal-data]
 package = "snarkvm-ledger-narwhal-data"
 path = "../ledger/narwhal/data"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.ledger-puzzle]
 package = "snarkvm-ledger-puzzle"
 path = "../ledger/puzzle"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.ledger-puzzle-epoch]
 package = "snarkvm-ledger-puzzle-epoch"
 path = "../ledger/puzzle/epoch"
-version = "=1.6.0"
+version = "=3.7.1"
 features = [ "synthesis" ]
 
 [dependencies.ledger-query]
 package = "snarkvm-ledger-query"
 path = "../ledger/query"
-version = "=1.6.0"
+version = "=3.7.1"
 default-features = false
 features = [ "query" ]
 
 [dependencies.ledger-store]
 package = "snarkvm-ledger-store"
 path = "../ledger/store"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.synthesizer-process]
 package = "snarkvm-synthesizer-process"
 path = "./process"
-version = "=1.6.0"
+version = "=3.7.1"
 optional = true
 
 [dependencies.synthesizer-program]
 package = "snarkvm-synthesizer-program"
 path = "./program"
-version = "=1.6.0"
+version = "=3.7.1"
 optional = true
 
 [dependencies.synthesizer-snark]
 package = "snarkvm-synthesizer-snark"
 path = "./snark"
-version = "=1.6.0"
+version = "=3.7.1"
 optional = true
 
 [dependencies.utilities]
 package = "snarkvm-utilities"
 path = "../utilities"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.aleo-std]
 version = "1.0.1"
diff --git a/synthesizer/process/Cargo.toml b/synthesizer/process/Cargo.toml
index 6e0493967..1c0e39498 100644
--- a/synthesizer/process/Cargo.toml
+++ b/synthesizer/process/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-synthesizer-process"
-version = "1.6.0"
+version = "3.7.1"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "A process for a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -59,50 +59,50 @@ harness = false
 [dependencies.algorithms]
 package = "snarkvm-algorithms"
 path = "../../algorithms"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.console]
 package = "snarkvm-console"
 path = "../../console"
-version = "=1.6.0"
+version = "=3.7.1"
 default-features = false
 features = [ "network", "program", "types" ]
 
 [dependencies.circuit]
 package = "snarkvm-circuit"
 path = "../../circuit"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.ledger-block]
 package = "snarkvm-ledger-block"
 path = "../../ledger/block"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.ledger-query]
 package = "snarkvm-ledger-query"
 path = "../../ledger/query"
-version = "=1.6.0"
+version = "=3.7.1"
 default-features = false
 
 [dependencies.ledger-store]
 package = "snarkvm-ledger-store"
 path = "../../ledger/store"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.synthesizer-program]
 package = "snarkvm-synthesizer-program"
 path = "../../synthesizer/program"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.synthesizer-snark]
 package = "snarkvm-synthesizer-snark"
 path = "../../synthesizer/snark"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.utilities]
 package = "snarkvm-utilities"
 path = "../../utilities"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.aleo-std]
 version = "1.0.1"
diff --git a/synthesizer/program/Cargo.toml b/synthesizer/program/Cargo.toml
index b98ad6fb8..94ae8ca8f 100644
--- a/synthesizer/program/Cargo.toml
+++ b/synthesizer/program/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-synthesizer-program"
-version = "1.6.0"
+version = "3.7.1"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Program for a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -31,12 +31,12 @@ wasm = [ "console/wasm" ]
 [dependencies.circuit]
 package = "snarkvm-circuit"
 path = "../../circuit"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.console]
 package = "snarkvm-console"
 path = "../../console"
-version = "=1.6.0"
+version = "=3.7.1"
 default-features = false
 features = [ "account", "network", "program", "types" ]
 
@@ -60,7 +60,7 @@ features = [ "preserve_order" ]
 [dependencies.synthesizer-snark]
 package = "snarkvm-synthesizer-snark"
 path = "../snark"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dev-dependencies.bincode]
 version = "1"
diff --git a/synthesizer/program/src/lib.rs b/synthesizer/program/src/lib.rs
index 26fa9fa16..a7f70a0e7 100644
--- a/synthesizer/program/src/lib.rs
+++ b/synthesizer/program/src/lib.rs
@@ -94,6 +94,7 @@ use console::{
 };
 
 use indexmap::{IndexMap, IndexSet};
+use std::collections::BTreeSet;
 
 #[derive(Copy, Clone, PartialEq, Eq, Hash)]
 enum ProgramDefinition {
@@ -738,6 +739,52 @@ impl<N: Network, Instruction: InstructionTrait<N>, Command: CommandTrait<N>> Pro
     }
 }
 
+impl<N: Network, Instruction: InstructionTrait<N>, Command: CommandTrait<N>> ProgramCore<N, Instruction, Command> {
+    /// Returns `true` if the program structure is well formed under the following rules:
+    ///  1. The program ID must not contain the keyword "aleo" in the program name.
+    ///  2. The record name must not contain the keyword "aleo".
+    ///  3. Record names must not be prefixes of other record names.
+    ///  4. Record entry names must not contain the keyword "aleo".
+    pub fn check_program_naming_structure(&self) -> Result<()> {
+        // 1. Check if the program ID contains the "aleo" substring
+        let program_id = self.id().name().to_string();
+        if program_id.contains("aleo") {
+            bail!("Program ID '{program_id}' can't contain the reserved keyword 'aleo'.");
+        }
+
+        // Fetch the record names in a sorted BTreeSet.
+        let record_names: BTreeSet<String> = self.records.keys().map(|name| name.to_string()).collect();
+
+        // 2. Check if any record name contains the "aleo" substring.
+        for record_name in &record_names {
+            if record_name.contains("aleo") {
+                bail!("Record name '{record_name}' can't contain the reserved keyword 'aleo'.");
+            }
+        }
+
+        // 3. Check if any of the record names are a prefix of another.
+        let mut record_names_iter = record_names.iter();
+        let mut previous_record_name = record_names_iter.next();
+        for record_name in record_names_iter {
+            if let Some(previous) = previous_record_name {
+                if record_name.starts_with(previous) {
+                    bail!("Record name '{previous}' can't be a prefix of record name '{record_name}'.");
+                }
+            }
+            previous_record_name = Some(record_name);
+        }
+
+        // 4. Check if any record entry names contain the "aleo" substring.
+        for record_entry_name in self.records.values().flat_map(|record_type| record_type.entries().keys()) {
+            if record_entry_name.to_string().contains("aleo") {
+                bail!("Record entry name '{record_entry_name}' can't contain the reserved keyword 'aleo'.");
+            }
+        }
+
+        Ok(())
+    }
+}
+
 impl<N: Network, Instruction: InstructionTrait<N>, Command: CommandTrait<N>> TypeName
     for ProgramCore<N, Instruction, Command>
 {
diff --git a/synthesizer/snark/Cargo.toml b/synthesizer/snark/Cargo.toml
index cfdc5ed58..2ac903752 100644
--- a/synthesizer/snark/Cargo.toml
+++ b/synthesizer/snark/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-synthesizer-snark"
-version = "1.6.0"
+version = "3.7.1"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "SNARK wrappers for a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -33,18 +33,18 @@ wasm = [ "console/wasm", "snarkvm-algorithms/wasm" ]
 [dependencies.circuit]
 package = "snarkvm-circuit"
 path = "../../circuit"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.console]
 package = "snarkvm-console"
 path = "../../console"
-version = "=1.6.0"
+version = "=3.7.1"
 default-features = false
 features = [ "network" ]
 
 [dependencies.snarkvm-algorithms]
 path = "../../algorithms"
-version = "=1.6.0"
+version = "=3.7.1"
 
 [dependencies.bincode]
 version = "1"
diff --git a/synthesizer/src/vm/verify.rs b/synthesizer/src/vm/verify.rs
index 7f7c9fa17..c2c32e53f 100644
--- a/synthesizer/src/vm/verify.rs
+++ b/synthesizer/src/vm/verify.rs
@@ -165,6 +165,10 @@ impl<N: Network, C: ConsensusStorage<N>> VM<N, C> {
                 let current_block_height = self.block_store().current_block_height();
                 let consensus_version = N::CONSENSUS_VERSION(current_block_height)?;
                 deployment.program().check_restricted_keywords_for_consensus_version(consensus_version)?;
+                // Perform additional program checks if the consensus version is V7 or beyond.
+                if self.block_store().current_block_height() >= N::CONSENSUS_HEIGHT(ConsensusVersion::V7)? {
+                    deployment.program().check_program_naming_structure()?;
+                }
                 // Verify the deployment if it has not been verified before.
                 if !is_partially_verified {
                     // Verify the deployment.
@@ -956,4 +960,160 @@ function compute:
         // Check that v2 transaction is invalid.
         assert!(new_vm.check_transaction(&transaction_v2, None, rng).is_err());
     }
+
+    #[cfg(feature = "test")]
+    #[test]
+    fn test_program_rules_migration() {
+        let rng = &mut TestRng::default();
+
+        // Initialize the VM.
+        let vm = crate::vm::test_helpers::sample_vm();
+        // Initialize the genesis block.
+        let genesis = crate::vm::test_helpers::sample_genesis_block(rng);
+        // Update the VM.
+        vm.add_next_block(&genesis).unwrap();
+
+        // Track the vm blocks.
+        let mut vm_blocks = vec![genesis.clone()];
+
+        // Fetch the private key.
+        let private_key = crate::vm::test_helpers::sample_genesis_private_key(rng);
+
+        // Advance the ledger past ConsensusV4 where the new varuna version starts to take place.
+        let transactions: [Transaction<CurrentNetwork>; 0] = [];
+        while vm.block_store().current_block_height() < CurrentNetwork::CONSENSUS_HEIGHT(ConsensusVersion::V4).unwrap()
+        {
+            // Call the function
+            let next_block = crate::vm::test_helpers::sample_next_block(&vm, &private_key, &transactions, rng).unwrap();
+            vm.add_next_block(&next_block).unwrap();
+            vm_blocks.push(next_block);
+        }
+
+        // Create a new program that contains "aleo" in the name.
+        let program_1 = Program::from_str(
+            r"
+program testing_1_aleo.aleo;
+
+record token:
+    owner as address.private;
+    amount as u64.private;
+
+function compute:
+    input r0 as u32.private;
+    add r0 r0 into r1;
+    output r1 as u32.public;",
+        )
+        .unwrap();
+
+        // Create a new program that contains "aleo" in the record name.
+        let program_2 = Program::from_str(
+            r"
+program testing_2.aleo;
+
+record token_aleo:
+    owner as address.private;
+    amount as u64.private;
+
+function compute:
+    input r0 as u32.private;
+    add r0 r0 into r1;
+    output r1 as u32.public;",
+        )
+        .unwrap();
+
+        // Create a new program with records names that have other record names as a prefix.
+        let program_3 = Program::from_str(
+            r"
+program testing_3.aleo;
+
+record token:
+    owner as address.private;
+    amount as u64.private;
+
+record token_2:
+    owner as address.private;
+    amount as u64.private;
+
+function compute:
+    input r0 as u32.private;
+    add r0 r0 into r1;
+    output r1 as u32.public;",
+        )
+        .unwrap();
+
+        // Create a new program with records that has an entry containing "aleo".
+        let program_4 = Program::from_str(
+            r"
+program testing_4.aleo;
+
+record token:
+    owner as address.private;
+    test_aleo as u64.private;
+    amount as u64.private;
+
+function compute:
+    input r0 as u32.private;
+    add r0 r0 into r1;
+    output r1 as u32.public;",
+        )
+        .unwrap();
+
+        // Create a deployment transaction for the first program.
+        let deploy_1 = vm.deploy(&private_key, &program_1, None, 0, None, rng).unwrap();
+
+        // Create a deployment transaction for the second program.
+        let deploy_2 = vm.deploy(&private_key, &program_2, None, 0, None, rng).unwrap();
+
+        // Create a deployment transaction for the third program.
+        let deploy_3 = vm.deploy(&private_key, &program_3, None, 0, None, rng).unwrap();
+
+        // Create a deployment transaction for the fourth program.
+        let deploy_4 = vm.deploy(&private_key, &program_4, None, 0, None, rng).unwrap();
+
+        // // Ensure that the deployments are valid.
+        assert!(vm.check_transaction(&deploy_1, None, rng).is_ok());
+        assert!(vm.check_transaction(&deploy_2, None, rng).is_ok());
+        assert!(vm.check_transaction(&deploy_3, None, rng).is_ok());
+        assert!(vm.check_transaction(&deploy_4, None, rng).is_ok());
+
+        // Advance the ledger past ConsensusVersion::V7
+        let transactions: [Transaction<CurrentNetwork>; 0] = [];
+        while vm.block_store().current_block_height() < CurrentNetwork::CONSENSUS_HEIGHT(ConsensusVersion::V7).unwrap()
+        {
+            // // Ensure that the deployments are valid.
+            assert!(vm.check_transaction(&deploy_1, None, rng).is_ok());
+            assert!(vm.check_transaction(&deploy_2, None, rng).is_ok());
+            assert!(vm.check_transaction(&deploy_3, None, rng).is_ok());
+            assert!(vm.check_transaction(&deploy_4, None, rng).is_ok());
+            // Call the function
+            let next_block = crate::vm::test_helpers::sample_next_block(&vm, &private_key, &transactions, rng).unwrap();
+            vm.add_next_block(&next_block).unwrap();
+            vm_blocks.push(next_block);
+        }
+
+        // Ensure that the deployments are no longer valid.
+        assert!(vm.check_transaction(&deploy_1, None, rng).is_err());
+        assert!(vm.check_transaction(&deploy_2, None, rng).is_err());
+        assert!(vm.check_transaction(&deploy_3, None, rng).is_err());
+        assert!(vm.check_transaction(&deploy_4, None, rng).is_err());
+
+        // Check that the next block will abort the deployments.
+        let deploy_1_tx_id = deploy_1.id();
+        let deploy_2_tx_id = deploy_2.id();
+        let deploy_3_tx_id = deploy_3.id();
+        let deploy_4_tx_id = deploy_4.id();
+        let next_block = crate::vm::test_helpers::sample_next_block(
+            &vm,
+            &private_key,
+            &[deploy_1, deploy_2, deploy_3, deploy_4],
+            rng,
+        )
+        .unwrap();
+        assert_eq!(next_block.aborted_transaction_ids(), &vec![
+            deploy_1_tx_id,
+            deploy_2_tx_id,
+            deploy_3_tx_id,
+            deploy_4_tx_id
+        ]);
+    }
 }
diff --git a/utilities/Cargo.toml b/utilities/Cargo.toml
index fdfc66b2d..351dbeb90 100644
--- a/utilities/Cargo.toml
+++ b/utilities/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-utilities"
-version = "1.6.0"
+version = "3.7.1"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Utilities for a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -25,7 +25,7 @@ edition = "2021"
 
 [dependencies.snarkvm-utilities-derives]
 path = "./derives"
-version = "=1.6.0"
+version = "=3.7.1"
 optional = true
 
 [dependencies.aleo-std]
diff --git a/utilities/derives/Cargo.toml b/utilities/derives/Cargo.toml
index cf39b4086..852f8c6c7 100644
--- a/utilities/derives/Cargo.toml
+++ b/utilities/derives/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-utilities-derives"
-version = "1.6.0"
+version = "3.7.1"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Canonical serialization for a decentralized virtual machine"
 homepage = "https://aleo.org"
diff --git a/wasm/Cargo.toml b/wasm/Cargo.toml
index aafcccfe1..9ade446d8 100644
--- a/wasm/Cargo.toml
+++ b/wasm/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-wasm"
-version = "1.6.0"
+version = "3.7.1"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "WASM for a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -51,54 +51,54 @@ utilities = [ "snarkvm-utilities" ]
 
 [dependencies.snarkvm-circuit-network]
 path = "../circuit/network"
-version = "=1.6.0"
+version = "=3.7.1"
 features = [ "wasm" ]
 optional = true
 
 [dependencies.snarkvm-console]
 path = "../console"
-version = "=1.6.0"
+version = "=3.7.1"
 features = [ "wasm" ]
 optional = true
 
 [dependencies.snarkvm-curves]
 path = "../curves"
-version = "=1.6.0"
+version = "=3.7.1"
 optional = true
 
 [dependencies.snarkvm-fields]
 path = "../fields"
-version = "=1.6.0"
+version = "=3.7.1"
 optional = true
 
 [dependencies.snarkvm-ledger-block]
 path = "../ledger/block"
-version = "=1.6.0"
+version = "=3.7.1"
 features = [ "wasm" ]
 optional = true
 
 [dependencies.snarkvm-ledger-query]
 path = "../ledger/query"
-version = "=1.6.0"
+version = "=3.7.1"
 features = [ "async", "wasm" ]
 optional = true
 
 [dependencies.snarkvm-ledger-store]
 path = "../ledger/store"
-version = "=1.6.0"
+version = "=3.7.1"
 features = [ "wasm" ]
 optional = true
 
 [dependencies.snarkvm-synthesizer]
 path = "../synthesizer"
-version = "=1.6.0"
+version = "=3.7.1"
 default-features = false
 features = [ "async", "wasm" ]
 optional = true
 
 [dependencies.snarkvm-utilities]
 path = "../utilities"
-version = "=1.6.0"
+version = "=3.7.1"
 features = [ "wasm" ]
 optional = true
 
