diff --git a/Cargo.lock b/Cargo.lock
index dcc187fe2..be5b0718d 100644
--- a/Cargo.lock
+++ b/Cargo.lock
@@ -1845,6 +1845,7 @@ name = "snarkvm-utilities"
 version = "0.7.1"
 dependencies = [
  "bincode",
+ "num-bigint",
  "rand",
  "rand_xorshift",
  "snarkvm-derives",
diff --git a/algorithms/src/crh/bowe_hopwood_pedersen.rs b/algorithms/src/crh/bowe_hopwood_pedersen.rs
index c971ca1f8..f9381062d 100644
--- a/algorithms/src/crh/bowe_hopwood_pedersen.rs
+++ b/algorithms/src/crh/bowe_hopwood_pedersen.rs
@@ -214,7 +214,7 @@ mod tests {
                 .unwrap();
         assert_eq!(
             &*output.to_string(),
-            "Affine(x=232405123812771034726439972860096518067116445442313271493943612938654881935, y=752634260468672343124870935373206613671657768711738358314821821547485346646)"
+            "Affine(x=1458830605996255967666145170206084970380287513737423487919697505288312101007, y=4724361822497728774087744092818831022870480949262013688485525440243906394966)"
         );
     }
 }
diff --git a/curves/src/bw6_761/tests.rs b/curves/src/bw6_761/tests.rs
index a985c5600..bac025af6 100644
--- a/curves/src/bw6_761/tests.rs
+++ b/curves/src/bw6_761/tests.rs
@@ -45,6 +45,7 @@ fn test_bw6_761_fr() {
     field_test(a, b);
     sqrt_field_test(a);
     primefield_test::<Fr>();
+    field_serialization_test::<Fq>();
 }
 
 #[test]
diff --git a/fields/src/tests_field.rs b/fields/src/tests_field.rs
index b21617746..6427bc11d 100644
--- a/fields/src/tests_field.rs
+++ b/fields/src/tests_field.rs
@@ -171,38 +171,43 @@ fn random_expansion_tests<F: Field, R: Rng>(rng: &mut R) {
     }
 }
 
-fn random_field_tests<F: Field>() {
+fn random_string_tests<F: PrimeField>() {
     let mut rng = XorShiftRng::seed_from_u64(1231275789u64);
 
-    random_negation_tests::<F, _>(&mut rng);
-    random_addition_tests::<F, _>(&mut rng);
-    random_subtraction_tests::<F, _>(&mut rng);
-    random_multiplication_tests::<F, _>(&mut rng);
-    random_inversion_tests::<F, _>(&mut rng);
-    random_doubling_tests::<F, _>(&mut rng);
-    random_squaring_tests::<F, _>(&mut rng);
-    random_expansion_tests::<F, _>(&mut rng);
-
-    assert!(F::zero().is_zero());
     {
-        let z = -F::zero();
-        assert!(z.is_zero());
+        let a = "84395729384759238745923745892374598234705297301958723458712394587103249587213984572934750213947582345792304758273458972349582734958273495872304598234";
+        let b = "38495729084572938457298347502349857029384609283450692834058293405982304598230458230495820394850293845098234059823049582309485203948502938452093482039";
+        let c = "3248875134290623212325429203829831876024364170316860259933542844758450336418538569901990710701240661702808867062612075657861768196242274635305077449545396068598317421057721935408562373834079015873933065667961469731886739181625866970316226171512545167081793907058686908697431878454091011239990119126";
+
+        let mut a = F::from_str(a).map_err(|_| ()).unwrap();
+        let b = F::from_str(b).map_err(|_| ()).unwrap();
+        let c = F::from_str(c).map_err(|_| ()).unwrap();
+
+        a *= &b;
+
+        assert_eq!(a, c);
     }
 
-    assert!(F::zero().inverse().is_none());
+    assert!(F::from_str("").is_err());
+    assert!(F::from_str("0").map_err(|_| ()).unwrap().is_zero());
+    assert!(F::from_str("00").is_err());
+    assert!(F::from_str("00000000000").is_err());
 
-    // Multiplication by zero
-    {
-        let a = F::rand(&mut rng) * F::zero();
-        assert!(a.is_zero());
+    for _ in 0..ITERATIONS {
+        let n: u64 = rng.gen();
+
+        let a = F::from_str(&format!("{}", n)).map_err(|_| ()).unwrap();
+        let b = F::from_repr(n.into()).unwrap();
+
+        assert_eq!(a, b);
     }
 
-    // Addition by zero
-    {
-        let mut a = F::rand(&mut rng);
-        let copy = a;
-        a += &F::zero();
-        assert_eq!(a, copy);
+    for _ in 0..ITERATIONS {
+        let a = F::rand(&mut rng);
+
+        let reference = a.to_string();
+        let candidate = F::from_str(&reference).map_err(|_| panic!()).unwrap().to_string();
+        assert_eq!(reference, candidate);
     }
 }
 
@@ -235,40 +240,6 @@ fn random_sqrt_tests<F: SquareRootField>() {
     }
 }
 
-pub fn from_str_test<F: PrimeField>() {
-    {
-        let a = "84395729384759238745923745892374598234705297301958723458712394587103249587213984572934750213947582345792304758273458972349582734958273495872304598234";
-        let b = "38495729084572938457298347502349857029384609283450692834058293405982304598230458230495820394850293845098234059823049582309485203948502938452093482039";
-        let c = "3248875134290623212325429203829831876024364170316860259933542844758450336418538569901990710701240661702808867062612075657861768196242274635305077449545396068598317421057721935408562373834079015873933065667961469731886739181625866970316226171512545167081793907058686908697431878454091011239990119126";
-
-        let mut a = F::from_str(a).map_err(|_| ()).unwrap();
-        let b = F::from_str(b).map_err(|_| ()).unwrap();
-        let c = F::from_str(c).map_err(|_| ()).unwrap();
-
-        a *= &b;
-
-        assert_eq!(a, c);
-    }
-
-    {
-        let mut rng = XorShiftRng::seed_from_u64(1231275789u64);
-
-        for _ in 0..ITERATIONS {
-            let n: u64 = rng.gen();
-
-            let a = F::from_str(&format!("{}", n)).map_err(|_| ()).unwrap();
-            let b = F::from_repr(n.into()).unwrap();
-
-            assert_eq!(a, b);
-        }
-    }
-
-    assert!(F::from_str("").is_err());
-    assert!(F::from_str("0").map_err(|_| ()).unwrap().is_zero());
-    assert!(F::from_str("00").is_err());
-    assert!(F::from_str("00000000000").is_err());
-}
-
 #[allow(clippy::eq_op)]
 pub fn field_test<F: Field>(a: F, b: F) {
     let zero = F::zero();
@@ -331,7 +302,37 @@ pub fn field_test<F: Field>(a: F, b: F) {
     // (a - b)^2 = (-(b - a))^2
     assert_eq!((a - b).square(), (-(b - a)).square());
 
-    random_field_tests::<F>();
+    let mut rng = XorShiftRng::seed_from_u64(1231275789u64);
+    random_negation_tests::<F, _>(&mut rng);
+    random_addition_tests::<F, _>(&mut rng);
+    random_subtraction_tests::<F, _>(&mut rng);
+    random_multiplication_tests::<F, _>(&mut rng);
+    random_inversion_tests::<F, _>(&mut rng);
+    random_doubling_tests::<F, _>(&mut rng);
+    random_squaring_tests::<F, _>(&mut rng);
+    random_expansion_tests::<F, _>(&mut rng);
+
+    assert!(F::zero().is_zero());
+    {
+        let z = -F::zero();
+        assert!(z.is_zero());
+    }
+
+    assert!(F::zero().inverse().is_none());
+
+    // Multiplication by zero
+    {
+        let a = F::rand(&mut rng) * F::zero();
+        assert!(a.is_zero());
+    }
+
+    // Addition by zero
+    {
+        let mut a = F::rand(&mut rng);
+        let copy = a;
+        a += &F::zero();
+        assert_eq!(a, copy);
+    }
 }
 
 pub fn fft_field_test<F: PrimeField + FftField>() {
@@ -373,6 +374,7 @@ pub fn primefield_test<F: PrimeField>() {
     assert_eq!(F::from_str("2").ok().unwrap(), two);
     assert_eq!(F::from_str(&two.to_string()).ok().unwrap(), two);
 
+    random_string_tests::<F>();
     fft_field_test::<F>();
 }
 
diff --git a/utilities/Cargo.toml b/utilities/Cargo.toml
index edd1191f6..b47f4a83e 100644
--- a/utilities/Cargo.toml
+++ b/utilities/Cargo.toml
@@ -25,6 +25,9 @@ optional = true
 [dependencies.bincode]
 version = "1.3.3"
 
+[dependencies.num-bigint]
+version = "0.4"
+
 [dependencies.rand]
 version = "0.8"
 default-features = false
diff --git a/utilities/src/biginteger/biginteger.rs b/utilities/src/biginteger/biginteger.rs
index 46e5cec1c..07e4aa76d 100644
--- a/utilities/src/biginteger/biginteger.rs
+++ b/utilities/src/biginteger/biginteger.rs
@@ -57,6 +57,9 @@ pub trait BigInteger:
     + AsRef<[u64]>
     + From<u64>
 {
+    /// The number of limbs used in this BigInteger.
+    const NUM_LIMBS: usize;
+
     /// Add another representation to this one, returning the carry bit.
     fn add_nocarry(&mut self, other: &Self) -> bool;
 
diff --git a/utilities/src/biginteger/macros.rs b/utilities/src/biginteger/macros.rs
index fb0e49502..4cbc28f11 100644
--- a/utilities/src/biginteger/macros.rs
+++ b/utilities/src/biginteger/macros.rs
@@ -26,6 +26,8 @@ macro_rules! bigint_impl {
         }
 
         impl BigInteger for $name {
+            const NUM_LIMBS: usize = $num_limbs;
+
             #[inline]
             fn add_nocarry(&mut self, other: &Self) -> bool {
                 let mut carry = 0;
@@ -167,16 +169,15 @@ macro_rules! bigint_impl {
             #[inline]
             fn from_bits_be(mut bits: Vec<bool>) -> Self {
                 let mut res = Self::default();
-                let mut acc: u64 = 0;
 
                 bits.reverse();
                 for (i, bits64) in bits.chunks(64).enumerate() {
+                    let mut acc: u64 = 0;
                     for bit in bits64.iter().rev() {
                         acc <<= 1;
                         acc += *bit as u64;
                     }
                     res.0[i] = acc;
-                    acc = 0;
                 }
                 res
             }
@@ -240,27 +241,9 @@ macro_rules! bigint_impl {
 
         impl Display for $name {
             fn fmt(&self, f: &mut ::std::fmt::Formatter<'_>) -> ::std::fmt::Result {
-                let mut is_nonzero = false;
-                for (i, limb) in self.0.iter().rev().enumerate() {
-                    if *limb > 0 {
-                        is_nonzero = true;
-                    }
-
-                    // Begin writing the limb, for its corresponding bits.
-                    if is_nonzero {
-                        let shift = (($num_limbs - (i + 1)) * 6) as u64;
-                        let shifter = 1 << shift;
-                        write!(f, "{}", shifter - 1 + *limb)?;
-                    }
-                }
-
-                // If the value is 0, then `is_nonzero` will still be `false` here,
-                // so proceed to write `0`.
-                if !is_nonzero {
-                    write!(f, "0")?;
-                }
-
-                Ok(())
+                // TODO: Implement a native version, without the unwrap.
+                let bytes = crate::to_bytes![self.0].unwrap();
+                write!(f, "{}", num_bigint::BigUint::from_bytes_le(&bytes))
             }
         }
 
diff --git a/utilities/src/biginteger/tests.rs b/utilities/src/biginteger/tests.rs
index e75eb4e5a..58ed1496a 100644
--- a/utilities/src/biginteger/tests.rs
+++ b/utilities/src/biginteger/tests.rs
@@ -77,6 +77,41 @@ fn biginteger_bytes_test<B: BigInteger>() {
     assert_eq!(x, y);
 }
 
+fn biginteger_to_string_test<B: BigInteger>() {
+    const ITERATIONS: u64 = 1_000_000;
+
+    let mut rng = XorShiftRng::seed_from_u64(1231275789u64);
+
+    // Sanity check the integers starting from 0 to ITERATIONS.
+    for integer in 0..ITERATIONS {
+        assert_eq!(format!("{}", integer), B::from(integer).to_string());
+    }
+
+    // If the BigInteger has more than one limb, sanity check
+    // that the second limb conversion is also correct.
+    if B::NUM_LIMBS > 1 {
+        let start = u64::MAX as u128;
+        for integer in start..(start + ITERATIONS as u128) {
+            let mut buffer = vec![0u8; 8 * B::NUM_LIMBS];
+            buffer
+                .iter_mut()
+                .zip(&(integer as u128).to_le_bytes())
+                .for_each(|(buf, val)| {
+                    *buf = *val;
+                });
+            assert_eq!(format!("{}", integer), B::read(&*buffer).unwrap().to_string());
+        }
+    }
+
+    // Sample random integers and check they match against num-bigint.
+    for _ in 0..ITERATIONS {
+        let candidate: B = UniformRand::rand(&mut rng);
+        let candidate_hex = format!("{:?}", candidate);
+        let reference = num_bigint::BigUint::parse_bytes(candidate_hex.as_bytes(), 16).unwrap();
+        assert_eq!(reference.to_str_radix(10), candidate.to_string());
+    }
+}
+
 fn test_biginteger<B: BigInteger>(zero: B) {
     let mut rng = XorShiftRng::seed_from_u64(1231275789u64);
     let a: B = UniformRand::rand(&mut rng);
@@ -84,6 +119,7 @@ fn test_biginteger<B: BigInteger>(zero: B) {
     biginteger_arithmetic_test(a, b, zero);
     biginteger_bytes_test::<B>();
     biginteger_bits_test::<B>();
+    biginteger_to_string_test::<B>();
 }
 
 #[test]
