diff --git a/Cargo.lock b/Cargo.lock
index 1e8f1f7a8..7724fb128 100644
--- a/Cargo.lock
+++ b/Cargo.lock
@@ -97,6 +97,12 @@ dependencies = [
  "colored",
 ]
 
+[[package]]
+name = "anes"
+version = "0.1.6"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "4b46cbb362ab8752921c97e041f5e366ee6297bd428a31275b9fcf1e380f7299"
+
 [[package]]
 name = "ansi_term"
 version = "0.12.1"
@@ -194,18 +200,6 @@ version = "0.4.0"
 source = "registry+https://github.com/rust-lang/crates.io-index"
 checksum = "771fe0050b883fcc3ea2359b1a96bcfbc090b7116eae7c3c512c7a083fdf23d3"
 
-[[package]]
-name = "bstr"
-version = "0.2.17"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "ba3569f383e8f1598449f1a423e72e99569137b47740b1da11ef19af3d5c3223"
-dependencies = [
- "lazy_static",
- "memchr",
- "regex-automata",
- "serde",
-]
-
 [[package]]
 name = "buf_redux"
 version = "0.8.4"
@@ -267,6 +261,33 @@ dependencies = [
  "envmnt",
 ]
 
+[[package]]
+name = "ciborium"
+version = "0.2.0"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "b0c137568cc60b904a7724001b35ce2630fd00d5d84805fbb608ab89509d788f"
+dependencies = [
+ "ciborium-io",
+ "ciborium-ll",
+ "serde",
+]
+
+[[package]]
+name = "ciborium-io"
+version = "0.2.0"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "346de753af073cc87b52b2083a506b38ac176a44cfb05497b622e27be899b369"
+
+[[package]]
+name = "ciborium-ll"
+version = "0.2.0"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "213030a2b5a4e0c0892b6652260cf6ccac84827b83a85a534e178e3906c4cf1b"
+dependencies = [
+ "ciborium-io",
+ "half",
+]
+
 [[package]]
 name = "cl-sys"
 version = "0.4.2"
@@ -286,17 +307,6 @@ dependencies = [
  "libc",
 ]
 
-[[package]]
-name = "clap"
-version = "2.34.0"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "a0610544180c38b88101fecf2dd634b174a62eef6946f84dfc6a7127512b381c"
-dependencies = [
- "bitflags",
- "textwrap 0.11.0",
- "unicode-width",
-]
-
 [[package]]
 name = "clap"
 version = "3.2.22"
@@ -311,7 +321,7 @@ dependencies = [
  "once_cell",
  "strsim",
  "termcolor",
- "textwrap 0.15.1",
+ "textwrap",
 ]
 
 [[package]]
@@ -413,15 +423,16 @@ dependencies = [
 
 [[package]]
 name = "criterion"
-version = "0.3.6"
+version = "0.4.0"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "b01d6de93b2b6c65e17c634a26653a29d107b3c98c607c765bf38d041531cd8f"
+checksum = "e7c76e09c1aae2bc52b3d2f29e13c6572553b30c4aa1b8a49fd70de6412654cb"
 dependencies = [
+ "anes",
  "atty",
  "cast",
- "clap 2.34.0",
+ "ciborium",
+ "clap",
  "criterion-plot",
- "csv",
  "itertools",
  "lazy_static",
  "num-traits",
@@ -430,7 +441,6 @@ dependencies = [
  "rayon",
  "regex",
  "serde",
- "serde_cbor",
  "serde_derive",
  "serde_json",
  "tinytemplate",
@@ -439,9 +449,9 @@ dependencies = [
 
 [[package]]
 name = "criterion-plot"
-version = "0.4.5"
+version = "0.5.0"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "2673cc8207403546f45f5fd319a974b1e6983ad1a3ee7e6041650013be041876"
+checksum = "6b50826342786a51a89e2da3a28f1c32b06e387201bc2d19791f622c673706b1"
 dependencies = [
  "cast",
  "itertools",
@@ -502,28 +512,6 @@ dependencies = [
  "typenum",
 ]
 
-[[package]]
-name = "csv"
-version = "1.1.6"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "22813a6dc45b335f9bade10bf7271dc477e81113e89eb251a0bc2a8a81c536e1"
-dependencies = [
- "bstr",
- "csv-core",
- "itoa 0.4.8",
- "ryu",
- "serde",
-]
-
-[[package]]
-name = "csv-core"
-version = "0.1.10"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "2b2466559f260f48ad25fe6317b3c8dac77b5bdb5763ac7d9d6103530663bc90"
-dependencies = [
- "memchr",
-]
-
 [[package]]
 name = "cuda-config"
 version = "0.1.0"
@@ -968,7 +956,7 @@ checksum = "75f43d41e26995c17e71ee126451dd3941010b0514a81a9d11f3b341debc2399"
 dependencies = [
  "bytes",
  "fnv",
- "itoa 1.0.3",
+ "itoa",
 ]
 
 [[package]]
@@ -1009,7 +997,7 @@ dependencies = [
  "http-body",
  "httparse",
  "httpdate",
- "itoa 1.0.3",
+ "itoa",
  "pin-project-lite",
  "socket2",
  "tokio",
@@ -1091,12 +1079,6 @@ dependencies = [
  "either",
 ]
 
-[[package]]
-name = "itoa"
-version = "0.4.8"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "b71991ff56294aa922b450139ee08b3bfc70982c6b2c7562771375cf73542dd4"
-
 [[package]]
 name = "itoa"
 version = "1.0.3"
@@ -1981,16 +1963,6 @@ dependencies = [
  "serde_derive",
 ]
 
-[[package]]
-name = "serde_cbor"
-version = "0.11.2"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "2bef2ebfde456fb76bbcf9f59315333decc4fda0b2b44b420243c11e0f5ec1f5"
-dependencies = [
- "half",
- "serde",
-]
-
 [[package]]
 name = "serde_derive"
 version = "1.0.145"
@@ -2008,7 +1980,7 @@ version = "1.0.85"
 source = "registry+https://github.com/rust-lang/crates.io-index"
 checksum = "e55a28e3aaef9d5ce0506d0a14dbba8054ddc7e499ef522dd8b26859ec9d4a44"
 dependencies = [
- "itoa 1.0.3",
+ "itoa",
  "ryu",
  "serde",
 ]
@@ -2020,7 +1992,7 @@ source = "registry+https://github.com/rust-lang/crates.io-index"
 checksum = "d3491c14715ca2294c4d6a88f15e84739788c1d030eed8c110436aafdaa2f3fd"
 dependencies = [
  "form_urlencoded",
- "itoa 1.0.3",
+ "itoa",
  "ryu",
  "serde",
 ]
@@ -2114,7 +2086,7 @@ version = "0.9.0"
 dependencies = [
  "anyhow",
  "bincode",
- "clap 3.2.22",
+ "clap",
  "colored",
  "indexmap",
  "once_cell",
@@ -2825,15 +2797,6 @@ dependencies = [
  "winapi",
 ]
 
-[[package]]
-name = "textwrap"
-version = "0.11.0"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "d326610f408c7a4eb6f51c37c330e496b08506c9457c9d34287ecc38809fb060"
-dependencies = [
- "unicode-width",
-]
-
 [[package]]
 name = "textwrap"
 version = "0.15.1"
@@ -2871,9 +2834,9 @@ dependencies = [
 
 [[package]]
 name = "time"
-version = "0.3.14"
+version = "0.3.15"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "3c3f9a28b618c3a6b9251b6908e9c99e04b9e5c02e6581ccbb67d59c34ef7f9b"
+checksum = "d634a985c4d4238ec39cacaed2e7ae552fbd3c476b552c1deac3021b7d7eaf0c"
 dependencies = [
  "libc",
  "num_threads",
diff --git a/algorithms/Cargo.toml b/algorithms/Cargo.toml
index 5ae88a85c..bb02334af 100644
--- a/algorithms/Cargo.toml
+++ b/algorithms/Cargo.toml
@@ -144,7 +144,7 @@ version = "1.1.0"
 version = "1"
 
 [dev-dependencies.criterion]
-version = "0.3.6"
+version = "0.4.0"
 
 [dev-dependencies.serde_json]
 version = "1"
diff --git a/circuit/environment/Cargo.toml b/circuit/environment/Cargo.toml
index 047a6d32b..9579ce360 100644
--- a/circuit/environment/Cargo.toml
+++ b/circuit/environment/Cargo.toml
@@ -68,7 +68,7 @@ path = "../../circuit"
 path = "../../console/types"
 
 [dev-dependencies.criterion]
-version = "0.3"
+version = "0.4"
 
 [features]
 default = [ "snarkvm-curves/default", "snarkvm-r1cs/default" ]
diff --git a/circuit/program/src/data/identifier/from_bits.rs b/circuit/program/src/data/identifier/from_bits.rs
index d321e7efa..a38d7756b 100644
--- a/circuit/program/src/data/identifier/from_bits.rs
+++ b/circuit/program/src/data/identifier/from_bits.rs
@@ -29,29 +29,17 @@ impl<A: Aleo> FromBits for Identifier<A> {
         // Recover the field element from the bits.
         let field = Field::from_bits_le(bits_le);
 
-        // Eject the bits in **little-endian** form, and convert the bits to bytes.
-        let bytes = bits_le
-            .to_bits_le()
-            .eject_value()
-            .chunks(8)
-            .map(|byte| match u8::from_bits_le(byte) {
-                Ok(byte) => byte,
-                Err(error) => A::halt(format!("Failed to recover an identifier from bits: {error}")),
-            })
-            .collect::<Vec<_>>();
-
-        // Recover the identifier length from the bits, by finding the first instance of a `0` byte,
-        // which is the null character '\0' in UTF-8, and an invalid character in an identifier.
-        let num_bytes = match bytes.iter().position(|&byte| byte == 0) {
-            Some(index) => index, // `index` is 0-indexed, and we exclude the null character.
-            None => bytes.len(),  // No null character found, so the identifier is the full length.
+        // Eject the bits in **little-endian** form, and determine the number of bytes.
+        let num_bytes = match console::Identifier::<A::Network>::from_bits_le(&bits_le.eject_value()) {
+            Ok(console_identifier) => console_identifier.size_in_bits() / 8,
+            Err(error) => A::halt(format!("Failed to recover an identifier from bits: {error}")),
         };
 
         // Ensure identifier fits within the data capacity of the base field.
         let max_bytes = A::BaseField::size_in_data_bits() / 8; // Note: This intentionally rounds down.
-        match num_bytes <= max_bytes {
+        match num_bytes as usize <= max_bytes {
             // Return the identifier.
-            true => Self(field, num_bytes as u8),
+            true => Self(field, num_bytes),
             false => A::halt("Identifier exceeds the maximum capacity allowed"),
         }
     }
diff --git a/circuit/program/src/data/identifier/mod.rs b/circuit/program/src/data/identifier/mod.rs
index 88b0194e9..8ad349d06 100644
--- a/circuit/program/src/data/identifier/mod.rs
+++ b/circuit/program/src/data/identifier/mod.rs
@@ -26,7 +26,7 @@ mod to_field;
 
 use snarkvm_circuit_network::Aleo;
 use snarkvm_circuit_types::{environment::prelude::*, Boolean, Field, U8};
-use snarkvm_utilities::{FromBits as FB, ToBits as TB};
+use snarkvm_utilities::ToBits as TB;
 
 /// An identifier is an **immutable** UTF-8 string,
 /// represented as a **constant** field element in the circuit.
diff --git a/circuit/types/boolean/Cargo.toml b/circuit/types/boolean/Cargo.toml
index 4010a72b0..5777a1028 100644
--- a/circuit/types/boolean/Cargo.toml
+++ b/circuit/types/boolean/Cargo.toml
@@ -22,7 +22,7 @@ path = "../../environment"
 version = "0.9.0"
 
 [dev-dependencies.criterion]
-version = "0.3"
+version = "0.4"
 
 [features]
 default = [ "enable_console" ]
diff --git a/console/algorithms/Cargo.toml b/console/algorithms/Cargo.toml
index 0814e0e87..29d65cde1 100644
--- a/console/algorithms/Cargo.toml
+++ b/console/algorithms/Cargo.toml
@@ -39,7 +39,7 @@ path = "../../curves"
 default-features = false
 
 [dev-dependencies.criterion]
-version = "0.3.6"
+version = "0.4.0"
 
 [dev-dependencies.expect-test]
 version = "1.2"
diff --git a/console/collections/Cargo.toml b/console/collections/Cargo.toml
index 61e31aa7c..2ccae0003 100644
--- a/console/collections/Cargo.toml
+++ b/console/collections/Cargo.toml
@@ -33,7 +33,7 @@ optional = true
 path = "../network"
 
 [dev-dependencies.criterion]
-version = "0.3.5"
+version = "0.4.0"
 
 [features]
 default = [ "parallel" ]
diff --git a/console/network/environment/src/helpers/sanitizer.rs b/console/network/environment/src/helpers/sanitizer.rs
index 9f02eb199..9f82bdced 100644
--- a/console/network/environment/src/helpers/sanitizer.rs
+++ b/console/network/environment/src/helpers/sanitizer.rs
@@ -18,9 +18,9 @@ use crate::ParserResult;
 
 use nom::{
     branch::alt,
-    bytes::complete::{tag, take_until},
+    bytes::complete::tag,
     character::complete::{anychar, char, line_ending, multispace1},
-    combinator::{cut, map, recognize, value},
+    combinator::{cut, map, recognize, value, verify},
     error::{ErrorKind, VerboseError, VerboseErrorKind},
     multi::fold_many0,
     sequence::{preceded, terminated},
@@ -48,12 +48,57 @@ impl Sanitizer {
     pub fn parse_comment(string: &str) -> ParserResult<&str> {
         preceded(
             char('/'),
-            alt((
-                preceded(char('/'), cut(Self::str_till_eol)),
-                preceded(char('*'), cut(terminated(take_until("*/"), tag("*/")))),
-            )),
+            alt((preceded(char('/'), cut(Self::str_till_eol)), preceded(char('*'), cut(Self::str_till_star_slash)))),
         )(string)
     }
+
+    /// Parse a safe character (in the sense explained below).
+    /// Returns an error if no character is found or a non-safe character is found.
+    /// The character is returned, along with the remaining input.
+    ///
+    /// This is used for otherwise unconstrained characters
+    /// in (line and block) comments and in string literals.
+    ///
+    /// We regard the following characters as safe:
+    /// - Horizontal tab (code 9).
+    /// - Line feed (code 10).
+    /// - Carriage return (code 13).
+    /// - Space (code 32).
+    /// - Visible ASCII (codes 33-126).
+    /// - Non-ASCII Unicode scalar values except bidi embeddings, overrides, and isolates.
+    ///
+    /// The Unicode bidi characters are well-known for presenting Trojan Source dangers.
+    /// The ASCII backspace (code 8) can be also used to make text look different from what it is,
+    /// and a similar danger may apply to delete (126).
+    /// Other ASCII control characters
+    /// (except for horizontal tab, space, line feed, and carriage return, which are allowed)
+    /// may or may not present dangers, but we see no good reason for allowing them.
+    /// At some point we may want disallow additional non-ASCII characters,
+    /// if we see no good reason to allow them.
+    ///
+    /// Note that we say 'Unicode scalar values' above,
+    /// because we read UTF-8-decoded characters,
+    /// and thus we will never encounter surrogate code points,
+    /// and we do not need to explicitly exclude them in this function.
+    ///
+    /// Note also that the `nom` documentation for `anychar` says that
+    /// it matches one byte as a character.
+    /// However, simple experiments show that it matches a Unicode character,
+    /// e.g. attempting to parse `"\u{4141}"` yields one CJK character and exhausts the input,
+    /// as opposed to returning `A` and leaving another `A` in the input.
+    pub fn parse_safe_char(string: &str) -> ParserResult<char> {
+        fn is_safe(ch: &char) -> bool {
+            let code = *ch as u32;
+            code == 9
+                || code == 10
+                || code == 13
+                || (32..=126).contains(&code)
+                || (128..=0x2029).contains(&code)
+                || (0x202f..=0x2065).contains(&code)
+                || (0x206a <= code)
+        }
+        verify(anychar, is_safe)(string)
+    }
 }
 
 impl Sanitizer {
@@ -102,8 +147,22 @@ impl Sanitizer {
     ///
     /// Discard any leading newline.
     fn str_till_eol(string: &str) -> ParserResult<&str> {
-        map(recognize(Self::till(alt((value((), tag("\\\n")), value((), anychar))), Self::eol)), |i| {
-            if i.as_bytes().last() == Some(&b'\n') { &i[0..i.len() - 1] } else { i }
+        map(
+            recognize(Self::till(alt((value((), tag("\\\n")), value((), Sanitizer::parse_safe_char))), Self::eol)),
+            |i| {
+                if i.as_bytes().last() == Some(&b'\n') { &i[0..i.len() - 1] } else { i }
+            },
+        )(string)
+    }
+
+    /// Parse a string until `*/` is encountered.
+    ///
+    /// This is used to parse the body of a block comment, after the opening `/*`.
+    ///
+    /// Return the body of the comment, i.e. what is between `/*` and `*/`.
+    fn str_till_star_slash(string: &str) -> ParserResult<&str> {
+        map(recognize(Self::till(value((), Sanitizer::parse_safe_char), tag("*/"))), |i| {
+            &i[0..i.len() - 2] // subtract 2 to discard the closing `*/`
         })(string)
     }
 
@@ -120,6 +179,65 @@ impl Sanitizer {
 mod tests {
     use super::*;
 
+    #[test]
+    fn test_parse_safe_char() {
+        // test correct acceptance of ASCII and non-ASCII:
+        assert_eq!(("", 'A'), Sanitizer::parse_safe_char("A").unwrap());
+        assert_eq!((" and more", 'A'), Sanitizer::parse_safe_char("A and more").unwrap());
+        assert_eq!(("", '\u{4141}'), Sanitizer::parse_safe_char("\u{4141}").unwrap());
+        assert_eq!((" and more", '\u{4141}'), Sanitizer::parse_safe_char("\u{4141} and more").unwrap());
+
+        // test rejection and acceptance of ASCII control characters:
+        assert!(Sanitizer::parse_safe_char("\x00").is_err());
+        assert!(Sanitizer::parse_safe_char("\x01").is_err());
+        assert!(Sanitizer::parse_safe_char("\x02").is_err());
+        assert!(Sanitizer::parse_safe_char("\x03").is_err());
+        assert!(Sanitizer::parse_safe_char("\x04").is_err());
+        assert!(Sanitizer::parse_safe_char("\x05").is_err());
+        assert!(Sanitizer::parse_safe_char("\x06").is_err());
+        assert!(Sanitizer::parse_safe_char("\x07").is_err());
+        assert!(Sanitizer::parse_safe_char("\x08").is_err());
+        assert!(Sanitizer::parse_safe_char("\x09").is_ok());
+        assert!(Sanitizer::parse_safe_char("\x0a").is_ok());
+        assert!(Sanitizer::parse_safe_char("\x0b").is_err());
+        assert!(Sanitizer::parse_safe_char("\x0c").is_err());
+        assert!(Sanitizer::parse_safe_char("\x0d").is_ok());
+        assert!(Sanitizer::parse_safe_char("\x0e").is_err());
+        assert!(Sanitizer::parse_safe_char("\x0f").is_err());
+        assert!(Sanitizer::parse_safe_char("\x10").is_err());
+        assert!(Sanitizer::parse_safe_char("\x11").is_err());
+        assert!(Sanitizer::parse_safe_char("\x12").is_err());
+        assert!(Sanitizer::parse_safe_char("\x13").is_err());
+        assert!(Sanitizer::parse_safe_char("\x14").is_err());
+        assert!(Sanitizer::parse_safe_char("\x15").is_err());
+        assert!(Sanitizer::parse_safe_char("\x16").is_err());
+        assert!(Sanitizer::parse_safe_char("\x17").is_err());
+        assert!(Sanitizer::parse_safe_char("\x18").is_err());
+        assert!(Sanitizer::parse_safe_char("\x19").is_err());
+        assert!(Sanitizer::parse_safe_char("\x1a").is_err());
+        assert!(Sanitizer::parse_safe_char("\x1b").is_err());
+        assert!(Sanitizer::parse_safe_char("\x1c").is_err());
+        assert!(Sanitizer::parse_safe_char("\x1d").is_err());
+        assert!(Sanitizer::parse_safe_char("\x1e").is_err());
+        assert!(Sanitizer::parse_safe_char("\x1f").is_err());
+        assert!(Sanitizer::parse_safe_char("\x7f").is_err());
+
+        // test rejection of bidi characters, and acceptance of the ones just above/below:
+        assert!(Sanitizer::parse_safe_char("\u{2029}").is_ok());
+        assert!(Sanitizer::parse_safe_char("\u{202a}").is_err());
+        assert!(Sanitizer::parse_safe_char("\u{202b}").is_err());
+        assert!(Sanitizer::parse_safe_char("\u{202c}").is_err());
+        assert!(Sanitizer::parse_safe_char("\u{202d}").is_err());
+        assert!(Sanitizer::parse_safe_char("\u{202e}").is_err());
+        assert!(Sanitizer::parse_safe_char("\u{202f}").is_ok());
+        assert!(Sanitizer::parse_safe_char("\u{2065}").is_ok());
+        assert!(Sanitizer::parse_safe_char("\u{2066}").is_err());
+        assert!(Sanitizer::parse_safe_char("\u{2067}").is_err());
+        assert!(Sanitizer::parse_safe_char("\u{2068}").is_err());
+        assert!(Sanitizer::parse_safe_char("\u{2069}").is_err());
+        assert!(Sanitizer::parse_safe_char("\u{206a}").is_ok());
+    }
+
     #[test]
     fn test_sanitize() {
         // Whitespaces
@@ -164,5 +282,15 @@ mod tests {
         assert_eq!(("hello world", "/* hello */\n"), Sanitizer::parse_comments("/* hello */\nhello world").unwrap());
         assert_eq!(("hello world", "/** hello */\n"), Sanitizer::parse_comments("/** hello */\nhello world").unwrap());
         assert_eq!(("/\nhello world", ""), Sanitizer::parse_comments("/\nhello world").unwrap());
+        assert_eq!(
+            ("hello world", "// hel\u{4141}lo\n"),
+            Sanitizer::parse_comments("// hel\u{4141}lo\nhello world").unwrap()
+        );
+        assert!(Sanitizer::parse_comments("// hel\x08lo\nhello world").is_err());
+        assert!(Sanitizer::parse_comments("// hel\u{2066}lo\nhello world").is_err());
+        assert!(Sanitizer::parse_comments("/* hel\x7flo */\nhello world").is_err());
+        assert!(Sanitizer::parse_comments("/* hel\u{202d}lo */\nhello world").is_err());
+        assert!(Sanitizer::parse_comments("/** hel\x00lo */\nhello world").is_err());
+        assert!(Sanitizer::parse_comments("/** hel\u{202a}lo */\nhello world").is_err());
     }
 }
diff --git a/console/network/environment/src/prelude.rs b/console/network/environment/src/prelude.rs
index 64feead5a..1540cf0cd 100644
--- a/console/network/environment/src/prelude.rs
+++ b/console/network/environment/src/prelude.rs
@@ -75,7 +75,7 @@ pub use nom::{
     branch::alt,
     bytes::{complete::tag, streaming::take},
     character::complete::{alpha1, alphanumeric1, char, one_of},
-    combinator::{map, map_res, opt, recognize},
+    combinator::{complete, map, map_res, opt, recognize},
     multi::{many0, many1, separated_list0, separated_list1},
     sequence::{pair, terminated},
 };
diff --git a/console/network/environment/src/traits/string.rs b/console/network/environment/src/traits/string.rs
index 58917f4fb..62d980c6e 100644
--- a/console/network/environment/src/traits/string.rs
+++ b/console/network/environment/src/traits/string.rs
@@ -40,8 +40,18 @@ pub mod string_parser {
     };
 
     /// The list of unsupported "invisible" code points.
-    const UNSUPPORTED_CODE_POINTS: [&str; 9] =
-        ["\u{202a}", "\u{202b}", "\u{202c}", "\u{202d}", "\u{202e}", "\u{2066}", "\u{2067}", "\u{2068}", "\u{2069}"];
+    /// See [Sanitizer::parse_safe_char] for an explanation.
+    ///
+    /// Instead of using this list to test whether its elements occur in a string literal,
+    /// we should probably rework the string literal parsing code
+    /// to call [Sanitizer::parse_safe_char],
+    /// and remove this list.
+    const UNSUPPORTED_CODE_POINTS: [&str; 39] = [
+        "\x00", "\x01", "\x02", "\x03", "\x04", "\x05", "\x06", "\x07", "\x08", "\x0b", "\x0c", "\x0e", "\x0f", "\x10",
+        "\x11", "\x12", "\x13", "\x14", "\x15", "\x16", "\x17", "\x18", "\x19", "\x1a", "\x1b", "\x1c", "\x1d", "\x1e",
+        "\x1f", "\x7f", "\u{202a}", "\u{202b}", "\u{202c}", "\u{202d}", "\u{202e}", "\u{2066}", "\u{2067}", "\u{2068}",
+        "\u{2069}",
+    ];
 
     /// Parse a unicode sequence, of the form u{XXXX}, where XXXX is 1 to 6
     /// hexadecimal numerals. We will combine this later with parse_escaped_char
@@ -186,3 +196,23 @@ pub mod string_parser {
         delimited(char('"'), build_string, char('"'))(input)
     }
 }
+
+#[test]
+fn test_parse_string() {
+    // to use parse_string_wrapper instead of string_parser::parse_string::<nom::error::VerboseError<&str>> in the tests below:
+    fn parse_string_wrapper(input: &str) -> crate::ParserResult<String> {
+        string_parser::parse_string(input)
+    }
+
+    // tests some correct string literals:
+    assert_eq!(("", String::from("")), parse_string_wrapper("\"\"").unwrap());
+    assert_eq!(("", String::from("abc")), parse_string_wrapper("\"abc\"").unwrap());
+    assert_eq!((" and more", String::from("abc")), parse_string_wrapper("\"abc\" and more").unwrap());
+    assert_eq!(("", String::from("\r")), parse_string_wrapper("\"\r\"").unwrap());
+    assert_eq!(("", String::from("4\u{4141}x\x09")), parse_string_wrapper("\"4\u{4141}x\x09\"").unwrap());
+
+    // test rejection of disallowed characters:
+    assert!(parse_string_wrapper("\"hel\x08lo\"").is_err());
+    assert!(parse_string_wrapper("\"hel\x1flo\"").is_err());
+    assert!(parse_string_wrapper("\"hel\u{2069}lo\"").is_err());
+}
diff --git a/console/program/src/data/identifier/from_bits.rs b/console/program/src/data/identifier/from_bits.rs
index 3f241fdf0..1f3485fc1 100644
--- a/console/program/src/data/identifier/from_bits.rs
+++ b/console/program/src/data/identifier/from_bits.rs
@@ -24,9 +24,6 @@ impl<N: Network> FromBits for Identifier<N> {
         // the final step checks the byte-aligned field element is within the data capacity.
         ensure!(bits_le.len() <= Field::<N>::size_in_bits(), "Identifier exceeds the maximum bits allowed");
 
-        // Recover the field element from the bits.
-        let field = Field::<N>::from_bits_le(bits_le)?;
-
         // Convert the bits to bytes, and parse the bytes as a UTF-8 string.
         let bytes = bits_le.chunks(8).map(u8::from_bits_le).collect::<Result<Vec<u8>>>()?;
 
@@ -37,13 +34,8 @@ impl<N: Network> FromBits for Identifier<N> {
             None => bytes.len(),  // No null character found, so the identifier is the full length.
         };
 
-        // Ensure identifier fits within the data capacity of the base field.
-        let max_bytes = Field::<N>::size_in_data_bits() / 8; // Note: This intentionally rounds down.
-        match num_bytes <= max_bytes {
-            // Return the identifier.
-            true => Ok(Self(field, u8::try_from(num_bytes).or_halt_with::<N>("Invalid identifier byte size"))),
-            false => bail!("Identifier exceeds the maximum capacity allowed"),
-        }
+        // Parse the bytes as a UTF-8 string.
+        Self::from_str(str::from_utf8(&bytes[0..num_bytes])?)
     }
 
     /// Initializes a new identifier from a list of big-endian bits *without* leading zeros.
diff --git a/console/program/src/data/identifier/mod.rs b/console/program/src/data/identifier/mod.rs
index 834d903d4..5deab8f57 100644
--- a/console/program/src/data/identifier/mod.rs
+++ b/console/program/src/data/identifier/mod.rs
@@ -102,4 +102,11 @@ pub(crate) mod tests {
         }
         Ok(())
     }
+
+    #[test]
+    fn test_identifier_try_from_illegal() {
+        assert!(Identifier::<CurrentNetwork>::try_from("123").is_err());
+        assert!(Identifier::<CurrentNetwork>::try_from("abc\x08def").is_err());
+        assert!(Identifier::<CurrentNetwork>::try_from("abc\u{202a}def").is_err());
+    }
 }
diff --git a/console/program/src/data/identifier/parse.rs b/console/program/src/data/identifier/parse.rs
index 91692b32b..d67c08b5d 100644
--- a/console/program/src/data/identifier/parse.rs
+++ b/console/program/src/data/identifier/parse.rs
@@ -37,20 +37,15 @@ impl<N: Network> FromStr for Identifier<N> {
 
     /// Reads in an identifier from a string.
     fn from_str(identifier: &str) -> Result<Self, Self::Err> {
-        // Ensure the identifier is not an empty string, and does not start with a number.
+        // Ensure the identifier is not an empty string, and starts with an ASCII letter.
         match identifier.chars().next() {
-            Some(character) => ensure!(!character.is_numeric(), "Identifier cannot start with a number"),
-            None => bail!("Identifier cannot be an empty string"),
+            Some(character) => ensure!(character.is_ascii_alphabetic(), "Identifier must start with a letter"),
+            None => bail!("Identifier cannot be empty"),
         }
 
-        // Ensure the identifier is alphanumeric and underscores.
-        if identifier.chars().any(|character| !character.is_alphanumeric() && character != '_') {
-            bail!("Identifier '{identifier}' must be alphanumeric and underscores")
-        }
-
-        // Ensure the identifier is not solely underscores.
-        if identifier.chars().all(|character| character == '_') {
-            bail!("Identifier cannot consist solely of underscores")
+        // Ensure the identifier consists of ASCII letters, ASCII digits, and underscores.
+        if identifier.chars().any(|character| !character.is_ascii_alphanumeric() && character != '_') {
+            bail!("Identifier '{identifier}' must consist of letters, digits, and underscores")
         }
 
         // Ensure identifier fits within the data capacity of the base field.
@@ -192,6 +187,9 @@ mod tests {
 
     #[test]
     fn test_from_str_fails() {
+        // Must be non-empty.
+        assert!(Identifier::<CurrentNetwork>::from_str("").is_err());
+
         // Must be alphanumeric or underscore.
         assert!(Identifier::<CurrentNetwork>::from_str("foo_bar~baz").is_err());
         assert!(Identifier::<CurrentNetwork>::from_str("foo_bar-baz").is_err());
@@ -210,6 +208,13 @@ mod tests {
         assert!(Identifier::<CurrentNetwork>::from_str("12").is_err());
         assert!(Identifier::<CurrentNetwork>::from_str("111").is_err());
 
+        // Must not start with underscore.
+        assert!(Identifier::<CurrentNetwork>::from_str("_foo").is_err());
+
+        // Must be ASCII.
+        assert!(Identifier::<CurrentNetwork>::from_str("\u{03b1}").is_err()); // Greek alpha
+        assert!(Identifier::<CurrentNetwork>::from_str("\u{03b2}").is_err()); // Greek beta
+
         // Must fit within the data capacity of a base field element.
         let identifier = Identifier::<CurrentNetwork>::from_str(
             "foo_bar_baz_qux_quux_quuz_corge_grault_garply_waldo_fred_plugh_xyzzy",
diff --git a/curves/Cargo.toml b/curves/Cargo.toml
index 9b10f74b9..76acb8bf7 100644
--- a/curves/Cargo.toml
+++ b/curves/Cargo.toml
@@ -60,7 +60,7 @@ version = "1.0"
 version = "1.3.3"
 
 [dev-dependencies.criterion]
-version = "0.3"
+version = "0.4"
 
 [dev-dependencies.rand]
 version = "0.8"
diff --git a/curves/src/traits/tests_field.rs b/curves/src/traits/tests_field.rs
index 4ad5e3211..837184dac 100644
--- a/curves/src/traits/tests_field.rs
+++ b/curves/src/traits/tests_field.rs
@@ -364,8 +364,8 @@ pub(crate) fn bench_sqrt<F: PrimeField + SquareRootField>(rng: &mut TestRng) {
     let avg_time_a = profile_a.into_iter().reduce(|a, b| a.checked_add(b).unwrap()).unwrap() / num_a;
     let avg_time_b = profile_b.into_iter().reduce(|a, b| a.checked_add(b).unwrap()).unwrap() / num_b;
 
-    println!("{avg_time_a} vs. {avg_time_b} ({}%)", (avg_time_b - avg_time_a) * 100 / avg_time_a);
-    assert!(avg_time_a < avg_time_b);
+    println!("{avg_time_a} vs. {avg_time_b} ({}%)", avg_time_b.saturating_sub(avg_time_a) * 100 / avg_time_a);
+    assert!(avg_time_a <= avg_time_b, "This is an acceptable failure. It is for a timer.");
 }
 
 /// Returns the square root using the Tonelli-Shanks algorithm.
diff --git a/rest/src/lib.rs b/rest/src/lib.rs
index eadce6309..9b0c08f31 100644
--- a/rest/src/lib.rs
+++ b/rest/src/lib.rs
@@ -28,7 +28,7 @@ pub use routes::*;
 mod start;
 pub use start::*;
 
-use snarkvm_compiler::{BlockStorage, Ledger, ProgramStorage, RecordsFilter, Transaction};
+use snarkvm_compiler::{BlockStorage, Ledger, Program, ProgramStorage, RecordsFilter, Transaction};
 use snarkvm_console::{
     account::{Address, ViewKey},
     prelude::Network,
@@ -40,7 +40,7 @@ use anyhow::Result;
 use indexmap::IndexMap;
 use parking_lot::RwLock;
 use serde::{Deserialize, Serialize};
-use std::sync::Arc;
+use std::{str::FromStr, sync::Arc};
 use tokio::{sync::mpsc, task::JoinHandle};
 use warp::{http::StatusCode, reject, reply, Filter, Rejection, Reply};
 
diff --git a/rest/src/routes.rs b/rest/src/routes.rs
index 4df09050e..227c3482d 100644
--- a/rest/src/routes.rs
+++ b/rest/src/routes.rs
@@ -25,7 +25,7 @@ struct BlockRange {
 
 impl<N: Network, B: BlockStorage<N>, P: ProgramStorage<N>> Server<N, B, P> {
     /// Initializes the routes, given the ledger and ledger sender.
-    #[allow(clippy::redundant_clone)]
+    #[allow(clippy::redundant_clone, opaque_hidden_inferred_bound)]
     pub fn routes(&self) -> impl Filter<Extract = impl Reply, Error = Rejection> + Clone {
         // GET /testnet3/latest/height
         let latest_height = warp::get()
@@ -94,9 +94,9 @@ impl<N: Network, B: BlockStorage<N>, P: ProgramStorage<N>> Server<N, B, P> {
 
         // GET /testnet3/statePath/{commitment}
         let get_state_path = warp::get()
-            .and(warp::path!("testnet3" / "statePath"))
-            .and(warp::body::content_length_limit(128))
-            .and(warp::body::json())
+            .and(warp::path!("testnet3" / "statePath" / ..))
+            .and(warp::path::param::<Field<N>>())
+            .and(warp::path::end())
             .and(with(self.ledger.clone()))
             .and_then(Self::get_state_path);
 
@@ -225,7 +225,13 @@ impl<N: Network, B: BlockStorage<N>, P: ProgramStorage<N>> Server<N, B, P> {
         program_id: ProgramID<N>,
         ledger: Arc<RwLock<Ledger<N, B, P>>>,
     ) -> Result<impl Reply, Rejection> {
-        Ok(reply::json(&ledger.read().get_program(program_id).or_reject()?))
+        let program = if program_id == ProgramID::<N>::from_str("credits.aleo").or_reject()? {
+            Program::<N>::credits().or_reject()?
+        } else {
+            ledger.read().get_program(program_id).or_reject()?
+        };
+
+        Ok(reply::json(&program))
     }
 
     /// Returns the list of current validators.
diff --git a/vm/compiler/Cargo.toml b/vm/compiler/Cargo.toml
index 38d796cfc..50c7f800c 100644
--- a/vm/compiler/Cargo.toml
+++ b/vm/compiler/Cargo.toml
@@ -87,7 +87,7 @@ version = "1.0"
 version = "1.0"
 
 [dependencies.time]
-version = "0.3.14"
+version = "0.3.15"
 
 [dependencies.tracing]
 version = "0.1"
diff --git a/vm/compiler/src/process/mod.rs b/vm/compiler/src/process/mod.rs
index a5a8b820b..7644378c9 100644
--- a/vm/compiler/src/process/mod.rs
+++ b/vm/compiler/src/process/mod.rs
@@ -775,11 +775,17 @@ closure execute:
     output r3 as field;
     output r2 as field;
 
+closure check_not_equal:
+    input r0 as field;
+    input r1 as field;
+    assert.neq r0 r1;
+
 function compute:
     input r0 as field.private;
     input r1 as field.public;
     input r2 as token.record;
     cast r2.owner r2.gates r2.token_amount into r3 as token.record;
+    call check_not_equal r0 r1;
     call execute r0 r1 into r4 r5 r6;
     output r3 as token.record;
     output r4 as field.private;
@@ -891,6 +897,16 @@ function mint:
     cast r0 0u64 r1 into r2 as token.record;
     output r2 as token.record;
 
+function produce_magic_number:
+    add 1234u64 0u64 into r0;
+    output r0 as u64.private;
+
+function check_magic_number:
+    input r0 as u64.private;
+    assert.eq r0 1234u64;
+
+function noop:
+
 function transfer:
     input r0 as token.record;
     input r1 as address.private;
@@ -917,9 +933,12 @@ function transfer:
     input r0 as token.aleo/token.record;
     input r1 as address.private;
     input r2 as u64.private;
-    call token.aleo/transfer r0 r1 r2 into r3 r4;
-    output r3 as token.aleo/token.record;
-    output r4 as token.aleo/token.record;",
+    call token.aleo/noop;
+    call token.aleo/produce_magic_number into r3;
+    call token.aleo/check_magic_number r3;
+    call token.aleo/transfer r0 r1 r2 into r4 r5;
+    output r4 as token.aleo/token.record;
+    output r5 as token.aleo/token.record;",
         )
         .unwrap();
         assert!(string.is_empty(), "Parser did not consume all of the string: '{string}'");
@@ -953,7 +972,7 @@ function transfer:
         let authorization = process
             .authorize::<CurrentAleo, _>(&caller0_private_key, program1.id(), function_name, &[r0, r1, r2], rng)
             .unwrap();
-        assert_eq!(authorization.len(), 2);
+        assert_eq!(authorization.len(), 5);
         println!("\nAuthorize\n{:#?}\n\n", authorization.to_vec_deque());
 
         let (output_a, output_b) = {
@@ -982,7 +1001,7 @@ function transfer:
         };
 
         // Check again to make sure we didn't modify the authorization before calling `evaluate`.
-        assert_eq!(authorization.len(), 2);
+        assert_eq!(authorization.len(), 5);
 
         // Compute the output value.
         let response = process.evaluate::<CurrentAleo>(authorization.replicate()).unwrap();
@@ -992,7 +1011,7 @@ function transfer:
         assert_eq!(output_b, candidate[1]);
 
         // Check again to make sure we didn't modify the authorization after calling `evaluate`.
-        assert_eq!(authorization.len(), 2);
+        assert_eq!(authorization.len(), 5);
 
         // Execute the request.
         let (response, execution) = process.execute::<CurrentAleo, _>(authorization, rng).unwrap();
@@ -1325,4 +1344,140 @@ finalize mint_public:
             store.get_value(program_id, &mapping_name, &Plaintext::from(Literal::Address(caller))).unwrap().unwrap();
         assert_eq!(candidate, Value::from_str("3u64").unwrap());
     }
+
+    #[test]
+    fn test_process_execute_call_mint_public() {
+        // Initialize a new program.
+        let (string, program0) = Program::<CurrentNetwork>::parse(
+            r"
+program token.aleo;
+
+// On-chain storage of an `account` map, with `owner` as the key,
+// and `amount` as the value.
+mapping account:
+    // The token owner.
+    key owner as address.public;
+    // The token amount.
+    value amount as u64.public;
+
+// The function `mint_public` issues the specified token amount
+// for the token receiver publicly on the network.
+function mint_public:
+    // Input the token receiver.
+    input r0 as address.public;
+    // Input the token amount.
+    input r1 as u64.public;
+    // Mint the tokens publicly.
+    finalize r0 r1;
+
+// The finalize scope of `mint_public` increments the
+// `account` of the token receiver by the specified amount.
+finalize mint_public:
+    // Input the token receiver.
+    input r0 as address.public;
+    // Input the token amount.
+    input r1 as u64.public;
+
+    // Increments `account[r0]` by `r1`.
+    // If `account[r0]` does not exist, it will be created.
+    // If `account[r0] + r1` overflows, `mint_public` is reverted.
+    increment account[r0] by r1;
+",
+        )
+        .unwrap();
+        assert!(string.is_empty(), "Parser did not consume all of the string: '{string}'");
+
+        // Declare the mapping.
+        let mapping_name = Identifier::from_str("account").unwrap();
+        // Declare the function name.
+        let function_name = Identifier::from_str("mint_public").unwrap();
+
+        // Initialize the RNG.
+        let rng = &mut TestRng::default();
+
+        // Construct the process.
+        let process = super::test_helpers::sample_process(&program0);
+        // Check that the circuit key can be synthesized.
+        process.synthesize_key::<CurrentAleo, _>(program0.id(), &function_name, rng).unwrap();
+
+        // Reset the process.
+        let mut process = Process::load().unwrap();
+
+        // Initialize a new program store.
+        let store = ProgramStore::<_, ProgramMemory<_>>::open(None).unwrap();
+
+        // Add the program to the process.
+        let deployment = process.deploy::<CurrentAleo, _>(&program0, rng).unwrap();
+        // Check that the deployment verifies.
+        process.verify_deployment::<CurrentAleo, _>(&deployment, rng).unwrap();
+        // Finalize the deployment.
+        process.finalize_deployment(&store, &deployment).unwrap();
+
+        // TODO (howardwu): Remove this. I call this to synthesize the proving key independent of the assignment from 'execute'.
+        //  In general, we should update all tests to utilize a presynthesized proving key, before execution, to test
+        //  the correctness of the synthesizer.
+        process.synthesize_key::<CurrentAleo, _>(program0.id(), &function_name, rng).unwrap();
+
+        // Initialize another program.
+        let (string, program1) = Program::<CurrentNetwork>::parse(
+            r"
+import token.aleo;
+
+program public_wallet.aleo;
+
+function mint:
+    input r0 as address.public;
+    input r1 as u64.public;
+    call token.aleo/mint_public r0 r1;",
+        )
+        .unwrap();
+        assert!(string.is_empty(), "Parser did not consume all of the string: '{string}'");
+
+        // Add the program to the process.
+        process.add_program(&program1).unwrap();
+
+        // Initialize the RNG.
+        let rng = &mut TestRng::default();
+
+        // Initialize caller.
+        let caller_private_key = PrivateKey::<CurrentNetwork>::new(rng).unwrap();
+        let caller = Address::try_from(&caller_private_key).unwrap();
+
+        // Declare the function name.
+        let function_name = Identifier::from_str("mint").unwrap();
+
+        // Declare the input value.
+        let r0 = Value::<CurrentNetwork>::from_str(&caller.to_string()).unwrap();
+        let r1 = Value::<CurrentNetwork>::from_str("100u64").unwrap();
+
+        // Authorize the function call.
+        let authorization = process
+            .authorize::<CurrentAleo, _>(&caller_private_key, program1.id(), function_name, &[r0, r1], rng)
+            .unwrap();
+        assert_eq!(authorization.len(), 2);
+
+        // Compute the output value.
+        let response = process.evaluate::<CurrentAleo>(authorization.replicate()).unwrap();
+        let candidate = response.outputs();
+        assert_eq!(0, candidate.len());
+
+        // Check again to make sure we didn't modify the authorization after calling `evaluate`.
+        assert_eq!(authorization.len(), 2);
+
+        // Execute the request.
+        let (response, execution) = process.execute::<CurrentAleo, _>(authorization, rng).unwrap();
+        let candidate = response.outputs();
+        assert_eq!(0, candidate.len());
+
+        // Verify the execution.
+        assert!(process.verify_execution(&execution).is_ok());
+
+        // Now, finalize the execution.
+        process.finalize_execution(&store, &execution).unwrap();
+
+        // Check the account balance.
+        let candidate =
+            store.get_value(program0.id(), &mapping_name, &Plaintext::from(Literal::Address(caller))).unwrap().unwrap();
+        assert_eq!(candidate, Value::from_str("100u64").unwrap());
+    }
 }
diff --git a/vm/compiler/src/program/instruction/operation/call.rs b/vm/compiler/src/program/instruction/operation/call.rs
index 630280419..679c8a64a 100644
--- a/vm/compiler/src/program/instruction/operation/call.rs
+++ b/vm/compiler/src/program/instruction/operation/call.rs
@@ -534,7 +534,7 @@ impl<N: Network> Parser for Call<N> {
         // Parse the whitespace from the string.
         let (string, _) = Sanitizer::parse_whitespaces(string)?;
         // Parse the operands from the string.
-        let (string, operands) = map_res(many1(parse_operand), |operands: Vec<Operand<N>>| {
+        let (string, operands) = map_res(many0(complete(parse_operand)), |operands: Vec<Operand<N>>| {
             // Ensure the number of operands is within the bounds.
             match operands.len() <= N::MAX_OPERANDS {
                 true => Ok(operands),
@@ -543,18 +543,28 @@ impl<N: Network> Parser for Call<N> {
         })(string)?;
         // Parse the whitespace from the string.
         let (string, _) = Sanitizer::parse_whitespaces(string)?;
-        // Parse the "into" from the string.
-        let (string, _) = tag("into")(string)?;
-        // Parse the whitespace from the string.
-        let (string, _) = Sanitizer::parse_whitespaces(string)?;
-        // Parse the destination register from the string.
-        let (string, destinations) = map_res(many1(parse_destination), |destinations: Vec<Register<N>>| {
-            // Ensure the number of destination registers is within the bounds.
-            match destinations.len() <= N::MAX_OPERANDS {
-                true => Ok(destinations),
-                false => Err(error("Failed to parse 'call' opcode: too many destination registers")),
+
+        // Optionally parse the "into" from the string.
+        let (string, destinations) = match opt(tag("into"))(string)? {
+            // If the "into" was not parsed, return the string and an empty vector of destinations.
+            (string, None) => (string, vec![]),
+            // If the "into" was parsed, parse the destinations from the string.
+            (string, Some(_)) => {
+                // Parse the whitespace from the string.
+                let (string, _) = Sanitizer::parse_whitespaces(string)?;
+                // Parse the destinations from the string.
+                let (string, destinations) =
+                    map_res(many0(complete(parse_destination)), |destinations: Vec<Register<N>>| {
+                        // Ensure the number of destinations is within the bounds.
+                        match destinations.len() <= N::MAX_OPERANDS {
+                            true => Ok(destinations),
+                            false => Err(error("Failed to parse 'call' opcode: too many destinations")),
+                        }
+                    })(string)?;
+                // Return the string and the destinations.
+                (string, destinations)
             }
-        })(string)?;
+        };
 
         Ok((string, Self { operator, operands, destinations }))
     }
@@ -589,20 +599,23 @@ impl<N: Network> Display for Call<N> {
     /// Prints the operation to a string.
     fn fmt(&self, f: &mut Formatter) -> fmt::Result {
         // Ensure the number of operands is within the bounds.
-        if self.operands.len().is_zero() || self.operands.len() > N::MAX_OPERANDS {
-            eprintln!("The number of operands must be nonzero and <= {}", N::MAX_OPERANDS);
+        if self.operands.len() > N::MAX_OPERANDS {
+            eprintln!("The number of operands must be <= {}", N::MAX_OPERANDS);
             return Err(fmt::Error);
         }
         // Ensure the number of destinations is within the bounds.
-        if self.destinations.len().is_zero() || self.destinations.len() > N::MAX_OPERANDS {
-            eprintln!("The number of destinations must be nonzero and <= {}", N::MAX_OPERANDS);
+        if self.destinations.len() > N::MAX_OPERANDS {
+            eprintln!("The number of destinations must be <= {}", N::MAX_OPERANDS);
             return Err(fmt::Error);
         }
         // Print the operation.
         write!(f, "{} {}", Self::opcode(), self.operator)?;
         self.operands.iter().try_for_each(|operand| write!(f, " {operand}"))?;
-        write!(f, " into")?;
-        self.destinations.iter().try_for_each(|destination| write!(f, " {destination}"))
+        if !self.destinations.is_empty() {
+            write!(f, " into")?;
+            self.destinations.iter().try_for_each(|destination| write!(f, " {destination}"))?;
+        }
+        Ok(())
     }
 }
 
@@ -615,8 +628,8 @@ impl<N: Network> FromBytes for Call<N> {
         // Read the number of operands.
         let num_operands = u8::read_le(&mut reader)? as usize;
         // Ensure the number of operands is within the bounds.
-        if num_operands.is_zero() || num_operands > N::MAX_OPERANDS {
-            return Err(error(format!("The number of operands must be nonzero and <= {}", N::MAX_OPERANDS)));
+        if num_operands > N::MAX_OPERANDS {
+            return Err(error(format!("The number of operands must be <= {}", N::MAX_OPERANDS)));
         }
 
         // Initialize the vector for the operands.
@@ -629,8 +642,8 @@ impl<N: Network> FromBytes for Call<N> {
         // Read the number of destination registers.
         let num_destinations = u8::read_le(&mut reader)? as usize;
         // Ensure the number of destinations is within the bounds.
-        if num_destinations.is_zero() || num_destinations > N::MAX_OPERANDS {
-            return Err(error(format!("The number of destinations must be nonzero and <= {}", N::MAX_OPERANDS)));
+        if num_destinations > N::MAX_OPERANDS {
+            return Err(error(format!("The number of destinations must be <= {}", N::MAX_OPERANDS)));
         }
 
         // Initialize the vector for the destinations.
@@ -649,12 +662,12 @@ impl<N: Network> ToBytes for Call<N> {
     /// Writes the operation to a buffer.
     fn write_le<W: Write>(&self, mut writer: W) -> IoResult<()> {
         // Ensure the number of operands is within the bounds.
-        if self.operands.len().is_zero() || self.operands.len() > N::MAX_OPERANDS {
-            return Err(error(format!("The number of operands must be nonzero and <= {}", N::MAX_OPERANDS)));
+        if self.operands.len() > N::MAX_OPERANDS {
+            return Err(error(format!("The number of operands must be <= {}", N::MAX_OPERANDS)));
         }
         // Ensure the number of destinations is within the bounds.
-        if self.destinations.len().is_zero() || self.destinations.len() > N::MAX_OPERANDS {
-            return Err(error(format!("The number of destinations must be nonzero and <= {}", N::MAX_OPERANDS)));
+        if self.destinations.len() > N::MAX_OPERANDS {
+            return Err(error(format!("The number of destinations must be <= {}", N::MAX_OPERANDS)));
         }
 
         // Write the name of the call.
@@ -673,35 +686,108 @@ impl<N: Network> ToBytes for Call<N> {
 #[cfg(test)]
 mod tests {
     use super::*;
-    use console::{network::Testnet3, program::Identifier};
+    use console::{
+        network::Testnet3,
+        program::{Address, Identifier, Literal, U64},
+    };
 
     type CurrentNetwork = Testnet3;
 
+    const TEST_CASES: &[&str] = &[
+        "call foo",
+        "call foo r0",
+        "call foo r0.owner",
+        "call foo r0 r1",
+        "call foo r0.owner r0.gates",
+        "call foo into r0",
+        "call foo into r0 r1",
+        "call foo into r0 r1 r2",
+        "call foo r0 into r1",
+        "call foo r0 r1 into r2",
+        "call foo r0 r1 into r2 r3",
+        "call foo r0 r1 r2 into r3 r4",
+        "call foo r0 r1 r2 into r3 r4 r5",
+    ];
+
+    fn check_parser(
+        string: &str,
+        expected_operator: CallOperator<CurrentNetwork>,
+        expected_operands: Vec<Operand<CurrentNetwork>>,
+        expected_destinations: Vec<Register<CurrentNetwork>>,
+    ) {
+        // Check that the parser works.
+        let (string, call) = Call::<CurrentNetwork>::parse(string).unwrap();
+
+        // Check that the entire string was consumed.
+        assert!(string.is_empty(), "Parser did not consume all of the string: '{string}'");
+
+        // Check that the operator is correct.
+        assert_eq!(call.operator, expected_operator, "The call operator is incorrect");
+
+        // Check that the operands are correct.
+        assert_eq!(call.operands.len(), expected_operands.len(), "The number of operands is incorrect");
+        for (i, (given, expected)) in call.operands.iter().zip(expected_operands.iter()).enumerate() {
+            assert_eq!(given, expected, "The {}-th operand is incorrect", i);
+        }
+
+        // Check that the destinations are correct.
+        assert_eq!(call.destinations.len(), expected_destinations.len(), "The number of destinations is incorrect");
+        for (i, (given, expected)) in call.destinations.iter().zip(expected_destinations.iter()).enumerate() {
+            assert_eq!(given, expected, "The {}-th destination is incorrect", i);
+        }
+    }
+
     #[test]
     fn test_parse() {
-        let (string, call) =
-            Call::<CurrentNetwork>::parse("call transfer r0.owner r0.gates r0.token_amount into r1 r2 r3").unwrap();
-        assert!(string.is_empty(), "Parser did not consume all of the string: '{string}'");
-        assert_eq!(call.operator, CallOperator::from_str("transfer").unwrap(), "The call operator is incorrect");
-        assert_eq!(call.operands.len(), 3, "The number of operands is incorrect");
-        assert_eq!(
-            call.operands[0],
-            Operand::Register(Register::Member(0, vec![Identifier::from_str("owner").unwrap()])),
-            "The first operand is incorrect"
+        check_parser(
+            "call transfer r0.owner r0.gates r0.token_amount into r1 r2 r3",
+            CallOperator::from_str("transfer").unwrap(),
+            vec![
+                Operand::Register(Register::Member(0, vec![Identifier::from_str("owner").unwrap()])),
+                Operand::Register(Register::Member(0, vec![Identifier::from_str("gates").unwrap()])),
+                Operand::Register(Register::Member(0, vec![Identifier::from_str("token_amount").unwrap()])),
+            ],
+            vec![Register::Locator(1), Register::Locator(2), Register::Locator(3)],
         );
-        assert_eq!(
-            call.operands[1],
-            Operand::Register(Register::Member(0, vec![Identifier::from_str("gates").unwrap()])),
-            "The second operand is incorrect"
+
+        check_parser(
+            "call mint_public aleo1wfyyj2uvwuqw0c0dqa5x70wrawnlkkvuepn4y08xyaqfqqwweqys39jayw 100u64",
+            CallOperator::from_str("mint_public").unwrap(),
+            vec![
+                Operand::Literal(Literal::Address(
+                    Address::from_str("aleo1wfyyj2uvwuqw0c0dqa5x70wrawnlkkvuepn4y08xyaqfqqwweqys39jayw").unwrap(),
+                )),
+                Operand::Literal(Literal::U64(U64::from_str("100u64").unwrap())),
+            ],
+            vec![],
         );
-        assert_eq!(
-            call.operands[2],
-            Operand::Register(Register::Member(0, vec![Identifier::from_str("token_amount").unwrap()])),
-            "The third operand is incorrect"
+
+        check_parser(
+            "call get_magic_number into r0",
+            CallOperator::from_str("get_magic_number").unwrap(),
+            vec![],
+            vec![Register::Locator(0)],
         );
-        assert_eq!(call.destinations.len(), 3, "The number of destinations is incorrect");
-        assert_eq!(call.destinations[0], Register::Locator(1), "The first destination register is incorrect");
-        assert_eq!(call.destinations[1], Register::Locator(2), "The second destination register is incorrect");
-        assert_eq!(call.destinations[2], Register::Locator(3), "The third destination register is incorrect");
+
+        check_parser("call noop", CallOperator::from_str("noop").unwrap(), vec![], vec![])
+    }
+
+    #[test]
+    fn test_display() {
+        for expected in TEST_CASES {
+            assert_eq!(Call::<CurrentNetwork>::from_str(expected).unwrap().to_string(), *expected);
+        }
+    }
+
+    #[test]
+    fn test_bytes() {
+        for case in TEST_CASES {
+            let expected = Call::<CurrentNetwork>::from_str(case).unwrap();
+
+            // Check the byte representation.
+            let expected_bytes = expected.to_bytes_le().unwrap();
+            assert_eq!(expected, Call::read_le(&expected_bytes[..]).unwrap());
+            assert!(Call::<CurrentNetwork>::read_le(&expected_bytes[1..]).is_err());
+        }
     }
 }
