diff --git a/src/network/server.rs b/src/network/server.rs
index 31d62ae2a..35d382120 100644
--- a/src/network/server.rs
+++ b/src/network/server.rs
@@ -297,12 +297,13 @@ impl<N: Network, E: Environment> Server<N, E> {
             loop {
                 info!(
                     r"
+
 ===========================================================================================================
                                   Aleo Testnet2 - Incentivization Period
 ===========================================================================================================
 
     In preparation for the incentivization period, testnet2 will automatically reset within 48 hours.
-    Please ensure your Aleo node is runnning either the `run-client.sh` or `run-miner.sh` script,
+    Please ensure your Aleo node is running either the `run-client.sh` or `run-miner.sh` script,
     in order to automatically upgrade to the incentivized testnet.
 
 ===========================================================================================================
