diff --git a/node/rest/src/lib.rs b/node/rest/src/lib.rs
index 17881a511..0b3b95c6d 100644
--- a/node/rest/src/lib.rs
+++ b/node/rest/src/lib.rs
@@ -153,6 +153,7 @@ impl<N: Network, C: ConsensusStorage<N>, R: Routing<N>> Rest<N, C, R> {
             .route(&format!("/{network}/block/:height_or_hash"), get(Self::get_block))
             // The path param here is actually only the height, but the name must match the route
             // above, otherwise there'll be a conflict at runtime.
+            .route(&format!("/{network}/block/:height_or_hash/header"), get(Self::get_block_header))
             .route(&format!("/{network}/block/:height_or_hash/transactions"), get(Self::get_block_transactions))
 
             // GET and POST ../transaction/..
diff --git a/node/rest/src/routes.rs b/node/rest/src/routes.rs
index d5d1b516b..0b59eafe3 100644
--- a/node/rest/src/routes.rs
+++ b/node/rest/src/routes.rs
@@ -125,6 +125,14 @@ impl<N: Network, C: ConsensusStorage<N>, R: Routing<N>> Rest<N, C, R> {
         Ok(ErasedJson::pretty(rest.ledger.get_height(&hash)?))
     }
 
+    // GET /<network>/block/{height}/header
+    pub(crate) async fn get_block_header(
+        State(rest): State<Self>,
+        Path(height): Path<u32>,
+    ) -> Result<ErasedJson, RestError> {
+        Ok(ErasedJson::pretty(rest.ledger.get_header(height)?))
+    }
+
     // GET /<network>/block/{height}/transactions
     pub(crate) async fn get_block_transactions(
         State(rest): State<Self>,
diff --git a/node/src/client/router.rs b/node/src/client/router.rs
index 08da9da72..4362f79ca 100644
--- a/node/src/client/router.rs
+++ b/node/src/client/router.rs
@@ -146,8 +146,9 @@ impl<N: Network, C: ConsensusStorage<N>> Client<N, C> {
     ) {
         // Process the message. Disconnect if the peer violated the protocol.
         if let Err(error) = self.inbound(peer_addr, message).await {
+            error!("Failed to process inbound message from '{peer_addr}' - {error}");
             if let Some(peer_ip) = self.router().resolve_to_listener(&peer_addr) {
-                warn!("Disconnecting from '{peer_ip}' - {error}");
+                warn!("Disconnecting from '{peer_ip}' for protocol violation");
                 Outbound::send(self, peer_ip, Message::Disconnect(DisconnectReason::ProtocolViolation.into()));
                 // Disconnect from this peer.
                 self.router().disconnect(peer_ip);
diff --git a/node/src/validator/router.rs b/node/src/validator/router.rs
index 2ee5a57e9..c9d1ab6d3 100644
--- a/node/src/validator/router.rs
+++ b/node/src/validator/router.rs
@@ -137,8 +137,9 @@ impl<N: Network, C: ConsensusStorage<N>> Validator<N, C> {
     ) {
         // Process the message. Disconnect if the peer violated the protocol.
         if let Err(error) = self.inbound(peer_addr, message).await {
+            error!("Failed to process inbound message from '{peer_addr}' - {error}");
             if let Some(peer_ip) = self.router().resolve_to_listener(&peer_addr) {
-                warn!("Disconnecting from '{peer_ip}' - {error}");
+                warn!("Disconnecting from '{peer_ip}' for protocol violation");
                 Outbound::send(self, peer_ip, Message::Disconnect(DisconnectReason::ProtocolViolation.into()));
                 // Disconnect from this peer.
                 self.router().disconnect(peer_ip);
