diff --git a/circuit/environment/src/canary_circuit.rs b/circuit/environment/src/canary_circuit.rs
new file mode 100644
index 000000000..230e51265
--- /dev/null
+++ b/circuit/environment/src/canary_circuit.rs
@@ -0,0 +1,409 @@
+// Copyright (C) 2019-2023 Aleo Systems Inc.
+// This file is part of the snarkVM library.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+// http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::{helpers::Constraint, Mode, *};
+
+use core::{
+    cell::{Cell, RefCell},
+    fmt,
+};
+
+type Field = <console::CanaryV0 as console::Environment>::Field;
+
+thread_local! {
+    static VARIABLE_LIMIT: Cell<Option<u64>> = Cell::new(None);
+    static CONSTRAINT_LIMIT: Cell<Option<u64>> = Cell::new(None);
+    pub(super) static CANARY_CIRCUIT: RefCell<R1CS<Field>> = RefCell::new(R1CS::new());
+    static IN_WITNESS: Cell<bool> = Cell::new(false);
+    static ZERO: LinearCombination<Field> = LinearCombination::zero();
+    static ONE: LinearCombination<Field> = LinearCombination::one();
+}
+
+#[derive(Copy, Clone, Debug, Eq, PartialEq, Hash)]
+pub struct CanaryCircuit;
+
+impl Environment for CanaryCircuit {
+    type Affine = <console::CanaryV0 as console::Environment>::Affine;
+    type BaseField = Field;
+    type Network = console::CanaryV0;
+    type ScalarField = <console::CanaryV0 as console::Environment>::Scalar;
+
+    /// Returns the `zero` constant.
+    fn zero() -> LinearCombination<Self::BaseField> {
+        ZERO.with(|zero| zero.clone())
+    }
+
+    /// Returns the `one` constant.
+    fn one() -> LinearCombination<Self::BaseField> {
+        ONE.with(|one| one.clone())
+    }
+
+    /// Returns a new variable of the given mode and value.
+    fn new_variable(mode: Mode, value: Self::BaseField) -> Variable<Self::BaseField> {
+        IN_WITNESS.with(|in_witness| {
+            // Ensure we are not in witness mode.
+            if !in_witness.get() {
+                // Ensure that we do not surpass the variable limit for the circuit.
+                VARIABLE_LIMIT.with(|variable_limit| {
+                    if let Some(limit) = variable_limit.get() {
+                        if Self::num_variables() > limit {
+                            Self::halt(format!("Surpassed the variable limit ({limit})"))
+                        }
+                    }
+                });
+                CANARY_CIRCUIT.with(|circuit| match mode {
+                    Mode::Constant => circuit.borrow_mut().new_constant(value),
+                    Mode::Public => circuit.borrow_mut().new_public(value),
+                    Mode::Private => circuit.borrow_mut().new_private(value),
+                })
+            } else {
+                Self::halt("Tried to initialize a new variable in witness mode")
+            }
+        })
+    }
+
+    /// Returns a new witness of the given mode and value.
+    fn new_witness<Fn: FnOnce() -> Output::Primitive, Output: Inject>(mode: Mode, logic: Fn) -> Output {
+        IN_WITNESS.with(|in_witness| {
+            // Set the entire environment to witness mode.
+            in_witness.replace(true);
+
+            // Run the logic.
+            let output = logic();
+
+            // Return the entire environment from witness mode.
+            in_witness.replace(false);
+
+            Inject::new(mode, output)
+        })
+    }
+
+    /// Enters a new scope for the environment.
+    fn scope<S: Into<String>, Fn, Output>(name: S, logic: Fn) -> Output
+    where
+        Fn: FnOnce() -> Output,
+    {
+        IN_WITNESS.with(|in_witness| {
+            // Ensure we are not in witness mode.
+            if !in_witness.get() {
+                CANARY_CIRCUIT.with(|circuit| {
+                    // Set the entire environment to the new scope.
+                    let name = name.into();
+                    if let Err(error) = circuit.borrow_mut().push_scope(&name) {
+                        Self::halt(error)
+                    }
+
+                    // Run the logic.
+                    let output = logic();
+
+                    // Return the entire environment to the previous scope.
+                    if let Err(error) = circuit.borrow_mut().pop_scope(name) {
+                        Self::halt(error)
+                    }
+
+                    output
+                })
+            } else {
+                Self::halt("Tried to initialize a new scope in witness mode")
+            }
+        })
+    }
+
+    /// Adds one constraint enforcing that `(A * B) == C`.
+    fn enforce<Fn, A, B, C>(constraint: Fn)
+    where
+        Fn: FnOnce() -> (A, B, C),
+        A: Into<LinearCombination<Self::BaseField>>,
+        B: Into<LinearCombination<Self::BaseField>>,
+        C: Into<LinearCombination<Self::BaseField>>,
+    {
+        IN_WITNESS.with(|in_witness| {
+            // Ensure we are not in witness mode.
+            if !in_witness.get() {
+                CANARY_CIRCUIT.with(|circuit| {
+                    // Ensure that we do not surpass the constraint limit for the circuit.
+                    CONSTRAINT_LIMIT.with(|constraint_limit| {
+                        if let Some(limit) = constraint_limit.get() {
+                            if circuit.borrow().num_constraints() > limit {
+                                Self::halt(format!("Surpassed the constraint limit ({limit})"))
+                            }
+                        }
+                    });
+
+                    let (a, b, c) = constraint();
+                    let (a, b, c) = (a.into(), b.into(), c.into());
+
+                    // Ensure the constraint is not comprised of constants.
+                    match a.is_constant() && b.is_constant() && c.is_constant() {
+                        true => {
+                            // Evaluate the constant constraint.
+                            assert_eq!(
+                                a.value() * b.value(),
+                                c.value(),
+                                "Constant constraint failed: ({a} * {b}) =?= {c}"
+                            );
+
+                            // match self.counter.scope().is_empty() {
+                            //     true => println!("Enforced constraint with constant terms: ({} * {}) =?= {}", a, b, c),
+                            //     false => println!(
+                            //         "Enforced constraint with constant terms ({}): ({} * {}) =?= {}",
+                            //         self.counter.scope(), a, b, c
+                            //     ),
+                            // }
+                        }
+                        false => {
+                            // Construct the constraint object.
+                            let constraint = Constraint(circuit.borrow().scope(), a, b, c);
+                            // Append the constraint.
+                            circuit.borrow_mut().enforce(constraint)
+                        }
+                    }
+                });
+            } else {
+                Self::halt("Tried to add a new constraint in witness mode")
+            }
+        })
+    }
+
+    /// Returns `true` if all constraints in the environment are satisfied.
+    fn is_satisfied() -> bool {
+        CANARY_CIRCUIT.with(|circuit| circuit.borrow().is_satisfied())
+    }
+
+    /// Returns `true` if all constraints in the current scope are satisfied.
+    fn is_satisfied_in_scope() -> bool {
+        CANARY_CIRCUIT.with(|circuit| circuit.borrow().is_satisfied_in_scope())
+    }
+
+    /// Returns the number of constants in the entire circuit.
+    fn num_constants() -> u64 {
+        CANARY_CIRCUIT.with(|circuit| circuit.borrow().num_constants())
+    }
+
+    /// Returns the number of public variables in the entire circuit.
+    fn num_public() -> u64 {
+        CANARY_CIRCUIT.with(|circuit| circuit.borrow().num_public())
+    }
+
+    /// Returns the number of private variables in the entire circuit.
+    fn num_private() -> u64 {
+        CANARY_CIRCUIT.with(|circuit| circuit.borrow().num_private())
+    }
+
+    /// Returns the number of constant, public, and private variables in the entire circuit.
+    fn num_variables() -> u64 {
+        CANARY_CIRCUIT.with(|circuit| circuit.borrow().num_variables())
+    }
+
+    /// Returns the number of constraints in the entire circuit.
+    fn num_constraints() -> u64 {
+        CANARY_CIRCUIT.with(|circuit| circuit.borrow().num_constraints())
+    }
+
+    /// Returns the number of nonzeros in the entire circuit.
+    fn num_nonzeros() -> (u64, u64, u64) {
+        CANARY_CIRCUIT.with(|circuit| circuit.borrow().num_nonzeros())
+    }
+
+    /// Returns the number of constants for the current scope.
+    fn num_constants_in_scope() -> u64 {
+        CANARY_CIRCUIT.with(|circuit| circuit.borrow().num_constants_in_scope())
+    }
+
+    /// Returns the number of public variables for the current scope.
+    fn num_public_in_scope() -> u64 {
+        CANARY_CIRCUIT.with(|circuit| circuit.borrow().num_public_in_scope())
+    }
+
+    /// Returns the number of private variables for the current scope.
+    fn num_private_in_scope() -> u64 {
+        CANARY_CIRCUIT.with(|circuit| circuit.borrow().num_private_in_scope())
+    }
+
+    /// Returns the number of constraints for the current scope.
+    fn num_constraints_in_scope() -> u64 {
+        CANARY_CIRCUIT.with(|circuit| circuit.borrow().num_constraints_in_scope())
+    }
+
+    /// Returns the number of nonzeros for the current scope.
+    fn num_nonzeros_in_scope() -> (u64, u64, u64) {
+        CANARY_CIRCUIT.with(|circuit| circuit.borrow().num_nonzeros_in_scope())
+    }
+
+    /// Returns the variable limit for the circuit, if one exists.
+    fn get_variable_limit() -> Option<u64> {
+        VARIABLE_LIMIT.with(|current_limit| current_limit.get())
+    }
+
+    /// Sets the variable limit for the circuit.
+    fn set_variable_limit(limit: Option<u64>) {
+        VARIABLE_LIMIT.with(|current_limit| current_limit.replace(limit));
+    }
+
+    /// Returns the constraint limit for the circuit, if one exists.
+    fn get_constraint_limit() -> Option<u64> {
+        CONSTRAINT_LIMIT.with(|current_limit| current_limit.get())
+    }
+
+    /// Sets the constraint limit for the circuit.
+    fn set_constraint_limit(limit: Option<u64>) {
+        CONSTRAINT_LIMIT.with(|current_limit| current_limit.replace(limit));
+    }
+
+    /// Halts the program from further synthesis, evaluation, and execution in the current environment.
+    fn halt<S: Into<String>, T>(message: S) -> T {
+        let error = message.into();
+        // eprintln!("{}", &error);
+        panic!("{}", &error)
+    }
+
+    /// Returns the R1CS circuit, resetting the circuit.
+    fn inject_r1cs(r1cs: R1CS<Self::BaseField>) {
+        CANARY_CIRCUIT.with(|circuit| {
+            // Ensure the circuit is empty before injecting.
+            assert_eq!(0, circuit.borrow().num_constants());
+            assert_eq!(1, circuit.borrow().num_public());
+            assert_eq!(0, circuit.borrow().num_private());
+            assert_eq!(1, circuit.borrow().num_variables());
+            assert_eq!(0, circuit.borrow().num_constraints());
+            // Inject the R1CS instance.
+            let r1cs = circuit.replace(r1cs);
+            // Ensure the circuit that was replaced is empty.
+            assert_eq!(0, r1cs.num_constants());
+            assert_eq!(1, r1cs.num_public());
+            assert_eq!(0, r1cs.num_private());
+            assert_eq!(1, r1cs.num_variables());
+            assert_eq!(0, r1cs.num_constraints());
+        })
+    }
+
+    /// Returns the R1CS circuit, resetting the circuit.
+    fn eject_r1cs_and_reset() -> R1CS<Self::BaseField> {
+        CANARY_CIRCUIT.with(|circuit| {
+            // Reset the witness mode.
+            IN_WITNESS.with(|in_witness| in_witness.replace(false));
+            // Reset the variable limit.
+            Self::set_variable_limit(None);
+            // Reset the constraint limit.
+            Self::set_constraint_limit(None);
+            // Eject the R1CS instance.
+            let r1cs = circuit.replace(R1CS::<<Self as Environment>::BaseField>::new());
+            // Ensure the circuit is now empty.
+            assert_eq!(0, circuit.borrow().num_constants());
+            assert_eq!(1, circuit.borrow().num_public());
+            assert_eq!(0, circuit.borrow().num_private());
+            assert_eq!(1, circuit.borrow().num_variables());
+            assert_eq!(0, circuit.borrow().num_constraints());
+            // Return the R1CS instance.
+            r1cs
+        })
+    }
+
+    /// Returns the R1CS assignment of the circuit, resetting the circuit.
+    fn eject_assignment_and_reset() -> Assignment<<Self::Network as console::Environment>::Field> {
+        CANARY_CIRCUIT.with(|circuit| {
+            // Reset the witness mode.
+            IN_WITNESS.with(|in_witness| in_witness.replace(false));
+            // Reset the variable limit.
+            Self::set_variable_limit(None);
+            // Reset the constraint limit.
+            Self::set_constraint_limit(None);
+            // Eject the R1CS instance.
+            let r1cs = circuit.replace(R1CS::<<Self as Environment>::BaseField>::new());
+            assert_eq!(0, circuit.borrow().num_constants());
+            assert_eq!(1, circuit.borrow().num_public());
+            assert_eq!(0, circuit.borrow().num_private());
+            assert_eq!(1, circuit.borrow().num_variables());
+            assert_eq!(0, circuit.borrow().num_constraints());
+            // Convert the R1CS instance to an assignment.
+            Assignment::from(r1cs)
+        })
+    }
+
+    /// Clears the circuit and initializes an empty environment.
+    fn reset() {
+        CANARY_CIRCUIT.with(|circuit| {
+            // Reset the witness mode.
+            IN_WITNESS.with(|in_witness| in_witness.replace(false));
+            // Reset the variable limit.
+            Self::set_variable_limit(None);
+            // Reset the constraint limit.
+            Self::set_constraint_limit(None);
+            // Reset the circuit.
+            *circuit.borrow_mut() = R1CS::<<Self as Environment>::BaseField>::new();
+            assert_eq!(0, circuit.borrow().num_constants());
+            assert_eq!(1, circuit.borrow().num_public());
+            assert_eq!(0, circuit.borrow().num_private());
+            assert_eq!(1, circuit.borrow().num_variables());
+            assert_eq!(0, circuit.borrow().num_constraints());
+        });
+    }
+}
+
+impl fmt::Display for CanaryCircuit {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        CANARY_CIRCUIT.with(|circuit| write!(f, "{}", circuit.borrow()))
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use snarkvm_circuit::prelude::*;
+
+    /// Compute 2^EXPONENT - 1, in a purposefully constraint-inefficient manner for testing.
+    fn create_example_circuit<E: Environment>() -> Field<E> {
+        let one = snarkvm_console_types::Field::<E::Network>::one();
+        let two = one + one;
+
+        const EXPONENT: u64 = 64;
+
+        // Compute 2^EXPONENT - 1, in a purposefully constraint-inefficient manner for testing.
+        let mut candidate = Field::<E>::new(Mode::Public, one);
+        let mut accumulator = Field::new(Mode::Private, two);
+        for _ in 0..EXPONENT {
+            candidate += &accumulator;
+            accumulator *= Field::new(Mode::Private, two);
+        }
+
+        assert_eq!((accumulator - Field::one()).eject_value(), candidate.eject_value());
+        assert_eq!(2, E::num_public());
+        assert_eq!(2 * EXPONENT + 1, E::num_private());
+        assert_eq!(EXPONENT, E::num_constraints());
+        assert!(E::is_satisfied());
+
+        candidate
+    }
+
+    #[test]
+    fn test_print_circuit() {
+        let _candidate = create_example_circuit::<CanaryCircuit>();
+        let output = format!("{CanaryCircuit}");
+        println!("{output}");
+    }
+
+    #[test]
+    fn test_circuit_scope() {
+        CanaryCircuit::scope("test_circuit_scope", || {
+            assert_eq!(0, CanaryCircuit::num_constants());
+            assert_eq!(1, CanaryCircuit::num_public());
+            assert_eq!(0, CanaryCircuit::num_private());
+            assert_eq!(0, CanaryCircuit::num_constraints());
+
+            assert_eq!(0, CanaryCircuit::num_constants_in_scope());
+            assert_eq!(0, CanaryCircuit::num_public_in_scope());
+            assert_eq!(0, CanaryCircuit::num_private_in_scope());
+            assert_eq!(0, CanaryCircuit::num_constraints_in_scope());
+        })
+    }
+}
diff --git a/circuit/environment/src/helpers/converter.rs b/circuit/environment/src/helpers/converter.rs
index 625b468f2..9ba96a414 100644
--- a/circuit/environment/src/helpers/converter.rs
+++ b/circuit/environment/src/helpers/converter.rs
@@ -12,7 +12,7 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-use crate::{Circuit, LinearCombination, TestnetCircuit, Variable, R1CS};
+use crate::{CanaryCircuit, Circuit, LinearCombination, TestnetCircuit, Variable, R1CS};
 use snarkvm_curves::edwards_bls12::Fq;
 use snarkvm_fields::PrimeField;
 
@@ -44,6 +44,16 @@ impl snarkvm_algorithms::r1cs::ConstraintSynthesizer<Fq> for TestnetCircuit {
     }
 }
 
+impl snarkvm_algorithms::r1cs::ConstraintSynthesizer<Fq> for CanaryCircuit {
+    /// Synthesizes the constraints from the environment into a `snarkvm_algorithms::r1cs`-compliant constraint system.
+    fn generate_constraints<CS: snarkvm_algorithms::r1cs::ConstraintSystem<Fq>>(
+        &self,
+        cs: &mut CS,
+    ) -> Result<(), snarkvm_algorithms::r1cs::SynthesisError> {
+        crate::canary_circuit::CANARY_CIRCUIT.with(|circuit| circuit.borrow().generate_constraints(cs))
+    }
+}
+
 impl<F: PrimeField> R1CS<F> {
     /// Synthesizes the constraints from the environment into a `snarkvm_algorithms::r1cs`-compliant constraint system.
     fn generate_constraints<CS: snarkvm_algorithms::r1cs::ConstraintSystem<F>>(
diff --git a/circuit/environment/src/lib.rs b/circuit/environment/src/lib.rs
index b47140bb9..eb6ac980c 100644
--- a/circuit/environment/src/lib.rs
+++ b/circuit/environment/src/lib.rs
@@ -19,6 +19,9 @@ extern crate snarkvm_circuit_environment_witness;
 
 pub use snarkvm_circuit_environment_witness::rename_selfs;
 
+pub mod canary_circuit;
+pub use canary_circuit::*;
+
 pub mod circuit;
 pub use circuit::*;
 
diff --git a/circuit/network/src/canary_v0.rs b/circuit/network/src/canary_v0.rs
new file mode 100644
index 000000000..94238bf55
--- /dev/null
+++ b/circuit/network/src/canary_v0.rs
@@ -0,0 +1,598 @@
+// Copyright (C) 2019-2023 Aleo Systems Inc.
+// This file is part of the snarkVM library.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+// http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::Aleo;
+use snarkvm_circuit_algorithms::{
+    Commit,
+    CommitUncompressed,
+    Hash,
+    HashMany,
+    HashToGroup,
+    HashToScalar,
+    HashUncompressed,
+    Keccak256,
+    Keccak384,
+    Keccak512,
+    Pedersen128,
+    Pedersen64,
+    Poseidon2,
+    Poseidon4,
+    Poseidon8,
+    Sha3_256,
+    Sha3_384,
+    Sha3_512,
+    BHP1024,
+    BHP256,
+    BHP512,
+    BHP768,
+};
+use snarkvm_circuit_collections::merkle_tree::MerklePath;
+use snarkvm_circuit_types::{
+    environment::{prelude::*, Assignment, CanaryCircuit, R1CS},
+    Boolean,
+    Field,
+    Group,
+    Scalar,
+};
+
+use core::fmt;
+
+type E = CanaryCircuit;
+
+thread_local! {
+    /// The group bases for the Aleo signature and encryption schemes.
+    static GENERATOR_G: Vec<Group<AleoCanaryV0>> = Vec::constant(<console::CanaryV0 as console::Network>::g_powers().to_vec());
+
+    /// The encryption domain as a constant field element.
+    static ENCRYPTION_DOMAIN: Field<AleoCanaryV0> = Field::constant(<console::CanaryV0 as console::Network>::encryption_domain());
+    /// The graph key domain as a constant field element.
+    static GRAPH_KEY_DOMAIN: Field<AleoCanaryV0> = Field::constant(<console::CanaryV0 as console::Network>::graph_key_domain());
+    /// The serial number domain as a constant field element.
+    static SERIAL_NUMBER_DOMAIN: Field<AleoCanaryV0> = Field::constant(<console::CanaryV0 as console::Network>::serial_number_domain());
+
+    /// The BHP hash function, which can take an input of up to 256 bits.
+    static BHP_256: BHP256<AleoCanaryV0> = BHP256::<AleoCanaryV0>::constant(console::CANARY_BHP_256.clone());
+    /// The BHP hash function, which can take an input of up to 512 bits.
+    static BHP_512: BHP512<AleoCanaryV0> = BHP512::<AleoCanaryV0>::constant(console::CANARY_BHP_512.clone());
+    /// The BHP hash function, which can take an input of up to 768 bits.
+    static BHP_768: BHP768<AleoCanaryV0> = BHP768::<AleoCanaryV0>::constant(console::CANARY_BHP_768.clone());
+    /// The BHP hash function, which can take an input of up to 1024 bits.
+    static BHP_1024: BHP1024<AleoCanaryV0> = BHP1024::<AleoCanaryV0>::constant(console::CANARY_BHP_1024.clone());
+
+    /// The Keccak hash function, which outputs 256 bits.
+    static KECCAK_256: Keccak256<AleoCanaryV0> = Keccak256::<AleoCanaryV0>::new();
+    /// The Keccak hash function, which outputs 384 bits.
+    static KECCAK_384: Keccak384<AleoCanaryV0> = Keccak384::<AleoCanaryV0>::new();
+    /// The Keccak hash function, which outputs 512 bits.
+    static KECCAK_512: Keccak512<AleoCanaryV0> = Keccak512::<AleoCanaryV0>::new();
+
+    /// The Pedersen hash function, which can take an input of up to 64 bits.
+    static PEDERSEN_64: Pedersen64<AleoCanaryV0> = Pedersen64::<AleoCanaryV0>::constant(console::CANARY_PEDERSEN_64.clone());
+    /// The Pedersen hash function, which can take an input of up to 128 bits.
+    static PEDERSEN_128: Pedersen128<AleoCanaryV0> = Pedersen128::<AleoCanaryV0>::constant(console::CANARY_PEDERSEN_128.clone());
+
+    /// The Poseidon hash function, using a rate of 2.
+    static POSEIDON_2: Poseidon2<AleoCanaryV0> = Poseidon2::<AleoCanaryV0>::constant(console::CANARY_POSEIDON_2.clone());
+    /// The Poseidon hash function, using a rate of 4.
+    static POSEIDON_4: Poseidon4<AleoCanaryV0> = Poseidon4::<AleoCanaryV0>::constant(console::CANARY_POSEIDON_4.clone());
+    /// The Poseidon hash function, using a rate of 8.
+    static POSEIDON_8: Poseidon8<AleoCanaryV0> = Poseidon8::<AleoCanaryV0>::constant(console::CANARY_POSEIDON_8.clone());
+
+    /// The SHA-3 hash function, which outputs 256 bits.
+    static SHA3_256: Sha3_256<AleoCanaryV0> = Sha3_256::<AleoCanaryV0>::new();
+    /// The SHA-3 hash function, which outputs 384 bits.
+    static SHA3_384: Sha3_384<AleoCanaryV0> = Sha3_384::<AleoCanaryV0>::new();
+    /// The SHA-3 hash function, which outputs 512 bits.
+    static SHA3_512: Sha3_512<AleoCanaryV0> = Sha3_512::<AleoCanaryV0>::new();
+}
+
+#[derive(Copy, Clone, Debug, Eq, PartialEq, Hash)]
+pub struct AleoCanaryV0;
+
+impl Aleo for AleoCanaryV0 {
+    /// Initializes the global constants for the Aleo environment.
+    fn initialize_global_constants() {
+        GENERATOR_G.with(|_| ());
+        ENCRYPTION_DOMAIN.with(|_| ());
+        GRAPH_KEY_DOMAIN.with(|_| ());
+        SERIAL_NUMBER_DOMAIN.with(|_| ());
+        BHP_256.with(|_| ());
+        BHP_512.with(|_| ());
+        BHP_768.with(|_| ());
+        BHP_1024.with(|_| ());
+        KECCAK_256.with(|_| ());
+        KECCAK_384.with(|_| ());
+        KECCAK_512.with(|_| ());
+        PEDERSEN_64.with(|_| ());
+        PEDERSEN_128.with(|_| ());
+        POSEIDON_2.with(|_| ());
+        POSEIDON_4.with(|_| ());
+        POSEIDON_8.with(|_| ());
+        SHA3_256.with(|_| ());
+        SHA3_384.with(|_| ());
+        SHA3_512.with(|_| ());
+    }
+
+    /// Returns the encryption domain as a constant field element.
+    fn encryption_domain() -> Field<Self> {
+        ENCRYPTION_DOMAIN.with(|domain| domain.clone())
+    }
+
+    /// Returns the graph key domain as a constant field element.
+    fn graph_key_domain() -> Field<Self> {
+        GRAPH_KEY_DOMAIN.with(|domain| domain.clone())
+    }
+
+    /// Returns the serial number domain as a constant field element.
+    fn serial_number_domain() -> Field<Self> {
+        SERIAL_NUMBER_DOMAIN.with(|domain| domain.clone())
+    }
+
+    /// Returns the scalar multiplication on the generator `G`.
+    #[inline]
+    fn g_scalar_multiply(scalar: &Scalar<Self>) -> Group<Self> {
+        GENERATOR_G.with(|bases| {
+            bases
+                .iter()
+                .zip_eq(&scalar.to_bits_le())
+                .fold(Group::zero(), |output, (base, bit)| Group::ternary(bit, &(&output + base), &output))
+        })
+    }
+
+    /// Returns a BHP commitment with an input hasher of 256-bits.
+    fn commit_bhp256(input: &[Boolean<Self>], randomizer: &Scalar<Self>) -> Field<Self> {
+        BHP_256.with(|bhp| bhp.commit(input, randomizer))
+    }
+
+    /// Returns a BHP commitment with an input hasher of 512-bits.
+    fn commit_bhp512(input: &[Boolean<Self>], randomizer: &Scalar<Self>) -> Field<Self> {
+        BHP_512.with(|bhp| bhp.commit(input, randomizer))
+    }
+
+    /// Returns a BHP commitment with an input hasher of 768-bits.
+    fn commit_bhp768(input: &[Boolean<Self>], randomizer: &Scalar<Self>) -> Field<Self> {
+        BHP_768.with(|bhp| bhp.commit(input, randomizer))
+    }
+
+    /// Returns a BHP commitment with an input hasher of 1024-bits.
+    fn commit_bhp1024(input: &[Boolean<Self>], randomizer: &Scalar<Self>) -> Field<Self> {
+        BHP_1024.with(|bhp| bhp.commit(input, randomizer))
+    }
+
+    /// Returns a Pedersen commitment for the given (up to) 64-bit input and randomizer.
+    fn commit_ped64(input: &[Boolean<Self>], randomizer: &Scalar<Self>) -> Field<Self> {
+        PEDERSEN_64.with(|pedersen| pedersen.commit(input, randomizer))
+    }
+
+    /// Returns a Pedersen commitment for the given (up to) 128-bit input and randomizer.
+    fn commit_ped128(input: &[Boolean<Self>], randomizer: &Scalar<Self>) -> Field<Self> {
+        PEDERSEN_128.with(|pedersen| pedersen.commit(input, randomizer))
+    }
+
+    /// Returns a BHP commitment with an input hasher of 256-bits.
+    fn commit_to_group_bhp256(input: &[Boolean<Self>], randomizer: &Scalar<Self>) -> Group<Self> {
+        BHP_256.with(|bhp| bhp.commit_uncompressed(input, randomizer))
+    }
+
+    /// Returns a BHP commitment with an input hasher of 512-bits.
+    fn commit_to_group_bhp512(input: &[Boolean<Self>], randomizer: &Scalar<Self>) -> Group<Self> {
+        BHP_512.with(|bhp| bhp.commit_uncompressed(input, randomizer))
+    }
+
+    /// Returns a BHP commitment with an input hasher of 768-bits.
+    fn commit_to_group_bhp768(input: &[Boolean<Self>], randomizer: &Scalar<Self>) -> Group<Self> {
+        BHP_768.with(|bhp| bhp.commit_uncompressed(input, randomizer))
+    }
+
+    /// Returns a BHP commitment with an input hasher of 1024-bits.
+    fn commit_to_group_bhp1024(input: &[Boolean<Self>], randomizer: &Scalar<Self>) -> Group<Self> {
+        BHP_1024.with(|bhp| bhp.commit_uncompressed(input, randomizer))
+    }
+
+    /// Returns a Pedersen commitment for the given (up to) 64-bit input and randomizer.
+    fn commit_to_group_ped64(input: &[Boolean<Self>], randomizer: &Scalar<Self>) -> Group<Self> {
+        PEDERSEN_64.with(|pedersen| pedersen.commit_uncompressed(input, randomizer))
+    }
+
+    /// Returns a Pedersen commitment for the given (up to) 128-bit input and randomizer.
+    fn commit_to_group_ped128(input: &[Boolean<Self>], randomizer: &Scalar<Self>) -> Group<Self> {
+        PEDERSEN_128.with(|pedersen| pedersen.commit_uncompressed(input, randomizer))
+    }
+
+    /// Returns the BHP hash with an input hasher of 256-bits.
+    fn hash_bhp256(input: &[Boolean<Self>]) -> Field<Self> {
+        BHP_256.with(|bhp| bhp.hash(input))
+    }
+
+    /// Returns the BHP hash with an input hasher of 512-bits.
+    fn hash_bhp512(input: &[Boolean<Self>]) -> Field<Self> {
+        BHP_512.with(|bhp| bhp.hash(input))
+    }
+
+    /// Returns the BHP hash with an input hasher of 768-bits.
+    fn hash_bhp768(input: &[Boolean<Self>]) -> Field<Self> {
+        BHP_768.with(|bhp| bhp.hash(input))
+    }
+
+    /// Returns the BHP hash with an input hasher of 1024-bits.
+    fn hash_bhp1024(input: &[Boolean<Self>]) -> Field<Self> {
+        BHP_1024.with(|bhp| bhp.hash(input))
+    }
+
+    /// Returns the Keccak hash with a 256-bit output.
+    fn hash_keccak256(input: &[Boolean<Self>]) -> Vec<Boolean<Self>> {
+        KECCAK_256.with(|keccak| keccak.hash(input))
+    }
+
+    /// Returns the Keccak hash with a 384-bit output.
+    fn hash_keccak384(input: &[Boolean<Self>]) -> Vec<Boolean<Self>> {
+        KECCAK_384.with(|keccak| keccak.hash(input))
+    }
+
+    /// Returns the Keccak hash with a 512-bit output.
+    fn hash_keccak512(input: &[Boolean<Self>]) -> Vec<Boolean<Self>> {
+        KECCAK_512.with(|keccak| keccak.hash(input))
+    }
+
+    /// Returns the Pedersen hash for a given (up to) 64-bit input.
+    fn hash_ped64(input: &[Boolean<Self>]) -> Field<Self> {
+        PEDERSEN_64.with(|pedersen| pedersen.hash(input))
+    }
+
+    /// Returns the Pedersen hash for a given (up to) 128-bit input.
+    fn hash_ped128(input: &[Boolean<Self>]) -> Field<Self> {
+        PEDERSEN_128.with(|pedersen| pedersen.hash(input))
+    }
+
+    /// Returns the Poseidon hash with an input rate of 2.
+    fn hash_psd2(input: &[Field<Self>]) -> Field<Self> {
+        POSEIDON_2.with(|poseidon| poseidon.hash(input))
+    }
+
+    /// Returns the Poseidon hash with an input rate of 4.
+    fn hash_psd4(input: &[Field<Self>]) -> Field<Self> {
+        POSEIDON_4.with(|poseidon| poseidon.hash(input))
+    }
+
+    /// Returns the Poseidon hash with an input rate of 8.
+    fn hash_psd8(input: &[Field<Self>]) -> Field<Self> {
+        POSEIDON_8.with(|poseidon| poseidon.hash(input))
+    }
+
+    /// Returns the SHA-3 hash with a 256-bit output.
+    fn hash_sha3_256(input: &[Boolean<Self>]) -> Vec<Boolean<Self>> {
+        SHA3_256.with(|sha3| sha3.hash(input))
+    }
+
+    /// Returns the SHA-3 hash with a 384-bit output.
+    fn hash_sha3_384(input: &[Boolean<Self>]) -> Vec<Boolean<Self>> {
+        SHA3_384.with(|sha3| sha3.hash(input))
+    }
+
+    /// Returns the SHA-3 hash with a 512-bit output.
+    fn hash_sha3_512(input: &[Boolean<Self>]) -> Vec<Boolean<Self>> {
+        SHA3_512.with(|sha3| sha3.hash(input))
+    }
+
+    /// Returns the extended Poseidon hash with an input rate of 2.
+    fn hash_many_psd2(input: &[Field<Self>], num_outputs: u16) -> Vec<Field<Self>> {
+        POSEIDON_2.with(|poseidon| poseidon.hash_many(input, num_outputs))
+    }
+
+    /// Returns the extended Poseidon hash with an input rate of 4.
+    fn hash_many_psd4(input: &[Field<Self>], num_outputs: u16) -> Vec<Field<Self>> {
+        POSEIDON_4.with(|poseidon| poseidon.hash_many(input, num_outputs))
+    }
+
+    /// Returns the extended Poseidon hash with an input rate of 8.
+    fn hash_many_psd8(input: &[Field<Self>], num_outputs: u16) -> Vec<Field<Self>> {
+        POSEIDON_8.with(|poseidon| poseidon.hash_many(input, num_outputs))
+    }
+
+    /// Returns the BHP hash with an input hasher of 256-bits.
+    fn hash_to_group_bhp256(input: &[Boolean<Self>]) -> Group<Self> {
+        BHP_256.with(|bhp| bhp.hash_uncompressed(input))
+    }
+
+    /// Returns the BHP hash with an input hasher of 512-bits.
+    fn hash_to_group_bhp512(input: &[Boolean<Self>]) -> Group<Self> {
+        BHP_512.with(|bhp| bhp.hash_uncompressed(input))
+    }
+
+    /// Returns the BHP hash with an input hasher of 768-bits.
+    fn hash_to_group_bhp768(input: &[Boolean<Self>]) -> Group<Self> {
+        BHP_768.with(|bhp| bhp.hash_uncompressed(input))
+    }
+
+    /// Returns the BHP hash with an input hasher of 1024-bits.
+    fn hash_to_group_bhp1024(input: &[Boolean<Self>]) -> Group<Self> {
+        BHP_1024.with(|bhp| bhp.hash_uncompressed(input))
+    }
+
+    /// Returns the Pedersen hash for a given (up to) 64-bit input.
+    fn hash_to_group_ped64(input: &[Boolean<Self>]) -> Group<Self> {
+        PEDERSEN_64.with(|pedersen| pedersen.hash_uncompressed(input))
+    }
+
+    /// Returns the Pedersen hash for a given (up to) 128-bit input.
+    fn hash_to_group_ped128(input: &[Boolean<Self>]) -> Group<Self> {
+        PEDERSEN_128.with(|pedersen| pedersen.hash_uncompressed(input))
+    }
+
+    /// Returns the Poseidon hash with an input rate of 2 on the affine curve.
+    fn hash_to_group_psd2(input: &[Field<Self>]) -> Group<Self> {
+        POSEIDON_2.with(|poseidon| poseidon.hash_to_group(input))
+    }
+
+    /// Returns the Poseidon hash with an input rate of 4 on the affine curve.
+    fn hash_to_group_psd4(input: &[Field<Self>]) -> Group<Self> {
+        POSEIDON_4.with(|poseidon| poseidon.hash_to_group(input))
+    }
+
+    /// Returns the Poseidon hash with an input rate of 8 on the affine curve.
+    fn hash_to_group_psd8(input: &[Field<Self>]) -> Group<Self> {
+        POSEIDON_8.with(|poseidon| poseidon.hash_to_group(input))
+    }
+
+    /// Returns the Poseidon hash with an input rate of 2 on the scalar field.
+    fn hash_to_scalar_psd2(input: &[Field<Self>]) -> Scalar<Self> {
+        POSEIDON_2.with(|poseidon| poseidon.hash_to_scalar(input))
+    }
+
+    /// Returns the Poseidon hash with an input rate of 4 on the scalar field.
+    fn hash_to_scalar_psd4(input: &[Field<Self>]) -> Scalar<Self> {
+        POSEIDON_4.with(|poseidon| poseidon.hash_to_scalar(input))
+    }
+
+    /// Returns the Poseidon hash with an input rate of 8 on the scalar field.
+    fn hash_to_scalar_psd8(input: &[Field<Self>]) -> Scalar<Self> {
+        POSEIDON_8.with(|poseidon| poseidon.hash_to_scalar(input))
+    }
+
+    /// Returns `true` if the given Merkle path is valid for the given root and leaf.
+    fn verify_merkle_path_bhp<const DEPTH: u8>(
+        path: &MerklePath<Self, DEPTH>,
+        root: &Field<Self>,
+        leaf: &Vec<Boolean<Self>>,
+    ) -> Boolean<Self> {
+        BHP_1024.with(|bhp1024| BHP_512.with(|bhp512| path.verify(bhp1024, bhp512, root, leaf)))
+    }
+
+    /// Returns `true` if the given Merkle path is valid for the given root and leaf.
+    fn verify_merkle_path_psd<const DEPTH: u8>(
+        path: &MerklePath<Self, DEPTH>,
+        root: &Field<Self>,
+        leaf: &Vec<Field<Self>>,
+    ) -> Boolean<Self> {
+        POSEIDON_4.with(|psd4| POSEIDON_2.with(|psd2| path.verify(psd4, psd2, root, leaf)))
+    }
+}
+
+impl Environment for AleoCanaryV0 {
+    type Affine = <E as Environment>::Affine;
+    type BaseField = <E as Environment>::BaseField;
+    type Network = <E as Environment>::Network;
+    type ScalarField = <E as Environment>::ScalarField;
+
+    /// Returns the `zero` constant.
+    fn zero() -> LinearCombination<Self::BaseField> {
+        E::zero()
+    }
+
+    /// Returns the `one` constant.
+    fn one() -> LinearCombination<Self::BaseField> {
+        E::one()
+    }
+
+    /// Returns a new variable of the given mode and value.
+    fn new_variable(mode: Mode, value: Self::BaseField) -> Variable<Self::BaseField> {
+        E::new_variable(mode, value)
+    }
+
+    /// Returns a new witness of the given mode and value.
+    fn new_witness<Fn: FnOnce() -> Output::Primitive, Output: Inject>(mode: Mode, logic: Fn) -> Output {
+        E::new_witness(mode, logic)
+    }
+
+    /// Enters a new scope for the environment.
+    fn scope<S: Into<String>, Fn, Output>(name: S, logic: Fn) -> Output
+    where
+        Fn: FnOnce() -> Output,
+    {
+        E::scope(name, logic)
+    }
+
+    /// Adds one constraint enforcing that `(A * B) == C`.
+    fn enforce<Fn, A, B, C>(constraint: Fn)
+    where
+        Fn: FnOnce() -> (A, B, C),
+        A: Into<LinearCombination<Self::BaseField>>,
+        B: Into<LinearCombination<Self::BaseField>>,
+        C: Into<LinearCombination<Self::BaseField>>,
+    {
+        E::enforce(constraint)
+    }
+
+    /// Returns `true` if all constraints in the environment are satisfied.
+    fn is_satisfied() -> bool {
+        E::is_satisfied()
+    }
+
+    /// Returns `true` if all constraints in the current scope are satisfied.
+    fn is_satisfied_in_scope() -> bool {
+        E::is_satisfied_in_scope()
+    }
+
+    /// Returns the number of constants in the entire circuit.
+    fn num_constants() -> u64 {
+        E::num_constants()
+    }
+
+    /// Returns the number of public variables in the entire circuit.
+    fn num_public() -> u64 {
+        E::num_public()
+    }
+
+    /// Returns the number of private variables in the entire circuit.
+    fn num_private() -> u64 {
+        E::num_private()
+    }
+
+    /// Returns the number of constant, public, and private variables in the entire circuit.
+    fn num_variables() -> u64 {
+        E::num_variables()
+    }
+
+    /// Returns the number of constraints in the entire circuit.
+    fn num_constraints() -> u64 {
+        E::num_constraints()
+    }
+
+    /// Returns the number of nonzeros in the entire circuit.
+    fn num_nonzeros() -> (u64, u64, u64) {
+        E::num_nonzeros()
+    }
+
+    /// Returns the number of constants for the current scope.
+    fn num_constants_in_scope() -> u64 {
+        E::num_constants_in_scope()
+    }
+
+    /// Returns the number of public variables for the current scope.
+    fn num_public_in_scope() -> u64 {
+        E::num_public_in_scope()
+    }
+
+    /// Returns the number of private variables for the current scope.
+    fn num_private_in_scope() -> u64 {
+        E::num_private_in_scope()
+    }
+
+    /// Returns the number of constraints for the current scope.
+    fn num_constraints_in_scope() -> u64 {
+        E::num_constraints_in_scope()
+    }
+
+    /// Returns the number of nonzeros for the current scope.
+    fn num_nonzeros_in_scope() -> (u64, u64, u64) {
+        E::num_nonzeros_in_scope()
+    }
+
+    /// Returns the variable limit for the circuit, if one exists.
+    fn get_variable_limit() -> Option<u64> {
+        E::get_variable_limit()
+    }
+
+    /// Sets the variable limit for the circuit.
+    fn set_variable_limit(limit: Option<u64>) {
+        E::set_variable_limit(limit)
+    }
+
+    /// Returns the constraint limit for the circuit, if one exists.
+    fn get_constraint_limit() -> Option<u64> {
+        E::get_constraint_limit()
+    }
+
+    /// Sets the constraint limit for the circuit.
+    fn set_constraint_limit(limit: Option<u64>) {
+        E::set_constraint_limit(limit)
+    }
+
+    /// Halts the program from further synthesis, evaluation, and execution in the current environment.
+    fn halt<S: Into<String>, T>(message: S) -> T {
+        E::halt(message)
+    }
+
+    /// Returns the R1CS circuit, resetting the circuit.
+    fn inject_r1cs(r1cs: R1CS<Self::BaseField>) {
+        E::inject_r1cs(r1cs)
+    }
+
+    /// Returns the R1CS circuit, resetting the circuit.
+    fn eject_r1cs_and_reset() -> R1CS<Self::BaseField> {
+        E::eject_r1cs_and_reset()
+    }
+
+    /// Returns the R1CS assignment of the circuit, resetting the circuit.
+    fn eject_assignment_and_reset() -> Assignment<<Self::Network as console::Environment>::Field> {
+        E::eject_assignment_and_reset()
+    }
+
+    /// Clears the circuit and initializes an empty environment.
+    fn reset() {
+        E::reset()
+    }
+}
+
+impl Display for AleoCanaryV0 {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        // TODO (howardwu): Find a better way to print the circuit.
+        fmt::Display::fmt(&CanaryCircuit, f)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use snarkvm_circuit_types::Field;
+
+    type CurrentAleo = AleoCanaryV0;
+
+    /// Compute 2^EXPONENT - 1, in a purposefully constraint-inefficient manner for testing.
+    fn create_example_circuit<E: Environment>() -> Field<E> {
+        let one = snarkvm_console_types::Field::<<E as Environment>::Network>::one();
+        let two = one + one;
+
+        const EXPONENT: u64 = 64;
+
+        // Compute 2^EXPONENT - 1, in a purposefully constraint-inefficient manner for testing.
+        let mut candidate = Field::<E>::new(Mode::Public, one);
+        let mut accumulator = Field::new(Mode::Private, two);
+        for _ in 0..EXPONENT {
+            candidate += &accumulator;
+            accumulator *= Field::new(Mode::Private, two);
+        }
+
+        assert_eq!((accumulator - Field::one()).eject_value(), candidate.eject_value());
+        assert_eq!(2, E::num_public());
+        assert_eq!(2 * EXPONENT + 1, E::num_private());
+        assert_eq!(EXPONENT, E::num_constraints());
+        assert!(E::is_satisfied());
+
+        candidate
+    }
+
+    #[test]
+    fn test_print_circuit() {
+        let circuit = CurrentAleo {};
+        let _candidate = create_example_circuit::<CurrentAleo>();
+        let output = format!("{circuit}");
+        println!("{output}");
+    }
+
+    #[test]
+    fn test_circuit_scope() {
+        CurrentAleo::scope("test_circuit_scope", || {
+            assert_eq!(0, CurrentAleo::num_constants());
+            assert_eq!(1, CurrentAleo::num_public());
+            assert_eq!(0, CurrentAleo::num_private());
+            assert_eq!(0, CurrentAleo::num_constraints());
+
+            assert_eq!(0, CurrentAleo::num_constants_in_scope());
+            assert_eq!(0, CurrentAleo::num_public_in_scope());
+            assert_eq!(0, CurrentAleo::num_private_in_scope());
+            assert_eq!(0, CurrentAleo::num_constraints_in_scope());
+        })
+    }
+}
diff --git a/circuit/network/src/lib.rs b/circuit/network/src/lib.rs
index 36c131a70..464e4e7ea 100644
--- a/circuit/network/src/lib.rs
+++ b/circuit/network/src/lib.rs
@@ -15,6 +15,9 @@
 #![forbid(unsafe_code)]
 #![allow(clippy::too_many_arguments)]
 
+pub mod canary_v0;
+pub use canary_v0::*;
+
 pub mod testnet_v0;
 pub use testnet_v0::*;
 
diff --git a/circuit/src/lib.rs b/circuit/src/lib.rs
index 78af7cfe6..48991bc63 100644
--- a/circuit/src/lib.rs
+++ b/circuit/src/lib.rs
@@ -27,7 +27,16 @@ pub mod modules {
     pub use snarkvm_circuit_collections::*;
 
     pub use snarkvm_circuit_environment as environment;
-    pub use snarkvm_circuit_environment::{Assignment, Circuit, Eject, Environment, Inject, Mode, TestnetCircuit};
+    pub use snarkvm_circuit_environment::{
+        Assignment,
+        CanaryCircuit,
+        Circuit,
+        Eject,
+        Environment,
+        Inject,
+        Mode,
+        TestnetCircuit,
+    };
 
     pub use snarkvm_circuit_network as network;
     pub use snarkvm_circuit_network::*;
diff --git a/console/network/src/canary_v0.rs b/console/network/src/canary_v0.rs
new file mode 100644
index 000000000..9e8d920f0
--- /dev/null
+++ b/console/network/src/canary_v0.rs
@@ -0,0 +1,492 @@
+// Copyright (C) 2019-2023 Aleo Systems Inc.
+// This file is part of the snarkVM library.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+// http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use super::*;
+use crate::TRANSACTION_PREFIX;
+use snarkvm_console_algorithms::{
+    Blake2Xs,
+    Keccak256,
+    Keccak384,
+    Keccak512,
+    Pedersen128,
+    Pedersen64,
+    Poseidon2,
+    Poseidon4,
+    Poseidon8,
+    Sha3_256,
+    Sha3_384,
+    Sha3_512,
+    BHP1024,
+    BHP256,
+    BHP512,
+    BHP768,
+};
+
+lazy_static! {
+    /// The group bases for the Aleo signature and encryption schemes.
+    static ref GENERATOR_G: Vec<Group<CanaryV0 >> = CanaryV0::new_bases("AleoAccountEncryptionAndSignatureScheme0");
+
+    /// The Varuna sponge parameters.
+    static ref VARUNA_FS_PARAMETERS: FiatShamirParameters<CanaryV0> = FiatShamir::<CanaryV0>::sample_parameters();
+
+    /// The encryption domain as a constant field element.
+    static ref ENCRYPTION_DOMAIN: Field<CanaryV0> = Field::<CanaryV0>::new_domain_separator("AleoSymmetricEncryption0");
+    /// The graph key domain as a constant field element.
+    static ref GRAPH_KEY_DOMAIN: Field<CanaryV0> = Field::<CanaryV0>::new_domain_separator("AleoGraphKey0");
+    /// The serial number domain as a constant field element.
+    static ref SERIAL_NUMBER_DOMAIN: Field<CanaryV0> = Field::<CanaryV0>::new_domain_separator("AleoSerialNumber0");
+
+    /// The BHP hash function, which can take an input of up to 256 bits.
+    pub static ref CANARY_BHP_256: BHP256<CanaryV0> = BHP256::<CanaryV0>::setup("AleoBHP256").expect("Failed to setup BHP256");
+    /// The BHP hash function, which can take an input of up to 512 bits.
+    pub static ref CANARY_BHP_512: BHP512<CanaryV0> = BHP512::<CanaryV0>::setup("AleoBHP512").expect("Failed to setup BHP512");
+    /// The BHP hash function, which can take an input of up to 768 bits.
+    pub static ref CANARY_BHP_768: BHP768<CanaryV0> = BHP768::<CanaryV0>::setup("AleoBHP768").expect("Failed to setup BHP768");
+    /// The BHP hash function, which can take an input of up to 1024 bits.
+    pub static ref CANARY_BHP_1024: BHP1024<CanaryV0> = BHP1024::<CanaryV0>::setup("AleoBHP1024").expect("Failed to setup BHP1024");
+
+    /// The Pedersen hash function, which can take an input of up to 64 bits.
+    pub static ref CANARY_PEDERSEN_64: Pedersen64<CanaryV0> = Pedersen64::<CanaryV0>::setup("AleoPedersen64");
+    /// The Pedersen hash function, which can take an input of up to 128 bits.
+    pub static ref CANARY_PEDERSEN_128: Pedersen128<CanaryV0> = Pedersen128::<CanaryV0>::setup("AleoPedersen128");
+
+    /// The Poseidon hash function, using a rate of 2.
+    pub static ref CANARY_POSEIDON_2: Poseidon2<CanaryV0> = Poseidon2::<CanaryV0>::setup("AleoPoseidon2").expect("Failed to setup Poseidon2");
+    /// The Poseidon hash function, using a rate of 4.
+    pub static ref CANARY_POSEIDON_4: Poseidon4<CanaryV0> = Poseidon4::<CanaryV0>::setup("AleoPoseidon4").expect("Failed to setup Poseidon4");
+    /// The Poseidon hash function, using a rate of 8.
+    pub static ref CANARY_POSEIDON_8: Poseidon8<CanaryV0> = Poseidon8::<CanaryV0>::setup("AleoPoseidon8").expect("Failed to setup Poseidon8");
+
+    pub static ref CANARY_CREDITS_PROVING_KEYS: IndexMap<String, Arc<VarunaProvingKey<Console>>> = {
+        let mut map = IndexMap::new();
+        snarkvm_parameters::insert_canary_credit_keys!(map, VarunaProvingKey<Console>, Prover);
+        map
+    };
+    pub static ref CANARY_CREDITS_VERIFYING_KEYS: IndexMap<String, Arc<VarunaVerifyingKey<Console>>> = {
+        let mut map = IndexMap::new();
+        snarkvm_parameters::insert_canary_credit_keys!(map, VarunaVerifyingKey<Console>, Verifier);
+        map
+    };
+}
+
+#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash, Serialize, Deserialize)]
+pub struct CanaryV0;
+
+impl CanaryV0 {
+    /// Initializes a new instance of group bases from a given input domain message.
+    fn new_bases(message: &str) -> Vec<Group<Self>> {
+        // Hash the given message to a point on the curve, to initialize the starting base.
+        let (base, _, _) = Blake2Xs::hash_to_curve::<<Self as Environment>::Affine>(message);
+
+        // Compute the bases up to the size of the scalar field (in bits).
+        let mut g = Group::<Self>::new(base);
+        let mut g_bases = Vec::with_capacity(Scalar::<Self>::size_in_bits());
+        for _ in 0..Scalar::<Self>::size_in_bits() {
+            g_bases.push(g);
+            g = g.double();
+        }
+        g_bases
+    }
+}
+
+impl Environment for CanaryV0 {
+    type Affine = <Console as Environment>::Affine;
+    type BigInteger = <Console as Environment>::BigInteger;
+    type Field = <Console as Environment>::Field;
+    type PairingCurve = <Console as Environment>::PairingCurve;
+    type Projective = <Console as Environment>::Projective;
+    type Scalar = <Console as Environment>::Scalar;
+
+    /// The coefficient `A` of the twisted Edwards curve.
+    const EDWARDS_A: Self::Field = Console::EDWARDS_A;
+    /// The coefficient `D` of the twisted Edwards curve.
+    const EDWARDS_D: Self::Field = Console::EDWARDS_D;
+    /// The coefficient `A` of the Montgomery curve.
+    const MONTGOMERY_A: Self::Field = Console::MONTGOMERY_A;
+    /// The coefficient `B` of the Montgomery curve.
+    const MONTGOMERY_B: Self::Field = Console::MONTGOMERY_B;
+}
+
+impl Network for CanaryV0 {
+    /// The block hash type.
+    type BlockHash = AleoID<Field<Self>, { hrp2!("ab") }>;
+    /// The ratification ID type.
+    type RatificationID = AleoID<Field<Self>, { hrp2!("ar") }>;
+    /// The state root type.
+    type StateRoot = AleoID<Field<Self>, { hrp2!("sr") }>;
+    /// The transaction ID type.
+    type TransactionID = AleoID<Field<Self>, { hrp2!(TRANSACTION_PREFIX) }>;
+    /// The transition ID type.
+    type TransitionID = AleoID<Field<Self>, { hrp2!("au") }>;
+
+    /// The network edition.
+    const EDITION: u16 = 0;
+    /// The genesis block coinbase target.
+    const GENESIS_COINBASE_TARGET: u64 = (1u64 << 10).saturating_sub(1);
+    /// The genesis block proof target.
+    const GENESIS_PROOF_TARGET: u64 = 1u64 << 8;
+    /// The fixed timestamp of the genesis block.
+    const GENESIS_TIMESTAMP: i64 = 1715776496 /* 2024-05-15 12:34:56 UTC */;
+    /// The network ID.
+    const ID: u16 = 2;
+    /// The function name for the inclusion circuit.
+    const INCLUSION_FUNCTION_NAME: &'static str = MainnetV0::INCLUSION_FUNCTION_NAME;
+    /// The maximum number of certificates in a batch.
+    const MAX_CERTIFICATES: u16 = 100;
+    /// The network name.
+    const NAME: &'static str = "Aleo Canary (v0)";
+
+    /// Returns the genesis block bytes.
+    fn genesis_bytes() -> &'static [u8] {
+        snarkvm_parameters::canary::GenesisBytes::load_bytes()
+    }
+
+    /// Returns the proving key for the given function name in `credits.aleo`.
+    fn get_credits_proving_key(function_name: String) -> Result<&'static Arc<VarunaProvingKey<Self>>> {
+        CANARY_CREDITS_PROVING_KEYS
+            .get(&function_name)
+            .ok_or_else(|| anyhow!("Proving key for credits.aleo/{function_name}' not found"))
+    }
+
+    /// Returns the verifying key for the given function name in `credits.aleo`.
+    fn get_credits_verifying_key(function_name: String) -> Result<&'static Arc<VarunaVerifyingKey<Self>>> {
+        CANARY_CREDITS_VERIFYING_KEYS
+            .get(&function_name)
+            .ok_or_else(|| anyhow!("Verifying key for credits.aleo/{function_name}' not found"))
+    }
+
+    /// Returns the `proving key` for the inclusion circuit.
+    fn inclusion_proving_key() -> &'static Arc<VarunaProvingKey<Self>> {
+        static INSTANCE: OnceCell<Arc<VarunaProvingKey<Console>>> = OnceCell::new();
+        INSTANCE.get_or_init(|| {
+            // Skipping the first byte, which is the encoded version.
+            Arc::new(
+                CircuitProvingKey::from_bytes_le(&snarkvm_parameters::canary::INCLUSION_PROVING_KEY[1..])
+                    .expect("Failed to load inclusion proving key."),
+            )
+        })
+    }
+
+    /// Returns the `verifying key` for the inclusion circuit.
+    fn inclusion_verifying_key() -> &'static Arc<VarunaVerifyingKey<Self>> {
+        static INSTANCE: OnceCell<Arc<VarunaVerifyingKey<Console>>> = OnceCell::new();
+        INSTANCE.get_or_init(|| {
+            // Skipping the first byte, which is the encoded version.
+            Arc::new(
+                CircuitVerifyingKey::from_bytes_le(&snarkvm_parameters::canary::INCLUSION_VERIFYING_KEY[1..])
+                    .expect("Failed to load inclusion verifying key."),
+            )
+        })
+    }
+
+    /// Returns the powers of `G`.
+    fn g_powers() -> &'static Vec<Group<Self>> {
+        &GENERATOR_G
+    }
+
+    /// Returns the scalar multiplication on the generator `G`.
+    fn g_scalar_multiply(scalar: &Scalar<Self>) -> Group<Self> {
+        GENERATOR_G
+            .iter()
+            .zip_eq(&scalar.to_bits_le())
+            .filter_map(|(base, bit)| match bit {
+                true => Some(base),
+                false => None,
+            })
+            .sum()
+    }
+
+    /// Returns the Varuna universal prover.
+    fn varuna_universal_prover() -> &'static UniversalProver<Self::PairingCurve> {
+        MainnetV0::varuna_universal_prover()
+    }
+
+    /// Returns the Varuna universal verifier.
+    fn varuna_universal_verifier() -> &'static UniversalVerifier<Self::PairingCurve> {
+        MainnetV0::varuna_universal_verifier()
+    }
+
+    /// Returns the sponge parameters used for the sponge in the Varuna SNARK.
+    fn varuna_fs_parameters() -> &'static FiatShamirParameters<Self> {
+        &VARUNA_FS_PARAMETERS
+    }
+
+    /// Returns the encryption domain as a constant field element.
+    fn encryption_domain() -> Field<Self> {
+        *ENCRYPTION_DOMAIN
+    }
+
+    /// Returns the graph key domain as a constant field element.
+    fn graph_key_domain() -> Field<Self> {
+        *GRAPH_KEY_DOMAIN
+    }
+
+    /// Returns the serial number domain as a constant field element.
+    fn serial_number_domain() -> Field<Self> {
+        *SERIAL_NUMBER_DOMAIN
+    }
+
+    /// Returns a BHP commitment with an input hasher of 256-bits and randomizer.
+    fn commit_bhp256(input: &[bool], randomizer: &Scalar<Self>) -> Result<Field<Self>> {
+        CANARY_BHP_256.commit(input, randomizer)
+    }
+
+    /// Returns a BHP commitment with an input hasher of 512-bits and randomizer.
+    fn commit_bhp512(input: &[bool], randomizer: &Scalar<Self>) -> Result<Field<Self>> {
+        CANARY_BHP_512.commit(input, randomizer)
+    }
+
+    /// Returns a BHP commitment with an input hasher of 768-bits and randomizer.
+    fn commit_bhp768(input: &[bool], randomizer: &Scalar<Self>) -> Result<Field<Self>> {
+        CANARY_BHP_768.commit(input, randomizer)
+    }
+
+    /// Returns a BHP commitment with an input hasher of 1024-bits and randomizer.
+    fn commit_bhp1024(input: &[bool], randomizer: &Scalar<Self>) -> Result<Field<Self>> {
+        CANARY_BHP_1024.commit(input, randomizer)
+    }
+
+    /// Returns a Pedersen commitment for the given (up to) 64-bit input and randomizer.
+    fn commit_ped64(input: &[bool], randomizer: &Scalar<Self>) -> Result<Field<Self>> {
+        CANARY_PEDERSEN_64.commit(input, randomizer)
+    }
+
+    /// Returns a Pedersen commitment for the given (up to) 128-bit input and randomizer.
+    fn commit_ped128(input: &[bool], randomizer: &Scalar<Self>) -> Result<Field<Self>> {
+        CANARY_PEDERSEN_128.commit(input, randomizer)
+    }
+
+    /// Returns a BHP commitment with an input hasher of 256-bits and randomizer.
+    fn commit_to_group_bhp256(input: &[bool], randomizer: &Scalar<Self>) -> Result<Group<Self>> {
+        CANARY_BHP_256.commit_uncompressed(input, randomizer)
+    }
+
+    /// Returns a BHP commitment with an input hasher of 512-bits and randomizer.
+    fn commit_to_group_bhp512(input: &[bool], randomizer: &Scalar<Self>) -> Result<Group<Self>> {
+        CANARY_BHP_512.commit_uncompressed(input, randomizer)
+    }
+
+    /// Returns a BHP commitment with an input hasher of 768-bits and randomizer.
+    fn commit_to_group_bhp768(input: &[bool], randomizer: &Scalar<Self>) -> Result<Group<Self>> {
+        CANARY_BHP_768.commit_uncompressed(input, randomizer)
+    }
+
+    /// Returns a BHP commitment with an input hasher of 1024-bits and randomizer.
+    fn commit_to_group_bhp1024(input: &[bool], randomizer: &Scalar<Self>) -> Result<Group<Self>> {
+        CANARY_BHP_1024.commit_uncompressed(input, randomizer)
+    }
+
+    /// Returns a Pedersen commitment for the given (up to) 64-bit input and randomizer.
+    fn commit_to_group_ped64(input: &[bool], randomizer: &Scalar<Self>) -> Result<Group<Self>> {
+        CANARY_PEDERSEN_64.commit_uncompressed(input, randomizer)
+    }
+
+    /// Returns a Pedersen commitment for the given (up to) 128-bit input and randomizer.
+    fn commit_to_group_ped128(input: &[bool], randomizer: &Scalar<Self>) -> Result<Group<Self>> {
+        CANARY_PEDERSEN_128.commit_uncompressed(input, randomizer)
+    }
+
+    /// Returns the BHP hash with an input hasher of 256-bits.
+    fn hash_bhp256(input: &[bool]) -> Result<Field<Self>> {
+        CANARY_BHP_256.hash(input)
+    }
+
+    /// Returns the BHP hash with an input hasher of 512-bits.
+    fn hash_bhp512(input: &[bool]) -> Result<Field<Self>> {
+        CANARY_BHP_512.hash(input)
+    }
+
+    /// Returns the BHP hash with an input hasher of 768-bits.
+    fn hash_bhp768(input: &[bool]) -> Result<Field<Self>> {
+        CANARY_BHP_768.hash(input)
+    }
+
+    /// Returns the BHP hash with an input hasher of 1024-bits.
+    fn hash_bhp1024(input: &[bool]) -> Result<Field<Self>> {
+        CANARY_BHP_1024.hash(input)
+    }
+
+    /// Returns the Keccak hash with a 256-bit output.
+    fn hash_keccak256(input: &[bool]) -> Result<Vec<bool>> {
+        Keccak256::default().hash(input)
+    }
+
+    /// Returns the Keccak hash with a 384-bit output.
+    fn hash_keccak384(input: &[bool]) -> Result<Vec<bool>> {
+        Keccak384::default().hash(input)
+    }
+
+    /// Returns the Keccak hash with a 512-bit output.
+    fn hash_keccak512(input: &[bool]) -> Result<Vec<bool>> {
+        Keccak512::default().hash(input)
+    }
+
+    /// Returns the Pedersen hash for a given (up to) 64-bit input.
+    fn hash_ped64(input: &[bool]) -> Result<Field<Self>> {
+        CANARY_PEDERSEN_64.hash(input)
+    }
+
+    /// Returns the Pedersen hash for a given (up to) 128-bit input.
+    fn hash_ped128(input: &[bool]) -> Result<Field<Self>> {
+        CANARY_PEDERSEN_128.hash(input)
+    }
+
+    /// Returns the Poseidon hash with an input rate of 2.
+    fn hash_psd2(input: &[Field<Self>]) -> Result<Field<Self>> {
+        CANARY_POSEIDON_2.hash(input)
+    }
+
+    /// Returns the Poseidon hash with an input rate of 4.
+    fn hash_psd4(input: &[Field<Self>]) -> Result<Field<Self>> {
+        CANARY_POSEIDON_4.hash(input)
+    }
+
+    /// Returns the Poseidon hash with an input rate of 8.
+    fn hash_psd8(input: &[Field<Self>]) -> Result<Field<Self>> {
+        CANARY_POSEIDON_8.hash(input)
+    }
+
+    /// Returns the SHA-3 hash with a 256-bit output.
+    fn hash_sha3_256(input: &[bool]) -> Result<Vec<bool>> {
+        Sha3_256::default().hash(input)
+    }
+
+    /// Returns the SHA-3 hash with a 384-bit output.
+    fn hash_sha3_384(input: &[bool]) -> Result<Vec<bool>> {
+        Sha3_384::default().hash(input)
+    }
+
+    /// Returns the SHA-3 hash with a 512-bit output.
+    fn hash_sha3_512(input: &[bool]) -> Result<Vec<bool>> {
+        Sha3_512::default().hash(input)
+    }
+
+    /// Returns the extended Poseidon hash with an input rate of 2.
+    fn hash_many_psd2(input: &[Field<Self>], num_outputs: u16) -> Vec<Field<Self>> {
+        CANARY_POSEIDON_2.hash_many(input, num_outputs)
+    }
+
+    /// Returns the extended Poseidon hash with an input rate of 4.
+    fn hash_many_psd4(input: &[Field<Self>], num_outputs: u16) -> Vec<Field<Self>> {
+        CANARY_POSEIDON_4.hash_many(input, num_outputs)
+    }
+
+    /// Returns the extended Poseidon hash with an input rate of 8.
+    fn hash_many_psd8(input: &[Field<Self>], num_outputs: u16) -> Vec<Field<Self>> {
+        CANARY_POSEIDON_8.hash_many(input, num_outputs)
+    }
+
+    /// Returns the BHP hash with an input hasher of 256-bits.
+    fn hash_to_group_bhp256(input: &[bool]) -> Result<Group<Self>> {
+        CANARY_BHP_256.hash_uncompressed(input)
+    }
+
+    /// Returns the BHP hash with an input hasher of 512-bits.
+    fn hash_to_group_bhp512(input: &[bool]) -> Result<Group<Self>> {
+        CANARY_BHP_512.hash_uncompressed(input)
+    }
+
+    /// Returns the BHP hash with an input hasher of 768-bits.
+    fn hash_to_group_bhp768(input: &[bool]) -> Result<Group<Self>> {
+        CANARY_BHP_768.hash_uncompressed(input)
+    }
+
+    /// Returns the BHP hash with an input hasher of 1024-bits.
+    fn hash_to_group_bhp1024(input: &[bool]) -> Result<Group<Self>> {
+        CANARY_BHP_1024.hash_uncompressed(input)
+    }
+
+    /// Returns the Pedersen hash for a given (up to) 64-bit input.
+    fn hash_to_group_ped64(input: &[bool]) -> Result<Group<Self>> {
+        CANARY_PEDERSEN_64.hash_uncompressed(input)
+    }
+
+    /// Returns the Pedersen hash for a given (up to) 128-bit input.
+    fn hash_to_group_ped128(input: &[bool]) -> Result<Group<Self>> {
+        CANARY_PEDERSEN_128.hash_uncompressed(input)
+    }
+
+    /// Returns the Poseidon hash with an input rate of 2 on the affine curve.
+    fn hash_to_group_psd2(input: &[Field<Self>]) -> Result<Group<Self>> {
+        CANARY_POSEIDON_2.hash_to_group(input)
+    }
+
+    /// Returns the Poseidon hash with an input rate of 4 on the affine curve.
+    fn hash_to_group_psd4(input: &[Field<Self>]) -> Result<Group<Self>> {
+        CANARY_POSEIDON_4.hash_to_group(input)
+    }
+
+    /// Returns the Poseidon hash with an input rate of 8 on the affine curve.
+    fn hash_to_group_psd8(input: &[Field<Self>]) -> Result<Group<Self>> {
+        CANARY_POSEIDON_8.hash_to_group(input)
+    }
+
+    /// Returns the Poseidon hash with an input rate of 2 on the scalar field.
+    fn hash_to_scalar_psd2(input: &[Field<Self>]) -> Result<Scalar<Self>> {
+        CANARY_POSEIDON_2.hash_to_scalar(input)
+    }
+
+    /// Returns the Poseidon hash with an input rate of 4 on the scalar field.
+    fn hash_to_scalar_psd4(input: &[Field<Self>]) -> Result<Scalar<Self>> {
+        CANARY_POSEIDON_4.hash_to_scalar(input)
+    }
+
+    /// Returns the Poseidon hash with an input rate of 8 on the scalar field.
+    fn hash_to_scalar_psd8(input: &[Field<Self>]) -> Result<Scalar<Self>> {
+        CANARY_POSEIDON_8.hash_to_scalar(input)
+    }
+
+    /// Returns a Merkle tree with a BHP leaf hasher of 1024-bits and a BHP path hasher of 512-bits.
+    fn merkle_tree_bhp<const DEPTH: u8>(leaves: &[Vec<bool>]) -> Result<BHPMerkleTree<Self, DEPTH>> {
+        MerkleTree::new(&*CANARY_BHP_1024, &*CANARY_BHP_512, leaves)
+    }
+
+    /// Returns a Merkle tree with a Poseidon leaf hasher with input rate of 4 and a Poseidon path hasher with input rate of 2.
+    fn merkle_tree_psd<const DEPTH: u8>(leaves: &[Vec<Field<Self>>]) -> Result<PoseidonMerkleTree<Self, DEPTH>> {
+        MerkleTree::new(&*CANARY_POSEIDON_4, &*CANARY_POSEIDON_2, leaves)
+    }
+
+    /// Returns `true` if the given Merkle path is valid for the given root and leaf.
+    fn verify_merkle_path_bhp<const DEPTH: u8>(
+        path: &MerklePath<Self, DEPTH>,
+        root: &Field<Self>,
+        leaf: &Vec<bool>,
+    ) -> bool {
+        path.verify(&*CANARY_BHP_1024, &*CANARY_BHP_512, root, leaf)
+    }
+
+    /// Returns `true` if the given Merkle path is valid for the given root and leaf.
+    fn verify_merkle_path_psd<const DEPTH: u8>(
+        path: &MerklePath<Self, DEPTH>,
+        root: &Field<Self>,
+        leaf: &Vec<Field<Self>>,
+    ) -> bool {
+        path.verify(&*CANARY_POSEIDON_4, &*CANARY_POSEIDON_2, root, leaf)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    type CurrentNetwork = CanaryV0;
+
+    #[test]
+    fn test_g_scalar_multiply() {
+        // Compute G^r.
+        let scalar = Scalar::rand(&mut TestRng::default());
+        let group = CurrentNetwork::g_scalar_multiply(&scalar);
+        assert_eq!(group, CurrentNetwork::g_powers()[0] * scalar);
+    }
+}
diff --git a/console/network/src/lib.rs b/console/network/src/lib.rs
index 306a77b6f..9d905ac76 100644
--- a/console/network/src/lib.rs
+++ b/console/network/src/lib.rs
@@ -25,6 +25,9 @@ pub use snarkvm_console_network_environment::*;
 mod helpers;
 pub use helpers::*;
 
+mod canary_v0;
+pub use canary_v0::*;
+
 mod mainnet_v0;
 pub use mainnet_v0::*;
 
diff --git a/ledger/query/src/query.rs b/ledger/query/src/query.rs
index a7c62f184..c5d0ac524 100644
--- a/ledger/query/src/query.rs
+++ b/ledger/query/src/query.rs
@@ -72,6 +72,9 @@ impl<N: Network, B: BlockStorage<N>> QueryTrait<N> for Query<N, B> {
                 console::network::TestnetV0::ID => {
                     Ok(Self::get_request(&format!("{url}/testnet/latest/stateRoot"))?.into_json()?)
                 }
+                console::network::CanaryV0::ID => {
+                    Ok(Self::get_request(&format!("{url}/canary/latest/stateRoot"))?.into_json()?)
+                }
                 _ => bail!("Unsupported network ID in inclusion query"),
             },
         }
@@ -89,6 +92,9 @@ impl<N: Network, B: BlockStorage<N>> QueryTrait<N> for Query<N, B> {
                 console::network::TestnetV0::ID => {
                     Ok(Self::get_request_async(&format!("{url}/testnet/latest/stateRoot")).await?.json().await?)
                 }
+                console::network::CanaryV0::ID => {
+                    Ok(Self::get_request_async(&format!("{url}/canary/latest/stateRoot")).await?.json().await?)
+                }
                 _ => bail!("Unsupported network ID in inclusion query"),
             },
         }
@@ -105,6 +111,9 @@ impl<N: Network, B: BlockStorage<N>> QueryTrait<N> for Query<N, B> {
                 console::network::TestnetV0::ID => {
                     Ok(Self::get_request(&format!("{url}/testnet/statePath/{commitment}"))?.into_json()?)
                 }
+                console::network::CanaryV0::ID => {
+                    Ok(Self::get_request(&format!("{url}/canary/statePath/{commitment}"))?.into_json()?)
+                }
                 _ => bail!("Unsupported network ID in inclusion query"),
             },
         }
@@ -122,6 +131,9 @@ impl<N: Network, B: BlockStorage<N>> QueryTrait<N> for Query<N, B> {
                 console::network::TestnetV0::ID => {
                     Ok(Self::get_request_async(&format!("{url}/testnet/statePath/{commitment}")).await?.json().await?)
                 }
+                console::network::CanaryV0::ID => {
+                    Ok(Self::get_request_async(&format!("{url}/canary/statePath/{commitment}")).await?.json().await?)
+                }
                 _ => bail!("Unsupported network ID in inclusion query"),
             },
         }
@@ -142,6 +154,9 @@ impl<N: Network, B: BlockStorage<N>> Query<N, B> {
                 console::network::TestnetV0::ID => {
                     Ok(Self::get_request(&format!("{url}/testnet/program/{program_id}"))?.into_json()?)
                 }
+                console::network::CanaryV0::ID => {
+                    Ok(Self::get_request(&format!("{url}/canary/program/{program_id}"))?.into_json()?)
+                }
                 _ => bail!("Unsupported network ID in inclusion query"),
             },
         }
@@ -161,6 +176,9 @@ impl<N: Network, B: BlockStorage<N>> Query<N, B> {
                 console::network::TestnetV0::ID => {
                     Ok(Self::get_request_async(&format!("{url}/testnet/program/{program_id}")).await?.json().await?)
                 }
+                console::network::CanaryV0::ID => {
+                    Ok(Self::get_request_async(&format!("{url}/canary/program/{program_id}")).await?.json().await?)
+                }
                 _ => bail!("Unsupported network ID in inclusion query"),
             },
         }
diff --git a/parameters/examples/inclusion.rs b/parameters/examples/inclusion.rs
index 4b69ffaa2..83c0f116b 100644
--- a/parameters/examples/inclusion.rs
+++ b/parameters/examples/inclusion.rs
@@ -16,7 +16,7 @@ use snarkvm_algorithms::crypto_hash::sha256::sha256;
 use snarkvm_circuit::{Aleo, Assignment};
 use snarkvm_console::{
     account::PrivateKey,
-    network::{MainnetV0, Network, TestnetV0},
+    network::{CanaryV0, MainnetV0, Network, TestnetV0},
     prelude::{One, ToBytes, Zero},
     program::{Plaintext, Record, StatePath},
     types::Field,
@@ -176,6 +176,9 @@ pub fn main() -> Result<()> {
         "testnet" => {
             inclusion::<TestnetV0, snarkvm_circuit::AleoTestnetV0>()?;
         }
+        "canary" => {
+            inclusion::<CanaryV0, snarkvm_circuit::AleoCanaryV0>()?;
+        }
         _ => panic!("Invalid network"),
     };
 
diff --git a/parameters/examples/setup.rs b/parameters/examples/setup.rs
index 557879d37..e324c9b0c 100644
--- a/parameters/examples/setup.rs
+++ b/parameters/examples/setup.rs
@@ -14,7 +14,7 @@
 
 use snarkvm_algorithms::crypto_hash::sha256::sha256;
 use snarkvm_circuit::Aleo;
-use snarkvm_console::network::{prelude::ToBytes, MainnetV0, Network, TestnetV0};
+use snarkvm_console::network::{prelude::ToBytes, CanaryV0, MainnetV0, Network, TestnetV0};
 use snarkvm_synthesizer::{Process, Program};
 
 use anyhow::Result;
@@ -152,6 +152,7 @@ pub fn main() -> Result<()> {
         "credits" => match args[2].as_str() {
             "mainnet" => credits_program::<MainnetV0, snarkvm_circuit::AleoV0>(),
             "testnet" => credits_program::<TestnetV0, snarkvm_circuit::AleoTestnetV0>(),
+            "canary" => credits_program::<CanaryV0, snarkvm_circuit::AleoCanaryV0>(),
             _ => panic!("Invalid network"),
         }?,
         _ => panic!("Invalid parameter"),
diff --git a/parameters/scripts/canary/credits.sh b/parameters/scripts/canary/credits.sh
new file mode 100755
index 000000000..5b816dfb8
--- /dev/null
+++ b/parameters/scripts/canary/credits.sh
@@ -0,0 +1,9 @@
+# Generate proving and verifying keys.
+
+# Inputs: program name
+
+cargo run --release --example setup credits canary -- --nocapture || exit
+
+mv *.metadata ../../src/canary/resources || exit
+mv *.prover.* ~/.aleo/resources || exit
+mv *.verifier ../../src/canary/resources || exit
diff --git a/parameters/scripts/canary/inclusion.sh b/parameters/scripts/canary/inclusion.sh
new file mode 100755
index 000000000..339e1615a
--- /dev/null
+++ b/parameters/scripts/canary/inclusion.sh
@@ -0,0 +1,9 @@
+# Generates the inclusion proving and verifying key.
+
+# Inputs: network
+
+cargo run --release --example inclusion canary -- --nocapture || exit
+
+mv inclusion.metadata ../../src/canary/resources || exit
+mv inclusion.prover.* ~/.aleo/resources || exit
+mv inclusion.verifier ../../src/canary/resources || exit
diff --git a/parameters/src/canary/genesis.rs b/parameters/src/canary/genesis.rs
new file mode 100644
index 000000000..d3a8b4935
--- /dev/null
+++ b/parameters/src/canary/genesis.rs
@@ -0,0 +1,32 @@
+// Copyright (C) 2019-2023 Aleo Systems Inc.
+// This file is part of the snarkVM library.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+// http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+pub struct GenesisBytes;
+
+impl GenesisBytes {
+    pub const fn load_bytes() -> &'static [u8] {
+        include_bytes!("./resources/block.genesis")
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_genesis_block() {
+        let bytes = GenesisBytes::load_bytes();
+        assert_eq!(14943, bytes.len() as u64, "Update me if serialization has changed");
+    }
+}
diff --git a/parameters/src/canary/mod.rs b/parameters/src/canary/mod.rs
new file mode 100644
index 000000000..02dd96dc8
--- /dev/null
+++ b/parameters/src/canary/mod.rs
@@ -0,0 +1,138 @@
+// Copyright (C) 2019-2023 Aleo Systems Inc.
+// This file is part of the snarkVM library.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+// http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+pub mod genesis;
+pub use genesis::*;
+
+const REMOTE_URL: &str = "https://s3-us-west-1.amazonaws.com/canary.parameters";
+
+// BondPublic
+impl_remote!(BondPublicProver, REMOTE_URL, "resources/", "bond_public", "prover");
+impl_local!(BondPublicVerifier, "resources/", "bond_public", "verifier");
+// UnbondPublic
+impl_remote!(UnbondPublicProver, REMOTE_URL, "resources/", "unbond_public", "prover");
+impl_local!(UnbondPublicVerifier, "resources/", "unbond_public", "verifier");
+// UnbondDelegatorAsValidator
+impl_remote!(UnbondDelegatorAsValidatorProver, REMOTE_URL, "resources/", "unbond_delegator_as_validator", "prover");
+impl_local!(UnbondDelegatorAsValidatorVerifier, "resources/", "unbond_delegator_as_validator", "verifier");
+// ClaimUnbondPublic
+impl_remote!(ClaimUnbondPublicProver, REMOTE_URL, "resources/", "claim_unbond_public", "prover");
+impl_local!(ClaimUnbondPublicVerifier, "resources/", "claim_unbond_public", "verifier");
+// SetValidatorState
+impl_remote!(SetValidatorStateProver, REMOTE_URL, "resources/", "set_validator_state", "prover");
+impl_local!(SetValidatorStateVerifier, "resources/", "set_validator_state", "verifier");
+// TransferPrivate
+impl_remote!(TransferPrivateProver, REMOTE_URL, "resources/", "transfer_private", "prover");
+impl_local!(TransferPrivateVerifier, "resources/", "transfer_private", "verifier");
+// TransferPublic
+impl_remote!(TransferPublicProver, REMOTE_URL, "resources/", "transfer_public", "prover");
+impl_local!(TransferPublicVerifier, "resources/", "transfer_public", "verifier");
+// TransferPublicAsSigner
+impl_remote!(TransferPublicAsSignerProver, REMOTE_URL, "resources/", "transfer_public_as_signer", "prover");
+impl_local!(TransferPublicAsSignerVerifier, "resources/", "transfer_public_as_signer", "verifier");
+// TransferPrivateToPublic
+impl_remote!(TransferPrivateToPublicProver, REMOTE_URL, "resources/", "transfer_private_to_public", "prover");
+impl_local!(TransferPrivateToPublicVerifier, "resources/", "transfer_private_to_public", "verifier");
+// TransferPublicToPrivate
+impl_remote!(TransferPublicToPrivateProver, REMOTE_URL, "resources/", "transfer_public_to_private", "prover");
+impl_local!(TransferPublicToPrivateVerifier, "resources/", "transfer_public_to_private", "verifier");
+// Join
+impl_remote!(JoinProver, REMOTE_URL, "resources/", "join", "prover");
+impl_local!(JoinVerifier, "resources/", "join", "verifier");
+// Split
+impl_remote!(SplitProver, REMOTE_URL, "resources/", "split", "prover");
+impl_local!(SplitVerifier, "resources/", "split", "verifier");
+// FeePrivate
+impl_remote!(FeePrivateProver, REMOTE_URL, "resources/", "fee_private", "prover");
+impl_local!(FeePrivateVerifier, "resources/", "fee_private", "verifier");
+// FeePublic
+impl_remote!(FeePublicProver, REMOTE_URL, "resources/", "fee_public", "prover");
+impl_local!(FeePublicVerifier, "resources/", "fee_public", "verifier");
+
+#[macro_export]
+macro_rules! insert_canary_credit_keys {
+    ($map:ident, $type:ident<$network:ident>, $variant:ident) => {{
+        paste::paste! {
+            let string = stringify!([<$variant:lower>]);
+            $crate::insert_canary_key!($map, string, $type<$network>, ("bond_public", $crate::canary::[<BondPublic $variant>]::load_bytes()));
+            $crate::insert_canary_key!($map, string, $type<$network>, ("unbond_public", $crate::canary::[<UnbondPublic $variant>]::load_bytes()));
+            $crate::insert_canary_key!($map, string, $type<$network>, ("unbond_delegator_as_validator", $crate::canary::[<UnbondDelegatorAsValidator $variant>]::load_bytes()));
+            $crate::insert_canary_key!($map, string, $type<$network>, ("claim_unbond_public", $crate::canary::[<ClaimUnbondPublic $variant>]::load_bytes()));
+            $crate::insert_canary_key!($map, string, $type<$network>, ("set_validator_state", $crate::canary::[<SetValidatorState $variant>]::load_bytes()));
+            $crate::insert_canary_key!($map, string, $type<$network>, ("transfer_private", $crate::canary::[<TransferPrivate $variant>]::load_bytes()));
+            $crate::insert_canary_key!($map, string, $type<$network>, ("transfer_public", $crate::canary::[<TransferPublic $variant>]::load_bytes()));
+            $crate::insert_canary_key!($map, string, $type<$network>, ("transfer_public_as_signer", $crate::canary::[<TransferPublicAsSigner $variant>]::load_bytes()));
+            $crate::insert_canary_key!($map, string, $type<$network>, ("transfer_private_to_public", $crate::canary::[<TransferPrivateToPublic $variant>]::load_bytes()));
+            $crate::insert_canary_key!($map, string, $type<$network>, ("transfer_public_to_private", $crate::canary::[<TransferPublicToPrivate $variant>]::load_bytes()));
+            $crate::insert_canary_key!($map, string, $type<$network>, ("join", $crate::canary::[<Join $variant>]::load_bytes()));
+            $crate::insert_canary_key!($map, string, $type<$network>, ("split", $crate::canary::[<Split $variant>]::load_bytes()));
+            $crate::insert_canary_key!($map, string, $type<$network>, ("fee_private", $crate::canary::[<FeePrivate $variant>]::load_bytes()));
+            $crate::insert_canary_key!($map, string, $type<$network>, ("fee_public", $crate::canary::[<FeePublic $variant>]::load_bytes()));
+        }
+    }};
+}
+
+#[macro_export]
+macro_rules! insert_canary_key {
+    ($map:ident, $string:tt, $type:ident<$network:ident>, ($name:tt, $circuit_key:expr)) => {{
+        // Load the circuit key bytes.
+        let key_bytes: Vec<u8> = $circuit_key.expect(&format!("Failed to load {} bytes", $string));
+        // Recover the circuit key.
+        let key = $type::<$network>::from_bytes_le(&key_bytes[1..]).expect(&format!("Failed to recover {}", $string));
+        // Insert the circuit key.
+        $map.insert($name.to_string(), std::sync::Arc::new(key));
+    }};
+}
+
+// Inclusion
+impl_remote!(InclusionProver, REMOTE_URL, "resources/", "inclusion", "prover");
+impl_local!(InclusionVerifier, "resources/", "inclusion", "verifier");
+
+/// The function name for the inclusion circuit.
+pub const NETWORK_INCLUSION_FUNCTION_NAME: &str = "inclusion";
+
+lazy_static! {
+    pub static ref INCLUSION_PROVING_KEY: Vec<u8> =
+        InclusionProver::load_bytes().expect("Failed to load inclusion proving key");
+    pub static ref INCLUSION_VERIFYING_KEY: Vec<u8> =
+        InclusionVerifier::load_bytes().expect("Failed to load inclusion verifying key");
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use wasm_bindgen_test::*;
+    wasm_bindgen_test_configure!(run_in_browser);
+
+    #[wasm_bindgen_test]
+    fn test_load_bytes() {
+        BondPublicVerifier::load_bytes().expect("Failed to load bond_public verifier");
+        UnbondPublicVerifier::load_bytes().expect("Failed to load unbond_public verifier");
+        UnbondDelegatorAsValidatorVerifier::load_bytes()
+            .expect("Failed to load unbond_delegator_as_validator verifier");
+        ClaimUnbondPublicVerifier::load_bytes().expect("Failed to load claim_unbond_public verifier");
+        SetValidatorStateVerifier::load_bytes().expect("Failed to load set_validator_state verifier");
+        TransferPrivateVerifier::load_bytes().expect("Failed to load transfer_private verifier");
+        TransferPublicVerifier::load_bytes().expect("Failed to load transfer_public verifier");
+        TransferPublicAsSignerVerifier::load_bytes().expect("Failed to load transfer_public_as_signer verifier");
+        TransferPrivateToPublicVerifier::load_bytes().expect("Failed to load transfer_private_to_public verifier");
+        TransferPublicToPrivateVerifier::load_bytes().expect("Failed to load transfer_public_to_private verifier");
+        FeePrivateProver::load_bytes().expect("Failed to load fee_private prover");
+        FeePrivateVerifier::load_bytes().expect("Failed to load fee_private verifier");
+        FeePublicProver::load_bytes().expect("Failed to load fee_public prover");
+        FeePublicVerifier::load_bytes().expect("Failed to load fee_public verifier");
+        InclusionProver::load_bytes().expect("Failed to load inclusion prover");
+        InclusionVerifier::load_bytes().expect("Failed to load inclusion verifier");
+    }
+}
diff --git a/parameters/src/canary/resources/block.genesis b/parameters/src/canary/resources/block.genesis
new file mode 100644
index 000000000..53847e49c
Binary files /dev/null and b/parameters/src/canary/resources/block.genesis differ
diff --git a/parameters/src/canary/resources/bond_public.metadata b/parameters/src/canary/resources/bond_public.metadata
new file mode 100644
index 000000000..8462ca243
--- /dev/null
+++ b/parameters/src/canary/resources/bond_public.metadata
@@ -0,0 +1,6 @@
+{
+  "prover_checksum": "ec245a43cd43f977cade50d77ec05b094b513a6bfcb219ff4c3571443b424128",
+  "prover_size": 29353850,
+  "verifier_checksum": "0efa4a4f9269397b1e7be9cdc2bbbf997155d6020f14d28de348fb9d4f5dc4fb",
+  "verifier_size": 673
+}
\ No newline at end of file
diff --git a/parameters/src/canary/resources/bond_public.verifier b/parameters/src/canary/resources/bond_public.verifier
new file mode 100644
index 000000000..de20c47e6
Binary files /dev/null and b/parameters/src/canary/resources/bond_public.verifier differ
diff --git a/parameters/src/canary/resources/claim_unbond_public.metadata b/parameters/src/canary/resources/claim_unbond_public.metadata
new file mode 100644
index 000000000..01ed51d3d
--- /dev/null
+++ b/parameters/src/canary/resources/claim_unbond_public.metadata
@@ -0,0 +1,6 @@
+{
+  "prover_checksum": "8868c181198d45ff5076b743ffb1703ef94c352d50fefd8bac8184f32acfc96a",
+  "prover_size": 16861332,
+  "verifier_checksum": "bf2e64d878f378f9f47764f5f1246a7bc71ef6a685dae6e91247f3b111bb15f7",
+  "verifier_size": 673
+}
\ No newline at end of file
diff --git a/parameters/src/canary/resources/claim_unbond_public.verifier b/parameters/src/canary/resources/claim_unbond_public.verifier
new file mode 100644
index 000000000..48fd9e6bf
Binary files /dev/null and b/parameters/src/canary/resources/claim_unbond_public.verifier differ
diff --git a/parameters/src/canary/resources/fee_private.metadata b/parameters/src/canary/resources/fee_private.metadata
new file mode 100644
index 000000000..d09bee560
--- /dev/null
+++ b/parameters/src/canary/resources/fee_private.metadata
@@ -0,0 +1,6 @@
+{
+  "prover_checksum": "b3640d77dd92a0a03aa3cb9e7e5d7241e10165d61066cc2cf8e3dd2aeb9bc0b2",
+  "prover_size": 66297972,
+  "verifier_checksum": "c1242f35a0943dbe9d3d68ae29c5e69871e2ed62a876792324f2583320a6b07a",
+  "verifier_size": 673
+}
\ No newline at end of file
diff --git a/parameters/src/canary/resources/fee_private.verifier b/parameters/src/canary/resources/fee_private.verifier
new file mode 100644
index 000000000..a261e295a
Binary files /dev/null and b/parameters/src/canary/resources/fee_private.verifier differ
diff --git a/parameters/src/canary/resources/fee_public.metadata b/parameters/src/canary/resources/fee_public.metadata
new file mode 100644
index 000000000..1476ce96c
--- /dev/null
+++ b/parameters/src/canary/resources/fee_public.metadata
@@ -0,0 +1,6 @@
+{
+  "prover_checksum": "1667bf508398fb94cb97dc7c99b647195492e0e8a9bf57fdb73fa3c294a3ceaf",
+  "prover_size": 29174402,
+  "verifier_checksum": "63936b8503ef8daf476e48d54106872fe813041a0b81e0f44c9714999866ebc4",
+  "verifier_size": 673
+}
\ No newline at end of file
diff --git a/parameters/src/canary/resources/fee_public.verifier b/parameters/src/canary/resources/fee_public.verifier
new file mode 100644
index 000000000..3fbae8a94
Binary files /dev/null and b/parameters/src/canary/resources/fee_public.verifier differ
diff --git a/parameters/src/canary/resources/inclusion.metadata b/parameters/src/canary/resources/inclusion.metadata
new file mode 100644
index 000000000..8df45d6ed
--- /dev/null
+++ b/parameters/src/canary/resources/inclusion.metadata
@@ -0,0 +1,6 @@
+{
+  "prover_checksum": "8faa4d3aaaa5e786d20b61f825e50901f9dcf470455d1d2994e091eefcc66a4e",
+  "prover_size": 233812212,
+  "verifier_checksum": "57bf79c7a493ee79f8ebed2c8e6e7e8a0a6d11b2f6edcba6e943768cb6a250cc",
+  "verifier_size": 673
+}
\ No newline at end of file
diff --git a/parameters/src/canary/resources/inclusion.verifier b/parameters/src/canary/resources/inclusion.verifier
new file mode 100644
index 000000000..d03795688
Binary files /dev/null and b/parameters/src/canary/resources/inclusion.verifier differ
diff --git a/parameters/src/canary/resources/join.metadata b/parameters/src/canary/resources/join.metadata
new file mode 100644
index 000000000..42ed8eb33
--- /dev/null
+++ b/parameters/src/canary/resources/join.metadata
@@ -0,0 +1,6 @@
+{
+  "prover_checksum": "11c8e1702e4644ac8b31e2ad67c97ebf7d2f84e7f82298bbc52d25fe69b26150",
+  "prover_size": 74722164,
+  "verifier_checksum": "ec6395b9c03b49c941fcba788c16f73ff4896c4ee1cbb0dd19c9d13d43f51479",
+  "verifier_size": 673
+}
\ No newline at end of file
diff --git a/parameters/src/canary/resources/join.verifier b/parameters/src/canary/resources/join.verifier
new file mode 100644
index 000000000..8a5b2fbde
Binary files /dev/null and b/parameters/src/canary/resources/join.verifier differ
diff --git a/parameters/src/canary/resources/set_validator_state.metadata b/parameters/src/canary/resources/set_validator_state.metadata
new file mode 100644
index 000000000..aba7a5b4f
--- /dev/null
+++ b/parameters/src/canary/resources/set_validator_state.metadata
@@ -0,0 +1,6 @@
+{
+  "prover_checksum": "e8fbb86008343485bb95c3e3e75d41b493ba654cc161d8b9f9c6901cb71d7a9d",
+  "prover_size": 17390188,
+  "verifier_checksum": "d0942ee642048e21cbf562737b4760524dd378ac17867e7a225736badb23f614",
+  "verifier_size": 673
+}
\ No newline at end of file
diff --git a/parameters/src/canary/resources/set_validator_state.verifier b/parameters/src/canary/resources/set_validator_state.verifier
new file mode 100644
index 000000000..b4b175c97
Binary files /dev/null and b/parameters/src/canary/resources/set_validator_state.verifier differ
diff --git a/parameters/src/canary/resources/split.metadata b/parameters/src/canary/resources/split.metadata
new file mode 100644
index 000000000..b1d962467
--- /dev/null
+++ b/parameters/src/canary/resources/split.metadata
@@ -0,0 +1,6 @@
+{
+  "prover_checksum": "5bf027bcf0476784edd3192c3c0731dbd65f955220ef3d6c060acc0e3d1a5555",
+  "prover_size": 75161428,
+  "verifier_checksum": "5eb3cea5fa953b36272b1a4f315df955253840acbb1ff900dd9e1126bb012437",
+  "verifier_size": 673
+}
\ No newline at end of file
diff --git a/parameters/src/canary/resources/split.verifier b/parameters/src/canary/resources/split.verifier
new file mode 100644
index 000000000..ccdd8a097
Binary files /dev/null and b/parameters/src/canary/resources/split.verifier differ
diff --git a/parameters/src/canary/resources/transfer_private.metadata b/parameters/src/canary/resources/transfer_private.metadata
new file mode 100644
index 000000000..f9cad9090
--- /dev/null
+++ b/parameters/src/canary/resources/transfer_private.metadata
@@ -0,0 +1,6 @@
+{
+  "prover_checksum": "3e4a7a7332889f9a2ad4fd1f1c8d7cb1d8c4b783dca7a832133a6a845fbc757a",
+  "prover_size": 75949172,
+  "verifier_checksum": "b50fcfdebda26d1fbc86f497e945f33118ca3f8d1c9ac95918199a8695408903",
+  "verifier_size": 673
+}
\ No newline at end of file
diff --git a/parameters/src/canary/resources/transfer_private.verifier b/parameters/src/canary/resources/transfer_private.verifier
new file mode 100644
index 000000000..867e4d93d
Binary files /dev/null and b/parameters/src/canary/resources/transfer_private.verifier differ
diff --git a/parameters/src/canary/resources/transfer_private_to_public.metadata b/parameters/src/canary/resources/transfer_private_to_public.metadata
new file mode 100644
index 000000000..0a224a18d
--- /dev/null
+++ b/parameters/src/canary/resources/transfer_private_to_public.metadata
@@ -0,0 +1,6 @@
+{
+  "prover_checksum": "c96d674e52911b0a8b2d0c6d07e5499b90783d8b8e6623d348fa33547835d64b",
+  "prover_size": 66299476,
+  "verifier_checksum": "38d0157f70c16badcfd472dd0c045f754b4ac6572e62be48a0053aaa1157f846",
+  "verifier_size": 673
+}
\ No newline at end of file
diff --git a/parameters/src/canary/resources/transfer_private_to_public.verifier b/parameters/src/canary/resources/transfer_private_to_public.verifier
new file mode 100644
index 000000000..5903d51a1
Binary files /dev/null and b/parameters/src/canary/resources/transfer_private_to_public.verifier differ
diff --git a/parameters/src/canary/resources/transfer_public.metadata b/parameters/src/canary/resources/transfer_public.metadata
new file mode 100644
index 000000000..3482a6c29
--- /dev/null
+++ b/parameters/src/canary/resources/transfer_public.metadata
@@ -0,0 +1,6 @@
+{
+  "prover_checksum": "f143e4c90526e0f08a43daa8977323177587778602b9a3dbb0426908eabe7874",
+  "prover_size": 28913482,
+  "verifier_checksum": "12da9fc8d7284177cc6bcdfb6c8b48217a11a6c70767e6e71f17ffb4e69f17ee",
+  "verifier_size": 673
+}
\ No newline at end of file
diff --git a/parameters/src/canary/resources/transfer_public.verifier b/parameters/src/canary/resources/transfer_public.verifier
new file mode 100644
index 000000000..8008e2d53
Binary files /dev/null and b/parameters/src/canary/resources/transfer_public.verifier differ
diff --git a/parameters/src/canary/resources/transfer_public_as_signer.metadata b/parameters/src/canary/resources/transfer_public_as_signer.metadata
new file mode 100644
index 000000000..b7d7d062d
--- /dev/null
+++ b/parameters/src/canary/resources/transfer_public_as_signer.metadata
@@ -0,0 +1,6 @@
+{
+  "prover_checksum": "e1f2676aeae5d639cc2212a514c50a5dd42efce8b28015ac56a4600200066713",
+  "prover_size": 28915282,
+  "verifier_checksum": "059827d436bd76af3c65c05b518dbcdd56370be4e5524b016c2f0dc702505eb4",
+  "verifier_size": 673
+}
\ No newline at end of file
diff --git a/parameters/src/canary/resources/transfer_public_as_signer.verifier b/parameters/src/canary/resources/transfer_public_as_signer.verifier
new file mode 100644
index 000000000..f511b174e
Binary files /dev/null and b/parameters/src/canary/resources/transfer_public_as_signer.verifier differ
diff --git a/parameters/src/canary/resources/transfer_public_to_private.metadata b/parameters/src/canary/resources/transfer_public_to_private.metadata
new file mode 100644
index 000000000..12d9e0df2
--- /dev/null
+++ b/parameters/src/canary/resources/transfer_public_to_private.metadata
@@ -0,0 +1,6 @@
+{
+  "prover_checksum": "cf5854ed4d0976aecd30f3753d9b3322fa369a4fa7e35748dbb27eb35df3791d",
+  "prover_size": 38413316,
+  "verifier_checksum": "db3285742eac86cc061e408cc2ec16c4318bcee98ccd5f19c358d572e83967e9",
+  "verifier_size": 673
+}
\ No newline at end of file
diff --git a/parameters/src/canary/resources/transfer_public_to_private.verifier b/parameters/src/canary/resources/transfer_public_to_private.verifier
new file mode 100644
index 000000000..90d7f127f
Binary files /dev/null and b/parameters/src/canary/resources/transfer_public_to_private.verifier differ
diff --git a/parameters/src/canary/resources/unbond_delegator_as_validator.metadata b/parameters/src/canary/resources/unbond_delegator_as_validator.metadata
new file mode 100644
index 000000000..b628839a8
--- /dev/null
+++ b/parameters/src/canary/resources/unbond_delegator_as_validator.metadata
@@ -0,0 +1,6 @@
+{
+  "prover_checksum": "1918cf7394c340549f64837dd0d279ce81db77bb8a0d43ccb1eb1f462766f046",
+  "prover_size": 27062250,
+  "verifier_checksum": "0b7889fbe15dc7cf42c43a37a3b979d17ad878ce5d2631bb251cecd8b2d60807",
+  "verifier_size": 673
+}
\ No newline at end of file
diff --git a/parameters/src/canary/resources/unbond_delegator_as_validator.verifier b/parameters/src/canary/resources/unbond_delegator_as_validator.verifier
new file mode 100644
index 000000000..0ec30397f
Binary files /dev/null and b/parameters/src/canary/resources/unbond_delegator_as_validator.verifier differ
diff --git a/parameters/src/canary/resources/unbond_public.metadata b/parameters/src/canary/resources/unbond_public.metadata
new file mode 100644
index 000000000..ae05372c4
--- /dev/null
+++ b/parameters/src/canary/resources/unbond_public.metadata
@@ -0,0 +1,6 @@
+{
+  "prover_checksum": "8f5781539f118a849fc1241777dd14489502d64f039ffc363ffbe6605ba330d2",
+  "prover_size": 17414380,
+  "verifier_checksum": "ac8c29499b4026ded24b2cfa0fff77b5b1619aacf4fe1a9ff487eda9fbd63316",
+  "verifier_size": 673
+}
\ No newline at end of file
diff --git a/parameters/src/canary/resources/unbond_public.verifier b/parameters/src/canary/resources/unbond_public.verifier
new file mode 100644
index 000000000..bd7e1af6b
Binary files /dev/null and b/parameters/src/canary/resources/unbond_public.verifier differ
diff --git a/parameters/src/lib.rs b/parameters/src/lib.rs
index 221b9c275..02ed021cf 100644
--- a/parameters/src/lib.rs
+++ b/parameters/src/lib.rs
@@ -30,6 +30,8 @@ pub mod macros;
 pub mod errors;
 pub use errors::*;
 
+pub mod canary;
+
 pub mod mainnet;
 
 pub mod testnet;
diff --git a/synthesizer/src/vm/helpers/macros.rs b/synthesizer/src/vm/helpers/macros.rs
index 3b002f09a..1859ec293 100644
--- a/synthesizer/src/vm/helpers/macros.rs
+++ b/synthesizer/src/vm/helpers/macros.rs
@@ -66,6 +66,10 @@ macro_rules! convert {
                 // Process the logic.
                 $logic!(console::network::TestnetV0, circuit::AleoTestnetV0)
             }
+            console::network::CanaryV0::ID => {
+                // Process the logic.
+                $logic!(console::network::CanaryV0, circuit::AleoCanaryV0)
+            }
             _ => bail!("Unsupported VM configuration for network: {}", N::ID),
         }
     }};
@@ -93,6 +97,14 @@ macro_rules! process {
                 // Process the logic.
                 $logic!(process.read(), console::network::TestnetV0, circuit::AleoTestnetV0)
             }
+            console::network::CanaryV0::ID => {
+                // Cast the process.
+                let process = (&$self.process as &dyn std::any::Any)
+                    .downcast_ref::<Arc<RwLock<Process<console::network::CanaryV0>>>>()
+                    .ok_or_else(|| anyhow!("Failed to downcast {}", stringify!($self.process)))?;
+                // Process the logic.
+                $logic!(process.read(), console::network::CanaryV0, circuit::AleoCanaryV0)
+            }
             _ => bail!("Unsupported VM configuration for network: {}", N::ID),
         }
     }};
