diff --git a/.cargo/audit.toml b/.cargo/audit.toml
index 35e07ce1b..6c2ff301b 100644
--- a/.cargo/audit.toml
+++ b/.cargo/audit.toml
@@ -5,4 +5,6 @@ ignore = [
   "RUSTSEC-2024-0436",
   # TODO replace encoding crate with encoding_rs
   "RUSTSEC-2021-0153",
+  # TODO find a suitable replacement for fxhash.
+  "RUSTSEC-2025-0057",
 ]
diff --git a/Cargo.lock b/Cargo.lock
index 1b1b2bdde..6e0f57c20 100644
--- a/Cargo.lock
+++ b/Cargo.lock
@@ -121,56 +121,12 @@ version = "0.1.6"
 source = "registry+https://github.com/rust-lang/crates.io-index"
 checksum = "4b46cbb362ab8752921c97e041f5e366ee6297bd428a31275b9fcf1e380f7299"
 
-[[package]]
-name = "anstream"
-version = "0.6.20"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "3ae563653d1938f79b1ab1b5e668c87c76a9930414574a6583a7b7e11a8e6192"
-dependencies = [
- "anstyle",
- "anstyle-parse",
- "anstyle-query",
- "anstyle-wincon",
- "colorchoice",
- "is_terminal_polyfill",
- "utf8parse",
-]
-
 [[package]]
 name = "anstyle"
 version = "1.0.11"
 source = "registry+https://github.com/rust-lang/crates.io-index"
 checksum = "862ed96ca487e809f1c8e5a8447f6ee2cf102f846893800b20cebdf541fc6bbd"
 
-[[package]]
-name = "anstyle-parse"
-version = "0.2.7"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "4e7644824f0aa2c7b9384579234ef10eb7efb6a0deb83f9630a49594dd9c15c2"
-dependencies = [
- "utf8parse",
-]
-
-[[package]]
-name = "anstyle-query"
-version = "1.1.4"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "9e231f6134f61b71076a3eab506c379d4f36122f2af15a9ff04415ea4c3339e2"
-dependencies = [
- "windows-sys 0.60.2",
-]
-
-[[package]]
-name = "anstyle-wincon"
-version = "3.0.10"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "3e0633414522a32ffaac8ac6cc8f748e090c5717661fddeea04219e2344f5f2a"
-dependencies = [
- "anstyle",
- "once_cell_polyfill",
- "windows-sys 0.60.2",
-]
-
 [[package]]
 name = "anyhow"
 version = "1.0.99"
@@ -267,7 +223,7 @@ dependencies = [
  "proc-macro2",
  "quote 1.0.40",
  "regex",
- "rustc-hash 1.1.0",
+ "rustc-hash",
  "shlex",
  "syn 2.0.106",
 ]
@@ -385,9 +341,9 @@ checksum = "37b2a672a2cb129a2e41c10b1224bb368f9f37a2b16b612598138befd7b37eb5"
 
 [[package]]
 name = "cc"
-version = "1.2.36"
+version = "1.2.37"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "5252b3d2648e5eedbc1a6f501e3c795e07025c1e93bbf8bbdd6eef7f447a6d54"
+checksum = "65193589c6404eb80b450d618eaf9a2cafaaafd57ecce47370519ef674a7bd44"
 dependencies = [
  "find-msvc-tools",
  "jobserver",
@@ -410,12 +366,6 @@ version = "1.0.3"
 source = "registry+https://github.com/rust-lang/crates.io-index"
 checksum = "2fd1289c04a9ea8cb22300a459a72a385d7c73d3259e2ed7dcb2af674838cfa9"
 
-[[package]]
-name = "cfg_aliases"
-version = "0.2.1"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "613afe47fcd5fac7ccf1db93babcb082c5994d996f20b8b159f2ad1658eb5724"
-
 [[package]]
 name = "ci_info"
 version = "0.10.2"
@@ -470,7 +420,6 @@ source = "registry+https://github.com/rust-lang/crates.io-index"
 checksum = "7eac00902d9d136acd712710d71823fb8ac8004ca445a89e73a41d45aa712931"
 dependencies = [
  "clap_builder",
- "clap_derive",
 ]
 
 [[package]]
@@ -479,22 +428,8 @@ version = "4.5.47"
 source = "registry+https://github.com/rust-lang/crates.io-index"
 checksum = "2ad9bbf750e73b5884fb8a211a9424a1906c1e156724260fdae972f31d70e1d6"
 dependencies = [
- "anstream",
  "anstyle",
  "clap_lex",
- "strsim",
-]
-
-[[package]]
-name = "clap_derive"
-version = "4.5.47"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "bbfd7eae0b0f1a6e63d4b13c9c478de77c2eb546fba158ad50b4203dc24b9f9c"
-dependencies = [
- "heck",
- "proc-macro2",
- "quote 1.0.40",
- "syn 2.0.106",
 ]
 
 [[package]]
@@ -503,12 +438,6 @@ version = "0.7.5"
 source = "registry+https://github.com/rust-lang/crates.io-index"
 checksum = "b94f61472cee1439c0b966b47e3aca9ae07e45d070759512cd390ea2bebc6675"
 
-[[package]]
-name = "colorchoice"
-version = "1.0.4"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "b05b61dc5112cbb17e4b6cd61790d9845d13888356391624cbe7e41efeac1e75"
-
 [[package]]
 name = "colored"
 version = "2.2.0"
@@ -528,19 +457,6 @@ dependencies = [
  "windows-sys 0.59.0",
 ]
 
-[[package]]
-name = "console"
-version = "0.15.11"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "054ccb5b10f9f2cbf51eb355ca1d05c2d279ce1804688d0db74b4733a5aeafd8"
-dependencies = [
- "encode_unicode",
- "libc",
- "once_cell",
- "unicode-width",
- "windows-sys 0.59.0",
-]
-
 [[package]]
 name = "constant_time_eq"
 version = "0.3.1"
@@ -805,12 +721,6 @@ version = "1.15.0"
 source = "registry+https://github.com/rust-lang/crates.io-index"
 checksum = "48c757948c5ede0e46177b7add2e67155f70e33c07fea8284df6576da70b3719"
 
-[[package]]
-name = "encode_unicode"
-version = "1.0.0"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "34aa73646ffb006b8f5147f3dc182bd4bcb190227ce861fc4a4844bf8e3cb2c0"
-
 [[package]]
 name = "encoding"
 version = "0.2.33"
@@ -1151,11 +1061,9 @@ source = "registry+https://github.com/rust-lang/crates.io-index"
 checksum = "26145e563e54f2cadc477553f1ec5ee650b00862f0a58bcd12cbdc5f0ea2d2f4"
 dependencies = [
  "cfg-if",
- "js-sys",
  "libc",
  "r-efi",
- "wasi 0.14.5+wasi-0.2.4",
- "wasm-bindgen",
+ "wasi 0.14.6+wasi-0.2.4",
 ]
 
 [[package]]
@@ -1222,12 +1130,6 @@ dependencies = [
  "foldhash",
 ]
 
-[[package]]
-name = "heck"
-version = "0.5.0"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "2304e00983f87ffb38b55b444b5e3b60a884b5d30c0fca7d82fe33449bbe55ea"
-
 [[package]]
 name = "hermit-abi"
 version = "0.5.2"
@@ -1325,7 +1227,6 @@ dependencies = [
  "tokio",
  "tokio-rustls",
  "tower-service",
- "webpki-roots",
 ]
 
 [[package]]
@@ -1346,9 +1247,9 @@ dependencies = [
 
 [[package]]
 name = "hyper-util"
-version = "0.1.16"
+version = "0.1.17"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "8d9b05277c7e8da2c93a568989bb6207bef0112e8d17df7a6eda4a3cf143bc5e"
+checksum = "3c6995591a8f1380fcb4ba966a252a4b29188d51d2b89e3a252f5305be65aea8"
 dependencies = [
  "base64",
  "bytes",
@@ -1499,19 +1400,6 @@ dependencies = [
  "serde",
 ]
 
-[[package]]
-name = "indicatif"
-version = "0.17.11"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "183b3088984b400f4cfac3620d5e076c84da5364016b4f49473de574b2586235"
-dependencies = [
- "console",
- "number_prefix",
- "portable-atomic",
- "unicode-width",
- "web-time",
-]
-
 [[package]]
 name = "io-uring"
 version = "0.7.10"
@@ -1539,12 +1427,6 @@ dependencies = [
  "serde",
 ]
 
-[[package]]
-name = "is_terminal_polyfill"
-version = "1.70.1"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "7943c866cc5cd64cbc25b2e01621d07fa8eb2a1a23160ee81ce38704e97b8ecf"
-
 [[package]]
 name = "itertools"
 version = "0.10.5"
@@ -1634,9 +1516,9 @@ checksum = "f9fbbcab51052fe104eb5e5d351cf728d30a5be1fe14d9be8a3b097481fb97de"
 
 [[package]]
 name = "libredox"
-version = "0.1.9"
+version = "0.1.10"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "391290121bad3d37fbddad76d8f5d1c1c314cfc646d143d7e07a3086ddff0ce3"
+checksum = "416f7e718bdb06000964960ffa43b4335ad4012ae8b99060261aa4a8088d5ccb"
 dependencies = [
  "bitflags 2.9.4",
  "libc",
@@ -1729,12 +1611,6 @@ dependencies = [
  "hashbrown 0.15.5",
 ]
 
-[[package]]
-name = "lru-slab"
-version = "0.1.2"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "112b39cec0b298b6c1999fee3e31427f74f676e4cb9879ed1a121b43661a4154"
-
 [[package]]
 name = "lz4-sys"
 version = "1.11.1+lz4-1.10.0"
@@ -1881,16 +1757,6 @@ dependencies = [
  "syn 2.0.106",
 ]
 
-[[package]]
-name = "num-format"
-version = "0.4.4"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "a652d9771a63711fd3c3deb670acfbe5c30a4072e664d7a3bf5a9e1056ac72c3"
-dependencies = [
- "arrayvec",
- "itoa",
-]
-
 [[package]]
 name = "num-integer"
 version = "0.1.46"
@@ -1920,12 +1786,6 @@ dependencies = [
  "libc",
 ]
 
-[[package]]
-name = "number_prefix"
-version = "0.4.0"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "830b246a0e5f20af87141b25c173cd1b609bd7779a4617d6ec582abaf90870f3"
-
 [[package]]
 name = "object"
 version = "0.36.7"
@@ -1941,12 +1801,6 @@ version = "1.21.3"
 source = "registry+https://github.com/rust-lang/crates.io-index"
 checksum = "42f5e15c9953c5e4ccceeb2e7382a716482c34515315f7b03532b8b4e8393d2d"
 
-[[package]]
-name = "once_cell_polyfill"
-version = "1.70.1"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "a4895175b425cb1f87721b59f0f286c2092bd4af812243672510e1ac53e2e0ad"
-
 [[package]]
 name = "oorandom"
 version = "11.1.5"
@@ -2159,70 +2013,6 @@ version = "1.2.3"
 source = "registry+https://github.com/rust-lang/crates.io-index"
 checksum = "a1d01941d82fa2ab50be1e79e6714289dd7cde78eba4c074bc5a4374f650dfe0"
 
-[[package]]
-name = "quick-xml"
-version = "0.37.5"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "331e97a1af0bf59823e6eadffe373d7b27f485be8748f71471c662c1f269b7fb"
-dependencies = [
- "memchr",
-]
-
-[[package]]
-name = "quinn"
-version = "0.11.9"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "b9e20a958963c291dc322d98411f541009df2ced7b5a4f2bd52337638cfccf20"
-dependencies = [
- "bytes",
- "cfg_aliases",
- "pin-project-lite",
- "quinn-proto",
- "quinn-udp",
- "rustc-hash 2.1.1",
- "rustls",
- "socket2",
- "thiserror 2.0.16",
- "tokio",
- "tracing",
- "web-time",
-]
-
-[[package]]
-name = "quinn-proto"
-version = "0.11.13"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "f1906b49b0c3bc04b5fe5d86a77925ae6524a19b816ae38ce1e426255f1d8a31"
-dependencies = [
- "bytes",
- "getrandom 0.3.3",
- "lru-slab",
- "rand 0.9.2",
- "ring",
- "rustc-hash 2.1.1",
- "rustls",
- "rustls-pki-types",
- "slab",
- "thiserror 2.0.16",
- "tinyvec",
- "tracing",
- "web-time",
-]
-
-[[package]]
-name = "quinn-udp"
-version = "0.5.14"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "addec6a0dcad8a8d96a771f815f0eaf55f9d1805756410b39f5fa81332574cbd"
-dependencies = [
- "cfg_aliases",
- "libc",
- "once_cell",
- "socket2",
- "tracing",
- "windows-sys 0.60.2",
-]
-
 [[package]]
 name = "quote"
 version = "0.3.15"
@@ -2409,9 +2199,7 @@ dependencies = [
  "base64",
  "bytes",
  "encoding_rs",
- "futures-channel",
  "futures-core",
- "futures-util",
  "h2",
  "http",
  "http-body",
@@ -2426,8 +2214,6 @@ dependencies = [
  "native-tls",
  "percent-encoding",
  "pin-project-lite",
- "quinn",
- "rustls",
  "rustls-pki-types",
  "serde",
  "serde_json",
@@ -2435,7 +2221,6 @@ dependencies = [
  "sync_wrapper",
  "tokio",
  "tokio-native-tls",
- "tokio-rustls",
  "tower",
  "tower-http",
  "tower-service",
@@ -2443,7 +2228,6 @@ dependencies = [
  "wasm-bindgen",
  "wasm-bindgen-futures",
  "web-sys",
- "webpki-roots",
 ]
 
 [[package]]
@@ -2482,12 +2266,6 @@ version = "1.1.0"
 source = "registry+https://github.com/rust-lang/crates.io-index"
 checksum = "08d43f7aa6b08d49f382cde6a7982047c3426db949b1424bc4b7ec9ae12c6ce2"
 
-[[package]]
-name = "rustc-hash"
-version = "2.1.1"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "357703d41365b4b27c590e3ed91eabb1b663f07c4c084095e60cbed4362dff0d"
-
 [[package]]
 name = "rustc_version"
 version = "0.4.1"
@@ -2553,15 +2331,14 @@ version = "1.12.0"
 source = "registry+https://github.com/rust-lang/crates.io-index"
 checksum = "229a4a4c221013e7e1f1a043678c5cc39fe5171437c88fb47151a21e6f5b5c79"
 dependencies = [
- "web-time",
  "zeroize",
 ]
 
 [[package]]
 name = "rustls-webpki"
-version = "0.103.5"
+version = "0.103.6"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "b5a37813727b78798e53c2bec3f5e8fe12a6d6f8389bf9ca7802add4c9905ad8"
+checksum = "8572f3c2cb9934231157b45499fc41e1f58c589fdfb81a844ba873265e80f8eb"
 dependencies = [
  "ring",
  "rustls-pki-types",
@@ -2652,55 +2429,35 @@ dependencies = [
 ]
 
 [[package]]
-name = "self-replace"
-version = "1.5.0"
+name = "semver"
+version = "1.0.27"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "03ec815b5eab420ab893f63393878d89c90fdd94c0bcc44c07abb8ad95552fb7"
-dependencies = [
- "fastrand",
- "tempfile",
- "windows-sys 0.52.0",
-]
+checksum = "d767eb0aabc880b29956c35734170f26ed551a859dbd361d140cdbeca61ab1e2"
 
 [[package]]
-name = "self_update"
-version = "0.42.0"
+name = "serde"
+version = "1.0.224"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "d832c086ece0dacc29fb2947bb4219b8f6e12fe9e40b7108f9e57c4224e47b5c"
+checksum = "6aaeb1e94f53b16384af593c71e20b095e958dab1d26939c1b70645c5cfbcc0b"
 dependencies = [
- "hyper",
- "indicatif",
- "log",
- "quick-xml",
- "regex",
- "reqwest",
- "self-replace",
- "semver",
- "serde_json",
- "tempfile",
- "urlencoding",
+ "serde_core",
+ "serde_derive",
 ]
 
 [[package]]
-name = "semver"
-version = "1.0.26"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "56e6fa9c48d24d85fb3de5ad847117517440f6beceb7798af16b4a87d616b8d0"
-
-[[package]]
-name = "serde"
-version = "1.0.219"
+name = "serde_core"
+version = "1.0.224"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "5f0e2c6ed6606019b4e29e69dbaba95b11854410e5347d525002456dbbb786b6"
+checksum = "32f39390fa6346e24defbcdd3d9544ba8a19985d0af74df8501fbfe9a64341ab"
 dependencies = [
  "serde_derive",
 ]
 
 [[package]]
 name = "serde_derive"
-version = "1.0.219"
+version = "1.0.224"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "5b0276cf7f2c73365f7157c8123c21cd9a50fbbd844757af28ca1f5925fc2a00"
+checksum = "87ff78ab5e8561c9a675bfc1785cb07ae721f0ee53329a595cefd8c04c2ac4e0"
 dependencies = [
  "proc-macro2",
  "quote 1.0.40",
@@ -2709,15 +2466,16 @@ dependencies = [
 
 [[package]]
 name = "serde_json"
-version = "1.0.143"
+version = "1.0.145"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "d401abef1d108fbd9cbaebc3e46611f4b1021f714a0597a71f41ee463f5f4a5a"
+checksum = "402a6f66d8c709116cf22f558eab210f5a50187f702eb4d7e5ef38d9a7f1c79c"
 dependencies = [
  "indexmap 2.11.1",
  "itoa",
  "memchr",
  "ryu",
  "serde",
+ "serde_core",
 ]
 
 [[package]]
@@ -2830,16 +2588,11 @@ dependencies = [
 name = "snarkvm"
 version = "4.2.1"
 dependencies = [
- "anstyle",
  "anyhow",
  "bincode",
- "clap",
- "colored 3.0.0",
  "dotenvy",
- "num-format",
  "rand 0.8.5",
  "rusty-hook",
- "self_update",
  "serde_json",
  "snarkvm",
  "snarkvm-algorithms",
@@ -2854,7 +2607,6 @@ dependencies = [
  "snarkvm-utilities",
  "snarkvm-wasm",
  "tempfile",
- "thiserror 2.0.16",
  "ureq",
  "walkdir",
 ]
@@ -3841,12 +3593,6 @@ version = "1.2.0"
 source = "registry+https://github.com/rust-lang/crates.io-index"
 checksum = "a8f112729512f8e442d81f95a8a7ddf2b7c6b8a1a6f509a95864142b30cab2d3"
 
-[[package]]
-name = "strsim"
-version = "0.11.1"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "7da8b5736845d9f2fcb837ea5d9e2628564b3b043a70948a3f0b778838c5fb4f"
-
 [[package]]
 name = "structmeta"
 version = "0.2.0"
@@ -4395,12 +4141,6 @@ dependencies = [
  "serde",
 ]
 
-[[package]]
-name = "urlencoding"
-version = "2.1.3"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "daf8dba3b7eb870caf1ddeed7bc9d2a049f3cfdfae7cb521b087cc33ae4c49da"
-
 [[package]]
 name = "utf-8"
 version = "0.7.6"
@@ -4413,12 +4153,6 @@ version = "1.0.4"
 source = "registry+https://github.com/rust-lang/crates.io-index"
 checksum = "b6c140620e7ffbb22c2dee59cafe6084a59b5ffc27a8859a5f0d494b5d52b6be"
 
-[[package]]
-name = "utf8parse"
-version = "0.2.2"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "06abde3611657adf66d383f00b093d7faecc7fa57071cce2578660c9f1010821"
-
 [[package]]
 name = "valuable"
 version = "0.1.1"
@@ -4473,18 +4207,18 @@ checksum = "ccf3ec651a847eb01de73ccad15eb7d99f80485de043efb2f370cd654f4ea44b"
 
 [[package]]
 name = "wasi"
-version = "0.14.5+wasi-0.2.4"
+version = "0.14.6+wasi-0.2.4"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "a4494f6290a82f5fe584817a676a34b9d6763e8d9d18204009fb31dceca98fd4"
+checksum = "7f71243a3f320c00a8459e455c046ce571229c2f31fd11645d9dc095e3068ca0"
 dependencies = [
  "wasip2",
 ]
 
 [[package]]
 name = "wasip2"
-version = "1.0.0+wasi-0.2.4"
+version = "1.0.1+wasi-0.2.4"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "03fa2761397e5bd52002cd7e73110c71af2109aca4e521a9f40473fe685b0a24"
+checksum = "0562428422c63773dad2c345a1882263bbf4d65cf3f42e90921f787ef5ad58e7"
 dependencies = [
  "wit-bindgen",
 ]
@@ -4595,16 +4329,6 @@ dependencies = [
  "wasm-bindgen",
 ]
 
-[[package]]
-name = "web-time"
-version = "1.1.0"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "5a6580f308b1fad9207618087a65c04e7a10bc77e02c8e84e9b00dd4b12fa0bb"
-dependencies = [
- "js-sys",
- "wasm-bindgen",
-]
-
 [[package]]
 name = "webpki-roots"
 version = "1.0.2"
@@ -4727,15 +4451,6 @@ dependencies = [
  "windows-targets 0.52.6",
 ]
 
-[[package]]
-name = "windows-sys"
-version = "0.60.2"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "f2f500e4d28234f72040990ec9d39e3a6b950f9f22d3dba18416c35882612bcb"
-dependencies = [
- "windows-targets 0.53.3",
-]
-
 [[package]]
 name = "windows-sys"
 version = "0.61.0"
@@ -4882,9 +4597,9 @@ checksum = "d135d17ab770252ad95e9a872d365cf3090e3be864a34ab46f48555993efc904"
 
 [[package]]
 name = "wit-bindgen"
-version = "0.45.1"
+version = "0.46.0"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "5c573471f125075647d03df72e026074b7203790d41351cd6edc96f46bcccd36"
+checksum = "f17a85883d4e6d00e8a97c586de764dabcc06133f7f1d55dce5cdc070ad7fe59"
 
 [[package]]
 name = "writeable"
diff --git a/Cargo.toml b/Cargo.toml
index 2a7fdf61e..8fbea44d5 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -91,11 +91,6 @@ members = [
 [lib]
 path = "vm/lib.rs"
 
-[[bin]]
-name = "snarkvm"
-path = "vm/cli/main.rs"
-required-features = [ "cli" ]
-
 [features]
 default = [
   "algorithms",
@@ -109,22 +104,10 @@ default = [
   "file",
 ]
 full = [
-  "cli",
   "default",
   "curves",
   "fields",
 ]
-cli = [
-  "circuit",
-  "utilities",
-  "package",
-  "dep:clap",
-  "dep:colored",
-  "dep:self_update",
-  "dep:thiserror",
-  "dep:ureq",
-  "dep:serde_json",
-]
 file = [
   "console",
   "synthesizer",
@@ -139,6 +122,7 @@ package = [
   "dep:ureq",
   "dep:dotenvy"
 ]
+
 async = [ "snarkvm-ledger/async", "snarkvm-synthesizer/async" ]
 cuda = [ "snarkvm-algorithms/cuda" ]
 dev-print = [ 
@@ -461,6 +445,9 @@ version = "0.6"
 [workspace.dependencies.colored]
 version = "3"
 
+[workspace.dependencies.dotenvy]
+version = "0.15"
+
 [workspace.dependencies.expect-test]
 version = "1.4.1"
 
@@ -583,54 +570,29 @@ optional = true
 workspace = true
 optional = true
 
-[dependencies.anstyle]
-version = "1"
-
 [dependencies.anyhow]
 workspace = true
 
-[dependencies.clap]
-version = "4.4"
-features = [ "derive", "color" ]
-optional = true
-
-[dependencies.colored]
-workspace = true
-optional = true
-
 [dependencies.dotenvy]
 version = "0.15"
 optional = true
 
-[dependencies.num-format]
-version = "0.4.4"
-
 [dependencies.rand]
 workspace = true
 
-[dependencies.self_update]
-version = "0.42"
-optional = true
-
 [dependencies.serde_json]
 workspace = true
 optional = true
 features = [ "preserve_order" ]
 
-[dependencies.thiserror]
-workspace = true
-optional = true
-
 [dependencies.ureq]
 workspace = true
-features = [ "json" ]
 optional = true
+features = [ "json" ]
 default-features = true
 
-# Some cli helper functions are used in the tests.
 [dev-dependencies.snarkvm]
 path = "."
-features = [ "cli" ]
 
 [dev-dependencies.bincode]
 workspace = true
diff --git a/README.md b/README.md
index 1afe3b395..0ff717b5b 100644
--- a/README.md
+++ b/README.md
@@ -51,36 +51,25 @@ We recommend installing Rust using [rustup](https://www.rustup.rs/). You can ins
   Download the [Windows 64-bit executable](https://win.rustup.rs/x86_64) or
   [Windows 32-bit executable](https://win.rustup.rs/i686) and follow the on-screen instructions.
 
-### 2.2.1 Build from Crates.io
+### 2.2 Using snarkVM as a Library
 
-We recommend installing `snarkvm` this way. In your terminal, run:
+snarkVM is primarily designed to be used as a library in Rust projects. Add it to your `Cargo.toml` with your favourite published version:
 
-```bash
-cargo install snarkvm
-```
-
-Now to use `snarkvm`, in your terminal, run:
-```bash
-snarkvm
+```toml
+[dependencies]
+snarkvm = "<major>.<minor>.<patch>"
 ```
 
-### 2.2.2 Build from Source Code
+### 2.3 Build from Source Code
 
-Alternatively, you can install `snarkvm` by building from the source code as follows:
+You can also build snarkVM from source:
 
 ```bash
 # Fetch the repository's development (staging) branch
 git clone --branch staging --single-branch https://github.com/ProvableHQ/snarkVM.git 
 cd snarkVM
-# Optional: pick a specific release of snarkOS
-git checkout v3.8.0
-# Install snarkVM
-cargo install --path . --features=cli
-```
-
-Now to use `snarkvm`, in your terminal, run:
-```bash
-snarkvm
+# Build the library
+cargo build --release
 ```
 
 ## 3. Contributors
diff --git a/vm/cli/cli.rs b/vm/cli/cli.rs
deleted file mode 100644
index effcd4744..000000000
--- a/vm/cli/cli.rs
+++ /dev/null
@@ -1,77 +0,0 @@
-// Copyright (c) 2019-2025 Provable Inc.
-// This file is part of the snarkVM library.
-
-// Licensed under the Apache License, Version 2.0 (the "License");
-// you may not use this file except in compliance with the License.
-// You may obtain a copy of the License at:
-
-// http://www.apache.org/licenses/LICENSE-2.0
-
-// Unless required by applicable law or agreed to in writing, software
-// distributed under the License is distributed on an "AS IS" BASIS,
-// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
-// See the License for the specific language governing permissions and
-// limitations under the License.
-
-use crate::cli::commands::{Clean, Execute, New, Run, Update};
-
-use anstyle::{AnsiColor, Color, Style};
-use anyhow::Result;
-use clap::{Parser, builder::Styles};
-
-const HEADER_COLOR: Option<Color> = Some(Color::Ansi(AnsiColor::Yellow));
-const LITERAL_COLOR: Option<Color> = Some(Color::Ansi(AnsiColor::Green));
-const STYLES: Styles = Styles::plain()
-    .header(Style::new().bold().fg_color(HEADER_COLOR))
-    .usage(Style::new().bold().fg_color(HEADER_COLOR))
-    .literal(Style::new().bold().fg_color(LITERAL_COLOR));
-
-#[derive(Debug, Parser)]
-#[clap(name = "snarkVM", author = "The Aleo Team <hello@aleo.org>", styles = STYLES)]
-pub struct CLI {
-    /// Specify the verbosity [options: 0, 1, 2, 3]
-    #[clap(default_value = "2", short, long)]
-    pub verbosity: u8,
-    /// Specify a subcommand.
-    #[clap(subcommand)]
-    pub command: Command,
-}
-
-#[derive(Debug, Parser)]
-pub enum Command {
-    #[clap(name = "clean")]
-    Clean(Clean),
-    #[clap(name = "execute")]
-    Execute(Execute),
-    #[clap(name = "new")]
-    New(New),
-    #[clap(name = "run")]
-    Run(Run),
-    #[clap(name = "update")]
-    Update(Update),
-}
-
-impl Command {
-    /// Parse the command.
-    pub fn parse(self) -> Result<String> {
-        match self {
-            Self::Clean(command) => command.parse(),
-            Self::Execute(command) => command.parse(),
-            Self::New(command) => command.parse(),
-            Self::Run(command) => command.parse(),
-            Self::Update(command) => command.parse(),
-        }
-    }
-}
-
-#[cfg(test)]
-mod tests {
-    use super::*;
-
-    // A test case recommended by clap (https://docs.rs/clap/latest/clap/_derive/_tutorial/index.html#testing).
-    #[test]
-    fn verify_cli() {
-        use clap::CommandFactory;
-        CLI::command().debug_assert()
-    }
-}
diff --git a/vm/cli/commands/clean.rs b/vm/cli/commands/clean.rs
deleted file mode 100644
index ae8c4f30b..000000000
--- a/vm/cli/commands/clean.rs
+++ /dev/null
@@ -1,36 +0,0 @@
-// Copyright (c) 2019-2025 Provable Inc.
-// This file is part of the snarkVM library.
-
-// Licensed under the Apache License, Version 2.0 (the "License");
-// you may not use this file except in compliance with the License.
-// You may obtain a copy of the License at:
-
-// http://www.apache.org/licenses/LICENSE-2.0
-
-// Unless required by applicable law or agreed to in writing, software
-// distributed under the License is distributed on an "AS IS" BASIS,
-// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
-// See the License for the specific language governing permissions and
-// limitations under the License.
-
-use super::*;
-
-/// Cleans the Aleo package build directory.
-#[derive(Debug, Parser)]
-pub struct Clean;
-
-impl Clean {
-    /// Cleans an Aleo package build directory.
-    pub fn parse(self) -> Result<String> {
-        // Derive the program directory path.
-        let path = std::env::current_dir()?;
-
-        // Clean the build directory.
-        Package::<CurrentNetwork>::clean(&path)?;
-
-        // Prepare the path string.
-        let path_string = format!("(in \"{}\")", path.join("build").display());
-
-        Ok(format!("✅ Cleaned the build directory {}", path_string.dimmed()))
-    }
-}
diff --git a/vm/cli/commands/execute.rs b/vm/cli/commands/execute.rs
deleted file mode 100644
index a26c69ded..000000000
--- a/vm/cli/commands/execute.rs
+++ /dev/null
@@ -1,118 +0,0 @@
-// Copyright (c) 2019-2025 Provable Inc.
-// This file is part of the snarkVM library.
-
-// Licensed under the Apache License, Version 2.0 (the "License");
-// you may not use this file except in compliance with the License.
-// You may obtain a copy of the License at:
-
-// http://www.apache.org/licenses/LICENSE-2.0
-
-// Unless required by applicable law or agreed to in writing, software
-// distributed under the License is distributed on an "AS IS" BASIS,
-// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
-// See the License for the specific language governing permissions and
-// limitations under the License.
-
-use super::*;
-
-use anyhow::Context;
-
-/// Executes an Aleo program function locally
-#[derive(Debug, Parser)]
-pub struct Execute {
-    /// The function name.
-    function: Identifier<CurrentNetwork>,
-    /// The function inputs.
-    inputs: Vec<Value<CurrentNetwork>>,
-    /// Uses the specified endpoint.
-    #[clap(default_value = "https://api.explorer.aleo.org/v1", long)]
-    endpoint: String,
-    /// Toggles offline mode.
-    #[clap(long)]
-    offline: bool,
-}
-
-impl Execute {
-    /// Compiles an Aleo program function with the specified name.
-    #[allow(clippy::format_in_format_args)]
-    pub fn parse(self) -> Result<String> {
-        // Derive the program directory path.
-        let path = std::env::current_dir()?;
-
-        // Load the package.
-        let package = Package::open(&path)?;
-        // Load the private key.
-        let private_key = crate::cli::helpers::dotenv_private_key()?;
-
-        // Initialize an RNG.
-        let rng = &mut rand::thread_rng();
-
-        // Execute the request.
-        let (response, execution, metrics) = package
-            .execute::<Aleo, _>(self.endpoint, &private_key, self.function, &self.inputs, rng)
-            .with_context(|| "Execution failed")?;
-
-        // TODO (howardwu): Include the option to execute a fee.
-        let fee = None;
-
-        // Construct the transaction.
-        let transaction =
-            Transaction::from_execution(execution, fee).with_context(|| "Failed to construct transaction")?;
-
-        // Count the number of times a function is called.
-        let mut program_frequency = HashMap::<String, usize>::new();
-        for metric in metrics.iter() {
-            // Prepare the function name string.
-            let function_name_string = format!("'{}/{}'", metric.program_id, metric.function_name).bold();
-
-            // Prepare the function constraints string
-            let function_constraints_string = format!(
-                "{function_name_string} - {} constraints",
-                metric.num_function_constraints.to_formatted_string(LOCALE)
-            );
-
-            // Increment the counter for the function call.
-            match program_frequency.get_mut(&function_constraints_string) {
-                Some(counter) => *counter += 1,
-                None => {
-                    let _ = program_frequency.insert(function_constraints_string, 1);
-                }
-            }
-        }
-
-        // Log the metrics.
-        use num_format::ToFormattedString;
-
-        println!("⛓  Constraints\n");
-        for (function_constraints, counter) in program_frequency {
-            // Log the constraints
-            let counter_string = match counter {
-                1 => "(called 1 time)".to_string().dimmed(),
-                counter => format!("(called {counter} times)").dimmed(),
-            };
-
-            println!(" •  {function_constraints} {counter_string}",)
-        }
-
-        // Log the outputs.
-        match response.outputs().len() {
-            0 => (),
-            1 => println!("\n➡️  Output\n"),
-            _ => println!("\n➡️  Outputs\n"),
-        };
-        for output in response.outputs() {
-            println!("{}", format!(" • {output}"));
-        }
-        println!();
-
-        // Print the transaction.
-        println!("{transaction}\n");
-
-        // Prepare the locator.
-        let locator = Locator::<CurrentNetwork>::from_str(&format!("{}/{}", package.program_id(), self.function))?;
-        // Prepare the path string.
-        let path_string = format!("(in \"{}\")", path.display());
-
-        Ok(format!("✅ Executed '{}' {}", locator.to_string().bold(), path_string.dimmed()))
-    }
-}
diff --git a/vm/cli/commands/mod.rs b/vm/cli/commands/mod.rs
deleted file mode 100644
index cdd3cb268..000000000
--- a/vm/cli/commands/mod.rs
+++ /dev/null
@@ -1,46 +0,0 @@
-// Copyright (c) 2019-2025 Provable Inc.
-// This file is part of the snarkVM library.
-
-// Licensed under the Apache License, Version 2.0 (the "License");
-// you may not use this file except in compliance with the License.
-// You may obtain a copy of the License at:
-
-// http://www.apache.org/licenses/LICENSE-2.0
-
-// Unless required by applicable law or agreed to in writing, software
-// distributed under the License is distributed on an "AS IS" BASIS,
-// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
-// See the License for the specific language governing permissions and
-// limitations under the License.
-
-pub mod clean;
-pub use clean::*;
-
-pub mod execute;
-pub use execute::*;
-
-pub mod new;
-pub use new::*;
-
-pub mod run;
-pub use run::*;
-
-pub mod update;
-pub use update::*;
-
-use crate::{
-    console::program::{Identifier, Locator, ProgramID, Value},
-    ledger::block::Transaction,
-    package::Package,
-};
-
-use anyhow::Result;
-use clap::Parser;
-use colored::Colorize;
-use core::str::FromStr;
-use std::collections::HashMap;
-
-pub const LOCALE: &num_format::Locale = &num_format::Locale::en;
-
-pub(crate) type CurrentNetwork = crate::prelude::MainnetV0;
-pub(crate) type Aleo = crate::circuit::AleoV0;
diff --git a/vm/cli/commands/new.rs b/vm/cli/commands/new.rs
deleted file mode 100644
index 4dc007918..000000000
--- a/vm/cli/commands/new.rs
+++ /dev/null
@@ -1,43 +0,0 @@
-// Copyright (c) 2019-2025 Provable Inc.
-// This file is part of the snarkVM library.
-
-// Licensed under the Apache License, Version 2.0 (the "License");
-// you may not use this file except in compliance with the License.
-// You may obtain a copy of the License at:
-
-// http://www.apache.org/licenses/LICENSE-2.0
-
-// Unless required by applicable law or agreed to in writing, software
-// distributed under the License is distributed on an "AS IS" BASIS,
-// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
-// See the License for the specific language governing permissions and
-// limitations under the License.
-
-use super::*;
-
-/// Create a new Aleo package.
-#[derive(Debug, Parser)]
-pub struct New {
-    /// The program name.
-    name: String,
-}
-
-impl New {
-    /// Creates an Aleo package with the specified name.
-    pub fn parse(self) -> Result<String> {
-        // Derive the program directory path.
-        let mut path = std::env::current_dir()?;
-        path.push(&self.name);
-
-        // Create the program ID from the name.
-        let id = ProgramID::<CurrentNetwork>::from_str(&format!("{}.aleo", self.name))?;
-
-        // Create the package.
-        Package::create(&path, &id)?;
-
-        // Prepare the path string.
-        let path_string = format!("(in \"{}\")", path.display());
-
-        Ok(format!("✅ Created an Aleo program '{}' {}", self.name.bold(), path_string.dimmed()))
-    }
-}
diff --git a/vm/cli/commands/run.rs b/vm/cli/commands/run.rs
deleted file mode 100644
index 43caf8cb0..000000000
--- a/vm/cli/commands/run.rs
+++ /dev/null
@@ -1,129 +0,0 @@
-// Copyright (c) 2019-2025 Provable Inc.
-// This file is part of the snarkVM library.
-
-// Licensed under the Apache License, Version 2.0 (the "License");
-// you may not use this file except in compliance with the License.
-// You may obtain a copy of the License at:
-
-// http://www.apache.org/licenses/LICENSE-2.0
-
-// Unless required by applicable law or agreed to in writing, software
-// distributed under the License is distributed on an "AS IS" BASIS,
-// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
-// See the License for the specific language governing permissions and
-// limitations under the License.
-
-use super::*;
-
-use anyhow::Context;
-
-/// Runs an Aleo program function
-#[derive(Debug, Parser)]
-pub struct Run {
-    /// The function name.
-    function: Identifier<CurrentNetwork>,
-    /// The function inputs.
-    inputs: Vec<Value<CurrentNetwork>>,
-}
-
-impl Run {
-    /// Compiles an Aleo program function with the specified name.
-    #[allow(clippy::format_in_format_args)]
-    pub fn parse(self) -> Result<String> {
-        // Derive the program directory path.
-        let path = std::env::current_dir()?;
-
-        // Load the package.
-        let package = Package::open(&path)?;
-        // Load the private key.
-        let private_key = crate::cli::helpers::dotenv_private_key()?;
-
-        // Initialize an RNG.
-        let rng = &mut rand::thread_rng();
-
-        // Execute the request.
-        let (response, metrics) =
-            package.run::<Aleo, _>(&private_key, self.function, &self.inputs, rng).with_context(|| {
-                format!("Running {}:{} with inputs {:?} failed", package.program_id(), &self.function, &self.inputs)
-            })?;
-
-        // Count the number of times a function is called.
-        let mut program_frequency = HashMap::<String, usize>::new();
-        for metric in metrics.iter() {
-            // Prepare the function name string.
-            let function_name_string = format!("'{}/{}'", metric.program_id, metric.function_name).bold();
-
-            // Prepare the function constraints string
-            let function_constraints_string = format!(
-                "{function_name_string} - {} constraints",
-                metric.num_function_constraints.to_formatted_string(LOCALE)
-            );
-
-            // Increment the counter for the function call.
-            match program_frequency.get_mut(&function_constraints_string) {
-                Some(counter) => *counter += 1,
-                None => {
-                    let _ = program_frequency.insert(function_constraints_string, 1);
-                }
-            }
-        }
-
-        // Log the metrics.
-        use num_format::ToFormattedString;
-
-        println!("⛓  Constraints\n");
-        for (function_constraints, counter) in program_frequency {
-            // Log the constraints
-            let counter_string = match counter {
-                1 => "(called 1 time)".to_string().dimmed(),
-                counter => format!("(called {counter} times)").dimmed(),
-            };
-
-            println!(" •  {function_constraints} {counter_string}",)
-        }
-
-        // Log the outputs.
-        match response.outputs().len() {
-            0 => (),
-            1 => println!("\n➡️  Output\n"),
-            _ => println!("\n➡️  Outputs\n"),
-        };
-        for output in response.outputs() {
-            println!("{}", format!(" • {output}"));
-        }
-        println!();
-
-        // Prepare the locator.
-        let locator = Locator::<CurrentNetwork>::from_str(&format!("{}/{}", package.program_id(), self.function))?;
-        // Prepare the path string.
-        let path_string = format!("(in \"{}\")", path.display());
-
-        Ok(format!("✅ Finished '{}' {}", locator.to_string().bold(), path_string.dimmed()))
-    }
-}
-
-#[cfg(test)]
-mod tests {
-    use super::*;
-    use crate::{
-        cli::{CLI, Command},
-        prelude::{Identifier, Value},
-    };
-
-    #[test]
-    fn clap_snarkvm_run() {
-        let arg_vec = vec!["snarkvm", "run", "hello", "1u32", "2u32", "foo.aleo"];
-        let cli = CLI::parse_from(&arg_vec);
-
-        if let Command::Run(run) = cli.command {
-            assert_eq!(run.function, Identifier::try_from(arg_vec[2]).unwrap());
-            assert_eq!(run.inputs, vec![
-                Value::try_from(arg_vec[3]).unwrap(),
-                Value::try_from(arg_vec[4]).unwrap(),
-                Value::try_from(arg_vec[5]).unwrap()
-            ]);
-        } else {
-            panic!("Unexpected result of clap parsing!");
-        }
-    }
-}
diff --git a/vm/cli/commands/update.rs b/vm/cli/commands/update.rs
deleted file mode 100644
index 7cb1adc04..000000000
--- a/vm/cli/commands/update.rs
+++ /dev/null
@@ -1,58 +0,0 @@
-// Copyright (c) 2019-2025 Provable Inc.
-// This file is part of the snarkVM library.
-
-// Licensed under the Apache License, Version 2.0 (the "License");
-// you may not use this file except in compliance with the License.
-// You may obtain a copy of the License at:
-
-// http://www.apache.org/licenses/LICENSE-2.0
-
-// Unless required by applicable law or agreed to in writing, software
-// distributed under the License is distributed on an "AS IS" BASIS,
-// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
-// See the License for the specific language governing permissions and
-// limitations under the License.
-
-use super::*;
-use crate::cli::helpers::Updater;
-
-/// Update SnarkVM to the latest version
-#[derive(Debug, Parser)]
-pub struct Update {
-    /// Lists all available versions of Aleo
-    #[clap(short = 'l', long)]
-    list: bool,
-    /// Suppress outputs to terminal
-    #[clap(short = 'q', long)]
-    quiet: bool,
-}
-
-impl Update {
-    pub fn parse(self) -> Result<String> {
-        match self.list {
-            true => match Updater::show_available_releases() {
-                Ok(output) => Ok(output),
-                Err(error) => Ok(format!("Failed to list the available versions of Aleo\n{error}\n")),
-            },
-            false => {
-                let result = Updater::update_to_latest_release(!self.quiet);
-                if !self.quiet {
-                    match result {
-                        Ok(status) => {
-                            if status.uptodate() {
-                                Ok("\nAleo is already on the latest version".to_string())
-                            } else if status.updated() {
-                                Ok(format!("\nAleo has updated to version {}", status.version()))
-                            } else {
-                                Ok(String::new())
-                            }
-                        }
-                        Err(e) => Ok(format!("\nFailed to update Aleo to the latest version\n{e}\n")),
-                    }
-                } else {
-                    Ok(String::new())
-                }
-            }
-        }
-    }
-}
diff --git a/vm/cli/errors.rs b/vm/cli/errors.rs
deleted file mode 100644
index 1a00e0a5b..000000000
--- a/vm/cli/errors.rs
+++ /dev/null
@@ -1,29 +0,0 @@
-// Copyright (c) 2019-2025 Provable Inc.
-// This file is part of the snarkVM library.
-
-// Licensed under the Apache License, Version 2.0 (the "License");
-// you may not use this file except in compliance with the License.
-// You may obtain a copy of the License at:
-
-// http://www.apache.org/licenses/LICENSE-2.0
-
-// Unless required by applicable law or agreed to in writing, software
-// distributed under the License is distributed on an "AS IS" BASIS,
-// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
-// See the License for the specific language governing permissions and
-// limitations under the License.
-
-#[derive(Debug, Error)]
-pub enum UpdaterError {
-    #[error("{}: {}", _0, _1)]
-    Crate(&'static str, String),
-
-    #[error("The current version {} is more recent than the release version {}", _0, _1)]
-    OldReleaseVersion(String, String),
-}
-
-impl From<self_update::errors::Error> for UpdaterError {
-    fn from(error: self_update::errors::Error) -> Self {
-        UpdaterError::Crate("self_update", error.to_string())
-    }
-}
diff --git a/vm/cli/helpers/env.rs b/vm/cli/helpers/env.rs
deleted file mode 100644
index dfd10c830..000000000
--- a/vm/cli/helpers/env.rs
+++ /dev/null
@@ -1,54 +0,0 @@
-// Copyright (c) 2019-2025 Provable Inc.
-// This file is part of the snarkVM library.
-
-// Licensed under the Apache License, Version 2.0 (the "License");
-// you may not use this file except in compliance with the License.
-// You may obtain a copy of the License at:
-
-// http://www.apache.org/licenses/LICENSE-2.0
-
-// Unless required by applicable law or agreed to in writing, software
-// distributed under the License is distributed on an "AS IS" BASIS,
-// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
-// See the License for the specific language governing permissions and
-// limitations under the License.
-
-use crate::{cli::CurrentNetwork, console::account::PrivateKey};
-
-use anyhow::{Result, anyhow};
-
-fn env_template() -> String {
-    r#"
-NETWORK=mainnet
-PRIVATE_KEY={{PASTE_YOUR_PRIVATE_KEY_HERE}}
-"#
-    .to_string()
-}
-
-/// Loads the environment variables from the .env file.
-fn dotenv_load() -> Result<()> {
-    // Load environment variables from .env file.
-    // Fails if .env file not found, not readable or invalid.
-    dotenvy::dotenv().map_err(|_| {
-        anyhow!(
-            "Missing a '.env' file. Create the '.env' file in your package's root directory with the following:\n\n{}\n",
-            env_template()
-        )
-    })?;
-    Ok(())
-}
-
-/// Returns the private key from the environment.
-pub fn dotenv_private_key() -> Result<PrivateKey<CurrentNetwork>> {
-    if cfg!(test) {
-        let rng = &mut crate::utilities::TestRng::fixed(123456789);
-        PrivateKey::<CurrentNetwork>::new(rng)
-    } else {
-        use std::str::FromStr;
-        dotenv_load()?;
-        // Load the private key from the environment.
-        let private_key = dotenvy::var("PRIVATE_KEY").map_err(|e| anyhow!("Missing PRIVATE_KEY - {e}"))?;
-        // Parse the private key.
-        PrivateKey::<CurrentNetwork>::from_str(&private_key)
-    }
-}
diff --git a/vm/cli/helpers/mod.rs b/vm/cli/helpers/mod.rs
deleted file mode 100644
index e7298afa0..000000000
--- a/vm/cli/helpers/mod.rs
+++ /dev/null
@@ -1,20 +0,0 @@
-// Copyright (c) 2019-2025 Provable Inc.
-// This file is part of the snarkVM library.
-
-// Licensed under the Apache License, Version 2.0 (the "License");
-// you may not use this file except in compliance with the License.
-// You may obtain a copy of the License at:
-
-// http://www.apache.org/licenses/LICENSE-2.0
-
-// Unless required by applicable law or agreed to in writing, software
-// distributed under the License is distributed on an "AS IS" BASIS,
-// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
-// See the License for the specific language governing permissions and
-// limitations under the License.
-
-pub mod env;
-pub use env::*;
-
-pub mod updater;
-pub use updater::*;
diff --git a/vm/cli/helpers/updater.rs b/vm/cli/helpers/updater.rs
deleted file mode 100644
index e15682ce1..000000000
--- a/vm/cli/helpers/updater.rs
+++ /dev/null
@@ -1,92 +0,0 @@
-// Copyright (c) 2019-2025 Provable Inc.
-// This file is part of the snarkVM library.
-
-// Licensed under the Apache License, Version 2.0 (the "License");
-// you may not use this file except in compliance with the License.
-// You may obtain a copy of the License at:
-
-// http://www.apache.org/licenses/LICENSE-2.0
-
-// Unless required by applicable law or agreed to in writing, software
-// distributed under the License is distributed on an "AS IS" BASIS,
-// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
-// See the License for the specific language governing permissions and
-// limitations under the License.
-
-use crate::cli::errors::UpdaterError;
-
-use anyhow::Result;
-use colored::Colorize;
-use self_update::{Status, backends::github, version::bump_is_greater};
-
-pub struct Updater;
-
-// TODO Add logic for users to easily select release versions.
-impl Updater {
-    const SNARKVM_BIN_NAME: &'static str = "snarkvm";
-    const SNARKVM_REPO_NAME: &'static str = "snarkvm";
-    const SNARKVM_REPO_OWNER: &'static str = "ProvableHQ";
-
-    /// Show all available releases for `snarkvm`.
-    #[allow(clippy::format_push_string)]
-    pub fn show_available_releases() -> Result<String> {
-        let releases = github::ReleaseList::configure()
-            .repo_owner(Self::SNARKVM_REPO_OWNER)
-            .repo_name(Self::SNARKVM_REPO_NAME)
-            .build()?
-            .fetch()?;
-
-        let mut output = "List of available versions\n".to_string();
-        for release in releases {
-            output += &format!("  * {}\n", release.version);
-        }
-        Ok(output)
-    }
-
-    /// Update `snarkvm` to the latest release.
-    pub fn update_to_latest_release(show_output: bool) -> Result<Status> {
-        let status = github::Update::configure()
-            .repo_owner(Self::SNARKVM_REPO_OWNER)
-            .repo_name(Self::SNARKVM_REPO_NAME)
-            .bin_name(Self::SNARKVM_BIN_NAME)
-            .current_version(env!("CARGO_PKG_VERSION"))
-            .show_download_progress(show_output)
-            .no_confirm(true)
-            .show_output(show_output)
-            .build()?
-            .update()?;
-
-        Ok(status)
-    }
-
-    /// Check if there is an available update for `snarkvm` and return the newest release.
-    pub fn update_available() -> Result<String, UpdaterError> {
-        let updater = github::Update::configure()
-            .repo_owner(Self::SNARKVM_REPO_OWNER)
-            .repo_name(Self::SNARKVM_REPO_NAME)
-            .bin_name(Self::SNARKVM_BIN_NAME)
-            .current_version(env!("CARGO_PKG_VERSION"))
-            .build()?;
-
-        let current_version = updater.current_version();
-        let latest_release = updater.get_latest_release()?;
-
-        if bump_is_greater(&current_version, &latest_release.version)? {
-            Ok(latest_release.version)
-        } else {
-            Err(UpdaterError::OldReleaseVersion(current_version, latest_release.version))
-        }
-    }
-
-    /// Returns the message to print on the CLI, if any.
-    pub fn print_cli() -> Option<String> {
-        if let Ok(latest_version) = Self::update_available() {
-            let mut output = "🟢 A new version is available! Run".bold().green().to_string();
-            output += &" `aleo update` ".bold().white();
-            output += &format!("to update to v{latest_version}.").bold().green();
-            Some(output)
-        } else {
-            None
-        }
-    }
-}
diff --git a/vm/cli/main.rs b/vm/cli/main.rs
deleted file mode 100644
index adfe0053a..000000000
--- a/vm/cli/main.rs
+++ /dev/null
@@ -1,65 +0,0 @@
-// Copyright (c) 2019-2025 Provable Inc.
-// This file is part of the snarkVM library.
-
-// Licensed under the Apache License, Version 2.0 (the "License");
-// you may not use this file except in compliance with the License.
-// You may obtain a copy of the License at:
-
-// http://www.apache.org/licenses/LICENSE-2.0
-
-// Unless required by applicable law or agreed to in writing, software
-// distributed under the License is distributed on an "AS IS" BASIS,
-// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
-// See the License for the specific language governing permissions and
-// limitations under the License.
-
-use snarkvm::{
-    cli::{CLI, Updater},
-    utilities::display_error,
-};
-
-use clap::Parser;
-
-use std::panic::catch_unwind;
-
-fn main() {
-    // Parse the given arguments.
-    let cli = CLI::parse();
-    // Run the updater.
-    if let Some(msg) = Updater::print_cli() {
-        println!("{msg}");
-    }
-
-    // Set a custom hook here to show "pretty" errors when panicking.
-    std::panic::set_hook(Box::new(|err| {
-        eprintln!("⚠️ {}", err.to_string().replace("panicked at", "snarkVM encountered an unexpected error at"));
-    }));
-
-    // Run the CLI.
-    // We use `catch_unwind` here to ensure a panic stops execution and not just a single thread.
-    // Note: `catch_unwind` can be nested without problems.
-    let result = catch_unwind(|| cli.command.parse());
-
-    // Process any errors (including panics).
-    match result {
-        Ok(Ok(output)) => {
-            println!("{output}\n");
-            std::process::exit(0);
-        }
-        Ok(Err(err)) => {
-            // A regular error occurred.
-            display_error(&err);
-            eprintln!();
-            eprintln!("Use `--help` for instructions on how to use this command");
-            std::process::exit(1);
-        }
-        Err(_) => {
-            eprintln!();
-            eprintln!("This is most likely a bug!");
-            eprintln!(
-                "Please report it to the snarkVM developers: https://github.com/ProvableHQ/snarkVM/issues/new?template=bug.md"
-            );
-            std::process::exit(1);
-        }
-    }
-}
diff --git a/vm/cli/mod.rs b/vm/cli/mod.rs
deleted file mode 100644
index 0f41eedb5..000000000
--- a/vm/cli/mod.rs
+++ /dev/null
@@ -1,26 +0,0 @@
-// Copyright (c) 2019-2025 Provable Inc.
-// This file is part of the snarkVM library.
-
-// Licensed under the Apache License, Version 2.0 (the "License");
-// you may not use this file except in compliance with the License.
-// You may obtain a copy of the License at:
-
-// http://www.apache.org/licenses/LICENSE-2.0
-
-// Unless required by applicable law or agreed to in writing, software
-// distributed under the License is distributed on an "AS IS" BASIS,
-// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
-// See the License for the specific language governing permissions and
-// limitations under the License.
-
-mod cli;
-pub use cli::*;
-
-mod commands;
-pub use commands::*;
-
-mod errors;
-pub use errors::*;
-
-pub mod helpers;
-pub use helpers::*;
diff --git a/vm/lib.rs b/vm/lib.rs
index ed4c83580..a330abbef 100644
--- a/vm/lib.rs
+++ b/vm/lib.rs
@@ -17,12 +17,6 @@
 #![allow(clippy::module_inception)]
 #![cfg_attr(test, allow(clippy::assertions_on_result_states))]
 
-#[cfg(feature = "cli")]
-#[macro_use]
-extern crate thiserror;
-
-#[cfg(feature = "cli")]
-pub mod cli;
 #[cfg(feature = "file")]
 pub mod file;
 #[cfg(feature = "package")]
diff --git a/vm/package/mod.rs b/vm/package/mod.rs
index 5d296e319..6db8fb20c 100644
--- a/vm/package/mod.rs
+++ b/vm/package/mod.rs
@@ -207,12 +207,50 @@ pub(crate) mod test_helpers {
 
     use std::{fs::File, io::Write};
 
+    use anyhow::anyhow;
+
     type CurrentNetwork = MainnetV0;
 
     fn temp_dir() -> PathBuf {
         tempfile::tempdir().expect("Failed to open temporary directory").keep()
     }
 
+    fn env_template() -> String {
+        r#"
+    NETWORK=mainnet
+    PRIVATE_KEY={{PASTE_YOUR_PRIVATE_KEY_HERE}}
+    "#
+        .to_string()
+    }
+
+    /// Loads the environment variables from the .env file.
+    fn dotenv_load() -> Result<()> {
+        // Load environment variables from .env file.
+        // Fails if .env file not found, not readable or invalid.
+        dotenvy::dotenv().map_err(|_| {
+            anyhow!(
+                "Missing a '.env' file. Create the '.env' file in your package's root directory with the following:\n\n{}\n",
+                env_template()
+            )
+        })?;
+        Ok(())
+    }
+
+    /// Returns the private key from the environment.
+    fn dotenv_private_key() -> Result<PrivateKey<CurrentNetwork>> {
+        if cfg!(test) {
+            let rng = &mut crate::utilities::TestRng::fixed(123456789);
+            PrivateKey::<CurrentNetwork>::new(rng)
+        } else {
+            use std::str::FromStr;
+            dotenv_load()?;
+            // Load the private key from the environment.
+            let private_key = dotenvy::var("PRIVATE_KEY").map_err(|e| anyhow!("Missing PRIVATE_KEY - {e}"))?;
+            // Parse the private key.
+            PrivateKey::<CurrentNetwork>::from_str(&private_key)
+        }
+    }
+
     /// Samples a (temporary) package containing a `token.aleo` program.
     pub(crate) fn sample_token_package() -> (PathBuf, Package<CurrentNetwork>) {
         // Initialize the program.
@@ -429,7 +467,7 @@ function main:
         match program_id.to_string().as_str() {
             "token.aleo" => {
                 // Sample a random private key.
-                let private_key = crate::cli::helpers::dotenv_private_key().unwrap();
+                let private_key = dotenv_private_key().unwrap();
                 let caller = Address::try_from(&private_key).unwrap();
 
                 // Initialize the function name.
@@ -443,7 +481,7 @@ function main:
             }
             "wallet.aleo" => {
                 // Initialize caller 0.
-                let caller0_private_key = crate::cli::helpers::dotenv_private_key().unwrap();
+                let caller0_private_key = dotenv_private_key().unwrap();
                 let caller0 = Address::try_from(&caller0_private_key).unwrap();
 
                 // Initialize caller 1.
@@ -465,7 +503,7 @@ function main:
             }
             "grandparent.aleo" => {
                 // Initialize caller 0.
-                let caller0_private_key = crate::cli::helpers::dotenv_private_key().unwrap();
+                let caller0_private_key = dotenv_private_key().unwrap();
 
                 // Initialize caller 1.
                 let caller1_private_key = PrivateKey::<CurrentNetwork>::new(rng).unwrap();
