diff --git a/synthesizer/program/src/closure/mod.rs b/synthesizer/program/src/closure/mod.rs
index 71879da70..aad825d1e 100644
--- a/synthesizer/program/src/closure/mod.rs
+++ b/synthesizer/program/src/closure/mod.rs
@@ -83,7 +83,7 @@ impl<N: Network, Instruction: InstructionTrait<N>> ClosureCore<N, Instruction> {
         ensure!(self.outputs.is_empty(), "Cannot add inputs after outputs have been added");
 
         // Ensure the maximum number of inputs has not been exceeded.
-        ensure!(self.inputs.len() <= N::MAX_INPUTS, "Cannot add more than {} inputs", N::MAX_INPUTS);
+        ensure!(self.inputs.len() < N::MAX_INPUTS, "Cannot add more than {} inputs", N::MAX_INPUTS);
         // Ensure the input statement was not previously added.
         ensure!(!self.inputs.contains(&input), "Cannot add duplicate input statement");
 
@@ -147,3 +147,78 @@ impl<N: Network, Instruction: InstructionTrait<N>> TypeName for ClosureCore<N, I
         "closure"
     }
 }
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    use crate::{Closure, Instruction};
+
+    type CurrentNetwork = console::network::Testnet3;
+
+    #[test]
+    fn test_add_input() {
+        // Initialize a new closure instance.
+        let name = Identifier::from_str("closure_core_test").unwrap();
+        let mut closure = Closure::<CurrentNetwork>::new(name);
+
+        // Ensure that an input can be added.
+        let input = Input::<CurrentNetwork>::from_str("input r0 as field;").unwrap();
+        assert!(closure.add_input(input.clone()).is_ok());
+
+        // Ensure that adding a duplicate input will fail.
+        assert!(closure.add_input(input).is_err());
+
+        // Ensure that adding more than the maximum number of inputs will fail.
+        for i in 1..CurrentNetwork::MAX_INPUTS * 2 {
+            let input = Input::<CurrentNetwork>::from_str(&format!("input r{i} as field;")).unwrap();
+
+            match closure.inputs.len() < CurrentNetwork::MAX_INPUTS {
+                true => assert!(closure.add_input(input).is_ok()),
+                false => assert!(closure.add_input(input).is_err()),
+            }
+        }
+    }
+
+    #[test]
+    fn test_add_instruction() {
+        // Initialize a new closure instance.
+        let name = Identifier::from_str("closure_core_test").unwrap();
+        let mut closure = Closure::<CurrentNetwork>::new(name);
+
+        // Ensure that an instruction can be added.
+        let instruction = Instruction::<CurrentNetwork>::from_str("add r0 r1 into r2;").unwrap();
+        assert!(closure.add_instruction(instruction.clone()).is_ok());
+
+        // Ensure that adding more than the maximum number of instructions will fail.
+        for i in 3..CurrentNetwork::MAX_INSTRUCTIONS * 2 {
+            let instruction = Instruction::<CurrentNetwork>::from_str(&format!("add r0 r1 into r{i};")).unwrap();
+
+            match closure.instructions.len() < CurrentNetwork::MAX_INSTRUCTIONS {
+                true => assert!(closure.add_instruction(instruction).is_ok()),
+                false => assert!(closure.add_instruction(instruction).is_err()),
+            }
+        }
+    }
+
+    #[test]
+    fn test_add_output() {
+        // Initialize a new closure instance.
+        let name = Identifier::from_str("closure_core_test").unwrap();
+        let mut closure = Closure::<CurrentNetwork>::new(name);
+
+        // Ensure that an output can be added.
+        let output = Output::<CurrentNetwork>::from_str("output r0 as field;").unwrap();
+        assert!(closure.add_output(output.clone()).is_ok());
+
+        // Ensure that adding more than the maximum number of outputs will fail.
+        for i in 1..CurrentNetwork::MAX_OUTPUTS * 2 {
+            let output = Output::<CurrentNetwork>::from_str(&format!("output r{i} as field;")).unwrap();
+
+            match closure.outputs.len() < CurrentNetwork::MAX_OUTPUTS {
+                true => assert!(closure.add_output(output).is_ok()),
+                false => assert!(closure.add_output(output).is_err()),
+            }
+        }
+    }
+}
diff --git a/synthesizer/program/src/finalize/mod.rs b/synthesizer/program/src/finalize/mod.rs
index ba3e4f395..2a6f57510 100644
--- a/synthesizer/program/src/finalize/mod.rs
+++ b/synthesizer/program/src/finalize/mod.rs
@@ -92,7 +92,7 @@ impl<N: Network, Command: CommandTrait<N>> FinalizeCore<N, Command> {
         ensure!(self.commands.is_empty(), "Cannot add inputs after commands have been added");
 
         // Ensure the maximum number of inputs has not been exceeded.
-        ensure!(self.inputs.len() <= N::MAX_INPUTS, "Cannot add more than {} inputs", N::MAX_INPUTS);
+        ensure!(self.inputs.len() < N::MAX_INPUTS, "Cannot add more than {} inputs", N::MAX_INPUTS);
         // Ensure the input statement was not previously added.
         ensure!(!self.inputs.contains(&input), "Cannot add duplicate input statement");
 
@@ -161,3 +161,109 @@ impl<N: Network, Command: CommandTrait<N>> TypeName for FinalizeCore<N, Command>
         "finalize"
     }
 }
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    use crate::{Command, Finalize};
+
+    type CurrentNetwork = console::network::Testnet3;
+
+    #[test]
+    fn test_add_input() {
+        // Initialize a new finalize instance.
+        let name = Identifier::from_str("finalize_core_test").unwrap();
+        let mut finalize = Finalize::<CurrentNetwork>::new(name);
+
+        // Ensure that an input can be added.
+        let input = Input::<CurrentNetwork>::from_str("input r0 as field.public;").unwrap();
+        assert!(finalize.add_input(input.clone()).is_ok());
+
+        // Ensure that adding a duplicate input will fail.
+        assert!(finalize.add_input(input).is_err());
+
+        // Ensure that adding more than the maximum number of inputs will fail.
+        for i in 1..CurrentNetwork::MAX_INPUTS * 2 {
+            let input = Input::<CurrentNetwork>::from_str(&format!("input r{i} as field.public;")).unwrap();
+
+            match finalize.inputs.len() < CurrentNetwork::MAX_INPUTS {
+                true => assert!(finalize.add_input(input).is_ok()),
+                false => assert!(finalize.add_input(input).is_err()),
+            }
+        }
+    }
+
+    #[test]
+    fn test_add_command() {
+        // Initialize a new finalize instance.
+        let name = Identifier::from_str("finalize_core_test").unwrap();
+        let mut finalize = Finalize::<CurrentNetwork>::new(name);
+
+        // Ensure that a command can be added.
+        let command = Command::<CurrentNetwork>::from_str("add r0 r1 into r2;").unwrap();
+        assert!(finalize.add_command(command.clone()).is_ok());
+
+        // Ensure that adding more than the maximum number of commands will fail.
+        for i in 3..CurrentNetwork::MAX_COMMANDS * 2 {
+            let command = Command::<CurrentNetwork>::from_str(&format!("add r0 r1 into r{i};")).unwrap();
+
+            match finalize.commands.len() < CurrentNetwork::MAX_COMMANDS {
+                true => assert!(finalize.add_command(command).is_ok()),
+                false => assert!(finalize.add_command(command).is_err()),
+            }
+        }
+
+        // Ensure that adding more than the maximum number of writes will fail.
+
+        // Initialize a new finalize instance.
+        let name = Identifier::from_str("finalize_core_test").unwrap();
+        let mut finalize = Finalize::<CurrentNetwork>::new(name);
+
+        for _ in 0..CurrentNetwork::MAX_WRITES * 2 {
+            let command = Command::<CurrentNetwork>::from_str("remove object[r0];").unwrap();
+
+            match finalize.commands.len() < CurrentNetwork::MAX_WRITES as usize {
+                true => assert!(finalize.add_command(command).is_ok()),
+                false => assert!(finalize.add_command(command).is_err()),
+            }
+        }
+    }
+
+    #[test]
+    fn test_add_command_duplicate_positions() {
+        // Initialize a new finalize instance.
+        let name = Identifier::from_str("finalize_core_test").unwrap();
+        let mut finalize = Finalize::<CurrentNetwork>::new(name);
+
+        // Ensure that a command can be added.
+        let command = Command::<CurrentNetwork>::from_str("position start;").unwrap();
+        assert!(finalize.add_command(command.clone()).is_ok());
+
+        // Ensure that adding a duplicate position will fail.
+        assert!(finalize.add_command(command).is_err());
+
+        // Helper method to convert a number to a unique string.
+        #[allow(clippy::cast_possible_truncation)]
+        fn to_unique_string(mut n: usize) -> String {
+            let mut s = String::new();
+            while n > 0 {
+                s.push((b'A' + (n % 26) as u8) as char);
+                n /= 26;
+            }
+            s.chars().rev().collect::<String>()
+        }
+
+        // Ensure that adding more than the maximum number of positions will fail.
+        for i in 1..u8::MAX as usize * 2 {
+            let position = to_unique_string(i);
+            println!("position: {}", position);
+            let command = Command::<CurrentNetwork>::from_str(&format!("position {position};")).unwrap();
+
+            match finalize.commands.len() < u8::MAX as usize {
+                true => assert!(finalize.add_command(command).is_ok()),
+                false => assert!(finalize.add_command(command).is_err()),
+            }
+        }
+    }
+}
diff --git a/synthesizer/program/src/function/mod.rs b/synthesizer/program/src/function/mod.rs
index 083375bc6..9f03a3c4a 100644
--- a/synthesizer/program/src/function/mod.rs
+++ b/synthesizer/program/src/function/mod.rs
@@ -201,3 +201,78 @@ impl<N: Network, Instruction: InstructionTrait<N>, Command: CommandTrait<N>> Typ
         "function"
     }
 }
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    use crate::{Function, Instruction};
+
+    type CurrentNetwork = console::network::Testnet3;
+
+    #[test]
+    fn test_add_input() {
+        // Initialize a new function instance.
+        let name = Identifier::from_str("function_core_test").unwrap();
+        let mut function = Function::<CurrentNetwork>::new(name);
+
+        // Ensure that an input can be added.
+        let input = Input::<CurrentNetwork>::from_str("input r0 as field.private;").unwrap();
+        assert!(function.add_input(input.clone()).is_ok());
+
+        // Ensure that adding a duplicate input will fail.
+        assert!(function.add_input(input).is_err());
+
+        // Ensure that adding more than the maximum number of inputs will fail.
+        for i in 1..CurrentNetwork::MAX_INPUTS * 2 {
+            let input = Input::<CurrentNetwork>::from_str(&format!("input r{i} as field.private;")).unwrap();
+
+            match function.inputs.len() < CurrentNetwork::MAX_INPUTS {
+                true => assert!(function.add_input(input).is_ok()),
+                false => assert!(function.add_input(input).is_err()),
+            }
+        }
+    }
+
+    #[test]
+    fn test_add_instruction() {
+        // Initialize a new function instance.
+        let name = Identifier::from_str("function_core_test").unwrap();
+        let mut function = Function::<CurrentNetwork>::new(name);
+
+        // Ensure that an instruction can be added.
+        let instruction = Instruction::<CurrentNetwork>::from_str("add r0 r1 into r2;").unwrap();
+        assert!(function.add_instruction(instruction.clone()).is_ok());
+
+        // Ensure that adding more than the maximum number of instructions will fail.
+        for i in 3..CurrentNetwork::MAX_INSTRUCTIONS * 2 {
+            let instruction = Instruction::<CurrentNetwork>::from_str(&format!("add r0 r1 into r{i};")).unwrap();
+
+            match function.instructions.len() < CurrentNetwork::MAX_INSTRUCTIONS {
+                true => assert!(function.add_instruction(instruction).is_ok()),
+                false => assert!(function.add_instruction(instruction).is_err()),
+            }
+        }
+    }
+
+    #[test]
+    fn test_add_output() {
+        // Initialize a new function instance.
+        let name = Identifier::from_str("function_core_test").unwrap();
+        let mut function = Function::<CurrentNetwork>::new(name);
+
+        // Ensure that an output can be added.
+        let output = Output::<CurrentNetwork>::from_str("output r0 as field.private;").unwrap();
+        assert!(function.add_output(output.clone()).is_ok());
+
+        // Ensure that adding more than the maximum number of outputs will fail.
+        for i in 1..CurrentNetwork::MAX_OUTPUTS * 2 {
+            let output = Output::<CurrentNetwork>::from_str(&format!("output r{i} as field.private;")).unwrap();
+
+            match function.outputs.len() < CurrentNetwork::MAX_OUTPUTS {
+                true => assert!(function.add_output(output).is_ok()),
+                false => assert!(function.add_output(output).is_err()),
+            }
+        }
+    }
+}
