diff --git a/node/bft/src/sync/mod.rs b/node/bft/src/sync/mod.rs
index f8c7bbc98..976e489a8 100644
--- a/node/bft/src/sync/mod.rs
+++ b/node/bft/src/sync/mod.rs
@@ -296,7 +296,7 @@ impl<N: Network> Sync<N> {
         Ok(())
     }
 
-    /// Syncs the storage with the given blocks.
+    /// Syncs the storage with blocks already received from peers.
     pub async fn sync_storage_with_blocks(&self) -> Result<()> {
         // Acquire the response lock.
         let _lock = self.response_lock.lock().await;
diff --git a/node/sync/src/block_sync.rs b/node/sync/src/block_sync.rs
index 7fe9534b2..7096c0452 100644
--- a/node/sync/src/block_sync.rs
+++ b/node/sync/src/block_sync.rs
@@ -78,11 +78,11 @@ impl BlockSyncMode {
 ///
 /// # State
 /// - When a request is inserted, the `requests` map and `request_timestamps` map insert an entry for the request height.
-/// - When a response is inserted, the `requests` map inserts the entry for the request height.
+/// - When a response is inserted, the `responses` map inserts the entry for the request height.
 /// - When a request is completed, the `requests` map still has the entry, but its `sync_ips` is empty;
 ///   the `request_timestamps` map remains unchanged.
 /// - When a response is removed/completed, the `requests` map and `request_timestamps` map also remove the entry for the request height.
-/// - When a request is timed out, the `requests`, `request_timestamps`, and `responses` map remove the entry for the request height;
+/// - When a request is timed out, the `requests`, `request_timestamps`, and `responses` map remove the entry for the request height.
 #[derive(Clone, Debug)]
 pub struct BlockSync<N: Network> {
     /// The block sync mode.
