diff --git a/Cargo.lock b/Cargo.lock
index cd23bc99b..0ff164c20 100644
--- a/Cargo.lock
+++ b/Cargo.lock
@@ -2661,9 +2661,9 @@ checksum = "c307a32c1c5c437f38c7fd45d753050587732ba8628319fbdf12a7e289ccc590"
 
 [[package]]
 name = "smallvec"
-version = "1.6.1"
+version = "1.7.0"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "fe0f37c9e8f3c5a4a66ad655a93c74daac4ad00c441533bf5c6e7990bb42604e"
+checksum = "1ecab6c735a6bb4139c0caafd0cc3635748bbb3acf4550e8138122099251f309"
 
 [[package]]
 name = "snarkos"
diff --git a/benchmarks/Cargo.toml b/benchmarks/Cargo.toml
index e7dd7ac08..f7810a70c 100644
--- a/benchmarks/Cargo.toml
+++ b/benchmarks/Cargo.toml
@@ -108,7 +108,7 @@ version = "0.8"
 version = "1"
 
 [dependencies.smallvec]
-version = "1.6"
+version = "1.7"
 
 [dependencies.criterion]
 version = "0.3.4"
diff --git a/consensus/Cargo.toml b/consensus/Cargo.toml
index 72b601444..91804b238 100644
--- a/consensus/Cargo.toml
+++ b/consensus/Cargo.toml
@@ -69,7 +69,7 @@ features = [ "serde" ]
 version = "0.4.2"
 
 [dependencies.smallvec]
-version = "1.6"
+version = "1.7"
 
 [dependencies.rand]
 version = "0.8"
diff --git a/network/src/peers/peers.rs b/network/src/peers/peers.rs
index 6aadf130f..52d9b7070 100644
--- a/network/src/peers/peers.rs
+++ b/network/src/peers/peers.rs
@@ -20,12 +20,13 @@ use std::{
     time::{Duration, Instant},
 };
 
+use itertools::Itertools;
 use rand::{prelude::SliceRandom, rngs::SmallRng, seq::IteratorRandom, SeedableRng};
 use tokio::task;
 
 use snarkos_metrics::{self as metrics, connections::*};
 
-use crate::{message::*, KnownNetworkMessage, NetworkError, Node, NodeType};
+use crate::{message::*, KnownNetworkMessage, NetworkError, Node, NodeType, Peer};
 use anyhow::*;
 
 impl Node {
@@ -224,7 +225,7 @@ impl Node {
         // Local address must be known by now.
         let own_address = self.expect_local_addr();
 
-        let random_peers = {
+        let random_peers: Vec<SocketAddr> = {
             trace!(
                 "Connecting to {} disconnected peers",
                 cmp::min(count, self.peer_book.get_disconnected_peer_count() as usize)
@@ -242,14 +243,30 @@ impl Node {
                 candidates.sort_unstable_by_key(|peer| peer.quality.last_connected);
             }
 
-            let addr_iter = candidates.iter().map(|peer| peer.address);
-
             if !self.is_of_type(NodeType::Client) {
-                addr_iter.take(count).collect()
+                candidates.into_iter().take(count).collect()
             } else {
-                addr_iter.choose_multiple(&mut SmallRng::from_entropy(), count)
+                // Floored if count is odd.
+                let random_count = count / 2;
+                let random_picks: Vec<Peer> = candidates
+                    .iter()
+                    .choose_multiple(&mut SmallRng::from_entropy(), random_count)
+                    .into_iter()
+                    .cloned()
+                    .collect();
+                candidates.sort_unstable_by(|x, y| y.quality.block_height.cmp(&x.quality.block_height));
+
+                candidates.truncate(count - random_count);
+                candidates
+                    .into_iter()
+                    .chain(random_picks.into_iter())
+                    .unique_by(|x| x.address)
+                    .collect::<Vec<Peer>>()
             }
-        };
+        }
+        .iter()
+        .map(|peer| peer.address)
+        .collect();
 
         for remote_address in random_peers {
             let node = self.clone();
@@ -323,8 +340,6 @@ impl Node {
     pub(crate) async fn send_peers(&self, remote_address: SocketAddr, time_received: Option<Instant>) {
         // Broadcast the sanitized list of connected peers back to the requesting peer.
 
-        use crate::Peer;
-
         let connected_peers = self.peer_book.connected_peers_snapshot().await;
 
         let basic_filter = |peer: &Peer| peer.address != remote_address;
diff --git a/network/src/sync/blocks/aggro.rs b/network/src/sync/blocks/aggro.rs
index f90ce3fa0..78b1bc8c8 100644
--- a/network/src/sync/blocks/aggro.rs
+++ b/network/src/sync/blocks/aggro.rs
@@ -14,24 +14,19 @@
 // You should have received a copy of the GNU General Public License
 // along with the snarkOS library. If not, see <https://www.gnu.org/licenses/>.
 
-use std::{sync::Arc, time::Duration};
+use std::sync::Arc;
 
 use crate::{Cache, Node, Payload, Peer, SyncBase, SyncInbound};
 use anyhow::*;
 use snarkos_metrics::wrapped_mpsc;
 use snarkos_storage::Digest;
-use tokio::{sync::RwLock, time::Instant};
+use tokio::sync::RwLock;
 
 /// Aggressive, continuous sync process that pulls peers entire canon trees.
 pub struct SyncAggro {
     base: SyncBase,
 }
 
-struct BlockLocatorHashes {
-    hashes: Vec<Digest>,
-    last_update: Instant,
-}
-
 impl SyncAggro {
     pub fn new(node: Node) -> (Self, wrapped_mpsc::Sender<SyncInbound>) {
         let (base, sender) = SyncBase::new(node);
@@ -64,16 +59,12 @@ impl SyncAggro {
 
         self.base.node.register_block_sync_attempt();
 
-        let block_locator_hashes = Arc::new(RwLock::new(BlockLocatorHashes {
-            hashes: SyncBase::block_locator_hashes(&self.base.node).await?,
-            last_update: Instant::now(),
-        }));
+        let block_locator_hashes = SyncBase::block_locator_hashes(&self.base.node).await?;
 
         let peer_syncs = {
-            let block_locator_hashes = block_locator_hashes.read().await;
             sync_nodes
                 .iter()
-                .map(|peer| (peer.clone(), block_locator_hashes.hashes.clone()))
+                .map(|peer| (peer.clone(), block_locator_hashes.clone()))
                 .collect()
         };
 
@@ -87,7 +78,7 @@ impl SyncAggro {
 
         let node = self.base.node.clone();
         self.base
-            .receive_messages(15, 3, |msg| {
+            .receive_messages(60, 3, |msg| {
                 match msg {
                     SyncInbound::BlockHashes(peer, hashes) => {
                         debug!("received {} sync hashes from {}", hashes.len(), peer);
@@ -98,20 +89,8 @@ impl SyncAggro {
                         let last_hash = hashes.last().unwrap().clone();
 
                         let node = node.clone();
-                        let block_locator_hashes = block_locator_hashes.clone();
                         let received_hashes = received_hashes.clone();
                         tokio::spawn(async move {
-                            if block_locator_hashes.read().await.last_update.elapsed() > Duration::from_secs(10) {
-                                match SyncBase::block_locator_hashes(&node).await {
-                                    Ok(hashes) => {
-                                        let mut target = block_locator_hashes.write().await;
-                                        target.hashes = hashes;
-                                        target.last_update = Instant::now();
-                                    }
-                                    Err(e) => warn!("sync failed to fetch block locator hashes: {:?}", e),
-                                }
-                            }
-
                             let mut hashes_trimmed = Vec::with_capacity(hashes.len());
                             {
                                 let received_hashes = received_hashes.read().await;
@@ -136,15 +115,15 @@ impl SyncAggro {
                                 .filter(|(_, status)| matches!(status, snarkos_storage::BlockStatus::Unknown))
                                 .map(|(hash, _)| hash)
                                 .collect();
-                            if blocks.is_empty() {
-                                return;
-                            }
                             debug!("requesting {} sync blocks from {}", blocks.len(), peer);
 
                             if let Some(peer) = node.peer_book.get_peer_handle(peer) {
                                 peer.expecting_sync_blocks(blocks.len() as u32).await;
-                                peer.send_payload(Payload::GetBlocks(blocks), None).await;
                                 peer.send_payload(Payload::GetSync(vec![last_hash]), None).await;
+                                if blocks.is_empty() {
+                                    return;
+                                }
+                                peer.send_payload(Payload::GetBlocks(blocks), None).await;
                             }
                         });
                     }
diff --git a/storage/Cargo.toml b/storage/Cargo.toml
index daf7f160c..a7778e210 100644
--- a/storage/Cargo.toml
+++ b/storage/Cargo.toml
@@ -53,7 +53,7 @@ version = "2.2.0"
 version = "0.9"
 
 [dependencies.smallvec]
-version = "1.6"
+version = "1.7"
 features = [ "write" ]
 
 [dependencies.anyhow]
diff --git a/storage/src/digest.rs b/storage/src/digest.rs
index ab6ffdddd..c2f42e90e 100644
--- a/storage/src/digest.rs
+++ b/storage/src/digest.rs
@@ -16,7 +16,7 @@
 
 use std::{fmt, io, ops::Deref};
 
-use serde::Serialize;
+use serde::{Deserialize, Serialize};
 use smallvec::SmallVec;
 use snarkvm_utilities::{ToBytes, Write};
 
@@ -45,6 +45,16 @@ impl Serialize for Digest {
     }
 }
 
+impl<'de> Deserialize<'de> for Digest {
+    fn deserialize<D: serde::Deserializer<'de>>(deserializer: D) -> Result<Self, D::Error> {
+        let name = String::deserialize(deserializer)?;
+
+        Ok(Self(InnerType::from(
+            hex::decode(&name).map_err(|e| serde::de::Error::custom(e))?,
+        )))
+    }
+}
+
 impl Deref for Digest {
     type Target = SmallVec<[u8; 64]>;
 
diff --git a/storage/src/storage/sync.rs b/storage/src/storage/sync.rs
index db0396df6..da216b6af 100644
--- a/storage/src/storage/sync.rs
+++ b/storage/src/storage/sync.rs
@@ -246,40 +246,16 @@ pub trait SyncStorage {
         }
 
         // Calculate the average distance between block hashes based on the desired number of locator hashes.
-        let mut proportional_step =
+        let proportional_step =
             (hash_index.min(oldest_fork_threshold as u32) / num_locator_hashes).min(crate::NUM_LOCATOR_HASHES - 1);
 
         // Provide hashes of blocks with indices descending quadratically while the quadratic step distance is
         // lower or close to the proportional step distance.
-        let num_quadratic_steps = (proportional_step as f32).log2() as u32;
+        let num_quadratic_steps = ((hash_index as f32).log2() - (proportional_step as f32).log2()) as u32;
 
         // The remaining hashes should have a proportional index distance between them.
         let num_proportional_steps = num_locator_hashes - num_quadratic_steps;
 
-        // Obtain a few hashes increasing the distance quadratically.
-        let mut quadratic_step = 2; // the size of the first quadratic step
-        for _ in 0..num_quadratic_steps {
-            let hash = self.get_block_hash_guarded(hash_index)?;
-            trace!("block locator hash -- quadratic: block# {}: {}", hash_index, hash);
-            block_locator_hashes.push(hash);
-            hash_index = hash_index.saturating_sub(quadratic_step);
-            quadratic_step *= 2;
-        }
-
-        // Update the size of the proportional step so that the hashes of the remaining blocks have the same distance
-        // between one another.
-        proportional_step =
-            (hash_index.min(oldest_fork_threshold as u32) / num_locator_hashes).min(crate::NUM_LOCATOR_HASHES - 1);
-
-        // Tweak: in order to avoid "jumping" by too many indices with the last step,
-        // increase the value of each step by 1 if the last step is too large. This
-        // can result in the final number of locator hashes being a bit lower, but
-        // it's preferable to having a large gap between values.
-        if hash_index - proportional_step * num_proportional_steps > 2 * proportional_step {
-            proportional_step += 1;
-        }
-
-        // Obtain the rest of hashes with a proportional distance between them.
         for _ in 0..num_proportional_steps {
             let hash = self.get_block_hash_guarded(hash_index)?;
             trace!("block locator hash -- proportional: block# {}: {}", hash_index, hash);
@@ -290,6 +266,19 @@ pub trait SyncStorage {
             hash_index = hash_index.saturating_sub(proportional_step);
         }
 
+        // Obtain a few hashes increasing the distance quadratically.
+        let mut quadratic_step = proportional_step * 2; // the size of the first quadratic step
+        for _ in 0..num_quadratic_steps {
+            if hash_index == 0 {
+                break;
+            }
+            let hash = self.get_block_hash_guarded(hash_index)?;
+            trace!("block locator hash -- quadratic: block# {}: {}", hash_index, hash);
+            block_locator_hashes.push(hash);
+            hash_index = hash_index.saturating_sub(quadratic_step);
+            quadratic_step *= 2;
+        }
+
         let hash = self.get_block_hash_guarded(0)?;
         trace!("block locator hash -- genesis: block# {}: {}", 0, hash);
         block_locator_hashes.push(hash);
diff --git a/testing/src/network/sync.rs b/testing/src/network/sync.rs
index cdd8ac022..e3b42e2e8 100644
--- a/testing/src/network/sync.rs
+++ b/testing/src/network/sync.rs
@@ -83,12 +83,15 @@ async fn block_initiator_side() {
     let sync = Payload::Sync(block_header_hashes);
     peer.write_message(&sync).await;
 
+    assert!(matches!(peer.read_payload().await.unwrap(), Payload::GetSync(_)));
+
     // make sure both GetBlock messages are received
     let payload = peer.read_payload().await.unwrap();
+
     let block_hashes = if let Payload::GetBlocks(block_hashes) = payload {
         block_hashes
     } else {
-        unreachable!();
+        panic!("unexpected payload in test: {:?}", payload);
     };
 
     assert!(block_hashes.contains(&block_1_header_hash) && block_hashes.contains(&block_2_header_hash));
