diff --git a/circuit/program/src/data/literal/cast/integer.rs b/circuit/program/src/data/literal/cast/integer.rs
index 6c5aae4e5..78655df65 100644
--- a/circuit/program/src/data/literal/cast/integer.rs
+++ b/circuit/program/src/data/literal/cast/integer.rs
@@ -79,7 +79,7 @@ impl<E: Environment, I0: IntegerType, I1: IntegerType> Cast<Integer<E, I1>> for
                 // Then instantiate the new integer from the lower bits.
                 false => {
                     Boolean::assert_bits_are_zero(&bits_le[(I1::BITS.saturating_sub(1) as usize)..]);
-                    Integer::<E, I1>::from_bits_le(&bits_le[..(I1::BITS.saturating_sub(1) as usize)])
+                    Integer::<E, I1>::from_bits_le(&bits_le[..(I1::BITS as usize)])
                 }
             },
             // If the source type is signed and the destination type is unsigned, perform the required checks.
@@ -88,7 +88,7 @@ impl<E: Environment, I0: IntegerType, I1: IntegerType> Cast<Integer<E, I1>> for
                 // Then instantiate the new integer from the lower bits.
                 true => {
                     E::assert(!&bits_le[I0::BITS.saturating_sub(1) as usize]);
-                    Integer::<E, I1>::from_bits_le(&bits_le[..(I0::BITS.saturating_sub(1) as usize)])
+                    Integer::<E, I1>::from_bits_le(&bits_le)
                 }
                 // If the source type is larger than the destination type, check that the upper bits are zero.
                 // Then instantiate the new integer from the lower bits.
diff --git a/circuit/types/group/src/lib.rs b/circuit/types/group/src/lib.rs
index 0983de04d..a0f636b43 100644
--- a/circuit/types/group/src/lib.rs
+++ b/circuit/types/group/src/lib.rs
@@ -401,6 +401,56 @@ mod tests {
 
     #[test]
     fn test_is_in_group() {
+        type PrimitiveField = console::Field<<Circuit as Environment>::Network>;
+
+        // First test the points that have low order.
+
+        // The two halves of (0,1) in group arithmetic are (0,1) and (0,-1).
+        let minus1_string = "8444461749428370424248824938781546531375899335154063827935233455917409239040field";
+        // The two halves of (0,-1) in group arithmetic are (q1x,0) and (q2x,0),
+        // where q1x = sqrt(-1) mod F_q  and  q2x = -sqrt(-1) mod F_q.
+        let q1x_string = "880904806456922042258150504921383618666682042621506879489field";
+        let q2x_string = "8444461749428370423367920132324624489117748830232680209268551413295902359552field";
+
+        // (0,1) is in the large prime subgroup.
+        let y1: Field<Circuit> = Field::new(Mode::Public, PrimitiveField::from_str("1field").unwrap());
+        let group0 = Group::<Circuit>::from_xy_coordinates_unchecked(Field::zero(), y1);
+        Circuit::scope("group0", || {
+            let group0_is_in_group = group0.is_in_group();
+            assert!(group0_is_in_group.eject_value());
+            assert_scope!(750, 0, 2253, 2253);
+        });
+        Circuit::reset();
+
+        // The other three low order points are on the curve but not in the large prime subgroup.
+        // Make sure is_in_group returns false for these.
+        let minus1: Field<Circuit> = Field::new(Mode::Public, PrimitiveField::from_str(minus1_string).unwrap());
+        let half0 = Group::<Circuit>::from_xy_coordinates_unchecked(Field::zero(), minus1);
+        Circuit::scope("half0", || {
+            let half0_is_not_in_group = !half0.is_in_group();
+            assert!(half0_is_not_in_group.eject_value());
+            assert_scope!(750, 0, 2253, 2253);
+        });
+        Circuit::reset();
+
+        let q1x: Field<Circuit> = Field::new(Mode::Public, PrimitiveField::from_str(q1x_string).unwrap());
+        let quarter1 = Group::<Circuit>::from_xy_coordinates_unchecked(q1x, Field::zero());
+        Circuit::scope("quarter1", || {
+            let quarter1_is_not_in_group = !quarter1.is_in_group();
+            assert!(quarter1_is_not_in_group.eject_value());
+            assert_scope!(750, 0, 2253, 2253);
+        });
+        Circuit::reset();
+
+        let q2x: Field<Circuit> = Field::new(Mode::Public, PrimitiveField::from_str(q2x_string).unwrap());
+        let quarter2 = Group::<Circuit>::from_xy_coordinates_unchecked(q2x, Field::zero());
+        Circuit::scope("quarter2", || {
+            let quarter2_is_not_in_group = !quarter2.is_in_group();
+            assert!(quarter2_is_not_in_group.eject_value());
+            assert_scope!(750, 0, 2253, 2253);
+        });
+        Circuit::reset();
+
         fn check_is_in_group(mode: Mode, num_constants: u64, num_public: u64, num_private: u64, num_constraints: u64) {
             let mut rng = TestRng::default();
 
diff --git a/synthesizer/process/src/stack/finalize_types/initialize.rs b/synthesizer/process/src/stack/finalize_types/initialize.rs
index e36fe3d55..25bb0c7ca 100644
--- a/synthesizer/process/src/stack/finalize_types/initialize.rs
+++ b/synthesizer/process/src/stack/finalize_types/initialize.rs
@@ -660,7 +660,27 @@ impl<N: Network> FinalizeTypes<N> {
                         }
                     }
                 }
-                "cast.lossy" => bail!("Instruction '{instruction}' is not supported yet."),
+                "cast.lossy" => {
+                    // Retrieve the cast operation.
+                    let operation = match instruction {
+                        Instruction::CastLossy(operation) => operation,
+                        _ => bail!("Instruction '{instruction}' is not a cast.lossy operation."),
+                    };
+
+                    // Ensure the instruction has one destination register.
+                    ensure!(
+                        instruction.destinations().len() == 1,
+                        "Instruction '{instruction}' has multiple destinations."
+                    );
+
+                    // Ensure the casted register type is valid and defined.
+                    match operation.cast_type() {
+                        CastType::Plaintext(PlaintextType::Literal(_)) => {
+                            ensure!(instruction.operands().len() == 1, "Expected 1 operand.");
+                        }
+                        _ => bail!("`cast.lossy` is only supported for casting to a literal type."),
+                    }
+                }
                 _ => bail!("Instruction '{instruction}' is not for opcode '{opcode}'."),
             },
             Opcode::Command(opcode) => {
diff --git a/synthesizer/process/src/stack/register_types/initialize.rs b/synthesizer/process/src/stack/register_types/initialize.rs
index 7b3265a1f..abfec5a48 100644
--- a/synthesizer/process/src/stack/register_types/initialize.rs
+++ b/synthesizer/process/src/stack/register_types/initialize.rs
@@ -554,7 +554,27 @@ impl<N: Network> RegisterTypes<N> {
                         }
                     }
                 }
-                "cast.lossy" => bail!("Instruction '{instruction}' is not supported yet."),
+                "cast.lossy" => {
+                    // Retrieve the cast operation.
+                    let operation = match instruction {
+                        Instruction::CastLossy(operation) => operation,
+                        _ => bail!("Instruction '{instruction}' is not a cast.lossy operation."),
+                    };
+
+                    // Ensure the instruction has one destination register.
+                    ensure!(
+                        instruction.destinations().len() == 1,
+                        "Instruction '{instruction}' has multiple destinations."
+                    );
+
+                    // Ensure the casted register type is valid and defined.
+                    match operation.cast_type() {
+                        CastType::Plaintext(PlaintextType::Literal(_)) => {
+                            ensure!(instruction.operands().len() == 1, "Expected 1 operand.");
+                        }
+                        _ => bail!("`cast.lossy` is only supported for casting to a literal type."),
+                    }
+                }
                 _ => bail!("Instruction '{instruction}' is not for opcode '{opcode}'."),
             },
             Opcode::Command(opcode) => {
diff --git a/synthesizer/process/src/trace/mod.rs b/synthesizer/process/src/trace/mod.rs
index 988be7ad2..69789b89f 100644
--- a/synthesizer/process/src/trace/mod.rs
+++ b/synthesizer/process/src/trace/mod.rs
@@ -314,9 +314,6 @@ impl<N: Network> Trace<N> {
             verifier_inputs.push((verifying_key, batch_inclusion_inputs));
         }
         // Verify the proof.
-        match VerifyingKey::verify_batch(locator, verifier_inputs, proof) {
-            true => Ok(()),
-            false => bail!("Failed to verify proof"),
-        }
+        VerifyingKey::verify_batch(locator, verifier_inputs, proof).map_err(|e| anyhow!("Failed to verify proof - {e}"))
     }
 }
diff --git a/synthesizer/program/src/logic/instruction/operation/cast.rs b/synthesizer/program/src/logic/instruction/operation/cast.rs
index a516daf42..2cf60e79a 100644
--- a/synthesizer/program/src/logic/instruction/operation/cast.rs
+++ b/synthesizer/program/src/logic/instruction/operation/cast.rs
@@ -206,9 +206,12 @@ impl<N: Network, const VARIANT: u8> CastOperation<N, VARIANT> {
         stack: &(impl StackMatches<N> + StackProgram<N>),
         registers: &mut (impl RegistersSigner<N> + RegistersLoad<N> + RegistersStore<N>),
     ) -> Result<()> {
-        // TODO (howardwu & d0cd): Re-enable after stabilizing.
+        // If the variant is `cast.lossy`, then check that the `cast_type` is a `PlaintextType::Literal`.
         if VARIANT == CastVariant::CastLossy as u8 {
-            bail!("cast.lossy is not supported (yet)")
+            ensure!(
+                matches!(self.cast_type, CastType::Plaintext(PlaintextType::Literal(..))),
+                "`cast.lossy` is only supported for casting to a literal type"
+            )
         }
 
         // Load the operands values.
@@ -333,9 +336,12 @@ impl<N: Network, const VARIANT: u8> CastOperation<N, VARIANT> {
         stack: &(impl StackMatches<N> + StackProgram<N>),
         registers: &mut (impl RegistersSignerCircuit<N, A> + RegistersLoadCircuit<N, A> + RegistersStoreCircuit<N, A>),
     ) -> Result<()> {
-        // TODO (howardwu & d0cd): Re-enable after stabilizing.
+        // If the variant is `cast.lossy`, then check that the `cast_type` is a `PlaintextType::Literal`.
         if VARIANT == CastVariant::CastLossy as u8 {
-            bail!("cast.lossy is not supported (yet)")
+            ensure!(
+                matches!(self.cast_type, CastType::Plaintext(PlaintextType::Literal(..))),
+                "`cast.lossy` is only supported for casting to a literal type"
+            )
         }
 
         use circuit::{Eject, Inject};
@@ -582,9 +588,12 @@ impl<N: Network, const VARIANT: u8> CastOperation<N, VARIANT> {
         stack: &(impl StackMatches<N> + StackProgram<N>),
         registers: &mut (impl RegistersLoad<N> + RegistersStore<N>),
     ) -> Result<()> {
-        // TODO (howardwu & d0cd): Re-enable after stabilizing.
+        // If the variant is `cast.lossy`, then check that the `cast_type` is a `PlaintextType::Literal`.
         if VARIANT == CastVariant::CastLossy as u8 {
-            bail!("cast.lossy is not supported (yet)")
+            ensure!(
+                matches!(self.cast_type, CastType::Plaintext(PlaintextType::Literal(..))),
+                "`cast.lossy` is only supported for casting to a literal type"
+            )
         }
 
         // Load the operands values.
@@ -641,6 +650,14 @@ impl<N: Network, const VARIANT: u8> CastOperation<N, VARIANT> {
         stack: &impl StackProgram<N>,
         input_types: &[RegisterType<N>],
     ) -> Result<Vec<RegisterType<N>>> {
+        // If the variant is `cast.lossy`, then check that the `cast_type` is a `PlaintextType::Literal`.
+        if VARIANT == CastVariant::CastLossy as u8 {
+            ensure!(
+                matches!(self.cast_type, CastType::Plaintext(PlaintextType::Literal(..))),
+                "`cast.lossy` is only supported for casting to a literal type"
+            )
+        }
+
         // Ensure the number of operands is correct.
         ensure!(
             input_types.len() == self.operands.len(),
diff --git a/synthesizer/snark/src/proving_key/mod.rs b/synthesizer/snark/src/proving_key/mod.rs
index 07b3cd9d8..5f13807c3 100644
--- a/synthesizer/snark/src/proving_key/mod.rs
+++ b/synthesizer/snark/src/proving_key/mod.rs
@@ -65,10 +65,12 @@ impl<N: Network> ProvingKey<N> {
         let timer = std::time::Instant::now();
 
         // Prepare the instances.
+        let num_expected_instances = assignments.len();
         let instances: BTreeMap<_, _> = assignments
             .iter()
             .map(|(proving_key, assignments)| (proving_key.deref(), assignments.as_slice()))
             .collect();
+        ensure!(instances.len() == num_expected_instances, "Incorrect number of proving keys for batch proof");
 
         // Retrieve the proving parameters.
         let universal_prover = N::varuna_universal_prover();
diff --git a/synthesizer/snark/src/verifying_key/mod.rs b/synthesizer/snark/src/verifying_key/mod.rs
index 700d79108..a9494acf7 100644
--- a/synthesizer/snark/src/verifying_key/mod.rs
+++ b/synthesizer/snark/src/verifying_key/mod.rs
@@ -61,13 +61,19 @@ impl<N: Network> VerifyingKey<N> {
 
     /// Returns `true` if the batch proof is valid for the given public inputs.
     #[allow(clippy::type_complexity)]
-    pub fn verify_batch(locator: &str, inputs: Vec<(VerifyingKey<N>, Vec<Vec<N::Field>>)>, proof: &Proof<N>) -> bool {
+    pub fn verify_batch(
+        locator: &str,
+        inputs: Vec<(VerifyingKey<N>, Vec<Vec<N::Field>>)>,
+        proof: &Proof<N>,
+    ) -> Result<()> {
         #[cfg(feature = "aleo-cli")]
         let timer = std::time::Instant::now();
 
         // Convert the instances.
+        let num_expected_keys = inputs.len();
         let keys_to_inputs: BTreeMap<_, _> =
             inputs.iter().map(|(verifying_key, inputs)| (verifying_key.deref(), inputs.as_slice())).collect();
+        ensure!(keys_to_inputs.len() == num_expected_keys, "Incorrect number of verifying keys for batch proof");
 
         // Retrieve the verification parameters.
         let universal_verifier = N::varuna_universal_verifier();
@@ -77,13 +83,16 @@ impl<N: Network> VerifyingKey<N> {
         match Varuna::<N>::verify_batch(universal_verifier, fiat_shamir, &keys_to_inputs, proof) {
             Ok(is_valid) => {
                 #[cfg(feature = "aleo-cli")]
-                println!("{}", format!(" • Verified '{locator}' (in {} ms)", timer.elapsed().as_millis()).dimmed());
-                is_valid
+                println!(
+                    "{}",
+                    format!(" • Verified '{locator}': {is_valid} (in {} ms)", timer.elapsed().as_millis()).dimmed()
+                );
+                if is_valid { Ok(()) } else { bail!("'verify_batch' failed") }
             }
             Err(error) => {
                 #[cfg(feature = "aleo-cli")]
                 println!("{}", format!(" • Verifier failed: {error}").dimmed());
-                false
+                bail!(error)
             }
         }
     }
diff --git a/synthesizer/tests/expectations/process/execute/lossy_casts.out b/synthesizer/tests/expectations/process/execute/lossy_casts.out
new file mode 100644
index 000000000..0a3a6279b
--- /dev/null
+++ b/synthesizer/tests/expectations/process/execute/lossy_casts.out
@@ -0,0 +1,62 @@
+errors: []
+outputs:
+- - 93754672984613362290390329026198290544u128
+  - 93754672984613362290390329026198290544i128
+  - 16982649547548967024u64
+  - -1464094526160584592i64
+  - 446349424u32
+  - 446349424i32
+  - 49264u16
+  - -16272i16
+  - 112u8
+  - 112i8
+  - 'false'
+  - 4764330560507562969049837239139509827871889319573451509694277586124473432540group
+- - 0u128
+  - 0i128
+  - 0u64
+  - 0i64
+  - 0u32
+  - 0i32
+  - 0u16
+  - 0i16
+  - 0u8
+  - 0i8
+  - 'false'
+  - 0group
+- - 1u128
+  - 1i128
+  - 1u64
+  - 1i64
+  - 1u32
+  - 1i32
+  - 1u16
+  - 1i16
+  - 1u8
+  - 1i8
+  - 'true'
+  - 1540945439182663264862696551825005342995406165131907382295858612069623286213group
+- - 0u128
+  - 0i128
+  - 0u64
+  - 0i64
+  - 0u32
+  - 0i32
+  - 0u16
+  - 0i16
+  - 0u8
+  - 0i8
+  - 'false'
+  - 340282366920938463463374607431768211456group
+- - 1u128
+  - 1i128
+  - 1u64
+  - 1i64
+  - 1u32
+  - 1i32
+  - 1u16
+  - 1i16
+  - 1u8
+  - 1i8
+  - 'true'
+  - 2207376757013908562861143127796680436121374163712359508096962653679714303999group
diff --git a/synthesizer/tests/tests/process/execute/lossy_casts.aleo b/synthesizer/tests/tests/process/execute/lossy_casts.aleo
new file mode 100644
index 000000000..6513c1bb0
--- /dev/null
+++ b/synthesizer/tests/tests/process/execute/lossy_casts.aleo
@@ -0,0 +1,50 @@
+/*
+randomness: 902384
+cases:
+  - program: lossy_casts.aleo
+    function: test_lossy_cast
+    inputs: [79363714989903307245735717098563574705733591463163614225748337416674727625843187853442697973404985688481508350822field]
+  - program: lossy_casts.aleo
+    function: test_lossy_cast
+    inputs: [0field]
+  - program: lossy_casts.aleo
+    function: test_lossy_cast
+    inputs: [1field]
+  - program: lossy_casts.aleo
+    function: test_lossy_cast
+    inputs: [340_282_366_920_938_463_463_374_607_431_768_211_456field]
+  - program: lossy_casts.aleo
+    function: test_lossy_cast
+    inputs: [340_282_366_920_938_463_463_374_607_431_768_211_457field]
+
+
+*/
+
+program lossy_casts.aleo;
+
+function test_lossy_cast:
+    input r0 as field.private;
+    cast.lossy r0 into r1 as u128;
+    cast.lossy r0 into r2 as i128;
+    cast.lossy r0 into r3 as u64;
+    cast.lossy r0 into r4 as i64;
+    cast.lossy r0 into r5 as u32;
+    cast.lossy r0 into r6 as i32;
+    cast.lossy r0 into r7 as u16;
+    cast.lossy r0 into r8 as i16;
+    cast.lossy r0 into r9 as u8;
+    cast.lossy r0 into r10 as i8;
+    cast.lossy r0 into r11 as boolean;
+    cast.lossy r0 into r12 as group;
+    output r1 as u128.private;
+    output r2 as i128.private;
+    output r3 as u64.private;
+    output r4 as i64.private;
+    output r5 as u32.private;
+    output r6 as i32.private;
+    output r7 as u16.private;
+    output r8 as i16.private;
+    output r9 as u8.private;
+    output r10 as i8.private;
+    output r11 as boolean.private;
+    output r12 as group.private;
