diff --git a/synthesizer/program/src/lib.rs b/synthesizer/program/src/lib.rs
index 26f9b7c2b..6e32c776f 100644
--- a/synthesizer/program/src/lib.rs
+++ b/synthesizer/program/src/lib.rs
@@ -575,6 +575,7 @@ impl<N: Network, Instruction: InstructionTrait<N>, Command: CommandTrait<N>> Pro
         "u64",
         "u128",
         "scalar",
+        "signature",
         "string",
         // Boolean
         "true",
