diff --git a/Cargo.lock b/Cargo.lock
index 798ed3926..a491fee20 100644
--- a/Cargo.lock
+++ b/Cargo.lock
@@ -2666,6 +2666,7 @@ name = "snarkvm-rest"
 version = "0.9.0"
 dependencies = [
  "anyhow",
+ "http",
  "indexmap",
  "parking_lot 0.12.1",
  "serde",
diff --git a/circuit/types/string/src/equal.rs b/circuit/types/string/src/equal.rs
index a30031784..60a20a0f6 100644
--- a/circuit/types/string/src/equal.rs
+++ b/circuit/types/string/src/equal.rs
@@ -25,28 +25,15 @@ impl<E: Environment> Equal<Self> for StringType<E> {
         let this = self.to_fields();
         let that = other.to_fields();
 
-        // Return `false` if the length of the strings are equal.
-        if this.len() != that.len() {
-            return Boolean::constant(false);
-        }
-
-        // Check if the string contents are equal.
-        this.iter().zip_eq(&that).fold(Boolean::constant(true), |acc, (a, b)| acc & a.is_equal(b))
+        // Check that the size in bytes of the two strings are equal.
+        self.size_in_bytes.is_equal(&other.size_in_bytes)
+            // Check that the string contents are equal.
+            & this.iter().zip(&that).fold(Boolean::constant(true), |acc, (a, b)| acc & a.is_equal(b))
     }
 
     /// Returns `true` if `self` and `other` are *not* equal.
     fn is_not_equal(&self, other: &Self) -> Self::Output {
-        // Convert each string type into fields.
-        let this = self.to_fields();
-        let that = other.to_fields();
-
-        // Return `true` if the length of the strings are *not* equal.
-        if this.len() != that.len() {
-            return Boolean::constant(true);
-        }
-
-        // Check if the string contents are *not* equal.
-        this.iter().zip_eq(&that).fold(Boolean::constant(false), |acc, (a, b)| acc | a.is_not_equal(b))
+        !self.is_equal(other)
     }
 }
 
@@ -123,31 +110,31 @@ mod tests {
 
     #[test]
     fn test_is_equal_constant() -> Result<()> {
-        check_is_equal(Mode::Constant, 8, 0, 0, 0)
+        check_is_equal(Mode::Constant, 9, 0, 0, 0)
     }
 
     #[test]
     fn test_is_equal_public() -> Result<()> {
-        check_is_equal(Mode::Public, 0, 0, 23, 31)
+        check_is_equal(Mode::Public, 0, 0, 26, 35)
     }
 
     #[test]
     fn test_is_equal_private() -> Result<()> {
-        check_is_equal(Mode::Private, 0, 0, 23, 31)
+        check_is_equal(Mode::Private, 0, 0, 26, 35)
     }
 
     #[test]
     fn test_is_not_equal_constant() -> Result<()> {
-        check_is_not_equal(Mode::Constant, 8, 0, 0, 0)
+        check_is_not_equal(Mode::Constant, 9, 0, 0, 0)
     }
 
     #[test]
     fn test_is_not_equal_public() -> Result<()> {
-        check_is_not_equal(Mode::Public, 0, 0, 23, 31)
+        check_is_not_equal(Mode::Public, 0, 0, 26, 35)
     }
 
     #[test]
     fn test_is_not_equal_private() -> Result<()> {
-        check_is_not_equal(Mode::Private, 0, 0, 23, 31)
+        check_is_not_equal(Mode::Private, 0, 0, 26, 35)
     }
 }
diff --git a/circuit/types/string/src/helpers/from_bits.rs b/circuit/types/string/src/helpers/from_bits.rs
index 3b74fb8ec..ea3a2bb10 100644
--- a/circuit/types/string/src/helpers/from_bits.rs
+++ b/circuit/types/string/src/helpers/from_bits.rs
@@ -21,22 +21,51 @@ impl<E: Environment> FromBits for StringType<E> {
 
     /// Initializes a new string from a list of little-endian bits *with* trailing zeros (to byte-alignment).
     fn from_bits_le(bits_le: &[Self::Boolean]) -> Self {
-        // Ensure the list of booleans is byte-aligned.
-        let num_bits = bits_le.len();
-        match num_bits % 8 == 0 {
-            true => StringType { mode: bits_le.eject_mode(), bytes: bits_le.chunks(8).map(U8::from_bits_le).collect() },
-            false => E::halt(format!("Attempted to instantiate a {num_bits}-bit string, which is not byte-aligned")),
-        }
+        // Determine the mode.
+        let mode = bits_le.eject_mode();
+        // Inject the string size in bytes.
+        let size_in_bytes = Self::inject_size_in_bytes(mode, bits_le);
+        // Return the string.
+        StringType { mode, bytes: bits_le.chunks(8).map(U8::from_bits_le).collect(), size_in_bytes }
     }
 
     /// Initializes a new scalar field element from a list of big-endian bits *with* leading zeros (to byte-alignment).
     fn from_bits_be(bits_be: &[Self::Boolean]) -> Self {
-        // Ensure the list of booleans is byte-aligned.
-        let num_bits = bits_be.len();
-        match num_bits % 8 == 0 {
-            true => StringType { mode: bits_be.eject_mode(), bytes: bits_be.chunks(8).map(U8::from_bits_be).collect() },
-            false => E::halt(format!("Attempted to instantiate a {num_bits}-bit string, which is not byte-aligned")),
+        // Determine the mode.
+        let mode = bits_be.eject_mode();
+        // Inject the string size in bytes.
+        let size_in_bytes = Self::inject_size_in_bytes(mode, bits_be);
+        // Return the string.
+        StringType { mode, bytes: bits_be.chunks(8).map(U8::from_bits_be).collect(), size_in_bytes }
+    }
+}
+
+impl<E: Environment> StringType<E> {
+    /// Checks the size of the given bits for the given mode, and returns the size (of the string) in bytes.
+    /// "Load-bearing witness allocation - Please do not optimize me." - Pratyush :)
+    fn inject_size_in_bytes(mode: Mode, bits: &[Boolean<E>]) -> Field<E> {
+        // Ensure the bits are byte-aligned.
+        let num_bits = bits.len();
+        if num_bits % 8 != 0 {
+            E::halt(format!("Attempted to instantiate a {num_bits}-bit string, which is not byte-aligned"))
         }
+
+        // Cast the number of bytes in the 'string' as a field element.
+        let num_bytes =
+            console::Field::from_u32(u32::try_from(num_bits / 8).unwrap_or_else(|error| E::halt(error.to_string())));
+
+        // Inject the number of bytes as a constant.
+        let expected_size_in_bytes = Field::constant(num_bytes);
+        // Inject the number of bytes as a witness.
+        let size_in_bytes = match mode.is_constant() {
+            true => expected_size_in_bytes.clone(),
+            false => Field::new(Mode::Private, num_bytes),
+        };
+        // Ensure the witness matches the constant.
+        E::assert_eq(&expected_size_in_bytes, &size_in_bytes);
+
+        // Return the size in bytes.
+        size_in_bytes
     }
 }
 
@@ -91,31 +120,31 @@ mod tests {
 
     #[test]
     fn test_from_bits_le_constant() {
-        check_from_bits_le(Mode::Constant, 0, 0, 0, 0);
+        check_from_bits_le(Mode::Constant, 1, 0, 0, 0);
     }
 
     #[test]
     fn test_from_bits_le_public() {
-        check_from_bits_le(Mode::Public, 0, 0, 0, 0);
+        check_from_bits_le(Mode::Public, 1, 0, 1, 1);
     }
 
     #[test]
     fn test_from_bits_le_private() {
-        check_from_bits_le(Mode::Private, 0, 0, 0, 0);
+        check_from_bits_le(Mode::Private, 1, 0, 1, 1);
     }
 
     #[test]
     fn test_from_bits_be_constant() {
-        check_from_bits_be(Mode::Constant, 0, 0, 0, 0);
+        check_from_bits_be(Mode::Constant, 1, 0, 0, 0);
     }
 
     #[test]
     fn test_from_bits_be_public() {
-        check_from_bits_be(Mode::Public, 0, 0, 0, 0);
+        check_from_bits_be(Mode::Public, 1, 0, 1, 1);
     }
 
     #[test]
     fn test_from_bits_be_private() {
-        check_from_bits_be(Mode::Private, 0, 0, 0, 0);
+        check_from_bits_be(Mode::Private, 1, 0, 1, 1);
     }
 }
diff --git a/circuit/types/string/src/lib.rs b/circuit/types/string/src/lib.rs
index 50ec181a2..1878ae908 100644
--- a/circuit/types/string/src/lib.rs
+++ b/circuit/types/string/src/lib.rs
@@ -27,12 +27,14 @@ use snarkvm_circuit_environment::assert_scope;
 
 use snarkvm_circuit_environment::prelude::*;
 use snarkvm_circuit_types_boolean::Boolean;
+use snarkvm_circuit_types_field::Field;
 use snarkvm_circuit_types_integers::U8;
 
 #[derive(Clone)]
 pub struct StringType<E: Environment> {
     mode: Mode,
     bytes: Vec<U8<E>>,
+    size_in_bytes: Field<E>,
 }
 
 impl<E: Environment> StringTrait for StringType<E> {}
@@ -43,7 +45,27 @@ impl<E: Environment> Inject for StringType<E> {
 
     /// Initializes a new instance of a string.
     fn new(mode: Mode, string: Self::Primitive) -> Self {
-        Self { mode, bytes: string.as_bytes().iter().map(|byte| U8::new(mode, console::Integer::new(*byte))).collect() }
+        // Cast the number of bytes in the 'string' as a field element.
+        let num_bytes =
+            console::Field::from_u32(u32::try_from(string.len()).unwrap_or_else(|error| E::halt(error.to_string())));
+
+        // "Load-bearing witness allocation - Please do not optimize me." - Pratyush :)
+
+        // Inject the number of bytes as a constant.
+        let expected_size_in_bytes = Field::constant(num_bytes);
+        // Inject the number of bytes as a witness.
+        let size_in_bytes = match mode.is_constant() {
+            true => expected_size_in_bytes.clone(),
+            false => Field::new(Mode::Private, num_bytes),
+        };
+        // Ensure the witness matches the constant.
+        E::assert_eq(&expected_size_in_bytes, &size_in_bytes);
+
+        Self {
+            mode,
+            bytes: string.as_bytes().iter().map(|byte| U8::new(mode, console::Integer::new(*byte))).collect(),
+            size_in_bytes,
+        }
     }
 }
 
diff --git a/rest/Cargo.toml b/rest/Cargo.toml
index 68f33f153..9ff4af733 100644
--- a/rest/Cargo.toml
+++ b/rest/Cargo.toml
@@ -23,6 +23,9 @@ include = [ "Cargo.toml", "src", "README.md", "LICENSE.md" ]
 license = "GPL-3.0"
 edition = "2021"
 
+[dependencies.http]
+version = "0.2.8"
+
 [dependencies.snarkvm-compiler]
 path = "../vm/compiler"
 version = "0.9.0"
@@ -52,4 +55,4 @@ version = "1"
 version = "0.1"
 
 [dependencies.warp]
-version = "0.3"
+version = "0.3.3"
diff --git a/rest/src/start.rs b/rest/src/start.rs
index 254fdee54..a27fb5ac6 100644
--- a/rest/src/start.rs
+++ b/rest/src/start.rs
@@ -15,6 +15,7 @@
 // along with the snarkVM library. If not, see <https://www.gnu.org/licenses/>.
 
 use super::*;
+use http::header::HeaderName;
 
 impl<N: Network, B: 'static + BlockStorage<N>, P: 'static + ProgramStorage<N>> Server<N, B, P> {
     /// Initializes a new instance of the server.
@@ -45,6 +46,11 @@ impl<N: Network, B: 'static + BlockStorage<N>, P: 'static + ProgramStorage<N>> S
         additional_routes: Option<impl Filter<Extract = impl Reply, Error = Rejection> + Clone + Sync + Send + 'static>,
         custom_port: Option<u16>,
     ) {
+        let cors = warp::cors()
+            .allow_any_origin()
+            .allow_header(HeaderName::from_static("content-type"))
+            .allow_methods(vec!["GET", "POST", "OPTIONS"]);
+
         // Initialize the routes.
         let routes = self.routes();
         // Spawn the server.
@@ -57,8 +63,8 @@ impl<N: Network, B: 'static + BlockStorage<N>, P: 'static + ProgramStorage<N>> S
             println!("\n🌐 Server is running at http://0.0.0.0:{}\n", ip.1);
             // Start the server, with optional additional routes.
             match additional_routes {
-                Some(additional_routes) => warp::serve(routes.or(additional_routes)).run(ip).await,
-                None => warp::serve(routes).run(ip).await,
+                Some(additional_routes) => warp::serve(routes.or(additional_routes).with(cors)).run(ip).await,
+                None => warp::serve(routes.with(cors)).run(ip).await,
             }
         })))
     }
