diff --git a/vm/compiler/src/program/instruction/operation/assert.rs b/vm/compiler/src/program/instruction/operation/assert.rs
index c3c653dc4..7fcbb0aad 100644
--- a/vm/compiler/src/program/instruction/operation/assert.rs
+++ b/vm/compiler/src/program/instruction/operation/assert.rs
@@ -348,6 +348,8 @@ mod tests {
         mode_a: &circuit::Mode,
         mode_b: &circuit::Mode,
     ) {
+        println!("Checking '{opcode}' for '{literal_a}.{mode_a}' and '{literal_b}.{mode_b}'");
+
         // Initialize the types.
         let type_a = literal_a.to_type();
         let type_b = literal_b.to_type();
diff --git a/vm/compiler/src/program/instruction/operation/is.rs b/vm/compiler/src/program/instruction/operation/is.rs
index 259b7b4dd..e3a6dcb34 100644
--- a/vm/compiler/src/program/instruction/operation/is.rs
+++ b/vm/compiler/src/program/instruction/operation/is.rs
@@ -364,6 +364,8 @@ mod tests {
     ) {
         use circuit::Eject;
 
+        println!("Checking '{opcode}' for '{literal_a}.{mode_a}' and '{literal_b}.{mode_b}'");
+
         // Initialize the types.
         let type_a = literal_a.to_type();
         let type_b = literal_b.to_type();
diff --git a/vm/package/is_build_required.rs b/vm/package/is_build_required.rs
index beff9eea3..8c6a6780b 100644
--- a/vm/package/is_build_required.rs
+++ b/vm/package/is_build_required.rs
@@ -59,7 +59,7 @@ impl<N: Network> Package<N> {
             }
         }
 
-        // Package hasn't changed, no need to build
+        // Skip building the package, as it has not changed.
         false
     }
 }
