diff --git a/synthesizer/src/program/finalize/mod.rs b/synthesizer/src/program/finalize/mod.rs
index 6318ee9fa..3d7b9afe1 100644
--- a/synthesizer/src/program/finalize/mod.rs
+++ b/synthesizer/src/program/finalize/mod.rs
@@ -121,18 +121,6 @@ impl<N: Network> Finalize<N> {
             Command::Instruction(Instruction::HashManyPSD8(_)) => {
                 bail!("`hash_many.psd8` is not supported in finalize.")
             }
-            Command::Instruction(Instruction::HashToGroupBHP256(_)) => Ok(200_000),
-            Command::Instruction(Instruction::HashToGroupBHP512(_)) => Ok(100_000),
-            Command::Instruction(Instruction::HashToGroupBHP768(_)) => Ok(100_000),
-            Command::Instruction(Instruction::HashToGroupBHP1024(_)) => Ok(100_000),
-            Command::Instruction(Instruction::HashToGroupPED64(_)) => Ok(10_000),
-            Command::Instruction(Instruction::HashToGroupPED128(_)) => Ok(20_000),
-            Command::Instruction(Instruction::HashToGroupPSD2(_)) => Ok(600_000),
-            Command::Instruction(Instruction::HashToGroupPSD4(_)) => Ok(700_000),
-            Command::Instruction(Instruction::HashToGroupPSD8(_)) => Ok(800_000),
-            Command::Instruction(Instruction::HashToScalarPSD2(_)) => Ok(60_000),
-            Command::Instruction(Instruction::HashToScalarPSD4(_)) => Ok(100_000),
-            Command::Instruction(Instruction::HashToScalarPSD8(_)) => Ok(200_000),
             Command::Instruction(Instruction::Inv(_)) => Ok(10_000),
             Command::Instruction(Instruction::IsEq(_)) => Ok(2_000),
             Command::Instruction(Instruction::IsNeq(_)) => Ok(2_000),
