diff --git a/.circleci/config.yml b/.circleci/config.yml
index 7f1db3113..128556d2b 100644
--- a/.circleci/config.yml
+++ b/.circleci/config.yml
@@ -645,13 +645,43 @@ jobs:
           workspace_member: synthesizer/program
           cache_key: snarkvm-synthesizer-program-cache
 
-  synthesizer-program-integration:
+  synthesizer-program-integration-keccak:
     docker:
       - image: cimg/rust:1.71.1
     resource_class: 2xlarge
     steps:
       - run_serial:
-          flags: --test '*'
+          flags: keccak --test '*'
+          workspace_member: synthesizer/program
+          cache_key: snarkvm-synthesizer-program-cache
+
+  synthesizer-program-integration-psd:
+    docker:
+      - image: cimg/rust:1.71.1
+    resource_class: 2xlarge
+    steps:
+      - run_serial:
+          flags: psd --test '*'
+          workspace_member: synthesizer/program
+          cache_key: snarkvm-synthesizer-program-cache
+
+  synthesizer-program-integration-sha:
+    docker:
+      - image: cimg/rust:1.71.1
+    resource_class: 2xlarge
+    steps:
+      - run_serial:
+          flags: sha --test '*'
+          workspace_member: synthesizer/program
+          cache_key: snarkvm-synthesizer-program-cache
+
+  synthesizer-program-integration-rest:
+    docker:
+      - image: cimg/rust:1.71.1
+    resource_class: 2xlarge
+    steps:
+      - run_serial:
+          flags: --test '*' -- --skip keccak --skip psd --skip sha
           workspace_member: synthesizer/program
           cache_key: snarkvm-synthesizer-program-cache
 
@@ -822,7 +852,10 @@ workflows:
       - synthesizer-integration
       - synthesizer-process
       - synthesizer-program
-      - synthesizer-program-integration
+      - synthesizer-program-integration-keccak
+      - synthesizer-program-integration-psd
+      - synthesizer-program-integration-sha
+      - synthesizer-program-integration-rest
       - synthesizer-snark
       - utilities
       - utilities-derives
diff --git a/rust-toolchain b/rust-toolchain
new file mode 100644
index 000000000..883bde307
--- /dev/null
+++ b/rust-toolchain
@@ -0,0 +1 @@
+1.71.1
\ No newline at end of file
diff --git a/synthesizer/program/tests/instruction/commit.rs b/synthesizer/program/tests/instruction/commit.rs
index bb4175883..cf75f8ef3 100644
--- a/synthesizer/program/tests/instruction/commit.rs
+++ b/synthesizer/program/tests/instruction/commit.rs
@@ -41,7 +41,7 @@ use synthesizer_process::{Process, Stack};
 type CurrentNetwork = Testnet3;
 type CurrentAleo = AleoV0;
 
-const ITERATIONS: usize = 100;
+const ITERATIONS: usize = 50;
 
 /// **Attention**: When changing this, also update in `src/logic/instruction/commit.rs`.
 fn valid_destination_types() -> &'static [LiteralType] {
diff --git a/synthesizer/program/tests/instruction/hash.rs b/synthesizer/program/tests/instruction/hash.rs
index 46ad7ea7f..20c7cf44a 100644
--- a/synthesizer/program/tests/instruction/hash.rs
+++ b/synthesizer/program/tests/instruction/hash.rs
@@ -50,7 +50,7 @@ use synthesizer_process::{Process, Stack};
 type CurrentNetwork = Testnet3;
 type CurrentAleo = AleoV0;
 
-const ITERATIONS: usize = 100;
+const ITERATIONS: usize = 50;
 
 /// **Attention**: When changing this, also update in `src/logic/instruction/hash.rs`.
 fn valid_destination_types<N: Network>() -> &'static [PlaintextType<N>] {
@@ -191,7 +191,7 @@ fn check_hash<const VARIANT: u8>(
 }
 
 macro_rules! test_hash {
-        ($name: tt, $hash:ident) => {
+        ($name: tt, $hash:ident, $iterations:expr) => {
             paste::paste! {
                 #[test]
                 fn [<test _ $name _ is _ consistent>]() {
@@ -206,7 +206,7 @@ macro_rules! test_hash {
                     // Prepare the test.
                     let modes = [circuit::Mode::Public, circuit::Mode::Private];
 
-                    for _ in 0..ITERATIONS {
+                    for _ in 0..$iterations {
                         let literals = sample_literals!(CurrentNetwork, &mut rng);
                         for literal in literals.iter() {
                             for mode in modes.iter() {
@@ -227,22 +227,22 @@ macro_rules! test_hash {
         };
     }
 
-test_hash!(hash_bhp256, HashBHP256);
-test_hash!(hash_bhp512, HashBHP512);
-test_hash!(hash_bhp768, HashBHP768);
-test_hash!(hash_bhp1024, HashBHP1024);
+test_hash!(hash_bhp256, HashBHP256, ITERATIONS);
+test_hash!(hash_bhp512, HashBHP512, ITERATIONS);
+test_hash!(hash_bhp768, HashBHP768, ITERATIONS);
+test_hash!(hash_bhp1024, HashBHP1024, ITERATIONS);
 
-test_hash!(hash_keccak256, HashKeccak256);
-test_hash!(hash_keccak384, HashKeccak384);
-test_hash!(hash_keccak512, HashKeccak512);
+test_hash!(hash_keccak256, HashKeccak256, 5);
+test_hash!(hash_keccak384, HashKeccak384, 5);
+test_hash!(hash_keccak512, HashKeccak512, 5);
 
-test_hash!(hash_psd2, HashPSD2);
-test_hash!(hash_psd4, HashPSD4);
-test_hash!(hash_psd8, HashPSD8);
+test_hash!(hash_psd2, HashPSD2, ITERATIONS);
+test_hash!(hash_psd4, HashPSD4, ITERATIONS);
+test_hash!(hash_psd8, HashPSD8, ITERATIONS);
 
-test_hash!(hash_sha3_256, HashSha3_256);
-test_hash!(hash_sha3_384, HashSha3_384);
-test_hash!(hash_sha3_512, HashSha3_512);
+test_hash!(hash_sha3_256, HashSha3_256, 5);
+test_hash!(hash_sha3_384, HashSha3_384, 5);
+test_hash!(hash_sha3_512, HashSha3_512, 5);
 
 // Note this test must be explicitly written, instead of using the macro, because HashPED64 fails on certain input types.
 #[test]
