diff --git a/console/network/src/lib.rs b/console/network/src/lib.rs
index f6da37a8d..8ccd00a64 100644
--- a/console/network/src/lib.rs
+++ b/console/network/src/lib.rs
@@ -87,7 +87,7 @@ pub trait Network:
     const STARTING_SUPPLY: u64 = 1_100_000_000_000_000; // 1.1B credits
 
     /// The anchor time per block in seconds, which must be greater than the round time per block.
-    const ANCHOR_TIME: u16 = 20;
+    const ANCHOR_TIME: u16 = 25;
     /// The coinbase puzzle degree.
     const COINBASE_PUZZLE_DEGREE: u32 = (1 << 13) - 1; // 8,191
     /// The maximum number of prover solutions that can be included per block.
diff --git a/parameters/src/testnet3/genesis.rs b/parameters/src/testnet3/genesis.rs
index aca52b63b..e126be108 100644
--- a/parameters/src/testnet3/genesis.rs
+++ b/parameters/src/testnet3/genesis.rs
@@ -29,6 +29,6 @@ mod tests {
     #[test]
     fn test_genesis_block() {
         let bytes = GenesisBytes::load_bytes();
-        assert_eq!(1775, bytes.len() as u64, "Update me if serialization has changed");
+        assert_eq!(1783, bytes.len() as u64, "Update me if serialization has changed");
     }
 }
diff --git a/parameters/src/testnet3/resources/block.genesis b/parameters/src/testnet3/resources/block.genesis
index 323cca17f..591677d28 100644
Binary files a/parameters/src/testnet3/resources/block.genesis and b/parameters/src/testnet3/resources/block.genesis differ
diff --git a/synthesizer/src/block/header/genesis.rs b/synthesizer/src/block/header/genesis.rs
index 6621b8aa6..bc95b8c2c 100644
--- a/synthesizer/src/block/header/genesis.rs
+++ b/synthesizer/src/block/header/genesis.rs
@@ -55,7 +55,7 @@ mod tests {
         // Previous state root, transactions root, and accumulator point size.
         (Field::<N>::size_in_bytes() * 3)
             // Metadata size.
-            + 2 + 4 + 8 + 8 + 8 + 8 + 8
+            + 2 + 4 + 8 + 8 + 8 + 8 + 8 + 8
             // Add an additional 4 bytes for versioning.
             + 2 + 2
     }
@@ -89,6 +89,8 @@ mod tests {
         assert_eq!(header.round(), 0);
         assert_eq!(header.coinbase_target(), CurrentNetwork::GENESIS_COINBASE_TARGET);
         assert_eq!(header.proof_target(), CurrentNetwork::GENESIS_PROOF_TARGET);
+        assert_eq!(header.last_coinbase_target(), CurrentNetwork::GENESIS_COINBASE_TARGET);
+        assert_eq!(header.last_coinbase_timestamp(), CurrentNetwork::GENESIS_TIMESTAMP);
         assert_eq!(header.timestamp(), CurrentNetwork::GENESIS_TIMESTAMP);
 
         // Ensure the genesis block does *not* contain the following.
diff --git a/synthesizer/src/block/header/merkle.rs b/synthesizer/src/block/header/merkle.rs
index 16cf4f187..e13303aaa 100644
--- a/synthesizer/src/block/header/merkle.rs
+++ b/synthesizer/src/block/header/merkle.rs
@@ -111,6 +111,7 @@ mod tests {
                     u32::rand(rng),
                     u64::rand(rng),
                     u64::rand(rng),
+                    u64::rand(rng),
                     rng.gen_range(0..i64::MAX),
                     rng.gen_range(0..i64::MAX),
                 )?,
diff --git a/synthesizer/src/block/header/metadata/bytes.rs b/synthesizer/src/block/header/metadata/bytes.rs
index df679f8e1..de1824dfd 100644
--- a/synthesizer/src/block/header/metadata/bytes.rs
+++ b/synthesizer/src/block/header/metadata/bytes.rs
@@ -33,12 +33,22 @@ impl<N: Network> FromBytes for Metadata<N> {
         let height = u32::read_le(&mut reader)?;
         let coinbase_target = u64::read_le(&mut reader)?;
         let proof_target = u64::read_le(&mut reader)?;
+        let last_coinbase_target = u64::read_le(&mut reader)?;
         let last_coinbase_timestamp = i64::read_le(&mut reader)?;
         let timestamp = i64::read_le(&mut reader)?;
 
         // Construct the metadata.
-        Self::new(network, round, height, coinbase_target, proof_target, last_coinbase_timestamp, timestamp)
-            .map_err(|e| error(e.to_string()))
+        Self::new(
+            network,
+            round,
+            height,
+            coinbase_target,
+            proof_target,
+            last_coinbase_target,
+            last_coinbase_timestamp,
+            timestamp,
+        )
+        .map_err(|e| error(e.to_string()))
     }
 }
 
@@ -55,6 +65,7 @@ impl<N: Network> ToBytes for Metadata<N> {
         self.height.write_le(&mut writer)?;
         self.coinbase_target.write_le(&mut writer)?;
         self.proof_target.write_le(&mut writer)?;
+        self.last_coinbase_target.write_le(&mut writer)?;
         self.last_coinbase_timestamp.write_le(&mut writer)?;
         self.timestamp.write_le(&mut writer)
     }
diff --git a/synthesizer/src/block/header/metadata/genesis.rs b/synthesizer/src/block/header/metadata/genesis.rs
index 9af2676f5..92bdc116a 100644
--- a/synthesizer/src/block/header/metadata/genesis.rs
+++ b/synthesizer/src/block/header/metadata/genesis.rs
@@ -25,11 +25,21 @@ impl<N: Network> Metadata<N> {
         let height = 0;
         let coinbase_target = N::GENESIS_COINBASE_TARGET;
         let proof_target = N::GENESIS_PROOF_TARGET;
+        let last_coinbase_target = N::GENESIS_COINBASE_TARGET;
         let last_coinbase_timestamp = N::GENESIS_TIMESTAMP;
         let timestamp = N::GENESIS_TIMESTAMP;
 
         // Return the genesis metadata.
-        Self::new(network, round, height, coinbase_target, proof_target, last_coinbase_timestamp, timestamp)
+        Self::new(
+            network,
+            round,
+            height,
+            coinbase_target,
+            proof_target,
+            last_coinbase_target,
+            last_coinbase_timestamp,
+            timestamp,
+        )
     }
 
     /// Returns `true` if the metadata is a genesis metadata.
@@ -44,6 +54,8 @@ impl<N: Network> Metadata<N> {
             && self.coinbase_target == N::GENESIS_COINBASE_TARGET
             // Ensure the proof target in the genesis block is `GENESIS_PROOF_TARGET`.
             && self.proof_target == N::GENESIS_PROOF_TARGET
+            // Ensure the last coinbase target in the genesis block is `GENESIS_COINBASE_TARGET`.
+            && self.last_coinbase_target == N::GENESIS_COINBASE_TARGET
             // Ensure the last coinbase timestamp in the genesis block is `GENESIS_TIMESTAMP`.
             && self.last_coinbase_timestamp == N::GENESIS_TIMESTAMP
             // Ensure the timestamp in the genesis block is `GENESIS_TIMESTAMP`.
@@ -62,7 +74,7 @@ mod tests {
     /// Update this method if the contents of the metadata have changed.
     fn get_expected_size<N: Network>() -> usize {
         // Metadata size.
-        2 + 4 + 8 + 8 + 8 + 8 + 8
+        2 + 4 + 8 + 8 + 8 + 8 + 8 + 8
             // Add an additional 2 bytes for versioning.
             + 2
     }
@@ -94,6 +106,7 @@ mod tests {
         assert_eq!(metadata.round(), 0);
         assert_eq!(metadata.coinbase_target(), CurrentNetwork::GENESIS_COINBASE_TARGET);
         assert_eq!(metadata.proof_target(), CurrentNetwork::GENESIS_PROOF_TARGET);
+        assert_eq!(metadata.last_coinbase_target(), CurrentNetwork::GENESIS_COINBASE_TARGET);
         assert_eq!(metadata.last_coinbase_timestamp(), CurrentNetwork::GENESIS_TIMESTAMP);
         assert_eq!(metadata.timestamp(), CurrentNetwork::GENESIS_TIMESTAMP);
     }
diff --git a/synthesizer/src/block/header/metadata/mod.rs b/synthesizer/src/block/header/metadata/mod.rs
index 741b3d27e..987760b9d 100644
--- a/synthesizer/src/block/header/metadata/mod.rs
+++ b/synthesizer/src/block/header/metadata/mod.rs
@@ -37,6 +37,8 @@ pub struct Metadata<N: Network> {
     coinbase_target: u64,
     /// The proof target for this block - 8 bytes.
     proof_target: u64,
+    /// The coinbase target for the last coinbase - 8 bytes.
+    last_coinbase_target: u64,
     /// The Unix timestamp (UTC) for the last coinbase - 8 bytes.
     last_coinbase_timestamp: i64,
     /// The Unix timestamp (UTC) for this block - 8 bytes.
@@ -47,12 +49,14 @@ pub struct Metadata<N: Network> {
 
 impl<N: Network> Metadata<N> {
     /// Initializes a new metadata with the given inputs.
+    #[allow(clippy::too_many_arguments)]
     pub fn new(
         network: u16,
         round: u64,
         height: u32,
         coinbase_target: u64,
         proof_target: u64,
+        last_coinbase_target: u64,
         last_coinbase_timestamp: i64,
         timestamp: i64,
     ) -> Result<Self> {
@@ -63,6 +67,7 @@ impl<N: Network> Metadata<N> {
             height,
             coinbase_target,
             proof_target,
+            last_coinbase_target,
             last_coinbase_timestamp,
             timestamp,
             _phantom: PhantomData,
@@ -89,6 +94,10 @@ impl<N: Network> Metadata<N> {
                     && self.coinbase_target >= N::GENESIS_COINBASE_TARGET
                     // Ensure the proof target is at or above the minimum.
                     && self.proof_target >= N::GENESIS_PROOF_TARGET
+                    // Ensure the coinbase target is larger than the proof target.
+                    && self.coinbase_target > self.proof_target
+                    // Ensure the last coinbase target is at or above the minimum.
+                    && self.last_coinbase_target >= N::GENESIS_COINBASE_TARGET
                     // Ensure the last coinbase timestamp is after the genesis timestamp.
                     && self.last_coinbase_timestamp >= N::GENESIS_TIMESTAMP
                     // Ensure the timestamp in the block is after the genesis timestamp.
@@ -124,6 +133,11 @@ impl<N: Network> Metadata<N> {
         self.proof_target
     }
 
+    /// Returns the coinbase target of the last coinbase.
+    pub const fn last_coinbase_target(&self) -> u64 {
+        self.last_coinbase_target
+    }
+
     /// Returns the Unix timestamp (UTC) of the last coinbase.
     pub const fn last_coinbase_timestamp(&self) -> i64 {
         self.last_coinbase_timestamp
diff --git a/synthesizer/src/block/header/metadata/serialize.rs b/synthesizer/src/block/header/metadata/serialize.rs
index 557ef756a..660d7e64d 100644
--- a/synthesizer/src/block/header/metadata/serialize.rs
+++ b/synthesizer/src/block/header/metadata/serialize.rs
@@ -21,12 +21,13 @@ impl<N: Network> Serialize for Metadata<N> {
     fn serialize<S: Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
         match serializer.is_human_readable() {
             true => {
-                let mut metadata = serializer.serialize_struct("Certificate", 6)?;
+                let mut metadata = serializer.serialize_struct("Metadata", 8)?;
                 metadata.serialize_field("network", &self.network)?;
                 metadata.serialize_field("round", &self.round)?;
                 metadata.serialize_field("height", &self.height)?;
                 metadata.serialize_field("coinbase_target", &self.coinbase_target)?;
                 metadata.serialize_field("proof_target", &self.proof_target)?;
+                metadata.serialize_field("last_coinbase_target", &self.last_coinbase_target)?;
                 metadata.serialize_field("last_coinbase_timestamp", &self.last_coinbase_timestamp)?;
                 metadata.serialize_field("timestamp", &self.timestamp)?;
                 metadata.end()
@@ -48,6 +49,7 @@ impl<'de, N: Network> Deserialize<'de> for Metadata<N> {
                     serde_json::from_value(metadata["height"].take()).map_err(de::Error::custom)?,
                     serde_json::from_value(metadata["coinbase_target"].take()).map_err(de::Error::custom)?,
                     serde_json::from_value(metadata["proof_target"].take()).map_err(de::Error::custom)?,
+                    serde_json::from_value(metadata["last_coinbase_target"].take()).map_err(de::Error::custom)?,
                     serde_json::from_value(metadata["last_coinbase_timestamp"].take()).map_err(de::Error::custom)?,
                     serde_json::from_value(metadata["timestamp"].take()).map_err(de::Error::custom)?,
                 )
diff --git a/synthesizer/src/block/header/metadata/to_bits.rs b/synthesizer/src/block/header/metadata/to_bits.rs
index f23fc1685..ccec4987d 100644
--- a/synthesizer/src/block/header/metadata/to_bits.rs
+++ b/synthesizer/src/block/header/metadata/to_bits.rs
@@ -25,6 +25,7 @@ impl<N: Network> ToBits for Metadata<N> {
             self.height.to_bits_le(),                  // 4 bytes
             self.coinbase_target.to_bits_le(),         // 8 bytes
             self.proof_target.to_bits_le(),            // 8 bytes
+            self.last_coinbase_target.to_bits_le(),    // 8 bytes
             self.last_coinbase_timestamp.to_bits_le(), // 8 bytes
             self.timestamp.to_bits_le(),               // 8 bytes
         ]
@@ -39,6 +40,7 @@ impl<N: Network> ToBits for Metadata<N> {
             self.height.to_bits_be(),                  // 4 bytes
             self.coinbase_target.to_bits_be(),         // 8 bytes
             self.proof_target.to_bits_be(),            // 8 bytes
+            self.last_coinbase_target.to_bits_be(),    // 8 bytes
             self.last_coinbase_timestamp.to_bits_be(), // 8 bytes
             self.timestamp.to_bits_be(),               // 8 bytes
         ]
diff --git a/synthesizer/src/block/header/metadata/to_hash.rs b/synthesizer/src/block/header/metadata/to_hash.rs
index 3acfb130a..4295d375c 100644
--- a/synthesizer/src/block/header/metadata/to_hash.rs
+++ b/synthesizer/src/block/header/metadata/to_hash.rs
@@ -20,9 +20,9 @@ impl<N: Network> Metadata<N> {
     /// Returns the metadata hash.
     pub fn to_hash(&self) -> Result<Field<N>> {
         // Construct the metadata bits (the last leaf in the Merkle tree).
-        let metadata_bits = self.to_bits_le(); // 368 bits
+        let metadata_bits = self.to_bits_le(); // 432 bits
         // Ensure the metadata bits is the correct size.
-        ensure!(metadata_bits.len() == 368, "Incorrect metadata size");
+        ensure!(metadata_bits.len() == 432, "Incorrect metadata size");
         // Hash the metadata bits.
         let metadata_hash = N::hash_bhp512(&metadata_bits)?;
         // Return the metadata hash.
diff --git a/synthesizer/src/block/header/mod.rs b/synthesizer/src/block/header/mod.rs
index dea3b3f42..de2466362 100644
--- a/synthesizer/src/block/header/mod.rs
+++ b/synthesizer/src/block/header/mod.rs
@@ -120,6 +120,11 @@ impl<N: Network> Header<N> {
         self.metadata.proof_target()
     }
 
+    /// Returns the coinbase target of the last coinbase.
+    pub const fn last_coinbase_target(&self) -> u64 {
+        self.metadata.last_coinbase_target()
+    }
+
     /// Returns the Unix timestamp (UTC) of the last coinbase.
     pub const fn last_coinbase_timestamp(&self) -> i64 {
         self.metadata.last_coinbase_timestamp()
diff --git a/synthesizer/src/block/mod.rs b/synthesizer/src/block/mod.rs
index 6e32a9a5f..a5c31f2bb 100644
--- a/synthesizer/src/block/mod.rs
+++ b/synthesizer/src/block/mod.rs
@@ -184,6 +184,11 @@ impl<N: Network> Block<N> {
         self.header.proof_target()
     }
 
+    /// Returns the coinbase target of the last coinbase.
+    pub const fn last_coinbase_target(&self) -> u64 {
+        self.header.last_coinbase_target()
+    }
+
     /// Returns the Unix timestamp (UTC) of the last coinbase.
     pub const fn last_coinbase_timestamp(&self) -> i64 {
         self.header.last_coinbase_timestamp()
