diff --git a/.cargo/config b/.cargo/config
index 81ae99b32..8dab6d1fb 100644
--- a/.cargo/config
+++ b/.cargo/config
@@ -3,3 +3,4 @@ rustflags = ["-C", "target-cpu=native"]
 
 [target.wasm32-unknown-unknown]
 runner = 'wasm-bindgen-test-runner'
+crate-type = ["cdylib"]
diff --git a/Cargo.lock b/Cargo.lock
index 57d9c6dec..308926c75 100644
--- a/Cargo.lock
+++ b/Cargo.lock
@@ -560,11 +560,10 @@ dependencies = [
 
 [[package]]
 name = "dirs"
-version = "2.0.2"
+version = "3.0.1"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "13aea89a5c93364a98e9b37b2fa237effbb694d5cfe01c5b70941f7eb087d5e3"
+checksum = "142995ed02755914747cc6ca76fc7e4583cd18578746716d0508ea6ed558b9ff"
 dependencies = [
- "cfg-if",
  "dirs-sys",
 ]
 
diff --git a/Cargo.toml b/Cargo.toml
index b92c70f13..b0c4e1851 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -55,7 +55,7 @@ snarkos-utilities = { path = "./utilities", version = "0.8.0" }
 
 clap = { version = "2.33.1" }
 colored = { version = "1.9" }
-dirs = { version = "2.0.2" }
+dirs = { version = "3.0.1" }
 env_logger = { version = "0.7.1" }
 hex = { version = "0.4.1" }
 log = { version = "0.4.8" }
diff --git a/algorithms/src/crh/pedersen.rs b/algorithms/src/crh/pedersen.rs
index 68b709359..48d37550d 100644
--- a/algorithms/src/crh/pedersen.rs
+++ b/algorithms/src/crh/pedersen.rs
@@ -7,7 +7,7 @@ use snarkos_models::{
 
 use rand::Rng;
 
-// #[cfg(feature = "pedersen-parallel")]
+#[cfg(feature = "pedersen-parallel")]
 use rayon::prelude::*;
 
 pub fn bytes_to_bits(bytes: &[u8]) -> Vec<bool> {
@@ -69,38 +69,38 @@ impl<G: Group, S: PedersenSize> CRH for PedersenCRH<G, S> {
 
         // Compute sum of h_i^{m_i} for all i.
         let result = {
-            // #[cfg(feature = "pedersen-parallel")]
-            // {
-            bytes_to_bits(input)
-                .par_chunks(S::WINDOW_SIZE)
-                .zip(&self.parameters.bases)
-                .map(|(bits, powers)| {
-                    let mut encoded = G::zero();
-                    for (bit, base) in bits.iter().zip(powers.iter()) {
-                        if *bit {
-                            encoded += base;
+            #[cfg(feature = "pedersen-parallel")]
+            {
+                bytes_to_bits(input)
+                    .par_chunks(S::WINDOW_SIZE)
+                    .zip(&self.parameters.bases)
+                    .map(|(bits, powers)| {
+                        let mut encoded = G::zero();
+                        for (bit, base) in bits.iter().zip(powers.iter()) {
+                            if *bit {
+                                encoded += base;
+                            }
                         }
-                    }
-                    encoded
-                })
-                .reduce(G::zero, |a, b| a + &b)
-            // }
-            // #[cfg(not(feature = "pedersen-parallel"))]
-            // {
-            //     bytes_to_bits(input)
-            //         .chunks(S::WINDOW_SIZE)
-            //         .zip(&self.parameters.bases)
-            //         .map(|(bits, powers)| {
-            //             let mut encoded = G::zero();
-            //             for (bit, base) in bits.iter().zip(powers.iter()) {
-            //                 if *bit {
-            //                     encoded += base;
-            //                 }
-            //             }
-            //             encoded
-            //         })
-            //         .fold(G::zero(), |a, b| a + &b)
-            // }
+                        encoded
+                    })
+                    .reduce(G::zero, |a, b| a + &b)
+            }
+            #[cfg(not(feature = "pedersen-parallel"))]
+            {
+                bytes_to_bits(input)
+                    .chunks(S::WINDOW_SIZE)
+                    .zip(&self.parameters.bases)
+                    .map(|(bits, powers)| {
+                        let mut encoded = G::zero();
+                        for (bit, base) in bits.iter().zip(powers.iter()) {
+                            if *bit {
+                                encoded += base;
+                            }
+                        }
+                        encoded
+                    })
+                    .fold(G::zero(), |a, b| a + &b)
+            }
         };
 
         Ok(result)
diff --git a/models/src/gadgets/r1cs/constraint_counter.rs b/models/src/gadgets/r1cs/constraint_counter.rs
new file mode 100644
index 000000000..195fcc6fa
--- /dev/null
+++ b/models/src/gadgets/r1cs/constraint_counter.rs
@@ -0,0 +1,81 @@
+use crate::{
+    curves::Field,
+    gadgets::r1cs::{ConstraintSystem, Index, LinearCombination, Variable},
+};
+use snarkos_errors::gadgets::SynthesisError;
+
+/// Constraint counter for testing purposes.
+pub struct ConstraintCounter {
+    pub num_inputs: usize,
+    pub num_aux: usize,
+    pub num_constraints: usize,
+}
+
+impl ConstraintCounter {
+    pub fn new() -> Self {
+        Self {
+            num_aux: 0,
+            num_inputs: 0,
+            num_constraints: 0,
+        }
+    }
+
+    pub fn num_constraints(&self) -> usize {
+        self.num_constraints
+    }
+}
+
+impl<ConstraintF: Field> ConstraintSystem<ConstraintF> for ConstraintCounter {
+    type Root = Self;
+
+    fn alloc<F, A, AR>(&mut self, _: A, _: F) -> Result<Variable, SynthesisError>
+    where
+        F: FnOnce() -> Result<ConstraintF, SynthesisError>,
+        A: FnOnce() -> AR,
+        AR: Into<String>,
+    {
+        let var = Variable::new_unchecked(Index::Aux(self.num_aux));
+        self.num_aux += 1;
+        Ok(var)
+    }
+
+    fn alloc_input<F, A, AR>(&mut self, _: A, _: F) -> Result<Variable, SynthesisError>
+    where
+        F: FnOnce() -> Result<ConstraintF, SynthesisError>,
+        A: FnOnce() -> AR,
+        AR: Into<String>,
+    {
+        let var = Variable::new_unchecked(Index::Input(self.num_inputs));
+        self.num_inputs += 1;
+
+        Ok(var)
+    }
+
+    fn enforce<A, AR, LA, LB, LC>(&mut self, _: A, _: LA, _: LB, _: LC)
+    where
+        A: FnOnce() -> AR,
+        AR: Into<String>,
+        LA: FnOnce(LinearCombination<ConstraintF>) -> LinearCombination<ConstraintF>,
+        LB: FnOnce(LinearCombination<ConstraintF>) -> LinearCombination<ConstraintF>,
+        LC: FnOnce(LinearCombination<ConstraintF>) -> LinearCombination<ConstraintF>,
+    {
+        self.num_constraints += 1;
+    }
+
+    fn push_namespace<NR, N>(&mut self, _: N)
+    where
+        NR: Into<String>,
+        N: FnOnce() -> NR,
+    {
+    }
+
+    fn pop_namespace(&mut self) {}
+
+    fn get_root(&mut self) -> &mut Self::Root {
+        self
+    }
+
+    fn num_constraints(&self) -> usize {
+        self.num_constraints
+    }
+}
diff --git a/models/src/gadgets/r1cs/mod.rs b/models/src/gadgets/r1cs/mod.rs
index ba6548591..1fc4e9a1d 100644
--- a/models/src/gadgets/r1cs/mod.rs
+++ b/models/src/gadgets/r1cs/mod.rs
@@ -1,4 +1,5 @@
 mod assignment;
+mod constraint_counter;
 mod constraint_system;
 mod impl_constraint_var;
 mod impl_lc;
@@ -7,6 +8,7 @@ mod test_fr;
 
 pub use crate::curves::to_field_vec::ToConstraintField;
 pub use assignment::*;
+pub use constraint_counter::ConstraintCounter;
 pub use constraint_system::{ConstraintSynthesizer, ConstraintSystem, Namespace};
 pub use test_constraint_system::TestConstraintSystem;
 pub use test_fr::*;
diff --git a/toolkit/src/lib.rs b/toolkit/src/lib.rs
index 95cab0664..508dd1da7 100644
--- a/toolkit/src/lib.rs
+++ b/toolkit/src/lib.rs
@@ -1,5 +1,7 @@
 #[macro_use]
 extern crate thiserror;
+
+#[cfg(target_arch = "wasm32")]
 #[macro_use]
 extern crate wasm_bindgen_test;
 
diff --git a/toolkit/src/wasm/mod.rs b/toolkit/src/wasm/mod.rs
index a1367f2c8..a53c216a2 100644
--- a/toolkit/src/wasm/mod.rs
+++ b/toolkit/src/wasm/mod.rs
@@ -1,5 +1,6 @@
 pub mod account;
 pub use account::*;
 
+#[cfg(target_arch = "wasm32")]
 #[cfg(test)]
 pub mod tests;
