diff --git a/Cargo.lock b/Cargo.lock
index 273975941..08bcb0824 100644
--- a/Cargo.lock
+++ b/Cargo.lock
@@ -3309,6 +3309,7 @@ version = "2.2.7"
 dependencies = [
  "async-trait",
  "indexmap 2.2.3",
+ "lru",
  "parking_lot",
  "rand",
  "snarkvm",
@@ -3517,7 +3518,7 @@ dependencies = [
 [[package]]
 name = "snarkvm"
 version = "0.16.19"
-source = "git+https://github.com/AleoHQ/snarkVM.git?rev=1cebaaf#1cebaaf6c1f95381e307d049e819e12a60491a83"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=46f2625#46f2625223c0ce162d43552f03426cad422f2aed"
 dependencies = [
  "anstyle",
  "anyhow",
@@ -3548,7 +3549,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-algorithms"
 version = "0.16.19"
-source = "git+https://github.com/AleoHQ/snarkVM.git?rev=1cebaaf#1cebaaf6c1f95381e307d049e819e12a60491a83"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=46f2625#46f2625223c0ce162d43552f03426cad422f2aed"
 dependencies = [
  "aleo-std",
  "anyhow",
@@ -3578,7 +3579,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-circuit"
 version = "0.16.19"
-source = "git+https://github.com/AleoHQ/snarkVM.git?rev=1cebaaf#1cebaaf6c1f95381e307d049e819e12a60491a83"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=46f2625#46f2625223c0ce162d43552f03426cad422f2aed"
 dependencies = [
  "snarkvm-circuit-account",
  "snarkvm-circuit-algorithms",
@@ -3592,7 +3593,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-circuit-account"
 version = "0.16.19"
-source = "git+https://github.com/AleoHQ/snarkVM.git?rev=1cebaaf#1cebaaf6c1f95381e307d049e819e12a60491a83"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=46f2625#46f2625223c0ce162d43552f03426cad422f2aed"
 dependencies = [
  "snarkvm-circuit-algorithms",
  "snarkvm-circuit-network",
@@ -3603,7 +3604,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-circuit-algorithms"
 version = "0.16.19"
-source = "git+https://github.com/AleoHQ/snarkVM.git?rev=1cebaaf#1cebaaf6c1f95381e307d049e819e12a60491a83"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=46f2625#46f2625223c0ce162d43552f03426cad422f2aed"
 dependencies = [
  "snarkvm-circuit-types",
  "snarkvm-console-algorithms",
@@ -3613,7 +3614,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-circuit-collections"
 version = "0.16.19"
-source = "git+https://github.com/AleoHQ/snarkVM.git?rev=1cebaaf#1cebaaf6c1f95381e307d049e819e12a60491a83"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=46f2625#46f2625223c0ce162d43552f03426cad422f2aed"
 dependencies = [
  "snarkvm-circuit-algorithms",
  "snarkvm-circuit-types",
@@ -3623,7 +3624,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-circuit-environment"
 version = "0.16.19"
-source = "git+https://github.com/AleoHQ/snarkVM.git?rev=1cebaaf#1cebaaf6c1f95381e307d049e819e12a60491a83"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=46f2625#46f2625223c0ce162d43552f03426cad422f2aed"
 dependencies = [
  "indexmap 2.2.3",
  "itertools 0.11.0",
@@ -3641,12 +3642,12 @@ dependencies = [
 [[package]]
 name = "snarkvm-circuit-environment-witness"
 version = "0.16.19"
-source = "git+https://github.com/AleoHQ/snarkVM.git?rev=1cebaaf#1cebaaf6c1f95381e307d049e819e12a60491a83"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=46f2625#46f2625223c0ce162d43552f03426cad422f2aed"
 
 [[package]]
 name = "snarkvm-circuit-network"
 version = "0.16.19"
-source = "git+https://github.com/AleoHQ/snarkVM.git?rev=1cebaaf#1cebaaf6c1f95381e307d049e819e12a60491a83"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=46f2625#46f2625223c0ce162d43552f03426cad422f2aed"
 dependencies = [
  "snarkvm-circuit-algorithms",
  "snarkvm-circuit-collections",
@@ -3657,7 +3658,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-circuit-program"
 version = "0.16.19"
-source = "git+https://github.com/AleoHQ/snarkVM.git?rev=1cebaaf#1cebaaf6c1f95381e307d049e819e12a60491a83"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=46f2625#46f2625223c0ce162d43552f03426cad422f2aed"
 dependencies = [
  "paste",
  "snarkvm-circuit-account",
@@ -3672,7 +3673,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-circuit-types"
 version = "0.16.19"
-source = "git+https://github.com/AleoHQ/snarkVM.git?rev=1cebaaf#1cebaaf6c1f95381e307d049e819e12a60491a83"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=46f2625#46f2625223c0ce162d43552f03426cad422f2aed"
 dependencies = [
  "snarkvm-circuit-environment",
  "snarkvm-circuit-types-address",
@@ -3687,7 +3688,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-circuit-types-address"
 version = "0.16.19"
-source = "git+https://github.com/AleoHQ/snarkVM.git?rev=1cebaaf#1cebaaf6c1f95381e307d049e819e12a60491a83"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=46f2625#46f2625223c0ce162d43552f03426cad422f2aed"
 dependencies = [
  "snarkvm-circuit-environment",
  "snarkvm-circuit-types-boolean",
@@ -3700,7 +3701,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-circuit-types-boolean"
 version = "0.16.19"
-source = "git+https://github.com/AleoHQ/snarkVM.git?rev=1cebaaf#1cebaaf6c1f95381e307d049e819e12a60491a83"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=46f2625#46f2625223c0ce162d43552f03426cad422f2aed"
 dependencies = [
  "snarkvm-circuit-environment",
  "snarkvm-console-types-boolean",
@@ -3709,7 +3710,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-circuit-types-field"
 version = "0.16.19"
-source = "git+https://github.com/AleoHQ/snarkVM.git?rev=1cebaaf#1cebaaf6c1f95381e307d049e819e12a60491a83"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=46f2625#46f2625223c0ce162d43552f03426cad422f2aed"
 dependencies = [
  "snarkvm-circuit-environment",
  "snarkvm-circuit-types-boolean",
@@ -3719,7 +3720,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-circuit-types-group"
 version = "0.16.19"
-source = "git+https://github.com/AleoHQ/snarkVM.git?rev=1cebaaf#1cebaaf6c1f95381e307d049e819e12a60491a83"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=46f2625#46f2625223c0ce162d43552f03426cad422f2aed"
 dependencies = [
  "snarkvm-circuit-environment",
  "snarkvm-circuit-types-boolean",
@@ -3731,7 +3732,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-circuit-types-integers"
 version = "0.16.19"
-source = "git+https://github.com/AleoHQ/snarkVM.git?rev=1cebaaf#1cebaaf6c1f95381e307d049e819e12a60491a83"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=46f2625#46f2625223c0ce162d43552f03426cad422f2aed"
 dependencies = [
  "snarkvm-circuit-environment",
  "snarkvm-circuit-types-boolean",
@@ -3743,7 +3744,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-circuit-types-scalar"
 version = "0.16.19"
-source = "git+https://github.com/AleoHQ/snarkVM.git?rev=1cebaaf#1cebaaf6c1f95381e307d049e819e12a60491a83"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=46f2625#46f2625223c0ce162d43552f03426cad422f2aed"
 dependencies = [
  "snarkvm-circuit-environment",
  "snarkvm-circuit-types-boolean",
@@ -3754,7 +3755,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-circuit-types-string"
 version = "0.16.19"
-source = "git+https://github.com/AleoHQ/snarkVM.git?rev=1cebaaf#1cebaaf6c1f95381e307d049e819e12a60491a83"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=46f2625#46f2625223c0ce162d43552f03426cad422f2aed"
 dependencies = [
  "snarkvm-circuit-environment",
  "snarkvm-circuit-types-boolean",
@@ -3766,7 +3767,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-console"
 version = "0.16.19"
-source = "git+https://github.com/AleoHQ/snarkVM.git?rev=1cebaaf#1cebaaf6c1f95381e307d049e819e12a60491a83"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=46f2625#46f2625223c0ce162d43552f03426cad422f2aed"
 dependencies = [
  "snarkvm-console-account",
  "snarkvm-console-algorithms",
@@ -3779,7 +3780,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-console-account"
 version = "0.16.19"
-source = "git+https://github.com/AleoHQ/snarkVM.git?rev=1cebaaf#1cebaaf6c1f95381e307d049e819e12a60491a83"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=46f2625#46f2625223c0ce162d43552f03426cad422f2aed"
 dependencies = [
  "bs58",
  "snarkvm-console-network",
@@ -3790,7 +3791,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-console-algorithms"
 version = "0.16.19"
-source = "git+https://github.com/AleoHQ/snarkVM.git?rev=1cebaaf#1cebaaf6c1f95381e307d049e819e12a60491a83"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=46f2625#46f2625223c0ce162d43552f03426cad422f2aed"
 dependencies = [
  "blake2s_simd",
  "smallvec",
@@ -3803,7 +3804,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-console-collections"
 version = "0.16.19"
-source = "git+https://github.com/AleoHQ/snarkVM.git?rev=1cebaaf#1cebaaf6c1f95381e307d049e819e12a60491a83"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=46f2625#46f2625223c0ce162d43552f03426cad422f2aed"
 dependencies = [
  "aleo-std",
  "rayon",
@@ -3814,7 +3815,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-console-network"
 version = "0.16.19"
-source = "git+https://github.com/AleoHQ/snarkVM.git?rev=1cebaaf#1cebaaf6c1f95381e307d049e819e12a60491a83"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=46f2625#46f2625223c0ce162d43552f03426cad422f2aed"
 dependencies = [
  "anyhow",
  "indexmap 2.2.3",
@@ -3837,7 +3838,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-console-network-environment"
 version = "0.16.19"
-source = "git+https://github.com/AleoHQ/snarkVM.git?rev=1cebaaf#1cebaaf6c1f95381e307d049e819e12a60491a83"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=46f2625#46f2625223c0ce162d43552f03426cad422f2aed"
 dependencies = [
  "anyhow",
  "bech32",
@@ -3855,7 +3856,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-console-program"
 version = "0.16.19"
-source = "git+https://github.com/AleoHQ/snarkVM.git?rev=1cebaaf#1cebaaf6c1f95381e307d049e819e12a60491a83"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=46f2625#46f2625223c0ce162d43552f03426cad422f2aed"
 dependencies = [
  "enum_index",
  "enum_index_derive",
@@ -3876,7 +3877,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-console-types"
 version = "0.16.19"
-source = "git+https://github.com/AleoHQ/snarkVM.git?rev=1cebaaf#1cebaaf6c1f95381e307d049e819e12a60491a83"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=46f2625#46f2625223c0ce162d43552f03426cad422f2aed"
 dependencies = [
  "snarkvm-console-network-environment",
  "snarkvm-console-types-address",
@@ -3891,7 +3892,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-console-types-address"
 version = "0.16.19"
-source = "git+https://github.com/AleoHQ/snarkVM.git?rev=1cebaaf#1cebaaf6c1f95381e307d049e819e12a60491a83"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=46f2625#46f2625223c0ce162d43552f03426cad422f2aed"
 dependencies = [
  "snarkvm-console-network-environment",
  "snarkvm-console-types-boolean",
@@ -3902,7 +3903,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-console-types-boolean"
 version = "0.16.19"
-source = "git+https://github.com/AleoHQ/snarkVM.git?rev=1cebaaf#1cebaaf6c1f95381e307d049e819e12a60491a83"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=46f2625#46f2625223c0ce162d43552f03426cad422f2aed"
 dependencies = [
  "snarkvm-console-network-environment",
 ]
@@ -3910,7 +3911,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-console-types-field"
 version = "0.16.19"
-source = "git+https://github.com/AleoHQ/snarkVM.git?rev=1cebaaf#1cebaaf6c1f95381e307d049e819e12a60491a83"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=46f2625#46f2625223c0ce162d43552f03426cad422f2aed"
 dependencies = [
  "snarkvm-console-network-environment",
  "snarkvm-console-types-boolean",
@@ -3920,7 +3921,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-console-types-group"
 version = "0.16.19"
-source = "git+https://github.com/AleoHQ/snarkVM.git?rev=1cebaaf#1cebaaf6c1f95381e307d049e819e12a60491a83"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=46f2625#46f2625223c0ce162d43552f03426cad422f2aed"
 dependencies = [
  "snarkvm-console-network-environment",
  "snarkvm-console-types-boolean",
@@ -3931,7 +3932,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-console-types-integers"
 version = "0.16.19"
-source = "git+https://github.com/AleoHQ/snarkVM.git?rev=1cebaaf#1cebaaf6c1f95381e307d049e819e12a60491a83"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=46f2625#46f2625223c0ce162d43552f03426cad422f2aed"
 dependencies = [
  "snarkvm-console-network-environment",
  "snarkvm-console-types-boolean",
@@ -3942,7 +3943,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-console-types-scalar"
 version = "0.16.19"
-source = "git+https://github.com/AleoHQ/snarkVM.git?rev=1cebaaf#1cebaaf6c1f95381e307d049e819e12a60491a83"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=46f2625#46f2625223c0ce162d43552f03426cad422f2aed"
 dependencies = [
  "snarkvm-console-network-environment",
  "snarkvm-console-types-boolean",
@@ -3953,7 +3954,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-console-types-string"
 version = "0.16.19"
-source = "git+https://github.com/AleoHQ/snarkVM.git?rev=1cebaaf#1cebaaf6c1f95381e307d049e819e12a60491a83"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=46f2625#46f2625223c0ce162d43552f03426cad422f2aed"
 dependencies = [
  "snarkvm-console-network-environment",
  "snarkvm-console-types-boolean",
@@ -3964,7 +3965,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-curves"
 version = "0.16.19"
-source = "git+https://github.com/AleoHQ/snarkVM.git?rev=1cebaaf#1cebaaf6c1f95381e307d049e819e12a60491a83"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=46f2625#46f2625223c0ce162d43552f03426cad422f2aed"
 dependencies = [
  "rand",
  "rayon",
@@ -3978,7 +3979,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-fields"
 version = "0.16.19"
-source = "git+https://github.com/AleoHQ/snarkVM.git?rev=1cebaaf#1cebaaf6c1f95381e307d049e819e12a60491a83"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=46f2625#46f2625223c0ce162d43552f03426cad422f2aed"
 dependencies = [
  "aleo-std",
  "anyhow",
@@ -3995,7 +3996,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-ledger"
 version = "0.16.19"
-source = "git+https://github.com/AleoHQ/snarkVM.git?rev=1cebaaf#1cebaaf6c1f95381e307d049e819e12a60491a83"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=46f2625#46f2625223c0ce162d43552f03426cad422f2aed"
 dependencies = [
  "aleo-std",
  "anyhow",
@@ -4020,7 +4021,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-ledger-authority"
 version = "0.16.19"
-source = "git+https://github.com/AleoHQ/snarkVM.git?rev=1cebaaf#1cebaaf6c1f95381e307d049e819e12a60491a83"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=46f2625#46f2625223c0ce162d43552f03426cad422f2aed"
 dependencies = [
  "anyhow",
  "rand",
@@ -4032,7 +4033,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-ledger-block"
 version = "0.16.19"
-source = "git+https://github.com/AleoHQ/snarkVM.git?rev=1cebaaf#1cebaaf6c1f95381e307d049e819e12a60491a83"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=46f2625#46f2625223c0ce162d43552f03426cad422f2aed"
 dependencies = [
  "indexmap 2.2.3",
  "rayon",
@@ -4051,7 +4052,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-ledger-coinbase"
 version = "0.16.19"
-source = "git+https://github.com/AleoHQ/snarkVM.git?rev=1cebaaf#1cebaaf6c1f95381e307d049e819e12a60491a83"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=46f2625#46f2625223c0ce162d43552f03426cad422f2aed"
 dependencies = [
  "aleo-std",
  "anyhow",
@@ -4071,7 +4072,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-ledger-committee"
 version = "0.16.19"
-source = "git+https://github.com/AleoHQ/snarkVM.git?rev=1cebaaf#1cebaaf6c1f95381e307d049e819e12a60491a83"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=46f2625#46f2625223c0ce162d43552f03426cad422f2aed"
 dependencies = [
  "anyhow",
  "indexmap 2.2.3",
@@ -4089,7 +4090,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-ledger-narwhal"
 version = "0.16.19"
-source = "git+https://github.com/AleoHQ/snarkVM.git?rev=1cebaaf#1cebaaf6c1f95381e307d049e819e12a60491a83"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=46f2625#46f2625223c0ce162d43552f03426cad422f2aed"
 dependencies = [
  "snarkvm-ledger-narwhal-batch-certificate",
  "snarkvm-ledger-narwhal-batch-header",
@@ -4102,7 +4103,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-ledger-narwhal-batch-certificate"
 version = "0.16.19"
-source = "git+https://github.com/AleoHQ/snarkVM.git?rev=1cebaaf#1cebaaf6c1f95381e307d049e819e12a60491a83"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=46f2625#46f2625223c0ce162d43552f03426cad422f2aed"
 dependencies = [
  "indexmap 2.2.3",
  "rayon",
@@ -4115,7 +4116,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-ledger-narwhal-batch-header"
 version = "0.16.19"
-source = "git+https://github.com/AleoHQ/snarkVM.git?rev=1cebaaf#1cebaaf6c1f95381e307d049e819e12a60491a83"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=46f2625#46f2625223c0ce162d43552f03426cad422f2aed"
 dependencies = [
  "indexmap 2.2.3",
  "serde_json",
@@ -4127,7 +4128,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-ledger-narwhal-data"
 version = "0.16.19"
-source = "git+https://github.com/AleoHQ/snarkVM.git?rev=1cebaaf#1cebaaf6c1f95381e307d049e819e12a60491a83"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=46f2625#46f2625223c0ce162d43552f03426cad422f2aed"
 dependencies = [
  "bytes",
  "serde_json",
@@ -4138,7 +4139,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-ledger-narwhal-subdag"
 version = "0.16.19"
-source = "git+https://github.com/AleoHQ/snarkVM.git?rev=1cebaaf#1cebaaf6c1f95381e307d049e819e12a60491a83"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=46f2625#46f2625223c0ce162d43552f03426cad422f2aed"
 dependencies = [
  "indexmap 2.2.3",
  "rayon",
@@ -4153,7 +4154,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-ledger-narwhal-transmission"
 version = "0.16.19"
-source = "git+https://github.com/AleoHQ/snarkVM.git?rev=1cebaaf#1cebaaf6c1f95381e307d049e819e12a60491a83"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=46f2625#46f2625223c0ce162d43552f03426cad422f2aed"
 dependencies = [
  "bytes",
  "serde_json",
@@ -4166,7 +4167,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-ledger-narwhal-transmission-id"
 version = "0.16.19"
-source = "git+https://github.com/AleoHQ/snarkVM.git?rev=1cebaaf#1cebaaf6c1f95381e307d049e819e12a60491a83"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=46f2625#46f2625223c0ce162d43552f03426cad422f2aed"
 dependencies = [
  "snarkvm-console",
  "snarkvm-ledger-coinbase",
@@ -4175,7 +4176,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-ledger-query"
 version = "0.16.19"
-source = "git+https://github.com/AleoHQ/snarkVM.git?rev=1cebaaf#1cebaaf6c1f95381e307d049e819e12a60491a83"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=46f2625#46f2625223c0ce162d43552f03426cad422f2aed"
 dependencies = [
  "async-trait",
  "reqwest",
@@ -4188,7 +4189,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-ledger-store"
 version = "0.16.19"
-source = "git+https://github.com/AleoHQ/snarkVM.git?rev=1cebaaf#1cebaaf6c1f95381e307d049e819e12a60491a83"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=46f2625#46f2625223c0ce162d43552f03426cad422f2aed"
 dependencies = [
  "aleo-std-storage",
  "anyhow",
@@ -4214,7 +4215,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-ledger-test-helpers"
 version = "0.16.19"
-source = "git+https://github.com/AleoHQ/snarkVM.git?rev=1cebaaf#1cebaaf6c1f95381e307d049e819e12a60491a83"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=46f2625#46f2625223c0ce162d43552f03426cad422f2aed"
 dependencies = [
  "once_cell",
  "snarkvm-circuit",
@@ -4229,7 +4230,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-metrics"
 version = "0.16.19"
-source = "git+https://github.com/AleoHQ/snarkVM.git?rev=1cebaaf#1cebaaf6c1f95381e307d049e819e12a60491a83"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=46f2625#46f2625223c0ce162d43552f03426cad422f2aed"
 dependencies = [
  "metrics",
  "metrics-exporter-prometheus",
@@ -4238,7 +4239,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-parameters"
 version = "0.16.19"
-source = "git+https://github.com/AleoHQ/snarkVM.git?rev=1cebaaf#1cebaaf6c1f95381e307d049e819e12a60491a83"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=46f2625#46f2625223c0ce162d43552f03426cad422f2aed"
 dependencies = [
  "aleo-std",
  "anyhow",
@@ -4263,7 +4264,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-synthesizer"
 version = "0.16.19"
-source = "git+https://github.com/AleoHQ/snarkVM.git?rev=1cebaaf#1cebaaf6c1f95381e307d049e819e12a60491a83"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=46f2625#46f2625223c0ce162d43552f03426cad422f2aed"
 dependencies = [
  "aleo-std",
  "anyhow",
@@ -4288,7 +4289,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-synthesizer-process"
 version = "0.16.19"
-source = "git+https://github.com/AleoHQ/snarkVM.git?rev=1cebaaf#1cebaaf6c1f95381e307d049e819e12a60491a83"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=46f2625#46f2625223c0ce162d43552f03426cad422f2aed"
 dependencies = [
  "aleo-std",
  "colored",
@@ -4311,7 +4312,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-synthesizer-program"
 version = "0.16.19"
-source = "git+https://github.com/AleoHQ/snarkVM.git?rev=1cebaaf#1cebaaf6c1f95381e307d049e819e12a60491a83"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=46f2625#46f2625223c0ce162d43552f03426cad422f2aed"
 dependencies = [
  "indexmap 2.2.3",
  "paste",
@@ -4325,7 +4326,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-synthesizer-snark"
 version = "0.16.19"
-source = "git+https://github.com/AleoHQ/snarkVM.git?rev=1cebaaf#1cebaaf6c1f95381e307d049e819e12a60491a83"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=46f2625#46f2625223c0ce162d43552f03426cad422f2aed"
 dependencies = [
  "bincode",
  "once_cell",
@@ -4338,7 +4339,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-utilities"
 version = "0.16.19"
-source = "git+https://github.com/AleoHQ/snarkVM.git?rev=1cebaaf#1cebaaf6c1f95381e307d049e819e12a60491a83"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=46f2625#46f2625223c0ce162d43552f03426cad422f2aed"
 dependencies = [
  "aleo-std",
  "anyhow",
@@ -4359,7 +4360,7 @@ dependencies = [
 [[package]]
 name = "snarkvm-utilities-derives"
 version = "0.16.19"
-source = "git+https://github.com/AleoHQ/snarkVM.git?rev=1cebaaf#1cebaaf6c1f95381e307d049e819e12a60491a83"
+source = "git+https://github.com/AleoHQ/snarkVM.git?rev=46f2625#46f2625223c0ce162d43552f03426cad422f2aed"
 dependencies = [
  "proc-macro2",
  "quote 1.0.35",
diff --git a/Cargo.toml b/Cargo.toml
index a6529d688..76177a3d8 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -46,7 +46,7 @@ default-features = false
 
 [workspace.dependencies.snarkvm]
 git = "https://github.com/AleoHQ/snarkVM.git"
-rev = "1cebaaf"
+rev = "46f2625"
 #version = "=0.16.18"
 features = [ "circuit", "console", "rocks" ]
 
diff --git a/README.md b/README.md
index 6fa2186ca..e482287ca 100644
--- a/README.md
+++ b/README.md
@@ -17,8 +17,8 @@
   * [2.1 Requirements](#21-requirements)
   * [2.2 Installation](#22-installation)
 * [3. Run an Aleo Node](#3-run-an-aleo-node)
-  * [3a. Run an Aleo Client](#3a-run-an-aleo-client)
-  * [3b. Run an Aleo Prover](#3b-run-an-aleo-prover)
+  * [3.1 Run an Aleo Client](#31-run-an-aleo-client)
+  * [3.2 Run an Aleo Prover](#32-run-an-aleo-prover)
 * [4. FAQs](#4-faqs)
 * [5. Command Line Interface](#5-command-line-interface)
 * [6. Development Guide](#6-development-guide)
@@ -96,7 +96,7 @@ Please ensure ports `4130/tcp` and `3030/tcp` are open on your router and OS fir
 
 ## 3. Run an Aleo Node
 
-## 3a. Run an Aleo Client
+## 3.1 Run an Aleo Client
 
 Start by following the instructions in the [Build Guide](#2-build-guide).
 
@@ -105,7 +105,7 @@ Next, to start a client node, from the `snarkOS` directory, run:
 ./run-client.sh
 ```
 
-## 3b. Run an Aleo Prover
+## 3.2 Run an Aleo Prover
 
 Start by following the instructions in the [Build Guide](#2-build-guide).
 
diff --git a/node/bft/events/src/certificate_response.rs b/node/bft/events/src/certificate_response.rs
index d81f28f1b..e24c3db3f 100644
--- a/node/bft/events/src/certificate_response.rs
+++ b/node/bft/events/src/certificate_response.rs
@@ -86,11 +86,20 @@ pub mod prop_tests {
         (Just(committee.clone()), any::<Selector>(), vec(any_transmission(), 0..16))
             .prop_map(|(committee, selector, transmissions)| {
                 let mut rng = TestRng::default();
-                let CommitteeContext(_, ValidatorSet(validators)) = committee;
+                let CommitteeContext(committee, ValidatorSet(validators)) = committee;
                 let signer = selector.select(validators);
                 let transmission_ids = transmissions.into_iter().map(|(id, _)| id).collect();
 
-                BatchHeader::new(&signer.private_key, 0, now(), transmission_ids, Default::default(), &mut rng).unwrap()
+                BatchHeader::new(
+                    &signer.private_key,
+                    0,
+                    now(),
+                    committee.id(),
+                    transmission_ids,
+                    Default::default(),
+                    &mut rng,
+                )
+                .unwrap()
             })
             .boxed()
     }
diff --git a/node/bft/events/src/helpers/codec.rs b/node/bft/events/src/helpers/codec.rs
index e9498427e..0d0c273d1 100644
--- a/node/bft/events/src/helpers/codec.rs
+++ b/node/bft/events/src/helpers/codec.rs
@@ -83,21 +83,11 @@ impl<N: Network> Decoder for EventCodec<N> {
             Some(bytes) => bytes,
             None => return Ok(None),
         };
-        #[cfg(feature = "metrics")]
-        let num_bytes = bytes.len() as f64;
+
         // Convert the bytes to an event, or fail if it is not valid.
         let reader = bytes.reader();
         match Event::read_le(reader) {
-            Ok(event) => {
-                #[cfg(feature = "metrics")]
-                metrics::histogram_label(
-                    metrics::tcp::TCP_GATEWAY,
-                    "event",
-                    String::from(event.name().clone()),
-                    num_bytes,
-                );
-                Ok(Some(event))
-            }
+            Ok(event) => Ok(Some(event)),
             Err(error) => {
                 error!("Failed to deserialize an event: {}", error);
                 Err(std::io::ErrorKind::InvalidData.into())
diff --git a/node/bft/ledger-service/Cargo.toml b/node/bft/ledger-service/Cargo.toml
index 8a37cdb97..40d0ca87b 100644
--- a/node/bft/ledger-service/Cargo.toml
+++ b/node/bft/ledger-service/Cargo.toml
@@ -18,7 +18,7 @@ edition = "2021"
 
 [features]
 default = [ ]
-ledger = [ "rand", "tokio", "tracing" ]
+ledger = [ "parking_lot", "rand", "tokio", "tracing" ]
 ledger-write = [ ]
 mock = [ "parking_lot", "tracing" ]
 prover = [ ]
@@ -32,6 +32,9 @@ version = "0.1"
 version = "2.1"
 features = [ "serde", "rayon" ]
 
+[dependencies.lru]
+version = "0.12"
+
 [dependencies.parking_lot]
 version = "0.12"
 optional = true
diff --git a/node/bft/ledger-service/src/ledger.rs b/node/bft/ledger-service/src/ledger.rs
index 1f42f909b..520822971 100644
--- a/node/bft/ledger-service/src/ledger.rs
+++ b/node/bft/ledger-service/src/ledger.rs
@@ -26,6 +26,8 @@ use snarkvm::{
 };
 
 use indexmap::IndexMap;
+use lru::LruCache;
+use parking_lot::Mutex;
 use std::{
     fmt,
     ops::Range,
@@ -35,10 +37,14 @@ use std::{
     },
 };
 
+/// The capacity of the LRU holiding the recently queried committees.
+const COMMITTEE_CACHE_SIZE: usize = 16;
+
 /// A core ledger service.
 pub struct CoreLedgerService<N: Network, C: ConsensusStorage<N>> {
     ledger: Ledger<N, C>,
     coinbase_verifying_key: Arc<CoinbaseVerifyingKey<N>>,
+    committee_cache: Arc<Mutex<LruCache<u64, Committee<N>>>>,
     shutdown: Arc<AtomicBool>,
 }
 
@@ -46,7 +52,8 @@ impl<N: Network, C: ConsensusStorage<N>> CoreLedgerService<N, C> {
     /// Initializes a new core ledger service.
     pub fn new(ledger: Ledger<N, C>, shutdown: Arc<AtomicBool>) -> Self {
         let coinbase_verifying_key = Arc::new(ledger.coinbase_puzzle().coinbase_verifying_key().clone());
-        Self { ledger, coinbase_verifying_key, shutdown }
+        let committee_cache = Arc::new(Mutex::new(LruCache::new(COMMITTEE_CACHE_SIZE.try_into().unwrap())));
+        Self { ledger, coinbase_verifying_key, committee_cache, shutdown }
     }
 }
 
@@ -127,9 +134,19 @@ impl<N: Network, C: ConsensusStorage<N>> LedgerService<N> for CoreLedgerService<
     /// Returns the committee for the given round.
     /// If the given round is in the future, then the current committee is returned.
     fn get_committee_for_round(&self, round: u64) -> Result<Committee<N>> {
+        // Check if the committee is already in the cache.
+        if let Some(committee) = self.committee_cache.lock().get(&round) {
+            return Ok(committee.clone());
+        }
+
         match self.ledger.get_committee_for_round(round)? {
             // Return the committee if it exists.
-            Some(committee) => Ok(committee),
+            Some(committee) => {
+                // Insert the committee into the cache.
+                self.committee_cache.lock().push(round, committee.clone());
+                // Return the committee.
+                Ok(committee)
+            }
             // Return the current committee if the round is in the future.
             None => {
                 // Retrieve the current committee.
diff --git a/node/bft/src/helpers/pending.rs b/node/bft/src/helpers/pending.rs
index 7c30e31f9..4ea0d322b 100644
--- a/node/bft/src/helpers/pending.rs
+++ b/node/bft/src/helpers/pending.rs
@@ -13,29 +13,43 @@
 // limitations under the License.
 
 use crate::MAX_FETCH_TIMEOUT_IN_MS;
+use snarkos_node_bft_ledger_service::LedgerService;
+use snarkvm::{console::network::Network, ledger::committee::Committee};
 
 use parking_lot::{Mutex, RwLock};
 use std::{
     collections::{HashMap, HashSet},
     hash::Hash,
     net::SocketAddr,
+    sync::Arc,
 };
 use time::OffsetDateTime;
 use tokio::sync::oneshot;
 
-#[cfg(not(test))]
-pub const NUM_REDUNDANT_REQUESTS: usize = 2;
-#[cfg(test)]
-pub const NUM_REDUNDANT_REQUESTS: usize = 10;
-
 const CALLBACK_TIMEOUT_IN_SECS: i64 = MAX_FETCH_TIMEOUT_IN_MS as i64 / 1000;
 
+/// Returns the maximum number of redundant requests for the number of validators in the specified round.
+pub fn max_redundant_requests<N: Network>(ledger: Arc<dyn LedgerService<N>>, round: u64) -> usize {
+    // Determine the number of validators in the committee lookback for the given round.
+    let num_validators = ledger
+        .get_committee_lookback_for_round(round)
+        .map(|committee| committee.num_members())
+        .ok()
+        .unwrap_or(Committee::<N>::MAX_COMMITTEE_SIZE as usize);
+
+    // Note: It is adequate to set this value to the availability threshold,
+    // as with high probability one will respond honestly (in the best and worst case
+    // with stake spread across the validators evenly and unevenly, respectively).
+    1 + num_validators.saturating_div(3)
+}
+
 #[derive(Debug)]
 pub struct Pending<T: PartialEq + Eq + Hash, V: Clone> {
     /// The map of pending `items` to `peer IPs` that have the item.
     pending: RwLock<HashMap<T, HashSet<SocketAddr>>>,
     /// The optional callback queue.
-    callbacks: Mutex<HashMap<T, Vec<(oneshot::Sender<V>, i64)>>>,
+    /// Each callback has a timeout and a flag indicating if it is associated with a sent request.
+    callbacks: Mutex<HashMap<T, Vec<(oneshot::Sender<V>, i64, bool)>>>,
 }
 
 impl<T: Copy + Clone + PartialEq + Eq + Hash, V: Clone> Default for Pending<T, V> {
@@ -85,12 +99,29 @@ impl<T: Copy + Clone + PartialEq + Eq + Hash, V: Clone> Pending<T, V> {
         self.callbacks.lock().get(&item).map_or(0, |callbacks| callbacks.len())
     }
 
+    /// Returns the number of pending sent requests for the specified `item`.
+    pub fn num_sent_requests(&self, item: impl Into<T>) -> usize {
+        let item = item.into();
+        // Clear the callbacks that have expired.
+        self.clear_expired_callbacks_for_item(item);
+        // Return the number of live callbacks.
+        self.callbacks
+            .lock()
+            .get(&item)
+            .map_or(0, |callbacks| callbacks.iter().filter(|(_, _, request_sent)| *request_sent).count())
+    }
+
     /// Inserts the specified `item` and `peer IP` to the pending queue,
     /// returning `true` if the `peer IP` was newly-inserted into the entry for the `item`.
     ///
     /// In addition, an optional `callback` may be provided, that is triggered upon removal.
     /// Note: The callback, if provided, is **always** inserted into the callback queue.
-    pub fn insert(&self, item: impl Into<T>, peer_ip: SocketAddr, callback: Option<oneshot::Sender<V>>) -> bool {
+    pub fn insert(
+        &self,
+        item: impl Into<T>,
+        peer_ip: SocketAddr,
+        callback: Option<(oneshot::Sender<V>, bool)>,
+    ) -> bool {
         let item = item.into();
         // Insert the peer IP into the pending queue.
         let result = self.pending.write().entry(item).or_default().insert(peer_ip);
@@ -99,8 +130,12 @@ impl<T: Copy + Clone + PartialEq + Eq + Hash, V: Clone> Pending<T, V> {
         self.clear_expired_callbacks_for_item(item);
 
         // If a callback is provided, insert it into the callback queue.
-        if let Some(callback) = callback {
-            self.callbacks.lock().entry(item).or_default().push((callback, OffsetDateTime::now_utc().unix_timestamp()));
+        if let Some((callback, request_sent)) = callback {
+            self.callbacks.lock().entry(item).or_default().push((
+                callback,
+                OffsetDateTime::now_utc().unix_timestamp(),
+                request_sent,
+            ));
         }
         // Return the result.
         result
@@ -117,7 +152,7 @@ impl<T: Copy + Clone + PartialEq + Eq + Hash, V: Clone> Pending<T, V> {
         if let Some(callbacks) = self.callbacks.lock().remove(&item) {
             if let Some(callback_value) = callback_value {
                 // Send a notification to the callback.
-                for (callback, _) in callbacks {
+                for (callback, _, _) in callbacks {
                     callback.send(callback_value.clone()).ok();
                 }
             }
@@ -133,7 +168,7 @@ impl<T: Copy + Clone + PartialEq + Eq + Hash, V: Clone> Pending<T, V> {
             // Fetch the current timestamp.
             let now = OffsetDateTime::now_utc().unix_timestamp();
             // Remove the callbacks that have expired.
-            callbacks.retain(|(_, timestamp)| now - *timestamp <= CALLBACK_TIMEOUT_IN_SECS);
+            callbacks.retain(|(_, timestamp, _)| now - *timestamp <= CALLBACK_TIMEOUT_IN_SECS);
         }
     }
 }
@@ -150,6 +185,8 @@ mod tests {
 
     type CurrentNetwork = snarkvm::prelude::MainnetV0;
 
+    const ITERATIONS: usize = 100;
+
     #[test]
     fn test_pending() {
         let rng = &mut TestRng::default();
@@ -233,13 +270,13 @@ mod tests {
         let (callback_sender_3, _) = oneshot::channel();
 
         // Insert the commitments.
-        assert!(pending.insert(commitment_1, addr_1, Some(callback_sender_1)));
-        assert!(pending.insert(commitment_1, addr_2, Some(callback_sender_2)));
+        assert!(pending.insert(commitment_1, addr_1, Some((callback_sender_1, true))));
+        assert!(pending.insert(commitment_1, addr_2, Some((callback_sender_2, true))));
 
         // Sleep for a few seconds.
         thread::sleep(Duration::from_secs(CALLBACK_TIMEOUT_IN_SECS as u64 - 1));
 
-        assert!(pending.insert(commitment_1, addr_3, Some(callback_sender_3)));
+        assert!(pending.insert(commitment_1, addr_3, Some((callback_sender_3, true))));
 
         // Check that the number of callbacks has not changed.
         assert_eq!(pending.num_callbacks(commitment_1), 3);
@@ -256,6 +293,37 @@ mod tests {
         // Ensure that the expired callbacks have been removed.
         assert_eq!(pending.num_callbacks(commitment_1), 0);
     }
+
+    #[test]
+    fn test_num_sent_requests() {
+        let rng = &mut TestRng::default();
+
+        // Initialize the ready queue.
+        let pending = Pending::<TransmissionID<CurrentNetwork>, ()>::new();
+
+        for _ in 0..ITERATIONS {
+            // Generate a commitment.
+            let commitment = TransmissionID::Solution(PuzzleCommitment::from_g1_affine(rng.gen()));
+            // Check if the number of sent requests is correct.
+            let mut expected_num_sent_requests = 0;
+            for i in 0..ITERATIONS {
+                // Generate a peer address.
+                let addr = SocketAddr::from(([127, 0, 0, 1], i as u16));
+                // Initialize a callback.
+                let (callback_sender, _) = oneshot::channel();
+                // Randomly determine if the callback is associated with a sent request.
+                let is_sent_request = rng.gen();
+                // Increment the expected number of sent requests.
+                if is_sent_request {
+                    expected_num_sent_requests += 1;
+                }
+                // Insert the commitment.
+                assert!(pending.insert(commitment, addr, Some((callback_sender, is_sent_request))));
+            }
+            // Ensure that the number of sent requests is correct.
+            assert_eq!(pending.num_sent_requests(commitment), expected_num_sent_requests);
+        }
+    }
 }
 
 #[cfg(test)]
diff --git a/node/bft/src/helpers/proposal.rs b/node/bft/src/helpers/proposal.rs
index 7b26e0f9f..1137ddaa6 100644
--- a/node/bft/src/helpers/proposal.rs
+++ b/node/bft/src/helpers/proposal.rs
@@ -205,6 +205,7 @@ mod prop_tests {
             &signer.private_key,
             committee.starting_round(),
             now(),
+            committee.id(),
             transmission_map.keys().cloned().collect(),
             Default::default(),
             &mut rng,
diff --git a/node/bft/src/helpers/storage.rs b/node/bft/src/helpers/storage.rs
index fba04dfcc..b390c5739 100644
--- a/node/bft/src/helpers/storage.rs
+++ b/node/bft/src/helpers/storage.rs
@@ -1107,6 +1107,7 @@ pub mod prop_tests {
         selector: Selector,
     ) {
         let CommitteeContext(committee, ValidatorSet(validators)) = context;
+        let committee_id = committee.id();
 
         // Initialize the storage.
         let ledger = Arc::new(MockLedgerService::new(committee));
@@ -1128,6 +1129,7 @@ pub mod prop_tests {
             &signer.private_key,
             0,
             now(),
+            committee_id,
             transmission_map.keys().cloned().collect(),
             Default::default(),
             &mut rng,
diff --git a/node/bft/src/primary.rs b/node/bft/src/primary.rs
index 8dea9e33a..ead1edaac 100644
--- a/node/bft/src/primary.rs
+++ b/node/bft/src/primary.rs
@@ -336,10 +336,10 @@ impl<N: Network> Primary<N> {
             bail!("Primary is safely skipping {}", format!("(round {round} was already certified)").dimmed());
         }
 
+        // Retrieve the committee to check against.
+        let committee_lookback = self.ledger.get_committee_lookback_for_round(round)?;
         // Check if the primary is connected to enough validators to reach quorum threshold.
         {
-            // Retrieve the committee to check against.
-            let committee_lookback = self.ledger.get_committee_lookback_for_round(round)?;
             // Retrieve the connected validator addresses.
             let mut connected_validators = self.gateway.connected_addresses();
             // Append the primary to the set.
@@ -451,6 +451,8 @@ impl<N: Network> Primary<N> {
 
         // Retrieve the private key.
         let private_key = *self.gateway.account().private_key();
+        // Retrieve the committee ID.
+        let committee_id = committee_lookback.id();
         // Prepare the transmission IDs.
         let transmission_ids = transmissions.keys().copied().collect();
         // Prepare the previous batch certificate IDs.
@@ -460,13 +462,13 @@ impl<N: Network> Primary<N> {
             &private_key,
             round,
             now(),
+            committee_id,
             transmission_ids,
             previous_certificate_ids,
             &mut rand::thread_rng()
         ))?;
         // Construct the proposal.
-        let proposal =
-            Proposal::new(self.ledger.get_committee_lookback_for_round(round)?, batch_header.clone(), transmissions)?;
+        let proposal = Proposal::new(committee_lookback, batch_header.clone(), transmissions)?;
         // Broadcast the batch to all validators for signing.
         self.gateway.broadcast(Event::BatchPropose(batch_header.into()));
         // Set the proposed batch.
@@ -521,10 +523,28 @@ impl<N: Network> Primary<N> {
             bail!("Invalid peer - proposed batch from myself ({batch_author})");
         }
 
+        // Ensure that the batch proposal's committee ID matches the expected committee ID.
+        let expected_committee_id = self.ledger.get_committee_lookback_for_round(batch_round)?.id();
+        if expected_committee_id != batch_header.committee_id() {
+            // Proceed to disconnect the validator.
+            self.gateway.disconnect(peer_ip);
+            bail!(
+                "Malicious peer - proposed batch has a different committee ID ({expected_committee_id} != {})",
+                batch_header.committee_id()
+            );
+        }
+
         // Retrieve the cached round and batch ID for this validator.
         if let Some((signed_round, signed_batch_id, signature)) =
             self.signed_proposals.read().get(&batch_author).copied()
         {
+            // If the signed round is ahead of the peer's batch round, then the validator is malicious.
+            if signed_round > batch_header.round() {
+                // Proceed to disconnect the validator.
+                self.gateway.disconnect(peer_ip);
+                bail!("Malicious peer - proposed a batch for a previous round ({})", batch_header.round());
+            }
+
             // If the round matches and the batch ID differs, then the validator is malicious.
             if signed_round == batch_header.round() && signed_batch_id != batch_header.batch_id() {
                 // Proceed to disconnect the validator.
@@ -630,7 +650,7 @@ impl<N: Network> Primary<N> {
         let BatchSignature { batch_id, signature } = batch_signature;
 
         // Retrieve the signer.
-        let signer = spawn_blocking!(Ok(signature.to_address()))?;
+        let signer = signature.to_address();
 
         // Ensure the batch signature is signed by the validator.
         if self.gateway.resolver().get_address(peer_ip).map_or(true, |address| address != signer) {
@@ -643,15 +663,16 @@ impl<N: Network> Primary<N> {
             bail!("Invalid peer - received a batch signature from myself ({signer})");
         }
 
-        let proposal = {
+        let self_ = self.clone();
+        let Some(proposal) = spawn_blocking!({
             // Acquire the write lock.
-            let mut proposed_batch = self.proposed_batch.write();
+            let mut proposed_batch = self_.proposed_batch.write();
             // Add the signature to the batch, and determine if the batch is ready to be certified.
             match proposed_batch.as_mut() {
                 Some(proposal) => {
                     // Ensure the batch ID matches the currently proposed batch ID.
                     if proposal.batch_id() != batch_id {
-                        match self.storage.contains_batch(batch_id) {
+                        match self_.storage.contains_batch(batch_id) {
                             true => bail!("This batch was already certified"),
                             false => bail!(
                                 "Unknown batch ID '{batch_id}', expected '{}' for round {}",
@@ -661,9 +682,9 @@ impl<N: Network> Primary<N> {
                         }
                     }
                     // Retrieve the committee lookback for the round.
-                    let committee_lookback = self.ledger.get_committee_lookback_for_round(proposal.round())?;
+                    let committee_lookback = self_.ledger.get_committee_lookback_for_round(proposal.round())?;
                     // Retrieve the address of the validator.
-                    let Some(signer) = self.gateway.resolver().get_address(peer_ip) else {
+                    let Some(signer) = self_.gateway.resolver().get_address(peer_ip) else {
                         bail!("Signature is from a disconnected validator");
                     };
                     // Add the signature to the batch.
@@ -672,17 +693,20 @@ impl<N: Network> Primary<N> {
                     // Check if the batch is ready to be certified.
                     if !proposal.is_quorum_threshold_reached(&committee_lookback) {
                         // If the batch is not ready to be certified, return early.
-                        return Ok(());
+                        return Ok(None);
                     }
                 }
                 // There is no proposed batch, so return early.
-                None => return Ok(()),
+                None => return Ok(None),
             };
             // Retrieve the batch proposal, clearing the proposed batch.
             match proposed_batch.take() {
-                Some(proposal) => proposal,
-                None => return Ok(()),
+                Some(proposal) => Ok(Some(proposal)),
+                None => Ok(None),
             }
+        })?
+        else {
+            return Ok(());
         };
 
         /* Proceeding to certify the batch. */
@@ -724,6 +748,8 @@ impl<N: Network> Primary<N> {
         let author = certificate.author();
         // Retrieve the batch certificate round.
         let certificate_round = certificate.round();
+        // Retrieve the batch certificate committee ID.
+        let committee_id = certificate.committee_id();
 
         // Ensure the batch certificate is from an authorized validator.
         if !self.gateway.is_authorized_validator_ip(peer_ip) {
@@ -753,6 +779,14 @@ impl<N: Network> Primary<N> {
         // Check if the certificates have reached the quorum threshold.
         let is_quorum = committee_lookback.is_quorum_threshold_reached(&authors);
 
+        // Ensure that the batch certificate's committee ID matches the expected committee ID.
+        let expected_committee_id = committee_lookback.id();
+        if expected_committee_id != committee_id {
+            // Proceed to disconnect the validator.
+            self.gateway.disconnect(peer_ip);
+            bail!("Batch certificate has a different committee ID ({expected_committee_id} != {committee_id})");
+        }
+
         // Determine if we are currently proposing a round that is relevant.
         // Note: This is important, because while our peers have advanced,
         // they may not be proposing yet, and thus still able to sign our proposed batch.
@@ -793,7 +827,8 @@ impl<N: Network> Primary<N> {
                     tokio::time::sleep(Duration::from_millis(PRIMARY_PING_IN_MS)).await;
 
                     // Retrieve the block locators.
-                    let block_locators = match self_.sync.get_block_locators() {
+                    let self__ = self_.clone();
+                    let block_locators = match spawn_blocking!(self__.sync.get_block_locators()) {
                         Ok(block_locators) => block_locators,
                         Err(e) => {
                             warn!("Failed to retrieve block locators - {e}");
@@ -900,6 +935,12 @@ impl<N: Network> Primary<N> {
                     debug!("Skipping batch proposal {}", "(node is syncing)".dimmed());
                     continue;
                 }
+                // A best-effort attempt to skip the scheduled batch proposal if
+                // round progression already triggered one.
+                if self_.propose_lock.try_lock().is_err() {
+                    trace!("Skipping batch proposal {}", "(node is already proposing)".dimmed());
+                    continue;
+                };
                 // If there is no proposed batch, attempt to propose a batch.
                 // Note: Do NOT spawn a task around this function call. Proposing a batch is a critical path,
                 // and only one batch needs be proposed at a time.
@@ -1114,7 +1155,7 @@ impl<N: Network> Primary<N> {
     /// Stores the certified batch and broadcasts it to all validators, returning the certificate.
     async fn store_and_broadcast_certificate(&self, proposal: &Proposal<N>, committee: &Committee<N>) -> Result<()> {
         // Create the batch certificate and transmissions.
-        let (certificate, transmissions) = proposal.to_certificate(committee)?;
+        let (certificate, transmissions) = tokio::task::block_in_place(|| proposal.to_certificate(committee))?;
         // Convert the transmissions into a HashMap.
         // Note: Do not change the `Proposal` to use a HashMap. The ordering there is necessary for safety.
         let transmissions = transmissions.into_iter().collect::<HashMap<_, _>>();
@@ -1525,8 +1566,16 @@ mod tests {
         ]
         .into();
         // Sign the batch header.
-        let batch_header =
-            BatchHeader::new(private_key, round, timestamp, transmission_ids, previous_certificate_ids, rng).unwrap();
+        let batch_header = BatchHeader::new(
+            private_key,
+            round,
+            timestamp,
+            committee.id(),
+            transmission_ids,
+            previous_certificate_ids,
+            rng,
+        )
+        .unwrap();
         // Construct the proposal.
         Proposal::new(committee, batch_header, transmissions).unwrap()
     }
@@ -1584,6 +1633,7 @@ mod tests {
             accounts.iter().find(|&(_, acct)| acct.address() == primary_address).map(|(_, acct)| acct.clone()).unwrap();
         let private_key = author.private_key();
 
+        let committee_id = Field::rand(rng);
         let (solution_commitment, solution) = sample_unconfirmed_solution(rng);
         let (transaction_id, transaction) = sample_unconfirmed_transaction(rng);
         let transmission_ids = [solution_commitment.into(), (&transaction_id).into()].into();
@@ -1593,8 +1643,16 @@ mod tests {
         ]
         .into();
 
-        let batch_header =
-            BatchHeader::new(private_key, round, timestamp, transmission_ids, previous_certificate_ids, rng).unwrap();
+        let batch_header = BatchHeader::new(
+            private_key,
+            round,
+            timestamp,
+            committee_id,
+            transmission_ids,
+            previous_certificate_ids,
+            rng,
+        )
+        .unwrap();
         let signatures = peer_signatures_for_batch(primary_address, accounts, batch_header.batch_id(), rng);
         let certificate = BatchCertificate::<CurrentNetwork>::from(batch_header, signatures).unwrap();
         (certificate, transmissions)
@@ -1840,7 +1898,7 @@ mod tests {
         );
     }
 
-    #[tokio::test]
+    #[tokio::test(flavor = "multi_thread")]
     async fn test_batch_signature_from_peer() {
         let mut rng = TestRng::default();
         let (primary, accounts) = primary_without_handlers(&mut rng).await;
@@ -1875,7 +1933,7 @@ mod tests {
         assert_eq!(primary.current_round(), round + 1);
     }
 
-    #[tokio::test]
+    #[tokio::test(flavor = "multi_thread")]
     async fn test_batch_signature_from_peer_in_round() {
         let round = 5;
         let mut rng = TestRng::default();
diff --git a/node/bft/src/sync/mod.rs b/node/bft/src/sync/mod.rs
index 46cc1885d..555bdb5a4 100644
--- a/node/bft/src/sync/mod.rs
+++ b/node/bft/src/sync/mod.rs
@@ -13,7 +13,7 @@
 // limitations under the License.
 
 use crate::{
-    helpers::{fmt_id, BFTSender, Pending, Storage, SyncReceiver, NUM_REDUNDANT_REQUESTS},
+    helpers::{fmt_id, max_redundant_requests, BFTSender, Pending, Storage, SyncReceiver},
     Gateway,
     Transport,
     MAX_FETCH_TIMEOUT_IN_MS,
@@ -410,19 +410,27 @@ impl<N: Network> Sync<N> {
     ) -> Result<BatchCertificate<N>> {
         // Initialize a oneshot channel.
         let (callback_sender, callback_receiver) = oneshot::channel();
+        // Determine how many sent requests are pending.
+        let num_sent_requests = self.pending.num_sent_requests(certificate_id);
+        // Determine the maximum number of redundant requests.
+        let num_redundant_requests = max_redundant_requests(self.ledger.clone(), self.storage.current_round());
+        // Determine if we should send a certificate request to the peer.
+        let should_send_request = num_sent_requests < num_redundant_requests;
+
         // Insert the certificate ID into the pending queue.
-        if self.pending.insert(certificate_id, peer_ip, Some(callback_sender)) {
-            // Determine how many requests are pending for the certificate.
-            let num_pending_requests = self.pending.num_callbacks(certificate_id);
-            // If the number of requests is less than or equal to the redundancy factor, send the certificate request to the peer.
-            if num_pending_requests <= NUM_REDUNDANT_REQUESTS {
-                // Send the certificate request to the peer.
-                if self.gateway.send(peer_ip, Event::CertificateRequest(certificate_id.into())).await.is_none() {
-                    bail!("Unable to fetch batch certificate {certificate_id} - failed to send request")
-                }
-            } else {
-                trace!("Skipped sending redundant request for certificate {} to '{peer_ip}'", fmt_id(certificate_id));
+        self.pending.insert(certificate_id, peer_ip, Some((callback_sender, should_send_request)));
+
+        // If the number of requests is less than or equal to the redundancy factor, send the certificate request to the peer.
+        if should_send_request {
+            // Send the certificate request to the peer.
+            if self.gateway.send(peer_ip, Event::CertificateRequest(certificate_id.into())).await.is_none() {
+                bail!("Unable to fetch batch certificate {certificate_id} - failed to send request")
             }
+        } else {
+            debug!(
+                "Skipped sending request for certificate {} to '{peer_ip}' ({num_sent_requests} redundant requests)",
+                fmt_id(certificate_id)
+            );
         }
         // Wait for the certificate to be fetched.
         match tokio::time::timeout(core::time::Duration::from_millis(MAX_FETCH_TIMEOUT_IN_MS), callback_receiver).await
diff --git a/node/bft/src/worker.rs b/node/bft/src/worker.rs
index 4e59a399e..e19670374 100644
--- a/node/bft/src/worker.rs
+++ b/node/bft/src/worker.rs
@@ -14,7 +14,7 @@
 
 use crate::{
     events::{Event, TransmissionRequest, TransmissionResponse},
-    helpers::{fmt_id, Pending, Ready, Storage, WorkerReceiver, NUM_REDUNDANT_REQUESTS},
+    helpers::{fmt_id, max_redundant_requests, Pending, Ready, Storage, WorkerReceiver},
     ProposedBatch,
     Transport,
     MAX_FETCH_TIMEOUT_IN_MS,
@@ -90,6 +90,11 @@ impl<N: Network> Worker<N> {
     pub const fn id(&self) -> u8 {
         self.id
     }
+
+    /// Returns a reference to the pending transmissions queue.
+    pub fn pending(&self) -> &Arc<Pending<TransmissionID<N>, Transmission<N>>> {
+        &self.pending
+    }
 }
 
 impl<N: Network> Worker<N> {
@@ -384,18 +389,27 @@ impl<N: Network> Worker<N> {
     ) -> Result<(TransmissionID<N>, Transmission<N>)> {
         // Initialize a oneshot channel.
         let (callback_sender, callback_receiver) = oneshot::channel();
+        // Determine how many sent requests are pending.
+        let num_sent_requests = self.pending.num_sent_requests(transmission_id);
+        // Determine the maximum number of redundant requests.
+        let num_redundant_requests = max_redundant_requests(self.ledger.clone(), self.storage.current_round());
+        // Determine if we should send a transmission request to the peer.
+        let should_send_request = num_sent_requests < num_redundant_requests;
+
         // Insert the transmission ID into the pending queue.
-        self.pending.insert(transmission_id, peer_ip, Some(callback_sender));
-        // Determine how many requests are pending for the transmission.
-        let num_pending_requests = self.pending.num_callbacks(transmission_id);
+        self.pending.insert(transmission_id, peer_ip, Some((callback_sender, should_send_request)));
+
         // If the number of requests is less than or equal to the the redundancy factor, send the transmission request to the peer.
-        if num_pending_requests <= NUM_REDUNDANT_REQUESTS {
+        if should_send_request {
             // Send the transmission request to the peer.
             if self.gateway.send(peer_ip, Event::TransmissionRequest(transmission_id.into())).await.is_none() {
                 bail!("Unable to fetch transmission - failed to send request")
             }
         } else {
-            trace!("Skipped sending redundant request for transmission {} to '{peer_ip}'", fmt_id(transmission_id));
+            debug!(
+                "Skipped sending request for transmission {} to '{peer_ip}' ({num_sent_requests} redundant requests)",
+                fmt_id(transmission_id)
+            );
         }
         // Wait for the transmission to be fetched.
         match timeout(Duration::from_millis(MAX_FETCH_TIMEOUT_IN_MS), callback_receiver).await {
@@ -527,15 +541,35 @@ mod tests {
         }
     }
 
+    #[tokio::test]
+    async fn test_max_redundant_requests() {
+        let rng = &mut TestRng::default();
+        // Sample a committee.
+        let committee = snarkvm::ledger::committee::test_helpers::sample_committee_for_round_and_size(0, 100, rng);
+        let committee_clone = committee.clone();
+        // Setup the mock ledger.
+        let mut mock_ledger = MockLedger::default();
+        mock_ledger.expect_current_committee().returning(move || Ok(committee.clone()));
+        mock_ledger.expect_get_committee_lookback_for_round().returning(move |_| Ok(committee_clone.clone()));
+        mock_ledger.expect_contains_transmission().returning(|_| Ok(false));
+        mock_ledger.expect_check_solution_basic().returning(|_, _| Ok(()));
+        let ledger: Arc<dyn LedgerService<CurrentNetwork>> = Arc::new(mock_ledger);
+
+        // Ensure the maximum number of redundant requests is correct and consistent across iterations.
+        assert_eq!(max_redundant_requests(ledger, 0), 34, "Update me if the formula changes");
+    }
+
     #[tokio::test]
     async fn test_process_transmission() {
         let rng = &mut TestRng::default();
         // Sample a committee.
         let committee = snarkvm::ledger::committee::test_helpers::sample_committee(rng);
+        let committee_clone = committee.clone();
         // Setup the mock gateway and ledger.
         let gateway = MockGateway::default();
         let mut mock_ledger = MockLedger::default();
         mock_ledger.expect_current_committee().returning(move || Ok(committee.clone()));
+        mock_ledger.expect_get_committee_lookback_for_round().returning(move |_| Ok(committee_clone.clone()));
         mock_ledger.expect_contains_transmission().returning(|_| Ok(false));
         mock_ledger.expect_check_solution_basic().returning(|_, _| Ok(()));
         let ledger: Arc<dyn LedgerService<CurrentNetwork>> = Arc::new(mock_ledger);
@@ -565,6 +599,7 @@ mod tests {
         let rng = &mut TestRng::default();
         // Sample a committee.
         let committee = snarkvm::ledger::committee::test_helpers::sample_committee(rng);
+        let committee_clone = committee.clone();
         // Setup the mock gateway and ledger.
         let mut gateway = MockGateway::default();
         gateway.expect_send().returning(|_, _| {
@@ -573,6 +608,7 @@ mod tests {
         });
         let mut mock_ledger = MockLedger::default();
         mock_ledger.expect_current_committee().returning(move || Ok(committee.clone()));
+        mock_ledger.expect_get_committee_lookback_for_round().returning(move |_| Ok(committee_clone.clone()));
         mock_ledger.expect_ensure_transmission_id_matches().returning(|_, _| Ok(()));
         let ledger: Arc<dyn LedgerService<CurrentNetwork>> = Arc::new(mock_ledger);
         // Initialize the storage.
@@ -600,6 +636,7 @@ mod tests {
         let rng = &mut TestRng::default();
         // Sample a committee.
         let committee = snarkvm::ledger::committee::test_helpers::sample_committee(rng);
+        let committee_clone = committee.clone();
         // Setup the mock gateway and ledger.
         let mut gateway = MockGateway::default();
         gateway.expect_send().returning(|_, _| {
@@ -608,6 +645,7 @@ mod tests {
         });
         let mut mock_ledger = MockLedger::default();
         mock_ledger.expect_current_committee().returning(move || Ok(committee.clone()));
+        mock_ledger.expect_get_committee_lookback_for_round().returning(move |_| Ok(committee_clone.clone()));
         mock_ledger.expect_contains_transmission().returning(|_| Ok(false));
         mock_ledger.expect_check_solution_basic().returning(|_, _| Ok(()));
         let ledger: Arc<dyn LedgerService<CurrentNetwork>> = Arc::new(mock_ledger);
@@ -638,6 +676,7 @@ mod tests {
         let rng = &mut TestRng::default();
         // Sample a committee.
         let committee = snarkvm::ledger::committee::test_helpers::sample_committee(rng);
+        let committee_clone = committee.clone();
         // Setup the mock gateway and ledger.
         let mut gateway = MockGateway::default();
         gateway.expect_send().returning(|_, _| {
@@ -646,6 +685,7 @@ mod tests {
         });
         let mut mock_ledger = MockLedger::default();
         mock_ledger.expect_current_committee().returning(move || Ok(committee.clone()));
+        mock_ledger.expect_get_committee_lookback_for_round().returning(move |_| Ok(committee_clone.clone()));
         mock_ledger.expect_contains_transmission().returning(|_| Ok(false));
         mock_ledger.expect_check_solution_basic().returning(|_, _| Err(anyhow!("")));
         let ledger: Arc<dyn LedgerService<CurrentNetwork>> = Arc::new(mock_ledger);
@@ -676,6 +716,7 @@ mod tests {
         let mut rng = &mut TestRng::default();
         // Sample a committee.
         let committee = snarkvm::ledger::committee::test_helpers::sample_committee(rng);
+        let committee_clone = committee.clone();
         // Setup the mock gateway and ledger.
         let mut gateway = MockGateway::default();
         gateway.expect_send().returning(|_, _| {
@@ -684,6 +725,7 @@ mod tests {
         });
         let mut mock_ledger = MockLedger::default();
         mock_ledger.expect_current_committee().returning(move || Ok(committee.clone()));
+        mock_ledger.expect_get_committee_lookback_for_round().returning(move |_| Ok(committee_clone.clone()));
         mock_ledger.expect_contains_transmission().returning(|_| Ok(false));
         mock_ledger.expect_check_transaction_basic().returning(|_, _| Ok(()));
         let ledger: Arc<dyn LedgerService<CurrentNetwork>> = Arc::new(mock_ledger);
@@ -714,6 +756,7 @@ mod tests {
         let mut rng = &mut TestRng::default();
         // Sample a committee.
         let committee = snarkvm::ledger::committee::test_helpers::sample_committee(rng);
+        let committee_clone = committee.clone();
         // Setup the mock gateway and ledger.
         let mut gateway = MockGateway::default();
         gateway.expect_send().returning(|_, _| {
@@ -722,6 +765,7 @@ mod tests {
         });
         let mut mock_ledger = MockLedger::default();
         mock_ledger.expect_current_committee().returning(move || Ok(committee.clone()));
+        mock_ledger.expect_get_committee_lookback_for_round().returning(move |_| Ok(committee_clone.clone()));
         mock_ledger.expect_contains_transmission().returning(|_| Ok(false));
         mock_ledger.expect_check_transaction_basic().returning(|_, _| Err(anyhow!("")));
         let ledger: Arc<dyn LedgerService<CurrentNetwork>> = Arc::new(mock_ledger);
@@ -746,6 +790,84 @@ mod tests {
         assert!(!worker.pending.contains(transmission_id));
         assert!(!worker.ready.contains(transmission_id));
     }
+
+    #[tokio::test]
+    async fn test_flood_transmission_requests() {
+        let mut rng = &mut TestRng::default();
+        // Sample a committee.
+        let committee = snarkvm::ledger::committee::test_helpers::sample_committee(rng);
+        let committee_clone = committee.clone();
+        // Setup the mock gateway and ledger.
+        let mut gateway = MockGateway::default();
+        gateway.expect_send().returning(|_, _| {
+            let (_tx, rx) = oneshot::channel();
+            Some(rx)
+        });
+        let mut mock_ledger = MockLedger::default();
+        mock_ledger.expect_current_committee().returning(move || Ok(committee.clone()));
+        mock_ledger.expect_get_committee_lookback_for_round().returning(move |_| Ok(committee_clone.clone()));
+        mock_ledger.expect_contains_transmission().returning(|_| Ok(false));
+        mock_ledger.expect_check_transaction_basic().returning(|_, _| Ok(()));
+        let ledger: Arc<dyn LedgerService<CurrentNetwork>> = Arc::new(mock_ledger);
+        // Initialize the storage.
+        let storage = Storage::<CurrentNetwork>::new(ledger.clone(), Arc::new(BFTMemoryService::new()), 1);
+
+        // Create the Worker.
+        let worker = Worker::new(0, Arc::new(gateway), storage, ledger, Default::default()).unwrap();
+        let transaction_id: <CurrentNetwork as Network>::TransactionID = Field::<CurrentNetwork>::rand(&mut rng).into();
+        let transmission_id = TransmissionID::Transaction(transaction_id);
+        let peer_ip = SocketAddr::from(([127, 0, 0, 1], 1234));
+
+        // Determine the number of redundant requests are sent.
+        let num_redundant_requests = max_redundant_requests(worker.ledger.clone(), worker.storage.current_round());
+        let num_flood_requests = num_redundant_requests * 10;
+        // Flood the pending queue with transmission requests.
+        for i in 1..=num_flood_requests {
+            let worker_ = worker.clone();
+            tokio::spawn(async move {
+                let _ = worker_.send_transmission_request(peer_ip, transmission_id).await;
+            });
+            tokio::time::sleep(Duration::from_millis(10)).await;
+            // Check that the number of sent requests does not exceed the maximum number of redundant requests.
+            assert!(worker.pending.num_sent_requests(transmission_id) <= num_redundant_requests);
+            assert_eq!(worker.pending.num_callbacks(transmission_id), i);
+        }
+        // Check that the number of sent requests does not exceed the maximum number of redundant requests.
+        assert_eq!(worker.pending.num_sent_requests(transmission_id), num_redundant_requests);
+        assert_eq!(worker.pending.num_callbacks(transmission_id), num_flood_requests);
+
+        // Let all the requests expire.
+        tokio::time::sleep(Duration::from_millis(MAX_FETCH_TIMEOUT_IN_MS + 1000)).await;
+        assert_eq!(worker.pending.num_sent_requests(transmission_id), 0);
+        assert_eq!(worker.pending.num_callbacks(transmission_id), 0);
+
+        // Flood the pending queue with transmission requests again.
+        for i in 1..=num_flood_requests {
+            let worker_ = worker.clone();
+            tokio::spawn(async move {
+                let _ = worker_.send_transmission_request(peer_ip, transmission_id).await;
+            });
+            tokio::time::sleep(Duration::from_millis(10)).await;
+            assert!(worker.pending.num_sent_requests(transmission_id) <= num_redundant_requests);
+            assert_eq!(worker.pending.num_callbacks(transmission_id), i);
+        }
+        // Check that the number of sent requests does not exceed the maximum number of redundant requests.
+        assert_eq!(worker.pending.num_sent_requests(transmission_id), num_redundant_requests);
+        assert_eq!(worker.pending.num_callbacks(transmission_id), num_flood_requests);
+
+        // Check that fulfilling a transmission request clears the pending queue.
+        let result = worker
+            .process_unconfirmed_transaction(
+                transaction_id,
+                Data::Buffer(Bytes::from((0..512).map(|_| rng.gen::<u8>()).collect::<Vec<_>>())),
+            )
+            .await;
+        assert!(result.is_ok());
+        assert_eq!(worker.pending.num_sent_requests(transmission_id), 0);
+        assert_eq!(worker.pending.num_callbacks(transmission_id), 0);
+        assert!(!worker.pending.contains(transmission_id));
+        assert!(worker.ready.contains(transmission_id));
+    }
 }
 
 #[cfg(test)]
diff --git a/node/bft/tests/bft_e2e.rs b/node/bft/tests/bft_e2e.rs
index cd2a442f5..22f2be82d 100644
--- a/node/bft/tests/bft_e2e.rs
+++ b/node/bft/tests/bft_e2e.rs
@@ -14,6 +14,8 @@
 
 #[allow(dead_code)]
 mod common;
+#[allow(dead_code)]
+mod components;
 
 use crate::common::primary::{TestNetwork, TestNetworkConfig};
 use deadline::deadline;
diff --git a/node/bft/tests/common/primary.rs b/node/bft/tests/common/primary.rs
index ef7eb9161..43b4ce8d3 100644
--- a/node/bft/tests/common/primary.rs
+++ b/node/bft/tests/common/primary.rs
@@ -324,7 +324,7 @@ impl TestNetwork {
 }
 
 // Initializes a new test committee.
-fn new_test_committee(n: u16) -> (Vec<Account<CurrentNetwork>>, Committee<CurrentNetwork>) {
+pub fn new_test_committee(n: u16) -> (Vec<Account<CurrentNetwork>>, Committee<CurrentNetwork>) {
     let mut accounts = Vec::with_capacity(n as usize);
     let mut members = IndexMap::with_capacity(n as usize);
     for i in 0..n {
@@ -346,7 +346,7 @@ fn genesis_cache() -> &'static Mutex<HashMap<Vec<u8>, Block<CurrentNetwork>>> {
     CACHE.get_or_init(|| Mutex::new(HashMap::new()))
 }
 
-fn genesis_block(
+pub fn genesis_block(
     genesis_private_key: PrivateKey<CurrentNetwork>,
     committee: Committee<CurrentNetwork>,
     public_balances: IndexMap<Address<CurrentNetwork>, u64>,
@@ -361,7 +361,7 @@ fn genesis_block(
     vm.genesis_quorum(&genesis_private_key, committee, public_balances, bonded_balances, rng).unwrap()
 }
 
-fn genesis_ledger(
+pub fn genesis_ledger(
     genesis_private_key: PrivateKey<CurrentNetwork>,
     committee: Committee<CurrentNetwork>,
     public_balances: IndexMap<Address<CurrentNetwork>, u64>,
diff --git a/node/bft/tests/components/mod.rs b/node/bft/tests/components/mod.rs
new file mode 100644
index 000000000..1a11df65a
--- /dev/null
+++ b/node/bft/tests/components/mod.rs
@@ -0,0 +1,80 @@
+// Copyright (C) 2019-2023 Aleo Systems Inc.
+// This file is part of the snarkOS library.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+// http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+pub mod pending;
+pub mod worker;
+
+use crate::common::{primary, CurrentNetwork, TranslucentLedgerService};
+use snarkos_account::Account;
+use snarkos_node_bft::{helpers::Storage, Gateway, Worker};
+use snarkos_node_bft_ledger_service::LedgerService;
+use snarkos_node_bft_storage_service::BFTMemoryService;
+use snarkvm::{
+    console::{account::Address, network::Network},
+    ledger::{narwhal::BatchHeader, store::helpers::memory::ConsensusMemory},
+    prelude::TestRng,
+};
+
+use indexmap::IndexMap;
+use parking_lot::RwLock;
+use std::{str::FromStr, sync::Arc};
+
+const ITERATIONS: u32 = 100;
+
+/// Samples a new ledger with the given number of nodes.
+pub fn sample_ledger(
+    num_nodes: u16,
+    rng: &mut TestRng,
+) -> Arc<TranslucentLedgerService<CurrentNetwork, ConsensusMemory<CurrentNetwork>>> {
+    let (accounts, committee) = primary::new_test_committee(num_nodes);
+    let bonded_balances: IndexMap<_, _> =
+        committee.members().iter().map(|(address, (amount, _))| (*address, (*address, *amount))).collect();
+    let gen_key = *accounts[0].private_key();
+    let public_balance_per_validator =
+        (1_500_000_000_000_000 - (num_nodes as u64) * 1_000_000_000_000) / (num_nodes as u64);
+    let mut balances = IndexMap::<Address<CurrentNetwork>, u64>::new();
+    for account in accounts.iter() {
+        balances.insert(account.address(), public_balance_per_validator);
+    }
+
+    let gen_ledger =
+        primary::genesis_ledger(gen_key, committee.clone(), balances.clone(), bonded_balances.clone(), rng);
+    Arc::new(TranslucentLedgerService::new(gen_ledger, Default::default()))
+}
+
+/// Samples a new storage with the given ledger.
+pub fn sample_storage<N: Network>(ledger: Arc<TranslucentLedgerService<N, ConsensusMemory<N>>>) -> Storage<N> {
+    Storage::new(ledger, Arc::new(BFTMemoryService::new()), BatchHeader::<N>::MAX_GC_ROUNDS as u64)
+}
+
+/// Samples a new gateway with the given ledger.
+pub fn sample_gateway<N: Network>(ledger: Arc<TranslucentLedgerService<N, ConsensusMemory<N>>>) -> Gateway<N> {
+    let num_nodes: u16 = ledger.current_committee().unwrap().num_members().try_into().unwrap();
+    let (accounts, _committee) = primary::new_test_committee(num_nodes);
+    let account = Account::from_str(&accounts[0].private_key().to_string()).unwrap();
+    // Initialize the gateway.
+    Gateway::new(account, ledger, None, &[], None).unwrap()
+}
+
+/// Samples a new worker with the given ledger.
+pub fn sample_worker<N: Network>(id: u8, ledger: Arc<TranslucentLedgerService<N, ConsensusMemory<N>>>) -> Worker<N> {
+    // Sample a storage.
+    let storage = sample_storage(ledger.clone());
+    // Sample a gateway.
+    let gateway = sample_gateway(ledger.clone());
+    // Sample a dummy proposed batch.
+    let proposed_batch = Arc::new(RwLock::new(None));
+    // Construct the worker instance.
+    Worker::new(id, Arc::new(gateway.clone()), storage.clone(), ledger, proposed_batch).unwrap()
+}
diff --git a/node/bft/tests/components/pending.rs b/node/bft/tests/components/pending.rs
new file mode 100644
index 000000000..b1405207d
--- /dev/null
+++ b/node/bft/tests/components/pending.rs
@@ -0,0 +1,29 @@
+// Copyright (C) 2019-2023 Aleo Systems Inc.
+// This file is part of the snarkOS library.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+// http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::components::sample_ledger;
+use snarkos_node_bft::helpers::max_redundant_requests;
+use snarkvm::prelude::TestRng;
+
+#[test]
+fn test_max_redundant_requests() {
+    const NUM_NODES: u16 = 100;
+
+    // Initialize the RNG.
+    let rng = &mut TestRng::default();
+    // Sample a ledger.
+    let ledger = sample_ledger(NUM_NODES, rng);
+    // Ensure the maximum number of redundant requests is correct and consistent across iterations.
+    assert_eq!(max_redundant_requests(ledger, 0), 34, "Update me if the formula changes");
+}
diff --git a/node/bft/tests/components/worker.rs b/node/bft/tests/components/worker.rs
new file mode 100644
index 000000000..279bc1bf9
--- /dev/null
+++ b/node/bft/tests/components/worker.rs
@@ -0,0 +1,141 @@
+// Copyright (C) 2019-2023 Aleo Systems Inc.
+// This file is part of the snarkOS library.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+// http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::{
+    common::CurrentNetwork,
+    components::{sample_ledger, sample_worker},
+};
+use snarkos_node_bft::helpers::max_redundant_requests;
+use snarkvm::{
+    ledger::narwhal::TransmissionID,
+    prelude::{Network, TestRng},
+};
+
+use std::net::SocketAddr;
+
+#[tokio::test]
+#[rustfmt::skip]
+async fn test_resend_transmission_request() {
+    const NUM_NODES: u16 = 100;
+
+    // Initialize the RNG.
+    let rng = &mut TestRng::default();
+    // Sample a ledger.
+    let ledger = sample_ledger(NUM_NODES, rng);
+    // Sample a worker.
+    let worker = sample_worker(0, ledger.clone());
+
+    // Prepare a dummy transmission ID.
+    let peer_ip = SocketAddr::from(([127, 0, 0, 1], 1234));
+    let transmission_id = TransmissionID::Transaction(<CurrentNetwork as Network>::TransactionID::default());
+
+    // Ensure the worker does not have the dummy transmission ID.
+    assert!(!worker.contains_transmission(transmission_id), "Transmission should not exist");
+
+    // Send a request to fetch the dummy transmission.
+    let worker_ = worker.clone();
+    tokio::spawn(async move { worker_.get_or_fetch_transmission(peer_ip, transmission_id).await });
+
+    tokio::time::sleep(std::time::Duration::from_millis(10)).await;
+
+    let pending = worker.pending();
+    // Ensure the transmission ID exists in the pending queue.
+    assert!(pending.contains(transmission_id), "Missing a transmission in the pending queue");
+    // Ensure the peer IP is in the pending queue for the transmission ID.
+    assert!(pending.contains_peer(transmission_id, peer_ip), "Missing a peer IP for transmission in the pending queue");
+    assert_eq!(pending.get(transmission_id), Some([peer_ip].into_iter().collect()), "Missing a peer IP for transmission in the pending queue");
+    // Ensure the number of callbacks is correct.
+    assert_eq!(pending.num_callbacks(transmission_id), 1, "Incorrect number of callbacks for transmission");
+    // Ensure the number of sent requests is correct.
+    assert_eq!(pending.num_sent_requests(transmission_id), 1, "Incorrect number of sent requests for transmission");
+
+    // Rebroadcast the same request to fetch the dummy transmission.
+    for i in 1..=10 {
+        let worker_ = worker.clone();
+        tokio::spawn(async move { worker_.get_or_fetch_transmission(peer_ip, transmission_id).await });
+
+        tokio::time::sleep(std::time::Duration::from_millis(10)).await;
+
+        // Ensure the transmission ID exists in the pending queue.
+        assert!(pending.contains(transmission_id), "Missing a transmission in the pending queue");
+        // Ensure the peer IP is in the pending queue for the transmission ID.
+        assert!(pending.contains_peer(transmission_id, peer_ip), "Missing a peer IP for transmission in the pending queue");
+        assert_eq!(pending.get(transmission_id), Some([peer_ip].into_iter().collect()), "Missing a peer IP for transmission in the pending queue");
+        // Ensure the number of callbacks is correct.
+        assert_eq!(pending.num_callbacks(transmission_id), 1 + i, "Incorrect number of callbacks for transmission");
+        // Ensure the number of sent requests is correct.
+        assert_eq!(pending.num_sent_requests(transmission_id), 1 + i, "Incorrect number of sent requests for transmission");
+    }
+}
+
+#[tokio::test]
+#[rustfmt::skip]
+async fn test_flood_transmission_requests() {
+    const NUM_NODES: u16 = 100;
+
+    // Initialize the RNG.
+    let rng = &mut TestRng::default();
+    // Sample a ledger.
+    let ledger = sample_ledger(NUM_NODES, rng);
+    // Sample a worker.
+    let worker = sample_worker(0, ledger.clone());
+
+    // Determine the maximum number of redundant requests.
+    let max_redundancy = max_redundant_requests(ledger.clone(), 0);
+    assert_eq!(max_redundancy, 34, "Update me if the formula changes");
+
+    // Prepare a dummy transmission ID.
+    let peer_ip = SocketAddr::from(([127, 0, 0, 1], 1234));
+    let transmission_id = TransmissionID::Transaction(<CurrentNetwork as Network>::TransactionID::default());
+
+    // Ensure the worker does not have the dummy transmission ID.
+    assert!(!worker.contains_transmission(transmission_id), "Transmission should not exist");
+
+    // Send the maximum number of redundant requests to fetch the dummy transmission.
+    for _ in 0..max_redundancy {
+        let worker_ = worker.clone();
+        tokio::spawn(async move { worker_.get_or_fetch_transmission(peer_ip, transmission_id).await });
+    }
+
+    tokio::time::sleep(std::time::Duration::from_millis(10)).await;
+
+    let pending = worker.pending();
+    // Ensure the transmission ID exists in the pending queue.
+    assert!(pending.contains(transmission_id), "Missing a transmission in the pending queue");
+    // Ensure the peer IP is in the pending queue for the transmission ID.
+    assert!(pending.contains_peer(transmission_id, peer_ip), "Missing a peer IP for transmission in the pending queue");
+    assert_eq!(pending.get(transmission_id), Some([peer_ip].into_iter().collect()), "Missing a peer IP for transmission in the pending queue");
+    // Ensure the number of callbacks is correct.
+    assert_eq!(pending.num_callbacks(transmission_id), max_redundancy, "Incorrect number of callbacks for transmission");
+    // Ensure the number of sent requests is correct.
+    assert_eq!(pending.num_sent_requests(transmission_id), max_redundancy, "Incorrect number of sent requests for transmission");
+
+    // Ensure any further redundant requests are not sent.
+    for i in 1..=20 {
+        let worker_ = worker.clone();
+        tokio::spawn(async move { worker_.get_or_fetch_transmission(peer_ip, transmission_id).await });
+
+        tokio::time::sleep(std::time::Duration::from_millis(10)).await;
+
+        // Ensure the transmission ID exists in the pending queue.
+        assert!(pending.contains(transmission_id), "Missing a transmission in the pending queue");
+        // Ensure the peer IP is in the pending queue for the transmission ID.
+        assert!(pending.contains_peer(transmission_id, peer_ip), "Missing a peer IP for transmission in the pending queue");
+        assert_eq!(pending.get(transmission_id), Some([peer_ip].into_iter().collect()), "Missing a peer IP for transmission in the pending queue");
+        // Ensure the number of callbacks is correct.
+        assert_eq!(pending.num_callbacks(transmission_id), max_redundancy + i, "Incorrect number of callbacks for transmission");
+        // Ensure the number of sent requests is correct.
+        assert_eq!(pending.num_sent_requests(transmission_id), max_redundancy, "Incorrect number of sent requests for transmission");
+    }
+}
diff --git a/node/consensus/src/lib.rs b/node/consensus/src/lib.rs
index 30bc32c7e..ea8d04d85 100644
--- a/node/consensus/src/lib.rs
+++ b/node/consensus/src/lib.rs
@@ -407,7 +407,7 @@ impl<N: Network> Consensus<N> {
             let block_latency = next_block_timestamp - current_block_timestamp;
 
             metrics::gauge(metrics::blocks::HEIGHT, next_block.height() as f64);
-            metrics::counter(metrics::blocks::TRANSACTIONS, next_block.transactions().len() as u64);
+            metrics::increment_gauge(metrics::blocks::TRANSACTIONS, next_block.transactions().len() as f64);
             metrics::gauge(metrics::consensus::LAST_COMMITTED_ROUND, next_block.round() as f64);
             metrics::gauge(metrics::consensus::COMMITTED_CERTIFICATES, num_committed_certificates as f64);
             metrics::histogram(metrics::consensus::CERTIFICATE_COMMIT_LATENCY, elapsed.as_secs_f64());
diff --git a/node/metrics/src/names.rs b/node/metrics/src/names.rs
index 819434184..bcf163a53 100644
--- a/node/metrics/src/names.rs
+++ b/node/metrics/src/names.rs
@@ -30,7 +30,7 @@ pub(super) const GAUGE_NAMES: [&str; 13] = [
     tcp::TCP_TASKS,
 ];
 
-pub(super) const HISTOGRAM_NAMES: [&str; 8] = [
+pub(super) const HISTOGRAM_NAMES: [&str; 7] = [
     bft::COMMIT_ROUNDS_LATENCY,
     consensus::CERTIFICATE_COMMIT_LATENCY,
     consensus::BLOCK_LATENCY,
@@ -38,7 +38,6 @@ pub(super) const HISTOGRAM_NAMES: [&str; 8] = [
     tcp::NOISE_CODEC_DECRYPTION_TIME,
     tcp::NOISE_CODEC_ENCRYPTION_SIZE,
     tcp::NOISE_CODEC_DECRYPTION_SIZE,
-    tcp::TCP_GATEWAY,
 ];
 
 pub mod bft {
@@ -75,5 +74,4 @@ pub mod tcp {
     pub const NOISE_CODEC_ENCRYPTION_SIZE: &str = "snarkos_tcp_noise_codec_encryption_size";
     pub const NOISE_CODEC_DECRYPTION_SIZE: &str = "snarkos_tcp_noise_codec_decryption_size";
     pub const TCP_TASKS: &str = "snarkos_tcp_tasks_total";
-    pub const TCP_GATEWAY: &str = "snarkos_tcp_gateway_messages_received";
 }
diff --git a/node/src/validator/mod.rs b/node/src/validator/mod.rs
index 7ac250f1a..efb8003f6 100644
--- a/node/src/validator/mod.rs
+++ b/node/src/validator/mod.rs
@@ -16,7 +16,7 @@ mod router;
 
 use crate::traits::NodeInterface;
 use snarkos_account::Account;
-use snarkos_node_bft::{helpers::init_primary_channels, ledger_service::CoreLedgerService};
+use snarkos_node_bft::{helpers::init_primary_channels, ledger_service::CoreLedgerService, spawn_blocking};
 use snarkos_node_consensus::Consensus;
 use snarkos_node_rest::Rest;
 use snarkos_node_router::{
@@ -386,15 +386,16 @@ impl<N: Network, C: ConsensusStorage<N>> Validator<N, C> {
                 // Prepare the inputs.
                 let inputs = [Value::from(Literal::Address(self_.address())), Value::from(Literal::U64(U64::new(1)))];
                 // Execute the transaction.
-                let transaction = match self_.ledger.vm().execute(
-                    self_.private_key(),
+                let self__ = self_.clone();
+                let transaction = match spawn_blocking!(self__.ledger.vm().execute(
+                    self__.private_key(),
                     locator,
                     inputs.into_iter(),
                     None,
                     10_000,
                     None,
                     &mut rand::thread_rng(),
-                ) {
+                )) {
                     Ok(transaction) => transaction,
                     Err(error) => {
                         error!("Transaction pool encountered an execution error - {error}");
