diff --git a/circuits/src/fields/base/ternary.rs b/circuits/src/fields/base/ternary.rs
index f6b20d8bd..fd893ae5a 100644
--- a/circuits/src/fields/base/ternary.rs
+++ b/circuits/src/fields/base/ternary.rs
@@ -29,22 +29,16 @@ impl<E: Environment> Ternary for BaseField<E> {
                 false => second.clone(),
             }
         }
-        // // Constant `first` and `second`
-        // else if first.is_constant() && second.is_constant() {
-        //     let not_condition = BaseField::from(&!condition);
-        //     let condition = BaseField::from(&condition);
-        //     (condition * first) + (not_condition * second)
-        // }
+        // Constant `first` and `second`
+        else if first.is_constant() && second.is_constant() {
+            let not_condition = BaseField::from(&!condition);
+            let condition = BaseField::from(&condition);
+            (condition * first) + (not_condition * second)
+        }
         // Variables
         else {
-            // Determine the variable mode.
-            let mode = match first.is_constant() && second.is_constant() {
-                true => Mode::Constant,
-                false => Mode::Private,
-            };
-
             // Initialize the witness.
-            let witness = BaseField::new(mode, match condition.eject_value() {
+            let witness = BaseField::new(Mode::Private, match condition.eject_value() {
                 true => first.eject_value(),
                 false => second.eject_value(),
             });
@@ -176,10 +170,10 @@ mod tests {
             let condition = Boolean::new(Mode::Public, true);
             Circuit::scoped("Public(true) ? Constant : Constant", |scope| {
                 let output = BaseField::ternary(&condition, &a, &b);
-                assert_eq!(1, scope.num_constants_in_scope());
+                assert_eq!(0, scope.num_constants_in_scope());
                 assert_eq!(0, scope.num_public_in_scope());
                 assert_eq!(0, scope.num_private_in_scope());
-                assert_eq!(1, scope.num_constraints_in_scope());
+                assert_eq!(0, scope.num_constraints_in_scope());
 
                 assert!(output.is_eq(&a).eject_value());
                 assert!(!output.is_eq(&b).eject_value());
@@ -188,10 +182,10 @@ mod tests {
             let condition = Boolean::new(Mode::Public, false);
             Circuit::scoped("Public(false) ? Constant : Constant", |scope| {
                 let output = BaseField::ternary(&condition, &a, &b);
-                assert_eq!(1, scope.num_constants_in_scope());
+                assert_eq!(0, scope.num_constants_in_scope());
                 assert_eq!(0, scope.num_public_in_scope());
                 assert_eq!(0, scope.num_private_in_scope());
-                assert_eq!(1, scope.num_constraints_in_scope());
+                assert_eq!(0, scope.num_constraints_in_scope());
 
                 assert!(!output.is_eq(&a).eject_value());
                 assert!(output.is_eq(&b).eject_value());
@@ -206,10 +200,10 @@ mod tests {
             let condition = Boolean::new(Mode::Private, true);
             Circuit::scoped("Private(true) ? Constant : Constant", |scope| {
                 let output = BaseField::ternary(&condition, &a, &b);
-                assert_eq!(1, scope.num_constants_in_scope());
+                assert_eq!(0, scope.num_constants_in_scope());
                 assert_eq!(0, scope.num_public_in_scope());
                 assert_eq!(0, scope.num_private_in_scope());
-                assert_eq!(1, scope.num_constraints_in_scope());
+                assert_eq!(0, scope.num_constraints_in_scope());
 
                 assert!(output.is_eq(&a).eject_value());
                 assert!(!output.is_eq(&b).eject_value());
@@ -218,10 +212,10 @@ mod tests {
             let condition = Boolean::new(Mode::Private, false);
             Circuit::scoped("Private(false) ? Constant : Constant", |scope| {
                 let output = BaseField::ternary(&condition, &a, &b);
-                assert_eq!(1, scope.num_constants_in_scope());
+                assert_eq!(0, scope.num_constants_in_scope());
                 assert_eq!(0, scope.num_public_in_scope());
                 assert_eq!(0, scope.num_private_in_scope());
-                assert_eq!(1, scope.num_constraints_in_scope());
+                assert_eq!(0, scope.num_constraints_in_scope());
 
                 assert!(!output.is_eq(&a).eject_value());
                 assert!(output.is_eq(&b).eject_value());
diff --git a/circuits/src/group/mul.rs b/circuits/src/group/mul.rs
index bd30c7591..7c3e9284e 100644
--- a/circuits/src/group/mul.rs
+++ b/circuits/src/group/mul.rs
@@ -278,9 +278,9 @@ mod tests {
             let b = ScalarField::<Circuit>::new(Mode::Public, scalar);
 
             let name = format!("Mul: a * b {}", i);
-            check_mul(&name, &expected, &a, &b, 2259, 0, 0, 502);
+            check_mul(&name, &expected, &a, &b, 757, 0, 2500, 2500);
             let name = format!("MulAssign: a * b {}", i);
-            check_mul_assign(&name, &expected, &a, &b, 2259, 0, 0, 502);
+            check_mul_assign(&name, &expected, &a, &b, 757, 0, 2500, 2500);
         }
     }
 
@@ -295,9 +295,9 @@ mod tests {
             let b = ScalarField::<Circuit>::new(Mode::Private, scalar);
 
             let name = format!("Mul: a * b {}", i);
-            check_mul(&name, &expected, &a, &b, 2259, 0, 0, 502);
+            check_mul(&name, &expected, &a, &b, 757, 0, 2500, 2500);
             let name = format!("MulAssign: a * b {}", i);
-            check_mul_assign(&name, &expected, &a, &b, 2259, 0, 0, 502);
+            check_mul_assign(&name, &expected, &a, &b, 757, 0, 2500, 2500);
         }
     }
 
diff --git a/circuits/src/group/ternary.rs b/circuits/src/group/ternary.rs
index 7419f7940..d47648c8d 100644
--- a/circuits/src/group/ternary.rs
+++ b/circuits/src/group/ternary.rs
@@ -168,14 +168,14 @@ mod tests {
         let condition = Boolean::<Circuit>::new(Mode::Public, false);
         let a = Affine::<Circuit>::new(Mode::Constant, first.to_x_coordinate(), None);
         let b = Affine::<Circuit>::new(Mode::Constant, second.to_x_coordinate(), None);
-        check_ternary("false ? Constant : Constant", expected, condition, a, b, 2, 0, 0, 2);
+        check_ternary("false ? Constant : Constant", expected, condition, a, b, 0, 0, 0, 0);
 
         // true ? Constant : Constant
         let expected = first;
         let condition = Boolean::<Circuit>::new(Mode::Public, true);
         let a = Affine::<Circuit>::new(Mode::Constant, first.to_x_coordinate(), None);
         let b = Affine::<Circuit>::new(Mode::Constant, second.to_x_coordinate(), None);
-        check_ternary("true ? Constant : Constant", expected, condition, a, b, 2, 0, 0, 2);
+        check_ternary("true ? Constant : Constant", expected, condition, a, b, 0, 0, 0, 0);
     }
 
     #[test]
@@ -222,14 +222,14 @@ mod tests {
         let condition = Boolean::<Circuit>::new(Mode::Private, false);
         let a = Affine::<Circuit>::new(Mode::Constant, first.to_x_coordinate(), None);
         let b = Affine::<Circuit>::new(Mode::Constant, second.to_x_coordinate(), None);
-        check_ternary("false ? Constant : Constant", expected, condition, a, b, 2, 0, 0, 2);
+        check_ternary("false ? Constant : Constant", expected, condition, a, b, 0, 0, 0, 0);
 
         // true ? Constant : Constant
         let expected = first;
         let condition = Boolean::<Circuit>::new(Mode::Private, true);
         let a = Affine::<Circuit>::new(Mode::Constant, first.to_x_coordinate(), None);
         let b = Affine::<Circuit>::new(Mode::Constant, second.to_x_coordinate(), None);
-        check_ternary("true ? Constant : Constant", expected, condition, a, b, 2, 0, 0, 2);
+        check_ternary("true ? Constant : Constant", expected, condition, a, b, 0, 0, 0, 0);
     }
 
     #[test]
