diff --git a/vm/cli/commands/run.rs b/vm/cli/commands/run.rs
index fc101e39f..8ba7e9940 100644
--- a/vm/cli/commands/run.rs
+++ b/vm/cli/commands/run.rs
@@ -20,10 +20,8 @@ pub const LOCALE: &num_format::Locale = &num_format::Locale::en;
 #[derive(Debug, Parser)]
 pub struct Run {
     /// The function name.
-    #[clap(long)]
     function: Identifier<CurrentNetwork>,
     /// The function inputs.
-    #[clap(long)]
     inputs: Vec<Value<CurrentNetwork>>,
     /// Uses the specified endpoint.
     #[clap(long)]
diff --git a/vm/file/readme_file.rs b/vm/file/readme_file.rs
index d248cfffe..687780ca8 100644
--- a/vm/file/readme_file.rs
+++ b/vm/file/readme_file.rs
@@ -45,7 +45,12 @@ impl README {
 
 To compile this Aleo program, run:
 ```bash
-aleo build
+snarkvm build
+```
+
+To execute this Aleo program, run:
+```bash
+snarkvm run hello
 ```
 "
         );
