diff --git a/synthesizer/src/process/trace/inclusion/execute.rs b/synthesizer/src/process/trace/inclusion/execute.rs
index 8e6526c2a..e82acb3d4 100644
--- a/synthesizer/src/process/trace/inclusion/execute.rs
+++ b/synthesizer/src/process/trace/inclusion/execute.rs
@@ -47,26 +47,22 @@ macro_rules! prepare_execution_impl {
                         let local_state_root = (*transaction_tree.root()).into();
 
                         // Construct the state path.
-                        let state_path = match task.is_local {
-                            true => {
+                        let state_path = match &task.local {
+                            Some((transaction_leaf, transition_path, transition_leaf)) => {
                                 // Compute the transaction path.
                                 let transaction_path =
-                                    transaction_tree.prove(transition_index, &transaction_leaf.to_bits_le())?;
-                                // Compute the transition leaf.
-                                let transition_leaf = task.leaf;
-                                // Compute the transition path.
-                                let transition_path = transition.to_path(&transition_leaf)?;
+                                    transaction_tree.prove(transaction_leaf.index() as usize, &transaction_leaf.to_bits_le())?;
                                 // Output the state path.
                                 StatePath::new_local(
                                     global_state_root,
                                     local_state_root,
                                     transaction_path,
-                                    transaction_leaf,
-                                    transition_path,
-                                    transition_leaf,
+                                    *transaction_leaf,
+                                    transition_path.clone(),
+                                    *transition_leaf,
                                 )?
                             }
-                            false => {
+                            None => {
                                 $query.$get_state_path_for_commitment(&task.commitment)
                                 $(.$await)?
                             }?
@@ -84,7 +80,7 @@ macro_rules! prepare_execution_impl {
                             task.gamma,
                             task.serial_number,
                             local_state_root,
-                            !task.is_local,
+                            task.local.is_none(), // Equivalent to 'is_global'
                         );
 
                         // Add the assignment to the assignments.
diff --git a/synthesizer/src/process/trace/inclusion/fee.rs b/synthesizer/src/process/trace/inclusion/fee.rs
index dfb191aee..1dd289819 100644
--- a/synthesizer/src/process/trace/inclusion/fee.rs
+++ b/synthesizer/src/process/trace/inclusion/fee.rs
@@ -53,7 +53,7 @@ macro_rules! prepare_fee_impl {
                         task.gamma,
                         task.serial_number,
                         local_state_root,
-                        !task.is_local,
+                        task.local.is_none(), // Equivalent to 'is_global'.
                     );
 
                     // Add the assignment to the assignments.
diff --git a/synthesizer/src/process/trace/inclusion/mod.rs b/synthesizer/src/process/trace/inclusion/mod.rs
index 4c6c5f2de..3511b06c5 100644
--- a/synthesizer/src/process/trace/inclusion/mod.rs
+++ b/synthesizer/src/process/trace/inclusion/mod.rs
@@ -24,7 +24,7 @@ use crate::{
 };
 use console::{
     network::prelude::*,
-    program::{InputID, StatePath, TransactionLeaf, TransitionLeaf, TRANSACTION_DEPTH},
+    program::{InputID, StatePath, TransactionLeaf, TransitionLeaf, TransitionPath, TRANSACTION_DEPTH},
     types::{Field, Group},
 };
 
@@ -38,18 +38,17 @@ struct InputTask<N: Network> {
     gamma: Group<N>,
     /// The serial number.
     serial_number: Field<N>,
-    /// The transition leaf.
-    leaf: TransitionLeaf<N>,
-    /// A boolean indicating whether the input was produced by the current transaction.
-    is_local: bool,
+    /// Contains the local transaction leaf, local transition path, and local transition leaf,
+    /// if this input is a record from a previous local transition.
+    local: Option<(TransactionLeaf<N>, TransitionPath<N>, TransitionLeaf<N>)>,
 }
 
 #[derive(Clone, Debug, Default)]
 pub(super) struct Inclusion<N: Network> {
-    /// A map of transition IDs to a list of input tasks.
+    /// A map of `transition IDs` to a list of `input tasks`.
     input_tasks: HashMap<N::TransitionID, Vec<InputTask<N>>>,
-    /// A map of commitments to (transition ID, output index) pairs.
-    output_commitments: HashMap<Field<N>, (N::TransitionID, u8)>,
+    /// A map of `commitments` to `(local transaction leaf, local transition path, local transition leaf)` pairs.
+    output_commitments: HashMap<Field<N>, (TransactionLeaf<N>, TransitionPath<N>, TransitionLeaf<N>)>,
 }
 
 impl<N: Network> Inclusion<N> {
@@ -65,11 +64,14 @@ impl<N: Network> Inclusion<N> {
             bail!("Inclusion expected the same number of input IDs as transition inputs")
         }
 
+        // Retrieve the transition index.
+        let transition_index = self.input_tasks.len() as u16;
+
         // Initialize the input tasks.
         let input_tasks = self.input_tasks.entry(*transition.id()).or_default();
 
         // Process the inputs.
-        for (index, (input, input_id)) in transition.inputs().iter().zip_eq(input_ids).enumerate() {
+        for input_id in input_ids {
             // Filter the inputs for records.
             if let InputID::Record(commitment, gamma, serial_number, ..) = input_id {
                 // Add the record to the input tasks.
@@ -77,8 +79,7 @@ impl<N: Network> Inclusion<N> {
                     commitment: *commitment,
                     gamma: *gamma,
                     serial_number: *serial_number,
-                    leaf: input.to_transition_leaf(index as u8),
-                    is_local: self.output_commitments.contains_key(commitment),
+                    local: self.output_commitments.get(commitment).cloned(),
                 });
             }
         }
@@ -87,8 +88,14 @@ impl<N: Network> Inclusion<N> {
         for (index, output) in transition.outputs().iter().enumerate() {
             // Filter the outputs for records.
             if let Output::Record(commitment, ..) = output {
-                // Add the record to the output commitments.
-                self.output_commitments.insert(*commitment, (*transition.id(), (input_ids.len() + index) as u8));
+                // Compute the transaction leaf.
+                let transaction_leaf = TransactionLeaf::new_execution(transition_index, **transition.id());
+                // Compute the transition leaf.
+                let transition_leaf = output.to_transition_leaf((input_ids.len() + index) as u8);
+                // Compute the transition path.
+                let transition_path = transition.to_path(&transition_leaf)?;
+                // Add the record's local Merklization to the output commitments.
+                self.output_commitments.insert(*commitment, (transaction_leaf, transition_path, transition_leaf));
             }
         }
 
diff --git a/synthesizer/src/vm/mod.rs b/synthesizer/src/vm/mod.rs
index 4c8556ca2..f0551f2d4 100644
--- a/synthesizer/src/vm/mod.rs
+++ b/synthesizer/src/vm/mod.rs
@@ -712,7 +712,8 @@ finalize getter:
 
     #[test]
     fn test_load_deployments_with_imports() {
-        let rng = &mut TestRng::default();
+        // NOTE: This seed was chosen for the CI's RNG to ensure that the test passes.
+        let rng = &mut TestRng::fixed(987654321);
 
         // Initialize a new caller.
         let caller_private_key = crate::vm::test_helpers::sample_genesis_private_key(rng);
@@ -806,4 +807,70 @@ function a:
         // Enforce that the VM can load properly with the imports.
         assert!(VM::from(vm.store.clone()).is_ok());
     }
+
+    #[test]
+    fn test_multiple_external_calls() {
+        let rng = &mut TestRng::default();
+
+        // Initialize a new caller.
+        let caller_private_key = crate::vm::test_helpers::sample_genesis_private_key(rng);
+        let caller_view_key = ViewKey::try_from(&caller_private_key).unwrap();
+        let address = Address::try_from(&caller_private_key).unwrap();
+
+        // Initialize the genesis block.
+        let genesis = crate::vm::test_helpers::sample_genesis_block(rng);
+
+        // Fetch the unspent records.
+        let records =
+            genesis.transitions().cloned().flat_map(Transition::into_records).take(3).collect::<IndexMap<_, _>>();
+        trace!("Unspent Records:\n{:#?}", records);
+        let record_0 = records.values().next().unwrap().decrypt(&caller_view_key).unwrap();
+        let record_1 = records.values().nth(1).unwrap().decrypt(&caller_view_key).unwrap();
+        let record_2 = records.values().nth(2).unwrap().decrypt(&caller_view_key).unwrap();
+
+        // Initialize the VM.
+        let vm = sample_vm();
+        // Update the VM.
+        vm.add_next_block(&genesis).unwrap();
+
+        // Deploy the program.
+        let program = Program::from_str(
+            r"
+import credits.aleo;
+
+program test_multiple_external_calls.aleo;
+
+function multitransfer:
+    input r0 as credits.aleo/credits.record;
+    input r1 as address.private;
+    input r2 as u64.private;
+    call credits.aleo/transfer_private r0 r1 r2 into r3 r4;
+    call credits.aleo/transfer_private r4 r1 r2 into r5 r6;
+    output r4 as credits.aleo/credits.record;
+    output r5 as credits.aleo/credits.record;
+    output r6 as credits.aleo/credits.record;
+    ",
+        )
+        .unwrap();
+        let deployment = vm.deploy(&caller_private_key, &program, (record_0, 1), None, rng).unwrap();
+        vm.add_next_block(&sample_next_block(&vm, &caller_private_key, &[deployment], rng).unwrap()).unwrap();
+
+        // Execute the programs.
+        let inputs = [
+            Value::<Testnet3>::Record(record_1),
+            Value::<Testnet3>::from_str(&address.to_string()).unwrap(),
+            Value::<Testnet3>::from_str("10u64").unwrap(),
+        ];
+        let execution = vm
+            .execute(
+                &caller_private_key,
+                ("test_multiple_external_calls.aleo", "multitransfer"),
+                inputs.into_iter(),
+                Some((record_2, 1)),
+                None,
+                rng,
+            )
+            .unwrap();
+        vm.add_next_block(&sample_next_block(&vm, &caller_private_key, &[execution], rng).unwrap()).unwrap();
+    }
 }
