diff --git a/node/router/src/lib.rs b/node/router/src/lib.rs
index 4cec32037..c393c7b89 100644
--- a/node/router/src/lib.rs
+++ b/node/router/src/lib.rs
@@ -128,7 +128,7 @@ pub struct Router<N: Network> {
 #[rustfmt::skip]
 impl<N: Network> Router<N> {
     /// The maximum duration in seconds permitted for establishing a connection with a node, before dropping the connection.
-    const CONNECTION_TIMEOUT_IN_MILLIS: u64 = 210; // 3.5 minutes
+    const CONNECTION_TIMEOUT_IN_MILLIS: u64 = 1000;
     /// The duration in seconds to sleep in between heartbeat executions.
     const HEARTBEAT_IN_SECS: u64 = 9; // 9 seconds
     /// The maximum number of candidate peers permitted to be stored in the node.
diff --git a/node/store/src/rocksdb/map.rs b/node/store/src/rocksdb/map.rs
index 205e29f37..992d883d9 100644
--- a/node/store/src/rocksdb/map.rs
+++ b/node/store/src/rocksdb/map.rs
@@ -56,7 +56,7 @@ impl<
                 // Prepare the prefixed key and serialized value.
                 let raw_key = self.create_prefixed_key(&key)?;
                 let raw_value = bincode::serialize(&value)?;
-                self.database.put(&raw_key, &raw_value)?;
+                self.database.put(raw_key, raw_value)?;
             }
         }
 
@@ -79,7 +79,7 @@ impl<
             false => {
                 // Prepare the prefixed key.
                 let raw_key = self.create_prefixed_key(key)?;
-                self.database.delete(&raw_key)?;
+                self.database.delete(raw_key)?;
             }
         }
 
