diff --git a/.resources/release-version b/.resources/release-version
index b02d37b0f..6eaf89433 100644
--- a/.resources/release-version
+++ b/.resources/release-version
@@ -1 +1 @@
-v2.0.2
\ No newline at end of file
+v2.0.0
\ No newline at end of file
diff --git a/src/rpc/rpc.rs b/src/rpc/rpc.rs
index 2b243bc5d..5fb5b669d 100644
--- a/src/rpc/rpc.rs
+++ b/src/rpc/rpc.rs
@@ -442,7 +442,7 @@ mod tests {
         LedgerState,
     };
     use snarkvm::{
-        dpc::{testnet2::Testnet2, AccountScheme, AleoAmount, Transaction, Transactions, Transition},
+        dpc::{testnet2::Testnet2, AleoAmount, Transaction, Transactions, Transition},
         prelude::{Account, Block, BlockHeader},
         utilities::ToBytes,
     };
