diff --git a/console/algorithms/src/poseidon/helpers/sponge.rs b/console/algorithms/src/poseidon/helpers/sponge.rs
index 6401bc1df..ffa6002ca 100644
--- a/console/algorithms/src/poseidon/helpers/sponge.rs
+++ b/console/algorithms/src/poseidon/helpers/sponge.rs
@@ -22,7 +22,7 @@ use snarkvm_console_types::{prelude::*, Field};
 use snarkvm_fields::PoseidonParameters;
 
 use smallvec::SmallVec;
-use std::sync::Arc;
+use std::{ops::DerefMut, sync::Arc};
 
 /// A duplex sponge based using the Poseidon permutation.
 ///
@@ -113,12 +113,14 @@ impl<E: Environment, const RATE: usize, const CAPACITY: usize> PoseidonSponge<E,
         // Full rounds apply the S Box (x^alpha) to every element of state
         if is_full_round {
             for elem in self.state.iter_mut() {
-                *elem = elem.pow(Field::from_u64(self.parameters.alpha));
+                let e = elem.deref_mut();
+                *e = e.pow([self.parameters.alpha]);
             }
         }
         // Partial rounds apply the S Box (x^alpha) to just the first element of state
         else {
-            self.state[0] = self.state[0].pow(Field::from_u64(self.parameters.alpha));
+            let e = self.state[0].deref_mut();
+            *e = e.pow(&[self.parameters.alpha]);
         }
     }
 
diff --git a/console/network/environment/src/prelude.rs b/console/network/environment/src/prelude.rs
index 1540cf0cd..4b237ec58 100644
--- a/console/network/environment/src/prelude.rs
+++ b/console/network/environment/src/prelude.rs
@@ -50,6 +50,7 @@ pub use core::{
         BitXor,
         BitXorAssign,
         Deref,
+        DerefMut,
         Div,
         DivAssign,
         Mul,
diff --git a/console/types/field/src/lib.rs b/console/types/field/src/lib.rs
index 614e82565..5973b89fb 100644
--- a/console/types/field/src/lib.rs
+++ b/console/types/field/src/lib.rs
@@ -107,3 +107,10 @@ impl<E: Environment> Deref for Field<E> {
         &self.field
     }
 }
+
+impl<E: Environment> DerefMut for Field<E> {
+    #[inline]
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.field
+    }
+}
diff --git a/curves/src/bls12_377/fq.rs b/curves/src/bls12_377/fq.rs
index 9abc25f57..3d29b6017 100644
--- a/curves/src/bls12_377/fq.rs
+++ b/curves/src/bls12_377/fq.rs
@@ -34,52 +34,52 @@ impl FftParameters for FqParameters {
     type BigInteger = BigInteger;
 
     #[rustfmt::skip]
-    const POWERS_OF_G: &'static [BigInteger] = &[
-        BigInteger([7385130745848031862, 701542640293853790, 7860911539678904085, 2081126899580975139, 1314744540578258676, 5505385934676702]),
-        BigInteger([5134937160227832732, 2527472353192255794, 5983990541424952770, 12889027417780705039, 1827101640268914115, 35921860030846829]),
-        BigInteger([17978844473626023153, 17876791103060992241, 14280117436792259086, 6075183100816533711, 14380438162445701085, 23634273502926963]),
-        BigInteger([3042313168607755652, 2898683936875278397, 18142852711952420851, 3299128674474358658, 5587069763096216413, 57910226810639103]),
-        BigInteger([8670062213291950983, 13231535783847415120, 13853374098618609475, 4375084111178468297, 15058863618213473729, 104302990866566346]),
-        BigInteger([6782246736218306282, 13202884221908298588, 17690457268828644541, 1977382937194332749, 1043164677067163225, 66006011576170137]),
-        BigInteger([14729707132561216529, 7604534147383175560, 8315993351605680848, 15266138444828105666, 6205165579526030584, 22022460061138669]),
-        BigInteger([6933372435512853879, 17628910870098511585, 10234767657922363124, 14806887497637517507, 16162400383743571186, 72444598743672739]),
-        BigInteger([8065937948938570075, 12349274400625613597, 9974581888245964682, 12097593823039447971, 5746133023516248483, 20558105834535994]),
-        BigInteger([6276868676555924960, 11048343163423705260, 13200118183172813340, 11888317846591950664, 12394546186160613838, 53596618024763257]),
-        BigInteger([13476941164363887320, 6529087504189619549, 13078956481460091614, 14557775281486146673, 5487124721358523033, 10819738202564788]),
-        BigInteger([13364818876935230670, 13269779243477965971, 15761372403346801792, 7427170436970270783, 8383781585990660556, 23078685542227851]),
-        BigInteger([17619255348814714837, 7566977047604138306, 16444835836418231544, 7587307705081412631, 13159035542584219759, 89641791553270776]),
-        BigInteger([13411296197887831162, 949773516635351564, 17264420250194156204, 7803749772625501627, 7722287160777812231, 92384713559664914]),
-        BigInteger([6568791503267352999, 11293051525851483970, 5706681168726160073, 11095454440956603812, 5019415166258911703, 35364342249184590]),
-        BigInteger([7063460146736001849, 3851981843724025120, 4404051240421452066, 8747482441359976867, 11928295146298204946, 15143911508191343]),
-        BigInteger([11326837853769895854, 12895752614942843615, 18082893035929964238, 6645302761511957642, 13726634636068843846, 44863353895094001]),
-        BigInteger([11417549255766836924, 6538651435163630262, 13199868510547063205, 2112495793012567303, 11909382513093093621, 75895599265178156]),
-        BigInteger([8401643149461645733, 12537573869478123761, 4341358729788488908, 10710884481768240625, 12971282953629475608, 45112253026753539]),
-        BigInteger([1152007377126551531, 9007601379630786885, 1437653439713848702, 8068372743214382487, 5083343468186829292, 60684180583630844]),
-        BigInteger([2248450227402695479, 14072713940474318953, 15235079680217091076, 10148034663319658399, 10948083137235044598, 49146251091271825]),
-        BigInteger([8841623610086649495, 5088757700890560195, 9378044910544196932, 10224534985235096112, 9183104370154873548, 23033958700638443]),
-        BigInteger([1631173013104702089, 7590789119163414124, 5824062274472647016, 17863642524987727381, 4255407151711471511, 64651884988346985]),
-        BigInteger([9310047075170506565, 2104747671456310692, 2644219065673981138, 4856291027038128876, 8180550020249516800, 106383048458869756]),
-        BigInteger([337370040796255196, 5211605300302596945, 7148574707434112052, 15794714402959134312, 15972085473587326639, 37109373511278407]),
-        BigInteger([11471650965284872747, 2409290254183797046, 12611606615659248625, 352120820464993320, 17145040817448456307, 76649058183435327]),
-        BigInteger([628222360320685885, 3316269393081920675, 8362470787126080828, 5039133592514299306, 14593231448297904645, 66448120101990539]),
-        BigInteger([17207968786480344296, 15238487662000849007, 13594549201811747429, 1023321563714992538, 1640889964080741822, 93804327810640416]),
-        BigInteger([7175959093718815137, 8275715052969182726, 6436112820578665714, 8138571789395785875, 6050826606617757421, 96865101988354652]),
-        BigInteger([4070177388613442533, 17956834252521987703, 6950522450157943907, 12697584603990425156, 16587871642483005244, 57471560345127555]),
-        BigInteger([10900985717020507711, 5295852191758189818, 11820298144983500636, 550003973110951803, 263308017339972560, 111125542229885816]),
-        BigInteger([6096269296977312113, 9933768294483763101, 3745456635322318425, 1234445854684469083, 6265727993591295891, 90344874383011427]),
-        BigInteger([14266056396848430283, 15657563994013446202, 2661332711830601499, 9394576428684896628, 2717774174090368664, 8155723651091523]),
-        BigInteger([10525954341863739652, 9230764219066024507, 1476128509584006807, 9208851952254982823, 3442027277949361561, 31949044617252679]),
-        BigInteger([511600968522897682, 7208634953489512620, 3530006720424357148, 14813261074251638594, 12630011612421075990, 88930841157402859]),
-        BigInteger([2873670194202000053, 316233994604135445, 3504079781682637435, 16475221371051119220, 5511724617596116755, 48306385732647534]),
-        BigInteger([18390923601239652091, 8383328236368904633, 16870035286970652539, 1713556008324362517, 10834798082861818717, 76626671700720801]),
-        BigInteger([6453880472035760162, 9279136170086455551, 5255158191281631096, 9402683820780075235, 13294404117759409916, 28911609878178747]),
-        BigInteger([15457678214934104727, 10257410602787968791, 10543920980663046204, 12873421307184168602, 3005171103040729111, 6608332313712721]),
-        BigInteger([7669310927744901814, 10180432966734195262, 7102068305190679471, 3394102346354943911, 7728984932024735354, 113393442359243117]),
-        BigInteger([332121547564755110, 4058914442193592743, 3261683794795430975, 14314300534785368151, 12189004563146915906, 76871270605085138]),
-        BigInteger([5666933051403538098, 9026061949184173996, 2715769564174870087, 15173756192814126609, 9536623932835594166, 62873430292373036]),
-        BigInteger([13313210803279475255, 648193293732931258, 17506987073671981935, 17186037581194467208, 9596589497284832987, 53781961547701120]),
-        BigInteger([7836246735827273146, 13870095715782950813, 14107419625140408226, 4952953651681455070, 12502319229901919751, 7811766342450446]),
-        BigInteger([7959784605059980387, 11049727206760807822, 13611648078312050530, 4612698621573455553, 13890566741340831409, 101342263646697418]),
+    const POWERS_OF_ROOTS_OF_UNITY: &'static [BigInteger] = &[
+        BigInteger([2022196864061697551, 17419102863309525423, 8564289679875062096, 17152078065055548215, 17966377291017729567, 68610905582439508]),
+        BigInteger([1097558877362951325, 10990216524069526157, 1703764964909126124, 5337477398168494451, 8388789446441546832, 83780176967333390]),
+        BigInteger([6021325183928272861, 3071743451473281050, 17793609045682579055, 4499483110946763311, 3694917660005904773, 35840689841924604]),
+        BigInteger([4148421423508782473, 16486417951907259381, 13681320370461884655, 9543156358084556689, 15654115908807492478, 85178773442619605]),
+        BigInteger([4440291260215189129, 8134083368238958870, 5850770822935137603, 18388642548453398432, 4033689956485938698, 96943229116346922]),
+        BigInteger([13321959743753581084, 13957307756377123667, 8022065601003020126, 15105587473044942004, 3436102915385715417, 16497947308509307]),
+        BigInteger([12740765449271843915, 13529810863058982449, 15988003695485397690, 5204094055301492127, 12204797504641322421, 54884696873855227]),
+        BigInteger([7200341240129221640, 128866105844736666, 1972097482351781478, 7508123272393265315, 18134316519918384483, 90386983337599854]),
+        BigInteger([11182828541229423812, 3352059565776569042, 745605881396983371, 10544136951768851167, 13944324842745283530, 76153606183323606]),
+        BigInteger([5451839093323299025, 11344095019039321738, 3696612540108785619, 849808749100130193, 7737394061630388518, 41447732009599376]),
+        BigInteger([4914435911578270728, 11442571726223078752, 200783168592117624, 5223206539372270413, 12471543125846282538, 14780043884677376]),
+        BigInteger([6835942096487989968, 9916740166863458597, 15149601694743014865, 3251390487756549529, 16202626875757167903, 78637381314787701]),
+        BigInteger([10024136201786834022, 2737410832464918544, 11569195233337582907, 6882399291950297037, 17956512286065493665, 87559046345038795]),
+        BigInteger([7787906257689135971, 4007925209220880779, 5799259275421890332, 5651393587803319504, 5361114559272044762, 78591763064139593]),
+        BigInteger([8749978828541193864, 7615081815133949072, 13070992874042179472, 1364907361524039183, 10109099343823847681, 104282651814875028]),
+        BigInteger([16730838284865363583, 15400461195736119534, 4032519462999757033, 15551203984340620330, 382650037072032689, 76641657953301822]),
+        BigInteger([8166159088317375095, 1704530912566839702, 13372111531284600631, 9715437978473519961, 4887487291063440487, 6286512958015546]),
+        BigInteger([1768534497835661159, 1244838626041538416, 9153540930636989172, 12437852256056737422, 2519703766117714353, 37580752840997918]),
+        BigInteger([13468864600499712370, 15009260878077565, 14291559461807896332, 9284492953958434183, 18315507429112290509, 41332741727179643]),
+        BigInteger([8269976866043356680, 16766307433795223092, 339184897717558508, 9029309210784216270, 16947438621423748249, 1154969810162894]),
+        BigInteger([10598383875620927263, 1261087173785984910, 11498737931675819777, 8568819297979713299, 1166319621891739467, 53485818962535716]),
+        BigInteger([1592802484188722575, 16097103819145948294, 7922347041881118053, 14182139815326903761, 1080552542780789890, 84770178757176689]),
+        BigInteger([17305426336106269005, 8862145762498523999, 13494918675279718853, 3717788643186382361, 10515687947609219968, 112422016356963276]),
+        BigInteger([10874618176231306138, 13631539808973743734, 11959456582258358795, 13957665559292951976, 7355794120017380469, 51180796759257049]),
+        BigInteger([9944918783779352295, 11674736477566853654, 4430181012666063989, 5365556697595744810, 13024803618277831802, 80199608722292197]),
+        BigInteger([11841292144630778493, 3327346977638983511, 9717945600717261357, 17521983842610332579, 17551948704477232338, 38629783630845674]),
+        BigInteger([13053061452543995362, 1573516732916099269, 7734072523617387500, 15879247754331170173, 4985704564687245967, 50017930811101333]),
+        BigInteger([5968032006797444201, 4356970211500493920, 14567417169078198555, 5589343791883175912, 13942140483976875481, 77532733089031308]),
+        BigInteger([1520419837524800053, 4505352024184739223, 11163238693972155221, 14849828818714008227, 2337273235727598583, 119043856166660542]),
+        BigInteger([17856003618470058414, 17971666744702441889, 17477339519519243281, 1190648955970084811, 8535065801186549950, 88202116378298476]),
+        BigInteger([3613548573764529932, 17312871435250531950, 14583594317345501622, 15714261421593485983, 3619881023171270578, 102701416776282826]),
+        BigInteger([2906047431783924967, 3246340428391821303, 9728912894597247012, 14160632400906136931, 17644388830484393402, 38484906188769515]),
+        BigInteger([17914929175761583691, 10159064175292721773, 17508169904588375801, 231704298035432102, 16645242204034044467, 72146101474673684]),
+        BigInteger([3958533722314001392, 10367882774307637822, 12047882728743649441, 405962725023303708, 2571821397758431564, 49080211914402704]),
+        BigInteger([590761718662001177, 7169147383307934023, 3111750625130109315, 3509576743298836380, 9392819985844593396, 2903460257980110]),
+        BigInteger([10894310132299876932, 7222012151301227611, 7759303741675735532, 18168258420254565546, 1036933145679827201, 84980589105836606]),
+        BigInteger([6223094491993811174, 10029382893184057307, 1782107595179881914, 4749850940526401858, 17559628252757726982, 20954496417456641]),
+        BigInteger([9716738584996322209, 2920283958606294680, 5037648889630024082, 10095866865449441757, 15798169953625136889, 75996191339425194]),
+        BigInteger([4358863173918119755, 15055260416528219604, 14884179962427098601, 16487815824678336974, 9163017981789726811, 95112432538757947]),
+        BigInteger([2011658397676046806, 15717875473168825780, 280862458342309401, 2045794038349364078, 15857429573515749114, 147064043817050]),
+        BigInteger([16760038912651623103, 790302385909476981, 6794528760947774914, 1897898981979607280, 9930859122245863541, 62625600058823796]),
+        BigInteger([6859520296144159969, 909173910965215762, 15967585281779797458, 10442593522875662584, 7918433044340205291, 104628130661177253]),
+        BigInteger([7212508657599325242, 1584945223518302852, 9963925208591887870, 11598974258151462965, 3264112703271587886, 20341848365764212]),
+        BigInteger([2142683228075401710, 5947621269725846341, 5696548463751032717, 13780290877820379740, 10702692037250602632, 33891554982109078]),
+        BigInteger([10965161018967488287, 18251363109856037426, 7036083669251591763, 16109345360066746489, 4679973768683352764, 96952949334633821]),
     ];
     #[rustfmt::skip]
     const TWO_ADICITY: u32 = 46u32;
@@ -191,26 +191,34 @@ impl PoseidonDefaultParameters for FqParameters {
 #[cfg(test)]
 mod tests {
     use super::*;
-    use snarkvm_fields::{Field, PrimeField};
+    use snarkvm_fields::{FftField, Field, PrimeField};
 
     #[test]
-    fn test_powers_of_g() {
+    fn test_powers_of_root_of_unity() {
         let two = Fq::from(2u8);
 
-        // Compute the expected powers of G.
-        let g = Fq::from_bigint(FqParameters::GENERATOR).unwrap().pow(FqParameters::T);
+        // Compute the expected powers of root of unity.
+        let root_of_unity = Fq::two_adic_root_of_unity();
         let powers = (0..FqParameters::TWO_ADICITY - 1)
-            .map(|i| g.pow(two.pow(Fq::from(i as u64).to_bigint()).to_bigint()).to_bigint())
+            .map(|i| root_of_unity.pow(two.pow(Fq::from(i as u64).to_bigint()).to_bigint()))
             .collect::<Vec<_>>();
+        assert_eq!(powers[0], Fq::two_adic_root_of_unity());
 
-        // Ensure the correct number of powers of G are present.
-        assert_eq!(FqParameters::POWERS_OF_G.len() as u64, (FqParameters::TWO_ADICITY - 1) as u64);
-        assert_eq!(FqParameters::POWERS_OF_G.len(), powers.len());
+        // Ensure the correct number of powers of root of unity are present.
+        assert_eq!(FqParameters::POWERS_OF_ROOTS_OF_UNITY.len() as u64, (FqParameters::TWO_ADICITY - 1) as u64);
+        assert_eq!(FqParameters::POWERS_OF_ROOTS_OF_UNITY.len(), powers.len());
 
         // Ensure the expected and candidate powers match.
-        for (expected, candidate) in powers.iter().zip(FqParameters::POWERS_OF_G.iter()) {
-            println!("{:?} =?= {:?}", expected, candidate);
-            assert_eq!(expected, candidate);
+        for (expected, candidate) in powers.iter().zip(FqParameters::POWERS_OF_ROOTS_OF_UNITY) {
+            // println!("BigInteger({:?}),", expected.0.0);
+            println!("{:?} =?= {:?}", expected.0, candidate);
+            assert_eq!(&expected.0, candidate);
         }
     }
+
+    #[test]
+    fn test_two_adic_root_of_unity() {
+        let expected = Fq::multiplicative_generator().pow(FqParameters::T);
+        assert_eq!(expected, Fq::two_adic_root_of_unity());
+    }
 }
diff --git a/curves/src/bls12_377/fr.rs b/curves/src/bls12_377/fr.rs
index 6ac88841d..8b01016a1 100644
--- a/curves/src/bls12_377/fr.rs
+++ b/curves/src/bls12_377/fr.rs
@@ -57,53 +57,53 @@ impl FftParameters for FrParameters {
     type BigInteger = BigInteger;
 
     #[rustfmt::skip]
-    const POWERS_OF_G: &'static [BigInteger] = &[
-        BigInteger([652114075796693467, 9243445511369889275, 4931911925841322402, 1202872507392441788]),
-        BigInteger([357278320042854805, 8679756035718620602, 24147076532703075, 386601462669077701]),
-        BigInteger([5821309058267914206, 930024428521638902, 15246330751380806139, 1156926535716486857]),
-        BigInteger([2632950189532968238, 5392535751966509925, 4941032747320406800, 606661836792814885]),
-        BigInteger([5797205653290972231, 12589652645565380589, 15756140458856054557, 1282965999660522983]),
-        BigInteger([8306707692583858925, 7335440127635038484, 80047149603102519, 999419446928193801]),
-        BigInteger([6956875524334552994, 13082051501117969572, 2938619391416456789, 304101838065490918]),
-        BigInteger([3134783599327598467, 16631357106850039899, 2627692882553178120, 1024359273584097450]),
-        BigInteger([12324015454749298429, 2042071882771645842, 7777946766613653260, 727393526865169839]),
-        BigInteger([9072551969993762373, 557399396585981236, 17262240838654823200, 748448633583709418]),
-        BigInteger([15745607881079653327, 10557409482890538181, 9754145699313741618, 1132312371963887010]),
-        BigInteger([7370900649931229405, 15276690995798305716, 13935407585395225192, 182158718734875642]),
-        BigInteger([6635723712198957876, 4459854046832330285, 2164992155939541098, 1061313009910823330]),
-        BigInteger([9866347584304354950, 13633573028704785617, 1490634813024064211, 1193920255548692291]),
-        BigInteger([12218842672787700946, 17502401118141046153, 3014192385295539746, 1278232049812798994]),
-        BigInteger([12946324424558980416, 14730603725897097862, 9093296730287121559, 1301088483395017614]),
-        BigInteger([13274457918106467667, 268045572525229395, 3096857327038933972, 791465315693154522]),
-        BigInteger([3037126518913560865, 2335854563834671212, 1382087909526519297, 252291075014288680]),
-        BigInteger([16107485408258731397, 1785252687040230207, 3542225586915910559, 939194661354943120]),
-        BigInteger([9097289758579870090, 13453801755062728985, 5601243069208241633, 755137600292698783]),
-        BigInteger([14650978806763043555, 1977920005064853667, 4380129240051161143, 696557356556620607]),
-        BigInteger([17358176504781348191, 14494404166002030320, 6366857619701608749, 307359208841810046]),
-        BigInteger([17831624937242459876, 3921541248490335982, 13076272174481132773, 203579329280294258]),
-        BigInteger([4827811091337682419, 2252535990376301773, 4834246201427258736, 176594086007515430]),
-        BigInteger([8224015011019235350, 3471968399227732089, 13945344924603372249, 63060411398793076]),
-        BigInteger([3722182982532906157, 12487091451829240696, 15973379993919888494, 933622446976592570]),
-        BigInteger([16918580269234120176, 9463814177789368183, 3230046188658025811, 700080667892842034]),
-        BigInteger([5528995323364708964, 4645547136665772307, 10740406109497613416, 1060204339835393709]),
-        BigInteger([9061129324622442628, 17862151211333623624, 184854872781714471, 22913953325058684]),
-        BigInteger([15739313880441174300, 6790670340938497210, 3392071052049343159, 164706088673074561]),
-        BigInteger([12245098422944482069, 9519432892459635938, 2900831850175999990, 1306683128084439414]),
-        BigInteger([18170289470227336398, 7960339046993766714, 1786191937113318291, 974715431061307237]),
-        BigInteger([14376073387189366589, 9083892742653733897, 7413368814740562107, 1229070466540458162]),
-        BigInteger([5040906166828092920, 1597932116119946833, 17873658496853978081, 1121069752338197226]),
-        BigInteger([3762856476273098914, 6524339332895362878, 17551910839959747290, 302725641766624789]),
-        BigInteger([117129974372144763, 12585572440832662321, 11599001791804262972, 228386842112432103]),
-        BigInteger([15180109745367336074, 4935483470718716768, 15393897344535774139, 138118718125512881]),
-        BigInteger([9320622113425242802, 1507038136306394275, 6249837229555579558, 707827944913503007]),
-        BigInteger([8510091037240551326, 2255407371696257485, 9086980551808272394, 1151908693325144507]),
-        BigInteger([8197034620141798687, 4334208919548785209, 13603393822593628560, 479050288045546609]),
-        BigInteger([6310980340485637787, 4470469036336467487, 10811712561993161098, 358887687171593677]),
-        BigInteger([1532731303765601343, 7737984259716464239, 18210871776416574727, 1086801557967392038]),
-        BigInteger([8069388237560588236, 1260332170967911028, 13997795246790076617, 325648630152787926]),
-        BigInteger([14375636960200474030, 3537274190141887733, 721323736305227360, 1057644659207653462]),
-        BigInteger([1117754170585778689, 2375847183000234400, 8144967840909054003, 522520945354137089]),
-        BigInteger([10311624665562349569, 14944710915545628673, 2588746559005780992, 0]),
+    const POWERS_OF_ROOTS_OF_UNITY: &'static [BigInteger] = &[
+        BigInteger([12646347781564978760, 6783048705277173164, 268534165941069093, 1121515446318641358]),
+        BigInteger([9908812190343590727, 989981212292874635, 13217848517394370016, 313220887743686251]),
+        BigInteger([4570942453055933478, 14327950181157151272, 8177734713484223276, 732395113570565582]),
+        BigInteger([7547026561639937065, 11733968610145256351, 8343216976141027051, 576989125389526455]),
+        BigInteger([15503117548499178639, 2308620569553928151, 16285062430639685160, 103769458534362861]),
+        BigInteger([8608407781632583430, 6271302370352858992, 1070338101025660973, 538942973387907962]),
+        BigInteger([6401736382236295208, 1153288719752352283, 14823640363840104239, 499246014633910619]),
+        BigInteger([17833850933386010261, 6090852137313170745, 12528041030915574153, 857737301371729770]),
+        BigInteger([1982048757083514280, 12158391317477180280, 13684570609270614441, 840786117129730731]),
+        BigInteger([18189047035874409266, 10063848728508956944, 18125989563719264563, 109154778261203350]),
+        BigInteger([4145860318443588715, 480045415555849201, 15185689511254420012, 997424544238423752]),
+        BigInteger([12380934610292147810, 14921791250714101079, 7953234348116523713, 740566716969780590]),
+        BigInteger([17523206016699816339, 8210268852326501515, 12102826939441359278, 69320453637394156]),
+        BigInteger([13184634841845083965, 5035641472528052881, 10848835132746879957, 1093292728521196096]),
+        BigInteger([3644591431389930536, 17535182174514028817, 2333780137994975086, 1211430046545197955]),
+        BigInteger([15764618857955124109, 17001304349044012093, 11249158843350648159, 1304234597518548509]),
+        BigInteger([17023740166556598291, 17327346947230052231, 15227607426000437488, 524356170404741496]),
+        BigInteger([15508095723619167236, 8916154698977858924, 4872918458681303657, 507246242966933127]),
+        BigInteger([13933261170268378635, 6975150440507062408, 285981162592700355, 909990229706841196]),
+        BigInteger([1796369225555089555, 18439169056424861976, 8554690303853538396, 739593369575156928]),
+        BigInteger([8849035624997270559, 11082174508102696985, 5797273297011026936, 458094565250654809]),
+        BigInteger([5944936778479914843, 369725202543584470, 22913969483319861, 1123237285519395196]),
+        BigInteger([8228193452833987056, 3247446711273005067, 7285244012699814841, 165660218100066147]),
+        BigInteger([16110965290454180518, 5418683100417587062, 11303731202360030209, 290121373407754357]),
+        BigInteger([10742665076645341070, 7056370150642984550, 12615465793934837989, 548498362185685512]),
+        BigInteger([8097456064175987351, 10585807130451162109, 5877015307964227995, 477368390704686837]),
+        BigInteger([15676770081073062471, 12262153699415403306, 7528677848461606057, 649917933945894399]),
+        BigInteger([18271947154240526633, 11103933895114620578, 16558660923152452318, 925331653431062360]),
+        BigInteger([3907589926240897769, 11045697709584246531, 90835143142482923, 870886311473863909]),
+        BigInteger([3528144693560331180, 11121597045360901952, 11405196888439666838, 670327805282996966]),
+        BigInteger([13456528212939925364, 9040660049171970165, 4374817133042500768, 422841521572866881]),
+        BigInteger([17220360707710309003, 7704697058569325505, 14703012148831508770, 113932329837224644]),
+        BigInteger([15456395681208503803, 6416534260312397167, 4409371950760456950, 855302230113322418]),
+        BigInteger([14648805411005782346, 659899539555743592, 13712138229698623536, 688628542983365223]),
+        BigInteger([17553691999655354057, 12414962864428128502, 12778610342023380950, 309906985245362430]),
+        BigInteger([13671771363231792901, 13550649581906397391, 14960300922341772840, 197239424530289859]),
+        BigInteger([14840183457516239942, 7651729362259515675, 11442121706837547178, 409184120412760908]),
+        BigInteger([5849560869621087513, 6649867980513935874, 16080766030443795594, 905356649678185654]),
+        BigInteger([13302147401216114106, 3611900969696901525, 3470557722594359823, 950175460673052306]),
+        BigInteger([16339580035428004219, 4845663562592066382, 6914211203719332907, 510216861370749871]),
+        BigInteger([2145222547635311650, 10230679325758658394, 7207487648880633871, 209805430256584658]),
+        BigInteger([9208812316068886018, 11169310505015993932, 8317257015293247865, 1090372203370214591]),
+        BigInteger([18408280738886485821, 14547869021980959191, 6047025381655759581, 725535874718029226]),
+        BigInteger([1670730338129557402, 7866724895305809984, 10026936948289003902, 781269171506367679]),
+        BigInteger([4882877962169421051, 8060560647985508350, 4729166814476724001, 1147730753089737444]),
+        BigInteger([5461406015399410446, 5014654494648953692, 8156709087178280082, 1299557346046566890]),
     ];
     #[rustfmt::skip]
     const TWO_ADICITY: u32 = 47;
@@ -207,26 +207,34 @@ impl PoseidonDefaultParameters for FrParameters {
 #[cfg(test)]
 mod tests {
     use super::*;
-    use snarkvm_fields::{Field, PrimeField};
+    use snarkvm_fields::{FftField, Field, PrimeField};
 
     #[test]
-    fn test_powers_of_g() {
+    fn test_powers_of_root_of_unity() {
         let two = Fr::from(2u8);
 
-        // Compute the expected powers of G.
-        let g = Fr::from_bigint(FrParameters::GENERATOR).unwrap().pow(FrParameters::T);
+        // Compute the expected powers of root of unity.
+        let root_of_unity = Fr::two_adic_root_of_unity();
         let powers = (0..FrParameters::TWO_ADICITY - 1)
-            .map(|i| g.pow(two.pow(Fr::from(i as u64).to_bigint()).to_bigint()).to_bigint())
+            .map(|i| root_of_unity.pow(two.pow(Fr::from(i as u64).to_bigint()).to_bigint()))
             .collect::<Vec<_>>();
+        assert_eq!(powers[0], Fr::two_adic_root_of_unity());
 
-        // Ensure the correct number of powers of G are present.
-        assert_eq!(FrParameters::POWERS_OF_G.len() as u64, (FrParameters::TWO_ADICITY - 1) as u64);
-        assert_eq!(FrParameters::POWERS_OF_G.len(), powers.len());
+        // Ensure the correct number of powers of root of unity are present.
+        assert_eq!(FrParameters::POWERS_OF_ROOTS_OF_UNITY.len() as u64, (FrParameters::TWO_ADICITY - 1) as u64);
+        assert_eq!(FrParameters::POWERS_OF_ROOTS_OF_UNITY.len(), powers.len());
 
         // Ensure the expected and candidate powers match.
-        for (expected, candidate) in powers.iter().zip(FrParameters::POWERS_OF_G.iter()) {
-            println!("{:?} =?= {:?}", expected, candidate);
-            assert_eq!(expected, candidate);
+        for (expected, candidate) in powers.iter().zip(FrParameters::POWERS_OF_ROOTS_OF_UNITY) {
+            // println!("BigInteger({:?}),", expected.0.0);
+            println!("{:?} =?= {:?}", expected.0, candidate);
+            assert_eq!(&expected.0, candidate);
         }
     }
+
+    #[test]
+    fn test_two_adic_root_of_unity() {
+        let expected = Fr::multiplicative_generator().pow(FrParameters::T);
+        assert_eq!(expected, Fr::two_adic_root_of_unity());
+    }
 }
diff --git a/curves/src/edwards_bls12/fr.rs b/curves/src/edwards_bls12/fr.rs
index 26bab57c8..cb0795f14 100644
--- a/curves/src/edwards_bls12/fr.rs
+++ b/curves/src/edwards_bls12/fr.rs
@@ -33,7 +33,7 @@ impl Fp256Parameters for FrParameters {}
 impl FftParameters for FrParameters {
     type BigInteger = BigInteger;
 
-    const POWERS_OF_G: &'static [BigInteger] = &[];
+    const POWERS_OF_ROOTS_OF_UNITY: &'static [BigInteger] = unimplemented!();
     const TWO_ADICITY: u32 = 1;
     #[rustfmt::skip]
     const TWO_ADIC_ROOT_OF_UNITY: BigInteger = BigInteger([
@@ -117,3 +117,15 @@ impl PoseidonDefaultParameters for FrParameters {
         PoseidonDefaultParametersEntry::new(8, 3, 8, 84, 0),
     ];
 }
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use snarkvm_fields::{FftField, Field};
+
+    #[test]
+    fn test_two_adic_root_of_unity() {
+        let expected = Fr::multiplicative_generator().pow(FrParameters::T);
+        assert_eq!(expected, Fr::two_adic_root_of_unity());
+    }
+}
diff --git a/fields/src/fp_256.rs b/fields/src/fp_256.rs
index 12a5d84f5..6e7edf54e 100644
--- a/fields/src/fp_256.rs
+++ b/fields/src/fp_256.rs
@@ -544,7 +544,6 @@ impl<P: Fp256Parameters> SquareRootField for Fp256<P> {
         }
     }
 
-    // Only works for p = 1 (mod 16).
     #[inline]
     fn sqrt(&self) -> Option<Self> {
         sqrt_impl!(Self, P, self)
diff --git a/fields/src/macros.rs b/fields/src/macros.rs
index 159c861b9..1539c5636 100644
--- a/fields/src/macros.rs
+++ b/fields/src/macros.rs
@@ -136,8 +136,7 @@ macro_rules! sqrt_impl {
                         let n_minus_one_minus_i = n - 1 - i;
                         s += 2u64.pow(n_minus_one_minus_i as u32);
                         if i > 0 {
-                            delta *= $Self::from_bigint($P::POWERS_OF_G[n_minus_one_minus_i as usize])
-                                .expect("precomputed powers of g should always convert properly");
+                            delta *= $Self($P::POWERS_OF_ROOTS_OF_UNITY[n_minus_one_minus_i as usize], PhantomData);
                         } else {
                             delta = -delta;
                         }
@@ -159,8 +158,7 @@ macro_rules! sqrt_impl {
                             }
                             q_bits.iter().enumerate().take(*l as usize).for_each(|(k, bit)| {
                                 if *bit {
-                                    gamma *= $Self::from_bigint($P::POWERS_OF_G[(kappa as usize) + k])
-                                        .expect("precomputed powers of g should always convert properly");
+                                    gamma *= $Self($P::POWERS_OF_ROOTS_OF_UNITY[(kappa as usize) + k], PhantomData);
                                 }
                             });
                         });
diff --git a/fields/src/traits/fft_parameters.rs b/fields/src/traits/fft_parameters.rs
index adcad409f..66bf204b6 100644
--- a/fields/src/traits/fft_parameters.rs
+++ b/fields/src/traits/fft_parameters.rs
@@ -41,6 +41,6 @@ pub trait FftParameters: 'static + Send + Sync + Sized {
     /// SMALL_SUBGROUP_BASE^SMALL_SUBGROUP_BASE_ADICITY)) Used for mixed-radix FFT.
     const LARGE_SUBGROUP_ROOT_OF_UNITY: Option<Self::BigInteger> = None;
 
-    /// `G^2^i` for `i := 0..TWO_ADICITY-1`
-    const POWERS_OF_G: &'static [Self::BigInteger];
+    /// `TWO_ADIC_ROOT_OF_UNITY^2^i` for `i := 0..TWO_ADICITY-1`
+    const POWERS_OF_ROOTS_OF_UNITY: &'static [Self::BigInteger];
 }
diff --git a/parameters/src/testnet3/genesis.rs b/parameters/src/testnet3/genesis.rs
index 75e716b8b..353018d25 100644
--- a/parameters/src/testnet3/genesis.rs
+++ b/parameters/src/testnet3/genesis.rs
@@ -29,6 +29,6 @@ mod tests {
     #[test]
     fn test_genesis_block() {
         let bytes = GenesisBytes::load_bytes();
-        assert_eq!(1740, bytes.len() as u64, "Update me if serialization has changed");
+        assert_eq!(1748, bytes.len() as u64, "Update me if serialization has changed");
     }
 }
diff --git a/parameters/src/testnet3/resources/block.genesis b/parameters/src/testnet3/resources/block.genesis
index c6267b2dd..8835cecb8 100644
Binary files a/parameters/src/testnet3/resources/block.genesis and b/parameters/src/testnet3/resources/block.genesis differ
diff --git a/synthesizer/src/block/bytes.rs b/synthesizer/src/block/bytes.rs
index 4d50cd567..1f1af658d 100644
--- a/synthesizer/src/block/bytes.rs
+++ b/synthesizer/src/block/bytes.rs
@@ -33,20 +33,20 @@ impl<N: Network> FromBytes for Block<N> {
         let header = FromBytes::read_le(&mut reader)?;
         let transactions = FromBytes::read_le(&mut reader)?;
 
-        // Write the coinbase proof.
+        // Write the coinbase.
         let coinbase_variant = u8::read_le(&mut reader)?;
-        let coinbase_proof = match coinbase_variant {
+        let coinbase = match coinbase_variant {
             0 => None,
             1 => Some(FromBytes::read_le(&mut reader)?),
-            _ => return Err(error("Invalid coinbase proof variant")),
+            _ => return Err(error("Invalid coinbase variant")),
         };
 
         // Write the signature.
         let signature = FromBytes::read_le(&mut reader)?;
 
         // Construct the block.
-        let block = Self::from(previous_hash, header, transactions, coinbase_proof, signature)
-            .map_err(|e| error(e.to_string()))?;
+        let block =
+            Self::from(previous_hash, header, transactions, coinbase, signature).map_err(|e| error(e.to_string()))?;
 
         // Ensure the block hash matches.
         match block_hash == block.hash() {
@@ -69,12 +69,12 @@ impl<N: Network> ToBytes for Block<N> {
         self.header.write_le(&mut writer)?;
         self.transactions.write_le(&mut writer)?;
 
-        // Write the coinbase proof.
-        match self.coinbase_proof {
+        // Write the coinbase solution.
+        match self.coinbase {
             None => 0u8.write_le(&mut writer)?,
-            Some(ref coinbase_proof) => {
+            Some(ref coinbase) => {
                 1u8.write_le(&mut writer)?;
-                coinbase_proof.write_le(&mut writer)?;
+                coinbase.write_le(&mut writer)?;
             }
         }
 
diff --git a/synthesizer/src/block/genesis.rs b/synthesizer/src/block/genesis.rs
index a2b3e5c4e..0e93ac422 100644
--- a/synthesizer/src/block/genesis.rs
+++ b/synthesizer/src/block/genesis.rs
@@ -44,11 +44,11 @@ impl<N: Network> Block<N> {
         // Prepare the previous block hash.
         let previous_hash = N::BlockHash::default();
 
-        // Prepare the coinbase proof.
-        let coinbase_proof = None; // The genesis block does not require a coinbase proof.
+        // Prepare the coinbase solution.
+        let coinbase_solution = None; // The genesis block does not require a coinbase solution.
 
         // Construct the block.
-        let block = Self::new(private_key, previous_hash, header, transactions, coinbase_proof, rng)?;
+        let block = Self::new(private_key, previous_hash, header, transactions, coinbase_solution, rng)?;
         // Ensure the block is valid genesis block.
         match block.is_genesis() {
             true => Ok(block),
@@ -64,8 +64,8 @@ impl<N: Network> Block<N> {
             && self.header.is_genesis()
             // Ensure there is 1 transaction in the genesis block.
             && self.transactions.len() == 1
-            // Ensure the coinbase proof does not exist.
-            && self.coinbase_proof.is_none()
+            // Ensure the coinbase solution does not exist.
+            && self.coinbase.is_none()
     }
 }
 
diff --git a/synthesizer/src/block/header/genesis.rs b/synthesizer/src/block/header/genesis.rs
index 69b5ba853..6621b8aa6 100644
--- a/synthesizer/src/block/header/genesis.rs
+++ b/synthesizer/src/block/header/genesis.rs
@@ -55,7 +55,7 @@ mod tests {
         // Previous state root, transactions root, and accumulator point size.
         (Field::<N>::size_in_bytes() * 3)
             // Metadata size.
-            + 2 + 4 + 8 + 8 + 8 + 8
+            + 2 + 4 + 8 + 8 + 8 + 8 + 8
             // Add an additional 4 bytes for versioning.
             + 2 + 2
     }
diff --git a/synthesizer/src/block/header/merkle.rs b/synthesizer/src/block/header/merkle.rs
index 485e35df2..10142040e 100644
--- a/synthesizer/src/block/header/merkle.rs
+++ b/synthesizer/src/block/header/merkle.rs
@@ -120,6 +120,7 @@ mod tests {
                     u64::rand(rng),
                     u64::rand(rng),
                     rng.gen_range(0..i64::MAX),
+                    rng.gen_range(0..i64::MAX),
                 )?,
             )?;
 
diff --git a/synthesizer/src/block/header/metadata/bytes.rs b/synthesizer/src/block/header/metadata/bytes.rs
index acbe87989..df679f8e1 100644
--- a/synthesizer/src/block/header/metadata/bytes.rs
+++ b/synthesizer/src/block/header/metadata/bytes.rs
@@ -33,10 +33,12 @@ impl<N: Network> FromBytes for Metadata<N> {
         let height = u32::read_le(&mut reader)?;
         let coinbase_target = u64::read_le(&mut reader)?;
         let proof_target = u64::read_le(&mut reader)?;
+        let last_coinbase_timestamp = i64::read_le(&mut reader)?;
         let timestamp = i64::read_le(&mut reader)?;
 
         // Construct the metadata.
-        Self::new(network, round, height, coinbase_target, proof_target, timestamp).map_err(|e| error(e.to_string()))
+        Self::new(network, round, height, coinbase_target, proof_target, last_coinbase_timestamp, timestamp)
+            .map_err(|e| error(e.to_string()))
     }
 }
 
@@ -53,6 +55,7 @@ impl<N: Network> ToBytes for Metadata<N> {
         self.height.write_le(&mut writer)?;
         self.coinbase_target.write_le(&mut writer)?;
         self.proof_target.write_le(&mut writer)?;
+        self.last_coinbase_timestamp.write_le(&mut writer)?;
         self.timestamp.write_le(&mut writer)
     }
 }
diff --git a/synthesizer/src/block/header/metadata/genesis.rs b/synthesizer/src/block/header/metadata/genesis.rs
index b9a407f8e..9af2676f5 100644
--- a/synthesizer/src/block/header/metadata/genesis.rs
+++ b/synthesizer/src/block/header/metadata/genesis.rs
@@ -25,10 +25,11 @@ impl<N: Network> Metadata<N> {
         let height = 0;
         let coinbase_target = N::GENESIS_COINBASE_TARGET;
         let proof_target = N::GENESIS_PROOF_TARGET;
+        let last_coinbase_timestamp = N::GENESIS_TIMESTAMP;
         let timestamp = N::GENESIS_TIMESTAMP;
 
         // Return the genesis metadata.
-        Self::new(network, round, height, coinbase_target, proof_target, timestamp)
+        Self::new(network, round, height, coinbase_target, proof_target, last_coinbase_timestamp, timestamp)
     }
 
     /// Returns `true` if the metadata is a genesis metadata.
@@ -43,6 +44,8 @@ impl<N: Network> Metadata<N> {
             && self.coinbase_target == N::GENESIS_COINBASE_TARGET
             // Ensure the proof target in the genesis block is `GENESIS_PROOF_TARGET`.
             && self.proof_target == N::GENESIS_PROOF_TARGET
+            // Ensure the last coinbase timestamp in the genesis block is `GENESIS_TIMESTAMP`.
+            && self.last_coinbase_timestamp == N::GENESIS_TIMESTAMP
             // Ensure the timestamp in the genesis block is `GENESIS_TIMESTAMP`.
             && self.timestamp == N::GENESIS_TIMESTAMP
     }
@@ -59,7 +62,7 @@ mod tests {
     /// Update this method if the contents of the metadata have changed.
     fn get_expected_size<N: Network>() -> usize {
         // Metadata size.
-        2 + 4 + 8 + 8 + 8 + 8
+        2 + 4 + 8 + 8 + 8 + 8 + 8
             // Add an additional 2 bytes for versioning.
             + 2
     }
@@ -91,6 +94,7 @@ mod tests {
         assert_eq!(metadata.round(), 0);
         assert_eq!(metadata.coinbase_target(), CurrentNetwork::GENESIS_COINBASE_TARGET);
         assert_eq!(metadata.proof_target(), CurrentNetwork::GENESIS_PROOF_TARGET);
+        assert_eq!(metadata.last_coinbase_timestamp(), CurrentNetwork::GENESIS_TIMESTAMP);
         assert_eq!(metadata.timestamp(), CurrentNetwork::GENESIS_TIMESTAMP);
     }
 }
diff --git a/synthesizer/src/block/header/metadata/mod.rs b/synthesizer/src/block/header/metadata/mod.rs
index cfb40e347..7ab083f64 100644
--- a/synthesizer/src/block/header/metadata/mod.rs
+++ b/synthesizer/src/block/header/metadata/mod.rs
@@ -37,6 +37,8 @@ pub struct Metadata<N: Network> {
     coinbase_target: u64,
     /// The proof target for this block - 8 bytes.
     proof_target: u64,
+    /// The Unix timestamp (UTC) for the last coinbase - 8 bytes.
+    last_coinbase_timestamp: i64,
     /// The Unix timestamp (UTC) for this block - 8 bytes.
     timestamp: i64,
     /// PhantomData.
@@ -51,10 +53,20 @@ impl<N: Network> Metadata<N> {
         height: u32,
         coinbase_target: u64,
         proof_target: u64,
+        last_coinbase_timestamp: i64,
         timestamp: i64,
     ) -> Result<Self> {
         // Construct a new metadata.
-        let metadata = Self { network, round, height, coinbase_target, proof_target, timestamp, _phantom: PhantomData };
+        let metadata = Self {
+            network,
+            round,
+            height,
+            coinbase_target,
+            proof_target,
+            last_coinbase_timestamp,
+            timestamp,
+            _phantom: PhantomData,
+        };
         // Ensure the header is valid.
         match metadata.is_valid() {
             true => Ok(metadata),
@@ -73,8 +85,14 @@ impl<N: Network> Metadata<N> {
                     && self.round != 0u64
                     // Ensure the height is nonzero.
                     && self.height != 0u32
-                    // Ensure the timestamp in the block is nonzero.
-                    && self.timestamp != 0i64
+                    // Ensure the coinbase target is at or above the minimum.
+                    && self.coinbase_target >= N::GENESIS_COINBASE_TARGET
+                    // Ensure the proof target is at or above the minimum.
+                    && self.proof_target >= N::GENESIS_PROOF_TARGET
+                    // Ensure the last coinbase timestamp is after the genesis timestamp.
+                    && self.last_coinbase_timestamp > N::GENESIS_TIMESTAMP
+                    // Ensure the timestamp in the block is after the genesis timestamp.
+                    && self.timestamp > N::GENESIS_TIMESTAMP
             }
         }
     }
@@ -106,6 +124,11 @@ impl<N: Network> Metadata<N> {
         self.proof_target
     }
 
+    /// Returns the Unix timestamp (UTC) of the last coinbase.
+    pub const fn last_coinbase_timestamp(&self) -> i64 {
+        self.last_coinbase_timestamp
+    }
+
     /// Returns the Unix timestamp (UTC) for this block.
     pub const fn timestamp(&self) -> i64 {
         self.timestamp
diff --git a/synthesizer/src/block/header/metadata/serialize.rs b/synthesizer/src/block/header/metadata/serialize.rs
index a773eb991..557ef756a 100644
--- a/synthesizer/src/block/header/metadata/serialize.rs
+++ b/synthesizer/src/block/header/metadata/serialize.rs
@@ -27,6 +27,7 @@ impl<N: Network> Serialize for Metadata<N> {
                 metadata.serialize_field("height", &self.height)?;
                 metadata.serialize_field("coinbase_target", &self.coinbase_target)?;
                 metadata.serialize_field("proof_target", &self.proof_target)?;
+                metadata.serialize_field("last_coinbase_timestamp", &self.last_coinbase_timestamp)?;
                 metadata.serialize_field("timestamp", &self.timestamp)?;
                 metadata.end()
             }
@@ -47,6 +48,7 @@ impl<'de, N: Network> Deserialize<'de> for Metadata<N> {
                     serde_json::from_value(metadata["height"].take()).map_err(de::Error::custom)?,
                     serde_json::from_value(metadata["coinbase_target"].take()).map_err(de::Error::custom)?,
                     serde_json::from_value(metadata["proof_target"].take()).map_err(de::Error::custom)?,
+                    serde_json::from_value(metadata["last_coinbase_timestamp"].take()).map_err(de::Error::custom)?,
                     serde_json::from_value(metadata["timestamp"].take()).map_err(de::Error::custom)?,
                 )
                 .map_err(de::Error::custom)?)
diff --git a/synthesizer/src/block/header/metadata/to_bits.rs b/synthesizer/src/block/header/metadata/to_bits.rs
index bbac84ad5..f23fc1685 100644
--- a/synthesizer/src/block/header/metadata/to_bits.rs
+++ b/synthesizer/src/block/header/metadata/to_bits.rs
@@ -20,12 +20,13 @@ impl<N: Network> ToBits for Metadata<N> {
     /// Returns the little-endian bits of the metadata.
     fn to_bits_le(&self) -> Vec<bool> {
         vec![
-            self.network.to_bits_le(),         // 2 bytes
-            self.round.to_bits_le(),           // 8 bytes
-            self.height.to_bits_le(),          // 4 bytes
-            self.coinbase_target.to_bits_le(), // 8 bytes
-            self.proof_target.to_bits_le(),    // 8 bytes
-            self.timestamp.to_bits_le(),       // 8 bytes
+            self.network.to_bits_le(),                 // 2 bytes
+            self.round.to_bits_le(),                   // 8 bytes
+            self.height.to_bits_le(),                  // 4 bytes
+            self.coinbase_target.to_bits_le(),         // 8 bytes
+            self.proof_target.to_bits_le(),            // 8 bytes
+            self.last_coinbase_timestamp.to_bits_le(), // 8 bytes
+            self.timestamp.to_bits_le(),               // 8 bytes
         ]
         .concat()
     }
@@ -33,12 +34,13 @@ impl<N: Network> ToBits for Metadata<N> {
     /// Returns the big-endian bits of the metadata.
     fn to_bits_be(&self) -> Vec<bool> {
         vec![
-            self.network.to_bits_be(),         // 2 bytes
-            self.round.to_bits_be(),           // 8 bytes
-            self.height.to_bits_be(),          // 4 bytes
-            self.coinbase_target.to_bits_be(), // 8 bytes
-            self.proof_target.to_bits_be(),    // 8 bytes
-            self.timestamp.to_bits_be(),       // 8 bytes
+            self.network.to_bits_be(),                 // 2 bytes
+            self.round.to_bits_be(),                   // 8 bytes
+            self.height.to_bits_be(),                  // 4 bytes
+            self.coinbase_target.to_bits_be(),         // 8 bytes
+            self.proof_target.to_bits_be(),            // 8 bytes
+            self.last_coinbase_timestamp.to_bits_be(), // 8 bytes
+            self.timestamp.to_bits_be(),               // 8 bytes
         ]
         .concat()
     }
diff --git a/synthesizer/src/block/header/metadata/to_hash.rs b/synthesizer/src/block/header/metadata/to_hash.rs
index 866801f1c..3acfb130a 100644
--- a/synthesizer/src/block/header/metadata/to_hash.rs
+++ b/synthesizer/src/block/header/metadata/to_hash.rs
@@ -20,9 +20,9 @@ impl<N: Network> Metadata<N> {
     /// Returns the metadata hash.
     pub fn to_hash(&self) -> Result<Field<N>> {
         // Construct the metadata bits (the last leaf in the Merkle tree).
-        let metadata_bits = self.to_bits_le(); // 304 bits
+        let metadata_bits = self.to_bits_le(); // 368 bits
         // Ensure the metadata bits is the correct size.
-        ensure!(metadata_bits.len() == 304, "Incorrect metadata size");
+        ensure!(metadata_bits.len() == 368, "Incorrect metadata size");
         // Hash the metadata bits.
         let metadata_hash = N::hash_bhp512(&metadata_bits)?;
         // Return the metadata hash.
diff --git a/synthesizer/src/block/header/mod.rs b/synthesizer/src/block/header/mod.rs
index f839c44ca..bda8deec2 100644
--- a/synthesizer/src/block/header/mod.rs
+++ b/synthesizer/src/block/header/mod.rs
@@ -125,6 +125,11 @@ impl<N: Network> Header<N> {
         self.metadata.proof_target()
     }
 
+    /// Returns the Unix timestamp (UTC) of the last coinbase.
+    pub const fn last_coinbase_timestamp(&self) -> i64 {
+        self.metadata.last_coinbase_timestamp()
+    }
+
     /// Returns the Unix timestamp (UTC) for this block.
     pub const fn timestamp(&self) -> i64 {
         self.metadata.timestamp()
diff --git a/synthesizer/src/block/mod.rs b/synthesizer/src/block/mod.rs
index 3220321a5..c9cd13af5 100644
--- a/synthesizer/src/block/mod.rs
+++ b/synthesizer/src/block/mod.rs
@@ -62,20 +62,20 @@ pub struct Block<N: Network> {
     header: Header<N>,
     /// The transactions in this block.
     transactions: Transactions<N>,
-    /// The coinbase proof.
-    coinbase_proof: Option<CoinbaseSolution<N>>,
+    /// The coinbase solution.
+    coinbase: Option<CoinbaseSolution<N>>,
     /// The signature for this block.
     signature: Signature<N>,
 }
 
 impl<N: Network> Block<N> {
-    /// Initializes a new block from a given previous hash, header, and transactions list.
+    /// Initializes a new block from a given previous hash, header, transactions, coinbase, and signature.
     pub fn new<R: Rng + CryptoRng>(
         private_key: &PrivateKey<N>,
         previous_hash: N::BlockHash,
         header: Header<N>,
         transactions: Transactions<N>,
-        coinbase_proof: Option<CoinbaseSolution<N>>,
+        coinbase: Option<CoinbaseSolution<N>>,
         rng: &mut R,
     ) -> Result<Self> {
         // Ensure the block is not empty.
@@ -85,15 +85,15 @@ impl<N: Network> Block<N> {
         // Sign the block hash.
         let signature = private_key.sign(&[block_hash], rng)?;
         // Construct the block.
-        Self::from(previous_hash, header, transactions, coinbase_proof, signature)
+        Self::from(previous_hash, header, transactions, coinbase, signature)
     }
 
-    /// Initializes a new block from a given previous hash, header, and transactions list.
+    /// Initializes a new block from a given previous hash, header, transactions, coinbase, and signature.
     pub fn from(
         previous_hash: N::BlockHash,
         header: Header<N>,
         transactions: Transactions<N>,
-        coinbase_proof: Option<CoinbaseSolution<N>>,
+        coinbase: Option<CoinbaseSolution<N>>,
         signature: Signature<N>,
     ) -> Result<Self> {
         // Ensure the block is not empty.
@@ -105,18 +105,18 @@ impl<N: Network> Block<N> {
         // Ensure the signature is valid.
         ensure!(signature.verify(&address, &[block_hash]), "Invalid signature for block {}", header.height());
 
-        // Ensure that coinbase accumulator matches the coinbase proof.
-        let expected_accumulator_point = match &coinbase_proof {
-            Some(coinbase_proof) => coinbase_proof.to_accumulator_point()?,
+        // Ensure that coinbase accumulator matches the coinbase solution.
+        let expected_accumulator_point = match &coinbase {
+            Some(coinbase_solution) => coinbase_solution.to_accumulator_point()?,
             None => Field::<N>::zero(),
         };
         ensure!(
             header.coinbase_accumulator_point() == expected_accumulator_point,
-            "The coinbase accumulator point in the block header does not correspond to the given coinbase proof"
+            "The coinbase accumulator point in the block header does not correspond to the given coinbase solution"
         );
 
         // Construct the block.
-        Ok(Self { block_hash: block_hash.into(), previous_hash, header, transactions, coinbase_proof, signature })
+        Ok(Self { block_hash: block_hash.into(), previous_hash, header, transactions, coinbase, signature })
     }
 }
 
@@ -131,9 +131,9 @@ impl<N: Network> Block<N> {
         self.previous_hash
     }
 
-    /// Returns the coinbase proof.
-    pub const fn coinbase_proof(&self) -> Option<&CoinbaseSolution<N>> {
-        self.coinbase_proof.as_ref()
+    /// Returns the coinbase solution.
+    pub const fn coinbase(&self) -> Option<&CoinbaseSolution<N>> {
+        self.coinbase.as_ref()
     }
 
     /// Returns the signature.
diff --git a/synthesizer/src/block/serialize.rs b/synthesizer/src/block/serialize.rs
index 9bfd6ae19..feeb18db3 100644
--- a/synthesizer/src/block/serialize.rs
+++ b/synthesizer/src/block/serialize.rs
@@ -27,8 +27,8 @@ impl<N: Network> Serialize for Block<N> {
                 block.serialize_field("header", &self.header)?;
                 block.serialize_field("transactions", &self.transactions)?;
 
-                if let Some(coinbase_proof) = self.coinbase_proof() {
-                    block.serialize_field("coinbase_proof", coinbase_proof)?;
+                if let Some(coinbase) = &self.coinbase {
+                    block.serialize_field("coinbase", coinbase)?;
                 }
 
                 block.serialize_field("signature", &self.signature)?;
@@ -53,8 +53,8 @@ impl<'de, N: Network> Deserialize<'de> for Block<N> {
                     serde_json::from_value(block["previous_hash"].take()).map_err(de::Error::custom)?,
                     serde_json::from_value(block["header"].take()).map_err(de::Error::custom)?,
                     serde_json::from_value(block["transactions"].take()).map_err(de::Error::custom)?,
-                    match block["coinbase_proof"].as_str() {
-                        Some(coinbase_proof) => Some(serde_json::from_str(coinbase_proof).map_err(de::Error::custom)?),
+                    match block["coinbase"].as_str() {
+                        Some(coinbase) => Some(serde_json::from_str(coinbase).map_err(de::Error::custom)?),
                         None => None,
                     },
                     serde_json::from_value(block["signature"].take()).map_err(de::Error::custom)?,
diff --git a/synthesizer/src/coinbase_puzzle/hash.rs b/synthesizer/src/coinbase_puzzle/hash.rs
index b963f0dee..b6b93807a 100644
--- a/synthesizer/src/coinbase_puzzle/hash.rs
+++ b/synthesizer/src/coinbase_puzzle/hash.rs
@@ -46,7 +46,7 @@ pub fn hash_to_polynomial<F: PrimeField>(input: &[u8], degree: u32) -> DensePoly
     DensePolynomial::from_coefficients_vec(coefficients)
 }
 
-pub fn hash_commitment<E: PairingEngine>(commitment: KZGCommitment<E>) -> Result<E::Fr> {
+pub fn hash_commitment<E: PairingEngine>(commitment: &KZGCommitment<E>) -> Result<E::Fr> {
     // Convert the commitment into bytes.
     let mut bytes = Vec::with_capacity(96);
     commitment.serialize_uncompressed(&mut bytes)?;
diff --git a/synthesizer/src/coinbase_puzzle/helpers/coinbase_solution/bytes.rs b/synthesizer/src/coinbase_puzzle/helpers/coinbase_solution/bytes.rs
index 1b8499572..6cfc453ae 100644
--- a/synthesizer/src/coinbase_puzzle/helpers/coinbase_solution/bytes.rs
+++ b/synthesizer/src/coinbase_puzzle/helpers/coinbase_solution/bytes.rs
@@ -29,7 +29,7 @@ impl<N: Network> FromBytes for CoinbaseSolution<N> {
 
         let proof = KZGProof::read_le(&mut reader)?;
 
-        Ok(Self { partial_solutions, proof })
+        Ok(Self::new(partial_solutions, proof))
     }
 }
 
diff --git a/synthesizer/src/coinbase_puzzle/helpers/coinbase_solution/mod.rs b/synthesizer/src/coinbase_puzzle/helpers/coinbase_solution/mod.rs
index f41b52652..f7f451bb8 100644
--- a/synthesizer/src/coinbase_puzzle/helpers/coinbase_solution/mod.rs
+++ b/synthesizer/src/coinbase_puzzle/helpers/coinbase_solution/mod.rs
@@ -56,7 +56,7 @@ impl<N: Network> CoinbaseSolution<N> {
     }
 
     /// Returns the cumulative sum of the prover solutions.
-    pub fn to_cumulative_target(&self) -> Result<u128> {
+    pub fn to_cumulative_proof_target(&self) -> Result<u128> {
         // Compute the cumulative target as a u128.
         self.partial_solutions.iter().try_fold(0u128, |cumulative, solution| {
             cumulative.checked_add(solution.to_target()? as u128).ok_or_else(|| anyhow!("Cumulative target overflowed"))
@@ -66,7 +66,7 @@ impl<N: Network> CoinbaseSolution<N> {
     /// Returns the accumulator challenge point.
     pub fn to_accumulator_point(&self) -> Result<Field<N>> {
         let mut challenge_points =
-            hash_commitments(self.partial_solutions.iter().map(|solution| solution.commitment()))?;
+            hash_commitments(self.partial_solutions.iter().map(|solution| *solution.commitment()))?;
         ensure!(challenge_points.len() == self.partial_solutions.len() + 1, "Invalid number of challenge points");
 
         // Pop the last challenge point as the accumulator challenge point.
diff --git a/synthesizer/src/coinbase_puzzle/helpers/mod.rs b/synthesizer/src/coinbase_puzzle/helpers/mod.rs
index 383eb50f1..4c3218b51 100644
--- a/synthesizer/src/coinbase_puzzle/helpers/mod.rs
+++ b/synthesizer/src/coinbase_puzzle/helpers/mod.rs
@@ -26,6 +26,9 @@ pub use partial_solution::*;
 mod prover_solution;
 pub use prover_solution::*;
 
+mod puzzle_commitment;
+pub use puzzle_commitment::*;
+
 use crate::coinbase_puzzle::{hash_commitment, hash_commitments, CoinbasePuzzle};
 use console::{account::Address, prelude::*, types::Field};
 use snarkvm_algorithms::{
@@ -41,8 +44,6 @@ use std::{
     io::{Read, Result as IoResult, Write},
 };
 
-/// The commitment to the polynomial.
-pub type PuzzleCommitment<N> = KZGCommitment<<N as Environment>::PairingCurve>;
 /// The proof of opening the polynomial, for the solution.
 pub type PuzzleProof<N> = KZGProof<<N as Environment>::PairingCurve>;
 
diff --git a/synthesizer/src/coinbase_puzzle/helpers/partial_solution/bytes.rs b/synthesizer/src/coinbase_puzzle/helpers/partial_solution/bytes.rs
index f84f5b343..1df4b0127 100644
--- a/synthesizer/src/coinbase_puzzle/helpers/partial_solution/bytes.rs
+++ b/synthesizer/src/coinbase_puzzle/helpers/partial_solution/bytes.rs
@@ -23,7 +23,7 @@ impl<N: Network> FromBytes for PartialSolution<N> {
         let nonce = u64::read_le(&mut reader)?;
         let commitment = KZGCommitment::read_le(&mut reader)?;
 
-        Ok(Self { address, nonce, commitment })
+        Ok(Self::new(address, nonce, commitment))
     }
 }
 
diff --git a/synthesizer/src/coinbase_puzzle/helpers/partial_solution/mod.rs b/synthesizer/src/coinbase_puzzle/helpers/partial_solution/mod.rs
index 393aa3a26..effcad908 100644
--- a/synthesizer/src/coinbase_puzzle/helpers/partial_solution/mod.rs
+++ b/synthesizer/src/coinbase_puzzle/helpers/partial_solution/mod.rs
@@ -34,8 +34,8 @@ pub struct PartialSolution<N: Network> {
 
 impl<N: Network> PartialSolution<N> {
     /// Initializes a new instance of the partial solution.
-    pub const fn new(address: Address<N>, nonce: u64, commitment: PuzzleCommitment<N>) -> Self {
-        Self { address, nonce, commitment }
+    pub fn new<C: Into<PuzzleCommitment<N>>>(address: Address<N>, nonce: u64, commitment: C) -> Self {
+        Self { address, nonce, commitment: commitment.into() }
     }
 
     /// Returns the address of the prover.
diff --git a/synthesizer/src/coinbase_puzzle/helpers/partial_solution/serialize.rs b/synthesizer/src/coinbase_puzzle/helpers/partial_solution/serialize.rs
index f10eb24e5..39e372d34 100644
--- a/synthesizer/src/coinbase_puzzle/helpers/partial_solution/serialize.rs
+++ b/synthesizer/src/coinbase_puzzle/helpers/partial_solution/serialize.rs
@@ -24,7 +24,7 @@ impl<N: Network> Serialize for PartialSolution<N> {
                 let mut partial_prover_solution = serializer.serialize_struct("PartialSolution", 3)?;
                 partial_prover_solution.serialize_field("address", &self.address)?;
                 partial_prover_solution.serialize_field("nonce", &self.nonce)?;
-                partial_prover_solution.serialize_field("commitment", &self.commitment.0)?;
+                partial_prover_solution.serialize_field("commitment", &self.commitment)?;
                 partial_prover_solution.end()
             }
             false => ToBytesSerializer::serialize_with_size_encoding(self, serializer),
@@ -41,10 +41,8 @@ impl<'de, N: Network> Deserialize<'de> for PartialSolution<N> {
                 Ok(Self::new(
                     serde_json::from_value(partial_prover_solution["address"].clone()).map_err(de::Error::custom)?,
                     serde_json::from_value(partial_prover_solution["nonce"].clone()).map_err(de::Error::custom)?,
-                    KZGCommitment(
-                        serde_json::from_value(partial_prover_solution["commitment"].clone())
-                            .map_err(de::Error::custom)?,
-                    ),
+                    serde_json::from_value::<PuzzleCommitment<N>>(partial_prover_solution["commitment"].clone())
+                        .map_err(de::Error::custom)?,
                 ))
             }
             false => FromBytesDeserializer::<Self>::deserialize_with_size_encoding(deserializer, "partial solution"),
diff --git a/synthesizer/src/coinbase_puzzle/helpers/prover_solution/bytes.rs b/synthesizer/src/coinbase_puzzle/helpers/prover_solution/bytes.rs
index ff3c57a06..b4734e6c3 100644
--- a/synthesizer/src/coinbase_puzzle/helpers/prover_solution/bytes.rs
+++ b/synthesizer/src/coinbase_puzzle/helpers/prover_solution/bytes.rs
@@ -22,7 +22,7 @@ impl<N: Network> FromBytes for ProverSolution<N> {
         let partial_solution: PartialSolution<N> = FromBytes::read_le(&mut reader)?;
         let proof = KZGProof::read_le(&mut reader)?;
 
-        Ok(Self { partial_solution, proof })
+        Ok(Self::new(partial_solution, proof))
     }
 }
 
diff --git a/synthesizer/src/coinbase_puzzle/helpers/prover_solution/mod.rs b/synthesizer/src/coinbase_puzzle/helpers/prover_solution/mod.rs
index d123588e9..9b0c78f5d 100644
--- a/synthesizer/src/coinbase_puzzle/helpers/prover_solution/mod.rs
+++ b/synthesizer/src/coinbase_puzzle/helpers/prover_solution/mod.rs
@@ -56,7 +56,7 @@ impl<N: Network> ProverSolution<N> {
         let prover_polynomial = self.partial_solution.to_prover_polynomial(epoch_challenge)?;
 
         // Compute the challenge point.
-        let challenge_point = hash_commitment(self.commitment())?;
+        let challenge_point = hash_commitment(&self.commitment())?;
 
         // Evaluate the epoch and prover polynomials at the challenge point.
         let epoch_evaluation = epoch_challenge.epoch_polynomial().evaluate(challenge_point);
diff --git a/synthesizer/src/coinbase_puzzle/helpers/puzzle_commitment/bytes.rs b/synthesizer/src/coinbase_puzzle/helpers/puzzle_commitment/bytes.rs
new file mode 100644
index 000000000..998a0575c
--- /dev/null
+++ b/synthesizer/src/coinbase_puzzle/helpers/puzzle_commitment/bytes.rs
@@ -0,0 +1,56 @@
+// Copyright (C) 2019-2022 Aleo Systems Inc.
+// This file is part of the snarkVM library.
+
+// The snarkVM library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// The snarkVM library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with the snarkVM library. If not, see <https://www.gnu.org/licenses/>.
+
+use super::*;
+
+impl<N: Network> FromBytes for PuzzleCommitment<N> {
+    /// Reads the puzzle commitment from the buffer.
+    fn read_le<R: Read>(mut reader: R) -> IoResult<Self> {
+        let commitment = KZGCommitment::read_le(&mut reader)?;
+
+        Ok(Self::new(commitment))
+    }
+}
+
+impl<N: Network> ToBytes for PuzzleCommitment<N> {
+    /// Writes the puzzle commitment to the buffer.
+    fn write_le<W: Write>(&self, mut writer: W) -> IoResult<()> {
+        self.commitment.write_le(&mut writer)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use console::network::Testnet3;
+
+    type CurrentNetwork = Testnet3;
+
+    #[test]
+    fn test_bytes() -> Result<()> {
+        let mut rng = TestRng::default();
+        // Sample a new puzzle commitment.
+        let expected = PuzzleCommitment::<CurrentNetwork>::new(KZGCommitment(rng.gen()));
+
+        // Check the byte representation.
+        let expected_bytes = expected.to_bytes_le()?;
+        assert_eq!(expected_bytes.len(), 48);
+        assert_eq!(expected, PuzzleCommitment::read_le(&expected_bytes[..])?);
+        assert!(PuzzleCommitment::<CurrentNetwork>::read_le(&expected_bytes[1..]).is_err());
+
+        Ok(())
+    }
+}
diff --git a/synthesizer/src/coinbase_puzzle/helpers/puzzle_commitment/mod.rs b/synthesizer/src/coinbase_puzzle/helpers/puzzle_commitment/mod.rs
new file mode 100644
index 000000000..fc61392bc
--- /dev/null
+++ b/synthesizer/src/coinbase_puzzle/helpers/puzzle_commitment/mod.rs
@@ -0,0 +1,50 @@
+// Copyright (C) 2019-2022 Aleo Systems Inc.
+// This file is part of the snarkVM library.
+
+// The snarkVM library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// The snarkVM library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with the snarkVM library. If not, see <https://www.gnu.org/licenses/>.
+
+mod bytes;
+mod serialize;
+mod string;
+
+use super::*;
+
+/// A coinbase puzzle commitment to a polynomial.
+#[derive(Copy, Clone, Eq, PartialEq, Hash)]
+pub struct PuzzleCommitment<N: Network> {
+    /// The commitment for the solution.
+    commitment: KZGCommitment<<N as Environment>::PairingCurve>,
+}
+
+impl<N: Network> PuzzleCommitment<N> {
+    /// Initializes a new instance of the puzzle commitment.
+    pub const fn new(commitment: KZGCommitment<<N as Environment>::PairingCurve>) -> Self {
+        Self { commitment }
+    }
+}
+
+impl<N: Network> From<KZGCommitment<<N as Environment>::PairingCurve>> for PuzzleCommitment<N> {
+    /// Initializes a new instance of the puzzle commitment.
+    fn from(commitment: KZGCommitment<<N as Environment>::PairingCurve>) -> Self {
+        Self::new(commitment)
+    }
+}
+
+impl<N: Network> Deref for PuzzleCommitment<N> {
+    type Target = KZGCommitment<<N as Environment>::PairingCurve>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.commitment
+    }
+}
diff --git a/synthesizer/src/coinbase_puzzle/helpers/puzzle_commitment/serialize.rs b/synthesizer/src/coinbase_puzzle/helpers/puzzle_commitment/serialize.rs
new file mode 100644
index 000000000..f5114d283
--- /dev/null
+++ b/synthesizer/src/coinbase_puzzle/helpers/puzzle_commitment/serialize.rs
@@ -0,0 +1,83 @@
+// Copyright (C) 2019-2022 Aleo Systems Inc.
+// This file is part of the snarkVM library.
+
+// The snarkVM library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// The snarkVM library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with the snarkVM library. If not, see <https://www.gnu.org/licenses/>.
+
+use super::*;
+
+impl<N: Network> Serialize for PuzzleCommitment<N> {
+    /// Serializes the puzzle commitment to a string or buffer.
+    fn serialize<S: Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
+        match serializer.is_human_readable() {
+            true => serializer.collect_str(self),
+            false => ToBytesSerializer::serialize_with_size_encoding(self, serializer),
+        }
+    }
+}
+
+impl<'de, N: Network> Deserialize<'de> for PuzzleCommitment<N> {
+    /// Deserializes the puzzle commitment from a string or buffer.
+    fn deserialize<D: Deserializer<'de>>(deserializer: D) -> Result<Self, D::Error> {
+        match deserializer.is_human_readable() {
+            true => FromStr::from_str(&String::deserialize(deserializer)?).map_err(de::Error::custom),
+            false => FromBytesDeserializer::<Self>::deserialize_with_size_encoding(deserializer, "puzzle commitment"),
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use console::network::Testnet3;
+
+    type CurrentNetwork = Testnet3;
+
+    #[test]
+    fn test_serde_json() -> Result<()> {
+        let mut rng = TestRng::default();
+
+        // Sample a new puzzle commitment.
+        let expected = PuzzleCommitment::<CurrentNetwork>::new(KZGCommitment(rng.gen()));
+
+        // Serialize
+        let expected_string = &expected.to_string();
+        let candidate_string = serde_json::to_string(&expected)?;
+        assert_eq!(expected, serde_json::from_str(&candidate_string)?);
+
+        // Deserialize
+        assert_eq!(expected, PuzzleCommitment::from_str(expected_string)?);
+        assert_eq!(expected, serde_json::from_str(&candidate_string)?);
+
+        Ok(())
+    }
+
+    #[test]
+    fn test_bincode() -> Result<()> {
+        let mut rng = TestRng::default();
+
+        // Sample a new puzzle commitment.
+        let expected = PuzzleCommitment::<CurrentNetwork>::new(KZGCommitment(rng.gen()));
+
+        // Serialize
+        let expected_bytes = expected.to_bytes_le()?;
+        let expected_bytes_with_size_encoding = bincode::serialize(&expected)?;
+        assert_eq!(&expected_bytes[..], &expected_bytes_with_size_encoding[8..]);
+
+        // Deserialize
+        assert_eq!(expected, PuzzleCommitment::read_le(&expected_bytes[..])?);
+        assert_eq!(expected, bincode::deserialize(&expected_bytes_with_size_encoding[..])?);
+
+        Ok(())
+    }
+}
diff --git a/synthesizer/src/coinbase_puzzle/helpers/puzzle_commitment/string.rs b/synthesizer/src/coinbase_puzzle/helpers/puzzle_commitment/string.rs
new file mode 100644
index 000000000..84d4f6f13
--- /dev/null
+++ b/synthesizer/src/coinbase_puzzle/helpers/puzzle_commitment/string.rs
@@ -0,0 +1,105 @@
+// Copyright (C) 2019-2022 Aleo Systems Inc.
+// This file is part of the snarkVM library.
+
+// The snarkVM library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// The snarkVM library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with the snarkVM library. If not, see <https://www.gnu.org/licenses/>.
+
+use super::*;
+
+static PUZZLE_COMMITMENT_PREFIX: &str = "puzzle";
+
+impl<N: Network> FromStr for PuzzleCommitment<N> {
+    type Err = Error;
+
+    /// Reads in the puzzle commitment string.
+    fn from_str(puzzle_commitment: &str) -> Result<Self, Self::Err> {
+        // Decode the puzzle commitment string from bech32m.
+        let (hrp, data, variant) = bech32::decode(puzzle_commitment)?;
+        if hrp != PUZZLE_COMMITMENT_PREFIX {
+            bail!("Failed to decode puzzle commitment: '{hrp}' is an invalid prefix")
+        } else if data.is_empty() {
+            bail!("Failed to decode puzzle commitment: data field is empty")
+        } else if variant != bech32::Variant::Bech32m {
+            bail!("Found a puzzle commitment that is not bech32m encoded: {puzzle_commitment}");
+        }
+        // Decode the puzzle commitment data from u5 to u8, and into the puzzle commitment.
+        Ok(Self::read_le(&Vec::from_base32(&data)?[..])?)
+    }
+}
+
+impl<N: Network> Debug for PuzzleCommitment<N> {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        Display::fmt(self, f)
+    }
+}
+
+impl<N: Network> Display for PuzzleCommitment<N> {
+    /// Writes the puzzle commitment as a bech32m string.
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        // Convert the puzzle commitment to bytes.
+        let bytes = self.to_bytes_le().map_err(|_| fmt::Error)?;
+        // Encode the bytes into bech32m.
+        let string = bech32::encode(PUZZLE_COMMITMENT_PREFIX, bytes.to_base32(), bech32::Variant::Bech32m)
+            .map_err(|_| fmt::Error)?;
+        // Output the string.
+        Display::fmt(&string, f)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use console::network::Testnet3;
+
+    type CurrentNetwork = Testnet3;
+
+    const ITERATIONS: u64 = 1_000;
+
+    #[test]
+    fn test_string() -> Result<()> {
+        // Ensure type and empty value fails.
+        assert!(PuzzleCommitment::<CurrentNetwork>::from_str(&format!("{PUZZLE_COMMITMENT_PREFIX}1")).is_err());
+        assert!(PuzzleCommitment::<CurrentNetwork>::from_str("").is_err());
+
+        let mut rng = TestRng::default();
+
+        for _ in 0..ITERATIONS {
+            // Sample a new puzzle commitment.
+            let expected = PuzzleCommitment::<CurrentNetwork>::new(KZGCommitment(rng.gen()));
+
+            // Check the string representation.
+            let candidate = format!("{expected}");
+            assert_eq!(expected, PuzzleCommitment::from_str(&candidate)?);
+            assert_eq!(PUZZLE_COMMITMENT_PREFIX, candidate.to_string().split('1').next().unwrap());
+        }
+        Ok(())
+    }
+
+    #[test]
+    fn test_display() -> Result<()> {
+        let mut rng = TestRng::default();
+
+        for _ in 0..ITERATIONS {
+            // Sample a new puzzle commitment.
+            let expected = PuzzleCommitment::<CurrentNetwork>::new(KZGCommitment(rng.gen()));
+
+            let candidate = expected.to_string();
+            assert_eq!(format!("{expected}"), candidate);
+            assert_eq!(PUZZLE_COMMITMENT_PREFIX, candidate.split('1').next().unwrap());
+
+            let candidate_recovered = PuzzleCommitment::<CurrentNetwork>::from_str(&candidate.to_string())?;
+            assert_eq!(expected, candidate_recovered);
+        }
+        Ok(())
+    }
+}
diff --git a/synthesizer/src/coinbase_puzzle/mod.rs b/synthesizer/src/coinbase_puzzle/mod.rs
index 064628242..4baffe980 100644
--- a/synthesizer/src/coinbase_puzzle/mod.rs
+++ b/synthesizer/src/coinbase_puzzle/mod.rs
@@ -132,7 +132,7 @@ impl<N: Network> CoinbasePuzzle<N> {
         };
         let (commitment, _rand) =
             KZG10::commit_lagrange(&pk.lagrange_basis(), &product_evaluations, None, &Default::default(), None)?;
-        let point = hash_commitment(commitment)?;
+        let point = hash_commitment(&commitment)?;
         let product_eval_at_point = polynomial.evaluate(point) * epoch_challenge.epoch_polynomial().evaluate(point);
 
         let proof = KZG10::open_lagrange(
@@ -190,7 +190,7 @@ impl<N: Network> CoinbasePuzzle<N> {
             .unzip();
 
         // Compute the challenge points.
-        let mut challenges = hash_commitments(partial_solutions.iter().map(|solution| solution.commitment()))?;
+        let mut challenges = hash_commitments(partial_solutions.iter().map(|solution| *solution.commitment()))?;
         ensure!(challenges.len() == partial_solutions.len() + 1, "Invalid number of challenge points");
 
         // Pop the last challenge as the accumulator challenge point.
@@ -267,7 +267,7 @@ impl<N: Network> CoinbasePuzzle<N> {
         }
 
         // Ensure the coinbase proof meets the required coinbase target.
-        if coinbase_solution.to_cumulative_target()? < coinbase_target as u128 {
+        if coinbase_solution.to_cumulative_proof_target()? < coinbase_target as u128 {
             bail!("The coinbase proof does not meet the coinbase target");
         }
 
@@ -283,7 +283,7 @@ impl<N: Network> CoinbasePuzzle<N> {
 
         // Compute the challenge points.
         let mut challenge_points =
-            hash_commitments(coinbase_solution.partial_solutions().iter().map(|solution| solution.commitment()))?;
+            hash_commitments(coinbase_solution.partial_solutions().iter().map(|solution| *solution.commitment()))?;
         ensure!(
             challenge_points.len() == coinbase_solution.partial_solutions().len() + 1,
             "Invalid number of challenge points"
diff --git a/synthesizer/src/program/instruction/operation/macros.rs b/synthesizer/src/program/instruction/operation/macros.rs
index 209f8b00c..ac377d323 100644
--- a/synthesizer/src/program/instruction/operation/macros.rs
+++ b/synthesizer/src/program/instruction/operation/macros.rs
@@ -710,14 +710,36 @@ mod tests {
 
                     // Check the operation on randomly-sampled values.
                     for i in 0..num_iterations {
+                        macro_rules! sample_value {
+                            (I8, I8) => { sample_value!(I128, I128) };
+                            (I16, I16) => { sample_value!(I128, I128) };
+                            (I32, I32) => { sample_value!(I128, I128) };
+                            (I64, I64) => { sample_value!(I128, I128) };
+                            (I128, I128) => {
+                                match i {
+                                    0 => ($input_a::zero(), $input_b::zero()),
+                                    1 => ($input_a::<CurrentNetwork>::rand(&mut rng), $input_b::zero()),
+                                    2 => ($input_a::zero(), $input_b::<CurrentNetwork>::rand(&mut rng)),
+                                    3 => ($input_a::MIN, $input_b::zero() - $input_b::one()),
+                                    4.. => ($input_b::<CurrentNetwork>::rand(&mut rng), $input_b::<CurrentNetwork>::rand(&mut rng))
+                                }
+                            };
+                            ($lhs:ident, $rhs:ident) => {
+                                match i {
+                                    0 => ($lhs::zero(), $rhs::zero()),
+                                    1 => ($lhs::<CurrentNetwork>::rand(&mut rng), $rhs::zero()),
+                                    2 => ($lhs::zero(), $rhs::<CurrentNetwork>::rand(&mut rng)),
+                                    3.. => ($lhs::<CurrentNetwork>::rand(&mut rng), $rhs::<CurrentNetwork>::rand(&mut rng))
+                                }
+                            }
+                        }
                         // Sample the first and second value.
                         #[allow(deprecated)]
-                        let (a, b) = match i {
-                            0 => ($input_a::zero(), $input_b::zero()),
-                            1 => ($input_a::<CurrentNetwork>::rand(&mut rng), $input_b::zero()),
-                            2 => ($input_a::zero(), $input_b::<CurrentNetwork>::rand(&mut rng)),
-                            3.. => ($input_a::<CurrentNetwork>::rand(&mut rng), $input_b::<CurrentNetwork>::rand(&mut rng))
-                        };
+                        let (a, b) = sample_value!($input_a, $input_b);
+
+                        // This flag is used to determine halting conditions.
+                        #[allow(deprecated)]
+                        let is_rhs_zero = (*b) == *$input_b::<CurrentNetwork>::zero();
 
                         // Initialize an indicator whether the operation should succeed or not.
                         #[allow(unused_mut)]
@@ -785,8 +807,10 @@ mod tests {
                                 let mut should_panic_on_halt = false;
                                 // If the operation is a shift operator, check if the mode of the RHS is a constant and if the shift amount exceeds the bitwidth.
                                 should_panic_on_halt |= is_shift_operator && shift_exceeds_bitwidth && mode_b.is_constant();
-                                // If the operation is a division operator, check if the mode of the RHS is a constant.
-                                should_panic_on_halt |= is_division_operator && mode_b.is_constant();
+                                // If the operation is a division operator, check if both operands are constant or if the RHS is a constant and zero.
+                                should_panic_on_halt |= is_division_operator && (
+                                    mode_b.is_constant() && (mode_a.is_constant() || is_rhs_zero)
+                                );
 
                                 // If this iteration should succeed, ensure the evaluated and executed outputs match the expected output.
                                 if should_succeed {
diff --git a/synthesizer/src/store/block/mod.rs b/synthesizer/src/store/block/mod.rs
index 1be3830a6..0bcfcb3c5 100644
--- a/synthesizer/src/store/block/mod.rs
+++ b/synthesizer/src/store/block/mod.rs
@@ -17,7 +17,7 @@
 use crate::{
     atomic_write_batch,
     block::{Block, Header, Transactions},
-    coinbase_puzzle::CoinbaseSolution,
+    coinbase_puzzle::{CoinbaseSolution, PuzzleCommitment},
     cow_to_cloned,
     cow_to_copied,
     store::{
@@ -63,8 +63,10 @@ pub trait BlockStorage<N: Network>: 'static + Clone + Send + Sync {
     type TransactionStorage: TransactionStorage<N, TransitionStorage = Self::TransitionStorage>;
     /// The transition storage.
     type TransitionStorage: TransitionStorage<N>;
-    /// The mapping of `block hash` to `block coinbase proof`.
-    type CoinbaseProofMap: for<'a> Map<'a, N::BlockHash, Option<CoinbaseSolution<N>>>;
+    /// The mapping of `block hash` to `block coinbase solution`.
+    type CoinbaseSolutionMap: for<'a> Map<'a, N::BlockHash, Option<CoinbaseSolution<N>>>;
+    /// The mapping of `puzzle commitment` to `block hash`.
+    type CoinbasePuzzleCommitmentMap: for<'a> Map<'a, PuzzleCommitment<N>, N::BlockHash>;
     /// The mapping of `block hash` to `block signature`.
     type SignatureMap: for<'a> Map<'a, N::BlockHash, Signature<N>>;
 
@@ -87,8 +89,10 @@ pub trait BlockStorage<N: Network>: 'static + Clone + Send + Sync {
     fn reverse_transactions_map(&self) -> &Self::ReverseTransactionsMap;
     /// Returns the transaction store.
     fn transaction_store(&self) -> &TransactionStore<N, Self::TransactionStorage>;
-    /// Returns the coinbase proof map.
-    fn coinbase_proof_map(&self) -> &Self::CoinbaseProofMap;
+    /// Returns the coinbase solution map.
+    fn coinbase_solution_map(&self) -> &Self::CoinbaseSolutionMap;
+    /// Returns the coinbase puzzle commitment map.
+    fn coinbase_puzzle_commitment_map(&self) -> &Self::CoinbasePuzzleCommitmentMap;
     /// Returns the signature map.
     fn signature_map(&self) -> &Self::SignatureMap;
 
@@ -112,7 +116,8 @@ pub trait BlockStorage<N: Network>: 'static + Clone + Send + Sync {
         self.transactions_map().start_atomic();
         self.reverse_transactions_map().start_atomic();
         self.transaction_store().start_atomic();
-        self.coinbase_proof_map().start_atomic();
+        self.coinbase_solution_map().start_atomic();
+        self.coinbase_puzzle_commitment_map().start_atomic();
         self.signature_map().start_atomic();
     }
 
@@ -126,7 +131,8 @@ pub trait BlockStorage<N: Network>: 'static + Clone + Send + Sync {
             || self.transactions_map().is_atomic_in_progress()
             || self.reverse_transactions_map().is_atomic_in_progress()
             || self.transaction_store().is_atomic_in_progress()
-            || self.coinbase_proof_map().is_atomic_in_progress()
+            || self.coinbase_solution_map().is_atomic_in_progress()
+            || self.coinbase_puzzle_commitment_map().is_atomic_in_progress()
             || self.signature_map().is_atomic_in_progress()
     }
 
@@ -140,7 +146,8 @@ pub trait BlockStorage<N: Network>: 'static + Clone + Send + Sync {
         self.transactions_map().abort_atomic();
         self.reverse_transactions_map().abort_atomic();
         self.transaction_store().abort_atomic();
-        self.coinbase_proof_map().abort_atomic();
+        self.coinbase_solution_map().abort_atomic();
+        self.coinbase_puzzle_commitment_map().abort_atomic();
         self.signature_map().abort_atomic();
     }
 
@@ -154,7 +161,8 @@ pub trait BlockStorage<N: Network>: 'static + Clone + Send + Sync {
         self.transactions_map().finish_atomic()?;
         self.reverse_transactions_map().finish_atomic()?;
         self.transaction_store().finish_atomic()?;
-        self.coinbase_proof_map().finish_atomic()?;
+        self.coinbase_solution_map().finish_atomic()?;
+        self.coinbase_puzzle_commitment_map().finish_atomic()?;
         self.signature_map().finish_atomic()
     }
 
@@ -184,8 +192,15 @@ pub trait BlockStorage<N: Network>: 'static + Clone + Send + Sync {
                 self.transaction_store().insert(transaction)?;
             }
 
-            // Store the block coinbase proof.
-            self.coinbase_proof_map().insert(block.hash(), block.coinbase_proof().cloned())?;
+            // Store the block coinbase solution.
+            self.coinbase_solution_map().insert(block.hash(), block.coinbase().cloned())?;
+
+            // Store the block coinbase puzzle commitment.
+            if let Some(coinbase) = block.coinbase() {
+                for puzzle_commitment in coinbase.partial_solutions().iter().map(|s| s.commitment()) {
+                    self.coinbase_puzzle_commitment_map().insert(puzzle_commitment, block.hash())?;
+                }
+            }
 
             // Store the block signature.
             self.signature_map().insert(block.hash(), *block.signature())?;
@@ -213,6 +228,13 @@ pub trait BlockStorage<N: Network>: 'static + Clone + Send + Sync {
             Some(transaction_ids) => transaction_ids,
             None => bail!("Failed to remove block: missing transactions for block '{block_height}' ('{block_hash}')"),
         };
+        // Retrieve the coinbase solution.
+        let coinbase = match self.coinbase_solution_map().get(block_hash)? {
+            Some(coinbase_solution) => cow_to_cloned!(coinbase_solution),
+            None => {
+                bail!("Failed to remove block: missing coinbase solution for block '{block_height}' ('{block_hash}')")
+            }
+        };
 
         atomic_write_batch!(self, {
             // Remove the (block height, state root) pair.
@@ -238,8 +260,15 @@ pub trait BlockStorage<N: Network>: 'static + Clone + Send + Sync {
                 self.transaction_store().remove(transaction_id)?;
             }
 
-            // Remove the block coinbase proof.
-            self.coinbase_proof_map().remove(block_hash)?;
+            // Remove the block coinbase solution.
+            self.coinbase_solution_map().remove(block_hash)?;
+
+            // Remove the block coinbase puzzle commitment.
+            if let Some(coinbase) = coinbase {
+                for puzzle_commitment in coinbase.partial_solutions().iter().map(|s| s.commitment()) {
+                    self.coinbase_puzzle_commitment_map().remove(&puzzle_commitment)?;
+                }
+            }
 
             // Remove the block signature.
             self.signature_map().remove(block_hash)?;
@@ -266,6 +295,17 @@ pub trait BlockStorage<N: Network>: 'static + Clone + Send + Sync {
         }
     }
 
+    /// Returns the block hash that contains the given `puzzle commitment`.
+    fn find_block_hash_from_puzzle_commitment(
+        &self,
+        puzzle_commitment: &PuzzleCommitment<N>,
+    ) -> Result<Option<N::BlockHash>> {
+        match self.coinbase_puzzle_commitment_map().get(puzzle_commitment)? {
+            Some(block_hash) => Ok(Some(cow_to_copied!(block_hash))),
+            None => Ok(None),
+        }
+    }
+
     /// Returns the state root that contains the given `block height`.
     fn get_state_root(&self, block_height: u32) -> Result<Option<Field<N>>> {
         match self.state_root_map().get(&block_height)? {
@@ -329,11 +369,11 @@ pub trait BlockStorage<N: Network>: 'static + Clone + Send + Sync {
         Ok(Some(Transactions::from(&transactions)))
     }
 
-    /// Returns the block coinbase proof for the given `block hash`.
-    fn get_block_coinbase_proof(&self, block_hash: &N::BlockHash) -> Result<Option<CoinbaseSolution<N>>> {
-        match self.coinbase_proof_map().get(block_hash)? {
-            Some(coinbase_proof) => Ok(cow_to_cloned!(coinbase_proof)),
-            None => bail!("Missing coinbase proof for block ('{block_hash}')"),
+    /// Returns the block coinbase solution for the given `block hash`.
+    fn get_block_coinbase(&self, block_hash: &N::BlockHash) -> Result<Option<CoinbaseSolution<N>>> {
+        match self.coinbase_solution_map().get(block_hash)? {
+            Some(coinbase_solution) => Ok(cow_to_cloned!(coinbase_solution)),
+            None => bail!("Missing coinbase solution for block ('{block_hash}')"),
         }
     }
 
@@ -373,10 +413,10 @@ pub trait BlockStorage<N: Network>: 'static + Clone + Send + Sync {
             Some(transactions) => transactions,
             None => bail!("Missing transactions for block {height} ('{block_hash}')"),
         };
-        // Retrieve the block coinbase proof.
-        let coinbase_proof = match self.get_block_coinbase_proof(block_hash) {
-            Ok(coinbase_proof) => coinbase_proof,
-            Err(_) => bail!("Missing coinbase proof for block {height} ('{block_hash}')"),
+        // Retrieve the block coinbase solution.
+        let coinbase = match self.get_block_coinbase(block_hash) {
+            Ok(coinbase_solution) => coinbase_solution,
+            Err(_) => bail!("Missing coinbase solution for block {height} ('{block_hash}')"),
         };
         // Retrieve the block signature.
         let signature = match self.get_block_signature(block_hash)? {
@@ -385,7 +425,7 @@ pub trait BlockStorage<N: Network>: 'static + Clone + Send + Sync {
         };
 
         // Return the block.
-        Ok(Some(Block::from(previous_hash, header, transactions, coinbase_proof, signature)?))
+        Ok(Some(Block::from(previous_hash, header, transactions, coinbase, signature)?))
     }
 }
 
@@ -408,8 +448,10 @@ pub struct BlockMemory<N: Network> {
     reverse_transactions_map: MemoryMap<N::TransactionID, N::BlockHash>,
     /// The transaction store.
     transaction_store: TransactionStore<N, TransactionMemory<N>>,
-    /// The coinbase proof map.
-    coinbase_proof_map: MemoryMap<N::BlockHash, Option<CoinbaseSolution<N>>>,
+    /// The coinbase solution map.
+    coinbase_solution_map: MemoryMap<N::BlockHash, Option<CoinbaseSolution<N>>>,
+    /// The coinbase puzzle commitment map.
+    coinbase_puzzle_commitment_map: MemoryMap<PuzzleCommitment<N>, N::BlockHash>,
     /// The signature map.
     signature_map: MemoryMap<N::BlockHash, Signature<N>>,
 }
@@ -425,7 +467,8 @@ impl<N: Network> BlockStorage<N> for BlockMemory<N> {
     type ReverseTransactionsMap = MemoryMap<N::TransactionID, N::BlockHash>;
     type TransactionStorage = TransactionMemory<N>;
     type TransitionStorage = TransitionMemory<N>;
-    type CoinbaseProofMap = MemoryMap<N::BlockHash, Option<CoinbaseSolution<N>>>;
+    type CoinbaseSolutionMap = MemoryMap<N::BlockHash, Option<CoinbaseSolution<N>>>;
+    type CoinbasePuzzleCommitmentMap = MemoryMap<PuzzleCommitment<N>, N::BlockHash>;
     type SignatureMap = MemoryMap<N::BlockHash, Signature<N>>;
 
     /// Initializes the block storage.
@@ -444,7 +487,8 @@ impl<N: Network> BlockStorage<N> for BlockMemory<N> {
             transactions_map: MemoryMap::default(),
             reverse_transactions_map: MemoryMap::default(),
             transaction_store,
-            coinbase_proof_map: MemoryMap::default(),
+            coinbase_solution_map: MemoryMap::default(),
+            coinbase_puzzle_commitment_map: MemoryMap::default(),
             signature_map: MemoryMap::default(),
         })
     }
@@ -489,9 +533,14 @@ impl<N: Network> BlockStorage<N> for BlockMemory<N> {
         &self.transaction_store
     }
 
-    /// Returns the coinbase proof map.
-    fn coinbase_proof_map(&self) -> &Self::CoinbaseProofMap {
-        &self.coinbase_proof_map
+    /// Returns the coinbase solution map.
+    fn coinbase_solution_map(&self) -> &Self::CoinbaseSolutionMap {
+        &self.coinbase_solution_map
+    }
+
+    /// Returns the coinbase puzzle commitment map.
+    fn coinbase_puzzle_commitment_map(&self) -> &Self::CoinbasePuzzleCommitmentMap {
+        &self.coinbase_puzzle_commitment_map
     }
 
     /// Returns the signature map.
@@ -579,6 +628,14 @@ impl<N: Network, B: BlockStorage<N>> BlockStore<N, B> {
     pub fn find_block_hash(&self, transaction_id: &N::TransactionID) -> Result<Option<N::BlockHash>> {
         self.storage.find_block_hash(transaction_id)
     }
+
+    /// Returns the block hash that contains the given `puzzle commitment`.
+    pub fn find_block_hash_from_puzzle_commitment(
+        &self,
+        puzzle_commitment: &PuzzleCommitment<N>,
+    ) -> Result<Option<N::BlockHash>> {
+        self.storage.find_block_hash_from_puzzle_commitment(puzzle_commitment)
+    }
 }
 
 impl<N: Network, B: BlockStorage<N>> BlockStore<N, B> {
@@ -612,9 +669,9 @@ impl<N: Network, B: BlockStorage<N>> BlockStore<N, B> {
         self.storage.get_block_transactions(block_hash)
     }
 
-    /// Returns the block coinbase proof for the given `block hash`.
-    pub fn get_block_coinbase_proof(&self, block_hash: &N::BlockHash) -> Result<Option<CoinbaseSolution<N>>> {
-        self.storage.get_block_coinbase_proof(block_hash)
+    /// Returns the block coinbase solution for the given `block hash`.
+    pub fn get_block_coinbase(&self, block_hash: &N::BlockHash) -> Result<Option<CoinbaseSolution<N>>> {
+        self.storage.get_block_coinbase(block_hash)
     }
 
     /// Returns the block signature for the given `block hash`.
@@ -643,6 +700,11 @@ impl<N: Network, B: BlockStorage<N>> BlockStore<N, B> {
     pub fn contains_block_hash(&self, block_hash: &N::BlockHash) -> Result<bool> {
         self.storage.reverse_id_map().contains_key(block_hash)
     }
+
+    /// Returns `true` if the given puzzle commitment exists.
+    pub fn contains_puzzle_commitment(&self, puzzle_commitment: &PuzzleCommitment<N>) -> Result<bool> {
+        self.storage.coinbase_puzzle_commitment_map().contains_key(puzzle_commitment)
+    }
 }
 
 impl<N: Network, B: BlockStorage<N>> BlockStore<N, B> {
@@ -660,6 +722,11 @@ impl<N: Network, B: BlockStorage<N>> BlockStore<N, B> {
     pub fn hashes(&self) -> impl '_ + Iterator<Item = Cow<'_, N::BlockHash>> {
         self.storage.reverse_id_map().keys()
     }
+
+    /// Returns an iterator over the puzzle commitments, for all blocks in `self`.
+    pub fn puzzle_commitments(&self) -> impl '_ + Iterator<Item = Cow<'_, PuzzleCommitment<N>>> {
+        self.storage.coinbase_puzzle_commitment_map().keys()
+    }
 }
 
 #[cfg(test)]
