diff --git a/Cargo.lock b/Cargo.lock
index 0a43632de..cc3e4d52d 100644
--- a/Cargo.lock
+++ b/Cargo.lock
@@ -1921,6 +1921,7 @@ dependencies = [
  "snarkvm-r1cs",
  "snarkvm-synthesizer",
  "snarkvm-utilities",
+ "snarkvm-wasm",
  "tempfile",
  "thiserror",
  "ureq",
@@ -2442,6 +2443,7 @@ dependencies = [
  "paste",
  "rand",
  "rayon",
+ "reqwest",
  "serde",
  "serde_json",
  "snarkvm-algorithms",
diff --git a/Cargo.toml b/Cargo.toml
index d658ccedb..3377c9d40 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -71,12 +71,19 @@ members = [
 path = "vm/lib.rs"
 
 [[bin]]
-name = "vm"
+name = "snarkvm"
 path = "vm/cli/main.rs"
 required-features = [ "cli" ]
 
 [features]
-default = [ "cli", "parallel", "parameters" ]
+default = [
+  "circuit",
+  "console",
+  "synthesizer",
+  "cli",
+  "parallel",
+  "parameters"
+]
 full = [
   "algorithms",
   "circuit",
@@ -96,9 +103,6 @@ cli = [
   "self_update",
   "serde_json",
   "thiserror",
-  "circuit",
-  "console",
-  "synthesizer"
 ]
 aleo-cli = [ "snarkvm-synthesizer/aleo-cli" ]
 cuda = [ "snarkvm-algorithms/cuda" ]
@@ -120,6 +124,7 @@ parameters = [ "snarkvm-parameters" ]
 r1cs = [ "snarkvm-r1cs" ]
 synthesizer = [ "snarkvm-synthesizer" ]
 utilities = [ "snarkvm-utilities" ]
+wasm = [ "snarkvm-wasm" ]
 
 [dependencies.snarkvm-algorithms]
 path = "./algorithms"
@@ -168,6 +173,11 @@ path = "./utilities"
 version = "0.9.6"
 optional = true
 
+[dependencies.snarkvm-wasm]
+path = "./wasm"
+version = "0.9.6"
+optional = true
+
 [dependencies.anyhow]
 version = "1.0.66"
 optional = true
diff --git a/circuit/program/src/request/verify.rs b/circuit/program/src/request/verify.rs
index e8d0aab52..4bb9d873a 100644
--- a/circuit/program/src/request/verify.rs
+++ b/circuit/program/src/request/verify.rs
@@ -340,7 +340,7 @@ mod tests {
             let input_record = console::Value::<<Circuit as Environment>::Network>::from_str(&record_string).unwrap();
             let input_external_record =
                 console::Value::<<Circuit as Environment>::Network>::from_str(&record_string).unwrap();
-            let inputs = vec![input_constant, input_public, input_private, input_record, input_external_record];
+            let inputs = [input_constant, input_public, input_private, input_record, input_external_record];
 
             // Construct the input types.
             let input_types = vec![
@@ -352,7 +352,8 @@ mod tests {
             ];
 
             // Compute the signed request.
-            let request = console::Request::sign(&private_key, program_id, function_name, &inputs, &input_types, rng)?;
+            let request =
+                console::Request::sign(&private_key, program_id, function_name, inputs.iter(), &input_types, rng)?;
             assert!(request.verify(&input_types));
 
             // Inject the request into a circuit.
diff --git a/console/network/environment/src/prelude.rs b/console/network/environment/src/prelude.rs
index 4b237ec58..9fe6f932f 100644
--- a/console/network/environment/src/prelude.rs
+++ b/console/network/environment/src/prelude.rs
@@ -76,7 +76,7 @@ pub use nom::{
     branch::alt,
     bytes::{complete::tag, streaming::take},
     character::complete::{alpha1, alphanumeric1, char, one_of},
-    combinator::{complete, map, map_res, opt, recognize},
+    combinator::{complete, fail, map, map_res, opt, recognize},
     multi::{many0, many1, separated_list0, separated_list1},
     sequence::{pair, terminated},
 };
diff --git a/console/network/src/testnet3.rs b/console/network/src/testnet3.rs
index 3b0f38dd5..f32d1b3ec 100644
--- a/console/network/src/testnet3.rs
+++ b/console/network/src/testnet3.rs
@@ -125,7 +125,7 @@ impl Network for Testnet3 {
     /// The function name for the inclusion circuit.
     const INCLUSION_FUNCTION_NAME: &'static str = snarkvm_parameters::testnet3::TESTNET3_INCLUSION_FUNCTION_NAME;
     /// The network name.
-    const NAME: &'static str = "Aleo Testnet3";
+    const NAME: &'static str = "Aleo Testnet 3";
 
     /// Returns the genesis block bytes.
     fn genesis_bytes() -> &'static [u8] {
diff --git a/console/program/src/data/identifier/mod.rs b/console/program/src/data/identifier/mod.rs
index 5deab8f57..8c11c9682 100644
--- a/console/program/src/data/identifier/mod.rs
+++ b/console/program/src/data/identifier/mod.rs
@@ -39,6 +39,31 @@ use snarkvm_console_types::{prelude::*, Field};
 #[derive(Copy, Clone)]
 pub struct Identifier<N: Network>(Field<N>, u8); // Number of bytes in the identifier.
 
+impl<N: Network> From<&Identifier<N>> for Identifier<N> {
+    /// Returns a copy of the identifier.
+    fn from(identifier: &Identifier<N>) -> Self {
+        *identifier
+    }
+}
+
+impl<N: Network> TryFrom<String> for Identifier<N> {
+    type Error = Error;
+
+    /// Initializes an identifier from a string.
+    fn try_from(identifier: String) -> Result<Self> {
+        Self::from_str(&identifier)
+    }
+}
+
+impl<N: Network> TryFrom<&String> for Identifier<N> {
+    type Error = Error;
+
+    /// Initializes an identifier from a string.
+    fn try_from(identifier: &String) -> Result<Self> {
+        Self::from_str(identifier)
+    }
+}
+
 impl<N: Network> TryFrom<&str> for Identifier<N> {
     type Error = Error;
 
diff --git a/console/program/src/data/record/is_owner.rs b/console/program/src/data/record/is_owner.rs
index e5ae5da0c..7e93d615b 100644
--- a/console/program/src/data/record/is_owner.rs
+++ b/console/program/src/data/record/is_owner.rs
@@ -17,11 +17,26 @@
 use super::*;
 
 impl<N: Network> Record<N, Ciphertext<N>> {
-    /// Decrypts `self` into plaintext using the given view key & nonce.
-    pub fn is_owner(&self, address: &Address<N>, view_key: &ViewKey<N>) -> bool {
+    /// Decrypts `self` into plaintext using the given view key.
+    pub fn is_owner(&self, view_key: &ViewKey<N>) -> bool {
+        // Compute the address.
+        let address = view_key.to_address();
+        // Check if the address is the owner.
+        self.is_owner_with_address_x_coordinate(view_key, &address.to_x_coordinate())
+    }
+
+    /// Decrypts `self` into plaintext using the x-coordinate of the address corresponding to the given view key.
+    pub fn is_owner_with_address_x_coordinate(&self, view_key: &ViewKey<N>, address_x_coordinate: &Field<N>) -> bool {
+        // In debug mode, check that the address corresponds to the given view key.
+        debug_assert_eq!(
+            &view_key.to_address().to_x_coordinate(),
+            address_x_coordinate,
+            "Failed to check record - view key and address do not match"
+        );
+
         match &self.owner {
             // If the owner is public, check if the address is the owner.
-            Owner::Public(owner) => owner == address,
+            Owner::Public(owner) => &owner.to_x_coordinate() == address_x_coordinate,
             // If the owner is private, decrypt the owner to check if it matches the address.
             Owner::Private(ciphertext) => {
                 // Compute the record view key.
@@ -38,7 +53,7 @@ impl<N: Network> Record<N, Ciphertext<N>> {
                 // together are an authenticated encryption scheme, we know that the ciphertext has not been malleated.
                 // Thus overall we know that if the x-coordinate matches that of `address`, then the underlying `address`es must also match.
                 // Therefore we can skip recomputing the address from `owner_x` and instead compare the x-coordinates directly.
-                owner_x == address.to_x_coordinate()
+                &owner_x == address_x_coordinate
             }
         }
     }
@@ -80,19 +95,15 @@ mod tests {
         // Encrypt the record.
         let ciphertext = record.encrypt(randomizer)?;
 
-        // Compute the address.
-        let address = Address::try_from(&view_key)?;
-
         // Ensure the record belongs to the owner.
-        assert!(ciphertext.is_owner(&address, &view_key));
+        assert!(ciphertext.is_owner(&view_key));
 
         // Sample a random view key and address.
         let private_key = PrivateKey::<N>::new(rng)?;
         let view_key = ViewKey::try_from(&private_key)?;
-        let address = Address::try_from(&private_key)?;
 
         // Ensure the random address is not the owner.
-        assert!(!ciphertext.is_owner(&address, &view_key));
+        assert!(!ciphertext.is_owner(&view_key));
 
         Ok(())
     }
diff --git a/console/program/src/data/value/mod.rs b/console/program/src/data/value/mod.rs
index c0c3b572d..4772d2c42 100644
--- a/console/program/src/data/value/mod.rs
+++ b/console/program/src/data/value/mod.rs
@@ -33,3 +33,46 @@ pub enum Value<N: Network> {
     /// A record value.
     Record(Record<N, Plaintext<N>>),
 }
+
+impl<N: Network> From<&Value<N>> for Value<N> {
+    /// Returns a clone of the value.
+    fn from(value: &Value<N>) -> Self {
+        value.clone()
+    }
+}
+
+impl<N: Network> TryFrom<Result<Value<N>>> for Value<N> {
+    type Error = Error;
+
+    /// Initializes a value from a result.
+    fn try_from(value: Result<Value<N>>) -> Result<Self> {
+        value
+    }
+}
+
+impl<N: Network> TryFrom<String> for Value<N> {
+    type Error = Error;
+
+    /// Initializes a value from a string.
+    fn try_from(value: String) -> Result<Self> {
+        Self::from_str(&value)
+    }
+}
+
+impl<N: Network> TryFrom<&String> for Value<N> {
+    type Error = Error;
+
+    /// Initializes a value from a string.
+    fn try_from(value: &String) -> Result<Self> {
+        Self::from_str(value)
+    }
+}
+
+impl<N: Network> TryFrom<&str> for Value<N> {
+    type Error = Error;
+
+    /// Initializes a value from a string.
+    fn try_from(value: &str) -> Result<Self> {
+        Self::from_str(value)
+    }
+}
diff --git a/console/program/src/data_types/record_type/parse.rs b/console/program/src/data_types/record_type/parse.rs
index 5ece81c1b..b18c81533 100644
--- a/console/program/src/data_types/record_type/parse.rs
+++ b/console/program/src/data_types/record_type/parse.rs
@@ -247,4 +247,24 @@ record message:
         );
         assert!(candidate.is_err());
     }
+
+    #[test]
+    fn test_parse_max_members() {
+        let mut string = "record message:\n    owner as address.private;\n    gates as u64.public;\n".to_string();
+        for i in 0..CurrentNetwork::MAX_DATA_ENTRIES {
+            string += &format!("    member_{} as field.private;\n", i);
+        }
+        let candidate = RecordType::<CurrentNetwork>::parse(&string);
+        assert!(candidate.is_ok());
+    }
+
+    #[test]
+    fn test_parse_too_many_members() {
+        let mut string = "record message:\n    owner as address.private;\n    gates as u64.public;\n".to_string();
+        for i in 0..=CurrentNetwork::MAX_DATA_ENTRIES {
+            string += &format!("    member_{} as field.private;\n", i);
+        }
+        let candidate = RecordType::<CurrentNetwork>::parse(&string);
+        assert!(candidate.is_err());
+    }
 }
diff --git a/console/program/src/data_types/struct_/parse.rs b/console/program/src/data_types/struct_/parse.rs
index 04f129573..84f3bc5c3 100644
--- a/console/program/src/data_types/struct_/parse.rs
+++ b/console/program/src/data_types/struct_/parse.rs
@@ -194,4 +194,22 @@ struct message:
             Struct::<CurrentNetwork>::parse("struct message:\n    first as field.public;\n    first as field.private;");
         assert!(candidate.is_err());
     }
+
+    #[test]
+    fn test_max_members() {
+        let mut string = "struct message:\n".to_string();
+        for i in 0..CurrentNetwork::MAX_DATA_ENTRIES {
+            string += &format!("    member_{} as field;\n", i);
+        }
+        assert!(Struct::<CurrentNetwork>::parse(&string).is_ok());
+    }
+
+    #[test]
+    fn test_too_many_members() {
+        let mut string = "struct message:\n".to_string();
+        for i in 0..=CurrentNetwork::MAX_DATA_ENTRIES {
+            string += &format!("    member_{} as field;\n", i);
+        }
+        assert!(Struct::<CurrentNetwork>::parse(&string).is_err());
+    }
 }
diff --git a/console/program/src/id/mod.rs b/console/program/src/id/mod.rs
index 3b1541142..12ec6c932 100644
--- a/console/program/src/id/mod.rs
+++ b/console/program/src/id/mod.rs
@@ -34,6 +34,13 @@ pub struct ProgramID<N: Network> {
     network: Identifier<N>,
 }
 
+impl<N: Network> From<&ProgramID<N>> for ProgramID<N> {
+    /// Returns a copy of the program ID.
+    fn from(program_id: &ProgramID<N>) -> Self {
+        *program_id
+    }
+}
+
 impl<N: Network> From<(Identifier<N>, Identifier<N>)> for ProgramID<N> {
     /// Initializes a program ID from a name and network-level domain identifier.
     fn from((name, network): (Identifier<N>, Identifier<N>)) -> Self {
@@ -41,6 +48,40 @@ impl<N: Network> From<(Identifier<N>, Identifier<N>)> for ProgramID<N> {
     }
 }
 
+impl<N: Network> TryFrom<String> for ProgramID<N> {
+    type Error = Error;
+
+    /// Initializes a program ID from a name and network-level domain identifier.
+    fn try_from(program_id: String) -> Result<Self> {
+        Self::from_str(&program_id)
+    }
+}
+
+impl<N: Network> TryFrom<&String> for ProgramID<N> {
+    type Error = Error;
+
+    /// Initializes a program ID from a name and network-level domain identifier.
+    fn try_from(program_id: &String) -> Result<Self> {
+        Self::from_str(program_id)
+    }
+}
+
+impl<N: Network> TryFrom<&str> for ProgramID<N> {
+    type Error = Error;
+
+    /// Initializes a program ID from a name and network-level domain identifier.
+    fn try_from(program_id: &str) -> Result<Self> {
+        // Split the program ID into a name and network-level domain.
+        let mut split = program_id.split('.');
+        // Parse the name and network.
+        if let (Some(name), Some(network), None) = (split.next(), split.next(), split.next()) {
+            Ok(Self { name: Identifier::from_str(name)?, network: Identifier::from_str(network)? })
+        } else {
+            bail!("Invalid program ID '{program_id}'")
+        }
+    }
+}
+
 impl<N: Network> ProgramID<N> {
     /// Returns the program name.
     #[inline]
diff --git a/console/program/src/request/mod.rs b/console/program/src/request/mod.rs
index 858fc53e4..781ecaeec 100644
--- a/console/program/src/request/mod.rs
+++ b/console/program/src/request/mod.rs
@@ -196,7 +196,7 @@ mod test_helpers {
                 let inputs = vec![input_constant, input_public, input_private, input_record, input_external_record];
 
                 // Construct the input types.
-                let input_types = vec![
+                let input_types = [
                     ValueType::from_str("amount.constant").unwrap(),
                     ValueType::from_str("amount.public").unwrap(),
                     ValueType::from_str("amount.private").unwrap(),
@@ -206,7 +206,7 @@ mod test_helpers {
 
                 // Compute the signed request.
                 let request =
-                    Request::sign(&private_key, program_id, function_name, &inputs, &input_types, rng).unwrap();
+                    Request::sign(&private_key, program_id, function_name, inputs.into_iter(), &input_types, rng).unwrap();
                 assert!(request.verify(&input_types));
                 request
             })
diff --git a/console/program/src/request/sign.rs b/console/program/src/request/sign.rs
index 767b6e24c..8b591d85b 100644
--- a/console/program/src/request/sign.rs
+++ b/console/program/src/request/sign.rs
@@ -24,7 +24,7 @@ impl<N: Network> Request<N> {
         private_key: &PrivateKey<N>,
         program_id: ProgramID<N>,
         function_name: Identifier<N>,
-        inputs: &[Value<N>],
+        inputs: impl ExactSizeIterator<Item = impl TryInto<Value<N>>>,
         input_types: &[ValueType<N>],
         rng: &mut R,
     ) -> Result<Self> {
@@ -78,11 +78,20 @@ impl<N: Network> Request<N> {
         message.extend([g_r, pk_sig, pr_sig, *caller].map(|point| point.to_x_coordinate()));
         message.extend([tvk, tcm, function_id]);
 
+        // Initialize a vector to store the prepared inputs.
+        let mut prepared_inputs = Vec::with_capacity(inputs.len());
         // Initialize a vector to store the input IDs.
         let mut input_ids = Vec::with_capacity(inputs.len());
 
         // Prepare the inputs.
-        for (index, (input, input_type)) in inputs.iter().zip_eq(input_types).enumerate() {
+        for (index, (input, input_type)) in inputs.zip_eq(input_types).enumerate() {
+            // Prepare the input.
+            let input = input.try_into().map_err(|_| {
+                anyhow!("Failed to parse input #{index} ('{input_type}') for '{program_id}/{function_name}'")
+            })?;
+            // Store the prepared input.
+            prepared_inputs.push(input.clone());
+
             match input_type {
                 // A constant input is hashed (using `tcm`) to a field element.
                 ValueType::Constant(..) => {
@@ -218,7 +227,7 @@ impl<N: Network> Request<N> {
             program_id,
             function_name,
             input_ids,
-            inputs: inputs.to_vec(),
+            inputs: prepared_inputs,
             signature: Signature::from((challenge, response, compute_key)),
             sk_tag,
             tvk,
diff --git a/console/program/src/request/verify.rs b/console/program/src/request/verify.rs
index cba202873..b544867ce 100644
--- a/console/program/src/request/verify.rs
+++ b/console/program/src/request/verify.rs
@@ -256,7 +256,7 @@ mod tests {
             let input_private = Value::from_str("{ token_amount: 9876543210u128 }").unwrap();
             let input_record = Value::from_str(&record_string).unwrap();
             let input_external_record = Value::from_str(&record_string).unwrap();
-            let inputs = vec![input_constant, input_public, input_private, input_record, input_external_record];
+            let inputs = [input_constant, input_public, input_private, input_record, input_external_record];
 
             // Construct the input types.
             let input_types = vec![
@@ -268,7 +268,8 @@ mod tests {
             ];
 
             // Compute the signed request.
-            let request = Request::sign(&private_key, program_id, function_name, &inputs, &input_types, rng).unwrap();
+            let request =
+                Request::sign(&private_key, program_id, function_name, inputs.into_iter(), &input_types, rng).unwrap();
             assert!(request.verify(&input_types));
         }
     }
diff --git a/synthesizer/Cargo.toml b/synthesizer/Cargo.toml
index 84f33245f..0bae1cf4e 100644
--- a/synthesizer/Cargo.toml
+++ b/synthesizer/Cargo.toml
@@ -91,6 +91,10 @@ version = "0.8"
 version = "1"
 optional = true
 
+[dependencies.reqwest]
+version = "0.11"
+features = [ "blocking", "json" ]
+
 [dependencies.serde]
 version = "1.0"
 
diff --git a/synthesizer/src/block/genesis.rs b/synthesizer/src/block/genesis.rs
index 542cf9c58..f708988eb 100644
--- a/synthesizer/src/block/genesis.rs
+++ b/synthesizer/src/block/genesis.rs
@@ -26,16 +26,12 @@ impl<N: Network> Block<N> {
     ) -> Result<Self> {
         // Prepare the caller.
         let caller = Address::try_from(private_key)?;
-        // Prepare the program ID.
-        let program_id = FromStr::from_str("credits.aleo")?;
-        // Prepare the function name.
-        let function_name = FromStr::from_str("mint")?;
         // Prepare the function inputs.
-        let inputs = [Value::from_str(&caller.to_string())?, Value::from_str(&format!("{}_u64", N::STARTING_SUPPLY))?];
+        let inputs = [caller.to_string(), format!("{}_u64", N::STARTING_SUPPLY)];
         // Authorize the call to start.
-        let authorization = vm.authorize(private_key, &program_id, function_name, &inputs, rng)?;
+        let authorization = vm.authorize(private_key, "credits.aleo", "mint", inputs, rng)?;
         // Execute the genesis function.
-        let transaction = Transaction::execute_authorization(vm, authorization, rng)?;
+        let transaction = Transaction::execute_authorization(vm, authorization, None, rng)?;
 
         // Prepare the transactions.
         let transactions = Transactions::from(&[transaction]);
diff --git a/synthesizer/src/block/mod.rs b/synthesizer/src/block/mod.rs
index d0a1c78c5..f3a68e4c6 100644
--- a/synthesizer/src/block/mod.rs
+++ b/synthesizer/src/block/mod.rs
@@ -39,7 +39,7 @@ use crate::{
 use console::{
     account::{Address, PrivateKey, Signature},
     network::prelude::*,
-    program::{Ciphertext, Record, Value},
+    program::{Ciphertext, Record},
     types::{Field, Group},
 };
 
@@ -310,14 +310,76 @@ impl<N: Network> Block<N> {
         self.transactions.commitments()
     }
 
+    /// Returns an iterator over the records, for all transition outputs that are records.
+    pub fn records(&self) -> impl '_ + Iterator<Item = (&Field<N>, &Record<N, Ciphertext<N>>)> {
+        self.transactions.records()
+    }
+
     /// Returns an iterator over the nonces, for all transition outputs that are records.
     pub fn nonces(&self) -> impl '_ + Iterator<Item = &Group<N>> {
         self.transactions.nonces()
     }
 
-    /// Returns an iterator over the fees, for all transitions.
-    pub fn fees(&self) -> impl '_ + Iterator<Item = &i64> {
-        self.transactions.fees()
+    /// Returns an iterator over the transaction fees, for all transactions.
+    pub fn transaction_fees(&self) -> impl '_ + Iterator<Item = Result<i64>> {
+        self.transactions.transaction_fees()
+    }
+}
+
+impl<N: Network> Block<N> {
+    /// Returns a consuming iterator over the transaction IDs, for all transactions in `self`.
+    pub fn into_transaction_ids(self) -> impl Iterator<Item = N::TransactionID> {
+        self.transactions.into_transaction_ids()
+    }
+
+    /// Returns a consuming iterator over all transactions in `self` that are deployments.
+    pub fn into_deployments(self) -> impl Iterator<Item = Deployment<N>> {
+        self.transactions.into_deployments()
+    }
+
+    /// Returns a consuming iterator over all transactions in `self` that are executions.
+    pub fn into_executions(self) -> impl Iterator<Item = Execution<N>> {
+        self.transactions.into_executions()
+    }
+
+    /// Returns a consuming iterator over all transitions.
+    pub fn into_transitions(self) -> impl Iterator<Item = Transition<N>> {
+        self.transactions.into_transitions()
+    }
+
+    /// Returns a consuming iterator over the transition IDs, for all transitions.
+    pub fn into_transition_ids(self) -> impl Iterator<Item = N::TransitionID> {
+        self.transactions.into_transition_ids()
+    }
+
+    /// Returns a consuming iterator over the transition public keys, for all transactions.
+    pub fn into_transition_public_keys(self) -> impl Iterator<Item = Group<N>> {
+        self.transactions.into_transition_public_keys()
+    }
+
+    /// Returns a consuming iterator over the tags, for all transition inputs that are records.
+    pub fn into_tags(self) -> impl Iterator<Item = Field<N>> {
+        self.transactions.into_tags()
+    }
+
+    /// Returns a consuming iterator over the serial numbers, for all transition inputs that are records.
+    pub fn into_serial_numbers(self) -> impl Iterator<Item = Field<N>> {
+        self.transactions.into_serial_numbers()
+    }
+
+    /// Returns a consuming iterator over the commitments, for all transition outputs that are records.
+    pub fn into_commitments(self) -> impl Iterator<Item = Field<N>> {
+        self.transactions.into_commitments()
+    }
+
+    /// Returns a consuming iterator over the records, for all transition outputs that are records.
+    pub fn into_records(self) -> impl Iterator<Item = (Field<N>, Record<N, Ciphertext<N>>)> {
+        self.transactions.into_records()
+    }
+
+    /// Returns a consuming iterator over the nonces, for all transition outputs that are records.
+    pub fn into_nonces(self) -> impl Iterator<Item = Group<N>> {
+        self.transactions.into_nonces()
     }
 }
 
@@ -342,17 +404,13 @@ pub(crate) mod test_helpers {
 
                 // Initialize the VM.
                 let vm = crate::vm::test_helpers::sample_vm();
-                // Prepare the program ID.
-                let program_id = FromStr::from_str("credits.aleo").unwrap();
-                // Prepare the function name.
-                let function_name = FromStr::from_str("mint").unwrap();
                 // Prepare the function inputs.
-                let inputs = [Value::from_str(&address.to_string()).unwrap(), Value::from_str("1_u64").unwrap()];
+                let inputs = [address.to_string(), "1_u64".to_string()];
                 // Authorize the call to start.
-                let authorization = vm.authorize(&private_key, &program_id, function_name, &inputs, rng).unwrap();
+                let authorization = vm.authorize(&private_key, "credits.aleo", "mint", inputs, rng).unwrap();
 
                 // Construct the transaction.
-                let transaction = Transaction::execute_authorization(&vm, authorization, rng).unwrap();
+                let transaction = Transaction::execute_authorization(&vm, authorization, None, rng).unwrap();
                 // Construct the transactions.
                 let transactions = Transactions::from(&[transaction.clone()]);
                 // Construct the block.
diff --git a/synthesizer/src/block/transaction/mod.rs b/synthesizer/src/block/transaction/mod.rs
index 23c00e4d0..1fcecc29a 100644
--- a/synthesizer/src/block/transaction/mod.rs
+++ b/synthesizer/src/block/transaction/mod.rs
@@ -25,6 +25,7 @@ use crate::{
     program::Program,
     vm::VM,
     ConsensusStorage,
+    Query,
 };
 use console::{
     account::PrivateKey,
@@ -84,42 +85,45 @@ impl<N: Network> Transaction<N> {
         private_key: &PrivateKey<N>,
         program: &Program<N>,
         (credits, fee_in_gates): (Record<N, Plaintext<N>>, u64),
+        query: Option<Query<N, C::BlockStorage>>,
         rng: &mut R,
     ) -> Result<Self> {
         // Compute the deployment.
         let deployment = vm.deploy(program, rng)?;
         // Compute the fee.
-        let (_, fee) = vm.execute_fee(private_key, credits, fee_in_gates, rng)?;
+        let (_, fee) = vm.execute_fee(private_key, credits, fee_in_gates, query, rng)?;
         // Initialize the transaction.
         Self::from_deployment(deployment, fee)
     }
 
-    /// Initializes a new execution transaction from an authorization.
+    /// Initializes a new execution transaction from an authorization, and an optional fee.
     pub fn execute_authorization<C: ConsensusStorage<N>, R: Rng + CryptoRng>(
         vm: &VM<N, C>,
         authorization: Authorization<N>,
+        query: Option<Query<N, C::BlockStorage>>,
         rng: &mut R,
     ) -> Result<Self> {
         // Compute the execution.
-        let (_, execution) = vm.execute(authorization, rng)?;
+        let (_, execution) = vm.execute(authorization, query, rng)?;
         // Initialize the transaction.
         Self::from_execution(execution, None)
     }
 
-    /// Initializes a new execution transaction from an authorization and additional fee.
+    /// Initializes a new execution transaction from an authorization, and an optional fee.
     pub fn execute_authorization_with_additional_fee<C: ConsensusStorage<N>, R: Rng + CryptoRng>(
         vm: &VM<N, C>,
         private_key: &PrivateKey<N>,
         authorization: Authorization<N>,
         additional_fee: Option<(Record<N, Plaintext<N>>, u64)>,
+        query: Option<Query<N, C::BlockStorage>>,
         rng: &mut R,
     ) -> Result<Self> {
         // Compute the execution.
-        let (_, execution) = vm.execute(authorization, rng)?;
+        let (_, execution) = vm.execute(authorization, query.clone(), rng)?;
         // Compute the additional fee, if it is present.
         let additional_fee = match additional_fee {
             Some((credits, additional_fee_in_gates)) => {
-                Some(vm.execute_fee(private_key, credits, additional_fee_in_gates, rng)?.1)
+                Some(vm.execute_fee(private_key, credits, additional_fee_in_gates, query, rng)?.1)
             }
             None => None,
         };
@@ -128,19 +132,21 @@ impl<N: Network> Transaction<N> {
     }
 
     /// Initializes a new execution transaction.
+    #[allow(clippy::too_many_arguments)]
     pub fn execute<C: ConsensusStorage<N>, R: Rng + CryptoRng>(
         vm: &VM<N, C>,
         private_key: &PrivateKey<N>,
-        program_id: &ProgramID<N>,
-        function_name: Identifier<N>,
-        inputs: &[Value<N>],
+        program_id: impl TryInto<ProgramID<N>>,
+        function_name: impl TryInto<Identifier<N>>,
+        inputs: impl ExactSizeIterator<Item = impl TryInto<Value<N>>>,
         additional_fee: Option<(Record<N, Plaintext<N>>, u64)>,
+        query: Option<Query<N, C::BlockStorage>>,
         rng: &mut R,
     ) -> Result<Self> {
         // Compute the authorization.
         let authorization = vm.authorize(private_key, program_id, function_name, inputs, rng)?;
         // Initialize the transaction.
-        Self::execute_authorization_with_additional_fee(vm, private_key, authorization, additional_fee, rng)
+        Self::execute_authorization_with_additional_fee(vm, private_key, authorization, additional_fee, query, rng)
     }
 }
 
diff --git a/synthesizer/src/block/transactions/mod.rs b/synthesizer/src/block/transactions/mod.rs
index dd4e0fbd1..12df591ca 100644
--- a/synthesizer/src/block/transactions/mod.rs
+++ b/synthesizer/src/block/transactions/mod.rs
@@ -195,27 +195,32 @@ impl<N: Network> Transactions<N> {
 
     /// Returns an iterator over the tags, for all transition inputs that are records.
     pub fn tags(&self) -> impl '_ + Iterator<Item = &Field<N>> {
-        self.transitions().flat_map(Transition::tags)
+        self.iter().flat_map(Transaction::tags)
     }
 
     /// Returns an iterator over the serial numbers, for all transition inputs that are records.
     pub fn serial_numbers(&self) -> impl '_ + Iterator<Item = &Field<N>> {
-        self.transitions().flat_map(Transition::serial_numbers)
+        self.iter().flat_map(Transaction::serial_numbers)
     }
 
     /// Returns an iterator over the commitments, for all transition outputs that are records.
     pub fn commitments(&self) -> impl '_ + Iterator<Item = &Field<N>> {
-        self.transitions().flat_map(Transition::commitments)
+        self.iter().flat_map(Transaction::commitments)
+    }
+
+    /// Returns an iterator over the records, for all transition outputs that are records.
+    pub fn records(&self) -> impl '_ + Iterator<Item = (&Field<N>, &Record<N, Ciphertext<N>>)> {
+        self.iter().flat_map(Transaction::records)
     }
 
     /// Returns an iterator over the nonces, for all transition outputs that are records.
     pub fn nonces(&self) -> impl '_ + Iterator<Item = &Group<N>> {
-        self.transitions().flat_map(Transition::nonces)
+        self.iter().flat_map(Transaction::nonces)
     }
 
-    /// Returns an iterator over the fees, for all transitions.
-    pub fn fees(&self) -> impl '_ + Iterator<Item = &i64> {
-        self.transitions().map(Transition::fee)
+    /// Returns an iterator over the transaction fees, for all transactions.
+    pub fn transaction_fees(&self) -> impl '_ + Iterator<Item = Result<i64>> {
+        self.iter().map(Transaction::fee)
     }
 }
 
@@ -268,22 +273,27 @@ impl<N: Network> Transactions<N> {
 
     /// Returns a consuming iterator over the tags, for all transition inputs that are records.
     pub fn into_tags(self) -> impl Iterator<Item = Field<N>> {
-        self.into_transitions().flat_map(Transition::into_tags)
+        self.into_iter().flat_map(Transaction::into_tags)
     }
 
     /// Returns a consuming iterator over the serial numbers, for all transition inputs that are records.
     pub fn into_serial_numbers(self) -> impl Iterator<Item = Field<N>> {
-        self.into_transitions().flat_map(Transition::into_serial_numbers)
+        self.into_iter().flat_map(Transaction::into_serial_numbers)
     }
 
     /// Returns a consuming iterator over the commitments, for all transition outputs that are records.
     pub fn into_commitments(self) -> impl Iterator<Item = Field<N>> {
-        self.into_transitions().flat_map(Transition::into_commitments)
+        self.into_iter().flat_map(Transaction::into_commitments)
+    }
+
+    /// Returns a consuming iterator over the records, for all transition outputs that are records.
+    pub fn into_records(self) -> impl Iterator<Item = (Field<N>, Record<N, Ciphertext<N>>)> {
+        self.into_iter().flat_map(Transaction::into_records)
     }
 
     /// Returns a consuming iterator over the nonces, for all transition outputs that are records.
     pub fn into_nonces(self) -> impl Iterator<Item = Group<N>> {
-        self.into_transitions().flat_map(Transition::into_nonces)
+        self.into_iter().flat_map(Transaction::into_nonces)
     }
 }
 
diff --git a/synthesizer/src/process/authorize.rs b/synthesizer/src/process/authorize.rs
index d11e1ece9..e2a3fa8f4 100644
--- a/synthesizer/src/process/authorize.rs
+++ b/synthesizer/src/process/authorize.rs
@@ -22,9 +22,9 @@ impl<N: Network> Process<N> {
     pub fn authorize<A: circuit::Aleo<Network = N>, R: Rng + CryptoRng>(
         &self,
         private_key: &PrivateKey<N>,
-        program_id: &ProgramID<N>,
-        function_name: Identifier<N>,
-        inputs: &[Value<N>],
+        program_id: impl TryInto<ProgramID<N>>,
+        function_name: impl TryInto<Identifier<N>>,
+        inputs: impl ExactSizeIterator<Item = impl TryInto<Value<N>>>,
         rng: &mut R,
     ) -> Result<Authorization<N>> {
         // Authorize the call.
diff --git a/synthesizer/src/process/execute_fee.rs b/synthesizer/src/process/execute_fee.rs
index 88b18f360..bd2e41c97 100644
--- a/synthesizer/src/process/execute_fee.rs
+++ b/synthesizer/src/process/execute_fee.rs
@@ -32,17 +32,17 @@ impl<N: Network> Process<N> {
         let function_name = Identifier::from_str("fee")?;
 
         // Retrieve the input types.
-        let input_types = self.get_program(&program_id)?.get_function(&function_name)?.input_types();
+        let input_types = self.get_program(program_id)?.get_function(&function_name)?.input_types();
         // Construct the inputs.
-        let inputs = vec![Value::Record(credits), Value::from_str(&format!("{}", U64::<N>::new(fee_in_gates)))?];
+        let inputs = [Value::Record(credits), Value::from_str(&format!("{}", U64::<N>::new(fee_in_gates)))?];
         // Compute the request.
-        let request = Request::sign(private_key, program_id, function_name, &inputs, &input_types, rng)?;
+        let request = Request::sign(private_key, program_id, function_name, inputs.iter(), &input_types, rng)?;
         // Initialize the authorization.
         let authorization = Authorization::new(&[request.clone()]);
         // Construct the call stack.
         let call_stack = CallStack::Authorize(vec![request], *private_key, authorization.clone());
         // Construct the authorization from the function.
-        let _response = self.get_stack(&program_id)?.execute_function::<A, R>(call_stack, rng)?;
+        let _response = self.get_stack(program_id)?.execute_function::<A, R>(call_stack, rng)?;
 
         // Retrieve the main request (without popping it).
         let request = authorization.peek_next()?;
diff --git a/synthesizer/src/process/mod.rs b/synthesizer/src/process/mod.rs
index c73187dc7..4f2132cde 100644
--- a/synthesizer/src/process/mod.rs
+++ b/synthesizer/src/process/mod.rs
@@ -169,37 +169,47 @@ impl<N: Network> Process<N> {
 
     /// Returns the stack for the given program ID.
     #[inline]
-    pub fn get_stack(&self, program_id: &ProgramID<N>) -> Result<&Stack<N>> {
+    pub fn get_stack(&self, program_id: impl TryInto<ProgramID<N>>) -> Result<&Stack<N>> {
+        // Prepare the program ID.
+        let program_id = program_id.try_into().map_err(|_| anyhow!("Invalid program ID"))?;
         // Retrieve the stack.
-        let stack = self.stacks.get(program_id).ok_or_else(|| anyhow!("Program '{program_id}' does not exist"))?;
+        let stack = self.stacks.get(&program_id).ok_or_else(|| anyhow!("Program '{program_id}' does not exist"))?;
         // Ensure the program ID matches.
-        ensure!(stack.program_id() == program_id, "Expected program '{}', found '{program_id}'", stack.program_id());
+        ensure!(stack.program_id() == &program_id, "Expected program '{}', found '{program_id}'", stack.program_id());
         // Return the stack.
         Ok(stack)
     }
 
     /// Returns the program for the given program ID.
     #[inline]
-    pub fn get_program(&self, program_id: &ProgramID<N>) -> Result<&Program<N>> {
+    pub fn get_program(&self, program_id: impl TryInto<ProgramID<N>>) -> Result<&Program<N>> {
         self.get_stack(program_id).map(Stack::program)
     }
 
     /// Returns the proving key for the given program ID and function name.
     #[inline]
-    pub fn get_proving_key(&self, program_id: &ProgramID<N>, function_name: &Identifier<N>) -> Result<ProvingKey<N>> {
+    pub fn get_proving_key(
+        &self,
+        program_id: impl TryInto<ProgramID<N>>,
+        function_name: impl TryInto<Identifier<N>>,
+    ) -> Result<ProvingKey<N>> {
+        // Prepare the function name.
+        let function_name = function_name.try_into().map_err(|_| anyhow!("Invalid function name"))?;
         // Return the proving key.
-        self.get_stack(program_id)?.get_proving_key(function_name)
+        self.get_stack(program_id)?.get_proving_key(&function_name)
     }
 
     /// Returns the verifying key for the given program ID and function name.
     #[inline]
     pub fn get_verifying_key(
         &self,
-        program_id: &ProgramID<N>,
-        function_name: &Identifier<N>,
+        program_id: impl TryInto<ProgramID<N>>,
+        function_name: impl TryInto<Identifier<N>>,
     ) -> Result<VerifyingKey<N>> {
+        // Prepare the function name.
+        let function_name = function_name.try_into().map_err(|_| anyhow!("Invalid function name"))?;
         // Return the verifying key.
-        self.get_stack(program_id)?.get_verifying_key(function_name)
+        self.get_stack(program_id)?.get_verifying_key(&function_name)
     }
 
     /// Inserts the given proving key, for the given program ID and function name.
@@ -241,11 +251,7 @@ impl<N: Network> Process<N> {
 pub(crate) mod test_helpers {
     use super::*;
     use crate::{Process, Program, Transition};
-    use console::{
-        account::PrivateKey,
-        network::Testnet3,
-        program::{Identifier, Value},
-    };
+    use console::{account::PrivateKey, network::Testnet3, program::Identifier};
 
     use once_cell::sync::OnceCell;
 
@@ -288,8 +294,8 @@ function compute:
                 process.synthesize_key::<CurrentAleo, _>(program.id(), &function_name, rng).unwrap();
 
                 // Get the proving and verifying key.
-                let proving_key = process.get_proving_key(program.id(), &function_name).unwrap();
-                let verifying_key = process.get_verifying_key(program.id(), &function_name).unwrap();
+                let proving_key = process.get_proving_key(program.id(), function_name).unwrap();
+                let verifying_key = process.get_verifying_key(program.id(), function_name).unwrap();
 
                 (function_name, proving_key, verifying_key)
             })
@@ -330,10 +336,7 @@ function compute:
                         &caller_private_key,
                         program.id(),
                         function_name,
-                        &[
-                            Value::<CurrentNetwork>::from_str("5u32").unwrap(),
-                            Value::<CurrentNetwork>::from_str("10u32").unwrap(),
-                        ],
+                        ["5u32", "10u32"].into_iter(),
                         rng,
                     )
                     .unwrap();
@@ -406,7 +409,7 @@ mod tests {
                 &caller_private_key,
                 program.id(),
                 Identifier::from_str("mint").unwrap(),
-                &[r0.clone(), r1.clone()],
+                [r0.clone(), r1.clone()].iter(),
                 rng,
             )
             .unwrap();
@@ -476,7 +479,7 @@ mod tests {
                 &caller_private_key,
                 program.id(),
                 Identifier::from_str("mint").unwrap(),
-                &[r0, r1],
+                [r0, r1].iter(),
                 rng,
             )
             .unwrap();
@@ -560,7 +563,13 @@ function hello_world:
 
         // Authorize the function call.
         let authorization = process
-            .authorize::<CurrentAleo, _>(&caller_private_key, program.id(), function_name, &[input_a, input_b], rng)
+            .authorize::<CurrentAleo, _>(
+                &caller_private_key,
+                program.id(),
+                function_name,
+                [input_a, input_b].iter(),
+                rng,
+            )
             .unwrap();
         assert_eq!(authorization.len(), 1);
         let request = authorization.peek_next().unwrap();
@@ -652,7 +661,7 @@ function hello_world:
 
         // Authorize the function call.
         let authorization = process
-            .authorize::<CurrentAleo, _>(&caller_private_key, program.id(), function_name, &[input], rng)
+            .authorize::<CurrentAleo, _>(&caller_private_key, program.id(), function_name, [input].iter(), rng)
             .unwrap();
         assert_eq!(authorization.len(), 1);
         let request = authorization.peek_next().unwrap();
@@ -719,8 +728,10 @@ function hello_world:
         let caller_private_key = PrivateKey::<CurrentNetwork>::new(rng).unwrap();
 
         // Authorize the function call.
-        let authorization =
-            process.authorize::<CurrentAleo, _>(&caller_private_key, program.id(), function_name, &[], rng).unwrap();
+        let inputs: &[Value<CurrentNetwork>] = &[];
+        let authorization = process
+            .authorize::<CurrentAleo, _>(&caller_private_key, program.id(), function_name, inputs.iter(), rng)
+            .unwrap();
         assert_eq!(authorization.len(), 1);
 
         // Declare the output value.
@@ -821,7 +832,7 @@ function compute:
 
         // Authorize the function call.
         let authorization = process
-            .authorize::<CurrentAleo, _>(&caller_private_key, program.id(), function_name, &[r0, r1, r2], rng)
+            .authorize::<CurrentAleo, _>(&caller_private_key, program.id(), function_name, [r0, r1, r2].iter(), rng)
             .unwrap();
         assert_eq!(authorization.len(), 1);
         let request = authorization.peek_next().unwrap();
@@ -965,7 +976,7 @@ function transfer:
 
         // Authorize the function call.
         let authorization = process
-            .authorize::<CurrentAleo, _>(&caller0_private_key, program1.id(), function_name, &[r0, r1, r2], rng)
+            .authorize::<CurrentAleo, _>(&caller0_private_key, program1.id(), function_name, [r0, r1, r2].iter(), rng)
             .unwrap();
         assert_eq!(authorization.len(), 5);
         println!("\nAuthorize\n{:#?}\n\n", authorization.to_vec_deque());
@@ -1099,7 +1110,7 @@ finalize compute:
 
         // Authorize the function call.
         let authorization = process
-            .authorize::<CurrentAleo, _>(&caller_private_key, program.id(), function_name, &[r0, r1, r2], rng)
+            .authorize::<CurrentAleo, _>(&caller_private_key, program.id(), function_name, [r0, r1, r2].iter(), rng)
             .unwrap();
         assert_eq!(authorization.len(), 1);
 
@@ -1196,7 +1207,7 @@ finalize compute:
 
         // Authorize the function call.
         let authorization = process
-            .authorize::<CurrentAleo, _>(&caller_private_key, program.id(), function_name, &[r0, r1, r2], rng)
+            .authorize::<CurrentAleo, _>(&caller_private_key, program.id(), function_name, [r0, r1, r2].iter(), rng)
             .unwrap();
         assert_eq!(authorization.len(), 1);
 
@@ -1311,7 +1322,7 @@ finalize mint_public:
 
         // Authorize the function call.
         let authorization = process
-            .authorize::<CurrentAleo, _>(&caller_private_key, program.id(), function_name, &[r0, r1], rng)
+            .authorize::<CurrentAleo, _>(&caller_private_key, program.id(), function_name, [r0, r1].iter(), rng)
             .unwrap();
         assert_eq!(authorization.len(), 1);
 
@@ -1447,7 +1458,7 @@ function mint:
 
         // Authorize the function call.
         let authorization = process
-            .authorize::<CurrentAleo, _>(&caller_private_key, program1.id(), function_name, &[r0, r1], rng)
+            .authorize::<CurrentAleo, _>(&caller_private_key, program1.id(), function_name, [r0, r1].iter(), rng)
             .unwrap();
         assert_eq!(authorization.len(), 2);
 
diff --git a/synthesizer/src/process/stack/authorize.rs b/synthesizer/src/process/stack/authorize.rs
index c2eb28b3b..b0f0da71f 100644
--- a/synthesizer/src/process/stack/authorize.rs
+++ b/synthesizer/src/process/stack/authorize.rs
@@ -22,13 +22,15 @@ impl<N: Network> Stack<N> {
     pub fn authorize<A: circuit::Aleo<Network = N>, R: Rng + CryptoRng>(
         &self,
         private_key: &PrivateKey<N>,
-        function_name: Identifier<N>,
-        inputs: &[Value<N>],
+        function_name: impl TryInto<Identifier<N>>,
+        inputs: impl ExactSizeIterator<Item = impl TryInto<Value<N>>>,
         rng: &mut R,
     ) -> Result<Authorization<N>> {
         // Ensure the program contains functions.
         ensure!(!self.program.functions().is_empty(), "Program '{}' has no functions", self.program.id());
 
+        // Prepare the function name.
+        let function_name = function_name.try_into().map_err(|_| anyhow!("Invalid function name"))?;
         // Retrieve the function.
         let function = self.get_function(&function_name)?;
         // Retrieve the input types.
diff --git a/synthesizer/src/process/stack/deploy.rs b/synthesizer/src/process/stack/deploy.rs
index 7434892a8..c8e2d44da 100644
--- a/synthesizer/src/process/stack/deploy.rs
+++ b/synthesizer/src/process/stack/deploy.rs
@@ -117,8 +117,14 @@ impl<N: Network> Stack<N> {
                 .collect::<Result<Vec<_>>>()?;
 
             // Compute the request, with a burner private key.
-            let request =
-                Request::sign(&burner_private_key, *program_id, *function.name(), &inputs, &input_types, rng)?;
+            let request = Request::sign(
+                &burner_private_key,
+                *program_id,
+                *function.name(),
+                inputs.into_iter(),
+                &input_types,
+                rng,
+            )?;
             // Initialize the assignments.
             let assignments = Assignments::<N>::default();
             // Initialize the call stack.
diff --git a/synthesizer/src/process/stack/helpers/synthesize.rs b/synthesizer/src/process/stack/helpers/synthesize.rs
index f2c77171b..feb98fb1e 100644
--- a/synthesizer/src/process/stack/helpers/synthesize.rs
+++ b/synthesizer/src/process/stack/helpers/synthesize.rs
@@ -53,7 +53,8 @@ impl<N: Network> Stack<N> {
             .collect::<Result<Vec<_>>>()?;
 
         // Compute the request, with a burner private key.
-        let request = Request::sign(&burner_private_key, *program_id, *function_name, &inputs, &input_types, rng)?;
+        let request =
+            Request::sign(&burner_private_key, *program_id, *function_name, inputs.into_iter(), &input_types, rng)?;
         // Initialize the authorization.
         let authorization = Authorization::new(&[request.clone()]);
         // Initialize the call stack.
diff --git a/synthesizer/src/process/stack/inclusion/mod.rs b/synthesizer/src/process/stack/inclusion/mod.rs
index e2b10d50e..d3b6cef3f 100644
--- a/synthesizer/src/process/stack/inclusion/mod.rs
+++ b/synthesizer/src/process/stack/inclusion/mod.rs
@@ -21,6 +21,7 @@ use crate::{
     Fee,
     Input,
     Output,
+    Program,
     Proof,
     ProvingKey,
     Stack,
@@ -37,6 +38,99 @@ use console::{
 use console::program::{Identifier, ProgramID};
 use std::collections::HashMap;
 
+#[derive(Clone)]
+pub enum Query<N: Network, B: BlockStorage<N>> {
+    /// The block store from the VM.
+    VM(BlockStore<N, B>),
+    /// The base URL of the node.
+    REST(String),
+}
+
+impl<N: Network, B: BlockStorage<N>> From<BlockStore<N, B>> for Query<N, B> {
+    fn from(block_store: BlockStore<N, B>) -> Self {
+        Self::VM(block_store)
+    }
+}
+
+impl<N: Network, B: BlockStorage<N>> From<&BlockStore<N, B>> for Query<N, B> {
+    fn from(block_store: &BlockStore<N, B>) -> Self {
+        Self::VM(block_store.clone())
+    }
+}
+
+impl<N: Network, B: BlockStorage<N>> From<reqwest::Url> for Query<N, B> {
+    fn from(url: reqwest::Url) -> Self {
+        Self::REST(url.to_string())
+    }
+}
+
+impl<N: Network, B: BlockStorage<N>> From<&reqwest::Url> for Query<N, B> {
+    fn from(url: &reqwest::Url) -> Self {
+        Self::REST(url.to_string())
+    }
+}
+
+impl<N: Network, B: BlockStorage<N>> From<String> for Query<N, B> {
+    fn from(url: String) -> Self {
+        Self::REST(url)
+    }
+}
+
+impl<N: Network, B: BlockStorage<N>> From<&String> for Query<N, B> {
+    fn from(url: &String) -> Self {
+        Self::REST(url.to_string())
+    }
+}
+
+impl<N: Network, B: BlockStorage<N>> From<&str> for Query<N, B> {
+    fn from(url: &str) -> Self {
+        Self::REST(url.to_string())
+    }
+}
+
+impl<N: Network, B: BlockStorage<N>> Query<N, B> {
+    /// Returns the program for the given program ID.
+    pub fn get_program(&self, program_id: &ProgramID<N>) -> Result<Program<N>> {
+        match self {
+            Self::VM(block_store) => {
+                block_store.get_program(program_id)?.ok_or_else(|| anyhow!("Program {program_id} not found in storage"))
+            }
+            Self::REST(url) => match N::ID {
+                3 => Ok(Self::get_request(&format!("{url}/testnet3/program/{program_id}"))?.json()?),
+                _ => bail!("Unsupported network ID in inclusion query"),
+            },
+        }
+    }
+
+    /// Returns the current state root.
+    pub fn current_state_root(&self) -> Result<N::StateRoot> {
+        match self {
+            Self::VM(block_store) => Ok(block_store.current_state_root()),
+            Self::REST(url) => match N::ID {
+                3 => Ok(Self::get_request(&format!("{url}/testnet3/latest/stateRoot"))?.json()?),
+                _ => bail!("Unsupported network ID in inclusion query"),
+            },
+        }
+    }
+
+    /// Returns a state path for the given `commitment`.
+    pub fn get_state_path_for_commitment(&self, commitment: &Field<N>) -> Result<StatePath<N>> {
+        match self {
+            Self::VM(block_store) => block_store.get_state_path_for_commitment(commitment),
+            Self::REST(url) => match N::ID {
+                3 => Ok(Self::get_request(&format!("{url}/testnet3/statePath/{commitment}"))?.json()?),
+                _ => bail!("Unsupported network ID in inclusion query"),
+            },
+        }
+    }
+
+    /// Performs a GET request to the given URL.
+    fn get_request(url: &str) -> Result<reqwest::blocking::Response> {
+        let response = reqwest::blocking::get(url)?;
+        if response.status().is_success() { Ok(response) } else { bail!("Failed to fetch from {}", url) }
+    }
+}
+
 #[derive(Clone, Debug)]
 struct InputTask<N: Network> {
     /// The commitment.
@@ -103,11 +197,13 @@ impl<N: Network> Inclusion<N> {
     }
 
     /// Returns the inclusion assignments for the given execution.
-    pub fn prepare_execution<B: BlockStorage<N>>(
+    pub fn prepare_execution<B: BlockStorage<N>, Q: Into<Query<N, B>>>(
         &self,
         execution: &Execution<N>,
-        block_store: &BlockStore<N, B>,
+        query: Q,
     ) -> Result<Vec<InclusionAssignment<N>>> {
+        let query = query.into();
+
         // Ensure the number of leaves is within the Merkle tree size.
         Transaction::check_execution_size(execution)?;
 
@@ -133,7 +229,7 @@ impl<N: Network> Inclusion<N> {
                         let state_path = match task.is_local {
                             true => {
                                 // Retrieve the global state root.
-                                let global_state_root = block_store.current_state_root();
+                                let global_state_root = query.current_state_root()?;
                                 // Compute the transaction path.
                                 let transaction_path =
                                     transaction_tree.prove(transition_index, &transaction_leaf.to_bits_le())?;
@@ -151,7 +247,7 @@ impl<N: Network> Inclusion<N> {
                                     transition_leaf,
                                 )?
                             }
-                            false => block_store.get_state_path_for_commitment(&task.commitment)?,
+                            false => query.get_state_path_for_commitment(&task.commitment)?,
                         };
 
                         // Ensure the global state root is the same across iterations.
@@ -242,10 +338,10 @@ impl<N: Network> Inclusion<N> {
     }
 
     /// Returns the inclusion assignments for the given fee transition.
-    pub fn prepare_fee<B: BlockStorage<N>>(
+    pub fn prepare_fee<B: BlockStorage<N>, Q: Into<Query<N, B>>>(
         &self,
         fee_transition: &Transition<N>,
-        block_store: &BlockStore<N, B>,
+        query: Q,
     ) -> Result<Vec<InclusionAssignment<N>>> {
         // Ensure the fee has the correct program ID.
         let fee_program_id = ProgramID::from_str("credits.aleo")?;
@@ -265,11 +361,13 @@ impl<N: Network> Inclusion<N> {
         // Process the input tasks.
         match self.input_tasks.get(fee_transition.id()) {
             Some(tasks) => {
+                let query = query.into();
+
                 for task in tasks {
                     // Retrieve the local state root.
                     let local_state_root = (*transaction_tree.root()).into();
                     // Construct the state path.
-                    let state_path = block_store.get_state_path_for_commitment(&task.commitment)?;
+                    let state_path = query.get_state_path_for_commitment(&task.commitment)?;
 
                     // Ensure the global state root is the same across iterations.
                     if *global_state_root != Field::zero() && global_state_root != state_path.global_state_root() {
diff --git a/synthesizer/src/program/instruction/operation/call.rs b/synthesizer/src/program/instruction/operation/call.rs
index 28cbdc16f..7742e8ab6 100644
--- a/synthesizer/src/program/instruction/operation/call.rs
+++ b/synthesizer/src/program/instruction/operation/call.rs
@@ -300,7 +300,7 @@ impl<N: Network> Call<N> {
                             &private_key,
                             *substack.program_id(),
                             *function.name(),
-                            &inputs,
+                            inputs.iter(),
                             &function.input_types(),
                             rng,
                         )?;
@@ -325,7 +325,7 @@ impl<N: Network> Call<N> {
                             &private_key,
                             *substack.program_id(),
                             *function.name(),
-                            &inputs,
+                            inputs.iter(),
                             &function.input_types(),
                             rng,
                         )?;
diff --git a/synthesizer/src/program/instruction/operation/cast.rs b/synthesizer/src/program/instruction/operation/cast.rs
index 252cb41d2..2e71577a8 100644
--- a/synthesizer/src/program/instruction/operation/cast.rs
+++ b/synthesizer/src/program/instruction/operation/cast.rs
@@ -446,13 +446,7 @@ impl<N: Network> Parser for Cast<N> {
         // Parse the opcode from the string.
         let (string, _) = tag(*Self::opcode())(string)?;
         // Parse the operands from the string.
-        let (string, operands) = map_res(many1(parse_operand), |operands: Vec<Operand<N>>| {
-            // Ensure the number of operands is within the bounds.
-            match operands.len() <= N::MAX_OPERANDS {
-                true => Ok(operands),
-                false => Err(error("Failed to parse 'cast' opcode: too many operands")),
-            }
-        })(string)?;
+        let (string, operands) = many1(parse_operand)(string)?;
         // Parse the whitespace from the string.
         let (string, _) = Sanitizer::parse_whitespaces(string)?;
         // Parse the "into" from the string.
@@ -469,8 +463,20 @@ impl<N: Network> Parser for Cast<N> {
         let (string, _) = Sanitizer::parse_whitespaces(string)?;
         // Parse the register type from the string.
         let (string, register_type) = RegisterType::parse(string)?;
-
-        Ok((string, Self { operands, destination, register_type }))
+        // Check that the number of operands does not exceed the maximum number of data entries.
+        let max_operands = match register_type {
+            RegisterType::Plaintext(_) => N::MAX_DATA_ENTRIES,
+            // Note that if the register type is a record, then we must account for `owner` and `gates` which are not data entries.
+            RegisterType::Record(_) | RegisterType::ExternalRecord(_) => N::MAX_DATA_ENTRIES + 2,
+        };
+        match operands.len() <= max_operands {
+            true => Ok((string, Self { operands, destination, register_type })),
+            false => {
+                map_res(fail, |_: ParserResult<Self>| Err(error("Failed to parse 'cast' opcode: too many operands")))(
+                    string,
+                )
+            }
+        }
     }
 }
 
@@ -503,8 +509,13 @@ impl<N: Network> Display for Cast<N> {
     /// Prints the operation to a string.
     fn fmt(&self, f: &mut Formatter) -> fmt::Result {
         // Ensure the number of operands is within the bounds.
-        if self.operands.len().is_zero() || self.operands.len() > N::MAX_OPERANDS {
-            eprintln!("The number of operands must be nonzero and <= {}", N::MAX_OPERANDS);
+        let max_operands = match self.register_type() {
+            RegisterType::Plaintext(_) => N::MAX_DATA_ENTRIES,
+            // Note that if the register type is a record, then we must account for `owner` and `gates` which are not data entries.
+            RegisterType::Record(_) | RegisterType::ExternalRecord(_) => N::MAX_DATA_ENTRIES + 2,
+        };
+        if self.operands.len().is_zero() || self.operands.len() > max_operands {
+            eprintln!("The number of operands must be nonzero and <= {}", max_operands);
             return Err(fmt::Error);
         }
         // Print the operation.
@@ -520,9 +531,10 @@ impl<N: Network> FromBytes for Cast<N> {
         // Read the number of operands.
         let num_operands = u8::read_le(&mut reader)? as usize;
 
-        // Ensure the number of operands is within the bounds.
-        if num_operands.is_zero() || num_operands > N::MAX_OPERANDS {
-            return Err(error(format!("The number of operands must be nonzero and <= {}", N::MAX_OPERANDS)));
+        // Ensure that the number of operands does not exceed the upper bound.
+        // Note: although a similar check is performed later, this check is performed to ensure that an exceedingly large number of operands is not allocated.
+        if num_operands.is_zero() || num_operands > N::MAX_DATA_ENTRIES + 2 {
+            return Err(error(format!("The number of operands must be nonzero and <= {}", N::MAX_DATA_ENTRIES + 2)));
         }
 
         // Initialize the vector for the operands.
@@ -538,6 +550,16 @@ impl<N: Network> FromBytes for Cast<N> {
         // Read the casted register type.
         let register_type = RegisterType::read_le(&mut reader)?;
 
+        // Ensure the number of operands is within the bounds for the register type.
+        let max_operands = match register_type {
+            RegisterType::Plaintext(_) => N::MAX_DATA_ENTRIES,
+            // Note that if the register type is a record, then we must account for `owner` and `gates` which are not data entries.
+            RegisterType::Record(_) | RegisterType::ExternalRecord(_) => N::MAX_DATA_ENTRIES + 2,
+        };
+        if num_operands.is_zero() || num_operands > max_operands {
+            return Err(error(format!("The number of operands must be nonzero and <= {}", max_operands)));
+        }
+
         // Return the operation.
         Ok(Self { operands, destination, register_type })
     }
@@ -547,8 +569,13 @@ impl<N: Network> ToBytes for Cast<N> {
     /// Writes the operation to a buffer.
     fn write_le<W: Write>(&self, mut writer: W) -> IoResult<()> {
         // Ensure the number of operands is within the bounds.
-        if self.operands.len().is_zero() || self.operands.len() > N::MAX_OPERANDS {
-            return Err(error(format!("The number of operands must be nonzero and <= {}", N::MAX_OPERANDS)));
+        let max_operands = match self.register_type() {
+            RegisterType::Plaintext(_) => N::MAX_DATA_ENTRIES,
+            // Note that if the register type is a record, then we must account for `owner` and `gates` which are not data entries.
+            RegisterType::Record(_) | RegisterType::ExternalRecord(_) => N::MAX_DATA_ENTRIES + 2,
+        };
+        if self.operands.len().is_zero() || self.operands.len() > max_operands {
+            return Err(error(format!("The number of operands must be nonzero and <= {}", max_operands)));
         }
 
         // Write the number of operands.
@@ -597,4 +624,74 @@ mod tests {
             "The value type is incorrect"
         );
     }
+
+    #[test]
+    fn test_parse_cast_into_plaintext_max_operands() {
+        let mut string = "cast ".to_string();
+        let mut operands = Vec::with_capacity(CurrentNetwork::MAX_DATA_ENTRIES);
+        for i in 0..CurrentNetwork::MAX_DATA_ENTRIES {
+            string.push_str(&format!("r{} ", i));
+            operands.push(Operand::Register(Register::Locator(i as u64)));
+        }
+        string.push_str(&format!("into r{} as foo", CurrentNetwork::MAX_DATA_ENTRIES));
+        let (string, cast) = Cast::<CurrentNetwork>::parse(&string).unwrap();
+        assert!(string.is_empty(), "Parser did not consume all of the string: '{string}'");
+        assert_eq!(cast.operands.len(), CurrentNetwork::MAX_DATA_ENTRIES, "The number of operands is incorrect");
+        assert_eq!(cast.operands, operands, "The operands are incorrect");
+        assert_eq!(
+            cast.destination,
+            Register::Locator(CurrentNetwork::MAX_DATA_ENTRIES as u64),
+            "The destination register is incorrect"
+        );
+        assert_eq!(
+            cast.register_type,
+            RegisterType::Plaintext(PlaintextType::Struct(Identifier::from_str("foo").unwrap())),
+            "The value type is incorrect"
+        );
+    }
+
+    #[test]
+    fn test_parse_cast_into_record_max_operands() {
+        let mut string = "cast ".to_string();
+        let mut operands = Vec::with_capacity(CurrentNetwork::MAX_DATA_ENTRIES + 2);
+        for i in 0..CurrentNetwork::MAX_DATA_ENTRIES + 2 {
+            string.push_str(&format!("r{} ", i));
+            operands.push(Operand::Register(Register::Locator(i as u64)));
+        }
+        string.push_str(&format!("into r{} as token.record", CurrentNetwork::MAX_DATA_ENTRIES + 2));
+        let (string, cast) = Cast::<CurrentNetwork>::parse(&string).unwrap();
+        assert!(string.is_empty(), "Parser did not consume all of the string: '{string}'");
+        assert_eq!(cast.operands.len(), CurrentNetwork::MAX_DATA_ENTRIES + 2, "The number of operands is incorrect");
+        assert_eq!(cast.operands, operands, "The operands are incorrect");
+        assert_eq!(
+            cast.destination,
+            Register::Locator((CurrentNetwork::MAX_DATA_ENTRIES + 2) as u64),
+            "The destination register is incorrect"
+        );
+        assert_eq!(
+            cast.register_type,
+            RegisterType::Record(Identifier::from_str("token").unwrap()),
+            "The value type is incorrect"
+        );
+    }
+
+    #[test]
+    fn test_parse_cast_into_record_too_many_operands() {
+        let mut string = "cast ".to_string();
+        for i in 0..=CurrentNetwork::MAX_DATA_ENTRIES + 2 {
+            string.push_str(&format!("r{} ", i));
+        }
+        string.push_str(&format!("into r{} as token.record", CurrentNetwork::MAX_DATA_ENTRIES + 3));
+        assert!(Cast::<CurrentNetwork>::parse(&string).is_err(), "Parser did not error");
+    }
+
+    #[test]
+    fn test_parse_cast_into_plaintext_too_many_operands() {
+        let mut string = "cast ".to_string();
+        for i in 0..=CurrentNetwork::MAX_DATA_ENTRIES {
+            string.push_str(&format!("r{} ", i));
+        }
+        string.push_str(&format!("into r{} as foo", CurrentNetwork::MAX_DATA_ENTRIES + 1));
+        assert!(Cast::<CurrentNetwork>::parse(&string).is_err(), "Parser did not error");
+    }
 }
diff --git a/synthesizer/src/program/mod.rs b/synthesizer/src/program/mod.rs
index 1b993ea51..886a31963 100644
--- a/synthesizer/src/program/mod.rs
+++ b/synthesizer/src/program/mod.rs
@@ -828,7 +828,7 @@ function swap:
         // Declare the function name.
         let function_name = Identifier::from_str("foo").unwrap();
         // Declare the function inputs.
-        let inputs = vec![
+        let inputs = [
             Value::<CurrentNetwork>::Plaintext(Plaintext::from_str("2field").unwrap()),
             Value::Plaintext(Plaintext::from_str("3field").unwrap()),
         ];
@@ -846,7 +846,7 @@ function swap:
 
             // Authorize the function call.
             let authorization = process
-                .authorize::<CurrentAleo, _>(&caller_private_key, program.id(), function_name, &inputs, rng)
+                .authorize::<CurrentAleo, _>(&caller_private_key, program.id(), function_name, inputs.iter(), rng)
                 .unwrap();
             assert_eq!(authorization.len(), 1);
             authorization
@@ -912,7 +912,7 @@ function compute:
 
             // Authorize the function call.
             let authorization = process
-                .authorize::<CurrentAleo, _>(&caller_private_key, program.id(), function_name, &[input], rng)
+                .authorize::<CurrentAleo, _>(&caller_private_key, program.id(), function_name, [input].iter(), rng)
                 .unwrap();
             assert_eq!(authorization.len(), 1);
             authorization
@@ -980,7 +980,7 @@ function compute:
 
         // Authorize the function call.
         let authorization = process
-            .authorize::<CurrentAleo, _>(&caller_private_key, program.id(), function_name, &[input], rng)
+            .authorize::<CurrentAleo, _>(&caller_private_key, program.id(), function_name, [input].iter(), rng)
             .unwrap();
         assert_eq!(authorization.len(), 1);
 
@@ -1066,7 +1066,7 @@ function compute:
                     &caller_private_key,
                     program.id(),
                     function_name,
-                    &[r0.clone(), r1.clone()],
+                    [r0.clone(), r1.clone()].iter(),
                     rng,
                 )
                 .unwrap();
@@ -1108,7 +1108,7 @@ function compute:
         let burner_private_key = PrivateKey::new(rng).unwrap();
         // Authorize the function call, with a burner private key.
         let authorization = process
-            .authorize::<CurrentAleo, _>(&burner_private_key, program.id(), function_name, &[r0, r1], rng)
+            .authorize::<CurrentAleo, _>(&burner_private_key, program.id(), function_name, [r0, r1].iter(), rng)
             .unwrap();
         assert_eq!(authorization.len(), 1);
 
@@ -1166,7 +1166,7 @@ function compute:
 
         // Authorize the function call.
         let authorization = process
-            .authorize::<CurrentAleo, _>(&caller_private_key, program.id(), function_name, &[input], rng)
+            .authorize::<CurrentAleo, _>(&caller_private_key, program.id(), function_name, [input].iter(), rng)
             .unwrap();
         assert_eq!(authorization.len(), 1);
         let request = authorization.peek_next().unwrap();
diff --git a/synthesizer/src/store/block/mod.rs b/synthesizer/src/store/block/mod.rs
index 13fbd493b..76ca6a453 100644
--- a/synthesizer/src/store/block/mod.rs
+++ b/synthesizer/src/store/block/mod.rs
@@ -29,11 +29,12 @@ use crate::{
         TransitionStorage,
         TransitionStore,
     },
+    Program,
 };
 use console::{
     account::Signature,
     network::prelude::*,
-    program::{BlockTree, HeaderLeaf, StatePath},
+    program::{BlockTree, HeaderLeaf, ProgramID, StatePath},
     types::Field,
 };
 
@@ -855,6 +856,11 @@ impl<N: Network, B: BlockStorage<N>> BlockStore<N, B> {
     pub fn get_block(&self, block_hash: &N::BlockHash) -> Result<Option<Block<N>>> {
         self.storage.get_block(block_hash)
     }
+
+    /// Returns the program for the given `program ID`.
+    pub fn get_program(&self, program_id: &ProgramID<N>) -> Result<Option<Program<N>>> {
+        self.storage.transaction_store().get_program(program_id)
+    }
 }
 
 impl<N: Network, B: BlockStorage<N>> BlockStore<N, B> {
diff --git a/synthesizer/src/vm/authorize.rs b/synthesizer/src/vm/authorize.rs
index ebdd699b3..650f89f00 100644
--- a/synthesizer/src/vm/authorize.rs
+++ b/synthesizer/src/vm/authorize.rs
@@ -22,11 +22,26 @@ impl<N: Network, C: ConsensusStorage<N>> VM<N, C> {
     pub fn authorize<R: Rng + CryptoRng>(
         &self,
         private_key: &PrivateKey<N>,
-        program_id: &ProgramID<N>,
-        function_name: Identifier<N>,
-        inputs: &[Value<N>],
+        program_id: impl TryInto<ProgramID<N>>,
+        function_name: impl TryInto<Identifier<N>>,
+        inputs: impl IntoIterator<IntoIter = impl ExactSizeIterator<Item = impl TryInto<Value<N>>>>,
         rng: &mut R,
     ) -> Result<Authorization<N>> {
+        // Prepare the program ID.
+        let program_id = program_id.try_into().map_err(|_| anyhow!("Invalid program ID"))?;
+        // Prepare the function name.
+        let function_name = function_name.try_into().map_err(|_| anyhow!("Invalid function name"))?;
+        // Prepare the inputs.
+        let inputs = inputs
+            .into_iter()
+            .enumerate()
+            .map(|(index, input)| {
+                input
+                    .try_into()
+                    .map_err(|_| anyhow!("Failed to parse input #{index} for '{program_id}/{function_name}'"))
+            })
+            .collect::<Result<Vec<_>>>()?;
+
         // Compute the core logic.
         macro_rules! logic {
             ($process:expr, $network:path, $aleo:path) => {{
@@ -34,13 +49,13 @@ impl<N: Network, C: ConsensusStorage<N>> VM<N, C> {
 
                 // Prepare the inputs.
                 let private_key = cast_ref!(&private_key as PrivateKey<$network>);
-                let program_id = cast_ref!(&program_id as ProgramID<$network>);
+                let program_id = cast_ref!(program_id as ProgramID<$network>);
                 let function_name = cast_ref!(function_name as Identifier<$network>);
                 let inputs = cast_ref!(inputs as Vec<Value<$network>>);
 
                 // Compute the authorization.
                 let authorization =
-                    $process.authorize::<$aleo, _>(private_key, program_id, function_name.clone(), inputs, rng)?;
+                    $process.authorize::<$aleo, _>(private_key, program_id, function_name, inputs.iter(), rng)?;
 
                 // Return the authorization.
                 Ok(cast_ref!(authorization as Authorization<N>).clone())
diff --git a/synthesizer/src/vm/execute.rs b/synthesizer/src/vm/execute.rs
index 5666b57e6..3c3b1ff5a 100644
--- a/synthesizer/src/vm/execute.rs
+++ b/synthesizer/src/vm/execute.rs
@@ -22,8 +22,15 @@ impl<N: Network, C: ConsensusStorage<N>> VM<N, C> {
     pub fn execute<R: Rng + CryptoRng>(
         &self,
         authorization: Authorization<N>,
+        query: Option<Query<N, C::BlockStorage>>,
         rng: &mut R,
     ) -> Result<(Response<N>, Execution<N>)> {
+        // Prepare the query.
+        let query = match query {
+            Some(query) => query,
+            None => Query::VM(self.block_store().clone()),
+        };
+
         // Compute the core logic.
         macro_rules! logic {
             ($process:expr, $network:path, $aleo:path) => {{
@@ -37,7 +44,7 @@ impl<N: Network, C: ConsensusStorage<N>> VM<N, C> {
                 let assignments = {
                     let execution = cast_ref!(execution as Execution<N>);
                     let inclusion = cast_ref!(inclusion as Inclusion<N>);
-                    inclusion.prepare_execution(execution, self.block_store())?
+                    inclusion.prepare_execution(execution, query)?
                 };
                 let assignments = cast_ref!(assignments as Vec<InclusionAssignment<$network>>);
 
@@ -62,8 +69,15 @@ impl<N: Network, C: ConsensusStorage<N>> VM<N, C> {
         private_key: &PrivateKey<N>,
         credits: Record<N, Plaintext<N>>,
         fee_in_gates: u64,
+        query: Option<Query<N, C::BlockStorage>>,
         rng: &mut R,
     ) -> Result<(Response<N>, Fee<N>)> {
+        // Prepare the query.
+        let query = match query {
+            Some(query) => query,
+            None => Query::VM(self.block_store().clone()),
+        };
+
         // Compute the core logic.
         macro_rules! logic {
             ($process:expr, $network:path, $aleo:path) => {{
@@ -81,7 +95,7 @@ impl<N: Network, C: ConsensusStorage<N>> VM<N, C> {
                 let assignments = {
                     let fee_transition = cast_ref!(fee_transition as Transition<N>);
                     let inclusion = cast_ref!(inclusion as Inclusion<N>);
-                    inclusion.prepare_fee(fee_transition, self.block_store())?
+                    inclusion.prepare_fee(fee_transition, query)?
                 };
                 let assignments = cast_ref!(assignments as Vec<InclusionAssignment<$network>>);
 
diff --git a/synthesizer/src/vm/mod.rs b/synthesizer/src/vm/mod.rs
index f76376e1b..adad2e725 100644
--- a/synthesizer/src/vm/mod.rs
+++ b/synthesizer/src/vm/mod.rs
@@ -27,7 +27,7 @@ use crate::{
     block::{Block, Transaction, Transactions, Transition},
     cast_ref,
     process,
-    process::{Authorization, Deployment, Execution, Fee, Inclusion, InclusionAssignment, Process},
+    process::{Authorization, Deployment, Execution, Fee, Inclusion, InclusionAssignment, Process, Query},
     program::Program,
     store::{BlockStore, ConsensusStorage, ConsensusStore, ProgramStore, TransactionStore, TransitionStore},
 };
@@ -52,12 +52,11 @@ impl<N: Network, C: ConsensusStorage<N>> VM<N, C> {
     /// Initializes the VM from storage.
     #[inline]
     pub fn from(store: ConsensusStore<N, C>) -> Result<Self> {
-        // Retrieve the transaction store.
-        let transaction_store = store.transaction_store();
-
         // Initialize a new process.
         let mut process = Process::load()?;
 
+        // Retrieve the transaction store.
+        let transaction_store = store.transaction_store();
         // Load the deployments from the store.
         for transaction_id in transaction_store.deployment_ids() {
             // Retrieve the deployment.
@@ -75,26 +74,7 @@ impl<N: Network, C: ConsensusStorage<N>> VM<N, C> {
     /// Returns `true` if a program with the given program ID exists.
     #[inline]
     pub fn contains_program(&self, program_id: &ProgramID<N>) -> bool {
-        // Compute the core logic.
-        macro_rules! logic {
-            ($process:expr, $network:path, $aleo:path) => {{
-                let task = || {
-                    // Prepare the program ID.
-                    let program_id = cast_ref!(&program_id as ProgramID<$network>);
-                    // Return `true` if the program ID exists.
-                    Ok($process.contains_program(program_id))
-                };
-                task()
-            }};
-        }
-        // Process the logic.
-        match process!(self, logic) {
-            Ok(contains_program) => contains_program,
-            Err(error) => {
-                warn!("Failed to check if program exists: {error}");
-                false
-            }
-        }
+        self.process.read().contains_program(program_id)
     }
 
     /// Adds the given block into the VM.
@@ -114,6 +94,12 @@ impl<N: Network, C: ConsensusStorage<N>> VM<N, C> {
         }
     }
 
+    /// Returns the process.
+    #[inline]
+    pub fn process(&self) -> Arc<RwLock<Process<N>>> {
+        self.process.clone()
+    }
+
     /// Returns the program store.
     #[inline]
     pub fn program_store(&self) -> &ProgramStore<N, C::ProgramStorage> {
@@ -166,7 +152,7 @@ pub(crate) mod test_helpers {
     use console::{
         account::{Address, ViewKey},
         network::Testnet3,
-        program::{Identifier, Value},
+        program::Value,
     };
 
     use indexmap::IndexMap;
@@ -275,7 +261,8 @@ function compute:
                 vm.add_next_block(&genesis).unwrap();
 
                 // Deploy.
-                let transaction = Transaction::deploy(&vm, &caller_private_key, &program, additional_fee, rng).unwrap();
+                let transaction =
+                    Transaction::deploy(&vm, &caller_private_key, &program, additional_fee, None, rng).unwrap();
                 // Verify.
                 assert!(vm.verify(&transaction));
                 // Return the transaction.
@@ -313,20 +300,21 @@ function compute:
                 let authorization = vm
                     .authorize(
                         &caller_private_key,
-                        &ProgramID::from_str("credits.aleo").unwrap(),
-                        Identifier::from_str("transfer").unwrap(),
-                        &[
+                        "credits.aleo",
+                        "transfer",
+                        [
                             Value::<CurrentNetwork>::Record(record),
                             Value::<CurrentNetwork>::from_str(&address.to_string()).unwrap(),
                             Value::<CurrentNetwork>::from_str("1u64").unwrap(),
-                        ],
+                        ]
+                        .into_iter(),
                         rng,
                     )
                     .unwrap();
                 assert_eq!(authorization.len(), 1);
 
                 // Execute.
-                let transaction = Transaction::execute_authorization(&vm, authorization, rng).unwrap();
+                let transaction = Transaction::execute_authorization(&vm, authorization, None, rng).unwrap();
                 // Verify.
                 assert!(vm.verify(&transaction));
                 // Return the transaction.
@@ -362,7 +350,7 @@ function compute:
                 vm.add_next_block(&genesis).unwrap();
 
                 // Execute.
-                let (_response, fee) = vm.execute_fee(&caller_private_key, record, 1u64, rng).unwrap();
+                let (_response, fee) = vm.execute_fee(&caller_private_key, record, 1u64, None, rng).unwrap();
                 // Verify.
                 Inclusion::verify_fee(&fee).unwrap();
                 // Return the fee.
diff --git a/vm/file/prover.rs b/vm/file/prover.rs
index 41b26bc8b..d9985bca7 100644
--- a/vm/file/prover.rs
+++ b/vm/file/prover.rs
@@ -241,7 +241,7 @@ function compute:
         process.synthesize_key::<CurrentAleo, _>(program.id(), &function_name, &mut TestRng::default()).unwrap();
 
         // Retrieve the proving key.
-        let proving_key = process.get_proving_key(program.id(), &function_name).unwrap();
+        let proving_key = process.get_proving_key(program.id(), function_name).unwrap();
 
         // Create the prover file at the path.
         let expected = ProverFile::create(&directory, &function_name, proving_key).unwrap();
diff --git a/vm/file/verifier.rs b/vm/file/verifier.rs
index efcae85fe..6b66164fe 100644
--- a/vm/file/verifier.rs
+++ b/vm/file/verifier.rs
@@ -241,7 +241,7 @@ function compute:
         process.synthesize_key::<CurrentAleo, _>(program.id(), &function_name, &mut TestRng::default()).unwrap();
 
         // Retrieve the verifying key.
-        let verifying_key = process.get_verifying_key(program.id(), &function_name).unwrap();
+        let verifying_key = process.get_verifying_key(program.id(), function_name).unwrap();
 
         // Create the verifier file at the path.
         let expected = VerifierFile::create(&directory, &function_name, verifying_key).unwrap();
diff --git a/vm/lib.rs b/vm/lib.rs
index 55ef0456e..8818c3966 100644
--- a/vm/lib.rs
+++ b/vm/lib.rs
@@ -37,8 +37,6 @@ pub use snarkvm_console as console;
 pub use snarkvm_curves as curves;
 #[cfg(feature = "fields")]
 pub use snarkvm_fields as fields;
-#[cfg(feature = "gadgets")]
-pub use snarkvm_gadgets as gadgets;
 #[cfg(feature = "parameters")]
 pub use snarkvm_parameters as parameters;
 #[cfg(feature = "r1cs")]
@@ -47,6 +45,8 @@ pub use snarkvm_r1cs as r1cs;
 pub use snarkvm_synthesizer as synthesizer;
 #[cfg(feature = "utilities")]
 pub use snarkvm_utilities as utilities;
+#[cfg(feature = "wasm")]
+pub use snarkvm_wasm as wasm;
 
 pub mod errors {
     #[cfg(feature = "algorithms")]
diff --git a/vm/package/run.rs b/vm/package/run.rs
index 110083ad3..67ccf4435 100644
--- a/vm/package/run.rs
+++ b/vm/package/run.rs
@@ -47,7 +47,7 @@ impl<N: Network> Package<N> {
         // Construct the process.
         let process = self.get_process()?;
         // Authorize the function call.
-        let authorization = process.authorize::<A, R>(private_key, program_id, function_name, inputs, rng)?;
+        let authorization = process.authorize::<A, R>(private_key, program_id, function_name, inputs.iter(), rng)?;
 
         // Retrieve the program.
         let program = process.get_program(program_id)?;
