diff --git a/vm/compiler/src/ledger/get.rs b/vm/compiler/src/ledger/get.rs
index 8f51e577a..c3fc2e4cf 100644
--- a/vm/compiler/src/ledger/get.rs
+++ b/vm/compiler/src/ledger/get.rs
@@ -78,7 +78,7 @@ impl<N: Network, B: BlockStorage<N>, P: ProgramStorage<N>> Ledger<N, B, P> {
     /// Returns the transaction for the given transaction id.
     pub fn get_transaction(&self, transaction_id: N::TransactionID) -> Result<Transaction<N>> {
         // Retrieve the transaction.
-        match self.transactions.get_transaction(transaction_id)? {
+        match self.transactions.get_transaction(&transaction_id)? {
             Some(transaction) => Ok(transaction),
             None => bail!("Missing transaction for id {transaction_id}"),
         }
diff --git a/vm/package/deploy.rs b/vm/package/deploy.rs
index 46738c52f..b75e353d7 100644
--- a/vm/package/deploy.rs
+++ b/vm/package/deploy.rs
@@ -30,8 +30,9 @@ impl<N: Network> TryFrom<Deployment<N>> for DeployRequest<N> {
     type Error = anyhow::Error;
 
     fn try_from(deployment: Deployment<N>) -> Result<Self, Self::Error> {
-        let program_id = deployment.program_id().clone();
+        let program_id = *deployment.program_id();
         let address = program_id.to_address()?;
+
         Ok(Self { deployment, address, program_id })
     }
 }
