diff --git a/.resources/release-version b/.resources/release-version
index 95bcd3811..1ddc0f604 100644
--- a/.resources/release-version
+++ b/.resources/release-version
@@ -1 +1 @@
-v4.2.2
\ No newline at end of file
+v4.3.0
\ No newline at end of file
diff --git a/Cargo.lock b/Cargo.lock
index 7ca61f79c..a6c052711 100644
--- a/Cargo.lock
+++ b/Cargo.lock
@@ -587,9 +587,9 @@ dependencies = [
 
 [[package]]
 name = "cc"
-version = "1.2.42"
+version = "1.2.43"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "81bbf3b3619004ad9bd139f62a9ab5cfe467f307455a0d307b0cf58bf070feaa"
+checksum = "739eb0f94557554b3ca9a86d2d37bebd49c5e6d0c1d2bda35ba5bdac830befc2"
 dependencies = [
  "find-msvc-tools",
  "jobserver",
@@ -1132,9 +1132,9 @@ dependencies = [
 
 [[package]]
 name = "deranged"
-version = "0.5.4"
+version = "0.5.5"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "a41953f86f8a05768a6cda24def994fd2f424b04ec5c719cf89989779f199071"
+checksum = "ececcb659e7ba858fb4f10388c250a7252eb0a27373f1a72b8748afdd248e587"
 dependencies = [
  "powerfmt",
  "serde_core",
@@ -2325,9 +2325,9 @@ dependencies = [
 
 [[package]]
 name = "js-sys"
-version = "0.3.81"
+version = "0.3.82"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "ec48937a97411dcb524a265206ccd4c90bb711fca92b2792c407f268825b9305"
+checksum = "b011eec8cc36da2aab2d5cff675ec18454fad408585853910a202391cf9f8e65"
 dependencies = [
  "once_cell",
  "wasm-bindgen",
@@ -2639,14 +2639,14 @@ dependencies = [
 
 [[package]]
 name = "mio"
-version = "1.0.4"
+version = "1.1.0"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "78bed444cc8a2160f01cbcf811ef18cac863ad68ae8ca62092e8db51d51c761c"
+checksum = "69d83b0086dc8ecf3ce9ae2874b2d1290252e2a30720bea58a5c6639b0092873"
 dependencies = [
  "libc",
  "log",
  "wasi",
- "windows-sys 0.59.0",
+ "windows-sys 0.61.2",
 ]
 
 [[package]]
@@ -3645,9 +3645,9 @@ dependencies = [
 
 [[package]]
 name = "rustls-pki-types"
-version = "1.12.0"
+version = "1.13.0"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "229a4a4c221013e7e1f1a043678c5cc39fe5171437c88fb47151a21e6f5b5c79"
+checksum = "94182ad936a0c91c324cd46c6511b9510ed16af436d7b5bab34beab0afd55f7a"
 dependencies = [
  "web-time",
  "zeroize",
@@ -3984,9 +3984,9 @@ dependencies = [
 
 [[package]]
 name = "signal-hook-mio"
-version = "0.2.4"
+version = "0.2.5"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "34db1a06d485c9142248b7a054f034b349b212551f3dfd19c94d45a754a217cd"
+checksum = "b75a19a7a740b25bc7944bdee6172368f988763b744e3d4dfe753f6b4ece40cc"
 dependencies = [
  "libc",
  "mio",
@@ -4068,7 +4068,7 @@ dependencies = [
 
 [[package]]
 name = "snarkos"
-version = "4.2.2"
+version = "4.3.0"
 dependencies = [
  "built",
  "clap",
@@ -4094,7 +4094,7 @@ dependencies = [
 
 [[package]]
 name = "snarkos-account"
-version = "4.2.2"
+version = "4.3.0"
 dependencies = [
  "colored 3.0.0",
  "snarkvm",
@@ -4102,7 +4102,7 @@ dependencies = [
 
 [[package]]
 name = "snarkos-cli"
-version = "4.2.2"
+version = "4.3.0"
 dependencies = [
  "aleo-std",
  "anstyle",
@@ -4144,7 +4144,7 @@ dependencies = [
 
 [[package]]
 name = "snarkos-display"
-version = "4.2.2"
+version = "4.3.0"
 dependencies = [
  "anyhow",
  "crossterm 0.29.0",
@@ -4156,7 +4156,7 @@ dependencies = [
 
 [[package]]
 name = "snarkos-node"
-version = "4.2.2"
+version = "4.3.0"
 dependencies = [
  "aleo-std",
  "anyhow",
@@ -4198,7 +4198,7 @@ dependencies = [
 
 [[package]]
 name = "snarkos-node-bft"
-version = "4.2.2"
+version = "4.3.0"
 dependencies = [
  "aleo-std",
  "anyhow",
@@ -4249,7 +4249,7 @@ dependencies = [
 
 [[package]]
 name = "snarkos-node-bft-events"
-version = "4.2.2"
+version = "4.3.0"
 dependencies = [
  "anyhow",
  "bytes",
@@ -4267,7 +4267,7 @@ dependencies = [
 
 [[package]]
 name = "snarkos-node-bft-ledger-service"
-version = "4.2.2"
+version = "4.3.0"
 dependencies = [
  "anyhow",
  "async-trait",
@@ -4284,7 +4284,7 @@ dependencies = [
 
 [[package]]
 name = "snarkos-node-bft-storage-service"
-version = "4.2.2"
+version = "4.3.0"
 dependencies = [
  "aleo-std",
  "anyhow",
@@ -4298,7 +4298,7 @@ dependencies = [
 
 [[package]]
 name = "snarkos-node-cdn"
-version = "4.2.2"
+version = "4.3.0"
 dependencies = [
  "anyhow",
  "bincode",
@@ -4319,7 +4319,7 @@ dependencies = [
 
 [[package]]
 name = "snarkos-node-consensus"
-version = "4.2.2"
+version = "4.3.0"
 dependencies = [
  "aleo-std",
  "anyhow",
@@ -4344,7 +4344,7 @@ dependencies = [
 
 [[package]]
 name = "snarkos-node-metrics"
-version = "4.2.2"
+version = "4.3.0"
 dependencies = [
  "locktick",
  "metrics-exporter-prometheus",
@@ -4373,7 +4373,7 @@ dependencies = [
 
 [[package]]
 name = "snarkos-node-rest"
-version = "4.2.2"
+version = "4.3.0"
 dependencies = [
  "anyhow",
  "axum 0.8.6",
@@ -4407,7 +4407,7 @@ dependencies = [
 
 [[package]]
 name = "snarkos-node-router"
-version = "4.2.2"
+version = "4.3.0"
 dependencies = [
  "aleo-std",
  "anyhow",
@@ -4443,7 +4443,7 @@ dependencies = [
 
 [[package]]
 name = "snarkos-node-router-messages"
-version = "4.2.2"
+version = "4.3.0"
 dependencies = [
  "bytes",
  "proptest",
@@ -4458,7 +4458,7 @@ dependencies = [
 
 [[package]]
 name = "snarkos-node-sync"
-version = "4.2.2"
+version = "4.3.0"
 dependencies = [
  "anyhow",
  "futures",
@@ -4483,7 +4483,7 @@ dependencies = [
 
 [[package]]
 name = "snarkos-node-sync-communication-service"
-version = "4.2.2"
+version = "4.3.0"
 dependencies = [
  "async-trait",
  "tokio",
@@ -4491,7 +4491,7 @@ dependencies = [
 
 [[package]]
 name = "snarkos-node-sync-locators"
-version = "4.2.2"
+version = "4.3.0"
 dependencies = [
  "anyhow",
  "indexmap 2.12.0",
@@ -4502,7 +4502,7 @@ dependencies = [
 
 [[package]]
 name = "snarkos-node-tcp"
-version = "4.2.2"
+version = "4.3.0"
 dependencies = [
  "async-trait",
  "bytes",
@@ -6394,9 +6394,9 @@ dependencies = [
 
 [[package]]
 name = "wasm-bindgen"
-version = "0.2.104"
+version = "0.2.105"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "c1da10c01ae9f1ae40cbfac0bac3b1e724b320abfcf52229f80b547c0d250e2d"
+checksum = "da95793dfc411fbbd93f5be7715b0578ec61fe87cb1a42b12eb625caa5c5ea60"
 dependencies = [
  "cfg-if",
  "once_cell",
@@ -6405,25 +6405,11 @@ dependencies = [
  "wasm-bindgen-shared",
 ]
 
-[[package]]
-name = "wasm-bindgen-backend"
-version = "0.2.104"
-source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "671c9a5a66f49d8a47345ab942e2cb93c7d1d0339065d4f8139c486121b43b19"
-dependencies = [
- "bumpalo",
- "log",
- "proc-macro2",
- "quote 1.0.41",
- "syn 2.0.108",
- "wasm-bindgen-shared",
-]
-
 [[package]]
 name = "wasm-bindgen-futures"
-version = "0.4.54"
+version = "0.4.55"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "7e038d41e478cc73bae0ff9b36c60cff1c98b8f38f8d7e8061e79ee63608ac5c"
+checksum = "551f88106c6d5e7ccc7cd9a16f312dd3b5d36ea8b4954304657d5dfba115d4a0"
 dependencies = [
  "cfg-if",
  "js-sys",
@@ -6434,9 +6420,9 @@ dependencies = [
 
 [[package]]
 name = "wasm-bindgen-macro"
-version = "0.2.104"
+version = "0.2.105"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "7ca60477e4c59f5f2986c50191cd972e3a50d8a95603bc9434501cf156a9a119"
+checksum = "04264334509e04a7bf8690f2384ef5265f05143a4bff3889ab7a3269adab59c2"
 dependencies = [
  "quote 1.0.41",
  "wasm-bindgen-macro-support",
@@ -6444,31 +6430,31 @@ dependencies = [
 
 [[package]]
 name = "wasm-bindgen-macro-support"
-version = "0.2.104"
+version = "0.2.105"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "9f07d2f20d4da7b26400c9f4a0511e6e0345b040694e8a75bd41d578fa4421d7"
+checksum = "420bc339d9f322e562942d52e115d57e950d12d88983a14c79b86859ee6c7ebc"
 dependencies = [
+ "bumpalo",
  "proc-macro2",
  "quote 1.0.41",
  "syn 2.0.108",
- "wasm-bindgen-backend",
  "wasm-bindgen-shared",
 ]
 
 [[package]]
 name = "wasm-bindgen-shared"
-version = "0.2.104"
+version = "0.2.105"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "bad67dc8b2a1a6e5448428adec4c3e84c43e561d8c9ee8a9e5aabeb193ec41d1"
+checksum = "76f218a38c84bcb33c25ec7059b07847d465ce0e0a76b995e134a45adcb6af76"
 dependencies = [
  "unicode-ident",
 ]
 
 [[package]]
 name = "web-sys"
-version = "0.3.81"
+version = "0.3.82"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "9367c417a924a74cae129e6a2ae3b47fabb1f8995595ab474029da749a8be120"
+checksum = "3a1f95c0d03a47f4ae1f7a64643a6bb97465d9b740f0fa8f90ea33915c99a9a1"
 dependencies = [
  "js-sys",
  "wasm-bindgen",
diff --git a/Cargo.toml b/Cargo.toml
index 574b2780d..7ded58938 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkos"
-version = "4.2.2"
+version = "4.3.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "A decentralized operating system"
 homepage = "https://aleo.org"
@@ -48,7 +48,7 @@ default-features = false
 #path = "../snarkVM"
 git = "https://github.com/ProvableHQ/snarkVM.git"
 rev = "31fcaaca"
-#version = "=4.2.1"
+#version = "=4.3.0"
 default-features = false
 
 [workspace.dependencies.anyhow]
@@ -171,48 +171,48 @@ version = "0.4"
 
 [workspace.dependencies.snarkos-account]
 path = "account"
-version = "=4.2.2"
+version = "=4.3.0"
 
 [workspace.dependencies.snarkos-cli]
 path = "cli"
-version = "=4.2.2"
+version = "=4.3.0"
 
 [workspace.dependencies.snarkos-display]
 path = "display"
-version = "=4.2.2"
+version = "=4.3.0"
 
 [workspace.dependencies.snarkos-node]
 path = "node"
-version = "=4.2.2"
+version = "=4.3.0"
 
 [workspace.dependencies.snarkos-node-bft]
 path = "node/bft"
-version = "=4.2.2"
+version = "=4.3.0"
 
 [workspace.dependencies.snarkos-node-bft-events]
 path = "node/bft/events"
-version = "=4.2.2"
+version = "=4.3.0"
 
 [workspace.dependencies.snarkos-node-bft-storage-service]
 path = "node/bft/storage-service"
-version = "=4.2.2"
+version = "=4.3.0"
 
 [workspace.dependencies.snarkos-node-bft-ledger-service]
 path = "node/bft/ledger-service"
-version = "=4.2.2"
+version = "=4.3.0"
 default-features = false
 
 [workspace.dependencies.snarkos-node-consensus]
 path = "node/consensus"
-version = "=4.2.2"
+version = "=4.3.0"
 
 [workspace.dependencies.snarkos-node-cdn]
 path = "node/cdn"
-version = "=4.2.2"
+version = "=4.3.0"
 
 [workspace.dependencies.snarkos-node-metrics]
 path = "node/metrics"
-version = "=4.2.2"
+version = "=4.3.0"
 
 [workspace.dependencies.snarkos-node-network]
 path = "node/network"
@@ -220,31 +220,31 @@ version = "=4.2.2"
 
 [workspace.dependencies.snarkos-node-router]
 path = "node/router"
-version = "=4.2.2"
+version = "=4.3.0"
 
 [workspace.dependencies.snarkos-node-router-messages]
 path = "node/router/messages"
-version = "=4.2.2"
+version = "=4.3.0"
 
 [workspace.dependencies.snarkos-node-rest]
 path = "node/rest"
-version = "=4.2.2"
+version = "=4.3.0"
 
 [workspace.dependencies.snarkos-node-sync]
 path = "node/sync"
-version = "=4.2.2"
+version = "=4.3.0"
 
 [workspace.dependencies.snarkos-node-sync-communication-service]
 path = "node/sync/communication-service"
-version = "=4.2.2"
+version = "=4.3.0"
 
 [workspace.dependencies.snarkos-node-sync-locators]
 path = "node/sync/locators"
-version = "=4.2.2"
+version = "=4.3.0"
 
 [workspace.dependencies.snarkos-node-tcp]
 path = "node/tcp"
-version = "=4.2.2"
+version = "=4.3.0"
 
 [[bin]]
 name = "snarkos"
diff --git a/account/Cargo.toml b/account/Cargo.toml
index d741c2c2f..599699ee9 100644
--- a/account/Cargo.toml
+++ b/account/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkos-account"
-version = "4.2.2"
+version = "4.3.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Account for a decentralized operating system"
 homepage = "https://aleo.org"
diff --git a/cli/Cargo.toml b/cli/Cargo.toml
index 11fa0403a..b4003a039 100644
--- a/cli/Cargo.toml
+++ b/cli/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkos-cli"
-version = "4.2.2"
+version = "4.3.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Command-line interface for snarkOS"
 homepage = "https://aleo.org"
diff --git a/display/Cargo.toml b/display/Cargo.toml
index 7d94718ff..1a3a6ce28 100644
--- a/display/Cargo.toml
+++ b/display/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkos-display"
-version = "4.2.2"
+version = "4.3.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "A display for a decentralized operating system"
 homepage = "https://aleo.org"
diff --git a/node/Cargo.toml b/node/Cargo.toml
index 8f33d515a..d2fcb411b 100644
--- a/node/Cargo.toml
+++ b/node/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkos-node"
-version = "4.2.2"
+version = "4.3.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "A node for a decentralized operating system"
 homepage = "https://aleo.org"
diff --git a/node/bft/Cargo.toml b/node/bft/Cargo.toml
index 46380f181..25e79280a 100644
--- a/node/bft/Cargo.toml
+++ b/node/bft/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkos-node-bft"
-version = "4.2.2"
+version = "4.3.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "A memory pool for a decentralized operating system"
 homepage = "https://aleo.org"
diff --git a/node/bft/events/Cargo.toml b/node/bft/events/Cargo.toml
index 014be41c9..c6a679318 100644
--- a/node/bft/events/Cargo.toml
+++ b/node/bft/events/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkos-node-bft-events"
-version = "4.2.2"
+version = "4.3.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Events for the gateway in a decentralized operating system"
 homepage = "https://aleo.org"
diff --git a/node/bft/ledger-service/Cargo.toml b/node/bft/ledger-service/Cargo.toml
index d468168c5..c8d65fa88 100644
--- a/node/bft/ledger-service/Cargo.toml
+++ b/node/bft/ledger-service/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkos-node-bft-ledger-service"
-version = "4.2.2"
+version = "4.3.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "A ledger service for the memory pool in a decentralized operating system"
 homepage = "https://aleo.org"
diff --git a/node/bft/storage-service/Cargo.toml b/node/bft/storage-service/Cargo.toml
index 059ee0532..acc87b01c 100644
--- a/node/bft/storage-service/Cargo.toml
+++ b/node/bft/storage-service/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkos-node-bft-storage-service"
-version = "4.2.2"
+version = "4.3.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "A storage service for the memory pool in a decentralized operating system"
 homepage = "https://aleo.org"
diff --git a/node/cdn/Cargo.toml b/node/cdn/Cargo.toml
index e62673010..3387d31d4 100644
--- a/node/cdn/Cargo.toml
+++ b/node/cdn/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkos-node-cdn"
-version = "4.2.2"
+version = "4.3.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "A CDN client for a decentralized operating system"
 homepage = "https://aleo.org"
diff --git a/node/consensus/Cargo.toml b/node/consensus/Cargo.toml
index 88c1e1c36..c24514966 100644
--- a/node/consensus/Cargo.toml
+++ b/node/consensus/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkos-node-consensus"
-version = "4.2.2"
+version = "4.3.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "A node consensus for a decentralized operating system"
 homepage = "https://aleo.org"
diff --git a/node/metrics/Cargo.toml b/node/metrics/Cargo.toml
index 7b61d58ce..fa3ac94af 100644
--- a/node/metrics/Cargo.toml
+++ b/node/metrics/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkos-node-metrics"
-version = "4.2.2"
+version = "4.3.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "A node for a decentralized operating system"
 homepage = "https://aleo.org"
diff --git a/node/rest/Cargo.toml b/node/rest/Cargo.toml
index 372eaefc6..d740ab56a 100644
--- a/node/rest/Cargo.toml
+++ b/node/rest/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkos-node-rest"
-version = "4.2.2"
+version = "4.3.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "A REST API server for a decentralized virtual machine"
 homepage = "https://aleo.org"
diff --git a/node/rest/src/routes.rs b/node/rest/src/routes.rs
index b87c5ffff..813e19e63 100644
--- a/node/rest/src/routes.rs
+++ b/node/rest/src/routes.rs
@@ -139,17 +139,23 @@ impl<N: Network, C: ConsensusStorage<N>, R: Routing<N>> Rest<N, C, R> {
     ) -> Result<ErasedJson, RestError> {
         // Manually parse the height or the height of the hash, axum doesn't support different types
         // for the same path param.
+        let id_name;
         let block = if let Ok(height) = height_or_hash.parse::<u32>() {
-            rest.ledger.get_block(height).with_context(|| "Failed to get block by height")?
+            id_name = "hash";
+            rest.ledger.try_get_block(height).with_context(|| "Failed to get block by height")?
         } else if let Ok(hash) = height_or_hash.parse::<N::BlockHash>() {
-            rest.ledger.get_block_by_hash(&hash).with_context(|| "Failed to get block by hash")?
+            id_name = "height";
+            rest.ledger.try_get_block_by_hash(&hash).with_context(|| "Failed to get block by hash")?
         } else {
             return Err(RestError::bad_request(anyhow!(
                 "invalid input, it is neither a block height nor a block hash"
             )));
         };
 
-        Ok(ErasedJson::pretty(block))
+        match block {
+            Some(block) => Ok(ErasedJson::pretty(block)),
+            None => Err(RestError::not_found(anyhow!("No block with {id_name} {height_or_hash} found"))),
+        }
     }
 
     /// GET /<network>/blocks?start={start_height}&end={end_height}
diff --git a/node/router/Cargo.toml b/node/router/Cargo.toml
index 171750882..b85d4d38d 100644
--- a/node/router/Cargo.toml
+++ b/node/router/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkos-node-router"
-version = "4.2.2"
+version = "4.3.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "A node router for a decentralized operating system"
 homepage = "https://aleo.org"
diff --git a/node/router/messages/Cargo.toml b/node/router/messages/Cargo.toml
index d02424d8d..32604f7a0 100644
--- a/node/router/messages/Cargo.toml
+++ b/node/router/messages/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkos-node-router-messages"
-version = "4.2.2"
+version = "4.3.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Node messages for a decentralized operating system"
 homepage = "https://aleo.org"
diff --git a/node/sync/Cargo.toml b/node/sync/Cargo.toml
index c482bdd74..a2a95ec25 100644
--- a/node/sync/Cargo.toml
+++ b/node/sync/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkos-node-sync"
-version = "4.2.2"
+version = "4.3.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "A synchronization module for a decentralized operating system"
 homepage = "https://aleo.org"
diff --git a/node/sync/communication-service/Cargo.toml b/node/sync/communication-service/Cargo.toml
index 92de4ae18..8ebc455d5 100644
--- a/node/sync/communication-service/Cargo.toml
+++ b/node/sync/communication-service/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkos-node-sync-communication-service"
-version = "4.2.2"
+version = "4.3.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "A communication service for the sync module in a decentralized operating system"
 homepage = "https://aleo.org"
diff --git a/node/sync/locators/Cargo.toml b/node/sync/locators/Cargo.toml
index 93cb52e9b..f5f62cd64 100644
--- a/node/sync/locators/Cargo.toml
+++ b/node/sync/locators/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkos-node-sync-locators"
-version = "4.2.2"
+version = "4.3.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Locators to synchronize a decentralized operating system"
 homepage = "https://aleo.org"
diff --git a/node/tcp/Cargo.toml b/node/tcp/Cargo.toml
index 833407a12..97d7d1723 100644
--- a/node/tcp/Cargo.toml
+++ b/node/tcp/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkos-node-tcp"
-version = "4.2.2"
+version = "4.3.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "A TCP stack for a decentralized operating system"
 homepage = "https://aleo.org"
