diff --git a/ledger/block/src/verify.rs b/ledger/block/src/verify.rs
index 2d9ab54a5..87144f032 100644
--- a/ledger/block/src/verify.rs
+++ b/ledger/block/src/verify.rs
@@ -35,13 +35,18 @@ impl<N: Network> Block<N> {
         current_epoch_challenge: &EpochChallenge<N>,
         current_timestamp: i64,
         ratified_finalize_operations: Vec<FinalizeOperation<N>>,
-    ) -> Result<Vec<N::TransactionID>> {
+    ) -> Result<(Vec<PuzzleCommitment<N>>, Vec<N::TransactionID>)> {
         // Ensure the block hash is correct.
         self.verify_hash(previous_block.height(), previous_block.hash())?;
 
         // Ensure the block authority is correct.
-        let (expected_round, expected_height, expected_timestamp, expected_existing_transaction_ids) =
-            self.verify_authority(previous_block.round(), previous_block.height(), current_committee)?;
+        let (
+            expected_round,
+            expected_height,
+            expected_timestamp,
+            expected_existing_solution_ids,
+            expected_existing_transaction_ids,
+        ) = self.verify_authority(previous_block.round(), previous_block.height(), current_committee)?;
 
         // Ensure the block solutions are correct.
         let (
@@ -94,8 +99,8 @@ impl<N: Network> Block<N> {
             current_timestamp,
         )?;
 
-        // Return the expected existing transaction ids.
-        Ok(expected_existing_transaction_ids)
+        // Return the expected existing solution IDs and transaction IDs.
+        Ok((expected_existing_solution_ids, expected_existing_transaction_ids))
     }
 }
 
@@ -139,7 +144,7 @@ impl<N: Network> Block<N> {
         previous_round: u64,
         previous_height: u32,
         current_committee: &Committee<N>,
-    ) -> Result<(u64, u32, i64, Vec<N::TransactionID>)> {
+    ) -> Result<(u64, u32, i64, Vec<PuzzleCommitment<N>>, Vec<N::TransactionID>)> {
         // Note: Do not remove this. This ensures that all blocks after genesis are quorum blocks.
         #[cfg(not(any(test, feature = "test")))]
         ensure!(self.authority.is_quorum(), "The next block must be a quorum block");
@@ -174,8 +179,8 @@ impl<N: Network> Block<N> {
         );
 
         // Ensure the block authority is correct.
-        // Determine the transaction IDs expected to be in previous blocks.
-        let expected_existing_transaction_ids = match &self.authority {
+        // Determine the solution IDs and transaction IDs that are expected to be in previous blocks.
+        let (expected_existing_solution_ids, expected_existing_transaction_ids) = match &self.authority {
             Authority::Beacon(signature) => {
                 // Retrieve the signer.
                 let signer = signature.to_address();
@@ -190,7 +195,7 @@ impl<N: Network> Block<N> {
                     "Signature is invalid in block {expected_height}"
                 );
 
-                vec![]
+                (vec![], vec![])
             }
             Authority::Quorum(subdag) => {
                 // Compute the expected leader.
@@ -221,7 +226,13 @@ impl<N: Network> Block<N> {
         };
 
         // Return success.
-        Ok((expected_round, expected_height, expected_timestamp, expected_existing_transaction_ids))
+        Ok((
+            expected_round,
+            expected_height,
+            expected_timestamp,
+            expected_existing_solution_ids,
+            expected_existing_transaction_ids,
+        ))
     }
 
     /// Ensures the block ratifications are correct.
@@ -523,13 +534,14 @@ impl<N: Network> Block<N> {
     }
 
     /// Checks that the transmission IDs in the given subdag matches the solutions and transactions in the block.
+    /// Returns the IDs of the transactions and solutions that should already exist in the ledger.
     pub(super) fn check_subdag_transmissions(
         subdag: &Subdag<N>,
         solutions: &Option<CoinbaseSolution<N>>,
         aborted_solution_ids: &[PuzzleCommitment<N>],
         transactions: &Transactions<N>,
         aborted_transaction_ids: &[N::TransactionID],
-    ) -> Result<Vec<N::TransactionID>> {
+    ) -> Result<(Vec<PuzzleCommitment<N>>, Vec<N::TransactionID>)> {
         // Prepare an iterator over the solution IDs.
         let mut solutions = solutions.as_ref().map(|s| s.deref()).into_iter().flatten().peekable();
         // Prepare an iterator over the unconfirmed transaction IDs.
@@ -565,7 +577,7 @@ impl<N: Network> Block<N> {
                         }
                         // Otherwise, add the solution ID to the aborted or existing list.
                         _ => {
-                            if !aborted_or_existing_solution_ids.insert(solution_id) {
+                            if !aborted_or_existing_solution_ids.insert(*solution_id) {
                                 bail!("Block contains a duplicate aborted solution ID (found '{solution_id}')");
                             }
                         }
@@ -594,11 +606,10 @@ impl<N: Network> Block<N> {
         // Ensure there are no more transactions in the block.
         ensure!(unconfirmed_transaction_ids.next().is_none(), "There exists more transactions than expected.");
 
-        // TODO: Move this check to be outside of this method, and check against the ledger for existence.
         // Ensure the aborted solution IDs match.
         for aborted_solution_id in aborted_solution_ids {
             // If the aborted transaction ID is not found, throw an error.
-            if !aborted_or_existing_solution_ids.contains(&aborted_solution_id) {
+            if !aborted_or_existing_solution_ids.contains(aborted_solution_id) {
                 bail!(
                     "Block contains an aborted solution ID that is not found in the subdag (found '{aborted_solution_id}')"
                 );
@@ -614,13 +625,17 @@ impl<N: Network> Block<N> {
             }
         }
 
-        // Retrieve the transaction ids that should already exist in the ledger.
+        // Retrieve the solution IDs that should already exist in the ledger.
+        let existing_solution_ids: Vec<_> = aborted_or_existing_solution_ids
+            .difference(&aborted_solution_ids.iter().copied().collect())
+            .copied()
+            .collect();
+        // Retrieve the transaction IDs that should already exist in the ledger.
         let existing_transaction_ids: Vec<_> = aborted_or_existing_transaction_ids
-            .iter()
-            .filter(|id| !aborted_transaction_ids.contains(id))
+            .difference(&aborted_transaction_ids.iter().copied().collect())
             .copied()
             .collect();
 
-        Ok(existing_transaction_ids)
+        Ok((existing_solution_ids, existing_transaction_ids))
     }
 }
diff --git a/ledger/narwhal/batch-header/src/lib.rs b/ledger/narwhal/batch-header/src/lib.rs
index f52e4d787..e7a8de8c7 100644
--- a/ledger/narwhal/batch-header/src/lib.rs
+++ b/ledger/narwhal/batch-header/src/lib.rs
@@ -96,7 +96,7 @@ impl<N: Network> BatchHeader<N> {
         // Ensure that the number of transmissions is within bounds.
         ensure!(
             transmission_ids.len() <= Self::MAX_TRANSMISSIONS_PER_BATCH,
-            "Invalid number of transmission ids ({})",
+            "Invalid number of transmission IDs ({})",
             transmission_ids.len()
         );
         // Ensure that the number of previous certificate IDs is within bounds.
@@ -165,7 +165,7 @@ impl<N: Network> BatchHeader<N> {
         // Ensure that the number of transmissions is within bounds.
         ensure!(
             transmission_ids.len() <= Self::MAX_TRANSMISSIONS_PER_BATCH,
-            "Invalid number of transmission ids ({})",
+            "Invalid number of transmission IDs ({})",
             transmission_ids.len()
         );
         // Ensure that the number of previous certificate IDs is within bounds.
diff --git a/ledger/src/check_next_block.rs b/ledger/src/check_next_block.rs
index b99b3c8ff..8203ccdaa 100644
--- a/ledger/src/check_next_block.rs
+++ b/ledger/src/check_next_block.rs
@@ -83,7 +83,7 @@ impl<N: Network, C: ConsensusStorage<N>> Ledger<N, C> {
             self.vm.check_speculate(state, block.ratifications(), block.solutions(), block.transactions())?;
 
         // Ensure the block is correct.
-        let expected_existing_transaction_ids = block.verify(
+        let (expected_existing_solution_ids, expected_existing_transaction_ids) = block.verify(
             &self.latest_block(),
             self.latest_state_root(),
             &self.latest_committee()?,
@@ -93,7 +93,14 @@ impl<N: Network, C: ConsensusStorage<N>> Ledger<N, C> {
             ratified_finalize_operations,
         )?;
 
-        // Ensure that each existing transaction id from the block exists in the ledger.
+        // Ensure that each existing solution ID from the block exists in the ledger.
+        for existing_solution_id in expected_existing_solution_ids {
+            if !self.contains_puzzle_commitment(&existing_solution_id)? {
+                bail!("Solution ID '{existing_solution_id}' does not exist in the ledger");
+            }
+        }
+
+        // Ensure that each existing transaction ID from the block exists in the ledger.
         for existing_transaction_id in expected_existing_transaction_ids {
             if !self.contains_transaction_id(&existing_transaction_id)? {
                 bail!("Transaction ID '{existing_transaction_id}' does not exist in the ledger");
diff --git a/ledger/src/tests.rs b/ledger/src/tests.rs
index 57acfb21f..7eaf4d2f8 100644
--- a/ledger/src/tests.rs
+++ b/ledger/src/tests.rs
@@ -972,7 +972,7 @@ function empty_function:
     // Add the block to the ledger.
     ledger.advance_to_next_block(&block).unwrap();
 
-    // Create a transaction with different transaction ids, but with a fixed transition id.
+    // Create a transaction with different transaction IDs, but with a fixed transition ID.
     let mut create_transaction_with_duplicate_transition_id = || -> Transaction<CurrentNetwork> {
         // Use a fixed seed RNG.
         let fixed_rng = &mut TestRng::from_seed(1);
diff --git a/synthesizer/src/vm/finalize.rs b/synthesizer/src/vm/finalize.rs
index 8fb676733..14ff59946 100644
--- a/synthesizer/src/vm/finalize.rs
+++ b/synthesizer/src/vm/finalize.rs
@@ -1249,7 +1249,7 @@ finalize transfer_public:
         assert!(matches!(candidate_transactions[0], ConfirmedTransaction::RejectedDeploy(..)));
         assert!(aborted_transaction_ids.is_empty());
 
-        // Check that the unconfirmed transaction id of the rejected deployment is correct.
+        // Check that the unconfirmed transaction ID of the rejected deployment is correct.
         assert_eq!(candidate_transactions[0].to_unconfirmed_transaction_id().unwrap(), deployment_transaction_id);
     }
 
diff --git a/synthesizer/src/vm/mod.rs b/synthesizer/src/vm/mod.rs
index 227ecf5f1..d1f80e603 100644
--- a/synthesizer/src/vm/mod.rs
+++ b/synthesizer/src/vm/mod.rs
@@ -101,7 +101,7 @@ impl<N: Network, C: ConsensusStorage<N>> VM<N, C> {
             transaction_store: &TransactionStore<N, T>,
             transaction_id: N::TransactionID,
         ) -> Result<Vec<(ProgramID<N>, Deployment<N>)>> {
-            // Retrieve the deployment from the transaction id.
+            // Retrieve the deployment from the transaction ID.
             let deployment = match transaction_store.get_deployment(&transaction_id)? {
                 Some(deployment) => deployment,
                 None => bail!("Deployment transaction '{transaction_id}' is not found in storage."),
@@ -123,11 +123,11 @@ impl<N: Network, C: ConsensusStorage<N>> VM<N, C> {
             for import_program_id in program.imports().keys() {
                 // Add the imports to the process if does not exist yet.
                 if !process.contains_program(import_program_id) {
-                    // Fetch the deployment transaction id.
+                    // Fetch the deployment transaction ID.
                     let Some(transaction_id) =
                         transaction_store.deployment_store().find_transaction_id_from_program_id(import_program_id)?
                     else {
-                        bail!("Transaction id for '{program_id}' is not found in storage.");
+                        bail!("Transaction ID for '{program_id}' is not found in storage.");
                     };
 
                     // Add the deployment and its imports found recursively.
@@ -947,7 +947,7 @@ function a:
             let deployment_transaction_ids =
                 vm.transaction_store().deployment_transaction_ids().map(|id| *id).collect::<Vec<_>>();
             // This assert check is here to ensure that we are properly loading imports even though any order will work for `VM::from`.
-            // Note: `deployment_transaction_ids` is sorted lexicographically by transaction id, so the order may change if we update internal methods.
+            // Note: `deployment_transaction_ids` is sorted lexicographically by transaction ID, so the order may change if we update internal methods.
             assert_eq!(
                 deployment_transaction_ids,
                 vec![deployment_4.id(), deployment_1.id(), deployment_2.id(), deployment_3.id()],
