diff --git a/dpc/tests/mod.rs b/dpc/tests/mod.rs
index e1f2f51d5..3a136fee3 100644
--- a/dpc/tests/mod.rs
+++ b/dpc/tests/mod.rs
@@ -44,8 +44,7 @@ fn test_posw_terminate() {
         block.to_coinbase_transaction().unwrap().to_records().next().unwrap(),
     );
 
-    let terminator = Arc::new(AtomicBool::new(false));
-    let thread_terminator = terminator.clone();
+    let thread_terminator = Arc::new(AtomicBool::new(false));
     std::thread::spawn(move || {
         std::thread::sleep(Duration::from_secs(1));
         thread_terminator.store(true, Ordering::SeqCst);
diff --git a/fields/src/lib.rs b/fields/src/lib.rs
index afa4e1b6c..5f87fa11c 100644
--- a/fields/src/lib.rs
+++ b/fields/src/lib.rs
@@ -103,8 +103,8 @@ pub fn batch_inversion_and_mul<F: Field>(v: &mut [F], coeff: &F) {
     let num_elem_per_thread = min_elements_per_thread.max(num_elems / num_cpus_available);
 
     // Batch invert in parallel, without copying the vector
-    v.par_chunks_mut(num_elem_per_thread).for_each(|mut chunk| {
-        serial_batch_inversion_and_mul(&mut chunk, coeff);
+    v.par_chunks_mut(num_elem_per_thread).for_each(|chunk| {
+        serial_batch_inversion_and_mul(chunk, coeff);
     });
 }
 
@@ -142,7 +142,7 @@ fn serial_batch_inversion_and_mul<F: Field>(v: &mut [F], coeff: &F) {
     {
         // tmp := tmp * f; f := tmp * s = 1/f
         let new_tmp = tmp * *f;
-        *f = tmp * &s;
+        *f = tmp * s;
         tmp = new_tmp;
     }
 }
diff --git a/marlin/src/ahp/matrices.rs b/marlin/src/ahp/matrices.rs
index 373b20a15..aa46bce21 100644
--- a/marlin/src/ahp/matrices.rs
+++ b/marlin/src/ahp/matrices.rs
@@ -155,22 +155,19 @@ pub(crate) fn arithmetize_matrix<F: PrimeField>(
     let a = a
         .iter()
         .enumerate()
-        .map(|(r, row)| row.iter().map(move |(f, i)| ((r, *i), *f)))
-        .flatten()
+        .flat_map(|(r, row)| row.iter().map(move |(f, i)| ((r, *i), *f)))
         .collect::<BTreeMap<(usize, usize), F>>();
 
     let b = b
         .iter()
         .enumerate()
-        .map(|(r, row)| row.iter().map(move |(f, i)| ((r, *i), *f)))
-        .flatten()
+        .flat_map(|(r, row)| row.iter().map(move |(f, i)| ((r, *i), *f)))
         .collect::<BTreeMap<(usize, usize), F>>();
 
     let c = c
         .iter()
         .enumerate()
-        .map(|(r, row)| row.iter().map(move |(f, i)| ((r, *i), *f)))
-        .flatten()
+        .flat_map(|(r, row)| row.iter().map(move |(f, i)| ((r, *i), *f)))
         .collect::<BTreeMap<(usize, usize), F>>();
 
     let eq_poly_vals_time = start_timer!(|| "Precomputing eq_poly_vals");
diff --git a/marlin/src/ahp/prover/prover.rs b/marlin/src/ahp/prover/prover.rs
index e44f4acca..7af995aa3 100644
--- a/marlin/src/ahp/prover/prover.rs
+++ b/marlin/src/ahp/prover/prover.rs
@@ -347,7 +347,7 @@ impl<F: PrimeField, MM: MarlinMode> AHPForR1CS<F, MM> {
         matrix_randomizers: &[F],
         input_domain: EvaluationDomain<F>,
         domain_h: EvaluationDomain<F>,
-        r_alpha_x_on_h: &Vec<F>,
+        r_alpha_x_on_h: &[F],
     ) -> Polynomial<F> {
         let mut t_evals_on_h = vec![F::zero(); domain_h.size()];
         for (matrix, eta) in matrices.zip(matrix_randomizers) {
diff --git a/marlin/src/constraints/ahp.rs b/marlin/src/constraints/ahp.rs
index b25f103f3..ba2270c67 100644
--- a/marlin/src/constraints/ahp.rs
+++ b/marlin/src/constraints/ahp.rs
@@ -646,7 +646,7 @@ impl<
 
         let gamma = gamma_ref;
 
-        let mut query_set_gadget = QuerySetVar::<TargetField, BaseField> { 0: HashSet::new() };
+        let mut query_set_gadget = QuerySetVar::<TargetField, BaseField>(HashSet::new());
 
         query_set_gadget.0.insert(("g_1".to_string(), LabeledPointVar {
             name: "beta".to_string(),
@@ -695,7 +695,7 @@ impl<
                 value: gamma.clone(),
             }));
 
-        let mut evaluations_gadget = EvaluationsVar::<TargetField, BaseField> { 0: HashMap::new() };
+        let mut evaluations_gadget = EvaluationsVar::<TargetField, BaseField>(HashMap::new());
 
         let zero = NonNativeFieldVar::<TargetField, BaseField>::zero(cs.ns(|| "nonnative_zero"))?;
 
diff --git a/marlin/src/fiat_shamir/fiat_shamir_algebraic_sponge_gadget.rs b/marlin/src/fiat_shamir/fiat_shamir_algebraic_sponge_gadget.rs
index c4f435493..d42290773 100644
--- a/marlin/src/fiat_shamir/fiat_shamir_algebraic_sponge_gadget.rs
+++ b/marlin/src/fiat_shamir/fiat_shamir_algebraic_sponge_gadget.rs
@@ -341,7 +341,7 @@ impl<
         elems: &[NonNativeFieldVar<TargetField, BaseField>],
         ty: OptimizationType,
     ) -> Result<(), SynthesisError> {
-        Self::push_gadgets_to_sponge(cs, &mut self.s, &elems.to_vec(), ty)
+        Self::push_gadgets_to_sponge(cs, &mut self.s, elems, ty)
     }
 
     fn absorb_native_field_elements<CS: ConstraintSystem<BaseField>>(
diff --git a/marlin/tests/test_verifier_constraints_on_groth16.rs b/marlin/tests/test_verifier_constraints_on_groth16.rs
index b379c41a3..50aecdab2 100644
--- a/marlin/tests/test_verifier_constraints_on_groth16.rs
+++ b/marlin/tests/test_verifier_constraints_on_groth16.rs
@@ -336,7 +336,7 @@ fn verifier_on_groth16() {
         // MulCircuit
         let a = Fr::rand(rng);
         let b = Fr::rand(rng);
-        let c = a * &b;
+        let c = a * b;
 
         let circ = MulCircuit {
             a: Some(a),
@@ -361,7 +361,7 @@ fn verifier_on_groth16() {
         // MulSquareCircuit
         let a = Fr::rand(rng);
         let b = Fr::rand(rng);
-        let c = (a * &b)
+        let c = (a * b)
             .square()
             .square()
             .square()
@@ -404,7 +404,7 @@ fn verifier_on_groth16() {
     let circ2 = RecursiveCircuit {
         vk: Some(vk2.clone()),
         proof: Some(proof2.clone()),
-        input: Some(input2.clone()),
+        input: Some(input2),
     };
 
     println!("\n\nDiffing constraints now\n\n");
@@ -431,9 +431,9 @@ fn verifier_on_groth16() {
     assert!(Groth16::verify(&rec_vk1, &rec_input2, &rec_proof2).unwrap());
 
     let circ3 = RecursiveCircuit {
-        vk: Some(vk2.clone()),
-        proof: Some(proof2.clone()),
-        input: Some(input1.clone()),
+        vk: Some(vk2),
+        proof: Some(proof2),
+        input: Some(input1),
     };
     let rec_input3 = circ2.public_input();
     let rec_proof3 = Groth16::<_, Vec<Fq>>::prove(&rec_pk1, &circ3, rng).unwrap();
diff --git a/polycommit/src/kzg10/data_structures.rs b/polycommit/src/kzg10/data_structures.rs
index d710216fa..9160c4f8b 100644
--- a/polycommit/src/kzg10/data_structures.rs
+++ b/polycommit/src/kzg10/data_structures.rs
@@ -379,7 +379,7 @@ impl<E: PairingEngine> PreparedCommitment<E> {
             cur.double_in_place();
         }
 
-        Self { 0: prepared_comm }
+        Self(prepared_comm)
     }
 }
 
diff --git a/polycommit/src/sonic_pc/data_structures.rs b/polycommit/src/sonic_pc/data_structures.rs
index 13b8c2a45..25d9c460b 100644
--- a/polycommit/src/sonic_pc/data_structures.rs
+++ b/polycommit/src/sonic_pc/data_structures.rs
@@ -45,7 +45,7 @@ impl<E: PairingEngine> Prepare<PreparedCommitment<E>> for Commitment<E> {
             cur.double_in_place();
         }
 
-        PreparedCommitment::<E> { 0: prepared_comm }
+        kzg10::PreparedCommitment::<E>(prepared_comm)
     }
 }
 
diff --git a/polycommit/src/sonic_pc/mod.rs b/polycommit/src/sonic_pc/mod.rs
index 684c42fda..e4d8be2b4 100644
--- a/polycommit/src/sonic_pc/mod.rs
+++ b/polycommit/src/sonic_pc/mod.rs
@@ -90,7 +90,7 @@ impl<E: PairingEngine> SonicKZG10<E> {
             comms.push(comm);
         }
         let comms = E::G1Projective::batch_normalization_into_affine(comms);
-        comms.into_iter().map(|c| Commitment { 0: c })
+        comms.into_iter().map(|c| kzg10::Commitment(c))
     }
 }
 
diff --git a/r1cs/src/optional_vec.rs b/r1cs/src/optional_vec.rs
index 5973cd312..f70294108 100644
--- a/r1cs/src/optional_vec.rs
+++ b/r1cs/src/optional_vec.rs
@@ -47,7 +47,7 @@ impl<T> OptionalVec<T> {
     /// of values, i.e. pushing it to its end.
     #[inline]
     pub fn insert(&mut self, elem: T) -> usize {
-        let idx = self.holes.pop().unwrap_or_else(|| self.values.len());
+        let idx = self.holes.pop().unwrap_or(self.values.len());
         if idx < self.values.len() {
             self.values[idx] = Some(elem);
         } else {
@@ -59,7 +59,7 @@ impl<T> OptionalVec<T> {
     /// Returns the index of the next value inserted into the `OptionalVec`.
     #[inline]
     pub fn next_idx(&self) -> usize {
-        self.holes.last().copied().unwrap_or_else(|| self.values.len())
+        self.holes.last().copied().unwrap_or(self.values.len())
     }
 
     /// Removes a value at the specified index; assumes that the index points to
diff --git a/utilities/src/bytes.rs b/utilities/src/bytes.rs
index 92886186c..cf6fc6626 100644
--- a/utilities/src/bytes.rs
+++ b/utilities/src/bytes.rs
@@ -348,10 +348,7 @@ impl<'a, T: 'a + ToBytes> ToBytes for &'a T {
 #[deprecated]
 #[inline]
 pub fn from_bytes_le_to_bits_le(bytes: &[u8]) -> impl Iterator<Item = bool> + DoubleEndedIterator<Item = bool> + '_ {
-    bytes
-        .iter()
-        .map(|byte| (0..8).map(move |i| (*byte >> i) & 1 == 1))
-        .flatten()
+    bytes.iter().flat_map(|byte| (0..8).map(move |i| (*byte >> i) & 1 == 1))
 }
 
 #[deprecated]
