diff --git a/ledger/block/src/lib.rs b/ledger/block/src/lib.rs
index 8b35d163a..bdb507d46 100644
--- a/ledger/block/src/lib.rs
+++ b/ledger/block/src/lib.rs
@@ -707,7 +707,7 @@ pub mod test_helpers {
         (block, transaction, private_key)
     }
 
-    pub(crate) fn sample_metadata(rng: &mut TestRng) -> Metadata<CurrentNetwork> {
+    pub(crate) fn sample_metadata() -> Metadata<CurrentNetwork> {
         let network = CurrentNetwork::ID;
         let round = u64::MAX;
         let height = u32::MAX;
@@ -719,7 +719,7 @@ pub mod test_helpers {
         let timestamp = i64::MAX - 1;
         let last_coinbase_timestamp = timestamp - 1;
         Metadata::new(
-            CurrentNetwork::ID,
+            network,
             round,
             height,
             cumulative_weight,
@@ -881,9 +881,9 @@ mod tests {
 
     #[test]
     fn test_serde_metadata() {
-        let rng = &mut TestRng::default();
-        let metadata = crate::test_helpers::sample_metadata(rng);
+        let metadata = crate::test_helpers::sample_metadata();
         let json_metadata = serde_json::to_string(&metadata).unwrap();
         let deserialized_metadata: Metadata<CurrentNetwork> = serde_json::from_str(&json_metadata).unwrap();
+        assert_eq!(metadata, deserialized_metadata);
     }
 }
diff --git a/synthesizer/process/src/cost.rs b/synthesizer/process/src/cost.rs
index fbb072c58..532637255 100644
--- a/synthesizer/process/src/cost.rs
+++ b/synthesizer/process/src/cost.rs
@@ -87,7 +87,7 @@ pub fn execution_cost_v1<N: Network>(process: &Process<N>, execution: &Execution
 
     // Get the finalize cost for the root transition.
     let stack = process.get_stack(transition.program_id())?;
-    let finalize_cost = cost_in_microcredits_v1(stack, transition.function_name())?;
+    let finalize_cost = cost_in_microcredits_v1(&stack, transition.function_name())?;
 
     // Compute the total cost in microcredits.
     let total_cost = storage_cost
