diff --git a/.cargo/release-version b/.cargo/release-version
index 1ddc0f604..686bba9ac 100644
--- a/.cargo/release-version
+++ b/.cargo/release-version
@@ -1 +1 @@
-v4.3.0
\ No newline at end of file
+v4.4.0
\ No newline at end of file
diff --git a/.circleci/semver-checks.sh b/.circleci/semver-checks.sh
index 1093a91dd..a33a84bca 100755
--- a/.circleci/semver-checks.sh
+++ b/.circleci/semver-checks.sh
@@ -1,10 +1,13 @@
 #! /bin/bash
 
+set -x
+
+BASELINE_REV=acd55ad100550 # UPDATE ME ON NECESSARY BREAKING CHANGES
+
 # Ensure that the command is installed.
 cargo install cargo-semver-checks@0.43.0 --locked
 
-BASELINE_REV=dec54170ce # UPDATE ME ON NECESSARY BREAKING CHANGES
+# Ensure we can find the baseline revision
+git fetch --unshallow || true
 
-# Exclude CLI as it has been removed
-cargo semver-checks --workspace --default-features \
-  --exclude=snarkvm-cli --baseline-rev $BASELINE_REV
+cargo semver-checks --workspace --default-features --baseline-rev $BASELINE_REV
diff --git a/Cargo.lock b/Cargo.lock
index 3e7e8a608..b1c25f0a2 100644
--- a/Cargo.lock
+++ b/Cargo.lock
@@ -2807,7 +2807,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm"
-version = "4.3.0"
+version = "4.4.0"
 dependencies = [
  "anyhow",
  "bincode",
@@ -2834,7 +2834,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-algorithms"
-version = "4.3.0"
+version = "4.4.0"
 dependencies = [
  "aleo-std",
  "anyhow",
@@ -2866,7 +2866,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-algorithms-cuda"
-version = "4.3.0"
+version = "4.4.0"
 dependencies = [
  "blst",
  "cc",
@@ -2876,7 +2876,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-circuit"
-version = "4.3.0"
+version = "4.4.0"
 dependencies = [
  "snarkvm-circuit-account",
  "snarkvm-circuit-algorithms",
@@ -2889,7 +2889,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-circuit-account"
-version = "4.3.0"
+version = "4.4.0"
 dependencies = [
  "anyhow",
  "snarkvm-circuit-network",
@@ -2900,7 +2900,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-circuit-algorithms"
-version = "4.3.0"
+version = "4.4.0"
 dependencies = [
  "anyhow",
  "snarkvm-circuit-types",
@@ -2912,7 +2912,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-circuit-collections"
-version = "4.3.0"
+version = "4.4.0"
 dependencies = [
  "anyhow",
  "snarkvm-circuit-algorithms",
@@ -2926,7 +2926,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-circuit-environment"
-version = "4.3.0"
+version = "4.4.0"
 dependencies = [
  "criterion",
  "indexmap 2.12.0",
@@ -2948,11 +2948,11 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-circuit-environment-witness"
-version = "4.3.0"
+version = "4.4.0"
 
 [[package]]
 name = "snarkvm-circuit-network"
-version = "4.3.0"
+version = "4.4.0"
 dependencies = [
  "snarkvm-circuit-algorithms",
  "snarkvm-circuit-collections",
@@ -2963,7 +2963,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-circuit-program"
-version = "4.3.0"
+version = "4.4.0"
 dependencies = [
  "anyhow",
  "rand 0.8.5",
@@ -2980,7 +2980,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-circuit-types"
-version = "4.3.0"
+version = "4.4.0"
 dependencies = [
  "snarkvm-circuit-environment",
  "snarkvm-circuit-types-address",
@@ -2995,7 +2995,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-circuit-types-address"
-version = "4.3.0"
+version = "4.4.0"
 dependencies = [
  "snarkvm-circuit-environment",
  "snarkvm-circuit-types-boolean",
@@ -3007,7 +3007,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-circuit-types-boolean"
-version = "4.3.0"
+version = "4.4.0"
 dependencies = [
  "criterion",
  "snarkvm-circuit-environment",
@@ -3016,7 +3016,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-circuit-types-field"
-version = "4.3.0"
+version = "4.4.0"
 dependencies = [
  "snarkvm-circuit-environment",
  "snarkvm-circuit-types-boolean",
@@ -3025,7 +3025,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-circuit-types-group"
-version = "4.3.0"
+version = "4.4.0"
 dependencies = [
  "snarkvm-circuit-environment",
  "snarkvm-circuit-types-boolean",
@@ -3037,7 +3037,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-circuit-types-integers"
-version = "4.3.0"
+version = "4.4.0"
 dependencies = [
  "paste",
  "snarkvm-circuit-environment",
@@ -3050,7 +3050,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-circuit-types-scalar"
-version = "4.3.0"
+version = "4.4.0"
 dependencies = [
  "snarkvm-circuit-environment",
  "snarkvm-circuit-types-boolean",
@@ -3060,7 +3060,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-circuit-types-string"
-version = "4.3.0"
+version = "4.4.0"
 dependencies = [
  "rand 0.8.5",
  "snarkvm-circuit-environment",
@@ -3073,7 +3073,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-console"
-version = "4.3.0"
+version = "4.4.0"
 dependencies = [
  "snarkvm-console-account",
  "snarkvm-console-algorithms",
@@ -3085,7 +3085,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-console-account"
-version = "4.3.0"
+version = "4.4.0"
 dependencies = [
  "bincode",
  "bs58",
@@ -3098,7 +3098,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-console-algorithms"
-version = "4.3.0"
+version = "4.4.0"
 dependencies = [
  "bincode",
  "blake2s_simd",
@@ -3118,7 +3118,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-console-collections"
-version = "4.3.0"
+version = "4.4.0"
 dependencies = [
  "aleo-std",
  "criterion",
@@ -3131,7 +3131,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-console-network"
-version = "4.3.0"
+version = "4.4.0"
 dependencies = [
  "anyhow",
  "enum-iterator",
@@ -3150,7 +3150,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-console-network-environment"
-version = "4.3.0"
+version = "4.4.0"
 dependencies = [
  "anyhow",
  "bech32",
@@ -3167,7 +3167,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-console-program"
-version = "4.3.0"
+version = "4.4.0"
 dependencies = [
  "bincode",
  "enum-iterator",
@@ -3188,7 +3188,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-console-types"
-version = "4.3.0"
+version = "4.4.0"
 dependencies = [
  "criterion",
  "snarkvm-console-network",
@@ -3204,7 +3204,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-console-types-address"
-version = "4.3.0"
+version = "4.4.0"
 dependencies = [
  "bincode",
  "serde_json",
@@ -3216,7 +3216,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-console-types-boolean"
-version = "4.3.0"
+version = "4.4.0"
 dependencies = [
  "bincode",
  "serde_json",
@@ -3225,7 +3225,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-console-types-field"
-version = "4.3.0"
+version = "4.4.0"
 dependencies = [
  "bincode",
  "serde_json",
@@ -3236,7 +3236,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-console-types-group"
-version = "4.3.0"
+version = "4.4.0"
 dependencies = [
  "bincode",
  "serde_json",
@@ -3248,7 +3248,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-console-types-integers"
-version = "4.3.0"
+version = "4.4.0"
 dependencies = [
  "bincode",
  "serde_json",
@@ -3260,7 +3260,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-console-types-scalar"
-version = "4.3.0"
+version = "4.4.0"
 dependencies = [
  "bincode",
  "serde_json",
@@ -3272,7 +3272,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-console-types-string"
-version = "4.3.0"
+version = "4.4.0"
 dependencies = [
  "bincode",
  "serde_json",
@@ -3284,7 +3284,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-curves"
-version = "4.3.0"
+version = "4.4.0"
 dependencies = [
  "bincode",
  "criterion",
@@ -3299,7 +3299,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-fields"
-version = "4.3.0"
+version = "4.4.0"
 dependencies = [
  "aleo-std",
  "anyhow",
@@ -3317,7 +3317,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-ledger"
-version = "4.3.0"
+version = "4.4.0"
 dependencies = [
  "aleo-std",
  "aleo-std-storage",
@@ -3352,7 +3352,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-ledger-authority"
-version = "4.3.0"
+version = "4.4.0"
 dependencies = [
  "anyhow",
  "bincode",
@@ -3365,7 +3365,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-ledger-block"
-version = "4.3.0"
+version = "4.4.0"
 dependencies = [
  "aleo-std",
  "anyhow",
@@ -3393,7 +3393,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-ledger-committee"
-version = "4.3.0"
+version = "4.4.0"
 dependencies = [
  "anyhow",
  "bincode",
@@ -3414,7 +3414,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-ledger-narwhal"
-version = "4.3.0"
+version = "4.4.0"
 dependencies = [
  "snarkvm-ledger-narwhal",
  "snarkvm-ledger-narwhal-batch-certificate",
@@ -3427,7 +3427,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-ledger-narwhal-batch-certificate"
-version = "4.3.0"
+version = "4.4.0"
 dependencies = [
  "bincode",
  "indexmap 2.12.0",
@@ -3441,7 +3441,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-ledger-narwhal-batch-header"
-version = "4.3.0"
+version = "4.4.0"
 dependencies = [
  "bincode",
  "indexmap 2.12.0",
@@ -3455,7 +3455,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-ledger-narwhal-data"
-version = "4.3.0"
+version = "4.4.0"
 dependencies = [
  "bytes",
  "serde_json",
@@ -3467,7 +3467,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-ledger-narwhal-subdag"
-version = "4.3.0"
+version = "4.4.0"
 dependencies = [
  "bincode",
  "indexmap 2.12.0",
@@ -3483,7 +3483,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-ledger-narwhal-transmission"
-version = "4.3.0"
+version = "4.4.0"
 dependencies = [
  "bincode",
  "bytes",
@@ -3496,7 +3496,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-ledger-narwhal-transmission-id"
-version = "4.3.0"
+version = "4.4.0"
 dependencies = [
  "bincode",
  "serde_json",
@@ -3506,7 +3506,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-ledger-puzzle"
-version = "4.3.0"
+version = "4.4.0"
 dependencies = [
  "aleo-std",
  "anyhow",
@@ -3527,7 +3527,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-ledger-puzzle-epoch"
-version = "4.3.0"
+version = "4.4.0"
 dependencies = [
  "aleo-std",
  "anyhow",
@@ -3549,7 +3549,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-ledger-query"
-version = "4.3.0"
+version = "4.4.0"
 dependencies = [
  "anyhow",
  "async-trait",
@@ -3565,7 +3565,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-ledger-store"
-version = "4.3.0"
+version = "4.4.0"
 dependencies = [
  "aleo-std",
  "aleo-std-storage",
@@ -3596,7 +3596,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-ledger-test-helpers"
-version = "4.3.0"
+version = "4.4.0"
 dependencies = [
  "aleo-std",
  "anyhow",
@@ -3613,14 +3613,14 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-metrics"
-version = "4.3.0"
+version = "4.4.0"
 dependencies = [
  "metrics",
 ]
 
 [[package]]
 name = "snarkvm-parameters"
-version = "4.3.0"
+version = "4.4.0"
 dependencies = [
  "aleo-std",
  "anyhow",
@@ -3652,7 +3652,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-synthesizer"
-version = "4.3.0"
+version = "4.4.0"
 dependencies = [
  "aleo-std",
  "anyhow",
@@ -3694,7 +3694,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-synthesizer-process"
-version = "4.3.0"
+version = "4.4.0"
 dependencies = [
  "aleo-std",
  "bincode",
@@ -3723,7 +3723,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-synthesizer-program"
-version = "4.3.0"
+version = "4.4.0"
 dependencies = [
  "bincode",
  "criterion",
@@ -3745,7 +3745,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-synthesizer-snark"
-version = "4.3.0"
+version = "4.4.0"
 dependencies = [
  "bincode",
  "serde_json",
@@ -3770,7 +3770,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-utilities"
-version = "4.3.0"
+version = "4.4.0"
 dependencies = [
  "aleo-std",
  "anyhow",
@@ -3793,7 +3793,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-utilities-derives"
-version = "4.3.0"
+version = "4.4.0"
 dependencies = [
  "proc-macro2",
  "quote 1.0.42",
@@ -3802,7 +3802,7 @@ dependencies = [
 
 [[package]]
 name = "snarkvm-wasm"
-version = "4.3.0"
+version = "4.4.0"
 dependencies = [
  "getrandom 0.2.16",
  "snarkvm-console",
diff --git a/Cargo.toml b/Cargo.toml
index 2f4f887ef..08d453961 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm"
-version = "4.3.0"
+version = "4.4.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "A decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -179,251 +179,251 @@ test_consensus_heights = [ "snarkvm-console/test_consensus_heights", "snarkvm-sy
 
 [workspace.dependencies.snarkvm-algorithms]
 path = "algorithms"
-version = "=4.3.0"
+version = "=4.4.0"
 default-features = false
 
 [workspace.dependencies.snarkvm-algorithms-cuda]
 path = "algorithms/cuda"
-version = "=4.3.0"
+version = "=4.4.0"
 
 [workspace.dependencies.snarkvm-circuit]
 path = "circuit"
-version = "=4.3.0"
+version = "=4.4.0"
 
 [workspace.dependencies.snarkvm-circuit-account]
 path = "circuit/account"
-version = "=4.3.0"
+version = "=4.4.0"
 
 [workspace.dependencies.snarkvm-circuit-algorithms]
 path = "circuit/algorithms"
-version = "=4.3.0"
+version = "=4.4.0"
 
 [workspace.dependencies.snarkvm-circuit-collections]
 path = "circuit/collections"
-version = "=4.3.0"
+version = "=4.4.0"
 
 [workspace.dependencies.snarkvm-circuit-environment]
 path = "circuit/environment"
-version = "=4.3.0"
+version = "=4.4.0"
 
 [workspace.dependencies.snarkvm-circuit-environment-witness]
 path = "circuit/environment/witness"
-version = "=4.3.0"
+version = "=4.4.0"
 default-features = false
 
 [workspace.dependencies.snarkvm-circuit-network]
 path = "circuit/network"
-version = "=4.3.0"
+version = "=4.4.0"
 
 [workspace.dependencies.snarkvm-circuit-program]
 path = "circuit/program"
-version = "=4.3.0"
+version = "=4.4.0"
 
 [workspace.dependencies.snarkvm-circuit-types]
 path = "circuit/types"
-version = "=4.3.0"
+version = "=4.4.0"
 
 [workspace.dependencies.snarkvm-circuit-types-address]
 path = "circuit/types/address"
-version = "=4.3.0"
+version = "=4.4.0"
 
 [workspace.dependencies.snarkvm-circuit-types-boolean]
 path = "circuit/types/boolean"
-version = "=4.3.0"
+version = "=4.4.0"
 
 [workspace.dependencies.snarkvm-circuit-types-field]
 path = "circuit/types/field"
-version = "=4.3.0"
+version = "=4.4.0"
 
 [workspace.dependencies.snarkvm-circuit-types-scalar]
 path = "circuit/types/scalar"
-version = "=4.3.0"
+version = "=4.4.0"
 
 [workspace.dependencies.snarkvm-circuit-types-string]
 path = "circuit/types/string"
-version = "=4.3.0"
+version = "=4.4.0"
 
 [workspace.dependencies.snarkvm-console]
 path = "console"
-version = "=4.3.0"
+version = "=4.4.0"
 default-features = false
 
 [workspace.dependencies.snarkvm-console-account]
 path = "console/account"
-version = "=4.3.0"
+version = "=4.4.0"
 
 [workspace.dependencies.snarkvm-console-algorithms]
 path = "console/algorithms"
-version = "=4.3.0"
+version = "=4.4.0"
 
 [workspace.dependencies.snarkvm-console-collections]
 path = "console/collections"
-version = "=4.3.0"
+version = "=4.4.0"
 
 [workspace.dependencies.snarkvm-console-network]
 path = "console/network"
-version = "=4.3.0"
+version = "=4.4.0"
 
 [workspace.dependencies.snarkvm-console-network-environment]
 path = "console/network/environment"
-version = "=4.3.0"
+version = "=4.4.0"
 
 [workspace.dependencies.snarkvm-console-program]
 path = "console/program"
-version = "=4.3.0"
+version = "=4.4.0"
 
 [workspace.dependencies.snarkvm-console-types]
 path = "console/types"
-version = "=4.3.0"
+version = "=4.4.0"
 default-features = false
 
 [workspace.dependencies.snarkvm-console-types-address]
 path = "console/types/address"
-version = "=4.3.0"
+version = "=4.4.0"
 
 [workspace.dependencies.snarkvm-console-types-boolean]
 path = "console/types/boolean"
-version = "=4.3.0"
+version = "=4.4.0"
 
 [workspace.dependencies.snarkvm-console-types-field]
 path = "console/types/field"
-version = "=4.3.0"
+version = "=4.4.0"
 
 [workspace.dependencies.snarkvm-console-types-group]
 path = "console/types/group"
-version = "=4.3.0"
+version = "=4.4.0"
 
 [workspace.dependencies.snarkvm-console-types-integers]
 path = "console/types/integers"
-version = "=4.3.0"
+version = "=4.4.0"
 
 [workspace.dependencies.snarkvm-console-types-scalar]
 path = "console/types/scalar"
-version = "=4.3.0"
+version = "=4.4.0"
 
 [workspace.dependencies.snarkvm-console-types-string]
 path = "console/types/string"
-version = "=4.3.0"
+version = "=4.4.0"
 
 [workspace.dependencies.snarkvm-circuit-types-group]
 path = "circuit/types/group"
-version = "=4.3.0"
+version = "=4.4.0"
 
 [workspace.dependencies.snarkvm-circuit-types-integers]
 path = "circuit/types/integers"
-version = "=4.3.0"
+version = "=4.4.0"
 
 [workspace.dependencies.snarkvm-curves]
 path = "curves"
-version = "=4.3.0"
+version = "=4.4.0"
 default-features = false
 
 [workspace.dependencies.snarkvm-fields]
 path = "fields"
-version = "=4.3.0"
+version = "=4.4.0"
 default-features = false
 
 [workspace.dependencies.snarkvm-ledger]
 path = "ledger"
-version = "=4.3.0"
+version = "=4.4.0"
 
 [workspace.dependencies.snarkvm-ledger-authority]
 path = "ledger/authority"
-version = "=4.3.0"
+version = "=4.4.0"
 
 [workspace.dependencies.snarkvm-ledger-block]
 path = "ledger/block"
-version = "=4.3.0"
+version = "=4.4.0"
 
 [workspace.dependencies.snarkvm-ledger-committee]
 path = "ledger/committee"
-version = "=4.3.0"
+version = "=4.4.0"
 default-features = false
 
 [workspace.dependencies.snarkvm-ledger-narwhal]
 path = "ledger/narwhal"
-version = "=4.3.0"
+version = "=4.4.0"
 
 [workspace.dependencies.snarkvm-ledger-narwhal-data]
 path = "ledger/narwhal/data"
-version = "=4.3.0"
+version = "=4.4.0"
 
 [workspace.dependencies.snarkvm-ledger-narwhal-batch-header]
 path = "ledger/narwhal/batch-header"
-version = "=4.3.0"
+version = "=4.4.0"
 
 [workspace.dependencies.snarkvm-ledger-narwhal-batch-certificate]
 path = "ledger/narwhal/batch-certificate"
-version = "=4.3.0"
+version = "=4.4.0"
 
 [workspace.dependencies.snarkvm-ledger-narwhal-subdag]
 path = "ledger/narwhal/subdag"
-version = "=4.3.0"
+version = "=4.4.0"
 
 [workspace.dependencies.snarkvm-ledger-narwhal-transmission]
 path = "ledger/narwhal/transmission"
-version = "=4.3.0"
+version = "=4.4.0"
 
 [workspace.dependencies.snarkvm-ledger-narwhal-transmission-id]
 path = "ledger/narwhal/transmission-id"
-version = "=4.3.0"
+version = "=4.4.0"
 
 [workspace.dependencies.snarkvm-ledger-puzzle]
 path = "ledger/puzzle"
-version = "=4.3.0"
+version = "=4.4.0"
 
 [workspace.dependencies.snarkvm-ledger-puzzle-epoch]
 path = "ledger/puzzle/epoch"
-version = "=4.3.0"
+version = "=4.4.0"
 
 [workspace.dependencies.snarkvm-ledger-query]
 path = "ledger/query"
-version = "=4.3.0"
+version = "=4.4.0"
 default-features = false
 
 [workspace.dependencies.snarkvm-ledger-store]
 path = "ledger/store"
-version = "=4.3.0"
+version = "=4.4.0"
 
 [workspace.dependencies.snarkvm-ledger-test-helpers]
 path = "ledger/test-helpers"
-version = "=4.3.0"
+version = "=4.4.0"
 
 [workspace.dependencies.snarkvm-metrics]
 path = "metrics"
-version = "=4.3.0"
+version = "=4.4.0"
 
 [workspace.dependencies.snarkvm-parameters]
 path = "parameters"
-version = "=4.3.0"
+version = "=4.4.0"
 
 [workspace.dependencies.snarkvm-synthesizer]
 path = "synthesizer"
-version = "=4.3.0"
+version = "=4.4.0"
 
 [workspace.dependencies.snarkvm-synthesizer-process]
 path = "synthesizer/process"
-version = "=4.3.0"
+version = "=4.4.0"
 
 [workspace.dependencies.snarkvm-synthesizer-program]
 path = "synthesizer/program"
-version = "=4.3.0"
+version = "=4.4.0"
 
 [workspace.dependencies.snarkvm-synthesizer-snark]
 path = "synthesizer/snark"
-version = "=4.3.0"
+version = "=4.4.0"
 
 [workspace.dependencies.snarkvm-utilities]
 path = "utilities"
-version = "=4.3.0"
+version = "=4.4.0"
 
 [workspace.dependencies.snarkvm-utilities-derives]
 path = "utilities/derives"
-version = "=4.3.0"
+version = "=4.4.0"
 
 [workspace.dependencies.snarkvm-wasm]
 path = "wasm"
-version = "=4.3.0"
+version = "=4.4.0"
 
 [workspace.dependencies.aleo-std]
 version = "1.0.1"
diff --git a/algorithms/Cargo.toml b/algorithms/Cargo.toml
index b5fc9d23d..56aadbb18 100644
--- a/algorithms/Cargo.toml
+++ b/algorithms/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-algorithms"
-version = "4.3.0"
+version = "4.4.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Algorithms for a decentralized virtual machine"
 homepage = "https://aleo.org"
diff --git a/algorithms/cuda/Cargo.toml b/algorithms/cuda/Cargo.toml
index 94d927df9..438b79675 100644
--- a/algorithms/cuda/Cargo.toml
+++ b/algorithms/cuda/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-algorithms-cuda"
-version = "4.3.0"
+version = "4.4.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Cuda optimizations for a decentralized virtual machine"
 homepage = "https://aleo.org"
diff --git a/circuit/Cargo.toml b/circuit/Cargo.toml
index 12720e068..3c1111b01 100644
--- a/circuit/Cargo.toml
+++ b/circuit/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-circuit"
-version = "4.3.0"
+version = "4.4.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Circuits for a decentralized virtual machine"
 homepage = "https://aleo.org"
diff --git a/circuit/account/Cargo.toml b/circuit/account/Cargo.toml
index 072549632..e36a95a26 100644
--- a/circuit/account/Cargo.toml
+++ b/circuit/account/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-circuit-account"
-version = "4.3.0"
+version = "4.4.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Account circuit library for a decentralized virtual machine"
 homepage = "https://aleo.org"
diff --git a/circuit/algorithms/Cargo.toml b/circuit/algorithms/Cargo.toml
index 819f006f9..62bd47086 100644
--- a/circuit/algorithms/Cargo.toml
+++ b/circuit/algorithms/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-circuit-algorithms"
-version = "4.3.0"
+version = "4.4.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Algorithm circuit library for a decentralized virtual machine"
 homepage = "https://aleo.org"
diff --git a/circuit/collections/Cargo.toml b/circuit/collections/Cargo.toml
index 0846e1385..8dbcf9448 100644
--- a/circuit/collections/Cargo.toml
+++ b/circuit/collections/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-circuit-collections"
-version = "4.3.0"
+version = "4.4.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Collections circuit library for a decentralized virtual machine"
 homepage = "https://aleo.org"
diff --git a/circuit/environment/Cargo.toml b/circuit/environment/Cargo.toml
index 1526cb203..627713b38 100644
--- a/circuit/environment/Cargo.toml
+++ b/circuit/environment/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-circuit-environment"
-version = "4.3.0"
+version = "4.4.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Circuit environment for a decentralized virtual machine"
 homepage = "https://aleo.org"
diff --git a/circuit/environment/witness/Cargo.toml b/circuit/environment/witness/Cargo.toml
index b0c974254..99bcec279 100644
--- a/circuit/environment/witness/Cargo.toml
+++ b/circuit/environment/witness/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-circuit-environment-witness"
-version = "4.3.0"
+version = "4.4.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "A procedural macro to construct a witness in an environment"
 homepage = "https://aleo.org"
diff --git a/circuit/network/Cargo.toml b/circuit/network/Cargo.toml
index 521c1b68d..9c1921024 100644
--- a/circuit/network/Cargo.toml
+++ b/circuit/network/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-circuit-network"
-version = "4.3.0"
+version = "4.4.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Network circuit library for a decentralized virtual machine"
 homepage = "https://aleo.org"
diff --git a/circuit/program/Cargo.toml b/circuit/program/Cargo.toml
index c15080efe..de44a9edd 100644
--- a/circuit/program/Cargo.toml
+++ b/circuit/program/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-circuit-program"
-version = "4.3.0"
+version = "4.4.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Program circuit library for a decentralized virtual machine"
 homepage = "https://aleo.org"
diff --git a/circuit/types/Cargo.toml b/circuit/types/Cargo.toml
index 93710b671..38562d18c 100644
--- a/circuit/types/Cargo.toml
+++ b/circuit/types/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-circuit-types"
-version = "4.3.0"
+version = "4.4.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Primitive circuit for a decentralized virtual machine"
 homepage = "https://aleo.org"
diff --git a/circuit/types/address/Cargo.toml b/circuit/types/address/Cargo.toml
index 0c62d5340..7878ae412 100644
--- a/circuit/types/address/Cargo.toml
+++ b/circuit/types/address/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-circuit-types-address"
-version = "4.3.0"
+version = "4.4.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Address circuit for a decentralized virtual machine"
 homepage = "https://aleo.org"
diff --git a/circuit/types/boolean/Cargo.toml b/circuit/types/boolean/Cargo.toml
index 95072acc0..587f1ed9f 100644
--- a/circuit/types/boolean/Cargo.toml
+++ b/circuit/types/boolean/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-circuit-types-boolean"
-version = "4.3.0"
+version = "4.4.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Boolean circuit for a decentralized virtual machine"
 homepage = "https://aleo.org"
diff --git a/circuit/types/field/Cargo.toml b/circuit/types/field/Cargo.toml
index b5471acb5..447c85c11 100644
--- a/circuit/types/field/Cargo.toml
+++ b/circuit/types/field/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-circuit-types-field"
-version = "4.3.0"
+version = "4.4.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Field circuit for a decentralized virtual machine"
 homepage = "https://aleo.org"
diff --git a/circuit/types/group/Cargo.toml b/circuit/types/group/Cargo.toml
index 9e6a90b14..5cedd7b50 100644
--- a/circuit/types/group/Cargo.toml
+++ b/circuit/types/group/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-circuit-types-group"
-version = "4.3.0"
+version = "4.4.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Group circuit for a decentralized virtual machine"
 homepage = "https://aleo.org"
diff --git a/circuit/types/integers/Cargo.toml b/circuit/types/integers/Cargo.toml
index 7a044b4e1..f5f0d63c8 100644
--- a/circuit/types/integers/Cargo.toml
+++ b/circuit/types/integers/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-circuit-types-integers"
-version = "4.3.0"
+version = "4.4.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Integer circuit for a decentralized virtual machine"
 homepage = "https://aleo.org"
diff --git a/circuit/types/scalar/Cargo.toml b/circuit/types/scalar/Cargo.toml
index ca7152703..d074381eb 100644
--- a/circuit/types/scalar/Cargo.toml
+++ b/circuit/types/scalar/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-circuit-types-scalar"
-version = "4.3.0"
+version = "4.4.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Scalar circuit for a decentralized virtual machine"
 homepage = "https://aleo.org"
diff --git a/circuit/types/string/Cargo.toml b/circuit/types/string/Cargo.toml
index 6df8a876f..89a1fe031 100644
--- a/circuit/types/string/Cargo.toml
+++ b/circuit/types/string/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-circuit-types-string"
-version = "4.3.0"
+version = "4.4.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "String circuit for a decentralized virtual machine"
 homepage = "https://aleo.org"
diff --git a/console/Cargo.toml b/console/Cargo.toml
index fd91f1913..75fd28db6 100644
--- a/console/Cargo.toml
+++ b/console/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-console"
-version = "4.3.0"
+version = "4.4.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Console environment for a decentralized virtual machine"
 homepage = "https://aleo.org"
diff --git a/console/account/Cargo.toml b/console/account/Cargo.toml
index e54def844..e22b2b0a9 100644
--- a/console/account/Cargo.toml
+++ b/console/account/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-console-account"
-version = "4.3.0"
+version = "4.4.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Account operations for a decentralized virtual machine"
 homepage = "https://aleo.org"
diff --git a/console/algorithms/Cargo.toml b/console/algorithms/Cargo.toml
index 1f0c14446..aa86afe9c 100644
--- a/console/algorithms/Cargo.toml
+++ b/console/algorithms/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-console-algorithms"
-version = "4.3.0"
+version = "4.4.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Console algorithms for a decentralized virtual machine"
 homepage = "https://aleo.org"
diff --git a/console/collections/Cargo.toml b/console/collections/Cargo.toml
index 9cd7b9c00..98e592f75 100644
--- a/console/collections/Cargo.toml
+++ b/console/collections/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-console-collections"
-version = "4.3.0"
+version = "4.4.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Collections for a decentralized virtual machine"
 homepage = "https://aleo.org"
diff --git a/console/network/Cargo.toml b/console/network/Cargo.toml
index aaad1857a..a452c2566 100644
--- a/console/network/Cargo.toml
+++ b/console/network/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-console-network"
-version = "4.3.0"
+version = "4.4.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Network console library for a decentralized virtual machine"
 homepage = "https://aleo.org"
diff --git a/console/network/environment/Cargo.toml b/console/network/environment/Cargo.toml
index 6089a7297..ab6dc4939 100644
--- a/console/network/environment/Cargo.toml
+++ b/console/network/environment/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-console-network-environment"
-version = "4.3.0"
+version = "4.4.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Environment console library for a decentralized virtual machine"
 homepage = "https://aleo.org"
diff --git a/console/network/src/consensus_heights.rs b/console/network/src/consensus_heights.rs
index 1f9151bc0..149b83f4d 100644
--- a/console/network/src/consensus_heights.rs
+++ b/console/network/src/consensus_heights.rs
@@ -45,7 +45,7 @@ pub enum ConsensusVersion {
     V10 = 10,
     /// V11: Expand array size limit to 512 and introduce ECDSA signature verification opcodes.
     V11 = 11,
-    /// V12: Prevent connection to forked nodes, disable StringType.
+    /// V12: Prevent connection to forked nodes, disable StringType, enable block timestamp.
     V12 = 12,
 }
 
@@ -105,7 +105,7 @@ pub const CANARY_V0_CONSENSUS_VERSION_HEIGHTS: [(ConsensusVersion, u32); NUM_CON
     (ConsensusVersion::V9, 8_028_000),
     (ConsensusVersion::V10, 8_600_000),
     (ConsensusVersion::V11, 9_510_000),
-    (ConsensusVersion::V12, 9_906_000),
+    (ConsensusVersion::V12, 10_030_000),
 ];
 
 /// The consensus version height for `MainnetV0`.
@@ -121,7 +121,7 @@ pub const MAINNET_V0_CONSENSUS_VERSION_HEIGHTS: [(ConsensusVersion, u32); NUM_CO
     (ConsensusVersion::V9, 10_272_000),
     (ConsensusVersion::V10, 11_205_000),
     (ConsensusVersion::V11, 12_870_000),
-    (ConsensusVersion::V12, 13_579_000),
+    (ConsensusVersion::V12, 13_815_000),
 ];
 
 /// The consensus version heights for `TestnetV0`.
@@ -137,7 +137,7 @@ pub const TESTNET_V0_CONSENSUS_VERSION_HEIGHTS: [(ConsensusVersion, u32); NUM_CO
     (ConsensusVersion::V9, 9_800_000),
     (ConsensusVersion::V10, 10_525_000),
     (ConsensusVersion::V11, 11_952_000),
-    (ConsensusVersion::V12, 12_522_984),
+    (ConsensusVersion::V12, 12_669_000),
 ];
 
 /// The consensus version heights when the `test_consensus_heights` feature is enabled.
diff --git a/console/program/Cargo.toml b/console/program/Cargo.toml
index 6f998f8a1..e97d66985 100644
--- a/console/program/Cargo.toml
+++ b/console/program/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-console-program"
-version = "4.3.0"
+version = "4.4.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Program operations for a decentralized virtual machine"
 homepage = "https://aleo.org"
diff --git a/console/types/Cargo.toml b/console/types/Cargo.toml
index a8703dc36..3888f9af1 100644
--- a/console/types/Cargo.toml
+++ b/console/types/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-console-types"
-version = "4.3.0"
+version = "4.4.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Console types for a decentralized virtual machine"
 homepage = "https://aleo.org"
diff --git a/console/types/address/Cargo.toml b/console/types/address/Cargo.toml
index 00da1930c..2b70c75fe 100644
--- a/console/types/address/Cargo.toml
+++ b/console/types/address/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-console-types-address"
-version = "4.3.0"
+version = "4.4.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Type operations for a decentralized virtual machine"
 homepage = "https://aleo.org"
diff --git a/console/types/boolean/Cargo.toml b/console/types/boolean/Cargo.toml
index 34f3c1673..ad9f050d1 100644
--- a/console/types/boolean/Cargo.toml
+++ b/console/types/boolean/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-console-types-boolean"
-version = "4.3.0"
+version = "4.4.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Type operations for a decentralized virtual machine"
 homepage = "https://aleo.org"
diff --git a/console/types/field/Cargo.toml b/console/types/field/Cargo.toml
index 0aeea33fc..1a166557d 100644
--- a/console/types/field/Cargo.toml
+++ b/console/types/field/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-console-types-field"
-version = "4.3.0"
+version = "4.4.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Type operations for a decentralized virtual machine"
 homepage = "https://aleo.org"
diff --git a/console/types/group/Cargo.toml b/console/types/group/Cargo.toml
index 2c0eb402a..95a3aed3b 100644
--- a/console/types/group/Cargo.toml
+++ b/console/types/group/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-console-types-group"
-version = "4.3.0"
+version = "4.4.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Type operations for a decentralized virtual machine"
 homepage = "https://aleo.org"
diff --git a/console/types/integers/Cargo.toml b/console/types/integers/Cargo.toml
index 599ab5a69..c37a7e498 100644
--- a/console/types/integers/Cargo.toml
+++ b/console/types/integers/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-console-types-integers"
-version = "4.3.0"
+version = "4.4.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Type operations for a decentralized virtual machine"
 homepage = "https://aleo.org"
diff --git a/console/types/scalar/Cargo.toml b/console/types/scalar/Cargo.toml
index 0b8c75ea1..a79f1d403 100644
--- a/console/types/scalar/Cargo.toml
+++ b/console/types/scalar/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-console-types-scalar"
-version = "4.3.0"
+version = "4.4.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Type operations for a decentralized virtual machine"
 homepage = "https://aleo.org"
diff --git a/console/types/string/Cargo.toml b/console/types/string/Cargo.toml
index 199a6910c..2e7186608 100644
--- a/console/types/string/Cargo.toml
+++ b/console/types/string/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-console-types-string"
-version = "4.3.0"
+version = "4.4.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Type operations for a decentralized virtual machine"
 homepage = "https://aleo.org"
diff --git a/curves/Cargo.toml b/curves/Cargo.toml
index 74341f061..a7deed106 100644
--- a/curves/Cargo.toml
+++ b/curves/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-curves"
-version = "4.3.0"
+version = "4.4.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Curves for a decentralized virtual machine"
 homepage = "https://aleo.org"
diff --git a/fields/Cargo.toml b/fields/Cargo.toml
index 750520707..07fd689fb 100644
--- a/fields/Cargo.toml
+++ b/fields/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-fields"
-version = "4.3.0"
+version = "4.4.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Fields for a decentralized virtual machine"
 homepage = "https://aleo.org"
diff --git a/ledger/Cargo.toml b/ledger/Cargo.toml
index a8b601767..b79eaf927 100644
--- a/ledger/Cargo.toml
+++ b/ledger/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-ledger"
-version = "4.3.0"
+version = "4.4.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "A node ledger for a decentralized virtual machine"
 homepage = "https://aleo.org"
diff --git a/ledger/authority/Cargo.toml b/ledger/authority/Cargo.toml
index e8419a252..09fa414e7 100644
--- a/ledger/authority/Cargo.toml
+++ b/ledger/authority/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-ledger-authority"
-version = "4.3.0"
+version = "4.4.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Data structures for a block authority in a decentralized virtual machine"
 homepage = "https://aleo.org"
diff --git a/ledger/block/Cargo.toml b/ledger/block/Cargo.toml
index bb6e50f76..9ffa48da8 100644
--- a/ledger/block/Cargo.toml
+++ b/ledger/block/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-ledger-block"
-version = "4.3.0"
+version = "4.4.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "A block for a decentralized virtual machine"
 homepage = "https://aleo.org"
diff --git a/ledger/committee/Cargo.toml b/ledger/committee/Cargo.toml
index 4d3d6174a..1569624ce 100644
--- a/ledger/committee/Cargo.toml
+++ b/ledger/committee/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-ledger-committee"
-version = "4.3.0"
+version = "4.4.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "A committee for a decentralized virtual machine"
 homepage = "https://aleo.org"
diff --git a/ledger/narwhal/Cargo.toml b/ledger/narwhal/Cargo.toml
index 8cb2f2c26..90e422395 100644
--- a/ledger/narwhal/Cargo.toml
+++ b/ledger/narwhal/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-ledger-narwhal"
-version = "4.3.0"
+version = "4.4.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Data structures for a Narwhal-style memory pool in a decentralized virtual machine"
 homepage = "https://aleo.org"
diff --git a/ledger/narwhal/batch-certificate/Cargo.toml b/ledger/narwhal/batch-certificate/Cargo.toml
index 062ea0bca..962ad1e9e 100644
--- a/ledger/narwhal/batch-certificate/Cargo.toml
+++ b/ledger/narwhal/batch-certificate/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-ledger-narwhal-batch-certificate"
-version = "4.3.0"
+version = "4.4.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "A batch certificate for a Narwhal-style memory pool in a decentralized virtual machine"
 homepage = "https://aleo.org"
diff --git a/ledger/narwhal/batch-header/Cargo.toml b/ledger/narwhal/batch-header/Cargo.toml
index 62601ead6..8ad6bbd22 100644
--- a/ledger/narwhal/batch-header/Cargo.toml
+++ b/ledger/narwhal/batch-header/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-ledger-narwhal-batch-header"
-version = "4.3.0"
+version = "4.4.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "A batch header for a Narwhal-style memory pool in a decentralized virtual machine"
 homepage = "https://aleo.org"
diff --git a/ledger/narwhal/data/Cargo.toml b/ledger/narwhal/data/Cargo.toml
index 93cb72b2c..cd43e709c 100644
--- a/ledger/narwhal/data/Cargo.toml
+++ b/ledger/narwhal/data/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-ledger-narwhal-data"
-version = "4.3.0"
+version = "4.4.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "A batch certificate for a Narwhal-style memory pool in a decentralized virtual machine"
 homepage = "https://aleo.org"
diff --git a/ledger/narwhal/subdag/Cargo.toml b/ledger/narwhal/subdag/Cargo.toml
index c05c1c171..9475ce89f 100644
--- a/ledger/narwhal/subdag/Cargo.toml
+++ b/ledger/narwhal/subdag/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-ledger-narwhal-subdag"
-version = "4.3.0"
+version = "4.4.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "A subdag for a Narwhal-style memory pool in a decentralized virtual machine"
 homepage = "https://aleo.org"
diff --git a/ledger/narwhal/transmission-id/Cargo.toml b/ledger/narwhal/transmission-id/Cargo.toml
index e8449a723..10ea11075 100644
--- a/ledger/narwhal/transmission-id/Cargo.toml
+++ b/ledger/narwhal/transmission-id/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-ledger-narwhal-transmission-id"
-version = "4.3.0"
+version = "4.4.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "A transmission ID for a Narwhal-style memory pool in a decentralized virtual machine"
 homepage = "https://aleo.org"
diff --git a/ledger/narwhal/transmission/Cargo.toml b/ledger/narwhal/transmission/Cargo.toml
index 11527dcc5..dad15c22e 100644
--- a/ledger/narwhal/transmission/Cargo.toml
+++ b/ledger/narwhal/transmission/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-ledger-narwhal-transmission"
-version = "4.3.0"
+version = "4.4.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "A transmission for a Narwhal-style memory pool in a decentralized virtual machine"
 homepage = "https://aleo.org"
diff --git a/ledger/puzzle/Cargo.toml b/ledger/puzzle/Cargo.toml
index b767d4384..c3ef091d5 100644
--- a/ledger/puzzle/Cargo.toml
+++ b/ledger/puzzle/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-ledger-puzzle"
-version = "4.3.0"
+version = "4.4.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Puzzle for a decentralized virtual machine"
 homepage = "https://aleo.org"
diff --git a/ledger/puzzle/epoch/Cargo.toml b/ledger/puzzle/epoch/Cargo.toml
index 7a11c9086..5e60ebcd7 100644
--- a/ledger/puzzle/epoch/Cargo.toml
+++ b/ledger/puzzle/epoch/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-ledger-puzzle-epoch"
-version = "4.3.0"
+version = "4.4.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Epoch puzzle for a decentralized virtual machine"
 homepage = "https://aleo.org"
diff --git a/ledger/query/Cargo.toml b/ledger/query/Cargo.toml
index 879cf7c49..b66bad990 100644
--- a/ledger/query/Cargo.toml
+++ b/ledger/query/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-ledger-query"
-version = "4.3.0"
+version = "4.4.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "A query for a decentralized virtual machine"
 homepage = "https://aleo.org"
diff --git a/ledger/src/check_next_block.rs b/ledger/src/check_next_block.rs
index 76119cc29..3d62898bb 100644
--- a/ledger/src/check_next_block.rs
+++ b/ledger/src/check_next_block.rs
@@ -19,25 +19,112 @@ use crate::narwhal::BatchHeader;
 
 use anyhow::{Context, bail};
 
+/// Wrapper for a block that has a valid subDAG, but where the block header,
+/// solutions, and transmissions have not been verified yet.
+///
+/// This type is created by `Ledger::check_block_subdag` and consumed by `Ledger::check_block_content`.
+#[derive(Clone, PartialEq, Eq)]
+pub struct PendingBlock<N: Network>(Block<N>);
+
+impl<N: Network> Deref for PendingBlock<N> {
+    type Target = Block<N>;
+
+    fn deref(&self) -> &Block<N> {
+        &self.0
+    }
+}
+
 impl<N: Network, C: ConsensusStorage<N>> Ledger<N, C> {
-    /// Checks the given block is valid next block.
+    /// Checks that the subDAG in a given block is valid, but does not fully verify the block.
     ///
-    /// # Panics
-    /// This function panics if called from an async context.
-    pub fn check_next_block<R: CryptoRng + Rng>(&self, block: &Block<N>, rng: &mut R) -> Result<()> {
-        let height = block.height();
-        let latest_block = self.latest_block();
+    /// # Arguments
+    /// * `block` - The block to check.
+    /// * `pending_block` - A sequence of blocks between the block to check and the current height of the ledger.
+    ///
+    /// # Notes
+    /// * This does *not* check that the header of the block is correct or execute/verify any of the transmissions contained within it.
+    ///
+    /// * In most cases, you want to use [`Self::check_next_block`] instead to perform a full verification.
+    ///
+    /// * This will reject any blocks with a height <= the current height and any blocks with a height >= the current height + GC.
+    ///   For the former, a valid block already exists and,
+    ///   for the latter, the comittte is still unknown.
+    pub fn check_block_subdag(&self, block: Block<N>, pending_blocks: &[PendingBlock<N>]) -> Result<PendingBlock<N>> {
+        self.check_block_subdag_inner(&block, pending_blocks)?;
+        Ok(PendingBlock(block))
+    }
+
+    fn check_block_subdag_inner(&self, block: &Block<N>, pending_blocks: &[PendingBlock<N>]) -> Result<()> {
+        // First check that the heights and hashes of the pending block sequence and of the new block are correct.
+        // The hash checks should be redundant, but we perform them out of extra caution.
+        let mut expected_height = self.latest_height() + 1;
+        for pending in pending_blocks {
+            if pending.height() != expected_height {
+                bail!(
+                    "Pending block has invalid height. Expected {expected_height}, but got {actual}.",
+                    actual = pending.height()
+                );
+            }
 
-        // Check that this is actually the next block.
-        if height != latest_block.height() + 1 {
-            bail!("Block height is {height}, but expected {}", latest_block.height() + 1);
+            if self.contains_block_hash(&pending.hash())? {
+                bail!("Hash for pending block '{}' already exists in the ledger", block.hash())
+            }
+
+            expected_height += 1;
         }
 
-        // Ensure the block hash does not already exist.
         if self.contains_block_hash(&block.hash())? {
             bail!("Block hash '{}' already exists in the ledger", block.hash())
         }
 
+        if block.height() != expected_height {
+            bail!("Block has invalid height. Expected {expected_height}, but got {}.", block.height());
+        }
+
+        // Ensure the certificates in the block subdag have met quorum requirements.
+        self.check_block_subdag_quorum(block)?;
+
+        // Determine if the block subdag is correctly constructed and is not a combination of multiple subdags.
+        self.check_block_subdag_atomicity(block)?;
+
+        // Ensure that all leaves of the subdag point to valid batches in other subdags/blocks.
+        self.check_block_subdag_leaves(block, pending_blocks)?;
+
+        Ok(())
+    }
+
+    /// Checks the given block is a valid next block with regard to the current state/height of the Ledger.
+    ///
+    /// # Panics
+    /// This function panics if called from an async context.
+    pub fn check_next_block<R: CryptoRng + Rng>(&self, block: &Block<N>, rng: &mut R) -> Result<()> {
+        self.check_block_subdag_inner(block, &[])?;
+        self.check_block_content_inner(block, rng)?;
+
+        Ok(())
+    }
+
+    /// Takes a pending block and performs the remaining checks to full verify it.
+    ///
+    /// # Arguments
+    /// This takes a [`PendingBlock`] as input, which is the output of a successful call to [`Self::check_block_subdag`].
+    /// The latter already verified the block's DAG and certificate signatures.
+    ///
+    /// # Return Value
+    /// This returns a [`Block`] on success representing the fully verified block.
+    ///
+    /// # Panics
+    /// This function panics if called from an async context.
+    pub fn check_block_content<R: CryptoRng + Rng>(&self, block: PendingBlock<N>, rng: &mut R) -> Result<Block<N>> {
+        self.check_block_content_inner(&block.0, rng)?;
+        Ok(block.0)
+    }
+
+    /// # Panics
+    /// This function panics if called from an async context.
+    fn check_block_content_inner<R: CryptoRng + Rng>(&self, block: &Block<N>, rng: &mut R) -> Result<()> {
+        let latest_block = self.latest_block();
+
         // Ensure the solutions do not already exist.
         for solution_id in block.solutions().solution_ids() {
             if self.contains_solution_id(solution_id)? {
@@ -115,15 +202,6 @@ impl<N: Network, C: ConsensusStorage<N>> Ledger<N, C> {
             }
         }
 
-        // Ensure the certificates in the block subdag have met quorum requirements.
-        self.check_block_subdag_quorum(block)?;
-
-        // Determine if the block subdag is correctly constructed and is not a combination of multiple subdags.
-        self.check_block_subdag_atomicity(block)?;
-
-        // Ensure that all leaves of the subdag point to valid batches in other subdags/blocks.
-        self.check_block_subdag_leaves(block)?;
-
         // Ensure that each existing solution ID from the block exists in the ledger.
         for existing_solution_id in expected_existing_solution_ids {
             if !self.contains_solution_id(&existing_solution_id)? {
@@ -145,12 +223,21 @@ impl<N: Network, C: ConsensusStorage<N>> Ledger<N, C> {
     ///
     /// This does not verify that the batches are signed correctly or that the edges are valid
     /// (only point to the previous round), as those checks already happened when the node received the batch.
-    fn check_block_subdag_leaves(&self, block: &Block<N>) -> Result<()> {
+    fn check_block_subdag_leaves(&self, block: &Block<N>, previous_blocks: &[PendingBlock<N>]) -> Result<()> {
         // Check if the block has a subdag.
         let Authority::Quorum(subdag) = block.authority() else {
             return Ok(());
         };
 
+        let previous_certs: HashSet<_> = previous_blocks
+            .iter()
+            .filter_map(|block| match block.authority() {
+                Authority::Quorum(subdag) => Some(subdag.certificate_ids()),
+                Authority::Beacon(_) => None,
+            })
+            .flatten()
+            .collect();
+
         // Store the IDs of all certificates in this subDAG.
         // This allows determining which edges point to other subDAGs/blocks.
         let subdag_certs: HashSet<_> = subdag.certificate_ids().collect();
@@ -173,7 +260,7 @@ impl<N: Network, C: ConsensusStorage<N>> Ledger<N, C> {
             }
 
             // Ensure that the certificate is associated with a previous block.
-            if !self.vm.block_store().contains_block_for_certificate(prev_id)? {
+            if !previous_certs.contains(prev_id) && !self.vm.block_store().contains_block_for_certificate(prev_id)? {
                 bail!(
                     "Batch(es) in the block point(s) to a certificate {prev_id} in round {prev_round} that is not associated with a previous block"
                 )
@@ -184,6 +271,8 @@ impl<N: Network, C: ConsensusStorage<N>> Ledger<N, C> {
     }
 
     /// Check that the certificates in the block subdag have met quorum requirements.
+    ///
+    /// Called by [`Self::check_block_subdag`]
     fn check_block_subdag_quorum(&self, block: &Block<N>) -> Result<()> {
         // Check if the block has a subdag.
         let subdag = match block.authority() {
@@ -226,6 +315,8 @@ impl<N: Network, C: ConsensusStorage<N>> Ledger<N, C> {
     }
 
     /// Checks that the block subdag can not be split into multiple valid subdags.
+    ///
+    /// Called by [`Self::check_block_subdag`]
     fn check_block_subdag_atomicity(&self, block: &Block<N>) -> Result<()> {
         // Returns `true` if there is a path from the previous certificate to the current certificate.
         fn is_linked<N: Network>(
diff --git a/ledger/src/lib.rs b/ledger/src/lib.rs
index a35cb8748..06d25113a 100644
--- a/ledger/src/lib.rs
+++ b/ledger/src/lib.rs
@@ -37,8 +37,10 @@ pub use helpers::*;
 
 pub use crate::block::*;
 
-mod advance;
 mod check_next_block;
+pub use check_next_block::PendingBlock;
+
+mod advance;
 mod check_transaction_basic;
 mod contains;
 mod find;
diff --git a/ledger/store/Cargo.toml b/ledger/store/Cargo.toml
index 311a0c1f5..12ad67fff 100644
--- a/ledger/store/Cargo.toml
+++ b/ledger/store/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-ledger-store"
-version = "4.3.0"
+version = "4.4.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "A data store for a decentralized virtual machine"
 homepage = "https://aleo.org"
diff --git a/ledger/test-helpers/Cargo.toml b/ledger/test-helpers/Cargo.toml
index fe02a8158..623303fc0 100644
--- a/ledger/test-helpers/Cargo.toml
+++ b/ledger/test-helpers/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-ledger-test-helpers"
-version = "4.3.0"
+version = "4.4.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Test helpers for a decentralized virtual machine"
 homepage = "https://aleo.org"
diff --git a/ledger/tests/helpers/mod.rs b/ledger/tests/helpers/mod.rs
new file mode 100644
index 000000000..f20ea88af
--- /dev/null
+++ b/ledger/tests/helpers/mod.rs
@@ -0,0 +1,300 @@
+// Copyright (c) 2019-2025 Provable Inc.
+// This file is part of the snarkVM library.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+
+// http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use aleo_std::StorageMode;
+use snarkvm_console::{
+    account::{Address, PrivateKey},
+    network::MainnetV0,
+    prelude::*,
+};
+use snarkvm_ledger::{Block, Ledger};
+use snarkvm_ledger_narwhal::{BatchCertificate, BatchHeader, Subdag};
+use snarkvm_ledger_store::ConsensusStore;
+use snarkvm_synthesizer::vm::VM;
+
+use indexmap::{IndexMap, IndexSet};
+use std::collections::{BTreeMap, HashMap};
+use time::OffsetDateTime;
+
+pub type CurrentNetwork = MainnetV0;
+
+#[cfg(not(feature = "rocks"))]
+pub type LedgerType<N> = snarkvm_ledger_store::helpers::memory::ConsensusMemory<N>;
+#[cfg(feature = "rocks")]
+pub type LedgerType<N> = snarkvm_ledger_store::helpers::rocksdb::ConsensusDB<N>;
+
+/// Helper to build chains with custom structures for testing.
+pub struct TestChainBuilder {
+    /// The keys of all validators.
+    private_keys: Vec<PrivateKey<CurrentNetwork>>,
+    /// The underlying ledger.
+    ledger: Ledger<CurrentNetwork, LedgerType<CurrentNetwork>>,
+    /// The round containing the leader certificate for the most recent block we generated.
+    last_block_round: u64,
+    /// The batch certificates of the last round we generated.
+    round_to_certificates: HashMap<u64, IndexMap<usize, BatchCertificate<CurrentNetwork>>>,
+    /// The batch certificate of the last leader (if any).
+    previous_leader_certificate: Option<BatchCertificate<CurrentNetwork>>,
+    /// The last round for each committee member where they created a batch.
+    /// Invariant: for any validator i, last_batch[i] <= last_committed_batch[i]
+    last_batch_round: HashMap<usize, u64>,
+    /// The last batch of a validator that was included in a block.
+    last_committed_batch_round: HashMap<usize, u64>,
+    /// The start of the test chain.
+    genesis_block: Block<CurrentNetwork>,
+}
+
+/// Additional options you can pass to the builder when generating blocks.
+#[derive(Default)]
+pub struct BlockOptions {
+    /// Do not include votes to the previous leader certificate
+    pub skip_votes: bool,
+    /// Do not generate certificates for the specific node indices (to simulate a partition).
+    pub skip_nodes: Vec<usize>,
+}
+
+impl TestChainBuilder {
+    pub fn new(committee_size: usize, rng: &mut TestRng) -> Self {
+        // Sample the genesis private key.
+        let private_key = PrivateKey::<CurrentNetwork>::new(rng).unwrap();
+        // Initialize the store.
+        let store = ConsensusStore::<_, LedgerType<_>>::open(StorageMode::new_test(None)).unwrap();
+        // Create a genesis block with a seeded RNG to reproduce the same genesis private keys.
+        let seed: u64 = rng.r#gen();
+        let genesis_rng = &mut TestRng::from_seed(seed);
+        let genesis_block = VM::from(store).unwrap().genesis_beacon(&private_key, genesis_rng).unwrap();
+
+        // Extract the private keys from the genesis committee by using the same RNG to sample private keys.
+        let genesis_rng = &mut TestRng::from_seed(seed);
+        let private_keys = (0..committee_size).map(|_| PrivateKey::new(genesis_rng).unwrap()).collect();
+
+        Self::from_genesis(private_keys, genesis_block)
+    }
+
+    /// Initialize the builder with the specified committee and gensis block
+    pub fn from_genesis(private_keys: Vec<PrivateKey<CurrentNetwork>>, genesis_block: Block<CurrentNetwork>) -> Self {
+        // Initialize the ledger with the genesis block.
+        let ledger = Ledger::<CurrentNetwork, LedgerType<CurrentNetwork>>::load(
+            genesis_block.clone(),
+            StorageMode::new_test(None),
+        )
+        .unwrap();
+
+        Self {
+            private_keys,
+            ledger,
+
+            genesis_block,
+            last_batch_round: Default::default(),
+            last_committed_batch_round: Default::default(),
+            last_block_round: 0,
+            round_to_certificates: Default::default(),
+            previous_leader_certificate: Default::default(),
+        }
+    }
+
+    /// Create multiple blocks, with fully-connected DAGs.
+    #[allow(dead_code)]
+    pub fn generate_blocks(&mut self, num_blocks: usize, rng: &mut TestRng) -> Vec<Block<CurrentNetwork>> {
+        self.generate_blocks_with_opts(num_blocks, &BlockOptions::default(), rng)
+    }
+
+    /// Create multiple blocks, with additional parameters.
+    pub fn generate_blocks_with_opts(
+        &mut self,
+        num_blocks: usize,
+        options: &BlockOptions,
+        rng: &mut TestRng,
+    ) -> Vec<Block<CurrentNetwork>> {
+        assert!(num_blocks > 0, "Need to build at least one block");
+
+        (0..num_blocks).map(|_| self.generate_block_with_opts(options, rng)).collect()
+    }
+
+    /// Create a new block, with a fully-connected DAG.
+    ///
+    /// This will "fill in" any gaps left in earlier rounds from non participating nodes.
+    pub fn generate_block(&mut self, rng: &mut TestRng) -> Block<CurrentNetwork> {
+        self.generate_block_with_opts(&BlockOptions::default(), rng)
+    }
+
+    /// Same as `generate_block` but with additional options/parameters.
+    pub fn generate_block_with_opts(&mut self, options: &BlockOptions, rng: &mut TestRng) -> Block<CurrentNetwork> {
+        assert!(
+            options.skip_nodes.len() * 3 < self.private_keys.len(),
+            "Cannot mark more than f nodes as unavailable/skipped"
+        );
+
+        let next_block_round = self.last_block_round + 2;
+
+        // SubDAGs can be at most GC rounds long.
+        let mut round = if next_block_round < BatchHeader::<CurrentNetwork>::MAX_GC_ROUNDS as u64 {
+            // Batches from genesis round cannot be included in any block that isn't genesis
+            1
+        } else {
+            next_block_round - BatchHeader::<CurrentNetwork>::MAX_GC_ROUNDS as u64
+        };
+
+        // =======================================
+        // Create certificates for the new block.
+        // =======================================
+        loop {
+            let mut created_anchor = false;
+
+            let previous_certificate_ids = if round == 1 {
+                IndexSet::default()
+            } else {
+                self.round_to_certificates
+                    .get(&(round - 1))
+                    .unwrap()
+                    .iter()
+                    .filter_map(|(_, cert)| {
+                        // If votes are skipped, remove previous leader cert from the set.
+                        let skip = if let Some(leader) = &self.previous_leader_certificate {
+                            options.skip_votes && leader.id() == cert.id()
+                        } else {
+                            false
+                        };
+
+                        if skip { None } else { Some(cert.id()) }
+                    })
+                    .collect()
+            };
+
+            let committee = self.ledger.get_committee_lookback_for_round(round).unwrap().unwrap_or_else(|| {
+                panic!("No committee for round {round}");
+            });
+
+            for (key1_idx, private_key_1) in self.private_keys.iter().enumerate() {
+                if options.skip_nodes.contains(&key1_idx) {
+                    continue;
+                }
+                // Don't recreate batches that already exist.
+                if self.last_batch_round.get(&key1_idx).unwrap_or(&0) >= &round {
+                    continue;
+                }
+
+                let batch_header = BatchHeader::new(
+                    private_key_1,
+                    round,
+                    OffsetDateTime::now_utc().unix_timestamp(),
+                    committee.id(),
+                    Default::default(),
+                    previous_certificate_ids.clone(),
+                    rng,
+                )
+                .unwrap();
+
+                // Add signatures for the batch header.
+                let signatures = self
+                    .private_keys
+                    .iter()
+                    .enumerate()
+                    .filter(|&(key2_idx, _)| key1_idx != key2_idx)
+                    .map(|(_, private_key_2)| private_key_2.sign(&[batch_header.batch_id()], rng).unwrap())
+                    .collect();
+
+                // Update the round at which this validator last created a batch.
+                self.last_batch_round.insert(key1_idx, round);
+
+                // Insert certificate into the round_to_certificates mapping.
+                self.round_to_certificates
+                    .entry(round)
+                    .or_default()
+                    .insert(key1_idx, BatchCertificate::from(batch_header, signatures).unwrap());
+
+                // Check if this batch was an anchor.
+                if round % 2 == 0 {
+                    let leader = committee.get_leader(round).unwrap();
+                    if leader == Address::try_from(private_key_1).unwrap() {
+                        created_anchor = true;
+                    }
+                }
+            }
+
+            // Anchor was confirmed by more than a third of the validators.
+            if created_anchor && round % 2 == 0 && self.last_block_round < round {
+                self.last_block_round = round;
+                break;
+            }
+
+            round += 1;
+        }
+
+        // ==============================================================
+        // Build a subdag from the new certificates and create the block.
+        // ==============================================================
+        let commit_round = round;
+
+        let leader_committee = self.ledger.get_committee_lookback_for_round(round).unwrap().unwrap();
+        let leader = leader_committee.get_leader(commit_round).unwrap();
+        let (leader_idx, leader_certificate) =
+            self.round_to_certificates.get(&commit_round).unwrap().iter().find(|(_, c)| c.author() == leader).unwrap();
+        let leader_idx = *leader_idx;
+        let leader_certificate = leader_certificate.clone();
+
+        // Construct the subdag for the new block.
+        let mut subdag_map = BTreeMap::new();
+
+        // Figure out what the earliest round for the subDAG could be.
+        let start_round = if commit_round < BatchHeader::<CurrentNetwork>::MAX_GC_ROUNDS as u64 {
+            1
+        } else {
+            commit_round - BatchHeader::<CurrentNetwork>::MAX_GC_ROUNDS as u64 + 2
+        };
+
+        for round in start_round..commit_round {
+            let mut to_insert = IndexSet::new();
+            for idx in 0..self.private_keys.len() {
+                // Some of the batches we in previous rounds might not be new,
+                // and already included in a previous block.
+                let cround = self.last_committed_batch_round.entry(idx).or_default();
+                // Batch already included in another block
+                if *cround >= round {
+                    continue;
+                }
+
+                if let Some(cert) = self.round_to_certificates.entry(round).or_default().get(&idx) {
+                    to_insert.insert(cert.clone());
+                    *cround = round;
+                }
+            }
+            if !to_insert.is_empty() {
+                subdag_map.insert(round, to_insert);
+            }
+        }
+
+        // Add the leader certificate.
+        // (special case, because it is the only cert included from the commit round)
+        subdag_map.insert(commit_round, [leader_certificate.clone()].into());
+        self.last_committed_batch_round.insert(leader_idx, commit_round);
+
+        // Construct the block.
+        let subdag = Subdag::from(subdag_map).unwrap();
+        let block = self.ledger.prepare_advance_to_next_quorum_block(subdag, Default::default(), rng).unwrap();
+        self.ledger.check_next_block(&block, rng).unwrap();
+
+        // Update th ledger state.
+        self.ledger.advance_to_next_block(&block).unwrap();
+        self.previous_leader_certificate = Some(leader_certificate.clone());
+
+        block
+    }
+
+    /// Return the genesis block associated with the test chain
+    pub fn genesis_block(&self) -> &Block<CurrentNetwork> {
+        &self.genesis_block
+    }
+}
diff --git a/ledger/tests/pending_blocks.rs b/ledger/tests/pending_blocks.rs
new file mode 100644
index 000000000..cb9263e76
--- /dev/null
+++ b/ledger/tests/pending_blocks.rs
@@ -0,0 +1,63 @@
+// Copyright (c) 2019-2025 Provable Inc.
+// This file is part of the snarkVM library.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+
+// http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+mod helpers;
+use helpers::{BlockOptions, CurrentNetwork, LedgerType, TestChainBuilder};
+
+use snarkvm_ledger::Ledger;
+
+use aleo_std::StorageMode;
+use snarkvm_utilities::TestRng;
+
+#[test]
+fn test_preprocess_block() {
+    let rng = &mut TestRng::default();
+
+    let mut builder = TestChainBuilder::new(4, rng);
+
+    // Construct the ledger.
+    let ledger = Ledger::<CurrentNetwork, LedgerType<CurrentNetwork>>::load(
+        builder.genesis_block().clone(),
+        StorageMode::new_test(None),
+    )
+    .unwrap();
+
+    // Generate a bunch of blocks that do not contain votes
+    let mut pending_blocks = vec![];
+
+    for block in builder.generate_blocks_with_opts(5, &BlockOptions { skip_votes: true, ..Default::default() }, rng) {
+        if !pending_blocks.is_empty() {
+            // We shoud only be able to pre-process a pending block if the previous
+            // blocks are applied to the ledger or in the pending set
+            assert!(ledger.check_block_subdag(block.clone(), &[]).is_err());
+        }
+
+        let pending_block = ledger.check_block_subdag(block, &pending_blocks).unwrap();
+        pending_blocks.push(pending_block);
+    }
+
+    // Now, create a "vote block" that contains sufficient votes to the previous leader block
+    let vote_block = builder.generate_block(rng);
+    assert!(ledger.check_next_block(&vote_block, rng).is_err());
+
+    for pending in pending_blocks.into_iter() {
+        let block = ledger.check_block_content(pending, rng).expect("Pending block should be accepted");
+        assert!(ledger.advance_to_next_block(&block).is_ok());
+    }
+
+    // Now the commit block should be accepted
+    assert!(ledger.check_next_block(&vote_block, rng).is_ok());
+    assert!(ledger.advance_to_next_block(&vote_block).is_ok());
+}
diff --git a/metrics/Cargo.toml b/metrics/Cargo.toml
index 20258cbf0..10840e674 100644
--- a/metrics/Cargo.toml
+++ b/metrics/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-metrics"
-version = "4.3.0"
+version = "4.4.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Metrics for a decentralized virtual machine"
 homepage = "https://aleo.org"
diff --git a/parameters/Cargo.toml b/parameters/Cargo.toml
index fec64be14..a7343a5d6 100644
--- a/parameters/Cargo.toml
+++ b/parameters/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-parameters"
-version = "4.3.0"
+version = "4.4.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Parameters for a decentralized virtual machine"
 homepage = "https://aleo.org"
diff --git a/synthesizer/Cargo.toml b/synthesizer/Cargo.toml
index ce156ba46..a26ba5909 100644
--- a/synthesizer/Cargo.toml
+++ b/synthesizer/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-synthesizer"
-version = "4.3.0"
+version = "4.4.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Synthesizer for a decentralized virtual machine"
 homepage = "https://aleo.org"
diff --git a/synthesizer/process/Cargo.toml b/synthesizer/process/Cargo.toml
index 2ba141357..e59958ac2 100644
--- a/synthesizer/process/Cargo.toml
+++ b/synthesizer/process/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-synthesizer-process"
-version = "4.3.0"
+version = "4.4.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "A process for a decentralized virtual machine"
 homepage = "https://aleo.org"
diff --git a/synthesizer/program/Cargo.toml b/synthesizer/program/Cargo.toml
index 3ed229352..28b890136 100644
--- a/synthesizer/program/Cargo.toml
+++ b/synthesizer/program/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-synthesizer-program"
-version = "4.3.0"
+version = "4.4.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Program for a decentralized virtual machine"
 homepage = "https://aleo.org"
@@ -37,7 +37,7 @@ features = [ "account", "network", "program", "types" ]
 
 [dependencies.snarkvm-utilities]
 path = "../../utilities"
-version = "=4.3.0"
+version = "=4.4.0"
 default-features = false
 
 [dependencies.enum-iterator]
diff --git a/synthesizer/snark/Cargo.toml b/synthesizer/snark/Cargo.toml
index a4c535c1a..75503780a 100644
--- a/synthesizer/snark/Cargo.toml
+++ b/synthesizer/snark/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-synthesizer-snark"
-version = "4.3.0"
+version = "4.4.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "SNARK wrappers for a decentralized virtual machine"
 homepage = "https://aleo.org"
diff --git a/utilities/Cargo.toml b/utilities/Cargo.toml
index b47dccee6..ee868ef69 100644
--- a/utilities/Cargo.toml
+++ b/utilities/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-utilities"
-version = "4.3.0"
+version = "4.4.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Utilities for a decentralized virtual machine"
 homepage = "https://aleo.org"
diff --git a/utilities/derives/Cargo.toml b/utilities/derives/Cargo.toml
index 152b8c9e8..e894a3742 100644
--- a/utilities/derives/Cargo.toml
+++ b/utilities/derives/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-utilities-derives"
-version = "4.3.0"
+version = "4.4.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Canonical serialization for a decentralized virtual machine"
 homepage = "https://aleo.org"
diff --git a/wasm/Cargo.toml b/wasm/Cargo.toml
index 839d41c83..4e06022c2 100644
--- a/wasm/Cargo.toml
+++ b/wasm/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkvm-wasm"
-version = "4.3.0"
+version = "4.4.0"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "WASM for a decentralized virtual machine"
 homepage = "https://aleo.org"
