diff --git a/src/node.rs b/src/node.rs
index 28d367fcb..f77b911ad 100644
--- a/src/node.rs
+++ b/src/node.rs
@@ -140,7 +140,7 @@ impl Node {
             }
             _ => {
                 println!("{}", crate::display::welcome_message());
-                println!("Starting a client node on {}.", N::NETWORK_NAME);
+                println!("Starting a {} node on {}.", E::NODE_TYPE, N::NETWORK_NAME);
                 println!("{}", crate::display::notification_message::<N>(None));
                 None
             }
