diff --git a/Cargo.lock b/Cargo.lock
index 478e9f63e..e07e6a4bf 100644
--- a/Cargo.lock
+++ b/Cargo.lock
@@ -700,6 +700,70 @@ version = "0.3.6"
 source = "registry+https://github.com/rust-lang/crates.io-index"
 checksum = "a357d28ed41a50f9c765dbfe56cbc04a64e53e5fc58ba79fbc34c10ef3df831f"
 
+[[package]]
+name = "encoding"
+version = "0.2.33"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "6b0d943856b990d12d3b55b359144ff341533e516d94098b1d3fc1ac666d36ec"
+dependencies = [
+ "encoding-index-japanese",
+ "encoding-index-korean",
+ "encoding-index-simpchinese",
+ "encoding-index-singlebyte",
+ "encoding-index-tradchinese",
+]
+
+[[package]]
+name = "encoding-index-japanese"
+version = "1.20141219.5"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "04e8b2ff42e9a05335dbf8b5c6f7567e5591d0d916ccef4e0b1710d32a0d0c91"
+dependencies = [
+ "encoding_index_tests",
+]
+
+[[package]]
+name = "encoding-index-korean"
+version = "1.20141219.5"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "4dc33fb8e6bcba213fe2f14275f0963fd16f0a02c878e3095ecfdf5bee529d81"
+dependencies = [
+ "encoding_index_tests",
+]
+
+[[package]]
+name = "encoding-index-simpchinese"
+version = "1.20141219.5"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "d87a7194909b9118fc707194baa434a4e3b0fb6a5a757c73c3adb07aa25031f7"
+dependencies = [
+ "encoding_index_tests",
+]
+
+[[package]]
+name = "encoding-index-singlebyte"
+version = "1.20141219.5"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "3351d5acffb224af9ca265f435b859c7c01537c0849754d3db3fdf2bfe2ae84a"
+dependencies = [
+ "encoding_index_tests",
+]
+
+[[package]]
+name = "encoding-index-tradchinese"
+version = "1.20141219.5"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "fd0e20d5688ce3cab59eb3ef3a2083a5c77bf496cb798dc6fcdb75f323890c18"
+dependencies = [
+ "encoding_index_tests",
+]
+
+[[package]]
+name = "encoding_index_tests"
+version = "0.1.4"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "a246d82be1c9d791c5dfde9a2bd045fc3cbba3fa2b11ad558f27d01712f00569"
+
 [[package]]
 name = "encoding_rs"
 version = "0.8.32"
@@ -959,9 +1023,9 @@ checksum = "d2fabcfbdc87f4758337ca535fb41a6d701b65693ce38287d856d1674551ec9b"
 
 [[package]]
 name = "h2"
-version = "0.3.18"
+version = "0.3.19"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "17f8a914c2987b688368b5138aa05321db91f4090cf26118185672ad588bce21"
+checksum = "d357c7ae988e7d2182f7d7871d0b963962420b0678b0997ce7de72001aeab782"
 dependencies = [
  "bytes",
  "fnv",
@@ -2060,9 +2124,9 @@ dependencies = [
 
 [[package]]
 name = "security-framework"
-version = "2.8.2"
+version = "2.9.0"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "a332be01508d814fed64bf28f798a146d73792121129962fdf335bb3c49a4254"
+checksum = "ca2855b3715770894e67cbfa3df957790aa0c9edc3bf06efa1a84d77fa0839d1"
 dependencies = [
  "bitflags",
  "core-foundation",
@@ -2073,9 +2137,9 @@ dependencies = [
 
 [[package]]
 name = "security-framework-sys"
-version = "2.8.0"
+version = "2.9.0"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "31c9bb296072e961fcbd8853511dd39c2d8be2deb1e17c6860b1d30732b323b4"
+checksum = "f51d0c0d83bec45f16480d0ce0058397a69e48fcdc52d1dc8855fb68acbd31a7"
 dependencies = [
  "core-foundation-sys",
  "libc",
@@ -2760,14 +2824,15 @@ dependencies = [
  "cfg-if",
  "colored",
  "curl",
+ "encoding",
  "hex",
  "indexmap",
  "itertools",
+ "js-sys",
  "lazy_static",
  "parking_lot",
  "paste",
  "rand",
- "reqwest",
  "serde_json",
  "sha2",
  "snarkvm-algorithms",
@@ -2777,7 +2842,8 @@ dependencies = [
  "snarkvm-synthesizer",
  "snarkvm-utilities",
  "thiserror",
- "wasm-bindgen-futures",
+ "wasm-bindgen-test",
+ "web-sys",
 ]
 
 [[package]]
@@ -2812,6 +2878,7 @@ dependencies = [
  "paste",
  "rand",
  "rayon",
+ "reqwest",
  "rocksdb",
  "serde",
  "serde_json",
diff --git a/ledger/src/get.rs b/ledger/src/get.rs
index 29c611a90..0d104df1e 100644
--- a/ledger/src/get.rs
+++ b/ledger/src/get.rs
@@ -149,6 +149,15 @@ impl<N: Network, C: ConsensusStorage<N>> Ledger<N, C> {
         }
     }
 
+    /// Returns the confirmed transaction for the given transaction ID.
+    pub fn get_confirmed_transaction(&self, transaction_id: N::TransactionID) -> Result<ConfirmedTransaction<N>> {
+        // Retrieve the confirmed transaction.
+        match self.vm.block_store().get_confirmed_transaction(&transaction_id)? {
+            Some(confirmed_transaction) => Ok(confirmed_transaction),
+            None => bail!("Missing confirmed transaction for ID {transaction_id}"),
+        }
+    }
+
     /// Returns the program for the given program ID.
     pub fn get_program(&self, program_id: ProgramID<N>) -> Result<Program<N>> {
         match self.vm.transaction_store().get_program(&program_id)? {
diff --git a/ledger/src/lib.rs b/ledger/src/lib.rs
index eba65f1ad..dd23840e5 100644
--- a/ledger/src/lib.rs
+++ b/ledger/src/lib.rs
@@ -34,7 +34,7 @@ use console::{
     types::{Field, Group},
 };
 use synthesizer::{
-    block::{Block, Header, Transaction, Transactions},
+    block::{Block, ConfirmedTransaction, Header, Transaction, Transactions},
     coinbase_puzzle::{CoinbaseSolution, EpochChallenge, PuzzleCommitment},
     process::Query,
     program::Program,
diff --git a/parameters/Cargo.toml b/parameters/Cargo.toml
index 070b790ba..cc7275a9c 100644
--- a/parameters/Cargo.toml
+++ b/parameters/Cargo.toml
@@ -50,6 +50,10 @@ version = "1.0"
 version = "2"
 optional = true
 
+[dependencies.encoding]
+version = "0.2"
+optional = true
+
 [dependencies.hex]
 version = "0.4.3"
 
@@ -59,6 +63,10 @@ version = "1.9"
 [dependencies.itertools]
 version = "0.10.3"
 
+[dependencies.js-sys]
+version = "0.3.61"
+optional = true
+
 [dependencies.lazy_static]
 version = "1.4"
 
@@ -71,10 +79,6 @@ version = "1"
 [dependencies.rand]
 version = "0.8"
 
-[dependencies.reqwest]
-version = "0.11"
-optional = true
-
 [dependencies.serde_json]
 version = "1"
 features = [ "preserve_order" ]
@@ -86,8 +90,9 @@ default-features = false
 [dependencies.thiserror]
 version = "1.0"
 
-[dependencies.wasm-bindgen-futures]
-version = "0.4"
+[dependencies.web-sys]
+version = "0.3.61"
+features = [ "XmlHttpRequest" ]
 optional = true
 
 [target."cfg(not(target_family = \"wasm\"))".dependencies.curl]
@@ -107,8 +112,8 @@ path = "../console"
 [dev-dependencies.snarkvm-synthesizer]
 path = "../synthesizer"
 
-[dev-dependencies.curl]
-version = "0.4.34"
+[dev-dependencies.wasm-bindgen-test]
+version = "0.3.34"
 
 [dev-dependencies.hex]
 version = "0.4.3"
@@ -120,4 +125,4 @@ version = "0.8"
 default = [ "colored", "remote" ]
 no_std_out = [ ]
 remote = [ "curl" ]
-wasm = [ "reqwest", "wasm-bindgen-futures" ]
+wasm = [ "encoding", "js-sys", "web-sys" ]
diff --git a/parameters/src/errors/parameter.rs b/parameters/src/errors/parameter.rs
index 4a022be05..2fcfe8845 100644
--- a/parameters/src/errors/parameter.rs
+++ b/parameters/src/errors/parameter.rs
@@ -32,6 +32,9 @@ pub enum ParameterError {
 
     #[error("Expected size of {}, found size of {}", _0, _1)]
     SizeMismatch(usize, usize),
+
+    #[error("{}", _0)]
+    Wasm(String),
 }
 
 #[cfg(not(feature = "wasm"))]
@@ -41,13 +44,6 @@ impl From<curl::Error> for ParameterError {
     }
 }
 
-#[cfg(feature = "wasm")]
-impl From<reqwest::Error> for ParameterError {
-    fn from(error: reqwest::Error) -> Self {
-        ParameterError::Crate("request::error", format!("{error:?}"))
-    }
-}
-
 impl From<std::io::Error> for ParameterError {
     fn from(error: std::io::Error) -> Self {
         ParameterError::Crate("std::io", format!("{error:?}"))
diff --git a/parameters/src/macros.rs b/parameters/src/macros.rs
index 9c7c8b03f..41d4a9f7f 100644
--- a/parameters/src/macros.rs
+++ b/parameters/src/macros.rs
@@ -90,20 +90,48 @@ macro_rules! impl_store_and_remote_fetch {
         }
 
         #[cfg(feature = "wasm")]
-        fn remote_fetch(
-            buffer: alloc::sync::Weak<parking_lot::RwLock<Vec<u8>>>,
-            url: &'static str,
-        ) -> Result<(), $crate::errors::ParameterError> {
-            // NOTE(julesdesmit): We spawn a local thread here in order to be
-            // able to accommodate the async syntax from reqwest.
-            wasm_bindgen_futures::spawn_local(async move {
-                let content = reqwest::get(url).await.unwrap().text().await.unwrap();
-
-                let buffer = buffer.upgrade().unwrap();
-                buffer.write().extend_from_slice(content.as_bytes());
-                drop(buffer);
-            });
-            Ok(())
+        fn remote_fetch(url: &str) -> Result<Vec<u8>, $crate::errors::ParameterError> {
+            // Use the browser's XmlHttpRequest object to download the parameter file synchronously.
+            //
+            // This method blocks the event loop while the parameters are downloaded, and should be
+            // executed in a web worker to prevent the main browser window from freezing.
+            let xhr = web_sys::XmlHttpRequest::new().map_err(|_| {
+                $crate::errors::ParameterError::Wasm("Download failed - XMLHttpRequest object not found".to_string())
+            })?;
+
+            // XmlHttpRequest if specified as synchronous cannot use the responseType property. It
+            // cannot thus download bytes directly and enforces a text encoding. To get back the
+            // original binary, a charset that does not corrupt the original bytes must be used.
+            xhr.override_mime_type("octet/binary; charset=ISO-8859-5").unwrap();
+
+            // Initialize and send the request.
+            xhr.open_with_async("GET", url, false).map_err(|_| {
+                $crate::errors::ParameterError::Wasm(
+                    "Download failed - This browser does not support synchronous requests".to_string(),
+                )
+            })?;
+            xhr.send().map_err(|_| {
+                $crate::errors::ParameterError::Wasm("Download failed - XMLHttpRequest failed".to_string())
+            })?;
+
+            // Wait for the response in a blocking fashion.
+            if xhr.response().is_ok() && xhr.status().unwrap() == 200 {
+                // Get the text from the response.
+                let rust_text = xhr
+                    .response_text()
+                    .map_err(|_| $crate::errors::ParameterError::Wasm("XMLHttpRequest failed".to_string()))?
+                    .ok_or($crate::errors::ParameterError::Wasm(
+                        "The request was successful but no parameters were received".to_string(),
+                    ))?;
+
+                // Re-encode the text back into bytes using the chosen encoding.
+                use encoding::Encoding;
+                encoding::all::ISO_8859_5
+                    .encode(&rust_text, encoding::EncoderTrap::Strict)
+                    .map_err(|_| $crate::errors::ParameterError::Wasm("Parameter decoding failed".to_string()))
+            } else {
+                Err($crate::errors::ParameterError::Wasm("Download failed - XMLHttpRequest failed".to_string()))
+            }
         }
     };
 }
@@ -173,19 +201,7 @@ macro_rules! impl_load_bytes_logic_remote {
                         }
                     }
                 } else if #[cfg(feature = "wasm")] {
-                    let buffer = alloc::sync::Arc::new(parking_lot::RwLock::new(vec![]));
-
-                    // NOTE(julesdesmit): I'm leaking memory here so that I can get a
-                    // static reference to the url, which is needed to pass it into
-                    // the local thread which downloads the file.
-                    let url = Box::leak(url.into_boxed_str());
-
-                    let buffer_clone = alloc::sync::Arc::downgrade(&buffer);
-                    Self::remote_fetch(buffer_clone, url)?;
-
-                    // Recover the bytes.
-                    let buffer = alloc::sync::Arc::try_unwrap(buffer).unwrap();
-                    let buffer = buffer.write().clone();
+                    let buffer = Self::remote_fetch(&url)?;
 
                     // Ensure the checksum matches.
                     let candidate_checksum = checksum!(&buffer);
diff --git a/parameters/src/testnet3/mod.rs b/parameters/src/testnet3/mod.rs
index b1f8165aa..b94d29b60 100644
--- a/parameters/src/testnet3/mod.rs
+++ b/parameters/src/testnet3/mod.rs
@@ -115,3 +115,23 @@ lazy_static! {
     pub static ref INCLUSION_VERIFYING_KEY: Vec<u8> =
         InclusionVerifier::load_bytes().expect("Failed to load inclusion verifying key");
 }
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use wasm_bindgen_test::*;
+    wasm_bindgen_test_configure!(run_in_browser);
+
+    #[wasm_bindgen_test]
+    fn test_transfer_flow() {
+        Degree16::load_bytes().expect("Failed to load degree 16");
+        Degree17::load_bytes().expect("Failed to load degree 17");
+        Degree18::load_bytes().expect("Failed to load degree 18");
+        Degree19::load_bytes().expect("Failed to load degree 19");
+        TransferVerifier::load_bytes().expect("Failed to load transfer verifier");
+        FeeProver::load_bytes().expect("Failed to load fee prover");
+        FeeVerifier::load_bytes().expect("Failed to load fee verifier");
+        InclusionProver::load_bytes().expect("Failed to load inclusion prover");
+        InclusionVerifier::load_bytes().expect("Failed to load inclusion verifier");
+    }
+}
diff --git a/synthesizer/Cargo.toml b/synthesizer/Cargo.toml
index e072fff6a..8993a1aae 100644
--- a/synthesizer/Cargo.toml
+++ b/synthesizer/Cargo.toml
@@ -53,7 +53,11 @@ serial = [
 ]
 setup = [ ]
 timer = [ "aleo-std/timer" ]
-wasm = [ ]
+wasm = [
+  "console/wasm",
+  "snarkvm-algorithms/wasm",
+  "snarkvm-utilities/wasm"
+]
 
 [dependencies.circuit]
 package = "snarkvm-circuit"
@@ -127,6 +131,10 @@ default-features = false
 features = [ "lz4" ]
 optional = true
 
+[dependencies.reqwest]
+version = "0.11.16"
+features = [ "json" ]
+
 [dependencies.serde]
 version = "1.0"
 
diff --git a/synthesizer/src/process/mod.rs b/synthesizer/src/process/mod.rs
index 5fa5e9a84..076763f00 100644
--- a/synthesizer/src/process/mod.rs
+++ b/synthesizer/src/process/mod.rs
@@ -159,6 +159,26 @@ impl<N: Network> Process<N> {
         Ok(process)
     }
 
+    /// Initializes a new process without downloading the 'credits.aleo' circuit keys (for web contexts).
+    #[inline]
+    #[cfg(feature = "wasm")]
+    pub fn load_web() -> Result<Self> {
+        // Initialize the process.
+        let mut process = Self { universal_srs: Arc::new(UniversalSRS::load()?), stacks: IndexMap::new() };
+
+        // Initialize the 'credits.aleo' program.
+        let program = Program::credits()?;
+
+        // Compute the 'credits.aleo' program stack.
+        let stack = Stack::new(&process, &program)?;
+
+        // Add the stack to the process.
+        process.add_stack(stack);
+
+        // Return the process.
+        Ok(process)
+    }
+
     /// Initializes a new process with a cache of previously used keys. This version is suitable for tests
     /// (which often use nested loops that keep reusing those), as their deserialization is slow.
     #[cfg(test)]
diff --git a/synthesizer/src/process/stack/inclusion/execute.rs b/synthesizer/src/process/stack/inclusion/execute.rs
new file mode 100644
index 000000000..c2d3ec75d
--- /dev/null
+++ b/synthesizer/src/process/stack/inclusion/execute.rs
@@ -0,0 +1,255 @@
+// Copyright (C) 2019-2023 Aleo Systems Inc.
+// This file is part of the snarkVM library.
+
+// The snarkVM library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// The snarkVM library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with the snarkVM library. If not, see <https://www.gnu.org/licenses/>.
+
+use super::*;
+
+macro_rules! prepare_execution_impl {
+    ($self:ident, $execution:ident, $query:ident, $current_state_root:ident, $get_state_path_for_commitment:ident $(, $await:ident)?) => {{
+        // Ensure the number of leaves is within the Merkle tree size.
+        Transaction::check_execution_size($execution)?;
+
+        // Ensure the inclusion proof in the execution is 'None'.
+        if $execution.inclusion_proof().is_some() {
+            bail!("Inclusion proof in the execution should not be set in 'Inclusion::prepare_execution'")
+        }
+
+        // Initialize an empty transaction tree.
+        let mut transaction_tree = N::merkle_tree_bhp::<TRANSACTION_DEPTH>(&[])?;
+        // Initialize a vector for the assignments.
+        let mut assignments = vec![];
+
+        // Retrieve the global state root.
+        let global_state_root = {
+            $query.$current_state_root()
+            $(.$await)?
+        }?;
+
+        // Ensure the global state root is not zero.
+        if *global_state_root == Field::zero() {
+            bail!("Inclusion expected the global state root in the execution to *not* be zero")
+        }
+
+        for (transition_index, transition) in $execution.transitions().enumerate() {
+            // Construct the transaction leaf.
+            let transaction_leaf = TransactionLeaf::new_execution(transition_index as u16, **transition.id());
+
+            // Process the input tasks.
+            match $self.input_tasks.get(transition.id()) {
+                Some(tasks) => {
+                    for task in tasks {
+                        // Retrieve the local state root.
+                        let local_state_root = (*transaction_tree.root()).into();
+
+                        // Construct the state path.
+                        let state_path = match task.is_local {
+                            true => {
+                                // Compute the transaction path.
+                                let transaction_path =
+                                    transaction_tree.prove(transition_index, &transaction_leaf.to_bits_le())?;
+                                // Compute the transition leaf.
+                                let transition_leaf = task.leaf;
+                                // Compute the transition path.
+                                let transition_path = transition.to_path(&transition_leaf)?;
+                                // Output the state path.
+                                StatePath::new_local(
+                                    global_state_root,
+                                    local_state_root,
+                                    transaction_path,
+                                    transaction_leaf,
+                                    transition_path,
+                                    transition_leaf,
+                                )?
+                            }
+                            false => {
+                                $query.$get_state_path_for_commitment(&task.commitment)
+                                $(.$await)?
+                            }?
+                        };
+
+                        // Ensure the global state root is the same across iterations.
+                        if global_state_root != state_path.global_state_root() {
+                            bail!("Inclusion expected the global state root to be the same across iterations")
+                        }
+
+                        // Construct the assignment for the state path.
+                        let assignment = InclusionAssignment::new(
+                            state_path,
+                            task.commitment,
+                            task.gamma,
+                            task.serial_number,
+                            local_state_root,
+                            !task.is_local,
+                        );
+
+                        // Add the assignment to the assignments.
+                        assignments.push(assignment);
+                    }
+                }
+                None => bail!("Missing input tasks for transition {} in inclusion", transition.id()),
+            }
+
+            // Insert the leaf into the transaction tree.
+            transaction_tree.append(&[transaction_leaf.to_bits_le()])?;
+        }
+
+        Ok((assignments, global_state_root))
+    }};
+}
+
+impl<N: Network> Inclusion<N> {
+    /// Returns the inclusion assignments for the given execution.
+    pub fn prepare_execution<B: BlockStorage<N>, Q: Into<Query<N, B>>>(
+        &self,
+        execution: &Execution<N>,
+        query: Q,
+    ) -> Result<(Vec<InclusionAssignment<N>>, N::StateRoot)> {
+        let query = query.into();
+        prepare_execution_impl!(self, execution, query, current_state_root, get_state_path_for_commitment)
+    }
+
+    /// Returns the inclusion assignments for the given execution.
+    pub async fn prepare_execution_async<B: BlockStorage<N>, Q: Into<Query<N, B>>>(
+        &self,
+        execution: &Execution<N>,
+        query: Q,
+    ) -> Result<(Vec<InclusionAssignment<N>>, N::StateRoot)> {
+        let query = query.into();
+        prepare_execution_impl!(self, execution, query, current_state_root_async, get_state_path_for_commitment_async, await)
+    }
+}
+
+impl<N: Network> Inclusion<N> {
+    /// Returns a new execution with an inclusion proof, for the given execution.
+    pub fn prove_execution<A: circuit::Aleo<Network = N>, R: Rng + CryptoRng>(
+        &self,
+        execution: Execution<N>,
+        assignments: &[InclusionAssignment<N>],
+        global_state_root: N::StateRoot,
+        rng: &mut R,
+    ) -> Result<Execution<N>> {
+        match assignments.is_empty() {
+            true => {
+                // Ensure the global state root is not zero.
+                if *global_state_root == Field::zero() {
+                    bail!("Inclusion expected the global state root in the execution to *not* be zero")
+                }
+
+                // Ensure the inclusion proof in the execution is 'None'.
+                if execution.inclusion_proof().is_some() {
+                    bail!("Inclusion expected the inclusion proof in the execution to be 'None'")
+                }
+                // Return the execution.
+                Execution::from(execution.into_transitions(), global_state_root, None)
+            }
+            false => {
+                // Fetch the inclusion proving key.
+                let proving_key = ProvingKey::<N>::new(N::inclusion_proving_key().clone());
+
+                // Compute the inclusion batch proof.
+                let (global_state_root, inclusion_proof) = Self::prove_batch::<A, R>(&proving_key, assignments, rng)?;
+                // Return the execution.
+                Execution::from(execution.into_transitions(), global_state_root, Some(inclusion_proof))
+            }
+        }
+    }
+
+    /// Checks the inclusion proof for the execution.
+    /// Note: This does *not* check that the global state root exists in the ledger.
+    pub fn verify_execution(execution: &Execution<N>) -> Result<()> {
+        // Retrieve the global state root.
+        let global_state_root = execution.global_state_root();
+
+        // Retrieve the inclusion proof.
+        let inclusion_proof = execution.inclusion_proof();
+
+        // Initialize an empty transaction tree.
+        let mut transaction_tree = N::merkle_tree_bhp::<TRANSACTION_DEPTH>(&[])?;
+        // Initialize a vector for the batch verifier inputs.
+        let mut batch_verifier_inputs = vec![];
+
+        // Construct the batch verifier inputs.
+        for (transition_index, transition) in execution.transitions().enumerate() {
+            // Retrieve the local state root.
+            let local_state_root = *transaction_tree.root();
+
+            // Iterate through the inputs.
+            for input in transition.inputs() {
+                // Filter the inputs for records.
+                if let Input::Record(serial_number, _) = input {
+                    // Add the public inputs to the batch verifier inputs.
+                    batch_verifier_inputs.push(vec![
+                        N::Field::one(),
+                        **global_state_root,
+                        *local_state_root,
+                        **serial_number,
+                    ]);
+                }
+            }
+
+            // Construct the transaction leaf.
+            let transaction_leaf = TransactionLeaf::new_execution(transition_index as u16, **transition.id());
+            // Insert the leaf into the transaction tree.
+            transaction_tree.append(&[transaction_leaf.to_bits_le()])?;
+        }
+
+        // If there are no batch verifier inputs, then ensure the inclusion proof is 'None'.
+        if batch_verifier_inputs.is_empty() && inclusion_proof.is_some() {
+            bail!("No input records in the execution. Expected the inclusion proof to be 'None'")
+        }
+        // If there are batch verifier inputs, then ensure the inclusion proof is 'Some'.
+        if !batch_verifier_inputs.is_empty() && inclusion_proof.is_none() {
+            bail!("Missing inclusion proof for the execution")
+        }
+
+        // Verify the inclusion proof.
+        if let Some(inclusion_proof) = inclusion_proof {
+            // Ensure the global state root is not zero.
+            if *global_state_root == Field::zero() {
+                bail!("Inclusion expected the global state root in the execution to *not* be zero")
+            }
+
+            // Fetch the inclusion verifying key.
+            let verifying_key = VerifyingKey::<N>::new(N::inclusion_verifying_key().clone());
+            // Verify the inclusion proof.
+            ensure!(
+                verifying_key.verify_batch(N::INCLUSION_FUNCTION_NAME, &batch_verifier_inputs, inclusion_proof),
+                "Inclusion proof is invalid"
+            );
+        }
+
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use snarkvm_utilities::TestRng;
+
+    #[test]
+    fn test_inclusion_verify_execution() {
+        let rng = &mut TestRng::default();
+        // Fetch an execution transaction.
+        let execution_transaction = crate::vm::test_helpers::sample_execution_transaction_with_fee(rng);
+
+        match execution_transaction {
+            Transaction::Execute(_, execution, _) => {
+                assert!(Inclusion::verify_execution(&execution).is_ok());
+            }
+            _ => panic!("Expected an execution transaction"),
+        }
+    }
+}
diff --git a/synthesizer/src/process/stack/inclusion/fee.rs b/synthesizer/src/process/stack/inclusion/fee.rs
new file mode 100644
index 000000000..c9a78dd6f
--- /dev/null
+++ b/synthesizer/src/process/stack/inclusion/fee.rs
@@ -0,0 +1,209 @@
+// Copyright (C) 2019-2023 Aleo Systems Inc.
+// This file is part of the snarkVM library.
+
+// The snarkVM library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// The snarkVM library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with the snarkVM library. If not, see <https://www.gnu.org/licenses/>.
+
+use super::*;
+
+macro_rules! prepare_fee_impl {
+    ($self:ident, $fee_transition:ident, $query:ident, $get_state_path_for_commitment:ident $(, $await:ident)?) => {{
+        // Ensure the fee has the correct program ID.
+        let fee_program_id = ProgramID::from_str("credits.aleo")?;
+        ensure!(*$fee_transition.program_id() == fee_program_id, "Incorrect program ID for fee");
+
+        // Ensure the fee has the correct function.
+        let fee_function = Identifier::from_str("fee")?;
+        ensure!(*$fee_transition.function_name() == fee_function, "Incorrect function name for fee");
+
+        // Initialize an empty transaction tree.
+        let transaction_tree = N::merkle_tree_bhp::<TRANSACTION_DEPTH>(&[])?;
+        // Initialize the global state root.
+        let mut global_state_root = N::StateRoot::default();
+        // Initialize a vector for the assignments.
+        let mut assignments = vec![];
+
+        // Process the input tasks.
+        match $self.input_tasks.get($fee_transition.id()) {
+            Some(tasks) => {
+                for task in tasks {
+                    // Retrieve the local state root.
+                    let local_state_root = (*transaction_tree.root()).into();
+                    // Construct the state path.
+                    let state_path = {
+                        $query.$get_state_path_for_commitment(&task.commitment)
+                        $(.$await)?
+                    }?;
+
+                    // Ensure the global state root is the same across iterations.
+                    if *global_state_root != Field::zero() && global_state_root != state_path.global_state_root() {
+                        bail!("Inclusion expected the global state root to be the same across iterations")
+                    }
+
+                    // Update the global state root.
+                    global_state_root = state_path.global_state_root();
+
+                    // Prepare the assignment for the state path.
+                    let assignment = InclusionAssignment::new(
+                        state_path,
+                        task.commitment,
+                        task.gamma,
+                        task.serial_number,
+                        local_state_root,
+                        !task.is_local,
+                    );
+
+                    // Add the assignment to the assignments.
+                    assignments.push(assignment);
+                }
+            }
+            None => bail!("Missing input state for fee transition {} in inclusion", $fee_transition.id()),
+        }
+
+        // Ensure the global state root is not zero.
+        if *global_state_root == Field::zero() {
+            bail!("Inclusion expected the global state root in the fee to *not* be zero")
+        }
+        // Ensure the assignments are not empty.
+        if assignments.is_empty() {
+            bail!("Inclusion expected the assignments for the fee to *not* be empty")
+        }
+        // Return the assignments.
+        Ok(assignments)
+    }};
+}
+
+impl<N: Network> Inclusion<N> {
+    /// Returns the inclusion assignments for the given fee transition.
+    pub fn prepare_fee<B: BlockStorage<N>, Q: Into<Query<N, B>>>(
+        &self,
+        fee_transition: &Transition<N>,
+        query: Q,
+    ) -> Result<Vec<InclusionAssignment<N>>> {
+        let query = query.into();
+        prepare_fee_impl!(self, fee_transition, query, get_state_path_for_commitment)
+    }
+
+    /// Returns the inclusion assignments for the given fee transition.
+    pub async fn prepare_fee_async<B: BlockStorage<N>, Q: Into<Query<N, B>>>(
+        &self,
+        fee_transition: &Transition<N>,
+        query: Q,
+    ) -> Result<Vec<InclusionAssignment<N>>> {
+        let query = query.into();
+        prepare_fee_impl!(self, fee_transition, query, get_state_path_for_commitment_async, await)
+    }
+
+    /// Returns a new fee with an inclusion proof, for the given transition.
+    pub fn prove_fee<A: circuit::Aleo<Network = N>, R: Rng + CryptoRng>(
+        &self,
+        fee_transition: Transition<N>,
+        assignments: &[InclusionAssignment<N>],
+        rng: &mut R,
+    ) -> Result<Fee<N>> {
+        // Ensure the fee has the correct program ID.
+        let fee_program_id = ProgramID::from_str("credits.aleo")?;
+        ensure!(*fee_transition.program_id() == fee_program_id, "Incorrect program ID for fee");
+
+        // Ensure the fee has the correct function.
+        let fee_function = Identifier::from_str("fee")?;
+        ensure!(*fee_transition.function_name() == fee_function, "Incorrect function name for fee");
+
+        // Ensure the assignments are not empty.
+        if assignments.is_empty() {
+            bail!("Inclusion expected the assignments for the fee to *not* be empty")
+        }
+
+        // Fetch the inclusion proving key.
+        let proving_key = ProvingKey::<N>::new(N::inclusion_proving_key().clone());
+
+        // Compute the inclusion batch proof.
+        let (global_state_root, inclusion_proof) = Self::prove_batch::<A, R>(&proving_key, assignments, rng)?;
+        // Return the fee.
+        Ok(Fee::from(fee_transition, global_state_root, Some(inclusion_proof)))
+    }
+
+    /// Checks the inclusion proof for the fee.
+    /// Note: This does *not* check that the global state root exists in the ledger.
+    pub fn verify_fee(fee: &Fee<N>) -> Result<()> {
+        // Retrieve the global state root.
+        let global_state_root = fee.global_state_root();
+        // Ensure the global state root is not zero.
+        if *global_state_root == Field::zero() {
+            bail!("Inclusion expected the global state root in the fee to *not* be zero")
+        }
+
+        // Retrieve the inclusion proof.
+        let inclusion_proof = match fee.inclusion_proof() {
+            Some(inclusion_proof) => inclusion_proof,
+            None => bail!("Inclusion expected the fee to contain an inclusion proof"),
+        };
+
+        // Initialize an empty transaction tree.
+        let transaction_tree = N::merkle_tree_bhp::<TRANSACTION_DEPTH>(&[])?;
+        // Initialize a vector for the batch verifier inputs.
+        let mut batch_verifier_inputs = vec![];
+
+        // Retrieve the local state root.
+        let local_state_root = *transaction_tree.root();
+
+        // Construct the batch verifier inputs.
+        for input in fee.inputs() {
+            // Filter the inputs for records.
+            if let Input::Record(serial_number, _) = input {
+                // Add the public inputs to the batch verifier inputs.
+                batch_verifier_inputs.push(vec![
+                    N::Field::one(),
+                    **global_state_root,
+                    *local_state_root,
+                    **serial_number,
+                ]);
+            }
+        }
+
+        // Ensure there are batch verifier inputs.
+        if batch_verifier_inputs.is_empty() {
+            bail!("Inclusion expected the fee to contain input records")
+        }
+
+        // Fetch the inclusion verifying key.
+        let verifying_key = VerifyingKey::<N>::new(N::inclusion_verifying_key().clone());
+        // Verify the inclusion proof.
+        ensure!(
+            verifying_key.verify_batch(N::INCLUSION_FUNCTION_NAME, &batch_verifier_inputs, inclusion_proof),
+            "Inclusion proof is invalid"
+        );
+
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use snarkvm_utilities::TestRng;
+
+    #[test]
+    fn test_inclusion_verify_fee() {
+        let rng = &mut TestRng::default();
+        // Fetch a deployment transaction.
+        let deployment_transaction = crate::vm::test_helpers::sample_deployment_transaction(rng);
+
+        match deployment_transaction {
+            Transaction::Deploy(_, _, _, fee) => {
+                assert!(Inclusion::verify_fee(&fee).is_ok());
+            }
+            _ => panic!("Expected a deployment transaction"),
+        }
+    }
+}
diff --git a/synthesizer/src/process/stack/inclusion/mod.rs b/synthesizer/src/process/stack/inclusion/mod.rs
index 0be9b2cb2..95d1f3078 100644
--- a/synthesizer/src/process/stack/inclusion/mod.rs
+++ b/synthesizer/src/process/stack/inclusion/mod.rs
@@ -14,18 +14,20 @@
 // You should have received a copy of the GNU General Public License
 // along with the snarkVM library. If not, see <https://www.gnu.org/licenses/>.
 
+mod execute;
+mod fee;
+
 #[cfg(debug_assertions)]
 use crate::Stack;
 use crate::{
     BlockStorage,
-    BlockStore,
     Execution,
     Fee,
     Input,
     Output,
-    Program,
     Proof,
     ProvingKey,
+    Query,
     Transaction,
     Transition,
     VerifyingKey,
@@ -38,97 +40,6 @@ use console::{
 
 use std::collections::HashMap;
 
-#[derive(Clone)]
-pub enum Query<N: Network, B: BlockStorage<N>> {
-    /// The block store from the VM.
-    VM(BlockStore<N, B>),
-    /// The base URL of the node.
-    REST(String),
-}
-
-impl<N: Network, B: BlockStorage<N>> From<BlockStore<N, B>> for Query<N, B> {
-    fn from(block_store: BlockStore<N, B>) -> Self {
-        Self::VM(block_store)
-    }
-}
-
-impl<N: Network, B: BlockStorage<N>> From<&BlockStore<N, B>> for Query<N, B> {
-    fn from(block_store: &BlockStore<N, B>) -> Self {
-        Self::VM(block_store.clone())
-    }
-}
-
-impl<N: Network, B: BlockStorage<N>> From<String> for Query<N, B> {
-    fn from(url: String) -> Self {
-        Self::REST(url)
-    }
-}
-
-impl<N: Network, B: BlockStorage<N>> From<&String> for Query<N, B> {
-    fn from(url: &String) -> Self {
-        Self::REST(url.to_string())
-    }
-}
-
-impl<N: Network, B: BlockStorage<N>> From<&str> for Query<N, B> {
-    fn from(url: &str) -> Self {
-        Self::REST(url.to_string())
-    }
-}
-
-impl<N: Network, B: BlockStorage<N>> Query<N, B> {
-    /// Returns the program for the given program ID.
-    pub fn get_program(&self, program_id: &ProgramID<N>) -> Result<Program<N>> {
-        match self {
-            Self::VM(block_store) => {
-                block_store.get_program(program_id)?.ok_or_else(|| anyhow!("Program {program_id} not found in storage"))
-            }
-            #[cfg(not(feature = "wasm"))]
-            Self::REST(url) => match N::ID {
-                3 => Ok(Self::get_request(&format!("{url}/testnet3/program/{program_id}"))?.into_json()?),
-                _ => bail!("Unsupported network ID in inclusion query"),
-            },
-            #[cfg(feature = "wasm")]
-            _ => bail!("External API calls not supported from WASM"),
-        }
-    }
-
-    /// Returns the current state root.
-    pub fn current_state_root(&self) -> Result<N::StateRoot> {
-        match self {
-            Self::VM(block_store) => Ok(block_store.current_state_root()),
-            #[cfg(not(feature = "wasm"))]
-            Self::REST(url) => match N::ID {
-                3 => Ok(Self::get_request(&format!("{url}/testnet3/latest/stateRoot"))?.into_json()?),
-                _ => bail!("Unsupported network ID in inclusion query"),
-            },
-            #[cfg(feature = "wasm")]
-            _ => bail!("External API calls not supported from WASM"),
-        }
-    }
-
-    /// Returns a state path for the given `commitment`.
-    pub fn get_state_path_for_commitment(&self, commitment: &Field<N>) -> Result<StatePath<N>> {
-        match self {
-            Self::VM(block_store) => block_store.get_state_path_for_commitment(commitment),
-            #[cfg(not(feature = "wasm"))]
-            Self::REST(url) => match N::ID {
-                3 => Ok(Self::get_request(&format!("{url}/testnet3/statePath/{commitment}"))?.into_json()?),
-                _ => bail!("Unsupported network ID in inclusion query"),
-            },
-            #[cfg(feature = "wasm")]
-            _ => bail!("External API calls not supported from WASM"),
-        }
-    }
-
-    /// Performs a GET request to the given URL.
-    #[cfg(not(feature = "wasm"))]
-    fn get_request(url: &str) -> Result<ureq::Response> {
-        let response = ureq::get(url).call()?;
-        if response.status() == 200 { Ok(response) } else { bail!("Failed to fetch from {}", url) }
-    }
-}
-
 #[derive(Clone, Debug)]
 struct InputTask<N: Network> {
     /// The commitment.
@@ -194,351 +105,6 @@ impl<N: Network> Inclusion<N> {
         Ok(())
     }
 
-    /// Returns the inclusion assignments for the given execution.
-    pub fn prepare_execution<B: BlockStorage<N>, Q: Into<Query<N, B>>>(
-        &self,
-        execution: &Execution<N>,
-        query: Q,
-    ) -> Result<(Vec<InclusionAssignment<N>>, N::StateRoot)> {
-        let query = query.into();
-
-        // Ensure the number of leaves is within the Merkle tree size.
-        Transaction::check_execution_size(execution)?;
-
-        // Ensure the inclusion proof in the execution is 'None'.
-        if execution.inclusion_proof().is_some() {
-            bail!("Inclusion proof in the execution should not be set in 'Inclusion::prepare_execution'")
-        }
-
-        // Initialize an empty transaction tree.
-        let mut transaction_tree = N::merkle_tree_bhp::<TRANSACTION_DEPTH>(&[])?;
-        // Initialize a vector for the assignments.
-        let mut assignments = vec![];
-
-        // Retrieve the global state root.
-        let global_state_root = query.current_state_root()?;
-
-        // Ensure the global state root is not zero.
-        if *global_state_root == Field::zero() {
-            bail!("Inclusion expected the global state root in the execution to *not* be zero")
-        }
-
-        for (transition_index, transition) in execution.transitions().enumerate() {
-            // Construct the transaction leaf.
-            let transaction_leaf = TransactionLeaf::new_execution(transition_index as u16, **transition.id());
-
-            // Process the input tasks.
-            match self.input_tasks.get(transition.id()) {
-                Some(tasks) => {
-                    for task in tasks {
-                        // Retrieve the local state root.
-                        let local_state_root = (*transaction_tree.root()).into();
-
-                        // Construct the state path.
-                        let state_path = match task.is_local {
-                            true => {
-                                // Compute the transaction path.
-                                let transaction_path =
-                                    transaction_tree.prove(transition_index, &transaction_leaf.to_bits_le())?;
-                                // Compute the transition leaf.
-                                let transition_leaf = task.leaf;
-                                // Compute the transition path.
-                                let transition_path = transition.to_path(&transition_leaf)?;
-                                // Output the state path.
-                                StatePath::new_local(
-                                    global_state_root,
-                                    local_state_root,
-                                    transaction_path,
-                                    transaction_leaf,
-                                    transition_path,
-                                    transition_leaf,
-                                )?
-                            }
-                            false => query.get_state_path_for_commitment(&task.commitment)?,
-                        };
-
-                        // Ensure the global state root is the same across iterations.
-                        if global_state_root != state_path.global_state_root() {
-                            bail!("Inclusion expected the global state root to be the same across iterations")
-                        }
-
-                        // Construct the assignment for the state path.
-                        let assignment = InclusionAssignment::new(
-                            state_path,
-                            task.commitment,
-                            task.gamma,
-                            task.serial_number,
-                            local_state_root,
-                            !task.is_local,
-                        );
-
-                        // Add the assignment to the assignments.
-                        assignments.push(assignment);
-                    }
-                }
-                None => bail!("Missing input tasks for transition {} in inclusion", transition.id()),
-            }
-
-            // Insert the leaf into the transaction tree.
-            transaction_tree.append(&[transaction_leaf.to_bits_le()])?;
-        }
-
-        Ok((assignments, global_state_root))
-    }
-
-    /// Returns a new execution with an inclusion proof, for the given execution.
-    pub fn prove_execution<A: circuit::Aleo<Network = N>, R: Rng + CryptoRng>(
-        &self,
-        execution: Execution<N>,
-        assignments: &[InclusionAssignment<N>],
-        global_state_root: N::StateRoot,
-        rng: &mut R,
-    ) -> Result<Execution<N>> {
-        match assignments.is_empty() {
-            true => {
-                // Ensure the global state root is not zero.
-                if *global_state_root == Field::zero() {
-                    bail!("Inclusion expected the global state root in the execution to *not* be zero")
-                }
-
-                // Ensure the inclusion proof in the execution is 'None'.
-                if execution.inclusion_proof().is_some() {
-                    bail!("Inclusion expected the inclusion proof in the execution to be 'None'")
-                }
-                // Return the execution.
-                Execution::from(execution.into_transitions(), global_state_root, None)
-            }
-            false => {
-                // Fetch the inclusion proving key.
-                let proving_key = ProvingKey::<N>::new(N::inclusion_proving_key().clone());
-
-                // Compute the inclusion batch proof.
-                let (global_state_root, inclusion_proof) = Self::prove_batch::<A, R>(&proving_key, assignments, rng)?;
-                // Return the execution.
-                Execution::from(execution.into_transitions(), global_state_root, Some(inclusion_proof))
-            }
-        }
-    }
-
-    /// Returns the inclusion assignments for the given fee transition.
-    pub fn prepare_fee<B: BlockStorage<N>, Q: Into<Query<N, B>>>(
-        &self,
-        fee_transition: &Transition<N>,
-        query: Q,
-    ) -> Result<Vec<InclusionAssignment<N>>> {
-        // Ensure the fee has the correct program ID.
-        let fee_program_id = ProgramID::from_str("credits.aleo")?;
-        ensure!(*fee_transition.program_id() == fee_program_id, "Incorrect program ID for fee");
-
-        // Ensure the fee has the correct function.
-        let fee_function = Identifier::from_str("fee")?;
-        ensure!(*fee_transition.function_name() == fee_function, "Incorrect function name for fee");
-
-        // Initialize an empty transaction tree.
-        let transaction_tree = N::merkle_tree_bhp::<TRANSACTION_DEPTH>(&[])?;
-        // Initialize the global state root.
-        let mut global_state_root = N::StateRoot::default();
-        // Initialize a vector for the assignments.
-        let mut assignments = vec![];
-
-        // Process the input tasks.
-        match self.input_tasks.get(fee_transition.id()) {
-            Some(tasks) => {
-                let query = query.into();
-
-                for task in tasks {
-                    // Retrieve the local state root.
-                    let local_state_root = (*transaction_tree.root()).into();
-                    // Construct the state path.
-                    let state_path = query.get_state_path_for_commitment(&task.commitment)?;
-
-                    // Ensure the global state root is the same across iterations.
-                    if *global_state_root != Field::zero() && global_state_root != state_path.global_state_root() {
-                        bail!("Inclusion expected the global state root to be the same across iterations")
-                    }
-
-                    // Update the global state root.
-                    global_state_root = state_path.global_state_root();
-
-                    // Prepare the assignment for the state path.
-                    let assignment = InclusionAssignment::new(
-                        state_path,
-                        task.commitment,
-                        task.gamma,
-                        task.serial_number,
-                        local_state_root,
-                        !task.is_local,
-                    );
-
-                    // Add the assignment to the assignments.
-                    assignments.push(assignment);
-                }
-            }
-            None => bail!("Missing input state for fee transition {} in inclusion", fee_transition.id()),
-        }
-
-        // Ensure the global state root is not zero.
-        if *global_state_root == Field::zero() {
-            bail!("Inclusion expected the global state root in the fee to *not* be zero")
-        }
-        // Ensure the assignments are not empty.
-        if assignments.is_empty() {
-            bail!("Inclusion expected the assignments for the fee to *not* be empty")
-        }
-        // Return the assignments.
-        Ok(assignments)
-    }
-
-    /// Returns a new fee with an inclusion proof, for the given transition.
-    pub fn prove_fee<A: circuit::Aleo<Network = N>, R: Rng + CryptoRng>(
-        &self,
-        fee_transition: Transition<N>,
-        assignments: &[InclusionAssignment<N>],
-        rng: &mut R,
-    ) -> Result<Fee<N>> {
-        // Ensure the fee has the correct program ID.
-        let fee_program_id = ProgramID::from_str("credits.aleo")?;
-        ensure!(*fee_transition.program_id() == fee_program_id, "Incorrect program ID for fee");
-
-        // Ensure the fee has the correct function.
-        let fee_function = Identifier::from_str("fee")?;
-        ensure!(*fee_transition.function_name() == fee_function, "Incorrect function name for fee");
-
-        // Ensure the assignments are not empty.
-        if assignments.is_empty() {
-            bail!("Inclusion expected the assignments for the fee to *not* be empty")
-        }
-
-        // Fetch the inclusion proving key.
-        let proving_key = ProvingKey::<N>::new(N::inclusion_proving_key().clone());
-
-        // Compute the inclusion batch proof.
-        let (global_state_root, inclusion_proof) = Self::prove_batch::<A, R>(&proving_key, assignments, rng)?;
-        // Return the fee.
-        Ok(Fee::from(fee_transition, global_state_root, Some(inclusion_proof)))
-    }
-
-    /// Checks the inclusion proof for the execution.
-    /// Note: This does *not* check that the global state root exists in the ledger.
-    pub fn verify_execution(execution: &Execution<N>) -> Result<()> {
-        // Retrieve the global state root.
-        let global_state_root = execution.global_state_root();
-
-        // Retrieve the inclusion proof.
-        let inclusion_proof = execution.inclusion_proof();
-
-        // Initialize an empty transaction tree.
-        let mut transaction_tree = N::merkle_tree_bhp::<TRANSACTION_DEPTH>(&[])?;
-        // Initialize a vector for the batch verifier inputs.
-        let mut batch_verifier_inputs = vec![];
-
-        // Construct the batch verifier inputs.
-        for (transition_index, transition) in execution.transitions().enumerate() {
-            // Retrieve the local state root.
-            let local_state_root = *transaction_tree.root();
-
-            // Iterate through the inputs.
-            for input in transition.inputs() {
-                // Filter the inputs for records.
-                if let Input::Record(serial_number, _) = input {
-                    // Add the public inputs to the batch verifier inputs.
-                    batch_verifier_inputs.push(vec![
-                        N::Field::one(),
-                        **global_state_root,
-                        *local_state_root,
-                        **serial_number,
-                    ]);
-                }
-            }
-
-            // Construct the transaction leaf.
-            let transaction_leaf = TransactionLeaf::new_execution(transition_index as u16, **transition.id());
-            // Insert the leaf into the transaction tree.
-            transaction_tree.append(&[transaction_leaf.to_bits_le()])?;
-        }
-
-        // If there are no batch verifier inputs, then ensure the inclusion proof is 'None'.
-        if batch_verifier_inputs.is_empty() && inclusion_proof.is_some() {
-            bail!("No input records in the execution. Expected the inclusion proof to be 'None'")
-        }
-        // If there are batch verifier inputs, then ensure the inclusion proof is 'Some'.
-        if !batch_verifier_inputs.is_empty() && inclusion_proof.is_none() {
-            bail!("Missing inclusion proof for the execution")
-        }
-
-        // Verify the inclusion proof.
-        if let Some(inclusion_proof) = inclusion_proof {
-            // Ensure the global state root is not zero.
-            if *global_state_root == Field::zero() {
-                bail!("Inclusion expected the global state root in the execution to *not* be zero")
-            }
-
-            // Fetch the inclusion verifying key.
-            let verifying_key = VerifyingKey::<N>::new(N::inclusion_verifying_key().clone());
-            // Verify the inclusion proof.
-            ensure!(
-                verifying_key.verify_batch(N::INCLUSION_FUNCTION_NAME, &batch_verifier_inputs, inclusion_proof),
-                "Inclusion proof is invalid"
-            );
-        }
-
-        Ok(())
-    }
-
-    /// Checks the inclusion proof for the fee.
-    /// Note: This does *not* check that the global state root exists in the ledger.
-    pub fn verify_fee(fee: &Fee<N>) -> Result<()> {
-        // Retrieve the global state root.
-        let global_state_root = fee.global_state_root();
-        // Ensure the global state root is not zero.
-        if *global_state_root == Field::zero() {
-            bail!("Inclusion expected the global state root in the fee to *not* be zero")
-        }
-
-        // Retrieve the inclusion proof.
-        let inclusion_proof = match fee.inclusion_proof() {
-            Some(inclusion_proof) => inclusion_proof,
-            None => bail!("Inclusion expected the fee to contain an inclusion proof"),
-        };
-
-        // Initialize an empty transaction tree.
-        let transaction_tree = N::merkle_tree_bhp::<TRANSACTION_DEPTH>(&[])?;
-        // Initialize a vector for the batch verifier inputs.
-        let mut batch_verifier_inputs = vec![];
-
-        // Retrieve the local state root.
-        let local_state_root = *transaction_tree.root();
-
-        // Construct the batch verifier inputs.
-        for input in fee.inputs() {
-            // Filter the inputs for records.
-            if let Input::Record(serial_number, _) = input {
-                // Add the public inputs to the batch verifier inputs.
-                batch_verifier_inputs.push(vec![
-                    N::Field::one(),
-                    **global_state_root,
-                    *local_state_root,
-                    **serial_number,
-                ]);
-            }
-        }
-
-        // Ensure there are batch verifier inputs.
-        if batch_verifier_inputs.is_empty() {
-            bail!("Inclusion expected the fee to contain input records")
-        }
-
-        // Fetch the inclusion verifying key.
-        let verifying_key = VerifyingKey::<N>::new(N::inclusion_verifying_key().clone());
-        // Verify the inclusion proof.
-        ensure!(
-            verifying_key.verify_batch(N::INCLUSION_FUNCTION_NAME, &batch_verifier_inputs, inclusion_proof),
-            "Inclusion proof is invalid"
-        );
-
-        Ok(())
-    }
-
     /// Returns the global state root and inclusion proof for the given assignments.
     fn prove_batch<A: circuit::Aleo<Network = N>, R: Rng + CryptoRng>(
         proving_key: &ProvingKey<N>,
@@ -648,37 +214,3 @@ impl<N: Network> InclusionAssignment<N> {
         Ok(A::eject_assignment_and_reset())
     }
 }
-
-#[cfg(test)]
-mod tests {
-    use super::*;
-    use snarkvm_utilities::TestRng;
-
-    #[test]
-    fn test_inclusion_verify_execution() {
-        let rng = &mut TestRng::default();
-        // Fetch an execution transaction.
-        let execution_transaction = crate::vm::test_helpers::sample_execution_transaction_with_fee(rng);
-
-        match execution_transaction {
-            Transaction::Execute(_, execution, _) => {
-                assert!(Inclusion::verify_execution(&execution).is_ok());
-            }
-            _ => panic!("Expected an execution transaction"),
-        }
-    }
-
-    #[test]
-    fn test_inclusion_verify_fee() {
-        let rng = &mut TestRng::default();
-        // Fetch a deployment transaction.
-        let deployment_transaction = crate::vm::test_helpers::sample_deployment_transaction(rng);
-
-        match deployment_transaction {
-            Transaction::Deploy(_, _, _, fee) => {
-                assert!(Inclusion::verify_fee(&fee).is_ok());
-            }
-            _ => panic!("Expected a deployment transaction"),
-        }
-    }
-}
diff --git a/synthesizer/src/process/stack/mod.rs b/synthesizer/src/process/stack/mod.rs
index 2409289ac..53d540b90 100644
--- a/synthesizer/src/process/stack/mod.rs
+++ b/synthesizer/src/process/stack/mod.rs
@@ -35,6 +35,9 @@ pub use finalize_types::*;
 mod inclusion;
 pub use inclusion::*;
 
+mod query;
+pub use query::*;
+
 mod register_types;
 pub use register_types::*;
 
diff --git a/synthesizer/src/process/stack/query/mod.rs b/synthesizer/src/process/stack/query/mod.rs
new file mode 100644
index 000000000..89a7d064f
--- /dev/null
+++ b/synthesizer/src/process/stack/query/mod.rs
@@ -0,0 +1,146 @@
+// Copyright (C) 2019-2023 Aleo Systems Inc.
+// This file is part of the snarkVM library.
+
+// The snarkVM library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// The snarkVM library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with the snarkVM library. If not, see <https://www.gnu.org/licenses/>.
+
+use crate::{BlockStorage, BlockStore, Program};
+use console::{
+    network::prelude::*,
+    program::{ProgramID, StatePath},
+    types::Field,
+};
+
+#[derive(Clone)]
+pub enum Query<N: Network, B: BlockStorage<N>> {
+    /// The block store from the VM.
+    VM(BlockStore<N, B>),
+    /// The base URL of the node.
+    REST(String),
+}
+
+impl<N: Network, B: BlockStorage<N>> From<BlockStore<N, B>> for Query<N, B> {
+    fn from(block_store: BlockStore<N, B>) -> Self {
+        Self::VM(block_store)
+    }
+}
+
+impl<N: Network, B: BlockStorage<N>> From<&BlockStore<N, B>> for Query<N, B> {
+    fn from(block_store: &BlockStore<N, B>) -> Self {
+        Self::VM(block_store.clone())
+    }
+}
+
+impl<N: Network, B: BlockStorage<N>> From<String> for Query<N, B> {
+    fn from(url: String) -> Self {
+        Self::REST(url)
+    }
+}
+
+impl<N: Network, B: BlockStorage<N>> From<&String> for Query<N, B> {
+    fn from(url: &String) -> Self {
+        Self::REST(url.to_string())
+    }
+}
+
+impl<N: Network, B: BlockStorage<N>> From<&str> for Query<N, B> {
+    fn from(url: &str) -> Self {
+        Self::REST(url.to_string())
+    }
+}
+
+impl<N: Network, B: BlockStorage<N>> Query<N, B> {
+    /// Returns the program for the given program ID.
+    pub fn get_program(&self, program_id: &ProgramID<N>) -> Result<Program<N>> {
+        match self {
+            Self::VM(block_store) => {
+                block_store.get_program(program_id)?.ok_or_else(|| anyhow!("Program {program_id} not found in storage"))
+            }
+            Self::REST(url) => match N::ID {
+                3 => Ok(Self::get_request(&format!("{url}/testnet3/program/{program_id}"))?.into_json()?),
+                _ => bail!("Unsupported network ID in inclusion query"),
+            },
+        }
+    }
+
+    /// Returns the program for the given program ID.
+    pub async fn get_program_async(&self, program_id: &ProgramID<N>) -> Result<Program<N>> {
+        match self {
+            Self::VM(block_store) => {
+                block_store.get_program(program_id)?.ok_or_else(|| anyhow!("Program {program_id} not found in storage"))
+            }
+            Self::REST(url) => match N::ID {
+                3 => Ok(Self::get_request_async(&format!("{url}/testnet3/program/{program_id}")).await?.json().await?),
+                _ => bail!("Unsupported network ID in inclusion query"),
+            },
+        }
+    }
+
+    /// Returns the current state root.
+    pub fn current_state_root(&self) -> Result<N::StateRoot> {
+        match self {
+            Self::VM(block_store) => Ok(block_store.current_state_root()),
+            Self::REST(url) => match N::ID {
+                3 => Ok(Self::get_request(&format!("{url}/testnet3/latest/stateRoot"))?.into_json()?),
+                _ => bail!("Unsupported network ID in inclusion query"),
+            },
+        }
+    }
+
+    /// Returns the current state root.
+    pub async fn current_state_root_async(&self) -> Result<N::StateRoot> {
+        match self {
+            Self::VM(block_store) => Ok(block_store.current_state_root()),
+            Self::REST(url) => match N::ID {
+                3 => Ok(Self::get_request_async(&format!("{url}/testnet3/latest/stateRoot")).await?.json().await?),
+                _ => bail!("Unsupported network ID in inclusion query"),
+            },
+        }
+    }
+
+    /// Returns a state path for the given `commitment`.
+    pub fn get_state_path_for_commitment(&self, commitment: &Field<N>) -> Result<StatePath<N>> {
+        match self {
+            Self::VM(block_store) => block_store.get_state_path_for_commitment(commitment),
+            Self::REST(url) => match N::ID {
+                3 => Ok(Self::get_request(&format!("{url}/testnet3/statePath/{commitment}"))?.into_json()?),
+                _ => bail!("Unsupported network ID in inclusion query"),
+            },
+        }
+    }
+
+    /// Returns a state path for the given `commitment`.
+    pub async fn get_state_path_for_commitment_async(&self, commitment: &Field<N>) -> Result<StatePath<N>> {
+        match self {
+            Self::VM(block_store) => block_store.get_state_path_for_commitment(commitment),
+            Self::REST(url) => match N::ID {
+                3 => {
+                    Ok(Self::get_request_async(&format!("{url}/testnet3/statePath/{commitment}")).await?.json().await?)
+                }
+                _ => bail!("Unsupported network ID in inclusion query"),
+            },
+        }
+    }
+
+    /// Performs a GET request to the given URL.
+    fn get_request(url: &str) -> Result<ureq::Response> {
+        let response = ureq::get(url).call()?;
+        if response.status() == 200 { Ok(response) } else { bail!("Failed to fetch from {url}") }
+    }
+
+    /// Performs a GET request to the given URL.
+    async fn get_request_async(url: &str) -> Result<reqwest::Response> {
+        let response = reqwest::get(url).await?;
+        if response.status() == 200 { Ok(response) } else { bail!("Failed to fetch from {url}") }
+    }
+}
diff --git a/synthesizer/src/program/instruction/operation/assert.rs b/synthesizer/src/program/instruction/operation/assert.rs
index 202f1ed01..56694f800 100644
--- a/synthesizer/src/program/instruction/operation/assert.rs
+++ b/synthesizer/src/program/instruction/operation/assert.rs
@@ -241,11 +241,12 @@ impl<N: Network, const VARIANT: u8> ToBytes for AssertInstruction<N, VARIANT> {
 #[cfg(test)]
 mod tests {
     use super::*;
-    use crate::{ProvingKey, Registers, Store, StoreCircuit, VerifyingKey};
+    use crate::{program::test_helpers::sample_registers, ProvingKey, VerifyingKey};
+
     use circuit::AleoV0;
     use console::{
         network::Testnet3,
-        program::{Literal, LiteralType},
+        program::{Identifier, Literal, LiteralType},
     };
 
     use std::collections::HashMap;
@@ -263,7 +264,6 @@ mod tests {
         cache: &mut HashMap<String, (ProvingKey<CurrentNetwork>, VerifyingKey<CurrentNetwork>)>,
     ) -> Result<(Stack<CurrentNetwork>, Vec<Operand<CurrentNetwork>>)> {
         use crate::{Process, Program};
-        use console::program::Identifier;
 
         // Initialize the opcode.
         let opcode = opcode.to_string();
@@ -296,53 +296,6 @@ mod tests {
         Ok((stack, operands))
     }
 
-    /// Samples the registers. Note: Do not replicate this for real program use, it is insecure.
-    fn sample_registers(
-        stack: &Stack<CurrentNetwork>,
-        literal_a: &Literal<CurrentNetwork>,
-        literal_b: &Literal<CurrentNetwork>,
-        mode_a: Option<circuit::Mode>,
-        mode_b: Option<circuit::Mode>,
-    ) -> Result<Registers<CurrentNetwork, CurrentAleo>> {
-        use crate::{Authorization, CallStack};
-        use console::program::{Identifier, Plaintext, Value};
-
-        // Initialize the function name.
-        let function_name = Identifier::from_str("run")?;
-
-        // Initialize the registers.
-        let mut registers = Registers::<CurrentNetwork, CurrentAleo>::new(
-            CallStack::evaluate(Authorization::new(&[]))?,
-            stack.get_register_types(&function_name)?.clone(),
-        );
-
-        // Initialize the registers.
-        let r0 = Register::Locator(0);
-        let r1 = Register::Locator(1);
-
-        // Initialize the console values.
-        let value_a = Value::Plaintext(Plaintext::from(literal_a));
-        let value_b = Value::Plaintext(Plaintext::from(literal_b));
-
-        // Store the values in the console registers.
-        registers.store(stack, &r0, value_a.clone())?;
-        registers.store(stack, &r1, value_b.clone())?;
-
-        if let (Some(mode_a), Some(mode_b)) = (mode_a, mode_b) {
-            use circuit::Inject;
-
-            // Initialize the circuit values.
-            let circuit_a = circuit::Value::new(mode_a, value_a);
-            let circuit_b = circuit::Value::new(mode_b, value_b);
-
-            // Store the values in the circuit registers.
-            registers.store_circuit(stack, &r0, circuit_a)?;
-            registers.store_circuit(stack, &r1, circuit_b)?;
-        }
-
-        Ok(registers)
-    }
-
     fn check_assert<const VARIANT: u8>(
         operation: impl FnOnce(Vec<Operand<CurrentNetwork>>) -> AssertInstruction<CurrentNetwork, VARIANT>,
         opcode: Opcode,
@@ -363,11 +316,13 @@ mod tests {
         let (stack, operands) = sample_stack(opcode, type_a, type_b, *mode_a, *mode_b, cache).unwrap();
         // Initialize the operation.
         let operation = operation(operands);
+        // Construct the function name.
+        let fn_name = Identifier::from_str("run").unwrap();
 
         /* First, check the operation *succeeds* when both operands are `literal_a.mode_a`. */
         {
             // Attempt to compute the valid operand case.
-            let mut registers = sample_registers(&stack, literal_a, literal_a, None, None).unwrap();
+            let mut registers = sample_registers(&stack, &fn_name, literal_a, literal_a, None, None).unwrap();
             let result_a = operation.evaluate(&stack, &mut registers);
 
             // Ensure the result is correct.
@@ -381,7 +336,8 @@ mod tests {
             }
 
             // Attempt to compute the valid operand case.
-            let mut registers = sample_registers(&stack, literal_a, literal_a, Some(*mode_a), Some(*mode_a)).unwrap();
+            let mut registers =
+                sample_registers(&stack, &fn_name, literal_a, literal_a, Some(*mode_a), Some(*mode_a)).unwrap();
             let result_b = operation.execute::<CurrentAleo>(&stack, &mut registers);
 
             // Ensure the result is correct.
@@ -416,7 +372,7 @@ mod tests {
         /* Next, check the mismatching literals *fail*. */
         if literal_a != literal_b {
             // Attempt to compute the valid operand case.
-            let mut registers = sample_registers(&stack, literal_a, literal_b, None, None).unwrap();
+            let mut registers = sample_registers(&stack, &fn_name, literal_a, literal_b, None, None).unwrap();
             let result_a = operation.evaluate(&stack, &mut registers);
 
             // Ensure the result is correct.
@@ -430,7 +386,8 @@ mod tests {
             }
 
             // Attempt to compute the valid operand case.
-            let mut registers = sample_registers(&stack, literal_a, literal_b, Some(*mode_a), Some(*mode_b)).unwrap();
+            let mut registers =
+                sample_registers(&stack, &fn_name, literal_a, literal_b, Some(*mode_a), Some(*mode_b)).unwrap();
             let result_b = operation.execute::<CurrentAleo>(&stack, &mut registers);
 
             // Ensure the result is correct.
diff --git a/synthesizer/src/program/instruction/operation/is.rs b/synthesizer/src/program/instruction/operation/is.rs
index 7390d674d..1235e6cf7 100644
--- a/synthesizer/src/program/instruction/operation/is.rs
+++ b/synthesizer/src/program/instruction/operation/is.rs
@@ -253,9 +253,10 @@ impl<N: Network, const VARIANT: u8> ToBytes for IsInstruction<N, VARIANT> {
 #[cfg(test)]
 mod tests {
     use super::*;
-    use crate::{ProvingKey, Registers, VerifyingKey};
+    use crate::{program::test_helpers::sample_registers, ProvingKey, VerifyingKey};
+
     use circuit::AleoV0;
-    use console::network::Testnet3;
+    use console::{network::Testnet3, program::Identifier};
 
     use std::collections::HashMap;
 
@@ -273,7 +274,6 @@ mod tests {
         cache: &mut HashMap<String, (ProvingKey<CurrentNetwork>, VerifyingKey<CurrentNetwork>)>,
     ) -> Result<(Stack<CurrentNetwork>, Vec<Operand<CurrentNetwork>>, Register<CurrentNetwork>)> {
         use crate::{Process, Program};
-        use console::program::Identifier;
 
         // Initialize the opcode.
         let opcode = opcode.to_string();
@@ -307,53 +307,6 @@ mod tests {
         Ok((stack, operands, r2))
     }
 
-    /// Samples the registers. Note: Do not replicate this for real program use, it is insecure.
-    fn sample_registers(
-        stack: &Stack<CurrentNetwork>,
-        literal_a: &Literal<CurrentNetwork>,
-        literal_b: &Literal<CurrentNetwork>,
-        mode_a: Option<circuit::Mode>,
-        mode_b: Option<circuit::Mode>,
-    ) -> Result<Registers<CurrentNetwork, CurrentAleo>> {
-        use crate::{Authorization, CallStack};
-        use console::program::Identifier;
-
-        // Initialize the function name.
-        let function_name = Identifier::from_str("run")?;
-
-        // Initialize the registers.
-        let mut registers = Registers::<CurrentNetwork, CurrentAleo>::new(
-            CallStack::evaluate(Authorization::new(&[]))?,
-            stack.get_register_types(&function_name)?.clone(),
-        );
-
-        // Initialize the registers.
-        let r0 = Register::Locator(0);
-        let r1 = Register::Locator(1);
-
-        // Initialize the console values.
-        let value_a = Value::Plaintext(Plaintext::from(literal_a));
-        let value_b = Value::Plaintext(Plaintext::from(literal_b));
-
-        // Store the values in the console registers.
-        registers.store(stack, &r0, value_a.clone())?;
-        registers.store(stack, &r1, value_b.clone())?;
-
-        if let (Some(mode_a), Some(mode_b)) = (mode_a, mode_b) {
-            use circuit::Inject;
-
-            // Initialize the circuit values.
-            let circuit_a = circuit::Value::new(mode_a, value_a);
-            let circuit_b = circuit::Value::new(mode_b, value_b);
-
-            // Store the values in the circuit registers.
-            registers.store_circuit(stack, &r0, circuit_a)?;
-            registers.store_circuit(stack, &r1, circuit_b)?;
-        }
-
-        Ok(registers)
-    }
-
     fn check_is<const VARIANT: u8>(
         operation: impl FnOnce(
             Vec<Operand<CurrentNetwork>>,
@@ -379,13 +332,15 @@ mod tests {
         let (stack, operands, destination) = sample_stack(opcode, type_a, type_b, *mode_a, *mode_b, cache).unwrap();
         // Initialize the operation.
         let operation = operation(operands, destination.clone());
+        // Initialize the function name.
+        let fn_name = Identifier::from_str("run").unwrap();
         // Initialize a destination operand.
         let destination_operand = Operand::Register(destination);
 
         /* First, check the operation *succeeds* when both operands are `literal_a.mode_a`. */
         {
             // Attempt to compute the valid operand case.
-            let mut registers = sample_registers(&stack, literal_a, literal_a, None, None).unwrap();
+            let mut registers = sample_registers(&stack, &fn_name, literal_a, literal_a, None, None).unwrap();
             operation.evaluate(&stack, &mut registers).unwrap();
 
             // Retrieve the output.
@@ -406,7 +361,8 @@ mod tests {
             }
 
             // Attempt to compute the valid operand case.
-            let mut registers = sample_registers(&stack, literal_a, literal_a, Some(*mode_a), Some(*mode_a)).unwrap();
+            let mut registers =
+                sample_registers(&stack, &fn_name, literal_a, literal_a, Some(*mode_a), Some(*mode_a)).unwrap();
             operation.execute::<CurrentAleo>(&stack, &mut registers).unwrap();
 
             // Retrieve the output.
@@ -448,7 +404,7 @@ mod tests {
         /* Next, check the mismatching literals *fail*. */
         if literal_a != literal_b {
             // Attempt to compute the valid operand case.
-            let mut registers = sample_registers(&stack, literal_a, literal_b, None, None).unwrap();
+            let mut registers = sample_registers(&stack, &fn_name, literal_a, literal_b, None, None).unwrap();
             operation.evaluate(&stack, &mut registers).unwrap();
 
             // Retrieve the output.
@@ -469,7 +425,8 @@ mod tests {
             }
 
             // Attempt to compute the valid operand case.
-            let mut registers = sample_registers(&stack, literal_a, literal_b, Some(*mode_a), Some(*mode_b)).unwrap();
+            let mut registers =
+                sample_registers(&stack, &fn_name, literal_a, literal_b, Some(*mode_a), Some(*mode_b)).unwrap();
             operation.execute::<CurrentAleo>(&stack, &mut registers).unwrap();
 
             // Retrieve the output.
diff --git a/synthesizer/src/program/instruction/operation/mod.rs b/synthesizer/src/program/instruction/operation/mod.rs
index 9d5c26e36..58934df08 100644
--- a/synthesizer/src/program/instruction/operation/mod.rs
+++ b/synthesizer/src/program/instruction/operation/mod.rs
@@ -769,3 +769,60 @@ crate::operation!(
         (U128, U128) => U128,
     }
 );
+
+#[cfg(test)]
+pub(crate) mod test_helpers {
+    use super::*;
+    use crate::{Authorization, CallStack, Registers, Stack, Store, StoreCircuit};
+
+    use circuit::AleoV0;
+    use console::{
+        network::Testnet3,
+        program::{Identifier, Literal, Plaintext, Register, Value},
+    };
+
+    type CurrentNetwork = Testnet3;
+    type CurrentAleo = AleoV0;
+
+    /// Samples the registers. Note: Do not replicate this for real program use, it is insecure.
+    pub fn sample_registers(
+        stack: &Stack<CurrentNetwork>,
+        function_name: &Identifier<CurrentNetwork>,
+        literal_a: &Literal<CurrentNetwork>,
+        literal_b: &Literal<CurrentNetwork>,
+        mode_a: Option<circuit::Mode>,
+        mode_b: Option<circuit::Mode>,
+    ) -> Result<Registers<CurrentNetwork, CurrentAleo>> {
+        // Initialize the registers.
+        let mut registers = Registers::<CurrentNetwork, CurrentAleo>::new(
+            CallStack::evaluate(Authorization::new(&[]))?,
+            stack.get_register_types(function_name)?.clone(),
+        );
+
+        // Initialize the registers.
+        let r0 = Register::Locator(0);
+        let r1 = Register::Locator(1);
+
+        // Initialize the console values.
+        let value_a = Value::Plaintext(Plaintext::from(literal_a));
+        let value_b = Value::Plaintext(Plaintext::from(literal_b));
+
+        // Store the values in the console registers.
+        registers.store(stack, &r0, value_a.clone())?;
+        registers.store(stack, &r1, value_b.clone())?;
+
+        if let (Some(mode_a), Some(mode_b)) = (mode_a, mode_b) {
+            use circuit::Inject;
+
+            // Initialize the circuit values.
+            let circuit_a = circuit::Value::new(mode_a, value_a);
+            let circuit_b = circuit::Value::new(mode_b, value_b);
+
+            // Store the values in the circuit registers.
+            registers.store_circuit(stack, &r0, circuit_a)?;
+            registers.store_circuit(stack, &r1, circuit_b)?;
+        }
+
+        Ok(registers)
+    }
+}
diff --git a/synthesizer/src/store/block/mod.rs b/synthesizer/src/store/block/mod.rs
index 9bb55da51..1d74e704c 100644
--- a/synthesizer/src/store/block/mod.rs
+++ b/synthesizer/src/store/block/mod.rs
@@ -860,6 +860,14 @@ impl<N: Network, B: BlockStorage<N>> BlockStore<N, B> {
         self.storage.get_block(block_hash)
     }
 
+    /// Returns the confirmed transaction for the given `transaction ID`.
+    pub fn get_confirmed_transaction(
+        &self,
+        transaction_id: &N::TransactionID,
+    ) -> Result<Option<ConfirmedTransaction<N>>> {
+        self.storage.get_confirmed_transaction(*transaction_id)
+    }
+
     /// Returns the program for the given `program ID`.
     pub fn get_program(&self, program_id: &ProgramID<N>) -> Result<Option<Program<N>>> {
         self.storage.transaction_store().get_program(program_id)
diff --git a/utilities/Cargo.toml b/utilities/Cargo.toml
index 9572f5bcf..52405b3ca 100644
--- a/utilities/Cargo.toml
+++ b/utilities/Cargo.toml
@@ -70,6 +70,7 @@ default-features = false
 
 [features]
 default = [ "aleo-std/cpu", "derive", "num_cpus", "std" ]
+derive = [ "snarkvm-utilities-derives" ]
 serial = [ "derive" ]
 std = [ ]
-derive = [ "snarkvm-utilities-derives" ]
+wasm = [ ]
diff --git a/utilities/src/parallel.rs b/utilities/src/parallel.rs
index 0ccd99597..0d77e73de 100644
--- a/utilities/src/parallel.rs
+++ b/utilities/src/parallel.rs
@@ -67,18 +67,18 @@ pub fn max_available_threads() -> usize {
 }
 
 #[inline(always)]
-#[cfg(not(feature = "serial"))]
+#[cfg(not(any(feature = "serial", feature = "wasm")))]
 pub fn execute_with_max_available_threads<T: Sync + Send>(f: impl FnOnce() -> T + Send) -> T {
     execute_with_threads(f, max_available_threads())
 }
 
 #[inline(always)]
-#[cfg(feature = "serial")]
+#[cfg(any(feature = "serial", feature = "wasm"))]
 pub fn execute_with_max_available_threads<T>(f: impl FnOnce() -> T + Send) -> T {
     f()
 }
 
-#[cfg(not(feature = "serial"))]
+#[cfg(not(any(feature = "serial", feature = "wasm")))]
 #[inline(always)]
 fn execute_with_threads<T: Sync + Send>(f: impl FnOnce() -> T + Send, num_threads: usize) -> T {
     let pool = rayon::ThreadPoolBuilder::new().num_threads(num_threads).build().unwrap();
