diff --git a/src/network/peers.rs b/src/network/peers.rs
index 8998e8d3a..4f8aec4cc 100644
--- a/src/network/peers.rs
+++ b/src/network/peers.rs
@@ -392,10 +392,8 @@ impl<N: Network, E: Environment> Peers<N, E> {
                 };
 
                 // Add the sync nodes to the list of candidate peers.
-                if number_of_connected_sync_nodes == 0 {
-                    let sync_nodes: Vec<SocketAddr> = E::SYNC_NODES.iter().map(|ip| ip.parse().unwrap()).collect();
-                    self.add_candidate_peers(&sync_nodes).await;
-                }
+                let sync_nodes: Vec<SocketAddr> = E::SYNC_NODES.iter().map(|ip| ip.parse().unwrap()).collect();
+                self.add_candidate_peers(&sync_nodes).await;
 
                 // Add the beacon nodes to the list of candidate peers.
                 let beacon_nodes: Vec<SocketAddr> = E::BEACON_NODES.iter().map(|ip| ip.parse().unwrap()).collect();
