diff --git a/node/bft/src/bft.rs b/node/bft/src/bft.rs
index bc72924df..7662f27b7 100644
--- a/node/bft/src/bft.rs
+++ b/node/bft/src/bft.rs
@@ -251,6 +251,7 @@ impl<N: Network> BFT<N> {
             // Update to the next round in storage.
             if let Err(e) = self.storage().increment_to_next_round(current_round) {
                 warn!("BFT failed to increment to the next round from round {current_round} - {e}");
+                return false;
             }
             // Update the timer for the leader certificate.
             self.leader_certificate_timer.store(now(), Ordering::SeqCst);
