diff --git a/.ci/utils.sh b/.ci/utils.sh
index fa323af02..740a558af 100644
--- a/.ci/utils.sh
+++ b/.ci/utils.sh
@@ -75,7 +75,7 @@ function check_heights() {
   fi
 }
 
-# Function checking that nodes created logs on disk.
+# Function checking that nodes created logs on disk and they contain no errors.
 function check_logs() {
   echo "Checking logs exist for all nodes..."
   local log_dir=$1
@@ -90,6 +90,11 @@ function check_logs() {
       echo "❌ Test failed! Validator #${validator_index} did not create any logs in \"$log_dir\"."
       return 1
     fi
+
+    if grep -q "ERROR" "$log_dir/validator-${validator_index}.log"; then
+      echo "❌ Test failed! Validator #${validator_index} logs contain errors."
+      return 1
+    fi
   done
 
   for ((client_index = 0; client_index < total_clients; client_index++)); do
@@ -97,6 +102,12 @@ function check_logs() {
       echo "❌ Test failed! Client #${client_index} did not create any logs in \"$log_dir\"."
       return 1
     fi
+
+    if grep -q "ERROR" "$log_dir/client-${client_index}.log"; then
+      echo "❌ Test failed! Client #${client_index} logs contain errors."
+      return 1
+    fi
+ 
   done
 
   return 0
diff --git a/Cargo.lock b/Cargo.lock
index 21cba5b1a..196bbcc91 100644
--- a/Cargo.lock
+++ b/Cargo.lock
@@ -587,9 +587,9 @@ dependencies = [
 
 [[package]]
 name = "cc"
-version = "1.2.47"
+version = "1.2.48"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "cd405d82c84ff7f35739f175f67d8b9fb7687a0e84ccdc78bd3568839827cf07"
+checksum = "c481bdbf0ed3b892f6f806287d72acd515b352a4ec27a208489b8c1bc839633a"
 dependencies = [
  "find-msvc-tools",
  "jobserver",
@@ -795,9 +795,9 @@ checksum = "7c74b8349d32d297c9134b8c88677813a227df8f779daa29bfc29c183fe3dca6"
 
 [[package]]
 name = "convert_case"
-version = "0.7.1"
+version = "0.10.0"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "bb402b8d4c85569410425650ce3eddc7d698ed96d39a73f941b08fb63082f1e7"
+checksum = "633458d4ef8c78b72454de2d54fd6ab2e60f9e02be22f3c6104cdc8a4e0fceb9"
 dependencies = [
  "unicode-segmentation",
 ]
@@ -1165,22 +1165,23 @@ dependencies = [
 
 [[package]]
 name = "derive_more"
-version = "2.0.1"
+version = "2.1.0"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "093242cf7570c207c83073cf82f79706fe7b8317e98620a47d5be7c3d8497678"
+checksum = "10b768e943bed7bf2cab53df09f4bc34bfd217cdb57d971e769874c9a6710618"
 dependencies = [
  "derive_more-impl",
 ]
 
 [[package]]
 name = "derive_more-impl"
-version = "2.0.1"
+version = "2.1.0"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "bda628edc44c4bb645fbe0f758797143e4e07926f7ebf4e9bdfbd3d2ce621df3"
+checksum = "6d286bfdaf75e988b4a78e013ecd79c581e06399ab53fbacd2d916c2f904f30b"
 dependencies = [
  "convert_case",
  "proc-macro2",
  "quote 1.0.42",
+ "rustc_version",
  "syn 2.0.111",
 ]
 
@@ -2048,9 +2049,9 @@ dependencies = [
 
 [[package]]
 name = "hyper-util"
-version = "0.1.18"
+version = "0.1.19"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "52e9a2a24dc5c6821e71a7030e1e14b7b632acac55c40e9d2e082c621261bb56"
+checksum = "727805d60e7938b76b826a6ef209eb70eaa1812794f9424d4a4e2d740662df5f"
 dependencies = [
  "base64 0.22.1",
  "bytes",
@@ -2252,9 +2253,9 @@ dependencies = [
 
 [[package]]
 name = "instability"
-version = "0.3.9"
+version = "0.3.10"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "435d80800b936787d62688c927b6490e887c7ef5ff9ce922c6c6050fca75eb9a"
+checksum = "6778b0196eefee7df739db78758e5cf9b37412268bfa5650bfeed028aed20d9c"
 dependencies = [
  "darling 0.20.11",
  "indoc",
@@ -2340,9 +2341,9 @@ dependencies = [
 
 [[package]]
 name = "js-sys"
-version = "0.3.82"
+version = "0.3.83"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "b011eec8cc36da2aab2d5cff675ec18454fad408585853910a202391cf9f8e65"
+checksum = "464a3709c7f55f1f721e5389aa6ea4e3bc6aba669353300af094b29ffbdde1d8"
 dependencies = [
  "once_cell",
  "wasm-bindgen",
@@ -2389,9 +2390,9 @@ checksum = "830d08ce1d1d941e6b30645f1a0eb5643013d835ce3779a5fc208261dbe10f55"
 
 [[package]]
 name = "libc"
-version = "0.2.177"
+version = "0.2.178"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "2874a2af47a2325c2001a6e6fad9b16a53b802102b528163885171cf92b15976"
+checksum = "37c93d8daa9d8a012fd8ab92f088405fb202ea0b6ab73ee2482ae66af4f42091"
 
 [[package]]
 name = "libgit2-sys"
@@ -2511,9 +2512,9 @@ dependencies = [
 
 [[package]]
 name = "log"
-version = "0.4.28"
+version = "0.4.29"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "34080505efa8e45a4b816c349525ebe327ceaa8559756f0356cba97ef3bf7432"
+checksum = "5e5032e24019045c762d3c0f28f5b6b8bbf38563a65908389bf7978758920897"
 
 [[package]]
 name = "lru"
@@ -2654,9 +2655,9 @@ dependencies = [
 
 [[package]]
 name = "mio"
-version = "1.1.0"
+version = "1.1.1"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "69d83b0086dc8ecf3ce9ae2874b2d1290252e2a30720bea58a5c6639b0092873"
+checksum = "a69bcab0ad47271a0234d9422b131806bf3968021e5dc9328caf2d4cd58557fc"
 dependencies = [
  "libc",
  "log",
@@ -3651,9 +3652,9 @@ dependencies = [
 
 [[package]]
 name = "rustls-pki-types"
-version = "1.13.0"
+version = "1.13.1"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "94182ad936a0c91c324cd46c6511b9510ed16af436d7b5bab34beab0afd55f7a"
+checksum = "708c0f9d5f54ba0272468c1d306a52c495b31fa155e91bc25371e6df7996908c"
 dependencies = [
  "web-time",
  "zeroize",
@@ -3912,9 +3913,9 @@ dependencies = [
 
 [[package]]
 name = "serde_with"
-version = "3.16.0"
+version = "3.16.1"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "10574371d41b0d9b2cff89418eda27da52bcaff2cc8741db26382a77c29131f1"
+checksum = "4fa237f2807440d238e0364a218270b98f767a00d3dada77b1c53ae88940e2e7"
 dependencies = [
  "base64 0.22.1",
  "chrono",
@@ -3931,9 +3932,9 @@ dependencies = [
 
 [[package]]
 name = "serde_with_macros"
-version = "3.16.0"
+version = "3.16.1"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "08a72d8216842fdd57820dc78d840bef99248e35fb2554ff923319e60f2d686b"
+checksum = "52a8e3ca0ca629121f70ab50f95249e5a6f925cc0f6ffe8256c45b728875706c"
 dependencies = [
  "darling 0.21.3",
  "proc-macro2",
@@ -4136,6 +4137,7 @@ dependencies = [
  "snarkos-node-cdn",
  "snarkos-node-metrics",
  "snarkos-node-rest",
+ "snarkos-utilities",
  "snarkvm",
  "sys-info",
  "tempfile",
@@ -4156,6 +4158,7 @@ dependencies = [
  "crossterm 0.29.0",
  "ratatui",
  "snarkos-node",
+ "snarkos-utilities",
  "snarkvm",
  "tokio",
 ]
@@ -4176,7 +4179,6 @@ dependencies = [
  "locktick",
  "lru 0.16.2",
  "num_cpus",
- "once_cell",
  "parking_lot",
  "paste",
  "pea2pea",
@@ -4193,6 +4195,7 @@ dependencies = [
  "snarkos-node-router",
  "snarkos-node-sync",
  "snarkos-node-tcp",
+ "snarkos-utilities",
  "snarkvm",
  "time",
  "tokio",
@@ -4241,6 +4244,7 @@ dependencies = [
  "snarkos-node-network",
  "snarkos-node-sync",
  "snarkos-node-tcp",
+ "snarkos-utilities",
  "snarkvm",
  "test-log",
  "test-strategy 0.4.3",
@@ -4263,7 +4267,6 @@ dependencies = [
  "indexmap 2.12.1",
  "proptest",
  "serde",
- "snarkos-node-network",
  "snarkos-node-sync-locators",
  "snarkvm",
  "test-strategy 0.4.3",
@@ -4284,6 +4287,7 @@ dependencies = [
  "rand 0.8.5",
  "rayon",
  "snarkos-node-metrics",
+ "snarkos-utilities",
  "snarkvm",
  "tokio",
  "tracing",
@@ -4318,6 +4322,7 @@ dependencies = [
  "serde",
  "serde_json",
  "snarkos-node-metrics",
+ "snarkos-utilities",
  "snarkvm",
  "tokio",
  "tokio-test",
@@ -4524,6 +4529,16 @@ dependencies = [
  "tracing",
 ]
 
+[[package]]
+name = "snarkos-utilities"
+version = "4.2.1"
+dependencies = [
+ "locktick",
+ "parking_lot",
+ "tokio",
+ "tracing",
+]
+
 [[package]]
 name = "snarkvm"
 version = "4.4.0"
@@ -5697,9 +5712,9 @@ dependencies = [
 
 [[package]]
 name = "test-log-macros"
-version = "0.2.18"
+version = "0.2.19"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "451b374529930d7601b1eef8d32bc79ae870b6079b069401709c2a8bf9e75f36"
+checksum = "be35209fd0781c5401458ab66e4f98accf63553e8fae7425503e92fdd319783b"
 dependencies = [
  "proc-macro2",
  "quote 1.0.42",
@@ -6132,9 +6147,9 @@ dependencies = [
 
 [[package]]
 name = "tracing"
-version = "0.1.41"
+version = "0.1.43"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "784e0ac535deb450455cbfa28a6f0df145ea1bb7ae51b821cf5e7927fdcfbdd0"
+checksum = "2d15d90a0b5c19378952d479dc858407149d7bb45a14de0142f6c534b16fc647"
 dependencies = [
  "log",
  "pin-project-lite",
@@ -6425,9 +6440,9 @@ dependencies = [
 
 [[package]]
 name = "wasm-bindgen"
-version = "0.2.105"
+version = "0.2.106"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "da95793dfc411fbbd93f5be7715b0578ec61fe87cb1a42b12eb625caa5c5ea60"
+checksum = "0d759f433fa64a2d763d1340820e46e111a7a5ab75f993d1852d70b03dbb80fd"
 dependencies = [
  "cfg-if",
  "once_cell",
@@ -6438,9 +6453,9 @@ dependencies = [
 
 [[package]]
 name = "wasm-bindgen-futures"
-version = "0.4.55"
+version = "0.4.56"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "551f88106c6d5e7ccc7cd9a16f312dd3b5d36ea8b4954304657d5dfba115d4a0"
+checksum = "836d9622d604feee9e5de25ac10e3ea5f2d65b41eac0d9ce72eb5deae707ce7c"
 dependencies = [
  "cfg-if",
  "js-sys",
@@ -6451,9 +6466,9 @@ dependencies = [
 
 [[package]]
 name = "wasm-bindgen-macro"
-version = "0.2.105"
+version = "0.2.106"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "04264334509e04a7bf8690f2384ef5265f05143a4bff3889ab7a3269adab59c2"
+checksum = "48cb0d2638f8baedbc542ed444afc0644a29166f1595371af4fecf8ce1e7eeb3"
 dependencies = [
  "quote 1.0.42",
  "wasm-bindgen-macro-support",
@@ -6461,9 +6476,9 @@ dependencies = [
 
 [[package]]
 name = "wasm-bindgen-macro-support"
-version = "0.2.105"
+version = "0.2.106"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "420bc339d9f322e562942d52e115d57e950d12d88983a14c79b86859ee6c7ebc"
+checksum = "cefb59d5cd5f92d9dcf80e4683949f15ca4b511f4ac0a6e14d4e1ac60c6ecd40"
 dependencies = [
  "bumpalo",
  "proc-macro2",
@@ -6474,18 +6489,18 @@ dependencies = [
 
 [[package]]
 name = "wasm-bindgen-shared"
-version = "0.2.105"
+version = "0.2.106"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "76f218a38c84bcb33c25ec7059b07847d465ce0e0a76b995e134a45adcb6af76"
+checksum = "cbc538057e648b67f72a982e708d485b2efa771e1ac05fec311f9f63e5800db4"
 dependencies = [
  "unicode-ident",
 ]
 
 [[package]]
 name = "web-sys"
-version = "0.3.82"
+version = "0.3.83"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "3a1f95c0d03a47f4ae1f7a64643a6bb97465d9b740f0fa8f90ea33915c99a9a1"
+checksum = "9b32828d774c412041098d182a8b38b16ea816958e07cf40eec2bc080ae137ac"
 dependencies = [
  "js-sys",
  "wasm-bindgen",
@@ -6848,18 +6863,18 @@ dependencies = [
 
 [[package]]
 name = "zerocopy"
-version = "0.8.30"
+version = "0.8.31"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "4ea879c944afe8a2b25fef16bb4ba234f47c694565e97383b36f3a878219065c"
+checksum = "fd74ec98b9250adb3ca554bdde269adf631549f51d8a8f8f0a10b50f1cb298c3"
 dependencies = [
  "zerocopy-derive",
 ]
 
 [[package]]
 name = "zerocopy-derive"
-version = "0.8.30"
+version = "0.8.31"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "cf955aa904d6040f70dc8e9384444cb1030aed272ba3cb09bbc4ab9e7c1f34f5"
+checksum = "d8a8d209fdf45cf5138cbb5a506f6b52522a25afccc534d1475dad8e31105c6a"
 dependencies = [
  "proc-macro2",
  "quote 1.0.42",
diff --git a/Cargo.toml b/Cargo.toml
index 1c82514cc..d0c3363ea 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -37,7 +37,8 @@ members = [
   "node/sync",
   "node/sync/communication-service",
   "node/sync/locators",
-  "node/tcp"
+  "node/tcp",
+  "utilities",
 ]
 
 [workspace.dependencies.aleo-std]
@@ -249,6 +250,10 @@ version = "=4.4.0"
 path = "node/tcp"
 version = "=4.4.0"
 
+[workspace.dependencies.snarkos-utilities]
+path = "utilities"
+version = "=4.2.1"
+
 [[bin]]
 name = "snarkos"
 path = "snarkos/main.rs"
diff --git a/cli/Cargo.toml b/cli/Cargo.toml
index 72da717c9..22c54809a 100644
--- a/cli/Cargo.toml
+++ b/cli/Cargo.toml
@@ -17,12 +17,15 @@ license = "Apache-2.0"
 edition = "2024"
 
 [features]
-async = []
+default = [ ]
+async = [ ]
 locktick = [
   "dep:locktick",
+  "snarkos-display/locktick",
   "snarkos-node/locktick",
   "snarkos-node-cdn/locktick",
   "snarkos-node-rest/locktick",
+  "snarkos-utilities/locktick",
   "snarkvm/locktick"
 ]
 metrics = [ "dep:snarkos-node-metrics", "snarkos-node/metrics" ]
@@ -40,7 +43,11 @@ test_network = [
   "test_consensus_heights",
   "snarkvm/dev-print"
 ]
-serial = [ ]
+serial = [
+  "snarkvm/serial",
+  "snarkos-node/serial",
+  "snarkos-display/serial",
+]
 tokio_console = [ "dep:console-subscriber", "tokio/tracing" ]
 
 [dependencies.aleo-std]
@@ -126,6 +133,9 @@ optional = true
 [dependencies.snarkos-node-rest]
 workspace = true
 
+[dependencies.snarkos-utilities]
+workspace = true
+
 [dependencies.snarkvm]
 workspace = true
 features = [ "parameters", "circuit", "package" ]
diff --git a/cli/src/commands/developer/execute.rs b/cli/src/commands/developer/execute.rs
index 0931fe3c0..c3eaf86b1 100644
--- a/cli/src/commands/developer/execute.rs
+++ b/cli/src/commands/developer/execute.rs
@@ -74,6 +74,8 @@ pub struct Execute {
     /// to fit the network type and query.
     /// For example, the base URL may extend to "http://mynode.com/testnet/transaction/unconfirmed/ID" to retrieve
     /// an unconfirmed transaction on the test network.
+    ///
+    /// The given value may also be a JSON serialized `StaticQuery` struct.
     #[clap(short, long, alias="query", default_value=DEFAULT_ENDPOINT, verbatim_doc_comment)]
     endpoint: Uri,
     /// The priority fee in microcredits.
@@ -103,6 +105,9 @@ pub struct Execute {
     /// Timeout in seconds when waiting for transaction confirmation. Default is 60 seconds.
     #[clap(long, default_value_t = 60, requires = "wait")]
     timeout: u64,
+    /// Send the transaction without checking if sufficient funds are available (intended for testing purposes only).
+    #[clap(long, hide = true)]
+    skip_funds_check: bool,
 }
 
 impl Drop for Execute {
@@ -122,7 +127,7 @@ impl Execute {
         // Specify the query
         let query = Query::<N, BlockMemory<N>>::from(endpoint.clone());
 
-        // TODO (kaimast): can this ever be true?
+        // Check if the query is a static query.
         let is_static_query = matches!(query, Query::STATIC(_));
 
         // Retrieve the private key.
@@ -155,11 +160,9 @@ impl Execute {
                 let height = query.current_block_height().with_context(|| "Failed to retrieve current block height")?;
                 let version = N::CONSENSUS_VERSION(height)?;
                 debug!("At block height {height} and consensus {version:?}");
-                let edition = Developer::get_latest_edition(&endpoint, &program_id)
-                    .with_context(|| format!("Failed to get latest edition for program {program_id}"))?;
 
                 // Load the program and it's imports into the process.
-                load_program(&query, &mut vm.process().write(), &program_id, edition)?;
+                load_program(&query, &mut vm.process().write(), &program_id, &endpoint)?;
             }
 
             // Prepare the fee.
@@ -184,7 +187,7 @@ impl Execute {
         };
 
         // Check if the public balance is sufficient.
-        if self.record.is_none() && !is_static_query {
+        if self.record.is_none() && !is_static_query && !self.skip_funds_check {
             // Fetch the public balance.
             let address = Address::try_from(&private_key)?;
             let public_balance = Developer::get_public_balance::<N>(&endpoint, &address)
@@ -239,10 +242,13 @@ fn load_program<N: Network>(
     query: &Query<N, BlockMemory<N>>,
     process: &mut Process<N>,
     program_id: &ProgramID<N>,
-    edition: u16,
+    endpoint: &Uri,
 ) -> Result<()> {
     // Fetch the program.
     let program = query.get_program(program_id).with_context(|| "Failed to fetch program")?;
+    // Fetch the latest edition of the program.
+    let edition = Developer::get_latest_edition(endpoint, program_id)
+        .with_context(|| format!("Failed to get latest edition for program {program_id}"))?;
 
     // Return early if the program is already loaded.
     if process.contains_program(program.id()) {
@@ -254,13 +260,14 @@ fn load_program<N: Network>(
         // Add the imports to the process if does not exist yet.
         if !process.contains_program(import_program_id) {
             // Recursively load the program and its imports.
-            load_program(query, process, import_program_id, edition)
+            load_program(query, process, import_program_id, endpoint)
                 .with_context(|| format!("Failed to load imported program {import_program_id}"))?;
         }
     }
 
     // Add the program to the process if it does not already exist.
     if !process.contains_program(program.id()) {
+        debug!("Adding program {program_id} with edition {edition}");
         process
             .add_programs_with_editions(&[(program, edition)])
             .with_context(|| format!("Failed to add program {program_id}"))?;
diff --git a/cli/src/commands/developer/mod.rs b/cli/src/commands/developer/mod.rs
index 535b91812..3e46f03d7 100644
--- a/cli/src/commands/developer/mod.rs
+++ b/cli/src/commands/developer/mod.rs
@@ -76,7 +76,8 @@ pub enum DeveloperCommand {
 }
 
 /// Use the Provable explorer's API by default.
-const DEFAULT_ENDPOINT: &str = "https://api.explorer.provable.com";
+/// Note, `v2` here refers to the explorer version, not the snarkOS API version.
+const DEFAULT_ENDPOINT: &str = "https://api.explorer.provable.com/v2";
 
 #[derive(Debug, Parser)]
 pub struct Developer {
diff --git a/cli/src/commands/developer/scan.rs b/cli/src/commands/developer/scan.rs
index 13b999b5b..d51210ec1 100644
--- a/cli/src/commands/developer/scan.rs
+++ b/cli/src/commands/developer/scan.rs
@@ -20,6 +20,8 @@ use crate::{
 };
 
 use snarkos_node_cdn::CDN_BASE_URL;
+use snarkos_utilities::SimpleStoppable;
+
 use snarkvm::{
     console::network::Network,
     prelude::{Ciphertext, Field, Plaintext, PrivateKey, Record, ViewKey, block::Block},
@@ -312,7 +314,7 @@ impl Scan {
         let rt = tokio::runtime::Runtime::new()?;
 
         // Create a placeholder shutdown flag.
-        let _shutdown = Default::default();
+        let _shutdown = SimpleStoppable::new();
 
         // Copy endpoint for background task.
         let endpoint = endpoint.clone();
diff --git a/cli/src/commands/start.rs b/cli/src/commands/start.rs
index 55d34da83..c35a9bcd7 100644
--- a/cli/src/commands/start.rs
+++ b/cli/src/commands/start.rs
@@ -24,6 +24,8 @@ use snarkos_node::{
     rest::DEFAULT_REST_PORT,
     router::DEFAULT_NODE_PORT,
 };
+use snarkos_utilities::SignalHandler;
+
 use snarkvm::{
     console::{
         account::{Address, PrivateKey},
@@ -57,7 +59,7 @@ use std::{
 };
 use tokio::{
     runtime::{self, Runtime},
-    sync::oneshot,
+    sync::mpsc,
     task,
 };
 use tracing::warn;
@@ -267,7 +269,7 @@ pub struct Start {
 }
 
 impl Start {
-    /// Starts the snarkOS node.
+    /// Starts the snarkOS node and blocks until it terminates.
     pub fn parse(self) -> Result<String> {
         // Prepare the shutdown flag.
         let shutdown: Arc<AtomicBool> = Default::default();
@@ -285,57 +287,19 @@ impl Start {
         // Initialize the runtime.
         Self::runtime().block_on(async move {
             // Error messages.
-            let node_parse_error = || "Failed to parse node arguments";
-            let display_start_error = || "Failed to initialize the display";
-
-            let (shutdown_tx, shutdown_rx) = oneshot::channel();
+            let node_parse_error = || "Failed to start node";
 
             // Clone the configurations.
-            let mut cli = self.clone();
-            // Parse the network.
-            match cli.network {
-                MainnetV0::ID => {
-                    // Parse the node from the configurations.
-                    let node = cli
-                        .parse_node::<MainnetV0>(shutdown.clone(), shutdown_tx)
-                        .await
-                        .with_context(node_parse_error)?;
-                    // If the display is enabled, render the display.
-                    if !cli.nodisplay {
-                        // Initialize the display.
-                        Display::start(node, log_receiver).with_context(display_start_error)?;
-                    }
-                }
-                TestnetV0::ID => {
-                    // Parse the node from the configurations.
-                    let node = cli
-                        .parse_node::<TestnetV0>(shutdown.clone(), shutdown_tx)
-                        .await
-                        .with_context(node_parse_error)?;
-                    // If the display is enabled, render the display.
-                    if !cli.nodisplay {
-                        // Initialize the display.
-                        Display::start(node, log_receiver).with_context(display_start_error)?;
-                    }
-                }
-                CanaryV0::ID => {
-                    // Parse the node from the configurations.
-                    let node = cli
-                        .parse_node::<CanaryV0>(shutdown.clone(), shutdown_tx)
-                        .await
-                        .with_context(node_parse_error)?;
-                    // If the display is enabled, render the display.
-                    if !cli.nodisplay {
-                        // Initialize the display.
-                        Display::start(node, log_receiver).with_context(display_start_error)?;
-                    }
-                }
+            let mut self_ = self.clone();
+
+            // Parse the node arguments, start it, and block until shutdown.
+            match self_.network {
+                MainnetV0::ID => self_.parse_node::<MainnetV0>(log_receiver).await.with_context(node_parse_error)?,
+                TestnetV0::ID => self_.parse_node::<TestnetV0>(log_receiver).await.with_context(node_parse_error)?,
+                CanaryV0::ID => self_.parse_node::<CanaryV0>(log_receiver).await.with_context(node_parse_error)?,
                 _ => panic!("Invalid network ID specified"),
             };
 
-            // Wait for the shutdown signal.
-            let _ = shutdown_rx.await;
-
             Ok(String::new())
         })
     }
@@ -600,9 +564,9 @@ impl Start {
         }
     }
 
-    /// Returns the node type corresponding to the given configurations.
+    /// Start the node and blocks until it terminates.
     #[rustfmt::skip]
-    async fn parse_node<N: Network>(&mut self, shutdown: Arc<AtomicBool>, shutdown_tx: oneshot::Sender<()>) -> Result<Node<N>> {
+    async fn parse_node<N: Network>(&mut self, log_receiver: mpsc::Receiver<Vec<u8>>) -> Result<()> {
         if !self.nobanner {
             // Print the welcome banner.
             println!("{}", crate::helpers::welcome_message());
@@ -742,24 +706,31 @@ impl Start {
             }
         };
 
-
         // TODO(kaimast): start the display earlier and show sync progress.
         if !self.nodisplay && !self.nocdn {
             println!("🪧 The terminal UI will not start until the node has finished syncing from the CDN. If this step takes too long, consider restarting with `--nodisplay`.");
         }
 
-        let shutdown_tx = Some(shutdown_tx);
+        // Register the signal handler.
+        let signal_handler = SignalHandler::new();
 
         // Initialize the node.
-        match node_type {
-            NodeType::Validator => Node::new_validator(node_ip, self.bft, rest_ip, self.rest_rps, account, &trusted_peers, &trusted_validators, genesis, cdn, storage_mode, self.trusted_peers_only, dev_txs, self.dev, shutdown.clone(), shutdown_tx).await,
-            NodeType::Prover => Node::new_prover(node_ip, account, &trusted_peers, genesis, storage_mode, self.trusted_peers_only, self.dev, shutdown.clone(), shutdown_tx).await,
-            NodeType::Client => Node::new_client(node_ip, rest_ip, self.rest_rps, account, &trusted_peers, genesis, cdn, storage_mode, self.trusted_peers_only, self.dev, shutdown, shutdown_tx).await,
+        let node = match node_type {
+            NodeType::Validator => Node::new_validator(node_ip, self.bft, rest_ip, self.rest_rps, account, &trusted_peers, &trusted_validators, genesis, cdn, storage_mode, self.trusted_peers_only, dev_txs, self.dev, signal_handler.clone()).await,
+            NodeType::Prover => Node::new_prover(node_ip, account, &trusted_peers, genesis, storage_mode, self.trusted_peers_only, self.dev, signal_handler.clone()).await,
+            NodeType::Client => Node::new_client(node_ip, rest_ip, self.rest_rps, account, &trusted_peers, genesis, cdn, storage_mode, self.trusted_peers_only, self.dev, signal_handler.clone()).await,
             NodeType::BootstrapClient => Node::new_bootstrap_client(node_ip, account, *genesis.header(), self.dev).await,
+        }?;
+
+        if !self.nodisplay {
+            Display::start(node.clone(), log_receiver, signal_handler.clone()).with_context(|| "Failed to start the display")?;
         }
+
+        node.wait_for_signals(&signal_handler).await;
+        Ok(())
     }
 
-    /// Returns a runtime for the node.
+    /// Starts a rayon thread pool and tokio runtime for the node, and returns the tokio `Runtime`.
     fn runtime() -> Runtime {
         // Retrieve the number of cores.
         let num_cores = num_cpus::get();
@@ -770,14 +741,16 @@ impl Start {
         let (num_tokio_worker_threads, max_tokio_blocking_threads, num_rayon_cores_global) =
             (2 * num_cores, 512, num_cores);
 
-        // Initialize the parallelization parameters.
+        // Set up the rayon thread pool.
+        // A custom panic handler is not needed here, as rayon propagates the panic to the calling thread by default (except for `rayon::spawn` which we do not use).
         rayon::ThreadPoolBuilder::new()
             .stack_size(8 * 1024 * 1024)
             .num_threads(num_rayon_cores_global)
             .build_global()
             .unwrap();
 
-        // Initialize the runtime configuration.
+        // Set up the tokio Runtime.
+        // TODO(kaimast): set up a panic handler here for each worker thread once [`tokio::runtime::Builder::unhandled_panic`](https://docs.rs/tokio/latest/tokio/runtime/struct.Builder.html#method.unhandled_panic) is stabilized.
         runtime::Builder::new_multi_thread()
             .enable_all()
             .thread_stack_size(8 * 1024 * 1024)
diff --git a/display/Cargo.toml b/display/Cargo.toml
index 551e18fef..ed9e7081f 100644
--- a/display/Cargo.toml
+++ b/display/Cargo.toml
@@ -16,6 +16,11 @@ categories = [ "cryptography", "cryptography::cryptocurrencies", "os" ]
 license = "Apache-2.0"
 edition = "2024"
 
+[features]
+default = [ ]
+locktick = [ "snarkos-node/locktick", "snarkos-utilities/locktick" ]
+serial = [ "snarkvm/serial" ]
+
 [dependencies.anyhow]
 workspace = true
 
@@ -28,6 +33,9 @@ version = "0.29"
 [dependencies.snarkos-node]
 workspace = true
 
+[dependencies.snarkos-utilities]
+workspace = true
+
 [dependencies.snarkvm]
 workspace = true
 
diff --git a/display/src/lib.rs b/display/src/lib.rs
index 1a0305859..e97be68ea 100644
--- a/display/src/lib.rs
+++ b/display/src/lib.rs
@@ -22,6 +22,8 @@ mod tabs;
 use tabs::Tabs;
 
 use snarkos_node::Node;
+use snarkos_utilities::Stoppable;
+
 use snarkvm::prelude::Network;
 
 use anyhow::Result;
@@ -41,6 +43,7 @@ use ratatui::{
 };
 use std::{
     io,
+    sync::Arc,
     thread,
     time::{Duration, Instant},
 };
@@ -67,7 +70,7 @@ fn content_style() -> Style {
 
 impl<N: Network> Display<N> {
     /// Initializes a new display.
-    pub fn start(node: Node<N>, log_receiver: Receiver<Vec<u8>>) -> Result<()> {
+    pub fn start(node: Node<N>, log_receiver: Receiver<Vec<u8>>, stoppable: Arc<dyn Stoppable>) -> Result<()> {
         // Initialize the display.
         enable_raw_mode()?;
         let mut stdout = io::stdout();
@@ -84,7 +87,7 @@ impl<N: Network> Display<N> {
         };
 
         // Render the display.
-        let res = display.render(&mut terminal);
+        let res = display.render(&mut terminal, stoppable);
 
         // Terminate the display.
         disable_raw_mode()?;
@@ -102,7 +105,7 @@ impl<N: Network> Display<N> {
 
 impl<N: Network> Display<N> {
     /// Renders the display.
-    fn render<B: Backend>(&mut self, terminal: &mut Terminal<B>) -> io::Result<()> {
+    fn render<B: Backend>(&mut self, terminal: &mut Terminal<B>, stoppable: Arc<dyn Stoppable>) -> io::Result<()> {
         let mut last_tick = Instant::now();
         loop {
             terminal.draw(|f| self.draw(f))?;
@@ -114,11 +117,7 @@ impl<N: Network> Display<N> {
                 if let Event::Key(key) = event::read()? {
                     match key.code {
                         KeyCode::Esc => {
-                            // // TODO (howardwu): @ljedrz to implement a wrapping scope for Display within Node/Server.
-                            // #[allow(unused_must_use)]
-                            // {
-                            //     self.node.shut_down();
-                            // }
+                            stoppable.stop();
                             return Ok(());
                         }
                         KeyCode::Left => self.tabs.previous(),
diff --git a/node/Cargo.toml b/node/Cargo.toml
index 23b298eb5..063755359 100644
--- a/node/Cargo.toml
+++ b/node/Cargo.toml
@@ -17,8 +17,7 @@ license = "Apache-2.0"
 edition = "2024"
 
 [features]
-default = [ "parallel" ]
-parallel = [ ] #TODO(kaimast): is this still used?
+default = [ ]
 timer = [ "aleo-std/timer" ]
 locktick = [
   "dep:locktick",
@@ -30,6 +29,7 @@ locktick = [
   "snarkos-node-router/locktick",
   "snarkos-node-sync/locktick",
   "snarkos-node-tcp/locktick",
+  "snarkos-utilities/locktick",
   "snarkvm/locktick"
 ]
 metrics = [
@@ -50,7 +50,10 @@ cuda = [
   "snarkos-node-router/cuda",
   "snarkos-node-sync/cuda"
 ]
-serial = [ "snarkos-node-bft/serial" ]
+serial = [
+  "snarkvm/serial",
+  "snarkos-node-bft/serial"
+]
 test = []
 
 [dependencies.aleo-std]
@@ -83,9 +86,6 @@ workspace = true
 [dependencies.num_cpus]
 workspace = true
 
-[dependencies.once_cell]
-workspace = true
-
 [dependencies.parking_lot]
 workspace = true
 
@@ -130,6 +130,9 @@ workspace = true
 [dependencies.snarkos-node-tcp]
 workspace = true
 
+[dependencies.snarkos-utilities]
+workspace = true
+
 [dependencies.snarkvm]
 workspace = true
 
diff --git a/node/bft/Cargo.toml b/node/bft/Cargo.toml
index 8e14863af..3ae7b8d5f 100644
--- a/node/bft/Cargo.toml
+++ b/node/bft/Cargo.toml
@@ -25,6 +25,7 @@ locktick = [
   "snarkos-node-bft-storage-service/locktick",
   "snarkos-node-sync/locktick",
   "snarkos-node-tcp/locktick",
+  "snarkos-utilities/locktick",
   "snarkvm/locktick"
 ]
 metrics = [
@@ -44,7 +45,10 @@ test = [
   "snarkos-node-bft-ledger-service/test",
   "snarkos-node-bft-storage-service/test"
 ]
-serial = [ "snarkos-node-bft-ledger-service/serial" ]
+serial = [
+  "snarkos-node-metrics/serial",
+  "snarkos-node-bft-ledger-service/serial"
+]
 
 [dependencies.aleo-std]
 workspace = true
@@ -120,6 +124,9 @@ workspace = true
 [dependencies.snarkos-node-tcp]
 workspace = true
 
+[dependencies.snarkos-utilities]
+workspace = true
+
 [dependencies.snarkvm]
 workspace = true
 features = [ "utilities" ]
diff --git a/node/bft/examples/simple_node.rs b/node/bft/examples/simple_node.rs
index a39770d8e..99536287f 100644
--- a/node/bft/examples/simple_node.rs
+++ b/node/bft/examples/simple_node.rs
@@ -19,7 +19,6 @@ extern crate tracing;
 #[cfg(feature = "metrics")]
 extern crate snarkos_node_metrics as metrics;
 
-use aleo_std::StorageMode;
 use snarkos_account::Account;
 use snarkos_node_bft::{
     BFT,
@@ -30,6 +29,9 @@ use snarkos_node_bft::{
 use snarkos_node_bft_ledger_service::TranslucentLedgerService;
 use snarkos_node_bft_storage_service::BFTMemoryService;
 use snarkos_node_sync::BlockSync;
+use snarkos_utilities::SimpleStoppable;
+
+use aleo_std::StorageMode;
 use snarkvm::{
     console::{account::PrivateKey, algorithms::BHP256, types::Address},
     ledger::{
@@ -64,7 +66,7 @@ use std::{
     net::{IpAddr, Ipv4Addr, SocketAddr},
     path::PathBuf,
     str::FromStr,
-    sync::{Arc, Mutex, OnceLock, atomic::AtomicBool},
+    sync::{Arc, Mutex, OnceLock},
 };
 use tokio::{net::TcpListener, sync::oneshot};
 use tracing_subscriber::{
@@ -233,7 +235,7 @@ fn create_ledger(
     }
     let mut rng = TestRng::default();
     let gen_ledger = genesis_ledger(*gen_key, committee.clone(), balances.clone(), node_id, &mut rng);
-    Arc::new(TranslucentLedgerService::new(gen_ledger, Arc::new(AtomicBool::new(false))))
+    Arc::new(TranslucentLedgerService::new(gen_ledger, SimpleStoppable::new()))
 }
 
 pub type CurrentLedger = Ledger<CurrentNetwork, ConsensusMemory<CurrentNetwork>>;
diff --git a/node/bft/ledger-service/Cargo.toml b/node/bft/ledger-service/Cargo.toml
index 8bcb6ea3e..aa3febe6f 100644
--- a/node/bft/ledger-service/Cargo.toml
+++ b/node/bft/ledger-service/Cargo.toml
@@ -20,12 +20,18 @@ edition = "2024"
 default = [ ]
 ledger = [ "parking_lot", "rand", "rayon", "tokio", "tracing" ]
 ledger-write = [ ]
-locktick = [ "dep:locktick", "snarkvm/locktick" ]
+locktick = [
+  "dep:locktick",
+  "snarkvm/locktick",
+  "snarkos-utilities/locktick"
+]
 metrics = [ "dep:snarkos-node-metrics", "snarkvm/metrics" ]
 mock = [ "parking_lot", "tracing" ]
 prover = [ ]
 cuda = [ "snarkvm/cuda" ]
-serial = [ ]
+serial = [
+  "snarkvm/serial"
+]
 test = [ "mock", "translucent" ]
 translucent = [ "ledger" ]
 
@@ -48,6 +54,9 @@ optional = true
 workspace = true
 optional = true
 
+[dependencies.snarkos-utilities]
+workspace = true
+
 [dependencies.parking_lot]
 workspace = true
 optional = true
diff --git a/node/bft/ledger-service/src/ledger.rs b/node/bft/ledger-service/src/ledger.rs
index 09ff58632..95c63d662 100644
--- a/node/bft/ledger-service/src/ledger.rs
+++ b/node/bft/ledger-service/src/ledger.rs
@@ -14,6 +14,9 @@
 // limitations under the License.
 
 use crate::{LedgerService, fmt_id, spawn_blocking};
+
+use snarkos_utilities::Stoppable;
+
 use snarkvm::{
     ledger::{
         Ledger,
@@ -49,28 +52,20 @@ use parking_lot::RwLock;
 #[cfg(not(feature = "serial"))]
 use rayon::prelude::*;
 
-use std::{
-    fmt,
-    io::Read,
-    ops::Range,
-    sync::{
-        Arc,
-        atomic::{AtomicBool, Ordering},
-    },
-};
+use std::{fmt, io::Read, ops::Range, sync::Arc};
 
 /// A core ledger service.
 #[allow(clippy::type_complexity)]
 pub struct CoreLedgerService<N: Network, C: ConsensusStorage<N>> {
     ledger: Ledger<N, C>,
     latest_leader: Arc<RwLock<Option<(u64, Address<N>)>>>,
-    shutdown: Arc<AtomicBool>,
+    stoppable: Arc<dyn Stoppable>,
 }
 
 impl<N: Network, C: ConsensusStorage<N>> CoreLedgerService<N, C> {
     /// Initializes a new core ledger service.
-    pub fn new(ledger: Ledger<N, C>, shutdown: Arc<AtomicBool>) -> Self {
-        Self { ledger, latest_leader: Default::default(), shutdown }
+    pub fn new(ledger: Ledger<N, C>, stoppable: Arc<dyn Stoppable>) -> Self {
+        Self { ledger, latest_leader: Default::default(), stoppable }
     }
 }
 
@@ -360,7 +355,7 @@ impl<N: Network, C: ConsensusStorage<N>> LedgerService<N> for CoreLedgerService<
     #[cfg(feature = "ledger-write")]
     fn advance_to_next_block(&self, block: &Block<N>) -> Result<()> {
         // If the Ctrl-C handler registered the signal, then skip advancing to the next block.
-        if self.shutdown.load(Ordering::Acquire) {
+        if self.stoppable.is_stopped() {
             bail!("Skipping advancing to block {} - The node is shutting down", block.height());
         }
         // Advance to the next block.
diff --git a/node/bft/ledger-service/src/translucent.rs b/node/bft/ledger-service/src/translucent.rs
index 7c3a5fcd2..b4ffb988d 100644
--- a/node/bft/ledger-service/src/translucent.rs
+++ b/node/bft/ledger-service/src/translucent.rs
@@ -14,8 +14,9 @@
 // limitations under the License.
 
 use crate::{CoreLedgerService, LedgerService};
-use async_trait::async_trait;
-use indexmap::IndexMap;
+
+use snarkos_utilities::Stoppable;
+
 use snarkvm::{
     ledger::{
         Ledger,
@@ -27,11 +28,10 @@ use snarkvm::{
     },
     prelude::{Address, ConsensusVersion, Field, Network, Result, narwhal::BatchCertificate},
 };
-use std::{
-    fmt,
-    ops::Range,
-    sync::{Arc, atomic::AtomicBool},
-};
+
+use async_trait::async_trait;
+use indexmap::IndexMap;
+use std::{fmt, ops::Range, sync::Arc};
 
 pub struct TranslucentLedgerService<N: Network, C: ConsensusStorage<N>> {
     inner: CoreLedgerService<N, C>,
@@ -46,8 +46,8 @@ impl<N: Network, C: ConsensusStorage<N>> fmt::Debug for TranslucentLedgerService
 
 impl<N: Network, C: ConsensusStorage<N>> TranslucentLedgerService<N, C> {
     /// Initializes a new ledger service wrapper.
-    pub fn new(ledger: Ledger<N, C>, shutdown: Arc<AtomicBool>) -> Self {
-        Self { inner: CoreLedgerService::new(ledger, shutdown) }
+    pub fn new(ledger: Ledger<N, C>, stoppable: Arc<dyn Stoppable>) -> Self {
+        Self { inner: CoreLedgerService::new(ledger, stoppable) }
     }
 }
 
diff --git a/node/bft/src/gateway.rs b/node/bft/src/gateway.rs
index 945d9df5d..3f38f1d1f 100644
--- a/node/bft/src/gateway.rs
+++ b/node/bft/src/gateway.rs
@@ -87,7 +87,7 @@ use std::{
     io,
     net::{Ipv4Addr, SocketAddr, SocketAddrV4},
     sync::Arc,
-    time::Duration,
+    time::{Duration, Instant},
 };
 use tokio::{
     net::TcpStream,
@@ -821,12 +821,14 @@ impl<N: Network> Gateway<N> {
     fn initialize_heartbeat(&self) {
         let self_clone = self.clone();
         self.spawn(async move {
+            let start = Instant::now();
             // Sleep briefly to ensure the other nodes are ready to connect.
             tokio::time::sleep(std::time::Duration::from_millis(1000)).await;
             info!("Starting the heartbeat of the gateway...");
             loop {
                 // Process a heartbeat in the gateway.
-                self_clone.heartbeat().await;
+                let uptime = start.elapsed();
+                self_clone.heartbeat(uptime).await;
                 // Sleep for the heartbeat interval.
                 tokio::time::sleep(Duration::from_secs(15)).await;
             }
@@ -854,10 +856,13 @@ impl<N: Network> Gateway<N> {
 }
 
 impl<N: Network> Gateway<N> {
+    /// The uptime after which nodes log a warning about missing validator connections.
+    const MISSING_VALIDATOR_CONNECTIONS_GRACE_PERIOD: Duration = Duration::from_secs(60);
+
     /// Handles the heartbeat request.
-    async fn heartbeat(&self) {
+    async fn heartbeat(&self, uptime: Duration) {
         // Log the connected validators.
-        self.log_connected_validators();
+        self.log_connected_validators(uptime);
         // Log the validator participation scores.
         #[cfg(feature = "telemetry")]
         self.log_participation_scores();
@@ -874,7 +879,7 @@ impl<N: Network> Gateway<N> {
     }
 
     /// Logs the connected validators.
-    fn log_connected_validators(&self) {
+    fn log_connected_validators(&self, uptime: Duration) {
         // Retrieve the connected validators and current committee.
         let connected_validators = self.connected_peers();
         let committee = match self.ledger.current_committee() {
@@ -937,7 +942,12 @@ impl<N: Network> Gateway<N> {
         }
 
         if !committee.is_quorum_threshold_reached(&connected_validator_addresses) {
-            error!("Not connected to a quorum of validators");
+            // Not being connected to a quorum of validators is begning during startup.
+            if uptime > Self::MISSING_VALIDATOR_CONNECTIONS_GRACE_PERIOD {
+                error!("Not connected to a quorum of validators");
+            } else {
+                debug!("Not connected to a quorum of validators");
+            }
         }
     }
 
diff --git a/node/bft/src/sync/mod.rs b/node/bft/src/sync/mod.rs
index 4d84da002..a8c097105 100644
--- a/node/bft/src/sync/mod.rs
+++ b/node/bft/src/sync/mod.rs
@@ -1020,6 +1020,7 @@ mod tests {
 
     use snarkos_account::Account;
     use snarkos_node_sync::BlockSync;
+    use snarkos_utilities::SimpleStoppable;
     use snarkvm::{
         console::{
             account::{Address, PrivateKey},
@@ -1072,7 +1073,7 @@ mod tests {
         let genesis_clone = genesis.clone();
         let ledger = spawn_blocking!(CurrentLedger::load(genesis_clone, StorageMode::new_test(None))).unwrap();
         // Initialize the ledger.
-        let core_ledger = Arc::new(CoreLedgerService::new(ledger.clone(), Default::default()));
+        let core_ledger = Arc::new(CoreLedgerService::new(ledger.clone(), SimpleStoppable::new()));
 
         // Sample 5 rounds of batch certificates starting at the genesis round from a static set of 4 authors.
         let (round_to_certificates_map, committee) = {
@@ -1252,10 +1253,10 @@ mod tests {
         spawn_blocking!(ledger.advance_to_next_block(&block))?;
 
         // Initialize the syncing ledger.
-        let syncing_ledger = spawn_blocking!(CurrentLedger::load(genesis, StorageMode::new_test(None))).unwrap();
-        let syncing_ledger = Arc::new(CoreLedgerService::new(syncing_ledger, Default::default()));
+        let syncing_ledger = spawn_blocking!(CurrentLedger::load(genesis, StorageMode::Test(None))).unwrap();
+        let syncing_ledger = Arc::new(CoreLedgerService::new(syncing_ledger, SimpleStoppable::new()));
+
         // Initialize the gateway.
-        let storage_mode = StorageMode::new_test(None);
         let gateway = Gateway::new(
             account.clone(),
             storage.clone(),
@@ -1263,7 +1264,7 @@ mod tests {
             None,
             &[],
             false,
-            storage_mode,
+            StorageMode::new_test(None),
             None,
         )?;
         // Initialize the block synchronization logic.
@@ -1315,7 +1316,7 @@ mod tests {
         // Initialize the ledger with the genesis block.
         let ledger = spawn_blocking!(CurrentLedger::load(genesis, StorageMode::new_test(None))).unwrap();
         // Initialize the ledger.
-        let core_ledger = Arc::new(CoreLedgerService::new(ledger, Default::default()));
+        let core_ledger = Arc::new(CoreLedgerService::new(ledger.clone(), SimpleStoppable::new()));
         // Sample rounds of batch certificates starting at the genesis round from a static set of 4 authors.
         let (round_to_certificates_map, committee) = {
             // Initialize the committee.
diff --git a/node/bft/tests/bft_e2e.rs b/node/bft/tests/bft_e2e.rs
index 0b89aa7ff..c1e788a95 100644
--- a/node/bft/tests/bft_e2e.rs
+++ b/node/bft/tests/bft_e2e.rs
@@ -156,10 +156,6 @@ async fn test_quorum_threshold() {
     const TARGET_ROUND: u64 = 4;
     let net = network.clone();
     deadline!(Duration::from_secs(20), move || { net.is_round_reached(TARGET_ROUND) });
-
-    tokio::task::spawn_blocking(move || {
-        drop(network);
-    });
 }
 
 #[tokio::test(flavor = "multi_thread")]
@@ -194,10 +190,6 @@ async fn test_quorum_break() {
 
     // Check the nodes have stopped advancing through the rounds.
     assert!(network.is_halted().await);
-
-    tokio::task::spawn_blocking(move || {
-        drop(network);
-    });
 }
 
 #[tokio::test(flavor = "multi_thread")]
@@ -259,10 +251,6 @@ async fn test_leader_election_consistency() {
         // Assert that all leaders are equal
         assert!(leaders.iter().all_equal());
     }
-
-    tokio::task::spawn_blocking(move || {
-        drop(network);
-    });
 }
 
 #[tokio::test(flavor = "multi_thread")]
diff --git a/node/bft/tests/common/primary.rs b/node/bft/tests/common/primary.rs
index 54dc04fd9..3e65ec77d 100644
--- a/node/bft/tests/common/primary.rs
+++ b/node/bft/tests/common/primary.rs
@@ -18,6 +18,7 @@ use crate::common::{
     TranslucentLedgerService,
     utils::{fire_unconfirmed_solutions, fire_unconfirmed_transactions, initialize_logger},
 };
+
 use snarkos_account::Account;
 use snarkos_node_bft::{
     BFT,
@@ -29,6 +30,8 @@ use snarkos_node_bft::{
 use snarkos_node_bft_storage_service::BFTMemoryService;
 use snarkos_node_network::PeerPoolHandling;
 use snarkos_node_sync::BlockSync;
+use snarkos_utilities::SimpleStoppable;
+
 use snarkvm::{
     console::{
         account::{Address, PrivateKey},
@@ -160,7 +163,7 @@ impl TestNetwork {
         for (id, account) in accounts.into_iter().enumerate() {
             let gen_ledger =
                 genesis_ledger(gen_key, committee.clone(), balances.clone(), bonded_balances.clone(), &mut rng);
-            let ledger = Arc::new(TranslucentLedgerService::new(gen_ledger, Default::default()));
+            let ledger = Arc::new(TranslucentLedgerService::new(gen_ledger, SimpleStoppable::new()));
             let storage = Storage::new(
                 ledger.clone(),
                 Arc::new(BFTMemoryService::new()),
diff --git a/node/bft/tests/common/utils.rs b/node/bft/tests/common/utils.rs
index 27d527a38..c9ba1d2ae 100644
--- a/node/bft/tests/common/utils.rs
+++ b/node/bft/tests/common/utils.rs
@@ -22,7 +22,9 @@ use snarkos_node_bft::{
     helpers::{PrimarySender, Storage},
 };
 
-use snarkos_node_bft_storage_service::BFTMemoryService;
+use snarkos_node_bft::storage_service::BFTMemoryService;
+use snarkos_utilities::SimpleStoppable;
+
 use snarkvm::{
     console::account::Address,
     ledger::{
@@ -202,7 +204,7 @@ pub fn sample_ledger(
 
     let gen_ledger =
         primary::genesis_ledger(gen_key, committee.clone(), balances.clone(), bonded_balances.clone(), rng);
-    Arc::new(TranslucentLedgerService::new(gen_ledger, Default::default()))
+    Arc::new(TranslucentLedgerService::new(gen_ledger, SimpleStoppable::new()))
 }
 
 /// Samples a new storage with the given ledger.
diff --git a/node/bft/tests/components/worker.rs b/node/bft/tests/components/worker.rs
index 921aa8ae2..aca874767 100644
--- a/node/bft/tests/components/worker.rs
+++ b/node/bft/tests/components/worker.rs
@@ -36,11 +36,10 @@ async fn test_resend_transmission_request() {
     // Initialize the accounts and the committee.
     let (accounts, committee) = new_test_committee(num_nodes, &mut rng);
     // Sample a ledger.
-    let accounts_ = accounts.clone();
-    let committee_ = committee.clone();
-    let ledger = tokio::task::spawn_blocking(move || {
-        sample_ledger(&accounts_, &committee_, &mut rng)
-    }).await.unwrap();
+    let ledger = {
+        let accounts = accounts.clone();
+        tokio::task::spawn_blocking(move || sample_ledger(&accounts, &committee, &mut rng)).await.unwrap()
+    };
     // Sample a worker.
     let worker = sample_worker(0, accounts[0].clone(), ledger.clone());
 
@@ -122,11 +121,10 @@ async fn test_flood_transmission_requests() {
     // Initialize the accounts and the committee.
     let (accounts, committee) = new_test_committee(num_nodes, &mut rng);
     // Sample a ledger.
-    let accounts_ = accounts.clone();
-    let committee_ = committee.clone();
-    let ledger = tokio::task::spawn_blocking(move || {
-        sample_ledger(&accounts_, &committee_, &mut rng)
-    }).await.unwrap();
+    let ledger = {
+        let accounts = accounts.clone();
+        tokio::task::spawn_blocking(move || sample_ledger(&accounts, &committee, &mut rng)).await.unwrap()
+    };
     // Sample a worker.
     let worker = sample_worker(0, accounts[0].clone(), ledger.clone());
 
diff --git a/node/bft/tests/narwhal_e2e.rs b/node/bft/tests/narwhal_e2e.rs
index 8072f2592..ea731fb70 100644
--- a/node/bft/tests/narwhal_e2e.rs
+++ b/node/bft/tests/narwhal_e2e.rs
@@ -108,10 +108,6 @@ async fn test_quorum_threshold() {
     const TARGET_ROUND: u64 = 4;
     let net = network.clone();
     deadline!(Duration::from_secs(20), move || { net.is_round_reached(TARGET_ROUND) });
-
-    tokio::task::spawn_blocking(move || {
-        drop(network);
-    });
 }
 
 #[tokio::test(flavor = "multi_thread")]
@@ -146,10 +142,6 @@ async fn test_quorum_break() {
 
     // Check the nodes have stopped advancing through the rounds.
     assert!(network.is_halted().await);
-
-    tokio::task::spawn_blocking(move || {
-        drop(network);
-    });
 }
 
 #[tokio::test(flavor = "multi_thread")]
@@ -186,8 +178,4 @@ async fn test_storage_coherence() {
     // check only up to 3 rounds before the the target round, because the network advances when
     // quorum is reached, not when all the nodes have completed the round and received certificates.
     assert!(network.is_certificate_round_coherent(1..TARGET_ROUND - 2));
-
-    tokio::task::spawn_blocking(move || {
-        drop(network);
-    });
 }
diff --git a/node/cdn/Cargo.toml b/node/cdn/Cargo.toml
index 1269103ed..35ac5c8ff 100644
--- a/node/cdn/Cargo.toml
+++ b/node/cdn/Cargo.toml
@@ -17,9 +17,17 @@ license = "Apache-2.0"
 edition = "2024"
 
 [features]
-default = [ "parallel" ]
-locktick = [ "dep:locktick", "snarkvm/locktick" ]
-parallel = [ "rayon" ]
+default = [ ]
+serial = [ 
+  "snarkos-node-metrics/locktick",
+  "snarkvm/serial",
+]
+locktick = [
+  "dep:locktick",
+  "snarkvm/locktick",
+  "snarkos-node-metrics/locktick",
+  "snarkos-utilities/locktick"
+]
 cuda = [ "snarkvm/cuda" ]
 metrics = [ "dep:snarkos-node-metrics" ]
 
@@ -45,9 +53,11 @@ workspace = true
 optional = true
 features = [ "metrics" ]
 
+[dependencies.snarkos-utilities]
+workspace = true
+
 [dependencies.rayon]
 workspace = true
-optional = true
 
 [dependencies.reqwest]
 version = "0.12"
diff --git a/node/cdn/src/blocks.rs b/node/cdn/src/blocks.rs
index 4b489d195..b984ac744 100644
--- a/node/cdn/src/blocks.rs
+++ b/node/cdn/src/blocks.rs
@@ -17,6 +17,8 @@
 // https://github.com/rust-lang/rust-clippy/issues/6446
 #![allow(clippy::await_holding_lock)]
 
+use snarkos_utilities::{SignalHandler, Stoppable};
+
 use snarkvm::prelude::{
     Deserialize,
     DeserializeOwned,
@@ -87,11 +89,11 @@ impl CdnBlockSync {
     pub fn new<N: Network, C: ConsensusStorage<N>>(
         base_url: http::Uri,
         ledger: Ledger<N, C>,
-        shutdown: Arc<AtomicBool>,
+        stoppable: Arc<SignalHandler>,
     ) -> Self {
         let task = {
             let base_url = base_url.clone();
-            tokio::spawn(async move { Self::worker(base_url, ledger, shutdown).await })
+            tokio::spawn(async move { Self::worker(base_url, ledger, stoppable).await })
         };
 
         debug!("Started sync from CDN at {base_url}");
@@ -119,13 +121,13 @@ impl CdnBlockSync {
     async fn worker<N: Network, C: ConsensusStorage<N>>(
         base_url: http::Uri,
         ledger: Ledger<N, C>,
-        shutdown: Arc<AtomicBool>,
+        stoppable: Arc<dyn Stoppable>,
     ) -> SyncResult {
         // Fetch the node height.
         let start_height = ledger.latest_height() + 1;
         // Load the blocks from the CDN into the ledger.
         let ledger_clone = ledger.clone();
-        let result = load_blocks(&base_url, start_height, None, shutdown, move |block: Block<N>| {
+        let result = load_blocks(&base_url, start_height, None, stoppable, move |block: Block<N>| {
             ledger_clone.advance_to_next_block(&block)
         })
         .await;
@@ -172,7 +174,7 @@ pub async fn load_blocks<N: Network>(
     base_url: &http::Uri,
     start_height: u32,
     end_height: Option<u32>,
-    shutdown: Arc<AtomicBool>,
+    stoppable: Arc<dyn Stoppable>,
     process: impl FnMut(Block<N>) -> Result<()> + Clone + Send + Sync + 'static,
 ) -> Result<u32, (u32, anyhow::Error)> {
     // Create a Client to maintain a connection pool throughout the sync.
@@ -225,16 +227,19 @@ pub async fn load_blocks<N: Network>(
     // Spawn a background task responsible for concurrent downloads.
     let pending_blocks_clone = pending_blocks.clone();
     let base_url = base_url.to_owned();
-    let shutdown_clone = shutdown.clone();
-    tokio::spawn(async move {
-        download_block_bundles(client, &base_url, cdn_start, cdn_end, pending_blocks_clone, shutdown_clone).await;
-    });
+
+    {
+        let stoppable = stoppable.clone();
+        tokio::spawn(async move {
+            download_block_bundles(client, &base_url, cdn_start, cdn_end, pending_blocks_clone, stoppable).await;
+        });
+    }
 
     // A loop for inserting the pending blocks into the ledger.
     let mut current_height = start_height.saturating_sub(1);
     while current_height < end_height - 1 {
         // If we are instructed to shut down, abort.
-        if shutdown.load(Ordering::Acquire) {
+        if stoppable.is_stopped() {
             info!("Stopping block sync at {} - shutting down", current_height);
             // We can shut down cleanly from here, as the node hasn't been started yet.
             std::process::exit(0);
@@ -269,12 +274,12 @@ pub async fn load_blocks<N: Network>(
 
         // Attempt to advance the ledger using the CDN block bundle.
         let mut process_clone = process.clone();
-        let shutdown_clone = shutdown.clone();
+        let stoppable_clone = stoppable.clone();
         current_height = tokio::task::spawn_blocking(move || {
             threadpool.install(|| {
                 for block in next_blocks.into_iter().filter(|b| (start_height..end_height).contains(&b.height())) {
                     // If we are instructed to shut down, abort.
-                    if shutdown_clone.load(Ordering::Relaxed) {
+                    if stoppable_clone.is_stopped() {
                         info!("Stopping block sync at {} - the node is shutting down", current_height);
                         // We can shut down cleanly from here, as the node hasn't been started yet.
                         std::process::exit(0);
@@ -314,7 +319,7 @@ async fn download_block_bundles<N: Network>(
     cdn_start: u32,
     cdn_end: u32,
     pending_blocks: Arc<TMutex<Vec<Block<N>>>>,
-    shutdown: Arc<AtomicBool>,
+    stoppable: Arc<dyn Stoppable>,
 ) {
     // Keep track of the number of concurrent requests.
     let active_requests: Arc<AtomicU32> = Default::default();
@@ -322,7 +327,7 @@ async fn download_block_bundles<N: Network>(
     let mut start = cdn_start;
     while start < cdn_end - 1 {
         // If we are instructed to shut down, stop downloading.
-        if shutdown.load(Ordering::Acquire) {
+        if stoppable.is_stopped() {
             break;
         }
 
@@ -356,7 +361,7 @@ async fn download_block_bundles<N: Network>(
             let base_url_clone = base_url.clone();
             let pending_blocks_clone = pending_blocks.clone();
             let active_requests_clone = active_requests.clone();
-            let shutdown_clone = shutdown.clone();
+            let stoppable_clone = stoppable.clone();
             tokio::spawn(async move {
                 // Increment the number of active requests.
                 active_requests_clone.fetch_add(1, Ordering::Relaxed);
@@ -392,7 +397,7 @@ async fn download_block_bundles<N: Network>(
                             attempts += 1;
                             if attempts > MAXIMUM_REQUEST_ATTEMPTS {
                                 warn!("Maximum number of requests to {blocks_url} reached - shutting down...");
-                                shutdown_clone.store(true, Ordering::Relaxed);
+                                stoppable_clone.stop();
                                 break;
                             }
                             tokio::time::sleep(Duration::from_secs(attempts as u64 * 10)).await;
@@ -553,8 +558,10 @@ fn log_progress<const OBJECTS_PER_FILE: u32>(
 
 #[cfg(test)]
 mod tests {
-    use super::{BLOCKS_PER_FILE, CDN_BASE_URL, cdn_height, log_progress};
-    use crate::load_blocks;
+    use super::{BLOCKS_PER_FILE, CDN_BASE_URL, cdn_height, load_blocks, log_progress};
+
+    use snarkos_utilities::SimpleStoppable;
+
     use snarkvm::prelude::{MainnetV0, block::Block};
 
     use http::Uri;
@@ -576,7 +583,7 @@ mod tests {
         let rt = tokio::runtime::Runtime::new().unwrap();
         rt.block_on(async {
             let completed_height =
-                load_blocks(&testnet_cdn_url, start, end, Default::default(), process).await.unwrap();
+                load_blocks(&testnet_cdn_url, start, end, SimpleStoppable::new(), process).await.unwrap();
             assert_eq!(blocks.read().len(), expected);
             if expected > 0 {
                 assert_eq!(blocks.read().last().unwrap().height(), completed_height);
diff --git a/node/consensus/Cargo.toml b/node/consensus/Cargo.toml
index 49753927f..7d6b75a66 100644
--- a/node/consensus/Cargo.toml
+++ b/node/consensus/Cargo.toml
@@ -28,8 +28,11 @@ locktick = [
 metrics = [ "dep:snarkos-node-metrics" ]
 telemetry = [ "snarkos-node-bft/telemetry" ]
 cuda = [ "snarkvm/cuda", "snarkos-account/cuda", "snarkos-node-bft-ledger-service/cuda" ]
-serial = [ "snarkos-node-bft-ledger-service/serial" ]
-
+serial = [ 
+  "snarkos-node-bft-ledger-service/serial",
+  "snarkos-node-metrics/serial",
+  "snarkvm/serial"
+]
 
 [dependencies.aleo-std]
 workspace = true
diff --git a/node/metrics/Cargo.toml b/node/metrics/Cargo.toml
index 1c72cf774..3f89b6a61 100644
--- a/node/metrics/Cargo.toml
+++ b/node/metrics/Cargo.toml
@@ -17,10 +17,13 @@ license = "Apache-2.0"
 edition = "2024"
 
 [features]
-default = [ "rayon", "snarkvm/metrics" ]
-locktick = [ "dep:locktick", "snarkvm/locktick" ]
+default = [ "metrics" ]
+locktick = [ 
+  "dep:locktick",
+  "snarkvm/locktick"
+]
 metrics = [ "snarkvm/metrics" ]
-serial = [ ]
+serial = [ "snarkvm/serial" ]
 
 [dependencies.locktick]
 workspace = true
@@ -35,7 +38,6 @@ workspace = true
 
 [dependencies.rayon]
 workspace = true
-optional = true
 
 [dependencies.snarkvm]
 workspace = true
diff --git a/node/router/Cargo.toml b/node/router/Cargo.toml
index 8dda700f9..a5e91536f 100644
--- a/node/router/Cargo.toml
+++ b/node/router/Cargo.toml
@@ -18,7 +18,12 @@ edition = "2024"
 
 [features]
 test = [ ]
-locktick = [ "dep:locktick", "snarkos-node-tcp/locktick", "snarkvm/locktick" ]
+locktick = [
+  "dep:locktick",
+  "snarkos-node-tcp/locktick",
+  "snarkvm/locktick",
+  "snarkos-node-network/locktick"
+]
 metrics = [ "dep:snarkos-node-metrics" ]
 cuda = [ "snarkvm/cuda", "snarkos-account/cuda" ]
 serial = [ "snarkos-node-bft-ledger-service/serial" ]
diff --git a/node/src/bootstrap_client/mod.rs b/node/src/bootstrap_client/mod.rs
index e594475d3..cca28578f 100644
--- a/node/src/bootstrap_client/mod.rs
+++ b/node/src/bootstrap_client/mod.rs
@@ -21,6 +21,7 @@ use crate::tcp::{self, Tcp};
 use snarkos_account::Account;
 use snarkos_node_network::{ConnectionMode, Peer, Resolver};
 use snarkos_node_tcp::{P2P, protocols::*};
+use snarkos_utilities::SignalHandler;
 use snarkvm::{
     ledger::committee::Committee,
     prelude::{Address, Field, Header, Network, PrivateKey, ViewKey},
@@ -230,4 +231,12 @@ impl<N: Network> BootstrapClient<N> {
             let _ = shutdown_tx.send(());
         }
     }
+
+    /// Blocks until a shutdown signal was received or manual shutdown was triggered.
+    pub async fn wait_for_signals(&self, handler: &SignalHandler) {
+        handler.wait_for_signals().await;
+
+        // If the node is already initialized, then shut it down.
+        self.shut_down().await;
+    }
 }
diff --git a/node/src/client/mod.rs b/node/src/client/mod.rs
index 80c87a8df..2ce274da8 100644
--- a/node/src/client/mod.rs
+++ b/node/src/client/mod.rs
@@ -15,11 +15,13 @@
 
 mod router;
 
-use crate::traits::NodeInterface;
+use crate::{
+    bft::{events::DataBlocks, helpers::fmt_id, ledger_service::CoreLedgerService, spawn_blocking},
+    cdn::CdnBlockSync,
+    traits::NodeInterface,
+};
 
 use snarkos_account::Account;
-use snarkos_node_bft::{events::DataBlocks, helpers::fmt_id, ledger_service::CoreLedgerService, spawn_blocking};
-use snarkos_node_cdn::CdnBlockSync;
 use snarkos_node_network::NodeType;
 use snarkos_node_rest::Rest;
 use snarkos_node_router::{
@@ -35,6 +37,8 @@ use snarkos_node_tcp::{
     P2P,
     protocols::{Disconnect, Handshake, OnConnect, Reading},
 };
+use snarkos_utilities::{SignalHandler, Stoppable};
+
 use snarkvm::{
     console::network::Network,
     ledger::{
@@ -62,7 +66,6 @@ use std::{
     sync::{
         Arc,
         atomic::{
-            AtomicBool,
             AtomicUsize,
             Ordering::{Acquire, Relaxed},
         },
@@ -70,7 +73,6 @@ use std::{
     time::Duration,
 };
 use tokio::{
-    sync::oneshot,
     task::JoinHandle,
     time::{sleep, timeout},
 };
@@ -124,10 +126,10 @@ pub struct Client<N: Network, C: ConsensusStorage<N>> {
     num_verifying_executions: Arc<AtomicUsize>,
     /// The spawned handles.
     handles: Arc<Mutex<Vec<JoinHandle<()>>>>,
-    /// The shutdown signal.
-    shutdown: Arc<AtomicBool>,
     /// Keeps track of sending pings.
     ping: Arc<Ping<N>>,
+    /// The signal handling logic.
+    signal_handler: Arc<SignalHandler>,
 }
 
 impl<N: Network, C: ConsensusStorage<N>> Client<N, C> {
@@ -143,12 +145,8 @@ impl<N: Network, C: ConsensusStorage<N>> Client<N, C> {
         storage_mode: StorageMode,
         trusted_peers_only: bool,
         dev: Option<u16>,
-        shutdown: Arc<AtomicBool>,
-        shutdown_tx: Option<oneshot::Sender<()>>,
+        signal_handler: Arc<SignalHandler>,
     ) -> Result<Self> {
-        // Initialize the signal handler.
-        let signal_node = Self::handle_signals(shutdown.clone(), shutdown_tx);
-
         // Initialize the ledger.
         let ledger = {
             let storage_mode = storage_mode.clone();
@@ -159,8 +157,7 @@ impl<N: Network, C: ConsensusStorage<N>> Client<N, C> {
         .with_context(|| "Failed to initialize the ledger")?;
 
         // Initialize the ledger service.
-        let ledger_service = Arc::new(CoreLedgerService::<N, C>::new(ledger.clone(), shutdown.clone()));
-
+        let ledger_service = Arc::new(CoreLedgerService::<N, C>::new(ledger.clone(), signal_handler.clone()));
         // Initialize the node router.
         let router = Router::new(
             node_ip,
@@ -198,13 +195,13 @@ impl<N: Network, C: ConsensusStorage<N>> Client<N, C> {
             num_verifying_deploys: Default::default(),
             num_verifying_executions: Default::default(),
             handles: Default::default(),
-            shutdown: shutdown.clone(),
+            signal_handler: signal_handler.clone(),
         };
 
         // Perform sync with CDN (if enabled).
         let cdn_sync = cdn.map(|base_url| {
             trace!("CDN sync is enabled");
-            Arc::new(CdnBlockSync::new(base_url, ledger.clone(), shutdown))
+            Arc::new(CdnBlockSync::new(base_url, ledger.clone(), signal_handler))
         });
 
         // Initialize the REST server.
@@ -236,8 +233,6 @@ impl<N: Network, C: ConsensusStorage<N>> Client<N, C> {
         node.initialize_execute_verification();
         // Initialize the notification message loop.
         node.handles.lock().push(crate::start_notification_message_loop());
-        // Pass the node to the signal handler.
-        let _ = signal_node.set(node.clone());
         // Return the node.
         Ok(node)
     }
@@ -269,11 +264,9 @@ impl<N: Network, C: ConsensusStorage<N>> Client<N, C> {
         // Start the block request generation loop (outgoing).
         let self_ = self.clone();
         self.spawn(async move {
-            while !self_.shutdown.load(std::sync::atomic::Ordering::Acquire) {
+            while !self_.signal_handler.is_stopped() {
                 // Perform the sync routine.
                 self_.try_issuing_block_requests().await;
-
-                // Rate limiting happens in [`Self::send_block_requests`] and no additional sleeps are needed here
             }
 
             info!("Stopped block request generation");
@@ -282,7 +275,7 @@ impl<N: Network, C: ConsensusStorage<N>> Client<N, C> {
         // Start the block response processing loop (incoming).
         let self_ = self.clone();
         self.spawn(async move {
-            while !self_.shutdown.load(std::sync::atomic::Ordering::Acquire) {
+            while !self_.signal_handler.is_stopped() {
                 // Wait until there is something to do or until the timeout.
                 let _ = timeout(Self::MAX_SYNC_INTERVAL, self_.sync.wait_for_block_responses()).await;
 
@@ -395,7 +388,7 @@ impl<N: Network, C: ConsensusStorage<N>> Client<N, C> {
         self.spawn(async move {
             loop {
                 // If the Ctrl-C handler registered the signal, stop the node.
-                if node.shutdown.load(Acquire) {
+                if node.signal_handler.is_stopped() {
                     info!("Shutting down solution verification");
                     break;
                 }
@@ -469,7 +462,7 @@ impl<N: Network, C: ConsensusStorage<N>> Client<N, C> {
         self.spawn(async move {
             loop {
                 // If the Ctrl-C handler registered the signal, stop the node.
-                if node.shutdown.load(Acquire) {
+                if node.signal_handler.is_stopped() {
                     info!("Shutting down deployment verification");
                     break;
                 }
@@ -537,7 +530,7 @@ impl<N: Network, C: ConsensusStorage<N>> Client<N, C> {
         self.spawn(async move {
             loop {
                 // If the Ctrl-C handler registered the signal, stop the node.
-                if node.shutdown.load(Acquire) {
+                if node.signal_handler.is_stopped() {
                     info!("Shutting down execution verification");
                     break;
                 }
@@ -615,7 +608,6 @@ impl<N: Network, C: ConsensusStorage<N>> NodeInterface<N> for Client<N, C> {
 
         // Shut down the node.
         trace!("Shutting down the node...");
-        self.shutdown.store(true, std::sync::atomic::Ordering::Release);
 
         // Abort the tasks.
         trace!("Shutting down the client...");
diff --git a/node/src/node.rs b/node/src/node.rs
index 14e10f0f7..93ffeb645 100644
--- a/node/src/node.rs
+++ b/node/src/node.rs
@@ -13,10 +13,19 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-use crate::{BootstrapClient, Client, Prover, Validator, traits::NodeInterface};
+use crate::{
+    BootstrapClient,
+    Client,
+    Prover,
+    Validator,
+    network::{NodeType, Peer, PeerPoolHandling},
+    router::Outbound,
+    traits::NodeInterface,
+};
+
 use snarkos_account::Account;
-use snarkos_node_network::{NodeType, Peer, PeerPoolHandling};
-use snarkos_node_router::Outbound;
+use snarkos_utilities::SignalHandler;
+
 use snarkvm::prelude::{
     Address,
     Header,
@@ -30,16 +39,12 @@ use snarkvm::prelude::{
 
 use aleo_std::StorageMode;
 use anyhow::Result;
+
 #[cfg(feature = "locktick")]
 use locktick::parking_lot::RwLock;
 #[cfg(not(feature = "locktick"))]
 use parking_lot::RwLock;
-use std::{
-    collections::HashMap,
-    net::SocketAddr,
-    sync::{Arc, atomic::AtomicBool},
-};
-use tokio::sync::oneshot;
+use std::{collections::HashMap, net::SocketAddr, sync::Arc};
 
 #[derive(Clone)]
 pub enum Node<N: Network> {
@@ -69,8 +74,7 @@ impl<N: Network> Node<N> {
         trusted_peers_only: bool,
         dev_txs: bool,
         dev: Option<u16>,
-        shutdown: Arc<AtomicBool>,
-        shutdown_tx: Option<oneshot::Sender<()>>,
+        signal_handler: Arc<SignalHandler>,
     ) -> Result<Self> {
         Ok(Self::Validator(Arc::new(
             Validator::new(
@@ -87,8 +91,7 @@ impl<N: Network> Node<N> {
                 trusted_peers_only,
                 dev_txs,
                 dev,
-                shutdown,
-                shutdown_tx,
+                signal_handler,
             )
             .await?,
         )))
@@ -103,8 +106,7 @@ impl<N: Network> Node<N> {
         storage_mode: StorageMode,
         trusted_peers_only: bool,
         dev: Option<u16>,
-        shutdown: Arc<AtomicBool>,
-        shutdown_tx: Option<oneshot::Sender<()>>,
+        signal_handler: Arc<SignalHandler>,
     ) -> Result<Self> {
         Ok(Self::Prover(Arc::new(
             Prover::new(
@@ -115,8 +117,7 @@ impl<N: Network> Node<N> {
                 storage_mode,
                 trusted_peers_only,
                 dev,
-                shutdown,
-                shutdown_tx,
+                signal_handler,
             )
             .await?,
         )))
@@ -134,8 +135,7 @@ impl<N: Network> Node<N> {
         storage_mode: StorageMode,
         trusted_peers_only: bool,
         dev: Option<u16>,
-        shutdown: Arc<AtomicBool>,
-        shutdown_tx: Option<oneshot::Sender<()>>,
+        signal_handler: Arc<SignalHandler>,
     ) -> Result<Self> {
         Ok(Self::Client(Arc::new(
             Client::new(
@@ -149,8 +149,7 @@ impl<N: Network> Node<N> {
                 storage_mode,
                 trusted_peers_only,
                 dev,
-                shutdown,
-                shutdown_tx,
+                signal_handler,
             )
             .await?,
         )))
@@ -277,4 +276,14 @@ impl<N: Network> Node<N> {
             Self::BootstrapClient(node) => node.shut_down().await,
         }
     }
+
+    /// Waits until the node receives a signal.
+    pub async fn wait_for_signals(&self, signal_handler: &SignalHandler) {
+        match self {
+            Self::Validator(node) => node.wait_for_signals(signal_handler).await,
+            Self::Prover(node) => node.wait_for_signals(signal_handler).await,
+            Self::Client(node) => node.wait_for_signals(signal_handler).await,
+            Self::BootstrapClient(node) => node.wait_for_signals(signal_handler).await,
+        }
+    }
 }
diff --git a/node/src/prover/mod.rs b/node/src/prover/mod.rs
index 433734ef8..5cedb19ae 100644
--- a/node/src/prover/mod.rs
+++ b/node/src/prover/mod.rs
@@ -15,10 +15,15 @@
 
 mod router;
 
-use crate::traits::NodeInterface;
+use crate::{
+    bft::ledger_service::ProverLedgerService,
+    sync::{BlockSync, Ping},
+    traits::NodeInterface,
+};
+
 use snarkos_account::Account;
-use snarkos_node_bft::ledger_service::ProverLedgerService;
 use snarkos_node_network::{NodeType, PeerPoolHandling};
+
 use snarkos_node_router::{
     Heartbeat,
     Inbound,
@@ -27,11 +32,12 @@ use snarkos_node_router::{
     Routing,
     messages::{Message, UnconfirmedSolution},
 };
-use snarkos_node_sync::{BlockSync, Ping};
 use snarkos_node_tcp::{
     P2P,
     protocols::{Disconnect, Handshake, OnConnect, Reading},
 };
+use snarkos_utilities::{SignalHandler, Stoppable};
+
 use snarkvm::{
     ledger::narwhal::Data,
     prelude::{
@@ -57,10 +63,10 @@ use std::{
     net::SocketAddr,
     sync::{
         Arc,
-        atomic::{AtomicBool, AtomicU8, Ordering},
+        atomic::{AtomicU8, Ordering},
     },
 };
-use tokio::{sync::oneshot, task::JoinHandle};
+use tokio::task::JoinHandle;
 
 /// A prover is a light node, capable of producing proofs for consensus.
 #[derive(Clone)]
@@ -83,10 +89,10 @@ pub struct Prover<N: Network, C: ConsensusStorage<N>> {
     max_puzzle_instances: u8,
     /// The spawned handles.
     handles: Arc<Mutex<Vec<JoinHandle<()>>>>,
-    /// The shutdown signal.
-    shutdown: Arc<AtomicBool>,
     /// Keeps track of sending pings.
     ping: Arc<Ping<N>>,
+    /// The signal handling logic.
+    signal_handler: Arc<SignalHandler>,
     /// PhantomData.
     _phantom: PhantomData<C>,
 }
@@ -101,12 +107,8 @@ impl<N: Network, C: ConsensusStorage<N>> Prover<N, C> {
         storage_mode: StorageMode,
         trusted_peers_only: bool,
         dev: Option<u16>,
-        shutdown: Arc<AtomicBool>,
-        shutdown_tx: Option<oneshot::Sender<()>>,
+        signal_handler: Arc<SignalHandler>,
     ) -> Result<Self> {
-        // Initialize the signal handler.
-        let signal_node = Self::handle_signals(shutdown.clone(), shutdown_tx);
-
         // Initialize the ledger service.
         let ledger_service = Arc::new(ProverLedgerService::new());
 
@@ -144,7 +146,7 @@ impl<N: Network, C: ConsensusStorage<N>> Prover<N, C> {
             max_puzzle_instances: u8::try_from(max_puzzle_instances)?,
             handles: Default::default(),
             ping,
-            shutdown,
+            signal_handler,
             _phantom: Default::default(),
         };
         // Initialize the routing.
@@ -153,8 +155,7 @@ impl<N: Network, C: ConsensusStorage<N>> Prover<N, C> {
         node.initialize_puzzle().await;
         // Initialize the notification message loop.
         node.handles.lock().push(crate::start_notification_message_loop());
-        // Pass the node to the signal handler.
-        let _ = signal_node.set(node.clone());
+
         // Return the node.
         Ok(node)
     }
@@ -172,7 +173,6 @@ impl<N: Network, C: ConsensusStorage<N>> NodeInterface<N> for Prover<N, C> {
 
         // Shut down the puzzle.
         debug!("Shutting down the puzzle...");
-        self.shutdown.store(true, Ordering::Release);
 
         // Abort the tasks.
         debug!("Shutting down the prover...");
@@ -243,7 +243,7 @@ impl<N: Network, C: ConsensusStorage<N>> Prover<N, C> {
             }
 
             // If the Ctrl-C handler registered the signal, stop the prover.
-            if self.shutdown.load(Ordering::Acquire) {
+            if self.signal_handler.is_stopped() {
                 debug!("Shutting down the puzzle...");
                 break;
             }
diff --git a/node/src/traits.rs b/node/src/traits.rs
index 735b37d0d..dd7a4155d 100644
--- a/node/src/traits.rs
+++ b/node/src/traits.rs
@@ -15,19 +15,10 @@
 
 use snarkos_node_network::{NodeType, PeerPoolHandling};
 use snarkos_node_router::Routing;
-use snarkvm::prelude::{Address, Network, PrivateKey, ViewKey};
 
-use once_cell::sync::OnceCell;
-use std::{
-    future::Future,
-    io,
-    sync::{
-        Arc,
-        atomic::{AtomicBool, Ordering},
-    },
-    time::Duration,
-};
-use tokio::sync::oneshot;
+use snarkos_utilities::SignalHandler;
+
+use snarkvm::prelude::{Address, Network, PrivateKey, ViewKey};
 
 #[async_trait]
 pub trait NodeInterface<N: Network>: Routing<N> {
@@ -56,69 +47,12 @@ pub trait NodeInterface<N: Network>: Routing<N> {
         self.router().is_dev()
     }
 
-    /// Handles OS signals for the node to intercept and perform a clean shutdown.
-    /// The optional `shutdown_flag` flag can be used to cleanly terminate the syncing process.
-    fn handle_signals(shutdown_flag: Arc<AtomicBool>, shutdown_tx: Option<oneshot::Sender<()>>) -> Arc<OnceCell<Self>> {
-        // In order for the signal handler to be started as early as possible, a reference to the node needs
-        // to be passed to it at a later time.
-        let node: Arc<OnceCell<Self>> = Default::default();
-
-        #[cfg(target_family = "unix")]
-        fn signal_listener() -> impl Future<Output = io::Result<()>> {
-            use tokio::signal::unix::{SignalKind, signal};
-
-            // Handle SIGINT, SIGTERM, SIGQUIT, and SIGHUP.
-            let mut s_int = signal(SignalKind::interrupt()).unwrap();
-            let mut s_term = signal(SignalKind::terminate()).unwrap();
-            let mut s_quit = signal(SignalKind::quit()).unwrap();
-            let mut s_hup = signal(SignalKind::hangup()).unwrap();
-
-            // Return when any of the signals above is received.
-            async move {
-                tokio::select!(
-                    _ = s_int.recv() => (),
-                    _ = s_term.recv() => (),
-                    _ = s_quit.recv() => (),
-                    _ = s_hup.recv() => (),
-                );
-                Ok(())
-            }
-        }
-        #[cfg(not(target_family = "unix"))]
-        fn signal_listener() -> impl Future<Output = io::Result<()>> {
-            tokio::signal::ctrl_c()
-        }
-
-        let node_clone = node.clone();
-        tokio::task::spawn(async move {
-            match signal_listener().await {
-                Ok(()) => {
-                    warn!("==========================================================================================");
-                    warn!("⚠️  Attention - Starting the graceful shutdown procedure (ETA: 30 seconds)...");
-                    warn!("⚠️  Attention - To avoid DATA CORRUPTION, do NOT interrupt snarkOS (or press Ctrl+C again)");
-                    warn!("⚠️  Attention - Please wait until the shutdown gracefully completes (ETA: 30 seconds)");
-                    warn!("==========================================================================================");
-
-                    match node_clone.get() {
-                        // If the node is already initialized, then shut it down.
-                        Some(node) => node.shut_down().await,
-                        // Otherwise, if the node is not yet initialized, then set the shutdown flag directly.
-                        None => shutdown_flag.store(true, Ordering::Relaxed),
-                    }
-
-                    // A best-effort attempt to let any ongoing activity conclude.
-                    tokio::time::sleep(Duration::from_secs(3)).await;
-
-                    // Terminate the process.
-                    if let Some(tx) = shutdown_tx {
-                        let _ = tx.send(());
-                    }
-                }
-                Err(error) => error!("tokio::signal::ctrl_c encountered an error: {}", error),
-            }
-        });
+    /// Blocks until a shutdown signal was received or manual shutdown was triggered.
+    async fn wait_for_signals(&self, handler: &SignalHandler) {
+        handler.wait_for_signals().await;
 
-        node
+        // If the node is already initialized, then shut it down.
+        self.shut_down().await;
     }
 
     /// Shuts down the node.
diff --git a/node/src/validator/mod.rs b/node/src/validator/mod.rs
index 7b8c3e530..c3d75d6e9 100644
--- a/node/src/validator/mod.rs
+++ b/node/src/validator/mod.rs
@@ -36,6 +36,8 @@ use snarkos_node_tcp::{
     P2P,
     protocols::{Disconnect, Handshake, OnConnect, Reading},
 };
+use snarkos_utilities::SignalHandler;
+
 use snarkvm::prelude::{
     Ledger,
     Network,
@@ -51,12 +53,8 @@ use core::future::Future;
 use locktick::parking_lot::Mutex;
 #[cfg(not(feature = "locktick"))]
 use parking_lot::Mutex;
-use std::{
-    net::SocketAddr,
-    sync::{Arc, atomic::AtomicBool},
-    time::Duration,
-};
-use tokio::{sync::oneshot, task::JoinHandle};
+use std::{net::SocketAddr, sync::Arc, time::Duration};
+use tokio::task::JoinHandle;
 
 /// A validator is a full node, capable of validating blocks.
 #[derive(Clone)]
@@ -73,8 +71,6 @@ pub struct Validator<N: Network, C: ConsensusStorage<N>> {
     sync: Arc<BlockSync<N>>,
     /// The spawned handles.
     handles: Arc<Mutex<Vec<JoinHandle<()>>>>,
-    /// The shutdown signal.
-    shutdown: Arc<AtomicBool>,
     /// Keeps track of sending pings.
     ping: Arc<Ping<N>>,
 }
@@ -95,12 +91,8 @@ impl<N: Network, C: ConsensusStorage<N>> Validator<N, C> {
         trusted_peers_only: bool,
         dev_txs: bool,
         dev: Option<u16>,
-        shutdown: Arc<AtomicBool>,
-        shutdown_tx: Option<oneshot::Sender<()>>,
+        signal_handler: Arc<SignalHandler>,
     ) -> Result<Self> {
-        // Initialize the signal handler.
-        let signal_node = Self::handle_signals(shutdown.clone(), shutdown_tx);
-
         // Initialize the ledger.
         let ledger = {
             let storage_mode = storage_mode.clone();
@@ -111,7 +103,7 @@ impl<N: Network, C: ConsensusStorage<N>> Validator<N, C> {
         .with_context(|| "Failed to initialize the ledger")?;
 
         // Initialize the ledger service.
-        let ledger_service = Arc::new(CoreLedgerService::new(ledger.clone(), shutdown.clone()));
+        let ledger_service = Arc::new(CoreLedgerService::new(ledger.clone(), signal_handler.clone()));
 
         // Initialize the node router.
         let router = Router::new(
@@ -155,11 +147,10 @@ impl<N: Network, C: ConsensusStorage<N>> Validator<N, C> {
             sync: sync.clone(),
             ping,
             handles: Default::default(),
-            shutdown: shutdown.clone(),
         };
 
         // Perform sync with CDN (if enabled).
-        let cdn_sync = cdn.map(|base_url| Arc::new(CdnBlockSync::new(base_url, ledger.clone(), shutdown)));
+        let cdn_sync = cdn.map(|base_url| Arc::new(CdnBlockSync::new(base_url, ledger.clone(), signal_handler)));
 
         // Initialize the transaction pool.
         node.initialize_transaction_pool(dev, dev_txs)?;
@@ -193,8 +184,7 @@ impl<N: Network, C: ConsensusStorage<N>> Validator<N, C> {
         node.initialize_routing().await;
         // Initialize the notification message loop.
         node.handles.lock().push(crate::start_notification_message_loop());
-        // Pass the node to the signal handler.
-        let _ = signal_node.set(node.clone());
+
         // Return the node.
         Ok(node)
     }
@@ -468,7 +458,6 @@ impl<N: Network, C: ConsensusStorage<N>> NodeInterface<N> for Validator<N, C> {
 
         // Shut down the node.
         trace!("Shutting down the node...");
-        self.shutdown.store(true, std::sync::atomic::Ordering::Release);
 
         // Abort the tasks.
         trace!("Shutting down the validator...");
@@ -538,8 +527,7 @@ mod tests {
             false,
             dev_txs,
             None,
-            Default::default(),
-            None,
+            SignalHandler::new(),
         )
         .await
         .unwrap();
diff --git a/node/tests/common/node.rs b/node/tests/common/node.rs
index 102133964..2e55778e8 100644
--- a/node/tests/common/node.rs
+++ b/node/tests/common/node.rs
@@ -14,8 +14,11 @@
 // limitations under the License.
 
 use crate::common::test_peer::sample_genesis_block;
+
 use snarkos_account::Account;
 use snarkos_node::{Client, Prover, Validator};
+use snarkos_utilities::SignalHandler;
+
 use snarkvm::prelude::{MainnetV0 as CurrentNetwork, store::helpers::memory::ConsensusMemory};
 
 use aleo_std::StorageMode;
@@ -33,8 +36,7 @@ pub async fn client() -> Client<CurrentNetwork, ConsensusMemory<CurrentNetwork>>
         StorageMode::new_test(None),
         false, // Connect to untrusted peers.
         None,
-        Default::default(),
-        None,
+        SignalHandler::new(),
     )
     .await
     .expect("couldn't create client instance")
@@ -49,8 +51,7 @@ pub async fn prover() -> Prover<CurrentNetwork, ConsensusMemory<CurrentNetwork>>
         StorageMode::new_test(None),
         false,
         None,
-        Default::default(),
-        None,
+        SignalHandler::new(),
     )
     .await
     .expect("couldn't create prover instance")
@@ -71,8 +72,7 @@ pub async fn validator() -> Validator<CurrentNetwork, ConsensusMemory<CurrentNet
         false, // This test requires validators to connect to peers.
         false, // No dev traffic in production mode.
         None,
-        Default::default(),
-        None,
+        SignalHandler::new(),
     )
     .await
     .expect("couldn't create validator instance")
diff --git a/utilities/Cargo.toml b/utilities/Cargo.toml
new file mode 100644
index 000000000..c03dc9317
--- /dev/null
+++ b/utilities/Cargo.toml
@@ -0,0 +1,32 @@
+[package]
+name = "snarkos-utilities"
+version = "4.2.1"
+authors = [ "The Aleo Team <hello@aleo.org>" ]
+description = "Utilities for a decentralized operating system"
+homepage = "https://aleo.org"
+repository = "https://github.com/ProvableHQ/snarkOS"
+keywords = [
+  "aleo",
+  "cryptography",
+  "blockchain",
+  "decentralized",
+  "zero-knowledge"
+]
+include = ["../LICENSE.md"]
+categories = [ "cryptography", "cryptography::cryptocurrencies", "os" ]
+license = "Apache-2.0"
+edition = "2024"
+
+[dependencies.tokio]
+workspace = true
+features = [ "macros", "signal", "sync" ]
+
+[dependencies.tracing]
+workspace = true
+
+[dependencies.locktick]
+optional = true
+workspace = true
+
+[dependencies.parking_lot]
+workspace = true
diff --git a/utilities/src/lib.rs b/utilities/src/lib.rs
new file mode 100644
index 000000000..de7548d27
--- /dev/null
+++ b/utilities/src/lib.rs
@@ -0,0 +1,19 @@
+// Copyright (c) 2019-2025 Provable Inc.
+// This file is part of the snarkOS library.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+
+// http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/// Utilities for signal and shutdown handling.
+pub mod signals;
+
+pub use signals::*;
diff --git a/utilities/src/signals.rs b/utilities/src/signals.rs
new file mode 100644
index 000000000..83809d1b2
--- /dev/null
+++ b/utilities/src/signals.rs
@@ -0,0 +1,155 @@
+// Copyright (c) 2019-2025 Provable Inc.
+// This file is part of the snarkOS library.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+
+// http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#[cfg(feature = "locktick")]
+use locktick::parking_lot::{Mutex, RwLock};
+#[cfg(not(feature = "locktick"))]
+use parking_lot::{Mutex, RwLock};
+
+use std::sync::{
+    Arc,
+    atomic::{AtomicBool, Ordering},
+};
+use tokio::sync::oneshot;
+
+use tracing::{debug, error, trace};
+
+/// Generic trait that can be queried for whether current process should be stopped.
+/// This is implemented by `SignalHandler` and `SimpleStoppable`.
+pub trait Stoppable: Send + Sync {
+    /// Initiates shutdown of the node.
+    fn stop(&self);
+
+    /// Returns `true` if the node is (in the process of being) stopped.
+    fn is_stopped(&self) -> bool;
+}
+
+/// Wrapper around `AtomicBool` that implements the `Stoppable` trait.
+///
+/// This is useful when no signal or complex shutdown handling is necessary (e.g., in a test environment).
+pub struct SimpleStoppable {
+    state: AtomicBool,
+}
+
+impl SimpleStoppable {
+    pub fn new() -> Arc<Self> {
+        Arc::new(Self { state: AtomicBool::new(false) })
+    }
+}
+
+impl Stoppable for SimpleStoppable {
+    fn stop(&self) {
+        self.state.store(true, Ordering::SeqCst);
+    }
+
+    fn is_stopped(&self) -> bool {
+        self.state.load(Ordering::SeqCst)
+    }
+}
+
+/// Helper for signal handling that implements the `Stoppable` trait.
+///
+/// This struct will set itself to "stopped" as soon as the process receives Ctrl+C.
+/// It can also be manually stopped (e.g., when the node encounters a fatal error).
+pub struct SignalHandler {
+    /// This sender is used to notify a waiting task that the node has been stopped.
+    /// If this is `None`, the node is in the process of shutting down.
+    stopped_sender: RwLock<Option<oneshot::Sender<()>>>,
+
+    /// This receiver is used to wait for the node to be stopped.
+    stopped_receiver: Mutex<Option<oneshot::Receiver<()>>>,
+}
+
+impl SignalHandler {
+    /// Spawns a background tasks that listens for Ctrl+C and returns `Self`.
+    pub fn new() -> Arc<Self> {
+        let (stopped_sender, stopped_receiver) = oneshot::channel();
+        let obj = Arc::new(Self {
+            stopped_sender: RwLock::new(Some(stopped_sender)),
+            stopped_receiver: Mutex::new(Some(stopped_receiver)),
+        });
+
+        {
+            let obj = obj.clone();
+            tokio::spawn(async move {
+                obj.handle_signals().await;
+            });
+        }
+
+        obj
+    }
+
+    /// Logic for the background task that waits for a signal.
+    async fn handle_signals(&self) {
+        #[cfg(target_family = "unix")]
+        let signal_listener = async move {
+            use tokio::signal::unix::{SignalKind, signal};
+
+            // Handle SIGINT, SIGTERM, SIGQUIT, and SIGHUP.
+            let mut s_int = signal(SignalKind::interrupt())?;
+            let mut s_term = signal(SignalKind::terminate())?;
+            let mut s_quit = signal(SignalKind::quit())?;
+            let mut s_hup = signal(SignalKind::hangup())?;
+
+            tokio::select!(
+                _ = s_int.recv() => trace!("Received SIGINT"),
+                _ = s_term.recv() => trace!("Received SIGTERM"),
+                _ = s_quit.recv() => trace!("Received SIGQUIT"),
+                _ = s_hup.recv() => trace!("Received SIGHUP"),
+            );
+
+            std::io::Result::<()>::Ok(())
+        };
+
+        #[cfg(not(target_family = "unix"))]
+        let signal_listener = async move {
+            tokio::signal::ctrl_c().await?;
+            std::io::Result::<()>::Ok(())
+        };
+
+        // Block until we receive a signal.
+        match signal_listener.await {
+            Ok(()) => debug!("Received signal, shutting down..."),
+            Err(error) => error!("tokio::signal encountered an error: {error}"),
+        }
+
+        self.stop();
+    }
+
+    /// Waits until the signal handler was invoked or the stopped flag was set some other way.
+    ///
+    /// Note: This can only be called once, and must not be called concurrently.
+    pub async fn wait_for_signals(&self) {
+        let Some(receiver) = self.stopped_receiver.lock().take() else {
+            panic!("wait_for_signals must be called at most once");
+        };
+
+        if let Err(err) = receiver.await {
+            error!("wait_for_signals encountered an error: {err}");
+        }
+    }
+}
+
+impl Stoppable for SignalHandler {
+    fn stop(&self) {
+        if let Some(stopped_sender) = self.stopped_sender.write().take() {
+            let _ = stopped_sender.send(());
+        }
+    }
+
+    fn is_stopped(&self) -> bool {
+        self.stopped_sender.read().is_none()
+    }
+}
