diff --git a/Cargo.lock b/Cargo.lock
index d7a5d1ff1..00d318a58 100644
--- a/Cargo.lock
+++ b/Cargo.lock
@@ -2272,7 +2272,6 @@ dependencies = [
  "snarkos-account",
  "snarkos-display",
  "snarkos-node",
- "snarkos-node-store",
  "snarkvm",
  "thiserror",
  "tokio",
diff --git a/cli/Cargo.toml b/cli/Cargo.toml
index 81be379c0..443f8cffb 100644
--- a/cli/Cargo.toml
+++ b/cli/Cargo.toml
@@ -65,9 +65,6 @@ path = "../display"
 [dependencies.snarkos-node]
 path = "../node"
 
-[dependencies.snarkos-node-store]
-path = "../node/store"
-
 [dependencies.snarkvm]
 workspace = true
 
diff --git a/cli/src/commands/clean.rs b/cli/src/commands/clean.rs
index 02f425ea1..ed6271aba 100644
--- a/cli/src/commands/clean.rs
+++ b/cli/src/commands/clean.rs
@@ -37,7 +37,7 @@ impl Clean {
     }
 
     /// Removes the specified ledger from storage.
-    pub(super) fn remove_ledger(network: u16, dev: Option<u16>) -> Result<String> {
+    fn remove_ledger(network: u16, dev: Option<u16>) -> Result<String> {
         // Construct the path to the ledger in storage.
         let path = aleo_std::aleo_ledger_dir(network, dev);
 
diff --git a/cli/src/commands/start.rs b/cli/src/commands/start.rs
index 52b58cd02..79c8b10d9 100644
--- a/cli/src/commands/start.rs
+++ b/cli/src/commands/start.rs
@@ -14,13 +14,10 @@
 // You should have received a copy of the GNU General Public License
 // along with the snarkOS library. If not, see <https://www.gnu.org/licenses/>.
 
-use crate::commands::Clean;
-
 use snarkos_account::Account;
 use snarkos_display::Display;
 use snarkos_node::Node;
-use snarkos_node_store::ConsensusDB;
-use snarkvm::prelude::{Block, ConsensusMemory, ConsensusStore, FromBytes, Network, PrivateKey, Testnet3, VM};
+use snarkvm::prelude::{Block, ConsensusMemory, ConsensusStore, Network, PrivateKey, Testnet3, VM};
 
 use anyhow::{bail, Result};
 use clap::Parser;
@@ -201,24 +198,6 @@ impl Start {
             false => Some(self.rest),
         };
 
-        // If development mode is not set, check that the existing genesis block is correct.
-        {
-            if self.dev.is_none() && genesis.is_none() {
-                let consensus = ConsensusStore::<N, ConsensusDB<N>>::open(None)?;
-                // Check if a genesis block exists.
-                if let Ok(Some(genesis_block_hash)) = consensus.block_store().get_block_hash(0) {
-                    // Check if the genesis block is correct.
-                    let expected_genesis_block_hash = Block::<N>::from_bytes_le(N::genesis_bytes())?.hash();
-                    // If the existing genesis block hash does not match the expected genesis block hash, then clear the ledger.
-                    if genesis_block_hash != expected_genesis_block_hash {
-                        drop(consensus);
-                        let remove_ledger = Clean::remove_ledger(N::ID, None)?;
-                        println!("{}", remove_ledger);
-                    }
-                }
-            }
-        }
-
         // Ensures only one of the four flags is set. If no flags are set, defaults to a client node.
         match (&self.beacon, &self.validator, &self.prover, &self.client) {
             (Some(private_key), None, None, None) => Node::new_beacon(self.node, rest_ip, PrivateKey::<N>::from_str(private_key)?, &trusted_peers, genesis, self.dev).await,
