diff --git a/Cargo.lock b/Cargo.lock
index 3875df1f0..97c6ebdf9 100644
--- a/Cargo.lock
+++ b/Cargo.lock
@@ -1847,6 +1847,7 @@ version = "1.0.91"
 source = "registry+https://github.com/rust-lang/crates.io-index"
 checksum = "877c235533714907a8c2464236f5c4b2a17262ef1bd71f38f35ea592c8da6883"
 dependencies = [
+ "indexmap",
  "itoa",
  "ryu",
  "serde",
diff --git a/Cargo.toml b/Cargo.toml
index 2ac18a8f4..624f06112 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -217,6 +217,7 @@ optional = true
 [dependencies.serde_json]
 version = "1.0"
 optional = true
+features = [ "preserve_order" ]
 
 [dependencies.thiserror]
 version = "1.0"
diff --git a/algorithms/Cargo.toml b/algorithms/Cargo.toml
index 423911ae4..368dbe4b6 100644
--- a/algorithms/Cargo.toml
+++ b/algorithms/Cargo.toml
@@ -154,6 +154,7 @@ version = "0.4.0"
 
 [dev-dependencies.serde_json]
 version = "1"
+features = [ "preserve_order" ]
 
 [dev-dependencies.serial_test]
 version = "1.0"
diff --git a/algorithms/src/fft/domain.rs b/algorithms/src/fft/domain.rs
index 000ed58d1..d6b210a74 100644
--- a/algorithms/src/fft/domain.rs
+++ b/algorithms/src/fft/domain.rs
@@ -346,8 +346,8 @@ impl<F: FftField> EvaluationDomain<F> {
     /// evaluations in the domain.
     /// Returns the evaluations of the product over the domain.
     #[must_use]
-    pub fn mul_polynomials_in_evaluation_domain(&self, self_evals: &[F], other_evals: &[F]) -> Vec<F> {
-        let mut result = self_evals.to_vec();
+    pub fn mul_polynomials_in_evaluation_domain(&self, self_evals: Vec<F>, other_evals: &[F]) -> Vec<F> {
+        let mut result = self_evals;
 
         cfg_iter_mut!(result).zip_eq(other_evals).for_each(|(a, b)| *a *= b);
 
diff --git a/console/account/Cargo.toml b/console/account/Cargo.toml
index 5279e20e8..301e8e649 100644
--- a/console/account/Cargo.toml
+++ b/console/account/Cargo.toml
@@ -32,6 +32,7 @@ version = "0.4.0"
 
 [dev-dependencies.serde_json]
 version = "1.0"
+features = [ "preserve_order" ]
 
 [features]
 default = [
diff --git a/console/algorithms/Cargo.toml b/console/algorithms/Cargo.toml
index b0672b084..3deefc2bd 100644
--- a/console/algorithms/Cargo.toml
+++ b/console/algorithms/Cargo.toml
@@ -52,3 +52,4 @@ version = "1.0"
 
 [dev-dependencies.serde_json]
 version = "1.0"
+features = [ "preserve_order" ]
diff --git a/console/program/Cargo.toml b/console/program/Cargo.toml
index 6c7b97fea..0184a6e17 100644
--- a/console/program/Cargo.toml
+++ b/console/program/Cargo.toml
@@ -50,6 +50,7 @@ version = "1.17.0"
 
 [dependencies.serde_json]
 version = "1.0"
+features = [ "preserve_order" ]
 
 [dev-dependencies.bincode]
 version = "1.3"
diff --git a/console/program/src/state_path/transaction_leaf/mod.rs b/console/program/src/state_path/transaction_leaf/mod.rs
index 902fdbd71..cfad7e680 100644
--- a/console/program/src/state_path/transaction_leaf/mod.rs
+++ b/console/program/src/state_path/transaction_leaf/mod.rs
@@ -39,6 +39,11 @@ impl<N: Network> TransactionLeaf<N> {
         Self { variant: 0, index, id }
     }
 
+    /// Initializes a new instance of `TransactionLeaf`.
+    pub const fn new_deployment_fee(index: u16, id: Field<N>) -> Self {
+        Self { variant: 1, index, id }
+    }
+
     /// Initializes a new instance of `TransactionLeaf`.
     pub const fn new_execution(index: u16, id: Field<N>) -> Self {
         Self { variant: 1, index, id }
diff --git a/console/types/address/Cargo.toml b/console/types/address/Cargo.toml
index 5f3869e12..1bd73aa8b 100644
--- a/console/types/address/Cargo.toml
+++ b/console/types/address/Cargo.toml
@@ -27,3 +27,4 @@ version = "1.3"
 
 [dev-dependencies.serde_json]
 version = "1.0"
+features = [ "preserve_order" ]
diff --git a/console/types/boolean/Cargo.toml b/console/types/boolean/Cargo.toml
index 5fa8f4edd..bd1a8056b 100644
--- a/console/types/boolean/Cargo.toml
+++ b/console/types/boolean/Cargo.toml
@@ -15,3 +15,4 @@ version = "1.3"
 
 [dev-dependencies.serde_json]
 version = "1.0"
+features = [ "preserve_order" ]
diff --git a/console/types/field/Cargo.toml b/console/types/field/Cargo.toml
index 5ec99b6fe..7621be607 100644
--- a/console/types/field/Cargo.toml
+++ b/console/types/field/Cargo.toml
@@ -19,3 +19,4 @@ version = "1.3"
 
 [dev-dependencies.serde_json]
 version = "1.0"
+features = [ "preserve_order" ]
diff --git a/console/types/group/Cargo.toml b/console/types/group/Cargo.toml
index fb3ae2f95..5890b7482 100644
--- a/console/types/group/Cargo.toml
+++ b/console/types/group/Cargo.toml
@@ -27,3 +27,4 @@ version = "1.3"
 
 [dev-dependencies.serde_json]
 version = "1.0"
+features = [ "preserve_order" ]
diff --git a/console/types/integers/Cargo.toml b/console/types/integers/Cargo.toml
index cda959afd..ae6558724 100644
--- a/console/types/integers/Cargo.toml
+++ b/console/types/integers/Cargo.toml
@@ -23,3 +23,4 @@ version = "1.3"
 
 [dev-dependencies.serde_json]
 version = "1.0"
+features = [ "preserve_order" ]
diff --git a/console/types/scalar/Cargo.toml b/console/types/scalar/Cargo.toml
index c009693a4..102789108 100644
--- a/console/types/scalar/Cargo.toml
+++ b/console/types/scalar/Cargo.toml
@@ -23,3 +23,4 @@ version = "1.3"
 
 [dev-dependencies.serde_json]
 version = "1.0"
+features = [ "preserve_order" ]
diff --git a/console/types/string/Cargo.toml b/console/types/string/Cargo.toml
index d8644f618..a91721e3d 100644
--- a/console/types/string/Cargo.toml
+++ b/console/types/string/Cargo.toml
@@ -27,3 +27,4 @@ version = "1.3"
 
 [dev-dependencies.serde_json]
 version = "1.0"
+features = [ "preserve_order" ]
diff --git a/parameters/Cargo.toml b/parameters/Cargo.toml
index c00491862..f4142581e 100644
--- a/parameters/Cargo.toml
+++ b/parameters/Cargo.toml
@@ -77,6 +77,7 @@ optional = true
 
 [dependencies.serde_json]
 version = "1"
+features = [ "preserve_order" ]
 
 [dependencies.sha2]
 version = "0.10"
diff --git a/synthesizer/Cargo.toml b/synthesizer/Cargo.toml
index 4d0f4c8dd..c015d9e64 100644
--- a/synthesizer/Cargo.toml
+++ b/synthesizer/Cargo.toml
@@ -116,6 +116,7 @@ version = "1.0"
 
 [dependencies.serde_json]
 version = "1.0"
+features = [ "preserve_order" ]
 
 [dependencies.tracing]
 version = "0.1"
diff --git a/synthesizer/src/block/transaction/merkle.rs b/synthesizer/src/block/transaction/merkle.rs
index f45dc5746..6920e264f 100644
--- a/synthesizer/src/block/transaction/merkle.rs
+++ b/synthesizer/src/block/transaction/merkle.rs
@@ -29,7 +29,7 @@ impl<N: Network> Transaction<N> {
                 // Check if the ID is the transition ID for the fee.
                 if *id == **fee.id() {
                     // Return the transaction leaf.
-                    return Ok(TransactionLeaf::new_deployment(
+                    return Ok(TransactionLeaf::new_deployment_fee(
                         deployment.program().functions().len() as u16, // The last index.
                         *id,
                     ));
@@ -110,7 +110,7 @@ impl<N: Network> Transaction<N> {
                 .to_bits_le())
             })
             .chain(
-                [Ok(TransactionLeaf::new_deployment(
+                [Ok(TransactionLeaf::new_deployment_fee(
                     program.functions().len() as u16, // The last index.
                     **fee.transition_id(),
                 )
diff --git a/synthesizer/src/coinbase_puzzle/mod.rs b/synthesizer/src/coinbase_puzzle/mod.rs
index 7c18c5130..4afac52eb 100644
--- a/synthesizer/src/coinbase_puzzle/mod.rs
+++ b/synthesizer/src/coinbase_puzzle/mod.rs
@@ -122,7 +122,7 @@ impl<N: Network> CoinbasePuzzle<N> {
         let product_evaluations = {
             let polynomial_evaluations = pk.product_domain.in_order_fft_with_pc(&polynomial, &pk.fft_precomputation);
             let product_evaluations = pk.product_domain.mul_polynomials_in_evaluation_domain(
-                &polynomial_evaluations,
+                polynomial_evaluations,
                 &epoch_challenge.epoch_polynomial_evaluations().evaluations,
             );
             product_evaluations
@@ -225,7 +225,7 @@ impl<N: Network> CoinbasePuzzle<N> {
             let accumulated_polynomial_evaluations =
                 pk.product_domain.in_order_fft_with_pc(&accumulated_prover_polynomial.coeffs, &pk.fft_precomputation);
             pk.product_domain.mul_polynomials_in_evaluation_domain(
-                &accumulated_polynomial_evaluations,
+                accumulated_polynomial_evaluations,
                 &epoch_challenge.epoch_polynomial_evaluations().evaluations,
             )
         };
diff --git a/synthesizer/src/process/deploy.rs b/synthesizer/src/process/deploy.rs
index 02c71e524..739491417 100644
--- a/synthesizer/src/process/deploy.rs
+++ b/synthesizer/src/process/deploy.rs
@@ -39,7 +39,7 @@ impl<N: Network> Process<N> {
         deployment
     }
 
-    /// Verifies the given deployment is well-formed.
+    /// Verifies the given deployment is ordered.
     #[inline]
     pub fn verify_deployment<A: circuit::Aleo<Network = N>, R: Rng + CryptoRng>(
         &self,
diff --git a/synthesizer/src/process/stack/deploy.rs b/synthesizer/src/process/stack/deploy.rs
index c559cf620..34b0be61b 100644
--- a/synthesizer/src/process/stack/deploy.rs
+++ b/synthesizer/src/process/stack/deploy.rs
@@ -25,8 +25,8 @@ impl<N: Network> Stack<N> {
         // Ensure the program contains functions.
         ensure!(!self.program.functions().is_empty(), "Program '{}' has no functions", self.program.id());
 
-        // Initialize a mapping for the bundle.
-        let mut bundle = IndexMap::with_capacity(self.program.functions().len());
+        // Initialize a vector for the verifying keys and certificates.
+        let mut verifying_keys = Vec::with_capacity(self.program.functions().len());
 
         for function_name in self.program.functions().keys() {
             // Synthesize the proving and verifying key.
@@ -44,13 +44,13 @@ impl<N: Network> Stack<N> {
             lap!(timer, "Certify the circuit");
 
             // Add the verifying key and certificate to the bundle.
-            bundle.insert(*function_name, (verifying_key, certificate));
+            verifying_keys.push((*function_name, (verifying_key, certificate)));
         }
 
         finish!(timer);
 
         // Return the deployment.
-        Deployment::new(N::EDITION, self.program.clone(), bundle)
+        Deployment::new(N::EDITION, self.program.clone(), verifying_keys)
     }
 
     /// Checks each function in the program on the given verifying key and certificate.
@@ -62,51 +62,21 @@ impl<N: Network> Stack<N> {
     ) -> Result<()> {
         let timer = timer!("Stack::verify_deployment");
 
-        // Retrieve the edition.
-        let edition = deployment.edition();
-        // Retrieve the program.
-        let program = &self.program;
-        // Retrieve the program ID.
-        let program_id = program.id();
-        // Retrieve the verifying keys.
-        let verifying_keys = deployment.verifying_keys();
-
         // Sanity Checks //
 
-        // Ensure the edition matches.
-        ensure!(edition == N::EDITION, "Deployed the wrong edition (expected '{}', found '{edition}').", N::EDITION);
-        // Ensure the program matches.
-        ensure!(program == deployment.program(), "The stack program does not match the deployment program");
-        // Ensure the program network-level domain (NLD) is correct.
-        ensure!(program_id.is_aleo(), "Program '{program_id}' has an incorrect network-level domain (NLD)");
-        // Ensure the program contains functions.
-        ensure!(!program.functions().is_empty(), "No functions present in the deployment for program '{program_id}'");
-        // Ensure the deployment contains verifying keys.
-        ensure!(!verifying_keys.is_empty(), "No verifying keys present in the deployment for program '{program_id}'");
+        // Ensure the deployment is ordered.
+        deployment.check_is_ordered()?;
+        // Ensure the program in the stack and deployment matches.
+        ensure!(&self.program == deployment.program(), "The stack program does not match the deployment program");
 
         // Check Verifying Keys //
 
-        // Ensure the number of verifying keys matches the number of program functions.
-        if verifying_keys.len() != program.functions().len() {
-            bail!("The number of verifying keys does not match the number of program functions");
-        }
-        lap!(timer, "Perform sanity checks");
-
-        // Ensure the program functions are in the same order as the verifying keys.
-        for ((function_name, function), candidate_name) in program.functions().iter().zip_eq(verifying_keys.keys()) {
-            // Ensure the function name is correct.
-            if function_name != function.name() {
-                bail!("The function key is '{function_name}', but the function name is '{}'", function.name())
-            }
-            // Ensure the function name with the verifying key is correct.
-            if candidate_name != function.name() {
-                bail!("The verifier key is '{candidate_name}', but the function name is '{}'", function.name())
-            }
-        }
-        lap!(timer, "Verify the function and verifying key ordering");
+        let program_id = self.program.id();
 
         // Iterate through the program functions.
-        for (function, (verifying_key, certificate)) in program.functions().values().zip_eq(verifying_keys.values()) {
+        for (function, (_, (verifying_key, certificate))) in
+            deployment.program().functions().values().zip_eq(deployment.verifying_keys())
+        {
             // Initialize a burner private key.
             let burner_private_key = PrivateKey::new(rng)?;
             // Compute the burner address.
@@ -147,7 +117,9 @@ impl<N: Network> Stack<N> {
             lap!(timer, "Synthesize the circuit");
             // Check the certificate.
             match assignments.read().last() {
-                None => bail!("The assignment for function '{}' is missing in '{program_id}'", function.name()),
+                None => {
+                    bail!("The assignment for function '{}' is missing in '{program_id}'", function.name())
+                }
                 Some(assignment) => {
                     // Ensure the certificate is valid.
                     if !certificate.verify(function.name(), assignment, verifying_key) {
diff --git a/synthesizer/src/process/stack/deployment/bytes.rs b/synthesizer/src/process/stack/deployment/bytes.rs
index 2e74291c1..6c81d9acf 100644
--- a/synthesizer/src/process/stack/deployment/bytes.rs
+++ b/synthesizer/src/process/stack/deployment/bytes.rs
@@ -33,8 +33,8 @@ impl<N: Network> FromBytes for Deployment<N> {
 
         // Read the number of entries in the bundle.
         let num_entries = u16::read_le(&mut reader)?;
-        // Read the bundle.
-        let mut bundle = IndexMap::with_capacity(num_entries as usize);
+        // Read the verifying keys.
+        let mut verifying_keys = Vec::with_capacity(num_entries as usize);
         for _ in 0..num_entries {
             // Read the identifier.
             let identifier = Identifier::<N>::read_le(&mut reader)?;
@@ -43,10 +43,11 @@ impl<N: Network> FromBytes for Deployment<N> {
             // Read the certificate.
             let certificate = Certificate::<N>::read_le(&mut reader)?;
             // Add the entry.
-            bundle.insert(identifier, (verifying_key, certificate));
+            verifying_keys.push((identifier, (verifying_key, certificate)));
         }
 
-        Ok(Self { edition, program, verifying_keys: bundle })
+        // Return the deployment.
+        Self::new(edition, program, verifying_keys).map_err(|err| error(format!("{err}")))
     }
 }
 
diff --git a/synthesizer/src/process/stack/deployment/mod.rs b/synthesizer/src/process/stack/deployment/mod.rs
index 166c6d8cc..73a15e035 100644
--- a/synthesizer/src/process/stack/deployment/mod.rs
+++ b/synthesizer/src/process/stack/deployment/mod.rs
@@ -24,8 +24,6 @@ use console::{
     program::{Identifier, ProgramID},
 };
 
-use indexmap::IndexMap;
-
 #[derive(Clone, PartialEq, Eq)]
 pub struct Deployment<N: Network> {
     /// The edition.
@@ -33,7 +31,7 @@ pub struct Deployment<N: Network> {
     /// The program.
     program: Program<N>,
     /// The mapping of function names to their verifying key and certificate.
-    verifying_keys: IndexMap<Identifier<N>, (VerifyingKey<N>, Certificate<N>)>,
+    verifying_keys: Vec<(Identifier<N>, (VerifyingKey<N>, Certificate<N>))>,
 }
 
 impl<N: Network> Deployment<N> {
@@ -41,9 +39,63 @@ impl<N: Network> Deployment<N> {
     pub fn new(
         edition: u16,
         program: Program<N>,
-        verifying_keys: IndexMap<Identifier<N>, (VerifyingKey<N>, Certificate<N>)>,
+        verifying_keys: Vec<(Identifier<N>, (VerifyingKey<N>, Certificate<N>))>,
     ) -> Result<Self> {
-        Ok(Self { edition, program, verifying_keys })
+        // Construct the deployment.
+        let deployment = Self { edition, program, verifying_keys };
+        // Ensure the deployment is ordered.
+        deployment.check_is_ordered()?;
+        // Return the deployment.
+        Ok(deployment)
+    }
+
+    /// Checks that the deployment is ordered.
+    pub fn check_is_ordered(&self) -> Result<()> {
+        let program_id = self.program.id();
+
+        // Ensure the edition matches.
+        ensure!(
+            self.edition == N::EDITION,
+            "Deployed the wrong edition (expected '{}', found '{}').",
+            N::EDITION,
+            self.edition
+        );
+        // Ensure the program network-level domain (NLD) is correct.
+        ensure!(program_id.is_aleo(), "Program '{program_id}' has an incorrect network-level domain (NLD)");
+        // Ensure the program contains functions.
+        ensure!(
+            !self.program.functions().is_empty(),
+            "No functions present in the deployment for program '{program_id}'"
+        );
+        // Ensure the deployment contains verifying keys.
+        ensure!(
+            !self.verifying_keys.is_empty(),
+            "No verifying keys present in the deployment for program '{program_id}'"
+        );
+
+        // Ensure the number of functions matches the number of verifying keys.
+        if self.program.functions().len() != self.verifying_keys.len() {
+            bail!("Deployment has an incorrect number of verifying keys, according to the program.");
+        }
+
+        // Ensure the function and verifying keys correspond.
+        for ((function_name, function), (name, _)) in self.program.functions().iter().zip_eq(&self.verifying_keys) {
+            // Ensure the function name is correct.
+            if function_name != function.name() {
+                bail!("The function key is '{function_name}', but the function name is '{}'", function.name())
+            }
+            // Ensure the function name with the verifying key is correct.
+            if name != function.name() {
+                bail!("The verifier key is '{name}', but the function name is '{}'", function.name())
+            }
+        }
+
+        ensure!(
+            !has_duplicates(self.verifying_keys.iter().map(|(name, ..)| name)),
+            "A duplicate function name was found"
+        );
+
+        Ok(())
     }
 
     /// Returns the edition.
@@ -62,7 +114,7 @@ impl<N: Network> Deployment<N> {
     }
 
     /// Returns the verifying keys.
-    pub const fn verifying_keys(&self) -> &IndexMap<Identifier<N>, (VerifyingKey<N>, Certificate<N>)> {
+    pub const fn verifying_keys(&self) -> &Vec<(Identifier<N>, (VerifyingKey<N>, Certificate<N>))> {
         &self.verifying_keys
     }
 }
diff --git a/synthesizer/src/process/stack/mod.rs b/synthesizer/src/process/stack/mod.rs
index c6d9340d2..2e471e7eb 100644
--- a/synthesizer/src/process/stack/mod.rs
+++ b/synthesizer/src/process/stack/mod.rs
@@ -412,6 +412,7 @@ impl<N: Network> PartialEq for Stack<N> {
         self.program == other.program
             && self.external_stacks == other.external_stacks
             && self.register_types == other.register_types
+            && self.finalize_types == other.finalize_types
     }
 }
 
diff --git a/synthesizer/src/store/block/mod.rs b/synthesizer/src/store/block/mod.rs
index 76ca6a453..7969a6c2f 100644
--- a/synthesizer/src/store/block/mod.rs
+++ b/synthesizer/src/store/block/mod.rs
@@ -333,7 +333,7 @@ pub trait BlockStorage<N: Network>: 'static + Clone + Send + Sync {
         // Find the transition that contains the commitment.
         let transition_id = self.transition_store().find_transition_id(commitment)?;
         // Find the transaction that contains the transition.
-        let transaction_id = match self.transaction_store().find_transaction_id(&transition_id)? {
+        let transaction_id = match self.transaction_store().find_transaction_id_from_transition_id(&transition_id)? {
             Some(transaction_id) => transaction_id,
             None => bail!("The transaction ID for commitment '{commitment}' is missing in storage"),
         };
diff --git a/synthesizer/src/store/transaction/deployment.rs b/synthesizer/src/store/transaction/deployment.rs
index 58384d29a..97f1197cb 100644
--- a/synthesizer/src/store/transaction/deployment.rs
+++ b/synthesizer/src/store/transaction/deployment.rs
@@ -36,7 +36,6 @@ use console::{
 
 use anyhow::Result;
 use core::marker::PhantomData;
-use indexmap::IndexMap;
 use std::borrow::Cow;
 
 /// A trait for deployment storage.
@@ -55,6 +54,8 @@ pub trait DeploymentStorage<N: Network>: Clone + Send + Sync {
     type CertificateMap: for<'a> Map<'a, (ProgramID<N>, Identifier<N>, u16), Certificate<N>>;
     /// The mapping of `transaction ID` to `(fee transition ID, global state root, inclusion proof)`.
     type FeeMap: for<'a> Map<'a, N::TransactionID, (N::TransitionID, N::StateRoot, Option<Proof<N>>)>;
+    /// The mapping of `fee transition ID` to `transaction ID`.
+    type ReverseFeeMap: for<'a> Map<'a, N::TransitionID, N::TransactionID>;
     /// The transition storage.
     type TransitionStorage: TransitionStorage<N>;
 
@@ -75,6 +76,8 @@ pub trait DeploymentStorage<N: Network>: Clone + Send + Sync {
     fn certificate_map(&self) -> &Self::CertificateMap;
     /// Returns the fee map.
     fn fee_map(&self) -> &Self::FeeMap;
+    /// Returns the reverse fee map.
+    fn reverse_fee_map(&self) -> &Self::ReverseFeeMap;
     /// Returns the transition storage.
     fn transition_store(&self) -> &TransitionStore<N, Self::TransitionStorage>;
 
@@ -92,6 +95,7 @@ pub trait DeploymentStorage<N: Network>: Clone + Send + Sync {
         self.verifying_key_map().start_atomic();
         self.certificate_map().start_atomic();
         self.fee_map().start_atomic();
+        self.reverse_fee_map().start_atomic();
         self.transition_store().start_atomic();
     }
 
@@ -104,6 +108,7 @@ pub trait DeploymentStorage<N: Network>: Clone + Send + Sync {
             || self.verifying_key_map().is_atomic_in_progress()
             || self.certificate_map().is_atomic_in_progress()
             || self.fee_map().is_atomic_in_progress()
+            || self.reverse_fee_map().is_atomic_in_progress()
             || self.transition_store().is_atomic_in_progress()
     }
 
@@ -116,6 +121,7 @@ pub trait DeploymentStorage<N: Network>: Clone + Send + Sync {
         self.verifying_key_map().abort_atomic();
         self.certificate_map().abort_atomic();
         self.fee_map().abort_atomic();
+        self.reverse_fee_map().abort_atomic();
         self.transition_store().abort_atomic();
     }
 
@@ -128,6 +134,7 @@ pub trait DeploymentStorage<N: Network>: Clone + Send + Sync {
         self.verifying_key_map().finish_atomic()?;
         self.certificate_map().finish_atomic()?;
         self.fee_map().finish_atomic()?;
+        self.reverse_fee_map().finish_atomic()?;
         self.transition_store().finish_atomic()
     }
 
@@ -141,23 +148,17 @@ pub trait DeploymentStorage<N: Network>: Clone + Send + Sync {
             }
         };
 
+        // Ensure the deployment is ordered.
+        if let Err(error) = deployment.check_is_ordered() {
+            bail!("Failed to insert malformed deployment transaction: {error}")
+        }
+
+        // Retrieve the edition.
+        let edition = deployment.edition();
         // Retrieve the program.
         let program = deployment.program();
         // Retrieve the program ID.
         let program_id = *program.id();
-        // Retrieve the edition.
-        let edition = deployment.edition();
-
-        // Ensure the number of functions matches the number of verifying keys.
-        if program.functions().len() != deployment.verifying_keys().len() {
-            bail!("Deployment has an incorrect number of verifying keys, according to the program.");
-        }
-        // Ensure the deployment contains the correct functions.
-        for function_name in program.functions().keys() {
-            if !deployment.verifying_keys().keys().contains(function_name) {
-                bail!("Deployment is missing a verifying key for function '{function_name}'.");
-            }
-        }
 
         atomic_write_batch!(self, {
             // Store the program ID.
@@ -183,6 +184,8 @@ pub trait DeploymentStorage<N: Network>: Clone + Send + Sync {
                 *transaction_id,
                 (*fee.transition_id(), fee.global_state_root(), fee.inclusion_proof().cloned()),
             )?;
+            self.reverse_fee_map().insert(*fee.transition_id(), *transaction_id)?;
+
             // Store the fee transition.
             self.transition_store().insert(fee)?;
 
@@ -236,6 +239,8 @@ pub trait DeploymentStorage<N: Network>: Clone + Send + Sync {
 
             // Remove the fee.
             self.fee_map().remove(transaction_id)?;
+            self.reverse_fee_map().remove(&transition_id)?;
+
             // Remove the fee transition.
             self.transition_store().remove(&transition_id)?;
 
@@ -246,7 +251,7 @@ pub trait DeploymentStorage<N: Network>: Clone + Send + Sync {
     }
 
     /// Returns the transaction ID that contains the given `program ID`.
-    fn find_transaction_id(&self, program_id: &ProgramID<N>) -> Result<Option<N::TransactionID>> {
+    fn find_transaction_id_from_program_id(&self, program_id: &ProgramID<N>) -> Result<Option<N::TransactionID>> {
         // Retrieve the edition.
         let edition = match self.get_edition(program_id)? {
             Some(edition) => edition,
@@ -259,6 +264,17 @@ pub trait DeploymentStorage<N: Network>: Clone + Send + Sync {
         }
     }
 
+    /// Returns the transaction ID that contains the given `transition ID`.
+    fn find_transaction_id_from_transition_id(
+        &self,
+        transition_id: &N::TransitionID,
+    ) -> Result<Option<N::TransactionID>> {
+        match self.reverse_fee_map().get(transition_id)? {
+            Some(transaction_id) => Ok(Some(cow_to_copied!(transaction_id))),
+            None => Ok(None),
+        }
+    }
+
     /// Returns the program ID for the given `transaction ID`.
     fn get_program_id(&self, transaction_id: &N::TransactionID) -> Result<Option<ProgramID<N>>> {
         // Retrieve the program ID.
@@ -345,7 +361,7 @@ pub trait DeploymentStorage<N: Network>: Clone + Send + Sync {
         };
 
         // Initialize a vector for the verifying keys and certificates.
-        let mut verifying_keys = IndexMap::new();
+        let mut verifying_keys = Vec::with_capacity(program.functions().len());
 
         // Retrieve the verifying keys and certificates.
         for function_name in program.functions().keys() {
@@ -360,7 +376,7 @@ pub trait DeploymentStorage<N: Network>: Clone + Send + Sync {
                 None => bail!("Failed to get the certificate for '{program_id}/{function_name}' (edition {edition})"),
             };
             // Add the verifying key and certificate to the deployment.
-            verifying_keys.insert(*function_name, (verifying_key, certificate));
+            verifying_keys.push((*function_name, (verifying_key, certificate)));
         }
 
         // Return the deployment.
@@ -422,6 +438,8 @@ pub struct DeploymentMemory<N: Network> {
     certificate_map: MemoryMap<(ProgramID<N>, Identifier<N>, u16), Certificate<N>>,
     /// The fee map.
     fee_map: MemoryMap<N::TransactionID, (N::TransitionID, N::StateRoot, Option<Proof<N>>)>,
+    /// The reverse fee map.
+    reverse_fee_map: MemoryMap<N::TransitionID, N::TransactionID>,
     /// The transition store.
     transition_store: TransitionStore<N, TransitionMemory<N>>,
 }
@@ -435,6 +453,7 @@ impl<N: Network> DeploymentStorage<N> for DeploymentMemory<N> {
     type VerifyingKeyMap = MemoryMap<(ProgramID<N>, Identifier<N>, u16), VerifyingKey<N>>;
     type CertificateMap = MemoryMap<(ProgramID<N>, Identifier<N>, u16), Certificate<N>>;
     type FeeMap = MemoryMap<N::TransactionID, (N::TransitionID, N::StateRoot, Option<Proof<N>>)>;
+    type ReverseFeeMap = MemoryMap<N::TransitionID, N::TransactionID>;
     type TransitionStorage = TransitionMemory<N>;
 
     /// Initializes the deployment storage.
@@ -447,6 +466,7 @@ impl<N: Network> DeploymentStorage<N> for DeploymentMemory<N> {
             verifying_key_map: MemoryMap::default(),
             certificate_map: MemoryMap::default(),
             fee_map: MemoryMap::default(),
+            reverse_fee_map: MemoryMap::default(),
             transition_store,
         })
     }
@@ -486,6 +506,11 @@ impl<N: Network> DeploymentStorage<N> for DeploymentMemory<N> {
         &self.fee_map
     }
 
+    /// Returns the reverse fee map.
+    fn reverse_fee_map(&self) -> &Self::ReverseFeeMap {
+        &self.reverse_fee_map
+    }
+
     /// Returns the transition store.
     fn transition_store(&self) -> &TransitionStore<N, Self::TransitionStorage> {
         &self.transition_store
@@ -603,8 +628,16 @@ impl<N: Network, D: DeploymentStorage<N>> DeploymentStore<N, D> {
 
 impl<N: Network, D: DeploymentStorage<N>> DeploymentStore<N, D> {
     /// Returns the transaction ID that deployed the given `program ID`.
-    pub fn find_transaction_id(&self, program_id: &ProgramID<N>) -> Result<Option<N::TransactionID>> {
-        self.storage.find_transaction_id(program_id)
+    pub fn find_transaction_id_from_program_id(&self, program_id: &ProgramID<N>) -> Result<Option<N::TransactionID>> {
+        self.storage.find_transaction_id_from_program_id(program_id)
+    }
+
+    /// Returns the transaction ID that deployed the given `transition ID`.
+    pub fn find_transaction_id_from_transition_id(
+        &self,
+        transition_id: &N::TransitionID,
+    ) -> Result<Option<N::TransactionID>> {
+        self.storage.find_transaction_id_from_transition_id(transition_id)
     }
 }
 
@@ -617,7 +650,7 @@ impl<N: Network, D: DeploymentStorage<N>> DeploymentStore<N, D> {
 
 impl<N: Network, D: DeploymentStorage<N>> DeploymentStore<N, D> {
     /// Returns an iterator over the deployment transaction IDs, for all deployments.
-    pub fn deployment_ids(&self) -> impl '_ + Iterator<Item = Cow<'_, N::TransactionID>> {
+    pub fn deployment_transaction_ids(&self) -> impl '_ + Iterator<Item = Cow<'_, N::TransactionID>> {
         self.storage.id_map().keys()
     }
 
@@ -710,21 +743,21 @@ mod tests {
         assert_eq!(None, candidate);
 
         // Ensure the transaction ID is not found.
-        let candidate = deployment_store.find_transaction_id(&program_id).unwrap();
+        let candidate = deployment_store.find_transaction_id_from_program_id(&program_id).unwrap();
         assert_eq!(None, candidate);
 
         // Insert the deployment.
         deployment_store.insert(&transaction).unwrap();
 
         // Find the transaction ID.
-        let candidate = deployment_store.find_transaction_id(&program_id).unwrap();
+        let candidate = deployment_store.find_transaction_id_from_program_id(&program_id).unwrap();
         assert_eq!(Some(transaction_id), candidate);
 
         // Remove the deployment.
         deployment_store.remove(&transaction_id).unwrap();
 
         // Ensure the transaction ID is not found.
-        let candidate = deployment_store.find_transaction_id(&program_id).unwrap();
+        let candidate = deployment_store.find_transaction_id_from_program_id(&program_id).unwrap();
         assert_eq!(None, candidate);
     }
 }
diff --git a/synthesizer/src/store/transaction/execution.rs b/synthesizer/src/store/transaction/execution.rs
index 229802bba..0b1dd350b 100644
--- a/synthesizer/src/store/transaction/execution.rs
+++ b/synthesizer/src/store/transaction/execution.rs
@@ -200,7 +200,10 @@ pub trait ExecutionStorage<N: Network>: Clone + Send + Sync {
     }
 
     /// Returns the transaction ID that contains the given `transition ID`.
-    fn find_transaction_id(&self, transition_id: &N::TransitionID) -> Result<Option<N::TransactionID>> {
+    fn find_transaction_id_from_transition_id(
+        &self,
+        transition_id: &N::TransitionID,
+    ) -> Result<Option<N::TransactionID>> {
         match self.reverse_id_map().get(transition_id)? {
             Some(transaction_id) => Ok(Some(cow_to_copied!(transaction_id))),
             None => Ok(None),
@@ -433,14 +436,17 @@ impl<N: Network, E: ExecutionStorage<N>> ExecutionStore<N, E> {
 
 impl<N: Network, E: ExecutionStorage<N>> ExecutionStore<N, E> {
     /// Returns the transaction ID that executed the given `transition ID`.
-    pub fn find_transaction_id(&self, transition_id: &N::TransitionID) -> Result<Option<N::TransactionID>> {
-        self.storage.find_transaction_id(transition_id)
+    pub fn find_transaction_id_from_transition_id(
+        &self,
+        transition_id: &N::TransitionID,
+    ) -> Result<Option<N::TransactionID>> {
+        self.storage.find_transaction_id_from_transition_id(transition_id)
     }
 }
 
 impl<N: Network, E: ExecutionStorage<N>> ExecutionStore<N, E> {
     /// Returns an iterator over the execution transaction IDs, for all executions.
-    pub fn execution_ids(&self) -> impl '_ + Iterator<Item = Cow<'_, N::TransactionID>> {
+    pub fn execution_transaction_ids(&self) -> impl '_ + Iterator<Item = Cow<'_, N::TransactionID>> {
         self.storage.id_map().keys()
     }
 }
@@ -506,21 +512,21 @@ mod tests {
 
         for transition_id in transition_ids {
             // Ensure the transaction ID is not found.
-            let candidate = execution_store.find_transaction_id(&transition_id).unwrap();
+            let candidate = execution_store.find_transaction_id_from_transition_id(&transition_id).unwrap();
             assert_eq!(None, candidate);
 
             // Insert the execution.
             execution_store.insert(&transaction).unwrap();
 
             // Find the transaction ID.
-            let candidate = execution_store.find_transaction_id(&transition_id).unwrap();
+            let candidate = execution_store.find_transaction_id_from_transition_id(&transition_id).unwrap();
             assert_eq!(Some(transaction_id), candidate);
 
             // Remove the execution.
             execution_store.remove(&transaction_id).unwrap();
 
             // Ensure the transaction ID is not found.
-            let candidate = execution_store.find_transaction_id(&transition_id).unwrap();
+            let candidate = execution_store.find_transaction_id_from_transition_id(&transition_id).unwrap();
             assert_eq!(None, candidate);
         }
     }
diff --git a/synthesizer/src/store/transaction/mod.rs b/synthesizer/src/store/transaction/mod.rs
index 5ead3ca67..b5df79d0c 100644
--- a/synthesizer/src/store/transaction/mod.rs
+++ b/synthesizer/src/store/transaction/mod.rs
@@ -160,13 +160,16 @@ pub trait TransactionStorage<N: Network>: Clone + Send + Sync {
     }
 
     /// Returns the transaction ID that contains the given `transition ID`.
-    fn find_transaction_id(&self, transition_id: &N::TransitionID) -> Result<Option<N::TransactionID>> {
-        self.execution_store().find_transaction_id(transition_id)
+    fn find_transaction_id_from_transition_id(
+        &self,
+        transition_id: &N::TransitionID,
+    ) -> Result<Option<N::TransactionID>> {
+        self.execution_store().find_transaction_id_from_transition_id(transition_id)
     }
 
     /// Returns the transaction ID that contains the given `program ID`.
-    fn find_deployment_id(&self, program_id: &ProgramID<N>) -> Result<Option<N::TransactionID>> {
-        self.deployment_store().find_transaction_id(program_id)
+    fn find_transaction_id_from_program_id(&self, program_id: &ProgramID<N>) -> Result<Option<N::TransactionID>> {
+        self.deployment_store().find_transaction_id_from_program_id(program_id)
     }
 
     /// Returns the transaction for the given `transaction ID`.
@@ -386,13 +389,24 @@ impl<N: Network, T: TransactionStorage<N>> TransactionStore<N, T> {
 
 impl<N: Network, T: TransactionStorage<N>> TransactionStore<N, T> {
     /// Returns the transaction ID that contains the given `program ID`.
-    pub fn find_deployment_id(&self, program_id: &ProgramID<N>) -> Result<Option<N::TransactionID>> {
-        self.storage.deployment_store().find_transaction_id(program_id)
+    pub fn find_transaction_id_from_program_id(&self, program_id: &ProgramID<N>) -> Result<Option<N::TransactionID>> {
+        self.storage.deployment_store().find_transaction_id_from_program_id(program_id)
     }
 
     /// Returns the transaction ID that contains the given `transition ID`.
-    pub fn find_transaction_id(&self, transition_id: &N::TransitionID) -> Result<Option<N::TransactionID>> {
-        self.storage.execution_store().find_transaction_id(transition_id)
+    pub fn find_transaction_id_from_transition_id(
+        &self,
+        transition_id: &N::TransitionID,
+    ) -> Result<Option<N::TransactionID>> {
+        // Check if the transaction id exists in the execution store.
+        let execution_transaction =
+            self.storage.execution_store().find_transaction_id_from_transition_id(transition_id);
+
+        // Check if the transaction id exists in the transition store.
+        match execution_transaction {
+            Ok(None) => self.storage.deployment_store().find_transaction_id_from_transition_id(transition_id),
+            _ => execution_transaction,
+        }
     }
 }
 
@@ -415,13 +429,13 @@ impl<N: Network, T: TransactionStorage<N>> TransactionStore<N, T> {
     }
 
     /// Returns an iterator over the deployment transaction IDs, for all deployments.
-    pub fn deployment_ids(&self) -> impl '_ + Iterator<Item = Cow<'_, N::TransactionID>> {
-        self.storage.deployment_store().deployment_ids()
+    pub fn deployment_transaction_ids(&self) -> impl '_ + Iterator<Item = Cow<'_, N::TransactionID>> {
+        self.storage.deployment_store().deployment_transaction_ids()
     }
 
     /// Returns an iterator over the execution transaction IDs, for all executions.
-    pub fn execution_ids(&self) -> impl '_ + Iterator<Item = Cow<'_, N::TransactionID>> {
-        self.storage.execution_store().execution_ids()
+    pub fn execution_transaction_ids(&self) -> impl '_ + Iterator<Item = Cow<'_, N::TransactionID>> {
+        self.storage.execution_store().execution_transaction_ids()
     }
 
     /// Returns an iterator over the program IDs, for all deployments.
@@ -514,21 +528,21 @@ mod tests {
 
         for transition_id in transition_ids {
             // Ensure the transaction ID is not found.
-            let candidate = transaction_store.find_transaction_id(&transition_id).unwrap();
+            let candidate = transaction_store.find_transaction_id_from_transition_id(&transition_id).unwrap();
             assert_eq!(None, candidate);
 
             // Insert the transaction.
             transaction_store.insert(&transaction).unwrap();
 
             // Find the transaction ID.
-            let candidate = transaction_store.find_transaction_id(&transition_id).unwrap();
+            let candidate = transaction_store.find_transaction_id_from_transition_id(&transition_id).unwrap();
             assert_eq!(Some(transaction_id), candidate);
 
             // Remove the transaction.
             transaction_store.remove(&transaction_id).unwrap();
 
             // Ensure the transaction ID is not found.
-            let candidate = transaction_store.find_transaction_id(&transition_id).unwrap();
+            let candidate = transaction_store.find_transaction_id_from_transition_id(&transition_id).unwrap();
             assert_eq!(None, candidate);
         }
     }
diff --git a/synthesizer/src/vm/mod.rs b/synthesizer/src/vm/mod.rs
index ab302b503..90e98b4a7 100644
--- a/synthesizer/src/vm/mod.rs
+++ b/synthesizer/src/vm/mod.rs
@@ -61,7 +61,7 @@ impl<N: Network, C: ConsensusStorage<N>> VM<N, C> {
         // Retrieve the transaction store.
         let transaction_store = store.transaction_store();
         // Load the deployments from the store.
-        for transaction_id in transaction_store.deployment_ids() {
+        for transaction_id in transaction_store.deployment_transaction_ids() {
             // Retrieve the deployment.
             match transaction_store.get_deployment(&transaction_id)? {
                 // Load the deployment.
diff --git a/synthesizer/src/vm/verify.rs b/synthesizer/src/vm/verify.rs
index 3d49c0546..7f8e267c4 100644
--- a/synthesizer/src/vm/verify.rs
+++ b/synthesizer/src/vm/verify.rs
@@ -242,6 +242,11 @@ mod tests {
 
         // Ensure the deployment is valid.
         assert!(vm.verify_deployment(&deployment));
+
+        // Ensure that deserialization doesn't break the transaction verification.
+        let serialized_deployment = deployment.to_string();
+        let deployment_transaction: Deployment<CurrentNetwork> = serde_json::from_str(&serialized_deployment).unwrap();
+        assert!(vm.verify_deployment(&deployment_transaction));
     }
 
     #[test]
@@ -260,6 +265,12 @@ mod tests {
                 assert!(Inclusion::verify_execution(&execution).is_ok());
                 // Verify the execution.
                 assert!(vm.verify_execution(&execution));
+
+                // Ensure that deserialization doesn't break the transaction verification.
+                let serialized_execution = execution.to_string();
+                let execution_transaction: Execution<CurrentNetwork> =
+                    serde_json::from_str(&serialized_execution).unwrap();
+                assert!(vm.verify_execution(&execution_transaction));
             }
             _ => panic!("Expected an execution transaction"),
         }
@@ -325,6 +336,13 @@ mod tests {
         // Add the deployment block.
         vm.add_next_block(&deployment_block).unwrap();
 
+        // Fetch the unspent records.
+        let records = deployment_block.records().collect::<indexmap::IndexMap<_, _>>();
+
+        // Prepare the additional fee.
+        let credits = records.values().next().unwrap().decrypt(&caller_view_key).unwrap();
+        let additional_fee = (credits, 10);
+
         // Authorize.
         let authorization = vm
             .authorize(
@@ -342,7 +360,15 @@ mod tests {
         assert_eq!(authorization.len(), 1);
 
         // Execute.
-        let transaction = Transaction::execute_authorization(&vm, authorization, None, rng).unwrap();
+        let transaction = Transaction::execute_authorization_with_additional_fee(
+            &vm,
+            &caller_private_key,
+            authorization,
+            Some(additional_fee),
+            None,
+            rng,
+        )
+        .unwrap();
 
         // Verify.
         assert!(vm.verify(&transaction));
