diff --git a/.circleci/semver-checks.sh b/.circleci/semver-checks.sh
index b860f4421..adcb57322 100755
--- a/.circleci/semver-checks.sh
+++ b/.circleci/semver-checks.sh
@@ -3,7 +3,7 @@
 # Ensure that the command is installed.
 cargo install cargo-semver-checks@0.43.0 --locked
 
-BASELINE_REV=65a079a87 # UPDATE ME ON NECESSARY BREAKING CHANGES
+BASELINE_REV=8970619a # UPDATE ME ON NECESSARY BREAKING CHANGES
 
 # Exclude CLI as it has been removed
 cargo semver-checks --workspace --default-features \
diff --git a/ledger/src/get.rs b/ledger/src/get.rs
index b330029f7..90e598d8f 100644
--- a/ledger/src/get.rs
+++ b/ledger/src/get.rs
@@ -254,12 +254,18 @@ impl<N: Network, C: ConsensusStorage<N>> Ledger<N, C> {
 
     /// Returns the program for the given `program ID` and `edition`.
     pub fn get_program_for_edition(&self, program_id: ProgramID<N>, edition: u16) -> Result<Program<N>> {
-        match self.vm.block_store().get_program_for_edition(&program_id, edition)? {
+        match self.try_get_program_for_edition(&program_id, edition)? {
             Some(program) => Ok(program),
             None => bail!("Missing program for ID {program_id} and edition {edition}"),
         }
     }
 
+    /// Returns the program for the given `program ID` and `edition`,
+    /// or `None` if no program of this ID and edition exists.
+    pub fn try_get_program_for_edition(&self, program_id: &ProgramID<N>, edition: u16) -> Result<Option<Program<N>>> {
+        self.vm.block_store().get_program_for_edition(program_id, edition)
+    }
+
     /// Returns the block solutions for the given block height.
     pub fn get_solutions(&self, height: u32) -> Result<Solutions<N>> {
         // If the height is 0, return the genesis block solutions.
diff --git a/ledger/store/src/transaction/deployment.rs b/ledger/store/src/transaction/deployment.rs
index c3a12e503..c7c28e9a3 100644
--- a/ledger/store/src/transaction/deployment.rs
+++ b/ledger/store/src/transaction/deployment.rs
@@ -392,13 +392,6 @@ pub trait DeploymentStorage<N: Network>: Clone + Send + Sync {
 
     /// Returns the latest edition for the given `program ID`.
     fn get_latest_edition_for_program(&self, program_id: &ProgramID<N>) -> Result<Option<u16>> {
-        // Check if the program ID is for 'credits.aleo'.
-        // This case is handled separately, as it is a default program of the VM.
-        // TODO (howardwu): After we update 'fee' rules and 'Ratify' in genesis, we can remove this.
-        if program_id == &ProgramID::from_str("credits.aleo")? {
-            return Ok(None);
-        }
-
         Ok(self.edition_map().get_confirmed(program_id)?.map(|x| *x))
     }
 
@@ -421,13 +414,6 @@ pub trait DeploymentStorage<N: Network>: Clone + Send + Sync {
 
     /// Returns the latest program for the given `program ID`.
     fn get_latest_program(&self, program_id: &ProgramID<N>) -> Result<Option<Program<N>>> {
-        // Check if the program ID is for 'credits.aleo'.
-        // This case is handled separately, as it is a default program of the VM.
-        // TODO (howardwu): After we update 'fee' rules and 'Ratify' in genesis, we can remove this.
-        if program_id == &ProgramID::from_str("credits.aleo")? {
-            return Ok(Some(Program::credits()?));
-        }
-
         // Retrieve the latest edition.
         let edition = match self.get_latest_edition_for_program(program_id)? {
             Some(edition) => edition,
@@ -442,18 +428,7 @@ pub trait DeploymentStorage<N: Network>: Clone + Send + Sync {
 
     /// Returns the program for the given `program ID` and `edition`.
     fn get_program_for_edition(&self, program_id: &ProgramID<N>, edition: u16) -> Result<Option<Program<N>>> {
-        // Check if the program ID is for 'credits.aleo'.
-        // This case is handled separately, as it is a default program of the VM.
-        // TODO (howardwu): After we update 'fee' rules and 'Ratify' in genesis, we can remove this.
-        if program_id == &ProgramID::from_str("credits.aleo")? {
-            return Ok(Some(Program::credits()?));
-        }
-
-        // Retrieve the program.
-        match self.program_map().get_confirmed(&(*program_id, edition))? {
-            Some(program) => Ok(Some(program.into_owned())),
-            None => bail!("Failed to get program '{program_id}' (edition {edition})"),
-        }
+        self.program_map().get_confirmed(&(*program_id, edition)).map(|p| p.map(|p| p.into_owned()))
     }
 
     /// Returns the latest verifying key for the given `program ID` and `function name`.
@@ -697,6 +672,12 @@ impl<N: Network, D: DeploymentStorage<N>> DeploymentStore<N, D> {
     pub fn open(fee_store: FeeStore<N, D::FeeStorage>) -> Result<Self> {
         // Initialize the deployment storage.
         let storage = D::open(fee_store)?;
+
+        // Insert `credits.aleo`, which is the default program.
+        let credits_id = ProgramID::from_str("credits.aleo")?;
+        storage.edition_map().insert(credits_id, 0)?;
+        storage.program_map().insert((credits_id, 0), Program::credits()?)?;
+
         // Return the deployment store.
         Ok(Self { storage, _phantom: PhantomData })
     }
