diff --git a/console/program/src/data/plaintext/mod.rs b/console/program/src/data/plaintext/mod.rs
index e9a40f28c..001d45958 100644
--- a/console/program/src/data/plaintext/mod.rs
+++ b/console/program/src/data/plaintext/mod.rs
@@ -33,7 +33,10 @@ use snarkvm_console_network::Network;
 use snarkvm_console_types::prelude::*;
 
 use indexmap::IndexMap;
-use std::sync::OnceLock;
+use std::{
+    hash::{Hash, Hasher},
+    sync::OnceLock,
+};
 
 #[derive(Clone)]
 pub enum Plaintext<N: Network> {
@@ -121,6 +124,25 @@ impl<N: Network> From<&Literal<N>> for Plaintext<N> {
     }
 }
 
+impl<N: Network> Hash for Plaintext<N> {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        match self {
+            Self::Literal(literal, _bits) => {
+                literal.hash(state);
+            }
+            Self::Struct(fields, _bits) => {
+                for (name, value) in fields {
+                    name.hash(state);
+                    value.hash(state);
+                }
+            }
+            Self::Array(array, _bits) => {
+                array.hash(state);
+            }
+        }
+    }
+}
+
 // A macro that derives implementations of `From` for arrays of a plaintext literals of various sizes.
 macro_rules! impl_plaintext_from_array {
     ($element:ident, $($size:literal),+) => {
diff --git a/ledger/Cargo.toml b/ledger/Cargo.toml
index 93ee33e9a..9c4bea1bd 100644
--- a/ledger/Cargo.toml
+++ b/ledger/Cargo.toml
@@ -62,6 +62,10 @@ async = [
   "snarkvm-ledger-query/async",
   "snarkvm-synthesizer/async"
 ]
+history = [
+  "snarkvm-ledger-store/history",
+  "snarkvm-synthesizer/history",
+]
 locktick = [
   "dep:locktick",
   "snarkvm-ledger-puzzle/locktick",
diff --git a/ledger/src/advance.rs b/ledger/src/advance.rs
index 6f756b27e..371808da8 100644
--- a/ledger/src/advance.rs
+++ b/ledger/src/advance.rs
@@ -21,6 +21,8 @@ impl<N: Network, C: ConsensusStorage<N>> Ledger<N, C> {
     /// Returns a candidate for the next block in the ledger, using a committed subdag and its transmissions.
     /// This candidate can then be passed to [`Ledger::advance_to_next_block`] to be added to the ledger.
     ///
+    /// The function will prevent concurrent update to the ledger, and may block if an update is currently in progress.
+    ///
     /// # Panics
     /// This function panics if called from an async context.
     pub fn prepare_advance_to_next_quorum_block<R: Rng + CryptoRng>(
@@ -28,14 +30,17 @@ impl<N: Network, C: ConsensusStorage<N>> Ledger<N, C> {
         subdag: Subdag<N>,
         transmissions: IndexMap<TransmissionID<N>, Transmission<N>>,
         rng: &mut R,
-    ) -> Result<Block<N>> {
+    ) -> Result<Block<N>, CheckBlockError<N>> {
         // Retrieve the latest block as the previous block (for the next block).
-        let previous_block = self.latest_block();
+        // Hold this lock while preparing the template, so that the latest block does not change mid-speculation.
+        let previous_block = self.current_block.read();
 
         // Decouple the transmissions into ratifications, solutions, and transactions.
         let (ratifications, solutions, transactions) = decouple_transmissions(transmissions.into_iter())?;
         // Currently, we do not support ratifications from the memory pool.
-        ensure!(ratifications.is_empty(), "Ratifications are currently unsupported from the memory pool");
+        if !ratifications.is_empty() {
+            return Err(anyhow!("Ratifications are currently unsupported from the memory pool").into());
+        }
         // Construct the block template.
         let (header, ratifications, solutions, aborted_solution_ids, transactions, aborted_transaction_ids) =
             self.construct_block_template(&previous_block, Some(&subdag), ratifications, solutions, transactions, rng)?;
@@ -51,6 +56,7 @@ impl<N: Network, C: ConsensusStorage<N>> Ledger<N, C> {
             transactions,
             aborted_transaction_ids,
         )
+        .map_err(|e| CheckBlockError::Other(e))
     }
 
     /// Returns a candidate for the next block in the ledger.
@@ -68,12 +74,15 @@ impl<N: Network, C: ConsensusStorage<N>> Ledger<N, C> {
         candidate_solutions: Vec<Solution<N>>,
         candidate_transactions: Vec<Transaction<N>>,
         rng: &mut R,
-    ) -> Result<Block<N>> {
+    ) -> Result<Block<N>, CheckBlockError<N>> {
         // Currently, we do not support ratifications from the memory pool.
-        ensure!(candidate_ratifications.is_empty(), "Ratifications are currently unsupported from the memory pool");
+        if !candidate_ratifications.is_empty() {
+            return Err(anyhow!("Ratifications are currently unsupported from the memory pool").into());
+        }
 
         // Retrieve the latest block as the previous block (for the next block).
-        let previous_block = self.latest_block();
+        // Hold this lock while preparing the template, so that the latest block does not change mid-speculation.
+        let previous_block = self.current_block.read();
 
         // Construct the block template.
         let (header, ratifications, solutions, aborted_solution_ids, transactions, aborted_transaction_ids) = self
@@ -98,20 +107,26 @@ impl<N: Network, C: ConsensusStorage<N>> Ledger<N, C> {
             aborted_transaction_ids,
             rng,
         )
+        .map_err(|e| CheckBlockError::Other(e))
     }
 
     /// Adds the given block as the next block in the ledger.
     ///
     /// This function expects a valid block, that either was created by a trusted source, or successfully passed
     /// the blocks checks (e.g. [`Ledger::check_next_block`]).
-    /// Note, that it is still possible that this function returns an error for a valid block, if there are concurrent tasks
-    /// updating the ledger.
+    ///
+    /// # Concurrency
+    /// It is guaranteed that at most one call this function is executing at any time and that no reads to the Ledger
+    /// (e.g, by [`Self::prepare_next_beacon_block`] or [`Self::check_next_block`]) execute while the ledger is advancing.
+    /// However, it is still possible that a *previously* valid block is rejected by a call to this function,
+    /// if the ledger advanced between calling `prepare_next_*_block` and passing it to this function.
     ///
     /// # Panics
     /// This function panics if called from an async context.
     pub fn advance_to_next_block(&self, block: &Block<N>) -> Result<()> {
-        // Acquire the write lock on the current block.
+        // Acquire a write lock to current_block to prevent current updates or reads to the block or state root.
         let mut current_block = self.current_block.write();
+
         // Check again for any possible race conditions.
         if current_block.is_genesis()? {
             // current block is initialized as the genesis block, but the ledger will
@@ -233,6 +248,12 @@ where
 impl<N: Network, C: ConsensusStorage<N>> Ledger<N, C> {
     /// Constructs a block template for the next block in the ledger.
     ///
+    /// # Concurrency
+    /// This function assumes that `previous_block` is identical to [`Self::current_block`]. To ensure this,
+    /// you should hold a read (or write) lock to the latter while calling this function.
+    /// Otherwise, it would be possible that the state root is inconsistent with the previous block, and for the
+    /// returned block template to be invalid.
+    ///
     /// # Panics
     /// This function panics if called from an async context.
     #[allow(clippy::type_complexity)]
@@ -244,17 +265,23 @@ impl<N: Network, C: ConsensusStorage<N>> Ledger<N, C> {
         candidate_solutions: Vec<Solution<N>>,
         candidate_transactions: Vec<Transaction<N>>,
         rng: &mut R,
-    ) -> Result<(Header<N>, Ratifications<N>, Solutions<N>, Vec<SolutionID<N>>, Transactions<N>, Vec<N::TransactionID>)>
-    {
+    ) -> Result<
+        (Header<N>, Ratifications<N>, Solutions<N>, Vec<SolutionID<N>>, Transactions<N>, Vec<N::TransactionID>),
+        CheckBlockError<N>,
+    > {
         // Construct the solutions.
         let (solutions, aborted_solutions, solutions_root, combined_proof_target) = match candidate_solutions.is_empty()
         {
             true => (None, vec![], Field::<N>::zero(), 0u128),
             false => {
-                // Retrieve the latest epoch hash.
-                let latest_epoch_hash = self.latest_epoch_hash()?;
+                // Retrieve the latest epoch hash directly from the previous block.
+                // (so we avoid read-locking current_block again)
+                let latest_epoch_hash = match self.current_epoch_hash.read().as_ref() {
+                    Some(epoch_hash) => *epoch_hash,
+                    None => self.get_epoch_hash(previous_block.height())?,
+                };
                 // Retrieve the latest proof target.
-                let latest_proof_target = self.latest_proof_target();
+                let latest_proof_target = previous_block.proof_target();
                 // Separate the candidate solutions into valid and aborted solutions.
                 let mut accepted_solutions: IndexMap<Address<N>, u64> = IndexMap::new();
                 let (valid_candidate_solutions, aborted_candidate_solutions) =
@@ -262,7 +289,8 @@ impl<N: Network, C: ConsensusStorage<N>> Ledger<N, C> {
                         let prover_address = solution.address();
                         let num_accepted_solutions = accepted_solutions.get(&prover_address).copied().unwrap_or(0);
                         // Check if the prover has reached their solution limit.
-                        if self.is_solution_limit_reached(&prover_address, num_accepted_solutions) {
+                        if self.is_solution_limit_reached_inner(previous_block, &prover_address, num_accepted_solutions)
+                        {
                             return false;
                         }
                         // Check if the solution is valid and update the number of accepted solutions.
@@ -314,7 +342,16 @@ impl<N: Network, C: ConsensusStorage<N>> Ledger<N, C> {
 
         // Compute the next round number.
         let next_round = match subdag {
-            Some(subdag) => subdag.anchor_round(),
+            Some(subdag) => {
+                if previous_block.round() >= subdag.anchor_round() {
+                    return Err(CheckBlockError::InvalidRound {
+                        new: subdag.anchor_round(),
+                        previous: previous_block.round(),
+                    });
+                }
+
+                subdag.anchor_round()
+            }
             None => previous_block.round().saturating_add(1),
         };
         // Compute the next height.
@@ -364,7 +401,8 @@ impl<N: Network, C: ConsensusStorage<N>> Ledger<N, C> {
             N::ANCHOR_HEIGHT,
             N::BLOCK_TIME,
             combined_proof_target,
-            u64::try_from(latest_cumulative_proof_target)?,
+            u64::try_from(latest_cumulative_proof_target)
+                .map_err(|e| anyhow!("Failed to convert cumulative proof target to u64 - {e}"))?,
             latest_coinbase_target,
         )?;
 
diff --git a/ledger/src/check_next_block.rs b/ledger/src/check_next_block.rs
index fcc15c38d..a9b27f092 100644
--- a/ledger/src/check_next_block.rs
+++ b/ledger/src/check_next_block.rs
@@ -49,6 +49,8 @@ pub enum CheckBlockError<N: Network> {
     BlockAlreadyExists { hash: N::BlockHash },
     #[error("Block has invalid height. Expected {expected}, but got {actual}")]
     InvalidHeight { expected: u32, actual: u32 },
+    #[error("Block has invalid round. Was {new}, but must be greater than previous round ({previous})")]
+    InvalidRound { new: u64, previous: u64 },
     #[error("Block has invalid hash")]
     InvalidHash,
     /// An error related to the given prefix of pending blocks.
@@ -148,7 +150,7 @@ impl<N: Network, C: ConsensusStorage<N>> Ledger<N, C> {
         self.check_block_subdag_quorum(block)?;
 
         // Determine if the block subdag is correctly constructed and is not a combination of multiple subdags.
-        self.check_block_subdag_atomicity(block)?;
+        self.check_block_subdag_atomicity(block, &latest_block)?;
 
         // Ensure that all leaves of the subdag point to valid batches in other subdags/blocks.
         self.check_block_subdag_leaves(block, prefix)?;
@@ -201,11 +203,20 @@ impl<N: Network, C: ConsensusStorage<N>> Ledger<N, C> {
         block: &Block<N>,
         rng: &mut R,
     ) -> Result<(), CheckBlockError<N>> {
-        let latest_block = self.current_block.read();
+        // Grab lock to the previous block here, to ensure it does not change mid-check.
+        let previous_block = self.current_block.read();
 
         // Ensure, again, that the ledger has not advanced yet. This prevents cryptic errors form appearing during the block check.
-        if block.height() != latest_block.height() + 1 {
-            return Err(CheckBlockError::InvalidHeight { expected: latest_block.height() + 1, actual: block.height() });
+        if block.height() != previous_block.height() + 1 {
+            return Err(CheckBlockError::InvalidHeight {
+                expected: previous_block.height() + 1,
+                actual: block.height(),
+            });
+        }
+
+        // Also ensure the round is valid, otherwise speculation on transactions will fail with a cryptic error.
+        if block.round() <= previous_block.round() {
+            return Err(CheckBlockError::InvalidRound { new: block.round(), previous: previous_block.round() });
         }
 
         // Ensure the solutions do not already exist.
@@ -229,7 +240,7 @@ impl<N: Network, C: ConsensusStorage<N>> Ledger<N, C> {
         )?;
 
         // Ensure speculation over the unconfirmed transactions is correct and ensure each transaction is well-formed and unique.
-        let time_since_last_block = block.timestamp().saturating_sub(self.latest_timestamp());
+        let time_since_last_block = block.timestamp().saturating_sub(previous_block.timestamp());
         let ratified_finalize_operations = self.vm.check_speculate(
             state,
             time_since_last_block,
@@ -253,15 +264,21 @@ impl<N: Network, C: ConsensusStorage<N>> Ledger<N, C> {
                 .ok_or(anyhow!("Failed to fetch committee lookback for round {penultimate_round}"))?
         };
 
+        // Get the latest epoch hash.
+        let latest_epoch_hash = match self.current_epoch_hash.read().as_ref() {
+            Some(epoch_hash) => *epoch_hash,
+            None => self.get_epoch_hash(previous_block.height())?,
+        };
+
         // Ensure the block is correct.
         let (expected_existing_solution_ids, expected_existing_transaction_ids) = block
             .verify(
-                &latest_block,
+                &previous_block,
                 self.latest_state_root(),
                 &previous_committee_lookback,
                 &committee_lookback,
                 self.puzzle(),
-                self.latest_epoch_hash()?,
+                latest_epoch_hash,
                 OffsetDateTime::now_utc().unix_timestamp(),
                 ratified_finalize_operations,
             )
@@ -274,7 +291,7 @@ impl<N: Network, C: ConsensusStorage<N>> Ledger<N, C> {
                 let prover_address = solution.address();
                 let num_accepted_solutions = *accepted_solutions.get(&prover_address).unwrap_or(&0);
                 // Check if the prover has reached their solution limit.
-                if self.is_solution_limit_reached(&prover_address, num_accepted_solutions) {
+                if self.is_solution_limit_reached_inner(&previous_block, &prover_address, num_accepted_solutions) {
                     return Err(CheckBlockError::SolutionLimitReached { prover_address });
                 }
                 // Track the already accepted solutions.
@@ -403,7 +420,9 @@ impl<N: Network, C: ConsensusStorage<N>> Ledger<N, C> {
     /// Checks that the block subdag can not be split into multiple valid subdags.
     ///
     /// Called by [`Self::check_block_subdag`]
-    fn check_block_subdag_atomicity(&self, block: &Block<N>) -> Result<()> {
+    fn check_block_subdag_atomicity(&self, block: &Block<N>, latest_block: &Block<N>) -> Result<()> {
+        let latest_round = latest_block.round();
+
         // Returns `true` if there is a path from the previous certificate to the current certificate.
         fn is_linked<N: Network>(
             subdag: &Subdag<N>,
@@ -432,8 +451,7 @@ impl<N: Network, C: ConsensusStorage<N>> Ledger<N, C> {
         };
 
         // Iterate over the rounds to find possible leader certificates.
-        for round in (self.latest_round().saturating_add(2)..=subdag.anchor_round().saturating_sub(2)).rev().step_by(2)
-        {
+        for round in (latest_round.saturating_add(2)..=subdag.anchor_round().saturating_sub(2)).rev().step_by(2) {
             // Retrieve the previous committee lookback.
             let previous_committee_lookback = self
                 .get_committee_lookback_for_round(round)?
diff --git a/ledger/src/is_solution_limit_reached.rs b/ledger/src/is_solution_limit_reached.rs
index 28c5e3eec..93a64e41c 100644
--- a/ledger/src/is_solution_limit_reached.rs
+++ b/ledger/src/is_solution_limit_reached.rs
@@ -83,12 +83,26 @@ pub fn maximum_allowed_solutions_per_epoch<N: Network>(prover_stake: u64, curren
 
 impl<N: Network, C: ConsensusStorage<N>> Ledger<N, C> {
     /// Returns the number of remaining solutions a prover can submit in the current epoch.
+    ///
+    /// # Locking
+    /// This function may deadlock if called while holding a write lock to the current block.
     pub fn num_remaining_solutions(&self, prover_address: &Address<N>, additional_solutions_in_block: u64) -> u64 {
+        self.num_remaining_solutions_inner(&self.current_block.read(), prover_address, additional_solutions_in_block)
+    }
+
+    /// Internal version of [`Self::num_remaining_solutions`] to be used when already holding a lock to the current block.
+    pub(super) fn num_remaining_solutions_inner(
+        &self,
+        latest_block: &Block<N>,
+        prover_address: &Address<N>,
+        additional_solutions_in_block: u64,
+    ) -> u64 {
         // Fetch the prover's stake.
         let prover_stake = self.get_bonded_amount(prover_address).unwrap_or(0);
 
         // Determine the maximum number of solutions allowed based on this prover's stake.
-        let maximum_allowed_solutions = maximum_allowed_solutions_per_epoch::<N>(prover_stake, self.latest_timestamp());
+        let maximum_allowed_solutions =
+            maximum_allowed_solutions_per_epoch::<N>(prover_stake, latest_block.timestamp());
 
         // Fetch the number of solutions the prover has earned rewards for in the current epoch.
         let prover_num_solutions_in_epoch = *self.epoch_provers_cache.read().get(prover_address).unwrap_or(&0);
@@ -101,9 +115,23 @@ impl<N: Network, C: ConsensusStorage<N>> Ledger<N, C> {
     }
 
     /// Returns `true` if the given prover address has reached their solution limit for the current epoch.
+    ///
+    /// # Locking
+    /// This function may deadlock if called while holding a write lock to the current block.
     pub fn is_solution_limit_reached(&self, prover_address: &Address<N>, additional_solutions_in_block: u64) -> bool {
+        self.is_solution_limit_reached_inner(&self.current_block.read(), prover_address, additional_solutions_in_block)
+    }
+
+    /// Internal version of [`Self::is_solution_limit_reached`] to be used when already holding a lock to the current block.
+    pub(super) fn is_solution_limit_reached_inner(
+        &self,
+        latest_block: &Block<N>,
+        prover_address: &Address<N>,
+        additional_solutions_in_block: u64,
+    ) -> bool {
         // Calculate the number of remaining solutions for the prover.
-        let num_remaining_solutions = self.num_remaining_solutions(prover_address, additional_solutions_in_block);
+        let num_remaining_solutions =
+            self.num_remaining_solutions_inner(latest_block, prover_address, additional_solutions_in_block);
 
         // If the number of remaining solutions is zero, the limit is reached.
         num_remaining_solutions == 0
diff --git a/ledger/src/lib.rs b/ledger/src/lib.rs
index 693fb3fac..1fc1d141c 100644
--- a/ledger/src/lib.rs
+++ b/ledger/src/lib.rs
@@ -133,8 +133,10 @@ pub struct InnerLedger<N: Network, C: ConsensusStorage<N>> {
     vm: VM<N, C>,
     /// The genesis block.
     genesis_block: Block<N>,
+
     /// The current epoch hash.
     current_epoch_hash: RwLock<Option<N::BlockHash>>,
+
     /// The committee resulting from all the on-chain staking activity.
     ///
     /// This includes any bonding and unbonding transactions in the latest block.
@@ -154,9 +156,10 @@ pub struct InnerLedger<N: Network, C: ConsensusStorage<N>> {
 
     /// The latest block that was added to the ledger.
     ///
-    /// This lock is also used as a way to prevent concurrent updates to the ledger, and to ensure that
-    /// the ledger does not advance while certain check happen.
+    /// This lock is also to ensure *atomicity* of calls to `[Ledger::advance`], i.e., to guarantee that
+    /// there cannot be multiple concurrent ledger advancements and that ledger state cannot be read while advancement happens.
     current_block: RwLock<Block<N>>,
+
     /// The recent committees of interest paired with their applicable rounds.
     ///
     /// Each entry consisting of a round `R` and a committee `C`,
diff --git a/ledger/src/tests.rs b/ledger/src/tests.rs
index 32536e038..403fc58c9 100644
--- a/ledger/src/tests.rs
+++ b/ledger/src/tests.rs
@@ -649,6 +649,41 @@ fn test_bond_and_unbond_validator() {
         )
         .unwrap();
 
+    // Procure mapping-related objects.
+    let program_id = ProgramID::<CurrentNetwork>::from_str("credits.aleo").unwrap();
+    let metadata_mapping_name = Identifier::from_str("metadata").unwrap();
+    let metadata_mapping_key =
+        Plaintext::<CurrentNetwork>::from_str("aleo1qqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqq3ljyzc")
+            .unwrap();
+
+    // Check the initial historical mapping values.
+    #[cfg(feature = "history")]
+    {
+        let initial_mapping_value = ledger
+            .vm()
+            .finalize_store()
+            .get_historical_mapping_value(program_id, metadata_mapping_name, metadata_mapping_key.clone(), 0)
+            .unwrap()
+            .unwrap();
+        assert_eq!(&*initial_mapping_value, &Value::<CurrentNetwork>::try_from("4u32").unwrap());
+
+        let initial_mapping_value_overshot = ledger
+            .vm()
+            .finalize_store()
+            .get_historical_mapping_value(program_id, metadata_mapping_name, metadata_mapping_key.clone(), 10)
+            .unwrap()
+            .unwrap();
+        assert_eq!(&*initial_mapping_value_overshot, &Value::<CurrentNetwork>::try_from("4u32").unwrap());
+
+        let initial_mapping_heights = ledger
+            .vm()
+            .finalize_store()
+            .get_mapping_update_heights(program_id, metadata_mapping_name, metadata_mapping_key.clone())
+            .unwrap()
+            .unwrap();
+        assert_eq!(&*initial_mapping_heights, &[0]);
+    }
+
     // Check that the next block is valid.
     ledger.check_next_block(&transfer_block, rng).unwrap();
 
@@ -683,19 +718,59 @@ fn test_bond_and_unbond_validator() {
     // Add the bond public block to the ledger.
     ledger.advance_to_next_block(&bond_validator_block).unwrap();
 
+    // Check the historical mapping values after the bonding.
+    #[cfg(feature = "history")]
+    {
+        let initial_mapping_value = ledger
+            .vm()
+            .finalize_store()
+            .get_historical_mapping_value(program_id, metadata_mapping_name, metadata_mapping_key.clone(), 0)
+            .unwrap()
+            .unwrap();
+        assert_eq!(&*initial_mapping_value, &Value::<CurrentNetwork>::try_from("4u32").unwrap());
+
+        let initial_mapping_value_overshot = ledger
+            .vm()
+            .finalize_store()
+            .get_historical_mapping_value(program_id, metadata_mapping_name, metadata_mapping_key.clone(), 1)
+            .unwrap()
+            .unwrap();
+        assert_eq!(&*initial_mapping_value_overshot, &Value::<CurrentNetwork>::try_from("4u32").unwrap());
+
+        let post_bond_mapping_value = ledger
+            .vm()
+            .finalize_store()
+            .get_historical_mapping_value(program_id, metadata_mapping_name, metadata_mapping_key.clone(), 2)
+            .unwrap()
+            .unwrap();
+        assert_eq!(&*post_bond_mapping_value, &Value::<CurrentNetwork>::try_from("5u32").unwrap());
+
+        let post_bond_mapping_value_overshot = ledger
+            .vm()
+            .finalize_store()
+            .get_historical_mapping_value(program_id, metadata_mapping_name, metadata_mapping_key.clone(), 5)
+            .unwrap()
+            .unwrap();
+        assert_eq!(&*post_bond_mapping_value_overshot, &Value::<CurrentNetwork>::try_from("5u32").unwrap());
+
+        let post_bond_mapping_heights = ledger
+            .vm()
+            .finalize_store()
+            .get_mapping_update_heights(program_id, metadata_mapping_name, metadata_mapping_key.clone())
+            .unwrap()
+            .unwrap();
+        assert_eq!(&*post_bond_mapping_heights, &[0, 2]);
+    }
+
     // Check that the committee is updated with the new member.
     let committee = ledger.latest_committee().unwrap();
     assert!(committee.is_committee_member(new_member_address));
 
     // Check that number of validators in the `metadata` mapping in `credits.aleo` is updated.
-    let program_id = ProgramID::<CurrentNetwork>::from_str("credits.aleo").unwrap();
-    let metadata_mapping_name = Identifier::from_str("metadata").unwrap();
-    let key = Plaintext::<CurrentNetwork>::from_str("aleo1qqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqq3ljyzc")
-        .unwrap();
     let num_validators = match ledger
         .vm()
         .finalize_store()
-        .get_value_confirmed(program_id, metadata_mapping_name, &key)
+        .get_value_confirmed(program_id, metadata_mapping_name, &metadata_mapping_key)
         .unwrap()
         .unwrap()
     {
@@ -734,19 +809,67 @@ fn test_bond_and_unbond_validator() {
     // Add the bond public block to the ledger.
     ledger.advance_to_next_block(&unbond_public_block).unwrap();
 
+    // Check the historical mapping values after the unbonding.
+    #[cfg(feature = "history")]
+    {
+        let initial_mapping_value = ledger
+            .vm()
+            .finalize_store()
+            .get_historical_mapping_value(program_id, metadata_mapping_name, metadata_mapping_key.clone(), 0)
+            .unwrap()
+            .unwrap();
+        assert_eq!(&*initial_mapping_value, &Value::<CurrentNetwork>::try_from("4u32").unwrap());
+
+        let initial_mapping_value_overshot = ledger
+            .vm()
+            .finalize_store()
+            .get_historical_mapping_value(program_id, metadata_mapping_name, metadata_mapping_key.clone(), 1)
+            .unwrap()
+            .unwrap();
+        assert_eq!(&*initial_mapping_value_overshot, &Value::<CurrentNetwork>::try_from("4u32").unwrap());
+
+        let post_bond_mapping_value = ledger
+            .vm()
+            .finalize_store()
+            .get_historical_mapping_value(program_id, metadata_mapping_name, metadata_mapping_key.clone(), 2)
+            .unwrap()
+            .unwrap();
+        assert_eq!(&*post_bond_mapping_value, &Value::<CurrentNetwork>::try_from("5u32").unwrap());
+
+        let post_unbond_mapping_value = ledger
+            .vm()
+            .finalize_store()
+            .get_historical_mapping_value(program_id, metadata_mapping_name, metadata_mapping_key.clone(), 3)
+            .unwrap()
+            .unwrap();
+        assert_eq!(&*post_unbond_mapping_value, &Value::<CurrentNetwork>::try_from("4u32").unwrap());
+
+        let post_unbond_mapping_value_overshot = ledger
+            .vm()
+            .finalize_store()
+            .get_historical_mapping_value(program_id, metadata_mapping_name, metadata_mapping_key.clone(), 100)
+            .unwrap()
+            .unwrap();
+        assert_eq!(&*post_unbond_mapping_value_overshot, &Value::<CurrentNetwork>::try_from("4u32").unwrap());
+
+        let post_unbond_mapping_heights = ledger
+            .vm()
+            .finalize_store()
+            .get_mapping_update_heights(program_id, metadata_mapping_name, metadata_mapping_key.clone())
+            .unwrap()
+            .unwrap();
+        assert_eq!(&*post_unbond_mapping_heights, &[0, 2, 3]);
+    }
+
     // Check that the committee does not include the new member.
     let committee = ledger.latest_committee().unwrap();
     assert!(!committee.is_committee_member(new_member_address));
 
-    // Check that number of validators in the `metadata` mapping in `credtis.aleo` is updated.
-    let program_id = ProgramID::<CurrentNetwork>::from_str("credits.aleo").unwrap();
-    let metadata_mapping_name = Identifier::from_str("metadata").unwrap();
-    let key = Plaintext::<CurrentNetwork>::from_str("aleo1qqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqq3ljyzc")
-        .unwrap();
+    // Check that number of validators in the `metadata` mapping in `credits.aleo` is updated.
     let num_validators = match ledger
         .vm()
         .finalize_store()
-        .get_value_confirmed(program_id, metadata_mapping_name, &key)
+        .get_value_confirmed(program_id, metadata_mapping_name, &metadata_mapping_key)
         .unwrap()
         .unwrap()
     {
diff --git a/ledger/store/Cargo.toml b/ledger/store/Cargo.toml
index 12ad67fff..d10ed5be6 100644
--- a/ledger/store/Cargo.toml
+++ b/ledger/store/Cargo.toml
@@ -18,6 +18,7 @@ edition = "2024"
 
 [features]
 default = [ "indexmap/rayon" ]
+history = [ ]
 locktick = [ "dep:locktick", "snarkvm-ledger-puzzle/locktick" ]
 rocks = [ "rocksdb", "smallvec" ]
 serial = [
diff --git a/ledger/store/src/helpers/memory/internal/map.rs b/ledger/store/src/helpers/memory/internal/map.rs
index 3992121a6..5e495a5c0 100644
--- a/ledger/store/src/helpers/memory/internal/map.rs
+++ b/ledger/store/src/helpers/memory/internal/map.rs
@@ -38,7 +38,7 @@ use std::{
 
 #[derive(Clone)]
 pub struct MemoryMap<
-    K: Copy + Clone + PartialEq + Eq + Hash + Serialize + for<'de> Deserialize<'de> + Send + Sync,
+    K: Clone + PartialEq + Eq + Hash + Serialize + for<'de> Deserialize<'de> + Send + Sync,
     V: Clone + Serialize + for<'de> Deserialize<'de> + Send + Sync,
 > {
     // The reason for using BTreeMap with binary keys is for the order of items to be the same as
@@ -51,7 +51,7 @@ pub struct MemoryMap<
 }
 
 impl<
-    K: Copy + Clone + PartialEq + Eq + Hash + Serialize + for<'de> Deserialize<'de> + Send + Sync,
+    K: Clone + PartialEq + Eq + Hash + Serialize + for<'de> Deserialize<'de> + Send + Sync,
     V: Clone + Serialize + for<'de> Deserialize<'de> + Send + Sync,
 > Default for MemoryMap<K, V>
 {
@@ -66,7 +66,7 @@ impl<
 }
 
 impl<
-    K: Copy + Clone + PartialEq + Eq + Hash + Serialize + for<'de> Deserialize<'de> + Send + Sync,
+    K: Clone + PartialEq + Eq + Hash + Serialize + for<'de> Deserialize<'de> + Send + Sync,
     V: Clone + Serialize + for<'de> Deserialize<'de> + Send + Sync,
 > FromIterator<(K, V)> for MemoryMap<K, V>
 {
@@ -87,7 +87,7 @@ impl<
 
 impl<
     'a,
-    K: 'a + Copy + Clone + PartialEq + Eq + Hash + Serialize + for<'de> Deserialize<'de> + Send + Sync,
+    K: 'a + Clone + PartialEq + Eq + Hash + Serialize + for<'de> Deserialize<'de> + Send + Sync,
     V: 'a + Clone + Serialize + for<'de> Deserialize<'de> + Send + Sync,
 > Map<'a, K, V> for MemoryMap<K, V>
 {
@@ -118,7 +118,7 @@ impl<
         match self.is_atomic_in_progress() {
             // If a batch is in progress, add the key-None pair to the batch.
             true => {
-                self.atomic_batch.lock().push((*key, None));
+                self.atomic_batch.lock().push((key.clone(), None));
             }
             // Otherwise, remove the key-value pair directly from the map.
             false => {
@@ -261,7 +261,7 @@ impl<
 
 impl<
     'a,
-    K: 'a + Copy + Clone + PartialEq + Eq + Hash + Serialize + for<'de> Deserialize<'de> + Send + Sync,
+    K: 'a + Clone + PartialEq + Eq + Hash + Serialize + for<'de> Deserialize<'de> + Send + Sync,
     V: 'a + Clone + Serialize + for<'de> Deserialize<'de> + Send + Sync,
 > MapRead<'a, K, V> for MemoryMap<K, V>
 {
@@ -383,7 +383,7 @@ impl<
 }
 
 impl<
-    K: Copy + Clone + PartialEq + Eq + Hash + Serialize + for<'de> Deserialize<'de> + Send + Sync,
+    K: Clone + PartialEq + Eq + Hash + Serialize + for<'de> Deserialize<'de> + Send + Sync,
     V: Clone + Serialize + for<'de> Deserialize<'de> + Send + Sync,
 > Deref for MemoryMap<K, V>
 {
diff --git a/ledger/store/src/helpers/memory/program.rs b/ledger/store/src/helpers/memory/program.rs
index 528399c22..4ca058e9e 100644
--- a/ledger/store/src/helpers/memory/program.rs
+++ b/ledger/store/src/helpers/memory/program.rs
@@ -29,6 +29,8 @@ use snarkvm_ledger_committee::Committee;
 
 use aleo_std_storage::StorageMode;
 use indexmap::IndexSet;
+#[cfg(feature = "history")]
+use std::sync::{Arc, atomic::AtomicU32};
 
 /// An in-memory finalize storage.
 #[derive(Clone)]
@@ -39,6 +41,15 @@ pub struct FinalizeMemory<N: Network> {
     program_id_map: MemoryMap<ProgramID<N>, IndexSet<Identifier<N>>>,
     /// The key-value map.
     key_value_map: NestedMemoryMap<(ProgramID<N>, Identifier<N>), Plaintext<N>, Value<N>>,
+    /// The historical mapping value update map.
+    #[cfg(feature = "history")]
+    mapping_update_map: MemoryMap<(ProgramID<N>, Identifier<N>, Plaintext<N>, u32), Value<N>>,
+    /// The historical mapping update heights map.
+    #[cfg(feature = "history")]
+    mapping_update_heights_map: MemoryMap<(ProgramID<N>, Identifier<N>, Plaintext<N>), Vec<u32>>,
+    /// The current block height.
+    #[cfg(feature = "history")]
+    block_height: Arc<AtomicU32>,
     /// The storage mode.
     storage_mode: StorageMode,
 }
@@ -48,6 +59,10 @@ impl<N: Network> FinalizeStorage<N> for FinalizeMemory<N> {
     type CommitteeStorage = CommitteeMemory<N>;
     type ProgramIDMap = MemoryMap<ProgramID<N>, IndexSet<Identifier<N>>>;
     type KeyValueMap = NestedMemoryMap<(ProgramID<N>, Identifier<N>), Plaintext<N>, Value<N>>;
+    #[cfg(feature = "history")]
+    type MappingUpdateMap = MemoryMap<(ProgramID<N>, Identifier<N>, Plaintext<N>, u32), Value<N>>;
+    #[cfg(feature = "history")]
+    type MappingUpdateHeightsMap = MemoryMap<(ProgramID<N>, Identifier<N>, Plaintext<N>), Vec<u32>>;
 
     /// Initializes the finalize storage.
     fn open<S: Into<StorageMode>>(storage: S) -> Result<Self> {
@@ -59,6 +74,12 @@ impl<N: Network> FinalizeStorage<N> for FinalizeMemory<N> {
             committee_store,
             program_id_map: MemoryMap::default(),
             key_value_map: NestedMemoryMap::default(),
+            #[cfg(feature = "history")]
+            mapping_update_map: MemoryMap::default(),
+            #[cfg(feature = "history")]
+            mapping_update_heights_map: MemoryMap::default(),
+            #[cfg(feature = "history")]
+            block_height: Default::default(),
             storage_mode: storage,
         })
     }
@@ -78,10 +99,27 @@ impl<N: Network> FinalizeStorage<N> for FinalizeMemory<N> {
         &self.key_value_map
     }
 
+    /// Returns the historical value map.
+    #[cfg(feature = "history")]
+    fn mapping_update_map(&self) -> &Self::MappingUpdateMap {
+        &self.mapping_update_map
+    }
+
+    #[cfg(feature = "history")]
+    fn mapping_update_heights_map(&self) -> &Self::MappingUpdateHeightsMap {
+        &self.mapping_update_heights_map
+    }
+
     /// Returns the storage mode.
     fn storage_mode(&self) -> &StorageMode {
         &self.storage_mode
     }
+
+    /// Returns the current block height.
+    #[cfg(feature = "history")]
+    fn current_block_height(&self) -> &AtomicU32 {
+        &self.block_height
+    }
 }
 
 /// An in-memory committee storage.
diff --git a/ledger/store/src/helpers/rocksdb/internal/id.rs b/ledger/store/src/helpers/rocksdb/internal/id.rs
index 4ef3f136e..41b76570f 100644
--- a/ledger/store/src/helpers/rocksdb/internal/id.rs
+++ b/ledger/store/src/helpers/rocksdb/internal/id.rs
@@ -202,6 +202,8 @@ pub enum TransitionMap {
 pub enum ProgramMap {
     ProgramID = DataID::ProgramIDMap as u16,
     KeyValueID = DataID::KeyValueMap as u16,
+    MappingUpdate = DataID::MappingUpdateMap as u16,
+    MappingUpdateHeights = DataID::MappingUpdateHeightsMap as u16,
 }
 
 /// The RocksDB map prefix for test-related entries.
@@ -303,6 +305,10 @@ enum DataID {
     IDEditionMap,
     // Track deployments that contain an optional checksum
     DeploymentChecksumMap,
+    // Historical mappings.
+    MappingUpdateMap,
+    // Historical mapping update heights.
+    MappingUpdateHeightsMap,
 
     // Testing
     #[cfg(test)]
diff --git a/ledger/store/src/helpers/rocksdb/internal/map.rs b/ledger/store/src/helpers/rocksdb/internal/map.rs
index 5f0636d7e..b303cf86e 100644
--- a/ledger/store/src/helpers/rocksdb/internal/map.rs
+++ b/ledger/store/src/helpers/rocksdb/internal/map.rs
@@ -59,7 +59,7 @@ impl<K: Serialize + DeserializeOwned, V: Serialize + DeserializeOwned> InnerData
 
 impl<
     'a,
-    K: 'a + Copy + Clone + Debug + PartialEq + Eq + Hash + Serialize + DeserializeOwned + Send + Sync,
+    K: 'a + Clone + Debug + PartialEq + Eq + Hash + Serialize + DeserializeOwned + Send + Sync,
     V: 'a + Clone + Serialize + DeserializeOwned + Send + Sync,
 > Map<'a, K, V> for DataMap<K, V>
 {
@@ -93,7 +93,7 @@ impl<
         match self.is_atomic_in_progress() {
             // If a batch is in progress, add the key to the batch.
             true => {
-                self.atomic_batch.lock().push((*key, None));
+                self.atomic_batch.lock().push((key.clone(), None));
             }
             // Otherwise, remove the key-value pair directly from the map.
             false => {
@@ -278,7 +278,7 @@ impl<
 
 impl<
     'a,
-    K: 'a + Copy + Clone + Debug + PartialEq + Eq + Hash + Serialize + DeserializeOwned + Send + Sync,
+    K: 'a + Clone + Debug + PartialEq + Eq + Hash + Serialize + DeserializeOwned + Send + Sync,
     V: 'a + Clone + Serialize + DeserializeOwned + Send + Sync,
 > MapRead<'a, K, V> for DataMap<K, V>
 {
diff --git a/ledger/store/src/helpers/rocksdb/program.rs b/ledger/store/src/helpers/rocksdb/program.rs
index b54536c6d..0f31aaede 100644
--- a/ledger/store/src/helpers/rocksdb/program.rs
+++ b/ledger/store/src/helpers/rocksdb/program.rs
@@ -29,6 +29,8 @@ use snarkvm_ledger_committee::Committee;
 
 use aleo_std_storage::StorageMode;
 use indexmap::IndexSet;
+#[cfg(feature = "history")]
+use std::sync::{Arc, atomic::AtomicU32};
 
 /// A RocksDB finalize storage.
 #[derive(Clone)]
@@ -39,6 +41,15 @@ pub struct FinalizeDB<N: Network> {
     program_id_map: DataMap<ProgramID<N>, IndexSet<Identifier<N>>>,
     /// The key-value map.
     key_value_map: NestedDataMap<(ProgramID<N>, Identifier<N>), Plaintext<N>, Value<N>>,
+    /// The historical mapping map.
+    #[cfg(feature = "history")]
+    mapping_update_map: DataMap<(ProgramID<N>, Identifier<N>, Plaintext<N>, u32), Value<N>>,
+    /// The historical mapping update heights map.
+    #[cfg(feature = "history")]
+    mapping_update_heights_map: DataMap<(ProgramID<N>, Identifier<N>, Plaintext<N>), Vec<u32>>,
+    /// The current block height.
+    #[cfg(feature = "history")]
+    block_height: Arc<AtomicU32>,
     /// The storage mode.
     storage_mode: StorageMode,
 }
@@ -48,6 +59,10 @@ impl<N: Network> FinalizeStorage<N> for FinalizeDB<N> {
     type CommitteeStorage = CommitteeDB<N>;
     type ProgramIDMap = DataMap<ProgramID<N>, IndexSet<Identifier<N>>>;
     type KeyValueMap = NestedDataMap<(ProgramID<N>, Identifier<N>), Plaintext<N>, Value<N>>;
+    #[cfg(feature = "history")]
+    type MappingUpdateMap = DataMap<(ProgramID<N>, Identifier<N>, Plaintext<N>, u32), Value<N>>;
+    #[cfg(feature = "history")]
+    type MappingUpdateHeightsMap = DataMap<(ProgramID<N>, Identifier<N>, Plaintext<N>), Vec<u32>>;
 
     /// Initializes the finalize storage.
     fn open<S: Into<StorageMode>>(storage: S) -> Result<Self> {
@@ -59,6 +74,12 @@ impl<N: Network> FinalizeStorage<N> for FinalizeDB<N> {
             committee_store,
             program_id_map: rocksdb::RocksDB::open_map(N::ID, storage.clone(), MapID::Program(ProgramMap::ProgramID))?,
             key_value_map: rocksdb::RocksDB::open_nested_map(N::ID, storage.clone(), MapID::Program(ProgramMap::KeyValueID))?,
+            #[cfg(feature = "history")]
+            mapping_update_map: rocksdb::RocksDB::open_map(N::ID, storage.clone(), MapID::Program(ProgramMap::MappingUpdate))?,
+            #[cfg(feature = "history")]
+            mapping_update_heights_map: rocksdb::RocksDB::open_map(N::ID, storage.clone(), MapID::Program(ProgramMap::MappingUpdateHeights))?,
+            #[cfg(feature = "history")]
+            block_height: Default::default(),
             storage_mode: storage,
         })
     }
@@ -78,10 +99,27 @@ impl<N: Network> FinalizeStorage<N> for FinalizeDB<N> {
         &self.key_value_map
     }
 
+    /// Returns the historical value map.
+    #[cfg(feature = "history")]
+    fn mapping_update_map(&self) -> &Self::MappingUpdateMap {
+        &self.mapping_update_map
+    }
+
+    #[cfg(feature = "history")]
+    fn mapping_update_heights_map(&self) -> &Self::MappingUpdateHeightsMap {
+        &self.mapping_update_heights_map
+    }
+
     /// Returns the storage mode.
     fn storage_mode(&self) -> &StorageMode {
         &self.storage_mode
     }
+
+    /// Returns the current block height.
+    #[cfg(feature = "history")]
+    fn current_block_height(&self) -> &AtomicU32 {
+        &self.block_height
+    }
 }
 
 /// A RocksDB committee storage.
diff --git a/ledger/store/src/helpers/traits/map.rs b/ledger/store/src/helpers/traits/map.rs
index 210a5ac12..06aef2120 100644
--- a/ledger/store/src/helpers/traits/map.rs
+++ b/ledger/store/src/helpers/traits/map.rs
@@ -21,7 +21,7 @@ use std::borrow::Cow;
 /// A trait representing map-like storage operations with read-write capabilities.
 pub trait Map<
     'a,
-    K: 'a + Copy + Clone + PartialEq + Eq + Hash + Serialize + Deserialize<'a> + Send + Sync,
+    K: 'a + Clone + PartialEq + Eq + Hash + Serialize + Deserialize<'a> + Send + Sync,
     V: 'a + Clone + Serialize + Deserialize<'a> + Send + Sync,
 >: Clone + MapRead<'a, K, V> + Send + Sync
 {
@@ -92,7 +92,7 @@ pub trait Map<
 /// A trait representing map-like storage operations with read-only capabilities.
 pub trait MapRead<
     'a,
-    K: 'a + Copy + Clone + PartialEq + Eq + Hash + Serialize + Deserialize<'a> + Sync,
+    K: 'a + Clone + PartialEq + Eq + Hash + Serialize + Deserialize<'a> + Sync,
     V: 'a + Clone + Serialize + Deserialize<'a> + Sync,
 >
 {
diff --git a/ledger/store/src/program/finalize.rs b/ledger/store/src/program/finalize.rs
index 1fc2f1316..767624e6f 100644
--- a/ledger/store/src/program/finalize.rs
+++ b/ledger/store/src/program/finalize.rs
@@ -29,6 +29,11 @@ use aleo_std_storage::StorageMode;
 use anyhow::Result;
 use core::marker::PhantomData;
 use indexmap::IndexSet;
+#[cfg(feature = "history")]
+use std::{
+    borrow::Cow,
+    sync::atomic::{AtomicU32, Ordering},
+};
 
 /// TODO (howardwu): Remove this.
 /// Returns the mapping ID for the given `program ID` and `mapping name`.
@@ -76,6 +81,12 @@ pub trait FinalizeStorage<N: Network>: 'static + Clone + Send + Sync {
     type ProgramIDMap: for<'a> Map<'a, ProgramID<N>, IndexSet<Identifier<N>>>;
     /// The mapping of `(program ID, mapping name)` to `[(key, value)]`.
     type KeyValueMap: for<'a> NestedMap<'a, (ProgramID<N>, Identifier<N>), Plaintext<N>, Value<N>>;
+    /// The mapping of `(program ID, mapping name, key, height)` to `value`.
+    #[cfg(feature = "history")]
+    type MappingUpdateMap: for<'a> Map<'a, (ProgramID<N>, Identifier<N>, Plaintext<N>, u32), Value<N>>;
+    /// The mapping of `(program ID, mapping name, key)` to [`height`].
+    #[cfg(feature = "history")]
+    type MappingUpdateHeightsMap: for<'a> Map<'a, (ProgramID<N>, Identifier<N>, Plaintext<N>), Vec<u32>>;
 
     /// Initializes the program state storage.
     fn open<S: Into<StorageMode>>(storage: S) -> Result<Self>;
@@ -86,6 +97,12 @@ pub trait FinalizeStorage<N: Network>: 'static + Clone + Send + Sync {
     fn program_id_map(&self) -> &Self::ProgramIDMap;
     /// Returns the key-value map.
     fn key_value_map(&self) -> &Self::KeyValueMap;
+    /// Returns the historical mapping value map.
+    #[cfg(feature = "history")]
+    fn mapping_update_map(&self) -> &Self::MappingUpdateMap;
+    /// Returns the historical mapping update heights map.
+    #[cfg(feature = "history")]
+    fn mapping_update_heights_map(&self) -> &Self::MappingUpdateHeightsMap;
 
     /// Returns the storage mode.
     fn storage_mode(&self) -> &StorageMode;
@@ -95,13 +112,25 @@ pub trait FinalizeStorage<N: Network>: 'static + Clone + Send + Sync {
         self.committee_store().start_atomic();
         self.program_id_map().start_atomic();
         self.key_value_map().start_atomic();
+        #[cfg(feature = "history")]
+        {
+            self.mapping_update_map().start_atomic();
+            self.mapping_update_heights_map().start_atomic();
+        }
     }
 
     /// Checks if an atomic batch is in progress.
     fn is_atomic_in_progress(&self) -> bool {
-        self.committee_store().is_atomic_in_progress()
+        let ret = self.committee_store().is_atomic_in_progress()
             || self.program_id_map().is_atomic_in_progress()
-            || self.key_value_map().is_atomic_in_progress()
+            || self.key_value_map().is_atomic_in_progress();
+        #[cfg(feature = "history")]
+        let ret = {
+            ret || self.mapping_update_map().is_atomic_in_progress()
+                || self.mapping_update_heights_map().is_atomic_in_progress()
+        };
+
+        ret
     }
 
     /// Checkpoints the atomic batch.
@@ -109,6 +138,11 @@ pub trait FinalizeStorage<N: Network>: 'static + Clone + Send + Sync {
         self.committee_store().atomic_checkpoint();
         self.program_id_map().atomic_checkpoint();
         self.key_value_map().atomic_checkpoint();
+        #[cfg(feature = "history")]
+        {
+            self.mapping_update_map().atomic_checkpoint();
+            self.mapping_update_heights_map().atomic_checkpoint();
+        }
     }
 
     /// Clears the latest atomic batch checkpoint.
@@ -116,6 +150,11 @@ pub trait FinalizeStorage<N: Network>: 'static + Clone + Send + Sync {
         self.committee_store().clear_latest_checkpoint();
         self.program_id_map().clear_latest_checkpoint();
         self.key_value_map().clear_latest_checkpoint();
+        #[cfg(feature = "history")]
+        {
+            self.mapping_update_map().clear_latest_checkpoint();
+            self.mapping_update_heights_map().clear_latest_checkpoint();
+        }
     }
 
     /// Rewinds the atomic batch to the previous checkpoint.
@@ -123,6 +162,11 @@ pub trait FinalizeStorage<N: Network>: 'static + Clone + Send + Sync {
         self.committee_store().atomic_rewind();
         self.program_id_map().atomic_rewind();
         self.key_value_map().atomic_rewind();
+        #[cfg(feature = "history")]
+        {
+            self.mapping_update_map().atomic_rewind();
+            self.mapping_update_heights_map().atomic_rewind();
+        }
     }
 
     /// Aborts an atomic batch write operation.
@@ -130,15 +174,30 @@ pub trait FinalizeStorage<N: Network>: 'static + Clone + Send + Sync {
         self.committee_store().abort_atomic();
         self.program_id_map().abort_atomic();
         self.key_value_map().abort_atomic();
+        #[cfg(feature = "history")]
+        {
+            self.mapping_update_map().abort_atomic();
+            self.mapping_update_heights_map().abort_atomic();
+        }
     }
 
     /// Finishes an atomic batch write operation.
     fn finish_atomic(&self) -> Result<()> {
         self.committee_store().finish_atomic()?;
         self.program_id_map().finish_atomic()?;
-        self.key_value_map().finish_atomic()
+        self.key_value_map().finish_atomic()?;
+        #[cfg(feature = "history")]
+        {
+            self.mapping_update_map().finish_atomic()?;
+            self.mapping_update_heights_map().finish_atomic()?;
+        }
+        Ok(())
     }
 
+    /// Returns the current block height.
+    #[cfg(feature = "history")]
+    fn current_block_height(&self) -> &AtomicU32;
+
     /// Initializes the given `program ID` and `mapping name` in storage.
     /// If the `mapping name` is already initialized, an error is returned.
     fn initialize_mapping(
@@ -196,6 +255,27 @@ pub trait FinalizeStorage<N: Network>: 'static + Clone + Send + Sync {
         let value_id = N::hash_bhp1024(&(key_id, N::hash_bhp1024(&value.to_bits_le())?).to_bits_le())?;
 
         atomic_batch_scope!(self, {
+            // Update the historical maps.
+            #[cfg(feature = "history")]
+            {
+                let current_height = self.current_block_height().load(Ordering::SeqCst);
+
+                // Insert the initial value as the first historical update.
+                self.mapping_update_map()
+                    .insert((program_id, mapping_name, key.clone(), current_height), value.clone())?;
+
+                // Ensure the update height history does not already exist.
+                let height_update_key = (program_id, mapping_name, key.clone());
+                if self.mapping_update_heights_map().contains_key_speculative(&height_update_key)? {
+                    bail!(
+                        "Illegal operation: the history of height updates for '{program_id}/{mapping_name}/{key}' already exists in storage"
+                    );
+                }
+
+                // Insert the first historical update height.
+                self.mapping_update_heights_map().insert(height_update_key, vec![current_height])?;
+            }
+
             // Update the key-value map with the new key-value.
             self.key_value_map().insert((program_id, mapping_name), key, value)?;
 
@@ -228,6 +308,25 @@ pub trait FinalizeStorage<N: Network>: 'static + Clone + Send + Sync {
         let value_id = N::hash_bhp1024(&(key_id, N::hash_bhp1024(&value.to_bits_le())?).to_bits_le())?;
 
         atomic_batch_scope!(self, {
+            // Update the historical maps.
+            #[cfg(feature = "history")]
+            {
+                let current_height = self.current_block_height().load(Ordering::SeqCst);
+
+                // Register the updated value at the current height.
+                self.mapping_update_map()
+                    .insert((program_id, mapping_name, key.clone(), current_height), value.clone())?;
+
+                // Obtain the list of past update heights.
+                let key = (program_id, mapping_name, key.clone());
+                let update_heights =
+                    self.mapping_update_heights_map().get_confirmed(&key)?.map(|list| list.into_owned());
+                let mut update_heights = update_heights.unwrap_or_default();
+                // Extend the historical update heights with the current height.
+                update_heights.push(current_height);
+                self.mapping_update_heights_map().insert(key, update_heights)?;
+            }
+
             // Update the key-value map with the new key-value.
             self.key_value_map().insert((program_id, mapping_name), key, value)?;
 
@@ -568,6 +667,51 @@ impl<N: Network, P: FinalizeStorage<N>> FinalizeStore<N, P> {
     pub fn storage_mode(&self) -> &StorageMode {
         self.storage.storage_mode()
     }
+
+    /// Returns the current block height.
+    #[cfg(feature = "history")]
+    pub fn current_block_height(&self) -> &AtomicU32 {
+        self.storage.current_block_height()
+    }
+
+    /// Returns the historical value of a mapping.
+    #[cfg(feature = "history")]
+    pub fn get_historical_mapping_value(
+        &self,
+        program_id: ProgramID<N>,
+        mapping_name: Identifier<N>,
+        mapping_key: Plaintext<N>,
+        height: u32,
+    ) -> Result<Option<Cow<'_, Value<N>>>, Error> {
+        // First, obtain the heights at which updates have happened.
+        let Some(update_heights) = self.get_mapping_update_heights(program_id, mapping_name, mapping_key.clone())?
+        else {
+            return Ok(None);
+        };
+
+        // Find the updated height which matches the desired point in history.
+        let Some(applicable_height) = (match update_heights.binary_search(&height) {
+            Ok(_) => Some(height),
+            Err(0) => None,
+            Err(idx) => update_heights.get(idx - 1).copied(),
+        }) else {
+            return Ok(None);
+        };
+
+        // Find the mapping value applicable at that height.
+        self.storage.mapping_update_map().get_confirmed(&(program_id, mapping_name, mapping_key, applicable_height))
+    }
+
+    /// Returns the heights at which past mapping updates occurred.
+    #[cfg(feature = "history")]
+    pub fn get_mapping_update_heights(
+        &self,
+        program_id: ProgramID<N>,
+        mapping_name: Identifier<N>,
+        mapping_key: Plaintext<N>,
+    ) -> Result<Option<Cow<'_, Vec<u32>>>, Error> {
+        self.storage.mapping_update_heights_map().get_confirmed(&(program_id, mapping_name, mapping_key))
+    }
 }
 
 impl<N: Network, P: FinalizeStorage<N>> FinalizeStore<N, P> {
diff --git a/synthesizer/Cargo.toml b/synthesizer/Cargo.toml
index edf2553c4..18c237138 100644
--- a/synthesizer/Cargo.toml
+++ b/synthesizer/Cargo.toml
@@ -33,7 +33,7 @@ locktick = [
 ]
 async = [ "snarkvm-ledger-query/async", "snarkvm-synthesizer-process/async" ]
 cuda = [ "snarkvm-algorithms/cuda" ]
-history = [ "serde" ]
+history = [ "snarkvm-ledger-store/history", "serde" ]
 rocks = [ "snarkvm-ledger-store/rocks" ]
 serial = [
   "snarkvm-console/serial",
diff --git a/synthesizer/src/vm/finalize.rs b/synthesizer/src/vm/finalize.rs
index c111ccb48..06b92043e 100644
--- a/synthesizer/src/vm/finalize.rs
+++ b/synthesizer/src/vm/finalize.rs
@@ -154,8 +154,12 @@ impl<N: Network, C: ConsensusStorage<N>> VM<N, C> {
             bail!("The ratifications after speculation do not match the ratifications in the block");
         }
         // Ensure the transactions after speculation match.
-        if transactions != &confirmed_transactions.into_iter().collect() {
-            bail!("The transactions after speculation do not match the transactions in the block");
+        let confirmed_transactions = confirmed_transactions.into_iter().collect();
+        if transactions != &confirmed_transactions {
+            let confirmed_transaction_ids = confirmed_transactions.transaction_ids().collect::<Vec<_>>();
+            bail!(
+                "The transactions after speculation do not match the transactions in the block. IDs: {confirmed_transaction_ids:?} - Transactions:{transactions:?}"
+            );
         }
         // Ensure there are no aborted transaction IDs from this speculation.
         // Note: There should be no aborted transactions, because we are checking a block,
@@ -275,6 +279,13 @@ impl<N: Network, C: ConsensusStorage<N>> VM<N, C> {
         // Determine the maximum number of aborted transactions allowed in a block.
         let max_aborted_transactions = Transactions::<N>::max_aborted_transactions();
 
+        // Update the block height used for the purposes of historical mapping accounting.
+        #[cfg(feature = "history")]
+        self.store
+            .finalize_store()
+            .current_block_height()
+            .store(state.block_height(), std::sync::atomic::Ordering::SeqCst);
+
         // Perform the finalize operation on the preset finalize mode.
         atomic_finalize!(self.finalize_store(), FinalizeMode::DryRun, {
             // Ensure the number of solutions does not exceed the maximum.
@@ -629,6 +640,13 @@ impl<N: Network, C: ConsensusStorage<N>> VM<N, C> {
 
         let timer = timer!("VM::atomic_finalize");
 
+        // Update the block height used for the purposes of historical mapping accounting.
+        #[cfg(feature = "history")]
+        self.store
+            .finalize_store()
+            .current_block_height()
+            .store(state.block_height(), std::sync::atomic::Ordering::SeqCst);
+
         // Perform the finalize operation on the preset finalize mode.
         atomic_finalize!(self.finalize_store(), FinalizeMode::RealRun, {
             // Initialize an iterator for ratifications before finalize.
@@ -1391,36 +1409,6 @@ impl<N: Network, C: ConsensusStorage<N>> VM<N, C> {
                     // Insert the next committee into storage.
                     store.committee_store().insert(state.block_height(), next_committee)?;
 
-                    #[cfg(all(feature = "history", feature = "rocks"))]
-                    {
-                        // When finalizing in `FinalizeMode::RealRun`, store the delegated and bonded mappings in history.
-                        if IS_FINALIZE {
-                            // Load a `History` object.
-                            let history = History::new(N::ID, store.storage_mode());
-
-                            // Write the delegated mapping as JSON.
-                            history.store_mapping(state.block_height(), MappingName::Delegated, &next_delegated_map)?;
-
-                            // Write the bonded mapping as JSON.
-                            history.store_mapping(state.block_height(), MappingName::Bonded, &next_bonded_map)?;
-
-                            // Write the metadata mapping as JSON.
-                            let metadata_mapping = Identifier::from_str("metadata")?;
-                            let metadata_map = store.get_mapping_speculative(program_id, metadata_mapping)?;
-                            history.store_mapping(state.block_height(), MappingName::Metadata, &metadata_map)?;
-
-                            // Write the unbonding mapping as JSON.
-                            let unbonding_mapping = Identifier::from_str("unbonding")?;
-                            let unbonding_map = store.get_mapping_speculative(program_id, unbonding_mapping)?;
-                            history.store_mapping(state.block_height(), MappingName::Unbonding, &unbonding_map)?;
-
-                            // Write the withdraw mapping as JSON.
-                            let withdraw_mapping = Identifier::from_str("withdraw")?;
-                            let withdraw_map = store.get_mapping_speculative(program_id, withdraw_mapping)?;
-                            history.store_mapping(state.block_height(), MappingName::Withdraw, &withdraw_map)?;
-                        }
-                    }
-
                     // Store the finalize operations for updating the committee and bonded mapping.
                     finalize_operations.extend(&[
                         // Replace the committee mapping in storage.
diff --git a/synthesizer/src/vm/helpers/history.rs b/synthesizer/src/vm/helpers/history.rs
deleted file mode 100644
index 9f3d4a9a6..000000000
--- a/synthesizer/src/vm/helpers/history.rs
+++ /dev/null
@@ -1,131 +0,0 @@
-// Copyright (c) 2019-2026 Provable Inc.
-// This file is part of the snarkVM library.
-
-// Licensed under the Apache License, Version 2.0 (the "License");
-// you may not use this file except in compliance with the License.
-// You may obtain a copy of the License at:
-
-// http://www.apache.org/licenses/LICENSE-2.0
-
-// Unless required by applicable law or agreed to in writing, software
-// distributed under the License is distributed on an "AS IS" BASIS,
-// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
-// See the License for the specific language governing permissions and
-// limitations under the License.
-
-use serde::{Deserialize, Serialize};
-
-use aleo_std::{StorageMode, aleo_ledger_dir};
-
-use anyhow::Result;
-use serde_json;
-use std::{
-    fmt::{Display, Formatter},
-    path::PathBuf,
-};
-
-/// Returns the path where a `history` directory may be stored.
-pub fn history_directory_path(network: u16, storage_mode: &StorageMode) -> PathBuf {
-    const HISTORY_DIRECTORY_NAME: &str = "history";
-
-    // Create the name of the history directory.
-    let directory_name = match &storage_mode {
-        StorageMode::Development(id) => format!(".{HISTORY_DIRECTORY_NAME}-{network}-{id}"),
-        StorageMode::Production | StorageMode::Custom(_) => format!("{HISTORY_DIRECTORY_NAME}-{network}"),
-        StorageMode::Test(_) => unimplemented!(),
-    };
-
-    // Obtain the path to the ledger.
-    let mut path = aleo_ledger_dir(network, storage_mode);
-    // Go to the folder right above the ledger.
-    path.pop();
-    // Append the history directory's name.
-    path.push(directory_name);
-
-    path
-}
-
-#[derive(Copy, Clone, Serialize, Deserialize)]
-#[serde(rename_all = "lowercase")]
-pub enum MappingName {
-    /// The `bonded` mapping.
-    Bonded,
-    /// The `delegated` mapping.
-    Delegated,
-    /// The `metadata` mapping.
-    Metadata,
-    /// The `unbonding` mapping.
-    Unbonding,
-    /// The `withdraw` mapping.
-    Withdraw,
-}
-
-impl Display for MappingName {
-    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
-        match self {
-            Self::Bonded => write!(f, "bonded"),
-            Self::Delegated => write!(f, "delegated"),
-            Self::Metadata => write!(f, "metadata"),
-            Self::Unbonding => write!(f, "unbonding"),
-            Self::Withdraw => write!(f, "withdraw"),
-        }
-    }
-}
-
-pub struct History {
-    /// The path to the history directory.
-    path: PathBuf,
-}
-
-impl History {
-    /// Initializes a new instance of `History`.
-    pub fn new(network: u16, storage_mode: &StorageMode) -> Self {
-        Self { path: history_directory_path(network, storage_mode) }
-    }
-
-    /// Stores a mapping from a given block in the history directory as JSON.
-    pub fn store_mapping<T>(&self, height: u32, mapping: MappingName, data: &T) -> Result<()>
-    where
-        T: Serialize + ?Sized,
-    {
-        // Get the path to the block directory.
-        let path = self.block_path(height);
-        // Create the block directory if it does not exist.
-        if !path.exists() {
-            std::fs::create_dir_all(&path)?;
-        }
-
-        // Write the entry to the block directory.
-        let path = path.join(format!("block-{height}-{mapping}.json"));
-        std::fs::write(path, serde_json::to_string_pretty(data)?)?;
-
-        Ok(())
-    }
-
-    /// Loads the JSON string for a mapping from a given block from the history directory.
-    pub fn load_mapping(&self, height: u32, mapping: MappingName) -> Result<String> {
-        // Get the path to the block directory.
-        let path = self.block_path(height);
-        // Get the path to the block file.
-        let path = path.join(format!("block-{height}-{mapping}.json"));
-
-        // Read the file.
-        let data = std::fs::read_to_string(path)?;
-
-        Ok(data)
-    }
-
-    // A helper function to get the path to the block directory.
-    fn block_path(&self, height: u32) -> PathBuf {
-        // Get the path the directory group.
-        let group = Self::group(height);
-        let path = self.path.join(format!("group-{group}"));
-        // Get the path to the block directory.
-        path.join(format!("block-{height}"))
-    }
-
-    // A helper function to calculate the group number for a given block height.
-    fn group(height: u32) -> u32 {
-        height.saturating_div(u16::MAX as u32)
-    }
-}
diff --git a/synthesizer/src/vm/helpers/mod.rs b/synthesizer/src/vm/helpers/mod.rs
index 8f2b3e470..59628e9e6 100644
--- a/synthesizer/src/vm/helpers/mod.rs
+++ b/synthesizer/src/vm/helpers/mod.rs
@@ -16,11 +16,6 @@
 pub(crate) mod committee;
 pub use committee::*;
 
-#[cfg(feature = "history")]
-mod history;
-#[cfg(feature = "history")]
-pub use history::*;
-
 mod macros;
 
 mod program;
diff --git a/synthesizer/src/vm/helpers/sequential_op.rs b/synthesizer/src/vm/helpers/sequential_op.rs
index 1e359f75d..a75bd32a9 100644
--- a/synthesizer/src/vm/helpers/sequential_op.rs
+++ b/synthesizer/src/vm/helpers/sequential_op.rs
@@ -32,7 +32,7 @@ impl<N: Network, C: ConsensusStorage<N>> VM<N, C> {
             // Sequentially process incoming operations.
             while let Ok(request) = request_rx.recv() {
                 let SequentialOperationRequest { op, response_tx } = request;
-                debug!("Sequentially processing operation '{op}'");
+                trace!("Sequentially processing operation '{op}'");
 
                 // Perform the queued operation.
                 let ret = match op {
@@ -54,7 +54,7 @@ impl<N: Network, C: ConsensusStorage<N>> VM<N, C> {
 
     /// Sends the given operation to the thread used for sequential processing.
     pub fn run_sequential_operation(&self, op: SequentialOperation<N>) -> Option<SequentialOperationResult<N>> {
-        debug!("Queuing operation '{op}' for sequential processing");
+        trace!("Queuing operation '{op}' for sequential processing");
 
         // Prepare a oneshot channel to obtain the result of the queued operation.
         let (response_tx, response_rx) = oneshot::channel();
