diff --git a/node/bft/src/primary.rs b/node/bft/src/primary.rs
index f41ae6242..22a330eda 100644
--- a/node/bft/src/primary.rs
+++ b/node/bft/src/primary.rs
@@ -1697,9 +1697,9 @@ impl<N: Network> Primary<N> {
             }
         }
 
-        // If there are no missing certificates, return early.
+        // If there are no certificates to fetch, return early with the existing unprocessed certificates.
         match fetch_certificates.is_empty() {
-            true => return Ok(Default::default()),
+            true => return Ok(missing_certificates),
             false => trace!(
                 "Fetching {} missing certificates for round {round} from '{peer_ip}'...",
                 fetch_certificates.len(),
diff --git a/node/bft/src/sync/mod.rs b/node/bft/src/sync/mod.rs
index 62f7f551f..ca2d23850 100644
--- a/node/bft/src/sync/mod.rs
+++ b/node/bft/src/sync/mod.rs
@@ -628,6 +628,7 @@ impl<N: Network> Sync<N> {
             );
         }
         // Wait for the certificate to be fetched.
+        // TODO (raychu86): Consider making the timeout dynamic based on network traffic and/or the number of validators.
         match tokio::time::timeout(Duration::from_millis(MAX_FETCH_TIMEOUT_IN_MS), callback_receiver).await {
             // If the certificate was fetched, return it.
             Ok(result) => Ok(result?),
diff --git a/node/bft/src/worker.rs b/node/bft/src/worker.rs
index 2e8d7ad55..002d75c8e 100644
--- a/node/bft/src/worker.rs
+++ b/node/bft/src/worker.rs
@@ -305,6 +305,18 @@ impl<N: Network> Worker<N> {
             // All other combinations are clearly invalid.
             _ => false,
         };
+        // If the transmission is a deserialized execution, verify it immediately.
+        // This takes heavy transaction verification out of the hot path during block generation.
+        if let (TransmissionID::Transaction(tx_id, _), Transmission::Transaction(tx)) = (transmission_id, &transmission)
+        {
+            if let Data::Object(Transaction::Execute(..)) = tx {
+                let self_ = self.clone();
+                let tx_ = tx.clone();
+                tokio::spawn(async move {
+                    let _ = self_.ledger.check_transaction_basic(tx_id, tx_).await;
+                });
+            }
+        }
         // If the transmission ID and transmission type matches, then insert the transmission into the ready queue.
         if is_well_formed && self.ready.insert(transmission_id, transmission) {
             trace!(
diff --git a/node/sync/locators/src/block_locators.rs b/node/sync/locators/src/block_locators.rs
index ba8981cd6..de0b9fee1 100644
--- a/node/sync/locators/src/block_locators.rs
+++ b/node/sync/locators/src/block_locators.rs
@@ -27,6 +27,45 @@ const RECENT_INTERVAL: u32 = 1; // 1 block intervals
 /// The interval between block checkpoints.
 pub const CHECKPOINT_INTERVAL: u32 = 10_000; // 10,000 block intervals
 
+/// Block locator maps.
+///
+/// This data structure is used by validators to advertise the blocks that
+/// they have and can provide to other validators to help them sync.
+/// Periodically, each validator broadcasts a [`PrimaryPing`],
+/// which contains a `BlockLocators` instance.
+/// Recall that blocks are indexed by their `u32` height, starting with 0 for the genesis block.
+/// The keys of the `recents` and `checkpoints` maps are the block heights;
+/// the values of the maps are the corresponding block hashes.
+///
+/// If a validator has `N` blocks, the `recents` and `checkpoints` maps are as follows:
+/// - The `recents` map contains entries for blocks at heights
+///   `N - 1 - (NUM_RECENT_BLOCKS - 1) * RECENT_INTERVAL`,
+///   `N - 1 - (NUM_RECENT_BLOCKS - 2) * RECENT_INTERVAL`,
+///   ...,
+///   `N - 1`.
+///   If any of the just listed heights are negative, there are no entries for them of course,
+///   and the `recents` map has fewer than `NUM_RECENT_BLOCKS` entries.
+///   If `RECENT_INTERVAL` is 1, the `recents` map contains entries
+///   for the last `NUM_RECENT_BLOCKS` blocks, i.e. from `N - NUM_RECENT_BLOCKS` to `N - 1`;
+///   if additionally `N < NUM_RECENT_BLOCKS`, the `recents` map contains
+///   entries for all the blocks, from `0` to `N - 1`.
+/// - The `checkpoints` map contains an entry for every `CHECKPOINT_INTERVAL`-th block,
+///   starting with 0 and not exceeding `N`, i.e. it has entries for blocks
+///   `0`, `CHECKPOINT_INTERVAL`, `2 * CHECKPOINT_INTERVAL`, ..., `k * CHECKPOINT_INTERVAL`,
+///   where `k` is the maximum integer such that `k * CHECKPOINT_INTERVAL <= N`.
+///
+/// The `recents` and `checkpoints` maps may have overlapping entries,
+/// e.g. if `N-1` is a multiple of `CHECKPOINT_INTERVAL`;
+/// but if `CHECKPOINT_INTERVAL` is much larger than `NUM_RECENT_BLOCKS`,
+/// there is no overlap most of the time.
+///
+/// `BlockLocators` instances are built, in the form described above,
+/// by [`BlockSync::get_block_locators()`].
+/// When a `BlockLocators` instance is received (in a [`PrimaryPing`]) by a validator,
+/// the maps may not have the form described below; the deserializer does not enforce that.
+/// However, before incorporating the `BlockLocators` instance information into its state,
+/// the validator checks that the maps have the correct form,
+/// via [`BlockLocators::ensure_is_valid()`].
 #[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
 pub struct BlockLocators<N: Network> {
     /// The map of recent blocks.
