diff --git a/ledger/benches/transaction.rs b/ledger/benches/transaction.rs
index e83254914..030d06133 100644
--- a/ledger/benches/transaction.rs
+++ b/ledger/benches/transaction.rs
@@ -104,7 +104,7 @@ fn execute(c: &mut Criterion) {
     // Retrieve the execution ID.
     let execution_id = execute_authorization.to_execution_id().unwrap();
     // Authorize the fee.
-    let fee_authorization = vm.authorize_fee_public(&private_key, 100000, 1000, execution_id, rng).unwrap();
+    let fee_authorization = vm.authorize_fee_public(&private_key, 300000, 1000, execution_id, rng).unwrap();
 
     c.bench_function("Transaction::Execute(transfer_public)", |b| {
         b.iter(|| {
