diff --git a/synthesizer/src/vm/mod.rs b/synthesizer/src/vm/mod.rs
index f4ba9d980..c71e4c94e 100644
--- a/synthesizer/src/vm/mod.rs
+++ b/synthesizer/src/vm/mod.rs
@@ -182,7 +182,7 @@ impl<N: Network, C: ConsensusStorage<N>> VM<N, C> {
             atomic_lock: Arc::new(Mutex::new(())),
             block_lock: Arc::new(Mutex::new(())),
             partially_verified_transactions: Arc::new(RwLock::new(LruCache::new(
-                NonZeroUsize::new(Transactions::<console::network::Testnet3>::MAX_TRANSACTIONS).unwrap(),
+                NonZeroUsize::new(Transactions::<N>::MAX_TRANSACTIONS).unwrap(),
             ))),
         })
     }
