diff --git a/.resources/release-version b/.resources/release-version
index 7740e0d83..760fde71e 100644
--- a/.resources/release-version
+++ b/.resources/release-version
@@ -1 +1 @@
-v1.3.14
\ No newline at end of file
+v1.3.15
\ No newline at end of file
diff --git a/Cargo.lock b/Cargo.lock
index 47caba039..f5e142724 100644
--- a/Cargo.lock
+++ b/Cargo.lock
@@ -1391,9 +1391,9 @@ checksum = "830d08ce1d1d941e6b30645f1a0eb5643013d835ce3779a5fc208261dbe10f55"
 
 [[package]]
 name = "libc"
-version = "0.2.101"
+version = "0.2.102"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "3cb00336871be5ed2c8ed44b60ae9959dc5b9f08539422ed43f09e34ecaeba21"
+checksum = "a2a5ac8f984bfcf3a823267e5fde638acc3325f6496633a5da6bb6eb2171e103"
 
 [[package]]
 name = "libloading"
@@ -1797,7 +1797,7 @@ version = "0.5.4"
 source = "registry+https://github.com/rust-lang/crates.io-index"
 checksum = "486ea01961c4a818096de679a8b740b26d9033146ac5291b1c98557658f8cdd9"
 dependencies = [
- "proc-macro-crate 1.0.0",
+ "proc-macro-crate 1.1.0",
  "proc-macro2",
  "quote",
  "syn",
@@ -2027,9 +2027,9 @@ dependencies = [
 
 [[package]]
 name = "proc-macro-crate"
-version = "1.0.0"
+version = "1.1.0"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "41fdbd1df62156fbc5945f4762632564d7d038153091c3fcf1067f6aef7cff92"
+checksum = "1ebace6889caf889b4d3f76becee12e90353f2b8c7d875534a71e5742f8f6f83"
 dependencies = [
  "thiserror",
  "toml",
@@ -2544,9 +2544,9 @@ dependencies = [
 
 [[package]]
 name = "serde_json"
-version = "1.0.67"
+version = "1.0.68"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "a7f9e390c27c3c0ce8bc5d725f6e4d30a29d26659494aa4b17535f7522c5c950"
+checksum = "0f690853975602e1bfe1ccbf50504d67174e3bcf340f23b5ea9992e0587a52d8"
 dependencies = [
  "itoa",
  "ryu",
@@ -2642,7 +2642,7 @@ checksum = "fe0f37c9e8f3c5a4a66ad655a93c74daac4ad00c441533bf5c6e7990bb42604e"
 
 [[package]]
 name = "snarkos"
-version = "1.3.14"
+version = "1.3.15"
 dependencies = [
  "anyhow",
  "capnpc",
@@ -2673,7 +2673,7 @@ dependencies = [
 
 [[package]]
 name = "snarkos-benchmarks"
-version = "1.3.14"
+version = "1.3.15"
 dependencies = [
  "criterion",
  "csv",
@@ -2702,7 +2702,7 @@ dependencies = [
 
 [[package]]
 name = "snarkos-consensus"
-version = "1.3.14"
+version = "1.3.15"
 dependencies = [
  "anyhow",
  "chrono",
@@ -2734,7 +2734,7 @@ dependencies = [
 
 [[package]]
 name = "snarkos-metrics"
-version = "1.3.14"
+version = "1.3.15"
 dependencies = [
  "circular-queue",
  "metrics",
@@ -2748,7 +2748,7 @@ dependencies = [
 
 [[package]]
 name = "snarkos-network"
-version = "1.3.14"
+version = "1.3.15"
 dependencies = [
  "anyhow",
  "arc-swap",
@@ -2793,7 +2793,7 @@ dependencies = [
 
 [[package]]
 name = "snarkos-parameters"
-version = "1.3.14"
+version = "1.3.15"
 dependencies = [
  "chrono",
  "curl",
@@ -2812,14 +2812,14 @@ dependencies = [
 
 [[package]]
 name = "snarkos-profiler"
-version = "1.3.14"
+version = "1.3.15"
 dependencies = [
  "colored",
 ]
 
 [[package]]
 name = "snarkos-rpc"
-version = "1.3.14"
+version = "1.3.15"
 dependencies = [
  "anyhow",
  "async-trait",
@@ -2851,7 +2851,7 @@ dependencies = [
 
 [[package]]
 name = "snarkos-storage"
-version = "1.3.14"
+version = "1.3.15"
 dependencies = [
  "anyhow",
  "async-trait",
@@ -2883,7 +2883,7 @@ dependencies = [
 
 [[package]]
 name = "snarkos-testing"
-version = "1.3.14"
+version = "1.3.15"
 dependencies = [
  "anyhow",
  "bincode",
@@ -3760,9 +3760,9 @@ checksum = "85e60b0d1b5f99db2556934e21937020776a5d31520bf169e851ac44e6420214"
 
 [[package]]
 name = "x25519-dalek"
-version = "1.1.1"
+version = "1.2.0"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "5a0c105152107e3b96f6a00a65e86ce82d9b125230e1c4302940eca58ff71f4f"
+checksum = "2392b6b94a576b4e2bf3c5b2757d63f10ada8020a2e4d08ac849ebcf6ea8e077"
 dependencies = [
  "curve25519-dalek",
  "rand_core 0.5.1",
diff --git a/Cargo.toml b/Cargo.toml
index 03f28161b..cf7f9f50e 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkos"
-version = "1.3.14"
+version = "1.3.15"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "A decentralized operating system"
 homepage = "https://aleo.org"
@@ -66,20 +66,20 @@ version = "=0.7.9"
 
 [dependencies.snarkos-consensus]
 path = "./consensus"
-version = "1.3.14"
+version = "1.3.15"
 
 [dependencies.snarkos-network]
 path = "./network"
-version = "1.3.14"
+version = "1.3.15"
 
 [dependencies.snarkos-rpc]
 path = "./rpc"
-version = "1.3.14"
+version = "1.3.15"
 
 [dependencies.snarkos-storage]
 path = "./storage"
 features = [ "rocksdb_storage" ]
-version = "1.3.14"
+version = "1.3.15"
 
 [dependencies.anyhow]
 version = "1.0"
diff --git a/benchmarks/Cargo.toml b/benchmarks/Cargo.toml
index 8a28bd502..8bd057b02 100644
--- a/benchmarks/Cargo.toml
+++ b/benchmarks/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkos-benchmarks"
-version = "1.3.14"
+version = "1.3.15"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Benchmarks for a decentralized operating system"
 homepage = "https://aleo.org"
@@ -72,19 +72,19 @@ version = "=0.7.9"
 
 [dependencies.snarkos-profiler]
 path = "../profiler"
-version = "1.3.14"
+version = "1.3.15"
 
 [dependencies.snarkos-network]
 path = "../network"
-version = "1.3.14"
+version = "1.3.15"
 
 [dependencies.snarkos-storage]
 path = "../storage"
-version = "1.3.14"
+version = "1.3.15"
 
 [dependencies.snarkos-testing]
 path = "../testing"
-version = "1.3.14"
+version = "1.3.15"
 
 [dependencies.csv]
 version = "1"
diff --git a/consensus/Cargo.toml b/consensus/Cargo.toml
index 06d73cff6..f5a43ca4b 100644
--- a/consensus/Cargo.toml
+++ b/consensus/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkos-consensus"
-version = "1.3.14"
+version = "1.3.15"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Consensus for a decentralized operating system"
 homepage = "https://aleo.org"
@@ -18,8 +18,8 @@ license = "GPL-3.0"
 edition = "2018"
 
 [features]
-default = []
-test = []
+default = [ ]
+test = [ ]
 
 [dependencies.snarkvm-algorithms]
 version = "=0.7.9"
@@ -47,15 +47,15 @@ version = "1"
 
 [dependencies.snarkos-metrics]
 path = "../metrics"
-version = "1.3.14"
+version = "1.3.15"
 
 [dependencies.snarkos-profiler]
 path = "../profiler"
-version = "1.3.14"
+version = "1.3.15"
 
 [dependencies.snarkos-storage]
 path = "../storage"
-version = "1.3.14"
+version = "1.3.15"
 default-features = false
 
 [dependencies.anyhow]
diff --git a/consensus/src/consensus/inner/agent.rs b/consensus/src/consensus/inner/agent.rs
index 3862ce640..e3d2eeb23 100644
--- a/consensus/src/consensus/inner/agent.rs
+++ b/consensus/src/consensus/inner/agent.rs
@@ -14,6 +14,8 @@
 // You should have received a copy of the GNU General Public License
 // along with the snarkOS library. If not, see <https://www.gnu.org/licenses/>.
 
+use snarkos_metrics::wrapped_mpsc;
+
 use crate::consensus::ConsensusMessageWrapped;
 
 use super::*;
@@ -42,14 +44,12 @@ impl ConsensusInner {
         Ok(())
     }
 
-    pub(in crate::consensus) async fn agent(mut self, mut receiver: mpsc::Receiver<ConsensusMessageWrapped>) {
+    pub(in crate::consensus) async fn agent(mut self, mut receiver: wrapped_mpsc::Receiver<ConsensusMessageWrapped>) {
         self.init()
             .await
             .expect("failed to initialize ledger & storage with genesis block");
 
         while let Some((message, response)) = receiver.recv().await {
-            metrics::decrement_gauge!(snarkos_metrics::queues::CONSENSUS, 1.0);
-
             match message {
                 ConsensusMessage::ReceiveTransaction(transaction) => {
                     let ret = self.receive_transaction(transaction).await;
diff --git a/consensus/src/consensus/inner/mod.rs b/consensus/src/consensus/inner/mod.rs
index b33a8d315..0368f55f7 100644
--- a/consensus/src/consensus/inner/mod.rs
+++ b/consensus/src/consensus/inner/mod.rs
@@ -41,7 +41,6 @@ use snarkvm_dpc::{
 };
 use snarkvm_posw::txids_to_roots;
 use snarkvm_utilities::has_duplicates;
-use tokio::sync::mpsc;
 
 use snarkos_metrics::misc::*;
 
diff --git a/consensus/src/consensus/mod.rs b/consensus/src/consensus/mod.rs
index 1f678cfed..abc114578 100644
--- a/consensus/src/consensus/mod.rs
+++ b/consensus/src/consensus/mod.rs
@@ -17,6 +17,7 @@
 use std::{convert::TryInto, sync::Arc};
 
 use rand::{thread_rng, Rng};
+use snarkos_metrics::wrapped_mpsc;
 use snarkos_storage::{Address, Digest, DynStorage, SerialBlock, SerialRecord, SerialTransaction, VMRecord};
 use snarkvm_algorithms::CRH;
 use snarkvm_dpc::{
@@ -32,7 +33,7 @@ use snarkvm_dpc::{
     ProgramScheme,
 };
 use snarkvm_utilities::{to_bytes_le, ToBytes};
-use tokio::sync::{mpsc, oneshot};
+use tokio::sync::oneshot;
 
 use crate::{error::ConsensusError, ConsensusParameters, DynLedger, MemoryPool};
 
@@ -50,7 +51,7 @@ pub struct Consensus {
     pub dpc: Arc<Testnet1DPC>,
     pub storage: DynStorage,
     genesis_block: SerialBlock,
-    sender: mpsc::Sender<ConsensusMessageWrapped>,
+    sender: wrapped_mpsc::Sender<ConsensusMessageWrapped>,
 }
 
 impl Consensus {
@@ -63,7 +64,7 @@ impl Consensus {
         storage: DynStorage,
         memory_pool: MemoryPool,
     ) -> Arc<Self> {
-        let (sender, receiver) = mpsc::channel(256);
+        let (sender, receiver) = wrapped_mpsc::channel(snarkos_metrics::queues::CONSENSUS, 256);
         let created = Arc::new(Self {
             parameters,
             dpc,
@@ -91,7 +92,6 @@ impl Consensus {
     async fn send<T: Send + Sync + 'static>(&self, message: ConsensusMessage) -> T {
         let (sender, receiver) = oneshot::channel();
         self.sender.send((message, sender)).await.ok();
-        metrics::increment_gauge!(snarkos_metrics::queues::CONSENSUS, 1.0);
         *receiver
             .await
             .ok()
diff --git a/metrics/Cargo.toml b/metrics/Cargo.toml
index b529a57cf..380981ef1 100644
--- a/metrics/Cargo.toml
+++ b/metrics/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkos-metrics"
-version = "1.3.14"
+version = "1.3.15"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Metrics for a decentralized operating system"
 homepage = "https://aleo.org"
@@ -43,7 +43,7 @@ features = [ "derive" ]
 
 [dependencies.tokio]
 version = "1"
-features = [ "macros", "rt-multi-thread" ]
+features = [ "macros", "rt-multi-thread", "sync" ]
 
 [dev-dependencies.snarkvm-derives]
 version = "=0.7.9"
diff --git a/metrics/src/lib.rs b/metrics/src/lib.rs
index 2a3723bb8..d92cfad8a 100644
--- a/metrics/src/lib.rs
+++ b/metrics/src/lib.rs
@@ -22,6 +22,9 @@ pub use names::*;
 pub mod snapshots;
 pub mod stats;
 
+pub mod wrapped_mpsc;
+pub use wrapped_mpsc::*;
+
 /// Re-export metrics macros
 pub use metrics::*;
 
diff --git a/metrics/src/wrapped_mpsc.rs b/metrics/src/wrapped_mpsc.rs
new file mode 100644
index 000000000..f141e8686
--- /dev/null
+++ b/metrics/src/wrapped_mpsc.rs
@@ -0,0 +1,127 @@
+// Copyright (C) 2019-2021 Aleo Systems Inc.
+// This file is part of the snarkOS library.
+
+// The snarkOS library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// The snarkOS library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with the snarkOS library. If not, see <https://www.gnu.org/licenses/>.
+
+use std::{
+    fmt,
+    sync::{
+        atomic::{AtomicUsize, Ordering},
+        Arc,
+    },
+};
+
+use tokio::sync::mpsc::{
+    self,
+    error::{SendError, TrySendError},
+};
+
+/// Wrapper over mpsc::Sender to track metrics
+pub struct Sender<T: Send> {
+    inner: mpsc::Sender<T>,
+    tracker: Arc<AtomicUsize>,
+    metrics_tracker: &'static str,
+}
+
+impl<T: Send> Clone for Sender<T> {
+    fn clone(&self) -> Self {
+        Self {
+            inner: self.inner.clone(),
+            tracker: self.tracker.clone(),
+            metrics_tracker: self.metrics_tracker,
+        }
+    }
+}
+
+impl<T: Send> fmt::Debug for Sender<T> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "sender for {}", self.metrics_tracker)
+    }
+}
+
+impl<T: Send> Sender<T> {
+    fn increment(&self) {
+        metrics::increment_gauge!(self.metrics_tracker, 1.0);
+        self.tracker.fetch_add(1, Ordering::SeqCst);
+    }
+
+    pub async fn send(&self, value: T) -> Result<(), SendError<T>> {
+        self.increment();
+        self.inner.send(value).await
+    }
+
+    pub fn try_send(&self, message: T) -> Result<(), TrySendError<T>> {
+        self.increment();
+        self.inner.try_send(message)
+    }
+
+    pub fn blocking_send(&self, value: T) -> Result<(), SendError<T>> {
+        self.increment();
+        self.inner.blocking_send(value)
+    }
+}
+
+/// Wrapper over mpsc::Receiver to track metrics
+#[derive(Debug)]
+pub struct Receiver<T: Send> {
+    inner: mpsc::Receiver<T>,
+    tracker: Arc<AtomicUsize>,
+    metrics_tracker: &'static str,
+}
+
+impl<T: Send> Receiver<T> {
+    fn maybe_decrement(&self, is_ok: bool) {
+        if is_ok {
+            metrics::decrement_gauge!(self.metrics_tracker, 1.0);
+            self.tracker.fetch_sub(1, Ordering::SeqCst);
+        }
+    }
+
+    pub async fn recv(&mut self) -> Option<T> {
+        let out = self.inner.recv().await;
+        self.maybe_decrement(out.is_some());
+        out
+    }
+
+    pub fn blocking_recv(&mut self) -> Option<T> {
+        let out = self.inner.blocking_recv();
+        self.maybe_decrement(out.is_some());
+        out
+    }
+}
+
+pub fn channel<T: Send>(metrics_tracker: &'static str, buffer: usize) -> (Sender<T>, Receiver<T>) {
+    let (sender, receiver) = mpsc::channel(buffer);
+    let tracker = Arc::new(AtomicUsize::new(0));
+
+    (
+        Sender {
+            inner: sender,
+            metrics_tracker,
+            tracker: tracker.clone(),
+        },
+        Receiver {
+            inner: receiver,
+            metrics_tracker,
+            tracker,
+        },
+    )
+}
+
+impl<T: Send> Drop for Receiver<T> {
+    fn drop(&mut self) {
+        let count = self.tracker.swap(0, Ordering::SeqCst) as f64;
+        metrics::decrement_gauge!(self.metrics_tracker, count);
+    }
+}
diff --git a/network/Cargo.toml b/network/Cargo.toml
index f6fa2764c..0c7cda58c 100644
--- a/network/Cargo.toml
+++ b/network/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkos-network"
-version = "1.3.14"
+version = "1.3.15"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Network for a decentralized operating system"
 homepage = "https://aleo.org"
@@ -19,7 +19,7 @@ edition = "2018"
 
 [features]
 prometheus = [ "snarkos-metrics/prometheus" ]
-test = []
+test = [ ]
 
 [dependencies.snarkvm-algorithms]
 version = "=0.7.9"
@@ -35,11 +35,11 @@ version = "=0.7.9"
 
 [dependencies.snarkos-consensus]
 path = "../consensus"
-version = "1.3.14"
+version = "1.3.15"
 
 [dependencies.snarkos-storage]
 path = "../storage"
-version = "1.3.14"
+version = "1.3.15"
 
 [dependencies.indexmap]
 version = "1.6"
@@ -83,7 +83,7 @@ version = "0.17"
 
 [dependencies.snarkos-metrics]
 path = "../metrics"
-version = "1.3.14"
+version = "1.3.15"
 
 [dependencies.defer]
 version = "0.1"
diff --git a/network/src/node.rs b/network/src/node.rs
index af6745c31..036036862 100644
--- a/network/src/node.rs
+++ b/network/src/node.rs
@@ -15,7 +15,7 @@
 // along with the snarkOS library. If not, see <https://www.gnu.org/licenses/>.
 
 use crate::*;
-use snarkos_metrics::{self as metrics, misc};
+use snarkos_metrics::{self as metrics, misc, wrapped_mpsc};
 
 use anyhow::*;
 use chrono::{DateTime, Utc};
@@ -34,7 +34,7 @@ use std::{
     thread,
 };
 use tokio::{
-    sync::{mpsc, Mutex, RwLock},
+    sync::{Mutex, RwLock},
     task,
     time::sleep,
 };
@@ -78,7 +78,7 @@ pub struct InnerNode {
     threads: DropJoin<thread::JoinHandle<()>>,
     /// An indicator of whether the node is shutting down.
     shutting_down: AtomicBool,
-    pub(crate) master_dispatch: RwLock<Option<mpsc::Sender<SyncInbound>>>,
+    pub(crate) master_dispatch: RwLock<Option<wrapped_mpsc::Sender<SyncInbound>>>,
     pub terminator: Arc<AtomicBool>,
 }
 
diff --git a/network/src/peers/peer/connector.rs b/network/src/peers/peer/connector.rs
index 669db4bb4..690e6fa88 100644
--- a/network/src/peers/peer/connector.rs
+++ b/network/src/peers/peer/connector.rs
@@ -19,9 +19,9 @@ use std::{
     time::Duration,
 };
 
-use tokio::{net::TcpStream, sync::mpsc, time::timeout};
+use tokio::{net::TcpStream, time::timeout};
 
-use snarkos_metrics::{self as metrics, connections::*};
+use snarkos_metrics::{self as metrics, connections::*, wrapped_mpsc};
 
 use crate::{NetworkError, Node, Peer, PeerEvent, PeerEventData, PeerHandle, Version};
 
@@ -30,8 +30,8 @@ use super::{network::PeerIOHandle, PeerAction};
 const CONNECTION_TIMEOUT_SECS: u64 = 3;
 
 impl Peer {
-    pub fn connect(mut self, node: Node, event_target: mpsc::Sender<PeerEvent>) {
-        let (sender, receiver) = mpsc::channel::<PeerAction>(64);
+    pub fn connect(mut self, node: Node, event_target: wrapped_mpsc::Sender<PeerEvent>) {
+        let (sender, receiver) = wrapped_mpsc::channel::<PeerAction>(snarkos_metrics::queues::PEER_EVENTS, 64);
         tokio::spawn(async move {
             self.set_connecting();
             match self.inner_connect(node.version()).await {
@@ -43,7 +43,6 @@ impl Peer {
                         })
                         .await
                         .ok();
-                    metrics::increment_gauge!(snarkos_metrics::queues::PEER_EVENTS, 1.0);
                     self.fail();
                     if !e.is_trivial() {
                         error!(
@@ -77,14 +76,10 @@ impl Peer {
                     event_target
                         .send(PeerEvent {
                             address: self.address,
-                            data: PeerEventData::Connected(PeerHandle {
-                                sender: sender.clone(),
-                                queued_outbound_message_count: self.queued_outbound_message_count.clone(),
-                            }),
+                            data: PeerEventData::Connected(PeerHandle { sender: sender.clone() }),
                         })
                         .await
                         .ok();
-                    metrics::increment_gauge!(snarkos_metrics::queues::PEER_EVENTS, 1.0);
 
                     if let Err(e) = self.run(node, network, receiver).await {
                         if !e.is_trivial() {
@@ -111,7 +106,6 @@ impl Peer {
                 })
                 .await
                 .ok();
-            metrics::increment_gauge!(snarkos_metrics::queues::PEER_EVENTS, 1.0);
         });
     }
 
diff --git a/network/src/peers/peer/outbound_handler.rs b/network/src/peers/peer/outbound_handler.rs
index 8415c383a..0b6c4bd9a 100644
--- a/network/src/peers/peer/outbound_handler.rs
+++ b/network/src/peers/peer/outbound_handler.rs
@@ -14,18 +14,12 @@
 // You should have received a copy of the GNU General Public License
 // along with the snarkOS library. If not, see <https://www.gnu.org/licenses/>.
 
-use std::{
-    sync::{
-        atomic::{AtomicUsize, Ordering},
-        Arc,
-    },
-    time::Instant,
-};
+use std::time::Instant;
 
 use snarkos_storage::SerialBlock;
-use tokio::sync::{mpsc, oneshot};
+use tokio::sync::oneshot;
 
-use snarkos_metrics::{self as metrics, queues::*};
+use snarkos_metrics::{self as metrics, wrapped_mpsc};
 
 use crate::{NetworkError, Payload, Peer};
 
@@ -45,8 +39,7 @@ pub(super) enum PeerAction {
 
 #[derive(Clone, Debug)]
 pub struct PeerHandle {
-    pub(super) sender: mpsc::Sender<PeerAction>,
-    pub queued_outbound_message_count: Arc<AtomicUsize>,
+    pub(super) sender: wrapped_mpsc::Sender<PeerAction>,
 }
 
 impl PeerHandle {
@@ -66,22 +59,14 @@ impl PeerHandle {
     }
 
     pub async fn send_payload(&self, payload: Payload, time_received: Option<Instant>) {
-        if self.sender.send(PeerAction::Send(payload, time_received)).await.is_ok() {
-            self.queued_outbound_message_count.fetch_add(1, Ordering::SeqCst);
-            metrics::increment_gauge!(OUTBOUND, 1.0);
-        }
+        self.sender.send(PeerAction::Send(payload, time_received)).await.ok();
     }
 
     pub async fn send_block(&self, payload: SerialBlock, height: u32) {
-        if self
-            .sender
+        self.sender
             .send(PeerAction::SendBlock(Box::new(payload), height))
             .await
-            .is_ok()
-        {
-            self.queued_outbound_message_count.fetch_add(1, Ordering::SeqCst);
-            metrics::increment_gauge!(OUTBOUND, 1.0);
-        }
+            .ok();
     }
 
     pub async fn cancel_sync(&self) {
@@ -118,9 +103,6 @@ impl Peer {
                 // check if they sent us the block recently
                 let block = block.serialize();
 
-                self.queued_outbound_message_count.fetch_sub(1, Ordering::SeqCst);
-                metrics::decrement_gauge!(OUTBOUND, 1.0);
-
                 if self.block_received_cache.contains(&block[..]) {
                     metrics::increment_counter!(metrics::outbound::ALL_CACHE_HITS);
                     return Ok(PeerResponse::None);
@@ -144,9 +126,6 @@ impl Peer {
                     self.quality.last_ping_sent = Some(Instant::now());
                 }
 
-                self.queued_outbound_message_count.fetch_sub(1, Ordering::SeqCst);
-                metrics::decrement_gauge!(OUTBOUND, 1.0);
-
                 network.write_payload(&message).await.map_err(|e| {
                     metrics::increment_counter!(metrics::outbound::ALL_FAILURES);
                     e
diff --git a/network/src/peers/peer/peer.rs b/network/src/peers/peer/peer.rs
index 2678fe4d9..75ce66019 100644
--- a/network/src/peers/peer/peer.rs
+++ b/network/src/peers/peer/peer.rs
@@ -17,15 +17,11 @@
 use anyhow::*;
 use chrono::Utc;
 use serde::{Deserialize, Serialize};
+use snarkos_metrics::wrapped_mpsc;
 use std::{
     net::SocketAddr,
-    sync::{
-        atomic::{AtomicUsize, Ordering},
-        Arc,
-    },
     time::{Duration, Instant},
 };
-use tokio::sync::mpsc;
 
 use super::PeerQuality;
 use crate::{message::Payload, BlockCache, NetworkError, Node};
@@ -37,10 +33,6 @@ pub struct Peer {
     pub address: SocketAddr,
     pub quality: PeerQuality,
     pub is_bootnode: bool,
-    #[serde(skip)]
-    pub queued_inbound_message_count: Arc<AtomicUsize>,
-    #[serde(skip)]
-    pub queued_outbound_message_count: Arc<AtomicUsize>,
     /// Whether this peer is routable or not.
     ///
     /// `None` indicates the node has never attempted a connection with this peer.
@@ -59,8 +51,6 @@ impl Peer {
             address,
             quality: Default::default(),
             is_bootnode,
-            queued_inbound_message_count: Default::default(),
-            queued_outbound_message_count: Default::default(),
 
             // Set to `None` since peer creation only ever happens before a connection to the peer,
             // therefore we don't know if its listener is routable or not.
@@ -113,12 +103,12 @@ impl Peer {
         &mut self,
         node: Node,
         mut network: PeerIOHandle,
-        mut receiver: mpsc::Receiver<PeerAction>,
+        mut receiver: wrapped_mpsc::Receiver<PeerAction>,
     ) -> Result<(), NetworkError> {
         let mut reader = network.take_reader();
 
-        let (sender, mut read_receiver) = mpsc::channel::<Result<Vec<u8>, NetworkError>>(8);
-        let queued_inbound_message_count = self.queued_inbound_message_count.clone();
+        let (sender, mut read_receiver) =
+            wrapped_mpsc::channel::<Result<Vec<u8>, NetworkError>>(snarkos_metrics::queues::INBOUND, 8);
 
         tokio::spawn(async move {
             loop {
@@ -128,9 +118,6 @@ impl Peer {
                     .is_err()
                 {
                     break;
-                } else {
-                    queued_inbound_message_count.fetch_add(1, Ordering::SeqCst);
-                    metrics::increment_gauge!(snarkos_metrics::queues::INBOUND, 1.0);
                 }
             }
         });
@@ -154,9 +141,6 @@ impl Peer {
                         break;
                     }
 
-                    self.queued_inbound_message_count.fetch_sub(1, Ordering::SeqCst);
-                    metrics::decrement_gauge!(snarkos_metrics::queues::INBOUND, 1.0);
-
                     let data = match data.unwrap() {
                         // decrypt
                         Ok(data) => network.read_payload(&data[..]),
@@ -178,11 +162,6 @@ impl Peer {
             }
         }
 
-        let queued_inbound_message_count = self.queued_inbound_message_count.swap(0, Ordering::SeqCst);
-        metrics::decrement_gauge!(snarkos_metrics::queues::INBOUND, queued_inbound_message_count as f64);
-        let queued_outbound_message_count = self.queued_outbound_message_count.swap(0, Ordering::SeqCst);
-        metrics::decrement_gauge!(snarkos_metrics::queues::OUTBOUND, queued_outbound_message_count as f64);
-
         Ok(())
     }
 
diff --git a/network/src/peers/peer/receiver.rs b/network/src/peers/peer/receiver.rs
index e8a09056d..823070e05 100644
--- a/network/src/peers/peer/receiver.rs
+++ b/network/src/peers/peer/receiver.rs
@@ -16,17 +16,22 @@
 
 use std::net::{IpAddr, Ipv4Addr, SocketAddr};
 
-use tokio::{net::TcpStream, sync::mpsc};
+use tokio::net::TcpStream;
 
-use snarkos_metrics::{self as metrics, connections::*};
+use snarkos_metrics::{self as metrics, connections::*, wrapped_mpsc};
 
 use crate::{NetworkError, Node, Peer, PeerEvent, PeerEventData, PeerHandle, Version};
 
 use super::{network::PeerIOHandle, PeerAction};
 
 impl Peer {
-    pub fn receive(remote_address: SocketAddr, node: Node, stream: TcpStream, event_target: mpsc::Sender<PeerEvent>) {
-        let (sender, receiver) = mpsc::channel::<PeerAction>(64);
+    pub fn receive(
+        remote_address: SocketAddr,
+        node: Node,
+        stream: TcpStream,
+        event_target: wrapped_mpsc::Sender<PeerEvent>,
+    ) {
+        let (sender, receiver) = wrapped_mpsc::channel::<PeerAction>(metrics::queues::OUTBOUND, 64);
         tokio::spawn(async move {
             let (mut peer, network) = match Peer::inner_receive(remote_address, stream, node.version()).await {
                 Err(e) => {
@@ -41,25 +46,19 @@ impl Peer {
                         })
                         .await
                         .ok();
-                    metrics::increment_gauge!(snarkos_metrics::queues::PEER_EVENTS, 1.0);
                     return;
                 }
                 Ok(x) => x,
             };
 
             peer.set_connected();
-            metrics::increment_gauge!(CONNECTED, 1.0);
             event_target
                 .send(PeerEvent {
                     address: peer.address,
-                    data: PeerEventData::Connected(PeerHandle {
-                        sender: sender.clone(),
-                        queued_outbound_message_count: peer.queued_outbound_message_count.clone(),
-                    }),
+                    data: PeerEventData::Connected(PeerHandle { sender: sender.clone() }),
                 })
                 .await
                 .ok();
-            metrics::increment_gauge!(snarkos_metrics::queues::PEER_EVENTS, 1.0);
 
             if let Err(e) = peer.run(node, network, receiver).await {
                 if !e.is_trivial() {
@@ -84,7 +83,6 @@ impl Peer {
                 })
                 .await
                 .ok();
-            metrics::increment_gauge!(snarkos_metrics::queues::PEER_EVENTS, 1.0);
         });
     }
 
diff --git a/network/src/peers/peer_book.rs b/network/src/peers/peer_book.rs
index c2c0df968..cdec37b78 100644
--- a/network/src/peers/peer_book.rs
+++ b/network/src/peers/peer_book.rs
@@ -26,9 +26,9 @@ use std::{
 
 use mpmc_map::MpmcMap;
 use rand::{prelude::IteratorRandom, rngs::SmallRng, SeedableRng};
-use tokio::{net::TcpStream, sync::mpsc};
+use tokio::net::TcpStream;
 
-use snarkos_metrics::{self as metrics, connections::*};
+use snarkos_metrics::{self as metrics, connections::*, wrapped_mpsc};
 
 use crate::{NetworkError, Node, Payload, Peer, PeerEvent, PeerEventData, PeerHandle};
 
@@ -40,7 +40,7 @@ pub struct PeerBook {
     disconnected_peers: MpmcMap<SocketAddr, Peer>,
     connected_peers: MpmcMap<SocketAddr, PeerHandle>,
     pending_connections: Arc<AtomicU32>,
-    peer_events: mpsc::Sender<PeerEvent>,
+    peer_events: wrapped_mpsc::Sender<PeerEvent>,
 }
 
 // to avoid circular reference to peer_events
@@ -52,10 +52,8 @@ struct PeerBookRef {
 
 impl PeerBookRef {
     // gets terminated when sender is dropped from PeerBook
-    async fn handle_peer_events(self, mut receiver: mpsc::Receiver<PeerEvent>) {
+    async fn handle_peer_events(self, mut receiver: wrapped_mpsc::Receiver<PeerEvent>) {
         while let Some(event) = receiver.recv().await {
-            metrics::decrement_gauge!(snarkos_metrics::queues::PEER_EVENTS, 1.0);
-
             match event.data {
                 PeerEventData::Connected(handle) => {
                     self.pending_connections.fetch_sub(1, Ordering::SeqCst);
@@ -67,11 +65,6 @@ impl PeerBookRef {
                 PeerEventData::Disconnect(peer) => {
                     self.connected_peers.remove(peer.address).await;
 
-                    let queued_inbound_message_count = peer.queued_inbound_message_count.swap(0, Ordering::SeqCst);
-                    metrics::decrement_gauge!(snarkos_metrics::queues::INBOUND, queued_inbound_message_count as f64);
-                    let queued_outbound_message_count = peer.queued_outbound_message_count.swap(0, Ordering::SeqCst);
-                    metrics::decrement_gauge!(snarkos_metrics::queues::OUTBOUND, queued_outbound_message_count as f64);
-
                     if self.disconnected_peers.insert(peer.address, *peer).await.is_none() {
                         metrics::increment_gauge!(DISCONNECTED, 1.0);
                     }
@@ -86,7 +79,7 @@ impl PeerBookRef {
 
 impl PeerBook {
     pub fn spawn() -> Self {
-        let (sender, receiver) = mpsc::channel(256);
+        let (sender, receiver) = wrapped_mpsc::channel(snarkos_metrics::queues::PEER_EVENTS, 256);
         let peers = PeerBook {
             disconnected_peers: Default::default(),
             connected_peers: Default::default(),
diff --git a/network/src/sync/blocks/aggro.rs b/network/src/sync/blocks/aggro.rs
index 09758aa66..f90ce3fa0 100644
--- a/network/src/sync/blocks/aggro.rs
+++ b/network/src/sync/blocks/aggro.rs
@@ -18,11 +18,9 @@ use std::{sync::Arc, time::Duration};
 
 use crate::{Cache, Node, Payload, Peer, SyncBase, SyncInbound};
 use anyhow::*;
+use snarkos_metrics::wrapped_mpsc;
 use snarkos_storage::Digest;
-use tokio::{
-    sync::{mpsc, RwLock},
-    time::Instant,
-};
+use tokio::{sync::RwLock, time::Instant};
 
 /// Aggressive, continuous sync process that pulls peers entire canon trees.
 pub struct SyncAggro {
@@ -35,7 +33,7 @@ struct BlockLocatorHashes {
 }
 
 impl SyncAggro {
-    pub fn new(node: Node) -> (Self, mpsc::Sender<SyncInbound>) {
+    pub fn new(node: Node) -> (Self, wrapped_mpsc::Sender<SyncInbound>) {
         let (base, sender) = SyncBase::new(node);
         let new = Self { base };
         (new, sender)
@@ -90,7 +88,6 @@ impl SyncAggro {
         let node = self.base.node.clone();
         self.base
             .receive_messages(15, 3, |msg| {
-                metrics::decrement_gauge!(snarkos_metrics::queues::SYNC_ITEMS, 1.0);
                 match msg {
                     SyncInbound::BlockHashes(peer, hashes) => {
                         debug!("received {} sync hashes from {}", hashes.len(), peer);
@@ -174,9 +171,3 @@ impl SyncAggro {
         Ok(())
     }
 }
-
-impl Drop for SyncAggro {
-    fn drop(&mut self) {
-        metrics::gauge!(snarkos_metrics::queues::SYNC_ITEMS, 0.0);
-    }
-}
diff --git a/network/src/sync/blocks/base.rs b/network/src/sync/blocks/base.rs
index f4ca2bb8a..4a8906e94 100644
--- a/network/src/sync/blocks/base.rs
+++ b/network/src/sync/blocks/base.rs
@@ -16,8 +16,8 @@
 
 use std::{net::SocketAddr, time::Duration};
 
+use snarkos_metrics::wrapped_mpsc;
 use snarkos_storage::Digest;
-use tokio::sync::mpsc;
 
 use crate::{Node, Peer, SyncInbound};
 use anyhow::*;
@@ -25,12 +25,12 @@ use anyhow::*;
 /// Base sync helpers
 pub struct SyncBase {
     pub node: Node,
-    pub incoming: mpsc::Receiver<SyncInbound>,
+    pub incoming: wrapped_mpsc::Receiver<SyncInbound>,
 }
 
 impl SyncBase {
-    pub fn new(node: Node) -> (Self, mpsc::Sender<SyncInbound>) {
-        let (sender, receiver) = mpsc::channel(256);
+    pub fn new(node: Node) -> (Self, wrapped_mpsc::Sender<SyncInbound>) {
+        let (sender, receiver) = wrapped_mpsc::channel(snarkos_metrics::queues::SYNC_ITEMS, 256);
         let new = Self {
             node,
             incoming: receiver,
@@ -124,7 +124,6 @@ impl SyncBase {
                     if msg.is_none() {
                         break;
                     }
-                    metrics::decrement_gauge!(snarkos_metrics::queues::SYNC_ITEMS, 1.0);
                     if handler(msg.unwrap()) {
                         break;
                     }
diff --git a/network/src/sync/blocks/batched.rs b/network/src/sync/blocks/batched.rs
index 860446724..4c36d2385 100644
--- a/network/src/sync/blocks/batched.rs
+++ b/network/src/sync/blocks/batched.rs
@@ -20,10 +20,10 @@ use crate::{Node, Payload, Peer, SyncBase, SyncInbound};
 use anyhow::*;
 use hash_hasher::{HashBuildHasher, HashedMap, HashedSet};
 use rand::{prelude::SliceRandom, rngs::SmallRng, SeedableRng};
+use snarkos_metrics::wrapped_mpsc;
 use snarkos_storage::Digest;
 use snarkvm_algorithms::crh::double_sha256;
 use snarkvm_dpc::BlockHeader;
-use tokio::sync::mpsc;
 
 /// Efficient but slow and fork-prone sync method that operates in distributed batches.
 pub struct SyncBatched {
@@ -37,7 +37,7 @@ struct SyncBlock {
 }
 
 impl SyncBatched {
-    pub fn new(node: Node) -> (Self, mpsc::Sender<SyncInbound>) {
+    pub fn new(node: Node) -> (Self, wrapped_mpsc::Sender<SyncInbound>) {
         let (base, sender) = SyncBase::new(node);
         let new = Self { base };
         (new, sender)
@@ -288,9 +288,3 @@ impl SyncBatched {
         Ok(())
     }
 }
-
-impl Drop for SyncBatched {
-    fn drop(&mut self) {
-        metrics::gauge!(snarkos_metrics::queues::SYNC_ITEMS, 0.0);
-    }
-}
diff --git a/network/src/sync/blocks/interface.rs b/network/src/sync/blocks/interface.rs
index c031652af..6aec2db06 100644
--- a/network/src/sync/blocks/interface.rs
+++ b/network/src/sync/blocks/interface.rs
@@ -95,7 +95,6 @@ impl Node {
                     .send(SyncInbound::Block(remote_address, block, height))
                     .await
                     .ok();
-                metrics::increment_gauge!(snarkos_metrics::queues::SYNC_ITEMS, 1.0);
             }
         }
         Ok(())
@@ -230,7 +229,6 @@ impl Node {
                 .send(SyncInbound::BlockHashes(remote_address, block_hashes))
                 .await
                 .ok();
-            metrics::increment_gauge!(snarkos_metrics::queues::SYNC_ITEMS, 1.0);
         }
     }
 }
diff --git a/parameters/Cargo.toml b/parameters/Cargo.toml
index b9373976a..9c01015a0 100644
--- a/parameters/Cargo.toml
+++ b/parameters/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkos-parameters"
-version = "1.3.14"
+version = "1.3.15"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Parameters for a decentralized operating system"
 homepage = "https://aleo.org"
diff --git a/profiler/Cargo.toml b/profiler/Cargo.toml
index 709f3defa..65914dfb1 100644
--- a/profiler/Cargo.toml
+++ b/profiler/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkos-profiler"
-version = "1.3.14"
+version = "1.3.15"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Profiler for a decentralized operating system"
 homepage = "https://aleo.org"
diff --git a/rpc/Cargo.toml b/rpc/Cargo.toml
index 212a405f5..ba2b88d4e 100644
--- a/rpc/Cargo.toml
+++ b/rpc/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkos-rpc"
-version = "1.3.14"
+version = "1.3.15"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "RPC for a decentralized operating system"
 homepage = "https://aleo.org"
@@ -29,15 +29,15 @@ version = "=0.7.9"
 
 [dependencies.snarkos-consensus]
 path = "../consensus"
-version = "1.3.14"
+version = "1.3.15"
 
 [dependencies.snarkos-network]
 path = "../network"
-version = "1.3.14"
+version = "1.3.15"
 
 [dependencies.snarkos-storage]
 path = "../storage"
-version = "1.3.14"
+version = "1.3.15"
 
 [dependencies.anyhow]
 version = "1.0.40"
@@ -72,7 +72,7 @@ version = "1.0.1"
 version = "18"
 
 [dependencies.snarkos-metrics]
-version = "1.3.14"
+version = "1.3.15"
 path = "../metrics"
 
 [dependencies.parking_lot]
diff --git a/rpc/README.md b/rpc/README.md
index 6e2c1ff2c..fe50b38c8 100644
--- a/rpc/README.md
+++ b/rpc/README.md
@@ -212,6 +212,8 @@ None
 | `density`                            | f64        | The density of the known network                                                          |
 | `algebraic_connectivity`             | f64        | The algebraic connectivity, aka the fiedler eigenvalue of the known network               |
 | `degree_centrality_delta`            | f64        | The difference between the largest and the smallest peer count in the network             |
+| `potential_tip`                      | Option     | The crawler's best guess at the canon block height and member addresses                   |
+| `potential_forks`                    | array      | The crawler's best guess at the forks in the network                                      |
 | `edges`                              | array      | The list of connections known by the node                                                 |
 | `vertices`                           | array      | The list of nodes known by the node                                                       |
 | `edges[i].source`                    | SocketAddr | One side of the crawled connection                                                        |
@@ -245,6 +247,8 @@ None
 | `launched`       | timestamp     | The timestamp of when the node was launched   |
 | `listening_addr` | SocketAddr    | The configured listening address of the node  |
 | `version`        | string        | The version of the client binary              |
+| `min_peers`      | number        | The minimum desired number of connected peers |
+| `max_peers`      | number        | The maximum allowed number of connected peers |
 
 ### Example
 ```ignore
@@ -260,43 +264,53 @@ None
 
 ### Response
 
-|             Parameter            | Type |                             Description                           |
-|:--------------------------------:|:----:|:-----------------------------------------------------------------:|
-| `connections.all_accepted`       | u64  | The number of connection requests the node has received           |
-| `connections.all_initiated`      | u64  | The number of connection requests the node has made               |
-| `connections.all_rejected`       | u64  | The number of connection requests the node has rejected           |
-| `connections.connected_peers`    | u16  | The number of currently connected peers                           |
-| `connections.connecting_peers`   | u16  | The number of currently connecting peers                          |
-| `connections.disconnected_peers` | u16  | The number of known disconnected peers                            |
-| `handshakes.failures_init`       | u64  | The number of failed handshakes as the initiator                  |
-| `handshakes.failures_resp`       | u64  | The number of failed handshakes as the responder                  |
-| `handshakes.successes_init`      | u64  | The number of successful handshakes as the initiator              |
-| `handshakes.successes_resp`      | u64  | The number of successful handshakes as the responder              |
-| `handshakes.timeouts_init`       | u64  | The number of handshake timeouts as the initiator                 |
-| `handshakes.timeouts_resp`       | u64  | The number of handshake timeouts as the responder                 |
-| `inbound.all_successes`          | u64  | The number of successfully processed inbound messages             |
-| `inbound.all_failures`           | u64  | The number of inbound messages that couldn't be processed         |
-| `inbound.blocks`                 | u64  | The number of all received Block messages                         |
-| `inbound.getblocks`              | u64  | The number of all received GetBlocks messages                     |
-| `inbound.getmemorypool`          | u64  | The number of all received GetMemoryPool messages                 |
-| `inbound.getpeers`               | u64  | The number of all received GetPeers messages                      |
-| `inbound.getsync`                | u64  | The number of all received GetSync messages                       |
-| `inbound.memorypool`             | u64  | The number of all received MemoryPool messages                    |
-| `inbound.peers`                  | u64  | The number of all received Peers messages                         |
-| `inbound.pings`                  | u64  | The number of all received Ping messages                          |
-| `inbound.pongs`                  | u64  | The number of all received Pong messages                          |
-| `inbound.syncs`                  | u64  | The number of all received Sync messages                          |
-| `inbound.syncblocks`             | u64  | The number of all received SyncBlock messages                     |
-| `inbound.transactions`           | u64  | The number of all received Transaction messages                   |
-| `inbound.unknown`                | u64  | The number of all received Unknown messages                       |
-| `misc.block_height`              | u32  | The current block height of the node                              |
-| `misc.blocks_mined`              | u32  | The number of blocks the node has mined                           |
-| `misc.duplicate_blocks`          | u64  | The number of duplicate blocks received                           |
-| `misc.duplicate_sync_blocks`     | u64  | The number of duplicate sync blocks received                      |
-| `outbound.all_successes`         | u64  | The number of successfully sent messages                          |
-| `outbound.all_failures`          | u64  | The number of failures to send messages                           |
-| `queues.inbound`                 | u32  | The number of messages queued in the common inbound channel       |
-| `queues.outbound`                | u32  | The number of messages queued in the individual outbound channels |
+| Parameter                        | Type | Description                                                                  |
+| :------------------------------: | :--: | :--------------------------------------------------------------------------: |
+| `connections.all_accepted`       | u64  | The number of connection requests the node has received                      |
+| `connections.all_initiated`      | u64  | The number of connection requests the node has made                          |
+| `connections.all_rejected`       | u64  | The number of connection requests the node has rejected                      |
+| `connections.average_duration`   | f64  | The average connection duration in seconds                                   |
+| `connections.connected_peers`    | u32  | The number of currently connected peers                                      |
+| `connections.connecting_peers`   | u32  | The number of currently connecting peers                                     |
+| `connections.disconnected_peers` | u32  | The number of known disconnected peers                                       |
+| `handshakes.failures_init`       | u64  | The number of failed handshakes as the initiator                             |
+| `handshakes.failures_resp`       | u64  | The number of failed handshakes as the responder                             |
+| `handshakes.successes_init`      | u64  | The number of successful handshakes as the initiator                         |
+| `handshakes.successes_resp`      | u64  | The number of successful handshakes as the responder                         |
+| `handshakes.timeouts_init`       | u64  | The number of handshake timeouts as the initiator                            |
+| `handshakes.timeouts_resp`       | u64  | The number of handshake timeouts as the responder                            |
+| `inbound.all_successes`          | u64  | The number of successfully processed inbound messages                        |
+| `inbound.all_failures`           | u64  | The number of inbound messages that couldn't be processed                    |
+| `inbound.blocks`                 | u64  | The number of all received Block messages                                    |
+| `inbound.getblocks`              | u64  | The number of all received GetBlocks messages                                |
+| `inbound.getmemorypool`          | u64  | The number of all received GetMemoryPool messages                            |
+| `inbound.getpeers`               | u64  | The number of all received GetPeers messages                                 |
+| `inbound.getsync`                | u64  | The number of all received GetSync messages                                  |
+| `inbound.memorypool`             | u64  | The number of all received MemoryPool messages                               |
+| `inbound.peers`                  | u64  | The number of all received Peers messages                                    |
+| `inbound.pings`                  | u64  | The number of all received Ping messages                                     |
+| `inbound.pongs`                  | u64  | The number of all received Pong messages                                     |
+| `inbound.syncs`                  | u64  | The number of all received Sync messages                                     |
+| `inbound.syncblocks`             | u64  | The number of all received SyncBlock messages                                |
+| `inbound.transactions`           | u64  | The number of all received Transaction messages                              |
+| `inbound.unknown`                | u64  | The number of all received Unknown messages                                  |
+| `internal_rtt.getpeer`           | f64  | The average internal RTT (query to response) for GetPeer messages in seconds |
+| `internal_rtt.getsync`           | f64  | The average internal RTT for GetSync messages in seconds                     |
+| `internal_rtt.getblocks`         | f64  | The average internal RTT for GetBlocks messages in seconds                   |
+| `internal_rtt.getmemorypool`     | f64  | The average internal RTT for GetMemoryPool messages in seconds               |
+| `misc.block_height`              | u32  | The current block height of the node                                         |
+| `misc.blocks_mined`              | u32  | The number of blocks the node has mined                                      |
+| `misc.block_processing_time`     | f64  | The block processing time in seconds                                         |
+| `misc.duplicate_blocks`          | u64  | The number of duplicate blocks received                                      |
+| `misc.duplicate_sync_blocks`     | u64  | The number of duplicate sync blocks received                                 |
+| `outbound.all_successes`         | u64  | The number of successfully sent messages                                     |
+| `outbound.all_failures`          | u64  | The number of failures to send messages                                      |
+| `queues.consensus`               | u64  | The number of queued consensus requests                                      |
+| `queues.inbound`                 | u64  | The number of messages queued in the common inbound channel                  |
+| `queues.outbound`                | u64  | The number of messages queued in the individual outbound channels            |
+| `queues.peer_events`             | u64  | The number of queued peer events                                             |
+| `queues.storage`                 | u64  | The number of queued storage requests                                        |
+| `queues.sync_items`              | u64  | The number of queued sync items                                              |
 
 ### Example
 ```ignore
diff --git a/storage/Cargo.toml b/storage/Cargo.toml
index 84cb8d2e2..73b56cc04 100644
--- a/storage/Cargo.toml
+++ b/storage/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkos-storage"
-version = "1.3.14"
+version = "1.3.15"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Storage for a decentralized operating system"
 homepage = "https://aleo.org"
@@ -27,7 +27,7 @@ test = [ ]
 [dependencies.rusqlite]
 version = "0.25"
 optional = true
-features = ["bundled"]
+features = [ "bundled" ]
 
 [dependencies.snarkvm-algorithms]
 version = "=0.7.9"
@@ -40,6 +40,7 @@ version = "=0.7.9"
 
 [dependencies.snarkos-metrics]
 path = "../metrics"
+version = "1.3.15"
 
 [dependencies.derivative]
 version = "2.2.0"
diff --git a/storage/src/storage/async_adapter.rs b/storage/src/storage/async_adapter.rs
index 5f65b976b..bde56c2de 100644
--- a/storage/src/storage/async_adapter.rs
+++ b/storage/src/storage/async_adapter.rs
@@ -17,8 +17,9 @@
 use std::{any::Any, fmt};
 
 use anyhow::*;
+use metrics::wrapped_mpsc;
 use snarkos_metrics::{self as metrics, queues};
-use tokio::sync::{mpsc, oneshot};
+use tokio::sync::oneshot;
 use tracing::log::trace;
 
 #[cfg(feature = "test")]
@@ -211,10 +212,9 @@ impl<S: SyncStorage + 'static> Agent<S> {
         }
     }
 
-    fn agent(mut self, mut receiver: mpsc::Receiver<MessageWrapper>) {
+    fn agent(mut self, mut receiver: wrapped_mpsc::Receiver<MessageWrapper>) {
         self.inner.init().expect("failed to initialize sync storage");
         while let Some((message, response)) = receiver.blocking_recv() {
-            metrics::decrement_gauge!(queues::STORAGE, 1.0);
             let out = self.handle_message(message);
             response.send(out).ok();
         }
@@ -224,12 +224,12 @@ impl<S: SyncStorage + 'static> Agent<S> {
 type MessageWrapper = (Message, oneshot::Sender<Box<dyn Any + Send + Sync>>);
 
 pub struct AsyncStorage {
-    sender: mpsc::Sender<MessageWrapper>,
+    sender: wrapped_mpsc::Sender<MessageWrapper>,
 }
 
 impl AsyncStorage {
     pub fn new<S: SyncStorage + Send + 'static>(inner: S) -> AsyncStorage {
-        let (sender, receiver) = mpsc::channel(256);
+        let (sender, receiver) = wrapped_mpsc::channel(queues::STORAGE, 256);
         std::thread::spawn(move || Agent::new(inner).agent(receiver));
         Self { sender }
     }
@@ -237,7 +237,6 @@ impl AsyncStorage {
     #[allow(clippy::ok_expect)]
     async fn send<T: Send + Sync + 'static>(&self, message: Message) -> T {
         let (sender, receiver) = oneshot::channel();
-        metrics::increment_gauge!(queues::STORAGE, 1.0);
         self.sender.send((message, sender)).await.ok();
         *receiver
             .await
diff --git a/testing/Cargo.toml b/testing/Cargo.toml
index 1fc695c4d..0b706996b 100644
--- a/testing/Cargo.toml
+++ b/testing/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "snarkos-testing"
-version = "1.3.14"
+version = "1.3.15"
 authors = [ "The Aleo Team <hello@aleo.org>" ]
 description = "Testing suite for a decentralized operating system"
 homepage = "https://aleo.org"
@@ -48,22 +48,22 @@ version = "=0.7.9"
 
 [dependencies.snarkos-consensus]
 path = "../consensus"
-version = "1.3.14"
+version = "1.3.15"
 features = [ "test" ]
 
 [dependencies.snarkos-network]
 path = "../network"
-version = "1.3.14"
+version = "1.3.15"
 optional = true
 features = [ "test" ]
 
 [dependencies.snarkos-parameters]
 path = "../parameters"
-version = "1.3.14"
+version = "1.3.15"
 
 [dependencies.snarkos-storage]
 path = "../storage"
-version = "1.3.14"
+version = "1.3.15"
 default-features = false
 features = [ "test" ]
 
