diff --git a/synthesizer/program/src/logic/command/get.rs b/synthesizer/program/src/logic/command/get.rs
index 3537877ff..ac56eb92c 100644
--- a/synthesizer/program/src/logic/command/get.rs
+++ b/synthesizer/program/src/logic/command/get.rs
@@ -134,8 +134,6 @@ pub struct Get<N: Network> {
     key: Operand<N>,
     /// The destination register.
     destination: Register<N>,
-    // TODO (howardwu): For mainnet - remove `is_old`.
-    is_old: bool,
 }
 
 impl<N: Network> PartialEq for Get<N> {
@@ -267,7 +265,7 @@ impl<N: Network> Parser for Get<N> {
         // Parse the ";" from the string.
         let (string, _) = tag(";")(string)?;
 
-        Ok((string, Self { mapping, key, destination, is_old: false }))
+        Ok((string, Self { mapping, key, destination }))
     }
 }
 
@@ -331,7 +329,7 @@ impl<N: Network> FromBytes for Get<N> {
         // Read the destination register.
         let destination = Register::read_le(&mut reader)?;
         // Return the command.
-        Ok(Self { mapping, key, destination, is_old: !first_byte.is_zero() })
+        Ok(Self { mapping, key, destination })
     }
 }
 
@@ -339,14 +337,10 @@ impl<N: Network> ToBytes for Get<N> {
     /// Writes the operation to a buffer.
     fn write_le<W: Write>(&self, mut writer: W) -> IoResult<()> {
         // Write the mapping name.
-        match self.is_old {
-            false => self.mapping.write_le(&mut writer)?,
-            true => match self.mapping {
-                MappingLocator::Resource(identifier) => identifier.write_le(&mut writer)?,
-                MappingLocator::Locator(_) => {
-                    return Err(error("Expected `MappingLocator::Resource` for `get`."));
-                }
-            },
+        // TODO (howardwu): For mainnet - Write `self.mapping` directly, instead of matching on the identifier case.
+        match &self.mapping {
+            MappingLocator::Locator(_) => self.mapping.write_le(&mut writer)?,
+            MappingLocator::Resource(identifier) => identifier.write_le(&mut writer)?,
         }
         // Write the key operand.
         self.key.write_le(&mut writer)?;
diff --git a/synthesizer/program/src/logic/command/get_or_use.rs b/synthesizer/program/src/logic/command/get_or_use.rs
index a78deaa1c..7c2ef6d7d 100644
--- a/synthesizer/program/src/logic/command/get_or_use.rs
+++ b/synthesizer/program/src/logic/command/get_or_use.rs
@@ -40,8 +40,6 @@ pub struct GetOrUse<N: Network> {
     default: Operand<N>,
     /// The destination register.
     destination: Register<N>,
-    // TODO (howardwu): For mainnet - remove `is_old`.
-    is_old: bool,
 }
 
 impl<N: Network> PartialEq for GetOrUse<N> {
@@ -186,7 +184,7 @@ impl<N: Network> Parser for GetOrUse<N> {
         // Parse the ";" from the string.
         let (string, _) = tag(";")(string)?;
 
-        Ok((string, Self { mapping, key, default, destination, is_old: false }))
+        Ok((string, Self { mapping, key, default, destination }))
     }
 }
 
@@ -252,7 +250,7 @@ impl<N: Network> FromBytes for GetOrUse<N> {
         // Read the destination register.
         let destination = Register::read_le(&mut reader)?;
         // Return the command.
-        Ok(Self { mapping, key, default, destination, is_old: !first_byte.is_zero() })
+        Ok(Self { mapping, key, default, destination })
     }
 }
 
@@ -260,14 +258,10 @@ impl<N: Network> ToBytes for GetOrUse<N> {
     /// Writes the operation to a buffer.
     fn write_le<W: Write>(&self, mut writer: W) -> IoResult<()> {
         // Write the mapping name.
-        match self.is_old {
-            false => self.mapping.write_le(&mut writer)?,
-            true => match self.mapping {
-                MappingLocator::Resource(identifier) => identifier.write_le(&mut writer)?,
-                MappingLocator::Locator(_) => {
-                    return Err(error("Expected `MappingLocator::Resource` for `get.or_use`."));
-                }
-            },
+        // TODO (howardwu): For mainnet - Write the `self.mapping` directly, instead of matching on the identifier case.
+        match &self.mapping {
+            MappingLocator::Locator(_) => self.mapping.write_le(&mut writer)?,
+            MappingLocator::Resource(identifier) => identifier.write_le(&mut writer)?,
         }
         // Write the key operand.
         self.key.write_le(&mut writer)?;
