diff --git a/network/tests/fuzzing.rs b/network/tests/fuzzing.rs
index e4315360d..c8f32d6af 100644
--- a/network/tests/fuzzing.rs
+++ b/network/tests/fuzzing.rs
@@ -15,7 +15,6 @@
 // along with the snarkOS library. If not, see <https://www.gnu.org/licenses/>.
 
 use snarkos_network::{MessageHeader, NodeType, Payload, Version};
-use snarkvm_dpc::BlockHeaderHash;
 
 use rand::{distributions::Standard, thread_rng, Rng};
 use snarkos_storage::Digest;
diff --git a/testing/src/network/sync.rs b/testing/src/network/sync.rs
index 04ca5c612..cdd8ac022 100644
--- a/testing/src/network/sync.rs
+++ b/testing/src/network/sync.rs
@@ -14,7 +14,7 @@
 // You should have received a copy of the GNU General Public License
 // along with the snarkOS library. If not, see <https://www.gnu.org/licenses/>.
 
-use snarkos_storage::{BlockStatus, Digest, VMBlock};
+use snarkos_storage::{BlockStatus, VMBlock};
 use snarkvm_utilities::to_bytes_le;
 use tokio::time::sleep;
 
